--- before/net/minecraft/entity/projectile/EntityFishHook.java
+++ after/net/minecraft/entity/projectile/EntityFishHook.java
@@ -280,8 +280,8 @@
     {
         ItemStack itemstack = this.angler.getHeldItemMainhand();
         ItemStack itemstack1 = this.angler.getHeldItemOffhand();
-        boolean flag = itemstack.getItem() == Items.FISHING_ROD;
-        boolean flag1 = itemstack1.getItem() == Items.FISHING_ROD;
+        boolean flag = itemstack.getItem() instanceof net.minecraft.item.ItemFishingRod;
+        boolean flag1 = itemstack1.getItem() instanceof net.minecraft.item.ItemFishingRod;
 
         if (!this.angler.isDead && this.angler.isEntityAlive() && (flag || flag1) && !(this.getDistanceSq(this.angler) > 1024.0))
         {
@@ -429,9 +429,9 @@
                 double d0 = this.posX + f1 * this.ticksCatchableDelay * 0.1F;
                 double d1 = MathHelper.floor(this.getEntityBoundingBox().minY) + 1.0F;
                 double d2 = this.posZ + f2 * this.ticksCatchableDelay * 0.1F;
-                Block block = worldserver.getBlockState(new BlockPos(d0, d1 - 1.0, d2)).getBlock();
+                IBlockState state = worldserver.getBlockState(new BlockPos(d0, d1 - 1.0, d2));
 
-                if (block == Blocks.WATER || block == Blocks.FLOWING_WATER)
+                if (state.getMaterial() == Material.WATER)
                 {
                     if (this.rand.nextFloat() < 0.15F)
                     {
@@ -499,9 +499,9 @@
                 double d4 = this.posX + MathHelper.sin(f6) * f7 * 0.1F;
                 double d5 = MathHelper.floor(this.getEntityBoundingBox().minY) + 1.0F;
                 double d6 = this.posZ + MathHelper.cos(f6) * f7 * 0.1F;
-                Block block1 = worldserver.getBlockState(new BlockPos((int)d4, (int)d5 - 1, (int)d6)).getBlock();
+                IBlockState state = worldserver.getBlockState(new BlockPos((int) d4, (int) d5 - 1, (int) d6));
 
-                if (block1 == Blocks.WATER || block1 == Blocks.FLOWING_WATER)
+                if (state.getMaterial() == Material.WATER)
                 {
                     worldserver.spawnParticle(EnumParticleTypes.WATER_SPLASH, d4, d5, d6, 2 + this.rand.nextInt(2), 0.1F, 0.0, 0.1F, 0.0);
                 }
@@ -541,6 +541,7 @@
         {
             int i = 0;
 
+            net.minecraftforge.event.entity.player.ItemFishedEvent event = null;
             if (this.caughtEntity != null)
             {
                 this.bringInHookedEntity();
@@ -550,12 +551,17 @@
             else if (this.ticksCatchable > 0)
             {
                 LootContext.Builder lootcontext$builder = new LootContext.Builder((WorldServer)this.world);
-                lootcontext$builder.withLuck(this.luck + this.angler.getLuck());
+                lootcontext$builder.withLuck((float)this.luck + this.angler.getLuck()).withPlayer(this.angler).withLootedEntity(this); // Forge: add player & looted entity to LootContext
+                List<ItemStack> result = this.world.getLootTableManager().getLootTableFromLocation(LootTableList.GAMEPLAY_FISHING).generateLootForPools(this.rand, lootcontext$builder.build());
+                event = new net.minecraftforge.event.entity.player.ItemFishedEvent(result, this.inGround ? 2 : 1, this);
+                net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+                if (event.isCanceled())
+                {
+                    this.setDead();
+                    return event.getRodDamage();
+                }
 
-                for (ItemStack itemstack : this.world
-                        .getLootTableManager()
-                        .getLootTableFromLocation(LootTableList.GAMEPLAY_FISHING)
-                        .generateLootForPools(this.rand, lootcontext$builder.build()))
+                for (ItemStack itemstack : result)
                 {
                     EntityItem entityitem = new EntityItem(this.world, this.posX, this.posY, this.posZ, itemstack);
                     double d0 = this.angler.posX - this.posX;
@@ -595,7 +601,7 @@
             }
 
             this.setDead();
-            return i;
+            return event == null ? i : event.getRodDamage();
         }
         else
         {
