--- before/net/minecraft/entity/projectile/EntityWitherSkull.java
+++ after/net/minecraft/entity/projectile/EntityWitherSkull.java
@@ -21,7 +21,7 @@
 
 public class EntityWitherSkull extends EntityFireball
 {
-    private static final DataParameter<Boolean> field_184565_e = EntityDataManager.func_187226_a(EntityWitherSkull.class, DataSerializers.field_187198_h);
+    private static final DataParameter<Boolean> field_184565_e = EntityDataManager.<Boolean>func_187226_a(EntityWitherSkull.class, DataSerializers.field_187198_h);
 
     public EntityWitherSkull(World p_i1793_1_)
     {
@@ -40,34 +40,29 @@
         EntityFireball.func_189743_a(p_189746_0_, "WitherSkull");
     }
 
-    @Override
     protected float func_82341_c()
     {
         return this.func_82342_d() ? 0.73F : super.func_82341_c();
     }
 
     @SideOnly(Side.CLIENT)
-    public EntityWitherSkull(
-        World p_i1795_1_, double p_i1795_2_, double p_i1795_4_, double p_i1795_6_, double p_i1795_8_, double p_i1795_10_, double p_i1795_12_
-    )
+    public EntityWitherSkull(World p_i1795_1_, double p_i1795_2_, double p_i1795_4_, double p_i1795_6_, double p_i1795_8_, double p_i1795_10_, double p_i1795_12_)
     {
         super(p_i1795_1_, p_i1795_2_, p_i1795_4_, p_i1795_6_, p_i1795_8_, p_i1795_10_, p_i1795_12_);
         this.func_70105_a(0.3125F, 0.3125F);
     }
 
-    @Override
     public boolean func_70027_ad()
     {
         return false;
     }
 
-    @Override
     public float func_180428_a(Explosion p_180428_1_, World p_180428_2_, BlockPos p_180428_3_, IBlockState p_180428_4_)
     {
         float f = super.func_180428_a(p_180428_1_, p_180428_2_, p_180428_3_, p_180428_4_);
         Block block = p_180428_4_.func_177230_c();
 
-        if (this.func_82342_d() && EntityWither.func_181033_a(block))
+        if (this.func_82342_d() && block.canEntityDestroy(p_180428_4_, p_180428_2_, p_180428_3_, this) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this.field_70235_a, p_180428_3_, p_180428_4_))
         {
             f = Math.min(0.8F, f);
         }
@@ -75,7 +70,6 @@
         return f;
     }
 
-    @Override
     protected void func_70227_a(RayTraceResult p_70227_1_)
     {
         if (!this.field_70170_p.field_72995_K)
@@ -121,43 +115,36 @@
                 }
             }
 
-            this.field_70170_p
-            .func_72885_a(
-                this, this.field_70165_t, this.field_70163_u, this.field_70161_v, 1.0F, false, this.field_70170_p.func_82736_K().func_82766_b("mobGriefing")
-            );
+            this.field_70170_p.func_72885_a(this, this.field_70165_t, this.field_70163_u, this.field_70161_v, 1.0F, false, net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this.field_70235_a));
             this.func_70106_y();
         }
     }
 
-    @Override
     public boolean func_70067_L()
     {
         return false;
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         return false;
     }
 
-    @Override
     protected void func_70088_a()
     {
-        this.field_70180_af.func_187214_a(field_184565_e, false);
+        this.field_70180_af.func_187214_a(field_184565_e, Boolean.valueOf(false));
     }
 
     public boolean func_82342_d()
     {
-        return this.field_70180_af.func_187225_a(field_184565_e);
+        return ((Boolean)this.field_70180_af.func_187225_a(field_184565_e)).booleanValue();
     }
 
     public void func_82343_e(boolean p_82343_1_)
     {
-        this.field_70180_af.func_187227_b(field_184565_e, p_82343_1_);
+        this.field_70180_af.func_187227_b(field_184565_e, Boolean.valueOf(p_82343_1_));
     }
 
-    @Override
     protected boolean func_184564_k()
     {
         return false;
