--- before/net/minecraft/entity/projectile/EntityLargeFireball.java
+++ after/net/minecraft/entity/projectile/EntityLargeFireball.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.projectile;
 
+import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.util.DamageSource;
@@ -19,9 +20,7 @@
     }
 
     @SideOnly(Side.CLIENT)
-    public EntityLargeFireball(
-        World p_i1768_1_, double p_i1768_2_, double p_i1768_4_, double p_i1768_6_, double p_i1768_8_, double p_i1768_10_, double p_i1768_12_
-    )
+    public EntityLargeFireball(World p_i1768_1_, double p_i1768_2_, double p_i1768_4_, double p_i1768_6_, double p_i1768_8_, double p_i1768_10_, double p_i1768_12_)
     {
         super(p_i1768_1_, p_i1768_2_, p_i1768_4_, p_i1768_6_, p_i1768_8_, p_i1768_10_, p_i1768_12_);
     }
@@ -31,7 +30,6 @@
         super(p_i1769_1_, p_i1769_2_, p_i1769_3_, p_i1769_5_, p_i1769_7_);
     }
 
-    @Override
     protected void func_70227_a(RayTraceResult p_70227_1_)
     {
         if (!this.field_70170_p.field_72995_K)
@@ -42,8 +40,8 @@
                 this.func_174815_a(this.field_70235_a, p_70227_1_.field_72308_g);
             }
 
-            boolean flag = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
-            this.field_70170_p.func_72885_a(null, this.field_70165_t, this.field_70163_u, this.field_70161_v, (float)this.field_92057_e, flag, flag);
+            boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this.field_70235_a);
+            this.field_70170_p.func_72885_a((Entity)null, this.field_70165_t, this.field_70163_u, this.field_70161_v, (float)this.field_92057_e, flag, flag);
             this.func_70106_y();
         }
     }
@@ -53,14 +51,12 @@
         EntityFireball.func_189743_a(p_189744_0_, "Fireball");
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
         p_70014_1_.func_74768_a("ExplosionPower", this.field_92057_e);
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
