--- before/net/minecraft/entity/projectile/EntityFireball.java
+++ after/net/minecraft/entity/projectile/EntityFireball.java
@@ -30,23 +30,21 @@
         this.func_70105_a(1.0F, 1.0F);
     }
 
-    @Override
     protected void func_70088_a()
     {
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_70112_a(double p_70112_1_)
     {
-        double d0 = this.func_174813_aQ().func_72320_b() * 4.0;
+        double d0 = this.func_174813_aQ().func_72320_b() * 4.0D;
 
         if (Double.isNaN(d0))
         {
-            d0 = 4.0;
+            d0 = 4.0D;
         }
 
-        d0 *= 64.0;
+        d0 = d0 * 64.0D;
         return p_70112_1_ < d0 * d0;
     }
 
@@ -57,9 +55,9 @@
         this.func_70012_b(p_i1760_2_, p_i1760_4_, p_i1760_6_, this.field_70177_z, this.field_70125_A);
         this.func_70107_b(p_i1760_2_, p_i1760_4_, p_i1760_6_);
         double d0 = (double)MathHelper.func_76133_a(p_i1760_8_ * p_i1760_8_ + p_i1760_10_ * p_i1760_10_ + p_i1760_12_ * p_i1760_12_);
-        this.field_70232_b = p_i1760_8_ / d0 * 0.1;
-        this.field_70233_c = p_i1760_10_ / d0 * 0.1;
-        this.field_70230_d = p_i1760_12_ / d0 * 0.1;
+        this.field_70232_b = p_i1760_8_ / d0 * 0.1D;
+        this.field_70233_c = p_i1760_10_ / d0 * 0.1D;
+        this.field_70230_d = p_i1760_12_ / d0 * 0.1D;
     }
 
     public EntityFireball(World p_i1761_1_, EntityLivingBase p_i1761_2_, double p_i1761_3_, double p_i1761_5_, double p_i1761_7_)
@@ -69,23 +67,21 @@
         this.func_70105_a(1.0F, 1.0F);
         this.func_70012_b(p_i1761_2_.field_70165_t, p_i1761_2_.field_70163_u, p_i1761_2_.field_70161_v, p_i1761_2_.field_70177_z, p_i1761_2_.field_70125_A);
         this.func_70107_b(this.field_70165_t, this.field_70163_u, this.field_70161_v);
-        this.field_70159_w = 0.0;
-        this.field_70181_x = 0.0;
-        this.field_70179_y = 0.0;
-        p_i1761_3_ += this.field_70146_Z.nextGaussian() * 0.4;
-        p_i1761_5_ += this.field_70146_Z.nextGaussian() * 0.4;
-        p_i1761_7_ += this.field_70146_Z.nextGaussian() * 0.4;
+        this.field_70159_w = 0.0D;
+        this.field_70181_x = 0.0D;
+        this.field_70179_y = 0.0D;
+        p_i1761_3_ = p_i1761_3_ + this.field_70146_Z.nextGaussian() * 0.4D;
+        p_i1761_5_ = p_i1761_5_ + this.field_70146_Z.nextGaussian() * 0.4D;
+        p_i1761_7_ = p_i1761_7_ + this.field_70146_Z.nextGaussian() * 0.4D;
         double d0 = (double)MathHelper.func_76133_a(p_i1761_3_ * p_i1761_3_ + p_i1761_5_ * p_i1761_5_ + p_i1761_7_ * p_i1761_7_);
-        this.field_70232_b = p_i1761_3_ / d0 * 0.1;
-        this.field_70233_c = p_i1761_5_ / d0 * 0.1;
-        this.field_70230_d = p_i1761_7_ / d0 * 0.1;
+        this.field_70232_b = p_i1761_3_ / d0 * 0.1D;
+        this.field_70233_c = p_i1761_5_ / d0 * 0.1D;
+        this.field_70230_d = p_i1761_7_ / d0 * 0.1D;
     }
 
-    @Override
     public void func_70071_h_()
     {
-        if (this.field_70170_p.field_72995_K
-                || (this.field_70235_a == null || !this.field_70235_a.field_70128_L) && this.field_70170_p.func_175667_e(new BlockPos(this)))
+        if (this.field_70170_p.field_72995_K || (this.field_70235_a == null || !this.field_70235_a.field_70128_L) && this.field_70170_p.func_175667_e(new BlockPos(this)))
         {
             super.func_70071_h_();
 
@@ -97,7 +93,7 @@
             ++this.field_70234_an;
             RayTraceResult raytraceresult = ProjectileHelper.func_188802_a(this, true, this.field_70234_an >= 25, this.field_70235_a);
 
-            if (raytraceresult != null)
+            if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
             {
                 this.func_70227_a(raytraceresult);
             }
@@ -113,16 +109,7 @@
                 for (int i = 0; i < 4; ++i)
                 {
                     float f1 = 0.25F;
-                    this.field_70170_p
-                    .func_175688_a(
-                        EnumParticleTypes.WATER_BUBBLE,
-                        this.field_70165_t - this.field_70159_w * 0.25,
-                        this.field_70163_u - this.field_70181_x * 0.25,
-                        this.field_70161_v - this.field_70179_y * 0.25,
-                        this.field_70159_w,
-                        this.field_70181_x,
-                        this.field_70179_y
-                    );
+                    this.field_70170_p.func_175688_a(EnumParticleTypes.WATER_BUBBLE, this.field_70165_t - this.field_70159_w * 0.25D, this.field_70163_u - this.field_70181_x * 0.25D, this.field_70161_v - this.field_70179_y * 0.25D, this.field_70159_w, this.field_70181_x, this.field_70179_y);
                 }
 
                 f = 0.8F;
@@ -134,7 +121,7 @@
             this.field_70159_w *= (double)f;
             this.field_70181_x *= (double)f;
             this.field_70179_y *= (double)f;
-            this.field_70170_p.func_175688_a(this.func_184563_j(), this.field_70165_t, this.field_70163_u + 0.5, this.field_70161_v, 0.0, 0.0, 0.0);
+            this.field_70170_p.func_175688_a(this.func_184563_j(), this.field_70165_t, this.field_70163_u + 0.5D, this.field_70161_v, 0.0D, 0.0D, 0.0D);
             this.func_70107_b(this.field_70165_t, this.field_70163_u, this.field_70161_v);
         }
         else
@@ -164,7 +151,6 @@
     {
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         p_70014_1_.func_74782_a("direction", this.func_70087_a(new double[] {this.field_70159_w, this.field_70181_x, this.field_70179_y}));
@@ -172,7 +158,6 @@
         p_70014_1_.func_74768_a("life", this.field_70236_j);
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         if (p_70037_1_.func_150297_b("power", 9))
@@ -202,19 +187,16 @@
         }
     }
 
-    @Override
     public boolean func_70067_L()
     {
         return true;
     }
 
-    @Override
     public float func_70111_Y()
     {
         return 1.0F;
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         if (this.func_180431_b(p_70097_1_))
@@ -234,9 +216,9 @@
                     this.field_70159_w = vec3d.field_72450_a;
                     this.field_70181_x = vec3d.field_72448_b;
                     this.field_70179_y = vec3d.field_72449_c;
-                    this.field_70232_b = this.field_70159_w * 0.1;
-                    this.field_70233_c = this.field_70181_x * 0.1;
-                    this.field_70230_d = this.field_70179_y * 0.1;
+                    this.field_70232_b = this.field_70159_w * 0.1D;
+                    this.field_70233_c = this.field_70181_x * 0.1D;
+                    this.field_70230_d = this.field_70179_y * 0.1D;
                 }
 
                 if (p_70097_1_.func_76346_g() instanceof EntityLivingBase)
@@ -253,14 +235,12 @@
         }
     }
 
-    @Override
     public float func_70013_c()
     {
         return 1.0F;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public int func_70070_b()
     {
         return 15728880;
