--- before/net/minecraft/entity/projectile/EntitySmallFireball.java
+++ after/net/minecraft/entity/projectile/EntitySmallFireball.java
@@ -23,9 +23,7 @@
         this.func_70105_a(0.3125F, 0.3125F);
     }
 
-    public EntitySmallFireball(
-        World p_i1772_1_, double p_i1772_2_, double p_i1772_4_, double p_i1772_6_, double p_i1772_8_, double p_i1772_10_, double p_i1772_12_
-    )
+    public EntitySmallFireball(World p_i1772_1_, double p_i1772_2_, double p_i1772_4_, double p_i1772_6_, double p_i1772_8_, double p_i1772_10_, double p_i1772_12_)
     {
         super(p_i1772_1_, p_i1772_2_, p_i1772_4_, p_i1772_6_, p_i1772_8_, p_i1772_10_, p_i1772_12_);
         this.func_70105_a(0.3125F, 0.3125F);
@@ -36,7 +34,6 @@
         EntityFireball.func_189743_a(p_189745_0_, "SmallFireball");
     }
 
-    @Override
     protected void func_70227_a(RayTraceResult p_70227_1_)
     {
         if (!this.field_70170_p.field_72995_K)
@@ -60,7 +57,7 @@
 
                 if (this.field_70235_a != null && this.field_70235_a instanceof EntityLiving)
                 {
-                    flag1 = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
+                    flag1 = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this.field_70235_a);
                 }
 
                 if (flag1)
@@ -78,13 +75,11 @@
         }
     }
 
-    @Override
     public boolean func_70067_L()
     {
         return false;
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         return false;
