--- before/net/minecraft/entity/projectile/EntityThrowable.java
+++ after/net/minecraft/entity/projectile/EntityThrowable.java
@@ -249,7 +249,7 @@
             {
                 this.setPortal(raytraceresult.getBlockPos());
             }
-            else
+            else if (!net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
             {
                 this.onImpact(raytraceresult);
             }
