--- before/net/minecraft/entity/projectile/EntityEvokerFangs.java
+++ after/net/minecraft/entity/projectile/EntityEvokerFangs.java
@@ -17,7 +17,7 @@
 {
     private int field_190553_a;
     private boolean field_190554_b;
-    private int field_190555_c = 22;
+    private int field_190555_c;
     private boolean field_190556_d;
     private EntityLivingBase field_190557_e;
     private UUID field_190558_f;
@@ -25,6 +25,7 @@
     public EntityEvokerFangs(World p_i47275_1_)
     {
         super(p_i47275_1_);
+        this.field_190555_c = 22;
         this.func_70105_a(0.5F, 0.8F);
     }
 
