--- before/net/minecraft/entity/projectile/EntityShulkerBullet.java
+++ after/net/minecraft/entity/projectile/EntityShulkerBullet.java
@@ -325,7 +325,7 @@
 
                 RayTraceResult raytraceresult = ProjectileHelper.forwardsRaycast(this, true, false, this.owner);
 
-                if (raytraceresult != null)
+                if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
                 {
                     this.bulletHit(raytraceresult);
                 }
