--- before/net/minecraft/entity/projectile/EntityFireball.java
+++ after/net/minecraft/entity/projectile/EntityFireball.java
@@ -97,7 +97,7 @@
             this.ticksInAir++;
             RayTraceResult raytraceresult = ProjectileHelper.forwardsRaycast(this, true, this.ticksInAir >= 25, this.shootingEntity);
 
-            if (raytraceresult != null)
+            if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
             {
                 this.onImpact(raytraceresult);
             }
