--- before/net/minecraft/entity/projectile/EntityArrow.java
+++ after/net/minecraft/entity/projectile/EntityArrow.java
@@ -270,7 +270,7 @@
                 }
             }
 
-            if (raytraceresult != null)
+            if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
             {
                 this.onHit(raytraceresult);
             }
