--- before/net/minecraft/entity/projectile/EntityThrowable.java
+++ after/net/minecraft/entity/projectile/EntityThrowable.java
@@ -25,9 +25,9 @@
 
 public abstract class EntityThrowable extends Entity implements IProjectile
 {
-    private int field_145788_c = -1;
-    private int field_145786_d = -1;
-    private int field_145787_e = -1;
+    private int field_145788_c;
+    private int field_145786_d;
+    private int field_145787_e;
     private Block field_174853_f;
     protected boolean field_174854_a;
     public int field_70191_b;
@@ -41,6 +41,9 @@
     public EntityThrowable(World p_i1776_1_)
     {
         super(p_i1776_1_);
+        this.field_145788_c = -1;
+        this.field_145786_d = -1;
+        this.field_145787_e = -1;
         this.func_70105_a(0.25F, 0.25F);
     }
 
@@ -52,35 +55,33 @@
 
     public EntityThrowable(World p_i1777_1_, EntityLivingBase p_i1777_2_)
     {
-        this(p_i1777_1_, p_i1777_2_.field_70165_t, p_i1777_2_.field_70163_u + (double)p_i1777_2_.func_70047_e() - 0.1F, p_i1777_2_.field_70161_v);
+        this(p_i1777_1_, p_i1777_2_.field_70165_t, p_i1777_2_.field_70163_u + (double)p_i1777_2_.func_70047_e() - 0.10000000149011612D, p_i1777_2_.field_70161_v);
         this.field_70192_c = p_i1777_2_;
     }
 
-    @Override
     protected void func_70088_a()
     {
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_70112_a(double p_70112_1_)
     {
-        double d0 = this.func_174813_aQ().func_72320_b() * 4.0;
+        double d0 = this.func_174813_aQ().func_72320_b() * 4.0D;
 
         if (Double.isNaN(d0))
         {
-            d0 = 4.0;
+            d0 = 4.0D;
         }
 
-        d0 *= 64.0;
+        d0 = d0 * 64.0D;
         return p_70112_1_ < d0 * d0;
     }
 
     public void func_184538_a(Entity p_184538_1_, float p_184538_2_, float p_184538_3_, float p_184538_4_, float p_184538_5_, float p_184538_6_)
     {
-        float f = -MathHelper.func_76126_a(p_184538_3_ * (float)(Math.PI / 180.0)) * MathHelper.func_76134_b(p_184538_2_ * (float)(Math.PI / 180.0));
-        float f1 = -MathHelper.func_76126_a((p_184538_2_ + p_184538_4_) * (float)(Math.PI / 180.0));
-        float f2 = MathHelper.func_76134_b(p_184538_3_ * (float)(Math.PI / 180.0)) * MathHelper.func_76134_b(p_184538_2_ * (float)(Math.PI / 180.0));
+        float f = -MathHelper.func_76126_a(p_184538_3_ * 0.017453292F) * MathHelper.func_76134_b(p_184538_2_ * 0.017453292F);
+        float f1 = -MathHelper.func_76126_a((p_184538_2_ + p_184538_4_) * 0.017453292F);
+        float f2 = MathHelper.func_76134_b(p_184538_3_ * 0.017453292F) * MathHelper.func_76134_b(p_184538_2_ * 0.017453292F);
         this.func_70186_c((double)f, (double)f1, (double)f2, p_184538_5_, p_184538_6_);
         this.field_70159_w += p_184538_1_.field_70159_w;
         this.field_70179_y += p_184538_1_.field_70179_y;
@@ -91,32 +92,30 @@
         }
     }
 
-    @Override
     public void func_70186_c(double p_70186_1_, double p_70186_3_, double p_70186_5_, float p_70186_7_, float p_70186_8_)
     {
         float f = MathHelper.func_76133_a(p_70186_1_ * p_70186_1_ + p_70186_3_ * p_70186_3_ + p_70186_5_ * p_70186_5_);
-        p_70186_1_ /= (double)f;
-        p_70186_3_ /= (double)f;
-        p_70186_5_ /= (double)f;
-        p_70186_1_ += this.field_70146_Z.nextGaussian() * 0.0075F * (double)p_70186_8_;
-        p_70186_3_ += this.field_70146_Z.nextGaussian() * 0.0075F * (double)p_70186_8_;
-        p_70186_5_ += this.field_70146_Z.nextGaussian() * 0.0075F * (double)p_70186_8_;
-        p_70186_1_ *= (double)p_70186_7_;
-        p_70186_3_ *= (double)p_70186_7_;
-        p_70186_5_ *= (double)p_70186_7_;
+        p_70186_1_ = p_70186_1_ / (double)f;
+        p_70186_3_ = p_70186_3_ / (double)f;
+        p_70186_5_ = p_70186_5_ / (double)f;
+        p_70186_1_ = p_70186_1_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
+        p_70186_3_ = p_70186_3_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
+        p_70186_5_ = p_70186_5_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
+        p_70186_1_ = p_70186_1_ * (double)p_70186_7_;
+        p_70186_3_ = p_70186_3_ * (double)p_70186_7_;
+        p_70186_5_ = p_70186_5_ * (double)p_70186_7_;
         this.field_70159_w = p_70186_1_;
         this.field_70181_x = p_70186_3_;
         this.field_70179_y = p_70186_5_;
         float f1 = MathHelper.func_76133_a(p_70186_1_ * p_70186_1_ + p_70186_5_ * p_70186_5_);
-        this.field_70177_z = (float)(MathHelper.func_181159_b(p_70186_1_, p_70186_5_) * 180.0F / (float)Math.PI);
-        this.field_70125_A = (float)(MathHelper.func_181159_b(p_70186_3_, (double)f1) * 180.0F / (float)Math.PI);
+        this.field_70177_z = (float)(MathHelper.func_181159_b(p_70186_1_, p_70186_5_) * (180D / Math.PI));
+        this.field_70125_A = (float)(MathHelper.func_181159_b(p_70186_3_, (double)f1) * (180D / Math.PI));
         this.field_70126_B = this.field_70177_z;
         this.field_70127_C = this.field_70125_A;
         this.field_70194_h = 0;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_70016_h(double p_70016_1_, double p_70016_3_, double p_70016_5_)
     {
         this.field_70159_w = p_70016_1_;
@@ -126,14 +125,13 @@
         if (this.field_70127_C == 0.0F && this.field_70126_B == 0.0F)
         {
             float f = MathHelper.func_76133_a(p_70016_1_ * p_70016_1_ + p_70016_5_ * p_70016_5_);
-            this.field_70177_z = (float)(MathHelper.func_181159_b(p_70016_1_, p_70016_5_) * 180.0F / (float)Math.PI);
-            this.field_70125_A = (float)(MathHelper.func_181159_b(p_70016_3_, (double)f) * 180.0F / (float)Math.PI);
+            this.field_70177_z = (float)(MathHelper.func_181159_b(p_70016_1_, p_70016_5_) * (180D / Math.PI));
+            this.field_70125_A = (float)(MathHelper.func_181159_b(p_70016_3_, (double)f) * (180D / Math.PI));
             this.field_70126_B = this.field_70177_z;
             this.field_70127_C = this.field_70125_A;
         }
     }
 
-    @Override
     public void func_70071_h_()
     {
         this.field_70142_S = this.field_70165_t;
@@ -148,8 +146,7 @@
 
         if (this.field_174854_a)
         {
-            if (this.field_70170_p.func_180495_p(new BlockPos(this.field_145788_c, this.field_145786_d, this.field_145787_e)).func_177230_c()
-                    == this.field_174853_f)
+            if (this.field_70170_p.func_180495_p(new BlockPos(this.field_145788_c, this.field_145786_d, this.field_145787_e)).func_177230_c() == this.field_174853_f)
             {
                 ++this.field_70194_h;
 
@@ -185,9 +182,8 @@
         }
 
         Entity entity = null;
-        List<Entity> list = this.field_70170_p
-                            .func_72839_b(this, this.func_174813_aQ().func_72321_a(this.field_70159_w, this.field_70181_x, this.field_70179_y).func_186662_g(1.0));
-        double d0 = 0.0;
+        List<Entity> list = this.field_70170_p.func_72839_b(this, this.func_174813_aQ().func_72321_a(this.field_70159_w, this.field_70181_x, this.field_70179_y).func_186662_g(1.0D));
+        double d0 = 0.0D;
         boolean flag = false;
 
         for (int i = 0; i < list.size(); ++i)
@@ -208,14 +204,14 @@
                 else
                 {
                     flag = false;
-                    AxisAlignedBB axisalignedbb = entity1.func_174813_aQ().func_186662_g(0.3F);
+                    AxisAlignedBB axisalignedbb = entity1.func_174813_aQ().func_186662_g(0.30000001192092896D);
                     RayTraceResult raytraceresult1 = axisalignedbb.func_72327_a(vec3d, vec3d1);
 
                     if (raytraceresult1 != null)
                     {
                         double d1 = vec3d.func_72436_e(raytraceresult1.field_72307_f);
 
-                        if (d1 < d0 || d0 == 0.0)
+                        if (d1 < d0 || d0 == 0.0D)
                         {
                             entity = entity1;
                             d0 = d1;
@@ -244,12 +240,11 @@
 
         if (raytraceresult != null)
         {
-            if (raytraceresult.field_72313_a == RayTraceResult.Type.BLOCK
-                    && this.field_70170_p.func_180495_p(raytraceresult.func_178782_a()).func_177230_c() == Blocks.field_150427_aO)
+            if (raytraceresult.field_72313_a == RayTraceResult.Type.BLOCK && this.field_70170_p.func_180495_p(raytraceresult.func_178782_a()).func_177230_c() == Blocks.field_150427_aO)
             {
                 this.func_181015_d(raytraceresult.func_178782_a());
             }
-            else
+            else if (!net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
             {
                 this.func_70184_a(raytraceresult);
             }
@@ -259,12 +254,11 @@
         this.field_70163_u += this.field_70181_x;
         this.field_70161_v += this.field_70179_y;
         float f = MathHelper.func_76133_a(this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y);
-        this.field_70177_z = (float)(MathHelper.func_181159_b(this.field_70159_w, this.field_70179_y) * 180.0F / (float)Math.PI);
-        this.field_70125_A = (float)(MathHelper.func_181159_b(this.field_70181_x, (double)f) * 180.0F / (float)Math.PI);
+        this.field_70177_z = (float)(MathHelper.func_181159_b(this.field_70159_w, this.field_70179_y) * (180D / Math.PI));
 
-        while (this.field_70125_A - this.field_70127_C < -180.0F)
+        for (this.field_70125_A = (float)(MathHelper.func_181159_b(this.field_70181_x, (double)f) * (180D / Math.PI)); this.field_70125_A - this.field_70127_C < -180.0F; this.field_70127_C -= 360.0F)
         {
-            this.field_70127_C -= 360.0F;
+            ;
         }
 
         while (this.field_70125_A - this.field_70127_C >= 180.0F)
@@ -292,16 +286,7 @@
             for (int j = 0; j < 4; ++j)
             {
                 float f3 = 0.25F;
-                this.field_70170_p
-                .func_175688_a(
-                    EnumParticleTypes.WATER_BUBBLE,
-                    this.field_70165_t - this.field_70159_w * 0.25,
-                    this.field_70163_u - this.field_70181_x * 0.25,
-                    this.field_70161_v - this.field_70179_y * 0.25,
-                    this.field_70159_w,
-                    this.field_70181_x,
-                    this.field_70179_y
-                );
+                this.field_70170_p.func_175688_a(EnumParticleTypes.WATER_BUBBLE, this.field_70165_t - this.field_70159_w * 0.25D, this.field_70163_u - this.field_70181_x * 0.25D, this.field_70161_v - this.field_70179_y * 0.25D, this.field_70159_w, this.field_70181_x, this.field_70179_y);
             }
 
             f1 = 0.8F;
@@ -330,7 +315,6 @@
     {
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         p_70014_1_.func_74768_a("xTile", this.field_145788_c);
@@ -349,7 +333,6 @@
         p_70014_1_.func_74778_a("ownerName", this.field_85053_h == null ? "" : this.field_85053_h);
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         this.field_145788_c = p_70037_1_.func_74762_e("xTile");
@@ -396,7 +379,7 @@
                         this.field_70192_c = (EntityLivingBase)entity;
                     }
                 }
-                catch (Throwable throwable)
+                catch (Throwable var2)
                 {
                     this.field_70192_c = null;
                 }
