--- before/net/minecraft/entity/projectile/EntityArrow.java
+++ after/net/minecraft/entity/projectile/EntityArrow.java
@@ -39,34 +39,37 @@
 
 public abstract class EntityArrow extends Entity implements IProjectile
 {
-    private static final Predicate<Entity> field_184553_f = Predicates.and(
-                    EntitySelectors.field_180132_d, EntitySelectors.field_94557_a, new Predicate<Entity>()
+    private static final Predicate<Entity> field_184553_f = Predicates.and(EntitySelectors.field_180132_d, EntitySelectors.field_94557_a, new Predicate<Entity>()
     {
         public boolean apply(@Nullable Entity p_apply_1_)
         {
             return p_apply_1_.func_70067_L();
         }
-    }
-            );
-    private static final DataParameter<Byte> field_184554_g = EntityDataManager.func_187226_a(EntityArrow.class, DataSerializers.field_187191_a);
-    private int field_145791_d = -1;
-    private int field_145792_e = -1;
-    private int field_145789_f = -1;
+    });
+    private static final DataParameter<Byte> field_184554_g = EntityDataManager.<Byte>func_187226_a(EntityArrow.class, DataSerializers.field_187191_a);
+    private int field_145791_d;
+    private int field_145792_e;
+    private int field_145789_f;
     private Block field_145790_g;
     private int field_70253_h;
     protected boolean field_70254_i;
     protected int field_184552_b;
-    public EntityArrow.PickupStatus field_70251_a = EntityArrow.PickupStatus.DISALLOWED;
+    public EntityArrow.PickupStatus field_70251_a;
     public int field_70249_b;
     public Entity field_70250_c;
     private int field_70252_j;
     private int field_70257_an;
-    private double field_70255_ao = 2.0;
+    private double field_70255_ao;
     private int field_70256_ap;
 
     public EntityArrow(World p_i1753_1_)
     {
         super(p_i1753_1_);
+        this.field_145791_d = -1;
+        this.field_145792_e = -1;
+        this.field_145789_f = -1;
+        this.field_70251_a = EntityArrow.PickupStatus.DISALLOWED;
+        this.field_70255_ao = 2.0D;
         this.func_70105_a(0.5F, 0.5F);
     }
 
@@ -78,7 +81,7 @@
 
     public EntityArrow(World p_i46777_1_, EntityLivingBase p_i46777_2_)
     {
-        this(p_i46777_1_, p_i46777_2_.field_70165_t, p_i46777_2_.field_70163_u + (double)p_i46777_2_.func_70047_e() - 0.1F, p_i46777_2_.field_70161_v);
+        this(p_i46777_1_, p_i46777_2_.field_70165_t, p_i46777_2_.field_70163_u + (double)p_i46777_2_.func_70047_e() - 0.10000000149011612D, p_i46777_2_.field_70161_v);
         this.field_70250_c = p_i46777_2_;
 
         if (p_i46777_2_ instanceof EntityPlayer)
@@ -88,31 +91,29 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_70112_a(double p_70112_1_)
     {
-        double d0 = this.func_174813_aQ().func_72320_b() * 10.0;
+        double d0 = this.func_174813_aQ().func_72320_b() * 10.0D;
 
         if (Double.isNaN(d0))
         {
-            d0 = 1.0;
+            d0 = 1.0D;
         }
 
-        d0 *= 64.0 * func_184183_bd();
+        d0 = d0 * 64.0D * func_184183_bd();
         return p_70112_1_ < d0 * d0;
     }
 
-    @Override
     protected void func_70088_a()
     {
-        this.field_70180_af.func_187214_a(field_184554_g, (byte)0);
+        this.field_70180_af.func_187214_a(field_184554_g, Byte.valueOf((byte)0));
     }
 
     public void func_184547_a(Entity p_184547_1_, float p_184547_2_, float p_184547_3_, float p_184547_4_, float p_184547_5_, float p_184547_6_)
     {
-        float f = -MathHelper.func_76126_a(p_184547_3_ * (float)(Math.PI / 180.0)) * MathHelper.func_76134_b(p_184547_2_ * (float)(Math.PI / 180.0));
-        float f1 = -MathHelper.func_76126_a(p_184547_2_ * (float)(Math.PI / 180.0));
-        float f2 = MathHelper.func_76134_b(p_184547_3_ * (float)(Math.PI / 180.0)) * MathHelper.func_76134_b(p_184547_2_ * (float)(Math.PI / 180.0));
+        float f = -MathHelper.func_76126_a(p_184547_3_ * 0.017453292F) * MathHelper.func_76134_b(p_184547_2_ * 0.017453292F);
+        float f1 = -MathHelper.func_76126_a(p_184547_2_ * 0.017453292F);
+        float f2 = MathHelper.func_76134_b(p_184547_3_ * 0.017453292F) * MathHelper.func_76134_b(p_184547_2_ * 0.017453292F);
         this.func_70186_c((double)f, (double)f1, (double)f2, p_184547_5_, p_184547_6_);
         this.field_70159_w += p_184547_1_.field_70159_w;
         this.field_70179_y += p_184547_1_.field_70179_y;
@@ -123,42 +124,37 @@
         }
     }
 
-    @Override
     public void func_70186_c(double p_70186_1_, double p_70186_3_, double p_70186_5_, float p_70186_7_, float p_70186_8_)
     {
         float f = MathHelper.func_76133_a(p_70186_1_ * p_70186_1_ + p_70186_3_ * p_70186_3_ + p_70186_5_ * p_70186_5_);
-        p_70186_1_ /= (double)f;
-        p_70186_3_ /= (double)f;
-        p_70186_5_ /= (double)f;
-        p_70186_1_ += this.field_70146_Z.nextGaussian() * 0.0075F * (double)p_70186_8_;
-        p_70186_3_ += this.field_70146_Z.nextGaussian() * 0.0075F * (double)p_70186_8_;
-        p_70186_5_ += this.field_70146_Z.nextGaussian() * 0.0075F * (double)p_70186_8_;
-        p_70186_1_ *= (double)p_70186_7_;
-        p_70186_3_ *= (double)p_70186_7_;
-        p_70186_5_ *= (double)p_70186_7_;
+        p_70186_1_ = p_70186_1_ / (double)f;
+        p_70186_3_ = p_70186_3_ / (double)f;
+        p_70186_5_ = p_70186_5_ / (double)f;
+        p_70186_1_ = p_70186_1_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
+        p_70186_3_ = p_70186_3_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
+        p_70186_5_ = p_70186_5_ + this.field_70146_Z.nextGaussian() * 0.007499999832361937D * (double)p_70186_8_;
+        p_70186_1_ = p_70186_1_ * (double)p_70186_7_;
+        p_70186_3_ = p_70186_3_ * (double)p_70186_7_;
+        p_70186_5_ = p_70186_5_ * (double)p_70186_7_;
         this.field_70159_w = p_70186_1_;
         this.field_70181_x = p_70186_3_;
         this.field_70179_y = p_70186_5_;
         float f1 = MathHelper.func_76133_a(p_70186_1_ * p_70186_1_ + p_70186_5_ * p_70186_5_);
-        this.field_70177_z = (float)(MathHelper.func_181159_b(p_70186_1_, p_70186_5_) * 180.0F / (float)Math.PI);
-        this.field_70125_A = (float)(MathHelper.func_181159_b(p_70186_3_, (double)f1) * 180.0F / (float)Math.PI);
+        this.field_70177_z = (float)(MathHelper.func_181159_b(p_70186_1_, p_70186_5_) * (180D / Math.PI));
+        this.field_70125_A = (float)(MathHelper.func_181159_b(p_70186_3_, (double)f1) * (180D / Math.PI));
         this.field_70126_B = this.field_70177_z;
         this.field_70127_C = this.field_70125_A;
         this.field_70252_j = 0;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
-    public void func_180426_a(
-        double p_180426_1_, double p_180426_3_, double p_180426_5_, float p_180426_7_, float p_180426_8_, int p_180426_9_, boolean p_180426_10_
-    )
+    public void func_180426_a(double p_180426_1_, double p_180426_3_, double p_180426_5_, float p_180426_7_, float p_180426_8_, int p_180426_9_, boolean p_180426_10_)
     {
         this.func_70107_b(p_180426_1_, p_180426_3_, p_180426_5_);
         this.func_70101_b(p_180426_7_, p_180426_8_);
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_70016_h(double p_70016_1_, double p_70016_3_, double p_70016_5_)
     {
         this.field_70159_w = p_70016_1_;
@@ -168,8 +164,8 @@
         if (this.field_70127_C == 0.0F && this.field_70126_B == 0.0F)
         {
             float f = MathHelper.func_76133_a(p_70016_1_ * p_70016_1_ + p_70016_5_ * p_70016_5_);
-            this.field_70125_A = (float)(MathHelper.func_181159_b(p_70016_3_, (double)f) * 180.0F / (float)Math.PI);
-            this.field_70177_z = (float)(MathHelper.func_181159_b(p_70016_1_, p_70016_5_) * 180.0F / (float)Math.PI);
+            this.field_70125_A = (float)(MathHelper.func_181159_b(p_70016_3_, (double)f) * (180D / Math.PI));
+            this.field_70177_z = (float)(MathHelper.func_181159_b(p_70016_1_, p_70016_5_) * (180D / Math.PI));
             this.field_70127_C = this.field_70125_A;
             this.field_70126_B = this.field_70177_z;
             this.func_70012_b(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70177_z, this.field_70125_A);
@@ -177,7 +173,6 @@
         }
     }
 
-    @Override
     public void func_70071_h_()
     {
         super.func_70071_h_();
@@ -185,8 +180,8 @@
         if (this.field_70127_C == 0.0F && this.field_70126_B == 0.0F)
         {
             float f = MathHelper.func_76133_a(this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y);
-            this.field_70177_z = (float)(MathHelper.func_181159_b(this.field_70159_w, this.field_70179_y) * 180.0F / (float)Math.PI);
-            this.field_70125_A = (float)(MathHelper.func_181159_b(this.field_70181_x, (double)f) * 180.0F / (float)Math.PI);
+            this.field_70177_z = (float)(MathHelper.func_181159_b(this.field_70159_w, this.field_70179_y) * (180D / Math.PI));
+            this.field_70125_A = (float)(MathHelper.func_181159_b(this.field_70181_x, (double)f) * (180D / Math.PI));
             this.field_70126_B = this.field_70177_z;
             this.field_70127_C = this.field_70125_A;
         }
@@ -199,8 +194,7 @@
         {
             AxisAlignedBB axisalignedbb = iblockstate.func_185890_d(this.field_70170_p, blockpos);
 
-            if (axisalignedbb != Block.field_185506_k
-                    && axisalignedbb.func_186670_a(blockpos).func_72318_a(new Vec3d(this.field_70165_t, this.field_70163_u, this.field_70161_v)))
+            if (axisalignedbb != Block.field_185506_k && axisalignedbb.func_186670_a(blockpos).func_72318_a(new Vec3d(this.field_70165_t, this.field_70163_u, this.field_70161_v)))
             {
                 this.field_70254_i = true;
             }
@@ -215,7 +209,7 @@
         {
             int j = block.func_176201_c(iblockstate);
 
-            if ((block != this.field_145790_g || j != this.field_70253_h) && !this.field_70170_p.func_184143_b(this.func_174813_aQ().func_186662_g(0.05)))
+            if ((block != this.field_145790_g || j != this.field_70253_h) && !this.field_70170_p.func_184143_b(this.func_174813_aQ().func_186662_g(0.05D)))
             {
                 this.field_70254_i = false;
                 this.field_70159_w *= (double)(this.field_70146_Z.nextFloat() * 0.2F);
@@ -248,9 +242,7 @@
 
             if (raytraceresult != null)
             {
-                vec3d = new Vec3d(
-                    raytraceresult.field_72307_f.field_72450_a, raytraceresult.field_72307_f.field_72448_b, raytraceresult.field_72307_f.field_72449_c
-                );
+                vec3d = new Vec3d(raytraceresult.field_72307_f.field_72450_a, raytraceresult.field_72307_f.field_72448_b, raytraceresult.field_72307_f.field_72449_c);
             }
 
             Entity entity = this.func_184551_a(vec3d1, vec3d);
@@ -270,7 +262,7 @@
                 }
             }
 
-            if (raytraceresult != null)
+            if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
             {
                 this.func_184549_a(raytraceresult);
             }
@@ -279,16 +271,7 @@
             {
                 for (int k = 0; k < 4; ++k)
                 {
-                    this.field_70170_p
-                    .func_175688_a(
-                        EnumParticleTypes.CRIT,
-                        this.field_70165_t + this.field_70159_w * (double)k / 4.0,
-                        this.field_70163_u + this.field_70181_x * (double)k / 4.0,
-                        this.field_70161_v + this.field_70179_y * (double)k / 4.0,
-                        -this.field_70159_w,
-                        -this.field_70181_x + 0.2,
-                        -this.field_70179_y
-                    );
+                    this.field_70170_p.func_175688_a(EnumParticleTypes.CRIT, this.field_70165_t + this.field_70159_w * (double)k / 4.0D, this.field_70163_u + this.field_70181_x * (double)k / 4.0D, this.field_70161_v + this.field_70179_y * (double)k / 4.0D, -this.field_70159_w, -this.field_70181_x + 0.2D, -this.field_70179_y);
                 }
             }
 
@@ -296,12 +279,11 @@
             this.field_70163_u += this.field_70181_x;
             this.field_70161_v += this.field_70179_y;
             float f4 = MathHelper.func_76133_a(this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y);
-            this.field_70177_z = (float)(MathHelper.func_181159_b(this.field_70159_w, this.field_70179_y) * 180.0F / (float)Math.PI);
-            this.field_70125_A = (float)(MathHelper.func_181159_b(this.field_70181_x, (double)f4) * 180.0F / (float)Math.PI);
+            this.field_70177_z = (float)(MathHelper.func_181159_b(this.field_70159_w, this.field_70179_y) * (180D / Math.PI));
 
-            while (this.field_70125_A - this.field_70127_C < -180.0F)
+            for (this.field_70125_A = (float)(MathHelper.func_181159_b(this.field_70181_x, (double)f4) * (180D / Math.PI)); this.field_70125_A - this.field_70127_C < -180.0F; this.field_70127_C -= 360.0F)
             {
-                this.field_70127_C -= 360.0F;
+                ;
             }
 
             while (this.field_70125_A - this.field_70127_C >= 180.0F)
@@ -329,16 +311,7 @@
                 for (int i = 0; i < 4; ++i)
                 {
                     float f3 = 0.25F;
-                    this.field_70170_p
-                    .func_175688_a(
-                        EnumParticleTypes.WATER_BUBBLE,
-                        this.field_70165_t - this.field_70159_w * 0.25,
-                        this.field_70163_u - this.field_70181_x * 0.25,
-                        this.field_70161_v - this.field_70179_y * 0.25,
-                        this.field_70159_w,
-                        this.field_70181_x,
-                        this.field_70179_y
-                    );
+                    this.field_70170_p.func_175688_a(EnumParticleTypes.WATER_BUBBLE, this.field_70165_t - this.field_70159_w * 0.25D, this.field_70163_u - this.field_70181_x * 0.25D, this.field_70161_v - this.field_70179_y * 0.25D, this.field_70159_w, this.field_70181_x, this.field_70179_y);
                 }
 
                 f1 = 0.6F;
@@ -355,7 +328,7 @@
 
             if (!this.func_189652_ae())
             {
-                this.field_70181_x -= 0.05F;
+                this.field_70181_x -= 0.05000000074505806D;
             }
 
             this.func_70107_b(this.field_70165_t, this.field_70163_u, this.field_70161_v);
@@ -369,9 +342,7 @@
 
         if (entity != null)
         {
-            float f = MathHelper.func_76133_a(
-                          this.field_70159_w * this.field_70159_w + this.field_70181_x * this.field_70181_x + this.field_70179_y * this.field_70179_y
-                      );
+            float f = MathHelper.func_76133_a(this.field_70159_w * this.field_70159_w + this.field_70181_x * this.field_70181_x + this.field_70179_y * this.field_70179_y);
             int i = MathHelper.func_76143_f((double)f * this.field_70255_ao);
 
             if (this.func_70241_g())
@@ -412,11 +383,7 @@
 
                         if (f1 > 0.0F)
                         {
-                            entitylivingbase.func_70024_g(
-                                this.field_70159_w * (double)this.field_70256_ap * 0.6F / (double)f1,
-                                0.1,
-                                this.field_70179_y * (double)this.field_70256_ap * 0.6F / (double)f1
-                            );
+                            entitylivingbase.func_70024_g(this.field_70159_w * (double)this.field_70256_ap * 0.6000000238418579D / (double)f1, 0.1D, this.field_70179_y * (double)this.field_70256_ap * 0.6000000238418579D / (double)f1);
                         }
                     }
 
@@ -428,10 +395,7 @@
 
                     this.func_184548_a(entitylivingbase);
 
-                    if (this.field_70250_c != null
-                            && entitylivingbase != this.field_70250_c
-                            && entitylivingbase instanceof EntityPlayer
-                            && this.field_70250_c instanceof EntityPlayerMP)
+                    if (this.field_70250_c != null && entitylivingbase != this.field_70250_c && entitylivingbase instanceof EntityPlayer && this.field_70250_c instanceof EntityPlayerMP)
                     {
                         ((EntityPlayerMP)this.field_70250_c).field_71135_a.func_147359_a(new SPacketChangeGameState(6, 0.0F));
                     }
@@ -446,15 +410,14 @@
             }
             else
             {
-                this.field_70159_w *= -0.1F;
-                this.field_70181_x *= -0.1F;
-                this.field_70179_y *= -0.1F;
+                this.field_70159_w *= -0.10000000149011612D;
+                this.field_70181_x *= -0.10000000149011612D;
+                this.field_70179_y *= -0.10000000149011612D;
                 this.field_70177_z += 180.0F;
                 this.field_70126_B += 180.0F;
                 this.field_70257_an = 0;
 
-                if (!this.field_70170_p.field_72995_K
-                        && this.field_70159_w * this.field_70159_w + this.field_70181_x * this.field_70181_x + this.field_70179_y * this.field_70179_y < 0.001F)
+                if (!this.field_70170_p.field_72995_K && this.field_70159_w * this.field_70159_w + this.field_70181_x * this.field_70181_x + this.field_70179_y * this.field_70179_y < 0.0010000000474974513D)
                 {
                     if (this.field_70251_a == EntityArrow.PickupStatus.ALLOWED)
                     {
@@ -477,12 +440,10 @@
             this.field_70159_w = (double)((float)(p_184549_1_.field_72307_f.field_72450_a - this.field_70165_t));
             this.field_70181_x = (double)((float)(p_184549_1_.field_72307_f.field_72448_b - this.field_70163_u));
             this.field_70179_y = (double)((float)(p_184549_1_.field_72307_f.field_72449_c - this.field_70161_v));
-            float f2 = MathHelper.func_76133_a(
-                           this.field_70159_w * this.field_70159_w + this.field_70181_x * this.field_70181_x + this.field_70179_y * this.field_70179_y
-                       );
-            this.field_70165_t -= this.field_70159_w / (double)f2 * 0.05F;
-            this.field_70163_u -= this.field_70181_x / (double)f2 * 0.05F;
-            this.field_70161_v -= this.field_70179_y / (double)f2 * 0.05F;
+            float f2 = MathHelper.func_76133_a(this.field_70159_w * this.field_70159_w + this.field_70181_x * this.field_70181_x + this.field_70179_y * this.field_70179_y);
+            this.field_70165_t -= this.field_70159_w / (double)f2 * 0.05000000074505806D;
+            this.field_70163_u -= this.field_70181_x / (double)f2 * 0.05000000074505806D;
+            this.field_70161_v -= this.field_70179_y / (double)f2 * 0.05000000074505806D;
             this.func_184185_a(SoundEvents.field_187731_t, 1.0F, 1.2F / (this.field_70146_Z.nextFloat() * 0.2F + 0.9F));
             this.field_70254_i = true;
             this.field_70249_b = 7;
@@ -495,7 +456,6 @@
         }
     }
 
-    @Override
     public void func_70091_d(MoverType p_70091_1_, double p_70091_2_, double p_70091_4_, double p_70091_6_)
     {
         super.func_70091_d(p_70091_1_, p_70091_2_, p_70091_4_, p_70091_6_);
@@ -516,9 +476,8 @@
     protected Entity func_184551_a(Vec3d p_184551_1_, Vec3d p_184551_2_)
     {
         Entity entity = null;
-        List<Entity> list = this.field_70170_p
-                            .func_175674_a(this, this.func_174813_aQ().func_72321_a(this.field_70159_w, this.field_70181_x, this.field_70179_y).func_186662_g(1.0), field_184553_f);
-        double d0 = 0.0;
+        List<Entity> list = this.field_70170_p.func_175674_a(this, this.func_174813_aQ().func_72321_a(this.field_70159_w, this.field_70181_x, this.field_70179_y).func_186662_g(1.0D), field_184553_f);
+        double d0 = 0.0D;
 
         for (int i = 0; i < list.size(); ++i)
         {
@@ -526,14 +485,14 @@
 
             if (entity1 != this.field_70250_c || this.field_70257_an >= 5)
             {
-                AxisAlignedBB axisalignedbb = entity1.func_174813_aQ().func_186662_g(0.3F);
+                AxisAlignedBB axisalignedbb = entity1.func_174813_aQ().func_186662_g(0.30000001192092896D);
                 RayTraceResult raytraceresult = axisalignedbb.func_72327_a(p_184551_1_, p_184551_2_);
 
                 if (raytraceresult != null)
                 {
                     double d1 = p_184551_1_.func_72436_e(raytraceresult.field_72307_f);
 
-                    if (d1 < d0 || d0 == 0.0)
+                    if (d1 < d0 || d0 == 0.0D)
                     {
                         entity = entity1;
                         d0 = d1;
@@ -554,7 +513,6 @@
         func_189657_a(p_189658_0_, "Arrow");
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         p_70014_1_.func_74768_a("xTile", this.field_145791_d);
@@ -571,7 +529,6 @@
         p_70014_1_.func_74757_a("crit", this.func_70241_g());
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         this.field_145791_d = p_70037_1_.func_74762_e("xTile");
@@ -609,13 +566,11 @@
         this.func_70243_d(p_70037_1_.func_74767_n("crit"));
     }
 
-    @Override
     public void func_70100_b_(EntityPlayer p_70100_1_)
     {
         if (!this.field_70170_p.field_72995_K && this.field_70254_i && this.field_70249_b <= 0)
         {
-            boolean flag = this.field_70251_a == EntityArrow.PickupStatus.ALLOWED
-                           || this.field_70251_a == EntityArrow.PickupStatus.CREATIVE_ONLY && p_70100_1_.field_71075_bZ.field_75098_d;
+            boolean flag = this.field_70251_a == EntityArrow.PickupStatus.ALLOWED || this.field_70251_a == EntityArrow.PickupStatus.CREATIVE_ONLY && p_70100_1_.field_71075_bZ.field_75098_d;
 
             if (this.field_70251_a == EntityArrow.PickupStatus.ALLOWED && !p_70100_1_.field_71071_by.func_70441_a(this.func_184550_j()))
             {
@@ -632,7 +587,6 @@
 
     protected abstract ItemStack func_184550_j();
 
-    @Override
     protected boolean func_70041_e_()
     {
         return false;
@@ -653,13 +607,11 @@
         this.field_70256_ap = p_70240_1_;
     }
 
-    @Override
     public boolean func_70075_an()
     {
         return false;
     }
 
-    @Override
     public float func_70047_e()
     {
         return 0.0F;
@@ -667,21 +619,21 @@
 
     public void func_70243_d(boolean p_70243_1_)
     {
-        byte b0 = this.field_70180_af.func_187225_a(field_184554_g);
+        byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184554_g)).byteValue();
 
         if (p_70243_1_)
         {
-            this.field_70180_af.func_187227_b(field_184554_g, (byte)(b0 | 1));
+            this.field_70180_af.func_187227_b(field_184554_g, Byte.valueOf((byte)(b0 | 1)));
         }
         else
         {
-            this.field_70180_af.func_187227_b(field_184554_g, (byte)(b0 & -2));
+            this.field_70180_af.func_187227_b(field_184554_g, Byte.valueOf((byte)(b0 & -2)));
         }
     }
 
     public boolean func_70241_g()
     {
-        byte b0 = this.field_70180_af.func_187225_a(field_184554_g);
+        byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184554_g)).byteValue();
         return (b0 & 1) != 0;
     }
 
@@ -689,13 +641,11 @@
     {
         int i = EnchantmentHelper.func_185284_a(Enchantments.field_185309_u, p_190547_1_);
         int j = EnchantmentHelper.func_185284_a(Enchantments.field_185310_v, p_190547_1_);
-        this.func_70239_b(
-            (double)(p_190547_2_ * 2.0F) + this.field_70146_Z.nextGaussian() * 0.25 + (double)((float)this.field_70170_p.func_175659_aa().func_151525_a() * 0.11F)
-        );
+        this.func_70239_b((double)(p_190547_2_ * 2.0F) + this.field_70146_Z.nextGaussian() * 0.25D + (double)((float)this.field_70170_p.func_175659_aa().func_151525_a() * 0.11F));
 
         if (i > 0)
         {
-            this.func_70239_b(this.func_70242_d() + (double)i * 0.5 + 0.5);
+            this.func_70239_b(this.func_70242_d() + (double)i * 0.5D + 0.5D);
         }
 
         if (j > 0)
