--- before/net/minecraft/entity/projectile/EntityDragonFireball.java
+++ after/net/minecraft/entity/projectile/EntityDragonFireball.java
@@ -46,7 +46,7 @@
         {
             if (!this.field_70170_p.field_72995_K)
             {
-                List<EntityLivingBase> list = this.field_70170_p.func_72872_a(EntityLivingBase.class, this.func_174813_aQ().func_72314_b(4.0D, 2.0D, 4.0D));
+                List<EntityLivingBase> list = this.field_70170_p.<EntityLivingBase>func_72872_a(EntityLivingBase.class, this.func_174813_aQ().func_72314_b(4.0D, 2.0D, 4.0D));
                 EntityAreaEffectCloud entityareaeffectcloud = new EntityAreaEffectCloud(this.field_70170_p, this.field_70165_t, this.field_70163_u, this.field_70161_v);
                 entityareaeffectcloud.func_184481_a(this.field_70235_a);
                 entityareaeffectcloud.func_184491_a(EnumParticleTypes.DRAGON_BREATH);
