--- before/net/minecraft/entity/projectile/EntityLargeFireball.java
+++ after/net/minecraft/entity/projectile/EntityLargeFireball.java
@@ -42,7 +42,7 @@
                 this.applyEnchantments(this.shootingEntity, result.entityHit);
             }
 
-            boolean flag = this.world.getGameRules().getBoolean("mobGriefing");
+            boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.shootingEntity);
             this.world.newExplosion(null, this.posX, this.posY, this.posZ, (float)this.explosionPower, flag, flag);
             this.setDead();
         }
