--- before/net/minecraft/entity/projectile/EntityShulkerBullet.java
+++ after/net/minecraft/entity/projectile/EntityShulkerBullet.java
@@ -50,16 +50,13 @@
         this.field_70145_X = true;
     }
 
-    @Override
     public SoundCategory func_184176_by()
     {
         return SoundCategory.HOSTILE;
     }
 
     @SideOnly(Side.CLIENT)
-    public EntityShulkerBullet(
-        World p_i46771_1_, double p_i46771_2_, double p_i46771_4_, double p_i46771_6_, double p_i46771_8_, double p_i46771_10_, double p_i46771_12_
-    )
+    public EntityShulkerBullet(World p_i46771_1_, double p_i46771_2_, double p_i46771_4_, double p_i46771_6_, double p_i46771_8_, double p_i46771_10_, double p_i46771_12_)
     {
         this(p_i46771_1_);
         this.func_70012_b(p_i46771_2_, p_i46771_4_, p_i46771_6_, this.field_70177_z, this.field_70125_A);
@@ -73,16 +70,15 @@
         this(p_i46772_1_);
         this.field_184570_a = p_i46772_2_;
         BlockPos blockpos = new BlockPos(p_i46772_2_);
-        double d0 = (double)blockpos.func_177958_n() + 0.5;
-        double d1 = (double)blockpos.func_177956_o() + 0.5;
-        double d2 = (double)blockpos.func_177952_p() + 0.5;
+        double d0 = (double)blockpos.func_177958_n() + 0.5D;
+        double d1 = (double)blockpos.func_177956_o() + 0.5D;
+        double d2 = (double)blockpos.func_177952_p() + 0.5D;
         this.func_70012_b(d0, d1, d2, this.field_70177_z, this.field_70125_A);
         this.field_184571_b = p_i46772_3_;
         this.field_184573_c = EnumFacing.UP;
         this.func_184569_a(p_i46772_4_);
     }
 
-    @Override
     protected void func_70014_b(NBTTagCompound p_70014_1_)
     {
         if (this.field_184570_a != null)
@@ -116,7 +112,6 @@
         p_70014_1_.func_74780_a("TZD", this.field_184579_g);
     }
 
-    @Override
     protected void func_70037_a(NBTTagCompound p_70037_1_)
     {
         this.field_184575_d = p_70037_1_.func_74762_e("Steps");
@@ -144,7 +139,6 @@
         }
     }
 
-    @Override
     protected void func_70088_a()
     {
     }
@@ -156,28 +150,28 @@
 
     private void func_184569_a(@Nullable EnumFacing.Axis p_184569_1_)
     {
-        double d0 = 0.5;
+        double d0 = 0.5D;
         BlockPos blockpos;
 
         if (this.field_184571_b == null)
         {
-            blockpos = new BlockPos(this).func_177977_b();
+            blockpos = (new BlockPos(this)).func_177977_b();
         }
         else
         {
-            d0 = (double)this.field_184571_b.field_70131_O * 0.5;
+            d0 = (double)this.field_184571_b.field_70131_O * 0.5D;
             blockpos = new BlockPos(this.field_184571_b.field_70165_t, this.field_184571_b.field_70163_u + d0, this.field_184571_b.field_70161_v);
         }
 
-        double d1 = (double)blockpos.func_177958_n() + 0.5;
+        double d1 = (double)blockpos.func_177958_n() + 0.5D;
         double d2 = (double)blockpos.func_177956_o() + d0;
-        double d3 = (double)blockpos.func_177952_p() + 0.5;
+        double d3 = (double)blockpos.func_177952_p() + 0.5D;
         EnumFacing enumfacing = null;
 
-        if (blockpos.func_177957_d(this.field_70165_t, this.field_70163_u, this.field_70161_v) >= 4.0)
+        if (blockpos.func_177957_d(this.field_70165_t, this.field_70163_u, this.field_70161_v) >= 4.0D)
         {
             BlockPos blockpos1 = new BlockPos(this);
-            List<EnumFacing> list = Lists.newArrayList();
+            List<EnumFacing> list = Lists.<EnumFacing>newArrayList();
 
             if (p_184569_1_ != EnumFacing.Axis.X)
             {
@@ -240,24 +234,23 @@
         double d4 = d3 - this.field_70161_v;
         double d5 = (double)MathHelper.func_76133_a(d6 * d6 + d7 * d7 + d4 * d4);
 
-        if (d5 == 0.0)
+        if (d5 == 0.0D)
         {
-            this.field_184577_e = 0.0;
-            this.field_184578_f = 0.0;
-            this.field_184579_g = 0.0;
+            this.field_184577_e = 0.0D;
+            this.field_184578_f = 0.0D;
+            this.field_184579_g = 0.0D;
         }
         else
         {
-            this.field_184577_e = d6 / d5 * 0.15;
-            this.field_184578_f = d7 / d5 * 0.15;
-            this.field_184579_g = d4 / d5 * 0.15;
+            this.field_184577_e = d6 / d5 * 0.15D;
+            this.field_184578_f = d7 / d5 * 0.15D;
+            this.field_184579_g = d4 / d5 * 0.15D;
         }
 
         this.field_70160_al = true;
         this.field_184575_d = 10 + this.field_70146_Z.nextInt(5) * 10;
     }
 
-    @Override
     public void func_70071_h_()
     {
         if (!this.field_70170_p.field_72995_K && this.field_70170_p.func_175659_aa() == EnumDifficulty.PEACEFUL)
@@ -272,10 +265,7 @@
             {
                 if (this.field_184571_b == null && this.field_184574_at != null)
                 {
-                    for (EntityLivingBase entitylivingbase : this.field_70170_p
-                            .func_72872_a(
-                                EntityLivingBase.class, new AxisAlignedBB(this.field_184576_au.func_177982_a(-2, -2, -2), this.field_184576_au.func_177982_a(2, 2, 2))
-                            ))
+                    for (EntityLivingBase entitylivingbase : this.field_70170_p.func_72872_a(EntityLivingBase.class, new AxisAlignedBB(this.field_184576_au.func_177982_a(-2, -2, -2), this.field_184576_au.func_177982_a(2, 2, 2))))
                     {
                         if (entitylivingbase.func_110124_au().equals(this.field_184574_at))
                         {
@@ -289,10 +279,7 @@
 
                 if (this.field_184570_a == null && this.field_184580_h != null)
                 {
-                    for (EntityLivingBase entitylivingbase1 : this.field_70170_p
-                            .func_72872_a(
-                                EntityLivingBase.class, new AxisAlignedBB(this.field_184572_as.func_177982_a(-2, -2, -2), this.field_184572_as.func_177982_a(2, 2, 2))
-                            ))
+                    for (EntityLivingBase entitylivingbase1 : this.field_70170_p.func_72872_a(EntityLivingBase.class, new AxisAlignedBB(this.field_184572_as.func_177982_a(-2, -2, -2), this.field_184572_as.func_177982_a(2, 2, 2))))
                     {
                         if (entitylivingbase1.func_110124_au().equals(this.field_184580_h))
                         {
@@ -304,28 +291,26 @@
                     this.field_184580_h = null;
                 }
 
-                if (this.field_184571_b == null
-                        || !this.field_184571_b.func_70089_S()
-                        || this.field_184571_b instanceof EntityPlayer && ((EntityPlayer)this.field_184571_b).func_175149_v())
+                if (this.field_184571_b == null || !this.field_184571_b.func_70089_S() || this.field_184571_b instanceof EntityPlayer && ((EntityPlayer)this.field_184571_b).func_175149_v())
                 {
                     if (!this.func_189652_ae())
                     {
-                        this.field_70181_x -= 0.04;
+                        this.field_70181_x -= 0.04D;
                     }
                 }
                 else
                 {
-                    this.field_184577_e = MathHelper.func_151237_a(this.field_184577_e * 1.025, -1.0, 1.0);
-                    this.field_184578_f = MathHelper.func_151237_a(this.field_184578_f * 1.025, -1.0, 1.0);
-                    this.field_184579_g = MathHelper.func_151237_a(this.field_184579_g * 1.025, -1.0, 1.0);
-                    this.field_70159_w += (this.field_184577_e - this.field_70159_w) * 0.2;
-                    this.field_70181_x += (this.field_184578_f - this.field_70181_x) * 0.2;
-                    this.field_70179_y += (this.field_184579_g - this.field_70179_y) * 0.2;
+                    this.field_184577_e = MathHelper.func_151237_a(this.field_184577_e * 1.025D, -1.0D, 1.0D);
+                    this.field_184578_f = MathHelper.func_151237_a(this.field_184578_f * 1.025D, -1.0D, 1.0D);
+                    this.field_184579_g = MathHelper.func_151237_a(this.field_184579_g * 1.025D, -1.0D, 1.0D);
+                    this.field_70159_w += (this.field_184577_e - this.field_70159_w) * 0.2D;
+                    this.field_70181_x += (this.field_184578_f - this.field_70181_x) * 0.2D;
+                    this.field_70179_y += (this.field_184579_g - this.field_70179_y) * 0.2D;
                 }
 
                 RayTraceResult raytraceresult = ProjectileHelper.func_188802_a(this, true, false, this.field_184570_a);
 
-                if (raytraceresult != null)
+                if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
                 {
                     this.func_184567_a(raytraceresult);
                 }
@@ -336,16 +321,7 @@
 
             if (this.field_70170_p.field_72995_K)
             {
-                this.field_70170_p
-                .func_175688_a(
-                    EnumParticleTypes.END_ROD,
-                    this.field_70165_t - this.field_70159_w,
-                    this.field_70163_u - this.field_70181_x + 0.15,
-                    this.field_70161_v - this.field_70179_y,
-                    0.0,
-                    0.0,
-                    0.0
-                );
+                this.field_70170_p.func_175688_a(EnumParticleTypes.END_ROD, this.field_70165_t - this.field_70159_w, this.field_70163_u - this.field_70181_x + 0.15D, this.field_70161_v - this.field_70179_y, 0.0D, 0.0D, 0.0D);
             }
             else if (this.field_184571_b != null && !this.field_184571_b.field_70128_L)
             {
@@ -372,9 +348,7 @@
                     {
                         BlockPos blockpos1 = new BlockPos(this.field_184571_b);
 
-                        if (enumfacing$axis == EnumFacing.Axis.X && blockpos.func_177958_n() == blockpos1.func_177958_n()
-                                || enumfacing$axis == EnumFacing.Axis.Z && blockpos.func_177952_p() == blockpos1.func_177952_p()
-                                || enumfacing$axis == EnumFacing.Axis.Y && blockpos.func_177956_o() == blockpos1.func_177956_o())
+                        if (enumfacing$axis == EnumFacing.Axis.X && blockpos.func_177958_n() == blockpos1.func_177958_n() || enumfacing$axis == EnumFacing.Axis.Z && blockpos.func_177952_p() == blockpos1.func_177952_p() || enumfacing$axis == EnumFacing.Axis.Y && blockpos.func_177956_o() == blockpos1.func_177956_o())
                         {
                             this.func_184569_a(enumfacing$axis);
                         }
@@ -384,27 +358,23 @@
         }
     }
 
-    @Override
     public boolean func_70027_ad()
     {
         return false;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_70112_a(double p_70112_1_)
     {
-        return p_70112_1_ < 16384.0;
+        return p_70112_1_ < 16384.0D;
     }
 
-    @Override
     public float func_70013_c()
     {
         return 1.0F;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public int func_70070_b()
     {
         return 15728880;
@@ -414,8 +384,7 @@
     {
         if (p_184567_1_.field_72308_g == null)
         {
-            ((WorldServer)this.field_70170_p)
-            .func_175739_a(EnumParticleTypes.EXPLOSION_LARGE, this.field_70165_t, this.field_70163_u, this.field_70161_v, 2, 0.2, 0.2, 0.2, 0.0);
+            ((WorldServer)this.field_70170_p).func_175739_a(EnumParticleTypes.EXPLOSION_LARGE, this.field_70165_t, this.field_70163_u, this.field_70161_v, 2, 0.2D, 0.2D, 0.2D, 0.0D);
             this.func_184185_a(SoundEvents.field_187775_eP, 1.0F, 1.0F);
         }
         else
@@ -436,20 +405,17 @@
         this.func_70106_y();
     }
 
-    @Override
     public boolean func_70067_L()
     {
         return true;
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         if (!this.field_70170_p.field_72995_K)
         {
             this.func_184185_a(SoundEvents.field_187777_eQ, 1.0F, 1.0F);
-            ((WorldServer)this.field_70170_p)
-            .func_175739_a(EnumParticleTypes.CRIT, this.field_70165_t, this.field_70163_u, this.field_70161_v, 15, 0.2, 0.2, 0.2, 0.0);
+            ((WorldServer)this.field_70170_p).func_175739_a(EnumParticleTypes.CRIT, this.field_70165_t, this.field_70163_u, this.field_70161_v, 15, 0.2D, 0.2D, 0.2D, 0.0D);
             this.func_70106_y();
         }
 
