--- before/net/minecraft/entity/projectile/ProjectileHelper.java
+++ after/net/minecraft/entity/projectile/ProjectileHelper.java
@@ -40,7 +40,7 @@
 
                 if (entity1.func_70067_L() && (p_188802_2_ || !entity1.func_70028_i(p_188802_3_)) && !entity1.field_70145_X)
                 {
-                    AxisAlignedBB axisalignedbb = entity1.func_174813_aQ().func_186662_g((double)0.3F);
+                    AxisAlignedBB axisalignedbb = entity1.func_174813_aQ().func_186662_g(0.30000001192092896D);
                     RayTraceResult raytraceresult1 = axisalignedbb.func_72327_a(vec3d, vec3d1);
 
                     if (raytraceresult1 != null)
@@ -71,10 +71,11 @@
         double d1 = p_188803_0_.field_70181_x;
         double d2 = p_188803_0_.field_70179_y;
         float f = MathHelper.func_76133_a(d0 * d0 + d2 * d2);
-        p_188803_0_.field_70177_z = (float)(MathHelper.func_181159_b(d2, d0) * (double)(180F / (float)Math.PI)) + 90.0F;
+        p_188803_0_.field_70177_z = (float)(MathHelper.func_181159_b(d2, d0) * (180D / Math.PI)) + 90.0F;
 
-        for (p_188803_0_.field_70125_A = (float)(MathHelper.func_181159_b((double)f, d1) * (double)(180F / (float)Math.PI)) - 90.0F; p_188803_0_.field_70125_A - p_188803_0_.field_70127_C < -180.0F; p_188803_0_.field_70127_C -= 360.0F)
+        for (p_188803_0_.field_70125_A = (float)(MathHelper.func_181159_b((double)f, d1) * (180D / Math.PI)) - 90.0F; p_188803_0_.field_70125_A - p_188803_0_.field_70127_C < -180.0F; p_188803_0_.field_70127_C -= 360.0F)
         {
+            ;
         }
 
         while (p_188803_0_.field_70125_A - p_188803_0_.field_70127_C >= 180.0F)
