--- before/net/minecraft/entity/player/EnumPlayerModelParts.java
+++ after/net/minecraft/entity/player/EnumPlayerModelParts.java
@@ -26,7 +26,7 @@
         this.field_179340_h = p_i45809_3_;
         this.field_179341_i = 1 << p_i45809_3_;
         this.field_179338_j = p_i45809_4_;
-        this.field_179339_k = new TextComponentTranslation("options.modelPart." + p_i45809_4_);
+        this.field_179339_k = new TextComponentTranslation("options.modelPart." + p_i45809_4_, new Object[0]);
     }
 
     public int func_179327_a()
