--- before/net/minecraft/entity/player/InventoryPlayer.java
+++ after/net/minecraft/entity/player/InventoryPlayer.java
@@ -304,7 +304,8 @@
 
         if (itemstack.isEmpty())
         {
-            itemstack = new ItemStack(item, 0, p_191973_2_.getMetadata());
+            itemstack = p_191973_2_.copy(); // Forge: Replace Item clone above to preserve item capabilities when picking the item up.
+            itemstack.setCount(0);
 
             if (p_191973_2_.hasTagCompound())
             {
@@ -375,6 +376,13 @@
                 }
             }
         }
+        for (ItemStack is : armorInventory) // FORGE: Tick armor on animation ticks
+        {
+            if (!is.isEmpty())
+            {
+                is.getItem().onArmorTick(player.world, player, is);
+            }
+        }
     }
 
     public boolean addItemStackToInventory(ItemStack itemStackIn)
@@ -452,6 +460,8 @@
                 CrashReportCategory crashreportcategory = crashreport.makeCategory("Item being added");
                 crashreportcategory.addCrashSection("Item ID", Item.getIdFromItem(p_191971_2_.getItem()));
                 crashreportcategory.addCrashSection("Item data", p_191971_2_.getMetadata());
+                crashreportcategory.addDetail("Registry Name", () -> String.valueOf(p_191971_2_.getItem().getRegistryName()));
+                crashreportcategory.addDetail("Item Class", () -> p_191971_2_.getItem().getClass().getName());
                 crashreportcategory.addDetail("Item name", new ICrashReportDetail<String>()
                 {
                     public String call() throws Exception
