--- before/net/minecraft/entity/player/InventoryPlayer.java
+++ after/net/minecraft/entity/player/InventoryPlayer.java
@@ -1,6 +1,7 @@
 package net.minecraft.entity.player;
 
 import java.util.Arrays;
+import java.util.Iterator;
 import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.block.state.IBlockState;
@@ -28,23 +29,25 @@
 
 public class InventoryPlayer implements IInventory
 {
-    public final NonNullList<ItemStack> field_70462_a = NonNullList.func_191197_a(36, ItemStack.field_190927_a);
-    public final NonNullList<ItemStack> field_70460_b = NonNullList.func_191197_a(4, ItemStack.field_190927_a);
-    public final NonNullList<ItemStack> field_184439_c = NonNullList.func_191197_a(1, ItemStack.field_190927_a);
-    private final List<NonNullList<ItemStack>> field_184440_g = Arrays.asList(this.field_70462_a, this.field_70460_b, this.field_184439_c);
+    public final NonNullList<ItemStack> field_70462_a = NonNullList.<ItemStack>func_191197_a(36, ItemStack.field_190927_a);
+    public final NonNullList<ItemStack> field_70460_b = NonNullList.<ItemStack>func_191197_a(4, ItemStack.field_190927_a);
+    public final NonNullList<ItemStack> field_184439_c = NonNullList.<ItemStack>func_191197_a(1, ItemStack.field_190927_a);
+    private final List<NonNullList<ItemStack>> field_184440_g;
     public int field_70461_c;
     public EntityPlayer field_70458_d;
-    private ItemStack field_70457_g = ItemStack.field_190927_a;
+    private ItemStack field_70457_g;
     private int field_194017_h;
 
     public InventoryPlayer(EntityPlayer p_i1750_1_)
     {
+        this.field_184440_g = Arrays.<NonNullList<ItemStack>>asList(this.field_70462_a, this.field_70460_b, this.field_184439_c);
+        this.field_70457_g = ItemStack.field_190927_a;
         this.field_70458_d = p_i1750_1_;
     }
 
     public ItemStack func_70448_g()
     {
-        return func_184435_e(this.field_70461_c) ? this.field_70462_a.get(this.field_70461_c) : ItemStack.field_190927_a;
+        return func_184435_e(this.field_70461_c) ? (ItemStack)this.field_70462_a.get(this.field_70461_c) : ItemStack.field_190927_a;
     }
 
     public static int func_70451_h()
@@ -54,25 +57,19 @@
 
     private boolean func_184436_a(ItemStack p_184436_1_, ItemStack p_184436_2_)
     {
-        return !p_184436_1_.func_190926_b()
-               && this.func_184431_b(p_184436_1_, p_184436_2_)
-               && p_184436_1_.func_77985_e()
-               && p_184436_1_.func_190916_E() < p_184436_1_.func_77976_d()
-               && p_184436_1_.func_190916_E() < this.func_70297_j_();
+        return !p_184436_1_.func_190926_b() && this.func_184431_b(p_184436_1_, p_184436_2_) && p_184436_1_.func_77985_e() && p_184436_1_.func_190916_E() < p_184436_1_.func_77976_d() && p_184436_1_.func_190916_E() < this.func_70297_j_();
     }
 
     private boolean func_184431_b(ItemStack p_184431_1_, ItemStack p_184431_2_)
     {
-        return p_184431_1_.func_77973_b() == p_184431_2_.func_77973_b()
-               && (!p_184431_1_.func_77981_g() || p_184431_1_.func_77960_j() == p_184431_2_.func_77960_j())
-               && ItemStack.func_77970_a(p_184431_1_, p_184431_2_);
+        return p_184431_1_.func_77973_b() == p_184431_2_.func_77973_b() && (!p_184431_1_.func_77981_g() || p_184431_1_.func_77960_j() == p_184431_2_.func_77960_j()) && ItemStack.func_77970_a(p_184431_1_, p_184431_2_);
     }
 
     public int func_70447_i()
     {
         for (int i = 0; i < this.field_70462_a.size(); ++i)
         {
-            if (this.field_70462_a.get(i).func_190926_b())
+            if (((ItemStack)this.field_70462_a.get(i)).func_190926_b())
             {
                 return i;
             }
@@ -96,7 +93,7 @@
             {
                 this.field_70461_c = this.func_184433_k();
 
-                if (!this.field_70462_a.get(this.field_70461_c).func_190926_b())
+                if (!((ItemStack)this.field_70462_a.get(this.field_70461_c)).func_190926_b())
                 {
                     int j = this.func_70447_i();
 
@@ -133,7 +130,7 @@
     {
         for (int i = 0; i < this.field_70462_a.size(); ++i)
         {
-            if (!this.field_70462_a.get(i).func_190926_b() && this.func_184431_b(p_184429_1_, this.field_70462_a.get(i)))
+            if (!((ItemStack)this.field_70462_a.get(i)).func_190926_b() && this.func_184431_b(p_184429_1_, this.field_70462_a.get(i)))
             {
                 return i;
             }
@@ -148,11 +145,7 @@
         {
             ItemStack itemstack = this.field_70462_a.get(i);
 
-            if (!this.field_70462_a.get(i).func_190926_b()
-                    && this.func_184431_b(p_194014_1_, this.field_70462_a.get(i))
-                    && !this.field_70462_a.get(i).func_77951_h()
-                    && !itemstack.func_77948_v()
-                    && !itemstack.func_82837_s())
+            if (!((ItemStack)this.field_70462_a.get(i)).func_190926_b() && this.func_184431_b(p_194014_1_, this.field_70462_a.get(i)) && !((ItemStack)this.field_70462_a.get(i)).func_77951_h() && !itemstack.func_77948_v() && !itemstack.func_82837_s())
             {
                 return i;
             }
@@ -167,7 +160,7 @@
         {
             int j = (this.field_70461_c + i) % 9;
 
-            if (this.field_70462_a.get(j).func_190926_b())
+            if (((ItemStack)this.field_70462_a.get(j)).func_190926_b())
             {
                 return j;
             }
@@ -177,7 +170,7 @@
         {
             int l = (this.field_70461_c + k) % 9;
 
-            if (!this.field_70462_a.get(l).func_77948_v())
+            if (!((ItemStack)this.field_70462_a.get(l)).func_77948_v())
             {
                 return l;
             }
@@ -199,11 +192,9 @@
             p_70453_1_ = -1;
         }
 
-        this.field_70461_c -= p_70453_1_;
-
-        while (this.field_70461_c < 0)
+        for (this.field_70461_c -= p_70453_1_; this.field_70461_c < 0; this.field_70461_c += 9)
         {
-            this.field_70461_c += 9;
+            ;
         }
 
         while (this.field_70461_c >= 9)
@@ -220,10 +211,7 @@
         {
             ItemStack itemstack = this.func_70301_a(j);
 
-            if (!itemstack.func_190926_b()
-                    && (p_174925_1_ == null || itemstack.func_77973_b() == p_174925_1_)
-                    && (p_174925_2_ <= -1 || itemstack.func_77960_j() == p_174925_2_)
-                    && (p_174925_4_ == null || NBTUtil.func_181123_a(p_174925_4_, itemstack.func_77978_p(), true)))
+            if (!itemstack.func_190926_b() && (p_174925_1_ == null || itemstack.func_77973_b() == p_174925_1_) && (p_174925_2_ <= -1 || itemstack.func_77960_j() == p_174925_2_) && (p_174925_4_ == null || NBTUtil.func_181123_a(p_174925_4_, itemstack.func_77978_p(), true)))
             {
                 int k = p_174925_3_ <= 0 ? itemstack.func_190916_E() : Math.min(p_174925_3_ - i, itemstack.func_190916_E());
                 i += k;
@@ -304,7 +292,8 @@
 
         if (itemstack.func_190926_b())
         {
-            itemstack = new ItemStack(item, 0, p_191973_2_.func_77960_j());
+            itemstack = p_191973_2_.func_77946_l(); // Forge: Replace Item clone above to preserve item capabilities when picking the item up.
+            itemstack.func_190920_e(0);
 
             if (p_191973_2_.func_77942_o())
             {
@@ -332,7 +321,7 @@
         }
         else
         {
-            i -= j;
+            i = i - j;
             itemstack.func_190917_f(j);
             itemstack.func_190915_d(5);
             return i;
@@ -369,10 +358,17 @@
         {
             for (int i = 0; i < nonnulllist.size(); ++i)
             {
-                if (!nonnulllist.get(i).func_190926_b())
+                if (!((ItemStack)nonnulllist.get(i)).func_190926_b())
                 {
-                    nonnulllist.get(i).func_77945_a(this.field_70458_d.field_70170_p, this.field_70458_d, i, this.field_70461_c == i);
+                    ((ItemStack)nonnulllist.get(i)).func_77945_a(this.field_70458_d.field_70170_p, this.field_70458_d, i, this.field_70461_c == i);
                 }
+            }
+        }
+        for (ItemStack is : field_70460_b) // FORGE: Tick armor on animation ticks
+        {
+            if (!is.func_190926_b())
+            {
+                is.func_77973_b().onArmorTick(field_70458_d.field_70170_p, field_70458_d, is);
             }
         }
     }
@@ -402,7 +398,7 @@
                     if (p_191971_1_ >= 0)
                     {
                         this.field_70462_a.set(p_191971_1_, p_191971_2_.func_77946_l());
-                        this.field_70462_a.get(p_191971_1_).func_190915_d(5);
+                        ((ItemStack)this.field_70462_a.get(p_191971_1_)).func_190915_d(5);
                         p_191971_2_.func_190920_e(0);
                         return true;
                     }
@@ -420,7 +416,7 @@
                 {
                     int i;
 
-                    do
+                    while (true)
                     {
                         i = p_191971_2_.func_190916_E();
 
@@ -432,8 +428,12 @@
                         {
                             p_191971_2_.func_190920_e(this.func_191973_d(p_191971_1_, p_191971_2_));
                         }
+
+                        if (p_191971_2_.func_190926_b() || p_191971_2_.func_190916_E() >= i)
+                        {
+                            break;
+                        }
                     }
-                    while (!p_191971_2_.func_190926_b() && p_191971_2_.func_190916_E() < i);
 
                     if (p_191971_2_.func_190916_E() == i && this.field_70458_d.field_71075_bZ.field_75098_d)
                     {
@@ -450,8 +450,10 @@
             {
                 CrashReport crashreport = CrashReport.func_85055_a(throwable, "Adding item to inventory");
                 CrashReportCategory crashreportcategory = crashreport.func_85058_a("Item being added");
-                crashreportcategory.func_71507_a("Item ID", Item.func_150891_b(p_191971_2_.func_77973_b()));
-                crashreportcategory.func_71507_a("Item data", p_191971_2_.func_77960_j());
+                crashreportcategory.func_71507_a("Item ID", Integer.valueOf(Item.func_150891_b(p_191971_2_.func_77973_b())));
+                crashreportcategory.func_71507_a("Item data", Integer.valueOf(p_191971_2_.func_77960_j()));
+                crashreportcategory.func_189529_a("Registry Name", () -> String.valueOf(p_191971_2_.func_77973_b().getRegistryName()));
+                crashreportcategory.func_189529_a("Item Class", () -> p_191971_2_.func_77973_b().getClass().getName());
                 crashreportcategory.func_189529_a("Item name", new ICrashReportDetail<String>()
                 {
                     public String call() throws Exception
@@ -493,7 +495,6 @@
         }
     }
 
-    @Override
     public ItemStack func_70298_a(int p_70298_1_, int p_70298_2_)
     {
         List<ItemStack> list = null;
@@ -509,7 +510,7 @@
             p_70298_1_ -= nonnulllist.size();
         }
 
-        return list != null && !list.get(p_70298_1_).func_190926_b() ? ItemStackHelper.func_188382_a(list, p_70298_1_, p_70298_2_) : ItemStack.field_190927_a;
+        return list != null && !((ItemStack)list.get(p_70298_1_)).func_190926_b() ? ItemStackHelper.func_188382_a(list, p_70298_1_, p_70298_2_) : ItemStack.field_190927_a;
     }
 
     public void func_184437_d(ItemStack p_184437_1_)
@@ -527,7 +528,6 @@
         }
     }
 
-    @Override
     public ItemStack func_70304_b(int p_70304_1_)
     {
         NonNullList<ItemStack> nonnulllist = null;
@@ -543,7 +543,7 @@
             p_70304_1_ -= nonnulllist1.size();
         }
 
-        if (nonnulllist != null && !nonnulllist.get(p_70304_1_).func_190926_b())
+        if (nonnulllist != null && !((ItemStack)nonnulllist.get(p_70304_1_)).func_190926_b())
         {
             ItemStack itemstack = nonnulllist.get(p_70304_1_);
             nonnulllist.set(p_70304_1_, ItemStack.field_190927_a);
@@ -555,7 +555,6 @@
         }
     }
 
-    @Override
     public void func_70299_a(int p_70299_1_, ItemStack p_70299_2_)
     {
         NonNullList<ItemStack> nonnulllist = null;
@@ -581,9 +580,9 @@
     {
         float f = 1.0F;
 
-        if (!this.field_70462_a.get(this.field_70461_c).func_190926_b())
+        if (!((ItemStack)this.field_70462_a.get(this.field_70461_c)).func_190926_b())
         {
-            f *= this.field_70462_a.get(this.field_70461_c).func_150997_a(p_184438_1_);
+            f *= ((ItemStack)this.field_70462_a.get(this.field_70461_c)).func_150997_a(p_184438_1_);
         }
 
         return f;
@@ -593,33 +592,33 @@
     {
         for (int i = 0; i < this.field_70462_a.size(); ++i)
         {
-            if (!this.field_70462_a.get(i).func_190926_b())
+            if (!((ItemStack)this.field_70462_a.get(i)).func_190926_b())
             {
                 NBTTagCompound nbttagcompound = new NBTTagCompound();
                 nbttagcompound.func_74774_a("Slot", (byte)i);
-                this.field_70462_a.get(i).func_77955_b(nbttagcompound);
+                ((ItemStack)this.field_70462_a.get(i)).func_77955_b(nbttagcompound);
                 p_70442_1_.func_74742_a(nbttagcompound);
             }
         }
 
         for (int j = 0; j < this.field_70460_b.size(); ++j)
         {
-            if (!this.field_70460_b.get(j).func_190926_b())
+            if (!((ItemStack)this.field_70460_b.get(j)).func_190926_b())
             {
                 NBTTagCompound nbttagcompound1 = new NBTTagCompound();
                 nbttagcompound1.func_74774_a("Slot", (byte)(j + 100));
-                this.field_70460_b.get(j).func_77955_b(nbttagcompound1);
+                ((ItemStack)this.field_70460_b.get(j)).func_77955_b(nbttagcompound1);
                 p_70442_1_.func_74742_a(nbttagcompound1);
             }
         }
 
         for (int k = 0; k < this.field_184439_c.size(); ++k)
         {
-            if (!this.field_184439_c.get(k).func_190926_b())
+            if (!((ItemStack)this.field_184439_c.get(k)).func_190926_b())
             {
                 NBTTagCompound nbttagcompound2 = new NBTTagCompound();
                 nbttagcompound2.func_74774_a("Slot", (byte)(k + 150));
-                this.field_184439_c.get(k).func_77955_b(nbttagcompound2);
+                ((ItemStack)this.field_184439_c.get(k)).func_77955_b(nbttagcompound2);
                 p_70442_1_.func_74742_a(nbttagcompound2);
             }
         }
@@ -657,13 +656,11 @@
         }
     }
 
-    @Override
     public int func_70302_i_()
     {
         return this.field_70462_a.size() + this.field_70460_b.size() + this.field_184439_c.size();
     }
 
-    @Override
     public boolean func_191420_l()
     {
         for (ItemStack itemstack : this.field_70462_a)
@@ -693,7 +690,6 @@
         return true;
     }
 
-    @Override
     public ItemStack func_70301_a(int p_70301_1_)
     {
         List<ItemStack> list = null;
@@ -709,28 +705,24 @@
             p_70301_1_ -= nonnulllist.size();
         }
 
-        return list == null ? ItemStack.field_190927_a : list.get(p_70301_1_);
+        return list == null ? ItemStack.field_190927_a : (ItemStack)list.get(p_70301_1_);
     }
 
-    @Override
     public String func_70005_c_()
     {
         return "container.inventory";
     }
 
-    @Override
     public boolean func_145818_k_()
     {
         return false;
     }
 
-    @Override
     public ITextComponent func_145748_c_()
     {
-        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_()));
+        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_(), new Object[0]));
     }
 
-    @Override
     public int func_70297_j_()
     {
         return 64;
@@ -757,7 +749,7 @@
 
     public void func_70449_g(float p_70449_1_)
     {
-        p_70449_1_ /= 4.0F;
+        p_70449_1_ = p_70449_1_ / 4.0F;
 
         if (p_70449_1_ < 1.0F)
         {
@@ -792,7 +784,6 @@
         }
     }
 
-    @Override
     public void func_70296_d()
     {
         ++this.field_194017_h;
@@ -814,7 +805,6 @@
         return this.field_70457_g;
     }
 
-    @Override
     public boolean func_70300_a(EntityPlayer p_70300_1_)
     {
         if (this.field_70458_d.field_70128_L)
@@ -823,37 +813,47 @@
         }
         else
         {
-            return !(p_70300_1_.func_70068_e(this.field_70458_d) > 64.0);
+            return p_70300_1_.func_70068_e(this.field_70458_d) <= 64.0D;
         }
     }
 
     public boolean func_70431_c(ItemStack p_70431_1_)
     {
+        label23:
+
         for (List<ItemStack> list : this.field_184440_g)
         {
-            for (ItemStack itemstack : list)
+            Iterator iterator = list.iterator();
+
+            while (true)
             {
+                if (!iterator.hasNext())
+                {
+                    continue label23;
+                }
+
+                ItemStack itemstack = (ItemStack)iterator.next();
+
                 if (!itemstack.func_190926_b() && itemstack.func_77969_a(p_70431_1_))
                 {
-                    return true;
+                    break;
                 }
             }
+
+            return true;
         }
 
         return false;
     }
 
-    @Override
     public void func_174889_b(EntityPlayer p_174889_1_)
     {
     }
 
-    @Override
     public void func_174886_c(EntityPlayer p_174886_1_)
     {
     }
 
-    @Override
     public boolean func_94041_b(int p_94041_1_, ItemStack p_94041_2_)
     {
         return true;
@@ -869,24 +869,20 @@
         this.field_70461_c = p_70455_1_.field_70461_c;
     }
 
-    @Override
     public int func_174887_a_(int p_174887_1_)
     {
         return 0;
     }
 
-    @Override
     public void func_174885_b(int p_174885_1_, int p_174885_2_)
     {
     }
 
-    @Override
     public int func_174890_g()
     {
         return 0;
     }
 
-    @Override
     public void func_174888_l()
     {
         for (List<ItemStack> list : this.field_184440_g)
