--- before/net/minecraft/realms/RealmsScreen.java
+++ after/net/minecraft/realms/RealmsScreen.java
@@ -98,7 +98,7 @@
     {
         for (int i = 0; i < this.proxy.func_154320_j().size(); ++i)
         {
-            this.proxy.func_154320_j().get(i).render(p_render_1_, p_render_2_, p_render_3_);
+            ((RealmsButton)this.proxy.func_154320_j().get(i)).render(p_render_1_, p_render_2_, p_render_3_);
         }
     }
 
