--- before/net/minecraft/realms/RealmsConnect.java
+++ after/net/minecraft/realms/RealmsConnect.java
@@ -38,6 +38,7 @@
 
                 try
                 {
+                    net.minecraftforge.fml.client.FMLClientHandler.instance().connectToRealmsServer(p_connect_1_, p_connect_2_);
                     inetaddress = InetAddress.getByName(p_connect_1_);
 
                     if (RealmsConnect.this.aborted)
@@ -59,7 +60,7 @@
                         return;
                     }
 
-                    RealmsConnect.this.connection.func_179290_a(new C00Handshake(p_connect_1_, p_connect_2_, EnumConnectionState.LOGIN));
+                    RealmsConnect.this.connection.func_179290_a(new C00Handshake(p_connect_1_, p_connect_2_, EnumConnectionState.LOGIN, true));
 
                     if (RealmsConnect.this.aborted)
                     {
@@ -78,7 +79,7 @@
                     }
 
                     RealmsConnect.LOGGER.error("Couldn't connect to world", (Throwable)unknownhostexception);
-                    Realms.setScreen(new DisconnectedRealmsScreen(RealmsConnect.this.onlineScreen, "connect.failed", new TextComponentTranslation("disconnect.genericReason", "Unknown host '" + p_connect_1_ + "'")));
+                    Realms.setScreen(new DisconnectedRealmsScreen(RealmsConnect.this.onlineScreen, "connect.failed", new TextComponentTranslation("disconnect.genericReason", new Object[] {"Unknown host '" + p_connect_1_ + "'"})));
                 }
                 catch (Exception exception)
                 {
@@ -98,7 +99,7 @@
                         s = s.replaceAll(s1, "");
                     }
 
-                    Realms.setScreen(new DisconnectedRealmsScreen(RealmsConnect.this.onlineScreen, "connect.failed", new TextComponentTranslation("disconnect.genericReason", s)));
+                    Realms.setScreen(new DisconnectedRealmsScreen(RealmsConnect.this.onlineScreen, "connect.failed", new TextComponentTranslation("disconnect.genericReason", new Object[] {s})));
                 }
             }
         }).start();
@@ -110,7 +111,7 @@
 
         if (this.connection != null && this.connection.func_150724_d())
         {
-            this.connection.func_150718_a(new TextComponentTranslation("disconnect.genericReason"));
+            this.connection.func_150718_a(new TextComponentTranslation("disconnect.genericReason", new Object[0]));
             this.connection.func_179293_l();
         }
     }
