--- before/net/minecraft/realms/RealmsVertexFormat.java
+++ after/net/minecraft/realms/RealmsVertexFormat.java
@@ -70,7 +70,7 @@
 
     public List<RealmsVertexFormatElement> getElements()
     {
-        List<RealmsVertexFormatElement> list = Lists.newArrayList();
+        List<RealmsVertexFormatElement> list = Lists.<RealmsVertexFormatElement>newArrayList();
 
         for (VertexFormatElement vertexformatelement : this.v.func_177343_g())
         {
