--- before/net/minecraft/potion/PotionUtils.java
+++ after/net/minecraft/potion/PotionUtils.java
@@ -5,6 +5,7 @@
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import javax.annotation.Nullable;
 import net.minecraft.entity.ai.attributes.AttributeModifier;
 import net.minecraft.entity.ai.attributes.IAttribute;
@@ -28,7 +29,7 @@
 
     public static List<PotionEffect> func_185186_a(PotionType p_185186_0_, Collection<PotionEffect> p_185186_1_)
     {
-        List<PotionEffect> list = Lists.newArrayList();
+        List<PotionEffect> list = Lists.<PotionEffect>newArrayList();
         list.addAll(p_185186_0_.func_185170_a());
         list.addAll(p_185186_1_);
         return list;
@@ -36,7 +37,7 @@
 
     public static List<PotionEffect> func_185185_a(@Nullable NBTTagCompound p_185185_0_)
     {
-        List<PotionEffect> list = Lists.newArrayList();
+        List<PotionEffect> list = Lists.<PotionEffect>newArrayList();
         list.addAll(func_185187_c(p_185185_0_).func_185170_a());
         func_185193_a(p_185185_0_, list);
         return list;
@@ -49,7 +50,7 @@
 
     public static List<PotionEffect> func_185192_b(@Nullable NBTTagCompound p_185192_0_)
     {
-        List<PotionEffect> list = Lists.newArrayList();
+        List<PotionEffect> list = Lists.<PotionEffect>newArrayList();
         func_185193_a(p_185192_0_, list);
         return list;
     }
@@ -179,7 +180,7 @@
         }
         else
         {
-            NBTTagCompound nbttagcompound = MoreObjects.firstNonNull(p_185184_0_.func_77978_p(), new NBTTagCompound());
+            NBTTagCompound nbttagcompound = (NBTTagCompound)MoreObjects.firstNonNull(p_185184_0_.func_77978_p(), new NBTTagCompound());
             NBTTagList nbttaglist = nbttagcompound.func_150295_c("CustomPotionEffects", 9);
 
             for (PotionEffect potioneffect : p_185184_1_)
@@ -197,7 +198,7 @@
     public static void func_185182_a(ItemStack p_185182_0_, List<String> p_185182_1_, float p_185182_2_)
     {
         List<PotionEffect> list = func_185189_a(p_185182_0_);
-        List<Tuple<String, AttributeModifier>> list1 = Lists.newArrayList();
+        List<Tuple<String, AttributeModifier>> list1 = Lists.<Tuple<String, AttributeModifier>>newArrayList();
 
         if (list.isEmpty())
         {
@@ -214,11 +215,11 @@
 
                 if (!map.isEmpty())
                 {
-                    for (Map.Entry<IAttribute, AttributeModifier> entry : map.entrySet())
+                    for (Entry<IAttribute, AttributeModifier> entry : map.entrySet())
                     {
                         AttributeModifier attributemodifier = entry.getValue();
                         AttributeModifier attributemodifier1 = new AttributeModifier(attributemodifier.func_111166_b(), potion.func_111183_a(potioneffect.func_76458_c(), attributemodifier), attributemodifier.func_111169_c());
-                        list1.add(new Tuple<>(entry.getKey().func_111108_a(), attributemodifier1));
+                        list1.add(new Tuple(((IAttribute)entry.getKey()).func_111108_a(), attributemodifier1));
                     }
                 }
 
@@ -269,7 +270,7 @@
                 }
                 else if (d0 < 0.0D)
                 {
-                    d1 *= -1.0D;
+                    d1 = d1 * -1.0D;
                     p_185182_1_.add(TextFormatting.RED + I18n.func_74837_a("attribute.modifier.take." + attributemodifier2.func_111169_c(), ItemStack.field_111284_a.format(d1), I18n.func_74838_a("attribute.name." + (String)tuple.func_76341_a())));
                 }
             }
