--- before/net/minecraft/advancements/critereon/BrewedPotionTrigger.java
+++ after/net/minecraft/advancements/critereon/BrewedPotionTrigger.java
@@ -20,7 +20,7 @@
 public class BrewedPotionTrigger implements ICriterionTrigger<BrewedPotionTrigger.Instance>
 {
     private static final ResourceLocation field_192176_a = new ResourceLocation("brewed_potion");
-    private final Map<PlayerAdvancements, BrewedPotionTrigger.Listeners> field_192177_b = Maps.newHashMap();
+    private final Map<PlayerAdvancements, BrewedPotionTrigger.Listeners> field_192177_b = Maps.<PlayerAdvancements, BrewedPotionTrigger.Listeners>newHashMap();
 
     public ResourceLocation func_192163_a()
     {
@@ -108,7 +108,7 @@
     static class Listeners
         {
             private final PlayerAdvancements field_192350_a;
-            private final Set<ICriterionTrigger.Listener<BrewedPotionTrigger.Instance>> field_192351_b = Sets.newHashSet();
+            private final Set<ICriterionTrigger.Listener<BrewedPotionTrigger.Instance>> field_192351_b = Sets.<ICriterionTrigger.Listener<BrewedPotionTrigger.Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements p_i47399_1_)
             {
@@ -136,11 +136,11 @@
 
                 for (ICriterionTrigger.Listener<BrewedPotionTrigger.Instance> listener : this.field_192351_b)
                 {
-                    if (listener.func_192158_a().func_192250_a(p_192348_1_))
+                    if (((BrewedPotionTrigger.Instance)listener.func_192158_a()).func_192250_a(p_192348_1_))
                     {
                         if (list == null)
                         {
-                            list = Lists.newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<BrewedPotionTrigger.Instance>>newArrayList();
                         }
 
                         list.add(listener);
