--- before/net/minecraft/advancements/critereon/ChangeDimensionTrigger.java
+++ after/net/minecraft/advancements/critereon/ChangeDimensionTrigger.java
@@ -19,7 +19,7 @@
 public class ChangeDimensionTrigger implements ICriterionTrigger<ChangeDimensionTrigger.Instance>
 {
     private static final ResourceLocation field_193144_a = new ResourceLocation("changed_dimension");
-    private final Map<PlayerAdvancements, ChangeDimensionTrigger.Listeners> field_193145_b = Maps.newHashMap();
+    private final Map<PlayerAdvancements, ChangeDimensionTrigger.Listeners> field_193145_b = Maps.<PlayerAdvancements, ChangeDimensionTrigger.Listeners>newHashMap();
 
     public ResourceLocation func_192163_a()
     {
@@ -106,7 +106,7 @@
     static class Listeners
         {
             private final PlayerAdvancements field_193235_a;
-            private final Set<ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance>> field_193236_b = Sets.newHashSet();
+            private final Set<ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance>> field_193236_b = Sets.<ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements p_i47476_1_)
             {
@@ -134,11 +134,11 @@
 
                 for (ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance> listener : this.field_193236_b)
                 {
-                    if (listener.func_192158_a().func_193190_a(p_193234_1_, p_193234_2_))
+                    if (((ChangeDimensionTrigger.Instance)listener.func_192158_a()).func_193190_a(p_193234_1_, p_193234_2_))
                     {
                         if (list == null)
                         {
-                            list = Lists.newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<ChangeDimensionTrigger.Instance>>newArrayList();
                         }
 
                         list.add(listener);
