--- before/net/minecraft/advancements/critereon/DamagePredicate.java
+++ after/net/minecraft/advancements/critereon/DamagePredicate.java
@@ -52,7 +52,7 @@
         {
             return false;
         }
-        else if (this.field_192370_e != null && this.field_192370_e != p_192365_5_)
+        else if (this.field_192370_e != null && this.field_192370_e.booleanValue() != p_192365_5_)
         {
             return false;
         }
