--- before/net/minecraft/advancements/critereon/DamageSourcePredicate.java
+++ after/net/minecraft/advancements/critereon/DamageSourcePredicate.java
@@ -52,31 +52,31 @@
         {
             return true;
         }
-        else if (this.field_192450_b != null && this.field_192450_b != p_193418_2_.func_76352_a())
-        {
-            return false;
-        }
-        else if (this.field_192451_c != null && this.field_192451_c != p_193418_2_.func_94541_c())
-        {
-            return false;
-        }
-        else if (this.field_192452_d != null && this.field_192452_d != p_193418_2_.func_76363_c())
-        {
-            return false;
-        }
-        else if (this.field_192453_e != null && this.field_192453_e != p_193418_2_.func_76357_e())
-        {
-            return false;
-        }
-        else if (this.field_192454_f != null && this.field_192454_f != p_193418_2_.func_151517_h())
-        {
-            return false;
-        }
-        else if (this.field_192455_g != null && this.field_192455_g != p_193418_2_.func_76347_k())
-        {
-            return false;
-        }
-        else if (this.field_192456_h != null && this.field_192456_h != p_193418_2_.func_82725_o())
+        else if (this.field_192450_b != null && this.field_192450_b.booleanValue() != p_193418_2_.func_76352_a())
+        {
+            return false;
+        }
+        else if (this.field_192451_c != null && this.field_192451_c.booleanValue() != p_193418_2_.func_94541_c())
+        {
+            return false;
+        }
+        else if (this.field_192452_d != null && this.field_192452_d.booleanValue() != p_193418_2_.func_76363_c())
+        {
+            return false;
+        }
+        else if (this.field_192453_e != null && this.field_192453_e.booleanValue() != p_193418_2_.func_76357_e())
+        {
+            return false;
+        }
+        else if (this.field_192454_f != null && this.field_192454_f.booleanValue() != p_193418_2_.func_151517_h())
+        {
+            return false;
+        }
+        else if (this.field_192455_g != null && this.field_192455_g.booleanValue() != p_193418_2_.func_76347_k())
+        {
+            return false;
+        }
+        else if (this.field_192456_h != null && this.field_192456_h.booleanValue() != p_193418_2_.func_82725_o())
         {
             return false;
         }
