--- before/net/minecraft/advancements/critereon/EffectsChangedTrigger.java
+++ after/net/minecraft/advancements/critereon/EffectsChangedTrigger.java
@@ -16,7 +16,7 @@
 public class EffectsChangedTrigger implements ICriterionTrigger<EffectsChangedTrigger.Instance>
 {
     private static final ResourceLocation field_193154_a = new ResourceLocation("effects_changed");
-    private final Map<PlayerAdvancements, EffectsChangedTrigger.Listeners> field_193155_b = Maps.newHashMap();
+    private final Map<PlayerAdvancements, EffectsChangedTrigger.Listeners> field_193155_b = Maps.<PlayerAdvancements, EffectsChangedTrigger.Listeners>newHashMap();
 
     public ResourceLocation func_192163_a()
     {
@@ -91,7 +91,7 @@
     static class Listeners
         {
             private final PlayerAdvancements field_193433_a;
-            private final Set<ICriterionTrigger.Listener<EffectsChangedTrigger.Instance>> field_193434_b = Sets.newHashSet();
+            private final Set<ICriterionTrigger.Listener<EffectsChangedTrigger.Instance>> field_193434_b = Sets.<ICriterionTrigger.Listener<EffectsChangedTrigger.Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements p_i47546_1_)
             {
@@ -119,11 +119,11 @@
 
                 for (ICriterionTrigger.Listener<EffectsChangedTrigger.Instance> listener : this.field_193434_b)
                 {
-                    if (listener.func_192158_a().func_193195_a(p_193432_1_))
+                    if (((EffectsChangedTrigger.Instance)listener.func_192158_a()).func_193195_a(p_193432_1_))
                     {
                         if (list == null)
                         {
-                            list = Lists.newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<EffectsChangedTrigger.Instance>>newArrayList();
                         }
 
                         list.add(listener);
