--- before/net/minecraft/advancements/critereon/MinMaxBounds.java
+++ after/net/minecraft/advancements/critereon/MinMaxBounds.java
@@ -19,25 +19,25 @@
 
     public boolean func_192514_a(float p_192514_1_)
     {
-        if (this.field_192517_b != null && this.field_192517_b > p_192514_1_)
+        if (this.field_192517_b != null && this.field_192517_b.floatValue() > p_192514_1_)
         {
             return false;
         }
         else
         {
-            return this.field_192518_c == null || !(this.field_192518_c < p_192514_1_);
+            return this.field_192518_c == null || this.field_192518_c.floatValue() >= p_192514_1_;
         }
     }
 
     public boolean func_192513_a(double p_192513_1_)
     {
-        if (this.field_192517_b != null && (double)(this.field_192517_b * this.field_192517_b) > p_192513_1_)
+        if (this.field_192517_b != null && (double)(this.field_192517_b.floatValue() * this.field_192517_b.floatValue()) > p_192513_1_)
         {
             return false;
         }
         else
         {
-            return this.field_192518_c == null || !((double)(this.field_192518_c * this.field_192518_c) < p_192513_1_);
+            return this.field_192518_c == null || (double)(this.field_192518_c.floatValue() * this.field_192518_c.floatValue()) >= p_192513_1_;
         }
     }
 
