--- before/net/minecraft/advancements/critereon/PlacedBlockTrigger.java
+++ after/net/minecraft/advancements/critereon/PlacedBlockTrigger.java
@@ -11,6 +11,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.Map.Entry;
 import javax.annotation.Nullable;
 import net.minecraft.advancements.ICriterionTrigger;
 import net.minecraft.advancements.PlayerAdvancements;
@@ -28,7 +29,7 @@
 public class PlacedBlockTrigger implements ICriterionTrigger<PlacedBlockTrigger.Instance>
 {
     private static final ResourceLocation field_193174_a = new ResourceLocation("placed_block");
-    private final Map<PlayerAdvancements, PlacedBlockTrigger.Listeners> field_193175_b = Maps.newHashMap();
+    private final Map<PlayerAdvancements, PlacedBlockTrigger.Listeners> field_193175_b = Maps.<PlayerAdvancements, PlacedBlockTrigger.Listeners>newHashMap();
 
     public ResourceLocation func_192163_a()
     {
@@ -95,7 +96,7 @@
 
             BlockStateContainer blockstatecontainer = block.func_176194_O();
 
-            for (Map.Entry<String, JsonElement> entry : JsonUtils.func_152754_s(p_192166_1_, "state").entrySet())
+            for (Entry<String, JsonElement> entry : JsonUtils.func_152754_s(p_192166_1_, "state").entrySet())
             {
                 IProperty<?> iproperty = blockstatecontainer.func_185920_a(entry.getKey());
 
@@ -114,7 +115,7 @@
 
                 if (map == null)
                 {
-                    map = Maps.newHashMap();
+                    map = Maps. < IProperty<?>, Object > newHashMap();
                 }
 
                 map.put(iproperty, optional.get());
@@ -163,7 +164,7 @@
                 {
                     if (this.field_193212_b != null)
                     {
-                        for (Map.Entry < IProperty<?>, Object > entry : this.field_193212_b.entrySet())
+                        for (Entry < IProperty<?>, Object > entry : this.field_193212_b.entrySet())
                         {
                             if (p_193210_1_.func_177229_b(entry.getKey()) != entry.getValue())
                             {
@@ -187,7 +188,7 @@
     static class Listeners
         {
             private final PlayerAdvancements field_193491_a;
-            private final Set<ICriterionTrigger.Listener<PlacedBlockTrigger.Instance>> field_193492_b = Sets.newHashSet();
+            private final Set<ICriterionTrigger.Listener<PlacedBlockTrigger.Instance>> field_193492_b = Sets.<ICriterionTrigger.Listener<PlacedBlockTrigger.Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements p_i47567_1_)
             {
@@ -215,11 +216,11 @@
 
                 for (ICriterionTrigger.Listener<PlacedBlockTrigger.Instance> listener : this.field_193492_b)
                 {
-                    if (listener.func_192158_a().func_193210_a(p_193489_1_, p_193489_2_, p_193489_3_, p_193489_4_))
+                    if (((PlacedBlockTrigger.Instance)listener.func_192158_a()).func_193210_a(p_193489_1_, p_193489_2_, p_193489_3_, p_193489_4_))
                     {
                         if (list == null)
                         {
-                            list = Lists.newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<PlacedBlockTrigger.Instance>>newArrayList();
                         }
 
                         list.add(listener);
