--- before/net/minecraft/advancements/critereon/TickTrigger.java
+++ after/net/minecraft/advancements/critereon/TickTrigger.java
@@ -15,7 +15,7 @@
 public class TickTrigger implements ICriterionTrigger<TickTrigger.Instance>
 {
     public static final ResourceLocation field_193183_a = new ResourceLocation("tick");
-    private final Map<PlayerAdvancements, TickTrigger.Listeners> field_193184_b = Maps.newHashMap();
+    private final Map<PlayerAdvancements, TickTrigger.Listeners> field_193184_b = Maps.<PlayerAdvancements, TickTrigger.Listeners>newHashMap();
 
     public ResourceLocation func_192163_a()
     {
@@ -81,7 +81,7 @@
     static class Listeners
         {
             private final PlayerAdvancements field_193504_a;
-            private final Set<ICriterionTrigger.Listener<TickTrigger.Instance>> field_193505_b = Sets.newHashSet();
+            private final Set<ICriterionTrigger.Listener<TickTrigger.Instance>> field_193505_b = Sets.<ICriterionTrigger.Listener<TickTrigger.Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements p_i47496_1_)
             {
