--- before/net/minecraft/advancements/critereon/UsedEnderEyeTrigger.java
+++ after/net/minecraft/advancements/critereon/UsedEnderEyeTrigger.java
@@ -17,7 +17,7 @@
 public class UsedEnderEyeTrigger implements ICriterionTrigger<UsedEnderEyeTrigger.Instance>
 {
     private static final ResourceLocation field_192242_a = new ResourceLocation("used_ender_eye");
-    private final Map<PlayerAdvancements, UsedEnderEyeTrigger.Listeners> field_192243_b = Maps.newHashMap();
+    private final Map<PlayerAdvancements, UsedEnderEyeTrigger.Listeners> field_192243_b = Maps.<PlayerAdvancements, UsedEnderEyeTrigger.Listeners>newHashMap();
 
     public ResourceLocation func_192163_a()
     {
@@ -94,7 +94,7 @@
     static class Listeners
         {
             private final PlayerAdvancements field_192547_a;
-            private final Set<ICriterionTrigger.Listener<UsedEnderEyeTrigger.Instance>> field_192548_b = Sets.newHashSet();
+            private final Set<ICriterionTrigger.Listener<UsedEnderEyeTrigger.Instance>> field_192548_b = Sets.<ICriterionTrigger.Listener<UsedEnderEyeTrigger.Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements p_i47450_1_)
             {
@@ -122,11 +122,11 @@
 
                 for (ICriterionTrigger.Listener<UsedEnderEyeTrigger.Instance> listener : this.field_192548_b)
                 {
-                    if (listener.func_192158_a().func_192288_a(p_192543_1_))
+                    if (((UsedEnderEyeTrigger.Instance)listener.func_192158_a()).func_192288_a(p_192543_1_))
                     {
                         if (list == null)
                         {
-                            list = Lists.newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<UsedEnderEyeTrigger.Instance>>newArrayList();
                         }
 
                         list.add(listener);
