--- before/net/minecraft/advancements/critereon/UsedTotemTrigger.java
+++ after/net/minecraft/advancements/critereon/UsedTotemTrigger.java
@@ -17,7 +17,7 @@
 public class UsedTotemTrigger implements ICriterionTrigger<UsedTotemTrigger.Instance>
 {
     private static final ResourceLocation field_193188_a = new ResourceLocation("used_totem");
-    private final Map<PlayerAdvancements, UsedTotemTrigger.Listeners> field_193189_b = Maps.newHashMap();
+    private final Map<PlayerAdvancements, UsedTotemTrigger.Listeners> field_193189_b = Maps.<PlayerAdvancements, UsedTotemTrigger.Listeners>newHashMap();
 
     public ResourceLocation func_192163_a()
     {
@@ -92,7 +92,7 @@
     static class Listeners
         {
             private final PlayerAdvancements field_193510_a;
-            private final Set<ICriterionTrigger.Listener<UsedTotemTrigger.Instance>> field_193511_b = Sets.newHashSet();
+            private final Set<ICriterionTrigger.Listener<UsedTotemTrigger.Instance>> field_193511_b = Sets.<ICriterionTrigger.Listener<UsedTotemTrigger.Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements p_i47565_1_)
             {
@@ -120,11 +120,11 @@
 
                 for (ICriterionTrigger.Listener<UsedTotemTrigger.Instance> listener : this.field_193511_b)
                 {
-                    if (listener.func_192158_a().func_193218_a(p_193509_1_))
+                    if (((UsedTotemTrigger.Instance)listener.func_192158_a()).func_193218_a(p_193509_1_))
                     {
                         if (list == null)
                         {
-                            list = Lists.newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<UsedTotemTrigger.Instance>>newArrayList();
                         }
 
                         list.add(listener);
