--- before/net/minecraft/advancements/critereon/VillagerTradeTrigger.java
+++ after/net/minecraft/advancements/critereon/VillagerTradeTrigger.java
@@ -18,7 +18,7 @@
 public class VillagerTradeTrigger implements ICriterionTrigger<VillagerTradeTrigger.Instance>
 {
     private static final ResourceLocation field_192237_a = new ResourceLocation("villager_trade");
-    private final Map<PlayerAdvancements, VillagerTradeTrigger.Listeners> field_192238_b = Maps.newHashMap();
+    private final Map<PlayerAdvancements, VillagerTradeTrigger.Listeners> field_192238_b = Maps.<PlayerAdvancements, VillagerTradeTrigger.Listeners>newHashMap();
 
     public ResourceLocation func_192163_a()
     {
@@ -103,7 +103,7 @@
     static class Listeners
         {
             private final PlayerAdvancements field_192541_a;
-            private final Set<ICriterionTrigger.Listener<VillagerTradeTrigger.Instance>> field_192542_b = Sets.newHashSet();
+            private final Set<ICriterionTrigger.Listener<VillagerTradeTrigger.Instance>> field_192542_b = Sets.<ICriterionTrigger.Listener<VillagerTradeTrigger.Instance>>newHashSet();
 
             public Listeners(PlayerAdvancements p_i47458_1_)
             {
@@ -131,11 +131,11 @@
 
                 for (ICriterionTrigger.Listener<VillagerTradeTrigger.Instance> listener : this.field_192542_b)
                 {
-                    if (listener.func_192158_a().func_192285_a(p_192537_1_, p_192537_2_, p_192537_3_))
+                    if (((VillagerTradeTrigger.Instance)listener.func_192158_a()).func_192285_a(p_192537_1_, p_192537_2_, p_192537_3_))
                     {
                         if (list == null)
                         {
-                            list = Lists.newArrayList();
+                            list = Lists.<ICriterionTrigger.Listener<VillagerTradeTrigger.Instance>>newArrayList();
                         }
 
                         list.add(listener);
