--- before/net/minecraft/item/crafting/CraftingManager.java
+++ after/net/minecraft/item/crafting/CraftingManager.java
@@ -6,6 +6,7 @@
 import com.google.gson.JsonParseException;
 import com.google.gson.JsonSyntaxException;
 import java.io.BufferedReader;
+import java.io.Closeable;
 import java.io.IOException;
 import java.io.Reader;
 import java.net.URI;
@@ -35,7 +36,7 @@
 {
     private static final Logger field_192422_a = LogManager.getLogger();
     private static int field_193381_c;
-    public static final RegistryNamespaced<ResourceLocation, IRecipe> field_193380_a = new RegistryNamespaced<>();
+    public static final RegistryNamespaced<ResourceLocation, IRecipe> field_193380_a = net.minecraftforge.registries.GameData.getWrapper(IRecipe.class);
 
     public static boolean func_193377_a()
     {
@@ -54,7 +55,7 @@
             func_193379_a("shulkerboxcoloring", new ShulkerBoxRecipes.ShulkerBoxColoring());
             return func_192420_c();
         }
-        catch (Throwable throwable)
+        catch (Throwable var1)
         {
             return false;
         }
@@ -63,18 +64,14 @@
     private static boolean func_192420_c()
     {
         FileSystem filesystem = null;
-        Gson gson = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
+        Gson gson = (new GsonBuilder()).setPrettyPrinting().disableHtmlEscaping().create();
+        boolean flag1;
 
         try
         {
             URL url = CraftingManager.class.getResource("/assets/.mcassetsroot");
 
-            if (url == null)
-            {
-                field_192422_a.error("Couldn't find .mcassetsroot");
-                return false;
-            }
-            else
+            if (url != null)
             {
                 URI uri = url.toURI();
                 Path path;
@@ -88,7 +85,8 @@
                     if (!"jar".equals(uri.getScheme()))
                     {
                         field_192422_a.error("Unsupported scheme " + uri + " trying to list all recipes");
-                        return false;
+                        boolean flag2 = false;
+                        return flag2;
                     }
 
                     filesystem = FileSystems.newFileSystem(uri, Collections.emptyMap());
@@ -110,18 +108,25 @@
 
                         try
                         {
-                            bufferedreader = Files.newBufferedReader(path1);
-                            func_193379_a(s, func_193376_a(JsonUtils.func_193839_a(gson, bufferedreader, JsonObject.class)));
-                        }
-                        catch (JsonParseException jsonparseexception)
-                        {
-                            field_192422_a.error("Parsing error loading recipe " + resourcelocation, (Throwable)jsonparseexception);
-                            return false;
-                        }
-                        catch (IOException ioexception)
-                        {
-                            field_192422_a.error("Couldn't read recipe " + resourcelocation + " from " + path1, (Throwable)ioexception);
-                            return false;
+                            boolean flag;
+
+                            try
+                            {
+                                bufferedreader = Files.newBufferedReader(path1);
+                                func_193379_a(s, func_193376_a((JsonObject)JsonUtils.func_193839_a(gson, bufferedreader, JsonObject.class)));
+                            }
+                            catch (JsonParseException jsonparseexception)
+                            {
+                                field_192422_a.error("Parsing error loading recipe " + resourcelocation, (Throwable)jsonparseexception);
+                                flag = false;
+                                return flag;
+                            }
+                            catch (IOException ioexception)
+                            {
+                                field_192422_a.error("Couldn't read recipe " + resourcelocation + " from " + path1, (Throwable)ioexception);
+                                flag = false;
+                                return flag;
+                            }
                         }
                         finally
                         {
@@ -132,16 +137,22 @@
 
                 return true;
             }
+
+            field_192422_a.error("Couldn't find .mcassetsroot");
+            flag1 = false;
         }
         catch (IOException | URISyntaxException urisyntaxexception)
         {
             field_192422_a.error("Couldn't get a list of all recipe files", (Throwable)urisyntaxexception);
-            return false;
+            flag1 = false;
+            return flag1;
         }
         finally
         {
-            IOUtils.closeQuietly(filesystem);
+            IOUtils.closeQuietly((Closeable)filesystem);
         }
+
+        return flag1;
     }
 
     private static IRecipe func_193376_a(JsonObject p_193376_0_)
@@ -162,12 +173,14 @@
         }
     }
 
-    public static void func_193379_a(String p_193379_0_, IRecipe p_193379_1_)
+    //Forge: Made private use GameData/Registry events!
+    private static void func_193379_a(String p_193379_0_, IRecipe p_193379_1_)
     {
         func_193372_a(new ResourceLocation(p_193379_0_), p_193379_1_);
     }
 
-    public static void func_193372_a(ResourceLocation p_193372_0_, IRecipe p_193372_1_)
+    //Forge: Made private use GameData/Registry events!
+    private static void func_193372_a(ResourceLocation p_193372_0_, IRecipe p_193372_1_)
     {
         if (field_193380_a.func_148741_d(p_193372_0_))
         {
@@ -216,9 +229,9 @@
             }
         }
 
-        NonNullList<ItemStack> nonnulllist = NonNullList.func_191197_a(p_180303_0_.func_70302_i_(), ItemStack.field_190927_a);
+        NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191197_a(p_180303_0_.func_70302_i_(), ItemStack.field_190927_a);
 
-        for (int i = 0; i < nonnulllist.size(); i++)
+        for (int i = 0; i < nonnulllist.size(); ++i)
         {
             nonnulllist.set(i, p_180303_0_.func_70301_a(i));
         }
@@ -232,11 +245,13 @@
         return field_193380_a.func_82594_a(p_193373_0_);
     }
 
+    @Deprecated //DO NOT USE THIS
     public static int func_193375_a(IRecipe p_193375_0_)
     {
         return field_193380_a.func_148757_b(p_193375_0_);
     }
 
+    @Deprecated //DO NOT USE THIS
     @Nullable
     public static IRecipe func_193374_a(int p_193374_0_)
     {
