--- before/net/minecraft/item/crafting/CraftingManager.java
+++ after/net/minecraft/item/crafting/CraftingManager.java
@@ -35,7 +35,7 @@
 {
     private static final Logger LOGGER = LogManager.getLogger();
     private static int nextAvailableId;
-    public static final RegistryNamespaced<ResourceLocation, IRecipe> REGISTRY = new RegistryNamespaced<>();
+    public static final RegistryNamespaced<ResourceLocation, IRecipe> REGISTRY = net.minecraftforge.registries.GameData.getWrapper(IRecipe.class);
 
     public static boolean init()
     {
@@ -162,12 +162,14 @@
         }
     }
 
-    public static void register(String name, IRecipe recipe)
+    //Forge: Made private use GameData/Registry events!
+    private static void register(String name, IRecipe recipe)
     {
         register(new ResourceLocation(name), recipe);
     }
 
-    public static void register(ResourceLocation name, IRecipe recipe)
+    //Forge: Made private use GameData/Registry events!
+    private static void register(ResourceLocation name, IRecipe recipe)
     {
         if (REGISTRY.containsKey(name))
         {
@@ -232,11 +234,13 @@
         return REGISTRY.getObject(name);
     }
 
+    @Deprecated //DO NOT USE THIS
     public static int getIDForRecipe(IRecipe recipe)
     {
         return REGISTRY.getIDForObject(recipe);
     }
 
+    @Deprecated //DO NOT USE THIS
     @Nullable
     public static IRecipe getRecipeById(int id)
     {
