--- before/net/minecraft/item/crafting/RecipesArmorDyes.java
+++ after/net/minecraft/item/crafting/RecipesArmorDyes.java
@@ -9,10 +9,8 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipesArmorDyes implements IRecipe
+public class RecipesArmorDyes extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
 {
     @Override
     public boolean matches(InventoryCrafting inv, World worldIn)
@@ -39,7 +37,7 @@
                 }
                 else
                 {
-                    if (itemstack1.getItem() != Items.DYE)
+                    if (!net.minecraftforge.oredict.DyeUtils.isDye(itemstack1))
                     {
                         return false;
                     }
@@ -94,12 +92,12 @@
                 }
                 else
                 {
-                    if (itemstack1.getItem() != Items.DYE)
+                    if (!net.minecraftforge.oredict.DyeUtils.isDye(itemstack1))
                     {
                         return ItemStack.EMPTY;
                     }
 
-                    float[] afloat = EnumDyeColor.byDyeDamage(itemstack1.getMetadata()).getColorComponentValues();
+                    float[] afloat = net.minecraftforge.oredict.DyeUtils.colorFromStack(itemstack1).get().getColorComponentValues();
                     int l1 = (int)(afloat[0] * 255.0F);
                     int i2 = (int)(afloat[1] * 255.0F);
                     int j2 = (int)(afloat[2] * 255.0F);
@@ -147,11 +145,7 @@
         for (int i = 0; i < nonnulllist.size(); i++)
         {
             ItemStack itemstack = inv.getStackInSlot(i);
-
-            if (itemstack.getItem().hasContainerItem())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.getItem().getContainerItem()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
@@ -163,7 +157,6 @@
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
     @Override
     public boolean canFit(int width, int height)
     {
