--- before/net/minecraft/item/crafting/RecipeBookCloning.java
+++ after/net/minecraft/item/crafting/RecipeBookCloning.java
@@ -6,10 +6,8 @@
 import net.minecraft.item.ItemWrittenBook;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipeBookCloning implements IRecipe
+public class RecipeBookCloning extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
 {
     @Override
     public boolean matches(InventoryCrafting inv, World worldIn)
@@ -132,7 +130,6 @@
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
     @Override
     public boolean canFit(int width, int height)
     {
