--- before/net/minecraft/item/crafting/RecipesArmorDyes.java
+++ after/net/minecraft/item/crafting/RecipesArmorDyes.java
@@ -9,16 +9,13 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipesArmorDyes implements IRecipe
+public class RecipesArmorDyes extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
 {
-    @Override
     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
     {
         ItemStack itemstack = ItemStack.field_190927_a;
-        List<ItemStack> list = Lists.newArrayList();
+        List<ItemStack> list = Lists.<ItemStack>newArrayList();
 
         for (int i = 0; i < p_77569_1_.func_70302_i_(); ++i)
         {
@@ -39,7 +36,7 @@
                 }
                 else
                 {
-                    if (itemstack1.func_77973_b() != Items.field_151100_aR)
+                    if (!net.minecraftforge.oredict.DyeUtils.isDye(itemstack1))
                     {
                         return false;
                     }
@@ -52,7 +49,6 @@
         return !itemstack.func_190926_b() && !list.isEmpty();
     }
 
-    @Override
     public ItemStack func_77572_b(InventoryCrafting p_77572_1_)
     {
         ItemStack itemstack = ItemStack.field_190927_a;
@@ -82,9 +78,9 @@
                     if (itemarmor.func_82816_b_(itemstack1))
                     {
                         int l = itemarmor.func_82814_b(itemstack);
-                        float f = (float)(l >> 16 & 0xFF) / 255.0F;
-                        float f1 = (float)(l >> 8 & 0xFF) / 255.0F;
-                        float f2 = (float)(l & 0xFF) / 255.0F;
+                        float f = (float)(l >> 16 & 255) / 255.0F;
+                        float f1 = (float)(l >> 8 & 255) / 255.0F;
+                        float f2 = (float)(l & 255) / 255.0F;
                         i = (int)((float)i + Math.max(f, Math.max(f1, f2)) * 255.0F);
                         aint[0] = (int)((float)aint[0] + f * 255.0F);
                         aint[1] = (int)((float)aint[1] + f1 * 255.0F);
@@ -94,12 +90,12 @@
                 }
                 else
                 {
-                    if (itemstack1.func_77973_b() != Items.field_151100_aR)
+                    if (!net.minecraftforge.oredict.DyeUtils.isDye(itemstack1))
                     {
                         return ItemStack.field_190927_a;
                     }
 
-                    float[] afloat = EnumDyeColor.func_176766_a(itemstack1.func_77960_j()).func_193349_f();
+                    float[] afloat = net.minecraftforge.oredict.DyeUtils.colorFromStack(itemstack1).get().func_193349_f();
                     int l1 = (int)(afloat[0] * 255.0F);
                     int i2 = (int)(afloat[1] * 255.0F);
                     int j2 = (int)(afloat[2] * 255.0F);
@@ -133,38 +129,29 @@
         }
     }
 
-    @Override
     public ItemStack func_77571_b()
     {
         return ItemStack.field_190927_a;
     }
 
-    @Override
     public NonNullList<ItemStack> func_179532_b(InventoryCrafting p_179532_1_)
     {
-        NonNullList<ItemStack> nonnulllist = NonNullList.func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
+        NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
 
         for (int i = 0; i < nonnulllist.size(); ++i)
         {
             ItemStack itemstack = p_179532_1_.func_70301_a(i);
-
-            if (itemstack.func_77973_b().func_77634_r())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
     }
 
-    @Override
     public boolean func_192399_d()
     {
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_194133_a(int p_194133_1_, int p_194133_2_)
     {
         return p_194133_1_ * p_194133_2_ >= 2;
