--- before/net/minecraft/item/crafting/RecipeRepairItem.java
+++ after/net/minecraft/item/crafting/RecipeRepairItem.java
@@ -7,10 +7,8 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipeRepairItem implements IRecipe
+public class RecipeRepairItem extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
 {
     @Override
     public boolean matches(InventoryCrafting inv, World worldIn)
@@ -32,7 +30,7 @@
                     if (itemstack.getItem() != itemstack1.getItem()
                             || itemstack1.getCount() != 1
                             || itemstack.getCount() != 1
-                            || !itemstack1.getItem().isDamageable())
+                            || !itemstack1.getItem().isRepairable())
                     {
                         return false;
                     }
@@ -63,7 +61,7 @@
                     if (itemstack.getItem() != itemstack1.getItem()
                             || itemstack1.getCount() != 1
                             || itemstack.getCount() != 1
-                            || !itemstack1.getItem().isDamageable())
+                            || !itemstack1.getItem().isRepairable())
                     {
                         return ItemStack.EMPTY;
                     }
@@ -79,13 +77,13 @@
             if (itemstack2.getItem() == itemstack3.getItem()
                     && itemstack2.getCount() == 1
                     && itemstack3.getCount() == 1
-                    && itemstack2.getItem().isDamageable())
+                    && itemstack2.getItem().isRepairable())
             {
-                Item item = itemstack2.getItem();
-                int j = item.getMaxDamage() - itemstack2.getItemDamage();
-                int k = item.getMaxDamage() - itemstack3.getItemDamage();
-                int l = j + k + item.getMaxDamage() * 5 / 100;
-                int i1 = item.getMaxDamage() - l;
+                // FORGE: Make itemstack sensitive // Item item = itemstack2.getItem();
+                int j = itemstack2.getMaxDamage() - itemstack2.getItemDamage();
+                int k = itemstack2.getMaxDamage() - itemstack3.getItemDamage();
+                int l = j + k + itemstack2.getMaxDamage() * 5 / 100;
+                int i1 = itemstack2.getMaxDamage() - l;
 
                 if (i1 < 0)
                 {
@@ -113,11 +111,7 @@
         for (int i = 0; i < nonnulllist.size(); i++)
         {
             ItemStack itemstack = inv.getStackInSlot(i);
-
-            if (itemstack.getItem().hasContainerItem())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.getItem().getContainerItem()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
@@ -129,7 +123,6 @@
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
     @Override
     public boolean canFit(int width, int height)
     {
