--- before/net/minecraft/item/crafting/RecipeRepairItem.java
+++ after/net/minecraft/item/crafting/RecipeRepairItem.java
@@ -7,15 +7,12 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipeRepairItem implements IRecipe
+public class RecipeRepairItem extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
 {
-    @Override
     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
     {
-        List<ItemStack> list = Lists.newArrayList();
+        List<ItemStack> list = Lists.<ItemStack>newArrayList();
 
         for (int i = 0; i < p_77569_1_.func_70302_i_(); ++i)
         {
@@ -29,10 +26,7 @@
                 {
                     ItemStack itemstack1 = list.get(0);
 
-                    if (itemstack.func_77973_b() != itemstack1.func_77973_b()
-                            || itemstack1.func_190916_E() != 1
-                            || itemstack.func_190916_E() != 1
-                            || !itemstack1.func_77973_b().func_77645_m())
+                    if (itemstack.func_77973_b() != itemstack1.func_77973_b() || itemstack1.func_190916_E() != 1 || itemstack.func_190916_E() != 1 || !itemstack1.func_77973_b().isRepairable())
                     {
                         return false;
                     }
@@ -43,10 +37,9 @@
         return list.size() == 2;
     }
 
-    @Override
     public ItemStack func_77572_b(InventoryCrafting p_77572_1_)
     {
-        List<ItemStack> list = Lists.newArrayList();
+        List<ItemStack> list = Lists.<ItemStack>newArrayList();
 
         for (int i = 0; i < p_77572_1_.func_70302_i_(); ++i)
         {
@@ -60,10 +53,7 @@
                 {
                     ItemStack itemstack1 = list.get(0);
 
-                    if (itemstack.func_77973_b() != itemstack1.func_77973_b()
-                            || itemstack1.func_190916_E() != 1
-                            || itemstack.func_190916_E() != 1
-                            || !itemstack1.func_77973_b().func_77645_m())
+                    if (itemstack.func_77973_b() != itemstack1.func_77973_b() || itemstack1.func_190916_E() != 1 || itemstack.func_190916_E() != 1 || !itemstack1.func_77973_b().isRepairable())
                     {
                         return ItemStack.field_190927_a;
                     }
@@ -76,16 +66,13 @@
             ItemStack itemstack2 = list.get(0);
             ItemStack itemstack3 = list.get(1);
 
-            if (itemstack2.func_77973_b() == itemstack3.func_77973_b()
-                    && itemstack2.func_190916_E() == 1
-                    && itemstack3.func_190916_E() == 1
-                    && itemstack2.func_77973_b().func_77645_m())
+            if (itemstack2.func_77973_b() == itemstack3.func_77973_b() && itemstack2.func_190916_E() == 1 && itemstack3.func_190916_E() == 1 && itemstack2.func_77973_b().isRepairable())
             {
-                Item item = itemstack2.func_77973_b();
-                int j = item.func_77612_l() - itemstack2.func_77952_i();
-                int k = item.func_77612_l() - itemstack3.func_77952_i();
-                int l = j + k + item.func_77612_l() * 5 / 100;
-                int i1 = item.func_77612_l() - l;
+                // FORGE: Make itemstack sensitive // Item item = itemstack2.getItem();
+                int j = itemstack2.func_77958_k() - itemstack2.func_77952_i();
+                int k = itemstack2.func_77958_k() - itemstack3.func_77952_i();
+                int l = j + k + itemstack2.func_77958_k() * 5 / 100;
+                int i1 = itemstack2.func_77958_k() - l;
 
                 if (i1 < 0)
                 {
@@ -99,38 +86,29 @@
         return ItemStack.field_190927_a;
     }
 
-    @Override
     public ItemStack func_77571_b()
     {
         return ItemStack.field_190927_a;
     }
 
-    @Override
     public NonNullList<ItemStack> func_179532_b(InventoryCrafting p_179532_1_)
     {
-        NonNullList<ItemStack> nonnulllist = NonNullList.func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
+        NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
 
         for (int i = 0; i < nonnulllist.size(); ++i)
         {
             ItemStack itemstack = p_179532_1_.func_70301_a(i);
-
-            if (itemstack.func_77973_b().func_77634_r())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
     }
 
-    @Override
     public boolean func_192399_d()
     {
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_194133_a(int p_194133_1_, int p_194133_2_)
     {
         return p_194133_1_ * p_194133_2_ >= 2;
