--- before/net/minecraft/item/crafting/RecipeBookCloning.java
+++ after/net/minecraft/item/crafting/RecipeBookCloning.java
@@ -6,18 +6,15 @@
 import net.minecraft.item.ItemWrittenBook;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipeBookCloning implements IRecipe
+public class RecipeBookCloning extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
 {
-    @Override
     public boolean matches(InventoryCrafting inv, World worldIn)
     {
         int i = 0;
         ItemStack itemstack = ItemStack.EMPTY;
 
-        for (int j = 0; j < inv.getSizeInventory(); j++)
+        for (int j = 0; j < inv.getSizeInventory(); ++j)
         {
             ItemStack itemstack1 = inv.getStackInSlot(j);
 
@@ -39,7 +36,7 @@
                         return false;
                     }
 
-                    i++;
+                    ++i;
                 }
             }
         }
@@ -47,13 +44,12 @@
         return !itemstack.isEmpty() && itemstack.hasTagCompound() && i > 0;
     }
 
-    @Override
     public ItemStack getCraftingResult(InventoryCrafting inv)
     {
         int i = 0;
         ItemStack itemstack = ItemStack.EMPTY;
 
-        for (int j = 0; j < inv.getSizeInventory(); j++)
+        for (int j = 0; j < inv.getSizeInventory(); ++j)
         {
             ItemStack itemstack1 = inv.getStackInSlot(j);
 
@@ -75,7 +71,7 @@
                         return ItemStack.EMPTY;
                     }
 
-                    i++;
+                    ++i;
                 }
             }
         }
@@ -99,18 +95,16 @@
         }
     }
 
-    @Override
     public ItemStack getRecipeOutput()
     {
         return ItemStack.EMPTY;
     }
 
-    @Override
     public NonNullList<ItemStack> getRemainingItems(InventoryCrafting inv)
     {
-        NonNullList<ItemStack> nonnulllist = NonNullList.withSize(inv.getSizeInventory(), ItemStack.EMPTY);
+        NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>withSize(inv.getSizeInventory(), ItemStack.EMPTY);
 
-        for (int i = 0; i < nonnulllist.size(); i++)
+        for (int i = 0; i < nonnulllist.size(); ++i)
         {
             ItemStack itemstack = inv.getStackInSlot(i);
 
@@ -126,14 +120,11 @@
         return nonnulllist;
     }
 
-    @Override
     public boolean isDynamic()
     {
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
     public boolean canFit(int width, int height)
     {
         return width >= 3 && height >= 3;
