--- before/net/minecraft/item/crafting/RecipeTippedArrow.java
+++ after/net/minecraft/item/crafting/RecipeTippedArrow.java
@@ -7,19 +7,16 @@
 import net.minecraft.potion.PotionUtils;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipeTippedArrow implements IRecipe
+public class RecipeTippedArrow extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
 {
-    @Override
     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
     {
         if (p_77569_1_.func_174922_i() == 3 && p_77569_1_.func_174923_h() == 3)
         {
-            for (int i = 0; i < p_77569_1_.func_174922_i(); i++)
+            for (int i = 0; i < p_77569_1_.func_174922_i(); ++i)
             {
-                for (int j = 0; j < p_77569_1_.func_174923_h(); j++)
+                for (int j = 0; j < p_77569_1_.func_174923_h(); ++j)
                 {
                     ItemStack itemstack = p_77569_1_.func_70463_b(i, j);
 
@@ -52,7 +49,6 @@
         }
     }
 
-    @Override
     public ItemStack func_77572_b(InventoryCrafting p_77572_1_)
     {
         ItemStack itemstack = p_77572_1_.func_70463_b(1, 1);
@@ -70,26 +66,21 @@
         }
     }
 
-    @Override
     public ItemStack func_77571_b()
     {
         return ItemStack.field_190927_a;
     }
 
-    @Override
     public NonNullList<ItemStack> func_179532_b(InventoryCrafting p_179532_1_)
     {
-        return NonNullList.func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
+        return NonNullList.<ItemStack>func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
     }
 
-    @Override
     public boolean func_192399_d()
     {
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_194133_a(int p_194133_1_, int p_194133_2_)
     {
         return p_194133_1_ >= 2 && p_194133_2_ >= 2;
