--- before/net/minecraft/item/crafting/ShapedRecipes.java
+++ after/net/minecraft/item/crafting/ShapedRecipes.java
@@ -19,10 +19,8 @@
 import net.minecraft.util.NonNullList;
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class ShapedRecipes implements IRecipe
+public class ShapedRecipes extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements net.minecraftforge.common.crafting.IShapedRecipe
 {
     public final int recipeWidth;
     public final int recipeHeight;
@@ -39,7 +37,6 @@
         this.recipeOutput = result;
     }
 
-    @SideOnly(Side.CLIENT)
     @Override
     public String getGroup()
     {
@@ -61,10 +58,7 @@
         {
             ItemStack itemstack = inv.getStackInSlot(i);
 
-            if (itemstack.getItem().hasContainerItem())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.getItem().getContainerItem()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
@@ -76,7 +70,6 @@
         return this.recipeItems;
     }
 
-    @SideOnly(Side.CLIENT)
     @Override
     public boolean canFit(int width, int height)
     {
@@ -86,9 +79,9 @@
     @Override
     public boolean matches(InventoryCrafting inv, World worldIn)
     {
-        for (int i = 0; i <= 3 - this.recipeWidth; i++)
+        for (int i = 0; i <= inv.getWidth() - this.recipeWidth; ++i)
         {
-            for (int j = 0; j <= 3 - this.recipeHeight; j++)
+            for (int j = 0; j <= inv.getHeight() - this.recipeHeight; ++j)
             {
                 if (this.checkMatch(inv, i, j, true))
                 {
@@ -107,9 +100,9 @@
 
     private boolean checkMatch(InventoryCrafting craftingInventory, int p_77573_2_, int p_77573_3_, boolean p_77573_4_)
     {
-        for (int i = 0; i < 3; i++)
+        for (int i = 0; i < craftingInventory.getWidth(); ++i)
         {
-            for (int j = 0; j < 3; j++)
+            for (int j = 0; j < craftingInventory.getHeight(); ++j)
             {
                 int k = i - p_77573_2_;
                 int l = j - p_77573_3_;
@@ -382,5 +375,17 @@
             int j = useCount ? JsonUtils.getInt(json, "count", 1) : 1;
             return new ItemStack(item, j, i);
         }
+    }
+
+    //================================================ FORGE START ================================================
+    @Override
+    public int getRecipeWidth()
+    {
+        return this.getWidth();
+    }
+    @Override
+    public int getRecipeHeight()
+    {
+        return this.getHeight();
     }
 }
