--- before/net/minecraft/item/crafting/ShapedRecipes.java
+++ after/net/minecraft/item/crafting/ShapedRecipes.java
@@ -19,10 +19,8 @@
 import net.minecraft.util.NonNullList;
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class ShapedRecipes implements IRecipe
+public class ShapedRecipes extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements net.minecraftforge.common.crafting.IShapedRecipe
 {
     public final int field_77576_b;
     public final int field_77577_c;
@@ -39,56 +37,45 @@
         this.field_77575_e = p_i47501_5_;
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
     public String func_193358_e()
     {
         return this.field_194137_e;
     }
 
-    @Override
     public ItemStack func_77571_b()
     {
         return this.field_77575_e;
     }
 
-    @Override
     public NonNullList<ItemStack> func_179532_b(InventoryCrafting p_179532_1_)
     {
-        NonNullList<ItemStack> nonnulllist = NonNullList.func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
+        NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
 
-        for (int i = 0; i < nonnulllist.size(); i++)
+        for (int i = 0; i < nonnulllist.size(); ++i)
         {
             ItemStack itemstack = p_179532_1_.func_70301_a(i);
 
-            if (itemstack.func_77973_b().func_77634_r())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
     }
 
-    @Override
     public NonNullList<Ingredient> func_192400_c()
     {
         return this.field_77574_d;
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_194133_a(int p_194133_1_, int p_194133_2_)
     {
         return p_194133_1_ >= this.field_77576_b && p_194133_2_ >= this.field_77577_c;
     }
 
-    @Override
     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
     {
-        for (int i = 0; i <= 3 - this.field_77576_b; i++)
+        for (int i = 0; i <= p_77569_1_.func_174922_i() - this.field_77576_b; ++i)
         {
-            for (int j = 0; j <= 3 - this.field_77577_c; j++)
+            for (int j = 0; j <= p_77569_1_.func_174923_h() - this.field_77577_c; ++j)
             {
                 if (this.func_77573_a(p_77569_1_, i, j, true))
                 {
@@ -107,9 +94,9 @@
 
     private boolean func_77573_a(InventoryCrafting p_77573_1_, int p_77573_2_, int p_77573_3_, boolean p_77573_4_)
     {
-        for (int i = 0; i < 3; i++)
+        for (int i = 0; i < p_77573_1_.func_174922_i(); ++i)
         {
-            for (int j = 0; j < 3; j++)
+            for (int j = 0; j < p_77573_1_.func_174923_h(); ++j)
             {
                 int k = i - p_77573_2_;
                 int l = j - p_77573_3_;
@@ -137,7 +124,6 @@
         return true;
     }
 
-    @Override
     public ItemStack func_77572_b(InventoryCrafting p_77572_1_)
     {
         return this.func_77571_b().func_77946_l();
@@ -167,13 +153,13 @@
 
     private static NonNullList<Ingredient> func_192402_a(String[] p_192402_0_, Map<String, Ingredient> p_192402_1_, int p_192402_2_, int p_192402_3_)
     {
-        NonNullList<Ingredient> nonnulllist = NonNullList.func_191197_a(p_192402_2_ * p_192402_3_, Ingredient.field_193370_a);
+        NonNullList<Ingredient> nonnulllist = NonNullList.<Ingredient>func_191197_a(p_192402_2_ * p_192402_3_, Ingredient.field_193370_a);
         Set<String> set = Sets.newHashSet(p_192402_1_.keySet());
         set.remove(" ");
 
-        for (int i = 0; i < p_192402_0_.length; i++)
+        for (int i = 0; i < p_192402_0_.length; ++i)
         {
-            for (int j = 0; j < p_192402_0_[i].length(); j++)
+            for (int j = 0; j < p_192402_0_[i].length(); ++j)
             {
                 String s = p_192402_0_[i].substring(j, j + 1);
                 Ingredient ingredient = p_192402_1_.get(s);
@@ -206,7 +192,7 @@
         int k = 0;
         int l = 0;
 
-        for (int i1 = 0; i1 < p_194134_0_.length; i1++)
+        for (int i1 = 0; i1 < p_194134_0_.length; ++i1)
         {
             String s = p_194134_0_[i1];
             i = Math.min(i, func_194135_a(s));
@@ -217,10 +203,10 @@
             {
                 if (k == i1)
                 {
-                    k++;
+                    ++k;
                 }
 
-                l++;
+                ++l;
             }
             else
             {
@@ -236,7 +222,7 @@
         {
             String[] astring = new String[p_194134_0_.length - l - k];
 
-            for (int k1 = 0; k1 < astring.length; k1++)
+            for (int k1 = 0; k1 < astring.length; ++k1)
             {
                 astring[k1] = p_194134_0_[k1 + k].substring(i, j + 1);
             }
@@ -247,11 +233,11 @@
 
     private static int func_194135_a(String p_194135_0_)
     {
-        int i = 0;
+        int i;
 
-        while (i < p_194135_0_.length() && p_194135_0_.charAt(i) == ' ')
+        for (i = 0; i < p_194135_0_.length() && p_194135_0_.charAt(i) == ' '; ++i)
         {
-            i++;
+            ;
         }
 
         return i;
@@ -259,11 +245,11 @@
 
     private static int func_194136_b(String p_194136_0_)
     {
-        int i = p_194136_0_.length() - 1;
+        int i;
 
-        while (i >= 0 && p_194136_0_.charAt(i) == ' ')
+        for (i = p_194136_0_.length() - 1; i >= 0 && p_194136_0_.charAt(i) == ' '; --i)
         {
-            i--;
+            ;
         }
 
         return i;
@@ -283,7 +269,7 @@
         }
         else
         {
-            for (int i = 0; i < astring.length; i++)
+            for (int i = 0; i < astring.length; ++i)
             {
                 String s = JsonUtils.func_151206_a(p_192407_0_.get(i), "pattern[" + i + "]");
 
@@ -306,13 +292,13 @@
 
     private static Map<String, Ingredient> func_192408_a(JsonObject p_192408_0_)
     {
-        Map<String, Ingredient> map = Maps.newHashMap();
+        Map<String, Ingredient> map = Maps.<String, Ingredient>newHashMap();
 
         for (Entry<String, JsonElement> entry : p_192408_0_.entrySet())
         {
-            if (entry.getKey().length() != 1)
+            if (((String)entry.getKey()).length() != 1)
             {
-                throw new JsonSyntaxException("Invalid key entry: '" + entry.getKey() + "' is an invalid symbol (must be 1 character only).");
+                throw new JsonSyntaxException("Invalid key entry: '" + (String)entry.getKey() + "' is an invalid symbol (must be 1 character only).");
             }
 
             if (" ".equals(entry.getKey()))
@@ -329,38 +315,41 @@
 
     public static Ingredient func_193361_a(@Nullable JsonElement p_193361_0_)
     {
-        if (p_193361_0_ == null || p_193361_0_.isJsonNull())
+        if (p_193361_0_ != null && !p_193361_0_.isJsonNull())
+        {
+            if (p_193361_0_.isJsonObject())
+            {
+                return Ingredient.func_193369_a(func_192405_a(p_193361_0_.getAsJsonObject(), false));
+            }
+            else if (!p_193361_0_.isJsonArray())
+            {
+                throw new JsonSyntaxException("Expected item to be object or array of objects");
+            }
+            else
+            {
+                JsonArray jsonarray = p_193361_0_.getAsJsonArray();
+
+                if (jsonarray.size() == 0)
+                {
+                    throw new JsonSyntaxException("Item array cannot be empty, at least one item must be defined");
+                }
+                else
+                {
+                    ItemStack[] aitemstack = new ItemStack[jsonarray.size()];
+
+                    for (int i = 0; i < jsonarray.size(); ++i)
+                    {
+                        aitemstack[i] = func_192405_a(JsonUtils.func_151210_l(jsonarray.get(i), "item"), false);
+                    }
+
+                    return Ingredient.func_193369_a(aitemstack);
+                }
+            }
+        }
+        else
         {
             throw new JsonSyntaxException("Item cannot be null");
         }
-        else if (p_193361_0_.isJsonObject())
-        {
-            return Ingredient.func_193369_a(func_192405_a(p_193361_0_.getAsJsonObject(), false));
-        }
-        else if (!p_193361_0_.isJsonArray())
-        {
-            throw new JsonSyntaxException("Expected item to be object or array of objects");
-        }
-        else
-        {
-            JsonArray jsonarray = p_193361_0_.getAsJsonArray();
-
-            if (jsonarray.size() == 0)
-            {
-                throw new JsonSyntaxException("Item array cannot be empty, at least one item must be defined");
-            }
-            else
-            {
-                ItemStack[] aitemstack = new ItemStack[jsonarray.size()];
-
-                for (int i = 0; i < jsonarray.size(); i++)
-                {
-                    aitemstack[i] = func_192405_a(JsonUtils.func_151210_l(jsonarray.get(i), "item"), false);
-                }
-
-                return Ingredient.func_193369_a(aitemstack);
-            }
-        }
     }
 
     public static ItemStack func_192405_a(JsonObject p_192405_0_, boolean p_192405_1_)
@@ -382,5 +371,17 @@
             int j = p_192405_1_ ? JsonUtils.func_151208_a(p_192405_0_, "count", 1) : 1;
             return new ItemStack(item, j, i);
         }
+    }
+
+    //================================================ FORGE START ================================================
+    @Override
+    public int getRecipeWidth()
+    {
+        return this.func_192403_f();
+    }
+    @Override
+    public int getRecipeHeight()
+    {
+        return this.func_192404_g();
     }
 }
