--- before/net/minecraft/item/crafting/ShieldRecipes.java
+++ after/net/minecraft/item/crafting/ShieldRecipes.java
@@ -6,20 +6,17 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class ShieldRecipes
 {
-    public static class Decoration implements IRecipe
+    public static class Decoration extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
         {
-            @Override
             public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
             {
                 ItemStack itemstack = ItemStack.field_190927_a;
                 ItemStack itemstack1 = ItemStack.field_190927_a;
 
-                for (int i = 0; i < p_77569_1_.func_70302_i_(); i++)
+                for (int i = 0; i < p_77569_1_.func_70302_i_(); ++i)
                 {
                     ItemStack itemstack2 = p_77569_1_.func_70301_a(i);
 
@@ -56,16 +53,22 @@
                     }
                 }
 
-                return !itemstack.func_190926_b() && !itemstack1.func_190926_b();
+                if (!itemstack.func_190926_b() && !itemstack1.func_190926_b())
+                {
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
             }
 
-            @Override
             public ItemStack func_77572_b(InventoryCrafting p_77572_1_)
             {
                 ItemStack itemstack = ItemStack.field_190927_a;
                 ItemStack itemstack1 = ItemStack.field_190927_a;
 
-                for (int i = 0; i < p_77572_1_.func_70302_i_(); i++)
+                for (int i = 0; i < p_77572_1_.func_70302_i_(); ++i)
                 {
                     ItemStack itemstack2 = p_77572_1_.func_70301_a(i);
 
@@ -96,18 +99,16 @@
                 }
             }
 
-            @Override
             public ItemStack func_77571_b()
             {
                 return ItemStack.field_190927_a;
             }
 
-            @Override
             public NonNullList<ItemStack> func_179532_b(InventoryCrafting p_179532_1_)
             {
-                NonNullList<ItemStack> nonnulllist = NonNullList.func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
+                NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
 
-                for (int i = 0; i < nonnulllist.size(); i++)
+                for (int i = 0; i < nonnulllist.size(); ++i)
                 {
                     ItemStack itemstack = p_179532_1_.func_70301_a(i);
 
@@ -120,14 +121,11 @@
                 return nonnulllist;
             }
 
-            @Override
             public boolean func_192399_d()
             {
                 return true;
             }
 
-            @SideOnly(Side.CLIENT)
-            @Override
             public boolean func_194133_a(int p_194133_1_, int p_194133_2_)
             {
                 return p_194133_1_ * p_194133_2_ >= 2;
