--- before/net/minecraft/item/crafting/ShulkerBoxRecipes.java
+++ after/net/minecraft/item/crafting/ShulkerBoxRecipes.java
@@ -8,14 +8,11 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class ShulkerBoxRecipes
 {
-    public static class ShulkerBoxColoring implements IRecipe
+    public static class ShulkerBoxColoring extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
         {
-            @Override
             public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
             {
                 int i = 0;
@@ -33,7 +30,7 @@
                         }
                         else
                         {
-                            if (itemstack.func_77973_b() != Items.field_151100_aR)
+                            if (!net.minecraftforge.oredict.DyeUtils.isDye(itemstack))
                             {
                                 return false;
                             }
@@ -51,7 +48,6 @@
                 return i == 1 && j == 1;
             }
 
-            @Override
             public ItemStack func_77572_b(InventoryCrafting p_77572_1_)
             {
                 ItemStack itemstack = ItemStack.field_190927_a;
@@ -67,14 +63,14 @@
                         {
                             itemstack = itemstack2;
                         }
-                        else if (itemstack2.func_77973_b() == Items.field_151100_aR)
+                        else if (net.minecraftforge.oredict.DyeUtils.isDye(itemstack2))
                         {
                             itemstack1 = itemstack2;
                         }
                     }
                 }
 
-                ItemStack itemstack3 = BlockShulkerBox.func_190953_b(EnumDyeColor.func_176766_a(itemstack1.func_77960_j()));
+                ItemStack itemstack3 = BlockShulkerBox.func_190953_b(net.minecraftforge.oredict.DyeUtils.colorFromStack(itemstack1).get());
 
                 if (itemstack.func_77942_o())
                 {
@@ -84,16 +80,14 @@
                 return itemstack3;
             }
 
-            @Override
             public ItemStack func_77571_b()
             {
                 return ItemStack.field_190927_a;
             }
 
-            @Override
             public NonNullList<ItemStack> func_179532_b(InventoryCrafting p_179532_1_)
             {
-                NonNullList<ItemStack> nonnulllist = NonNullList.func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
+                NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
 
                 for (int i = 0; i < nonnulllist.size(); ++i)
                 {
@@ -108,14 +102,11 @@
                 return nonnulllist;
             }
 
-            @Override
             public boolean func_192399_d()
             {
                 return true;
             }
 
-            @SideOnly(Side.CLIENT)
-            @Override
             public boolean func_194133_a(int p_194133_1_, int p_194133_2_)
             {
                 return p_194133_1_ * p_194133_2_ >= 2;
