--- before/net/minecraft/client/renderer/debug/DebugRendererWater.java
+++ after/net/minecraft/client/renderer/debug/DebugRendererWater.java
@@ -58,8 +58,8 @@
 
             if (iblockstate1.func_177230_c() == Blocks.field_150355_j || iblockstate1.func_177230_c() == Blocks.field_150358_i)
             {
-                Integer integer = iblockstate1.func_177229_b(BlockLiquid.field_176367_b);
-                double d1 = integer > 7 ? 0.9D : 1.0D - 0.11D * (double)integer.intValue();
+                Integer integer = (Integer)iblockstate1.func_177229_b(BlockLiquid.field_176367_b);
+                double d1 = integer.intValue() > 7 ? 0.9D : 1.0D - 0.11D * (double)integer.intValue();
                 String s = iblockstate1.func_177230_c() == Blocks.field_150358_i ? "f" : "s";
                 DebugRenderer.func_190076_a(s + " " + integer, (double)blockpos2.func_177958_n() + 0.5D, (double)blockpos2.func_177956_o() + d1, (double)blockpos2.func_177952_p() + 0.5D, p_190060_1_, -16777216);
             }
