--- before/net/minecraft/client/renderer/debug/DebugRendererPathfinding.java
+++ after/net/minecraft/client/renderer/debug/DebugRendererPathfinding.java
@@ -20,9 +20,9 @@
 public class DebugRendererPathfinding implements DebugRenderer.IDebugRenderer
 {
     private final Minecraft field_188290_a;
-    private final Map<Integer, Path> field_188291_b = Maps.newHashMap();
-    private final Map<Integer, Float> field_188292_c = Maps.newHashMap();
-    private final Map<Integer, Long> field_188293_d = Maps.newHashMap();
+    private final Map<Integer, Path> field_188291_b = Maps.<Integer, Path>newHashMap();
+    private final Map<Integer, Float> field_188292_c = Maps.<Integer, Float>newHashMap();
+    private final Map<Integer, Long> field_188293_d = Maps.<Integer, Long>newHashMap();
     private EntityPlayer field_190068_e;
     private double field_190069_f;
     private double field_190070_g;
@@ -35,9 +35,9 @@
 
     public void func_188289_a(int p_188289_1_, Path p_188289_2_, float p_188289_3_)
     {
-        this.field_188291_b.put(p_188289_1_, p_188289_2_);
-        this.field_188293_d.put(p_188289_1_, System.currentTimeMillis());
-        this.field_188292_c.put(p_188289_1_, p_188289_3_);
+        this.field_188291_b.put(Integer.valueOf(p_188289_1_), p_188289_2_);
+        this.field_188293_d.put(Integer.valueOf(p_188289_1_), Long.valueOf(System.currentTimeMillis()));
+        this.field_188292_c.put(Integer.valueOf(p_188289_1_), Float.valueOf(p_188289_3_));
     }
 
     public void func_190060_a(float p_190060_1_, long p_190060_2_)
@@ -59,11 +59,11 @@
             for (Integer integer : this.field_188291_b.keySet())
             {
                 Path path = this.field_188291_b.get(integer);
-                float f = this.field_188292_c.get(integer);
+                float f = ((Float)this.field_188292_c.get(integer)).floatValue();
                 this.func_190067_a(p_190060_1_, path);
                 PathPoint pathpoint = path.func_189964_i();
 
-                if (!(this.func_190066_a(pathpoint) > 40.0F))
+                if (this.func_190066_a(pathpoint) <= 40.0F)
                 {
                     RenderGlobal.func_189696_b((new AxisAlignedBB((double)((float)pathpoint.field_75839_a + 0.25F), (double)((float)pathpoint.field_75837_b + 0.25F), (double)pathpoint.field_75838_c + 0.25D, (double)((float)pathpoint.field_75839_a + 0.75F), (double)((float)pathpoint.field_75837_b + 0.75F), (double)((float)pathpoint.field_75838_c + 0.75F))).func_72317_d(-this.field_190069_f, -this.field_190070_g, -this.field_190071_h), 0.0F, 1.0F, 0.0F, 0.5F);
 
@@ -71,7 +71,7 @@
                     {
                         PathPoint pathpoint1 = path.func_75877_a(j);
 
-                        if (!(this.func_190066_a(pathpoint1) > 40.0F))
+                        if (this.func_190066_a(pathpoint1) <= 40.0F)
                         {
                             float f1 = j == path.func_75873_e() ? 1.0F : 0.0F;
                             float f2 = j == path.func_75873_e() ? 0.0F : 1.0F;
@@ -87,7 +87,7 @@
 
                 for (PathPoint pathpoint3 : path1.func_189965_h())
                 {
-                    if (!(this.func_190066_a(pathpoint3) > 40.0F))
+                    if (this.func_190066_a(pathpoint3) <= 40.0F)
                     {
                         DebugRenderer.func_190076_a(String.format("%s", pathpoint3.field_186287_m), (double)pathpoint3.field_75839_a + 0.5D, (double)pathpoint3.field_75837_b + 0.75D, (double)pathpoint3.field_75838_c + 0.5D, p_190060_1_, -65536);
                         DebugRenderer.func_190076_a(String.format("%.2f", pathpoint3.field_186286_l), (double)pathpoint3.field_75839_a + 0.5D, (double)pathpoint3.field_75837_b + 0.25D, (double)pathpoint3.field_75838_c + 0.5D, p_190060_1_, -65536);
@@ -96,7 +96,7 @@
 
                 for (PathPoint pathpoint4 : path1.func_189966_g())
                 {
-                    if (!(this.func_190066_a(pathpoint4) > 40.0F))
+                    if (this.func_190066_a(pathpoint4) <= 40.0F)
                     {
                         DebugRenderer.func_190076_a(String.format("%s", pathpoint4.field_186287_m), (double)pathpoint4.field_75839_a + 0.5D, (double)pathpoint4.field_75837_b + 0.75D, (double)pathpoint4.field_75838_c + 0.5D, p_190060_1_, -16776961);
                         DebugRenderer.func_190076_a(String.format("%.2f", pathpoint4.field_186286_l), (double)pathpoint4.field_75839_a + 0.5D, (double)pathpoint4.field_75837_b + 0.25D, (double)pathpoint4.field_75838_c + 0.5D, p_190060_1_, -16776961);
@@ -107,7 +107,7 @@
                 {
                     PathPoint pathpoint2 = path1.func_75877_a(k);
 
-                    if (!(this.func_190066_a(pathpoint2) > 40.0F))
+                    if (this.func_190066_a(pathpoint2) <= 40.0F)
                     {
                         DebugRenderer.func_190076_a(String.format("%s", pathpoint2.field_186287_m), (double)pathpoint2.field_75839_a + 0.5D, (double)pathpoint2.field_75837_b + 0.75D, (double)pathpoint2.field_75838_c + 0.5D, p_190060_1_, -1);
                         DebugRenderer.func_190076_a(String.format("%.2f", pathpoint2.field_186286_l), (double)pathpoint2.field_75839_a + 0.5D, (double)pathpoint2.field_75837_b + 0.25D, (double)pathpoint2.field_75838_c + 0.5D, p_190060_1_, -1);
@@ -115,9 +115,9 @@
                 }
             }
 
-            for (Integer integer2 : this.field_188293_d.keySet().toArray(new Integer[0]))
+            for (Integer integer2 : (Integer[])this.field_188293_d.keySet().toArray(new Integer[0]))
             {
-                if (i - this.field_188293_d.get(integer2) > 20000L)
+                if (i - ((Long)this.field_188293_d.get(integer2)).longValue() > 20000L)
                 {
                     this.field_188291_b.remove(integer2);
                     this.field_188293_d.remove(integer2);
@@ -140,7 +140,7 @@
         {
             PathPoint pathpoint = p_190067_2_.func_75877_a(i);
 
-            if (!(this.func_190066_a(pathpoint) > 40.0F))
+            if (this.func_190066_a(pathpoint) <= 40.0F)
             {
                 float f = (float)i / (float)p_190067_2_.func_75874_d() * 0.33F;
                 int j = i == 0 ? 0 : MathHelper.func_181758_c(f, 0.9F, 0.9F);
