--- before/net/minecraft/server/network/NetHandlerLoginServer.java
+++ after/net/minecraft/server/network/NetHandlerLoginServer.java
@@ -56,7 +56,6 @@
         field_147329_d.nextBytes(this.field_147330_e);
     }
 
-    @Override
     public void func_73660_a()
     {
         if (this.field_147328_g == NetHandlerLoginServer.LoginState.READY_TO_ACCEPT)
@@ -70,14 +69,14 @@
             if (entityplayermp == null)
             {
                 this.field_147328_g = NetHandlerLoginServer.LoginState.READY_TO_ACCEPT;
-                this.field_147327_f.func_184103_al().func_72355_a(this.field_147333_a, this.field_181025_l);
+                net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.fmlServerHandshake(this.field_147327_f.func_184103_al(), this.field_147333_a, this.field_181025_l);
                 this.field_181025_l = null;
             }
         }
 
         if (this.field_147336_h++ == 600)
         {
-            this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.slow_login"));
+            this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.slow_login", new Object[0]));
         }
     }
 
@@ -106,7 +105,7 @@
 
         if (s != null)
         {
-            this.func_194026_b(new TextComponentTranslation(s));
+            this.func_194026_b(new TextComponentTranslation(s, new Object[0]));
         }
         else
         {
@@ -133,12 +132,11 @@
             }
             else
             {
-                this.field_147327_f.func_184103_al().func_72355_a(this.field_147333_a, this.field_147327_f.func_184103_al().func_148545_a(this.field_147337_i));
+                net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.fmlServerHandshake(this.field_147327_f.func_184103_al(), this.field_147333_a, this.field_147327_f.func_184103_al().func_148545_a(this.field_147337_i));
             }
         }
     }
 
-    @Override
     public void func_147231_a(ITextComponent p_147231_1_)
     {
         field_147332_c.info("{} lost connection: {}", this.func_147317_d(), p_147231_1_.func_150260_c());
@@ -146,12 +144,9 @@
 
     public String func_147317_d()
     {
-        return this.field_147337_i != null
-               ? this.field_147337_i + " (" + this.field_147333_a.func_74430_c() + ")"
-               : String.valueOf(this.field_147333_a.func_74430_c());
+        return this.field_147337_i != null ? this.field_147337_i + " (" + this.field_147333_a.func_74430_c() + ")" : String.valueOf((Object)this.field_147333_a.func_74430_c());
     }
 
-    @Override
     public void func_147316_a(CPacketLoginStart p_147316_1_)
     {
         Validate.validState(this.field_147328_g == NetHandlerLoginServer.LoginState.HELLO, "Unexpected hello packet");
@@ -168,7 +163,6 @@
         }
     }
 
-    @Override
     public void func_147315_a(CPacketEncryptionResponse p_147315_1_)
     {
         Validate.validState(this.field_147328_g == NetHandlerLoginServer.LoginState.KEY, "Unexpected key packet");
@@ -183,31 +177,20 @@
             this.field_147335_k = p_147315_1_.func_149300_a(privatekey);
             this.field_147328_g = NetHandlerLoginServer.LoginState.AUTHENTICATING;
             this.field_147333_a.func_150727_a(this.field_147335_k);
-            (new Thread("User Authenticator #" + field_147331_b.incrementAndGet())
+            (new Thread(net.minecraftforge.fml.common.thread.SidedThreadGroups.SERVER, "User Authenticator #" + field_147331_b.incrementAndGet())
             {
-                @Override
                 public void run()
                 {
                     GameProfile gameprofile = NetHandlerLoginServer.this.field_147337_i;
 
                     try
                     {
-                        String s = new BigInteger(
-                            CryptManager.func_75895_a(
-                                "", NetHandlerLoginServer.this.field_147327_f.func_71250_E().getPublic(), NetHandlerLoginServer.this.field_147335_k
-                            )
-                        )
-                        .toString(16);
-                        NetHandlerLoginServer.this.field_147337_i = NetHandlerLoginServer.this.field_147327_f
-                                .func_147130_as()
-                                .hasJoinedServer(new GameProfile(null, gameprofile.getName()), s, this.func_191235_a());
+                        String s = (new BigInteger(CryptManager.func_75895_a("", NetHandlerLoginServer.this.field_147327_f.func_71250_E().getPublic(), NetHandlerLoginServer.this.field_147335_k))).toString(16);
+                        NetHandlerLoginServer.this.field_147337_i = NetHandlerLoginServer.this.field_147327_f.func_147130_as().hasJoinedServer(new GameProfile((UUID)null, gameprofile.getName()), s, this.func_191235_a());
 
                         if (NetHandlerLoginServer.this.field_147337_i != null)
                         {
-                            NetHandlerLoginServer.field_147332_c
-                            .info(
-                                "UUID of player {} is {}", NetHandlerLoginServer.this.field_147337_i.getName(), NetHandlerLoginServer.this.field_147337_i.getId()
-                            );
+                            NetHandlerLoginServer.field_147332_c.info("UUID of player {} is {}", NetHandlerLoginServer.this.field_147337_i.getName(), NetHandlerLoginServer.this.field_147337_i.getId());
                             NetHandlerLoginServer.this.field_147328_g = NetHandlerLoginServer.LoginState.READY_TO_ACCEPT;
                         }
                         else if (NetHandlerLoginServer.this.field_147327_f.func_71264_H())
@@ -218,11 +201,11 @@
                         }
                         else
                         {
-                            NetHandlerLoginServer.this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.unverified_username"));
-                            NetHandlerLoginServer.field_147332_c.error("Username '{}' tried to join with an invalid session", gameprofile.getName());
+                            NetHandlerLoginServer.this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.unverified_username", new Object[0]));
+                            NetHandlerLoginServer.field_147332_c.error("Username '{}' tried to join with an invalid session", (Object)gameprofile.getName());
                         }
                     }
-                    catch (AuthenticationUnavailableException authenticationunavailableexception)
+                    catch (AuthenticationUnavailableException var3)
                     {
                         if (NetHandlerLoginServer.this.field_147327_f.func_71264_H())
                         {
@@ -232,7 +215,7 @@
                         }
                         else
                         {
-                            NetHandlerLoginServer.this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.authservers_down"));
+                            NetHandlerLoginServer.this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.authservers_down", new Object[0]));
                             NetHandlerLoginServer.field_147332_c.error("Couldn't verify username because servers are unavailable");
                         }
                     }
@@ -241,12 +224,9 @@
                 private InetAddress func_191235_a()
                 {
                     SocketAddress socketaddress = NetHandlerLoginServer.this.field_147333_a.func_74430_c();
-                    return NetHandlerLoginServer.this.field_147327_f.func_190518_ac() && socketaddress instanceof InetSocketAddress
-                           ? ((InetSocketAddress)socketaddress).getAddress()
-                           : null;
+                    return NetHandlerLoginServer.this.field_147327_f.func_190518_ac() && socketaddress instanceof InetSocketAddress ? ((InetSocketAddress)socketaddress).getAddress() : null;
                 }
-            })
-            .start();
+            }).start();
         }
     }
 
