--- before/net/minecraft/server/network/NetHandlerLoginServer.java
+++ after/net/minecraft/server/network/NetHandlerLoginServer.java
@@ -69,14 +69,14 @@
             if (entityplayermp == null)
             {
                 this.field_147328_g = NetHandlerLoginServer.LoginState.READY_TO_ACCEPT;
-                this.field_147327_f.func_184103_al().func_72355_a(this.field_147333_a, this.field_181025_l);
+                net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.fmlServerHandshake(this.field_147327_f.func_184103_al(), this.field_147333_a, this.field_181025_l);
                 this.field_181025_l = null;
             }
         }
 
         if (this.field_147336_h++ == 600)
         {
-            this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.slow_login"));
+            this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.slow_login", new Object[0]));
         }
     }
 
@@ -105,7 +105,7 @@
 
         if (s != null)
         {
-            this.func_194026_b(new TextComponentTranslation(s));
+            this.func_194026_b(new TextComponentTranslation(s, new Object[0]));
         }
         else
         {
@@ -132,7 +132,7 @@
             }
             else
             {
-                this.field_147327_f.func_184103_al().func_72355_a(this.field_147333_a, this.field_147327_f.func_184103_al().func_148545_a(this.field_147337_i));
+                net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.fmlServerHandshake(this.field_147327_f.func_184103_al(), this.field_147333_a, this.field_147327_f.func_184103_al().func_148545_a(this.field_147337_i));
             }
         }
     }
@@ -177,7 +177,7 @@
             this.field_147335_k = p_147315_1_.func_149300_a(privatekey);
             this.field_147328_g = NetHandlerLoginServer.LoginState.AUTHENTICATING;
             this.field_147333_a.func_150727_a(this.field_147335_k);
-            (new Thread("User Authenticator #" + field_147331_b.incrementAndGet())
+            (new Thread(net.minecraftforge.fml.common.thread.SidedThreadGroups.SERVER, "User Authenticator #" + field_147331_b.incrementAndGet())
             {
                 public void run()
                 {
@@ -201,11 +201,11 @@
                         }
                         else
                         {
-                            NetHandlerLoginServer.this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.unverified_username"));
+                            NetHandlerLoginServer.this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.unverified_username", new Object[0]));
                             NetHandlerLoginServer.field_147332_c.error("Username '{}' tried to join with an invalid session", (Object)gameprofile.getName());
                         }
                     }
-                    catch (AuthenticationUnavailableException authenticationunavailableexception)
+                    catch (AuthenticationUnavailableException var3)
                     {
                         if (NetHandlerLoginServer.this.field_147327_f.func_71264_H())
                         {
@@ -215,7 +215,7 @@
                         }
                         else
                         {
-                            NetHandlerLoginServer.this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.authservers_down"));
+                            NetHandlerLoginServer.this.func_194026_b(new TextComponentTranslation("multiplayer.disconnect.authservers_down", new Object[0]));
                             NetHandlerLoginServer.field_147332_c.error("Couldn't verify username because servers are unavailable");
                         }
                     }
