--- before/net/minecraft/server/network/NetHandlerHandshakeTCP.java
+++ after/net/minecraft/server/network/NetHandlerHandshakeTCP.java
@@ -23,6 +23,8 @@
     @Override
     public void processHandshake(C00Handshake packetIn)
     {
+        if (!net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerHandshake(packetIn, this.networkManager)) return;
+
         switch (packetIn.getRequestedState())
         {
             case LOGIN:
