--- before/net/minecraft/util/CooldownTracker.java
+++ after/net/minecraft/util/CooldownTracker.java
@@ -3,6 +3,7 @@
 import com.google.common.collect.Maps;
 import java.util.Iterator;
 import java.util.Map;
+import java.util.Map.Entry;
 import net.minecraft.item.Item;
 import net.minecraft.util.math.MathHelper;
 import net.minecraftforge.fml.relauncher.Side;
@@ -10,7 +11,7 @@
 
 public class CooldownTracker
 {
-    private final Map<Item, CooldownTracker.Cooldown> field_185147_a = Maps.newHashMap();
+    private final Map<Item, CooldownTracker.Cooldown> field_185147_a = Maps.<Item, CooldownTracker.Cooldown>newHashMap();
     private int field_185148_b;
 
     public boolean func_185141_a(Item p_185141_1_)
@@ -40,11 +41,11 @@
 
         if (!this.field_185147_a.isEmpty())
         {
-            Iterator<Map.Entry<Item, CooldownTracker.Cooldown>> iterator = this.field_185147_a.entrySet().iterator();
+            Iterator<Entry<Item, CooldownTracker.Cooldown>> iterator = this.field_185147_a.entrySet().iterator();
 
             while (iterator.hasNext())
             {
-                Map.Entry<Item, CooldownTracker.Cooldown> entry = iterator.next();
+                Entry<Item, CooldownTracker.Cooldown> entry = (Entry)iterator.next();
 
                 if ((entry.getValue()).field_185138_b <= this.field_185148_b)
                 {
