--- before/net/minecraft/util/CryptManager.java
+++ after/net/minecraft/util/CryptManager.java
@@ -103,11 +103,13 @@
             KeyFactory keyfactory = KeyFactory.getInstance("RSA");
             return keyfactory.generatePublic(encodedkeyspec);
         }
-        catch (NoSuchAlgorithmException nosuchalgorithmexception)
+        catch (NoSuchAlgorithmException var3)
         {
+            ;
         }
-        catch (InvalidKeySpecException invalidkeyspecexception)
+        catch (InvalidKeySpecException var4)
         {
+            ;
         }
 
         field_180198_a.error("Public key reconstitute failed!");
