--- before/net/minecraft/util/EnumFacing.java
+++ after/net/minecraft/util/EnumFacing.java
@@ -12,8 +12,6 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.Vec3i;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public enum EnumFacing implements IStringSerializable
 {
@@ -33,17 +31,9 @@
     private final Vec3i field_176756_m;
     public static final EnumFacing[] field_82609_l = new EnumFacing[6];
     public static final EnumFacing[] field_176754_o = new EnumFacing[4];
-    private static final Map<String, EnumFacing> field_176761_p = Maps.newHashMap();
+    private static final Map<String, EnumFacing> field_176761_p = Maps.<String, EnumFacing>newHashMap();
 
-    private EnumFacing(
-        int p_i46016_3_,
-        int p_i46016_4_,
-        int p_i46016_5_,
-        String p_i46016_6_,
-        EnumFacing.AxisDirection p_i46016_7_,
-        EnumFacing.Axis p_i46016_8_,
-        Vec3i p_i46016_9_
-    )
+    private EnumFacing(int p_i46016_3_, int p_i46016_4_, int p_i46016_5_, String p_i46016_6_, EnumFacing.AxisDirection p_i46016_7_, EnumFacing.Axis p_i46016_8_, Vec3i p_i46016_9_)
     {
         this.field_176748_g = p_i46016_3_;
         this.field_176760_i = p_i46016_5_;
@@ -74,7 +64,6 @@
         return func_82600_a(this.field_176759_h);
     }
 
-    @SideOnly(Side.CLIENT)
     public EnumFacing func_176732_a(EnumFacing.Axis p_176732_1_)
     {
         switch (p_176732_1_)
@@ -125,7 +114,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     private EnumFacing func_176744_n()
     {
         switch (this)
@@ -145,7 +133,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     private EnumFacing func_176738_p()
     {
         switch (this)
@@ -207,10 +194,9 @@
     }
 
     @Nullable
-    @SideOnly(Side.CLIENT)
     public static EnumFacing func_176739_a(String p_176739_0_)
     {
-        return p_176739_0_ == null ? null : field_176761_p.get(p_176739_0_.toLowerCase(Locale.ROOT));
+        return p_176739_0_ == null ? null : (EnumFacing)field_176761_p.get(p_176739_0_.toLowerCase(Locale.ROOT));
     }
 
     public static EnumFacing func_82600_a(int p_82600_0_)
@@ -225,7 +211,7 @@
 
     public static EnumFacing func_176733_a(double p_176733_0_)
     {
-        return func_176731_b(MathHelper.func_76128_c(p_176733_0_ / 90.0 + 0.5) & 3);
+        return func_176731_b(MathHelper.func_76128_c(p_176733_0_ / 90.0D + 0.5D) & 3);
     }
 
     public float func_185119_l()
@@ -238,7 +224,6 @@
         return values()[p_176741_0_.nextInt(values().length)];
     }
 
-    @SideOnly(Side.CLIENT)
     public static EnumFacing func_176737_a(float p_176737_0_, float p_176737_1_, float p_176737_2_)
     {
         EnumFacing enumfacing = NORTH;
@@ -246,9 +231,7 @@
 
         for (EnumFacing enumfacing1 : values())
         {
-            float f1 = p_176737_0_ * (float)enumfacing1.field_176756_m.func_177958_n()
-                       + p_176737_1_ * (float)enumfacing1.field_176756_m.func_177956_o()
-                       + p_176737_2_ * (float)enumfacing1.field_176756_m.func_177952_p();
+            float f1 = p_176737_0_ * (float)enumfacing1.field_176756_m.func_177958_n() + p_176737_1_ * (float)enumfacing1.field_176756_m.func_177956_o() + p_176737_2_ * (float)enumfacing1.field_176756_m.func_177952_p();
 
             if (f1 > f)
             {
@@ -260,13 +243,11 @@
         return enumfacing;
     }
 
-    @Override
     public String toString()
     {
         return this.field_176757_j;
     }
 
-    @Override
     public String func_176610_l()
     {
         return this.field_176757_j;
@@ -287,17 +268,16 @@
 
     public static EnumFacing func_190914_a(BlockPos p_190914_0_, EntityLivingBase p_190914_1_)
     {
-        if (Math.abs(p_190914_1_.field_70165_t - (double)((float)p_190914_0_.func_177958_n() + 0.5F)) < 2.0
-                && Math.abs(p_190914_1_.field_70161_v - (double)((float)p_190914_0_.func_177952_p() + 0.5F)) < 2.0)
+        if (Math.abs(p_190914_1_.field_70165_t - (double)((float)p_190914_0_.func_177958_n() + 0.5F)) < 2.0D && Math.abs(p_190914_1_.field_70161_v - (double)((float)p_190914_0_.func_177952_p() + 0.5F)) < 2.0D)
         {
             double d0 = p_190914_1_.field_70163_u + (double)p_190914_1_.func_70047_e();
 
-            if (d0 - (double)p_190914_0_.func_177956_o() > 2.0)
+            if (d0 - (double)p_190914_0_.func_177956_o() > 2.0D)
             {
                 return UP;
             }
 
-            if ((double)p_190914_0_.func_177956_o() - d0 > 0.0)
+            if ((double)p_190914_0_.func_177956_o() - d0 > 0.0D)
             {
                 return DOWN;
             }
@@ -306,7 +286,6 @@
         return p_190914_1_.func_174811_aO().func_176734_d();
     }
 
-    @SideOnly(Side.CLIENT)
     public Vec3i func_176730_m()
     {
         return this.field_176756_m;
@@ -332,7 +311,7 @@
         Y("y", EnumFacing.Plane.VERTICAL),
         Z("z", EnumFacing.Plane.HORIZONTAL);
 
-        private static final Map<String, EnumFacing.Axis> field_176725_d = Maps.newHashMap();
+        private static final Map<String, EnumFacing.Axis> field_176725_d = Maps.<String, EnumFacing.Axis>newHashMap();
         private final String field_176726_e;
         private final EnumFacing.Plane field_176723_f;
 
@@ -343,10 +322,9 @@
         }
 
         @Nullable
-        @SideOnly(Side.CLIENT)
         public static EnumFacing.Axis func_176717_a(String p_176717_0_)
         {
-            return p_176717_0_ == null ? null : field_176725_d.get(p_176717_0_.toLowerCase(Locale.ROOT));
+            return p_176717_0_ == null ? null : (EnumFacing.Axis)field_176725_d.get(p_176717_0_.toLowerCase(Locale.ROOT));
         }
 
         public String func_176719_a()
@@ -364,7 +342,6 @@
             return this.field_176723_f == EnumFacing.Plane.HORIZONTAL;
         }
 
-        @Override
         public String toString()
         {
             return this.field_176726_e;
@@ -380,7 +357,6 @@
             return this.field_176723_f;
         }
 
-        @Override
         public String func_176610_l()
         {
             return this.field_176726_e;
@@ -413,7 +389,6 @@
             return this.field_179528_c;
         }
 
-        @Override
         public String toString()
         {
             return this.field_179525_d;
@@ -448,10 +423,9 @@
             return p_apply_1_ != null && p_apply_1_.func_176740_k().func_176716_d() == this;
         }
 
-        @Override
         public Iterator<EnumFacing> iterator()
         {
-            return Iterators.forArray(this.func_179516_a());
+            return Iterators.<EnumFacing>forArray(this.func_179516_a());
         }
     }
 }
