--- before/net/minecraft/util/EnumFacing.java
+++ after/net/minecraft/util/EnumFacing.java
@@ -12,8 +12,6 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.Vec3i;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public enum EnumFacing implements IStringSerializable
 {
@@ -74,7 +72,6 @@
         return byIndex(this.opposite);
     }
 
-    @SideOnly(Side.CLIENT)
     public EnumFacing rotateAround(EnumFacing.Axis axis)
     {
         switch (axis)
@@ -125,7 +122,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     private EnumFacing rotateX()
     {
         switch (this)
@@ -145,7 +141,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     private EnumFacing rotateZ()
     {
         switch (this)
@@ -207,7 +202,6 @@
     }
 
     @Nullable
-    @SideOnly(Side.CLIENT)
     public static EnumFacing byName(String name)
     {
         return name == null ? null : NAME_LOOKUP.get(name.toLowerCase(Locale.ROOT));
@@ -238,7 +232,6 @@
         return values()[rand.nextInt(values().length)];
     }
 
-    @SideOnly(Side.CLIENT)
     public static EnumFacing getFacingFromVector(float x, float y, float z)
     {
         EnumFacing enumfacing = NORTH;
@@ -306,7 +299,6 @@
         return placer.getHorizontalFacing().getOpposite();
     }
 
-    @SideOnly(Side.CLIENT)
     public Vec3i getDirectionVec()
     {
         return this.directionVec;
@@ -343,7 +335,6 @@
         }
 
         @Nullable
-        @SideOnly(Side.CLIENT)
         public static EnumFacing.Axis byName(String name)
         {
             return name == null ? null : NAME_LOOKUP.get(name.toLowerCase(Locale.ROOT));
