--- before/net/minecraft/util/HttpUtil.java
+++ after/net/minecraft/util/HttpUtil.java
@@ -19,6 +19,7 @@
 import java.net.URL;
 import java.net.URLEncoder;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.concurrent.Executors;
 import java.util.concurrent.atomic.AtomicInteger;
 import javax.annotation.Nullable;
@@ -40,7 +41,7 @@
     {
         StringBuilder stringbuilder = new StringBuilder();
 
-        for (Map.Entry<String, Object> entry : p_76179_0_.entrySet())
+        for (Entry<String, Object> entry : p_76179_0_.entrySet())
         {
             if (stringbuilder.length() > 0)
             {
@@ -125,9 +126,9 @@
     }
 
     @SideOnly(Side.CLIENT)
-    public static ListenableFuture func_180192_a(final File p_180192_0_, final String p_180192_1_, final Map<String, String> p_180192_2_, final int p_180192_3_, @Nullable final IProgressUpdate p_180192_4_, final Proxy p_180192_5_)
+    public static ListenableFuture<Object> func_180192_a(final File p_180192_0_, final String p_180192_1_, final Map<String, String> p_180192_2_, final int p_180192_3_, @Nullable final IProgressUpdate p_180192_4_, final Proxy p_180192_5_)
     {
-        return field_180193_a.submit(new Runnable()
+        ListenableFuture<?> listenablefuture = field_180193_a.submit(new Runnable()
         {
             public void run()
             {
@@ -152,7 +153,7 @@
                         float f = 0.0F;
                         float f1 = (float)p_180192_2_.entrySet().size();
 
-                        for (Map.Entry<String, String> entry : p_180192_2_.entrySet())
+                        for (Entry<String, String> entry : p_180192_2_.entrySet())
                         {
                             httpurlconnection.setRequestProperty(entry.getKey(), entry.getValue());
 
@@ -185,7 +186,7 @@
                                 return;
                             }
 
-                            HttpUtil.field_151227_b.warn("Deleting {} as it does not match what we currently have ({} vs our {}).", p_180192_0_, i, j);
+                            HttpUtil.field_151227_b.warn("Deleting {} as it does not match what we currently have ({} vs our {}).", p_180192_0_, Integer.valueOf(i), Long.valueOf(j));
                             FileUtils.deleteQuietly(p_180192_0_);
                         }
                         else if (p_180192_0_.getParentFile() != null)
@@ -279,6 +280,7 @@
                 }
             }
         });
+        return (ListenableFuture<Object>) listenablefuture;
     }
 
     @SideOnly(Side.CLIENT)
@@ -301,8 +303,9 @@
                     serversocket.close();
                 }
             }
-            catch (IOException ioexception)
+            catch (IOException var8)
             {
+                ;
             }
         }
 
