--- before/net/minecraft/util/NonNullList.java
+++ after/net/minecraft/util/NonNullList.java
@@ -15,7 +15,7 @@
 
     public static <E> NonNullList<E> func_191196_a()
     {
-        return new NonNullList<>();
+        return new NonNullList<E>();
     }
 
     public static <E> NonNullList<E> func_191197_a(int p_191197_0_, E p_191197_1_)
@@ -23,12 +23,12 @@
         Validate.notNull(p_191197_1_);
         Object[] aobject = new Object[p_191197_0_];
         Arrays.fill(aobject, p_191197_1_);
-        return new NonNullList<>(Arrays.asList((E[])aobject), p_191197_1_);
+        return new NonNullList<E>(Arrays.asList((E[])aobject), p_191197_1_);
     }
 
     public static <E> NonNullList<E> func_193580_a(E p_193580_0_, E... p_193580_1_)
     {
-        return new NonNullList<>(Arrays.asList(p_193580_1_), p_193580_0_);
+        return new NonNullList<E>(Arrays.asList(p_193580_1_), p_193580_0_);
     }
 
     protected NonNullList()
