--- before/net/minecraft/util/SoundCategory.java
+++ after/net/minecraft/util/SoundCategory.java
@@ -17,7 +17,7 @@
     AMBIENT("ambient"),
     VOICE("voice");
 
-    private static final Map<String, SoundCategory> field_187961_k = Maps.newHashMap();
+    private static final Map<String, SoundCategory> field_187961_k = Maps.<String, SoundCategory>newHashMap();
     private final String field_187962_l;
 
     private SoundCategory(String p_i46833_3_)
