--- before/net/minecraft/network/NetHandlerPlayServer.java
+++ after/net/minecraft/network/NetHandlerPlayServer.java
@@ -370,6 +370,7 @@
                 d7 = d4 - this.lowestRiddenY1 - 1.0E-6;
                 d8 = d5 - this.lowestRiddenZ1;
                 entity.move(MoverType.PLAYER, d6, d7, d8);
+                double d11 = d7;
                 d6 = d3 - entity.posX;
                 d7 = d4 - entity.posY;
 
@@ -389,11 +390,13 @@
                 }
 
                 entity.setPositionAndRotation(d3, d4, d5, f, f1);
+                this.player.setPositionAndRotation(d3, d4, d5, this.player.rotationYaw, this.player.rotationPitch); // Forge - Resync player position on vehicle moving
                 boolean flag2 = worldserver.getCollisionBoxes(entity, entity.getEntityBoundingBox().shrink(0.0625)).isEmpty();
 
                 if (flag && (flag1 || !flag2))
                 {
                     entity.setPositionAndRotation(d0, d1, d2, f, f1);
+                    this.player.setPositionAndRotation(d0, d1, d2, this.player.rotationYaw, this.player.rotationPitch); // Forge - Resync player position on vehicle moving
                     this.netManager.sendPacket(new SPacketMoveVehicle(entity));
                     return;
                 }
@@ -401,7 +404,7 @@
                 this.server.getPlayerList().serverUpdateMovingPlayer(this.player);
                 this.player
                 .addMovementStat(this.player.posX - d0, this.player.posY - d1, this.player.posZ - d2);
-                this.vehicleFloating = d7 >= -0.03125
+                this.vehicleFloating = d11 >= -0.03125
                                       && !this.server.isFlightAllowed()
                                       && !worldserver.checkBlockCollision(entity.getEntityBoundingBox().grow(0.0625).expand(0.0, -0.55, 0.0));
                 this.lowestRiddenX1 = entity.posX;
@@ -600,6 +603,7 @@
 
                             this.player.move(MoverType.PLAYER, d7, d8, d9);
                             this.player.onGround = packetIn.isOnGround();
+                            double d12 = d8;
                             d7 = d4 - this.player.posX;
                             d8 = d5 - this.player.posY;
 
@@ -637,7 +641,7 @@
                                 }
                             }
 
-                            this.floating = d8 >= -0.03125;
+                            this.floating = d12 >= -0.03125;
                             this.floating = this.floating & (!this.server.isFlightAllowed() && !this.player.capabilities.allowFlying);
                             this.floating = this.floating
                                                   & (
@@ -743,7 +747,10 @@
                 double d2 = this.player.posZ - (blockpos.getZ() + 0.5);
                 double d3 = d0 * d0 + d1 * d1 + d2 * d2;
 
-                if (d3 > 36.0)
+                double dist = player.getEntityAttribute(EntityPlayer.REACH_DISTANCE).getAttributeValue() + 1;
+                dist *= dist;
+
+                if (d3 > dist)
                 {
                     return;
                 }
@@ -803,8 +810,10 @@
         if (blockpos.getY() < this.server.getBuildLimit() - 1
                 || enumfacing != EnumFacing.UP && blockpos.getY() < this.server.getBuildLimit())
         {
+            double dist = player.getEntityAttribute(EntityPlayer.REACH_DISTANCE).getAttributeValue() + 3;
+            dist *= dist;
             if (this.targetPos == null
-                    && this.player.getDistanceSq(blockpos.getX() + 0.5, blockpos.getY() + 0.5, blockpos.getZ() + 0.5) < 64.0
+                    && this.player.getDistanceSq(blockpos.getX() + 0.5, blockpos.getY() + 0.5, blockpos.getZ() + 0.5) < dist
                     && !this.server.isBlockProtected(worldserver, blockpos, this.player)
                     && worldserver.getWorldBorder().contains(blockpos))
             {
@@ -880,19 +889,20 @@
                 {
                     this.player.setPositionAndUpdate(entity.posX, entity.posY, entity.posZ);
                 }
-                else
+                else if (net.minecraftforge.common.ForgeHooks.onTravelToDimension(this.player, entity.dimension))
                 {
+                    int prevDimension = this.player.dimension;
                     WorldServer worldserver1 = this.player.getServerWorld();
                     WorldServer worldserver2 = (WorldServer)entity.world;
                     this.player.dimension = entity.dimension;
                     this.sendPacket(
                         new SPacketRespawn(
                             this.player.dimension,
-                            worldserver1.getDifficulty(),
-                            worldserver1.getWorldInfo().getTerrainType(),
+                            worldserver2.getDifficulty(),
+                            worldserver2.getWorldInfo().getTerrainType(),
                             this.player.interactionManager.getGameType()
                         )
-                    );
+                    ); // Forge: Use new dimensions information
                     this.server.getPlayerList().updatePermissionLevel(this.player);
                     worldserver1.removeEntityDangerously(this.player);
                     this.player.isDead = false;
@@ -911,6 +921,7 @@
                     this.player.interactionManager.setWorld(worldserver2);
                     this.server.getPlayerList().updateTimeAndWeatherForPlayer(this.player, worldserver2);
                     this.server.getPlayerList().syncPlayerInventory(this.player);
+                    net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerChangedDimensionEvent(this.player, prevDimension, this.player.dimension);
                 }
             }
         }
@@ -1036,7 +1047,9 @@
             }
             else
             {
-                ITextComponent itextcomponent = new TextComponentTranslation("chat.type.text", this.player.getDisplayName(), s);
+                ITextComponent itextcomponent = new TextComponentTranslation("chat.type.text", this.player.getDisplayName(), net.minecraftforge.common.ForgeHooks.newChatWithLinks(s));
+                itextcomponent = net.minecraftforge.common.ForgeHooks.onServerChatEvent(this, s, itextcomponent);
+                if (itextcomponent == null) return;
                 this.server.getPlayerList().sendMessage(itextcomponent, false);
             }
 
@@ -1175,6 +1188,7 @@
                 else if (packetIn.getAction() == CPacketUseEntity.Action.INTERACT_AT)
                 {
                     EnumHand enumhand1 = packetIn.getHand();
+                    if(net.minecraftforge.common.ForgeHooks.onInteractEntityAt(player, entity, packetIn.getHitVec(), enumhand1) != null) return;
                     entity.applyPlayerInteraction(this.player, packetIn.getHitVec(), enumhand1);
                 }
                 else if (packetIn.getAction() == CPacketUseEntity.Action.ATTACK)
@@ -1216,7 +1230,7 @@
                         return;
                     }
 
-                    this.player = this.server.getPlayerList().recreatePlayerEntity(this.player, 0, false);
+                    this.player = this.server.getPlayerList().recreatePlayerEntity(this.player, player.dimension, false);
 
                     if (this.server.isHardcore())
                     {
@@ -1264,7 +1278,7 @@
                                        .openContainer
                                        .slotClick(packetIn.getSlotId(), packetIn.getUsedButton(), packetIn.getClickType(), this.player);
 
-                if (ItemStack.areItemStacksEqual(packetIn.getClickedItem(), itemstack2))
+                if (ItemStack.areItemStacksEqualUsingNBTShareTag(packetIn.getClickedItem(), itemstack2))
                 {
                     this.player.connection.sendPacket(new SPacketConfirmTransaction(packetIn.getWindowId(), packetIn.getActionNumber(), true));
                     this.player.isChangingQuantityOnly = true;
