--- before/net/minecraft/network/NetHandlerPlayServer.java
+++ after/net/minecraft/network/NetHandlerPlayServer.java
@@ -133,7 +133,7 @@
     private long field_194404_h;
     private int field_147374_l;
     private int field_147375_m;
-    private final IntHashMap<Short> field_147372_n = new IntHashMap<>();
+    private final IntHashMap<Short> field_147372_n = new IntHashMap<Short>();
     private double field_184349_l;
     private double field_184350_m;
     private double field_184351_n;
@@ -167,22 +167,20 @@
         p_i1530_3_.field_71135_a = this;
     }
 
-    @Override
     public void func_73660_a()
     {
         this.func_184342_d();
         this.field_147369_b.func_71127_g();
-        this.field_147369_b
-        .func_70080_a(this.field_184349_l, this.field_184350_m, this.field_184351_n, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A);
-        this.field_147368_e++;
+        this.field_147369_b.func_70080_a(this.field_184349_l, this.field_184350_m, this.field_184351_n, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A);
+        ++this.field_147368_e;
         this.field_184348_G = this.field_184347_F;
 
         if (this.field_184344_B)
         {
             if (++this.field_147365_f > 80)
             {
-                field_147370_c.warn("{} was kicked for floating too long!", this.field_147369_b.func_70005_c_());
-                this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.flying"));
+                field_147370_c.warn("{} was kicked for floating too long!", (Object)this.field_147369_b.func_70005_c_());
+                this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.flying", new Object[0]));
                 return;
             }
         }
@@ -207,8 +205,8 @@
             {
                 if (++this.field_184346_E > 80)
                 {
-                    field_147370_c.warn("{} was kicked for floating a vehicle too long!", this.field_147369_b.func_70005_c_());
-                    this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.flying"));
+                    field_147370_c.warn("{} was kicked for floating a vehicle too long!", (Object)this.field_147369_b.func_70005_c_());
+                    this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.flying", new Object[0]));
                     return;
                 }
             }
@@ -232,7 +230,7 @@
         {
             if (this.field_194403_g)
             {
-                this.func_194028_b(new TextComponentTranslation("disconnect.timeout"));
+                this.func_194028_b(new TextComponentTranslation("disconnect.timeout", new Object[0]));
             }
             else
             {
@@ -247,19 +245,17 @@
 
         if (this.field_147374_l > 0)
         {
-            this.field_147374_l--;
+            --this.field_147374_l;
         }
 
         if (this.field_147375_m > 0)
         {
-            this.field_147375_m--;
+            --this.field_147375_m;
         }
 
-        if (this.field_147369_b.func_154331_x() > 0L
-                && this.field_147367_d.func_143007_ar() > 0
-                && MinecraftServer.func_130071_aq() - this.field_147369_b.func_154331_x() > (long)(this.field_147367_d.func_143007_ar() * 1000 * 60))
+        if (this.field_147369_b.func_154331_x() > 0L && this.field_147367_d.func_143007_ar() > 0 && MinecraftServer.func_130071_aq() - this.field_147369_b.func_154331_x() > (long)(this.field_147367_d.func_143007_ar() * 1000 * 60))
         {
-            this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.idling"));
+            this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.idling", new Object[0]));
         }
     }
 
@@ -282,7 +278,6 @@
     {
         this.field_147371_a.func_179288_a(new SPacketDisconnect(p_194028_1_), new GenericFutureListener < Future <? super Void >> ()
         {
-            @Override
             public void operationComplete(Future <? super Void > p_operationComplete_1_) throws Exception
             {
                 NetHandlerPlayServer.this.field_147371_a.func_150718_a(p_194028_1_);
@@ -291,7 +286,6 @@
         this.field_147371_a.func_150721_g();
         Futures.getUnchecked(this.field_147367_d.func_152344_a(new Runnable()
         {
-            @Override
             public void run()
             {
                 NetHandlerPlayServer.this.field_147371_a.func_179293_l();
@@ -299,7 +293,6 @@
         }));
     }
 
-    @Override
     public void func_147358_a(CPacketInput p_147358_1_)
     {
         PacketThreadUtil.func_180031_a(p_147358_1_, this, this.field_147369_b.func_71121_q());
@@ -308,34 +301,28 @@
 
     private static boolean func_183006_b(CPacketPlayer p_183006_0_)
     {
-        return Doubles.isFinite(p_183006_0_.func_186997_a(0.0))
-               && Doubles.isFinite(p_183006_0_.func_186996_b(0.0))
-               && Doubles.isFinite(p_183006_0_.func_187000_c(0.0))
-               && Floats.isFinite(p_183006_0_.func_186998_b(0.0F))
-               && Floats.isFinite(p_183006_0_.func_186999_a(0.0F))
-               ? Math.abs(p_183006_0_.func_186997_a(0.0)) > 3.0E7
-               || Math.abs(p_183006_0_.func_186996_b(0.0)) > 3.0E7
-               || Math.abs(p_183006_0_.func_187000_c(0.0)) > 3.0E7
-               : true;
+        if (Doubles.isFinite(p_183006_0_.func_186997_a(0.0D)) && Doubles.isFinite(p_183006_0_.func_186996_b(0.0D)) && Doubles.isFinite(p_183006_0_.func_187000_c(0.0D)) && Floats.isFinite(p_183006_0_.func_186998_b(0.0F)) && Floats.isFinite(p_183006_0_.func_186999_a(0.0F)))
+        {
+            return Math.abs(p_183006_0_.func_186997_a(0.0D)) > 3.0E7D || Math.abs(p_183006_0_.func_186996_b(0.0D)) > 3.0E7D || Math.abs(p_183006_0_.func_187000_c(0.0D)) > 3.0E7D;
+        }
+        else
+        {
+            return true;
+        }
     }
 
     private static boolean func_184341_b(CPacketVehicleMove p_184341_0_)
     {
-        return !Doubles.isFinite(p_184341_0_.func_187004_a())
-               || !Doubles.isFinite(p_184341_0_.func_187002_b())
-               || !Doubles.isFinite(p_184341_0_.func_187003_c())
-               || !Floats.isFinite(p_184341_0_.func_187005_e())
-               || !Floats.isFinite(p_184341_0_.func_187006_d());
+        return !Doubles.isFinite(p_184341_0_.func_187004_a()) || !Doubles.isFinite(p_184341_0_.func_187002_b()) || !Doubles.isFinite(p_184341_0_.func_187003_c()) || !Floats.isFinite(p_184341_0_.func_187005_e()) || !Floats.isFinite(p_184341_0_.func_187006_d());
     }
 
-    @Override
     public void func_184338_a(CPacketVehicleMove p_184338_1_)
     {
         PacketThreadUtil.func_180031_a(p_184338_1_, this, this.field_147369_b.func_71121_q());
 
         if (func_184341_b(p_184338_1_))
         {
-            this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.invalid_vehicle_movement"));
+            this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.invalid_vehicle_movement", new Object[0]));
         }
         else
         {
@@ -358,52 +345,52 @@
                 double d9 = entity.field_70159_w * entity.field_70159_w + entity.field_70181_x * entity.field_70181_x + entity.field_70179_y * entity.field_70179_y;
                 double d10 = d6 * d6 + d7 * d7 + d8 * d8;
 
-                if (d10 - d9 > 100.0 && (!this.field_147367_d.func_71264_H() || !this.field_147367_d.func_71214_G().equals(entity.func_70005_c_())))
+                if (d10 - d9 > 100.0D && (!this.field_147367_d.func_71264_H() || !this.field_147367_d.func_71214_G().equals(entity.func_70005_c_())))
                 {
-                    field_147370_c.warn("{} (vehicle of {}) moved too quickly! {},{},{}", entity.func_70005_c_(), this.field_147369_b.func_70005_c_(), d6, d7, d8);
+                    field_147370_c.warn("{} (vehicle of {}) moved too quickly! {},{},{}", entity.func_70005_c_(), this.field_147369_b.func_70005_c_(), Double.valueOf(d6), Double.valueOf(d7), Double.valueOf(d8));
                     this.field_147371_a.func_179290_a(new SPacketMoveVehicle(entity));
                     return;
                 }
 
-                boolean flag = worldserver.func_184144_a(entity, entity.func_174813_aQ().func_186664_h(0.0625)).isEmpty();
+                boolean flag = worldserver.func_184144_a(entity, entity.func_174813_aQ().func_186664_h(0.0625D)).isEmpty();
                 d6 = d3 - this.field_184359_v;
-                d7 = d4 - this.field_184360_w - 1.0E-6;
+                d7 = d4 - this.field_184360_w - 1.0E-6D;
                 d8 = d5 - this.field_184361_x;
                 entity.func_70091_d(MoverType.PLAYER, d6, d7, d8);
+                double d11 = d7;
                 d6 = d3 - entity.field_70165_t;
                 d7 = d4 - entity.field_70163_u;
 
-                if (d7 > -0.5 || d7 < 0.5)
+                if (d7 > -0.5D || d7 < 0.5D)
                 {
-                    d7 = 0.0;
+                    d7 = 0.0D;
                 }
 
                 d8 = d5 - entity.field_70161_v;
                 d10 = d6 * d6 + d7 * d7 + d8 * d8;
                 boolean flag1 = false;
 
-                if (d10 > 0.0625)
+                if (d10 > 0.0625D)
                 {
                     flag1 = true;
-                    field_147370_c.warn("{} moved wrongly!", entity.func_70005_c_());
+                    field_147370_c.warn("{} moved wrongly!", (Object)entity.func_70005_c_());
                 }
 
                 entity.func_70080_a(d3, d4, d5, f, f1);
-                boolean flag2 = worldserver.func_184144_a(entity, entity.func_174813_aQ().func_186664_h(0.0625)).isEmpty();
+                this.field_147369_b.func_70080_a(d3, d4, d5, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A); // Forge - Resync player position on vehicle moving
+                boolean flag2 = worldserver.func_184144_a(entity, entity.func_174813_aQ().func_186664_h(0.0625D)).isEmpty();
 
                 if (flag && (flag1 || !flag2))
                 {
                     entity.func_70080_a(d0, d1, d2, f, f1);
+                    this.field_147369_b.func_70080_a(d0, d1, d2, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A); // Forge - Resync player position on vehicle moving
                     this.field_147371_a.func_179290_a(new SPacketMoveVehicle(entity));
                     return;
                 }
 
                 this.field_147367_d.func_184103_al().func_72358_d(this.field_147369_b);
-                this.field_147369_b
-                .func_71000_j(this.field_147369_b.field_70165_t - d0, this.field_147369_b.field_70163_u - d1, this.field_147369_b.field_70161_v - d2);
-                this.field_184345_D = d7 >= -0.03125
-                                      && !this.field_147367_d.func_71231_X()
-                                      && !worldserver.func_72829_c(entity.func_174813_aQ().func_186662_g(0.0625).func_72321_a(0.0, -0.55, 0.0));
+                this.field_147369_b.func_71000_j(this.field_147369_b.field_70165_t - d0, this.field_147369_b.field_70163_u - d1, this.field_147369_b.field_70161_v - d2);
+                this.field_184345_D = d11 >= -0.03125D && !this.field_147367_d.func_71231_X() && !worldserver.func_72829_c(entity.func_174813_aQ().func_186662_g(0.0625D).func_72321_a(0.0D, -0.55D, 0.0D));
                 this.field_184359_v = entity.field_70165_t;
                 this.field_184360_w = entity.field_70163_u;
                 this.field_184361_x = entity.field_70161_v;
@@ -411,21 +398,13 @@
         }
     }
 
-    @Override
     public void func_184339_a(CPacketConfirmTeleport p_184339_1_)
     {
         PacketThreadUtil.func_180031_a(p_184339_1_, this, this.field_147369_b.func_71121_q());
 
         if (p_184339_1_.func_186987_a() == this.field_184363_z)
         {
-            this.field_147369_b
-            .func_70080_a(
-                this.field_184362_y.field_72450_a,
-                this.field_184362_y.field_72448_b,
-                this.field_184362_y.field_72449_c,
-                this.field_147369_b.field_70177_z,
-                this.field_147369_b.field_70125_A
-            );
+            this.field_147369_b.func_70080_a(this.field_184362_y.field_72450_a, this.field_184362_y.field_72448_b, this.field_184362_y.field_72449_c, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A);
 
             if (this.field_147369_b.func_184850_K())
             {
@@ -439,7 +418,6 @@
         }
     }
 
-    @Override
     public void func_191984_a(CPacketRecipeInfo p_191984_1_)
     {
         PacketThreadUtil.func_180031_a(p_191984_1_, this, this.field_147369_b.func_71121_q());
@@ -455,7 +433,6 @@
         }
     }
 
-    @Override
     public void func_194027_a(CPacketSeenAdvancements p_194027_1_)
     {
         PacketThreadUtil.func_180031_a(p_194027_1_, this, this.field_147369_b.func_71121_q());
@@ -472,14 +449,13 @@
         }
     }
 
-    @Override
     public void func_147347_a(CPacketPlayer p_147347_1_)
     {
         PacketThreadUtil.func_180031_a(p_147347_1_, this, this.field_147369_b.func_71121_q());
 
         if (func_183006_b(p_147347_1_))
         {
-            this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.invalid_player_movement"));
+            this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.invalid_player_movement", new Object[0]));
         }
         else
         {
@@ -497,13 +473,7 @@
                     if (this.field_147368_e - this.field_184343_A > 20)
                     {
                         this.field_184343_A = this.field_147368_e;
-                        this.func_147364_a(
-                            this.field_184362_y.field_72450_a,
-                            this.field_184362_y.field_72448_b,
-                            this.field_184362_y.field_72449_c,
-                            this.field_147369_b.field_70177_z,
-                            this.field_147369_b.field_70125_A
-                        );
+                        this.func_147364_a(this.field_184362_y.field_72450_a, this.field_184362_y.field_72448_b, this.field_184362_y.field_72449_c, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A);
                     }
                 }
                 else
@@ -512,14 +482,7 @@
 
                     if (this.field_147369_b.func_184218_aH())
                     {
-                        this.field_147369_b
-                        .func_70080_a(
-                            this.field_147369_b.field_70165_t,
-                            this.field_147369_b.field_70163_u,
-                            this.field_147369_b.field_70161_v,
-                            p_147347_1_.func_186999_a(this.field_147369_b.field_70177_z),
-                            p_147347_1_.func_186998_b(this.field_147369_b.field_70125_A)
-                        );
+                        this.field_147369_b.func_70080_a(this.field_147369_b.field_70165_t, this.field_147369_b.field_70163_u, this.field_147369_b.field_70161_v, p_147347_1_.func_186999_a(this.field_147369_b.field_70177_z), p_147347_1_.func_186998_b(this.field_147369_b.field_70125_A));
                         this.field_147367_d.func_184103_al().func_72358_d(this.field_147369_b);
                     }
                     else
@@ -536,99 +499,76 @@
                         double d7 = d4 - this.field_184349_l;
                         double d8 = d5 - this.field_184350_m;
                         double d9 = d6 - this.field_184351_n;
-                        double d10 = this.field_147369_b.field_70159_w * this.field_147369_b.field_70159_w
-                                     + this.field_147369_b.field_70181_x * this.field_147369_b.field_70181_x
-                                     + this.field_147369_b.field_70179_y * this.field_147369_b.field_70179_y;
+                        double d10 = this.field_147369_b.field_70159_w * this.field_147369_b.field_70159_w + this.field_147369_b.field_70181_x * this.field_147369_b.field_70181_x + this.field_147369_b.field_70179_y * this.field_147369_b.field_70179_y;
                         double d11 = d7 * d7 + d8 * d8 + d9 * d9;
 
                         if (this.field_147369_b.func_70608_bn())
                         {
-                            if (d11 > 1.0)
+                            if (d11 > 1.0D)
                             {
-                                this.func_147364_a(
-                                    this.field_147369_b.field_70165_t,
-                                    this.field_147369_b.field_70163_u,
-                                    this.field_147369_b.field_70161_v,
-                                    p_147347_1_.func_186999_a(this.field_147369_b.field_70177_z),
-                                    p_147347_1_.func_186998_b(this.field_147369_b.field_70125_A)
-                                );
+                                this.func_147364_a(this.field_147369_b.field_70165_t, this.field_147369_b.field_70163_u, this.field_147369_b.field_70161_v, p_147347_1_.func_186999_a(this.field_147369_b.field_70177_z), p_147347_1_.func_186998_b(this.field_147369_b.field_70125_A));
                             }
                         }
                         else
                         {
-                            this.field_184347_F++;
+                            ++this.field_184347_F;
                             int i = this.field_184347_F - this.field_184348_G;
 
                             if (i > 5)
                             {
-                                field_147370_c.debug("{} is sending move packets too frequently ({} packets since last tick)", this.field_147369_b.func_70005_c_(), i);
+                                field_147370_c.debug("{} is sending move packets too frequently ({} packets since last tick)", this.field_147369_b.func_70005_c_(), Integer.valueOf(i));
                                 i = 1;
                             }
 
-                            if (!this.field_147369_b.func_184850_K()
-                                    && (
-                                        !this.field_147369_b.func_71121_q().func_82736_K().func_82766_b("disableElytraMovementCheck")
-                                        || !this.field_147369_b.func_184613_cA()
-                                    ))
+                            if (!this.field_147369_b.func_184850_K() && (!this.field_147369_b.func_71121_q().func_82736_K().func_82766_b("disableElytraMovementCheck") || !this.field_147369_b.func_184613_cA()))
                             {
                                 float f2 = this.field_147369_b.func_184613_cA() ? 300.0F : 100.0F;
 
-                                if (d11 - d10 > (double)(f2 * (float)i)
-                                        && (!this.field_147367_d.func_71264_H() || !this.field_147367_d.func_71214_G().equals(this.field_147369_b.func_70005_c_())))
+                                if (d11 - d10 > (double)(f2 * (float)i) && (!this.field_147367_d.func_71264_H() || !this.field_147367_d.func_71214_G().equals(this.field_147369_b.func_70005_c_())))
                                 {
-                                    field_147370_c.warn("{} moved too quickly! {},{},{}", this.field_147369_b.func_70005_c_(), d7, d8, d9);
-                                    this.func_147364_a(
-                                        this.field_147369_b.field_70165_t,
-                                        this.field_147369_b.field_70163_u,
-                                        this.field_147369_b.field_70161_v,
-                                        this.field_147369_b.field_70177_z,
-                                        this.field_147369_b.field_70125_A
-                                    );
+                                    field_147370_c.warn("{} moved too quickly! {},{},{}", this.field_147369_b.func_70005_c_(), Double.valueOf(d7), Double.valueOf(d8), Double.valueOf(d9));
+                                    this.func_147364_a(this.field_147369_b.field_70165_t, this.field_147369_b.field_70163_u, this.field_147369_b.field_70161_v, this.field_147369_b.field_70177_z, this.field_147369_b.field_70125_A);
                                     return;
                                 }
                             }
 
-                            boolean flag2 = worldserver.func_184144_a(this.field_147369_b, this.field_147369_b.func_174813_aQ().func_186664_h(0.0625)).isEmpty();
+                            boolean flag2 = worldserver.func_184144_a(this.field_147369_b, this.field_147369_b.func_174813_aQ().func_186664_h(0.0625D)).isEmpty();
                             d7 = d4 - this.field_184352_o;
                             d8 = d5 - this.field_184353_p;
                             d9 = d6 - this.field_184354_q;
 
-                            if (this.field_147369_b.field_70122_E && !p_147347_1_.func_149465_i() && d8 > 0.0)
+                            if (this.field_147369_b.field_70122_E && !p_147347_1_.func_149465_i() && d8 > 0.0D)
                             {
                                 this.field_147369_b.func_70664_aZ();
                             }
 
                             this.field_147369_b.func_70091_d(MoverType.PLAYER, d7, d8, d9);
                             this.field_147369_b.field_70122_E = p_147347_1_.func_149465_i();
+                            double d12 = d8;
                             d7 = d4 - this.field_147369_b.field_70165_t;
                             d8 = d5 - this.field_147369_b.field_70163_u;
 
-                            if (d8 > -0.5 || d8 < 0.5)
+                            if (d8 > -0.5D || d8 < 0.5D)
                             {
-                                d8 = 0.0;
+                                d8 = 0.0D;
                             }
 
                             d9 = d6 - this.field_147369_b.field_70161_v;
                             d11 = d7 * d7 + d8 * d8 + d9 * d9;
                             boolean flag = false;
 
-                            if (!this.field_147369_b.func_184850_K()
-                                    && d11 > 0.0625
-                                    && !this.field_147369_b.func_70608_bn()
-                                    && !this.field_147369_b.field_71134_c.func_73083_d()
-                                    && this.field_147369_b.field_71134_c.func_73081_b() != GameType.SPECTATOR)
+                            if (!this.field_147369_b.func_184850_K() && d11 > 0.0625D && !this.field_147369_b.func_70608_bn() && !this.field_147369_b.field_71134_c.func_73083_d() && this.field_147369_b.field_71134_c.func_73081_b() != GameType.SPECTATOR)
                             {
                                 flag = true;
-                                field_147370_c.warn("{} moved wrongly!", this.field_147369_b.func_70005_c_());
+                                field_147370_c.warn("{} moved wrongly!", (Object)this.field_147369_b.func_70005_c_());
                             }
 
                             this.field_147369_b.func_70080_a(d4, d5, d6, f, f1);
-                            this.field_147369_b
-                            .func_71000_j(this.field_147369_b.field_70165_t - d0, this.field_147369_b.field_70163_u - d1, this.field_147369_b.field_70161_v - d2);
+                            this.field_147369_b.func_71000_j(this.field_147369_b.field_70165_t - d0, this.field_147369_b.field_70163_u - d1, this.field_147369_b.field_70161_v - d2);
 
                             if (!this.field_147369_b.field_70145_X && !this.field_147369_b.func_70608_bn())
                             {
-                                boolean flag1 = worldserver.func_184144_a(this.field_147369_b, this.field_147369_b.func_174813_aQ().func_186664_h(0.0625)).isEmpty();
+                                boolean flag1 = worldserver.func_184144_a(this.field_147369_b, this.field_147369_b.func_174813_aQ().func_186664_h(0.0625D)).isEmpty();
 
                                 if (flag2 && (flag || !flag1))
                                 {
@@ -637,14 +577,9 @@
                                 }
                             }
 
-                            this.field_184344_B = d8 >= -0.03125;
-                            this.field_184344_B = this.field_184344_B & (!this.field_147367_d.func_71231_X() && !this.field_147369_b.field_71075_bZ.field_75101_c);
-                            this.field_184344_B = this.field_184344_B
-                                                  & (
-                                                      !this.field_147369_b.func_70644_a(MobEffects.field_188424_y)
-                                                      && !this.field_147369_b.func_184613_cA()
-                                                      && !worldserver.func_72829_c(this.field_147369_b.func_174813_aQ().func_186662_g(0.0625).func_72321_a(0.0, -0.55, 0.0))
-                                                  );
+                            this.field_184344_B = d12 >= -0.03125D;
+                            this.field_184344_B &= !this.field_147367_d.func_71231_X() && !this.field_147369_b.field_71075_bZ.field_75101_c;
+                            this.field_184344_B &= !this.field_147369_b.func_70644_a(MobEffects.field_188424_y) && !this.field_147369_b.func_184613_cA() && !worldserver.func_72829_c(this.field_147369_b.func_174813_aQ().func_186662_g(0.0625D).func_72321_a(0.0D, -0.55D, 0.0D));
                             this.field_147369_b.field_70122_E = p_147347_1_.func_149465_i();
                             this.field_147367_d.func_184103_al().func_72358_d(this.field_147369_b);
                             this.field_147369_b.func_71122_b(this.field_147369_b.field_70163_u - d3, p_147347_1_.func_149465_i());
@@ -663,13 +598,11 @@
         this.func_175089_a(p_147364_1_, p_147364_3_, p_147364_5_, p_147364_7_, p_147364_8_, Collections.emptySet());
     }
 
-    public void func_175089_a(
-        double p_175089_1_, double p_175089_3_, double p_175089_5_, float p_175089_7_, float p_175089_8_, Set<SPacketPlayerPosLook.EnumFlags> p_175089_9_
-    )
+    public void func_175089_a(double p_175089_1_, double p_175089_3_, double p_175089_5_, float p_175089_7_, float p_175089_8_, Set<SPacketPlayerPosLook.EnumFlags> p_175089_9_)
     {
-        double d0 = p_175089_9_.contains(SPacketPlayerPosLook.EnumFlags.X) ? this.field_147369_b.field_70165_t : 0.0;
-        double d1 = p_175089_9_.contains(SPacketPlayerPosLook.EnumFlags.Y) ? this.field_147369_b.field_70163_u : 0.0;
-        double d2 = p_175089_9_.contains(SPacketPlayerPosLook.EnumFlags.Z) ? this.field_147369_b.field_70161_v : 0.0;
+        double d0 = p_175089_9_.contains(SPacketPlayerPosLook.EnumFlags.X) ? this.field_147369_b.field_70165_t : 0.0D;
+        double d1 = p_175089_9_.contains(SPacketPlayerPosLook.EnumFlags.Y) ? this.field_147369_b.field_70163_u : 0.0D;
+        double d2 = p_175089_9_.contains(SPacketPlayerPosLook.EnumFlags.Z) ? this.field_147369_b.field_70161_v : 0.0D;
         this.field_184362_y = new Vec3d(p_175089_1_ + d0, p_175089_3_ + d1, p_175089_5_ + d2);
         float f = p_175089_7_;
         float f1 = p_175089_8_;
@@ -691,12 +624,9 @@
 
         this.field_184343_A = this.field_147368_e;
         this.field_147369_b.func_70080_a(this.field_184362_y.field_72450_a, this.field_184362_y.field_72448_b, this.field_184362_y.field_72449_c, f, f1);
-        this.field_147369_b
-        .field_71135_a
-        .func_147359_a(new SPacketPlayerPosLook(p_175089_1_, p_175089_3_, p_175089_5_, p_175089_7_, p_175089_8_, p_175089_9_, this.field_184363_z));
+        this.field_147369_b.field_71135_a.func_147359_a(new SPacketPlayerPosLook(p_175089_1_, p_175089_3_, p_175089_5_, p_175089_7_, p_175089_8_, p_175089_9_, this.field_184363_z));
     }
 
-    @Override
     public void func_147345_a(CPacketPlayerDigging p_147345_1_)
     {
         PacketThreadUtil.func_180031_a(p_147345_1_, this, this.field_147369_b.func_71121_q());
@@ -738,12 +668,15 @@
             case START_DESTROY_BLOCK:
             case ABORT_DESTROY_BLOCK:
             case STOP_DESTROY_BLOCK:
-                double d0 = this.field_147369_b.field_70165_t - ((double)blockpos.func_177958_n() + 0.5);
-                double d1 = this.field_147369_b.field_70163_u - ((double)blockpos.func_177956_o() + 0.5) + 1.5;
-                double d2 = this.field_147369_b.field_70161_v - ((double)blockpos.func_177952_p() + 0.5);
+                double d0 = this.field_147369_b.field_70165_t - ((double)blockpos.func_177958_n() + 0.5D);
+                double d1 = this.field_147369_b.field_70163_u - ((double)blockpos.func_177956_o() + 0.5D) + 1.5D;
+                double d2 = this.field_147369_b.field_70161_v - ((double)blockpos.func_177952_p() + 0.5D);
                 double d3 = d0 * d0 + d1 * d1 + d2 * d2;
 
-                if (d3 > 36.0)
+                double dist = field_147369_b.func_110148_a(EntityPlayer.REACH_DISTANCE).func_111126_e() + 1;
+                dist *= dist;
+
+                if (d3 > dist)
                 {
                     return;
                 }
@@ -789,7 +722,6 @@
         }
     }
 
-    @Override
     public void func_184337_a(CPacketPlayerTryUseItemOnBlock p_184337_1_)
     {
         PacketThreadUtil.func_180031_a(p_184337_1_, this, this.field_147369_b.func_71121_q());
@@ -800,34 +732,18 @@
         EnumFacing enumfacing = p_184337_1_.func_187024_b();
         this.field_147369_b.func_143004_u();
 
-        if (blockpos.func_177956_o() < this.field_147367_d.func_71207_Z() - 1
-                || enumfacing != EnumFacing.UP && blockpos.func_177956_o() < this.field_147367_d.func_71207_Z())
+        if (blockpos.func_177956_o() < this.field_147367_d.func_71207_Z() - 1 || enumfacing != EnumFacing.UP && blockpos.func_177956_o() < this.field_147367_d.func_71207_Z())
         {
-            if (this.field_184362_y == null
-                    && this.field_147369_b
-                    .func_70092_e((double)blockpos.func_177958_n() + 0.5, (double)blockpos.func_177956_o() + 0.5, (double)blockpos.func_177952_p() + 0.5)
-                    < 64.0
-                    && !this.field_147367_d.func_175579_a(worldserver, blockpos, this.field_147369_b)
-                    && worldserver.func_175723_af().func_177746_a(blockpos))
+            double dist = field_147369_b.func_110148_a(EntityPlayer.REACH_DISTANCE).func_111126_e() + 3;
+            dist *= dist;
+            if (this.field_184362_y == null && this.field_147369_b.func_70092_e((double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 0.5D, (double)blockpos.func_177952_p() + 0.5D) < dist && !this.field_147367_d.func_175579_a(worldserver, blockpos, this.field_147369_b) && worldserver.func_175723_af().func_177746_a(blockpos))
             {
-                this.field_147369_b
-                .field_71134_c
-                .func_187251_a(
-                    this.field_147369_b,
-                    worldserver,
-                    itemstack,
-                    enumhand,
-                    blockpos,
-                    enumfacing,
-                    p_184337_1_.func_187026_d(),
-                    p_184337_1_.func_187025_e(),
-                    p_184337_1_.func_187020_f()
-                );
+                this.field_147369_b.field_71134_c.func_187251_a(this.field_147369_b, worldserver, itemstack, enumhand, blockpos, enumfacing, p_184337_1_.func_187026_d(), p_184337_1_.func_187025_e(), p_184337_1_.func_187020_f());
             }
         }
         else
         {
-            TextComponentTranslation textcomponenttranslation = new TextComponentTranslation("build.tooHigh", this.field_147367_d.func_71207_Z());
+            TextComponentTranslation textcomponenttranslation = new TextComponentTranslation("build.tooHigh", new Object[] {this.field_147367_d.func_71207_Z()});
             textcomponenttranslation.func_150256_b().func_150238_a(TextFormatting.RED);
             this.field_147369_b.field_71135_a.func_147359_a(new SPacketChat(textcomponenttranslation, ChatType.GAME_INFO));
         }
@@ -836,7 +752,6 @@
         this.field_147369_b.field_71135_a.func_147359_a(new SPacketBlockChange(worldserver, blockpos.func_177972_a(enumfacing)));
     }
 
-    @Override
     public void func_147346_a(CPacketPlayerTryUseItem p_147346_1_)
     {
         PacketThreadUtil.func_180031_a(p_147346_1_, this, this.field_147369_b.func_71121_q());
@@ -851,7 +766,6 @@
         }
     }
 
-    @Override
     public void func_175088_a(CPacketSpectate p_175088_1_)
     {
         PacketThreadUtil.func_180031_a(p_175088_1_, this, this.field_147369_b.func_71121_q());
@@ -882,19 +796,13 @@
                 {
                     this.field_147369_b.func_70634_a(entity.field_70165_t, entity.field_70163_u, entity.field_70161_v);
                 }
-                else
+                else if (net.minecraftforge.common.ForgeHooks.onTravelToDimension(this.field_147369_b, entity.field_71093_bK))
                 {
+                    int prevDimension = this.field_147369_b.field_71093_bK;
                     WorldServer worldserver1 = this.field_147369_b.func_71121_q();
                     WorldServer worldserver2 = (WorldServer)entity.field_70170_p;
                     this.field_147369_b.field_71093_bK = entity.field_71093_bK;
-                    this.func_147359_a(
-                        new SPacketRespawn(
-                            this.field_147369_b.field_71093_bK,
-                            worldserver1.func_175659_aa(),
-                            worldserver1.func_72912_H().func_76067_t(),
-                            this.field_147369_b.field_71134_c.func_73081_b()
-                        )
-                    );
+                    this.func_147359_a(new SPacketRespawn(this.field_147369_b.field_71093_bK, worldserver2.func_175659_aa(), worldserver2.func_72912_H().func_76067_t(), this.field_147369_b.field_71134_c.func_73081_b())); // Forge: Use new dimensions information
                     this.field_147367_d.func_184103_al().func_187243_f(this.field_147369_b);
                     worldserver1.func_72973_f(this.field_147369_b);
                     this.field_147369_b.field_70128_L = false;
@@ -913,17 +821,16 @@
                     this.field_147369_b.field_71134_c.func_73080_a(worldserver2);
                     this.field_147367_d.func_184103_al().func_72354_b(this.field_147369_b, worldserver2);
                     this.field_147367_d.func_184103_al().func_72385_f(this.field_147369_b);
+                    net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerChangedDimensionEvent(this.field_147369_b, prevDimension, this.field_147369_b.field_71093_bK);
                 }
             }
         }
     }
 
-    @Override
     public void func_175086_a(CPacketResourcePackStatus p_175086_1_)
     {
     }
 
-    @Override
     public void func_184340_a(CPacketSteerBoat p_184340_1_)
     {
         PacketThreadUtil.func_180031_a(p_184340_1_, this, this.field_147369_b.func_71121_q());
@@ -935,12 +842,11 @@
         }
     }
 
-    @Override
     public void func_147231_a(ITextComponent p_147231_1_)
     {
         field_147370_c.info("{} lost connection: {}", this.field_147369_b.func_70005_c_(), p_147231_1_.func_150260_c());
         this.field_147367_d.func_147132_au();
-        TextComponentTranslation textcomponenttranslation = new TextComponentTranslation("multiplayer.player.left", this.field_147369_b.func_145748_c_());
+        TextComponentTranslation textcomponenttranslation = new TextComponentTranslation("multiplayer.player.left", new Object[] {this.field_147369_b.func_145748_c_()});
         textcomponenttranslation.func_150256_b().func_150238_a(TextFormatting.YELLOW);
         this.field_147367_d.func_184103_al().func_148539_a(textcomponenttranslation);
         this.field_147369_b.func_71123_m();
@@ -990,7 +896,6 @@
         }
     }
 
-    @Override
     public void func_147355_a(CPacketHeldItemChange p_147355_1_)
     {
         PacketThreadUtil.func_180031_a(p_147355_1_, this, this.field_147369_b.func_71121_q());
@@ -1002,18 +907,17 @@
         }
         else
         {
-            field_147370_c.warn("{} tried to set an invalid carried item", this.field_147369_b.func_70005_c_());
+            field_147370_c.warn("{} tried to set an invalid carried item", (Object)this.field_147369_b.func_70005_c_());
         }
     }
 
-    @Override
     public void func_147354_a(CPacketChatMessage p_147354_1_)
     {
         PacketThreadUtil.func_180031_a(p_147354_1_, this, this.field_147369_b.func_71121_q());
 
         if (this.field_147369_b.func_147096_v() == EntityPlayer.EnumChatVisibility.HIDDEN)
         {
-            TextComponentTranslation textcomponenttranslation = new TextComponentTranslation("chat.cannotSend");
+            TextComponentTranslation textcomponenttranslation = new TextComponentTranslation("chat.cannotSend", new Object[0]);
             textcomponenttranslation.func_150256_b().func_150238_a(TextFormatting.RED);
             this.func_147359_a(new SPacketChat(textcomponenttranslation));
         }
@@ -1023,11 +927,11 @@
             String s = p_147354_1_.func_149439_c();
             s = StringUtils.normalizeSpace(s);
 
-            for (int i = 0; i < s.length(); i++)
+            for (int i = 0; i < s.length(); ++i)
             {
                 if (!ChatAllowedCharacters.func_71566_a(s.charAt(i)))
                 {
-                    this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.illegal_characters"));
+                    this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.illegal_characters", new Object[0]));
                     return;
                 }
             }
@@ -1038,7 +942,9 @@
             }
             else
             {
-                ITextComponent itextcomponent = new TextComponentTranslation("chat.type.text", this.field_147369_b.func_145748_c_(), s);
+                ITextComponent itextcomponent = new TextComponentTranslation("chat.type.text", this.field_147369_b.func_145748_c_(), net.minecraftforge.common.ForgeHooks.newChatWithLinks(s));
+                itextcomponent = net.minecraftforge.common.ForgeHooks.onServerChatEvent(this, s, itextcomponent);
+                if (itextcomponent == null) return;
                 this.field_147367_d.func_184103_al().func_148544_a(itextcomponent, false);
             }
 
@@ -1046,7 +952,7 @@
 
             if (this.field_147374_l > 200 && !this.field_147367_d.func_184103_al().func_152596_g(this.field_147369_b.func_146103_bH()))
             {
-                this.func_194028_b(new TextComponentTranslation("disconnect.spam"));
+                this.func_194028_b(new TextComponentTranslation("disconnect.spam", new Object[0]));
             }
         }
     }
@@ -1056,7 +962,6 @@
         this.field_147367_d.func_71187_D().func_71556_a(this.field_147369_b, p_147361_1_);
     }
 
-    @Override
     public void func_175087_a(CPacketAnimation p_175087_1_)
     {
         PacketThreadUtil.func_180031_a(p_175087_1_, this, this.field_147369_b.func_71121_q());
@@ -1064,7 +969,6 @@
         this.field_147369_b.func_184609_a(p_175087_1_.func_187018_a());
     }
 
-    @Override
     public void func_147357_a(CPacketEntityAction p_147357_1_)
     {
         PacketThreadUtil.func_180031_a(p_147357_1_, this, this.field_147369_b.func_71121_q());
@@ -1126,10 +1030,7 @@
                 break;
             case START_FALL_FLYING:
 
-                if (!this.field_147369_b.field_70122_E
-                        && this.field_147369_b.field_70181_x < 0.0
-                        && !this.field_147369_b.func_184613_cA()
-                        && !this.field_147369_b.func_70090_H())
+                if (!this.field_147369_b.field_70122_E && this.field_147369_b.field_70181_x < 0.0D && !this.field_147369_b.func_184613_cA() && !this.field_147369_b.func_70090_H())
                 {
                     ItemStack itemstack = this.field_147369_b.func_184582_a(EntityEquipmentSlot.CHEST);
 
@@ -1149,7 +1050,6 @@
         }
     }
 
-    @Override
     public void func_147340_a(CPacketUseEntity p_147340_1_)
     {
         PacketThreadUtil.func_180031_a(p_147340_1_, this, this.field_147369_b.func_71121_q());
@@ -1160,11 +1060,11 @@
         if (entity != null)
         {
             boolean flag = this.field_147369_b.func_70685_l(entity);
-            double d0 = 36.0;
+            double d0 = 36.0D;
 
             if (!flag)
             {
-                d0 = 9.0;
+                d0 = 9.0D;
             }
 
             if (this.field_147369_b.func_70068_e(entity) < d0)
@@ -1177,13 +1077,14 @@
                 else if (p_147340_1_.func_149565_c() == CPacketUseEntity.Action.INTERACT_AT)
                 {
                     EnumHand enumhand1 = p_147340_1_.func_186994_b();
+                    if(net.minecraftforge.common.ForgeHooks.onInteractEntityAt(field_147369_b, entity, p_147340_1_.func_179712_b(), enumhand1) != null) return;
                     entity.func_184199_a(this.field_147369_b, p_147340_1_.func_179712_b(), enumhand1);
                 }
                 else if (p_147340_1_.func_149565_c() == CPacketUseEntity.Action.ATTACK)
                 {
                     if (entity instanceof EntityItem || entity instanceof EntityXPOrb || entity instanceof EntityArrow || entity == this.field_147369_b)
                     {
-                        this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.invalid_entity_attacked"));
+                        this.func_194028_b(new TextComponentTranslation("multiplayer.disconnect.invalid_entity_attacked", new Object[0]));
                         this.field_147367_d.func_71236_h("Player " + this.field_147369_b.func_70005_c_() + " tried to attack an invalid entity");
                         return;
                     }
@@ -1194,7 +1095,6 @@
         }
     }
 
-    @Override
     public void func_147342_a(CPacketClientStatus p_147342_1_)
     {
         PacketThreadUtil.func_180031_a(p_147342_1_, this, this.field_147369_b.func_71121_q());
@@ -1218,7 +1118,7 @@
                         return;
                     }
 
-                    this.field_147369_b = this.field_147367_d.func_184103_al().func_72368_a(this.field_147369_b, 0, false);
+                    this.field_147369_b = this.field_147367_d.func_184103_al().func_72368_a(this.field_147369_b, field_147369_b.field_71093_bK, false);
 
                     if (this.field_147367_d.func_71199_h())
                     {
@@ -1233,40 +1133,35 @@
         }
     }
 
-    @Override
     public void func_147356_a(CPacketCloseWindow p_147356_1_)
     {
         PacketThreadUtil.func_180031_a(p_147356_1_, this, this.field_147369_b.func_71121_q());
         this.field_147369_b.func_71128_l();
     }
 
-    @Override
     public void func_147351_a(CPacketClickWindow p_147351_1_)
     {
         PacketThreadUtil.func_180031_a(p_147351_1_, this, this.field_147369_b.func_71121_q());
         this.field_147369_b.func_143004_u();
 
-        if (this.field_147369_b.field_71070_bA.field_75152_c == p_147351_1_.func_149548_c()
-                && this.field_147369_b.field_71070_bA.func_75129_b(this.field_147369_b))
+        if (this.field_147369_b.field_71070_bA.field_75152_c == p_147351_1_.func_149548_c() && this.field_147369_b.field_71070_bA.func_75129_b(this.field_147369_b))
         {
             if (this.field_147369_b.func_175149_v())
             {
-                NonNullList<ItemStack> nonnulllist = NonNullList.func_191196_a();
+                NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191196_a();
 
-                for (int i = 0; i < this.field_147369_b.field_71070_bA.field_75151_b.size(); i++)
+                for (int i = 0; i < this.field_147369_b.field_71070_bA.field_75151_b.size(); ++i)
                 {
-                    nonnulllist.add(this.field_147369_b.field_71070_bA.field_75151_b.get(i).func_75211_c());
+                    nonnulllist.add(((Slot)this.field_147369_b.field_71070_bA.field_75151_b.get(i)).func_75211_c());
                 }
 
                 this.field_147369_b.func_71110_a(this.field_147369_b.field_71070_bA, nonnulllist);
             }
             else
             {
-                ItemStack itemstack2 = this.field_147369_b
-                                       .field_71070_bA
-                                       .func_184996_a(p_147351_1_.func_149544_d(), p_147351_1_.func_149543_e(), p_147351_1_.func_186993_f(), this.field_147369_b);
+                ItemStack itemstack2 = this.field_147369_b.field_71070_bA.func_184996_a(p_147351_1_.func_149544_d(), p_147351_1_.func_149543_e(), p_147351_1_.func_186993_f(), this.field_147369_b);
 
-                if (ItemStack.func_77989_b(p_147351_1_.func_149546_g(), itemstack2))
+                if (ItemStack.areItemStacksEqualUsingNBTShareTag(p_147351_1_.func_149546_g(), itemstack2))
                 {
                     this.field_147369_b.field_71135_a.func_147359_a(new SPacketConfirmTransaction(p_147351_1_.func_149548_c(), p_147351_1_.func_149547_f(), true));
                     this.field_147369_b.field_71137_h = true;
@@ -1276,14 +1171,14 @@
                 }
                 else
                 {
-                    this.field_147372_n.func_76038_a(this.field_147369_b.field_71070_bA.field_75152_c, p_147351_1_.func_149547_f());
+                    this.field_147372_n.func_76038_a(this.field_147369_b.field_71070_bA.field_75152_c, Short.valueOf(p_147351_1_.func_149547_f()));
                     this.field_147369_b.field_71135_a.func_147359_a(new SPacketConfirmTransaction(p_147351_1_.func_149548_c(), p_147351_1_.func_149547_f(), false));
                     this.field_147369_b.field_71070_bA.func_75128_a(this.field_147369_b, false);
-                    NonNullList<ItemStack> nonnulllist1 = NonNullList.func_191196_a();
+                    NonNullList<ItemStack> nonnulllist1 = NonNullList.<ItemStack>func_191196_a();
 
-                    for (int j = 0; j < this.field_147369_b.field_71070_bA.field_75151_b.size(); j++)
+                    for (int j = 0; j < this.field_147369_b.field_71070_bA.field_75151_b.size(); ++j)
                     {
-                        ItemStack itemstack = this.field_147369_b.field_71070_bA.field_75151_b.get(j).func_75211_c();
+                        ItemStack itemstack = ((Slot)this.field_147369_b.field_71070_bA.field_75151_b.get(j)).func_75211_c();
                         ItemStack itemstack1 = itemstack.func_190926_b() ? ItemStack.field_190927_a : itemstack;
                         nonnulllist1.add(itemstack1);
                     }
@@ -1294,36 +1189,29 @@
         }
     }
 
-    @Override
     public void func_194308_a(CPacketPlaceRecipe p_194308_1_)
     {
         PacketThreadUtil.func_180031_a(p_194308_1_, this, this.field_147369_b.func_71121_q());
         this.field_147369_b.func_143004_u();
 
-        if (!this.field_147369_b.func_175149_v()
-                && this.field_147369_b.field_71070_bA.field_75152_c == p_194308_1_.func_194318_a()
-                && this.field_147369_b.field_71070_bA.func_75129_b(this.field_147369_b))
+        if (!this.field_147369_b.func_175149_v() && this.field_147369_b.field_71070_bA.field_75152_c == p_194308_1_.func_194318_a() && this.field_147369_b.field_71070_bA.func_75129_b(this.field_147369_b))
         {
             this.field_194309_H.func_194327_a(this.field_147369_b, p_194308_1_.func_194317_b(), p_194308_1_.func_194319_c());
         }
     }
 
-    @Override
     public void func_147338_a(CPacketEnchantItem p_147338_1_)
     {
         PacketThreadUtil.func_180031_a(p_147338_1_, this, this.field_147369_b.func_71121_q());
         this.field_147369_b.func_143004_u();
 
-        if (this.field_147369_b.field_71070_bA.field_75152_c == p_147338_1_.func_149539_c()
-                && this.field_147369_b.field_71070_bA.func_75129_b(this.field_147369_b)
-                && !this.field_147369_b.func_175149_v())
+        if (this.field_147369_b.field_71070_bA.field_75152_c == p_147338_1_.func_149539_c() && this.field_147369_b.field_71070_bA.func_75129_b(this.field_147369_b) && !this.field_147369_b.func_175149_v())
         {
             this.field_147369_b.field_71070_bA.func_75140_a(this.field_147369_b, p_147338_1_.func_149537_d());
             this.field_147369_b.field_71070_bA.func_75142_b();
         }
     }
 
-    @Override
     public void func_147344_a(CPacketCreativeInventoryAction p_147344_1_)
     {
         PacketThreadUtil.func_180031_a(p_147344_1_, this, this.field_147369_b.func_71121_q());
@@ -1382,23 +1270,17 @@
         }
     }
 
-    @Override
     public void func_147339_a(CPacketConfirmTransaction p_147339_1_)
     {
         PacketThreadUtil.func_180031_a(p_147339_1_, this, this.field_147369_b.func_71121_q());
         Short oshort = this.field_147372_n.func_76041_a(this.field_147369_b.field_71070_bA.field_75152_c);
 
-        if (oshort != null
-                && p_147339_1_.func_149533_d() == oshort
-                && this.field_147369_b.field_71070_bA.field_75152_c == p_147339_1_.func_149532_c()
-                && !this.field_147369_b.field_71070_bA.func_75129_b(this.field_147369_b)
-                && !this.field_147369_b.func_175149_v())
+        if (oshort != null && p_147339_1_.func_149533_d() == oshort.shortValue() && this.field_147369_b.field_71070_bA.field_75152_c == p_147339_1_.func_149532_c() && !this.field_147369_b.field_71070_bA.func_75129_b(this.field_147369_b) && !this.field_147369_b.func_175149_v())
         {
             this.field_147369_b.field_71070_bA.func_75128_a(this.field_147369_b, true);
         }
     }
 
-    @Override
     public void func_147343_a(CPacketUpdateSign p_147343_1_)
     {
         PacketThreadUtil.func_180031_a(p_147343_1_, this, this.field_147369_b.func_71121_q());
@@ -1426,7 +1308,7 @@
 
             String[] astring = p_147343_1_.func_187017_b();
 
-            for (int i = 0; i < astring.length; i++)
+            for (int i = 0; i < astring.length; ++i)
             {
                 tileentitysign.field_145915_a[i] = new TextComponentString(TextFormatting.func_110646_a(astring[i]));
             }
@@ -1436,7 +1318,6 @@
         }
     }
 
-    @Override
     public void func_147353_a(CPacketKeepAlive p_147353_1_)
     {
         if (this.field_194403_g && p_147353_1_.func_149460_c() == this.field_194404_h)
@@ -1447,7 +1328,7 @@
         }
         else if (!this.field_147369_b.func_70005_c_().equals(this.field_147367_d.func_71214_G()))
         {
-            this.func_194028_b(new TextComponentTranslation("disconnect.timeout"));
+            this.func_194028_b(new TextComponentTranslation("disconnect.timeout", new Object[0]));
         }
     }
 
@@ -1456,36 +1337,31 @@
         return System.nanoTime() / 1000000L;
     }
 
-    @Override
     public void func_147348_a(CPacketPlayerAbilities p_147348_1_)
     {
         PacketThreadUtil.func_180031_a(p_147348_1_, this, this.field_147369_b.func_71121_q());
         this.field_147369_b.field_71075_bZ.field_75100_b = p_147348_1_.func_149488_d() && this.field_147369_b.field_71075_bZ.field_75101_c;
     }
 
-    @Override
     public void func_147341_a(CPacketTabComplete p_147341_1_)
     {
         PacketThreadUtil.func_180031_a(p_147341_1_, this, this.field_147369_b.func_71121_q());
-        List<String> list = Lists.newArrayList();
+        List<String> list = Lists.<String>newArrayList();
 
-        for (String s : this.field_147367_d
-                .func_184104_a(this.field_147369_b, p_147341_1_.func_149419_c(), p_147341_1_.func_179709_b(), p_147341_1_.func_186989_c()))
+        for (String s : this.field_147367_d.func_184104_a(this.field_147369_b, p_147341_1_.func_149419_c(), p_147341_1_.func_179709_b(), p_147341_1_.func_186989_c()))
         {
             list.add(s);
         }
 
-        this.field_147369_b.field_71135_a.func_147359_a(new SPacketTabComplete(list.toArray(new String[list.size()])));
+        this.field_147369_b.field_71135_a.func_147359_a(new SPacketTabComplete((String[])list.toArray(new String[list.size()])));
     }
 
-    @Override
     public void func_147352_a(CPacketClientSettings p_147352_1_)
     {
         PacketThreadUtil.func_180031_a(p_147352_1_, this, this.field_147369_b.func_71121_q());
         this.field_147369_b.func_147100_a(p_147352_1_);
     }
 
-    @Override
     public void func_147349_a(CPacketCustomPayload p_147349_1_)
     {
         PacketThreadUtil.func_180031_a(p_147349_1_, this, this.field_147369_b.func_71121_q());
@@ -1558,7 +1434,7 @@
                     itemstack2.func_77983_a("title", new NBTTagString(itemstack3.func_77978_p().func_74779_i("title")));
                     NBTTagList nbttaglist = itemstack3.func_77978_p().func_150295_c("pages", 8);
 
-                    for (int i = 0; i < nbttaglist.func_74745_c(); i++)
+                    for (int i = 0; i < nbttaglist.func_74745_c(); ++i)
                     {
                         String s1 = nbttaglist.func_150307_f(i);
                         ITextComponent itextcomponent = new TextComponentString(s1);
@@ -1596,13 +1472,13 @@
         {
             if (!this.field_147367_d.func_82356_Z())
             {
-                this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.notEnabled"));
+                this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.notEnabled", new Object[0]));
                 return;
             }
 
             if (!this.field_147369_b.func_189808_dh())
             {
-                this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.notAllowed"));
+                this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.notAllowed", new Object[0]));
                 return;
             }
 
@@ -1615,9 +1491,7 @@
 
                 if (l == 0)
                 {
-                    TileEntity tileentity = this.field_147369_b
-                                            .field_70170_p
-                                            .func_175625_s(new BlockPos(packetbuffer2.readInt(), packetbuffer2.readInt(), packetbuffer2.readInt()));
+                    TileEntity tileentity = this.field_147369_b.field_70170_p.func_175625_s(new BlockPos(packetbuffer2.readInt(), packetbuffer2.readInt(), packetbuffer2.readInt()));
 
                     if (tileentity instanceof TileEntityCommandBlock)
                     {
@@ -1644,11 +1518,11 @@
 
                     if (!flag2)
                     {
-                        commandblockbaselogic1.func_145750_b(null);
+                        commandblockbaselogic1.func_145750_b((ITextComponent)null);
                     }
 
                     commandblockbaselogic1.func_145756_e();
-                    this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.setCommand.success", s6));
+                    this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.setCommand.success", new Object[] {s6}));
                 }
             }
             catch (Exception exception4)
@@ -1660,13 +1534,13 @@
         {
             if (!this.field_147367_d.func_82356_Z())
             {
-                this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.notEnabled"));
+                this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.notEnabled", new Object[0]));
                 return;
             }
 
             if (!this.field_147369_b.func_189808_dh())
             {
-                this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.notAllowed"));
+                this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.notAllowed", new Object[0]));
                 return;
             }
 
@@ -1693,39 +1567,21 @@
 
                 if (commandblockbaselogic != null)
                 {
-                    EnumFacing enumfacing = this.field_147369_b.field_70170_p.func_180495_p(blockpos1).func_177229_b(BlockCommandBlock.field_185564_a);
+                    EnumFacing enumfacing = (EnumFacing)this.field_147369_b.field_70170_p.func_180495_p(blockpos1).func_177229_b(BlockCommandBlock.field_185564_a);
 
                     switch (tileentitycommandblock$mode)
                     {
                         case SEQUENCE:
                             IBlockState iblockstate3 = Blocks.field_185777_dd.func_176223_P();
-                            this.field_147369_b
-                            .field_70170_p
-                            .func_180501_a(
-                                blockpos1,
-                                iblockstate3.func_177226_a(BlockCommandBlock.field_185564_a, enumfacing).func_177226_a(BlockCommandBlock.field_185565_b, flag),
-                                2
-                            );
+                            this.field_147369_b.field_70170_p.func_180501_a(blockpos1, iblockstate3.func_177226_a(BlockCommandBlock.field_185564_a, enumfacing).func_177226_a(BlockCommandBlock.field_185565_b, Boolean.valueOf(flag)), 2);
                             break;
                         case AUTO:
                             IBlockState iblockstate2 = Blocks.field_185776_dc.func_176223_P();
-                            this.field_147369_b
-                            .field_70170_p
-                            .func_180501_a(
-                                blockpos1,
-                                iblockstate2.func_177226_a(BlockCommandBlock.field_185564_a, enumfacing).func_177226_a(BlockCommandBlock.field_185565_b, flag),
-                                2
-                            );
+                            this.field_147369_b.field_70170_p.func_180501_a(blockpos1, iblockstate2.func_177226_a(BlockCommandBlock.field_185564_a, enumfacing).func_177226_a(BlockCommandBlock.field_185565_b, Boolean.valueOf(flag)), 2);
                             break;
                         case REDSTONE:
                             IBlockState iblockstate = Blocks.field_150483_bI.func_176223_P();
-                            this.field_147369_b
-                            .field_70170_p
-                            .func_180501_a(
-                                blockpos1,
-                                iblockstate.func_177226_a(BlockCommandBlock.field_185564_a, enumfacing).func_177226_a(BlockCommandBlock.field_185565_b, flag),
-                                2
-                            );
+                            this.field_147369_b.field_70170_p.func_180501_a(blockpos1, iblockstate.func_177226_a(BlockCommandBlock.field_185564_a, enumfacing).func_177226_a(BlockCommandBlock.field_185565_b, Boolean.valueOf(flag)), 2);
                     }
 
                     tileentity2.func_145829_t();
@@ -1735,7 +1591,7 @@
 
                     if (!flag3)
                     {
-                        commandblockbaselogic.func_145750_b(null);
+                        commandblockbaselogic.func_145750_b((ITextComponent)null);
                     }
 
                     tileentitycommandblock.func_184253_b(flag1);
@@ -1743,7 +1599,7 @@
 
                     if (!net.minecraft.util.StringUtils.func_151246_b(s7))
                     {
-                        this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.setCommand.success", s7));
+                        this.field_147369_b.func_145747_a(new TextComponentTranslation("advMode.setCommand.success", new Object[] {s7}));
                     }
                 }
             }
@@ -1846,37 +1702,37 @@
                     {
                         if (tileentitystructure.func_184419_m())
                         {
-                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.save_success", s4), false);
+                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.save_success", new Object[] {s4}), false);
                         }
                         else
                         {
-                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.save_failure", s4), false);
+                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.save_failure", new Object[] {s4}), false);
                         }
                     }
                     else if (l1 == 3)
                     {
                         if (!tileentitystructure.func_189709_F())
                         {
-                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.load_not_found", s4), false);
+                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.load_not_found", new Object[] {s4}), false);
                         }
                         else if (tileentitystructure.func_184412_n())
                         {
-                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.load_success", s4), false);
+                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.load_success", new Object[] {s4}), false);
                         }
                         else
                         {
-                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.load_prepare", s4), false);
+                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.load_prepare", new Object[] {s4}), false);
                         }
                     }
                     else if (l1 == 4)
                     {
                         if (tileentitystructure.func_184417_l())
                         {
-                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.size_success", s4), false);
+                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.size_success", new Object[] {s4}), false);
                         }
                         else
                         {
-                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.size_failure"), false);
+                            this.field_147369_b.func_146105_b(new TextComponentTranslation("structure_block.size_failure", new Object[0]), false);
                         }
                     }
 
@@ -1897,15 +1753,7 @@
             {
                 int j1 = packetbuffer6.func_150792_a();
                 this.field_147369_b.field_71071_by.func_184430_d(j1);
-                this.field_147369_b
-                .field_71135_a
-                .func_147359_a(
-                    new SPacketSetSlot(
-                        -2,
-                        this.field_147369_b.field_71071_by.field_70461_c,
-                        this.field_147369_b.field_71071_by.func_70301_a(this.field_147369_b.field_71071_by.field_70461_c)
-                    )
-                );
+                this.field_147369_b.field_71135_a.func_147359_a(new SPacketSetSlot(-2, this.field_147369_b.field_71071_by.field_70461_c, this.field_147369_b.field_71071_by.func_70301_a(this.field_147369_b.field_71071_by.field_70461_c)));
                 this.field_147369_b.field_71135_a.func_147359_a(new SPacketSetSlot(-2, j1, this.field_147369_b.field_71071_by.func_70301_a(j1)));
                 this.field_147369_b.field_71135_a.func_147359_a(new SPacketHeldItemChange(this.field_147369_b.field_71071_by.field_70461_c));
             }
