--- before/net/minecraft/network/NettyPacketEncoder.java
+++ after/net/minecraft/network/NettyPacketEncoder.java
@@ -22,7 +22,7 @@
 
     protected void encode(ChannelHandlerContext p_encode_1_, Packet<?> p_encode_2_, ByteBuf p_encode_3_) throws IOException, Exception
     {
-        EnumConnectionState enumconnectionstate = p_encode_1_.channel().attr(NetworkManager.field_150739_c).get();
+        EnumConnectionState enumconnectionstate = (EnumConnectionState)p_encode_1_.channel().attr(NetworkManager.field_150739_c).get();
 
         if (enumconnectionstate == null)
         {
@@ -44,7 +44,7 @@
             else
             {
                 PacketBuffer packetbuffer = new PacketBuffer(p_encode_3_);
-                packetbuffer.func_150787_b(integer);
+                packetbuffer.func_150787_b(integer.intValue());
 
                 try
                 {
@@ -52,7 +52,7 @@
                 }
                 catch (Throwable throwable)
                 {
-                    field_150798_a.error(throwable);
+                    throw throwable; // Forge: throw this instead of logging it, to prevent corrupt packets from being sent to the client where they are much harder to debug.
                 }
             }
         }
