--- before/net/minecraft/network/NetworkManager.java
+++ after/net/minecraft/network/NetworkManager.java
@@ -88,7 +88,12 @@
         this.field_179294_g = p_i46004_1_;
     }
 
+    public EnumPacketDirection getDirection()
+    {
+        return this.field_179294_g;
+    }
+
     @Override
     public void channelActive(ChannelHandlerContext p_channelActive_1_) throws Exception
     {
         super.channelActive(p_channelActive_1_);
@@ -208,9 +213,9 @@
     private void func_150732_b(final Packet<?> p_150732_1_, @Nullable final GenericFutureListener <? extends Future <? super Void >> [] p_150732_2_)
     {
         final EnumConnectionState enumconnectionstate = EnumConnectionState.func_150752_a(p_150732_1_);
         final EnumConnectionState enumconnectionstate1 = this.field_150746_k.attr(field_150739_c).get();
 
-        if (enumconnectionstate1 != enumconnectionstate)
+        if (enumconnectionstate1 != enumconnectionstate && !( p_150732_1_ instanceof net.minecraftforge.fml.common.network.internal.FMLProxyPacket))
         {
             field_150735_g.debug("Disabled auto read");
             this.field_150746_k.config().setAutoRead(false);
@@ -218,7 +223,7 @@
 
         if (this.field_150746_k.eventLoop().inEventLoop())
         {
-            if (enumconnectionstate != enumconnectionstate1)
+            if (enumconnectionstate != enumconnectionstate1 && !( p_150732_1_ instanceof net.minecraftforge.fml.common.network.internal.FMLProxyPacket))
             {
                 this.func_150723_a(enumconnectionstate);
             }
@@ -236,10 +241,10 @@
         {
             this.field_150746_k.eventLoop().execute(new Runnable()
             {
                 @Override
                 public void run()
                 {
-                    if (enumconnectionstate != enumconnectionstate1)
+                    if (enumconnectionstate != enumconnectionstate1  && !( p_150732_1_ instanceof net.minecraftforge.fml.common.network.internal.FMLProxyPacket))
                     {
                         NetworkManager.this.func_150723_a(enumconnectionstate);
                     }
@@ -317,6 +322,7 @@
     @SideOnly(Side.CLIENT)
     public static NetworkManager func_181124_a(InetAddress p_181124_0_, int p_181124_1_, boolean p_181124_2_)
     {
+        if (p_181124_0_ instanceof java.net.Inet6Address) System.setProperty("java.net.preferIPv4Stack", "false");
         final NetworkManager networkmanager = new NetworkManager(EnumPacketDirection.CLIENTBOUND);
         Class <? extends SocketChannel > oclass;
         LazyLoadBase <? extends EventLoopGroup > lazyloadbase;
@@ -471,10 +477,15 @@
                 }
                 else if (this.func_150729_e() != null)
                 {
                     this.func_150729_e().func_147231_a(new TextComponentTranslation("multiplayer.disconnect.generic"));
                 }
             }
         }
+    }
+
+    public Channel channel()
+    {
+        return field_150746_k;
     }
 
     static class InboundHandlerTuplePacketListener
