--- before/net/minecraft/network/NetworkManager.java
+++ after/net/minecraft/network/NetworkManager.java
@@ -88,6 +88,11 @@
         this.direction = packetDirection;
     }
 
+    public EnumPacketDirection getDirection()
+    {
+        return this.direction;
+    }
+
     @Override
     public void channelActive(ChannelHandlerContext p_channelActive_1_) throws Exception
     {
@@ -210,7 +215,7 @@
         final EnumConnectionState enumconnectionstate = EnumConnectionState.getFromPacket(inPacket);
         final EnumConnectionState enumconnectionstate1 = this.channel.attr(PROTOCOL_ATTRIBUTE_KEY).get();
 
-        if (enumconnectionstate1 != enumconnectionstate)
+        if (enumconnectionstate1 != enumconnectionstate && !( inPacket instanceof net.minecraftforge.fml.common.network.internal.FMLProxyPacket))
         {
             LOGGER.debug("Disabled auto read");
             this.channel.config().setAutoRead(false);
@@ -218,7 +223,7 @@
 
         if (this.channel.eventLoop().inEventLoop())
         {
-            if (enumconnectionstate != enumconnectionstate1)
+            if (enumconnectionstate != enumconnectionstate1 && !( inPacket instanceof net.minecraftforge.fml.common.network.internal.FMLProxyPacket))
             {
                 this.setConnectionState(enumconnectionstate);
             }
@@ -239,7 +244,7 @@
                 @Override
                 public void run()
                 {
-                    if (enumconnectionstate != enumconnectionstate1)
+                    if (enumconnectionstate != enumconnectionstate1  && !( inPacket instanceof net.minecraftforge.fml.common.network.internal.FMLProxyPacket))
                     {
                         NetworkManager.this.setConnectionState(enumconnectionstate);
                     }
@@ -317,6 +322,7 @@
     @SideOnly(Side.CLIENT)
     public static NetworkManager createNetworkManagerAndConnect(InetAddress address, int serverPort, boolean useNativeTransport)
     {
+        if (address instanceof java.net.Inet6Address) System.setProperty("java.net.preferIPv4Stack", "false");
         final NetworkManager networkmanager = new NetworkManager(EnumPacketDirection.CLIENTBOUND);
         Class <? extends SocketChannel > oclass;
         LazyLoadBase <? extends EventLoopGroup > lazyloadbase;
@@ -475,6 +481,11 @@
                 }
             }
         }
+    }
+
+    public Channel channel()
+    {
+        return channel;
     }
 
     static class InboundHandlerTuplePacketListener
