--- before/net/minecraft/network/PacketBuffer.java
+++ after/net/minecraft/network/PacketBuffer.java
@@ -330,7 +330,7 @@
 
             if (stack.getItem().isDamageable() || stack.getItem().getShareTag())
             {
-                nbttagcompound = stack.getTagCompound();
+                nbttagcompound = stack.getItem().getNBTShareTag(stack);
             }
 
             this.writeCompoundTag(nbttagcompound);
@@ -352,7 +352,7 @@
             int j = this.readByte();
             int k = this.readShort();
             ItemStack itemstack = new ItemStack(Item.getItemById(i), j, k);
-            itemstack.setTagCompound(this.readCompoundTag());
+            itemstack.getItem().readNBTShareTag(itemstack, this.readCompoundTag());
             return itemstack;
         }
     }
