--- before/net/minecraft/client/renderer/block/model/multipart/Multipart.java
+++ after/net/minecraft/client/renderer/block/model/multipart/Multipart.java
@@ -33,7 +33,7 @@
 
     public Set<VariantList> getVariants()
     {
-        Set<VariantList> set = Sets.newHashSet();
+        Set<VariantList> set = Sets.<VariantList>newHashSet();
 
         for (Selector selector : this.selectors)
         {
@@ -53,7 +53,6 @@
         return this.stateContainer;
     }
 
-    @Override
     public boolean equals(Object p_equals_1_)
     {
         if (this == p_equals_1_)
@@ -81,7 +80,6 @@
         }
     }
 
-    @Override
     public int hashCode()
     {
         return 31 * this.selectors.hashCode() + (this.stateContainer == null ? 0 : this.stateContainer.hashCode());
@@ -97,11 +95,11 @@
 
             private List<Selector> getSelectors(JsonDeserializationContext context, JsonArray elements)
             {
-                List<Selector> list = Lists.newArrayList();
+                List<Selector> list = Lists.<Selector>newArrayList();
 
                 for (JsonElement jsonelement : elements)
                 {
-                    list.add(context.deserialize(jsonelement, Selector.class));
+                    list.add((Selector)context.deserialize(jsonelement, Selector.class));
                 }
 
                 return list;
