--- before/net/minecraft/world/chunk/storage/AnvilSaveConverter.java
+++ after/net/minecraft/world/chunk/storage/AnvilSaveConverter.java
@@ -49,7 +49,7 @@
     {
         if (this.field_75808_a != null && this.field_75808_a.exists() && this.field_75808_a.isDirectory())
         {
-            List<WorldSummary> list = Lists.newArrayList();
+            List<WorldSummary> list = Lists.<WorldSummary>newArrayList();
             File[] afile = this.field_75808_a.listFiles();
 
             for (File file1 : afile)
@@ -115,9 +115,9 @@
     public boolean func_75805_a(String p_75805_1_, IProgressUpdate p_75805_2_)
     {
         p_75805_2_.func_73718_a(0);
-        List<File> list = Lists.newArrayList();
-        List<File> list1 = Lists.newArrayList();
-        List<File> list2 = Lists.newArrayList();
+        List<File> list = Lists.<File>newArrayList();
+        List<File> list1 = Lists.<File>newArrayList();
+        List<File> list2 = Lists.<File>newArrayList();
         File file1 = new File(this.field_75808_a, p_75805_1_);
         File file2 = new File(file1, "DIM-1");
         File file3 = new File(file1, "DIM1");
