--- before/net/minecraft/world/chunk/storage/RegionFileCache.java
+++ after/net/minecraft/world/chunk/storage/RegionFileCache.java
@@ -9,7 +9,7 @@
 
 public class RegionFileCache
 {
-    private static final Map<File, RegionFile> field_76553_a = Maps.newHashMap();
+    private static final Map<File, RegionFile> field_76553_a = Maps.<File, RegionFile>newHashMap();
 
     public static synchronized RegionFile func_76550_a(File p_76550_0_, int p_76550_1_, int p_76550_2_)
     {
