--- before/net/minecraft/entity/item/EntityEnderPearl.java
+++ after/net/minecraft/entity/item/EntityEnderPearl.java
@@ -45,7 +45,6 @@
         EntityThrowable.registerFixesThrowable(fixer, "ThrownEnderpearl");
     }
 
-    @Override
     protected void onImpact(RayTraceResult result)
     {
         EntityLivingBase entitylivingbase = this.getThrower();
@@ -86,18 +85,9 @@
             }
         }
 
-        for (int i = 0; i < 32; i++)
+        for (int i = 0; i < 32; ++i)
         {
-            this.world
-            .spawnParticle(
-                EnumParticleTypes.PORTAL,
-                this.posX,
-                this.posY + this.rand.nextDouble() * 2.0,
-                this.posZ,
-                this.rand.nextGaussian(),
-                0.0,
-                this.rand.nextGaussian()
-            );
+            this.world.spawnParticle(EnumParticleTypes.PORTAL, this.posX, this.posY + this.rand.nextDouble() * 2.0D, this.posZ, this.rand.nextGaussian(), 0.0D, this.rand.nextGaussian());
         }
 
         if (!this.world.isRemote)
@@ -106,21 +96,16 @@
             {
                 EntityPlayerMP entityplayermp = (EntityPlayerMP)entitylivingbase;
 
-                if (entityplayermp.connection.getNetworkManager().isChannelOpen()
-                        && entityplayermp.world == this.world
-                        && !entityplayermp.isPlayerSleeping())
+                if (entityplayermp.connection.getNetworkManager().isChannelOpen() && entityplayermp.world == this.world && !entityplayermp.isPlayerSleeping())
                 {
+                    net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(entityplayermp, this.posX, this.posY, this.posZ, 5.0F);
+                    if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event))
+                    { // Don't indent to lower patch size
                     if (this.rand.nextFloat() < 0.05F && this.world.getGameRules().getBoolean("doMobSpawning"))
                     {
                         EntityEndermite entityendermite = new EntityEndermite(this.world);
                         entityendermite.setSpawnedByPlayer(true);
-                        entityendermite.setLocationAndAngles(
-                            entitylivingbase.posX,
-                            entitylivingbase.posY,
-                            entitylivingbase.posZ,
-                            entitylivingbase.rotationYaw,
-                            entitylivingbase.rotationPitch
-                        );
+                        entityendermite.setLocationAndAngles(entitylivingbase.posX, entitylivingbase.posY, entitylivingbase.posZ, entitylivingbase.rotationYaw, entitylivingbase.rotationPitch);
                         this.world.spawnEntity(entityendermite);
                     }
 
@@ -129,9 +114,10 @@
                         entitylivingbase.dismountRidingEntity();
                     }
 
-                    entitylivingbase.setPositionAndUpdate(this.posX, this.posY, this.posZ);
+                    entitylivingbase.setPositionAndUpdate(event.getTargetX(), event.getTargetY(), event.getTargetZ());
                     entitylivingbase.fallDistance = 0.0F;
-                    entitylivingbase.attackEntityFrom(DamageSource.FALL, 5.0F);
+                    entitylivingbase.attackEntityFrom(DamageSource.FALL, event.getAttackDamage());
+                    }
                 }
             }
             else if (entitylivingbase != null)
@@ -144,7 +130,6 @@
         }
     }
 
-    @Override
     public void onUpdate()
     {
         EntityLivingBase entitylivingbase = this.getThrower();
@@ -160,14 +145,13 @@
     }
 
     @Nullable
-    @Override
-    public Entity changeDimension(int dimensionIn)
+    public Entity changeDimension(int dimensionIn, net.minecraftforge.common.util.ITeleporter teleporter)
     {
         if (this.thrower.dimension != dimensionIn)
         {
             this.thrower = null;
         }
 
-        return super.changeDimension(dimensionIn);
+        return super.changeDimension(dimensionIn, teleporter);
     }
 }
