--- before/net/minecraft/entity/item/EntityPainting.java
+++ after/net/minecraft/entity/item/EntityPainting.java
@@ -29,7 +29,7 @@
     public EntityPainting(World p_i45849_1_, BlockPos p_i45849_2_, EnumFacing p_i45849_3_)
     {
         super(p_i45849_1_, p_i45849_2_);
-        List<EntityPainting.EnumArt> list = Lists.newArrayList();
+        List<EntityPainting.EnumArt> list = Lists.<EntityPainting.EnumArt>newArrayList();
         int i = 0;
 
         for (EntityPainting.EnumArt entitypainting$enumart : EntityPainting.EnumArt.values())
