--- before/net/minecraft/entity/item/EntityFallingBlock.java
+++ after/net/minecraft/entity/item/EntityFallingBlock.java
@@ -177,6 +177,7 @@
                 {
                     IBlockState iblockstate = this.world.getBlockState(blockpos1);
 
+                    if (this.world.isAirBlock(new BlockPos(this.posX, this.posY - 0.01F, this.posZ))) //Forge: Don't indent below.
                     if (!flag1
                             && BlockFalling.canFallThrough(
                                 this.world.getBlockState(new BlockPos(this.posX, this.posY - 0.01F, this.posZ))
@@ -196,7 +197,7 @@
 
                         if (!this.dontSetBlock)
                         {
-                            if (this.world.mayPlace(block, blockpos1, true, EnumFacing.UP, null)
+                            if (this.world.mayPlace(block, blockpos1, true, EnumFacing.UP, this)
                                     && (flag1 || !BlockFalling.canFallThrough(this.world.getBlockState(blockpos1.down())))
                                     && this.world.setBlockState(blockpos1, this.fallTile, 3))
                             {
@@ -205,7 +206,7 @@
                                     ((BlockFalling)block).onEndFalling(this.world, blockpos1, this.fallTile, iblockstate);
                                 }
 
-                                if (this.tileEntityData != null && block instanceof ITileEntityProvider)
+                                if (this.tileEntityData != null && block.hasTileEntity(this.fallTile))
                                 {
                                     TileEntity tileentity = this.world.getTileEntity(blockpos1);
 
