--- before/net/minecraft/entity/item/EntityMinecartMobSpawner.java
+++ after/net/minecraft/entity/item/EntityMinecartMobSpawner.java
@@ -36,6 +36,10 @@
         {
             return new BlockPos(EntityMinecartMobSpawner.this);
         }
+        @Override  
+        public net.minecraft.entity.Entity getSpawnerEntity() {
+            return EntityMinecartMobSpawner.this;
+        }
     };
 
     public EntityMinecartMobSpawner(World worldIn)
