--- before/net/minecraft/entity/item/EntityBoat.java
+++ after/net/minecraft/entity/item/EntityBoat.java
@@ -475,6 +475,12 @@
                             blockpos$pooledmutableblockpos.setPos(l1, k1, i2);
                             IBlockState iblockstate = this.world.getBlockState(blockpos$pooledmutableblockpos);
 
+                            Boolean result = iblockstate.getBlock().isAABBInsideMaterial(world, blockpos$pooledmutableblockpos, new AxisAlignedBB(blockpos$pooledmutableblockpos), Material.WATER);
+                            if (result != null) {
+                                if (!result) continue;
+                                f = Math.max(f, iblockstate.getBlock().getBlockLiquidHeight(world, blockpos$pooledmutableblockpos, iblockstate, Material.WATER) + blockpos$pooledmutableblockpos.getY());
+                            }
+                            
                             if (iblockstate.getMaterial() == Material.WATER)
                             {
                                 f = Math.max(f, BlockLiquid.getBlockLiquidHeight(iblockstate, this.world, blockpos$pooledmutableblockpos));
@@ -550,7 +556,7 @@
 
                                 if (!list.isEmpty())
                                 {
-                                    f += iblockstate.getBlock().slipperiness;
+                                    f += iblockstate.getBlock().getSlipperiness(iblockstate, this.world, blockpos$pooledmutableblockpos, this);
                                     k1++;
                                 }
 
@@ -592,7 +598,16 @@
                     {
                         blockpos$pooledmutableblockpos.setPos(k1, l1, i2);
                         IBlockState iblockstate = this.world.getBlockState(blockpos$pooledmutableblockpos);
+                        
+                        Boolean result = iblockstate.getBlock().isAABBInsideMaterial(world, blockpos$pooledmutableblockpos, axisalignedbb, Material.WATER);
+                        if (result != null) {
+                            if (!result) continue;
 
+                            float f = iblockstate.getBlock().getBlockLiquidHeight(world, blockpos$pooledmutableblockpos, iblockstate, Material.WATER) + blockpos$pooledmutableblockpos.getY();
+                            this.waterLevel = Math.max(f, this.waterLevel);
+                            flag |= axisalignedbb.minY < (double)f;
+                        }
+                        
                         if (iblockstate.getMaterial() == Material.WATER)
                         {
                             float f = BlockLiquid.getLiquidHeight(iblockstate, this.world, blockpos$pooledmutableblockpos);
@@ -636,6 +651,18 @@
                         blockpos$pooledmutableblockpos.setPos(k1, l1, i2);
                         IBlockState iblockstate = this.world.getBlockState(blockpos$pooledmutableblockpos);
 
+                        Boolean result = iblockstate.getBlock().isAABBInsideMaterial(world, blockpos$pooledmutableblockpos, axisalignedbb, Material.WATER);
+                        if (result != null) {
+                            if (!result) continue;
+
+                            if(iblockstate.getBlock().getBlockLiquidHeight(world, blockpos$pooledmutableblockpos, iblockstate, Material.WATER) > 0)
+                            {
+                                blockpos$pooledmutableblockpos.release();
+                                return EntityBoat.Status.UNDER_FLOWING_WATER;
+                            } else
+                                continue;
+                        }
+
                         if (iblockstate.getMaterial() == Material.WATER
                                 && d0 < BlockLiquid.getLiquidHeight(iblockstate, this.world, blockpos$pooledmutableblockpos))
                         {
@@ -967,6 +994,22 @@
         this.rightInputDown = p_184442_2_;
         this.forwardInputDown = p_184442_3_;
         this.backInputDown = p_184442_4_;
+    }
+
+    // Forge: Fix MC-119811 by instantly completing lerp on board
+    @Override
+    protected void addPassenger(Entity passenger)
+    {
+        super.addPassenger(passenger);
+        if(this.canPassengerSteer() && this.lerpSteps > 0)
+        {
+            this.lerpSteps = 0;
+            this.posX = this.lerpX;
+            this.posY = this.lerpY;
+            this.posZ = this.lerpZ;
+            this.rotationYaw = (float)this.lerpYaw;
+            this.rotationPitch = (float)this.lerpPitch;
+        }
     }
 
     public static enum Status
