--- before/net/minecraft/entity/item/EntityMinecartTNT.java
+++ after/net/minecraft/entity/item/EntityMinecartTNT.java
@@ -66,7 +66,7 @@
         {
             double d0 = this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y;
 
-            if (d0 >= (double)0.01F)
+            if (d0 >= 0.009999999776482582D)
             {
                 this.func_94103_c(d0);
             }
@@ -94,7 +94,7 @@
     {
         double d0 = this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y;
 
-        if (!p_94095_1_.func_76347_k() && !p_94095_1_.func_94541_c() && !(d0 >= (double)0.01F))
+        if (!p_94095_1_.func_76347_k() && !p_94095_1_.func_94541_c() && d0 < 0.009999999776482582D)
         {
             super.func_94095_a(p_94095_1_);
 
