--- before/net/minecraft/entity/item/EntityEnderEye.java
+++ after/net/minecraft/entity/item/EntityEnderEye.java
@@ -40,7 +40,7 @@
             d0 = 4.0D;
         }
 
-        d0 *= 64.0D;
+        d0 = d0 * 64.0D;
         return p_70112_1_ < d0 * d0;
     }
 
@@ -88,8 +88,8 @@
         if (this.field_70127_C == 0.0F && this.field_70126_B == 0.0F)
         {
             float f = MathHelper.func_76133_a(p_70016_1_ * p_70016_1_ + p_70016_5_ * p_70016_5_);
-            this.field_70177_z = (float)(MathHelper.func_181159_b(p_70016_1_, p_70016_5_) * (double)(180F / (float)Math.PI));
-            this.field_70125_A = (float)(MathHelper.func_181159_b(p_70016_3_, (double)f) * (double)(180F / (float)Math.PI));
+            this.field_70177_z = (float)(MathHelper.func_181159_b(p_70016_1_, p_70016_5_) * (180D / Math.PI));
+            this.field_70125_A = (float)(MathHelper.func_181159_b(p_70016_3_, (double)f) * (180D / Math.PI));
             this.field_70126_B = this.field_70177_z;
             this.field_70127_C = this.field_70125_A;
         }
@@ -105,10 +105,11 @@
         this.field_70163_u += this.field_70181_x;
         this.field_70161_v += this.field_70179_y;
         float f = MathHelper.func_76133_a(this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y);
-        this.field_70177_z = (float)(MathHelper.func_181159_b(this.field_70159_w, this.field_70179_y) * (double)(180F / (float)Math.PI));
+        this.field_70177_z = (float)(MathHelper.func_181159_b(this.field_70159_w, this.field_70179_y) * (180D / Math.PI));
 
-        for (this.field_70125_A = (float)(MathHelper.func_181159_b(this.field_70181_x, (double)f) * (double)(180F / (float)Math.PI)); this.field_70125_A - this.field_70127_C < -180.0F; this.field_70127_C -= 360.0F)
+        for (this.field_70125_A = (float)(MathHelper.func_181159_b(this.field_70181_x, (double)f) * (180D / Math.PI)); this.field_70125_A - this.field_70127_C < -180.0F; this.field_70127_C -= 360.0F)
         {
+            ;
         }
 
         while (this.field_70125_A - this.field_70127_C >= 180.0F)
@@ -148,11 +149,11 @@
 
             if (this.field_70163_u < this.field_70225_c)
             {
-                this.field_70181_x += (1.0D - this.field_70181_x) * (double)0.015F;
+                this.field_70181_x += (1.0D - this.field_70181_x) * 0.014999999664723873D;
             }
             else
             {
-                this.field_70181_x += (-1.0D - this.field_70181_x) * (double)0.015F;
+                this.field_70181_x += (-1.0D - this.field_70181_x) * 0.014999999664723873D;
             }
         }
 
