--- before/net/minecraft/entity/item/EntityBoat.java
+++ after/net/minecraft/entity/item/EntityBoat.java
@@ -39,16 +39,12 @@
 
 public class EntityBoat extends Entity
 {
-    private static final DataParameter<Integer> field_184460_a = EntityDataManager.func_187226_a(EntityBoat.class, DataSerializers.field_187192_b);
-    private static final DataParameter<Integer> field_184462_b = EntityDataManager.func_187226_a(EntityBoat.class, DataSerializers.field_187192_b);
-    private static final DataParameter<Float> field_184464_c = EntityDataManager.func_187226_a(EntityBoat.class, DataSerializers.field_187193_c);
-    private static final DataParameter<Integer> field_184466_d = EntityDataManager.func_187226_a(EntityBoat.class, DataSerializers.field_187192_b);
-    private static final DataParameter<Boolean>[] field_184468_e = new DataParameter[]
-        {
-            EntityDataManager.func_187226_a(EntityBoat.class, DataSerializers.field_187198_h),
-            EntityDataManager.func_187226_a(EntityBoat.class, DataSerializers.field_187198_h)
-        };
-    private final float[] field_184470_f = new float[2];
+    private static final DataParameter<Integer> field_184460_a = EntityDataManager.<Integer>func_187226_a(EntityBoat.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Integer> field_184462_b = EntityDataManager.<Integer>func_187226_a(EntityBoat.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Float> field_184464_c = EntityDataManager.<Float>func_187226_a(EntityBoat.class, DataSerializers.field_187193_c);
+    private static final DataParameter<Integer> field_184466_d = EntityDataManager.<Integer>func_187226_a(EntityBoat.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Boolean>[] field_184468_e = new DataParameter[] {EntityDataManager.func_187226_a(EntityBoat.class, DataSerializers.field_187198_h), EntityDataManager.func_187226_a(EntityBoat.class, DataSerializers.field_187198_h)};
+    private final float[] field_184470_f;
     private float field_184472_g;
     private float field_184474_h;
     private float field_184475_as;
@@ -71,6 +67,7 @@
     public EntityBoat(World p_i1704_1_)
     {
         super(p_i1704_1_);
+        this.field_184470_f = new float[2];
         this.field_70156_m = true;
         this.func_70105_a(1.375F, 0.5625F);
     }
@@ -79,98 +76,93 @@
     {
         this(p_i1705_1_);
         this.func_70107_b(p_i1705_2_, p_i1705_4_, p_i1705_6_);
-        this.field_70159_w = 0.0;
-        this.field_70181_x = 0.0;
-        this.field_70179_y = 0.0;
+        this.field_70159_w = 0.0D;
+        this.field_70181_x = 0.0D;
+        this.field_70179_y = 0.0D;
         this.field_70169_q = p_i1705_2_;
         this.field_70167_r = p_i1705_4_;
         this.field_70166_s = p_i1705_6_;
     }
 
-    @Override
     protected boolean func_70041_e_()
     {
         return false;
     }
 
-    @Override
     protected void func_70088_a()
     {
-        this.field_70180_af.func_187214_a(field_184460_a, 0);
-        this.field_70180_af.func_187214_a(field_184462_b, 1);
-        this.field_70180_af.func_187214_a(field_184464_c, 0.0F);
-        this.field_70180_af.func_187214_a(field_184466_d, EntityBoat.Type.OAK.ordinal());
+        this.field_70180_af.func_187214_a(field_184460_a, Integer.valueOf(0));
+        this.field_70180_af.func_187214_a(field_184462_b, Integer.valueOf(1));
+        this.field_70180_af.func_187214_a(field_184464_c, Float.valueOf(0.0F));
+        this.field_70180_af.func_187214_a(field_184466_d, Integer.valueOf(EntityBoat.Type.OAK.ordinal()));
 
         for (DataParameter<Boolean> dataparameter : field_184468_e)
         {
-            this.field_70180_af.func_187214_a(dataparameter, false);
+            this.field_70180_af.func_187214_a(dataparameter, Boolean.valueOf(false));
         }
     }
 
     @Nullable
-    @Override
     public AxisAlignedBB func_70114_g(Entity p_70114_1_)
     {
         return p_70114_1_.func_70104_M() ? p_70114_1_.func_174813_aQ() : null;
     }
 
     @Nullable
-    @Override
     public AxisAlignedBB func_70046_E()
     {
         return this.func_174813_aQ();
     }
 
-    @Override
     public boolean func_70104_M()
     {
         return true;
     }
 
-    @Override
     public double func_70042_X()
     {
-        return -0.1;
+        return -0.1D;
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         if (this.func_180431_b(p_70097_1_))
         {
             return false;
         }
-        else if (this.field_70170_p.field_72995_K || this.field_70128_L)
-        {
-            return true;
-        }
-        else if (p_70097_1_ instanceof EntityDamageSourceIndirect && p_70097_1_.func_76346_g() != null && this.func_184196_w(p_70097_1_.func_76346_g()))
-        {
-            return false;
-        }
-        else
-        {
-            this.func_70269_c(-this.func_70267_i());
-            this.func_70265_b(10);
-            this.func_70266_a(this.func_70271_g() + p_70097_2_ * 10.0F);
-            this.func_70018_K();
-            boolean flag = p_70097_1_.func_76346_g() instanceof EntityPlayer && ((EntityPlayer)p_70097_1_.func_76346_g()).field_71075_bZ.field_75098_d;
+        else if (!this.field_70170_p.field_72995_K && !this.field_70128_L)
+        {
+            if (p_70097_1_ instanceof EntityDamageSourceIndirect && p_70097_1_.func_76346_g() != null && this.func_184196_w(p_70097_1_.func_76346_g()))
+            {
+                return false;
+            }
+            else
+            {
+                this.func_70269_c(-this.func_70267_i());
+                this.func_70265_b(10);
+                this.func_70266_a(this.func_70271_g() + p_70097_2_ * 10.0F);
+                this.func_70018_K();
+                boolean flag = p_70097_1_.func_76346_g() instanceof EntityPlayer && ((EntityPlayer)p_70097_1_.func_76346_g()).field_71075_bZ.field_75098_d;
 
-            if (flag || this.func_70271_g() > 40.0F)
-            {
-                if (!flag && this.field_70170_p.func_82736_K().func_82766_b("doEntityDrops"))
+                if (flag || this.func_70271_g() > 40.0F)
                 {
-                    this.func_145778_a(this.func_184455_j(), 1, 0.0F);
+                    if (!flag && this.field_70170_p.func_82736_K().func_82766_b("doEntityDrops"))
+                    {
+                        this.func_145778_a(this.func_184455_j(), 1, 0.0F);
+                    }
+
+                    this.func_70106_y();
                 }
 
-                this.func_70106_y();
+                return true;
             }
-
+        }
+        else
+        {
             return true;
         }
     }
 
-    @Override
     public void func_70108_f(Entity p_70108_1_)
     {
         if (p_70108_1_ instanceof EntityBoat)
@@ -207,7 +199,6 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_70057_ab()
     {
         this.func_70269_c(-this.func_70267_i());
@@ -215,17 +206,13 @@
         this.func_70266_a(this.func_70271_g() * 11.0F);
     }
 
-    @Override
     public boolean func_70067_L()
     {
         return !this.field_70128_L;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
-    public void func_180426_a(
-        double p_180426_1_, double p_180426_3_, double p_180426_5_, float p_180426_7_, float p_180426_8_, int p_180426_9_, boolean p_180426_10_
-    )
+    public void func_180426_a(double p_180426_1_, double p_180426_3_, double p_180426_5_, float p_180426_7_, float p_180426_8_, int p_180426_9_, boolean p_180426_10_)
     {
         this.field_70281_h = p_180426_1_;
         this.field_184477_av = p_180426_3_;
@@ -235,13 +222,11 @@
         this.field_184476_at = 10;
     }
 
-    @Override
     public EnumFacing func_184172_bi()
     {
         return this.func_174811_aO().func_176746_e();
     }
 
-    @Override
     public void func_70071_h_()
     {
         this.field_184471_aG = this.field_184469_aF;
@@ -253,7 +238,7 @@
         }
         else
         {
-            this.field_184474_h++;
+            ++this.field_184474_h;
         }
 
         if (!this.field_70170_p.field_72995_K && this.field_184474_h >= 60.0F)
@@ -296,18 +281,16 @@
         }
         else
         {
-            this.field_70159_w = 0.0;
-            this.field_70181_x = 0.0;
-            this.field_70179_y = 0.0;
+            this.field_70159_w = 0.0D;
+            this.field_70181_x = 0.0D;
+            this.field_70179_y = 0.0D;
         }
 
-        for (int i = 0; i <= 1; i++)
+        for (int i = 0; i <= 1; ++i)
         {
             if (this.func_184457_a(i))
             {
-                if (!this.func_174814_R()
-                        && (double)(this.field_184470_f[i] % (float)(Math.PI * 2)) <= (float)(Math.PI / 4)
-                        && ((double)this.field_184470_f[i] + (float)(Math.PI / 8)) % (float)(Math.PI * 2) >= (float)(Math.PI / 4))
+                if (!this.func_174814_R() && (double)(this.field_184470_f[i] % ((float)Math.PI * 2F)) <= (Math.PI / 4D) && ((double)this.field_184470_f[i] + 0.39269909262657166D) % (Math.PI * 2D) >= (Math.PI / 4D))
                 {
                     SoundEvent soundevent = this.func_193047_k();
 
@@ -316,21 +299,11 @@
                         Vec3d vec3d = this.func_70676_i(1.0F);
                         double d0 = i == 1 ? -vec3d.field_72449_c : vec3d.field_72449_c;
                         double d1 = i == 1 ? vec3d.field_72450_a : -vec3d.field_72450_a;
-                        this.field_70170_p
-                        .func_184148_a(
-                            null,
-                            this.field_70165_t + d0,
-                            this.field_70163_u,
-                            this.field_70161_v + d1,
-                            soundevent,
-                            this.func_184176_by(),
-                            1.0F,
-                            0.8F + 0.4F * this.field_70146_Z.nextFloat()
-                        );
+                        this.field_70170_p.func_184148_a((EntityPlayer)null, this.field_70165_t + d0, this.field_70163_u, this.field_70161_v + d1, soundevent, this.func_184176_by(), 1.0F, 0.8F + 0.4F * this.field_70146_Z.nextFloat());
                     }
                 }
 
-                this.field_184470_f[i] = (float)((double)this.field_184470_f[i] + (float)(Math.PI / 8));
+                this.field_184470_f[i] = (float)((double)this.field_184470_f[i] + 0.39269909262657166D);
             }
             else
             {
@@ -339,25 +312,19 @@
         }
 
         this.func_145775_I();
-        List<Entity> list = this.field_70170_p.func_175674_a(this, this.func_174813_aQ().func_72314_b(0.2F, -0.01F, 0.2F), EntitySelectors.func_188442_a(this));
+        List<Entity> list = this.field_70170_p.func_175674_a(this, this.func_174813_aQ().func_72314_b(0.20000000298023224D, -0.009999999776482582D, 0.20000000298023224D), EntitySelectors.func_188442_a(this));
 
         if (!list.isEmpty())
         {
             boolean flag = !this.field_70170_p.field_72995_K && !(this.func_184179_bs() instanceof EntityPlayer);
 
-            for (int j = 0; j < list.size(); j++)
+            for (int j = 0; j < list.size(); ++j)
             {
                 Entity entity = list.get(j);
 
                 if (!entity.func_184196_w(this))
                 {
-                    if (flag
-                            && this.func_184188_bt().size() < 2
-                            && !entity.func_184218_aH()
-                            && entity.field_70130_N < this.field_70130_N
-                            && entity instanceof EntityLivingBase
-                            && !(entity instanceof EntityWaterMob)
-                            && !(entity instanceof EntityPlayer))
+                    if (flag && this.func_184188_bt().size() < 2 && !entity.func_184218_aH() && entity.field_70130_N < this.field_70130_N && entity instanceof EntityLivingBase && !(entity instanceof EntityWaterMob) && !(entity instanceof EntityPlayer))
                     {
                         entity.func_184220_m(this);
                     }
@@ -397,7 +364,7 @@
             double d3 = MathHelper.func_76138_g(this.field_70273_g - (double)this.field_70177_z);
             this.field_70177_z = (float)((double)this.field_70177_z + d3 / (double)this.field_184476_at);
             this.field_70125_A = (float)((double)this.field_70125_A + (this.field_184479_ay - (double)this.field_70125_A) / (double)this.field_184476_at);
-            this.field_184476_at--;
+            --this.field_184476_at;
             this.func_70107_b(d0, d1, d2);
             this.func_70101_b(this.field_70177_z, this.field_70125_A);
         }
@@ -405,18 +372,14 @@
 
     public void func_184445_a(boolean p_184445_1_, boolean p_184445_2_)
     {
-        this.field_70180_af.func_187227_b(field_184468_e[0], p_184445_1_);
-        this.field_70180_af.func_187227_b(field_184468_e[1], p_184445_2_);
+        this.field_70180_af.func_187227_b(field_184468_e[0], Boolean.valueOf(p_184445_1_));
+        this.field_70180_af.func_187227_b(field_184468_e[1], Boolean.valueOf(p_184445_2_));
     }
 
     @SideOnly(Side.CLIENT)
     public float func_184448_a(int p_184448_1_, float p_184448_2_)
     {
-        return this.func_184457_a(p_184448_1_)
-               ? (float)MathHelper.func_151238_b(
-                   (double)this.field_184470_f[p_184448_1_] - (float)(Math.PI / 8), (double)this.field_184470_f[p_184448_1_], (double)p_184448_2_
-               )
-               : 0.0F;
+        return this.func_184457_a(p_184448_1_) ? (float)MathHelper.func_151238_b((double)this.field_184470_f[p_184448_1_] - 0.39269909262657166D, (double)this.field_184470_f[p_184448_1_], (double)p_184448_2_) : 0.0F;
     }
 
     private EntityBoat.Status func_184449_t()
@@ -461,48 +424,54 @@
 
         try
         {
-            label87:
+            label108:
 
-            for (int k1 = k; k1 < l; k1++)
+            for (int k1 = k; k1 < l; ++k1)
             {
                 float f = 0.0F;
                 int l1 = i;
 
                 while (true)
                 {
-                    if (l1 < j)
-                    {
-                        for (int i2 = i1; i2 < j1; i2++)
-                        {
-                            blockpos$pooledmutableblockpos.func_181079_c(l1, k1, i2);
-                            IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos$pooledmutableblockpos);
-
-                            if (iblockstate.func_185904_a() == Material.field_151586_h)
-                            {
-                                f = Math.max(f, BlockLiquid.func_190973_f(iblockstate, this.field_70170_p, blockpos$pooledmutableblockpos));
-                            }
-
-                            if (f >= 1.0F)
-                            {
-                                continue label87;
-                            }
-                        }
-
-                        l1++;
-                    }
-                    else
+                    if (l1 >= j)
                     {
                         if (f < 1.0F)
                         {
-                            return (float)blockpos$pooledmutableblockpos.func_177956_o() + f;
+                            float f2 = (float)blockpos$pooledmutableblockpos.func_177956_o() + f;
+                            return f2;
                         }
 
                         break;
                     }
+
+                    for (int i2 = i1; i2 < j1; ++i2)
+                    {
+                        blockpos$pooledmutableblockpos.func_181079_c(l1, k1, i2);
+                        IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos$pooledmutableblockpos);
+
+                        Boolean result = iblockstate.func_177230_c().isAABBInsideMaterial(field_70170_p, blockpos$pooledmutableblockpos, new AxisAlignedBB(blockpos$pooledmutableblockpos), Material.field_151586_h);
+                        if (result != null) {
+                            if (!result) continue;
+                            f = Math.max(f, iblockstate.func_177230_c().getBlockLiquidHeight(field_70170_p, blockpos$pooledmutableblockpos, iblockstate, Material.field_151586_h) + blockpos$pooledmutableblockpos.func_177956_o());
+                        }
+
+                        if (iblockstate.func_185904_a() == Material.field_151586_h)
+                        {
+                            f = Math.max(f, BlockLiquid.func_190973_f(iblockstate, this.field_70170_p, blockpos$pooledmutableblockpos));
+                        }
+
+                        if (f >= 1.0F)
+                        {
+                            continue label108;
+                        }
+                    }
+
+                    ++l1;
                 }
             }
 
-            return (float)(l + 1);
+            float f1 = (float)(l + 1);
+            return f1;
         }
         finally
         {
@@ -513,36 +482,29 @@
     public float func_184441_l()
     {
         AxisAlignedBB axisalignedbb = this.func_174813_aQ();
-        AxisAlignedBB axisalignedbb1 = new AxisAlignedBB(
-            axisalignedbb.field_72340_a,
-            axisalignedbb.field_72338_b - 0.001,
-            axisalignedbb.field_72339_c,
-            axisalignedbb.field_72336_d,
-            axisalignedbb.field_72338_b,
-            axisalignedbb.field_72334_f
-        );
+        AxisAlignedBB axisalignedbb1 = new AxisAlignedBB(axisalignedbb.field_72340_a, axisalignedbb.field_72338_b - 0.001D, axisalignedbb.field_72339_c, axisalignedbb.field_72336_d, axisalignedbb.field_72338_b, axisalignedbb.field_72334_f);
         int i = MathHelper.func_76128_c(axisalignedbb1.field_72340_a) - 1;
         int j = MathHelper.func_76143_f(axisalignedbb1.field_72336_d) + 1;
         int k = MathHelper.func_76128_c(axisalignedbb1.field_72338_b) - 1;
         int l = MathHelper.func_76143_f(axisalignedbb1.field_72337_e) + 1;
         int i1 = MathHelper.func_76128_c(axisalignedbb1.field_72339_c) - 1;
         int j1 = MathHelper.func_76143_f(axisalignedbb1.field_72334_f) + 1;
-        List<AxisAlignedBB> list = Lists.newArrayList();
+        List<AxisAlignedBB> list = Lists.<AxisAlignedBB>newArrayList();
         float f = 0.0F;
         int k1 = 0;
         BlockPos.PooledMutableBlockPos blockpos$pooledmutableblockpos = BlockPos.PooledMutableBlockPos.func_185346_s();
 
         try
         {
-            for (int l1 = i; l1 < j; l1++)
+            for (int l1 = i; l1 < j; ++l1)
             {
-                for (int i2 = i1; i2 < j1; i2++)
+                for (int i2 = i1; i2 < j1; ++i2)
                 {
                     int j2 = (l1 != i && l1 != j - 1 ? 0 : 1) + (i2 != i1 && i2 != j1 - 1 ? 0 : 1);
 
                     if (j2 != 2)
                     {
-                        for (int k2 = k; k2 < l; k2++)
+                        for (int k2 = k; k2 < l; ++k2)
                         {
                             if (j2 <= 0 || k2 != k && k2 != l - 1)
                             {
@@ -552,8 +514,8 @@
 
                                 if (!list.isEmpty())
                                 {
-                                    f += iblockstate.func_177230_c().field_149765_K;
-                                    k1++;
+                                    f += iblockstate.func_177230_c().getSlipperiness(iblockstate, this.field_70170_p, blockpos$pooledmutableblockpos, this);
+                                    ++k1;
                                 }
 
                                 list.clear();
@@ -577,7 +539,7 @@
         int i = MathHelper.func_76128_c(axisalignedbb.field_72340_a);
         int j = MathHelper.func_76143_f(axisalignedbb.field_72336_d);
         int k = MathHelper.func_76128_c(axisalignedbb.field_72338_b);
-        int l = MathHelper.func_76143_f(axisalignedbb.field_72338_b + 0.001);
+        int l = MathHelper.func_76143_f(axisalignedbb.field_72338_b + 0.001D);
         int i1 = MathHelper.func_76128_c(axisalignedbb.field_72339_c);
         int j1 = MathHelper.func_76143_f(axisalignedbb.field_72334_f);
         boolean flag = false;
@@ -586,14 +548,23 @@
 
         try
         {
-            for (int k1 = i; k1 < j; k1++)
+            for (int k1 = i; k1 < j; ++k1)
             {
-                for (int l1 = k; l1 < l; l1++)
+                for (int l1 = k; l1 < l; ++l1)
                 {
-                    for (int i2 = i1; i2 < j1; i2++)
+                    for (int i2 = i1; i2 < j1; ++i2)
                     {
                         blockpos$pooledmutableblockpos.func_181079_c(k1, l1, i2);
                         IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos$pooledmutableblockpos);
+
+                        Boolean result = iblockstate.func_177230_c().isAABBInsideMaterial(field_70170_p, blockpos$pooledmutableblockpos, axisalignedbb, Material.field_151586_h);
+                        if (result != null) {
+                            if (!result) continue;
+                            
+                            float f = iblockstate.func_177230_c().getBlockLiquidHeight(field_70170_p, blockpos$pooledmutableblockpos, iblockstate, Material.field_151586_h) + blockpos$pooledmutableblockpos.func_177956_o();
+                            this.field_184465_aD = Math.max((double)f, this.field_184465_aD);
+                            flag |= axisalignedbb.field_72338_b < (double)f;
+                        }
 
                         if (iblockstate.func_185904_a() == Material.field_151586_h)
                         {
@@ -617,7 +588,7 @@
     private EntityBoat.Status func_184444_v()
     {
         AxisAlignedBB axisalignedbb = this.func_174813_aQ();
-        double d0 = axisalignedbb.field_72337_e + 0.001;
+        double d0 = axisalignedbb.field_72337_e + 0.001D;
         int i = MathHelper.func_76128_c(axisalignedbb.field_72340_a);
         int j = MathHelper.func_76143_f(axisalignedbb.field_72336_d);
         int k = MathHelper.func_76128_c(axisalignedbb.field_72337_e);
@@ -629,21 +600,33 @@
 
         try
         {
-            for (int k1 = i; k1 < j; k1++)
+            for (int k1 = i; k1 < j; ++k1)
             {
-                for (int l1 = k; l1 < l; l1++)
+                for (int l1 = k; l1 < l; ++l1)
                 {
-                    for (int i2 = i1; i2 < j1; i2++)
+                    for (int i2 = i1; i2 < j1; ++i2)
                     {
                         blockpos$pooledmutableblockpos.func_181079_c(k1, l1, i2);
                         IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos$pooledmutableblockpos);
 
-                        if (iblockstate.func_185904_a() == Material.field_151586_h
-                                && d0 < (double)BlockLiquid.func_190972_g(iblockstate, this.field_70170_p, blockpos$pooledmutableblockpos))
-                        {
-                            if (iblockstate.func_177229_b(BlockLiquid.field_176367_b) != 0)
-                            {
+                        Boolean result = iblockstate.func_177230_c().isAABBInsideMaterial(field_70170_p, blockpos$pooledmutableblockpos, axisalignedbb, Material.field_151586_h);
+                        if (result != null) {
+                            if (!result) continue;
+                            
+                            if(iblockstate.func_177230_c().getBlockLiquidHeight(field_70170_p, blockpos$pooledmutableblockpos, iblockstate, Material.field_151586_h) > 0)
+                            {   
+                                blockpos$pooledmutableblockpos.func_185344_t();
                                 return EntityBoat.Status.UNDER_FLOWING_WATER;
+                            } else
+                                continue;
+                        }
+
+                        if (iblockstate.func_185904_a() == Material.field_151586_h && d0 < (double)BlockLiquid.func_190972_g(iblockstate, this.field_70170_p, blockpos$pooledmutableblockpos))
+                        {
+                            if (((Integer)iblockstate.func_177229_b(BlockLiquid.field_176367_b)).intValue() != 0)
+                            {
+                                EntityBoat.Status entityboat$status = EntityBoat.Status.UNDER_FLOWING_WATER;
+                                return entityboat$status;
                             }
 
                             flag = true;
@@ -662,19 +645,17 @@
 
     private void func_184450_w()
     {
-        double d0 = -0.04F;
-        double d1 = this.func_189652_ae() ? 0.0 : -0.04F;
-        double d2 = 0.0;
+        double d0 = -0.03999999910593033D;
+        double d1 = this.func_189652_ae() ? 0.0D : -0.03999999910593033D;
+        double d2 = 0.0D;
         this.field_184472_g = 0.05F;
 
-        if (this.field_184471_aG == EntityBoat.Status.IN_AIR
-                && this.field_184469_aF != EntityBoat.Status.IN_AIR
-                && this.field_184469_aF != EntityBoat.Status.ON_LAND)
+        if (this.field_184471_aG == EntityBoat.Status.IN_AIR && this.field_184469_aF != EntityBoat.Status.IN_AIR && this.field_184469_aF != EntityBoat.Status.ON_LAND)
         {
             this.field_184465_aD = this.func_174813_aQ().field_72338_b + (double)this.field_70131_O;
-            this.func_70107_b(this.field_70165_t, (double)(this.func_184451_k() - this.field_70131_O) + 0.101, this.field_70161_v);
-            this.field_70181_x = 0.0;
-            this.field_184473_aH = 0.0;
+            this.func_70107_b(this.field_70165_t, (double)(this.func_184451_k() - this.field_70131_O) + 0.101D, this.field_70161_v);
+            this.field_70181_x = 0.0D;
+            this.field_184473_aH = 0.0D;
             this.field_184469_aF = EntityBoat.Status.IN_WATER;
         }
         else
@@ -686,12 +667,12 @@
             }
             else if (this.field_184469_aF == EntityBoat.Status.UNDER_FLOWING_WATER)
             {
-                d1 = -7.0E-4;
+                d1 = -7.0E-4D;
                 this.field_184472_g = 0.9F;
             }
             else if (this.field_184469_aF == EntityBoat.Status.UNDER_WATER)
             {
-                d2 = 0.01F;
+                d2 = 0.009999999776482582D;
                 this.field_184472_g = 0.45F;
             }
             else if (this.field_184469_aF == EntityBoat.Status.IN_AIR)
@@ -708,17 +689,17 @@
                 }
             }
 
-            this.field_70159_w = this.field_70159_w * (double)this.field_184472_g;
-            this.field_70179_y = this.field_70179_y * (double)this.field_184472_g;
-            this.field_184475_as = this.field_184475_as * this.field_184472_g;
+            this.field_70159_w *= (double)this.field_184472_g;
+            this.field_70179_y *= (double)this.field_184472_g;
+            this.field_184475_as *= this.field_184472_g;
             this.field_70181_x += d1;
 
-            if (d2 > 0.0)
+            if (d2 > 0.0D)
             {
-                double d3 = 0.65;
-                this.field_70181_x += d2 * 0.06153846016296973;
-                double d4 = 0.75;
-                this.field_70181_x *= 0.75;
+                double d3 = 0.65D;
+                this.field_70181_x += d2 * 0.06153846016296973D;
+                double d4 = 0.75D;
+                this.field_70181_x *= 0.75D;
             }
         }
     }
@@ -736,7 +717,7 @@
 
             if (this.field_184459_aA)
             {
-                this.field_184475_as++;
+                ++this.field_184475_as;
             }
 
             if (this.field_184459_aA != this.field_184480_az && !this.field_184461_aB && !this.field_184463_aC)
@@ -744,7 +725,7 @@
                 f += 0.005F;
             }
 
-            this.field_70177_z = this.field_70177_z + this.field_184475_as;
+            this.field_70177_z += this.field_184475_as;
 
             if (this.field_184461_aB)
             {
@@ -756,21 +737,18 @@
                 f -= 0.005F;
             }
 
-            this.field_70159_w = this.field_70159_w + (double)(MathHelper.func_76126_a(-this.field_70177_z * (float)(Math.PI / 180.0)) * f);
-            this.field_70179_y = this.field_70179_y + (double)(MathHelper.func_76134_b(this.field_70177_z * (float)(Math.PI / 180.0)) * f);
-            this.func_184445_a(
-                this.field_184459_aA && !this.field_184480_az || this.field_184461_aB, this.field_184480_az && !this.field_184459_aA || this.field_184461_aB
-            );
+            this.field_70159_w += (double)(MathHelper.func_76126_a(-this.field_70177_z * 0.017453292F) * f);
+            this.field_70179_y += (double)(MathHelper.func_76134_b(this.field_70177_z * 0.017453292F) * f);
+            this.func_184445_a(this.field_184459_aA && !this.field_184480_az || this.field_184461_aB, this.field_184480_az && !this.field_184459_aA || this.field_184461_aB);
         }
     }
 
-    @Override
     public void func_184232_k(Entity p_184232_1_)
     {
         if (this.func_184196_w(p_184232_1_))
         {
             float f = 0.0F;
-            float f1 = (float)((this.field_70128_L ? 0.01F : this.func_70042_X()) + p_184232_1_.func_70033_W());
+            float f1 = (float)((this.field_70128_L ? 0.009999999776482582D : this.func_70042_X()) + p_184232_1_.func_70033_W());
 
             if (this.func_184188_bt().size() > 1)
             {
@@ -787,13 +765,13 @@
 
                 if (p_184232_1_ instanceof EntityAnimal)
                 {
-                    f = (float)((double)f + 0.2);
+                    f = (float)((double)f + 0.2D);
                 }
             }
 
-            Vec3d vec3d = new Vec3d((double)f, 0.0, 0.0).func_178785_b(-this.field_70177_z * (float)(Math.PI / 180.0) - (float)(Math.PI / 2));
+            Vec3d vec3d = (new Vec3d((double)f, 0.0D, 0.0D)).func_178785_b(-this.field_70177_z * 0.017453292F - ((float)Math.PI / 2F));
             p_184232_1_.func_70107_b(this.field_70165_t + vec3d.field_72450_a, this.field_70163_u + (double)f1, this.field_70161_v + vec3d.field_72449_c);
-            p_184232_1_.field_70177_z = p_184232_1_.field_70177_z + this.field_184475_as;
+            p_184232_1_.field_70177_z += this.field_184475_as;
             p_184232_1_.func_70034_d(p_184232_1_.func_70079_am() + this.field_184475_as);
             this.func_184454_a(p_184232_1_);
 
@@ -817,19 +795,16 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_184190_l(Entity p_184190_1_)
     {
         this.func_184454_a(p_184190_1_);
     }
 
-    @Override
     protected void func_70014_b(NBTTagCompound p_70014_1_)
     {
         p_70014_1_.func_74778_a("Type", this.func_184453_r().func_184980_a());
     }
 
-    @Override
     protected void func_70037_a(NBTTagCompound p_70037_1_)
     {
         if (p_70037_1_.func_150297_b("Type", 8))
@@ -838,7 +813,6 @@
         }
     }
 
-    @Override
     public boolean func_184230_a(EntityPlayer p_184230_1_, EnumHand p_184230_2_)
     {
         if (p_184230_1_.func_70093_af())
@@ -856,7 +830,6 @@
         }
     }
 
-    @Override
     protected void func_184231_a(double p_184231_1_, boolean p_184231_3_, IBlockState p_184231_4_, BlockPos p_184231_5_)
     {
         this.field_184473_aH = this.field_70181_x;
@@ -881,12 +854,12 @@
 
                         if (this.field_70170_p.func_82736_K().func_82766_b("doEntityDrops"))
                         {
-                            for (int i = 0; i < 3; i++)
+                            for (int i = 0; i < 3; ++i)
                             {
                                 this.func_70099_a(new ItemStack(Item.func_150898_a(Blocks.field_150344_f), 1, this.func_184453_r().func_184982_b()), 0.0F);
                             }
 
-                            for (int j = 0; j < 2; j++)
+                            for (int j = 0; j < 2; ++j)
                             {
                                 this.func_145778_a(Items.field_151055_y, 1, 0.0F);
                             }
@@ -896,7 +869,7 @@
 
                 this.field_70143_R = 0.0F;
             }
-            else if (this.field_70170_p.func_180495_p(new BlockPos(this).func_177977_b()).func_185904_a() != Material.field_151586_h && p_184231_1_ < 0.0)
+            else if (this.field_70170_p.func_180495_p((new BlockPos(this)).func_177977_b()).func_185904_a() != Material.field_151586_h && p_184231_1_ < 0.0D)
             {
                 this.field_70143_R = (float)((double)this.field_70143_R - p_184231_1_);
             }
@@ -905,61 +878,59 @@
 
     public boolean func_184457_a(int p_184457_1_)
     {
-        return this.field_70180_af.func_187225_a(field_184468_e[p_184457_1_]) && this.func_184179_bs() != null;
+        return ((Boolean)this.field_70180_af.func_187225_a(field_184468_e[p_184457_1_])).booleanValue() && this.func_184179_bs() != null;
     }
 
     public void func_70266_a(float p_70266_1_)
     {
-        this.field_70180_af.func_187227_b(field_184464_c, p_70266_1_);
+        this.field_70180_af.func_187227_b(field_184464_c, Float.valueOf(p_70266_1_));
     }
 
     public float func_70271_g()
     {
-        return this.field_70180_af.func_187225_a(field_184464_c);
+        return ((Float)this.field_70180_af.func_187225_a(field_184464_c)).floatValue();
     }
 
     public void func_70265_b(int p_70265_1_)
     {
-        this.field_70180_af.func_187227_b(field_184460_a, p_70265_1_);
+        this.field_70180_af.func_187227_b(field_184460_a, Integer.valueOf(p_70265_1_));
     }
 
     public int func_70268_h()
     {
-        return this.field_70180_af.func_187225_a(field_184460_a);
+        return ((Integer)this.field_70180_af.func_187225_a(field_184460_a)).intValue();
     }
 
     public void func_70269_c(int p_70269_1_)
     {
-        this.field_70180_af.func_187227_b(field_184462_b, p_70269_1_);
+        this.field_70180_af.func_187227_b(field_184462_b, Integer.valueOf(p_70269_1_));
     }
 
     public int func_70267_i()
     {
-        return this.field_70180_af.func_187225_a(field_184462_b);
+        return ((Integer)this.field_70180_af.func_187225_a(field_184462_b)).intValue();
     }
 
     public void func_184458_a(EntityBoat.Type p_184458_1_)
     {
-        this.field_70180_af.func_187227_b(field_184466_d, p_184458_1_.ordinal());
+        this.field_70180_af.func_187227_b(field_184466_d, Integer.valueOf(p_184458_1_.ordinal()));
     }
 
     public EntityBoat.Type func_184453_r()
     {
-        return EntityBoat.Type.func_184979_a(this.field_70180_af.func_187225_a(field_184466_d));
+        return EntityBoat.Type.func_184979_a(((Integer)this.field_70180_af.func_187225_a(field_184466_d)).intValue());
     }
 
-    @Override
     protected boolean func_184219_q(Entity p_184219_1_)
     {
         return this.func_184188_bt().size() < 2;
     }
 
     @Nullable
-    @Override
     public Entity func_184179_bs()
     {
         List<Entity> list = this.func_184188_bt();
-        return list.isEmpty() ? null : list.get(0);
+        return list.isEmpty() ? null : (Entity)list.get(0);
     }
 
     @SideOnly(Side.CLIENT)
@@ -1008,7 +979,6 @@
             return this.field_184991_h;
         }
 
-        @Override
         public String toString()
         {
             return this.field_184990_g;
@@ -1026,7 +996,7 @@
 
         public static EntityBoat.Type func_184981_a(String p_184981_0_)
         {
-            for (int i = 0; i < values().length; i++)
+            for (int i = 0; i < values().length; ++i)
             {
                 if (values()[i].func_184980_a().equals(p_184981_0_))
                 {
@@ -1035,6 +1005,22 @@
             }
 
             return values()[0];
+        }
+    }
+
+    // Forge: Fix MC-119811 by instantly completing lerp on board
+    @Override
+    protected void func_184200_o(Entity passenger)
+    {
+        super.func_184200_o(passenger);
+        if(this.func_184186_bw() && this.field_184476_at > 0)
+        {
+            this.field_184476_at = 0;
+            this.field_70165_t = this.field_70281_h;
+            this.field_70163_u = this.field_184477_av;
+            this.field_70161_v = this.field_184478_aw;
+            this.field_70177_z = (float)this.field_70273_g;
+            this.field_70125_A = (float)this.field_184479_ay;
         }
     }
 }
