--- before/net/minecraft/entity/item/EntityItemFrame.java
+++ after/net/minecraft/entity/item/EntityItemFrame.java
@@ -148,13 +148,14 @@
     {
         if (!stack.isEmpty())
         {
-            if (stack.getItem() == Items.FILLED_MAP)
+            if (stack.getItem() instanceof net.minecraft.item.ItemMap)
             {
                 MapData mapdata = ((ItemMap)stack.getItem()).getMapData(stack, this.world);
                 mapdata.mapDecorations.remove("frame-" + this.getEntityId());
             }
 
             stack.setItemFrame(null);
+            this.setDisplayedItem(ItemStack.EMPTY); //Forge: Fix MC-124833 Pistons duplicating Items.
         }
     }
 
