--- before/net/minecraft/entity/item/EntityTNTPrimed.java
+++ after/net/minecraft/entity/item/EntityTNTPrimed.java
@@ -13,14 +13,15 @@
 
 public class EntityTNTPrimed extends Entity
 {
-    private static final DataParameter<Integer> field_184537_a = EntityDataManager.func_187226_a(EntityTNTPrimed.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Integer> field_184537_a = EntityDataManager.<Integer>func_187226_a(EntityTNTPrimed.class, DataSerializers.field_187192_b);
     @Nullable
     private EntityLivingBase field_94084_b;
-    private int field_70516_a = 80;
+    private int field_70516_a;
 
     public EntityTNTPrimed(World p_i1729_1_)
     {
         super(p_i1729_1_);
+        this.field_70516_a = 80;
         this.field_70156_m = true;
         this.field_70178_ae = true;
         this.func_70105_a(0.98F, 0.98F);
@@ -30,9 +31,9 @@
     {
         this(p_i1730_1_);
         this.func_70107_b(p_i1730_2_, p_i1730_4_, p_i1730_6_);
-        float f = (float)(Math.random() * (double)((float)Math.PI * 2F));
+        float f = (float)(Math.random() * (Math.PI * 2D));
         this.field_70159_w = (double)(-((float)Math.sin((double)f)) * 0.02F);
-        this.field_70181_x = (double)0.2F;
+        this.field_70181_x = 0.20000000298023224D;
         this.field_70179_y = (double)(-((float)Math.cos((double)f)) * 0.02F);
         this.func_184534_a(80);
         this.field_70169_q = p_i1730_2_;
@@ -43,7 +44,7 @@
 
     protected void func_70088_a()
     {
-        this.field_70180_af.func_187214_a(field_184537_a, 80);
+        this.field_70180_af.func_187214_a(field_184537_a, Integer.valueOf(80));
     }
 
     protected boolean func_70041_e_()
@@ -64,18 +65,18 @@
 
         if (!this.func_189652_ae())
         {
-            this.field_70181_x -= (double)0.04F;
+            this.field_70181_x -= 0.03999999910593033D;
         }
 
         this.func_70091_d(MoverType.SELF, this.field_70159_w, this.field_70181_x, this.field_70179_y);
-        this.field_70159_w *= (double)0.98F;
-        this.field_70181_x *= (double)0.98F;
-        this.field_70179_y *= (double)0.98F;
+        this.field_70159_w *= 0.9800000190734863D;
+        this.field_70181_x *= 0.9800000190734863D;
+        this.field_70179_y *= 0.9800000190734863D;
 
         if (this.field_70122_E)
         {
-            this.field_70159_w *= (double)0.7F;
-            this.field_70179_y *= (double)0.7F;
+            this.field_70159_w *= 0.699999988079071D;
+            this.field_70179_y *= 0.699999988079071D;
             this.field_70181_x *= -0.5D;
         }
 
@@ -126,7 +127,7 @@
 
     public void func_184534_a(int p_184534_1_)
     {
-        this.field_70180_af.func_187227_b(field_184537_a, p_184534_1_);
+        this.field_70180_af.func_187227_b(field_184537_a, Integer.valueOf(p_184534_1_));
         this.field_70516_a = p_184534_1_;
     }
 
@@ -140,7 +141,7 @@
 
     public int func_184535_k()
     {
-        return this.field_70180_af.func_187225_a(field_184537_a);
+        return ((Integer)this.field_70180_af.func_187225_a(field_184537_a)).intValue();
     }
 
     public int func_184536_l()
