--- before/net/minecraft/entity/item/EntityEnderPearl.java
+++ after/net/minecraft/entity/item/EntityEnderPearl.java
@@ -110,6 +110,9 @@
                         && entityplayermp.world == this.world
                         && !entityplayermp.isPlayerSleeping())
                 {
+                    net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(entityplayermp, this.posX, this.posY, this.posZ, 5.0F);
+                    if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event))
+                    { // Don't indent to lower patch size
                     if (this.rand.nextFloat() < 0.05F && this.world.getGameRules().getBoolean("doMobSpawning"))
                     {
                         EntityEndermite entityendermite = new EntityEndermite(this.world);
@@ -129,9 +132,10 @@
                         entitylivingbase.dismountRidingEntity();
                     }
 
-                    entitylivingbase.setPositionAndUpdate(this.posX, this.posY, this.posZ);
+                    entitylivingbase.setPositionAndUpdate(event.getTargetX(), event.getTargetY(), event.getTargetZ());
                     entitylivingbase.fallDistance = 0.0F;
-                    entitylivingbase.attackEntityFrom(DamageSource.FALL, 5.0F);
+                    entitylivingbase.attackEntityFrom(DamageSource.FALL, event.getAttackDamage());
+                    }
                 }
             }
             else if (entitylivingbase != null)
@@ -161,13 +165,13 @@
 
     @Nullable
     @Override
-    public Entity changeDimension(int dimensionIn)
+    public Entity changeDimension(int dimensionIn, net.minecraftforge.common.util.ITeleporter teleporter)
     {
         if (this.thrower.dimension != dimensionIn)
         {
             this.thrower = null;
         }
 
-        return super.changeDimension(dimensionIn);
+        return super.changeDimension(dimensionIn, teleporter);
     }
 }
