--- before/net/minecraft/entity/item/EntityMinecartHopper.java
+++ after/net/minecraft/entity/item/EntityMinecartHopper.java
@@ -7,6 +7,7 @@
 import net.minecraft.init.Blocks;
 import net.minecraft.inventory.Container;
 import net.minecraft.inventory.ContainerHopper;
+import net.minecraft.inventory.IInventory;
 import net.minecraft.item.Item;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.tileentity.IHopper;
@@ -34,33 +35,29 @@
         super(p_i1721_1_, p_i1721_2_, p_i1721_4_, p_i1721_6_);
     }
 
-    @Override
     public EntityMinecart.Type func_184264_v()
     {
         return EntityMinecart.Type.HOPPER;
     }
 
-    @Override
     public IBlockState func_180457_u()
     {
         return Blocks.field_150438_bZ.func_176223_P();
     }
 
-    @Override
     public int func_94085_r()
     {
         return 1;
     }
 
-    @Override
     public int func_70302_i_()
     {
         return 5;
     }
 
-    @Override
     public boolean func_184230_a(EntityPlayer p_184230_1_, EnumHand p_184230_2_)
     {
+        if (super.func_184230_a(p_184230_1_, p_184230_2_)) return true;
         if (!this.field_70170_p.field_72995_K)
         {
             p_184230_1_.func_71007_a(this);
@@ -69,7 +66,6 @@
         return true;
     }
 
-    @Override
     public void func_96095_a(int p_96095_1_, int p_96095_2_, int p_96095_3_, boolean p_96095_4_)
     {
         boolean flag = !p_96095_4_;
@@ -90,31 +86,26 @@
         this.field_96113_a = p_96110_1_;
     }
 
-    @Override
     public World func_145831_w()
     {
         return this.field_70170_p;
     }
 
-    @Override
     public double func_96107_aA()
     {
         return this.field_70165_t;
     }
 
-    @Override
     public double func_96109_aB()
     {
-        return this.field_70163_u + 0.5;
+        return this.field_70163_u + 0.5D;
     }
 
-    @Override
     public double func_96108_aC()
     {
         return this.field_70161_v;
     }
 
-    @Override
     public void func_70071_h_()
     {
         super.func_70071_h_();
@@ -153,19 +144,17 @@
         }
         else
         {
-            List<EntityItem> list = this.field_70170_p
-                                    .func_175647_a(EntityItem.class, this.func_174813_aQ().func_72314_b(0.25, 0.0, 0.25), EntitySelectors.field_94557_a);
+            List<EntityItem> list = this.field_70170_p.<EntityItem>func_175647_a(EntityItem.class, this.func_174813_aQ().func_72314_b(0.25D, 0.0D, 0.25D), EntitySelectors.field_94557_a);
 
             if (!list.isEmpty())
             {
-                TileEntityHopper.func_145898_a(null, this, list.get(0));
+                TileEntityHopper.func_145898_a((IInventory)null, this, list.get(0));
             }
 
             return false;
         }
     }
 
-    @Override
     public void func_94095_a(DamageSource p_94095_1_)
     {
         super.func_94095_a(p_94095_1_);
@@ -181,7 +170,6 @@
         EntityMinecartContainer.func_190574_b(p_189682_0_, EntityMinecartHopper.class);
     }
 
-    @Override
     protected void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
@@ -189,7 +177,6 @@
         p_70014_1_.func_74757_a("Enabled", this.field_96113_a);
     }
 
-    @Override
     protected void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
@@ -207,13 +194,11 @@
         return this.field_98044_b > 0;
     }
 
-    @Override
     public String func_174875_k()
     {
         return "minecraft:hopper";
     }
 
-    @Override
     public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
     {
         return new ContainerHopper(p_174876_1_, this, p_174876_2_);
