--- before/net/minecraft/entity/item/EntityFireworkRocket.java
+++ after/net/minecraft/entity/item/EntityFireworkRocket.java
@@ -26,8 +26,8 @@
 
 public class EntityFireworkRocket extends Entity
 {
-    private static final DataParameter<ItemStack> field_184566_a = EntityDataManager.func_187226_a(EntityFireworkRocket.class, DataSerializers.field_187196_f);
-    private static final DataParameter<Integer> field_191512_b = EntityDataManager.func_187226_a(EntityFireworkRocket.class, DataSerializers.field_187192_b);
+    private static final DataParameter<ItemStack> field_184566_a = EntityDataManager.<ItemStack>func_187226_a(EntityFireworkRocket.class, DataSerializers.field_187196_f);
+    private static final DataParameter<Integer> field_191512_b = EntityDataManager.<Integer>func_187226_a(EntityFireworkRocket.class, DataSerializers.field_187192_b);
     private int field_92056_a;
     private int field_92055_b;
     private EntityLivingBase field_191513_e;
@@ -41,7 +41,7 @@
     protected void func_70088_a()
     {
         this.field_70180_af.func_187214_a(field_184566_a, ItemStack.field_190927_a);
-        this.field_70180_af.func_187214_a(field_191512_b, 0);
+        this.field_70180_af.func_187214_a(field_191512_b, Integer.valueOf(0));
     }
 
     @SideOnly(Side.CLIENT)
@@ -81,7 +81,7 @@
     public EntityFireworkRocket(World p_i47367_1_, ItemStack p_i47367_2_, EntityLivingBase p_i47367_3_)
     {
         this(p_i47367_1_, p_i47367_3_.field_70165_t, p_i47367_3_.field_70163_u, p_i47367_3_.field_70161_v, p_i47367_2_);
-        this.field_70180_af.func_187227_b(field_191512_b, p_i47367_3_.func_145782_y());
+        this.field_70180_af.func_187227_b(field_191512_b, Integer.valueOf(p_i47367_3_.func_145782_y()));
         this.field_191513_e = p_i47367_3_;
     }
 
@@ -95,8 +95,8 @@
         if (this.field_70127_C == 0.0F && this.field_70126_B == 0.0F)
         {
             float f = MathHelper.func_76133_a(p_70016_1_ * p_70016_1_ + p_70016_5_ * p_70016_5_);
-            this.field_70177_z = (float)(MathHelper.func_181159_b(p_70016_1_, p_70016_5_) * (double)(180F / (float)Math.PI));
-            this.field_70125_A = (float)(MathHelper.func_181159_b(p_70016_3_, (double)f) * (double)(180F / (float)Math.PI));
+            this.field_70177_z = (float)(MathHelper.func_181159_b(p_70016_1_, p_70016_5_) * (180D / Math.PI));
+            this.field_70125_A = (float)(MathHelper.func_181159_b(p_70016_3_, (double)f) * (180D / Math.PI));
             this.field_70126_B = this.field_70177_z;
             this.field_70127_C = this.field_70125_A;
         }
@@ -113,7 +113,7 @@
         {
             if (this.field_191513_e == null)
             {
-                Entity entity = this.field_70170_p.func_73045_a(this.field_70180_af.func_187225_a(field_191512_b));
+                Entity entity = this.field_70170_p.func_73045_a(((Integer)this.field_70180_af.func_187225_a(field_191512_b)).intValue());
 
                 if (entity instanceof EntityLivingBase)
                 {
@@ -148,10 +148,11 @@
         }
 
         float f = MathHelper.func_76133_a(this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y);
-        this.field_70177_z = (float)(MathHelper.func_181159_b(this.field_70159_w, this.field_70179_y) * (double)(180F / (float)Math.PI));
+        this.field_70177_z = (float)(MathHelper.func_181159_b(this.field_70159_w, this.field_70179_y) * (180D / Math.PI));
 
-        for (this.field_70125_A = (float)(MathHelper.func_181159_b(this.field_70181_x, (double)f) * (double)(180F / (float)Math.PI)); this.field_70125_A - this.field_70127_C < -180.0F; this.field_70127_C -= 360.0F)
+        for (this.field_70125_A = (float)(MathHelper.func_181159_b(this.field_70181_x, (double)f) * (180D / Math.PI)); this.field_70125_A - this.field_70127_C < -180.0F; this.field_70127_C -= 360.0F)
         {
+            ;
         }
 
         while (this.field_70125_A - this.field_70127_C >= 180.0F)
@@ -195,7 +196,7 @@
     private void func_191510_k()
     {
         float f = 0.0F;
-        ItemStack itemstack = this.field_70180_af.func_187225_a(field_184566_a);
+        ItemStack itemstack = (ItemStack)this.field_70180_af.func_187225_a(field_184566_a);
         NBTTagCompound nbttagcompound = itemstack.func_190926_b() ? null : itemstack.func_179543_a("Fireworks");
         NBTTagList nbttaglist = nbttagcompound != null ? nbttagcompound.func_150295_c("Explosions", 10) : null;
 
@@ -216,7 +217,7 @@
 
             for (EntityLivingBase entitylivingbase : this.field_70170_p.func_72872_a(EntityLivingBase.class, this.func_174813_aQ().func_186662_g(5.0D)))
             {
-                if (entitylivingbase != this.field_191513_e && !(this.func_70068_e(entitylivingbase) > 25.0D))
+                if (entitylivingbase != this.field_191513_e && this.func_70068_e(entitylivingbase) <= 25.0D)
                 {
                     boolean flag = false;
 
@@ -243,7 +244,7 @@
 
     public boolean func_191511_j()
     {
-        return this.field_70180_af.func_187225_a(field_191512_b) > 0;
+        return ((Integer)this.field_70180_af.func_187225_a(field_191512_b)).intValue() > 0;
     }
 
     @SideOnly(Side.CLIENT)
@@ -251,7 +252,7 @@
     {
         if (p_70103_1_ == 17 && this.field_70170_p.field_72995_K)
         {
-            ItemStack itemstack = this.field_70180_af.func_187225_a(field_184566_a);
+            ItemStack itemstack = (ItemStack)this.field_70180_af.func_187225_a(field_184566_a);
             NBTTagCompound nbttagcompound = itemstack.func_190926_b() ? null : itemstack.func_179543_a("Fireworks");
             this.field_70170_p.func_92088_a(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70159_w, this.field_70181_x, this.field_70179_y, nbttagcompound);
         }
@@ -261,14 +262,14 @@
 
     public static void func_189656_a(DataFixer p_189656_0_)
     {
-        p_189656_0_.func_188258_a(FixTypes.ENTITY, new ItemStackData(EntityFireworkRocket.class, "FireworksItem"));
+        p_189656_0_.func_188258_a(FixTypes.ENTITY, new ItemStackData(EntityFireworkRocket.class, new String[] {"FireworksItem"}));
     }
 
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         p_70014_1_.func_74768_a("Life", this.field_92056_a);
         p_70014_1_.func_74768_a("LifeTime", this.field_92055_b);
-        ItemStack itemstack = this.field_70180_af.func_187225_a(field_184566_a);
+        ItemStack itemstack = (ItemStack)this.field_70180_af.func_187225_a(field_184566_a);
 
         if (!itemstack.func_190926_b())
         {
