--- before/net/minecraft/entity/item/EntityMinecartEmpty.java
+++ after/net/minecraft/entity/item/EntityMinecartEmpty.java
@@ -22,9 +22,10 @@
         EntityMinecart.func_189669_a(p_189673_0_, EntityMinecartEmpty.class);
     }
 
-    @Override
     public boolean func_184230_a(EntityPlayer p_184230_1_, EnumHand p_184230_2_)
     {
+        if (super.func_184230_a(p_184230_1_, p_184230_2_)) return true;
+
         if (p_184230_1_.func_70093_af())
         {
             return false;
@@ -44,7 +45,6 @@
         }
     }
 
-    @Override
     public void func_96095_a(int p_96095_1_, int p_96095_2_, int p_96095_3_, boolean p_96095_4_)
     {
         if (p_96095_4_)
@@ -64,7 +64,6 @@
         }
     }
 
-    @Override
     public EntityMinecart.Type func_184264_v()
     {
         return EntityMinecart.Type.RIDEABLE;
