--- before/net/minecraft/client/renderer/culling/ClippingHelperImpl.java
+++ after/net/minecraft/client/renderer/culling/ClippingHelperImpl.java
@@ -1,6 +1,5 @@
 package net.minecraft.client.renderer.culling;
 
-import java.nio.Buffer;
 import java.nio.FloatBuffer;
 import net.minecraft.client.renderer.GLAllocation;
 import net.minecraft.client.renderer.GlStateManager;
@@ -33,16 +32,16 @@
 
     public void func_78560_b()
     {
-        ((Buffer)this.field_78561_f).clear();
-        ((Buffer)this.field_78562_g).clear();
-        ((Buffer)this.field_78564_h).clear();
+        this.field_78561_f.clear();
+        this.field_78562_g.clear();
+        this.field_78564_h.clear();
         GlStateManager.func_179111_a(2983, this.field_78561_f);
         GlStateManager.func_179111_a(2982, this.field_78562_g);
         float[] afloat = this.field_178625_b;
         float[] afloat1 = this.field_178626_c;
-        ((Buffer)this.field_78561_f).flip().limit(16);
+        this.field_78561_f.flip().limit(16);
         this.field_78561_f.get(afloat);
-        ((Buffer)this.field_78562_g).flip().limit(16);
+        this.field_78562_g.flip().limit(16);
         this.field_78562_g.get(afloat1);
         this.field_78554_d[0] = afloat1[0] * afloat[0] + afloat1[1] * afloat[4] + afloat1[2] * afloat[8] + afloat1[3] * afloat[12];
         this.field_78554_d[1] = afloat1[0] * afloat[1] + afloat1[1] * afloat[5] + afloat1[2] * afloat[9] + afloat1[3] * afloat[13];
