--- before/net/minecraft/client/renderer/culling/ClippingHelper.java
+++ after/net/minecraft/client/renderer/culling/ClippingHelper.java
@@ -22,7 +22,7 @@
         {
             float[] afloat = this.field_78557_a[i];
 
-            if (!(this.func_178624_a(afloat, p_78553_1_, p_78553_3_, p_78553_5_) > 0.0D) && !(this.func_178624_a(afloat, p_78553_7_, p_78553_3_, p_78553_5_) > 0.0D) && !(this.func_178624_a(afloat, p_78553_1_, p_78553_9_, p_78553_5_) > 0.0D) && !(this.func_178624_a(afloat, p_78553_7_, p_78553_9_, p_78553_5_) > 0.0D) && !(this.func_178624_a(afloat, p_78553_1_, p_78553_3_, p_78553_11_) > 0.0D) && !(this.func_178624_a(afloat, p_78553_7_, p_78553_3_, p_78553_11_) > 0.0D) && !(this.func_178624_a(afloat, p_78553_1_, p_78553_9_, p_78553_11_) > 0.0D) && !(this.func_178624_a(afloat, p_78553_7_, p_78553_9_, p_78553_11_) > 0.0D))
+            if (this.func_178624_a(afloat, p_78553_1_, p_78553_3_, p_78553_5_) <= 0.0D && this.func_178624_a(afloat, p_78553_7_, p_78553_3_, p_78553_5_) <= 0.0D && this.func_178624_a(afloat, p_78553_1_, p_78553_9_, p_78553_5_) <= 0.0D && this.func_178624_a(afloat, p_78553_7_, p_78553_9_, p_78553_5_) <= 0.0D && this.func_178624_a(afloat, p_78553_1_, p_78553_3_, p_78553_11_) <= 0.0D && this.func_178624_a(afloat, p_78553_7_, p_78553_3_, p_78553_11_) <= 0.0D && this.func_178624_a(afloat, p_78553_1_, p_78553_9_, p_78553_11_) <= 0.0D && this.func_178624_a(afloat, p_78553_7_, p_78553_9_, p_78553_11_) <= 0.0D)
             {
                 return false;
             }
