--- before/net/minecraft/client/gui/achievement/GuiStats.java
+++ after/net/minecraft/client/gui/achievement/GuiStats.java
@@ -488,7 +488,7 @@
         public StatsBlock(Minecraft p_i47554_2_)
         {
             super(p_i47554_2_);
-            this.field_148219_m = Lists.newArrayList();
+            this.field_148219_m = Lists.<StatCrafting>newArrayList();
 
             for (StatCrafting statcrafting : StatList.field_188096_e)
             {
@@ -714,7 +714,7 @@
         public StatsItem(Minecraft p_i47552_2_)
         {
             super(p_i47552_2_);
-            this.field_148219_m = Lists.newArrayList();
+            this.field_148219_m = Lists.<StatCrafting>newArrayList();
 
             for (StatCrafting statcrafting : StatList.field_188095_d)
             {
@@ -897,7 +897,7 @@
     @SideOnly(Side.CLIENT)
     class StatsMobsList extends GuiSlot
     {
-        private final List<EntityList.EntityEggInfo> field_148222_l = Lists.newArrayList();
+        private final List<EntityList.EntityEggInfo> field_148222_l = Lists.<EntityList.EntityEggInfo>newArrayList();
 
         public StatsMobsList(Minecraft p_i47551_2_)
         {
