--- before/net/minecraft/entity/boss/EntityWither.java
+++ after/net/minecraft/entity/boss/EntityWither.java
@@ -303,7 +303,7 @@
                     this.posZ,
                     7.0F,
                     false,
-                    this.world.getGameRules().getBoolean("mobGriefing")
+                    net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this)
                 );
                 this.world.playBroadcastSound(1023, new BlockPos(this), 0);
             }
@@ -407,7 +407,7 @@
             {
                 this.blockBreakCounter--;
 
-                if (this.blockBreakCounter == 0 && this.world.getGameRules().getBoolean("mobGriefing"))
+                if (this.blockBreakCounter == 0 && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this))
                 {
                     int i1 = MathHelper.floor(this.posY);
                     int l1 = MathHelper.floor(this.posX);
@@ -427,7 +427,9 @@
                                 IBlockState iblockstate = this.world.getBlockState(blockpos);
                                 Block block = iblockstate.getBlock();
 
-                                if (iblockstate.getMaterial() != Material.AIR && canDestroyBlock(block))
+                                if (!block.isAir(iblockstate, this.world, blockpos)
+                                       && block.canEntityDestroy(iblockstate, world, blockpos, this)
+                                       && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, iblockstate))
                                 {
                                     flag = this.world.destroyBlock(blockpos, true) || flag;
                                 }
