--- before/net/minecraft/entity/boss/EntityWither.java
+++ after/net/minecraft/entity/boss/EntityWither.java
@@ -52,11 +52,11 @@
 
 public class EntityWither extends EntityMob implements IRangedAttackMob
 {
-    private static final DataParameter<Integer> field_184741_a = EntityDataManager.func_187226_a(EntityWither.class, DataSerializers.field_187192_b);
-    private static final DataParameter<Integer> field_184742_b = EntityDataManager.func_187226_a(EntityWither.class, DataSerializers.field_187192_b);
-    private static final DataParameter<Integer> field_184743_c = EntityDataManager.func_187226_a(EntityWither.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Integer> field_184741_a = EntityDataManager.<Integer>func_187226_a(EntityWither.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Integer> field_184742_b = EntityDataManager.<Integer>func_187226_a(EntityWither.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Integer> field_184743_c = EntityDataManager.<Integer>func_187226_a(EntityWither.class, DataSerializers.field_187192_b);
     private static final DataParameter<Integer>[] field_184745_bv = new DataParameter[] {field_184741_a, field_184742_b, field_184743_c};
-    private static final DataParameter<Integer> field_184746_bw = EntityDataManager.func_187226_a(EntityWither.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Integer> field_184746_bw = EntityDataManager.<Integer>func_187226_a(EntityWither.class, DataSerializers.field_187192_b);
     private final float[] field_82220_d = new float[2];
     private final float[] field_82221_e = new float[2];
     private final float[] field_82217_f = new float[2];
@@ -64,15 +64,12 @@
     private final int[] field_82223_h = new int[2];
     private final int[] field_82224_i = new int[2];
     private int field_82222_j;
-    private final BossInfoServer field_184744_bE = (BossInfoServer)new BossInfoServer(this.func_145748_c_(), BossInfo.Color.PURPLE, BossInfo.Overlay.PROGRESS)
-            .func_186741_a(true);
+    private final BossInfoServer field_184744_bE = (BossInfoServer)(new BossInfoServer(this.func_145748_c_(), BossInfo.Color.PURPLE, BossInfo.Overlay.PROGRESS)).func_186741_a(true);
     private static final Predicate<Entity> field_82219_bJ = new Predicate<Entity>()
     {
         public boolean apply(@Nullable Entity p_apply_1_)
         {
-            return p_apply_1_ instanceof EntityLivingBase
-                   && ((EntityLivingBase)p_apply_1_).func_70668_bt() != EnumCreatureAttribute.UNDEAD
-                   && ((EntityLivingBase)p_apply_1_).func_190631_cK();
+            return p_apply_1_ instanceof EntityLivingBase && ((EntityLivingBase)p_apply_1_).func_70668_bt() != EnumCreatureAttribute.UNDEAD && ((EntityLivingBase)p_apply_1_).func_190631_cK();
         }
     };
 
@@ -86,27 +83,25 @@
         this.field_70728_aV = 50;
     }
 
-    @Override
     protected void func_184651_r()
     {
         this.field_70714_bg.func_75776_a(0, new EntityWither.AIDoNothing());
         this.field_70714_bg.func_75776_a(1, new EntityAISwimming(this));
-        this.field_70714_bg.func_75776_a(2, new EntityAIAttackRanged(this, 1.0, 40, 20.0F));
-        this.field_70714_bg.func_75776_a(5, new EntityAIWanderAvoidWater(this, 1.0));
+        this.field_70714_bg.func_75776_a(2, new EntityAIAttackRanged(this, 1.0D, 40, 20.0F));
+        this.field_70714_bg.func_75776_a(5, new EntityAIWanderAvoidWater(this, 1.0D));
         this.field_70714_bg.func_75776_a(6, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
         this.field_70714_bg.func_75776_a(7, new EntityAILookIdle(this));
-        this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, false));
-        this.field_70715_bh.func_75776_a(2, new EntityAINearestAttackableTarget<>(this, EntityLiving.class, 0, false, false, field_82219_bJ));
+        this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, false, new Class[0]));
+        this.field_70715_bh.func_75776_a(2, new EntityAINearestAttackableTarget(this, EntityLiving.class, 0, false, false, field_82219_bJ));
     }
 
-    @Override
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184741_a, 0);
-        this.field_70180_af.func_187214_a(field_184742_b, 0);
-        this.field_70180_af.func_187214_a(field_184743_c, 0);
-        this.field_70180_af.func_187214_a(field_184746_bw, 0);
+        this.field_70180_af.func_187214_a(field_184741_a, Integer.valueOf(0));
+        this.field_70180_af.func_187214_a(field_184742_b, Integer.valueOf(0));
+        this.field_70180_af.func_187214_a(field_184743_c, Integer.valueOf(0));
+        this.field_70180_af.func_187214_a(field_184746_bw, Integer.valueOf(0));
     }
 
     public static void func_189782_b(DataFixer p_189782_0_)
@@ -114,14 +109,12 @@
         EntityLiving.func_189752_a(p_189782_0_, EntityWither.class);
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
         p_70014_1_.func_74768_a("Invul", this.func_82212_n());
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
@@ -133,35 +126,30 @@
         }
     }
 
-    @Override
     public void func_96094_a(String p_96094_1_)
     {
         super.func_96094_a(p_96094_1_);
         this.field_184744_bE.func_186739_a(this.func_145748_c_());
     }
 
-    @Override
     protected SoundEvent func_184639_G()
     {
         return SoundEvents.field_187925_gy;
     }
 
-    @Override
     protected SoundEvent func_184601_bQ(DamageSource p_184601_1_)
     {
         return SoundEvents.field_187851_gB;
     }
 
-    @Override
     protected SoundEvent func_184615_bR()
     {
         return SoundEvents.field_187849_gA;
     }
 
-    @Override
     public void func_70636_d()
     {
-        this.field_70181_x *= 0.6F;
+        this.field_70181_x *= 0.6000000238418579D;
 
         if (!this.field_70170_p.field_72995_K && this.func_82203_t(0) > 0)
         {
@@ -169,43 +157,43 @@
 
             if (entity != null)
             {
-                if (this.field_70163_u < entity.field_70163_u || !this.func_82205_o() && this.field_70163_u < entity.field_70163_u + 5.0)
+                if (this.field_70163_u < entity.field_70163_u || !this.func_82205_o() && this.field_70163_u < entity.field_70163_u + 5.0D)
                 {
-                    if (this.field_70181_x < 0.0)
+                    if (this.field_70181_x < 0.0D)
                     {
-                        this.field_70181_x = 0.0;
+                        this.field_70181_x = 0.0D;
                     }
 
-                    this.field_70181_x = this.field_70181_x + (0.5 - this.field_70181_x) * 0.6F;
+                    this.field_70181_x += (0.5D - this.field_70181_x) * 0.6000000238418579D;
                 }
 
                 double d0 = entity.field_70165_t - this.field_70165_t;
                 double d1 = entity.field_70161_v - this.field_70161_v;
                 double d3 = d0 * d0 + d1 * d1;
 
-                if (d3 > 9.0)
+                if (d3 > 9.0D)
                 {
                     double d5 = (double)MathHelper.func_76133_a(d3);
-                    this.field_70159_w = this.field_70159_w + (d0 / d5 * 0.5 - this.field_70159_w) * 0.6F;
-                    this.field_70179_y = this.field_70179_y + (d1 / d5 * 0.5 - this.field_70179_y) * 0.6F;
+                    this.field_70159_w += (d0 / d5 * 0.5D - this.field_70159_w) * 0.6000000238418579D;
+                    this.field_70179_y += (d1 / d5 * 0.5D - this.field_70179_y) * 0.6000000238418579D;
                 }
             }
         }
 
-        if (this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y > 0.05F)
+        if (this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y > 0.05000000074505806D)
         {
-            this.field_70177_z = (float)MathHelper.func_181159_b(this.field_70179_y, this.field_70159_w) * (180.0F / (float)Math.PI) - 90.0F;
+            this.field_70177_z = (float)MathHelper.func_181159_b(this.field_70179_y, this.field_70159_w) * (180F / (float)Math.PI) - 90.0F;
         }
 
         super.func_70636_d();
 
-        for (int i = 0; i < 2; i++)
+        for (int i = 0; i < 2; ++i)
         {
             this.field_82218_g[i] = this.field_82221_e[i];
             this.field_82217_f[i] = this.field_82220_d[i];
         }
 
-        for (int j = 0; j < 2; j++)
+        for (int j = 0; j < 2; ++j)
         {
             int k = this.func_82203_t(j + 1);
             Entity entity1 = null;
@@ -224,8 +212,8 @@
                 double d7 = entity1.field_70163_u + (double)entity1.func_70047_e() - d12;
                 double d8 = entity1.field_70161_v - d13;
                 double d9 = (double)MathHelper.func_76133_a(d6 * d6 + d8 * d8);
-                float f = (float)(MathHelper.func_181159_b(d8, d6) * 180.0F / (float)Math.PI) - 90.0F;
-                float f1 = (float)(-(MathHelper.func_181159_b(d7, d9) * 180.0F / (float)Math.PI));
+                float f = (float)(MathHelper.func_181159_b(d8, d6) * (180D / Math.PI)) - 90.0F;
+                float f1 = (float)(-(MathHelper.func_181159_b(d7, d9) * (180D / Math.PI)));
                 this.field_82220_d[j] = this.func_82204_b(this.field_82220_d[j], f1, 40.0F);
                 this.field_82221_e[j] = this.func_82204_b(this.field_82221_e[j], f, 10.0F);
             }
@@ -237,56 +225,28 @@
 
         boolean flag = this.func_82205_o();
 
-        for (int l = 0; l < 3; l++)
+        for (int l = 0; l < 3; ++l)
         {
             double d10 = this.func_82214_u(l);
             double d2 = this.func_82208_v(l);
             double d4 = this.func_82213_w(l);
-            this.field_70170_p
-            .func_175688_a(
-                EnumParticleTypes.SMOKE_NORMAL,
-                d10 + this.field_70146_Z.nextGaussian() * 0.3F,
-                d2 + this.field_70146_Z.nextGaussian() * 0.3F,
-                d4 + this.field_70146_Z.nextGaussian() * 0.3F,
-                0.0,
-                0.0,
-                0.0
-            );
+            this.field_70170_p.func_175688_a(EnumParticleTypes.SMOKE_NORMAL, d10 + this.field_70146_Z.nextGaussian() * 0.30000001192092896D, d2 + this.field_70146_Z.nextGaussian() * 0.30000001192092896D, d4 + this.field_70146_Z.nextGaussian() * 0.30000001192092896D, 0.0D, 0.0D, 0.0D);
 
             if (flag && this.field_70170_p.field_73012_v.nextInt(4) == 0)
             {
-                this.field_70170_p
-                .func_175688_a(
-                    EnumParticleTypes.SPELL_MOB,
-                    d10 + this.field_70146_Z.nextGaussian() * 0.3F,
-                    d2 + this.field_70146_Z.nextGaussian() * 0.3F,
-                    d4 + this.field_70146_Z.nextGaussian() * 0.3F,
-                    0.7F,
-                    0.7F,
-                    0.5
-                );
+                this.field_70170_p.func_175688_a(EnumParticleTypes.SPELL_MOB, d10 + this.field_70146_Z.nextGaussian() * 0.30000001192092896D, d2 + this.field_70146_Z.nextGaussian() * 0.30000001192092896D, d4 + this.field_70146_Z.nextGaussian() * 0.30000001192092896D, 0.699999988079071D, 0.699999988079071D, 0.5D);
             }
         }
 
         if (this.func_82212_n() > 0)
         {
-            for (int i1 = 0; i1 < 3; i1++)
+            for (int i1 = 0; i1 < 3; ++i1)
             {
-                this.field_70170_p
-                .func_175688_a(
-                    EnumParticleTypes.SPELL_MOB,
-                    this.field_70165_t + this.field_70146_Z.nextGaussian(),
-                    this.field_70163_u + (double)(this.field_70146_Z.nextFloat() * 3.3F),
-                    this.field_70161_v + this.field_70146_Z.nextGaussian(),
-                    0.7F,
-                    0.7F,
-                    0.9F
-                );
+                this.field_70170_p.func_175688_a(EnumParticleTypes.SPELL_MOB, this.field_70165_t + this.field_70146_Z.nextGaussian(), this.field_70163_u + (double)(this.field_70146_Z.nextFloat() * 3.3F), this.field_70161_v + this.field_70146_Z.nextGaussian(), 0.699999988079071D, 0.699999988079071D, 0.8999999761581421D);
             }
         }
     }
 
-    @Override
     protected void func_70619_bc()
     {
         if (this.func_82212_n() > 0)
@@ -295,16 +255,7 @@
 
             if (j1 <= 0)
             {
-                this.field_70170_p
-                .func_72885_a(
-                    this,
-                    this.field_70165_t,
-                    this.field_70163_u + (double)this.func_70047_e(),
-                    this.field_70161_v,
-                    7.0F,
-                    false,
-                    this.field_70170_p.func_82736_K().func_82766_b("mobGriefing")
-                );
+                this.field_70170_p.func_72885_a(this, this.field_70165_t, this.field_70163_u + (double)this.func_70047_e(), this.field_70161_v, 7.0F, false, net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this));
                 this.field_70170_p.func_175669_a(1023, new BlockPos(this), 0);
             }
 
@@ -319,22 +270,28 @@
         {
             super.func_70619_bc();
 
-            for (int i = 1; i < 3; i++)
+            for (int i = 1; i < 3; ++i)
             {
                 if (this.field_70173_aa >= this.field_82223_h[i - 1])
                 {
                     this.field_82223_h[i - 1] = this.field_70173_aa + 10 + this.field_70146_Z.nextInt(10);
 
-                    if ((this.field_70170_p.func_175659_aa() == EnumDifficulty.NORMAL || this.field_70170_p.func_175659_aa() == EnumDifficulty.HARD)
-                            && this.field_82224_i[i - 1]++ > 15)
+                    if (this.field_70170_p.func_175659_aa() == EnumDifficulty.NORMAL || this.field_70170_p.func_175659_aa() == EnumDifficulty.HARD)
                     {
-                        float f = 10.0F;
-                        float f1 = 5.0F;
-                        double d0 = MathHelper.func_82716_a(this.field_70146_Z, this.field_70165_t - 10.0, this.field_70165_t + 10.0);
-                        double d1 = MathHelper.func_82716_a(this.field_70146_Z, this.field_70163_u - 5.0, this.field_70163_u + 5.0);
-                        double d2 = MathHelper.func_82716_a(this.field_70146_Z, this.field_70161_v - 10.0, this.field_70161_v + 10.0);
-                        this.func_82209_a(i + 1, d0, d1, d2, true);
-                        this.field_82224_i[i - 1] = 0;
+                        int j3 = i - 1;
+                        int k3 = this.field_82224_i[i - 1];
+                        this.field_82224_i[j3] = this.field_82224_i[i - 1] + 1;
+
+                        if (k3 > 15)
+                        {
+                            float f = 10.0F;
+                            float f1 = 5.0F;
+                            double d0 = MathHelper.func_82716_a(this.field_70146_Z, this.field_70165_t - 10.0D, this.field_70165_t + 10.0D);
+                            double d1 = MathHelper.func_82716_a(this.field_70146_Z, this.field_70163_u - 5.0D, this.field_70163_u + 5.0D);
+                            double d2 = MathHelper.func_82716_a(this.field_70146_Z, this.field_70161_v - 10.0D, this.field_70161_v + 10.0D);
+                            this.func_82209_a(i + 1, d0, d1, d2, true);
+                            this.field_82224_i[i - 1] = 0;
+                        }
                     }
 
                     int k1 = this.func_82203_t(i);
@@ -343,31 +300,29 @@
                     {
                         Entity entity = this.field_70170_p.func_73045_a(k1);
 
-                        if (entity == null || !entity.func_70089_S() || this.func_70068_e(entity) > 900.0 || !this.func_70685_l(entity))
-                        {
-                            this.func_82211_c(i, 0);
-                        }
-                        else if (entity instanceof EntityPlayer && ((EntityPlayer)entity).field_71075_bZ.field_75102_a)
-                        {
-                            this.func_82211_c(i, 0);
+                        if (entity != null && entity.func_70089_S() && this.func_70068_e(entity) <= 900.0D && this.func_70685_l(entity))
+                        {
+                            if (entity instanceof EntityPlayer && ((EntityPlayer)entity).field_71075_bZ.field_75102_a)
+                            {
+                                this.func_82211_c(i, 0);
+                            }
+                            else
+                            {
+                                this.func_82216_a(i + 1, (EntityLivingBase)entity);
+                                this.field_82223_h[i - 1] = this.field_70173_aa + 40 + this.field_70146_Z.nextInt(20);
+                                this.field_82224_i[i - 1] = 0;
+                            }
                         }
                         else
                         {
-                            this.func_82216_a(i + 1, (EntityLivingBase)entity);
-                            this.field_82223_h[i - 1] = this.field_70173_aa + 40 + this.field_70146_Z.nextInt(20);
-                            this.field_82224_i[i - 1] = 0;
+                            this.func_82211_c(i, 0);
                         }
                     }
                     else
                     {
-                        List<EntityLivingBase> list = this.field_70170_p
-                                                      .func_175647_a(
-                                                          EntityLivingBase.class,
-                                                          this.func_174813_aQ().func_72314_b(20.0, 8.0, 20.0),
-                                                          Predicates.and(field_82219_bJ, EntitySelectors.field_180132_d)
-                                                      );
+                        List<EntityLivingBase> list = this.field_70170_p.<EntityLivingBase>func_175647_a(EntityLivingBase.class, this.func_174813_aQ().func_72314_b(20.0D, 8.0D, 20.0D), Predicates.and(field_82219_bJ, EntitySelectors.field_180132_d));
 
-                        for (int j2 = 0; j2 < 10 && !list.isEmpty(); j2++)
+                        for (int j2 = 0; j2 < 10 && !list.isEmpty(); ++j2)
                         {
                             EntityLivingBase entitylivingbase = list.get(this.field_70146_Z.nextInt(list.size()));
 
@@ -405,20 +360,20 @@
 
             if (this.field_82222_j > 0)
             {
-                this.field_82222_j--;
+                --this.field_82222_j;
 
-                if (this.field_82222_j == 0 && this.field_70170_p.func_82736_K().func_82766_b("mobGriefing"))
+                if (this.field_82222_j == 0 && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this))
                 {
                     int i1 = MathHelper.func_76128_c(this.field_70163_u);
                     int l1 = MathHelper.func_76128_c(this.field_70165_t);
                     int i2 = MathHelper.func_76128_c(this.field_70161_v);
                     boolean flag = false;
 
-                    for (int k2 = -1; k2 <= 1; k2++)
+                    for (int k2 = -1; k2 <= 1; ++k2)
                     {
-                        for (int l2 = -1; l2 <= 1; l2++)
+                        for (int l2 = -1; l2 <= 1; ++l2)
                         {
-                            for (int j = 0; j <= 3; j++)
+                            for (int j = 0; j <= 3; ++j)
                             {
                                 int i3 = l1 + k2;
                                 int k = i1 + j;
@@ -427,7 +382,7 @@
                                 IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos);
                                 Block block = iblockstate.func_177230_c();
 
-                                if (iblockstate.func_185904_a() != Material.field_151579_a && func_181033_a(block))
+                                if (!block.isAir(iblockstate, this.field_70170_p, blockpos) && block.canEntityDestroy(iblockstate, field_70170_p, blockpos, this) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, iblockstate))
                                 {
                                     flag = this.field_70170_p.func_175655_b(blockpos, true) || flag;
                                 }
@@ -437,7 +392,7 @@
 
                     if (flag)
                     {
-                        this.field_70170_p.func_180498_a(null, 1022, new BlockPos(this), 0);
+                        this.field_70170_p.func_180498_a((EntityPlayer)null, 1022, new BlockPos(this), 0);
                     }
                 }
             }
@@ -453,17 +408,7 @@
 
     public static boolean func_181033_a(Block p_181033_0_)
     {
-        return p_181033_0_ != Blocks.field_150357_h
-               && p_181033_0_ != Blocks.field_150384_bq
-               && p_181033_0_ != Blocks.field_150378_br
-               && p_181033_0_ != Blocks.field_150483_bI
-               && p_181033_0_ != Blocks.field_185776_dc
-               && p_181033_0_ != Blocks.field_185777_dd
-               && p_181033_0_ != Blocks.field_180401_cv
-               && p_181033_0_ != Blocks.field_185779_df
-               && p_181033_0_ != Blocks.field_189881_dj
-               && p_181033_0_ != Blocks.field_180384_M
-               && p_181033_0_ != Blocks.field_185775_db;
+        return p_181033_0_ != Blocks.field_150357_h && p_181033_0_ != Blocks.field_150384_bq && p_181033_0_ != Blocks.field_150378_br && p_181033_0_ != Blocks.field_150483_bI && p_181033_0_ != Blocks.field_185776_dc && p_181033_0_ != Blocks.field_185777_dd && p_181033_0_ != Blocks.field_180401_cv && p_181033_0_ != Blocks.field_185779_df && p_181033_0_ != Blocks.field_189881_dj && p_181033_0_ != Blocks.field_180384_M && p_181033_0_ != Blocks.field_185775_db;
     }
 
     public void func_82206_m()
@@ -472,19 +417,16 @@
         this.func_70606_j(this.func_110138_aP() / 3.0F);
     }
 
-    @Override
     public void func_70110_aj()
     {
     }
 
-    @Override
     public void func_184178_b(EntityPlayerMP p_184178_1_)
     {
         super.func_184178_b(p_184178_1_);
         this.field_184744_bE.func_186760_a(p_184178_1_);
     }
 
-    @Override
     public void func_184203_c(EntityPlayerMP p_184203_1_)
     {
         super.func_184203_c(p_184203_1_);
@@ -499,15 +441,15 @@
         }
         else
         {
-            float f = (this.field_70761_aq + (float)(180 * (p_82214_1_ - 1))) * (float)(Math.PI / 180.0);
+            float f = (this.field_70761_aq + (float)(180 * (p_82214_1_ - 1))) * 0.017453292F;
             float f1 = MathHelper.func_76134_b(f);
-            return this.field_70165_t + (double)f1 * 1.3;
+            return this.field_70165_t + (double)f1 * 1.3D;
         }
     }
 
     private double func_82208_v(int p_82208_1_)
     {
-        return p_82208_1_ <= 0 ? this.field_70163_u + 3.0 : this.field_70163_u + 2.2;
+        return p_82208_1_ <= 0 ? this.field_70163_u + 3.0D : this.field_70163_u + 2.2D;
     }
 
     private double func_82213_w(int p_82213_1_)
@@ -518,9 +460,9 @@
         }
         else
         {
-            float f = (this.field_70761_aq + (float)(180 * (p_82213_1_ - 1))) * (float)(Math.PI / 180.0);
+            float f = (this.field_70761_aq + (float)(180 * (p_82213_1_ - 1))) * 0.017453292F;
             float f1 = MathHelper.func_76126_a(f);
-            return this.field_70161_v + (double)f1 * 1.3;
+            return this.field_70161_v + (double)f1 * 1.3D;
         }
     }
 
@@ -543,18 +485,12 @@
 
     private void func_82216_a(int p_82216_1_, EntityLivingBase p_82216_2_)
     {
-        this.func_82209_a(
-            p_82216_1_,
-            p_82216_2_.field_70165_t,
-            p_82216_2_.field_70163_u + (double)p_82216_2_.func_70047_e() * 0.5,
-            p_82216_2_.field_70161_v,
-            p_82216_1_ == 0 && this.field_70146_Z.nextFloat() < 0.001F
-        );
+        this.func_82209_a(p_82216_1_, p_82216_2_.field_70165_t, p_82216_2_.field_70163_u + (double)p_82216_2_.func_70047_e() * 0.5D, p_82216_2_.field_70161_v, p_82216_1_ == 0 && this.field_70146_Z.nextFloat() < 0.001F);
     }
 
     private void func_82209_a(int p_82209_1_, double p_82209_2_, double p_82209_4_, double p_82209_6_, boolean p_82209_8_)
     {
-        this.field_70170_p.func_180498_a(null, 1024, new BlockPos(this), 0);
+        this.field_70170_p.func_180498_a((EntityPlayer)null, 1024, new BlockPos(this), 0);
         double d0 = this.func_82214_u(p_82209_1_);
         double d1 = this.func_82208_v(p_82209_1_);
         double d2 = this.func_82213_w(p_82209_1_);
@@ -574,66 +510,63 @@
         this.field_70170_p.func_72838_d(entitywitherskull);
     }
 
-    @Override
     public void func_82196_d(EntityLivingBase p_82196_1_, float p_82196_2_)
     {
         this.func_82216_a(0, p_82196_1_);
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         if (this.func_180431_b(p_70097_1_))
         {
             return false;
         }
-        else if (p_70097_1_ == DamageSource.field_76369_e || p_70097_1_.func_76346_g() instanceof EntityWither)
-        {
-            return false;
-        }
-        else if (this.func_82212_n() > 0 && p_70097_1_ != DamageSource.field_76380_i)
-        {
-            return false;
-        }
-        else
-        {
-            if (this.func_82205_o())
-            {
-                Entity entity = p_70097_1_.func_76364_f();
-
-                if (entity instanceof EntityArrow)
-                {
-                    return false;
-                }
-            }
-
-            Entity entity1 = p_70097_1_.func_76346_g();
-
-            if (entity1 != null
-                    && !(entity1 instanceof EntityPlayer)
-                    && entity1 instanceof EntityLivingBase
-                    && ((EntityLivingBase)entity1).func_70668_bt() == this.func_70668_bt())
+        else if (p_70097_1_ != DamageSource.field_76369_e && !(p_70097_1_.func_76346_g() instanceof EntityWither))
+        {
+            if (this.func_82212_n() > 0 && p_70097_1_ != DamageSource.field_76380_i)
             {
                 return false;
             }
             else
             {
-                if (this.field_82222_j <= 0)
-                {
-                    this.field_82222_j = 20;
-                }
-
-                for (int i = 0; i < this.field_82224_i.length; i++)
-                {
-                    this.field_82224_i[i] = this.field_82224_i[i] + 3;
-                }
-
-                return super.func_70097_a(p_70097_1_, p_70097_2_);
+                if (this.func_82205_o())
+                {
+                    Entity entity = p_70097_1_.func_76364_f();
+
+                    if (entity instanceof EntityArrow)
+                    {
+                        return false;
+                    }
+                }
+
+                Entity entity1 = p_70097_1_.func_76346_g();
+
+                if (entity1 != null && !(entity1 instanceof EntityPlayer) && entity1 instanceof EntityLivingBase && ((EntityLivingBase)entity1).func_70668_bt() == this.func_70668_bt())
+                {
+                    return false;
+                }
+                else
+                {
+                    if (this.field_82222_j <= 0)
+                    {
+                        this.field_82222_j = 20;
+                    }
+
+                    for (int i = 0; i < this.field_82224_i.length; ++i)
+                    {
+                        this.field_82224_i[i] += 3;
+                    }
+
+                    return super.func_70097_a(p_70097_1_, p_70097_2_);
+                }
             }
         }
+        else
+        {
+            return false;
+        }
     }
 
-    @Override
     protected void func_70628_a(boolean p_70628_1_, int p_70628_2_)
     {
         EntityItem entityitem = this.func_145779_a(Items.field_151156_bN, 1);
@@ -644,37 +577,32 @@
         }
     }
 
-    @Override
     protected void func_70623_bb()
     {
         this.field_70708_bq = 0;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public int func_70070_b()
     {
         return 15728880;
     }
 
-    @Override
     public void func_180430_e(float p_180430_1_, float p_180430_2_)
     {
     }
 
-    @Override
     public void func_70690_d(PotionEffect p_70690_1_)
     {
     }
 
-    @Override
     protected void func_110147_ax()
     {
         super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(300.0);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.6F);
-        this.func_110148_a(SharedMonsterAttributes.field_111265_b).func_111128_a(40.0);
-        this.func_110148_a(SharedMonsterAttributes.field_188791_g).func_111128_a(4.0);
+        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(300.0D);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.6000000238418579D);
+        this.func_110148_a(SharedMonsterAttributes.field_111265_b).func_111128_a(40.0D);
+        this.func_110148_a(SharedMonsterAttributes.field_188791_g).func_111128_a(4.0D);
     }
 
     @SideOnly(Side.CLIENT)
@@ -691,22 +619,22 @@
 
     public int func_82212_n()
     {
-        return this.field_70180_af.func_187225_a(field_184746_bw);
+        return ((Integer)this.field_70180_af.func_187225_a(field_184746_bw)).intValue();
     }
 
     public void func_82215_s(int p_82215_1_)
     {
-        this.field_70180_af.func_187227_b(field_184746_bw, p_82215_1_);
+        this.field_70180_af.func_187227_b(field_184746_bw, Integer.valueOf(p_82215_1_));
     }
 
     public int func_82203_t(int p_82203_1_)
     {
-        return this.field_70180_af.func_187225_a(field_184745_bv[p_82203_1_]);
+        return ((Integer)this.field_70180_af.func_187225_a(field_184745_bv[p_82203_1_])).intValue();
     }
 
     public void func_82211_c(int p_82211_1_, int p_82211_2_)
     {
-        this.field_70180_af.func_187227_b(field_184745_bv[p_82211_1_], p_82211_2_);
+        this.field_70180_af.func_187227_b(field_184745_bv[p_82211_1_], Integer.valueOf(p_82211_2_));
     }
 
     public boolean func_82205_o()
@@ -714,25 +642,21 @@
         return this.func_110143_aJ() <= this.func_110138_aP() / 2.0F;
     }
 
-    @Override
     public EnumCreatureAttribute func_70668_bt()
     {
         return EnumCreatureAttribute.UNDEAD;
     }
 
-    @Override
     protected boolean func_184228_n(Entity p_184228_1_)
     {
         return false;
     }
 
-    @Override
     public boolean func_184222_aU()
     {
         return false;
     }
 
-    @Override
     public void func_184724_a(boolean p_184724_1_)
     {
     }
@@ -744,7 +668,6 @@
             this.func_75248_a(7);
         }
 
-        @Override
         public boolean func_75250_a()
         {
             return EntityWither.this.func_82212_n() > 0;
