--- before/net/minecraft/entity/boss/EntityDragon.java
+++ after/net/minecraft/entity/boss/EntityDragon.java
@@ -53,18 +53,18 @@
 public class EntityDragon extends EntityLiving implements IEntityMultiPart, IMob
 {
     private static final Logger field_184675_bH = LogManager.getLogger();
-    public static final DataParameter<Integer> field_184674_a = EntityDataManager.func_187226_a(EntityDragon.class, DataSerializers.field_187192_b);
+    public static final DataParameter<Integer> field_184674_a = EntityDataManager.<Integer>func_187226_a(EntityDragon.class, DataSerializers.field_187192_b);
     public double[][] field_70979_e = new double[64][3];
     public int field_70976_f = -1;
     public MultiPartEntityPart[] field_70977_g;
-    public MultiPartEntityPart field_70986_h;
-    public MultiPartEntityPart field_184673_bv;
-    public MultiPartEntityPart field_70987_i;
-    public MultiPartEntityPart field_70985_j;
-    public MultiPartEntityPart field_70984_by;
-    public MultiPartEntityPart field_70982_bz;
-    public MultiPartEntityPart field_70983_bA;
-    public MultiPartEntityPart field_70990_bB;
+    public MultiPartEntityPart field_70986_h = new MultiPartEntityPart(this, "head", 6.0F, 6.0F);
+    public MultiPartEntityPart field_184673_bv = new MultiPartEntityPart(this, "neck", 6.0F, 6.0F);
+    public MultiPartEntityPart field_70987_i = new MultiPartEntityPart(this, "body", 8.0F, 8.0F);
+    public MultiPartEntityPart field_70985_j = new MultiPartEntityPart(this, "tail", 4.0F, 4.0F);
+    public MultiPartEntityPart field_70984_by = new MultiPartEntityPart(this, "tail", 4.0F, 4.0F);
+    public MultiPartEntityPart field_70982_bz = new MultiPartEntityPart(this, "tail", 4.0F, 4.0F);
+    public MultiPartEntityPart field_70983_bA = new MultiPartEntityPart(this, "wing", 4.0F, 4.0F);
+    public MultiPartEntityPart field_70990_bB = new MultiPartEntityPart(this, "wing", 4.0F, 4.0F);
     public float field_70991_bC;
     public float field_70988_bD;
     public boolean field_70994_bF;
@@ -81,25 +81,7 @@
     public EntityDragon(World p_i1700_1_)
     {
         super(p_i1700_1_);
-        this.field_70986_h = new MultiPartEntityPart(this, "head", 6.0F, 6.0F);
-        this.field_184673_bv = new MultiPartEntityPart(this, "neck", 6.0F, 6.0F);
-        this.field_70987_i = new MultiPartEntityPart(this, "body", 8.0F, 8.0F);
-        this.field_70985_j = new MultiPartEntityPart(this, "tail", 4.0F, 4.0F);
-        this.field_70984_by = new MultiPartEntityPart(this, "tail", 4.0F, 4.0F);
-        this.field_70982_bz = new MultiPartEntityPart(this, "tail", 4.0F, 4.0F);
-        this.field_70983_bA = new MultiPartEntityPart(this, "wing", 4.0F, 4.0F);
-        this.field_70990_bB = new MultiPartEntityPart(this, "wing", 4.0F, 4.0F);
-        this.field_70977_g = new MultiPartEntityPart[]
-        {
-            this.field_70986_h,
-            this.field_184673_bv,
-            this.field_70987_i,
-            this.field_70985_j,
-            this.field_70984_by,
-            this.field_70982_bz,
-            this.field_70983_bA,
-            this.field_70990_bB
-        };
+        this.field_70977_g = new MultiPartEntityPart[] {this.field_70986_h, this.field_184673_bv, this.field_70987_i, this.field_70985_j, this.field_70984_by, this.field_70982_bz, this.field_70983_bA, this.field_70990_bB};
         this.func_70606_j(this.func_110138_aP());
         this.func_70105_a(16.0F, 8.0F);
         this.field_70145_X = true;
@@ -119,18 +101,16 @@
         this.field_184677_bJ = new PhaseManager(this);
     }
 
-    @Override
     protected void func_110147_ax()
     {
         super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(200.0);
+        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(200.0D);
     }
 
-    @Override
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.func_184212_Q().func_187214_a(field_184674_a, PhaseList.field_188751_k.func_188740_b());
+        this.func_184212_Q().func_187214_a(field_184674_a, Integer.valueOf(PhaseList.field_188751_k.func_188740_b()));
     }
 
     public double[] func_70974_a(int p_70974_1_, float p_70974_2_)
@@ -154,7 +134,6 @@
         return adouble;
     }
 
-    @Override
     public void func_70636_d()
     {
         if (this.field_70170_p.field_72995_K)
@@ -163,37 +142,17 @@
 
             if (!this.func_174814_R())
             {
-                float f = MathHelper.func_76134_b(this.field_70988_bD * (float)(Math.PI * 2));
-                float f1 = MathHelper.func_76134_b(this.field_70991_bC * (float)(Math.PI * 2));
+                float f = MathHelper.func_76134_b(this.field_70988_bD * ((float)Math.PI * 2F));
+                float f1 = MathHelper.func_76134_b(this.field_70991_bC * ((float)Math.PI * 2F));
 
                 if (f1 <= -0.3F && f >= -0.3F)
                 {
-                    this.field_70170_p
-                    .func_184134_a(
-                        this.field_70165_t,
-                        this.field_70163_u,
-                        this.field_70161_v,
-                        SoundEvents.field_187524_aN,
-                        this.func_184176_by(),
-                        5.0F,
-                        0.8F + this.field_70146_Z.nextFloat() * 0.3F,
-                        false
-                    );
+                    this.field_70170_p.func_184134_a(this.field_70165_t, this.field_70163_u, this.field_70161_v, SoundEvents.field_187524_aN, this.func_184176_by(), 5.0F, 0.8F + this.field_70146_Z.nextFloat() * 0.3F, false);
                 }
 
                 if (!this.field_184677_bJ.func_188756_a().func_188654_a() && --this.field_184678_bK < 0)
                 {
-                    this.field_70170_p
-                    .func_184134_a(
-                        this.field_70165_t,
-                        this.field_70163_u,
-                        this.field_70161_v,
-                        SoundEvents.field_187525_aO,
-                        this.func_184176_by(),
-                        2.5F,
-                        0.8F + this.field_70146_Z.nextFloat() * 0.3F,
-                        false
-                    );
+                    this.field_70170_p.func_184134_a(this.field_70165_t, this.field_70163_u, this.field_70161_v, SoundEvents.field_187525_aO, this.func_184176_by(), 2.5F, 0.8F + this.field_70146_Z.nextFloat() * 0.3F, false);
                     this.field_184678_bK = 200 + this.field_70146_Z.nextInt(200);
                 }
             }
@@ -206,22 +165,13 @@
             float f12 = (this.field_70146_Z.nextFloat() - 0.5F) * 8.0F;
             float f13 = (this.field_70146_Z.nextFloat() - 0.5F) * 4.0F;
             float f15 = (this.field_70146_Z.nextFloat() - 0.5F) * 8.0F;
-            this.field_70170_p
-            .func_175688_a(
-                EnumParticleTypes.EXPLOSION_LARGE,
-                this.field_70165_t + (double)f12,
-                this.field_70163_u + 2.0 + (double)f13,
-                this.field_70161_v + (double)f15,
-                0.0,
-                0.0,
-                0.0
-            );
+            this.field_70170_p.func_175688_a(EnumParticleTypes.EXPLOSION_LARGE, this.field_70165_t + (double)f12, this.field_70163_u + 2.0D + (double)f13, this.field_70161_v + (double)f15, 0.0D, 0.0D, 0.0D);
         }
         else
         {
             this.func_70969_j();
             float f11 = 0.2F / (MathHelper.func_76133_a(this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y) * 10.0F + 1.0F);
-            f11 *= (float)Math.pow(2.0, this.field_70181_x);
+            f11 = f11 * (float)Math.pow(2.0D, this.field_70181_x);
 
             if (this.field_184677_bJ.func_188756_a().func_188654_a())
             {
@@ -299,44 +249,34 @@
                         double d3 = d6 * d6 + d7 * d7 + d8 * d8;
                         float f5 = iphase.func_188651_f();
                         d7 = MathHelper.func_151237_a(d7 / (double)MathHelper.func_76133_a(d6 * d6 + d8 * d8), (double)(-f5), (double)f5);
-                        this.field_70181_x += d7 * 0.1F;
+                        this.field_70181_x += d7 * 0.10000000149011612D;
                         this.field_70177_z = MathHelper.func_76142_g(this.field_70177_z);
-                        double d4 = MathHelper.func_151237_a(
-                                        MathHelper.func_76138_g(180.0 - MathHelper.func_181159_b(d6, d8) * 180.0F / (float)Math.PI - (double)this.field_70177_z), -50.0, 50.0
-                                    );
-                        Vec3d vec3d1 = new Vec3d(
-                            vec3d.field_72450_a - this.field_70165_t, vec3d.field_72448_b - this.field_70163_u, vec3d.field_72449_c - this.field_70161_v
-                        )
-                        .func_72432_b();
-                        Vec3d vec3d2 = new Vec3d(
-                            (double)MathHelper.func_76126_a(this.field_70177_z * (float)(Math.PI / 180.0)),
-                            this.field_70181_x,
-                            (double)(-MathHelper.func_76134_b(this.field_70177_z * (float)(Math.PI / 180.0)))
-                        )
-                        .func_72432_b();
+                        double d4 = MathHelper.func_151237_a(MathHelper.func_76138_g(180.0D - MathHelper.func_181159_b(d6, d8) * (180D / Math.PI) - (double)this.field_70177_z), -50.0D, 50.0D);
+                        Vec3d vec3d1 = (new Vec3d(vec3d.field_72450_a - this.field_70165_t, vec3d.field_72448_b - this.field_70163_u, vec3d.field_72449_c - this.field_70161_v)).func_72432_b();
+                        Vec3d vec3d2 = (new Vec3d((double)MathHelper.func_76126_a(this.field_70177_z * 0.017453292F), this.field_70181_x, (double)(-MathHelper.func_76134_b(this.field_70177_z * 0.017453292F)))).func_72432_b();
                         float f7 = Math.max(((float)vec3d2.func_72430_b(vec3d1) + 0.5F) / 1.5F, 0.0F);
                         this.field_70704_bt *= 0.8F;
                         this.field_70704_bt = (float)((double)this.field_70704_bt + d4 * (double)iphase.func_188653_h());
                         this.field_70177_z += this.field_70704_bt * 0.1F;
-                        float f8 = (float)(2.0 / (d3 + 1.0));
+                        float f8 = (float)(2.0D / (d3 + 1.0D));
                         float f9 = 0.06F;
                         this.func_191958_b(0.0F, 0.0F, -1.0F, 0.06F * (f7 * f8 + (1.0F - f8)));
 
                         if (this.field_70994_bF)
                         {
-                            this.func_70091_d(MoverType.SELF, this.field_70159_w * 0.8F, this.field_70181_x * 0.8F, this.field_70179_y * 0.8F);
+                            this.func_70091_d(MoverType.SELF, this.field_70159_w * 0.800000011920929D, this.field_70181_x * 0.800000011920929D, this.field_70179_y * 0.800000011920929D);
                         }
                         else
                         {
                             this.func_70091_d(MoverType.SELF, this.field_70159_w, this.field_70181_x, this.field_70179_y);
                         }
 
-                        Vec3d vec3d3 = new Vec3d(this.field_70159_w, this.field_70181_x, this.field_70179_y).func_72432_b();
+                        Vec3d vec3d3 = (new Vec3d(this.field_70159_w, this.field_70181_x, this.field_70179_y)).func_72432_b();
                         float f10 = ((float)vec3d3.func_72430_b(vec3d2) + 1.0F) / 2.0F;
                         f10 = 0.8F + 0.15F * f10;
                         this.field_70159_w *= (double)f10;
                         this.field_70179_y *= (double)f10;
-                        this.field_70181_x *= 0.91F;
+                        this.field_70181_x *= 0.9100000262260437D;
                     }
                 }
 
@@ -364,56 +304,35 @@
                     avec3d[j] = new Vec3d(this.field_70977_g[j].field_70165_t, this.field_70977_g[j].field_70163_u, this.field_70977_g[j].field_70161_v);
                 }
 
-                float f14 = (float)(this.func_70974_a(5, 1.0F)[1] - this.func_70974_a(10, 1.0F)[1]) * 10.0F * (float)(Math.PI / 180.0);
+                float f14 = (float)(this.func_70974_a(5, 1.0F)[1] - this.func_70974_a(10, 1.0F)[1]) * 10.0F * 0.017453292F;
                 float f16 = MathHelper.func_76134_b(f14);
                 float f2 = MathHelper.func_76126_a(f14);
-                float f17 = this.field_70177_z * (float)(Math.PI / 180.0);
+                float f17 = this.field_70177_z * 0.017453292F;
                 float f3 = MathHelper.func_76126_a(f17);
                 float f18 = MathHelper.func_76134_b(f17);
                 this.field_70987_i.func_70071_h_();
-                this.field_70987_i
-                .func_70012_b(this.field_70165_t + (double)(f3 * 0.5F), this.field_70163_u, this.field_70161_v - (double)(f18 * 0.5F), 0.0F, 0.0F);
+                this.field_70987_i.func_70012_b(this.field_70165_t + (double)(f3 * 0.5F), this.field_70163_u, this.field_70161_v - (double)(f18 * 0.5F), 0.0F, 0.0F);
                 this.field_70983_bA.func_70071_h_();
-                this.field_70983_bA
-                .func_70012_b(this.field_70165_t + (double)(f18 * 4.5F), this.field_70163_u + 2.0, this.field_70161_v + (double)(f3 * 4.5F), 0.0F, 0.0F);
+                this.field_70983_bA.func_70012_b(this.field_70165_t + (double)(f18 * 4.5F), this.field_70163_u + 2.0D, this.field_70161_v + (double)(f3 * 4.5F), 0.0F, 0.0F);
                 this.field_70990_bB.func_70071_h_();
-                this.field_70990_bB
-                .func_70012_b(this.field_70165_t - (double)(f18 * 4.5F), this.field_70163_u + 2.0, this.field_70161_v - (double)(f3 * 4.5F), 0.0F, 0.0F);
+                this.field_70990_bB.func_70012_b(this.field_70165_t - (double)(f18 * 4.5F), this.field_70163_u + 2.0D, this.field_70161_v - (double)(f3 * 4.5F), 0.0F, 0.0F);
 
                 if (!this.field_70170_p.field_72995_K && this.field_70737_aN == 0)
                 {
-                    this.func_70970_a(
-                        this.field_70170_p.func_72839_b(this, this.field_70983_bA.func_174813_aQ().func_72314_b(4.0, 2.0, 4.0).func_72317_d(0.0, -2.0, 0.0))
-                    );
-                    this.func_70970_a(
-                        this.field_70170_p.func_72839_b(this, this.field_70990_bB.func_174813_aQ().func_72314_b(4.0, 2.0, 4.0).func_72317_d(0.0, -2.0, 0.0))
-                    );
-                    this.func_70971_b(this.field_70170_p.func_72839_b(this, this.field_70986_h.func_174813_aQ().func_186662_g(1.0)));
-                    this.func_70971_b(this.field_70170_p.func_72839_b(this, this.field_184673_bv.func_174813_aQ().func_186662_g(1.0)));
+                    this.func_70970_a(this.field_70170_p.func_72839_b(this, this.field_70983_bA.func_174813_aQ().func_72314_b(4.0D, 2.0D, 4.0D).func_72317_d(0.0D, -2.0D, 0.0D)));
+                    this.func_70970_a(this.field_70170_p.func_72839_b(this, this.field_70990_bB.func_174813_aQ().func_72314_b(4.0D, 2.0D, 4.0D).func_72317_d(0.0D, -2.0D, 0.0D)));
+                    this.func_70971_b(this.field_70170_p.func_72839_b(this, this.field_70986_h.func_174813_aQ().func_186662_g(1.0D)));
+                    this.func_70971_b(this.field_70170_p.func_72839_b(this, this.field_184673_bv.func_174813_aQ().func_186662_g(1.0D)));
                 }
 
                 double[] adouble = this.func_70974_a(5, 1.0F);
-                float f19 = MathHelper.func_76126_a(this.field_70177_z * (float)(Math.PI / 180.0) - this.field_70704_bt * 0.01F);
-                float f4 = MathHelper.func_76134_b(this.field_70177_z * (float)(Math.PI / 180.0) - this.field_70704_bt * 0.01F);
+                float f19 = MathHelper.func_76126_a(this.field_70177_z * 0.017453292F - this.field_70704_bt * 0.01F);
+                float f4 = MathHelper.func_76134_b(this.field_70177_z * 0.017453292F - this.field_70704_bt * 0.01F);
                 this.field_70986_h.func_70071_h_();
                 this.field_184673_bv.func_70071_h_();
                 float f20 = this.func_184662_q(1.0F);
-                this.field_70986_h
-                .func_70012_b(
-                    this.field_70165_t + (double)(f19 * 6.5F * f16),
-                    this.field_70163_u + (double)f20 + (double)(f2 * 6.5F),
-                    this.field_70161_v - (double)(f4 * 6.5F * f16),
-                    0.0F,
-                    0.0F
-                );
-                this.field_184673_bv
-                .func_70012_b(
-                    this.field_70165_t + (double)(f19 * 5.5F * f16),
-                    this.field_70163_u + (double)f20 + (double)(f2 * 5.5F),
-                    this.field_70161_v - (double)(f4 * 5.5F * f16),
-                    0.0F,
-                    0.0F
-                );
+                this.field_70986_h.func_70012_b(this.field_70165_t + (double)(f19 * 6.5F * f16), this.field_70163_u + (double)f20 + (double)(f2 * 6.5F), this.field_70161_v - (double)(f4 * 6.5F * f16), 0.0F, 0.0F);
+                this.field_184673_bv.func_70012_b(this.field_70165_t + (double)(f19 * 5.5F * f16), this.field_70163_u + (double)f20 + (double)(f2 * 5.5F), this.field_70161_v - (double)(f4 * 5.5F * f16), 0.0F, 0.0F);
 
                 for (int k = 0; k < 3; ++k)
                 {
@@ -435,26 +354,18 @@
                     }
 
                     double[] adouble1 = this.func_70974_a(12 + k * 2, 1.0F);
-                    float f21 = this.field_70177_z * (float)(Math.PI / 180.0) + this.func_70973_b(adouble1[0] - adouble[0]) * (float)(Math.PI / 180.0);
+                    float f21 = this.field_70177_z * 0.017453292F + this.func_70973_b(adouble1[0] - adouble[0]) * 0.017453292F;
                     float f6 = MathHelper.func_76126_a(f21);
                     float f22 = MathHelper.func_76134_b(f21);
                     float f23 = 1.5F;
                     float f24 = (float)(k + 1) * 2.0F;
                     multipartentitypart.func_70071_h_();
-                    multipartentitypart.func_70012_b(
-                        this.field_70165_t - (double)((f3 * 1.5F + f6 * f24) * f16),
-                        this.field_70163_u + (adouble1[1] - adouble[1]) - (double)((f24 + 1.5F) * f2) + 1.5,
-                        this.field_70161_v + (double)((f18 * 1.5F + f22 * f24) * f16),
-                        0.0F,
-                        0.0F
-                    );
+                    multipartentitypart.func_70012_b(this.field_70165_t - (double)((f3 * 1.5F + f6 * f24) * f16), this.field_70163_u + (adouble1[1] - adouble[1]) - (double)((f24 + 1.5F) * f2) + 1.5D, this.field_70161_v + (double)((f18 * 1.5F + f22 * f24) * f16), 0.0F, 0.0F);
                 }
 
                 if (!this.field_70170_p.field_72995_K)
                 {
-                    this.field_70994_bF = this.func_70972_a(this.field_70986_h.func_174813_aQ())
-                                          | this.func_70972_a(this.field_184673_bv.func_174813_aQ())
-                                          | this.func_70972_a(this.field_70987_i.func_174813_aQ());
+                    this.field_70994_bF = this.func_70972_a(this.field_70986_h.func_174813_aQ()) | this.func_70972_a(this.field_184673_bv.func_174813_aQ()) | this.func_70972_a(this.field_70987_i.func_174813_aQ());
 
                     if (this.field_184676_bI != null)
                     {
@@ -478,7 +389,7 @@
 
         if (this.field_184677_bJ.func_188756_a().func_188654_a())
         {
-            d0 = -1.0;
+            d0 = -1.0D;
         }
         else
         {
@@ -506,7 +417,7 @@
 
         if (this.field_70146_Z.nextInt(10) == 0)
         {
-            List<EntityEnderCrystal> list = this.field_70170_p.func_72872_a(EntityEnderCrystal.class, this.func_174813_aQ().func_186662_g(32.0));
+            List<EntityEnderCrystal> list = this.field_70170_p.<EntityEnderCrystal>func_72872_a(EntityEnderCrystal.class, this.func_174813_aQ().func_186662_g(32.0D));
             EntityEnderCrystal entityendercrystal = null;
             double d0 = Double.MAX_VALUE;
 
@@ -527,8 +438,8 @@
 
     private void func_70970_a(List<Entity> p_70970_1_)
     {
-        double d0 = (this.field_70987_i.func_174813_aQ().field_72340_a + this.field_70987_i.func_174813_aQ().field_72336_d) / 2.0;
-        double d1 = (this.field_70987_i.func_174813_aQ().field_72339_c + this.field_70987_i.func_174813_aQ().field_72334_f) / 2.0;
+        double d0 = (this.field_70987_i.func_174813_aQ().field_72340_a + this.field_70987_i.func_174813_aQ().field_72336_d) / 2.0D;
+        double d1 = (this.field_70987_i.func_174813_aQ().field_72339_c + this.field_70987_i.func_174813_aQ().field_72334_f) / 2.0D;
 
         for (Entity entity : p_70970_1_)
         {
@@ -537,7 +448,7 @@
                 double d2 = entity.field_70165_t - d0;
                 double d3 = entity.field_70161_v - d1;
                 double d4 = d2 * d2 + d3 * d3;
-                entity.func_70024_g(d2 / d4 * 4.0, 0.2F, d3 / d4 * 4.0);
+                entity.func_70024_g(d2 / d4 * 4.0D, 0.20000000298023224D, d3 / d4 * 4.0D);
 
                 if (!this.field_184677_bJ.func_188756_a().func_188654_a() && ((EntityLivingBase)entity).func_142015_aE() < entity.field_70173_aa - 2)
                 {
@@ -588,28 +499,22 @@
                     IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos);
                     Block block = iblockstate.func_177230_c();
 
-                    if (iblockstate.func_185904_a() != Material.field_151579_a && iblockstate.func_185904_a() != Material.field_151581_o)
+                    if (!block.isAir(iblockstate, this.field_70170_p, blockpos) && iblockstate.func_185904_a() != Material.field_151581_o)
                     {
-                        if (!this.field_70170_p.func_82736_K().func_82766_b("mobGriefing"))
-                        {
-                            flag = true;
-                        }
-                        else if (block == Blocks.field_180401_cv
-                                 || block == Blocks.field_150343_Z
-                                 || block == Blocks.field_150377_bs
-                                 || block == Blocks.field_150357_h
-                                 || block == Blocks.field_150384_bq
-                                 || block == Blocks.field_150378_br)
-                        {
-                            flag = true;
-                        }
-                        else if (block != Blocks.field_150483_bI
-                                 && block != Blocks.field_185776_dc
-                                 && block != Blocks.field_185777_dd
-                                 && block != Blocks.field_150411_aY
-                                 && block != Blocks.field_185775_db)
-                        {
-                            flag1 = this.field_70170_p.func_175698_g(blockpos) || flag1;
+                        if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this))
+                        {
+                            flag = true;
+                        }
+                        else if (block.canEntityDestroy(iblockstate, this.field_70170_p, blockpos, this) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, iblockstate))
+                        {
+                            if (block != Blocks.field_150483_bI && block != Blocks.field_185776_dc && block != Blocks.field_185777_dd && block != Blocks.field_150411_aY && block != Blocks.field_185775_db)
+                            {
+                                flag1 = this.field_70170_p.func_175698_g(blockpos) || flag1;
+                            }
+                            else
+                            {
+                                flag = true;
+                            }
                         }
                         else
                         {
@@ -625,13 +530,12 @@
             double d0 = p_70972_1_.field_72340_a + (p_70972_1_.field_72336_d - p_70972_1_.field_72340_a) * (double)this.field_70146_Z.nextFloat();
             double d1 = p_70972_1_.field_72338_b + (p_70972_1_.field_72337_e - p_70972_1_.field_72338_b) * (double)this.field_70146_Z.nextFloat();
             double d2 = p_70972_1_.field_72339_c + (p_70972_1_.field_72334_f - p_70972_1_.field_72339_c) * (double)this.field_70146_Z.nextFloat();
-            this.field_70170_p.func_175688_a(EnumParticleTypes.EXPLOSION_LARGE, d0, d1, d2, 0.0, 0.0, 0.0);
+            this.field_70170_p.func_175688_a(EnumParticleTypes.EXPLOSION_LARGE, d0, d1, d2, 0.0D, 0.0D, 0.0D);
         }
 
         return flag;
     }
 
-    @Override
     public boolean func_70965_a(MultiPartEntityPart p_70965_1_, DamageSource p_70965_2_, float p_70965_3_)
     {
         p_70965_3_ = this.field_184677_bJ.func_188756_a().func_188656_a(p_70965_1_, p_70965_2_, p_70965_3_);
@@ -674,7 +578,6 @@
         }
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         if (p_70097_1_ instanceof EntityDamageSource && ((EntityDamageSource)p_70097_1_).func_180139_w())
@@ -690,7 +593,6 @@
         return super.func_70097_a(p_82195_1_, p_82195_2_);
     }
 
-    @Override
     public void func_174812_G()
     {
         this.func_70106_y();
@@ -702,7 +604,6 @@
         }
     }
 
-    @Override
     protected void func_70609_aI()
     {
         if (this.field_184676_bI != null)
@@ -717,16 +618,7 @@
             float f = (this.field_70146_Z.nextFloat() - 0.5F) * 8.0F;
             float f1 = (this.field_70146_Z.nextFloat() - 0.5F) * 4.0F;
             float f2 = (this.field_70146_Z.nextFloat() - 0.5F) * 8.0F;
-            this.field_70170_p
-            .func_175688_a(
-                EnumParticleTypes.EXPLOSION_HUGE,
-                this.field_70165_t + (double)f,
-                this.field_70163_u + 2.0 + (double)f1,
-                this.field_70161_v + (double)f2,
-                0.0,
-                0.0,
-                0.0
-            );
+            this.field_70170_p.func_175688_a(EnumParticleTypes.EXPLOSION_HUGE, this.field_70165_t + (double)f, this.field_70163_u + 2.0D + (double)f1, this.field_70161_v + (double)f2, 0.0D, 0.0D, 0.0D);
         }
 
         boolean flag = this.field_70170_p.func_82736_K().func_82766_b("doMobLoot");
@@ -741,7 +633,9 @@
         {
             if (this.field_70995_bG > 150 && this.field_70995_bG % 5 == 0 && flag)
             {
-                this.func_184668_a(MathHelper.func_76141_d((float)i * 0.08F));
+                int b = MathHelper.func_76141_d((float)i * 0.08F);
+                b = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.field_70717_bb, b);
+                if(b>0) this.func_184668_a(b);
             }
 
             if (this.field_70995_bG == 1)
@@ -750,7 +644,7 @@
             }
         }
 
-        this.func_70091_d(MoverType.SELF, 0.0, 0.1F, 0.0);
+        this.func_70091_d(MoverType.SELF, 0.0D, 0.10000000149011612D, 0.0D);
         this.field_70177_z += 20.0F;
         this.field_70761_aq = this.field_70177_z;
 
@@ -758,7 +652,9 @@
         {
             if (flag)
             {
-                this.func_184668_a(MathHelper.func_76141_d((float)i * 0.2F));
+                int b = MathHelper.func_76141_d((float)i * 0.2F);
+                b = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.field_70717_bb, b);
+                if(b>0) this.func_184668_a(b);
             }
 
             if (this.field_184676_bI != null)
@@ -792,21 +688,21 @@
 
                 if (i < 12)
                 {
-                    l = (int)(60.0F * MathHelper.func_76134_b(2.0F * ((float) - Math.PI + (float)(Math.PI / 12) * (float)i)));
-                    i1 = (int)(60.0F * MathHelper.func_76126_a(2.0F * ((float) - Math.PI + (float)(Math.PI / 12) * (float)i)));
+                    l = (int)(60.0F * MathHelper.func_76134_b(2.0F * (-(float)Math.PI + 0.2617994F * (float)i)));
+                    i1 = (int)(60.0F * MathHelper.func_76126_a(2.0F * (-(float)Math.PI + 0.2617994F * (float)i)));
                 }
                 else if (i < 20)
                 {
                     int lvt_3_1_ = i - 12;
-                    l = (int)(40.0F * MathHelper.func_76134_b(2.0F * ((float) - Math.PI + (float)(Math.PI / 8) * (float)lvt_3_1_)));
-                    i1 = (int)(40.0F * MathHelper.func_76126_a(2.0F * ((float) - Math.PI + (float)(Math.PI / 8) * (float)lvt_3_1_)));
+                    l = (int)(40.0F * MathHelper.func_76134_b(2.0F * (-(float)Math.PI + 0.3926991F * (float)lvt_3_1_)));
+                    i1 = (int)(40.0F * MathHelper.func_76126_a(2.0F * (-(float)Math.PI + 0.3926991F * (float)lvt_3_1_)));
                     j += 10;
                 }
                 else
                 {
                     int k1 = i - 20;
-                    l = (int)(20.0F * MathHelper.func_76134_b(2.0F * ((float) - Math.PI + (float)(Math.PI / 4) * (float)k1)));
-                    i1 = (int)(20.0F * MathHelper.func_76126_a(2.0F * ((float) - Math.PI + (float)(Math.PI / 4) * (float)k1)));
+                    l = (int)(20.0F * MathHelper.func_76134_b(2.0F * (-(float)Math.PI + ((float)Math.PI / 4F) * (float)k1)));
+                    i1 = (int)(20.0F * MathHelper.func_76126_a(2.0F * (-(float)Math.PI + ((float)Math.PI / 4F) * (float)k1)));
                 }
 
                 int j1 = Math.max(this.field_70170_p.func_181545_F() + 10, this.field_70170_p.func_175672_r(new BlockPos(l, 0, i1)).func_177956_o() + j);
@@ -969,7 +865,7 @@
         }
         else
         {
-            field_184675_bH.debug("Failed to find path from {} to {}", p_184666_1_, p_184666_2_);
+            field_184675_bH.debug("Failed to find path from {} to {}", Integer.valueOf(p_184666_1_), Integer.valueOf(p_184666_2_));
 
             if (p_184666_3_ != null)
             {
@@ -1008,14 +904,12 @@
         EntityLiving.func_189752_a(p_189755_0_, EntityDragon.class);
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
         p_70014_1_.func_74768_a("DragonPhase", this.field_184677_bJ.func_188756_a().func_188652_i().func_188740_b());
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
@@ -1026,55 +920,46 @@
         }
     }
 
-    @Override
     protected void func_70623_bb()
     {
     }
 
-    @Override
     public Entity[] func_70021_al()
     {
         return this.field_70977_g;
     }
 
-    @Override
     public boolean func_70067_L()
     {
         return false;
     }
 
-    @Override
     public World func_82194_d()
     {
         return this.field_70170_p;
     }
 
-    @Override
     public SoundCategory func_184176_by()
     {
         return SoundCategory.HOSTILE;
     }
 
-    @Override
     protected SoundEvent func_184639_G()
     {
         return SoundEvents.field_187521_aK;
     }
 
-    @Override
     protected SoundEvent func_184601_bQ(DamageSource p_184601_1_)
     {
         return SoundEvents.field_187526_aP;
     }
 
-    @Override
     protected float func_70599_aP()
     {
         return 5.0F;
     }
 
     @Nullable
-    @Override
     protected ResourceLocation func_184647_J()
     {
         return LootTableList.field_191189_ay;
@@ -1087,24 +972,27 @@
         PhaseList <? extends IPhase > phaselist = iphase.func_188652_i();
         double d0;
 
-        if (phaselist == PhaseList.field_188744_d || phaselist == PhaseList.field_188745_e)
+        if (phaselist != PhaseList.field_188744_d && phaselist != PhaseList.field_188745_e)
+        {
+            if (iphase.func_188654_a())
+            {
+                d0 = (double)p_184667_1_;
+            }
+            else if (p_184667_1_ == 6)
+            {
+                d0 = 0.0D;
+            }
+            else
+            {
+                d0 = p_184667_3_[1] - p_184667_2_[1];
+            }
+        }
+        else
         {
             BlockPos blockpos = this.field_70170_p.func_175672_r(WorldGenEndPodium.field_186139_a);
             float f = Math.max(MathHelper.func_76133_a(this.func_174831_c(blockpos)) / 4.0F, 1.0F);
             d0 = (double)((float)p_184667_1_ / f);
         }
-        else if (iphase.func_188654_a())
-        {
-            d0 = (double)p_184667_1_;
-        }
-        else if (p_184667_1_ == 6)
-        {
-            d0 = 0.0;
-        }
-        else
-        {
-            d0 = p_184667_3_[1] - p_184667_2_[1];
-        }
 
         return (float)d0;
     }
@@ -1115,28 +1003,31 @@
         PhaseList <? extends IPhase > phaselist = iphase.func_188652_i();
         Vec3d vec3d;
 
-        if (phaselist == PhaseList.field_188744_d || phaselist == PhaseList.field_188745_e)
-        {
-            BlockPos blockpos = this.field_70170_p.func_175672_r(WorldGenEndPodium.field_186139_a);
-            float f5 = Math.max(MathHelper.func_76133_a(this.func_174831_c(blockpos)) / 4.0F, 1.0F);
-            float f2 = 6.0F / f5;
-            float f3 = this.field_70125_A;
-            float f4 = 1.5F;
-            this.field_70125_A = -f2 * 1.5F * 5.0F;
-            vec3d = this.func_70676_i(p_184665_1_);
-            this.field_70125_A = f3;
-        }
-        else if (iphase.func_188654_a())
-        {
-            float f = this.field_70125_A;
-            float f1 = 1.5F;
-            this.field_70125_A = -45.0F;
-            vec3d = this.func_70676_i(p_184665_1_);
-            this.field_70125_A = f;
+        if (phaselist != PhaseList.field_188744_d && phaselist != PhaseList.field_188745_e)
+        {
+            if (iphase.func_188654_a())
+            {
+                float f4 = this.field_70125_A;
+                float f5 = 1.5F;
+                this.field_70125_A = -45.0F;
+                vec3d = this.func_70676_i(p_184665_1_);
+                this.field_70125_A = f4;
+            }
+            else
+            {
+                vec3d = this.func_70676_i(p_184665_1_);
+            }
         }
         else
         {
+            BlockPos blockpos = this.field_70170_p.func_175672_r(WorldGenEndPodium.field_186139_a);
+            float f = Math.max(MathHelper.func_76133_a(this.func_174831_c(blockpos)) / 4.0F, 1.0F);
+            float f1 = 6.0F / f;
+            float f2 = this.field_70125_A;
+            float f3 = 1.5F;
+            this.field_70125_A = -f1 * 1.5F * 5.0F;
             vec3d = this.func_70676_i(p_184665_1_);
+            this.field_70125_A = f2;
         }
 
         return vec3d;
@@ -1152,7 +1043,7 @@
         }
         else
         {
-            entityplayer = this.field_70170_p.func_184139_a(p_184672_2_, 64.0, 64.0);
+            entityplayer = this.field_70170_p.func_184139_a(p_184672_2_, 64.0D, 64.0D);
         }
 
         if (p_184672_1_ == this.field_70992_bH)
@@ -1163,12 +1054,11 @@
         this.field_184677_bJ.func_188756_a().func_188655_a(p_184672_1_, p_184672_2_, p_184672_3_, entityplayer);
     }
 
-    @Override
     public void func_184206_a(DataParameter<?> p_184206_1_)
     {
         if (field_184674_a.equals(p_184206_1_) && this.field_70170_p.field_72995_K)
         {
-            this.field_184677_bJ.func_188758_a(PhaseList.func_188738_a(this.func_184212_Q().func_187225_a(field_184674_a)));
+            this.field_184677_bJ.func_188758_a(PhaseList.func_188738_a(((Integer)this.func_184212_Q().func_187225_a(field_184674_a)).intValue()));
         }
 
         super.func_184206_a(p_184206_1_);
@@ -1185,18 +1075,15 @@
         return this.field_184676_bI;
     }
 
-    @Override
     public void func_70690_d(PotionEffect p_70690_1_)
     {
     }
 
-    @Override
     protected boolean func_184228_n(Entity p_184228_1_)
     {
         return false;
     }
 
-    @Override
     public boolean func_184222_aU()
     {
         return false;
