--- before/net/minecraft/entity/boss/EntityDragon.java
+++ after/net/minecraft/entity/boss/EntityDragon.java
@@ -588,28 +588,27 @@
                     IBlockState iblockstate = this.world.getBlockState(blockpos);
                     Block block = iblockstate.getBlock();
 
-                    if (iblockstate.getMaterial() != Material.AIR && iblockstate.getMaterial() != Material.FIRE)
+                    if (!block.isAir(iblockstate, this.world, blockpos) && iblockstate.getMaterial() != Material.FIRE)
                     {
-                        if (!this.world.getGameRules().getBoolean("mobGriefing"))
-                        {
-                            flag = true;
-                        }
-                        else if (block == Blocks.BARRIER
-                                 || block == Blocks.OBSIDIAN
-                                 || block == Blocks.END_STONE
-                                 || block == Blocks.BEDROCK
-                                 || block == Blocks.END_PORTAL
-                                 || block == Blocks.END_PORTAL_FRAME)
-                        {
-                            flag = true;
-                        }
-                        else if (block != Blocks.COMMAND_BLOCK
-                                 && block != Blocks.REPEATING_COMMAND_BLOCK
-                                 && block != Blocks.CHAIN_COMMAND_BLOCK
-                                 && block != Blocks.IRON_BARS
-                                 && block != Blocks.END_GATEWAY)
-                        {
-                            flag1 = this.world.setBlockToAir(blockpos) || flag1;
+                        if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this))
+                        {
+                            flag = true;
+                        }
+                        else if (block.canEntityDestroy(iblockstate, this.world, blockpos, this)
+                                 && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, iblockstate))
+                        {
+                            if (block != Blocks.COMMAND_BLOCK
+                                && block != Blocks.REPEATING_COMMAND_BLOCK
+                                && block != Blocks.CHAIN_COMMAND_BLOCK
+                                && block != Blocks.IRON_BARS
+                                && block != Blocks.END_GATEWAY)
+                            {
+                                flag1 = this.world.setBlockToAir(blockpos) || flag1;
+                            }
+                            else
+                            {
+                                flag = true;
+                            }
                         }
                         else
                         {
