--- before/net/minecraft/client/model/ModelCreeper.java
+++ after/net/minecraft/client/model/ModelCreeper.java
@@ -60,8 +60,8 @@
 
     public void func_78087_a(float p_78087_1_, float p_78087_2_, float p_78087_3_, float p_78087_4_, float p_78087_5_, float p_78087_6_, Entity p_78087_7_)
     {
-        this.field_78135_a.field_78796_g = p_78087_4_ * ((float)Math.PI / 180F);
-        this.field_78135_a.field_78795_f = p_78087_5_ * ((float)Math.PI / 180F);
+        this.field_78135_a.field_78796_g = p_78087_4_ * 0.017453292F;
+        this.field_78135_a.field_78795_f = p_78087_5_ * 0.017453292F;
         this.field_78131_d.field_78795_f = MathHelper.func_76134_b(p_78087_1_ * 0.6662F) * 1.4F * p_78087_2_;
         this.field_78132_e.field_78795_f = MathHelper.func_76134_b(p_78087_1_ * 0.6662F + (float)Math.PI) * 1.4F * p_78087_2_;
         this.field_78129_f.field_78795_f = MathHelper.func_76134_b(p_78087_1_ * 0.6662F + (float)Math.PI) * 1.4F * p_78087_2_;
