--- before/net/minecraft/client/model/ModelMinecart.java
+++ after/net/minecraft/client/model/ModelMinecart.java
@@ -34,10 +34,10 @@
         this.field_78154_a[4].func_78790_a(-8.0F, -9.0F, -1.0F, 16, 8, 2, 0.0F);
         this.field_78154_a[4].func_78793_a(0.0F, 4.0F, 7.0F);
         this.field_78154_a[0].field_78795_f = ((float)Math.PI / 2F);
-        this.field_78154_a[1].field_78796_g = ((float)Math.PI * 1.5F);
+        this.field_78154_a[1].field_78796_g = ((float)Math.PI * 3F / 2F);
         this.field_78154_a[2].field_78796_g = ((float)Math.PI / 2F);
         this.field_78154_a[3].field_78796_g = (float)Math.PI;
-        this.field_78154_a[5].field_78795_f = (-(float)Math.PI / 2F);
+        this.field_78154_a[5].field_78795_f = -((float)Math.PI / 2F);
     }
 
     public void func_78088_a(Entity p_78088_1_, float p_78088_2_, float p_78088_3_, float p_78088_4_, float p_78088_5_, float p_78088_6_, float p_78088_7_)
