--- before/net/minecraft/client/model/ModelEnderMite.java
+++ after/net/minecraft/client/model/ModelEnderMite.java
@@ -11,10 +11,11 @@
     private static final int[][] field_178716_a = new int[][] {{4, 3, 2}, {6, 4, 5}, {3, 3, 1}, {1, 2, 1}};
     private static final int[][] field_178714_b = new int[][] {{0, 0}, {0, 5}, {0, 14}, {0, 18}};
     private static final int field_178715_c = field_178716_a.length;
-    private final ModelRenderer[] field_178713_d = new ModelRenderer[field_178715_c];
+    private final ModelRenderer[] field_178713_d;
 
     public ModelEnderMite()
     {
+        this.field_178713_d = new ModelRenderer[field_178715_c];
         float f = -3.5F;
 
         for (int i = 0; i < this.field_178713_d.length; ++i)
