--- before/net/minecraft/client/model/ModelElytra.java
+++ after/net/minecraft/client/model/ModelElytra.java
@@ -62,12 +62,12 @@
             }
 
             f = f4 * 0.34906584F + (1.0F - f4) * f;
-            f1 = f4 * (-(float)Math.PI / 2F) + (1.0F - f4) * f1;
+            f1 = f4 * -((float)Math.PI / 2F) + (1.0F - f4) * f1;
         }
         else if (p_78087_7_.func_70093_af())
         {
             f = ((float)Math.PI * 2F / 9F);
-            f1 = (-(float)Math.PI / 4F);
+            f1 = -((float)Math.PI / 4F);
             f2 = 3.0F;
             f3 = 0.08726646F;
         }
