--- before/net/minecraft/client/model/ModelWither.java
+++ after/net/minecraft/client/model/ModelWither.java
@@ -62,8 +62,8 @@
         this.field_82905_a[1].field_78795_f = (0.065F + 0.05F * f) * (float)Math.PI;
         this.field_82905_a[2].func_78793_a(-2.0F, 6.9F + MathHelper.func_76134_b(this.field_82905_a[1].field_78795_f) * 10.0F, -0.5F + MathHelper.func_76126_a(this.field_82905_a[1].field_78795_f) * 10.0F);
         this.field_82905_a[2].field_78795_f = (0.265F + 0.1F * f) * (float)Math.PI;
-        this.field_82904_b[0].field_78796_g = p_78087_4_ * ((float)Math.PI / 180F);
-        this.field_82904_b[0].field_78795_f = p_78087_5_ * ((float)Math.PI / 180F);
+        this.field_82904_b[0].field_78796_g = p_78087_4_ * 0.017453292F;
+        this.field_82904_b[0].field_78795_f = p_78087_5_ * 0.017453292F;
     }
 
     public void func_78086_a(EntityLivingBase p_78086_1_, float p_78086_2_, float p_78086_3_, float p_78086_4_)
@@ -72,8 +72,8 @@
 
         for (int i = 1; i < 3; ++i)
         {
-            this.field_82904_b[i].field_78796_g = (entitywither.func_82207_a(i - 1) - p_78086_1_.field_70761_aq) * ((float)Math.PI / 180F);
-            this.field_82904_b[i].field_78795_f = entitywither.func_82210_r(i - 1) * ((float)Math.PI / 180F);
+            this.field_82904_b[i].field_78796_g = (entitywither.func_82207_a(i - 1) - p_78086_1_.field_70761_aq) * 0.017453292F;
+            this.field_82904_b[i].field_78795_f = entitywither.func_82210_r(i - 1) * 0.017453292F;
         }
     }
 }
