--- before/net/minecraft/client/model/ModelBox.java
+++ after/net/minecraft/client/model/ModelBox.java
@@ -4,7 +4,6 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-@SideOnly(Side.CLIENT)
 public class ModelBox
 {
     private final PositionTextureVertex[] field_78253_h;
@@ -35,12 +34,12 @@
         float f = p_i46301_4_ + (float)p_i46301_7_;
         float f1 = p_i46301_5_ + (float)p_i46301_8_;
         float f2 = p_i46301_6_ + (float)p_i46301_9_;
-        p_i46301_4_ -= p_i46301_10_;
-        p_i46301_5_ -= p_i46301_10_;
-        p_i46301_6_ -= p_i46301_10_;
-        f += p_i46301_10_;
-        f1 += p_i46301_10_;
-        f2 += p_i46301_10_;
+        p_i46301_4_ = p_i46301_4_ - p_i46301_10_;
+        p_i46301_5_ = p_i46301_5_ - p_i46301_10_;
+        p_i46301_6_ = p_i46301_6_ - p_i46301_10_;
+        f = f + p_i46301_10_;
+        f1 = f1 + p_i46301_10_;
+        f2 = f2 + p_i46301_10_;
 
         if (p_i46301_11_)
         {
@@ -81,6 +80,7 @@
         }
     }
 
+    @SideOnly(Side.CLIENT)
     public void func_178780_a(BufferBuilder p_178780_1_, float p_178780_2_)
     {
         for (TexturedQuad texturedquad : this.field_78254_i)
