--- before/net/minecraft/client/model/ModelBox.java
+++ after/net/minecraft/client/model/ModelBox.java
@@ -4,7 +4,6 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-@SideOnly(Side.CLIENT)
 public class ModelBox
 {
     private final PositionTextureVertex[] field_78253_h;
@@ -17,47 +16,12 @@
     public final float field_78246_f;
     public String field_78247_g;
 
-    public ModelBox(
-        ModelRenderer p_i46359_1_,
-        int p_i46359_2_,
-        int p_i46359_3_,
-        float p_i46359_4_,
-        float p_i46359_5_,
-        float p_i46359_6_,
-        int p_i46359_7_,
-        int p_i46359_8_,
-        int p_i46359_9_,
-        float p_i46359_10_
-    )
+    public ModelBox(ModelRenderer p_i46359_1_, int p_i46359_2_, int p_i46359_3_, float p_i46359_4_, float p_i46359_5_, float p_i46359_6_, int p_i46359_7_, int p_i46359_8_, int p_i46359_9_, float p_i46359_10_)
     {
-        this(
-            p_i46359_1_,
-            p_i46359_2_,
-            p_i46359_3_,
-            p_i46359_4_,
-            p_i46359_5_,
-            p_i46359_6_,
-            p_i46359_7_,
-            p_i46359_8_,
-            p_i46359_9_,
-            p_i46359_10_,
-            p_i46359_1_.field_78809_i
-        );
+        this(p_i46359_1_, p_i46359_2_, p_i46359_3_, p_i46359_4_, p_i46359_5_, p_i46359_6_, p_i46359_7_, p_i46359_8_, p_i46359_9_, p_i46359_10_, p_i46359_1_.field_78809_i);
     }
 
-    public ModelBox(
-        ModelRenderer p_i46301_1_,
-        int p_i46301_2_,
-        int p_i46301_3_,
-        float p_i46301_4_,
-        float p_i46301_5_,
-        float p_i46301_6_,
-        int p_i46301_7_,
-        int p_i46301_8_,
-        int p_i46301_9_,
-        float p_i46301_10_,
-        boolean p_i46301_11_
-    )
+    public ModelBox(ModelRenderer p_i46301_1_, int p_i46301_2_, int p_i46301_3_, float p_i46301_4_, float p_i46301_5_, float p_i46301_6_, int p_i46301_7_, int p_i46301_8_, int p_i46301_9_, float p_i46301_10_, boolean p_i46301_11_)
     {
         this.field_78252_a = p_i46301_4_;
         this.field_78250_b = p_i46301_5_;
@@ -70,12 +34,12 @@
         float f = p_i46301_4_ + (float)p_i46301_7_;
         float f1 = p_i46301_5_ + (float)p_i46301_8_;
         float f2 = p_i46301_6_ + (float)p_i46301_9_;
-        p_i46301_4_ -= p_i46301_10_;
-        p_i46301_5_ -= p_i46301_10_;
-        p_i46301_6_ -= p_i46301_10_;
-        f += p_i46301_10_;
-        f1 += p_i46301_10_;
-        f2 += p_i46301_10_;
+        p_i46301_4_ = p_i46301_4_ - p_i46301_10_;
+        p_i46301_5_ = p_i46301_5_ - p_i46301_10_;
+        p_i46301_6_ = p_i46301_6_ - p_i46301_10_;
+        f = f + p_i46301_10_;
+        f1 = f1 + p_i46301_10_;
+        f2 = f2 + p_i46301_10_;
 
         if (p_i46301_11_)
         {
@@ -100,60 +64,12 @@
         this.field_78253_h[5] = positiontexturevertex4;
         this.field_78253_h[6] = positiontexturevertex5;
         this.field_78253_h[7] = positiontexturevertex6;
-        this.field_78254_i[0] = new TexturedQuad(
-            new PositionTextureVertex[] {positiontexturevertex4, positiontexturevertex, positiontexturevertex1, positiontexturevertex5},
-            p_i46301_2_ + p_i46301_9_ + p_i46301_7_,
-            p_i46301_3_ + p_i46301_9_,
-            p_i46301_2_ + p_i46301_9_ + p_i46301_7_ + p_i46301_9_,
-            p_i46301_3_ + p_i46301_9_ + p_i46301_8_,
-            p_i46301_1_.field_78801_a,
-            p_i46301_1_.field_78799_b
-        );
-        this.field_78254_i[1] = new TexturedQuad(
-            new PositionTextureVertex[] {positiontexturevertex7, positiontexturevertex3, positiontexturevertex6, positiontexturevertex2},
-            p_i46301_2_,
-            p_i46301_3_ + p_i46301_9_,
-            p_i46301_2_ + p_i46301_9_,
-            p_i46301_3_ + p_i46301_9_ + p_i46301_8_,
-            p_i46301_1_.field_78801_a,
-            p_i46301_1_.field_78799_b
-        );
-        this.field_78254_i[2] = new TexturedQuad(
-            new PositionTextureVertex[] {positiontexturevertex4, positiontexturevertex3, positiontexturevertex7, positiontexturevertex},
-            p_i46301_2_ + p_i46301_9_,
-            p_i46301_3_,
-            p_i46301_2_ + p_i46301_9_ + p_i46301_7_,
-            p_i46301_3_ + p_i46301_9_,
-            p_i46301_1_.field_78801_a,
-            p_i46301_1_.field_78799_b
-        );
-        this.field_78254_i[3] = new TexturedQuad(
-            new PositionTextureVertex[] {positiontexturevertex1, positiontexturevertex2, positiontexturevertex6, positiontexturevertex5},
-            p_i46301_2_ + p_i46301_9_ + p_i46301_7_,
-            p_i46301_3_ + p_i46301_9_,
-            p_i46301_2_ + p_i46301_9_ + p_i46301_7_ + p_i46301_7_,
-            p_i46301_3_,
-            p_i46301_1_.field_78801_a,
-            p_i46301_1_.field_78799_b
-        );
-        this.field_78254_i[4] = new TexturedQuad(
-            new PositionTextureVertex[] {positiontexturevertex, positiontexturevertex7, positiontexturevertex2, positiontexturevertex1},
-            p_i46301_2_ + p_i46301_9_,
-            p_i46301_3_ + p_i46301_9_,
-            p_i46301_2_ + p_i46301_9_ + p_i46301_7_,
-            p_i46301_3_ + p_i46301_9_ + p_i46301_8_,
-            p_i46301_1_.field_78801_a,
-            p_i46301_1_.field_78799_b
-        );
-        this.field_78254_i[5] = new TexturedQuad(
-            new PositionTextureVertex[] {positiontexturevertex3, positiontexturevertex4, positiontexturevertex5, positiontexturevertex6},
-            p_i46301_2_ + p_i46301_9_ + p_i46301_7_ + p_i46301_9_,
-            p_i46301_3_ + p_i46301_9_,
-            p_i46301_2_ + p_i46301_9_ + p_i46301_7_ + p_i46301_9_ + p_i46301_7_,
-            p_i46301_3_ + p_i46301_9_ + p_i46301_8_,
-            p_i46301_1_.field_78801_a,
-            p_i46301_1_.field_78799_b
-        );
+        this.field_78254_i[0] = new TexturedQuad(new PositionTextureVertex[] {positiontexturevertex4, positiontexturevertex, positiontexturevertex1, positiontexturevertex5}, p_i46301_2_ + p_i46301_9_ + p_i46301_7_, p_i46301_3_ + p_i46301_9_, p_i46301_2_ + p_i46301_9_ + p_i46301_7_ + p_i46301_9_, p_i46301_3_ + p_i46301_9_ + p_i46301_8_, p_i46301_1_.field_78801_a, p_i46301_1_.field_78799_b);
+        this.field_78254_i[1] = new TexturedQuad(new PositionTextureVertex[] {positiontexturevertex7, positiontexturevertex3, positiontexturevertex6, positiontexturevertex2}, p_i46301_2_, p_i46301_3_ + p_i46301_9_, p_i46301_2_ + p_i46301_9_, p_i46301_3_ + p_i46301_9_ + p_i46301_8_, p_i46301_1_.field_78801_a, p_i46301_1_.field_78799_b);
+        this.field_78254_i[2] = new TexturedQuad(new PositionTextureVertex[] {positiontexturevertex4, positiontexturevertex3, positiontexturevertex7, positiontexturevertex}, p_i46301_2_ + p_i46301_9_, p_i46301_3_, p_i46301_2_ + p_i46301_9_ + p_i46301_7_, p_i46301_3_ + p_i46301_9_, p_i46301_1_.field_78801_a, p_i46301_1_.field_78799_b);
+        this.field_78254_i[3] = new TexturedQuad(new PositionTextureVertex[] {positiontexturevertex1, positiontexturevertex2, positiontexturevertex6, positiontexturevertex5}, p_i46301_2_ + p_i46301_9_ + p_i46301_7_, p_i46301_3_ + p_i46301_9_, p_i46301_2_ + p_i46301_9_ + p_i46301_7_ + p_i46301_7_, p_i46301_3_, p_i46301_1_.field_78801_a, p_i46301_1_.field_78799_b);
+        this.field_78254_i[4] = new TexturedQuad(new PositionTextureVertex[] {positiontexturevertex, positiontexturevertex7, positiontexturevertex2, positiontexturevertex1}, p_i46301_2_ + p_i46301_9_, p_i46301_3_ + p_i46301_9_, p_i46301_2_ + p_i46301_9_ + p_i46301_7_, p_i46301_3_ + p_i46301_9_ + p_i46301_8_, p_i46301_1_.field_78801_a, p_i46301_1_.field_78799_b);
+        this.field_78254_i[5] = new TexturedQuad(new PositionTextureVertex[] {positiontexturevertex3, positiontexturevertex4, positiontexturevertex5, positiontexturevertex6}, p_i46301_2_ + p_i46301_9_ + p_i46301_7_ + p_i46301_9_, p_i46301_3_ + p_i46301_9_, p_i46301_2_ + p_i46301_9_ + p_i46301_7_ + p_i46301_9_ + p_i46301_7_, p_i46301_3_ + p_i46301_9_ + p_i46301_8_, p_i46301_1_.field_78801_a, p_i46301_1_.field_78799_b);
 
         if (p_i46301_11_)
         {
@@ -164,6 +80,7 @@
         }
     }
 
+    @SideOnly(Side.CLIENT)
     public void func_178780_a(BufferBuilder p_178780_1_, float p_178780_2_)
     {
         for (TexturedQuad texturedquad : this.field_78254_i)
