--- before/net/minecraft/client/model/ModelDragonHead.java
+++ after/net/minecraft/client/model/ModelDragonHead.java
@@ -53,8 +53,8 @@
     public void func_78088_a(Entity p_78088_1_, float p_78088_2_, float p_78088_3_, float p_78088_4_, float p_78088_5_, float p_78088_6_, float p_78088_7_)
     {
         this.field_187071_b.field_78795_f = (float)(Math.sin((double)(p_78088_2_ * (float)Math.PI * 0.2F)) + 1.0D) * 0.2F;
-        this.field_187070_a.field_78796_g = p_78088_5_ * ((float)Math.PI / 180F);
-        this.field_187070_a.field_78795_f = p_78088_6_ * ((float)Math.PI / 180F);
+        this.field_187070_a.field_78796_g = p_78088_5_ * 0.017453292F;
+        this.field_187070_a.field_78795_f = p_78088_6_ * 0.017453292F;
         GlStateManager.func_179109_b(0.0F, -0.374375F, 0.0F);
         GlStateManager.func_179152_a(0.75F, 0.75F, 0.75F);
         this.field_187070_a.func_78785_a(p_78088_7_);
