--- before/net/minecraft/client/model/ModelBed.java
+++ after/net/minecraft/client/model/ModelBed.java
@@ -32,7 +32,7 @@
         this.field_193774_c[3].field_78795_f = ((float)Math.PI / 2F);
         this.field_193774_c[0].field_78808_h = 0.0F;
         this.field_193774_c[1].field_78808_h = ((float)Math.PI / 2F);
-        this.field_193774_c[2].field_78808_h = ((float)Math.PI * 1.5F);
+        this.field_193774_c[2].field_78808_h = ((float)Math.PI * 3F / 2F);
         this.field_193774_c[3].field_78808_h = (float)Math.PI;
     }
 
