--- before/net/minecraft/client/model/ModelRenderer.java
+++ after/net/minecraft/client/model/ModelRenderer.java
@@ -9,7 +9,6 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-@SideOnly(Side.CLIENT)
 public class ModelRenderer
 {
     public float textureWidth = 64.0F;
@@ -117,6 +116,7 @@
         this.rotationPointZ = rotationPointZIn;
     }
 
+    @SideOnly(Side.CLIENT)
     public void render(float scale)
     {
         if (!this.isHidden)
@@ -195,6 +195,7 @@
         }
     }
 
+    @SideOnly(Side.CLIENT)
     public void renderWithRotation(float scale)
     {
         if (!this.isHidden)
@@ -230,6 +231,7 @@
         }
     }
 
+    @SideOnly(Side.CLIENT)
     public void postRender(float scale)
     {
         if (!this.isHidden)
@@ -268,6 +270,7 @@
         }
     }
 
+    @SideOnly(Side.CLIENT)
     private void compileDisplayList(float scale)
     {
         this.displayList = GLAllocation.generateDisplayLists(1);
