--- before/net/minecraft/client/model/ModelWitch.java
+++ after/net/minecraft/client/model/ModelWitch.java
@@ -49,9 +49,9 @@
         this.field_82898_f.field_82908_p = 0.0F;
         this.field_82898_f.field_82907_q = 0.0F;
         float f = 0.01F * (float)(p_78087_7_.func_145782_y() % 10);
-        this.field_82898_f.field_78795_f = MathHelper.func_76126_a((float)p_78087_7_.field_70173_aa * f) * 4.5F * ((float)Math.PI / 180F);
+        this.field_82898_f.field_78795_f = MathHelper.func_76126_a((float)p_78087_7_.field_70173_aa * f) * 4.5F * 0.017453292F;
         this.field_82898_f.field_78796_g = 0.0F;
-        this.field_82898_f.field_78808_h = MathHelper.func_76134_b((float)p_78087_7_.field_70173_aa * f) * 2.5F * ((float)Math.PI / 180F);
+        this.field_82898_f.field_78808_h = MathHelper.func_76134_b((float)p_78087_7_.field_70173_aa * f) * 2.5F * 0.017453292F;
 
         if (this.field_82900_g)
         {
