--- before/net/minecraft/client/model/ModelBase.java
+++ after/net/minecraft/client/model/ModelBase.java
@@ -10,14 +10,13 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-@SideOnly(Side.CLIENT)
 public abstract class ModelBase
 {
     public float swingProgress;
     public boolean isRiding;
     public boolean isChild = true;
-    public List<ModelRenderer> boxList = Lists.newArrayList();
-    private final Map<String, TextureOffset> modelTextureMap = Maps.newHashMap();
+    public List<ModelRenderer> boxList = Lists.<ModelRenderer>newArrayList();
+    private final Map<String, TextureOffset> modelTextureMap = Maps.<String, TextureOffset>newHashMap();
     public int textureWidth = 64;
     public int textureHeight = 32;
 
