--- before/net/minecraft/client/model/ModelDragon.java
+++ after/net/minecraft/client/model/ModelDragon.java
@@ -132,9 +132,9 @@
         {
             double[] adouble1 = entitydragon.func_70974_a(5 - i, this.field_78223_m);
             float f9 = (float)Math.cos((double)((float)i * 0.45F + f8)) * 0.15F;
-            this.field_78219_b.field_78796_g = this.func_78214_a(adouble1[0] - adouble[0]) * ((float)Math.PI / 180F) * 1.5F;
-            this.field_78219_b.field_78795_f = f9 + entitydragon.func_184667_a(i, adouble, adouble1) * ((float)Math.PI / 180F) * 1.5F * 5.0F;
-            this.field_78219_b.field_78808_h = -this.func_78214_a(adouble1[0] - (double)f7) * ((float)Math.PI / 180F) * 1.5F;
+            this.field_78219_b.field_78796_g = this.func_78214_a(adouble1[0] - adouble[0]) * 0.017453292F * 1.5F;
+            this.field_78219_b.field_78795_f = f9 + entitydragon.func_184667_a(i, adouble, adouble1) * 0.017453292F * 1.5F * 5.0F;
+            this.field_78219_b.field_78808_h = -this.func_78214_a(adouble1[0] - (double)f7) * 0.017453292F * 1.5F;
             this.field_78219_b.field_78797_d = f2;
             this.field_78219_b.field_78798_e = f3;
             this.field_78219_b.field_78800_c = f4;
@@ -148,9 +148,9 @@
         this.field_78221_a.field_78798_e = f3;
         this.field_78221_a.field_78800_c = f4;
         double[] adouble2 = entitydragon.func_70974_a(0, this.field_78223_m);
-        this.field_78221_a.field_78796_g = this.func_78214_a(adouble2[0] - adouble[0]) * ((float)Math.PI / 180F);
-        this.field_78221_a.field_78795_f = this.func_78214_a((double)entitydragon.func_184667_a(6, adouble, adouble2)) * ((float)Math.PI / 180F) * 1.5F * 5.0F;
-        this.field_78221_a.field_78808_h = -this.func_78214_a(adouble2[0] - (double)f7) * ((float)Math.PI / 180F);
+        this.field_78221_a.field_78796_g = this.func_78214_a(adouble2[0] - adouble[0]) * 0.017453292F;
+        this.field_78221_a.field_78795_f = this.func_78214_a((double)entitydragon.func_184667_a(6, adouble, adouble2)) * 0.017453292F * 1.5F * 5.0F;
+        this.field_78221_a.field_78808_h = -this.func_78214_a(adouble2[0] - (double)f7) * 0.017453292F;
         this.field_78221_a.func_78785_a(p_78088_7_);
         GlStateManager.func_179094_E();
         GlStateManager.func_179109_b(0.0F, 1.0F, 0.0F);
@@ -197,10 +197,10 @@
         for (int k = 0; k < 12; ++k)
         {
             adouble2 = entitydragon.func_70974_a(12 + k, this.field_78223_m);
-            f10 = (float)((double)f10 + Math.sin((double)((float)k * 0.45F + f8)) * (double)0.05F);
-            this.field_78219_b.field_78796_g = (this.func_78214_a(adouble2[0] - adouble[0]) * 1.5F + 180.0F) * ((float)Math.PI / 180F);
-            this.field_78219_b.field_78795_f = f10 + (float)(adouble2[1] - adouble[1]) * ((float)Math.PI / 180F) * 1.5F * 5.0F;
-            this.field_78219_b.field_78808_h = this.func_78214_a(adouble2[0] - (double)f7) * ((float)Math.PI / 180F) * 1.5F;
+            f10 = (float)((double)f10 + Math.sin((double)((float)k * 0.45F + f8)) * 0.05000000074505806D);
+            this.field_78219_b.field_78796_g = (this.func_78214_a(adouble2[0] - adouble[0]) * 1.5F + 180.0F) * 0.017453292F;
+            this.field_78219_b.field_78795_f = f10 + (float)(adouble2[1] - adouble[1]) * 0.017453292F * 1.5F * 5.0F;
+            this.field_78219_b.field_78808_h = this.func_78214_a(adouble2[0] - (double)f7) * 0.017453292F * 1.5F;
             this.field_78219_b.field_78797_d = f2;
             this.field_78219_b.field_78798_e = f3;
             this.field_78219_b.field_78800_c = f4;
