--- before/net/minecraft/client/model/ModelSilverfish.java
+++ after/net/minecraft/client/model/ModelSilverfish.java
@@ -8,7 +8,7 @@
 @SideOnly(Side.CLIENT)
 public class ModelSilverfish extends ModelBase
 {
-    private final ModelRenderer[] field_78171_a;
+    private final ModelRenderer[] field_78171_a = new ModelRenderer[7];
     private final ModelRenderer[] field_78169_b;
     private final float[] field_78170_c = new float[7];
     private static final int[][] field_78167_d = new int[][] {{3, 2, 2}, {4, 3, 2}, {6, 4, 3}, {3, 3, 3}, {2, 2, 3}, {2, 1, 2}, {1, 1, 2}};
@@ -16,7 +16,6 @@
 
     public ModelSilverfish()
     {
-        this.field_78171_a = new ModelRenderer[7];
         float f = -3.5F;
 
         for (int i = 0; i < this.field_78171_a.length; ++i)
