--- before/net/minecraft/client/renderer/vertex/VertexFormat.java
+++ after/net/minecraft/client/renderer/vertex/VertexFormat.java
@@ -17,6 +17,7 @@
     private int colorElementOffset = -1;
     private final List<Integer> uvOffsetsById = Lists.newArrayList();
     private int normalElementOffset = -1;
+    private int hashCode;
 
     public VertexFormat(VertexFormat vertexFormatIn)
     {
@@ -42,6 +43,7 @@
         this.uvOffsetsById.clear();
         this.normalElementOffset = -1;
         this.vertexSize = 0;
+        this.hashCode = 0;
     }
 
     @SuppressWarnings("incomplete-switch")
@@ -70,6 +72,7 @@
             }
 
             this.vertexSize = this.vertexSize + element.getSize();
+            this.hashCode = 0;
             return this;
         }
     }
@@ -198,8 +201,11 @@
     @Override
     public int hashCode()
     {
+        if (this.hashCode != 0) return this.hashCode;
         int i = this.elements.hashCode();
         i = 31 * i + this.offsets.hashCode();
-        return 31 * i + this.vertexSize;
+        i = 31 * i + this.vertexSize;
+        this.hashCode = i;
+        return i;
     }
 }
