--- before/net/minecraft/scoreboard/IScoreCriteria.java
+++ after/net/minecraft/scoreboard/IScoreCriteria.java
@@ -6,7 +6,7 @@
 
 public interface IScoreCriteria
 {
-    Map<String, IScoreCriteria> field_96643_a = Maps.newHashMap();
+    Map<String, IScoreCriteria> field_96643_a = Maps.<String, IScoreCriteria>newHashMap();
     IScoreCriteria field_96641_b = new ScoreCriteria("dummy");
     IScoreCriteria field_178791_c = new ScoreCriteria("trigger");
     IScoreCriteria field_96642_c = new ScoreCriteria("deathCount");
@@ -32,7 +32,7 @@
         INTEGER("integer"),
         HEARTS("hearts");
 
-        private static final Map<String, IScoreCriteria.EnumRenderType> field_178801_c = Maps.newHashMap();
+        private static final Map<String, IScoreCriteria.EnumRenderType> field_178801_c = Maps.<String, IScoreCriteria.EnumRenderType>newHashMap();
         private final String field_178798_d;
 
         private EnumRenderType(String p_i45548_3_)
