--- before/net/minecraft/scoreboard/Scoreboard.java
+++ after/net/minecraft/scoreboard/Scoreboard.java
@@ -13,12 +13,12 @@
 
 public class Scoreboard
 {
-    private final Map<String, ScoreObjective> field_96545_a = Maps.newHashMap();
-    private final Map<IScoreCriteria, List<ScoreObjective>> field_96543_b = Maps.newHashMap();
-    private final Map<String, Map<ScoreObjective, Score>> field_96544_c = Maps.newHashMap();
+    private final Map<String, ScoreObjective> field_96545_a = Maps.<String, ScoreObjective>newHashMap();
+    private final Map<IScoreCriteria, List<ScoreObjective>> field_96543_b = Maps.<IScoreCriteria, List<ScoreObjective>>newHashMap();
+    private final Map<String, Map<ScoreObjective, Score>> field_96544_c = Maps.<String, Map<ScoreObjective, Score>>newHashMap();
     private final ScoreObjective[] field_96541_d = new ScoreObjective[19];
-    private final Map<String, ScorePlayerTeam> field_96542_e = Maps.newHashMap();
-    private final Map<String, ScorePlayerTeam> field_96540_f = Maps.newHashMap();
+    private final Map<String, ScorePlayerTeam> field_96542_e = Maps.<String, ScorePlayerTeam>newHashMap();
+    private final Map<String, ScorePlayerTeam> field_96540_f = Maps.<String, ScorePlayerTeam>newHashMap();
     private static String[] field_178823_g;
 
     @Nullable
@@ -44,11 +44,11 @@
             else
             {
                 scoreobjective = new ScoreObjective(this, p_96535_1_, p_96535_2_);
-                List<ScoreObjective> list = this.field_96543_b.get(p_96535_2_);
+                List<ScoreObjective> list = (List)this.field_96543_b.get(p_96535_2_);
 
                 if (list == null)
                 {
-                    list = Lists.newArrayList();
+                    list = Lists.<ScoreObjective>newArrayList();
                     this.field_96543_b.put(p_96535_2_, list);
                 }
 
@@ -62,13 +62,13 @@
 
     public Collection<ScoreObjective> func_96520_a(IScoreCriteria p_96520_1_)
     {
-        Collection<ScoreObjective> collection = this.field_96543_b.get(p_96520_1_);
+        Collection<ScoreObjective> collection = (Collection)this.field_96543_b.get(p_96520_1_);
         return collection == null ? Lists.newArrayList() : Lists.newArrayList(collection);
     }
 
     public boolean func_178819_b(String p_178819_1_, ScoreObjective p_178819_2_)
     {
-        Map<ScoreObjective, Score> map = this.field_96544_c.get(p_178819_1_);
+        Map<ScoreObjective, Score> map = (Map)this.field_96544_c.get(p_178819_1_);
 
         if (map == null)
         {
@@ -89,11 +89,11 @@
         }
         else
         {
-            Map<ScoreObjective, Score> map = this.field_96544_c.get(p_96529_1_);
+            Map<ScoreObjective, Score> map = (Map)this.field_96544_c.get(p_96529_1_);
 
             if (map == null)
             {
-                map = Maps.newHashMap();
+                map = Maps.<ScoreObjective, Score>newHashMap();
                 this.field_96544_c.put(p_96529_1_, map);
             }
 
@@ -111,7 +111,7 @@
 
     public Collection<Score> func_96534_i(ScoreObjective p_96534_1_)
     {
-        List<Score> list = Lists.newArrayList();
+        List<Score> list = Lists.<Score>newArrayList();
 
         for (Map<ScoreObjective, Score> map : this.field_96544_c.values())
         {
@@ -141,7 +141,7 @@
     {
         if (p_178822_2_ == null)
         {
-            Map<ScoreObjective, Score> map = this.field_96544_c.remove(p_178822_1_);
+            Map<ScoreObjective, Score> map = (Map)this.field_96544_c.remove(p_178822_1_);
 
             if (map != null)
             {
@@ -150,7 +150,7 @@
         }
         else
         {
-            Map<ScoreObjective, Score> map2 = this.field_96544_c.get(p_178822_1_);
+            Map<ScoreObjective, Score> map2 = (Map)this.field_96544_c.get(p_178822_1_);
 
             if (map2 != null)
             {
@@ -158,7 +158,7 @@
 
                 if (map2.size() < 1)
                 {
-                    Map<ScoreObjective, Score> map1 = this.field_96544_c.remove(p_178822_1_);
+                    Map<ScoreObjective, Score> map1 = (Map)this.field_96544_c.remove(p_178822_1_);
 
                     if (map1 != null)
                     {
@@ -176,7 +176,7 @@
     public Collection<Score> func_96528_e()
     {
         Collection<Map<ScoreObjective, Score>> collection = this.field_96544_c.values();
-        List<Score> list = Lists.newArrayList();
+        List<Score> list = Lists.<Score>newArrayList();
 
         for (Map<ScoreObjective, Score> map : collection)
         {
@@ -188,11 +188,11 @@
 
     public Map<ScoreObjective, Score> func_96510_d(String p_96510_1_)
     {
-        Map<ScoreObjective, Score> map = this.field_96544_c.get(p_96510_1_);
+        Map<ScoreObjective, Score> map = (Map)this.field_96544_c.get(p_96510_1_);
 
         if (map == null)
         {
-            map = Maps.newHashMap();
+            map = Maps.<ScoreObjective, Score>newHashMap();
         }
 
         return map;
@@ -206,11 +206,11 @@
         {
             if (this.func_96539_a(i) == p_96519_1_)
             {
-                this.func_96530_a(i, null);
+                this.func_96530_a(i, (ScoreObjective)null);
             }
         }
 
-        List<ScoreObjective> list = this.field_96543_b.get(p_96519_1_.func_96680_c());
+        List<ScoreObjective> list = (List)this.field_96543_b.get(p_96519_1_.func_96680_c());
 
         if (list != null)
         {
@@ -459,7 +459,7 @@
         if (p_181140_1_ != null && !(p_181140_1_ instanceof EntityPlayer) && !p_181140_1_.func_70089_S())
         {
             String s = p_181140_1_.func_189512_bd();
-            this.func_178822_d(s, null);
+            this.func_178822_d(s, (ScoreObjective)null);
             this.func_96524_g(s);
         }
     }
