--- before/net/minecraft/scoreboard/ScorePlayerTeam.java
+++ after/net/minecraft/scoreboard/ScorePlayerTeam.java
@@ -12,7 +12,7 @@
 {
     private final Scoreboard field_96677_a;
     private final String field_96675_b;
-    private final Set<String> field_96676_c = Sets.newHashSet();
+    private final Set<String> field_96676_c = Sets.<String>newHashSet();
     private String field_96673_d;
     private String field_96674_e = "";
     private String field_96671_f = "";
