--- before/net/minecraft/scoreboard/Team.java
+++ after/net/minecraft/scoreboard/Team.java
@@ -49,13 +49,13 @@
         HIDE_FOR_OTHER_TEAMS("pushOtherTeams", 2),
         HIDE_FOR_OWN_TEAM("pushOwnTeam", 3);
 
-        private static final Map<String, Team.CollisionRule> field_186695_g = Maps.newHashMap();
+        private static final Map<String, Team.CollisionRule> field_186695_g = Maps.<String, Team.CollisionRule>newHashMap();
         public final String field_186693_e;
         public final int field_186694_f;
 
         public static String[] func_186687_a()
         {
-            return field_186695_g.keySet().toArray(new String[field_186695_g.size()]);
+            return (String[])field_186695_g.keySet().toArray(new String[field_186695_g.size()]);
         }
 
         @Nullable
@@ -86,13 +86,13 @@
         HIDE_FOR_OTHER_TEAMS("hideForOtherTeams", 2),
         HIDE_FOR_OWN_TEAM("hideForOwnTeam", 3);
 
-        private static final Map<String, Team.EnumVisible> field_186697_g = Maps.newHashMap();
+        private static final Map<String, Team.EnumVisible> field_186697_g = Maps.<String, Team.EnumVisible>newHashMap();
         public final String field_178830_e;
         public final int field_178827_f;
 
         public static String[] func_178825_a()
         {
-            return field_186697_g.keySet().toArray(new String[field_186697_g.size()]);
+            return (String[])field_186697_g.keySet().toArray(new String[field_186697_g.size()]);
         }
 
         @Nullable
