--- original/net/minecraft/client/gui/GuiScreenDemo.java
+++ modified/net/minecraft/client/gui/GuiScreenDemo.java
@@ -35,9 +35,7 @@
                 button.enabled = false;
 
                 try {
-                    Class<?> oclass = Class.forName("java.awt.Desktop");
-                    Object object = oclass.getMethod("getDesktop").invoke(null);
-                    oclass.getMethod("browse", URI.class).invoke(object, new URI("http://www.minecraft.net/store?source=demo"));
+                    com.cleanroommc.client.sdl.SDL.desktop().open(new URI("http://www.minecraft.net/store?source=demo"));
                 } catch (Throwable throwable) {
                     LOGGER.error("Couldn't open link", throwable);
                 }
@@ -75,23 +73,23 @@
         j += 12;
         GameSettings gamesettings = this.mc.gameSettings;
         this.fontRenderer
-            .drawString(
-                I18n.format(
-                    "demo.help.movementShort",
-                    GameSettings.getKeyDisplayString(gamesettings.keyBindForward.getKeyCode()),
-                    GameSettings.getKeyDisplayString(gamesettings.keyBindLeft.getKeyCode()),
-                    GameSettings.getKeyDisplayString(gamesettings.keyBindBack.getKeyCode()),
-                    GameSettings.getKeyDisplayString(gamesettings.keyBindRight.getKeyCode())
-                ),
-                i,
-                j,
-                5197647
-            );
+        .drawString(
+            I18n.format(
+                "demo.help.movementShort",
+                gamesettings.keyBindForward.getDisplayName(),
+                gamesettings.keyBindLeft.getDisplayName(),
+                gamesettings.keyBindBack.getDisplayName(),
+                gamesettings.keyBindRight.getDisplayName()
+            ),
+            i,
+            j,
+            5197647
+        );
         this.fontRenderer.drawString(I18n.format("demo.help.movementMouse"), i, j + 12, 5197647);
         this.fontRenderer
-            .drawString(I18n.format("demo.help.jump", GameSettings.getKeyDisplayString(gamesettings.keyBindJump.getKeyCode())), i, j + 24, 5197647);
+        .drawString(I18n.format("demo.help.jump", gamesettings.keyBindJump.getDisplayName()), i, j + 24, 5197647);
         this.fontRenderer
-            .drawString(I18n.format("demo.help.inventory", GameSettings.getKeyDisplayString(gamesettings.keyBindInventory.getKeyCode())), i, j + 36, 5197647);
+        .drawString(I18n.format("demo.help.inventory", gamesettings.keyBindInventory.getDisplayName()), i, j + 36, 5197647);
         this.fontRenderer.drawSplitString(I18n.format("demo.help.fullWrapped"), i, j + 68, 218, 2039583);
         super.drawScreen(mouseX, mouseY, partialTicks);
     }
