--- original/net/minecraft/client/gui/GuiTextField.java
+++ modified/net/minecraft/client/gui/GuiTextField.java
@@ -519,6 +519,11 @@
         }
 
         this.isFocused = isFocusedIn;
+        com.cleanroommc.client.sdl.Window window = com.cleanroommc.client.sdl.Window.main();
+        if (window != null) {
+            window.textInput(this.isFocused);
+        }
+
         if (Minecraft.getMinecraft().currentScreen != null) {
             Minecraft.getMinecraft().currentScreen.setFocused(isFocusedIn);
         }
@@ -592,4 +597,8 @@
     public void setVisible(boolean isVisible) {
         this.visible = isVisible;
     }
+
+    public int getLineScrollOffset() {
+        return lineScrollOffset;
+    }
 }
