--- before/net/minecraft/item/EnumRarity.java
+++ after/net/minecraft/item/EnumRarity.java
@@ -2,7 +2,7 @@
 
 import net.minecraft.util.text.TextFormatting;
 
-public enum EnumRarity
+public enum EnumRarity implements net.minecraftforge.common.IRarity
 {
     COMMON(TextFormatting.WHITE, "Common"),
     UNCOMMON(TextFormatting.YELLOW, "Uncommon"),
@@ -16,5 +16,17 @@
     {
         this.color = color;
         this.rarityName = name;
+    }
+
+    @Override
+    public TextFormatting getColor()
+    {
+        return this.color;
+    }
+
+    @Override
+    public String getName()
+    {
+        return this.rarityName;
     }
 }
