--- before/net/minecraft/item/ItemBoat.java
+++ after/net/minecraft/item/ItemBoat.java
@@ -39,10 +39,10 @@
         double d1 = p_77659_2_.field_70167_r + (p_77659_2_.field_70163_u - p_77659_2_.field_70167_r) * 1.0D + (double)p_77659_2_.func_70047_e();
         double d2 = p_77659_2_.field_70166_s + (p_77659_2_.field_70161_v - p_77659_2_.field_70166_s) * 1.0D;
         Vec3d vec3d = new Vec3d(d0, d1, d2);
-        float f3 = MathHelper.func_76134_b(-f2 * ((float)Math.PI / 180F) - (float)Math.PI);
-        float f4 = MathHelper.func_76126_a(-f2 * ((float)Math.PI / 180F) - (float)Math.PI);
-        float f5 = -MathHelper.func_76134_b(-f1 * ((float)Math.PI / 180F));
-        float f6 = MathHelper.func_76126_a(-f1 * ((float)Math.PI / 180F));
+        float f3 = MathHelper.func_76134_b(-f2 * 0.017453292F - (float)Math.PI);
+        float f4 = MathHelper.func_76126_a(-f2 * 0.017453292F - (float)Math.PI);
+        float f5 = -MathHelper.func_76134_b(-f1 * 0.017453292F);
+        float f6 = MathHelper.func_76126_a(-f1 * 0.017453292F);
         float f7 = f4 * f5;
         float f8 = f3 * f5;
         double d3 = 5.0D;
@@ -51,7 +51,7 @@
 
         if (raytraceresult == null)
         {
-            return new ActionResult<>(EnumActionResult.PASS, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
         }
         else
         {
@@ -76,11 +76,11 @@
 
             if (flag)
             {
-                return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
             }
             else if (raytraceresult.field_72313_a != RayTraceResult.Type.BLOCK)
             {
-                return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
             }
             else
             {
@@ -92,7 +92,7 @@
 
                 if (!p_77659_1_.func_184144_a(entityboat, entityboat.func_174813_aQ().func_186662_g(-0.1D)).isEmpty())
                 {
-                    return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
                 }
                 else
                 {
@@ -107,7 +107,7 @@
                     }
 
                     p_77659_2_.func_71029_a(StatList.func_188057_b(this));
-                    return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);
                 }
             }
         }
