--- before/net/minecraft/item/ItemBlockSpecial.java
+++ after/net/minecraft/item/ItemBlockSpecial.java
@@ -52,10 +52,10 @@
 
         if (!itemstack.isEmpty()
                 && player.canPlayerEdit(pos, facing, itemstack)
-                && worldIn.mayPlace(this.block, pos, false, facing, null))
+                && worldIn.mayPlace(this.block, pos, false, facing, player))
         {
             IBlockState iblockstate1 = this.block
-                                       .getStateForPlacement(worldIn, pos, facing, hitX, hitY, hitZ, 0, player);
+                                       .getStateForPlacement(worldIn, pos, facing, hitX, hitY, hitZ, 0, player, hand);
 
             if (!worldIn.setBlockState(pos, iblockstate1, 11))
             {
@@ -76,7 +76,7 @@
                     }
                 }
 
-                SoundType soundtype = this.block.getSoundType();
+                SoundType soundtype = iblockstate1.getBlock().getSoundType(iblockstate1, worldIn, pos, player);
                 worldIn.playSound(
                     player,
                     pos,
@@ -93,5 +93,15 @@
         {
             return EnumActionResult.FAIL;
         }
+    }
+
+    public Block getBlock()
+    {
+        return this.getBlockRaw() == null ? null : this.getBlockRaw().delegate.get();
+    }
+
+    private Block getBlockRaw()
+    {
+        return this.block;
     }
 }
