--- before/net/minecraft/item/ItemEndCrystal.java
+++ after/net/minecraft/item/ItemEndCrystal.java
@@ -47,7 +47,7 @@
             {
                 BlockPos blockpos1 = blockpos.func_177984_a();
                 boolean flag = !p_180614_2_.func_175623_d(blockpos) && !p_180614_2_.func_180495_p(blockpos).func_177230_c().func_176200_f(p_180614_2_, blockpos);
-                flag |= !p_180614_2_.func_175623_d(blockpos1) && !p_180614_2_.func_180495_p(blockpos1).func_177230_c().func_176200_f(p_180614_2_, blockpos1);
+                flag = flag | (!p_180614_2_.func_175623_d(blockpos1) && !p_180614_2_.func_180495_p(blockpos1).func_177230_c().func_176200_f(p_180614_2_, blockpos1));
 
                 if (flag)
                 {
