--- before/net/minecraft/item/ItemCompass.java
+++ after/net/minecraft/item/ItemCompass.java
@@ -46,7 +46,7 @@
                     {
                         double d1 = flag ? (double)entity.field_70177_z : this.func_185094_a((EntityItemFrame)entity);
                         d1 = MathHelper.func_191273_b(d1 / 360.0D, 1.0D);
-                        double d2 = this.func_185092_a(p_185085_2_, entity) / (double)((float)Math.PI * 2F);
+                        double d2 = this.func_185092_a(p_185085_2_, entity) / (Math.PI * 2D);
                         d0 = 0.5D - (d1 - 0.25D - d2);
                     }
                     else
