--- before/net/minecraft/item/ItemLilyPad.java
+++ after/net/minecraft/item/ItemLilyPad.java
@@ -54,6 +54,15 @@
                         && iblockstate.getValue(BlockLiquid.LEVEL) == 0
                         && worldIn.isAirBlock(blockpos1))
                 {
+                    // special case for handling block placement with water lilies
+                    net.minecraftforge.common.util.BlockSnapshot blocksnapshot = net.minecraftforge.common.util.BlockSnapshot.getBlockSnapshot(worldIn, blockpos1);
+                    worldIn.setBlockState(blockpos1, Blocks.WATERLILY.getDefaultState());
+                    if (net.minecraftforge.event.ForgeEventFactory.onPlayerBlockPlace(playerIn, blocksnapshot, net.minecraft.util.EnumFacing.UP, handIn).isCanceled())
+                    {
+                        blocksnapshot.restore(true, false);
+                        return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
+                    }
+
                     worldIn.setBlockState(blockpos1, Blocks.WATERLILY.getDefaultState(), 11);
 
                     if (playerIn instanceof EntityPlayerMP)
