--- before/net/minecraft/item/ItemElytra.java
+++ after/net/minecraft/item/ItemElytra.java
@@ -54,11 +54,11 @@
         {
             p_77659_2_.func_184201_a(entityequipmentslot, itemstack.func_77946_l());
             itemstack.func_190920_e(0);
-            return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);
         }
         else
         {
-            return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
         }
     }
 }
