--- before/net/minecraft/item/ItemMonsterPlacer.java
+++ after/net/minecraft/item/ItemMonsterPlacer.java
@@ -11,6 +11,7 @@
 import net.minecraft.entity.EntityList;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.entity.IEntityLivingData;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.init.Blocks;
 import net.minecraft.nbt.NBTTagCompound;
@@ -39,7 +40,6 @@
         this.func_77637_a(CreativeTabs.field_78026_f);
     }
 
-    @Override
     public String func_77653_i(ItemStack p_77653_1_)
     {
         String s = ("" + I18n.func_74838_a(this.func_77658_a() + ".name")).trim();
@@ -53,17 +53,7 @@
         return s;
     }
 
-    @Override
-    public EnumActionResult func_180614_a(
-        EntityPlayer p_180614_1_,
-        World p_180614_2_,
-        BlockPos p_180614_3_,
-        EnumHand p_180614_4_,
-        EnumFacing p_180614_5_,
-        float p_180614_6_,
-        float p_180614_7_,
-        float p_180614_8_
-    )
+    public EnumActionResult func_180614_a(EntityPlayer p_180614_1_, World p_180614_2_, BlockPos p_180614_3_, EnumHand p_180614_4_, EnumFacing p_180614_5_, float p_180614_6_, float p_180614_7_, float p_180614_8_)
     {
         ItemStack itemstack = p_180614_1_.func_184586_b(p_180614_4_);
 
@@ -102,13 +92,7 @@
 
             BlockPos blockpos = p_180614_3_.func_177972_a(p_180614_5_);
             double d0 = this.func_190909_a(p_180614_2_, blockpos);
-            Entity entity = func_77840_a(
-                                p_180614_2_,
-                                func_190908_h(itemstack),
-                                (double)blockpos.func_177958_n() + 0.5,
-                                (double)blockpos.func_177956_o() + d0,
-                                (double)blockpos.func_177952_p() + 0.5
-                            );
+            Entity entity = func_77840_a(p_180614_2_, func_190908_h(itemstack), (double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + d0, (double)blockpos.func_177952_p() + 0.5D);
 
             if (entity != null)
             {
@@ -131,12 +115,12 @@
 
     protected double func_190909_a(World p_190909_1_, BlockPos p_190909_2_)
     {
-        AxisAlignedBB axisalignedbb = new AxisAlignedBB(p_190909_2_).func_72321_a(0.0, -1.0, 0.0);
-        List<AxisAlignedBB> list = p_190909_1_.func_184144_a(null, axisalignedbb);
+        AxisAlignedBB axisalignedbb = (new AxisAlignedBB(p_190909_2_)).func_72321_a(0.0D, -1.0D, 0.0D);
+        List<AxisAlignedBB> list = p_190909_1_.func_184144_a((Entity)null, axisalignedbb);
 
         if (list.isEmpty())
         {
-            return 0.0;
+            return 0.0D;
         }
         else
         {
@@ -161,9 +145,7 @@
 
             if (nbttagcompound != null && nbttagcompound.func_150297_b("EntityTag", 10))
             {
-                if (!p_185079_0_.field_72995_K
-                        && p_185079_3_.func_184213_bq()
-                        && (p_185079_1_ == null || !minecraftserver.func_184103_al().func_152596_g(p_185079_1_.func_146103_bH())))
+                if (!p_185079_0_.field_72995_K && p_185079_3_.func_184213_bq() && (p_185079_1_ == null || !minecraftserver.func_184103_al().func_152596_g(p_185079_1_.func_146103_bH())))
                 {
                     return;
                 }
@@ -177,14 +159,13 @@
         }
     }
 
-    @Override
     public ActionResult<ItemStack> func_77659_a(World p_77659_1_, EntityPlayer p_77659_2_, EnumHand p_77659_3_)
     {
         ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
 
         if (p_77659_1_.field_72995_K)
         {
-            return new ActionResult<>(EnumActionResult.PASS, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
         }
         else
         {
@@ -196,21 +177,15 @@
 
                 if (!(p_77659_1_.func_180495_p(blockpos).func_177230_c() instanceof BlockLiquid))
                 {
-                    return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
                 }
                 else if (p_77659_1_.func_175660_a(p_77659_2_, blockpos) && p_77659_2_.func_175151_a(blockpos, raytraceresult.field_178784_b, itemstack))
                 {
-                    Entity entity = func_77840_a(
-                                        p_77659_1_,
-                                        func_190908_h(itemstack),
-                                        (double)blockpos.func_177958_n() + 0.5,
-                                        (double)blockpos.func_177956_o() + 0.5,
-                                        (double)blockpos.func_177952_p() + 0.5
-                                    );
+                    Entity entity = func_77840_a(p_77659_1_, func_190908_h(itemstack), (double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 0.5D, (double)blockpos.func_177952_p() + 0.5D);
 
                     if (entity == null)
                     {
-                        return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                        return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
                     }
                     else
                     {
@@ -227,17 +202,17 @@
                         }
 
                         p_77659_2_.func_71029_a(StatList.func_188057_b(this));
-                        return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
+                        return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);
                     }
                 }
                 else
                 {
-                    return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
                 }
             }
             else
             {
-                return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
             }
         }
     }
@@ -259,7 +234,8 @@
                     entity.func_70012_b(p_77840_2_, p_77840_4_, p_77840_6_, MathHelper.func_76142_g(p_77840_0_.field_73012_v.nextFloat() * 360.0F), 0.0F);
                     entityliving.field_70759_as = entityliving.field_70177_z;
                     entityliving.field_70761_aq = entityliving.field_70177_z;
-                    entityliving.func_180482_a(p_77840_0_.func_175649_E(new BlockPos(entityliving)), null);
+                    if (net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(entityliving, p_77840_0_, (float) p_77840_2_, (float) p_77840_4_, (float) p_77840_6_, null)) return null;
+                    entityliving.func_180482_a(p_77840_0_.func_175649_E(new BlockPos(entityliving)), (IEntityLivingData)null);
                     p_77840_0_.func_72838_d(entity);
                     entityliving.func_70642_aH();
                 }
@@ -273,7 +249,6 @@
         }
     }
 
-    @Override
     public void func_150895_a(CreativeTabs p_150895_1_, NonNullList<ItemStack> p_150895_2_)
     {
         if (this.func_194125_a(p_150895_1_))
