--- before/net/minecraft/item/ItemPotion.java
+++ after/net/minecraft/item/ItemPotion.java
@@ -100,7 +100,7 @@
     public ActionResult<ItemStack> func_77659_a(World p_77659_1_, EntityPlayer p_77659_2_, EnumHand p_77659_3_)
     {
         p_77659_2_.func_184598_c(p_77659_3_);
-        return new ActionResult<>(EnumActionResult.SUCCESS, p_77659_2_.func_184586_b(p_77659_3_));
+        return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, p_77659_2_.func_184586_b(p_77659_3_));
     }
 
     public String func_77653_i(ItemStack p_77653_1_)
@@ -114,12 +114,6 @@
         PotionUtils.func_185182_a(p_77624_1_, p_77624_3_, 1.0F);
     }
 
-    @SideOnly(Side.CLIENT)
-    public boolean func_77636_d(ItemStack p_77636_1_)
-    {
-        return super.func_77636_d(p_77636_1_) || !PotionUtils.func_185189_a(p_77636_1_).isEmpty();
-    }
-
     public void func_150895_a(CreativeTabs p_150895_1_, NonNullList<ItemStack> p_150895_2_)
     {
         if (this.func_194125_a(p_150895_1_))
@@ -132,5 +126,11 @@
                 }
             }
         }
+    }
+
+    @SideOnly(Side.CLIENT)
+    public boolean func_77636_d(ItemStack p_77636_1_)
+    {
+        return super.func_77636_d(p_77636_1_) || !PotionUtils.func_185189_a(p_77636_1_).isEmpty();
     }
 }
