--- before/net/minecraft/item/ItemSlab.java
+++ after/net/minecraft/item/ItemSlab.java
@@ -82,7 +82,7 @@
                             && worldIn.checkNoEntityCollision(axisalignedbb.offset(pos))
                             && worldIn.setBlockState(pos, iblockstate1, 11))
                     {
-                        SoundType soundtype = this.doubleSlab.getSoundType();
+                        SoundType soundtype = this.doubleSlab.getSoundType(iblockstate1, worldIn, pos, player);
                         worldIn.playSound(
                             player,
                             pos,
@@ -155,7 +155,7 @@
                         && worldIn.checkNoEntityCollision(axisalignedbb.offset(pos))
                         && worldIn.setBlockState(pos, iblockstate1, 11))
                 {
-                    SoundType soundtype = this.doubleSlab.getSoundType();
+                    SoundType soundtype = this.doubleSlab.getSoundType(iblockstate1, worldIn, pos, player);
                     worldIn.playSound(
                         player,
                         pos,
