--- before/net/minecraft/item/ItemBucket.java
+++ after/net/minecraft/item/ItemBucket.java
@@ -35,20 +35,21 @@
         this.func_77637_a(CreativeTabs.field_78026_f);
     }
 
-    @Override
     public ActionResult<ItemStack> func_77659_a(World p_77659_1_, EntityPlayer p_77659_2_, EnumHand p_77659_3_)
     {
         boolean flag = this.field_77876_a == Blocks.field_150350_a;
         ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
         RayTraceResult raytraceresult = this.func_77621_a(p_77659_1_, p_77659_2_, flag);
+        ActionResult<ItemStack> ret = net.minecraftforge.event.ForgeEventFactory.onBucketUse(p_77659_2_, p_77659_1_, itemstack, raytraceresult);
+        if (ret != null) return ret;
 
         if (raytraceresult == null)
         {
-            return new ActionResult<>(EnumActionResult.PASS, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
         }
         else if (raytraceresult.field_72313_a != RayTraceResult.Type.BLOCK)
         {
-            return new ActionResult<>(EnumActionResult.PASS, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
         }
         else
         {
@@ -56,36 +57,36 @@
 
             if (!p_77659_1_.func_175660_a(p_77659_2_, blockpos))
             {
-                return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
             }
             else if (flag)
             {
                 if (!p_77659_2_.func_175151_a(blockpos.func_177972_a(raytraceresult.field_178784_b), raytraceresult.field_178784_b, itemstack))
                 {
-                    return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
                 }
                 else
                 {
                     IBlockState iblockstate = p_77659_1_.func_180495_p(blockpos);
                     Material material = iblockstate.func_185904_a();
 
-                    if (material == Material.field_151586_h && iblockstate.func_177229_b(BlockLiquid.field_176367_b) == 0)
+                    if (material == Material.field_151586_h && ((Integer)iblockstate.func_177229_b(BlockLiquid.field_176367_b)).intValue() == 0)
                     {
                         p_77659_1_.func_180501_a(blockpos, Blocks.field_150350_a.func_176223_P(), 11);
                         p_77659_2_.func_71029_a(StatList.func_188057_b(this));
                         p_77659_2_.func_184185_a(SoundEvents.field_187630_M, 1.0F, 1.0F);
-                        return new ActionResult<>(EnumActionResult.SUCCESS, this.func_150910_a(itemstack, p_77659_2_, Items.field_151131_as));
+                        return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, this.func_150910_a(itemstack, p_77659_2_, Items.field_151131_as));
                     }
-                    else if (material == Material.field_151587_i && iblockstate.func_177229_b(BlockLiquid.field_176367_b) == 0)
+                    else if (material == Material.field_151587_i && ((Integer)iblockstate.func_177229_b(BlockLiquid.field_176367_b)).intValue() == 0)
                     {
                         p_77659_2_.func_184185_a(SoundEvents.field_187633_N, 1.0F, 1.0F);
                         p_77659_1_.func_180501_a(blockpos, Blocks.field_150350_a.func_176223_P(), 11);
                         p_77659_2_.func_71029_a(StatList.func_188057_b(this));
-                        return new ActionResult<>(EnumActionResult.SUCCESS, this.func_150910_a(itemstack, p_77659_2_, Items.field_151129_at));
+                        return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, this.func_150910_a(itemstack, p_77659_2_, Items.field_151129_at));
                     }
                     else
                     {
-                        return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                        return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
                     }
                 }
             }
@@ -96,7 +97,7 @@
 
                 if (!p_77659_2_.func_175151_a(blockpos1, raytraceresult.field_178784_b, itemstack))
                 {
-                    return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
                 }
                 else if (this.func_180616_a(p_77659_2_, p_77659_1_, blockpos1))
                 {
@@ -106,13 +107,11 @@
                     }
 
                     p_77659_2_.func_71029_a(StatList.func_188057_b(this));
-                    return !p_77659_2_.field_71075_bZ.field_75098_d
-                           ? new ActionResult<>(EnumActionResult.SUCCESS, new ItemStack(Items.field_151133_ar))
-                           : new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
+                    return !p_77659_2_.field_71075_bZ.field_75098_d ? new ActionResult(EnumActionResult.SUCCESS, new ItemStack(Items.field_151133_ar)) : new ActionResult(EnumActionResult.SUCCESS, itemstack);
                 }
                 else
                 {
-                    return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
                 }
             }
         }
@@ -168,20 +167,11 @@
                     int l = p_180616_3_.func_177958_n();
                     int i = p_180616_3_.func_177956_o();
                     int j = p_180616_3_.func_177952_p();
-                    p_180616_2_.func_184133_a(
-                        p_180616_1_,
-                        p_180616_3_,
-                        SoundEvents.field_187646_bt,
-                        SoundCategory.BLOCKS,
-                        0.5F,
-                        2.6F + (p_180616_2_.field_73012_v.nextFloat() - p_180616_2_.field_73012_v.nextFloat()) * 0.8F
-                    );
+                    p_180616_2_.func_184133_a(p_180616_1_, p_180616_3_, SoundEvents.field_187646_bt, SoundCategory.BLOCKS, 0.5F, 2.6F + (p_180616_2_.field_73012_v.nextFloat() - p_180616_2_.field_73012_v.nextFloat()) * 0.8F);
 
-                    for (int k = 0; k < 8; k++)
+                    for (int k = 0; k < 8; ++k)
                     {
-                        p_180616_2_.func_175688_a(
-                            EnumParticleTypes.SMOKE_LARGE, (double)l + Math.random(), (double)i + Math.random(), (double)j + Math.random(), 0.0, 0.0, 0.0
-                        );
+                        p_180616_2_.func_175688_a(EnumParticleTypes.SMOKE_LARGE, (double)l + Math.random(), (double)i + Math.random(), (double)j + Math.random(), 0.0D, 0.0D, 0.0D);
                     }
                 }
                 else
@@ -198,6 +188,18 @@
 
                 return true;
             }
+        }
+    }
+
+    @Override
+    public net.minecraftforge.common.capabilities.ICapabilityProvider initCapabilities(ItemStack stack, @Nullable net.minecraft.nbt.NBTTagCompound nbt) {
+        if (this.getClass() == ItemBucket.class)
+        {
+            return new net.minecraftforge.fluids.capability.wrappers.FluidBucketWrapper(stack);
+        }
+        else
+        {
+            return super.initCapabilities(stack, nbt);
         }
     }
 }
