--- before/net/minecraft/item/ItemRedstone.java
+++ after/net/minecraft/item/ItemRedstone.java
@@ -35,7 +35,7 @@
         ItemStack itemstack = player.getHeldItem(hand);
 
         if (player.canPlayerEdit(blockpos, facing, itemstack)
-                && worldIn.mayPlace(worldIn.getBlockState(blockpos).getBlock(), blockpos, false, facing, null)
+                && worldIn.mayPlace(worldIn.getBlockState(blockpos).getBlock(), blockpos, false, facing, player)
                 && Blocks.REDSTONE_WIRE.canPlaceBlockAt(worldIn, blockpos))
         {
             worldIn.setBlockState(blockpos, Blocks.REDSTONE_WIRE.getDefaultState());
