--- before/net/minecraft/item/ItemAxe.java
+++ after/net/minecraft/item/ItemAxe.java
@@ -32,7 +32,13 @@
         this.attackSpeed = ATTACK_SPEEDS[material.ordinal()];
     }
 
-    @Override
+    protected ItemAxe(Item.ToolMaterial material, float damage, float speed)
+    {
+        super(material, EFFECTIVE_ON);
+        this.attackDamage = damage;
+        this.attackSpeed = speed;
+    }
+
     public float getDestroySpeed(ItemStack stack, IBlockState state)
     {
         Material material = state.getMaterial();
