--- before/net/minecraft/item/ItemSkull.java
+++ after/net/minecraft/item/ItemSkull.java
@@ -52,13 +52,18 @@
         }
         else
         {
+            if (worldIn.getBlockState(pos).getBlock().isReplaceable(worldIn, pos))
+            {
+                facing = EnumFacing.UP;
+                pos = pos.down();
+            }
             IBlockState iblockstate = worldIn.getBlockState(pos);
             Block block = iblockstate.getBlock();
             boolean flag = block.isReplaceable(worldIn, pos);
 
             if (!flag)
             {
-                if (!worldIn.getBlockState(pos).getMaterial().isSolid())
+                if (!worldIn.getBlockState(pos).getMaterial().isSolid() && !worldIn.isSideSolid(pos, facing, true))
                 {
                     return EnumActionResult.FAIL;
                 }
