--- before/net/minecraft/item/ItemWritableBook.java
+++ after/net/minecraft/item/ItemWritableBook.java
@@ -21,7 +21,7 @@
         ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
         p_77659_2_.func_184814_a(itemstack, p_77659_3_);
         p_77659_2_.func_71029_a(StatList.func_188057_b(this));
-        return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
+        return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);
     }
 
     public static boolean func_150930_a(NBTTagCompound p_150930_0_)
