--- before/net/minecraft/item/ItemMinecart.java
+++ after/net/minecraft/item/ItemMinecart.java
@@ -32,7 +32,7 @@
             BlockPos blockpos = source.getBlockPos().offset(enumfacing);
             IBlockState iblockstate = world.getBlockState(blockpos);
             BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = iblockstate.getBlock() instanceof BlockRailBase
-                    ? iblockstate.getValue(((BlockRailBase)iblockstate.getBlock()).getShapeProperty())
+                    ? ((BlockRailBase)iblockstate.getBlock()).getRailDirection(world, blockpos, iblockstate, null)
                     : BlockRailBase.EnumRailDirection.NORTH_SOUTH;
             double d3;
 
@@ -56,7 +56,7 @@
 
                 IBlockState iblockstate1 = world.getBlockState(blockpos.down());
                 BlockRailBase.EnumRailDirection blockrailbase$enumraildirection1 = iblockstate1.getBlock() instanceof BlockRailBase
-                        ? iblockstate1.getValue(((BlockRailBase)iblockstate1.getBlock()).getShapeProperty())
+                        ? ((BlockRailBase)iblockstate1.getBlock()).getRailDirection(world, blockpos.down(), iblockstate1, null)
                         : BlockRailBase.EnumRailDirection.NORTH_SOUTH;
 
                 if (enumfacing != EnumFacing.DOWN && blockrailbase$enumraildirection1.isAscending())
@@ -121,7 +121,7 @@
             if (!worldIn.isRemote)
             {
                 BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = iblockstate.getBlock() instanceof BlockRailBase
-                        ? iblockstate.getValue(((BlockRailBase)iblockstate.getBlock()).getShapeProperty())
+                        ? ((BlockRailBase)iblockstate.getBlock()).getRailDirection(worldIn, pos, iblockstate, null)
                         : BlockRailBase.EnumRailDirection.NORTH_SOUTH;
                 double d0 = 0.0;
 
