--- before/net/minecraft/item/ItemArmor.java
+++ after/net/minecraft/item/ItemArmor.java
@@ -21,6 +21,7 @@
 import net.minecraft.util.ActionResult;
 import net.minecraft.util.EntitySelectors;
 import net.minecraft.util.EnumActionResult;
+import net.minecraft.util.EnumFacing;
 import net.minecraft.util.EnumHand;
 import net.minecraft.util.SoundEvent;
 import net.minecraft.util.math.AxisAlignedBB;
@@ -32,23 +33,10 @@
 public class ItemArmor extends Item
 {
     private static final int[] field_77882_bY = new int[] {13, 15, 16, 11};
-    private static final UUID[] field_185084_n = new UUID[]
-        {
-            UUID.fromString("845DB27C-C624-495F-8C9F-6020A9A58B6B"),
-            UUID.fromString("D8499B04-0E66-4726-AB29-64469D734E0D"),
-            UUID.fromString("9F3D476D-C118-4544-8365-64846904B48E"),
-            UUID.fromString("2AD3F246-FEE1-4E67-B886-69FD380BB150")
-        };
-    public static final String[] field_94603_a = new String[]
-        {
-            "minecraft:items/empty_armor_slot_boots",
-            "minecraft:items/empty_armor_slot_leggings",
-            "minecraft:items/empty_armor_slot_chestplate",
-            "minecraft:items/empty_armor_slot_helmet"
-        };
+    private static final UUID[] field_185084_n = new UUID[] {UUID.fromString("845DB27C-C624-495F-8C9F-6020A9A58B6B"), UUID.fromString("D8499B04-0E66-4726-AB29-64469D734E0D"), UUID.fromString("9F3D476D-C118-4544-8365-64846904B48E"), UUID.fromString("2AD3F246-FEE1-4E67-B886-69FD380BB150")};
+    public static final String[] field_94603_a = new String[] {"minecraft:items/empty_armor_slot_boots", "minecraft:items/empty_armor_slot_leggings", "minecraft:items/empty_armor_slot_chestplate", "minecraft:items/empty_armor_slot_helmet"};
     public static final IBehaviorDispenseItem field_96605_cw = new BehaviorDefaultDispenseItem()
     {
-        @Override
         protected ItemStack func_82487_b(IBlockSource p_82487_1_, ItemStack p_82487_2_)
         {
             ItemStack itemstack = ItemArmor.func_185082_a(p_82487_1_, p_82487_2_);
@@ -63,11 +51,8 @@
 
     public static ItemStack func_185082_a(IBlockSource p_185082_0_, ItemStack p_185082_1_)
     {
-        BlockPos blockpos = p_185082_0_.func_180699_d().func_177972_a(p_185082_0_.func_189992_e().func_177229_b(BlockDispenser.field_176441_a));
-        List<EntityLivingBase> list = p_185082_0_.func_82618_k()
-                                      .func_175647_a(
-                                          EntityLivingBase.class, new AxisAlignedBB(blockpos), Predicates.and(EntitySelectors.field_180132_d, new EntitySelectors.ArmoredMob(p_185082_1_))
-                                      );
+        BlockPos blockpos = p_185082_0_.func_180699_d().func_177972_a((EnumFacing)p_185082_0_.func_189992_e().func_177229_b(BlockDispenser.field_176441_a));
+        List<EntityLivingBase> list = p_185082_0_.func_82618_k().<EntityLivingBase>func_175647_a(EntityLivingBase.class, new AxisAlignedBB(blockpos), Predicates.and(EntitySelectors.field_180132_d, new EntitySelectors.ArmoredMob(p_185082_1_)));
 
         if (list.isEmpty())
         {
@@ -108,7 +93,6 @@
         return this.field_77881_a;
     }
 
-    @Override
     public int func_77619_b()
     {
         return this.field_77878_bZ.func_78045_a();
@@ -128,9 +112,7 @@
         else
         {
             NBTTagCompound nbttagcompound = p_82816_1_.func_77978_p();
-            return nbttagcompound != null && nbttagcompound.func_150297_b("display", 10)
-                   ? nbttagcompound.func_74775_l("display").func_150297_b("color", 3)
-                   : false;
+            return nbttagcompound != null && nbttagcompound.func_150297_b("display", 10) ? nbttagcompound.func_74775_l("display").func_150297_b("color", 3) : false;
         }
     }
 
@@ -203,13 +185,13 @@
         }
     }
 
-    @Override
     public boolean func_82789_a(ItemStack p_82789_1_, ItemStack p_82789_2_)
     {
-        return this.field_77878_bZ.func_151685_b() == p_82789_2_.func_77973_b() ? true : super.func_82789_a(p_82789_1_, p_82789_2_);
+        ItemStack mat = this.field_77878_bZ.getRepairItemStack();
+        if (!mat.func_190926_b() && net.minecraftforge.oredict.OreDictionary.itemMatches(mat,p_82789_2_,false)) return true;
+        return super.func_82789_a(p_82789_1_, p_82789_2_);
     }
 
-    @Override
     public ActionResult<ItemStack> func_77659_a(World p_77659_1_, EntityPlayer p_77659_2_, EnumHand p_77659_3_)
     {
         ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
@@ -220,34 +202,40 @@
         {
             p_77659_2_.func_184201_a(entityequipmentslot, itemstack.func_77946_l());
             itemstack.func_190920_e(0);
-            return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);
         }
         else
         {
-            return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
         }
     }
 
-    @Override
     public Multimap<String, AttributeModifier> func_111205_h(EntityEquipmentSlot p_111205_1_)
     {
         Multimap<String, AttributeModifier> multimap = super.func_111205_h(p_111205_1_);
 
         if (p_111205_1_ == this.field_77881_a)
         {
-            multimap.put(
-                SharedMonsterAttributes.field_188791_g.func_111108_a(),
-                new AttributeModifier(field_185084_n[p_111205_1_.func_188454_b()], "Armor modifier", (double)this.field_77879_b, 0)
-            );
-            multimap.put(
-                SharedMonsterAttributes.field_189429_h.func_111108_a(),
-                new AttributeModifier(field_185084_n[p_111205_1_.func_188454_b()], "Armor toughness", (double)this.field_189415_e, 0)
-            );
+            multimap.put(SharedMonsterAttributes.field_188791_g.func_111108_a(), new AttributeModifier(field_185084_n[p_111205_1_.func_188454_b()], "Armor modifier", (double)this.field_77879_b, 0));
+            multimap.put(SharedMonsterAttributes.field_189429_h.func_111108_a(), new AttributeModifier(field_185084_n[p_111205_1_.func_188454_b()], "Armor toughness", (double)this.field_189415_e, 0));
         }
 
         return multimap;
     }
 
+    /**
+     * Determines if this armor will be rendered with the secondary 'overlay' texture.
+     * If this is true, the first texture will be rendered using a tint of the color
+     * specified by getColor(ItemStack)
+     *
+     * @param stack The stack
+     * @return true/false
+     */
+    public boolean hasOverlay(ItemStack stack)
+    {
+        return this.field_77878_bZ == ItemArmor.ArmorMaterial.LEATHER || func_82814_b(stack) != 0x00FFFFFF;
+    }
+
     public static enum ArmorMaterial
     {
         LEATHER("leather", 5, new int[]{1, 2, 3, 1}, 15, SoundEvents.field_187728_s, 0.0F),
@@ -262,6 +250,8 @@
         private final int field_78055_h;
         private final SoundEvent field_185020_j;
         private final float field_189417_k;
+        //Added by forge for custom Armor materials.
+        public ItemStack repairMaterial = ItemStack.field_190927_a;
 
         private ArmorMaterial(String p_i47117_3_, int p_i47117_4_, int[] p_i47117_5_, int p_i47117_6_, SoundEvent p_i47117_7_, float p_i47117_8_)
         {
@@ -293,6 +283,7 @@
             return this.field_185020_j;
         }
 
+        @Deprecated // Use getRepairItemStack below
         public Item func_151685_b()
         {
             if (this == LEATHER)
@@ -326,6 +317,22 @@
         public float func_189416_e()
         {
             return this.field_189417_k;
+        }
+
+        public ArmorMaterial setRepairItem(ItemStack stack)
+        {
+            if (!this.repairMaterial.func_190926_b()) throw new RuntimeException("Repair material has already been set");
+            if (this == LEATHER || this == CHAIN || this == GOLD || this == IRON || this == DIAMOND) throw new RuntimeException("Can not change vanilla armor repair materials");
+            this.repairMaterial = stack;
+            return this;
+        }
+
+        public ItemStack getRepairItemStack()
+        {
+            if (!repairMaterial.func_190926_b()) return repairMaterial;
+            Item ret = this.func_151685_b();
+            if (ret != null) repairMaterial = new ItemStack(ret,1,net.minecraftforge.oredict.OreDictionary.WILDCARD_VALUE);
+            return repairMaterial;
         }
     }
 }
