--- before/net/minecraft/item/ItemBed.java
+++ after/net/minecraft/item/ItemBed.java
@@ -29,17 +29,7 @@
         this.func_77627_a(true);
     }
 
-    @Override
-    public EnumActionResult func_180614_a(
-        EntityPlayer p_180614_1_,
-        World p_180614_2_,
-        BlockPos p_180614_3_,
-        EnumHand p_180614_4_,
-        EnumFacing p_180614_5_,
-        float p_180614_6_,
-        float p_180614_7_,
-        float p_180614_8_
-    )
+    public EnumActionResult func_180614_a(EntityPlayer p_180614_1_, World p_180614_2_, BlockPos p_180614_3_, EnumHand p_180614_4_, EnumFacing p_180614_5_, float p_180614_6_, float p_180614_7_, float p_180614_8_)
     {
         if (p_180614_2_.field_72995_K)
         {
@@ -60,7 +50,7 @@
                 p_180614_3_ = p_180614_3_.func_177984_a();
             }
 
-            int i = MathHelper.func_76128_c((double)(p_180614_1_.field_70177_z * 4.0F / 360.0F) + 0.5) & 3;
+            int i = MathHelper.func_76128_c((double)(p_180614_1_.field_70177_z * 4.0F / 360.0F) + 0.5D) & 3;
             EnumFacing enumfacing = EnumFacing.func_176731_b(i);
             BlockPos blockpos = p_180614_3_.func_177972_a(enumfacing);
             ItemStack itemstack = p_180614_1_.func_184586_b(p_180614_4_);
@@ -72,27 +62,13 @@
                 boolean flag2 = flag || p_180614_2_.func_175623_d(p_180614_3_);
                 boolean flag3 = flag1 || p_180614_2_.func_175623_d(blockpos);
 
-                if (flag2
-                        && flag3
-                        && p_180614_2_.func_180495_p(p_180614_3_.func_177977_b()).func_185896_q()
-                        && p_180614_2_.func_180495_p(blockpos.func_177977_b()).func_185896_q())
+                if (flag2 && flag3 && p_180614_2_.func_180495_p(p_180614_3_.func_177977_b()).func_185896_q() && p_180614_2_.func_180495_p(blockpos.func_177977_b()).func_185896_q())
                 {
-                    IBlockState iblockstate2 = Blocks.field_150324_C
-                                               .func_176223_P()
-                                               .func_177226_a(BlockBed.field_176471_b, false)
-                                               .func_177226_a(BlockBed.field_185512_D, enumfacing)
-                                               .func_177226_a(BlockBed.field_176472_a, BlockBed.EnumPartType.FOOT);
+                    IBlockState iblockstate2 = Blocks.field_150324_C.func_176223_P().func_177226_a(BlockBed.field_176471_b, Boolean.valueOf(false)).func_177226_a(BlockBed.field_185512_D, enumfacing).func_177226_a(BlockBed.field_176472_a, BlockBed.EnumPartType.FOOT);
                     p_180614_2_.func_180501_a(p_180614_3_, iblockstate2, 10);
                     p_180614_2_.func_180501_a(blockpos, iblockstate2.func_177226_a(BlockBed.field_176472_a, BlockBed.EnumPartType.HEAD), 10);
-                    SoundType soundtype = iblockstate2.func_177230_c().func_185467_w();
-                    p_180614_2_.func_184133_a(
-                        null,
-                        p_180614_3_,
-                        soundtype.func_185841_e(),
-                        SoundCategory.BLOCKS,
-                        (soundtype.func_185843_a() + 1.0F) / 2.0F,
-                        soundtype.func_185847_b() * 0.8F
-                    );
+                    SoundType soundtype = iblockstate2.func_177230_c().getSoundType(iblockstate2, p_180614_2_, p_180614_3_, p_180614_1_);
+                    p_180614_2_.func_184133_a((EntityPlayer)null, p_180614_3_, soundtype.func_185841_e(), SoundCategory.BLOCKS, (soundtype.func_185843_a() + 1.0F) / 2.0F, soundtype.func_185847_b() * 0.8F);
                     TileEntity tileentity = p_180614_2_.func_175625_s(blockpos);
 
                     if (tileentity instanceof TileEntityBed)
@@ -130,13 +106,11 @@
         }
     }
 
-    @Override
     public String func_77667_c(ItemStack p_77667_1_)
     {
         return super.func_77658_a() + "." + EnumDyeColor.func_176764_b(p_77667_1_.func_77960_j()).func_176762_d();
     }
 
-    @Override
     public void func_150895_a(CreativeTabs p_150895_1_, NonNullList<ItemStack> p_150895_2_)
     {
         if (this.func_194125_a(p_150895_1_))
