--- before/net/minecraft/item/ItemRecord.java
+++ after/net/minecraft/item/ItemRecord.java
@@ -23,7 +23,7 @@
 
 public class ItemRecord extends Item
 {
-    private static final Map<SoundEvent, ItemRecord> field_150928_b = Maps.newHashMap();
+    private static final Map<SoundEvent, ItemRecord> field_150928_b = Maps.<SoundEvent, ItemRecord>newHashMap();
     private final SoundEvent field_185076_b;
     private final String field_185077_c;
 
@@ -40,7 +40,7 @@
     {
         IBlockState iblockstate = p_180614_2_.func_180495_p(p_180614_3_);
 
-        if (iblockstate.func_177230_c() == Blocks.field_150421_aI && !iblockstate.func_177229_b(BlockJukebox.field_176432_a))
+        if (iblockstate.func_177230_c() == Blocks.field_150421_aI && !((Boolean)iblockstate.func_177229_b(BlockJukebox.field_176432_a)).booleanValue())
         {
             if (!p_180614_2_.field_72995_K)
             {
