--- before/net/minecraft/item/ItemSnow.java
+++ after/net/minecraft/item/ItemSnow.java
@@ -23,17 +23,7 @@
         this.func_77656_e(0);
     }
 
-    @Override
-    public EnumActionResult func_180614_a(
-        EntityPlayer p_180614_1_,
-        World p_180614_2_,
-        BlockPos p_180614_3_,
-        EnumHand p_180614_4_,
-        EnumFacing p_180614_5_,
-        float p_180614_6_,
-        float p_180614_7_,
-        float p_180614_8_
-    )
+    public EnumActionResult func_180614_a(EntityPlayer p_180614_1_, World p_180614_2_, BlockPos p_180614_3_, EnumHand p_180614_4_, EnumFacing p_180614_5_, float p_180614_6_, float p_180614_7_, float p_180614_8_)
     {
         ItemStack itemstack = p_180614_1_.func_184586_b(p_180614_4_);
 
@@ -43,35 +33,26 @@
             Block block = iblockstate.func_177230_c();
             BlockPos blockpos = p_180614_3_;
 
-            if ((p_180614_5_ != EnumFacing.UP || block != this.field_150939_a) && !block.func_176200_f(p_180614_2_, p_180614_3_))
+            if ((p_180614_5_ != EnumFacing.UP || block != this.getBlockRaw()) && !block.func_176200_f(p_180614_2_, p_180614_3_))
             {
                 blockpos = p_180614_3_.func_177972_a(p_180614_5_);
                 iblockstate = p_180614_2_.func_180495_p(blockpos);
                 block = iblockstate.func_177230_c();
             }
 
-            if (block == this.field_150939_a)
+            if (block == this.getBlockRaw())
             {
-                int i = iblockstate.func_177229_b(BlockSnow.field_176315_a);
+                int i = ((Integer)iblockstate.func_177229_b(BlockSnow.field_176315_a)).intValue();
 
                 if (i < 8)
                 {
-                    IBlockState iblockstate1 = iblockstate.func_177226_a(BlockSnow.field_176315_a, i + 1);
+                    IBlockState iblockstate1 = iblockstate.func_177226_a(BlockSnow.field_176315_a, Integer.valueOf(i + 1));
                     AxisAlignedBB axisalignedbb = iblockstate1.func_185890_d(p_180614_2_, blockpos);
 
-                    if (axisalignedbb != Block.field_185506_k
-                            && p_180614_2_.func_72855_b(axisalignedbb.func_186670_a(blockpos))
-                            && p_180614_2_.func_180501_a(blockpos, iblockstate1, 10))
+                    if (axisalignedbb != Block.field_185506_k && p_180614_2_.func_72855_b(axisalignedbb.func_186670_a(blockpos)) && p_180614_2_.func_180501_a(blockpos, iblockstate1, 10))
                     {
-                        SoundType soundtype = this.field_150939_a.func_185467_w();
-                        p_180614_2_.func_184133_a(
-                            p_180614_1_,
-                            blockpos,
-                            soundtype.func_185841_e(),
-                            SoundCategory.BLOCKS,
-                            (soundtype.func_185843_a() + 1.0F) / 2.0F,
-                            soundtype.func_185847_b() * 0.8F
-                        );
+                        SoundType soundtype = this.getBlockRaw().getSoundType(iblockstate1, p_180614_2_, p_180614_3_, p_180614_1_);
+                        p_180614_2_.func_184133_a(p_180614_1_, blockpos, soundtype.func_185841_e(), SoundCategory.BLOCKS, (soundtype.func_185843_a() + 1.0F) / 2.0F, soundtype.func_185847_b() * 0.8F);
 
                         if (p_180614_1_ instanceof EntityPlayerMP)
                         {
@@ -92,9 +73,14 @@
         }
     }
 
-    @Override
     public int func_77647_b(int p_77647_1_)
     {
         return p_77647_1_;
+    }
+
+    public boolean func_179222_a(World world, BlockPos pos, EnumFacing side, EntityPlayer player, ItemStack stack)
+    {
+        IBlockState state = world.func_180495_p(pos);
+        return (state.func_177230_c() != net.minecraft.init.Blocks.field_150431_aC || ((Integer)state.func_177229_b(BlockSnow.field_176315_a)) > 7) ? super.func_179222_a(world, pos, side, player, stack) : true;
     }
 }
