--- before/net/minecraft/item/ItemBucket.java
+++ after/net/minecraft/item/ItemBucket.java
@@ -41,6 +41,8 @@
         boolean flag = this.containedBlock == Blocks.AIR;
         ItemStack itemstack = playerIn.getHeldItem(handIn);
         RayTraceResult raytraceresult = this.rayTrace(worldIn, playerIn, flag);
+        ActionResult<ItemStack> ret = net.minecraftforge.event.ForgeEventFactory.onBucketUse(playerIn, worldIn, itemstack, raytraceresult);
+        if (ret != null) return ret;
 
         if (raytraceresult == null)
         {
@@ -196,6 +198,18 @@
 
                 return true;
             }
+        }
+    }
+
+    @Override
+    public net.minecraftforge.common.capabilities.ICapabilityProvider initCapabilities(ItemStack stack, @Nullable net.minecraft.nbt.NBTTagCompound nbt) {
+        if (this.getClass() == ItemBucket.class)
+        {
+            return new net.minecraftforge.fluids.capability.wrappers.FluidBucketWrapper(stack);
+        }
+        else
+        {
+            return super.initCapabilities(stack, nbt);
         }
     }
 }
