--- before/net/minecraft/item/ItemDoor.java
+++ after/net/minecraft/item/ItemDoor.java
@@ -58,7 +58,7 @@
                 int j = enumfacing.getZOffset();
                 boolean flag = i < 0 && hitZ < 0.5F || i > 0 && hitZ > 0.5F || j < 0 && hitX > 0.5F || j > 0 && hitX < 0.5F;
                 placeDoor(worldIn, pos, enumfacing, this.block, flag);
-                SoundType soundtype = this.block.getSoundType();
+                SoundType soundtype = worldIn.getBlockState(pos).getBlock().getSoundType(worldIn.getBlockState(pos), worldIn, pos, player);
                 worldIn.playSound(
                     player,
                     pos,
