--- before/net/minecraft/item/ItemMonsterPlacer.java
+++ after/net/minecraft/item/ItemMonsterPlacer.java
@@ -11,6 +11,7 @@
 import net.minecraft.entity.EntityList;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.entity.IEntityLivingData;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.init.Blocks;
 import net.minecraft.nbt.NBTTagCompound;
@@ -39,7 +40,6 @@
         this.setCreativeTab(CreativeTabs.MISC);
     }
 
-    @Override
     public String getItemStackDisplayName(ItemStack stack)
     {
         String s = ("" + I18n.translateToLocal(this.getTranslationKey() + ".name")).trim();
@@ -53,17 +53,7 @@
         return s;
     }
 
-    @Override
-    public EnumActionResult onItemUse(
-        EntityPlayer player,
-        World worldIn,
-        BlockPos pos,
-        EnumHand hand,
-        EnumFacing facing,
-        float hitX,
-        float hitY,
-        float hitZ
-    )
+    public EnumActionResult onItemUse(EntityPlayer player, World worldIn, BlockPos pos, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ)
     {
         ItemStack itemstack = player.getHeldItem(hand);
 
@@ -102,9 +92,7 @@
 
             BlockPos blockpos = pos.offset(facing);
             double d0 = this.getYOffset(worldIn, blockpos);
-            Entity entity = spawnCreature(
-                                worldIn, getNamedIdFrom(itemstack), blockpos.getX() + 0.5, blockpos.getY() + d0, blockpos.getZ() + 0.5
-                            );
+            Entity entity = spawnCreature(worldIn, getNamedIdFrom(itemstack), (double)blockpos.getX() + 0.5D, (double)blockpos.getY() + d0, (double)blockpos.getZ() + 0.5D);
 
             if (entity != null)
             {
@@ -127,12 +115,12 @@
 
     protected double getYOffset(World p_190909_1_, BlockPos p_190909_2_)
     {
-        AxisAlignedBB axisalignedbb = new AxisAlignedBB(p_190909_2_).expand(0.0, -1.0, 0.0);
-        List<AxisAlignedBB> list = p_190909_1_.getCollisionBoxes(null, axisalignedbb);
+        AxisAlignedBB axisalignedbb = (new AxisAlignedBB(p_190909_2_)).expand(0.0D, -1.0D, 0.0D);
+        List<AxisAlignedBB> list = p_190909_1_.getCollisionBoxes((Entity)null, axisalignedbb);
 
         if (list.isEmpty())
         {
-            return 0.0;
+            return 0.0D;
         }
         else
         {
@@ -143,7 +131,7 @@
                 d0 = Math.max(axisalignedbb1.maxY, d0);
             }
 
-            return d0 - p_190909_2_.getY();
+            return d0 - (double)p_190909_2_.getY();
         }
     }
 
@@ -157,9 +145,7 @@
 
             if (nbttagcompound != null && nbttagcompound.hasKey("EntityTag", 10))
             {
-                if (!entityWorld.isRemote
-                        && targetEntity.ignoreItemEntityData()
-                        && (player == null || !minecraftserver.getPlayerList().canSendCommands(player.getGameProfile())))
+                if (!entityWorld.isRemote && targetEntity.ignoreItemEntityData() && (player == null || !minecraftserver.getPlayerList().canSendCommands(player.getGameProfile())))
                 {
                     return;
                 }
@@ -173,14 +159,13 @@
         }
     }
 
-    @Override
     public ActionResult<ItemStack> onItemRightClick(World worldIn, EntityPlayer playerIn, EnumHand handIn)
     {
         ItemStack itemstack = playerIn.getHeldItem(handIn);
 
         if (worldIn.isRemote)
         {
-            return new ActionResult<>(EnumActionResult.PASS, itemstack);
+            return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
         }
         else
         {
@@ -192,17 +177,15 @@
 
                 if (!(worldIn.getBlockState(blockpos).getBlock() instanceof BlockLiquid))
                 {
-                    return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
                 }
                 else if (worldIn.isBlockModifiable(playerIn, blockpos) && playerIn.canPlayerEdit(blockpos, raytraceresult.sideHit, itemstack))
                 {
-                    Entity entity = spawnCreature(
-                                        worldIn, getNamedIdFrom(itemstack), blockpos.getX() + 0.5, blockpos.getY() + 0.5, blockpos.getZ() + 0.5
-                                    );
+                    Entity entity = spawnCreature(worldIn, getNamedIdFrom(itemstack), (double)blockpos.getX() + 0.5D, (double)blockpos.getY() + 0.5D, (double)blockpos.getZ() + 0.5D);
 
                     if (entity == null)
                     {
-                        return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                        return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
                     }
                     else
                     {
@@ -219,17 +202,17 @@
                         }
 
                         playerIn.addStat(StatList.getObjectUseStats(this));
-                        return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
+                        return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);
                     }
                 }
                 else
                 {
-                    return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                    return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
                 }
             }
             else
             {
-                return new ActionResult<>(EnumActionResult.PASS, itemstack);
+                return new ActionResult<ItemStack>(EnumActionResult.PASS, itemstack);
             }
         }
     }
@@ -241,7 +224,7 @@
         {
             Entity entity = null;
 
-            for (int i = 0; i < 1; i++)
+            for (int i = 0; i < 1; ++i)
             {
                 entity = EntityList.createEntityByIDFromName(entityID, worldIn);
 
@@ -251,7 +234,8 @@
                     entity.setLocationAndAngles(x, y, z, MathHelper.wrapDegrees(worldIn.rand.nextFloat() * 360.0F), 0.0F);
                     entityliving.rotationYawHead = entityliving.rotationYaw;
                     entityliving.renderYawOffset = entityliving.rotationYaw;
-                    entityliving.onInitialSpawn(worldIn.getDifficultyForLocation(new BlockPos(entityliving)), null);
+                    if (net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(entityliving, worldIn, (float) x, (float) y, (float) z, null)) return null;
+                    entityliving.onInitialSpawn(worldIn.getDifficultyForLocation(new BlockPos(entityliving)), (IEntityLivingData)null);
                     worldIn.spawnEntity(entity);
                     entityliving.playLivingSound();
                 }
@@ -265,7 +249,6 @@
         }
     }
 
-    @Override
     public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items)
     {
         if (this.isInCreativeTab(tab))
