--- before/net/minecraft/item/ItemFlintAndSteel.java
+++ after/net/minecraft/item/ItemFlintAndSteel.java
@@ -44,7 +44,7 @@
         }
         else
         {
-            if (worldIn.getBlockState(pos).getMaterial() == Material.AIR)
+            if (worldIn.isAirBlock(pos))
             {
                 worldIn.playSound(
                     player, pos, SoundEvents.ITEM_FLINTANDSTEEL_USE, SoundCategory.BLOCKS, 1.0F, itemRand.nextFloat() * 0.4F + 0.8F
