--- before/net/minecraft/item/ItemMap.java
+++ after/net/minecraft/item/ItemMap.java
@@ -42,7 +42,7 @@
         worldIn.setData(s, mapdata);
         mapdata.scale = scale;
         mapdata.calculateMapCenter(worldX, worldZ, mapdata.scale);
-        mapdata.dimension = (byte)worldIn.provider.getDimensionType().getId();
+        mapdata.dimension = worldIn.provider.getDimension();
         mapdata.trackingPosition = trackingPosition;
         mapdata.unlimitedTracking = unlimitedTracking;
         mapdata.markDirty();
@@ -70,7 +70,7 @@
             mapdata = new MapData(s);
             mapdata.scale = 3;
             mapdata.calculateMapCenter(worldIn.getWorldInfo().getSpawnX(), worldIn.getWorldInfo().getSpawnZ(), mapdata.scale);
-            mapdata.dimension = (byte)worldIn.provider.getDimensionType().getId();
+            mapdata.dimension = worldIn.provider.getDimension();
             mapdata.markDirty();
             worldIn.setData(s, mapdata);
         }
@@ -80,7 +80,7 @@
 
     public void updateMapData(World worldIn, Entity viewer, MapData data)
     {
-        if (worldIn.provider.getDimensionType().getId() == data.dimension && viewer instanceof EntityPlayer)
+        if (worldIn.provider.getDimension() == data.dimension && viewer instanceof EntityPlayer)
         {
             int i = 1 << data.scale;
             int j = data.xCenter;
@@ -252,13 +252,13 @@
 
     public static void renderBiomePreviewMap(World worldIn, ItemStack map)
     {
-        if (map.getItem() == Items.FILLED_MAP)
+        if (map.getItem() instanceof ItemMap)
         {
-            MapData mapdata = Items.FILLED_MAP.getMapData(map, worldIn);
+            MapData mapdata = ((ItemMap) map.getItem()).getMapData(map, worldIn);
 
             if (mapdata != null)
             {
-                if (worldIn.provider.getDimensionType().getId() == mapdata.dimension)
+                if (worldIn.provider.getDimension() == mapdata.dimension)
                 {
                     int i = 1 << mapdata.scale;
                     int j = mapdata.xCenter;
