--- before/net/minecraft/world/gen/layer/GenLayerBiomeEdge.java
+++ after/net/minecraft/world/gen/layer/GenLayerBiomeEdge.java
@@ -27,12 +27,12 @@
                 {
                     if (k == Biome.func_185362_a(Biomes.field_76769_d))
                     {
-                        int l = aint[j + 1 + (i + 1 - 1) * (p_75904_3_ + 2)];
-                        int i1 = aint[j + 1 + 1 + (i + 1) * (p_75904_3_ + 2)];
-                        int j1 = aint[j + 1 - 1 + (i + 1) * (p_75904_3_ + 2)];
-                        int k1 = aint[j + 1 + (i + 1 + 1) * (p_75904_3_ + 2)];
+                        int l1 = aint[j + 1 + (i + 1 - 1) * (p_75904_3_ + 2)];
+                        int i2 = aint[j + 1 + 1 + (i + 1) * (p_75904_3_ + 2)];
+                        int j2 = aint[j + 1 - 1 + (i + 1) * (p_75904_3_ + 2)];
+                        int k2 = aint[j + 1 + (i + 1 + 1) * (p_75904_3_ + 2)];
 
-                        if (l != Biome.func_185362_a(Biomes.field_76774_n) && i1 != Biome.func_185362_a(Biomes.field_76774_n) && j1 != Biome.func_185362_a(Biomes.field_76774_n) && k1 != Biome.func_185362_a(Biomes.field_76774_n))
+                        if (l1 != Biome.func_185362_a(Biomes.field_76774_n) && i2 != Biome.func_185362_a(Biomes.field_76774_n) && j2 != Biome.func_185362_a(Biomes.field_76774_n) && k2 != Biome.func_185362_a(Biomes.field_76774_n))
                         {
                             aint1[j + i * p_75904_3_] = k;
                         }
@@ -43,14 +43,14 @@
                     }
                     else if (k == Biome.func_185362_a(Biomes.field_76780_h))
                     {
-                        int l1 = aint[j + 1 + (i + 1 - 1) * (p_75904_3_ + 2)];
-                        int i2 = aint[j + 1 + 1 + (i + 1) * (p_75904_3_ + 2)];
-                        int j2 = aint[j + 1 - 1 + (i + 1) * (p_75904_3_ + 2)];
-                        int k2 = aint[j + 1 + (i + 1 + 1) * (p_75904_3_ + 2)];
+                        int l = aint[j + 1 + (i + 1 - 1) * (p_75904_3_ + 2)];
+                        int i1 = aint[j + 1 + 1 + (i + 1) * (p_75904_3_ + 2)];
+                        int j1 = aint[j + 1 - 1 + (i + 1) * (p_75904_3_ + 2)];
+                        int k1 = aint[j + 1 + (i + 1 + 1) * (p_75904_3_ + 2)];
 
-                        if (l1 != Biome.func_185362_a(Biomes.field_76769_d) && i2 != Biome.func_185362_a(Biomes.field_76769_d) && j2 != Biome.func_185362_a(Biomes.field_76769_d) && k2 != Biome.func_185362_a(Biomes.field_76769_d) && l1 != Biome.func_185362_a(Biomes.field_150584_S) && i2 != Biome.func_185362_a(Biomes.field_150584_S) && j2 != Biome.func_185362_a(Biomes.field_150584_S) && k2 != Biome.func_185362_a(Biomes.field_150584_S) && l1 != Biome.func_185362_a(Biomes.field_76774_n) && i2 != Biome.func_185362_a(Biomes.field_76774_n) && j2 != Biome.func_185362_a(Biomes.field_76774_n) && k2 != Biome.func_185362_a(Biomes.field_76774_n))
+                        if (l != Biome.func_185362_a(Biomes.field_76769_d) && i1 != Biome.func_185362_a(Biomes.field_76769_d) && j1 != Biome.func_185362_a(Biomes.field_76769_d) && k1 != Biome.func_185362_a(Biomes.field_76769_d) && l != Biome.func_185362_a(Biomes.field_150584_S) && i1 != Biome.func_185362_a(Biomes.field_150584_S) && j1 != Biome.func_185362_a(Biomes.field_150584_S) && k1 != Biome.func_185362_a(Biomes.field_150584_S) && l != Biome.func_185362_a(Biomes.field_76774_n) && i1 != Biome.func_185362_a(Biomes.field_76774_n) && j1 != Biome.func_185362_a(Biomes.field_76774_n) && k1 != Biome.func_185362_a(Biomes.field_76774_n))
                         {
-                            if (l1 != Biome.func_185362_a(Biomes.field_76782_w) && k2 != Biome.func_185362_a(Biomes.field_76782_w) && i2 != Biome.func_185362_a(Biomes.field_76782_w) && j2 != Biome.func_185362_a(Biomes.field_76782_w))
+                            if (l != Biome.func_185362_a(Biomes.field_76782_w) && k1 != Biome.func_185362_a(Biomes.field_76782_w) && i1 != Biome.func_185362_a(Biomes.field_76782_w) && j1 != Biome.func_185362_a(Biomes.field_76782_w))
                             {
                                 aint1[j + i * p_75904_3_] = k;
                             }
