--- before/net/minecraft/world/gen/layer/IntCache.java
+++ after/net/minecraft/world/gen/layer/IntCache.java
@@ -6,10 +6,10 @@
 public class IntCache
 {
     private static int field_76451_a = 256;
-    private static final List<int[]> field_76449_b = Lists.newArrayList();
-    private static final List<int[]> field_76450_c = Lists.newArrayList();
-    private static final List<int[]> field_76447_d = Lists.newArrayList();
-    private static final List<int[]> field_76448_e = Lists.newArrayList();
+    private static final List<int[]> field_76449_b = Lists.<int[]>newArrayList();
+    private static final List<int[]> field_76450_c = Lists.<int[]>newArrayList();
+    private static final List<int[]> field_76447_d = Lists.<int[]>newArrayList();
+    private static final List<int[]> field_76448_e = Lists.<int[]>newArrayList();
 
     public static synchronized int[] func_76445_a(int p_76445_0_)
     {
