--- original/net/minecraft/item/crafting/RecipesBanners.java
+++ modified/net/minecraft/item/crafting/RecipesBanners.java
@@ -12,11 +12,9 @@
 import net.minecraft.tileentity.TileEntityBanner;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class RecipesBanners {
-    public static class RecipeAddPattern implements IRecipe {
+    public static class RecipeAddPattern extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe {
         /** Used to check if a recipe matches current crafting inventory */
         @Override
         public boolean matches(InventoryCrafting inv, World worldIn) {
@@ -60,8 +58,9 @@
 
                 for (int j = 0; j < inv.getSizeInventory(); j++) {
                     ItemStack itemstack2 = inv.getStackInSlot(j);
-                    if (itemstack2.getItem() == Items.DYE) {
-                        k = itemstack2.getMetadata();
+                    int color = net.minecraftforge.oredict.DyeUtils.rawDyeDamageFromStack(itemstack2);
+                    if (color != -1) {
+                        k = color;
                         break;
                     }
                 }
@@ -96,9 +95,7 @@
 
             for (int i = 0; i < nonnulllist.size(); i++) {
                 ItemStack itemstack = inv.getStackInSlot(i);
-                if (itemstack.getItem().hasContainerItem()) {
-                    nonnulllist.set(i, new ItemStack(itemstack.getItem().getContainerItem()));
-                }
+                nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
             }
 
             return nonnulllist;
@@ -116,7 +113,7 @@
                         for (int i = 0; i < p_190933_1_.getSizeInventory() && flag; i++) {
                             ItemStack itemstack = p_190933_1_.getStackInSlot(i);
                             if (!itemstack.isEmpty() && itemstack.getItem() != Items.BANNER) {
-                                if (itemstack.getItem() == Items.DYE) {
+                                if (net.minecraftforge.oredict.DyeUtils.isDye(itemstack)) {
                                     if (flag2) {
                                         flag = false;
                                         break;
@@ -145,7 +142,7 @@
                             int i1 = k % 3;
                             ItemStack itemstack1 = p_190933_1_.getStackInSlot(k);
                             if (!itemstack1.isEmpty() && itemstack1.getItem() != Items.BANNER) {
-                                if (itemstack1.getItem() != Items.DYE) {
+                                if (!net.minecraftforge.oredict.DyeUtils.isDye(itemstack1)) {
                                     flag = false;
                                     break;
                                 }
@@ -186,14 +183,13 @@
         }
 
         /** Used to determine if this recipe can fit in a grid of the given width/height */
-        @SideOnly(Side.CLIENT)
         @Override
         public boolean canFit(int width, int height) {
             return width >= 3 && height >= 3;
         }
     }
 
-    public static class RecipeDuplicatePattern implements IRecipe {
+    public static class RecipeDuplicatePattern extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe {
         /** Used to check if a recipe matches current crafting inventory */
         @Override
         public boolean matches(InventoryCrafting inv, World worldIn) {
@@ -272,8 +268,8 @@
             for (int i = 0; i < nonnulllist.size(); i++) {
                 ItemStack itemstack = inv.getStackInSlot(i);
                 if (!itemstack.isEmpty()) {
-                    if (itemstack.getItem().hasContainerItem()) {
-                        nonnulllist.set(i, new ItemStack(itemstack.getItem().getContainerItem()));
+                    if (itemstack.getItem().hasContainerItem(itemstack)) {
+                        nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
                     } else if (itemstack.hasTagCompound() && TileEntityBanner.getPatterns(itemstack) > 0) {
                         ItemStack itemstack1 = itemstack.copy();
                         itemstack1.setCount(1);
@@ -292,7 +288,6 @@
         }
 
         /** Used to determine if this recipe can fit in a grid of the given width/height */
-        @SideOnly(Side.CLIENT)
         @Override
         public boolean canFit(int width, int height) {
             return width * height >= 2;
