--- original/net/minecraft/item/crafting/RecipeRepairItem.java
+++ modified/net/minecraft/item/crafting/RecipeRepairItem.java
@@ -8,8 +8,7 @@
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
 
-public class RecipeRepairItem implements IRecipe {
-    /** Used to check if a recipe matches current crafting inventory */
+public class RecipeRepairItem extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<IRecipe> implements IRecipe {
     @Override
     public boolean matches(InventoryCrafting inv, World worldIn) {
         List<ItemStack> list = Lists.newArrayList();
@@ -21,9 +20,9 @@
                 if (list.size() > 1) {
                     ItemStack itemstack1 = list.get(0);
                     if (itemstack.getItem() != itemstack1.getItem()
-                        || itemstack1.getCount() != 1
-                        || itemstack.getCount() != 1
-                        || !itemstack1.getItem().isDamageable()) {
+                            || itemstack1.getCount() != 1
+                            || itemstack.getCount() != 1
+                            || !itemstack1.getItem().isRepairable()) {
                         return false;
                     }
                 }
@@ -45,9 +44,9 @@
                 if (list.size() > 1) {
                     ItemStack itemstack1 = list.get(0);
                     if (itemstack.getItem() != itemstack1.getItem()
-                        || itemstack1.getCount() != 1
-                        || itemstack.getCount() != 1
-                        || !itemstack1.getItem().isDamageable()) {
+                            || itemstack1.getCount() != 1
+                            || itemstack.getCount() != 1
+                            || !itemstack1.getItem().isRepairable()) {
                         return ItemStack.EMPTY;
                     }
                 }
@@ -58,14 +57,15 @@
             ItemStack itemstack2 = list.get(0);
             ItemStack itemstack3 = list.get(1);
             if (itemstack2.getItem() == itemstack3.getItem()
-                && itemstack2.getCount() == 1
-                && itemstack3.getCount() == 1
-                && itemstack2.getItem().isDamageable()) {
-                Item item = itemstack2.getItem();
-                int j = item.getMaxDamage() - itemstack2.getItemDamage();
-                int k = item.getMaxDamage() - itemstack3.getItemDamage();
-                int l = j + k + item.getMaxDamage() * 5 / 100;
-                int i1 = item.getMaxDamage() - l;
+                    && itemstack2.getCount() == 1
+                    && itemstack3.getCount() == 1
+                    && itemstack2.getItem().isRepairable()) {
+                // FORGE: Make itemstack sensitive // Item item = itemstack2.getItem();
+                int j = itemstack2.getMaxDamage() - itemstack2.getItemDamage();
+                int k = itemstack2.getMaxDamage() - itemstack3.getItemDamage();
+                int l = j + k + itemstack2.getMaxDamage() * 5 / 100;
+                int i1 = itemstack2.getMaxDamage() - l;
+
                 if (i1 < 0) {
                     i1 = 0;
                 }
@@ -89,9 +89,7 @@
 
         for (int i = 0; i < nonnulllist.size(); i++) {
             ItemStack itemstack = inv.getStackInSlot(i);
-            if (itemstack.getItem().hasContainerItem()) {
-                nonnulllist.set(i, new ItemStack(itemstack.getItem().getContainerItem()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
@@ -103,7 +101,6 @@
         return true;
     }
 
-    /** Used to determine if this recipe can fit in a grid of the given width/height */
     @Override
     public boolean canFit(int width, int height) {
         return width * height >= 2;
