--- before/net/minecraft/world/end/DragonFightManager.java
+++ after/net/minecraft/world/end/DragonFightManager.java
@@ -87,6 +87,7 @@
 
             this.dragonKilled = compound.getBoolean("DragonKilled");
             this.previouslyKilled = compound.getBoolean("PreviouslyKilled");
+            this.scanForLegacyFight = !compound.getBoolean("LegacyScanPerformed"); // Forge: fix MC-105080
 
             if (compound.getBoolean("IsRespawning"))
             {
@@ -140,6 +141,7 @@
 
         nbttagcompound.setBoolean("DragonKilled", this.dragonKilled);
         nbttagcompound.setBoolean("PreviouslyKilled", this.previouslyKilled);
+        nbttagcompound.setBoolean("LegacyScanPerformed", !this.scanForLegacyFight); // Forge: fix MC-105080
 
         if (this.exitPortalLocation != null)
         {
@@ -604,5 +606,15 @@
                 entityendercrystal.setBeamTarget(null);
             }
         }
+    }
+
+    public void addPlayer(EntityPlayerMP player)
+    {
+        this.bossInfo.addPlayer(player);
+    }
+
+    public void removePlayer(EntityPlayerMP player)
+    {
+        this.bossInfo.removePlayer(player);
     }
 }
