--- before/net/minecraft/world/end/DragonFightManager.java
+++ after/net/minecraft/world/end/DragonFightManager.java
@@ -8,6 +8,7 @@
 import com.google.common.collect.Range;
 import com.google.common.collect.Sets;
 import java.util.Collections;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Random;
 import java.util.Set;
@@ -50,16 +51,10 @@
 public class DragonFightManager
 {
     private static final Logger field_186107_a = LogManager.getLogger();
-    private static final Predicate<EntityPlayerMP> field_186108_b = Predicates.and(
-                    EntitySelectors.field_94557_a, EntitySelectors.func_188443_a(0.0, 128.0, 0.0, 192.0)
-                );
-    private final BossInfoServer field_186109_c = (BossInfoServer)new BossInfoServer(
-                new TextComponentTranslation("entity.EnderDragon.name"), BossInfo.Color.PINK, BossInfo.Overlay.PROGRESS
-            )
-            .func_186742_b(true)
-            .func_186743_c(true);
+    private static final Predicate<EntityPlayerMP> field_186108_b = Predicates.<EntityPlayerMP>and(EntitySelectors.field_94557_a, EntitySelectors.func_188443_a(0.0D, 128.0D, 0.0D, 192.0D));
+    private final BossInfoServer field_186109_c = (BossInfoServer)(new BossInfoServer(new TextComponentTranslation("entity.EnderDragon.name", new Object[0]), BossInfo.Color.PINK, BossInfo.Overlay.PROGRESS)).func_186742_b(true).func_186743_c(true);
     private final WorldServer field_186110_d;
-    private final List<Integer> field_186111_e = Lists.newArrayList();
+    private final List<Integer> field_186111_e = Lists.<Integer>newArrayList();
     private final BlockPattern field_186112_f;
     private int field_186113_g;
     private int field_186114_h;
@@ -87,6 +82,7 @@
 
             this.field_186117_k = p_i46669_2_.func_74767_n("DragonKilled");
             this.field_186118_l = p_i46669_2_.func_74767_n("PreviouslyKilled");
+            this.field_186120_n = !p_i46669_2_.func_74767_n("LegacyScanPerformed"); // Forge: fix MC-105080
 
             if (p_i46669_2_.func_74767_n("IsRespawning"))
             {
@@ -108,25 +104,18 @@
         {
             NBTTagList nbttaglist = p_i46669_2_.func_150295_c("Gateways", 3);
 
-            for (int i = 0; i < nbttaglist.func_74745_c(); i++)
+            for (int i = 0; i < nbttaglist.func_74745_c(); ++i)
             {
-                this.field_186111_e.add(nbttaglist.func_186858_c(i));
+                this.field_186111_e.add(Integer.valueOf(nbttaglist.func_186858_c(i)));
             }
         }
         else
         {
-            this.field_186111_e.addAll(ContiguousSet.create(Range.closedOpen(0, 20), DiscreteDomain.integers()));
+            this.field_186111_e.addAll(ContiguousSet.create(Range.closedOpen(Integer.valueOf(0), Integer.valueOf(20)), DiscreteDomain.integers()));
             Collections.shuffle(this.field_186111_e, new Random(p_i46669_1_.func_72905_C()));
         }
 
-        this.field_186112_f = FactoryBlockPattern.func_177660_a()
-                              .func_177659_a("       ", "       ", "       ", "   #   ", "       ", "       ", "       ")
-                              .func_177659_a("       ", "       ", "       ", "   #   ", "       ", "       ", "       ")
-                              .func_177659_a("       ", "       ", "       ", "   #   ", "       ", "       ", "       ")
-                              .func_177659_a("  ###  ", " #   # ", "#     #", "#  #  #", "#     #", " #   # ", "  ###  ")
-                              .func_177659_a("       ", "  ###  ", " ##### ", " ##### ", " ##### ", "  ###  ", "       ")
-                              .func_177662_a('#', BlockWorldState.func_177510_a(BlockMatcher.func_177642_a(Blocks.field_150357_h)))
-                              .func_177661_b();
+        this.field_186112_f = FactoryBlockPattern.func_177660_a().func_177659_a("       ", "       ", "       ", "   #   ", "       ", "       ", "       ").func_177659_a("       ", "       ", "       ", "   #   ", "       ", "       ", "       ").func_177659_a("       ", "       ", "       ", "   #   ", "       ", "       ", "       ").func_177659_a("  ###  ", " #   # ", "#     #", "#  #  #", "#     #", " #   # ", "  ###  ").func_177659_a("       ", "  ###  ", " ##### ", " ##### ", " ##### ", "  ###  ", "       ").func_177662_a('#', BlockWorldState.func_177510_a(BlockMatcher.func_177642_a(Blocks.field_150357_h))).func_177661_b();
     }
 
     public NBTTagCompound func_186088_a()
@@ -140,6 +129,7 @@
 
         nbttagcompound.func_74757_a("DragonKilled", this.field_186117_k);
         nbttagcompound.func_74757_a("PreviouslyKilled", this.field_186118_l);
+        nbttagcompound.func_74757_a("LegacyScanPerformed", !this.field_186120_n); // Forge: fix MC-105080
 
         if (this.field_186121_o != null)
         {
@@ -147,9 +137,11 @@
         }
 
         NBTTagList nbttaglist = new NBTTagList();
+        Iterator iterator = this.field_186111_e.iterator();
 
-        for (int i : this.field_186111_e)
+        while (iterator.hasNext())
         {
+            int i = ((Integer)iterator.next()).intValue();
             nbttaglist.func_74742_a(new NBTTagInt(i));
         }
 
@@ -198,7 +190,7 @@
                 {
                     EntityDragon entitydragon = list.get(0);
                     this.field_186119_m = entitydragon.func_110124_au();
-                    field_186107_a.info("Found that there's a dragon still alive ({})", entitydragon);
+                    field_186107_a.info("Found that there's a dragon still alive ({})", (Object)entitydragon);
                     this.field_186117_k = false;
 
                     if (!flag)
@@ -241,7 +233,7 @@
                     else
                     {
                         field_186107_a.debug("Haven't seen our dragon, but found another one to use.");
-                        this.field_186119_m = list1.get(0).func_110124_au();
+                        this.field_186119_m = ((EntityDragon)list1.get(0)).func_110124_au();
                     }
 
                     this.field_186113_g = 0;
@@ -286,9 +278,9 @@
 
     private boolean func_186104_g()
     {
-        for (int i = -8; i <= 8; i++)
+        for (int i = -8; i <= 8; ++i)
         {
-            for (int j = -8; j <= 8; j++)
+            for (int j = -8; j <= 8; ++j)
             {
                 Chunk chunk = this.field_186110_d.func_72964_e(i, j);
 
@@ -308,9 +300,9 @@
     @Nullable
     private BlockPattern.PatternHelper func_186091_h()
     {
-        for (int i = -8; i <= 8; i++)
+        for (int i = -8; i <= 8; ++i)
         {
-            for (int j = -8; j <= 8; j++)
+            for (int j = -8; j <= 8; ++j)
             {
                 Chunk chunk = this.field_186110_d.func_72964_e(i, j);
 
@@ -338,12 +330,9 @@
 
         int k = this.field_186110_d.func_175645_m(WorldGenEndPodium.field_186139_a).func_177956_o();
 
-        for (int l = k; l >= 0; l--)
+        for (int l = k; l >= 0; --l)
         {
-            BlockPattern.PatternHelper blockpattern$patternhelper1 = this.field_186112_f
-                    .func_177681_a(
-                        this.field_186110_d, new BlockPos(WorldGenEndPodium.field_186139_a.func_177958_n(), l, WorldGenEndPodium.field_186139_a.func_177952_p())
-                    );
+            BlockPattern.PatternHelper blockpattern$patternhelper1 = this.field_186112_f.func_177681_a(this.field_186110_d, new BlockPos(WorldGenEndPodium.field_186139_a.func_177958_n(), l, WorldGenEndPodium.field_186139_a.func_177952_p()));
 
             if (blockpattern$patternhelper1 != null)
             {
@@ -361,9 +350,9 @@
 
     private void func_186103_i()
     {
-        for (int i = -8; i <= 8; i++)
+        for (int i = -8; i <= 8; ++i)
         {
-            for (int j = -8; j <= 8; j++)
+            for (int j = -8; j <= 8; ++j)
             {
                 this.field_186110_d.func_72964_e(i, j);
             }
@@ -372,7 +361,7 @@
 
     private void func_186100_j()
     {
-        Set<EntityPlayerMP> set = Sets.newHashSet();
+        Set<EntityPlayerMP> set = Sets.<EntityPlayerMP>newHashSet();
 
         for (EntityPlayerMP entityplayermp : this.field_186110_d.func_175661_b(EntityPlayerMP.class, field_186108_b))
         {
@@ -396,10 +385,10 @@
 
         for (WorldGenSpikes.EndSpike worldgenspikes$endspike : BiomeEndDecorator.func_185426_a(this.field_186110_d))
         {
-            this.field_186114_h = this.field_186114_h + this.field_186110_d.func_72872_a(EntityEnderCrystal.class, worldgenspikes$endspike.func_186153_f()).size();
+            this.field_186114_h += this.field_186110_d.func_72872_a(EntityEnderCrystal.class, worldgenspikes$endspike.func_186153_f()).size();
         }
 
-        field_186107_a.debug("Found {} end crystals still alive", this.field_186114_h);
+        field_186107_a.debug("Found {} end crystals still alive", (int)this.field_186114_h);
     }
 
     public void func_186096_a(EntityDragon p_186096_1_)
@@ -425,9 +414,9 @@
     {
         if (!this.field_186111_e.isEmpty())
         {
-            int i = this.field_186111_e.remove(this.field_186111_e.size() - 1);
-            int j = (int)(96.0 * Math.cos(2.0 * (-Math.PI + (Math.PI / 20) * (double)i)));
-            int k = (int)(96.0 * Math.sin(2.0 * (-Math.PI + (Math.PI / 20) * (double)i)));
+            int i = ((Integer)this.field_186111_e.remove(this.field_186111_e.size() - 1)).intValue();
+            int j = (int)(96.0D * Math.cos(2.0D * (-Math.PI + 0.15707963267948966D * (double)i)));
+            int k = (int)(96.0D * Math.sin(2.0D * (-Math.PI + 0.15707963267948966D * (double)i)));
             this.func_186089_a(new BlockPos(j, 75, k));
         }
     }
@@ -435,7 +424,7 @@
     private void func_186089_a(BlockPos p_186089_1_)
     {
         this.field_186110_d.func_175718_b(3000, p_186089_1_, 0);
-        new WorldGenEndGateway().func_180709_b(this.field_186110_d, new Random(), p_186089_1_);
+        (new WorldGenEndGateway()).func_180709_b(this.field_186110_d, new Random(), p_186089_1_);
     }
 
     private void func_186094_a(boolean p_186094_1_)
@@ -444,14 +433,9 @@
 
         if (this.field_186121_o == null)
         {
-            this.field_186121_o = this.field_186110_d.func_175672_r(WorldGenEndPodium.field_186139_a).func_177977_b();
-
-            while (
-                this.field_186110_d.func_180495_p(this.field_186121_o).func_177230_c() == Blocks.field_150357_h
-                && this.field_186121_o.func_177956_o() > this.field_186110_d.func_181545_F()
-            )
+            for (this.field_186121_o = this.field_186110_d.func_175672_r(WorldGenEndPodium.field_186139_a).func_177977_b(); this.field_186110_d.func_180495_p(this.field_186121_o).func_177230_c() == Blocks.field_150357_h && this.field_186121_o.func_177956_o() > this.field_186110_d.func_181545_F(); this.field_186121_o = this.field_186121_o.func_177977_b())
             {
-                this.field_186121_o = this.field_186121_o.func_177977_b();
+                ;
             }
         }
 
@@ -463,7 +447,7 @@
         this.field_186110_d.func_175726_f(new BlockPos(0, 128, 0));
         EntityDragon entitydragon = new EntityDragon(this.field_186110_d);
         entitydragon.func_184670_cT().func_188758_a(PhaseList.field_188741_a);
-        entitydragon.func_70012_b(0.0, 128.0, 0.0, this.field_186110_d.field_73012_v.nextFloat() * 360.0F, 0.0F);
+        entitydragon.func_70012_b(0.0D, 128.0D, 0.0D, this.field_186110_d.field_73012_v.nextFloat() * 360.0F, 0.0F);
         this.field_186110_d.func_72838_d(entitydragon);
         this.field_186119_m = entitydragon.func_110124_au();
         return entitydragon;
@@ -539,13 +523,12 @@
                 blockpos = this.field_186121_o;
             }
 
-            List<EntityEnderCrystal> list1 = Lists.newArrayList();
+            List<EntityEnderCrystal> list1 = Lists.<EntityEnderCrystal>newArrayList();
             BlockPos blockpos1 = blockpos.func_177981_b(1);
 
             for (EnumFacing enumfacing : EnumFacing.Plane.HORIZONTAL)
             {
-                List<EntityEnderCrystal> list = this.field_186110_d
-                                                .func_72872_a(EntityEnderCrystal.class, new AxisAlignedBB(blockpos1.func_177967_a(enumfacing, 2)));
+                List<EntityEnderCrystal> list = this.field_186110_d.func_72872_a(EntityEnderCrystal.class, new AxisAlignedBB(blockpos1.func_177967_a(enumfacing, 2)));
 
                 if (list.isEmpty())
                 {
@@ -564,21 +547,17 @@
     {
         if (this.field_186117_k && this.field_186122_p == null)
         {
-            for (BlockPattern.PatternHelper blockpattern$patternhelper = this.func_186091_h();
-                    blockpattern$patternhelper != null;
-                    blockpattern$patternhelper = this.func_186091_h()
-                )
+            for (BlockPattern.PatternHelper blockpattern$patternhelper = this.func_186091_h(); blockpattern$patternhelper != null; blockpattern$patternhelper = this.func_186091_h())
             {
-                for (int i = 0; i < this.field_186112_f.func_177684_c(); i++)
+                for (int i = 0; i < this.field_186112_f.func_177684_c(); ++i)
                 {
-                    for (int j = 0; j < this.field_186112_f.func_177685_b(); j++)
+                    for (int j = 0; j < this.field_186112_f.func_177685_b(); ++j)
                     {
-                        for (int k = 0; k < this.field_186112_f.func_185922_a(); k++)
+                        for (int k = 0; k < this.field_186112_f.func_185922_a(); ++k)
                         {
                             BlockWorldState blockworldstate = blockpattern$patternhelper.func_177670_a(i, j, k);
 
-                            if (blockworldstate.func_177509_a().func_177230_c() == Blocks.field_150357_h
-                                    || blockworldstate.func_177509_a().func_177230_c() == Blocks.field_150384_bq)
+                            if (blockworldstate.func_177509_a().func_177230_c() == Blocks.field_150357_h || blockworldstate.func_177509_a().func_177230_c() == Blocks.field_150384_bq)
                             {
                                 this.field_186110_d.func_175656_a(blockworldstate.func_177508_d(), Blocks.field_150377_bs.func_176223_P());
                             }
@@ -601,8 +580,18 @@
             for (EntityEnderCrystal entityendercrystal : this.field_186110_d.func_72872_a(EntityEnderCrystal.class, worldgenspikes$endspike.func_186153_f()))
             {
                 entityendercrystal.func_184224_h(false);
-                entityendercrystal.func_184516_a(null);
+                entityendercrystal.func_184516_a((BlockPos)null);
             }
         }
+    }
+
+    public void addPlayer(EntityPlayerMP player)
+    {
+        this.field_186109_c.func_186760_a(player);
+    }
+
+    public void removePlayer(EntityPlayerMP player)
+    {
+        this.field_186109_c.func_186761_b(player);
     }
 }
