--- before/net/minecraft/client/gui/spectator/categories/SpectatorDetails.java
+++ after/net/minecraft/client/gui/spectator/categories/SpectatorDetails.java
@@ -24,7 +24,7 @@
 
     public ISpectatorMenuObject func_178680_a(int p_178680_1_)
     {
-        return p_178680_1_ >= 0 && p_178680_1_ < this.field_178682_b.size() ? MoreObjects.firstNonNull(this.field_178682_b.get(p_178680_1_), SpectatorMenu.field_178657_a) : SpectatorMenu.field_178657_a;
+        return p_178680_1_ >= 0 && p_178680_1_ < this.field_178682_b.size() ? (ISpectatorMenuObject)MoreObjects.firstNonNull(this.field_178682_b.get(p_178680_1_), SpectatorMenu.field_178657_a) : SpectatorMenu.field_178657_a;
     }
 
     public int func_178681_b()
