--- before/net/minecraft/client/gui/spectator/categories/TeleportToTeam.java
+++ after/net/minecraft/client/gui/spectator/categories/TeleportToTeam.java
@@ -26,7 +26,7 @@
 @SideOnly(Side.CLIENT)
 public class TeleportToTeam implements ISpectatorMenuView, ISpectatorMenuObject
 {
-    private final List<ISpectatorMenuObject> field_178672_a = Lists.newArrayList();
+    private final List<ISpectatorMenuObject> field_178672_a = Lists.<ISpectatorMenuObject>newArrayList();
 
     public TeleportToTeam()
     {
@@ -45,7 +45,7 @@
 
     public ITextComponent func_178670_b()
     {
-        return new TextComponentTranslation("spectatorMenu.team_teleport.prompt");
+        return new TextComponentTranslation("spectatorMenu.team_teleport.prompt", new Object[0]);
     }
 
     public void func_178661_a(SpectatorMenu p_178661_1_)
@@ -55,7 +55,7 @@
 
     public ITextComponent func_178664_z_()
     {
-        return new TextComponentTranslation("spectatorMenu.team_teleport");
+        return new TextComponentTranslation("spectatorMenu.team_teleport", new Object[0]);
     }
 
     public void func_178663_a(float p_178663_1_, int p_178663_2_)
@@ -87,7 +87,7 @@
         public TeamSelectionObject(ScorePlayerTeam p_i45492_2_)
         {
             this.field_178676_b = p_i45492_2_;
-            this.field_178675_d = Lists.newArrayList();
+            this.field_178675_d = Lists.<NetworkPlayerInfo>newArrayList();
 
             for (String s : p_i45492_2_.func_96670_d())
             {
@@ -105,7 +105,7 @@
             }
             else
             {
-                String s1 = this.field_178675_d.get((new Random()).nextInt(this.field_178675_d.size())).func_178845_a().getName();
+                String s1 = ((NetworkPlayerInfo)this.field_178675_d.get((new Random()).nextInt(this.field_178675_d.size()))).func_178845_a().getName();
                 this.field_178677_c = AbstractClientPlayer.func_110311_f(s1);
                 AbstractClientPlayer.func_110304_a(this.field_178677_c, s1);
             }
