--- before/net/minecraft/world/gen/structure/MapGenStronghold.java
+++ after/net/minecraft/world/gen/structure/MapGenStronghold.java
@@ -29,7 +29,15 @@
 
         for (Biome biome : Biome.field_185377_q)
         {
-            if (biome != null && biome.func_185355_j() > 0.0F)
+            if (biome != null && biome.func_185355_j() > 0.0F && !net.minecraftforge.common.BiomeManager.strongHoldBiomesBlackList.contains(biome))
+            {
+                this.field_151546_e.add(biome);
+            }
+        }
+
+        for (Biome biome : net.minecraftforge.common.BiomeManager.strongHoldBiomes)
+        {
+            if (!this.field_151546_e.contains(biome))
             {
                 this.field_151546_e.add(biome);
             }
