--- before/net/minecraft/world/gen/structure/MapGenStronghold.java
+++ after/net/minecraft/world/gen/structure/MapGenStronghold.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.gen.structure;
 
 import com.google.common.collect.Lists;
+import it.unimi.dsi.fastutil.objects.ObjectIterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Random;
@@ -15,17 +16,28 @@
 {
     public final List<Biome> field_151546_e;
     private boolean field_75056_f;
-    private ChunkPos[] field_75057_g = new ChunkPos[128];
-    private double field_82671_h = 32.0;
-    private int field_82672_i = 3;
+    private ChunkPos[] field_75057_g;
+    private double field_82671_h;
+    private int field_82672_i;
 
     public MapGenStronghold()
     {
-        this.field_151546_e = Lists.newArrayList();
+        this.field_75057_g = new ChunkPos[128];
+        this.field_82671_h = 32.0D;
+        this.field_82672_i = 3;
+        this.field_151546_e = Lists.<Biome>newArrayList();
 
         for (Biome biome : Biome.field_185377_q)
         {
-            if (biome != null && biome.func_185355_j() > 0.0F)
+            if (biome != null && biome.func_185355_j() > 0.0F && !net.minecraftforge.common.BiomeManager.strongHoldBiomesBlackList.contains(biome))
+            {
+                this.field_151546_e.add(biome);
+            }
+        }
+
+        for (Biome biome : net.minecraftforge.common.BiomeManager.strongHoldBiomes)
+        {
+            if (!this.field_151546_e.contains(biome))
             {
                 this.field_151546_e.add(biome);
             }
@@ -38,28 +50,26 @@
 
         for (Entry<String, String> entry : p_i2068_1_.entrySet())
         {
-            if (entry.getKey().equals("distance"))
+            if (((String)entry.getKey()).equals("distance"))
             {
-                this.field_82671_h = MathHelper.func_82713_a(entry.getValue(), this.field_82671_h, 1.0);
+                this.field_82671_h = MathHelper.func_82713_a(entry.getValue(), this.field_82671_h, 1.0D);
             }
-            else if (entry.getKey().equals("count"))
+            else if (((String)entry.getKey()).equals("count"))
             {
                 this.field_75057_g = new ChunkPos[MathHelper.func_82714_a(entry.getValue(), this.field_75057_g.length, 1)];
             }
-            else if (entry.getKey().equals("spread"))
+            else if (((String)entry.getKey()).equals("spread"))
             {
                 this.field_82672_i = MathHelper.func_82714_a(entry.getValue(), this.field_82672_i, 1);
             }
         }
     }
 
-    @Override
     public String func_143025_a()
     {
         return "Stronghold";
     }
 
-    @Override
     public BlockPos func_180706_b(World p_180706_1_, BlockPos p_180706_2_, boolean p_180706_3_)
     {
         if (!this.field_75056_f)
@@ -92,7 +102,6 @@
         return blockpos;
     }
 
-    @Override
     protected boolean func_75047_a(int p_75047_1_, int p_75047_2_)
     {
         if (!this.field_75056_f)
@@ -116,9 +125,12 @@
     {
         this.func_143027_a(this.field_75039_c);
         int i = 0;
+        ObjectIterator lvt_2_1_ = this.field_75053_d.values().iterator();
 
-        for (StructureStart structurestart : this.field_75053_d.values())
+        while (lvt_2_1_.hasNext())
         {
+            StructureStart structurestart = (StructureStart)lvt_2_1_.next();
+
             if (i < this.field_75057_g.length)
             {
                 this.field_75057_g[i++] = new ChunkPos(structurestart.func_143019_e(), structurestart.func_143018_f());
@@ -127,7 +139,7 @@
 
         Random random = new Random();
         random.setSeed(this.field_75039_c.func_72905_C());
-        double d1 = random.nextDouble() * Math.PI * 2.0;
+        double d1 = random.nextDouble() * Math.PI * 2.0D;
         int j = 0;
         int k = 0;
         int l = this.field_75053_d.size();
@@ -136,7 +148,7 @@
         {
             for (int i1 = 0; i1 < this.field_75057_g.length; ++i1)
             {
-                double d0 = 4.0 * this.field_82671_h + this.field_82671_h * (double)j * 6.0 + (random.nextDouble() - 0.5) * this.field_82671_h * 2.5;
+                double d0 = 4.0D * this.field_82671_h + this.field_82671_h * (double)j * 6.0D + (random.nextDouble() - 0.5D) * this.field_82671_h * 2.5D;
                 int j1 = (int)Math.round(Math.cos(d1) * d0);
                 int k1 = (int)Math.round(Math.sin(d1) * d0);
                 BlockPos blockpos = this.field_75039_c.func_72959_q().func_180630_a((j1 << 4) + 8, (k1 << 4) + 8, 112, this.field_151546_e, random);
@@ -152,31 +164,28 @@
                     this.field_75057_g[i1] = new ChunkPos(j1, k1);
                 }
 
-                d1 += (Math.PI * 2) / (double)this.field_82672_i;
+                d1 += (Math.PI * 2D) / (double)this.field_82672_i;
+                ++k;
 
-                if (++k == this.field_82672_i)
+                if (k == this.field_82672_i)
                 {
                     ++j;
                     k = 0;
                     this.field_82672_i += 2 * this.field_82672_i / (j + 1);
                     this.field_82672_i = Math.min(this.field_82672_i, this.field_75057_g.length - i1);
-                    d1 += random.nextDouble() * Math.PI * 2.0;
+                    d1 += random.nextDouble() * Math.PI * 2.0D;
                 }
             }
         }
     }
 
-    @Override
     protected StructureStart func_75049_b(int p_75049_1_, int p_75049_2_)
     {
-        MapGenStronghold.Start mapgenstronghold$start = new MapGenStronghold.Start(this.field_75039_c, this.field_75038_b, p_75049_1_, p_75049_2_);
+        MapGenStronghold.Start mapgenstronghold$start;
 
-        while (
-            mapgenstronghold$start.func_186161_c().isEmpty()
-            || ((StructureStrongholdPieces.Stairs2)mapgenstronghold$start.func_186161_c().get(0)).field_75025_b == null
-        )
+        for (mapgenstronghold$start = new MapGenStronghold.Start(this.field_75039_c, this.field_75038_b, p_75049_1_, p_75049_2_); mapgenstronghold$start.func_186161_c().isEmpty() || ((StructureStrongholdPieces.Stairs2)mapgenstronghold$start.func_186161_c().get(0)).field_75025_b == null; mapgenstronghold$start = new MapGenStronghold.Start(this.field_75039_c, this.field_75038_b, p_75049_1_, p_75049_2_))
         {
-            mapgenstronghold$start = new MapGenStronghold.Start(this.field_75039_c, this.field_75038_b, p_75049_1_, p_75049_2_);
+            ;
         }
 
         return mapgenstronghold$start;
@@ -192,9 +201,7 @@
             {
                 super(p_i2067_3_, p_i2067_4_);
                 StructureStrongholdPieces.func_75198_a();
-                StructureStrongholdPieces.Stairs2 structurestrongholdpieces$stairs2 = new StructureStrongholdPieces.Stairs2(
-                    0, p_i2067_2_, (p_i2067_3_ << 4) + 2, (p_i2067_4_ << 4) + 2
-                );
+                StructureStrongholdPieces.Stairs2 structurestrongholdpieces$stairs2 = new StructureStrongholdPieces.Stairs2(0, p_i2067_2_, (p_i2067_3_ << 4) + 2, (p_i2067_4_ << 4) + 2);
                 this.field_75075_a.add(structurestrongholdpieces$stairs2);
                 structurestrongholdpieces$stairs2.func_74861_a(structurestrongholdpieces$stairs2, this.field_75075_a, p_i2067_2_);
                 List<StructureComponent> list = structurestrongholdpieces$stairs2.field_75026_c;
