--- before/net/minecraft/world/gen/structure/MapGenEndCity.java
+++ after/net/minecraft/world/gen/structure/MapGenEndCity.java
@@ -41,10 +41,10 @@
         int k = p_75047_1_ / 20;
         int l = p_75047_2_ / 20;
         Random random = this.field_75039_c.func_72843_D(k, l, 10387313);
-        k *= 20;
-        l *= 20;
-        k += (random.nextInt(9) + random.nextInt(9)) / 2;
-        l += (random.nextInt(9) + random.nextInt(9)) / 2;
+        k = k * 20;
+        l = l * 20;
+        k = k + (random.nextInt(9) + random.nextInt(9)) / 2;
+        l = l + (random.nextInt(9) + random.nextInt(9)) / 2;
 
         if (i == k && j == l && this.field_186133_d.func_185961_c(i, j))
         {
@@ -95,7 +95,8 @@
         int l = chunkprimer.func_186138_a(7, 7 + j);
         int i1 = chunkprimer.func_186138_a(7 + i, 7);
         int j1 = chunkprimer.func_186138_a(7 + i, 7 + j);
-        return Math.min(Math.min(k, l), Math.min(i1, j1));
+        int k1 = Math.min(Math.min(k, l), Math.min(i1, j1));
+        return k1;
     }
 
     public static class Start extends StructureStart
