--- before/net/minecraft/world/gen/structure/MapGenStructure.java
+++ after/net/minecraft/world/gen/structure/MapGenStructure.java
@@ -155,12 +155,12 @@
     {
         if (this.structureData == null && worldIn != null)
         {
-            this.structureData = (MapGenStructureData)worldIn.loadData(MapGenStructureData.class, this.getStructureName());
+            this.structureData = (MapGenStructureData)worldIn.getPerWorldStorage().getOrLoadData(MapGenStructureData.class, this.getStructureName());
 
             if (this.structureData == null)
             {
                 this.structureData = new MapGenStructureData(this.getStructureName());
-                worldIn.setData(this.getStructureName(), this.structureData);
+                worldIn.getPerWorldStorage().setData(this.getStructureName(), this.structureData);
             }
             else
             {
