--- before/net/minecraft/world/gen/structure/MapGenStronghold.java
+++ after/net/minecraft/world/gen/structure/MapGenStronghold.java
@@ -15,8 +15,8 @@
 {
     public final List<Biome> allowedBiomes;
     private boolean ranBiomeCheck;
-    private ChunkPos[] structureCoords = new ChunkPos[128];
-    private double distance = 32.0;
+    private ChunkPos[] structureCoords= new ChunkPos[128];
+    private double distance = 32.0D;
     private int spread = 3;
 
     public MapGenStronghold()
@@ -25,7 +25,15 @@
 
         for (Biome biome : Biome.REGISTRY)
         {
-            if (biome != null && biome.getBaseHeight() > 0.0F)
+            if (biome != null && biome.getBaseHeight() > 0.0F && !net.minecraftforge.common.BiomeManager.strongHoldBiomesBlackList.contains(biome))
+            {
+                this.allowedBiomes.add(biome);
+            }
+        }
+
+        for (Biome biome : net.minecraftforge.common.BiomeManager.strongHoldBiomes)
+        {
+            if (!this.allowedBiomes.contains(biome))
             {
                 this.allowedBiomes.add(biome);
             }
