--- before/net/minecraft/world/gen/structure/MapGenScatteredFeature.java
+++ after/net/minecraft/world/gen/structure/MapGenScatteredFeature.java
@@ -5,6 +5,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Random;
+import java.util.Map.Entry;
 import net.minecraft.entity.monster.EntityWitch;
 import net.minecraft.init.Biomes;
 import net.minecraft.util.math.BlockPos;
@@ -14,13 +15,16 @@
 
 public class MapGenScatteredFeature extends MapGenStructure
 {
-    private static final List<Biome> field_75061_e = Arrays.asList(Biomes.field_76769_d, Biomes.field_76786_s, Biomes.field_76782_w, Biomes.field_76792_x, Biomes.field_76780_h, Biomes.field_76774_n, Biomes.field_150584_S);
-    private final List<Biome.SpawnListEntry> field_82668_f = Lists.newArrayList();
-    private int field_82669_g = 32;
-    private final int field_82670_h = 8;
+    private static final List<Biome> field_75061_e = Arrays.<Biome>asList(Biomes.field_76769_d, Biomes.field_76786_s, Biomes.field_76782_w, Biomes.field_76792_x, Biomes.field_76780_h, Biomes.field_76774_n, Biomes.field_150584_S);
+    private final List<Biome.SpawnListEntry> field_82668_f;
+    private int field_82669_g;
+    private final int field_82670_h;
 
     public MapGenScatteredFeature()
     {
+        this.field_82668_f = Lists.<Biome.SpawnListEntry>newArrayList();
+        this.field_82669_g = 32;
+        this.field_82670_h = 8;
         this.field_82668_f.add(new Biome.SpawnListEntry(EntityWitch.class, 1, 1, 1));
     }
 
@@ -28,9 +32,9 @@
     {
         this();
 
-        for (Map.Entry<String, String> entry : p_i2061_1_.entrySet())
+        for (Entry<String, String> entry : p_i2061_1_.entrySet())
         {
-            if (entry.getKey().equals("distance"))
+            if (((String)entry.getKey()).equals("distance"))
             {
                 this.field_82669_g = MathHelper.func_82714_a(entry.getValue(), this.field_82669_g, 9);
             }
@@ -60,10 +64,10 @@
         int k = p_75047_1_ / this.field_82669_g;
         int l = p_75047_2_ / this.field_82669_g;
         Random random = this.field_75039_c.func_72843_D(k, l, 14357617);
-        k *= this.field_82669_g;
-        l *= this.field_82669_g;
-        k += random.nextInt(this.field_82669_g - 8);
-        l += random.nextInt(this.field_82669_g - 8);
+        k = k * this.field_82669_g;
+        l = l * this.field_82669_g;
+        k = k + random.nextInt(this.field_82669_g - 8);
+        l = l + random.nextInt(this.field_82669_g - 8);
 
         if (i == k && j == l)
         {
