--- before/net/minecraft/world/gen/structure/MapGenMineshaft.java
+++ after/net/minecraft/world/gen/structure/MapGenMineshaft.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.gen.structure;
 
 import java.util.Map;
+import java.util.Map.Entry;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
@@ -22,9 +23,9 @@
 
     public MapGenMineshaft(Map<String, String> p_i2034_1_)
     {
-        for (Map.Entry<String, String> entry : p_i2034_1_.entrySet())
+        for (Entry<String, String> entry : p_i2034_1_.entrySet())
         {
-            if (entry.getKey().equals("chance"))
+            if (((String)entry.getKey()).equals("chance"))
             {
                 this.field_82673_e = MathHelper.func_82712_a(entry.getValue(), this.field_82673_e);
             }
