--- before/net/minecraft/world/gen/structure/MapGenNetherBridge.java
+++ after/net/minecraft/world/gen/structure/MapGenNetherBridge.java
@@ -14,7 +14,7 @@
 
 public class MapGenNetherBridge extends MapGenStructure
 {
-    private final List<Biome.SpawnListEntry> field_75060_e = Lists.newArrayList();
+    private final List<Biome.SpawnListEntry> field_75060_e = Lists.<Biome.SpawnListEntry>newArrayList();
 
     public MapGenNetherBridge()
     {
