--- before/net/minecraft/world/gen/structure/StructureOceanMonument.java
+++ after/net/minecraft/world/gen/structure/StructureOceanMonument.java
@@ -7,6 +7,7 @@
 import java.util.Map;
 import java.util.Random;
 import java.util.Set;
+import java.util.Map.Entry;
 import net.minecraft.entity.monster.EntityGuardian;
 import net.minecraft.init.Biomes;
 import net.minecraft.nbt.NBTTagCompound;
@@ -20,27 +21,29 @@
 
 public class StructureOceanMonument extends MapGenStructure
 {
-    private int field_175800_f = 32;
-    private int field_175801_g = 5;
-    public static final List<Biome> field_175802_d = Arrays.asList(Biomes.field_76771_b, Biomes.field_150575_M, Biomes.field_76781_i, Biomes.field_76776_l, Biomes.field_76777_m);
-    public static final List<Biome> field_186134_b = Arrays.asList(Biomes.field_150575_M);
-    private static final List<Biome.SpawnListEntry> field_175803_h = Lists.newArrayList();
+    private int field_175800_f;
+    private int field_175801_g;
+    public static final List<Biome> field_175802_d = Arrays.<Biome>asList(Biomes.field_76771_b, Biomes.field_150575_M, Biomes.field_76781_i, Biomes.field_76776_l, Biomes.field_76777_m);
+    public static final List<Biome> field_186134_b = Arrays.<Biome>asList(Biomes.field_150575_M);
+    private static final List<Biome.SpawnListEntry> field_175803_h = Lists.<Biome.SpawnListEntry>newArrayList();
 
     public StructureOceanMonument()
     {
+        this.field_175800_f = 32;
+        this.field_175801_g = 5;
     }
 
     public StructureOceanMonument(Map<String, String> p_i45608_1_)
     {
         this();
 
-        for (Map.Entry<String, String> entry : p_i45608_1_.entrySet())
+        for (Entry<String, String> entry : p_i45608_1_.entrySet())
         {
-            if (entry.getKey().equals("spacing"))
+            if (((String)entry.getKey()).equals("spacing"))
             {
                 this.field_175800_f = MathHelper.func_82714_a(entry.getValue(), this.field_175800_f, 1);
             }
-            else if (entry.getKey().equals("separation"))
+            else if (((String)entry.getKey()).equals("separation"))
             {
                 this.field_175801_g = MathHelper.func_82714_a(entry.getValue(), this.field_175801_g, 1);
             }
@@ -70,10 +73,10 @@
         int k = p_75047_1_ / this.field_175800_f;
         int l = p_75047_2_ / this.field_175800_f;
         Random random = this.field_75039_c.func_72843_D(k, l, 10387313);
-        k *= this.field_175800_f;
-        l *= this.field_175800_f;
-        k += (random.nextInt(this.field_175800_f - this.field_175801_g) + random.nextInt(this.field_175800_f - this.field_175801_g)) / 2;
-        l += (random.nextInt(this.field_175800_f - this.field_175801_g) + random.nextInt(this.field_175800_f - this.field_175801_g)) / 2;
+        k = k * this.field_175800_f;
+        l = l * this.field_175800_f;
+        k = k + (random.nextInt(this.field_175800_f - this.field_175801_g) + random.nextInt(this.field_175800_f - this.field_175801_g)) / 2;
+        l = l + (random.nextInt(this.field_175800_f - this.field_175801_g) + random.nextInt(this.field_175800_f - this.field_175801_g)) / 2;
 
         if (i == k && j == l)
         {
@@ -116,7 +119,7 @@
 
     public static class StartMonument extends StructureStart
         {
-            private final Set<ChunkPos> field_175791_c = Sets.newHashSet();
+            private final Set<ChunkPos> field_175791_c = Sets.<ChunkPos>newHashSet();
             private boolean field_175790_d;
 
             public StartMonument()
