--- before/net/minecraft/world/gen/structure/StructureStart.java
+++ after/net/minecraft/world/gen/structure/StructureStart.java
@@ -11,7 +11,7 @@
 
 public abstract class StructureStart
 {
-    protected List<StructureComponent> field_75075_a = Lists.newLinkedList();
+    protected List<StructureComponent> field_75075_a = Lists.<StructureComponent>newLinkedList();
     protected StructureBoundingBox field_75074_b;
     private int field_143024_c;
     private int field_143023_d;
@@ -63,6 +63,10 @@
 
     public NBTTagCompound func_143021_a(int p_143021_1_, int p_143021_2_)
     {
+        if (MapGenStructureIO.func_143033_a(this) == null) // This is just a more friendly error instead of the 'Null String' below
+        {
+            throw new RuntimeException("StructureStart \"" + this.getClass().getName() + "\" missing ID Mapping, Modder see MapGenStructureIO");
+        }
         NBTTagCompound nbttagcompound = new NBTTagCompound();
         nbttagcompound.func_74778_a("id", MapGenStructureIO.func_143033_a(this));
         nbttagcompound.func_74768_a("ChunkX", p_143021_1_);
@@ -96,9 +100,10 @@
 
         NBTTagList nbttaglist = p_143020_2_.func_150295_c("Children", 10);
 
-        for (int i = 0; i < nbttaglist.func_74745_c(); i++)
+        for (int i = 0; i < nbttaglist.func_74745_c(); ++i)
         {
-            this.field_75075_a.add(MapGenStructureIO.func_143032_b(nbttaglist.func_150305_b(i), p_143020_1_));
+            StructureComponent tmp = MapGenStructureIO.func_143032_b(nbttaglist.func_150305_b(i), p_143020_1_);
+            if (tmp != null) this.field_75075_a.add(tmp); //Forge: Prevent NPEs further down the line when a componenet can't be loaded.
         }
 
         this.func_143017_b(p_143020_2_);
