--- before/net/minecraft/world/gen/structure/StructureStrongholdPieces.java
+++ after/net/minecraft/world/gen/structure/StructureStrongholdPieces.java
@@ -67,7 +67,7 @@
 
     public static void func_75198_a()
     {
-        field_75206_c = Lists.newArrayList();
+        field_75206_c = Lists.<StructureStrongholdPieces.PieceWeight>newArrayList();
 
         for (StructureStrongholdPieces.PieceWeight structurestrongholdpieces$pieceweight : field_75205_b)
         {
@@ -204,7 +204,15 @@
             }
 
             StructureBoundingBox structureboundingbox = StructureStrongholdPieces.Corridor.func_175869_a(p_175955_1_, p_175955_2_, p_175955_3_, p_175955_4_, p_175955_5_, p_175955_6_);
-            return structureboundingbox != null && structureboundingbox.field_78895_b > 1 ? new StructureStrongholdPieces.Corridor(p_175955_7_, p_175955_2_, structureboundingbox, p_175955_6_) : null;
+
+            if (structureboundingbox != null && structureboundingbox.field_78895_b > 1)
+            {
+                return new StructureStrongholdPieces.Corridor(p_175955_7_, p_175955_2_, structureboundingbox, p_175955_6_);
+            }
+            else
+            {
+                return null;
+            }
         }
     }
 
@@ -871,18 +879,18 @@
                     flag &= aboolean[l];
                 }
 
-                this.func_175811_a(p_74875_1_, iblockstate4.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[0]), 4, 3, 8, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate4.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[1]), 5, 3, 8, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate4.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[2]), 6, 3, 8, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[3]), 4, 3, 12, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[4]), 5, 3, 12, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[5]), 6, 3, 12, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate1.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[6]), 3, 3, 9, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate1.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[7]), 3, 3, 10, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate1.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[8]), 3, 3, 11, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate2.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[9]), 7, 3, 9, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate2.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[10]), 7, 3, 10, p_74875_3_);
-                this.func_175811_a(p_74875_1_, iblockstate2.func_177226_a(BlockEndPortalFrame.field_176507_b, aboolean[11]), 7, 3, 11, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate4.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[0])), 4, 3, 8, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate4.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[1])), 5, 3, 8, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate4.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[2])), 6, 3, 8, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[3])), 4, 3, 12, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[4])), 5, 3, 12, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[5])), 6, 3, 12, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate1.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[6])), 3, 3, 9, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate1.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[7])), 3, 3, 10, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate1.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[8])), 3, 3, 11, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate2.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[9])), 7, 3, 9, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate2.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[10])), 7, 3, 10, p_74875_3_);
+                this.func_175811_a(p_74875_1_, iblockstate2.func_177226_a(BlockEndPortalFrame.field_176507_b, Boolean.valueOf(aboolean[11])), 7, 3, 11, p_74875_3_);
 
                 if (flag)
                 {
@@ -1267,7 +1275,7 @@
         {
             public StructureStrongholdPieces.PieceWeight field_75027_a;
             public StructureStrongholdPieces.PortalRoom field_75025_b;
-            public List<StructureComponent> field_75026_c = Lists.newArrayList();
+            public List<StructureComponent> field_75026_c = Lists.<StructureComponent>newArrayList();
 
             public Stairs2()
             {
