--- before/net/minecraft/world/gen/structure/StructureMineshaftPieces.java
+++ after/net/minecraft/world/gen/structure/StructureMineshaftPieces.java
@@ -683,7 +683,7 @@
 
     public static class Room extends StructureMineshaftPieces.Peice
         {
-            private final List<StructureBoundingBox> field_74949_a = Lists.newLinkedList();
+            private final List<StructureBoundingBox> field_74949_a = Lists.<StructureBoundingBox>newLinkedList();
 
             public Room()
             {
@@ -708,9 +708,9 @@
 
                 int k;
 
-                for (k = 0; k < this.field_74887_e.func_78883_b(); k += 4)
+                for (k = 0; k < this.field_74887_e.func_78883_b(); k = k + 4)
                 {
-                    k += p_74861_3_.nextInt(this.field_74887_e.func_78883_b());
+                    k = k + p_74861_3_.nextInt(this.field_74887_e.func_78883_b());
 
                     if (k + 3 > this.field_74887_e.func_78883_b())
                     {
@@ -726,9 +726,9 @@
                     }
                 }
 
-                for (k = 0; k < this.field_74887_e.func_78883_b(); k += 4)
+                for (k = 0; k < this.field_74887_e.func_78883_b(); k = k + 4)
                 {
-                    k += p_74861_3_.nextInt(this.field_74887_e.func_78883_b());
+                    k = k + p_74861_3_.nextInt(this.field_74887_e.func_78883_b());
 
                     if (k + 3 > this.field_74887_e.func_78883_b())
                     {
@@ -744,9 +744,9 @@
                     }
                 }
 
-                for (k = 0; k < this.field_74887_e.func_78880_d(); k += 4)
+                for (k = 0; k < this.field_74887_e.func_78880_d(); k = k + 4)
                 {
-                    k += p_74861_3_.nextInt(this.field_74887_e.func_78880_d());
+                    k = k + p_74861_3_.nextInt(this.field_74887_e.func_78880_d());
 
                     if (k + 3 > this.field_74887_e.func_78880_d())
                     {
@@ -762,9 +762,9 @@
                     }
                 }
 
-                for (k = 0; k < this.field_74887_e.func_78880_d(); k += 4)
+                for (k = 0; k < this.field_74887_e.func_78880_d(); k = k + 4)
                 {
-                    k += p_74861_3_.nextInt(this.field_74887_e.func_78880_d());
+                    k = k + p_74861_3_.nextInt(this.field_74887_e.func_78880_d());
 
                     if (k + 3 > this.field_74887_e.func_78880_d())
                     {
