--- before/net/minecraft/init/Blocks.java
+++ after/net/minecraft/init/Blocks.java
@@ -319,7 +319,7 @@
         }
         else
         {
-            field_185780_dg = Sets.newHashSet();
+            field_185780_dg = Sets.<Block>newHashSet();
             field_150350_a = func_180383_a("air");
             field_150348_b = func_180383_a("stone");
             field_150349_c = (BlockGrass)func_180383_a("grass");
