--- before/net/minecraft/init/PotionTypes.java
+++ after/net/minecraft/init/PotionTypes.java
@@ -67,7 +67,7 @@
         }
         else
         {
-            field_185228_K = Sets.newHashSet();
+            field_185228_K = Sets.<PotionType>newHashSet();
             field_185229_a = func_185217_a("empty");
             field_185230_b = func_185217_a("water");
             field_185231_c = func_185217_a("mundane");
