--- original/net/minecraft/stats/StatisticsManagerServer.java
+++ modified/net/minecraft/stats/StatisticsManagerServer.java
@@ -69,7 +69,8 @@
     }
 
     public Map<StatBase, TupleIntJsonSerializable> parseJson(String p_150881_1_) {
-        JsonElement jsonelement = new JsonParser().parse(p_150881_1_);
+        JsonElement jsonelement = JsonParser.parseString(p_150881_1_);
+
         if (!jsonelement.isJsonObject()) {
             return Maps.newHashMap();
         }
