--- before/net/minecraft/util/registry/RegistryNamespacedDefaultedByKey.java
+++ after/net/minecraft/util/registry/RegistryNamespacedDefaultedByKey.java
@@ -39,28 +39,28 @@
     @Nonnull
     public K func_177774_c(V p_177774_1_)
     {
-        K k = super.func_177774_c(p_177774_1_);
+        K k = (K)super.func_177774_c(p_177774_1_);
         return (K)(k == null ? this.field_148760_d : k);
     }
 
     @Nonnull
     public V func_82594_a(@Nullable K p_82594_1_)
     {
-        V v = super.func_82594_a(p_82594_1_);
+        V v = (V)super.func_82594_a(p_82594_1_);
         return (V)(v == null ? this.field_148761_e : v);
     }
 
     @Nonnull
     public V func_148754_a(int p_148754_1_)
     {
-        V v = super.func_148754_a(p_148754_1_);
+        V v = (V)super.func_148754_a(p_148754_1_);
         return (V)(v == null ? this.field_148761_e : v);
     }
 
     @Nonnull
     public V func_186801_a(Random p_186801_1_)
     {
-        V v = super.func_186801_a(p_186801_1_);
+        V v = (V)super.func_186801_a(p_186801_1_);
         return (V)(v == null ? this.field_148761_e : v);
     }
 }
