--- original/net/minecraft/client/renderer/tileentity/TileEntitySpecialRenderer.java
+++ modified/net/minecraft/client/renderer/tileentity/TileEntitySpecialRenderer.java
@@ -75,6 +75,8 @@
         return false;
     }
 
+    public void renderTileEntityFast(T te, double x, double y, double z, float partialTicks, int destroyStage, float partial, net.minecraft.client.renderer.BufferBuilder buffer) {}
+
     protected void drawNameplate(T te, String str, double x, double y, double z, int maxDistance) {
         Entity entity = this.rendererDispatcher.entity;
         double d0 = te.getDistanceSq(entity.posX, entity.posY, entity.posZ);
