--- before/net/minecraft/pathfinding/PathFinder.java
+++ after/net/minecraft/pathfinding/PathFinder.java
@@ -11,7 +11,7 @@
 public class PathFinder
 {
     private final PathHeap field_75866_b = new PathHeap();
-    private final Set<PathPoint> field_186337_b = Sets.newHashSet();
+    private final Set<PathPoint> field_186337_b = Sets.<PathPoint>newHashSet();
     private final PathPoint[] field_75864_d = new PathPoint[32];
     private final NodeProcessor field_176190_c;
 
@@ -108,7 +108,15 @@
             }
         }
 
-        return pathpoint == p_186335_1_ ? null : this.func_75853_a(p_186335_1_, pathpoint);
+        if (pathpoint == p_186335_1_)
+        {
+            return null;
+        }
+        else
+        {
+            Path path = this.func_75853_a(p_186335_1_, pathpoint);
+            return path;
+        }
     }
 
     private Path func_75853_a(PathPoint p_75853_1_, PathPoint p_75853_2_)
