--- before/net/minecraft/pathfinding/PathNavigateClimber.java
+++ after/net/minecraft/pathfinding/PathNavigateClimber.java
@@ -57,7 +57,7 @@
         {
             if (this.targetPosition != null)
             {
-                double d0 = (double)(this.entity.width * this.entity.width);
+                double d0 = Math.max((this.entity.width * this.entity.width), 1.0D);
 
                 if (!(this.entity.getDistanceSqToCenter(this.targetPosition) < d0)
                         && (
