--- before/net/minecraft/pathfinding/PathNavigateSwimmer.java
+++ after/net/minecraft/pathfinding/PathNavigateSwimmer.java
@@ -43,7 +43,7 @@
         {
             Vec3d vec3d1 = this.field_75514_c.func_75881_a(this.field_75515_a, j);
 
-            if (!(vec3d1.func_72436_e(vec3d) > 36.0D) && this.func_75493_a(vec3d, vec3d1, 0, 0, 0))
+            if (vec3d1.func_72436_e(vec3d) <= 36.0D && this.func_75493_a(vec3d, vec3d1, 0, 0, 0))
             {
                 this.field_75514_c.func_75872_c(j);
                 break;
