--- before/net/minecraft/pathfinding/PathHeap.java
+++ after/net/minecraft/pathfinding/PathHeap.java
@@ -72,7 +72,7 @@
             i = p_75847_1_ - 1 >> 1;
             PathPoint pathpoint1 = this.field_75852_a[i];
 
-            if (!(f < pathpoint1.field_75834_g))
+            if (f >= pathpoint1.field_75834_g)
             {
                 break;
             }
@@ -118,7 +118,7 @@
 
             if (f1 < f2)
             {
-                if (!(f1 < f))
+                if (f1 >= f)
                 {
                     break;
                 }
@@ -129,7 +129,7 @@
             }
             else
             {
-                if (!(f2 < f))
+                if (f2 >= f)
                 {
                     break;
                 }
