--- before/net/minecraft/world/storage/loot/properties/EntityPropertyManager.java
+++ after/net/minecraft/world/storage/loot/properties/EntityPropertyManager.java
@@ -6,8 +6,8 @@
 
 public class EntityPropertyManager
 {
-    private static final Map < ResourceLocation, EntityProperty.Serializer<? >> field_186647_a = Maps.newHashMap();
-    private static final Map < Class <? extends EntityProperty > , EntityProperty.Serializer<? >> field_186648_b = Maps.newHashMap();
+    private static final Map < ResourceLocation, EntityProperty.Serializer<? >> field_186647_a = Maps. < ResourceLocation, EntityProperty.Serializer<? >> newHashMap();
+    private static final Map < Class <? extends EntityProperty > , EntityProperty.Serializer<? >> field_186648_b = Maps. < Class <? extends EntityProperty > , EntityProperty.Serializer<? >> newHashMap();
 
     public static <T extends EntityProperty> void func_186644_a(EntityProperty.Serializer <? extends T > p_186644_0_)
     {
@@ -31,7 +31,7 @@
 
     public static EntityProperty.Serializer<?> func_186646_a(ResourceLocation p_186646_0_)
     {
-        EntityProperty.Serializer<?> serializer = field_186647_a.get(p_186646_0_);
+        EntityProperty.Serializer<?> serializer = (EntityProperty.Serializer)field_186647_a.get(p_186646_0_);
 
         if (serializer == null)
         {
@@ -45,7 +45,7 @@
 
     public static <T extends EntityProperty> EntityProperty.Serializer<T> func_186645_a(T p_186645_0_)
     {
-        EntityProperty.Serializer<?> serializer = field_186648_b.get(p_186645_0_.getClass());
+        EntityProperty.Serializer<?> serializer = (EntityProperty.Serializer)field_186648_b.get(p_186645_0_.getClass());
 
         if (serializer == null)
         {
