--- original/net/minecraft/client/renderer/chunk/ChunkRenderDispatcher.java
+++ modified/net/minecraft/client/renderer/chunk/ChunkRenderDispatcher.java
@@ -42,9 +42,15 @@
     private final ChunkRenderWorker renderWorker;
 
     public ChunkRenderDispatcher() {
+        this(-1);
+    }
+
+    public ChunkRenderDispatcher(int countRenderBuilders) {
         int i = Math.max(1, (int)(Runtime.getRuntime().maxMemory() * 0.3) / 10485760);
         int j = Math.max(1, MathHelper.clamp(Runtime.getRuntime().availableProcessors(), 1, i / 5));
-        this.countRenderBuilders = MathHelper.clamp(j * 10, 1, i);
+        if(countRenderBuilders < 0) countRenderBuilders = MathHelper.clamp(j * 10, 1, i);
+        this.countRenderBuilders = countRenderBuilders;
+
         if (j > 1) {
             for (int k = 0; k < j; k++) {
                 ChunkRenderWorker chunkrenderworker = new ChunkRenderWorker(this);
