--- before/net/minecraft/stats/StatisticsManager.java
+++ after/net/minecraft/stats/StatisticsManager.java
@@ -7,7 +7,7 @@
 
 public class StatisticsManager
 {
-    protected final Map<StatBase, TupleIntJsonSerializable> field_150875_a = Maps.newConcurrentMap();
+    protected final Map<StatBase, TupleIntJsonSerializable> field_150875_a = Maps.<StatBase, TupleIntJsonSerializable>newConcurrentMap();
 
     public void func_150871_b(EntityPlayer p_150871_1_, StatBase p_150871_2_, int p_150871_3_)
     {
