--- before/net/minecraft/server/integrated/IntegratedServer.java
+++ after/net/minecraft/server/integrated/IntegratedServer.java
@@ -7,6 +7,7 @@
 import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
 import java.io.File;
 import java.io.IOException;
+import java.net.InetAddress;
 import net.minecraft.client.ClientBrandRetriever;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.multiplayer.ThreadLanServerPing;
@@ -45,26 +46,9 @@
     private boolean field_71346_p;
     private ThreadLanServerPing field_71345_q;
 
-    public IntegratedServer(
-        Minecraft p_i46523_1_,
-        String p_i46523_2_,
-        String p_i46523_3_,
-        WorldSettings p_i46523_4_,
-        YggdrasilAuthenticationService p_i46523_5_,
-        MinecraftSessionService p_i46523_6_,
-        GameProfileRepository p_i46523_7_,
-        PlayerProfileCache p_i46523_8_
-    )
+    public IntegratedServer(Minecraft p_i46523_1_, String p_i46523_2_, String p_i46523_3_, WorldSettings p_i46523_4_, YggdrasilAuthenticationService p_i46523_5_, MinecraftSessionService p_i46523_6_, GameProfileRepository p_i46523_7_, PlayerProfileCache p_i46523_8_)
     {
-        super(
-            new File(p_i46523_1_.field_71412_D, "saves"),
-            p_i46523_1_.func_110437_J(),
-            p_i46523_1_.func_184126_aj(),
-            p_i46523_5_,
-            p_i46523_6_,
-            p_i46523_7_,
-            p_i46523_8_
-        );
+        super(new File(p_i46523_1_.field_71412_D, "saves"), p_i46523_1_.func_110437_J(), p_i46523_1_.func_184126_aj(), p_i46523_5_, p_i46523_6_, p_i46523_7_, p_i46523_8_);
         this.func_71224_l(p_i46523_1_.func_110432_I().func_111285_a());
         this.func_71261_m(p_i46523_2_);
         this.func_71246_n(p_i46523_3_);
@@ -76,18 +60,14 @@
         this.field_71350_m = this.func_71242_L() ? WorldServerDemo.field_73071_a : p_i46523_4_;
     }
 
-    @Override
     public ServerCommandManager func_175582_h()
     {
         return new IntegratedServerCommandManager(this);
     }
 
-    @Override
     public void func_71247_a(String p_71247_1_, String p_71247_2_, long p_71247_3_, WorldType p_71247_5_, String p_71247_6_)
     {
         this.func_71237_c(p_71247_1_);
-        this.field_71305_c = new WorldServer[3];
-        this.field_71312_k = new long[this.field_71305_c.length][100];
         ISaveHandler isavehandler = this.func_71254_M().func_75804_a(p_71247_1_, true);
         this.func_175584_a(this.func_71270_I(), isavehandler);
         WorldInfo worldinfo = isavehandler.func_75757_d();
@@ -101,7 +81,8 @@
             worldinfo.func_76062_a(p_71247_2_);
         }
 
-        for (int i = 0; i < this.field_71305_c.length; i++)
+        if (false) { //Forge: Dead Code, implement below.
+        for (int i = 0; i < this.field_71305_c.length; ++i)
         {
             int j = 0;
 
@@ -119,26 +100,41 @@
             {
                 if (this.func_71242_L())
                 {
-                    this.field_71305_c[i] = (WorldServer)new WorldServerDemo(this, isavehandler, worldinfo, j, this.field_71304_b).func_175643_b();
+                    this.field_71305_c[i] = (WorldServer)(new WorldServerDemo(this, isavehandler, worldinfo, j, this.field_71304_b)).func_175643_b();
                 }
                 else
                 {
-                    this.field_71305_c[i] = (WorldServer)new WorldServer(this, isavehandler, worldinfo, j, this.field_71304_b).func_175643_b();
+                    this.field_71305_c[i] = (WorldServer)(new WorldServer(this, isavehandler, worldinfo, j, this.field_71304_b)).func_175643_b();
                 }
 
                 this.field_71305_c[i].func_72963_a(this.field_71350_m);
             }
             else
             {
-                this.field_71305_c[i] = (WorldServer)new WorldServerMulti(this, isavehandler, j, this.field_71305_c[0], this.field_71304_b).func_175643_b();
+                this.field_71305_c[i] = (WorldServer)(new WorldServerMulti(this, isavehandler, j, this.field_71305_c[0], this.field_71304_b)).func_175643_b();
             }
 
             this.field_71305_c[i].func_72954_a(new ServerWorldEventHandler(this, this.field_71305_c[i]));
         }
-
-        this.func_184103_al().func_72364_a(this.field_71305_c);
-
-        if (this.field_71305_c[0].func_72912_H().func_176130_y() == null)
+        }// Forge: End Dead Code
+
+        WorldServer overWorld = (func_71242_L() ? (WorldServer)(new WorldServerDemo(this, isavehandler, worldinfo, 0, this.field_71304_b)).func_175643_b() :
+                                            (WorldServer)(new WorldServer(this, isavehandler, worldinfo, 0, this.field_71304_b)).func_175643_b());
+        overWorld.func_72963_a(this.field_71350_m);
+        for (int dim : net.minecraftforge.common.DimensionManager.getStaticDimensionIDs())
+        {
+            WorldServer world = (dim == 0 ? overWorld : (WorldServer)(new WorldServerMulti(this, isavehandler, dim, overWorld, this.field_71304_b)).func_175643_b());
+            world.func_72954_a(new ServerWorldEventHandler(this, world));
+            if (!this.func_71264_H())
+            {
+                world.func_72912_H().func_76060_a(func_71265_f());
+            }
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Load(world));
+        }
+
+        this.func_184103_al().func_72364_a(new WorldServer[]{ overWorld });
+
+        if (overWorld.func_72912_H().func_176130_y() == null)
         {
             this.func_147139_a(this.field_71349_l.field_71474_y.field_74318_M);
         }
@@ -146,7 +142,6 @@
         this.func_71222_d();
     }
 
-    @Override
     public boolean func_71197_b() throws IOException
     {
         field_147148_h.info("Starting integrated minecraft server version 1.12.2");
@@ -157,14 +152,12 @@
         this.func_71245_h(true);
         field_147148_h.info("Generating keypair");
         this.func_71253_a(CryptManager.func_75891_b());
-        this.func_71247_a(
-            this.func_71270_I(), this.func_71221_J(), this.field_71350_m.func_77160_d(), this.field_71350_m.func_77165_h(), this.field_71350_m.func_82749_j()
-        );
+        if (!net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerAboutToStart(this)) return false;
+        this.func_71247_a(this.func_71270_I(), this.func_71221_J(), this.field_71350_m.func_77160_d(), this.field_71350_m.func_77165_h(), this.field_71350_m.func_82749_j());
         this.func_71205_p(this.func_71214_G() + " - " + this.field_71305_c[0].func_72912_H().func_76065_j());
-        return true;
+        return net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerStarting(this);
     }
 
-    @Override
     public void func_71217_p()
     {
         boolean flag = this.field_71348_o;
@@ -193,7 +186,7 @@
 
             if (this.field_71349_l.field_71474_y.field_151451_c != this.func_184103_al().func_72395_o())
             {
-                field_147148_h.info("Changing view distance to {}, from {}", this.field_71349_l.field_71474_y.field_151451_c, this.func_184103_al().func_72395_o());
+                field_147148_h.info("Changing view distance to {}, from {}", Integer.valueOf(this.field_71349_l.field_71474_y.field_151451_c), Integer.valueOf(this.func_184103_al().func_72395_o()));
                 this.func_184103_al().func_152611_a(this.field_71349_l.field_71474_y.field_151451_c);
             }
 
@@ -209,7 +202,7 @@
                 }
                 else if (worldinfo.func_176123_z() && !worldinfo1.func_176123_z())
                 {
-                    field_147148_h.info("Locking difficulty to {}", worldinfo.func_176130_y());
+                    field_147148_h.info("Locking difficulty to {}", (Object)worldinfo.func_176130_y());
 
                     for (WorldServer worldserver : this.field_71305_c)
                     {
@@ -223,73 +216,62 @@
         }
     }
 
-    @Override
     public boolean func_71225_e()
     {
         return false;
     }
 
-    @Override
     public GameType func_71265_f()
     {
         return this.field_71350_m.func_77162_e();
     }
 
-    @Override
     public EnumDifficulty func_147135_j()
     {
+        if (this.field_71349_l.field_71441_e == null) return this.field_71349_l.field_71474_y.field_74318_M; // Fix NPE just in case.
         return this.field_71349_l.field_71441_e.func_72912_H().func_176130_y();
     }
 
-    @Override
     public boolean func_71199_h()
     {
         return this.field_71350_m.func_77158_f();
     }
 
-    @Override
     public boolean func_181034_q()
     {
         return true;
     }
 
-    @Override
     public boolean func_183002_r()
     {
         return true;
     }
 
-    @Override
     public void func_71267_a(boolean p_71267_1_)
     {
         super.func_71267_a(p_71267_1_);
     }
 
-    @Override
     public File func_71238_n()
     {
         return this.field_71349_l.field_71412_D;
     }
 
-    @Override
     public boolean func_71262_S()
     {
         return false;
     }
 
-    @Override
     public boolean func_181035_ah()
     {
         return false;
     }
 
-    @Override
     public void func_71228_a(CrashReport p_71228_1_)
     {
         this.field_71349_l.func_71404_a(p_71228_1_);
     }
 
-    @Override
     public CrashReport func_71230_b(CrashReport p_71230_1_)
     {
         p_71230_1_ = super.func_71230_b(p_71230_1_);
@@ -300,10 +282,7 @@
                 return "Integrated Server (map_client.txt)";
             }
         });
-        p_71230_1_.func_85056_g()
-        .func_189529_a(
-            "Is Modded",
-            new ICrashReportDetail<String>()
+        p_71230_1_.func_85056_g().func_189529_a("Is Modded", new ICrashReportDetail<String>()
         {
             public String call() throws Exception
             {
@@ -323,18 +302,14 @@
                     }
                     else
                     {
-                        return Minecraft.class.getSigners() == null
-                               ? "Very likely; Jar signature invalidated"
-                               : "Probably not. Jar signature remains and both client + server brands are untouched.";
+                        return Minecraft.class.getSigners() == null ? "Very likely; Jar signature invalidated" : "Probably not. Jar signature remains and both client + server brands are untouched.";
                     }
                 }
             }
-        }
-        );
+        });
         return p_71230_1_;
     }
 
-    @Override
     public void func_147139_a(EnumDifficulty p_147139_1_)
     {
         super.func_147139_a(p_147139_1_);
@@ -345,20 +320,17 @@
         }
     }
 
-    @Override
     public void func_70000_a(Snooper p_70000_1_)
     {
         super.func_70000_a(p_70000_1_);
         p_70000_1_.func_152768_a("snooper_partner", this.field_71349_l.func_71378_E().func_80006_f());
     }
 
-    @Override
     public boolean func_70002_Q()
     {
         return Minecraft.func_71410_x().func_70002_Q();
     }
 
-    @Override
     public String func_71206_a(GameType p_71206_1_, boolean p_71206_2_)
     {
         try
@@ -369,8 +341,9 @@
             {
                 i = HttpUtil.func_76181_a();
             }
-            catch (IOException ioexception)
+            catch (IOException var5)
             {
+                ;
             }
 
             if (i <= 0)
@@ -378,8 +351,8 @@
                 i = 25564;
             }
 
-            this.func_147137_ag().func_151265_a(null, i);
-            field_147148_h.info("Started on {}", i);
+            this.func_147137_ag().func_151265_a((InetAddress)null, i);
+            field_147148_h.info("Started on {}", (int)i);
             this.field_71346_p = true;
             this.field_71345_q = new ThreadLanServerPing(this.func_71273_Y(), i + "");
             this.field_71345_q.start();
@@ -388,13 +361,12 @@
             this.field_71349_l.field_71439_g.func_184839_n(p_71206_2_ ? 4 : 0);
             return i + "";
         }
-        catch (IOException ioexception1)
+        catch (IOException var6)
         {
             return null;
         }
     }
 
-    @Override
     public void func_71260_j()
     {
         super.func_71260_j();
@@ -406,12 +378,11 @@
         }
     }
 
-    @Override
     public void func_71263_m()
     {
+        if (func_71278_l())
         Futures.getUnchecked(this.func_152344_a(new Runnable()
         {
-            @Override
             public void run()
             {
                 for (EntityPlayerMP entityplayermp : Lists.newArrayList(IntegratedServer.this.func_184103_al().func_181057_v()))
@@ -437,20 +408,17 @@
         return this.field_71346_p;
     }
 
-    @Override
     public void func_71235_a(GameType p_71235_1_)
     {
         super.func_71235_a(p_71235_1_);
         this.func_184103_al().func_152604_a(p_71235_1_);
     }
 
-    @Override
     public boolean func_82356_Z()
     {
         return true;
     }
 
-    @Override
     public int func_110455_j()
     {
         return 4;
