--- before/net/minecraft/entity/boss/dragon/phase/PhaseSittingFlaming.java
+++ after/net/minecraft/entity/boss/dragon/phase/PhaseSittingFlaming.java
@@ -91,7 +91,7 @@
                 MathHelper.floor(d0), MathHelper.floor(d2), MathHelper.floor(d1)
             );
 
-            while (this.dragon.world.isAirBlock(blockpos$mutableblockpos))
+            while (this.dragon.world.isAirBlock(blockpos$mutableblockpos) && d2 >= 0) //Forge: Fix infinite loop if ground is missing.
             {
                 blockpos$mutableblockpos.setPos(MathHelper.floor(d0), MathHelper.floor(--d2), MathHelper.floor(d1));
             }
