--- before/net/minecraft/entity/boss/dragon/phase/PhaseHoldingPattern.java
+++ after/net/minecraft/entity/boss/dragon/phase/PhaseHoldingPattern.java
@@ -100,7 +100,7 @@
 
             if (this.field_188661_a.func_184664_cU() != null && this.field_188661_a.func_184664_cU().func_186092_c() >= 0)
             {
-                k %= 12;
+                k = k % 12;
 
                 if (k < 0)
                 {
@@ -109,9 +109,9 @@
             }
             else
             {
-                k -= 12;
-                k &= 7;
-                k += 12;
+                k = k - 12;
+                k = k & 7;
+                k = k + 12;
             }
 
             this.field_188677_b = this.field_188661_a.func_184666_a(j, k, (PathPoint)null);
@@ -128,7 +128,7 @@
     private void func_188674_a(EntityPlayer p_188674_1_)
     {
         this.field_188661_a.func_184670_cT().func_188758_a(PhaseList.field_188742_b);
-        this.field_188661_a.func_184670_cT().func_188757_b(PhaseList.field_188742_b).func_188686_a(p_188674_1_);
+        ((PhaseStrafePlayer)this.field_188661_a.func_184670_cT().func_188757_b(PhaseList.field_188742_b)).func_188686_a(p_188674_1_);
     }
 
     private void func_188676_k()
@@ -141,11 +141,15 @@
             double d1 = vec3d.field_72449_c;
             double d2;
 
-            do
+            while (true)
             {
                 d2 = vec3d.field_72448_b + (double)(this.field_188661_a.func_70681_au().nextFloat() * 20.0F);
+
+                if (d2 >= vec3d.field_72448_b)
+                {
+                    break;
+                }
             }
-            while (d2 < vec3d.field_72448_b);
 
             this.field_188678_c = new Vec3d(d0, d2, d1);
         }
