--- before/net/minecraft/entity/boss/dragon/phase/PhaseDying.java
+++ after/net/minecraft/entity/boss/dragon/phase/PhaseDying.java
@@ -40,7 +40,7 @@
 
         double d0 = this.field_188672_b.func_186679_c(this.field_188661_a.field_70165_t, this.field_188661_a.field_70163_u, this.field_188661_a.field_70161_v);
 
-        if (!(d0 < 100.0D) && !(d0 > 22500.0D) && !this.field_188661_a.field_70123_F && !this.field_188661_a.field_70124_G)
+        if (d0 >= 100.0D && d0 <= 22500.0D && !this.field_188661_a.field_70123_F && !this.field_188661_a.field_70124_G)
         {
             this.field_188661_a.func_70606_j(1.0F);
         }
