--- before/net/minecraft/entity/boss/dragon/phase/PhaseLandingApproach.java
+++ after/net/minecraft/entity/boss/dragon/phase/PhaseLandingApproach.java
@@ -92,11 +92,15 @@
             double d1 = vec3d.field_72449_c;
             double d2;
 
-            do
+            while (true)
             {
                 d2 = vec3d.field_72448_b + (double)(this.field_188661_a.func_70681_au().nextFloat() * 20.0F);
+
+                if (d2 >= vec3d.field_72448_b)
+                {
+                    break;
+                }
             }
-            while (d2 < vec3d.field_72448_b);
 
             this.field_188684_c = new Vec3d(d0, d2, d1);
         }
