--- before/net/minecraft/enchantment/EnchantmentFrostWalker.java
+++ after/net/minecraft/enchantment/EnchantmentFrostWalker.java
@@ -67,6 +67,7 @@
                         IBlockState iblockstate1 = worldIn.getBlockState(blockpos$mutableblockpos1);
 
                         if (iblockstate1.getMaterial() == Material.WATER
+                                && (iblockstate1.getBlock() == net.minecraft.init.Blocks.WATER || iblockstate1.getBlock() == net.minecraft.init.Blocks.FLOWING_WATER)
                                 && iblockstate1.getValue(BlockLiquid.LEVEL) == 0
                                 && worldIn.mayPlace(Blocks.FROSTED_ICE, blockpos$mutableblockpos1, false, EnumFacing.DOWN, null))
                         {
