--- before/net/minecraft/enchantment/EnchantmentThorns.java
+++ after/net/minecraft/enchantment/EnchantmentThorns.java
@@ -56,12 +56,12 @@
 
             if (!itemstack.isEmpty())
             {
-                itemstack.damageItem(3, user);
+                damageArmor(itemstack, 3, user);
             }
         }
         else if (!itemstack.isEmpty())
         {
-            itemstack.damageItem(1, user);
+            damageArmor(itemstack, 1, user);
         }
     }
 
@@ -73,5 +73,26 @@
     public static int getDamage(int level, Random rnd)
     {
         return level > 10 ? level - 10 : 1 + rnd.nextInt(4);
+    }
+
+    private void damageArmor(ItemStack stack, int amount, EntityLivingBase entity)
+    {
+        int slot = -1;
+        int x = 0;
+        for (ItemStack i : entity.getArmorInventoryList())
+        {
+            if (i == stack){
+                slot = x;
+                break;
+            }
+            x++;
+        }
+        if (slot == -1 || !(stack.getItem() instanceof net.minecraftforge.common.ISpecialArmor))
+        {
+            stack.damageItem(1, entity);
+            return;
+        }
+        net.minecraftforge.common.ISpecialArmor armor = (net.minecraftforge.common.ISpecialArmor)stack.getItem();
+        armor.damageArmor(entity, stack, DamageSource.causeThornsDamage(entity), amount, slot);
     }
 }
