--- before/net/minecraft/enchantment/EnumEnchantmentType.java
+++ after/net/minecraft/enchantment/EnumEnchantmentType.java
@@ -15,7 +15,6 @@
 public enum EnumEnchantmentType
 {
     ALL {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             for (EnumEnchantmentType enumenchantmenttype : EnumEnchantmentType.values())
@@ -30,77 +29,66 @@
         }
     },
     ARMOR {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             return p_77557_1_ instanceof ItemArmor;
         }
     },
     ARMOR_FEET {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             return p_77557_1_ instanceof ItemArmor && ((ItemArmor)p_77557_1_).field_77881_a == EntityEquipmentSlot.FEET;
         }
     },
     ARMOR_LEGS {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             return p_77557_1_ instanceof ItemArmor && ((ItemArmor)p_77557_1_).field_77881_a == EntityEquipmentSlot.LEGS;
         }
     },
     ARMOR_CHEST {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             return p_77557_1_ instanceof ItemArmor && ((ItemArmor)p_77557_1_).field_77881_a == EntityEquipmentSlot.CHEST;
         }
     },
     ARMOR_HEAD {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             return p_77557_1_ instanceof ItemArmor && ((ItemArmor)p_77557_1_).field_77881_a == EntityEquipmentSlot.HEAD;
         }
     },
     WEAPON {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             return p_77557_1_ instanceof ItemSword;
         }
     },
     DIGGER {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             return p_77557_1_ instanceof ItemTool;
         }
     },
     FISHING_ROD {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             return p_77557_1_ instanceof ItemFishingRod;
         }
     },
     BREAKABLE {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             return p_77557_1_.func_77645_m();
         }
     },
     BOW {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             return p_77557_1_ instanceof ItemBow;
         }
     },
     WEARABLE {
-        @Override
         public boolean func_77557_a(Item p_77557_1_)
         {
             boolean flag = p_77557_1_ instanceof ItemBlock && ((ItemBlock)p_77557_1_).func_179223_d() instanceof BlockPumpkin;
@@ -112,5 +100,13 @@
     {
     }
 
-    public abstract boolean func_77557_a(Item p_77557_1_);
+    private com.google.common.base.Predicate<Item> delegate = null;
+    private EnumEnchantmentType(com.google.common.base.Predicate<Item> delegate)
+    {
+        this.delegate = delegate;
+    }
+    public boolean func_77557_a(Item p_77557_1_)
+    {
+        return this.delegate == null ? false : this.delegate.apply(p_77557_1_);
+    }
 }
