--- original/net/minecraft/entity/EntityFlying.java
+++ modified/net/minecraft/entity/EntityFlying.java
@@ -35,34 +35,28 @@
         } else {
             float f = 0.91F;
             if (this.onGround) {
-                f = this.world
-                        .getBlockState(
-                            new BlockPos(
-                                MathHelper.floor(this.posX),
-                                MathHelper.floor(this.getEntityBoundingBox().minY) - 1,
-                                MathHelper.floor(this.posZ)
-                            )
-                        )
-                        .getBlock()
-                        .slipperiness
-                    * 0.91F;
+                BlockPos underPos =
+                        new BlockPos(
+                            MathHelper.floor(this.posX),
+                            MathHelper.floor(this.getEntityBoundingBox().minY) - 1,
+                            MathHelper.floor(this.posZ)
+                        );
+                IBlockState underState = this.world.getBlockState(underPos);
+                f = underState.getBlock().getSlipperiness(underState, this.world, underPos, this) * 0.91F;
             }
 
             float f1 = 0.16277136F / (f * f * f);
             this.moveRelative(strafe, vertical, forward, this.onGround ? 0.1F * f1 : 0.02F);
             f = 0.91F;
             if (this.onGround) {
-                f = this.world
-                        .getBlockState(
-                            new BlockPos(
-                                MathHelper.floor(this.posX),
-                                MathHelper.floor(this.getEntityBoundingBox().minY) - 1,
-                                MathHelper.floor(this.posZ)
-                            )
-                        )
-                        .getBlock()
-                        .slipperiness
-                    * 0.91F;
+                BlockPos underPos =
+                        new BlockPos(
+                            MathHelper.floor(this.posX),
+                            MathHelper.floor(this.getEntityBoundingBox().minY) - 1,
+                            MathHelper.floor(this.posZ)
+                        );
+                IBlockState underState = this.world.getBlockState(underPos);
+                f = underState.getBlock().getSlipperiness(underState, this.world, underPos, this) * 0.91F;
             }
 
             this.move(MoverType.SELF, this.motionX, this.motionY, this.motionZ);
