--- before/net/minecraft/world/BossInfo.java
+++ after/net/minecraft/world/BossInfo.java
@@ -2,8 +2,6 @@
 
 import java.util.UUID;
 import net.minecraft.util.text.ITextComponent;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public abstract class BossInfo
 {
@@ -55,7 +53,6 @@
         return this.color;
     }
 
-    @SideOnly(Side.CLIENT)
     public void setColor(BossInfo.Color colorIn)
     {
         this.color = colorIn;
@@ -66,7 +63,6 @@
         return this.overlay;
     }
 
-    @SideOnly(Side.CLIENT)
     public void setOverlay(BossInfo.Overlay overlayIn)
     {
         this.overlay = overlayIn;
