--- before/net/minecraft/world/BossInfoServer.java
+++ after/net/minecraft/world/BossInfoServer.java
@@ -9,21 +9,20 @@
 import net.minecraft.network.play.server.SPacketUpdateBossInfo;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.text.ITextComponent;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class BossInfoServer extends BossInfo
 {
-    private final Set<EntityPlayerMP> field_186762_h = Sets.newHashSet();
-    private final Set<EntityPlayerMP> field_186763_i = Collections.unmodifiableSet(this.field_186762_h);
-    private boolean field_186764_j = true;
+    private final Set<EntityPlayerMP> field_186762_h = Sets.<EntityPlayerMP>newHashSet();
+    private final Set<EntityPlayerMP> field_186763_i;
+    private boolean field_186764_j;
 
     public BossInfoServer(ITextComponent p_i46839_1_, BossInfo.Color p_i46839_2_, BossInfo.Overlay p_i46839_3_)
     {
         super(MathHelper.func_188210_a(), p_i46839_1_, p_i46839_2_, p_i46839_3_);
+        this.field_186763_i = Collections.<EntityPlayerMP>unmodifiableSet(this.field_186762_h);
+        this.field_186764_j = true;
     }
 
-    @Override
     public void func_186735_a(float p_186735_1_)
     {
         if (p_186735_1_ != this.field_186750_b)
@@ -33,8 +32,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
     public void func_186745_a(BossInfo.Color p_186745_1_)
     {
         if (p_186745_1_ != this.field_186751_c)
@@ -44,8 +41,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
     public void func_186746_a(BossInfo.Overlay p_186746_1_)
     {
         if (p_186746_1_ != this.field_186752_d)
@@ -55,7 +50,6 @@
         }
     }
 
-    @Override
     public BossInfo func_186741_a(boolean p_186741_1_)
     {
         if (p_186741_1_ != this.field_186753_e)
@@ -67,7 +61,6 @@
         return this;
     }
 
-    @Override
     public BossInfo func_186742_b(boolean p_186742_1_)
     {
         if (p_186742_1_ != this.field_186754_f)
@@ -79,7 +72,6 @@
         return this;
     }
 
-    @Override
     public BossInfo func_186743_c(boolean p_186743_1_)
     {
         if (p_186743_1_ != this.field_186755_g)
@@ -91,7 +83,6 @@
         return this;
     }
 
-    @Override
     public void func_186739_a(ITextComponent p_186739_1_)
     {
         if (!Objects.equal(p_186739_1_, this.field_186749_a))
@@ -138,8 +129,7 @@
 
             for (EntityPlayerMP entityplayermp : this.field_186762_h)
             {
-                entityplayermp.field_71135_a
-                .func_147359_a(new SPacketUpdateBossInfo(p_186758_1_ ? SPacketUpdateBossInfo.Operation.ADD : SPacketUpdateBossInfo.Operation.REMOVE, this));
+                entityplayermp.field_71135_a.func_147359_a(new SPacketUpdateBossInfo(p_186758_1_ ? SPacketUpdateBossInfo.Operation.ADD : SPacketUpdateBossInfo.Operation.REMOVE, this));
             }
         }
     }
