--- before/net/minecraft/world/ChunkCache.java
+++ after/net/minecraft/world/ChunkCache.java
@@ -59,10 +59,9 @@
     }
 
     @Nullable
-    @Override
     public TileEntity func_175625_s(BlockPos p_175625_1_)
     {
-        return this.func_190300_a(p_175625_1_, Chunk.EnumCreateEntityType.IMMEDIATE);
+        return this.func_190300_a(p_175625_1_, Chunk.EnumCreateEntityType.CHECK); // Forge: don't modify world from other threads
     }
 
     @Nullable
@@ -70,11 +69,11 @@
     {
         int i = (p_190300_1_.func_177958_n() >> 4) - this.field_72818_a;
         int j = (p_190300_1_.func_177952_p() >> 4) - this.field_72816_b;
+        if (!withinBounds(i, j)) return null;
         return this.field_72817_c[i][j].func_177424_a(p_190300_1_, p_190300_2_);
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public int func_175626_b(BlockPos p_175626_1_, int p_175626_2_)
     {
         int i = this.func_175629_a(EnumSkyBlock.SKY, p_175626_1_);
@@ -88,7 +87,6 @@
         return i << 20 | j << 4;
     }
 
-    @Override
     public IBlockState func_180495_p(BlockPos p_180495_1_)
     {
         if (p_180495_1_.func_177956_o() >= 0 && p_180495_1_.func_177956_o() < 256)
@@ -111,11 +109,11 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public Biome func_180494_b(BlockPos p_180494_1_)
     {
         int i = (p_180494_1_.func_177958_n() >> 4) - this.field_72818_a;
         int j = (p_180494_1_.func_177952_p() >> 4) - this.field_72816_b;
+        if (!withinBounds(i, j)) return net.minecraft.init.Biomes.field_76772_c;
         return this.field_72817_c[i][j].func_177411_a(p_180494_1_, this.field_72815_e.func_72959_q());
     }
 
@@ -153,6 +151,7 @@
             {
                 int i = (p_175629_2_.func_177958_n() >> 4) - this.field_72818_a;
                 int j = (p_175629_2_.func_177952_p() >> 4) - this.field_72816_b;
+                if (!withinBounds(i, j)) return p_175629_1_.field_77198_c;
                 return this.field_72817_c[i][j].func_177413_a(p_175629_1_, p_175629_2_);
             }
         }
@@ -162,10 +161,10 @@
         }
     }
 
-    @Override
     public boolean func_175623_d(BlockPos p_175623_1_)
     {
-        return this.func_180495_p(p_175623_1_).func_185904_a() == Material.field_151579_a;
+        IBlockState state = this.func_180495_p(p_175623_1_);
+        return state.func_177230_c().isAir(state, this, p_175623_1_);
     }
 
     @SideOnly(Side.CLIENT)
@@ -175,6 +174,7 @@
         {
             int i = (p_175628_2_.func_177958_n() >> 4) - this.field_72818_a;
             int j = (p_175628_2_.func_177952_p() >> 4) - this.field_72816_b;
+            if (!withinBounds(i, j)) return p_175628_1_.field_77198_c;
             return this.field_72817_c[i][j].func_177413_a(p_175628_1_, p_175628_2_);
         }
         else
@@ -183,16 +183,31 @@
         }
     }
 
-    @Override
     public int func_175627_a(BlockPos p_175627_1_, EnumFacing p_175627_2_)
     {
         return this.func_180495_p(p_175627_1_).func_185893_b(this, p_175627_1_, p_175627_2_);
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public WorldType func_175624_G()
     {
         return this.field_72815_e.func_175624_G();
+    }
+
+    @Override
+    public boolean isSideSolid(BlockPos pos, EnumFacing side, boolean _default)
+    {
+        int x = (pos.func_177958_n() >> 4) - this.field_72818_a;
+        int z = (pos.func_177952_p() >> 4) - this.field_72816_b;
+        if (pos.func_177956_o() < 0 || pos.func_177956_o() >= 256) return _default;
+        if (!withinBounds(x, z)) return _default;
+
+        IBlockState state = func_180495_p(pos);
+        return state.func_177230_c().isSideSolid(state, this, pos, side);
+    }
+
+    private boolean withinBounds(int x, int z)
+    {
+        return x >= 0 && x < field_72817_c.length && z >= 0 && z < field_72817_c[x].length && field_72817_c[x][z] != null;
     }
 }
