--- before/net/minecraft/world/ColorizerFoliage.java
+++ after/net/minecraft/world/ColorizerFoliage.java
@@ -15,7 +15,7 @@
 
     public static int func_77470_a(double p_77470_0_, double p_77470_2_)
     {
-        p_77470_2_ *= p_77470_0_;
+        p_77470_2_ = p_77470_2_ * p_77470_0_;
         int i = (int)((1.0D - p_77470_0_) * 255.0D);
         int j = (int)((1.0D - p_77470_2_) * 255.0D);
         return field_77471_a[j << 8 | i];
