--- before/net/minecraft/world/Explosion.java
+++ after/net/minecraft/world/Explosion.java
@@ -41,6 +41,7 @@
     private final float size;
     private final List<BlockPos> affectedBlockPositions = Lists.newArrayList();
     private final Map<EntityPlayer, Vec3d> playerKnockbackMap = Maps.newHashMap();
+    private final Vec3d position;
 
     @SideOnly(Side.CLIENT)
     public Explosion(
@@ -86,6 +87,7 @@
         this.z = z;
         this.causesFire = causesFire;
         this.damagesTerrain = damagesTerrain;
+        position = new Vec3d(this.x, this.y, this.z);
     }
 
     public void doExplosionA()
@@ -122,7 +124,7 @@
                             {
                                 float f2 = this.exploder != null
                                            ? this.exploder.getExplosionResistance(this, this.world, blockpos, iblockstate)
-                                           : iblockstate.getBlock().getExplosionResistance(null);
+                                           : iblockstate.getBlock().getExplosionResistance(world, blockpos, null, this);
                                 f -= (f2 + 0.3F) * 0.3F;
                             }
 
@@ -149,6 +151,7 @@
         int j2 = MathHelper.floor(this.z - f3 - 1.0);
         int j1 = MathHelper.floor(this.z + f3 + 1.0);
         List<Entity> list = this.world.getEntitiesWithinAABBExcludingEntity(this.exploder, new AxisAlignedBB(k1, i2, j2, l1, i1, j1));
+        net.minecraftforge.event.ForgeEventFactory.onExplosionDetonate(this.world, this, list, f3);
         Vec3d vec3d = new Vec3d(this.x, this.y, this.z);
 
         for (int k2 = 0; k2 < list.size(); k2++)
@@ -267,8 +270,7 @@
                         block.dropBlockAsItemWithChance(this.world, blockpos, this.world.getBlockState(blockpos), 1.0F / this.size, 0);
                     }
 
-                    this.world.setBlockState(blockpos, Blocks.AIR.getDefaultState(), 3);
-                    block.onExplosionDestroy(this.world, blockpos, this);
+                    block.onBlockExploded(this.world, blockpos, this);
                 }
             }
         }
@@ -318,4 +320,6 @@
     {
         return this.affectedBlockPositions;
     }
+
+    public Vec3d getPosition(){ return this.position; }
 }
