--- before/net/minecraft/world/GameRules.java
+++ after/net/minecraft/world/GameRules.java
@@ -6,7 +6,7 @@
 
 public class GameRules
 {
-    private final TreeMap<String, GameRules.Value> field_82771_a = new TreeMap<>();
+    private final TreeMap<String, GameRules.Value> field_82771_a = new TreeMap<String, GameRules.Value>();
 
     public GameRules()
     {
@@ -97,7 +97,7 @@
     public String[] func_82763_b()
     {
         Set<String> set = this.field_82771_a.keySet();
-        return set.toArray(new String[set.size()]);
+        return (String[])set.toArray(new String[set.size()]);
     }
 
     public boolean func_82765_e(String p_82765_1_)
@@ -135,16 +135,18 @@
                 {
                     this.field_82761_c = Integer.parseInt(p_82757_1_);
                 }
-                catch (NumberFormatException numberformatexception1)
+                catch (NumberFormatException var4)
                 {
+                    ;
                 }
 
                 try
                 {
                     this.field_82759_d = Double.parseDouble(p_82757_1_);
                 }
-                catch (NumberFormatException numberformatexception)
+                catch (NumberFormatException var3)
                 {
+                    ;
                 }
             }
 
