--- before/net/minecraft/world/ServerWorldEventHandler.java
+++ after/net/minecraft/world/ServerWorldEventHandler.java
@@ -98,7 +98,7 @@
             y,
             z,
             volume > 1.0F ? 16.0F * volume : 16.0,
-            this.world.provider.getDimensionType().getId(),
+            this.world.provider.getDimension(),
             new SPacketSoundEffect(soundIn, category, x, y, z, volume, pitch)
         );
     }
@@ -135,7 +135,7 @@
             blockPosIn.getY(),
             blockPosIn.getZ(),
             64.0,
-            this.world.provider.getDimensionType().getId(),
+            this.world.provider.getDimension(),
             new SPacketEffect(type, blockPosIn, data, false)
         );
     }
