--- before/net/minecraft/world/Teleporter.java
+++ after/net/minecraft/world/Teleporter.java
@@ -15,11 +15,11 @@
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.util.math.MathHelper;
 
-public class Teleporter
+public class Teleporter implements net.minecraftforge.common.util.ITeleporter
 {
     protected final WorldServer field_85192_a;
     protected final Random field_77187_a;
-    protected final Long2ObjectMap<Teleporter.PortalPosition> field_85191_c = new Long2ObjectOpenHashMap<>(4096);
+    protected final Long2ObjectMap<Teleporter.PortalPosition> field_85191_c = new Long2ObjectOpenHashMap<Teleporter.PortalPosition>(4096);
 
     public Teleporter(WorldServer p_i1963_1_)
     {
@@ -55,23 +55,22 @@
                         int j2 = j + l1;
                         int k2 = k + k1 * 0 - j1 * 1;
                         boolean flag = l1 < 0;
-                        this.field_85192_a
-                        .func_175656_a(new BlockPos(i2, j2, k2), flag ? Blocks.field_150343_Z.func_176223_P() : Blocks.field_150350_a.func_176223_P());
+                        this.field_85192_a.func_175656_a(new BlockPos(i2, j2, k2), flag ? Blocks.field_150343_Z.func_176223_P() : Blocks.field_150350_a.func_176223_P());
                     }
                 }
             }
 
             p_180266_1_.func_70012_b((double)i, (double)j, (double)k, p_180266_1_.field_70177_z, 0.0F);
-            p_180266_1_.field_70159_w = 0.0;
-            p_180266_1_.field_70181_x = 0.0;
-            p_180266_1_.field_70179_y = 0.0;
+            p_180266_1_.field_70159_w = 0.0D;
+            p_180266_1_.field_70181_x = 0.0D;
+            p_180266_1_.field_70179_y = 0.0D;
         }
     }
 
     public boolean func_180620_b(Entity p_180620_1_, float p_180620_2_)
     {
         int i = 128;
-        double d0 = -1.0;
+        double d0 = -1.0D;
         int j = MathHelper.func_76128_c(p_180620_1_.field_70165_t);
         int k = MathHelper.func_76128_c(p_180620_1_.field_70161_v);
         boolean flag = true;
@@ -80,8 +79,8 @@
 
         if (this.field_85191_c.containsKey(l))
         {
-            Teleporter.PortalPosition teleporter$portalposition = this.field_85191_c.get(l);
-            d0 = 0.0;
+            Teleporter.PortalPosition teleporter$portalposition = (Teleporter.PortalPosition)this.field_85191_c.get(l);
+            d0 = 0.0D;
             blockpos = teleporter$portalposition;
             teleporter$portalposition.field_85087_d = this.field_85192_a.func_82737_E();
             flag = false;
@@ -96,26 +95,20 @@
 
                 for (int j1 = -128; j1 <= 128; ++j1)
                 {
-                    for (BlockPos blockpos1 = blockpos3.func_177982_a(i1, this.field_85192_a.func_72940_L() - 1 - blockpos3.func_177956_o(), j1);
-                            blockpos1.func_177956_o() >= 0;
-                            blockpos1 = blockpos2
-                        )
+                    for (BlockPos blockpos1 = blockpos3.func_177982_a(i1, this.field_85192_a.func_72940_L() - 1 - blockpos3.func_177956_o(), j1); blockpos1.func_177956_o() >= 0; blockpos1 = blockpos2)
                     {
                         blockpos2 = blockpos1.func_177977_b();
 
                         if (this.field_85192_a.func_180495_p(blockpos1).func_177230_c() == Blocks.field_150427_aO)
                         {
-                            for (blockpos2 = blockpos1.func_177977_b();
-                                    this.field_85192_a.func_180495_p(blockpos2).func_177230_c() == Blocks.field_150427_aO;
-                                    blockpos2 = blockpos2.func_177977_b()
-                                )
+                            for (blockpos2 = blockpos1.func_177977_b(); this.field_85192_a.func_180495_p(blockpos2).func_177230_c() == Blocks.field_150427_aO; blockpos2 = blockpos2.func_177977_b())
                             {
                                 blockpos1 = blockpos2;
                             }
 
                             double d1 = blockpos1.func_177951_i(blockpos3);
 
-                            if (d0 < 0.0 || d1 < d0)
+                            if (d0 < 0.0D || d1 < d0)
                             {
                                 d0 = d1;
                                 blockpos = blockpos1;
@@ -126,22 +119,19 @@
             }
         }
 
-        if (d0 >= 0.0)
+        if (d0 >= 0.0D)
         {
             if (flag)
             {
                 this.field_85191_c.put(l, new Teleporter.PortalPosition(blockpos, this.field_85192_a.func_82737_E()));
             }
 
-            double d5 = (double)blockpos.func_177958_n() + 0.5;
-            double d7 = (double)blockpos.func_177952_p() + 0.5;
+            double d5 = (double)blockpos.func_177958_n() + 0.5D;
+            double d7 = (double)blockpos.func_177952_p() + 0.5D;
             BlockPattern.PatternHelper blockpattern$patternhelper = Blocks.field_150427_aO.func_181089_f(this.field_85192_a, blockpos);
             boolean flag1 = blockpattern$patternhelper.func_177669_b().func_176746_e().func_176743_c() == EnumFacing.AxisDirection.NEGATIVE;
-            double d2 = blockpattern$patternhelper.func_177669_b().func_176740_k() == EnumFacing.Axis.X
-                        ? (double)blockpattern$patternhelper.func_181117_a().func_177952_p()
-                        : (double)blockpattern$patternhelper.func_181117_a().func_177958_n();
-            double d6 = (double)(blockpattern$patternhelper.func_181117_a().func_177956_o() + 1)
-                        - p_180620_1_.func_181014_aG().field_72448_b * (double)blockpattern$patternhelper.func_181119_e();
+            double d2 = blockpattern$patternhelper.func_177669_b().func_176740_k() == EnumFacing.Axis.X ? (double)blockpattern$patternhelper.func_181117_a().func_177952_p() : (double)blockpattern$patternhelper.func_181117_a().func_177958_n();
+            double d6 = (double)(blockpattern$patternhelper.func_181117_a().func_177956_o() + 1) - p_180620_1_.func_181014_aG().field_72448_b * (double)blockpattern$patternhelper.func_181119_e();
 
             if (flag1)
             {
@@ -150,17 +140,11 @@
 
             if (blockpattern$patternhelper.func_177669_b().func_176740_k() == EnumFacing.Axis.X)
             {
-                d7 = d2
-                     + (1.0 - p_180620_1_.func_181014_aG().field_72450_a)
-                     * (double)blockpattern$patternhelper.func_181118_d()
-                     * (double)blockpattern$patternhelper.func_177669_b().func_176746_e().func_176743_c().func_179524_a();
+                d7 = d2 + (1.0D - p_180620_1_.func_181014_aG().field_72450_a) * (double)blockpattern$patternhelper.func_181118_d() * (double)blockpattern$patternhelper.func_177669_b().func_176746_e().func_176743_c().func_179524_a();
             }
             else
             {
-                d5 = d2
-                     + (1.0 - p_180620_1_.func_181014_aG().field_72450_a)
-                     * (double)blockpattern$patternhelper.func_181118_d()
-                     * (double)blockpattern$patternhelper.func_177669_b().func_176746_e().func_176743_c().func_179524_a();
+                d5 = d2 + (1.0D - p_180620_1_.func_181014_aG().field_72450_a) * (double)blockpattern$patternhelper.func_181118_d() * (double)blockpattern$patternhelper.func_177669_b().func_176746_e().func_176743_c().func_179524_a();
             }
 
             float f = 0.0F;
@@ -193,9 +177,7 @@
             double d4 = p_180620_1_.field_70179_y;
             p_180620_1_.field_70159_w = d3 * (double)f + d4 * (double)f3;
             p_180620_1_.field_70179_y = d3 * (double)f2 + d4 * (double)f1;
-            p_180620_1_.field_70177_z = p_180620_2_
-                                        - (float)(p_180620_1_.func_181012_aH().func_176734_d().func_176736_b() * 90)
-                                        + (float)(blockpattern$patternhelper.func_177669_b().func_176736_b() * 90);
+            p_180620_1_.field_70177_z = p_180620_2_ - (float)(p_180620_1_.func_181012_aH().func_176734_d().func_176736_b() * 90) + (float)(blockpattern$patternhelper.func_177669_b().func_176736_b() * 90);
 
             if (p_180620_1_ instanceof EntityPlayerMP)
             {
@@ -217,7 +199,7 @@
     public boolean func_85188_a(Entity p_85188_1_)
     {
         int i = 16;
-        double d0 = -1.0;
+        double d0 = -1.0D;
         int j = MathHelper.func_76128_c(p_85188_1_.field_70165_t);
         int k = MathHelper.func_76128_c(p_85188_1_.field_70163_u);
         int l = MathHelper.func_76128_c(p_85188_1_.field_70161_v);
@@ -230,12 +212,12 @@
 
         for (int j2 = j - 16; j2 <= j + 16; ++j2)
         {
-            double d1 = (double)j2 + 0.5 - p_85188_1_.field_70165_t;
+            double d1 = (double)j2 + 0.5D - p_85188_1_.field_70165_t;
 
             for (int l2 = l - 16; l2 <= l + 16; ++l2)
             {
-                double d2 = (double)l2 + 0.5 - p_85188_1_.field_70161_v;
-                label296:
+                double d2 = (double)l2 + 0.5D - p_85188_1_.field_70161_v;
+                label293:
 
                 for (int j3 = this.field_85192_a.func_72940_L() - 1; j3 >= 0; --j3)
                 {
@@ -268,19 +250,18 @@
                                         int k5 = l2 + (k4 - 1) * i4 - j4 * l3;
                                         blockpos$mutableblockpos.func_181079_c(i5, j5, k5);
 
-                                        if (l4 < 0 && !this.field_85192_a.func_180495_p(blockpos$mutableblockpos).func_185904_a().func_76220_a()
-                                                || l4 >= 0 && !this.field_85192_a.func_175623_d(blockpos$mutableblockpos))
+                                        if (l4 < 0 && !this.field_85192_a.func_180495_p(blockpos$mutableblockpos).func_185904_a().func_76220_a() || l4 >= 0 && !this.field_85192_a.func_175623_d(blockpos$mutableblockpos))
                                         {
-                                            continue label296;
+                                            continue label293;
                                         }
                                     }
                                 }
                             }
 
-                            double d5 = (double)j3 + 0.5 - p_85188_1_.field_70163_u;
+                            double d5 = (double)j3 + 0.5D - p_85188_1_.field_70163_u;
                             double d7 = d1 * d1 + d5 * d5 + d2 * d2;
 
-                            if (d0 < 0.0 || d7 < d0)
+                            if (d0 < 0.0D || d7 < d0)
                             {
                                 d0 = d7;
                                 i1 = j2;
@@ -294,16 +275,16 @@
             }
         }
 
-        if (d0 < 0.0)
+        if (d0 < 0.0D)
         {
             for (int l5 = j - 16; l5 <= j + 16; ++l5)
             {
-                double d3 = (double)l5 + 0.5 - p_85188_1_.field_70165_t;
+                double d3 = (double)l5 + 0.5D - p_85188_1_.field_70165_t;
 
                 for (int j6 = l - 16; j6 <= l + 16; ++j6)
                 {
-                    double d4 = (double)j6 + 0.5 - p_85188_1_.field_70161_v;
-                    label233:
+                    double d4 = (double)j6 + 0.5D - p_85188_1_.field_70161_v;
+                    label231:
 
                     for (int i7 = this.field_85192_a.func_72940_L() - 1; i7 >= 0; --i7)
                     {
@@ -328,18 +309,17 @@
                                         int j13 = j6 + (j10 - 1) * j9;
                                         blockpos$mutableblockpos.func_181079_c(j12, i13, j13);
 
-                                        if (j11 < 0 && !this.field_85192_a.func_180495_p(blockpos$mutableblockpos).func_185904_a().func_76220_a()
-                                                || j11 >= 0 && !this.field_85192_a.func_175623_d(blockpos$mutableblockpos))
+                                        if (j11 < 0 && !this.field_85192_a.func_180495_p(blockpos$mutableblockpos).func_185904_a().func_76220_a() || j11 >= 0 && !this.field_85192_a.func_175623_d(blockpos$mutableblockpos))
                                         {
-                                            continue label233;
+                                            continue label231;
                                         }
                                     }
                                 }
 
-                                double d6 = (double)i7 + 0.5 - p_85188_1_.field_70163_u;
+                                double d6 = (double)i7 + 0.5D - p_85188_1_.field_70163_u;
                                 double d8 = d3 * d3 + d6 * d6 + d4 * d4;
 
-                                if (d0 < 0.0 || d8 < d0)
+                                if (d0 < 0.0D || d8 < d0)
                                 {
                                     d0 = d8;
                                     i1 = l5;
@@ -366,7 +346,7 @@
             i3 = -i3;
         }
 
-        if (d0 < 0.0)
+        if (d0 < 0.0D)
         {
             j1 = MathHelper.func_76125_a(j1, 70, this.field_85192_a.func_72940_L() - 10);
             k2 = j1;
@@ -381,16 +361,13 @@
                         int k10 = k2 + k8;
                         int k11 = k6 + (l7 - 1) * i3 - j7 * l6;
                         boolean flag = k8 < 0;
-                        this.field_85192_a
-                        .func_175656_a(new BlockPos(k9, k10, k11), flag ? Blocks.field_150343_Z.func_176223_P() : Blocks.field_150350_a.func_176223_P());
+                        this.field_85192_a.func_175656_a(new BlockPos(k9, k10, k11), flag ? Blocks.field_150343_Z.func_176223_P() : Blocks.field_150350_a.func_176223_P());
                     }
                 }
             }
         }
 
-        IBlockState iblockstate = Blocks.field_150427_aO
-                                  .func_176223_P()
-                                  .func_177226_a(BlockPortal.field_176550_a, l6 == 0 ? EnumFacing.Axis.Z : EnumFacing.Axis.X);
+        IBlockState iblockstate = Blocks.field_150427_aO.func_176223_P().func_177226_a(BlockPortal.field_176550_a, l6 == 0 ? EnumFacing.Axis.Z : EnumFacing.Axis.X);
 
         for (int i8 = 0; i8 < 4; ++i8)
         {
@@ -431,7 +408,7 @@
 
             while (objectiterator.hasNext())
             {
-                Teleporter.PortalPosition teleporter$portalposition = objectiterator.next();
+                Teleporter.PortalPosition teleporter$portalposition = (Teleporter.PortalPosition)objectiterator.next();
 
                 if (teleporter$portalposition == null || teleporter$portalposition.field_85087_d < i)
                 {
@@ -450,5 +427,14 @@
             super(p_i45747_2_.func_177958_n(), p_i45747_2_.func_177956_o(), p_i45747_2_.func_177952_p());
             this.field_85087_d = p_i45747_3_;
         }
+    }
+
+    @Override
+    public void placeEntity(World world, Entity entity, float yaw)
+    {
+        if (entity instanceof EntityPlayerMP)
+            func_180266_a(entity, yaw);
+        else
+            func_180620_b(entity, yaw);
     }
 }
