--- before/net/minecraft/world/WorldEntitySpawner.java
+++ after/net/minecraft/world/WorldEntitySpawner.java
@@ -81,15 +81,17 @@
                         && (enumcreaturetype.getPeacefulCreature() || spawnHostileMobs)
                         && (!enumcreaturetype.getAnimal() || spawnOnSetTickRate))
                 {
-                    int k4 = worldServerIn.countEntities(enumcreaturetype.getCreatureClass());
+                    int k4 = worldServerIn.countEntities(enumcreaturetype, true);
                     int l4 = enumcreaturetype.getMaxNumberOfCreature() * i / MOB_COUNT_DIV;
 
                     if (k4 <= l4)
                     {
+                        java.util.ArrayList<ChunkPos> shuffled = com.google.common.collect.Lists.newArrayList(this.eligibleChunksForSpawning);
+                        java.util.Collections.shuffle(shuffled);
                         BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
                         label134:
 
-                        for (ChunkPos chunkpos1 : this.eligibleChunksForSpawning)
+                        for (ChunkPos chunkpos1 : shuffled)
                         {
                             BlockPos blockpos = getRandomChunkPosition(worldServerIn, chunkpos1.x, chunkpos1.z);
                             int k1 = blockpos.getX();
@@ -141,7 +143,7 @@
 
                                                 try
                                                 {
-                                                    entityliving = biome$spawnlistentry.entityClass.getConstructor(World.class).newInstance(worldServerIn);
+                                                    entityliving = biome$spawnlistentry.newInstance(worldServerIn);
                                                 }
                                                 catch (Exception exception)
                                                 {
@@ -151,8 +153,10 @@
 
                                                 entityliving.setLocationAndAngles(f, i3, f1, worldServerIn.rand.nextFloat() * 360.0F, 0.0F);
 
-                                                if (entityliving.getCanSpawnHere() && entityliving.isNotColliding())
+                                                net.minecraftforge.fml.common.eventhandler.Event.Result canSpawn = net.minecraftforge.event.ForgeEventFactory.canEntitySpawn(entityliving, worldServerIn, f, i3, f1, false);
+                                                if (canSpawn == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW || (canSpawn == net.minecraftforge.fml.common.eventhandler.Event.Result.DEFAULT && (entityliving.getCanSpawnHere() && entityliving.isNotColliding())))
                                                 {
+                                                    if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(entityliving, worldServerIn, f, i3, f1))
                                                     ientitylivingdata = entityliving.onInitialSpawn(worldServerIn.getDifficultyForLocation(new BlockPos(entityliving)), ientitylivingdata);
 
                                                     if (entityliving.isNotColliding())
@@ -165,7 +169,7 @@
                                                         entityliving.setDead();
                                                     }
 
-                                                    if (j2 >= entityliving.getMaxSpawnedInChunk())
+                                                    if (j2 >= net.minecraftforge.event.ForgeEventFactory.getMaxSpawnPackSize(entityliving))
                                                     {
                                                         continue label134;
                                                     }
@@ -206,9 +210,13 @@
         {
             return false;
         }
+        else if (state.getMaterial().isLiquid())
+        {
+            return false;
+        }
         else
         {
-            return state.getMaterial().isLiquid() ? false : !BlockRailBase.isRailBlock(state);
+            return !BlockRailBase.isRailBlock(state);
         }
     }
 
@@ -220,6 +228,13 @@
         }
         else
         {
+            return spawnPlacementTypeIn.canSpawnAt(worldIn, pos);
+        }
+    }
+
+    public static boolean canCreatureTypeSpawnBody(EntityLiving.SpawnPlacementType spawnPlacementTypeIn, World worldIn, BlockPos pos)
+    {
+        {
             IBlockState iblockstate = worldIn.getBlockState(pos);
 
             if (spawnPlacementTypeIn == EntityLiving.SpawnPlacementType.IN_WATER)
@@ -231,8 +246,9 @@
             else
             {
                 BlockPos blockpos = pos.down();
+                IBlockState state = worldIn.getBlockState(blockpos);
 
-                if (!worldIn.getBlockState(blockpos).isTopSolid())
+                if (!state.getBlock().canCreatureSpawn(state, worldIn, blockpos, spawnPlacementTypeIn))
                 {
                     return false;
                 }
@@ -276,7 +292,7 @@
 
                             try
                             {
-                                entityliving = biome$spawnlistentry.entityClass.getConstructor(World.class).newInstance(worldIn);
+                                entityliving = biome$spawnlistentry.newInstance(worldIn);
                             }
                             catch (Exception exception)
                             {
@@ -284,6 +300,7 @@
                                 continue;
                             }
 
+                            if (net.minecraftforge.event.ForgeEventFactory.canEntitySpawn(entityliving, worldIn, j + 0.5f, (float) blockpos.getY(), k +0.5f, false) == net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) continue;
                             entityliving.setLocationAndAngles(j + 0.5F, blockpos.getY(), k + 0.5F, randomIn.nextFloat() * 360.0F, 0.0F);
                             worldIn.spawnEntity(entityliving);
                             ientitylivingdata = entityliving.onInitialSpawn(worldIn.getDifficultyForLocation(new BlockPos(entityliving)), ientitylivingdata);
