--- before/net/minecraft/world/WorldProviderHell.java
+++ after/net/minecraft/world/WorldProviderHell.java
@@ -21,7 +21,7 @@
     @SideOnly(Side.CLIENT)
     public Vec3d func_76562_b(float p_76562_1_, float p_76562_2_)
     {
-        return new Vec3d((double)0.2F, (double)0.03F, (double)0.03F);
+        return new Vec3d(0.20000000298023224D, 0.029999999329447746D, 0.029999999329447746D);
     }
 
     protected void func_76556_a()
