--- before/net/minecraft/world/WorldProviderSurface.java
+++ after/net/minecraft/world/WorldProviderSurface.java
@@ -2,15 +2,13 @@
 
 public class WorldProviderSurface extends WorldProvider
 {
-    @Override
     public DimensionType getDimensionType()
     {
         return DimensionType.OVERWORLD;
     }
 
-    @Override
     public boolean canDropChunk(int x, int z)
     {
-        return !this.world.isSpawnChunk(x, z);
+        return !this.world.isSpawnChunk(x, z) || !this.world.provider.getDimensionType().shouldLoadSpawn();
     }
 }
