--- before/net/minecraft/client/tutorial/CraftPlanksStep.java
+++ after/net/minecraft/client/tutorial/CraftPlanksStep.java
@@ -16,8 +16,8 @@
 @SideOnly(Side.CLIENT)
 public class CraftPlanksStep implements ITutorialStep
 {
-    private static final ITextComponent field_193286_a = new TextComponentTranslation("tutorial.craft_planks.title");
-    private static final ITextComponent field_193287_b = new TextComponentTranslation("tutorial.craft_planks.description");
+    private static final ITextComponent field_193286_a = new TextComponentTranslation("tutorial.craft_planks.title", new Object[0]);
+    private static final ITextComponent field_193287_b = new TextComponentTranslation("tutorial.craft_planks.description", new Object[0]);
     private final Tutorial field_193288_c;
     private TutorialToast field_193289_d;
     private int field_193290_e;
