--- before/net/minecraft/client/tutorial/PunchTreeStep.java
+++ after/net/minecraft/client/tutorial/PunchTreeStep.java
@@ -21,8 +21,8 @@
 public class PunchTreeStep implements ITutorialStep
 {
     private static final Set<Block> field_193274_a = Sets.newHashSet(Blocks.field_150364_r, Blocks.field_150363_s);
-    private static final ITextComponent field_193275_b = new TextComponentTranslation("tutorial.punch_tree.title");
-    private static final ITextComponent field_193276_c = new TextComponentTranslation("tutorial.punch_tree.description", Tutorial.func_193291_a("attack"));
+    private static final ITextComponent field_193275_b = new TextComponentTranslation("tutorial.punch_tree.title", new Object[0]);
+    private static final ITextComponent field_193276_c = new TextComponentTranslation("tutorial.punch_tree.description", new Object[] {Tutorial.func_193291_a("attack")});
     private final Tutorial field_193277_d;
     private TutorialToast field_193278_e;
     private int field_193279_f;
