--- before/net/minecraft/block/BlockBanner.java
+++ after/net/minecraft/block/BlockBanner.java
@@ -3,6 +3,7 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.properties.PropertyInteger;
 import net.minecraft.block.state.BlockFaceShape;
@@ -99,16 +100,9 @@
     @Override
     public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune)
     {
-        ItemStack itemstack = this.getTileDataItemStack(worldIn, pos);
-
-        if (itemstack.isEmpty())
         {
             super.dropBlockAsItemWithChance(worldIn, pos, state, chance, fortune);
         }
-        else
-        {
-            spawnAsEntity(worldIn, pos, itemstack);
-        }
     }
 
     @Override
@@ -138,6 +132,23 @@
     public BlockFaceShape getBlockFaceShape(IBlockAccess worldIn, IBlockState state, BlockPos pos, EnumFacing face)
     {
         return BlockFaceShape.UNDEFINED;
+    }
+
+    @Override
+    public void getDrops(net.minecraft.util.NonNullList<ItemStack> drops, IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
+        TileEntity te = world.getTileEntity(pos);
+
+        if (te instanceof TileEntityBanner)
+        {
+            TileEntityBanner tileentitybanner = (TileEntityBanner)te;
+            ItemStack itemstack = tileentitybanner.getItem();
+            drops.add(itemstack);
+        }
+        else
+        {
+            drops.add(new ItemStack(Items.BANNER, 1, 0));
+        }
     }
 
     public static class BlockBannerHanging extends BlockBanner
