--- before/net/minecraft/block/BlockCactus.java
+++ after/net/minecraft/block/BlockCactus.java
@@ -2,6 +2,7 @@
 
 import java.util.Random;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyInteger;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
@@ -19,7 +20,7 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class BlockCactus extends Block
+public class BlockCactus extends Block implements net.minecraftforge.common.IPlantable
 {
     public static final PropertyInteger field_176587_a = PropertyInteger.func_177719_a("age", 0, 15);
     protected static final AxisAlignedBB field_185593_b = new AxisAlignedBB(0.0625D, 0.0D, 0.0625D, 0.9375D, 0.9375D, 0.9375D);
@@ -28,13 +29,14 @@
     protected BlockCactus()
     {
         super(Material.field_151570_A);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176587_a, 0));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176587_a, Integer.valueOf(0)));
         this.func_149675_a(true);
         this.func_149647_a(CreativeTabs.field_78031_c);
     }
 
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
     {
+        if (!p_180650_1_.func_175697_a(p_180650_2_, 1)) return; // Forge: prevent growing cactus from loading unloaded chunks with block update
         BlockPos blockpos = p_180650_2_.func_177984_a();
 
         if (p_180650_1_.func_175623_d(blockpos))
@@ -43,22 +45,27 @@
 
             for (i = 1; p_180650_1_.func_180495_p(p_180650_2_.func_177979_c(i)).func_177230_c() == this; ++i)
             {
+                ;
             }
 
             if (i < 3)
             {
-                int j = p_180650_3_.func_177229_b(field_176587_a);
+                int j = ((Integer)p_180650_3_.func_177229_b(field_176587_a)).intValue();
 
+                if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_180650_1_, blockpos, p_180650_3_, true))
+                {
                 if (j == 15)
                 {
                     p_180650_1_.func_175656_a(blockpos, this.func_176223_P());
-                    IBlockState iblockstate = p_180650_3_.func_177226_a(field_176587_a, 0);
+                    IBlockState iblockstate = p_180650_3_.func_177226_a(field_176587_a, Integer.valueOf(0));
                     p_180650_1_.func_180501_a(p_180650_2_, iblockstate, 4);
                     iblockstate.func_189546_a(p_180650_1_, blockpos, this, p_180650_2_);
                 }
                 else
                 {
-                    p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_.func_177226_a(field_176587_a, j + 1), 4);
+                    p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_.func_177226_a(field_176587_a, Integer.valueOf(j + 1)), 4);
+                }
+                net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_180650_1_, p_180650_2_, p_180650_3_, p_180650_1_.func_180495_p(p_180650_2_));
                 }
             }
         }
@@ -110,8 +117,8 @@
             }
         }
 
-        Block block = p_176586_1_.func_180495_p(p_176586_2_.func_177977_b()).func_177230_c();
-        return block == Blocks.field_150434_aF || block == Blocks.field_150354_m && !p_176586_1_.func_180495_p(p_176586_2_.func_177984_a()).func_185904_a().func_76224_d();
+        IBlockState state = p_176586_1_.func_180495_p(p_176586_2_.func_177977_b());
+        return state.func_177230_c().canSustainPlant(state, p_176586_1_, p_176586_2_.func_177977_b(), EnumFacing.UP, this) && !p_176586_1_.func_180495_p(p_176586_2_.func_177984_a()).func_185904_a().func_76224_d();
     }
 
     public void func_180634_a(World p_180634_1_, BlockPos p_180634_2_, IBlockState p_180634_3_, Entity p_180634_4_)
@@ -119,25 +126,37 @@
         p_180634_4_.func_70097_a(DamageSource.field_76367_g, 1.0F);
     }
 
+    public IBlockState func_176203_a(int p_176203_1_)
+    {
+        return this.func_176223_P().func_177226_a(field_176587_a, Integer.valueOf(p_176203_1_));
+    }
+
     @SideOnly(Side.CLIENT)
     public BlockRenderLayer func_180664_k()
     {
         return BlockRenderLayer.CUTOUT;
     }
 
-    public IBlockState func_176203_a(int p_176203_1_)
-    {
-        return this.func_176223_P().func_177226_a(field_176587_a, p_176203_1_);
-    }
-
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176587_a);
+        return ((Integer)p_176201_1_.func_177229_b(field_176587_a)).intValue();
+    }
+
+    @Override
+    public net.minecraftforge.common.EnumPlantType getPlantType(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return net.minecraftforge.common.EnumPlantType.Desert;
+    }
+
+    @Override
+    public IBlockState getPlant(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return func_176223_P();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176587_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176587_a});
     }
 
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
