--- before/net/minecraft/block/BlockChorusPlant.java
+++ after/net/minecraft/block/BlockChorusPlant.java
@@ -5,6 +5,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
@@ -36,7 +37,7 @@
     {
         super(Material.field_151585_k, MapColor.field_151678_z);
         this.func_149647_a(CreativeTabs.field_78031_c);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_185609_a, false).func_177226_a(field_185610_b, false).func_177226_a(field_185611_c, false).func_177226_a(field_185612_d, false).func_177226_a(field_185613_e, false).func_177226_a(field_185614_f, false));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_185609_a, Boolean.valueOf(false)).func_177226_a(field_185610_b, Boolean.valueOf(false)).func_177226_a(field_185611_c, Boolean.valueOf(false)).func_177226_a(field_185612_d, Boolean.valueOf(false)).func_177226_a(field_185613_e, Boolean.valueOf(false)).func_177226_a(field_185614_f, Boolean.valueOf(false)));
     }
 
     public IBlockState func_176221_a(IBlockState p_176221_1_, IBlockAccess p_176221_2_, BlockPos p_176221_3_)
@@ -47,19 +48,19 @@
         Block block3 = p_176221_2_.func_180495_p(p_176221_3_.func_177974_f()).func_177230_c();
         Block block4 = p_176221_2_.func_180495_p(p_176221_3_.func_177968_d()).func_177230_c();
         Block block5 = p_176221_2_.func_180495_p(p_176221_3_.func_177976_e()).func_177230_c();
-        return p_176221_1_.func_177226_a(field_185614_f, block == this || block == Blocks.field_185766_cS || block == Blocks.field_150377_bs).func_177226_a(field_185613_e, block1 == this || block1 == Blocks.field_185766_cS).func_177226_a(field_185609_a, block2 == this || block2 == Blocks.field_185766_cS).func_177226_a(field_185610_b, block3 == this || block3 == Blocks.field_185766_cS).func_177226_a(field_185611_c, block4 == this || block4 == Blocks.field_185766_cS).func_177226_a(field_185612_d, block5 == this || block5 == Blocks.field_185766_cS);
+        return p_176221_1_.func_177226_a(field_185614_f, Boolean.valueOf(block == this || block == Blocks.field_185766_cS || block == Blocks.field_150377_bs)).func_177226_a(field_185613_e, Boolean.valueOf(block1 == this || block1 == Blocks.field_185766_cS)).func_177226_a(field_185609_a, Boolean.valueOf(block2 == this || block2 == Blocks.field_185766_cS)).func_177226_a(field_185610_b, Boolean.valueOf(block3 == this || block3 == Blocks.field_185766_cS)).func_177226_a(field_185611_c, Boolean.valueOf(block4 == this || block4 == Blocks.field_185766_cS)).func_177226_a(field_185612_d, Boolean.valueOf(block5 == this || block5 == Blocks.field_185766_cS));
     }
 
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
         p_185496_1_ = p_185496_1_.func_185899_b(p_185496_2_, p_185496_3_);
         float f = 0.1875F;
-        float f1 = p_185496_1_.func_177229_b(field_185612_d) ? 0.0F : 0.1875F;
-        float f2 = p_185496_1_.func_177229_b(field_185614_f) ? 0.0F : 0.1875F;
-        float f3 = p_185496_1_.func_177229_b(field_185609_a) ? 0.0F : 0.1875F;
-        float f4 = p_185496_1_.func_177229_b(field_185610_b) ? 1.0F : 0.8125F;
-        float f5 = p_185496_1_.func_177229_b(field_185613_e) ? 1.0F : 0.8125F;
-        float f6 = p_185496_1_.func_177229_b(field_185611_c) ? 1.0F : 0.8125F;
+        float f1 = ((Boolean)p_185496_1_.func_177229_b(field_185612_d)).booleanValue() ? 0.0F : 0.1875F;
+        float f2 = ((Boolean)p_185496_1_.func_177229_b(field_185614_f)).booleanValue() ? 0.0F : 0.1875F;
+        float f3 = ((Boolean)p_185496_1_.func_177229_b(field_185609_a)).booleanValue() ? 0.0F : 0.1875F;
+        float f4 = ((Boolean)p_185496_1_.func_177229_b(field_185610_b)).booleanValue() ? 1.0F : 0.8125F;
+        float f5 = ((Boolean)p_185496_1_.func_177229_b(field_185613_e)).booleanValue() ? 1.0F : 0.8125F;
+        float f6 = ((Boolean)p_185496_1_.func_177229_b(field_185611_c)).booleanValue() ? 1.0F : 0.8125F;
         return new AxisAlignedBB((double)f1, (double)f2, (double)f3, (double)f4, (double)f5, (double)f6);
     }
 
@@ -74,32 +75,32 @@
         float f1 = 0.8125F;
         func_185492_a(p_185477_3_, p_185477_4_, p_185477_5_, new AxisAlignedBB(0.1875D, 0.1875D, 0.1875D, 0.8125D, 0.8125D, 0.8125D));
 
-        if (p_185477_1_.func_177229_b(field_185612_d))
+        if (((Boolean)p_185477_1_.func_177229_b(field_185612_d)).booleanValue())
         {
             func_185492_a(p_185477_3_, p_185477_4_, p_185477_5_, new AxisAlignedBB(0.0D, 0.1875D, 0.1875D, 0.1875D, 0.8125D, 0.8125D));
         }
 
-        if (p_185477_1_.func_177229_b(field_185610_b))
+        if (((Boolean)p_185477_1_.func_177229_b(field_185610_b)).booleanValue())
         {
             func_185492_a(p_185477_3_, p_185477_4_, p_185477_5_, new AxisAlignedBB(0.8125D, 0.1875D, 0.1875D, 1.0D, 0.8125D, 0.8125D));
         }
 
-        if (p_185477_1_.func_177229_b(field_185613_e))
+        if (((Boolean)p_185477_1_.func_177229_b(field_185613_e)).booleanValue())
         {
             func_185492_a(p_185477_3_, p_185477_4_, p_185477_5_, new AxisAlignedBB(0.1875D, 0.8125D, 0.1875D, 0.8125D, 1.0D, 0.8125D));
         }
 
-        if (p_185477_1_.func_177229_b(field_185614_f))
+        if (((Boolean)p_185477_1_.func_177229_b(field_185614_f)).booleanValue())
         {
             func_185492_a(p_185477_3_, p_185477_4_, p_185477_5_, new AxisAlignedBB(0.1875D, 0.0D, 0.1875D, 0.8125D, 0.1875D, 0.8125D));
         }
 
-        if (p_185477_1_.func_177229_b(field_185609_a))
+        if (((Boolean)p_185477_1_.func_177229_b(field_185609_a)).booleanValue())
         {
             func_185492_a(p_185477_3_, p_185477_4_, p_185477_5_, new AxisAlignedBB(0.1875D, 0.1875D, 0.0D, 0.8125D, 0.8125D, 0.1875D));
         }
 
-        if (p_185477_1_.func_177229_b(field_185611_c))
+        if (((Boolean)p_185477_1_.func_177229_b(field_185611_c)).booleanValue())
         {
             func_185492_a(p_185477_3_, p_185477_4_, p_185477_5_, new AxisAlignedBB(0.1875D, 0.1875D, 0.8125D, 0.8125D, 0.8125D, 1.0D));
         }
@@ -183,7 +184,7 @@
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_185609_a, field_185610_b, field_185611_c, field_185612_d, field_185613_e, field_185614_f);
+        return new BlockStateContainer(this, new IProperty[] {field_185609_a, field_185610_b, field_185611_c, field_185612_d, field_185613_e, field_185614_f});
     }
 
     public boolean func_176205_b(IBlockAccess p_176205_1_, BlockPos p_176205_2_)
