--- before/net/minecraft/block/BlockWorkbench.java
+++ after/net/minecraft/block/BlockWorkbench.java
@@ -62,7 +62,7 @@
 
             public ITextComponent func_145748_c_()
             {
-                return new TextComponentTranslation(Blocks.field_150462_ai.func_149739_a() + ".name");
+                return new TextComponentTranslation(Blocks.field_150462_ai.func_149739_a() + ".name", new Object[0]);
             }
 
             public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
