--- before/net/minecraft/block/BlockPressurePlateWeighted.java
+++ after/net/minecraft/block/BlockPressurePlateWeighted.java
@@ -2,6 +2,7 @@
 
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyInteger;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
@@ -26,7 +27,7 @@
     protected BlockPressurePlateWeighted(Material p_i46380_1_, int p_i46380_2_, MapColor p_i46380_3_)
     {
         super(p_i46380_1_, p_i46380_3_);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176579_a, 0));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176579_a, Integer.valueOf(0)));
         this.field_150068_a = p_i46380_2_;
     }
 
@@ -57,12 +58,12 @@
 
     protected int func_176576_e(IBlockState p_176576_1_)
     {
-        return p_176576_1_.func_177229_b(field_176579_a);
+        return ((Integer)p_176576_1_.func_177229_b(field_176579_a)).intValue();
     }
 
     protected IBlockState func_176575_a(IBlockState p_176575_1_, int p_176575_2_)
     {
-        return p_176575_1_.func_177226_a(field_176579_a, p_176575_2_);
+        return p_176575_1_.func_177226_a(field_176579_a, Integer.valueOf(p_176575_2_));
     }
 
     public int func_149738_a(World p_149738_1_)
@@ -72,16 +73,16 @@
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176579_a, p_176203_1_);
+        return this.func_176223_P().func_177226_a(field_176579_a, Integer.valueOf(p_176203_1_));
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176579_a);
+        return ((Integer)p_176201_1_.func_177229_b(field_176579_a)).intValue();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176579_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176579_a});
     }
 }
