--- before/net/minecraft/block/BlockLiquid.java
+++ after/net/minecraft/block/BlockLiquid.java
@@ -3,11 +3,13 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyInteger;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.Entity;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.init.Blocks;
 import net.minecraft.init.Items;
 import net.minecraft.init.SoundEvents;
@@ -33,24 +35,21 @@
     protected BlockLiquid(Material p_i45413_1_)
     {
         super(p_i45413_1_);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176367_b, 0));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176367_b, Integer.valueOf(0)));
         this.func_149675_a(true);
     }
 
-    @Override
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
         return field_185505_j;
     }
 
     @Nullable
-    @Override
     public AxisAlignedBB func_180646_a(IBlockState p_180646_1_, IBlockAccess p_180646_2_, BlockPos p_180646_3_)
     {
         return field_185506_k;
     }
 
-    @Override
     public boolean func_176205_b(IBlockAccess p_176205_1_, BlockPos p_176205_2_)
     {
         return this.field_149764_J != Material.field_151587_i;
@@ -68,7 +67,7 @@
 
     protected int func_189542_i(IBlockState p_189542_1_)
     {
-        return p_189542_1_.func_185904_a() == this.field_149764_J ? p_189542_1_.func_177229_b(field_176367_b) : -1;
+        return p_189542_1_.func_185904_a() == this.field_149764_J ? ((Integer)p_189542_1_.func_177229_b(field_176367_b)).intValue() : -1;
     }
 
     protected int func_189545_x(IBlockState p_189545_1_)
@@ -77,22 +76,19 @@
         return i >= 8 ? 0 : i;
     }
 
-    @Override
     public boolean func_149686_d(IBlockState p_149686_1_)
     {
         return false;
     }
 
-    @Override
     public boolean func_149662_c(IBlockState p_149662_1_)
     {
         return false;
     }
 
-    @Override
     public boolean func_176209_a(IBlockState p_176209_1_, boolean p_176209_2_)
     {
-        return p_176209_2_ && p_176209_1_.func_177229_b(field_176367_b) == 0;
+        return p_176209_2_ && ((Integer)p_176209_1_.func_177229_b(field_176367_b)).intValue() == 0;
     }
 
     private boolean func_176212_b(IBlockAccess p_176212_1_, BlockPos p_176212_2_, EnumFacing p_176212_3_)
@@ -121,7 +117,6 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_176225_a(IBlockState p_176225_1_, IBlockAccess p_176225_2_, BlockPos p_176225_3_, EnumFacing p_176225_4_)
     {
         if (p_176225_2_.func_180495_p(p_176225_3_.func_177972_a(p_176225_4_)).func_185904_a() == this.field_149764_J)
@@ -134,38 +129,16 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
-    public boolean func_176364_g(IBlockAccess p_176364_1_, BlockPos p_176364_2_)
-    {
-        for (int i = -1; i <= 1; ++i)
-        {
-            for (int j = -1; j <= 1; ++j)
-            {
-                IBlockState iblockstate = p_176364_1_.func_180495_p(p_176364_2_.func_177982_a(i, 0, j));
-
-                if (iblockstate.func_185904_a() != this.field_149764_J && !iblockstate.func_185913_b())
-                {
-                    return true;
-                }
-            }
-        }
-
-        return false;
-    }
-
-    @Override
     public EnumBlockRenderType func_149645_b(IBlockState p_149645_1_)
     {
         return EnumBlockRenderType.LIQUID;
     }
 
-    @Override
     public Item func_180660_a(IBlockState p_180660_1_, Random p_180660_2_, int p_180660_3_)
     {
         return Items.field_190931_a;
     }
 
-    @Override
     public int func_149745_a(Random p_149745_1_)
     {
         return 0;
@@ -173,9 +146,9 @@
 
     protected Vec3d func_189543_a(IBlockAccess p_189543_1_, BlockPos p_189543_2_, IBlockState p_189543_3_)
     {
-        double d0 = 0.0;
-        double d1 = 0.0;
-        double d2 = 0.0;
+        double d0 = 0.0D;
+        double d1 = 0.0D;
+        double d2 = 0.0D;
         int i = this.func_189545_x(p_189543_3_);
         BlockPos.PooledMutableBlockPos blockpos$pooledmutableblockpos = BlockPos.PooledMutableBlockPos.func_185346_s();
 
@@ -210,16 +183,15 @@
 
         Vec3d vec3d = new Vec3d(d0, d1, d2);
 
-        if (p_189543_3_.func_177229_b(field_176367_b) >= 8)
+        if (((Integer)p_189543_3_.func_177229_b(field_176367_b)).intValue() >= 8)
         {
             for (EnumFacing enumfacing1 : EnumFacing.Plane.HORIZONTAL)
             {
                 blockpos$pooledmutableblockpos.func_189533_g(p_189543_2_).func_189536_c(enumfacing1);
 
-                if (this.func_176212_b(p_189543_1_, blockpos$pooledmutableblockpos, enumfacing1)
-                        || this.func_176212_b(p_189543_1_, blockpos$pooledmutableblockpos.func_177984_a(), enumfacing1))
+                if (this.func_176212_b(p_189543_1_, blockpos$pooledmutableblockpos, enumfacing1) || this.func_176212_b(p_189543_1_, blockpos$pooledmutableblockpos.func_177984_a(), enumfacing1))
                 {
-                    vec3d = vec3d.func_72432_b().func_72441_c(0.0, -6.0, 0.0);
+                    vec3d = vec3d.func_72432_b().func_72441_c(0.0D, -6.0D, 0.0D);
                     break;
                 }
             }
@@ -229,13 +201,11 @@
         return vec3d.func_72432_b();
     }
 
-    @Override
     public Vec3d func_176197_a(World p_176197_1_, BlockPos p_176197_2_, Entity p_176197_3_, Vec3d p_176197_4_)
     {
         return p_176197_4_.func_178787_e(this.func_189543_a(p_176197_1_, p_176197_2_, p_176197_1_.func_180495_p(p_176197_2_)));
     }
 
-    @Override
     public int func_149738_a(World p_149738_1_)
     {
         if (this.field_149764_J == Material.field_151586_h)
@@ -253,27 +223,91 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
+    public boolean func_176364_g(IBlockAccess p_176364_1_, BlockPos p_176364_2_)
+    {
+        for (int i = -1; i <= 1; ++i)
+        {
+            for (int j = -1; j <= 1; ++j)
+            {
+                IBlockState iblockstate = p_176364_1_.func_180495_p(p_176364_2_.func_177982_a(i, 0, j));
+
+                if (iblockstate.func_185904_a() != this.field_149764_J && !iblockstate.func_185913_b())
+                {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    public void func_176213_c(World p_176213_1_, BlockPos p_176213_2_, IBlockState p_176213_3_)
+    {
+        this.func_176365_e(p_176213_1_, p_176213_2_, p_176213_3_);
+    }
+
+    public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
+    {
+        this.func_176365_e(p_189540_2_, p_189540_3_, p_189540_1_);
+    }
+
+    @SideOnly(Side.CLIENT)
     public int func_185484_c(IBlockState p_185484_1_, IBlockAccess p_185484_2_, BlockPos p_185484_3_)
     {
         int i = p_185484_2_.func_175626_b(p_185484_3_, 0);
         int j = p_185484_2_.func_175626_b(p_185484_3_.func_177984_a(), 0);
-        int k = i & 0xFF;
-        int l = j & 0xFF;
-        int i1 = i >> 16 & 0xFF;
-        int j1 = j >> 16 & 0xFF;
+        int k = i & 255;
+        int l = j & 255;
+        int i1 = i >> 16 & 255;
+        int j1 = j >> 16 & 255;
         return (k > l ? k : l) | (i1 > j1 ? i1 : j1) << 16;
     }
 
+    public boolean func_176365_e(World p_176365_1_, BlockPos p_176365_2_, IBlockState p_176365_3_)
+    {
+        if (this.field_149764_J == Material.field_151587_i)
+        {
+            boolean flag = false;
+
+            for (EnumFacing enumfacing : EnumFacing.values())
+            {
+                if (enumfacing != EnumFacing.DOWN && p_176365_1_.func_180495_p(p_176365_2_.func_177972_a(enumfacing)).func_185904_a() == Material.field_151586_h)
+                {
+                    flag = true;
+                    break;
+                }
+            }
+
+            if (flag)
+            {
+                Integer integer = (Integer)p_176365_3_.func_177229_b(field_176367_b);
+
+                if (integer.intValue() == 0)
+                {
+                    p_176365_1_.func_175656_a(p_176365_2_, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(p_176365_1_, p_176365_2_, p_176365_2_, Blocks.field_150343_Z.func_176223_P()));
+                    this.func_180688_d(p_176365_1_, p_176365_2_);
+                    return true;
+                }
+
+                if (integer.intValue() <= 4)
+                {
+                    p_176365_1_.func_175656_a(p_176365_2_, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(p_176365_1_, p_176365_2_, p_176365_2_, Blocks.field_150347_e.func_176223_P()));
+                    this.func_180688_d(p_176365_1_, p_176365_2_);
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
     @SideOnly(Side.CLIENT)
-    @Override
     public BlockRenderLayer func_180664_k()
     {
         return this.field_149764_J == Material.field_151586_h ? BlockRenderLayer.TRANSLUCENT : BlockRenderLayer.SOLID;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_180655_c(IBlockState p_180655_1_, World p_180655_2_, BlockPos p_180655_3_, Random p_180655_4_)
     {
         double d0 = (double)p_180655_3_.func_177958_n();
@@ -282,72 +316,35 @@
 
         if (this.field_149764_J == Material.field_151586_h)
         {
-            int i = p_180655_1_.func_177229_b(field_176367_b);
+            int i = ((Integer)p_180655_1_.func_177229_b(field_176367_b)).intValue();
 
             if (i > 0 && i < 8)
             {
                 if (p_180655_4_.nextInt(64) == 0)
                 {
-                    p_180655_2_.func_184134_a(
-                        d0 + 0.5,
-                        d1 + 0.5,
-                        d2 + 0.5,
-                        SoundEvents.field_187917_gq,
-                        SoundCategory.BLOCKS,
-                        p_180655_4_.nextFloat() * 0.25F + 0.75F,
-                        p_180655_4_.nextFloat() + 0.5F,
-                        false
-                    );
+                    p_180655_2_.func_184134_a(d0 + 0.5D, d1 + 0.5D, d2 + 0.5D, SoundEvents.field_187917_gq, SoundCategory.BLOCKS, p_180655_4_.nextFloat() * 0.25F + 0.75F, p_180655_4_.nextFloat() + 0.5F, false);
                 }
             }
             else if (p_180655_4_.nextInt(10) == 0)
             {
-                p_180655_2_.func_175688_a(
-                    EnumParticleTypes.SUSPENDED,
-                    d0 + (double)p_180655_4_.nextFloat(),
-                    d1 + (double)p_180655_4_.nextFloat(),
-                    d2 + (double)p_180655_4_.nextFloat(),
-                    0.0,
-                    0.0,
-                    0.0
-                );
+                p_180655_2_.func_175688_a(EnumParticleTypes.SUSPENDED, d0 + (double)p_180655_4_.nextFloat(), d1 + (double)p_180655_4_.nextFloat(), d2 + (double)p_180655_4_.nextFloat(), 0.0D, 0.0D, 0.0D);
             }
         }
 
-        if (this.field_149764_J == Material.field_151587_i
-                && p_180655_2_.func_180495_p(p_180655_3_.func_177984_a()).func_185904_a() == Material.field_151579_a
-                && !p_180655_2_.func_180495_p(p_180655_3_.func_177984_a()).func_185914_p())
+        if (this.field_149764_J == Material.field_151587_i && p_180655_2_.func_180495_p(p_180655_3_.func_177984_a()).func_185904_a() == Material.field_151579_a && !p_180655_2_.func_180495_p(p_180655_3_.func_177984_a()).func_185914_p())
         {
             if (p_180655_4_.nextInt(100) == 0)
             {
                 double d8 = d0 + (double)p_180655_4_.nextFloat();
                 double d4 = d1 + p_180655_1_.func_185900_c(p_180655_2_, p_180655_3_).field_72337_e;
                 double d6 = d2 + (double)p_180655_4_.nextFloat();
-                p_180655_2_.func_175688_a(EnumParticleTypes.LAVA, d8, d4, d6, 0.0, 0.0, 0.0);
-                p_180655_2_.func_184134_a(
-                    d8,
-                    d4,
-                    d6,
-                    SoundEvents.field_187662_cZ,
-                    SoundCategory.BLOCKS,
-                    0.2F + p_180655_4_.nextFloat() * 0.2F,
-                    0.9F + p_180655_4_.nextFloat() * 0.15F,
-                    false
-                );
+                p_180655_2_.func_175688_a(EnumParticleTypes.LAVA, d8, d4, d6, 0.0D, 0.0D, 0.0D);
+                p_180655_2_.func_184134_a(d8, d4, d6, SoundEvents.field_187662_cZ, SoundCategory.BLOCKS, 0.2F + p_180655_4_.nextFloat() * 0.2F, 0.9F + p_180655_4_.nextFloat() * 0.15F, false);
             }
 
             if (p_180655_4_.nextInt(200) == 0)
             {
-                p_180655_2_.func_184134_a(
-                    d0,
-                    d1,
-                    d2,
-                    SoundEvents.field_187656_cX,
-                    SoundCategory.BLOCKS,
-                    0.2F + p_180655_4_.nextFloat() * 0.2F,
-                    0.9F + p_180655_4_.nextFloat() * 0.15F,
-                    false
-                );
+                p_180655_2_.func_184134_a(d0, d1, d2, SoundEvents.field_187656_cX, SoundCategory.BLOCKS, 0.2F + p_180655_4_.nextFloat() * 0.2F, 0.9F + p_180655_4_.nextFloat() * 0.15F, false);
             }
         }
 
@@ -358,16 +355,16 @@
             if (!material.func_76230_c() && !material.func_76224_d())
             {
                 double d3 = d0 + (double)p_180655_4_.nextFloat();
-                double d5 = d1 - 1.05;
+                double d5 = d1 - 1.05D;
                 double d7 = d2 + (double)p_180655_4_.nextFloat();
 
                 if (this.field_149764_J == Material.field_151586_h)
                 {
-                    p_180655_2_.func_175688_a(EnumParticleTypes.DRIP_WATER, d3, d5, d7, 0.0, 0.0, 0.0);
+                    p_180655_2_.func_175688_a(EnumParticleTypes.DRIP_WATER, d3, d5, d7, 0.0D, 0.0D, 0.0D);
                 }
                 else
                 {
-                    p_180655_2_.func_175688_a(EnumParticleTypes.DRIP_LAVA, d3, d5, d7, 0.0, 0.0, 0.0);
+                    p_180655_2_.func_175688_a(EnumParticleTypes.DRIP_LAVA, d3, d5, d7, 0.0D, 0.0D, 0.0D);
                 }
             }
         }
@@ -377,59 +374,7 @@
     public static float func_189544_a(IBlockAccess p_189544_0_, BlockPos p_189544_1_, Material p_189544_2_, IBlockState p_189544_3_)
     {
         Vec3d vec3d = func_176361_a(p_189544_2_).func_189543_a(p_189544_0_, p_189544_1_, p_189544_3_);
-        return vec3d.field_72450_a == 0.0 && vec3d.field_72449_c == 0.0
-               ? -1000.0F
-               : (float)MathHelper.func_181159_b(vec3d.field_72449_c, vec3d.field_72450_a) - (float)(Math.PI / 2);
-    }
-
-    @Override
-    public void func_176213_c(World p_176213_1_, BlockPos p_176213_2_, IBlockState p_176213_3_)
-    {
-        this.func_176365_e(p_176213_1_, p_176213_2_, p_176213_3_);
-    }
-
-    @Override
-    public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
-    {
-        this.func_176365_e(p_189540_2_, p_189540_3_, p_189540_1_);
-    }
-
-    public boolean func_176365_e(World p_176365_1_, BlockPos p_176365_2_, IBlockState p_176365_3_)
-    {
-        if (this.field_149764_J == Material.field_151587_i)
-        {
-            boolean flag = false;
-
-            for (EnumFacing enumfacing : EnumFacing.values())
-            {
-                if (enumfacing != EnumFacing.DOWN && p_176365_1_.func_180495_p(p_176365_2_.func_177972_a(enumfacing)).func_185904_a() == Material.field_151586_h)
-                {
-                    flag = true;
-                    break;
-                }
-            }
-
-            if (flag)
-            {
-                Integer integer = p_176365_3_.func_177229_b(field_176367_b);
-
-                if (integer == 0)
-                {
-                    p_176365_1_.func_175656_a(p_176365_2_, Blocks.field_150343_Z.func_176223_P());
-                    this.func_180688_d(p_176365_1_, p_176365_2_);
-                    return true;
-                }
-
-                if (integer <= 4)
-                {
-                    p_176365_1_.func_175656_a(p_176365_2_, Blocks.field_150347_e.func_176223_P());
-                    this.func_180688_d(p_176365_1_, p_176365_2_);
-                    return true;
-                }
-            }
-        }
-
-        return false;
+        return vec3d.field_72450_a == 0.0D && vec3d.field_72449_c == 0.0D ? -1000.0F : (float)MathHelper.func_181159_b(vec3d.field_72449_c, vec3d.field_72450_a) - ((float)Math.PI / 2F);
     }
 
     protected void func_180688_d(World p_180688_1_, BlockPos p_180688_2_)
@@ -437,37 +382,27 @@
         double d0 = (double)p_180688_2_.func_177958_n();
         double d1 = (double)p_180688_2_.func_177956_o();
         double d2 = (double)p_180688_2_.func_177952_p();
-        p_180688_1_.func_184133_a(
-            null,
-            p_180688_2_,
-            SoundEvents.field_187659_cY,
-            SoundCategory.BLOCKS,
-            0.5F,
-            2.6F + (p_180688_1_.field_73012_v.nextFloat() - p_180688_1_.field_73012_v.nextFloat()) * 0.8F
-        );
+        p_180688_1_.func_184133_a((EntityPlayer)null, p_180688_2_, SoundEvents.field_187659_cY, SoundCategory.BLOCKS, 0.5F, 2.6F + (p_180688_1_.field_73012_v.nextFloat() - p_180688_1_.field_73012_v.nextFloat()) * 0.8F);
 
         for (int i = 0; i < 8; ++i)
         {
-            p_180688_1_.func_175688_a(EnumParticleTypes.SMOKE_LARGE, d0 + Math.random(), d1 + 1.2, d2 + Math.random(), 0.0, 0.0, 0.0);
+            p_180688_1_.func_175688_a(EnumParticleTypes.SMOKE_LARGE, d0 + Math.random(), d1 + 1.2D, d2 + Math.random(), 0.0D, 0.0D, 0.0D);
         }
     }
 
-    @Override
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176367_b, p_176203_1_);
+        return this.func_176223_P().func_177226_a(field_176367_b, Integer.valueOf(p_176203_1_));
     }
 
-    @Override
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176367_b);
+        return ((Integer)p_176201_1_.func_177229_b(field_176367_b)).intValue();
     }
 
-    @Override
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176367_b);
+        return new BlockStateContainer(this, new IProperty[] {field_176367_b});
     }
 
     public static BlockDynamicLiquid func_176361_a(Material p_176361_0_)
@@ -504,7 +439,7 @@
 
     public static float func_190973_f(IBlockState p_190973_0_, IBlockAccess p_190973_1_, BlockPos p_190973_2_)
     {
-        int i = p_190973_0_.func_177229_b(field_176367_b);
+        int i = ((Integer)p_190973_0_.func_177229_b(field_176367_b)).intValue();
         return (i & 7) == 0 && p_190973_1_.func_180495_p(p_190973_2_.func_177984_a()).func_185904_a() == Material.field_151586_h ? 1.0F : 1.0F - func_149801_b(i);
     }
 
@@ -514,8 +449,38 @@
     }
 
     @Override
+    public float getBlockLiquidHeight(World world, BlockPos pos, IBlockState state, Material material)
+    {
+        return BlockLiquid.func_190973_f(state, world, pos);
+    }
+
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
     {
         return BlockFaceShape.UNDEFINED;
+    }
+
+    @Override
+    @SideOnly (Side.CLIENT)
+    public Vec3d getFogColor(World world, BlockPos pos, IBlockState state, Entity entity, Vec3d originalColor, float partialTicks)
+    {
+        Vec3d viewport = net.minecraft.client.renderer.ActiveRenderInfo.func_178806_a(entity, partialTicks);
+
+        if (state.func_185904_a().func_76224_d())
+        {
+            float height = 0.0F;
+            if (state.func_177230_c() instanceof BlockLiquid)
+            {
+                height = func_149801_b(state.func_177229_b(field_176367_b)) - 0.11111111F;
+            }
+            float f1 = (float) (pos.func_177956_o() + 1) - height;
+            if (viewport.field_72448_b > (double)f1)
+            {
+                BlockPos upPos = pos.func_177984_a();
+                IBlockState upState = world.func_180495_p(upPos);
+                return upState.func_177230_c().getFogColor(world, upPos, upState, entity, originalColor, partialTicks);
+            }
+        }
+
+        return super.getFogColor(world, pos, state, entity, originalColor, partialTicks);
     }
 }
