--- before/net/minecraft/block/BlockAnvil.java
+++ after/net/minecraft/block/BlockAnvil.java
@@ -1,6 +1,7 @@
 package net.minecraft.block;
 
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.properties.PropertyInteger;
 import net.minecraft.block.state.BlockFaceShape;
@@ -42,7 +43,7 @@
     protected BlockAnvil()
     {
         super(Material.field_151574_g);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176506_a, EnumFacing.NORTH).func_177226_a(field_176505_b, 0));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176506_a, EnumFacing.NORTH).func_177226_a(field_176505_b, Integer.valueOf(0)));
         this.func_149713_g(0);
         this.func_149647_a(CreativeTabs.field_78031_c);
     }
@@ -68,9 +69,9 @@
 
         try
         {
-            return super.func_180642_a(p_180642_1_, p_180642_2_, p_180642_3_, p_180642_4_, p_180642_5_, p_180642_6_, p_180642_7_, p_180642_8_).func_177226_a(field_176506_a, enumfacing).func_177226_a(field_176505_b, p_180642_7_ >> 2);
+            return super.func_180642_a(p_180642_1_, p_180642_2_, p_180642_3_, p_180642_4_, p_180642_5_, p_180642_6_, p_180642_7_, p_180642_8_).func_177226_a(field_176506_a, enumfacing).func_177226_a(field_176505_b, Integer.valueOf(p_180642_7_ >> 2));
         }
-        catch (IllegalArgumentException illegalargumentexception)
+        catch (IllegalArgumentException var11)
         {
             if (!p_180642_1_.field_72995_K)
             {
@@ -78,11 +79,11 @@
 
                 if (p_180642_8_ instanceof EntityPlayer)
                 {
-                    p_180642_8_.func_145747_a(new TextComponentTranslation("Invalid damage property. Please pick in [0, 1, 2]"));
+                    p_180642_8_.func_145747_a(new TextComponentTranslation("Invalid damage property. Please pick in [0, 1, 2]", new Object[0]));
                 }
             }
 
-            return super.func_180642_a(p_180642_1_, p_180642_2_, p_180642_3_, p_180642_4_, p_180642_5_, p_180642_6_, 0, p_180642_8_).func_177226_a(field_176506_a, enumfacing).func_177226_a(field_176505_b, 0);
+            return super.func_180642_a(p_180642_1_, p_180642_2_, p_180642_3_, p_180642_4_, p_180642_5_, p_180642_6_, 0, p_180642_8_).func_177226_a(field_176506_a, enumfacing).func_177226_a(field_176505_b, Integer.valueOf(0));
         }
     }
 
@@ -98,12 +99,12 @@
 
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        return p_180651_1_.func_177229_b(field_176505_b);
+        return ((Integer)p_180651_1_.func_177229_b(field_176505_b)).intValue();
     }
 
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
-        EnumFacing enumfacing = p_185496_1_.func_177229_b(field_176506_a);
+        EnumFacing enumfacing = (EnumFacing)p_185496_1_.func_177229_b(field_176506_a);
         return enumfacing.func_176740_k() == EnumFacing.Axis.X ? field_185760_c : field_185761_d;
     }
 
@@ -137,24 +138,25 @@
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176506_a, EnumFacing.func_176731_b(p_176203_1_ & 3)).func_177226_a(field_176505_b, (p_176203_1_ & 15) >> 2);
+        return this.func_176223_P().func_177226_a(field_176506_a, EnumFacing.func_176731_b(p_176203_1_ & 3)).func_177226_a(field_176505_b, Integer.valueOf((p_176203_1_ & 15) >> 2));
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_176506_a).func_176736_b();
-        return i | p_176201_1_.func_177229_b(field_176505_b) << 2;
+        i = i | ((EnumFacing)p_176201_1_.func_177229_b(field_176506_a)).func_176736_b();
+        i = i | ((Integer)p_176201_1_.func_177229_b(field_176505_b)).intValue() << 2;
+        return i;
     }
 
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177230_c() != this ? p_185499_1_ : p_185499_1_.func_177226_a(field_176506_a, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_176506_a)));
+        return p_185499_1_.func_177230_c() != this ? p_185499_1_ : p_185499_1_.func_177226_a(field_176506_a, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_176506_a)));
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176506_a, field_176505_b);
+        return new BlockStateContainer(this, new IProperty[] {field_176506_a, field_176505_b});
     }
 
     public static class Anvil implements IInteractionObject
@@ -180,7 +182,7 @@
 
             public ITextComponent func_145748_c_()
             {
-                return new TextComponentTranslation(Blocks.field_150467_bQ.func_149739_a() + ".name");
+                return new TextComponentTranslation(Blocks.field_150467_bQ.func_149739_a() + ".name", new Object[0]);
             }
 
             public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
