--- before/net/minecraft/block/BlockMycelium.java
+++ after/net/minecraft/block/BlockMycelium.java
@@ -40,7 +40,8 @@
     {
         if (!worldIn.isRemote)
         {
-            if (worldIn.getLightFromNeighbors(pos.up()) < 4 && worldIn.getBlockState(pos.up()).getLightOpacity() > 2)
+            if (!worldIn.isAreaLoaded(pos, 2)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light and spreading
+            if (worldIn.getLightFromNeighbors(pos.up()) < 4 && worldIn.getBlockState(pos.up()).getLightOpacity(worldIn, pos.up()) > 2)
             {
                 worldIn.setBlockState(pos, Blocks.DIRT.getDefaultState().withProperty(BlockDirt.VARIANT, BlockDirt.DirtType.DIRT));
             }
@@ -57,7 +58,7 @@
                         if (iblockstate.getBlock() == Blocks.DIRT
                                 && iblockstate.getValue(BlockDirt.VARIANT) == BlockDirt.DirtType.DIRT
                                 && worldIn.getLightFromNeighbors(blockpos.up()) >= 4
-                                && iblockstate1.getLightOpacity() <= 2)
+                                && iblockstate1.getLightOpacity(worldIn, blockpos.up()) <= 2)
                         {
                             worldIn.setBlockState(blockpos, this.getDefaultState());
                         }
