--- before/net/minecraft/block/BlockDynamicLiquid.java
+++ after/net/minecraft/block/BlockDynamicLiquid.java
@@ -21,15 +21,13 @@
 
     private void func_180690_f(World p_180690_1_, BlockPos p_180690_2_, IBlockState p_180690_3_)
     {
-        p_180690_1_.func_180501_a(
-            p_180690_2_, func_176363_b(this.field_149764_J).func_176223_P().func_177226_a(field_176367_b, p_180690_3_.func_177229_b(field_176367_b)), 2
-        );
+        p_180690_1_.func_180501_a(p_180690_2_, func_176363_b(this.field_149764_J).func_176223_P().func_177226_a(field_176367_b, p_180690_3_.func_177229_b(field_176367_b)), 2);
     }
 
-    @Override
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
     {
-        int i = p_180650_3_.func_177229_b(field_176367_b);
+        if (!p_180650_1_.func_175697_a(p_180650_2_, this.func_185698_b(p_180650_1_))) return; // Forge: avoid loading unloaded chunks
+        int i = ((Integer)p_180650_3_.func_177229_b(field_176367_b)).intValue();
         int j = 1;
 
         if (this.field_149764_J == Material.field_151587_i && !p_180650_1_.field_73011_w.func_177500_n())
@@ -70,7 +68,7 @@
                 }
             }
 
-            if (this.field_149815_a >= 2 && this.field_149764_J == Material.field_151586_h)
+            if (this.field_149815_a >= 2 && net.minecraftforge.event.ForgeEventFactory.canCreateFluidSource(p_180650_1_, p_180650_2_, p_180650_3_, this.field_149764_J == Material.field_151586_h))
             {
                 IBlockState iblockstate = p_180650_1_.func_180495_p(p_180650_2_.func_177977_b());
 
@@ -78,7 +76,7 @@
                 {
                     i1 = 0;
                 }
-                else if (iblockstate.func_185904_a() == this.field_149764_J && iblockstate.func_177229_b(field_176367_b) == 0)
+                else if (iblockstate.func_185904_a() == this.field_149764_J && ((Integer)iblockstate.func_177229_b(field_176367_b)).intValue() == 0)
                 {
                     i1 = 0;
                 }
@@ -103,7 +101,7 @@
                 }
                 else
                 {
-                    p_180650_3_ = p_180650_3_.func_177226_a(field_176367_b, i1);
+                    p_180650_3_ = p_180650_3_.func_177226_a(field_176367_b, Integer.valueOf(i1));
                     p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_, 2);
                     p_180650_1_.func_175684_a(p_180650_2_, this, k);
                     p_180650_1_.func_175685_c(p_180650_2_, this, false);
@@ -119,10 +117,9 @@
 
         if (this.func_176373_h(p_180650_1_, p_180650_2_.func_177977_b(), iblockstate1))
         {
-            if (this.field_149764_J == Material.field_151587_i
-                    && p_180650_1_.func_180495_p(p_180650_2_.func_177977_b()).func_185904_a() == Material.field_151586_h)
+            if (this.field_149764_J == Material.field_151587_i && p_180650_1_.func_180495_p(p_180650_2_.func_177977_b()).func_185904_a() == Material.field_151586_h)
             {
-                p_180650_1_.func_175656_a(p_180650_2_.func_177977_b(), Blocks.field_150348_b.func_176223_P());
+                p_180650_1_.func_175656_a(p_180650_2_.func_177977_b(), net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(p_180650_1_, p_180650_2_.func_177977_b(), p_180650_2_, Blocks.field_150348_b.func_176223_P()));
                 this.func_180688_d(p_180650_1_, p_180650_2_.func_177977_b());
                 return;
             }
@@ -170,11 +167,12 @@
                 }
                 else
                 {
+                    if (p_176375_3_.func_177230_c() != Blocks.field_150431_aC) //Forge: Vanilla has a 'bug' where snowballs don't drop like every other block. So special case because ewww...
                     p_176375_3_.func_177230_c().func_176226_b(p_176375_1_, p_176375_2_, p_176375_3_, 0);
                 }
             }
 
-            p_176375_1_.func_180501_a(p_176375_2_, this.func_176223_P().func_177226_a(field_176367_b, p_176375_4_), 3);
+            p_176375_1_.func_180501_a(p_176375_2_, this.func_176223_P().func_177226_a(field_176367_b, Integer.valueOf(p_176375_4_)), 3);
         }
     }
 
@@ -189,10 +187,9 @@
                 BlockPos blockpos = p_176374_2_.func_177972_a(enumfacing);
                 IBlockState iblockstate = p_176374_1_.func_180495_p(blockpos);
 
-                if (!this.func_176372_g(p_176374_1_, blockpos, iblockstate)
-                        && (iblockstate.func_185904_a() != this.field_149764_J || iblockstate.func_177229_b(field_176367_b) > 0))
+                if (!this.func_176372_g(p_176374_1_, blockpos, iblockstate) && (iblockstate.func_185904_a() != this.field_149764_J || ((Integer)iblockstate.func_177229_b(field_176367_b)).intValue() > 0))
                 {
-                    if (!this.func_176372_g(p_176374_1_, blockpos.func_177977_b(), iblockstate))
+                    if (!this.func_176372_g(p_176374_1_, blockpos.func_177977_b(), p_176374_1_.func_180495_p(blockpos.func_177977_b())))
                     {
                         return p_176374_3_;
                     }
@@ -221,15 +218,14 @@
     private Set<EnumFacing> func_176376_e(World p_176376_1_, BlockPos p_176376_2_)
     {
         int i = 1000;
-        Set<EnumFacing> set = EnumSet.noneOf(EnumFacing.class);
+        Set<EnumFacing> set = EnumSet.<EnumFacing>noneOf(EnumFacing.class);
 
         for (EnumFacing enumfacing : EnumFacing.Plane.HORIZONTAL)
         {
             BlockPos blockpos = p_176376_2_.func_177972_a(enumfacing);
             IBlockState iblockstate = p_176376_1_.func_180495_p(blockpos);
 
-            if (!this.func_176372_g(p_176376_1_, blockpos, iblockstate)
-                    && (iblockstate.func_185904_a() != this.field_149764_J || iblockstate.func_177229_b(field_176367_b) > 0))
+            if (!this.func_176372_g(p_176376_1_, blockpos, iblockstate) && (iblockstate.func_185904_a() != this.field_149764_J || ((Integer)iblockstate.func_177229_b(field_176367_b)).intValue() > 0))
             {
                 int j;
 
@@ -260,16 +256,17 @@
 
     private boolean func_176372_g(World p_176372_1_, BlockPos p_176372_2_, IBlockState p_176372_3_)
     {
-        Block block = p_176372_1_.func_180495_p(p_176372_2_).func_177230_c();
+        Block block = p_176372_3_.func_177230_c(); //Forge: state must be valid for position
+        Material mat = p_176372_3_.func_185904_a();
 
-        if (block instanceof BlockDoor || block == Blocks.field_150472_an || block == Blocks.field_150468_ap || block == Blocks.field_150436_aH)
+        if (!(block instanceof BlockDoor) && block != Blocks.field_150472_an && block != Blocks.field_150468_ap && block != Blocks.field_150436_aH)
+        {
+            return mat != Material.field_151567_E && mat != Material.field_189963_J ? mat.func_76230_c() : true;
+        }
+        else
         {
             return true;
         }
-        else
-        {
-            return block.field_149764_J != Material.field_151567_E && block.field_149764_J != Material.field_189963_J ? block.field_149764_J.func_76230_c() : true;
-        }
     }
 
     protected int func_176371_a(World p_176371_1_, BlockPos p_176371_2_, int p_176371_3_)
@@ -284,7 +281,7 @@
         {
             if (i == 0)
             {
-                this.field_149815_a++;
+                ++this.field_149815_a;
             }
 
             if (i >= 8)
@@ -302,7 +299,6 @@
         return material != this.field_149764_J && material != Material.field_151587_i && !this.func_176372_g(p_176373_1_, p_176373_2_, p_176373_3_);
     }
 
-    @Override
     public void func_176213_c(World p_176213_1_, BlockPos p_176213_2_, IBlockState p_176213_3_)
     {
         if (!this.func_176365_e(p_176213_1_, p_176213_2_, p_176213_3_))
