--- before/net/minecraft/block/BlockConcretePowder.java
+++ after/net/minecraft/block/BlockConcretePowder.java
@@ -2,6 +2,7 @@
 
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
@@ -17,7 +18,7 @@
 
 public class BlockConcretePowder extends BlockFalling
 {
-    public static final PropertyEnum<EnumDyeColor> field_192426_a = PropertyEnum.func_177709_a("color", EnumDyeColor.class);
+    public static final PropertyEnum<EnumDyeColor> field_192426_a = PropertyEnum.<EnumDyeColor>func_177709_a("color", EnumDyeColor.class);
 
     public BlockConcretePowder()
     {
@@ -78,7 +79,7 @@
 
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        return p_180651_1_.func_177229_b(field_192426_a).func_176765_a();
+        return ((EnumDyeColor)p_180651_1_.func_177229_b(field_192426_a)).func_176765_a();
     }
 
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
@@ -91,7 +92,7 @@
 
     public MapColor func_180659_g(IBlockState p_180659_1_, IBlockAccess p_180659_2_, BlockPos p_180659_3_)
     {
-        return MapColor.func_193558_a(p_180659_1_.func_177229_b(field_192426_a));
+        return MapColor.func_193558_a((EnumDyeColor)p_180659_1_.func_177229_b(field_192426_a));
     }
 
     public IBlockState func_176203_a(int p_176203_1_)
@@ -101,11 +102,11 @@
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_192426_a).func_176765_a();
+        return ((EnumDyeColor)p_176201_1_.func_177229_b(field_192426_a)).func_176765_a();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_192426_a);
+        return new BlockStateContainer(this, new IProperty[] {field_192426_a});
     }
 }
