--- before/net/minecraft/block/BlockRailDetector.java
+++ after/net/minecraft/block/BlockRailDetector.java
@@ -183,6 +183,8 @@
     {
         if (blockState.getValue(POWERED))
         {
+            List<EntityMinecart> carts = this.findMinecarts(worldIn, pos, EntityMinecart.class);
+            if (!carts.isEmpty() && carts.getFirst().getComparatorLevel() > -1) return carts.getFirst().getComparatorLevel();
             List<EntityMinecartCommandBlock> list = this.findMinecarts(worldIn, pos, EntityMinecartCommandBlock.class);
 
             if (!list.isEmpty())
