--- before/net/minecraft/block/BlockDoor.java
+++ after/net/minecraft/block/BlockDoor.java
@@ -4,6 +4,7 @@
 import net.minecraft.block.material.EnumPushReaction;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.properties.PropertyEnum;
@@ -33,9 +34,9 @@
 {
     public static final PropertyDirection field_176520_a = BlockHorizontal.field_185512_D;
     public static final PropertyBool field_176519_b = PropertyBool.func_177716_a("open");
-    public static final PropertyEnum<BlockDoor.EnumHingePosition> field_176521_M = PropertyEnum.func_177709_a("hinge", BlockDoor.EnumHingePosition.class);
+    public static final PropertyEnum<BlockDoor.EnumHingePosition> field_176521_M = PropertyEnum.<BlockDoor.EnumHingePosition>func_177709_a("hinge", BlockDoor.EnumHingePosition.class);
     public static final PropertyBool field_176522_N = PropertyBool.func_177716_a("powered");
-    public static final PropertyEnum<BlockDoor.EnumDoorHalf> field_176523_O = PropertyEnum.func_177709_a("half", BlockDoor.EnumDoorHalf.class);
+    public static final PropertyEnum<BlockDoor.EnumDoorHalf> field_176523_O = PropertyEnum.<BlockDoor.EnumDoorHalf>func_177709_a("half", BlockDoor.EnumDoorHalf.class);
     protected static final AxisAlignedBB field_185658_f = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 1.0D, 1.0D, 0.1875D);
     protected static final AxisAlignedBB field_185659_g = new AxisAlignedBB(0.0D, 0.0D, 0.8125D, 1.0D, 1.0D, 1.0D);
     protected static final AxisAlignedBB field_185656_B = new AxisAlignedBB(0.8125D, 0.0D, 0.0D, 1.0D, 1.0D, 1.0D);
@@ -44,14 +45,14 @@
     protected BlockDoor(Material p_i45402_1_)
     {
         super(p_i45402_1_);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176520_a, EnumFacing.NORTH).func_177226_a(field_176519_b, false).func_177226_a(field_176521_M, BlockDoor.EnumHingePosition.LEFT).func_177226_a(field_176522_N, false).func_177226_a(field_176523_O, BlockDoor.EnumDoorHalf.LOWER));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176520_a, EnumFacing.NORTH).func_177226_a(field_176519_b, Boolean.valueOf(false)).func_177226_a(field_176521_M, BlockDoor.EnumHingePosition.LEFT).func_177226_a(field_176522_N, Boolean.valueOf(false)).func_177226_a(field_176523_O, BlockDoor.EnumDoorHalf.LOWER));
     }
 
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
         p_185496_1_ = p_185496_1_.func_185899_b(p_185496_2_, p_185496_3_);
-        EnumFacing enumfacing = p_185496_1_.func_177229_b(field_176520_a);
-        boolean flag = !p_185496_1_.func_177229_b(field_176519_b);
+        EnumFacing enumfacing = (EnumFacing)p_185496_1_.func_177229_b(field_176520_a);
+        boolean flag = !((Boolean)p_185496_1_.func_177229_b(field_176519_b)).booleanValue();
         boolean flag1 = p_185496_1_.func_177229_b(field_176521_M) == BlockDoor.EnumHingePosition.RIGHT;
 
         switch (enumfacing)
@@ -150,7 +151,7 @@
                 p_180639_3_ = iblockstate.func_177231_a(field_176519_b);
                 p_180639_1_.func_180501_a(blockpos, p_180639_3_, 10);
                 p_180639_1_.func_175704_b(blockpos, p_180639_2_);
-                p_180639_1_.func_180498_a(p_180639_4_, p_180639_3_.func_177229_b(field_176519_b) ? this.func_185655_g() : this.func_185654_e(), p_180639_2_, 0);
+                p_180639_1_.func_180498_a(p_180639_4_, ((Boolean)p_180639_3_.func_177229_b(field_176519_b)).booleanValue() ? this.func_185655_g() : this.func_185654_e(), p_180639_2_, 0);
                 return true;
             }
         }
@@ -165,9 +166,9 @@
             BlockPos blockpos = iblockstate.func_177229_b(field_176523_O) == BlockDoor.EnumDoorHalf.LOWER ? p_176512_2_ : p_176512_2_.func_177977_b();
             IBlockState iblockstate1 = p_176512_2_ == blockpos ? iblockstate : p_176512_1_.func_180495_p(blockpos);
 
-            if (iblockstate1.func_177230_c() == this && iblockstate1.func_177229_b(field_176519_b) != p_176512_3_)
+            if (iblockstate1.func_177230_c() == this && ((Boolean)iblockstate1.func_177229_b(field_176519_b)).booleanValue() != p_176512_3_)
             {
-                p_176512_1_.func_180501_a(blockpos, iblockstate1.func_177226_a(field_176519_b, p_176512_3_), 10);
+                p_176512_1_.func_180501_a(blockpos, iblockstate1.func_177226_a(field_176519_b, Boolean.valueOf(p_176512_3_)), 10);
                 p_176512_1_.func_175704_b(blockpos, p_176512_2_);
                 p_176512_1_.func_180498_a((EntityPlayer)null, p_176512_3_ ? this.func_185655_g() : this.func_185654_e(), p_176512_2_, 0);
             }
@@ -202,7 +203,7 @@
                 flag1 = true;
             }
 
-            if (!p_189540_2_.func_180495_p(p_189540_3_.func_177977_b()).func_185896_q())
+            if (!p_189540_2_.func_180495_p(p_189540_3_.func_177977_b()).isSideSolid(p_189540_2_,  p_189540_3_.func_177977_b(), EnumFacing.UP))
             {
                 p_189540_2_.func_175698_g(p_189540_3_);
                 flag1 = true;
@@ -224,13 +225,13 @@
             {
                 boolean flag = p_189540_2_.func_175640_z(p_189540_3_) || p_189540_2_.func_175640_z(blockpos1);
 
-                if (p_189540_4_ != this && (flag || p_189540_4_.func_176223_P().func_185897_m()) && flag != iblockstate1.func_177229_b(field_176522_N))
+                if (p_189540_4_ != this && (flag || p_189540_4_.func_176223_P().func_185897_m()) && flag != ((Boolean)iblockstate1.func_177229_b(field_176522_N)).booleanValue())
                 {
-                    p_189540_2_.func_180501_a(blockpos1, iblockstate1.func_177226_a(field_176522_N, flag), 2);
+                    p_189540_2_.func_180501_a(blockpos1, iblockstate1.func_177226_a(field_176522_N, Boolean.valueOf(flag)), 2);
 
-                    if (flag != p_189540_1_.func_177229_b(field_176519_b))
+                    if (flag != ((Boolean)p_189540_1_.func_177229_b(field_176519_b)).booleanValue())
                     {
-                        p_189540_2_.func_180501_a(p_189540_3_, p_189540_1_.func_177226_a(field_176519_b, flag), 2);
+                        p_189540_2_.func_180501_a(p_189540_3_, p_189540_1_.func_177226_a(field_176519_b, Boolean.valueOf(flag)), 2);
                         p_189540_2_.func_175704_b(p_189540_3_, p_189540_3_);
                         p_189540_2_.func_180498_a((EntityPlayer)null, flag ? this.func_185655_g() : this.func_185654_e(), p_189540_3_, 0);
                     }
@@ -246,13 +247,14 @@
 
     public boolean func_176196_c(World p_176196_1_, BlockPos p_176196_2_)
     {
-        if (p_176196_2_.func_177956_o() >= 255)
+        if (p_176196_2_.func_177956_o() >= p_176196_1_.func_72800_K() - 1)
         {
             return false;
         }
         else
         {
-            return p_176196_1_.func_180495_p(p_176196_2_.func_177977_b()).func_185896_q() && super.func_176196_c(p_176196_1_, p_176196_2_) && super.func_176196_c(p_176196_1_, p_176196_2_.func_177984_a());
+            IBlockState state = p_176196_1_.func_180495_p(p_176196_2_.func_177977_b());
+            return (state.func_185896_q() || state.func_193401_d(p_176196_1_, p_176196_2_.func_177977_b(), EnumFacing.UP) == BlockFaceShape.SOLID) && super.func_176196_c(p_176196_1_, p_176196_2_) && super.func_176196_c(p_176196_1_, p_176196_2_.func_177984_a());
         }
     }
 
@@ -363,17 +365,17 @@
 
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177229_b(field_176523_O) != BlockDoor.EnumDoorHalf.LOWER ? p_185499_1_ : p_185499_1_.func_177226_a(field_176520_a, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_176520_a)));
+        return p_185499_1_.func_177229_b(field_176523_O) != BlockDoor.EnumDoorHalf.LOWER ? p_185499_1_ : p_185499_1_.func_177226_a(field_176520_a, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_176520_a)));
     }
 
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_2_ == Mirror.NONE ? p_185471_1_ : p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_176520_a))).func_177231_a(field_176521_M);
+        return p_185471_2_ == Mirror.NONE ? p_185471_1_ : p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_176520_a))).func_177231_a(field_176521_M);
     }
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return (p_176203_1_ & 8) > 0 ? this.func_176223_P().func_177226_a(field_176523_O, BlockDoor.EnumDoorHalf.UPPER).func_177226_a(field_176521_M, (p_176203_1_ & 1) > 0 ? BlockDoor.EnumHingePosition.RIGHT : BlockDoor.EnumHingePosition.LEFT).func_177226_a(field_176522_N, (p_176203_1_ & 2) > 0) : this.func_176223_P().func_177226_a(field_176523_O, BlockDoor.EnumDoorHalf.LOWER).func_177226_a(field_176520_a, EnumFacing.func_176731_b(p_176203_1_ & 3).func_176735_f()).func_177226_a(field_176519_b, (p_176203_1_ & 4) > 0);
+        return (p_176203_1_ & 8) > 0 ? this.func_176223_P().func_177226_a(field_176523_O, BlockDoor.EnumDoorHalf.UPPER).func_177226_a(field_176521_M, (p_176203_1_ & 1) > 0 ? BlockDoor.EnumHingePosition.RIGHT : BlockDoor.EnumHingePosition.LEFT).func_177226_a(field_176522_N, Boolean.valueOf((p_176203_1_ & 2) > 0)) : this.func_176223_P().func_177226_a(field_176523_O, BlockDoor.EnumDoorHalf.LOWER).func_177226_a(field_176520_a, EnumFacing.func_176731_b(p_176203_1_ & 3).func_176735_f()).func_177226_a(field_176519_b, Boolean.valueOf((p_176203_1_ & 4) > 0));
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
@@ -382,23 +384,23 @@
 
         if (p_176201_1_.func_177229_b(field_176523_O) == BlockDoor.EnumDoorHalf.UPPER)
         {
-            i |= 8;
+            i = i | 8;
 
             if (p_176201_1_.func_177229_b(field_176521_M) == BlockDoor.EnumHingePosition.RIGHT)
             {
                 i |= 1;
             }
 
-            if (p_176201_1_.func_177229_b(field_176522_N))
+            if (((Boolean)p_176201_1_.func_177229_b(field_176522_N)).booleanValue())
             {
                 i |= 2;
             }
         }
         else
         {
-            i |= p_176201_1_.func_177229_b(field_176520_a).func_176746_e().func_176736_b();
+            i = i | ((EnumFacing)p_176201_1_.func_177229_b(field_176520_a)).func_176746_e().func_176736_b();
 
-            if (p_176201_1_.func_177229_b(field_176519_b))
+            if (((Boolean)p_176201_1_.func_177229_b(field_176519_b)).booleanValue())
             {
                 i |= 4;
             }
@@ -439,7 +441,7 @@
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176523_O, field_176520_a, field_176519_b, field_176521_M, field_176522_N);
+        return new BlockStateContainer(this, new IProperty[] {field_176523_O, field_176520_a, field_176519_b, field_176521_M, field_176522_N});
     }
 
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
