--- before/net/minecraft/block/BlockDropper.java
+++ after/net/minecraft/block/BlockDropper.java
@@ -46,7 +46,7 @@
             {
                 ItemStack itemstack = tileentitydispenser.getStackInSlot(i);
 
-                if (!itemstack.isEmpty())
+                if (!itemstack.isEmpty() && net.minecraftforge.items.VanillaInventoryCodeHooks.dropperInsertHook(worldIn, pos, tileentitydispenser, i, itemstack))
                 {
                     EnumFacing enumfacing = worldIn.getBlockState(pos).getValue(FACING);
                     BlockPos blockpos = pos.offset(enumfacing);
