--- before/net/minecraft/block/BlockPane.java
+++ after/net/minecraft/block/BlockPane.java
@@ -147,22 +147,10 @@
     @Override
     public IBlockState getActualState(IBlockState state, IBlockAccess worldIn, BlockPos pos)
     {
-        return state.withProperty(
-                   NORTH,
-                   this.attachesTo(worldIn, worldIn.getBlockState(pos.north()), pos.north(), EnumFacing.SOUTH)
-               )
-               .withProperty(
-                   SOUTH,
-                   this.attachesTo(worldIn, worldIn.getBlockState(pos.south()), pos.south(), EnumFacing.NORTH)
-               )
-               .withProperty(
-                   WEST,
-                   this.attachesTo(worldIn, worldIn.getBlockState(pos.west()), pos.west(), EnumFacing.EAST)
-               )
-               .withProperty(
-                   EAST,
-                   this.attachesTo(worldIn, worldIn.getBlockState(pos.east()), pos.east(), EnumFacing.WEST)
-               );
+        return state.withProperty(NORTH, canPaneConnectTo(worldIn, pos, EnumFacing.NORTH))
+               .withProperty(SOUTH, canPaneConnectTo(worldIn, pos, EnumFacing.SOUTH))
+               .withProperty(WEST,  canPaneConnectTo(worldIn, pos, EnumFacing.WEST))
+               .withProperty(EAST,  canPaneConnectTo(worldIn, pos, EnumFacing.EAST));
     }
 
     @Override
@@ -282,6 +270,24 @@
     {
         return new BlockStateContainer(this, NORTH, EAST, WEST, SOUTH);
     }
+
+    /* ======================================== FORGE START ======================================== */
+
+    @Override
+    public boolean canBeConnectedTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
+    {
+        BlockPos offset = pos.offset(facing);
+        return attachesTo(world, world.getBlockState(offset), offset, facing.getOpposite());
+    }
+
+    public boolean canPaneConnectTo(IBlockAccess world, BlockPos pos, EnumFacing dir)
+    {
+        BlockPos other = pos.offset(dir);
+        IBlockState state = world.getBlockState(other);
+        return state.getBlock().canBeConnectedTo(world, other, dir.getOpposite()) || attachesTo(world, state, other, dir.getOpposite());
+    }
+
+    /* ======================================== FORGE END ======================================== */
 
     @Override
     public BlockFaceShape getBlockFaceShape(IBlockAccess worldIn, IBlockState state, BlockPos pos, EnumFacing face)
