--- before/net/minecraft/block/BlockStoneSlab.java
+++ after/net/minecraft/block/BlockStoneSlab.java
@@ -21,7 +21,7 @@
 public abstract class BlockStoneSlab extends BlockSlab
 {
     public static final PropertyBool field_176555_b = PropertyBool.func_177716_a("seamless");
-    public static final PropertyEnum<BlockStoneSlab.EnumType> field_176556_M = PropertyEnum.func_177709_a("variant", BlockStoneSlab.EnumType.class);
+    public static final PropertyEnum<BlockStoneSlab.EnumType> field_176556_M = PropertyEnum.<BlockStoneSlab.EnumType>func_177709_a("variant", BlockStoneSlab.EnumType.class);
 
     public BlockStoneSlab()
     {
@@ -30,7 +30,7 @@
 
         if (this.func_176552_j())
         {
-            iblockstate = iblockstate.func_177226_a(field_176555_b, false);
+            iblockstate = iblockstate.func_177226_a(field_176555_b, Boolean.valueOf(false));
         }
         else
         {
@@ -48,7 +48,7 @@
 
     public ItemStack func_185473_a(World p_185473_1_, BlockPos p_185473_2_, IBlockState p_185473_3_)
     {
-        return new ItemStack(Blocks.field_150333_U, 1, p_185473_3_.func_177229_b(field_176556_M).func_176624_a());
+        return new ItemStack(Blocks.field_150333_U, 1, ((BlockStoneSlab.EnumType)p_185473_3_.func_177229_b(field_176556_M)).func_176624_a());
     }
 
     public String func_150002_b(int p_150002_1_)
@@ -83,7 +83,7 @@
 
         if (this.func_176552_j())
         {
-            iblockstate = iblockstate.func_177226_a(field_176555_b, (p_176203_1_ & 8) != 0);
+            iblockstate = iblockstate.func_177226_a(field_176555_b, Boolean.valueOf((p_176203_1_ & 8) != 0));
         }
         else
         {
@@ -96,11 +96,11 @@
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_176556_M).func_176624_a();
+        i = i | ((BlockStoneSlab.EnumType)p_176201_1_.func_177229_b(field_176556_M)).func_176624_a();
 
         if (this.func_176552_j())
         {
-            if (p_176201_1_.func_177229_b(field_176555_b))
+            if (((Boolean)p_176201_1_.func_177229_b(field_176555_b)).booleanValue())
             {
                 i |= 8;
             }
@@ -115,17 +115,17 @@
 
     protected BlockStateContainer func_180661_e()
     {
-        return this.func_176552_j() ? new BlockStateContainer(this, field_176555_b, field_176556_M) : new BlockStateContainer(this, field_176554_a, field_176556_M);
+        return this.func_176552_j() ? new BlockStateContainer(this, new IProperty[] {field_176555_b, field_176556_M}) : new BlockStateContainer(this, new IProperty[] {field_176554_a, field_176556_M});
     }
 
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        return p_180651_1_.func_177229_b(field_176556_M).func_176624_a();
+        return ((BlockStoneSlab.EnumType)p_180651_1_.func_177229_b(field_176556_M)).func_176624_a();
     }
 
     public MapColor func_180659_g(IBlockState p_180659_1_, IBlockAccess p_180659_2_, BlockPos p_180659_3_)
     {
-        return p_180659_1_.func_177229_b(field_176556_M).func_181074_c();
+        return ((BlockStoneSlab.EnumType)p_180659_1_.func_177229_b(field_176556_M)).func_181074_c();
     }
 
     public static enum EnumType implements IStringSerializable
