--- before/net/minecraft/block/BlockStaticLiquid.java
+++ after/net/minecraft/block/BlockStaticLiquid.java
@@ -54,22 +54,22 @@
                     {
                         blockpos = blockpos.add(rand.nextInt(3) - 1, 1, rand.nextInt(3) - 1);
 
-                        if (blockpos.getY() >= 0 && blockpos.getY() < 256 && !worldIn.isBlockLoaded(blockpos))
+                        if (blockpos.getY() >= 0 && blockpos.getY() < worldIn.getHeight() && !worldIn.isBlockLoaded(blockpos))
                         {
                             return;
                         }
 
-                        Block block = worldIn.getBlockState(blockpos).getBlock();
+                        IBlockState block = worldIn.getBlockState(blockpos);
 
-                        if (block.material == Material.AIR)
+                        if (block.getBlock().isAir(block, worldIn, blockpos))
                         {
                             if (this.isSurroundingBlockFlammable(worldIn, blockpos))
                             {
-                                worldIn.setBlockState(blockpos, Blocks.FIRE.getDefaultState());
+                                worldIn.setBlockState(blockpos, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(worldIn, blockpos, pos, Blocks.FIRE.getDefaultState()));
                                 return;
                             }
                         }
-                        else if (block.material.blocksMovement())
+                        else if (block.getMaterial().blocksMovement())
                         {
                             return;
                         }
@@ -88,7 +88,7 @@
 
                         if (worldIn.isAirBlock(blockpos1.up()) && this.getCanBlockBurn(worldIn, blockpos1))
                         {
-                            worldIn.setBlockState(blockpos1.up(), Blocks.FIRE.getDefaultState());
+                            worldIn.setBlockState(blockpos1.up(), net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(worldIn, blockpos1.up(), pos, Blocks.FIRE.getDefaultState()));
                         }
                     }
                 }
