--- before/net/minecraft/block/BlockRedstoneComparator.java
+++ after/net/minecraft/block/BlockRedstoneComparator.java
@@ -5,6 +5,7 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockStateContainer;
@@ -35,78 +36,56 @@
 public class BlockRedstoneComparator extends BlockRedstoneDiode implements ITileEntityProvider
 {
     public static final PropertyBool field_176464_a = PropertyBool.func_177716_a("powered");
-    public static final PropertyEnum<BlockRedstoneComparator.Mode> field_176463_b = PropertyEnum.func_177709_a("mode", BlockRedstoneComparator.Mode.class);
+    public static final PropertyEnum<BlockRedstoneComparator.Mode> field_176463_b = PropertyEnum.<BlockRedstoneComparator.Mode>func_177709_a("mode", BlockRedstoneComparator.Mode.class);
 
     public BlockRedstoneComparator(boolean p_i45399_1_)
     {
         super(p_i45399_1_);
-        this.func_180632_j(
-            this.field_176227_L
-            .func_177621_b()
-            .func_177226_a(field_185512_D, EnumFacing.NORTH)
-            .func_177226_a(field_176464_a, false)
-            .func_177226_a(field_176463_b, BlockRedstoneComparator.Mode.COMPARE)
-        );
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_185512_D, EnumFacing.NORTH).func_177226_a(field_176464_a, Boolean.valueOf(false)).func_177226_a(field_176463_b, BlockRedstoneComparator.Mode.COMPARE));
         this.field_149758_A = true;
     }
 
-    @Override
     public String func_149732_F()
     {
         return I18n.func_74838_a("item.comparator.name");
     }
 
-    @Override
     public Item func_180660_a(IBlockState p_180660_1_, Random p_180660_2_, int p_180660_3_)
     {
         return Items.field_151132_bS;
     }
 
-    @Override
     public ItemStack func_185473_a(World p_185473_1_, BlockPos p_185473_2_, IBlockState p_185473_3_)
     {
         return new ItemStack(Items.field_151132_bS);
     }
 
-    @Override
     protected int func_176403_d(IBlockState p_176403_1_)
     {
         return 2;
     }
 
-    @Override
     protected IBlockState func_180674_e(IBlockState p_180674_1_)
     {
-        Boolean obool = p_180674_1_.func_177229_b(field_176464_a);
-        BlockRedstoneComparator.Mode blockredstonecomparator$mode = p_180674_1_.func_177229_b(field_176463_b);
-        EnumFacing enumfacing = p_180674_1_.func_177229_b(field_185512_D);
-        return Blocks.field_150455_bV
-               .func_176223_P()
-               .func_177226_a(field_185512_D, enumfacing)
-               .func_177226_a(field_176464_a, obool)
-               .func_177226_a(field_176463_b, blockredstonecomparator$mode);
+        Boolean obool = (Boolean)p_180674_1_.func_177229_b(field_176464_a);
+        BlockRedstoneComparator.Mode blockredstonecomparator$mode = (BlockRedstoneComparator.Mode)p_180674_1_.func_177229_b(field_176463_b);
+        EnumFacing enumfacing = (EnumFacing)p_180674_1_.func_177229_b(field_185512_D);
+        return Blocks.field_150455_bV.func_176223_P().func_177226_a(field_185512_D, enumfacing).func_177226_a(field_176464_a, obool).func_177226_a(field_176463_b, blockredstonecomparator$mode);
     }
 
-    @Override
     protected IBlockState func_180675_k(IBlockState p_180675_1_)
     {
-        Boolean obool = p_180675_1_.func_177229_b(field_176464_a);
-        BlockRedstoneComparator.Mode blockredstonecomparator$mode = p_180675_1_.func_177229_b(field_176463_b);
-        EnumFacing enumfacing = p_180675_1_.func_177229_b(field_185512_D);
-        return Blocks.field_150441_bU
-               .func_176223_P()
-               .func_177226_a(field_185512_D, enumfacing)
-               .func_177226_a(field_176464_a, obool)
-               .func_177226_a(field_176463_b, blockredstonecomparator$mode);
+        Boolean obool = (Boolean)p_180675_1_.func_177229_b(field_176464_a);
+        BlockRedstoneComparator.Mode blockredstonecomparator$mode = (BlockRedstoneComparator.Mode)p_180675_1_.func_177229_b(field_176463_b);
+        EnumFacing enumfacing = (EnumFacing)p_180675_1_.func_177229_b(field_185512_D);
+        return Blocks.field_150441_bU.func_176223_P().func_177226_a(field_185512_D, enumfacing).func_177226_a(field_176464_a, obool).func_177226_a(field_176463_b, blockredstonecomparator$mode);
     }
 
-    @Override
     protected boolean func_176406_l(IBlockState p_176406_1_)
     {
-        return this.field_149914_a || p_176406_1_.func_177229_b(field_176464_a);
+        return this.field_149914_a || ((Boolean)p_176406_1_.func_177229_b(field_176464_a)).booleanValue();
     }
 
-    @Override
     protected int func_176408_a(IBlockAccess p_176408_1_, BlockPos p_176408_2_, IBlockState p_176408_3_)
     {
         TileEntity tileentity = p_176408_1_.func_175625_s(p_176408_2_);
@@ -115,12 +94,9 @@
 
     private int func_176460_j(World p_176460_1_, BlockPos p_176460_2_, IBlockState p_176460_3_)
     {
-        return p_176460_3_.func_177229_b(field_176463_b) == BlockRedstoneComparator.Mode.SUBTRACT
-               ? Math.max(this.func_176397_f(p_176460_1_, p_176460_2_, p_176460_3_) - this.func_176407_c(p_176460_1_, p_176460_2_, p_176460_3_), 0)
-               : this.func_176397_f(p_176460_1_, p_176460_2_, p_176460_3_);
+        return p_176460_3_.func_177229_b(field_176463_b) == BlockRedstoneComparator.Mode.SUBTRACT ? Math.max(this.func_176397_f(p_176460_1_, p_176460_2_, p_176460_3_) - this.func_176407_c(p_176460_1_, p_176460_2_, p_176460_3_), 0) : this.func_176397_f(p_176460_1_, p_176460_2_, p_176460_3_);
     }
 
-    @Override
     protected boolean func_176404_e(World p_176404_1_, BlockPos p_176404_2_, IBlockState p_176404_3_)
     {
         int i = this.func_176397_f(p_176404_1_, p_176404_2_, p_176404_3_);
@@ -136,15 +112,22 @@
         else
         {
             int j = this.func_176407_c(p_176404_1_, p_176404_2_, p_176404_3_);
-            return j == 0 ? true : i >= j;
+
+            if (j == 0)
+            {
+                return true;
+            }
+            else
+            {
+                return i >= j;
+            }
         }
     }
 
-    @Override
     protected int func_176397_f(World p_176397_1_, BlockPos p_176397_2_, IBlockState p_176397_3_)
     {
         int i = super.func_176397_f(p_176397_1_, p_176397_2_, p_176397_3_);
-        EnumFacing enumfacing = p_176397_3_.func_177229_b(field_185512_D);
+        EnumFacing enumfacing = (EnumFacing)p_176397_3_.func_177229_b(field_185512_D);
         BlockPos blockpos = p_176397_2_.func_177972_a(enumfacing);
         IBlockState iblockstate = p_176397_1_.func_180495_p(blockpos);
 
@@ -178,39 +161,17 @@
     @Nullable
     private EntityItemFrame func_176461_a(World p_176461_1_, final EnumFacing p_176461_2_, BlockPos p_176461_3_)
     {
-        List<EntityItemFrame> list = p_176461_1_.func_175647_a(
-                                         EntityItemFrame.class,
-                                         new AxisAlignedBB(
-                                             (double)p_176461_3_.func_177958_n(),
-                                             (double)p_176461_3_.func_177956_o(),
-                                             (double)p_176461_3_.func_177952_p(),
-                                             (double)(p_176461_3_.func_177958_n() + 1),
-                                             (double)(p_176461_3_.func_177956_o() + 1),
-                                             (double)(p_176461_3_.func_177952_p() + 1)
-                                         ),
-                                         new Predicate<Entity>()
+        List<EntityItemFrame> list = p_176461_1_.<EntityItemFrame>func_175647_a(EntityItemFrame.class, new AxisAlignedBB((double)p_176461_3_.func_177958_n(), (double)p_176461_3_.func_177956_o(), (double)p_176461_3_.func_177952_p(), (double)(p_176461_3_.func_177958_n() + 1), (double)(p_176461_3_.func_177956_o() + 1), (double)(p_176461_3_.func_177952_p() + 1)), new Predicate<Entity>()
         {
             public boolean apply(@Nullable Entity p_apply_1_)
             {
                 return p_apply_1_ != null && p_apply_1_.func_174811_aO() == p_176461_2_;
             }
-        }
-                                     );
-        return list.size() == 1 ? list.get(0) : null;
+        });
+        return list.size() == 1 ? (EntityItemFrame)list.get(0) : null;
     }
 
-    @Override
-    public boolean func_180639_a(
-        World p_180639_1_,
-        BlockPos p_180639_2_,
-        IBlockState p_180639_3_,
-        EntityPlayer p_180639_4_,
-        EnumHand p_180639_5_,
-        EnumFacing p_180639_6_,
-        float p_180639_7_,
-        float p_180639_8_,
-        float p_180639_9_
-    )
+    public boolean func_180639_a(World p_180639_1_, BlockPos p_180639_2_, IBlockState p_180639_3_, EntityPlayer p_180639_4_, EnumHand p_180639_5_, EnumFacing p_180639_6_, float p_180639_7_, float p_180639_8_, float p_180639_9_)
     {
         if (!p_180639_4_.field_71075_bZ.field_75099_e)
         {
@@ -227,7 +188,6 @@
         }
     }
 
-    @Override
     protected void func_176398_g(World p_176398_1_, BlockPos p_176398_2_, IBlockState p_176398_3_)
     {
         if (!p_176398_1_.func_175691_a(p_176398_2_, this))
@@ -270,36 +230,33 @@
 
             if (flag && !flag1)
             {
-                p_176462_1_.func_180501_a(p_176462_2_, p_176462_3_.func_177226_a(field_176464_a, false), 2);
+                p_176462_1_.func_180501_a(p_176462_2_, p_176462_3_.func_177226_a(field_176464_a, Boolean.valueOf(false)), 2);
             }
             else if (!flag && flag1)
             {
-                p_176462_1_.func_180501_a(p_176462_2_, p_176462_3_.func_177226_a(field_176464_a, true), 2);
+                p_176462_1_.func_180501_a(p_176462_2_, p_176462_3_.func_177226_a(field_176464_a, Boolean.valueOf(true)), 2);
             }
 
             this.func_176400_h(p_176462_1_, p_176462_2_, p_176462_3_);
         }
     }
 
-    @Override
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
     {
         if (this.field_149914_a)
         {
-            p_180650_1_.func_180501_a(p_180650_2_, this.func_180675_k(p_180650_3_).func_177226_a(field_176464_a, true), 4);
+            p_180650_1_.func_180501_a(p_180650_2_, this.func_180675_k(p_180650_3_).func_177226_a(field_176464_a, Boolean.valueOf(true)), 4);
         }
 
         this.func_176462_k(p_180650_1_, p_180650_2_, p_180650_3_);
     }
 
-    @Override
     public void func_176213_c(World p_176213_1_, BlockPos p_176213_2_, IBlockState p_176213_3_)
     {
         super.func_176213_c(p_176213_1_, p_176213_2_, p_176213_3_);
         p_176213_1_.func_175690_a(p_176213_2_, this.func_149915_a(p_176213_1_, 0));
     }
 
-    @Override
     public void func_180663_b(World p_180663_1_, BlockPos p_180663_2_, IBlockState p_180663_3_)
     {
         super.func_180663_b(p_180663_1_, p_180663_2_, p_180663_3_);
@@ -307,7 +264,6 @@
         this.func_176400_h(p_180663_1_, p_180663_2_, p_180663_3_);
     }
 
-    @Override
     public boolean func_189539_a(IBlockState p_189539_1_, World p_189539_2_, BlockPos p_189539_3_, int p_189539_4_, int p_189539_5_)
     {
         super.func_189539_a(p_189539_1_, p_189539_2_, p_189539_3_, p_189539_4_, p_189539_5_);
@@ -315,28 +271,22 @@
         return tileentity == null ? false : tileentity.func_145842_c(p_189539_4_, p_189539_5_);
     }
 
-    @Override
     public TileEntity func_149915_a(World p_149915_1_, int p_149915_2_)
     {
         return new TileEntityComparator();
     }
 
-    @Override
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P()
-               .func_177226_a(field_185512_D, EnumFacing.func_176731_b(p_176203_1_))
-               .func_177226_a(field_176464_a, (p_176203_1_ & 8) > 0)
-               .func_177226_a(field_176463_b, (p_176203_1_ & 4) > 0 ? BlockRedstoneComparator.Mode.SUBTRACT : BlockRedstoneComparator.Mode.COMPARE);
+        return this.func_176223_P().func_177226_a(field_185512_D, EnumFacing.func_176731_b(p_176203_1_)).func_177226_a(field_176464_a, Boolean.valueOf((p_176203_1_ & 8) > 0)).func_177226_a(field_176463_b, (p_176203_1_ & 4) > 0 ? BlockRedstoneComparator.Mode.SUBTRACT : BlockRedstoneComparator.Mode.COMPARE);
     }
 
-    @Override
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_185512_D).func_176736_b();
+        i = i | ((EnumFacing)p_176201_1_.func_177229_b(field_185512_D)).func_176736_b();
 
-        if (p_176201_1_.func_177229_b(field_176464_a))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176464_a)).booleanValue())
         {
             i |= 8;
         }
@@ -349,40 +299,39 @@
         return i;
     }
 
-    @Override
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177226_a(field_185512_D, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_185512_D)));
+        return p_185499_1_.func_177226_a(field_185512_D, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_185512_D)));
     }
 
-    @Override
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_185512_D)));
+        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_185512_D)));
     }
 
-    @Override
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_185512_D, field_176463_b, field_176464_a);
-    }
-
-    @Override
-    public IBlockState func_180642_a(
-        World p_180642_1_,
-        BlockPos p_180642_2_,
-        EnumFacing p_180642_3_,
-        float p_180642_4_,
-        float p_180642_5_,
-        float p_180642_6_,
-        int p_180642_7_,
-        EntityLivingBase p_180642_8_
-    )
-    {
-        return this.func_176223_P()
-               .func_177226_a(field_185512_D, p_180642_8_.func_174811_aO().func_176734_d())
-               .func_177226_a(field_176464_a, false)
-               .func_177226_a(field_176463_b, BlockRedstoneComparator.Mode.COMPARE);
+        return new BlockStateContainer(this, new IProperty[] {field_185512_D, field_176463_b, field_176464_a});
+    }
+
+    public IBlockState func_180642_a(World p_180642_1_, BlockPos p_180642_2_, EnumFacing p_180642_3_, float p_180642_4_, float p_180642_5_, float p_180642_6_, int p_180642_7_, EntityLivingBase p_180642_8_)
+    {
+        return this.func_176223_P().func_177226_a(field_185512_D, p_180642_8_.func_174811_aO().func_176734_d()).func_177226_a(field_176464_a, Boolean.valueOf(false)).func_177226_a(field_176463_b, BlockRedstoneComparator.Mode.COMPARE);
+    }
+
+    @Override
+    public void onNeighborChange(IBlockAccess world, BlockPos pos, BlockPos neighbor)
+    {
+        if (pos.func_177956_o() == neighbor.func_177956_o() && world instanceof World && !((World) world).field_72995_K)
+        {
+            func_189540_a(world.func_180495_p(pos), (World)world, pos, world.func_180495_p(neighbor).func_177230_c(), neighbor);
+        }
+    }
+
+    @Override
+    public boolean getWeakChanges(IBlockAccess world, BlockPos pos)
+    {
+        return true;
     }
 
     public static enum Mode implements IStringSerializable
@@ -397,13 +346,11 @@
             this.field_177041_c = p_i45731_3_;
         }
 
-        @Override
         public String toString()
         {
             return this.field_177041_c;
         }
 
-        @Override
         public String func_176610_l()
         {
             return this.field_177041_c;
