--- before/net/minecraft/block/BlockTrapDoor.java
+++ after/net/minecraft/block/BlockTrapDoor.java
@@ -2,6 +2,7 @@
 
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.properties.PropertyEnum;
@@ -28,7 +29,7 @@
 {
     public static final PropertyDirection field_176284_a = BlockHorizontal.field_185512_D;
     public static final PropertyBool field_176283_b = PropertyBool.func_177716_a("open");
-    public static final PropertyEnum<BlockTrapDoor.DoorHalf> field_176285_M = PropertyEnum.func_177709_a("half", BlockTrapDoor.DoorHalf.class);
+    public static final PropertyEnum<BlockTrapDoor.DoorHalf> field_176285_M = PropertyEnum.<BlockTrapDoor.DoorHalf>func_177709_a("half", BlockTrapDoor.DoorHalf.class);
     protected static final AxisAlignedBB field_185734_d = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 0.1875D, 1.0D, 1.0D);
     protected static final AxisAlignedBB field_185735_e = new AxisAlignedBB(0.8125D, 0.0D, 0.0D, 1.0D, 1.0D, 1.0D);
     protected static final AxisAlignedBB field_185736_f = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 1.0D, 1.0D, 0.1875D);
@@ -39,7 +40,7 @@
     protected BlockTrapDoor(Material p_i45434_1_)
     {
         super(p_i45434_1_);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176284_a, EnumFacing.NORTH).func_177226_a(field_176283_b, false).func_177226_a(field_176285_M, BlockTrapDoor.DoorHalf.BOTTOM));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176284_a, EnumFacing.NORTH).func_177226_a(field_176283_b, Boolean.valueOf(false)).func_177226_a(field_176285_M, BlockTrapDoor.DoorHalf.BOTTOM));
         this.func_149647_a(CreativeTabs.field_78028_d);
     }
 
@@ -47,7 +48,7 @@
     {
         AxisAlignedBB axisalignedbb;
 
-        if (p_185496_1_.func_177229_b(field_176283_b))
+        if (((Boolean)p_185496_1_.func_177229_b(field_176283_b)).booleanValue())
         {
             switch ((EnumFacing)p_185496_1_.func_177229_b(field_176284_a))
             {
@@ -89,7 +90,7 @@
 
     public boolean func_176205_b(IBlockAccess p_176205_1_, BlockPos p_176205_2_)
     {
-        return !p_176205_1_.func_180495_p(p_176205_2_).func_177229_b(field_176283_b);
+        return !((Boolean)p_176205_1_.func_180495_p(p_176205_2_).func_177229_b(field_176283_b)).booleanValue();
     }
 
     public boolean func_180639_a(World p_180639_1_, BlockPos p_180639_2_, IBlockState p_180639_3_, EntityPlayer p_180639_4_, EnumHand p_180639_5_, EnumFacing p_180639_6_, float p_180639_7_, float p_180639_8_, float p_180639_9_)
@@ -102,7 +103,7 @@
         {
             p_180639_3_ = p_180639_3_.func_177231_a(field_176283_b);
             p_180639_1_.func_180501_a(p_180639_2_, p_180639_3_, 2);
-            this.func_185731_a(p_180639_4_, p_180639_1_, p_180639_2_, p_180639_3_.func_177229_b(field_176283_b));
+            this.func_185731_a(p_180639_4_, p_180639_1_, p_180639_2_, ((Boolean)p_180639_3_.func_177229_b(field_176283_b)).booleanValue());
             return true;
         }
     }
@@ -129,11 +130,11 @@
 
             if (flag || p_189540_4_.func_176223_P().func_185897_m())
             {
-                boolean flag1 = p_189540_1_.func_177229_b(field_176283_b);
+                boolean flag1 = ((Boolean)p_189540_1_.func_177229_b(field_176283_b)).booleanValue();
 
                 if (flag1 != flag)
                 {
-                    p_189540_2_.func_180501_a(p_189540_3_, p_189540_1_.func_177226_a(field_176283_b, flag), 2);
+                    p_189540_2_.func_180501_a(p_189540_3_, p_189540_1_.func_177226_a(field_176283_b, Boolean.valueOf(flag)), 2);
                     this.func_185731_a((EntityPlayer)null, p_189540_2_, p_189540_3_, flag);
                 }
             }
@@ -146,18 +147,18 @@
 
         if (p_180642_3_.func_176740_k().func_176722_c())
         {
-            iblockstate = iblockstate.func_177226_a(field_176284_a, p_180642_3_).func_177226_a(field_176283_b, false);
+            iblockstate = iblockstate.func_177226_a(field_176284_a, p_180642_3_).func_177226_a(field_176283_b, Boolean.valueOf(false));
             iblockstate = iblockstate.func_177226_a(field_176285_M, p_180642_5_ > 0.5F ? BlockTrapDoor.DoorHalf.TOP : BlockTrapDoor.DoorHalf.BOTTOM);
         }
         else
         {
-            iblockstate = iblockstate.func_177226_a(field_176284_a, p_180642_8_.func_174811_aO().func_176734_d()).func_177226_a(field_176283_b, false);
+            iblockstate = iblockstate.func_177226_a(field_176284_a, p_180642_8_.func_174811_aO().func_176734_d()).func_177226_a(field_176283_b, Boolean.valueOf(false));
             iblockstate = iblockstate.func_177226_a(field_176285_M, p_180642_3_ == EnumFacing.UP ? BlockTrapDoor.DoorHalf.BOTTOM : BlockTrapDoor.DoorHalf.TOP);
         }
 
         if (p_180642_1_.func_175640_z(p_180642_2_))
         {
-            iblockstate = iblockstate.func_177226_a(field_176283_b, true);
+            iblockstate = iblockstate.func_177226_a(field_176283_b, Boolean.valueOf(true));
         }
 
         return iblockstate;
@@ -202,7 +203,7 @@
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176284_a, func_176281_b(p_176203_1_)).func_177226_a(field_176283_b, (p_176203_1_ & 4) != 0).func_177226_a(field_176285_M, (p_176203_1_ & 8) == 0 ? BlockTrapDoor.DoorHalf.BOTTOM : BlockTrapDoor.DoorHalf.TOP);
+        return this.func_176223_P().func_177226_a(field_176284_a, func_176281_b(p_176203_1_)).func_177226_a(field_176283_b, Boolean.valueOf((p_176203_1_ & 4) != 0)).func_177226_a(field_176285_M, (p_176203_1_ & 8) == 0 ? BlockTrapDoor.DoorHalf.BOTTOM : BlockTrapDoor.DoorHalf.TOP);
     }
 
     @SideOnly(Side.CLIENT)
@@ -214,9 +215,9 @@
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= func_176282_a(p_176201_1_.func_177229_b(field_176284_a));
+        i = i | func_176282_a((EnumFacing)p_176201_1_.func_177229_b(field_176284_a));
 
-        if (p_176201_1_.func_177229_b(field_176283_b))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176283_b)).booleanValue())
         {
             i |= 4;
         }
@@ -231,22 +232,34 @@
 
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177226_a(field_176284_a, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_176284_a)));
+        return p_185499_1_.func_177226_a(field_176284_a, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_176284_a)));
     }
 
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_176284_a)));
+        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_176284_a)));
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176284_a, field_176283_b, field_176285_M);
+        return new BlockStateContainer(this, new IProperty[] {field_176284_a, field_176283_b, field_176285_M});
     }
 
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
     {
-        return (p_193383_4_ == EnumFacing.UP && p_193383_2_.func_177229_b(field_176285_M) == BlockTrapDoor.DoorHalf.TOP || p_193383_4_ == EnumFacing.DOWN && p_193383_2_.func_177229_b(field_176285_M) == BlockTrapDoor.DoorHalf.BOTTOM) && !p_193383_2_.func_177229_b(field_176283_b) ? BlockFaceShape.SOLID : BlockFaceShape.UNDEFINED;
+        return (p_193383_4_ == EnumFacing.UP && p_193383_2_.func_177229_b(field_176285_M) == BlockTrapDoor.DoorHalf.TOP || p_193383_4_ == EnumFacing.DOWN && p_193383_2_.func_177229_b(field_176285_M) == BlockTrapDoor.DoorHalf.BOTTOM) && !((Boolean)p_193383_2_.func_177229_b(field_176283_b)).booleanValue() ? BlockFaceShape.SOLID : BlockFaceShape.UNDEFINED;
+    }
+
+    @Override
+    public boolean isLadder(IBlockState state, IBlockAccess world, BlockPos pos, EntityLivingBase entity)
+    {
+        if (state.func_177229_b(field_176283_b))
+        {
+            IBlockState down = world.func_180495_p(pos.func_177977_b());
+            if (down.func_177230_c() == net.minecraft.init.Blocks.field_150468_ap)
+                return down.func_177229_b(BlockLadder.field_176382_a) == state.func_177229_b(field_176284_a);
+        }
+        return false;
     }
 
     public static enum DoorHalf implements IStringSerializable
