--- before/net/minecraft/block/BlockStem.java
+++ after/net/minecraft/block/BlockStem.java
@@ -3,10 +3,12 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.properties.PropertyInteger;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
+import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.init.Blocks;
 import net.minecraft.init.Items;
 import net.minecraft.item.Item;
@@ -23,36 +25,24 @@
     public static final PropertyInteger field_176484_a = PropertyInteger.func_177719_a("age", 0, 7);
     public static final PropertyDirection field_176483_b = BlockTorch.field_176596_a;
     private final Block field_149877_a;
-    protected static final AxisAlignedBB[] field_185521_d = new AxisAlignedBB[]
-        {
-            new AxisAlignedBB(0.375, 0.0, 0.375, 0.625, 0.125, 0.625),
-            new AxisAlignedBB(0.375, 0.0, 0.375, 0.625, 0.25, 0.625),
-            new AxisAlignedBB(0.375, 0.0, 0.375, 0.625, 0.375, 0.625),
-            new AxisAlignedBB(0.375, 0.0, 0.375, 0.625, 0.5, 0.625),
-            new AxisAlignedBB(0.375, 0.0, 0.375, 0.625, 0.625, 0.625),
-            new AxisAlignedBB(0.375, 0.0, 0.375, 0.625, 0.75, 0.625),
-            new AxisAlignedBB(0.375, 0.0, 0.375, 0.625, 0.875, 0.625),
-            new AxisAlignedBB(0.375, 0.0, 0.375, 0.625, 1.0, 0.625)
-        };
+    protected static final AxisAlignedBB[] field_185521_d = new AxisAlignedBB[] {new AxisAlignedBB(0.375D, 0.0D, 0.375D, 0.625D, 0.125D, 0.625D), new AxisAlignedBB(0.375D, 0.0D, 0.375D, 0.625D, 0.25D, 0.625D), new AxisAlignedBB(0.375D, 0.0D, 0.375D, 0.625D, 0.375D, 0.625D), new AxisAlignedBB(0.375D, 0.0D, 0.375D, 0.625D, 0.5D, 0.625D), new AxisAlignedBB(0.375D, 0.0D, 0.375D, 0.625D, 0.625D, 0.625D), new AxisAlignedBB(0.375D, 0.0D, 0.375D, 0.625D, 0.75D, 0.625D), new AxisAlignedBB(0.375D, 0.0D, 0.375D, 0.625D, 0.875D, 0.625D), new AxisAlignedBB(0.375D, 0.0D, 0.375D, 0.625D, 1.0D, 0.625D)};
 
     protected BlockStem(Block p_i45430_1_)
     {
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176484_a, 0).func_177226_a(field_176483_b, EnumFacing.UP));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176484_a, Integer.valueOf(0)).func_177226_a(field_176483_b, EnumFacing.UP));
         this.field_149877_a = p_i45430_1_;
         this.func_149675_a(true);
-        this.func_149647_a(null);
+        this.func_149647_a((CreativeTabs)null);
     }
 
-    @Override
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
-        return field_185521_d[p_185496_1_.func_177229_b(field_176484_a)];
+        return field_185521_d[((Integer)p_185496_1_.func_177229_b(field_176484_a)).intValue()];
     }
 
-    @Override
     public IBlockState func_176221_a(IBlockState p_176221_1_, IBlockAccess p_176221_2_, BlockPos p_176221_3_)
     {
-        int i = p_176221_1_.func_177229_b(field_176484_a);
+        int i = ((Integer)p_176221_1_.func_177229_b(field_176484_a)).intValue();
         p_176221_1_ = p_176221_1_.func_177226_a(field_176483_b, EnumFacing.UP);
 
         for (EnumFacing enumfacing : EnumFacing.Plane.HORIZONTAL)
@@ -67,29 +57,28 @@
         return p_176221_1_;
     }
 
-    @Override
     protected boolean func_185514_i(IBlockState p_185514_1_)
     {
         return p_185514_1_.func_177230_c() == Blocks.field_150458_ak;
     }
 
-    @Override
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
     {
         super.func_180650_b(p_180650_1_, p_180650_2_, p_180650_3_, p_180650_4_);
 
+        if (!p_180650_1_.func_175697_a(p_180650_2_, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
         if (p_180650_1_.func_175671_l(p_180650_2_.func_177984_a()) >= 9)
         {
             float f = BlockCrops.func_180672_a(this, p_180650_1_, p_180650_2_);
 
-            if (p_180650_4_.nextInt((int)(25.0F / f) + 1) == 0)
+            if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_180650_1_, p_180650_2_, p_180650_3_, p_180650_4_.nextInt((int)(25.0F / f) + 1) == 0))
             {
-                int i = p_180650_3_.func_177229_b(field_176484_a);
+                int i = ((Integer)p_180650_3_.func_177229_b(field_176484_a)).intValue();
 
                 if (i < 7)
                 {
-                    p_180650_3_ = p_180650_3_.func_177226_a(field_176484_a, i + 1);
-                    p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_, 2);
+                    IBlockState newState = p_180650_3_.func_177226_a(field_176484_a, Integer.valueOf(i + 1));
+                    p_180650_1_.func_180501_a(p_180650_2_, newState, 2);
                 }
                 else
                 {
@@ -102,42 +91,45 @@
                     }
 
                     p_180650_2_ = p_180650_2_.func_177972_a(EnumFacing.Plane.HORIZONTAL.func_179518_a(p_180650_4_));
-                    Block block = p_180650_1_.func_180495_p(p_180650_2_.func_177977_b()).func_177230_c();
+                    IBlockState soil = p_180650_1_.func_180495_p(p_180650_2_.func_177977_b());
+                    Block block = soil.func_177230_c();
 
-                    if (p_180650_1_.func_180495_p(p_180650_2_).func_177230_c().field_149764_J == Material.field_151579_a
-                            && (block == Blocks.field_150458_ak || block == Blocks.field_150346_d || block == Blocks.field_150349_c))
+                    if (p_180650_1_.func_175623_d(p_180650_2_) && (block.canSustainPlant(soil, p_180650_1_, p_180650_2_.func_177977_b(), EnumFacing.UP, this) || block == Blocks.field_150346_d || block == Blocks.field_150349_c))
                     {
                         p_180650_1_.func_175656_a(p_180650_2_, this.field_149877_a.func_176223_P());
                     }
                 }
+                net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_180650_1_, p_180650_2_, p_180650_3_, p_180650_1_.func_180495_p(p_180650_2_));
             }
         }
     }
 
     public void func_176482_g(World p_176482_1_, BlockPos p_176482_2_, IBlockState p_176482_3_)
     {
-        int i = p_176482_3_.func_177229_b(field_176484_a) + MathHelper.func_76136_a(p_176482_1_.field_73012_v, 2, 5);
-        p_176482_1_.func_180501_a(p_176482_2_, p_176482_3_.func_177226_a(field_176484_a, Math.min(7, i)), 2);
+        int i = ((Integer)p_176482_3_.func_177229_b(field_176484_a)).intValue() + MathHelper.func_76136_a(p_176482_1_.field_73012_v, 2, 5);
+        p_176482_1_.func_180501_a(p_176482_2_, p_176482_3_.func_177226_a(field_176484_a, Integer.valueOf(Math.min(7, i))), 2);
     }
 
-    @Override
     public void func_180653_a(World p_180653_1_, BlockPos p_180653_2_, IBlockState p_180653_3_, float p_180653_4_, int p_180653_5_)
     {
         super.func_180653_a(p_180653_1_, p_180653_2_, p_180653_3_, p_180653_4_, p_180653_5_);
+    }
 
-        if (!p_180653_1_.field_72995_K)
+    @Override
+    public void getDrops(net.minecraft.util.NonNullList<ItemStack> drops, IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
         {
             Item item = this.func_176481_j();
 
             if (item != null)
             {
-                int i = p_180653_3_.func_177229_b(field_176484_a);
+                int i = ((Integer)state.func_177229_b(field_176484_a)).intValue();
 
                 for (int j = 0; j < 3; ++j)
                 {
-                    if (p_180653_1_.field_73012_v.nextInt(15) <= i)
+                    if (RANDOM.nextInt(15) <= i)
                     {
-                        func_180635_a(p_180653_1_, p_180653_2_, new ItemStack(item));
+                        drops.add(new ItemStack(item));
                     }
                 }
             }
@@ -157,52 +149,44 @@
         }
     }
 
-    @Override
     public Item func_180660_a(IBlockState p_180660_1_, Random p_180660_2_, int p_180660_3_)
     {
         return Items.field_190931_a;
     }
 
-    @Override
     public ItemStack func_185473_a(World p_185473_1_, BlockPos p_185473_2_, IBlockState p_185473_3_)
     {
         Item item = this.func_176481_j();
         return item == null ? ItemStack.field_190927_a : new ItemStack(item);
     }
 
-    @Override
     public boolean func_176473_a(World p_176473_1_, BlockPos p_176473_2_, IBlockState p_176473_3_, boolean p_176473_4_)
     {
-        return p_176473_3_.func_177229_b(field_176484_a) != 7;
+        return ((Integer)p_176473_3_.func_177229_b(field_176484_a)).intValue() != 7;
     }
 
-    @Override
     public boolean func_180670_a(World p_180670_1_, Random p_180670_2_, BlockPos p_180670_3_, IBlockState p_180670_4_)
     {
         return true;
     }
 
-    @Override
     public void func_176474_b(World p_176474_1_, Random p_176474_2_, BlockPos p_176474_3_, IBlockState p_176474_4_)
     {
         this.func_176482_g(p_176474_1_, p_176474_3_, p_176474_4_);
     }
 
-    @Override
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176484_a, p_176203_1_);
+        return this.func_176223_P().func_177226_a(field_176484_a, Integer.valueOf(p_176203_1_));
     }
 
-    @Override
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176484_a);
+        return ((Integer)p_176201_1_.func_177229_b(field_176484_a)).intValue();
     }
 
-    @Override
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176484_a, field_176483_b);
+        return new BlockStateContainer(this, new IProperty[] {field_176484_a, field_176483_b});
     }
 }
