--- before/net/minecraft/block/BlockPistonBase.java
+++ after/net/minecraft/block/BlockPistonBase.java
@@ -279,7 +279,7 @@
                 }
 
                 if (!flag1
-                        && iblockstate.getMaterial() != Material.AIR
+                        && !iblockstate.getBlock().isAir(iblockstate, worldIn, blockpos)
                         && canPush(iblockstate, worldIn, blockpos, enumfacing.getOpposite(), false, enumfacing)
                         && (iblockstate.getPushReaction() == EnumPushReaction.NORMAL || block == Blocks.PISTON || block == Blocks.STICKY_PISTON))
                 {
@@ -353,7 +353,7 @@
                     return false;
                 }
 
-                return !block.hasTileEntity();
+                return !block.hasTileEntity(blockStateIn);
             }
             else
             {
@@ -399,7 +399,9 @@
             {
                 BlockPos blockpos1 = list2.get(j);
                 IBlockState iblockstate = worldIn.getBlockState(blockpos1);
-                iblockstate.getBlock().dropBlockAsItem(worldIn, blockpos1, iblockstate, 0);
+                // Forge: With our change to how snowballs are dropped this needs to disallow to mimic vanilla behavior.
+                float chance = iblockstate.getBlock() instanceof BlockSnow ? -1.0f : 1.0f;
+                iblockstate.getBlock().dropBlockAsItemWithChance(worldIn, blockpos1, iblockstate, chance, 0);
                 worldIn.setBlockState(blockpos1, Blocks.AIR.getDefaultState(), 4);
                 aiblockstate[--k] = iblockstate;
             }
@@ -493,6 +495,14 @@
     {
         return new BlockStateContainer(this, FACING, EXTENDED);
     }
+
+    /* ======================================== FORGE START =====================================*/
+    public boolean rotateBlock(World world, BlockPos pos, EnumFacing axis)
+    {
+        IBlockState state = world.getBlockState(pos);
+        return !state.getValue(EXTENDED) && super.rotateBlock(world, pos, axis);
+    }
+    /* ======================================== FORGE END =======================================*/
 
     @Override
     public BlockFaceShape getBlockFaceShape(IBlockAccess worldIn, IBlockState state, BlockPos pos, EnumFacing face)
