--- before/net/minecraft/block/BlockHopper.java
+++ after/net/minecraft/block/BlockHopper.java
@@ -5,6 +5,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.state.BlockFaceShape;
@@ -52,7 +53,7 @@
     public BlockHopper()
     {
         super(Material.field_151573_f, MapColor.field_151665_m);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176430_a, EnumFacing.DOWN).func_177226_a(field_176429_b, true));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176430_a, EnumFacing.DOWN).func_177226_a(field_176429_b, Boolean.valueOf(true)));
         this.func_149647_a(CreativeTabs.field_78028_d);
     }
 
@@ -79,7 +80,7 @@
             enumfacing = EnumFacing.DOWN;
         }
 
-        return this.func_176223_P().func_177226_a(field_176430_a, enumfacing).func_177226_a(field_176429_b, true);
+        return this.func_176223_P().func_177226_a(field_176430_a, enumfacing).func_177226_a(field_176429_b, Boolean.valueOf(true));
     }
 
     public TileEntity func_149915_a(World p_149915_1_, int p_149915_2_)
@@ -141,9 +142,9 @@
     {
         boolean flag = !p_176427_1_.func_175640_z(p_176427_2_);
 
-        if (flag != p_176427_3_.func_177229_b(field_176429_b))
+        if (flag != ((Boolean)p_176427_3_.func_177229_b(field_176429_b)).booleanValue())
         {
-            p_176427_1_.func_180501_a(p_176427_2_, p_176427_3_.func_177226_a(field_176429_b, flag), 4);
+            p_176427_1_.func_180501_a(p_176427_2_, p_176427_3_.func_177226_a(field_176429_b, Boolean.valueOf(flag)), 4);
         }
     }
 
@@ -175,17 +176,17 @@
         return false;
     }
 
-    @SideOnly(Side.CLIENT)
-    public boolean func_176225_a(IBlockState p_176225_1_, IBlockAccess p_176225_2_, BlockPos p_176225_3_, EnumFacing p_176225_4_)
-    {
-        return true;
-    }
-
     public static EnumFacing func_176428_b(int p_176428_0_)
     {
         return EnumFacing.func_82600_a(p_176428_0_ & 7);
     }
 
+    @SideOnly(Side.CLIENT)
+    public boolean func_176225_a(IBlockState p_176225_1_, IBlockAccess p_176225_2_, BlockPos p_176225_3_, EnumFacing p_176225_4_)
+    {
+        return true;
+    }
+
     public static boolean func_149917_c(int p_149917_0_)
     {
         return (p_149917_0_ & 8) != 8;
@@ -209,15 +210,15 @@
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176430_a, func_176428_b(p_176203_1_)).func_177226_a(field_176429_b, func_149917_c(p_176203_1_));
+        return this.func_176223_P().func_177226_a(field_176430_a, func_176428_b(p_176203_1_)).func_177226_a(field_176429_b, Boolean.valueOf(func_149917_c(p_176203_1_)));
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_176430_a).func_176745_a();
+        i = i | ((EnumFacing)p_176201_1_.func_177229_b(field_176430_a)).func_176745_a();
 
-        if (!p_176201_1_.func_177229_b(field_176429_b))
+        if (!((Boolean)p_176201_1_.func_177229_b(field_176429_b)).booleanValue())
         {
             i |= 8;
         }
@@ -227,17 +228,17 @@
 
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177226_a(field_176430_a, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_176430_a)));
+        return p_185499_1_.func_177226_a(field_176430_a, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_176430_a)));
     }
 
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_176430_a)));
+        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_176430_a)));
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176430_a, field_176429_b);
+        return new BlockStateContainer(this, new IProperty[] {field_176430_a, field_176429_b});
     }
 
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
