--- before/net/minecraft/block/BlockPurpurSlab.java
+++ after/net/minecraft/block/BlockPurpurSlab.java
@@ -17,7 +17,7 @@
 
 public abstract class BlockPurpurSlab extends BlockSlab
 {
-    public static final PropertyEnum<BlockPurpurSlab.Variant> field_185678_d = PropertyEnum.func_177709_a("variant", BlockPurpurSlab.Variant.class);
+    public static final PropertyEnum<BlockPurpurSlab.Variant> field_185678_d = PropertyEnum.<BlockPurpurSlab.Variant>func_177709_a("variant", BlockPurpurSlab.Variant.class);
 
     public BlockPurpurSlab()
     {
@@ -69,7 +69,7 @@
 
     protected BlockStateContainer func_180661_e()
     {
-        return this.func_176552_j() ? new BlockStateContainer(this, field_185678_d) : new BlockStateContainer(this, field_176554_a, field_185678_d);
+        return this.func_176552_j() ? new BlockStateContainer(this, new IProperty[] {field_185678_d}) : new BlockStateContainer(this, new IProperty[] {field_176554_a, field_185678_d});
     }
 
     public String func_150002_b(int p_150002_1_)
