--- before/net/minecraft/block/BlockPistonMoving.java
+++ after/net/minecraft/block/BlockPistonMoving.java
@@ -4,6 +4,7 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockFaceShape;
@@ -35,17 +36,11 @@
     public BlockPistonMoving()
     {
         super(Material.field_76233_E);
-        this.func_180632_j(
-            this.field_176227_L
-            .func_177621_b()
-            .func_177226_a(field_176426_a, EnumFacing.NORTH)
-            .func_177226_a(field_176425_b, BlockPistonExtension.EnumPistonType.DEFAULT)
-        );
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176426_a, EnumFacing.NORTH).func_177226_a(field_176425_b, BlockPistonExtension.EnumPistonType.DEFAULT));
         this.func_149711_c(-1.0F);
     }
 
     @Nullable
-    @Override
     public TileEntity func_149915_a(World p_149915_1_, int p_149915_2_)
     {
         return null;
@@ -56,7 +51,6 @@
         return new TileEntityPiston(p_185588_0_, p_185588_1_, p_185588_2_, p_185588_3_);
     }
 
-    @Override
     public void func_180663_b(World p_180663_1_, BlockPos p_180663_2_, IBlockState p_180663_3_)
     {
         TileEntity tileentity = p_180663_1_.func_175625_s(p_180663_2_);
@@ -71,54 +65,38 @@
         }
     }
 
-    @Override
     public boolean func_176196_c(World p_176196_1_, BlockPos p_176196_2_)
     {
         return false;
     }
 
-    @Override
     public boolean func_176198_a(World p_176198_1_, BlockPos p_176198_2_, EnumFacing p_176198_3_)
     {
         return false;
     }
 
-    @Override
     public void func_176206_d(World p_176206_1_, BlockPos p_176206_2_, IBlockState p_176206_3_)
     {
-        BlockPos blockpos = p_176206_2_.func_177972_a(p_176206_3_.func_177229_b(field_176426_a).func_176734_d());
+        BlockPos blockpos = p_176206_2_.func_177972_a(((EnumFacing)p_176206_3_.func_177229_b(field_176426_a)).func_176734_d());
         IBlockState iblockstate = p_176206_1_.func_180495_p(blockpos);
 
-        if (iblockstate.func_177230_c() instanceof BlockPistonBase && iblockstate.func_177229_b(BlockPistonBase.field_176320_b))
+        if (iblockstate.func_177230_c() instanceof BlockPistonBase && ((Boolean)iblockstate.func_177229_b(BlockPistonBase.field_176320_b)).booleanValue())
         {
             p_176206_1_.func_175698_g(blockpos);
         }
     }
 
-    @Override
     public boolean func_149662_c(IBlockState p_149662_1_)
     {
         return false;
     }
 
-    @Override
     public boolean func_149686_d(IBlockState p_149686_1_)
     {
         return false;
     }
 
-    @Override
-    public boolean func_180639_a(
-        World p_180639_1_,
-        BlockPos p_180639_2_,
-        IBlockState p_180639_3_,
-        EntityPlayer p_180639_4_,
-        EnumHand p_180639_5_,
-        EnumFacing p_180639_6_,
-        float p_180639_7_,
-        float p_180639_8_,
-        float p_180639_9_
-    )
+    public boolean func_180639_a(World p_180639_1_, BlockPos p_180639_2_, IBlockState p_180639_3_, EntityPlayer p_180639_4_, EnumHand p_180639_5_, EnumFacing p_180639_6_, float p_180639_7_, float p_180639_8_, float p_180639_9_)
     {
         if (!p_180639_1_.field_72995_K && p_180639_1_.func_175625_s(p_180639_2_) == null)
         {
@@ -131,16 +109,14 @@
         }
     }
 
-    @Override
     public Item func_180660_a(IBlockState p_180660_1_, Random p_180660_2_, int p_180660_3_)
     {
         return Items.field_190931_a;
     }
 
-    @Override
     public void func_180653_a(World p_180653_1_, BlockPos p_180653_2_, IBlockState p_180653_3_, float p_180653_4_, int p_180653_5_)
     {
-        if (!p_180653_1_.field_72995_K)
+        if (false && !p_180653_1_.field_72995_K) // Forge: Noop this out
         {
             TileEntityPiston tileentitypiston = this.func_185589_c(p_180653_1_, p_180653_2_);
 
@@ -150,16 +126,15 @@
                 iblockstate.func_177230_c().func_176226_b(p_180653_1_, p_180653_2_, iblockstate, 0);
             }
         }
+        super.func_180653_a(p_180653_1_, p_180653_2_, p_180653_3_, 1, p_180653_5_); // mimic vanilla behavior from above and ignore chance
     }
 
     @Nullable
-    @Override
     public RayTraceResult func_180636_a(IBlockState p_180636_1_, World p_180636_2_, BlockPos p_180636_3_, Vec3d p_180636_4_, Vec3d p_180636_5_)
     {
         return null;
     }
 
-    @Override
     public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
     {
         if (!p_189540_2_.field_72995_K)
@@ -169,23 +144,13 @@
     }
 
     @Nullable
-    @Override
     public AxisAlignedBB func_180646_a(IBlockState p_180646_1_, IBlockAccess p_180646_2_, BlockPos p_180646_3_)
     {
         TileEntityPiston tileentitypiston = this.func_185589_c(p_180646_2_, p_180646_3_);
         return tileentitypiston == null ? null : tileentitypiston.func_184321_a(p_180646_2_, p_180646_3_);
     }
 
-    @Override
-    public void func_185477_a(
-        IBlockState p_185477_1_,
-        World p_185477_2_,
-        BlockPos p_185477_3_,
-        AxisAlignedBB p_185477_4_,
-        List<AxisAlignedBB> p_185477_5_,
-        @Nullable Entity p_185477_6_,
-        boolean p_185477_7_
-    )
+    public void func_185477_a(IBlockState p_185477_1_, World p_185477_2_, BlockPos p_185477_3_, AxisAlignedBB p_185477_4_, List<AxisAlignedBB> p_185477_5_, @Nullable Entity p_185477_6_, boolean p_185477_7_)
     {
         TileEntityPiston tileentitypiston = this.func_185589_c(p_185477_2_, p_185477_3_);
 
@@ -195,7 +160,6 @@
         }
     }
 
-    @Override
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
         TileEntityPiston tileentitypiston = this.func_185589_c(p_185496_2_, p_185496_3_);
@@ -209,37 +173,30 @@
         return tileentity instanceof TileEntityPiston ? (TileEntityPiston)tileentity : null;
     }
 
-    @Override
     public ItemStack func_185473_a(World p_185473_1_, BlockPos p_185473_2_, IBlockState p_185473_3_)
     {
         return ItemStack.field_190927_a;
     }
 
-    @Override
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P()
-               .func_177226_a(field_176426_a, BlockPistonExtension.func_176322_b(p_176203_1_))
-               .func_177226_a(field_176425_b, (p_176203_1_ & 8) > 0 ? BlockPistonExtension.EnumPistonType.STICKY : BlockPistonExtension.EnumPistonType.DEFAULT);
+        return this.func_176223_P().func_177226_a(field_176426_a, BlockPistonExtension.func_176322_b(p_176203_1_)).func_177226_a(field_176425_b, (p_176203_1_ & 8) > 0 ? BlockPistonExtension.EnumPistonType.STICKY : BlockPistonExtension.EnumPistonType.DEFAULT);
     }
 
-    @Override
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177226_a(field_176426_a, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_176426_a)));
+        return p_185499_1_.func_177226_a(field_176426_a, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_176426_a)));
     }
 
-    @Override
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_176426_a)));
+        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_176426_a)));
     }
 
-    @Override
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_176426_a).func_176745_a();
+        i = i | ((EnumFacing)p_176201_1_.func_177229_b(field_176426_a)).func_176745_a();
 
         if (p_176201_1_.func_177229_b(field_176425_b) == BlockPistonExtension.EnumPistonType.STICKY)
         {
@@ -249,13 +206,22 @@
         return i;
     }
 
-    @Override
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176426_a, field_176425_b);
+        return new BlockStateContainer(this, new IProperty[] {field_176426_a, field_176425_b});
     }
 
     @Override
+    public void getDrops(net.minecraft.util.NonNullList<net.minecraft.item.ItemStack> drops, IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
+        TileEntityPiston tileentitypiston = this.func_185589_c(world, pos);
+        if (tileentitypiston != null)
+        {
+            IBlockState pushed = tileentitypiston.func_174927_b();
+            drops.addAll(pushed.func_177230_c().getDrops(world, pos, pushed, fortune)); // use the old method until it gets removed, for backward compatibility
+        }
+    }
+
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
     {
         return BlockFaceShape.UNDEFINED;
