--- before/net/minecraft/block/BlockCarrot.java
+++ after/net/minecraft/block/BlockCarrot.java
@@ -23,6 +23,6 @@
 
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
-        return field_185533_a[p_185496_1_.func_177229_b(this.func_185524_e())];
+        return field_185533_a[((Integer)p_185496_1_.func_177229_b(this.func_185524_e())).intValue()];
     }
 }
