--- before/net/minecraft/block/BlockDirt.java
+++ after/net/minecraft/block/BlockDirt.java
@@ -2,6 +2,7 @@
 
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockStateContainer;
@@ -17,19 +18,19 @@
 
 public class BlockDirt extends Block
 {
-    public static final PropertyEnum<BlockDirt.DirtType> field_176386_a = PropertyEnum.func_177709_a("variant", BlockDirt.DirtType.class);
+    public static final PropertyEnum<BlockDirt.DirtType> field_176386_a = PropertyEnum.<BlockDirt.DirtType>func_177709_a("variant", BlockDirt.DirtType.class);
     public static final PropertyBool field_176385_b = PropertyBool.func_177716_a("snowy");
 
     protected BlockDirt()
     {
         super(Material.field_151578_c);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176386_a, BlockDirt.DirtType.DIRT).func_177226_a(field_176385_b, false));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176386_a, BlockDirt.DirtType.DIRT).func_177226_a(field_176385_b, Boolean.valueOf(false)));
         this.func_149647_a(CreativeTabs.field_78030_b);
     }
 
     public MapColor func_180659_g(IBlockState p_180659_1_, IBlockAccess p_180659_2_, BlockPos p_180659_3_)
     {
-        return p_180659_1_.func_177229_b(field_176386_a).func_181066_d();
+        return ((BlockDirt.DirtType)p_180659_1_.func_177229_b(field_176386_a)).func_181066_d();
     }
 
     public IBlockState func_176221_a(IBlockState p_176221_1_, IBlockAccess p_176221_2_, BlockPos p_176221_3_)
@@ -37,7 +38,7 @@
         if (p_176221_1_.func_177229_b(field_176386_a) == BlockDirt.DirtType.PODZOL)
         {
             Block block = p_176221_2_.func_180495_p(p_176221_3_.func_177984_a()).func_177230_c();
-            p_176221_1_ = p_176221_1_.func_177226_a(field_176385_b, block == Blocks.field_150433_aE || block == Blocks.field_150431_aC);
+            p_176221_1_ = p_176221_1_.func_177226_a(field_176385_b, Boolean.valueOf(block == Blocks.field_150433_aE || block == Blocks.field_150431_aC));
         }
 
         return p_176221_1_;
@@ -52,7 +53,7 @@
 
     public ItemStack func_185473_a(World p_185473_1_, BlockPos p_185473_2_, IBlockState p_185473_3_)
     {
-        return new ItemStack(this, 1, p_185473_3_.func_177229_b(field_176386_a).func_176925_a());
+        return new ItemStack(this, 1, ((BlockDirt.DirtType)p_185473_3_.func_177229_b(field_176386_a)).func_176925_a());
     }
 
     public IBlockState func_176203_a(int p_176203_1_)
@@ -62,17 +63,17 @@
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176386_a).func_176925_a();
+        return ((BlockDirt.DirtType)p_176201_1_.func_177229_b(field_176386_a)).func_176925_a();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176386_a, field_176385_b);
+        return new BlockStateContainer(this, new IProperty[] {field_176386_a, field_176385_b});
     }
 
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        BlockDirt.DirtType blockdirt$dirttype = p_180651_1_.func_177229_b(field_176386_a);
+        BlockDirt.DirtType blockdirt$dirttype = (BlockDirt.DirtType)p_180651_1_.func_177229_b(field_176386_a);
 
         if (blockdirt$dirttype == BlockDirt.DirtType.PODZOL)
         {
