--- before/net/minecraft/block/BlockSnow.java
+++ after/net/minecraft/block/BlockSnow.java
@@ -111,7 +111,7 @@
         {
             BlockFaceShape blockfaceshape = iblockstate.getBlockFaceShape(worldIn, pos.down(), EnumFacing.UP);
             return blockfaceshape == BlockFaceShape.SOLID
-                   || iblockstate.getMaterial() == Material.LEAVES
+                   || iblockstate.getBlock().isLeaves(iblockstate, worldIn, pos.down())
                    || block == this && iblockstate.getValue(LAYERS) == 8;
         }
         else
@@ -130,7 +130,6 @@
     {
         if (!this.canPlaceBlockAt(worldIn, pos))
         {
-            this.dropBlockAsItem(worldIn, pos, state, 0);
             worldIn.setBlockToAir(pos);
             return false;
         }
@@ -145,9 +144,8 @@
         World worldIn, EntityPlayer player, BlockPos pos, IBlockState state, @Nullable TileEntity te, ItemStack stack
     )
     {
-        spawnAsEntity(worldIn, pos, new ItemStack(Items.SNOWBALL, state.getValue(LAYERS) + 1, 0));
+        super.harvestBlock(worldIn, player, pos, state, te, stack);
         worldIn.setBlockToAir(pos);
-        player.addStat(StatList.getBlockStats(this));
     }
 
     @Override
@@ -159,7 +157,7 @@
     @Override
     public int quantityDropped(Random random)
     {
-        return 0;
+        return 1;
     }
 
     @Override
@@ -167,7 +165,6 @@
     {
         if (worldIn.getLightFor(EnumSkyBlock.BLOCK, pos) > 11)
         {
-            this.dropBlockAsItem(worldIn, pos, worldIn.getBlockState(pos), 0);
             worldIn.setBlockToAir(pos);
         }
     }
@@ -205,6 +202,12 @@
     public int getMetaFromState(IBlockState state)
     {
         return state.getValue(LAYERS) - 1;
+    }
+
+    @Override
+    public int quantityDropped(IBlockState state, int fortune, Random random)
+    {
+        return state.getValue(LAYERS) + 1;
     }
 
     @Override
