--- before/net/minecraft/block/BlockOldLeaf.java
+++ after/net/minecraft/block/BlockOldLeaf.java
@@ -2,6 +2,7 @@
 
 import com.google.common.base.Predicate;
 import javax.annotation.Nullable;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
@@ -18,28 +19,19 @@
 
 public class BlockOldLeaf extends BlockLeaves
 {
-    public static final PropertyEnum<BlockPlanks.EnumType> field_176239_P = PropertyEnum.func_177708_a(
-                    "variant", BlockPlanks.EnumType.class, new Predicate<BlockPlanks.EnumType>()
+    public static final PropertyEnum<BlockPlanks.EnumType> field_176239_P = PropertyEnum.<BlockPlanks.EnumType>func_177708_a("variant", BlockPlanks.EnumType.class, new Predicate<BlockPlanks.EnumType>()
     {
         public boolean apply(@Nullable BlockPlanks.EnumType p_apply_1_)
         {
             return p_apply_1_.func_176839_a() < 4;
         }
-    }
-            );
+    });
 
     public BlockOldLeaf()
     {
-        this.func_180632_j(
-            this.field_176227_L
-            .func_177621_b()
-            .func_177226_a(field_176239_P, BlockPlanks.EnumType.OAK)
-            .func_177226_a(field_176236_b, true)
-            .func_177226_a(field_176237_a, true)
-        );
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176239_P, BlockPlanks.EnumType.OAK).func_177226_a(field_176236_b, Boolean.valueOf(true)).func_177226_a(field_176237_a, Boolean.valueOf(true)));
     }
 
-    @Override
     protected void func_176234_a(World p_176234_1_, BlockPos p_176234_2_, IBlockState p_176234_3_, int p_176234_4_)
     {
         if (p_176234_3_.func_177229_b(field_176239_P) == BlockPlanks.EnumType.OAK && p_176234_1_.field_73012_v.nextInt(p_176234_4_) == 0)
@@ -48,13 +40,11 @@
         }
     }
 
-    @Override
     protected int func_176232_d(IBlockState p_176232_1_)
     {
         return p_176232_1_.func_177229_b(field_176239_P) == BlockPlanks.EnumType.JUNGLE ? 40 : super.func_176232_d(p_176232_1_);
     }
 
-    @Override
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
     {
         p_149666_2_.add(new ItemStack(this, 1, BlockPlanks.EnumType.OAK.func_176839_a()));
@@ -63,33 +53,27 @@
         p_149666_2_.add(new ItemStack(this, 1, BlockPlanks.EnumType.JUNGLE.func_176839_a()));
     }
 
-    @Override
     protected ItemStack func_180643_i(IBlockState p_180643_1_)
     {
-        return new ItemStack(Item.func_150898_a(this), 1, p_180643_1_.func_177229_b(field_176239_P).func_176839_a());
+        return new ItemStack(Item.func_150898_a(this), 1, ((BlockPlanks.EnumType)p_180643_1_.func_177229_b(field_176239_P)).func_176839_a());
     }
 
-    @Override
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P()
-               .func_177226_a(field_176239_P, this.func_176233_b(p_176203_1_))
-               .func_177226_a(field_176237_a, (p_176203_1_ & 4) == 0)
-               .func_177226_a(field_176236_b, (p_176203_1_ & 8) > 0);
+        return this.func_176223_P().func_177226_a(field_176239_P, this.func_176233_b(p_176203_1_)).func_177226_a(field_176237_a, Boolean.valueOf((p_176203_1_ & 4) == 0)).func_177226_a(field_176236_b, Boolean.valueOf((p_176203_1_ & 8) > 0));
     }
 
-    @Override
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_176239_P).func_176839_a();
+        i = i | ((BlockPlanks.EnumType)p_176201_1_.func_177229_b(field_176239_P)).func_176839_a();
 
-        if (!p_176201_1_.func_177229_b(field_176237_a))
+        if (!((Boolean)p_176201_1_.func_177229_b(field_176237_a)).booleanValue())
         {
             i |= 4;
         }
 
-        if (p_176201_1_.func_177229_b(field_176236_b))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176236_b)).booleanValue())
         {
             i |= 8;
         }
@@ -97,37 +81,36 @@
         return i;
     }
 
-    @Override
     public BlockPlanks.EnumType func_176233_b(int p_176233_1_)
     {
         return BlockPlanks.EnumType.func_176837_a((p_176233_1_ & 3) % 4);
     }
 
-    @Override
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176239_P, field_176236_b, field_176237_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176239_P, field_176236_b, field_176237_a});
     }
 
-    @Override
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        return p_180651_1_.func_177229_b(field_176239_P).func_176839_a();
+        return ((BlockPlanks.EnumType)p_180651_1_.func_177229_b(field_176239_P)).func_176839_a();
     }
 
-    @Override
-    public void func_180657_a(
-        World p_180657_1_, EntityPlayer p_180657_2_, BlockPos p_180657_3_, IBlockState p_180657_4_, @Nullable TileEntity p_180657_5_, ItemStack p_180657_6_
-    )
+    public void func_180657_a(World p_180657_1_, EntityPlayer p_180657_2_, BlockPos p_180657_3_, IBlockState p_180657_4_, @Nullable TileEntity p_180657_5_, ItemStack p_180657_6_)
     {
         if (!p_180657_1_.field_72995_K && p_180657_6_.func_77973_b() == Items.field_151097_aZ)
         {
             p_180657_2_.func_71029_a(StatList.func_188055_a(this));
-            func_180635_a(p_180657_1_, p_180657_3_, new ItemStack(Item.func_150898_a(this), 1, p_180657_4_.func_177229_b(field_176239_P).func_176839_a()));
         }
         else
         {
             super.func_180657_a(p_180657_1_, p_180657_2_, p_180657_3_, p_180657_4_, p_180657_5_, p_180657_6_);
         }
+    }
+
+    @Override
+    public NonNullList<ItemStack> onSheared(ItemStack item, net.minecraft.world.IBlockAccess world, BlockPos pos, int fortune)
+    {
+        return NonNullList.func_191197_a(1, new ItemStack(this, 1, world.func_180495_p(pos).func_177229_b(field_176239_P).func_176839_a()));
     }
 }
