--- before/net/minecraft/block/BlockStaticLiquid.java
+++ after/net/minecraft/block/BlockStaticLiquid.java
@@ -21,7 +21,6 @@
         }
     }
 
-    @Override
     public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
     {
         if (!this.func_176365_e(p_189540_2_, p_189540_3_, p_189540_1_))
@@ -37,7 +36,6 @@
         p_176370_1_.func_175684_a(p_176370_2_, blockdynamicliquid, this.func_149738_a(p_176370_1_));
     }
 
-    @Override
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
     {
         if (this.field_149764_J == Material.field_151587_i)
@@ -54,22 +52,22 @@
                     {
                         blockpos = blockpos.func_177982_a(p_180650_4_.nextInt(3) - 1, 1, p_180650_4_.nextInt(3) - 1);
 
-                        if (blockpos.func_177956_o() >= 0 && blockpos.func_177956_o() < 256 && !p_180650_1_.func_175667_e(blockpos))
+                        if (blockpos.func_177956_o() >= 0 && blockpos.func_177956_o() < p_180650_1_.func_72800_K() && !p_180650_1_.func_175667_e(blockpos))
                         {
                             return;
                         }
 
-                        Block block = p_180650_1_.func_180495_p(blockpos).func_177230_c();
+                        IBlockState block = p_180650_1_.func_180495_p(blockpos);
 
-                        if (block.field_149764_J == Material.field_151579_a)
+                        if (block.func_177230_c().isAir(block, p_180650_1_, blockpos))
                         {
                             if (this.func_176369_e(p_180650_1_, blockpos))
                             {
-                                p_180650_1_.func_175656_a(blockpos, Blocks.field_150480_ab.func_176223_P());
+                                p_180650_1_.func_175656_a(blockpos, net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(p_180650_1_, blockpos, p_180650_2_, Blocks.field_150480_ab.func_176223_P()));
                                 return;
                             }
                         }
-                        else if (block.field_149764_J.func_76230_c())
+                        else if (block.func_185904_a().func_76230_c())
                         {
                             return;
                         }
@@ -88,7 +86,7 @@
 
                         if (p_180650_1_.func_175623_d(blockpos1.func_177984_a()) && this.func_176368_m(p_180650_1_, blockpos1))
                         {
-                            p_180650_1_.func_175656_a(blockpos1.func_177984_a(), Blocks.field_150480_ab.func_176223_P());
+                            p_180650_1_.func_175656_a(blockpos1.func_177984_a(), net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(p_180650_1_, blockpos1.func_177984_a(), p_180650_2_, Blocks.field_150480_ab.func_176223_P()));
                         }
                     }
                 }
@@ -111,8 +109,6 @@
 
     private boolean func_176368_m(World p_176368_1_, BlockPos p_176368_2_)
     {
-        return p_176368_2_.func_177956_o() >= 0 && p_176368_2_.func_177956_o() < 256 && !p_176368_1_.func_175667_e(p_176368_2_)
-               ? false
-               : p_176368_1_.func_180495_p(p_176368_2_).func_185904_a().func_76217_h();
+        return p_176368_2_.func_177956_o() >= 0 && p_176368_2_.func_177956_o() < 256 && !p_176368_1_.func_175667_e(p_176368_2_) ? false : p_176368_1_.func_180495_p(p_176368_2_).func_185904_a().func_76217_h();
     }
 }
