--- before/net/minecraft/block/BlockOldLeaf.java
+++ after/net/minecraft/block/BlockOldLeaf.java
@@ -123,11 +123,16 @@
         if (!worldIn.isRemote && stack.getItem() == Items.SHEARS)
         {
             player.addStat(StatList.getBlockStats(this));
-            spawnAsEntity(worldIn, pos, new ItemStack(Item.getItemFromBlock(this), 1, state.getValue(VARIANT).getMetadata()));
         }
         else
         {
             super.harvestBlock(worldIn, player, pos, state, te, stack);
         }
+    }
+
+    @Override
+    public NonNullList<ItemStack> onSheared(ItemStack item, net.minecraft.world.IBlockAccess world, BlockPos pos, int fortune)
+    {
+        return NonNullList.withSize(1, new ItemStack(this, 1, world.getBlockState(pos).getValue(VARIANT).getMetadata()));
     }
 }
