--- before/net/minecraft/block/BlockSponge.java
+++ after/net/minecraft/block/BlockSponge.java
@@ -5,6 +5,7 @@
 import java.util.Queue;
 import java.util.Random;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
@@ -28,7 +29,7 @@
     protected BlockSponge()
     {
         super(Material.field_151583_m);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176313_a, false));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176313_a, Boolean.valueOf(false)));
         this.func_149647_a(CreativeTabs.field_78030_b);
     }
 
@@ -39,7 +40,7 @@
 
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        return p_180651_1_.func_177229_b(field_176313_a) ? 1 : 0;
+        return ((Boolean)p_180651_1_.func_177229_b(field_176313_a)).booleanValue() ? 1 : 0;
     }
 
     public void func_176213_c(World p_176213_1_, BlockPos p_176213_2_, IBlockState p_176213_3_)
@@ -55,25 +56,25 @@
 
     protected void func_176311_e(World p_176311_1_, BlockPos p_176311_2_, IBlockState p_176311_3_)
     {
-        if (!p_176311_3_.func_177229_b(field_176313_a) && this.func_176312_d(p_176311_1_, p_176311_2_))
+        if (!((Boolean)p_176311_3_.func_177229_b(field_176313_a)).booleanValue() && this.func_176312_d(p_176311_1_, p_176311_2_))
         {
-            p_176311_1_.func_180501_a(p_176311_2_, p_176311_3_.func_177226_a(field_176313_a, true), 2);
+            p_176311_1_.func_180501_a(p_176311_2_, p_176311_3_.func_177226_a(field_176313_a, Boolean.valueOf(true)), 2);
             p_176311_1_.func_175718_b(2001, p_176311_2_, Block.func_149682_b(Blocks.field_150355_j));
         }
     }
 
     private boolean func_176312_d(World p_176312_1_, BlockPos p_176312_2_)
     {
-        Queue<Tuple<BlockPos, Integer>> queue = Lists.newLinkedList();
-        List<BlockPos> list = Lists.newArrayList();
-        queue.add(new Tuple<>(p_176312_2_, 0));
+        Queue<Tuple<BlockPos, Integer>> queue = Lists.<Tuple<BlockPos, Integer>>newLinkedList();
+        List<BlockPos> list = Lists.<BlockPos>newArrayList();
+        queue.add(new Tuple(p_176312_2_, Integer.valueOf(0)));
         int i = 0;
 
         while (!queue.isEmpty())
         {
-            Tuple<BlockPos, Integer> tuple = queue.poll();
+            Tuple<BlockPos, Integer> tuple = (Tuple)queue.poll();
             BlockPos blockpos = tuple.func_76341_a();
-            int j = tuple.func_76340_b();
+            int j = ((Integer)tuple.func_76340_b()).intValue();
 
             for (EnumFacing enumfacing : EnumFacing.values())
             {
@@ -87,7 +88,7 @@
 
                     if (j < 6)
                     {
-                        queue.add(new Tuple<>(blockpos1, j + 1));
+                        queue.add(new Tuple(blockpos1, j + 1));
                     }
                 }
             }
@@ -114,23 +115,23 @@
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176313_a, (p_176203_1_ & 1) == 1);
+        return this.func_176223_P().func_177226_a(field_176313_a, Boolean.valueOf((p_176203_1_ & 1) == 1));
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176313_a) ? 1 : 0;
+        return ((Boolean)p_176201_1_.func_177229_b(field_176313_a)).booleanValue() ? 1 : 0;
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176313_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176313_a});
     }
 
     @SideOnly(Side.CLIENT)
     public void func_180655_c(IBlockState p_180655_1_, World p_180655_2_, BlockPos p_180655_3_, Random p_180655_4_)
     {
-        if (p_180655_1_.func_177229_b(field_176313_a))
+        if (((Boolean)p_180655_1_.func_177229_b(field_176313_a)).booleanValue())
         {
             EnumFacing enumfacing = EnumFacing.func_176741_a(p_180655_4_);
 
@@ -142,13 +143,13 @@
 
                 if (enumfacing == EnumFacing.DOWN)
                 {
-                    d1 -= 0.05D;
+                    d1 = d1 - 0.05D;
                     d0 += p_180655_4_.nextDouble();
                     d2 += p_180655_4_.nextDouble();
                 }
                 else
                 {
-                    d1 += p_180655_4_.nextDouble() * 0.8D;
+                    d1 = d1 + p_180655_4_.nextDouble() * 0.8D;
 
                     if (enumfacing.func_176740_k() == EnumFacing.Axis.X)
                     {
