--- before/net/minecraft/block/BlockChorusFlower.java
+++ after/net/minecraft/block/BlockChorusFlower.java
@@ -4,6 +4,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyInteger;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
@@ -56,7 +57,7 @@
             {
                 int i = state.getValue(AGE);
 
-                if (i < 5 && rand.nextInt(1) == 0)
+                if (i < 5 &&  net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, blockpos, state, rand.nextInt(1) == 0))
                 {
                     boolean flag = false;
                     boolean flag1 = false;
@@ -147,6 +148,7 @@
                     {
                         this.placeDeadFlower(worldIn, pos);
                     }
+                    net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state, worldIn.getBlockState(pos));
                 }
             }
         }
