--- before/net/minecraft/block/BlockButton.java
+++ after/net/minecraft/block/BlockButton.java
@@ -100,7 +100,7 @@
         IBlockState iblockstate = worldIn.getBlockState(blockpos);
         boolean flag = iblockstate.getBlockFaceShape(worldIn, blockpos, direction) == BlockFaceShape.SOLID;
         Block block = iblockstate.getBlock();
-        return direction == EnumFacing.UP ? block == Blocks.HOPPER || !isExceptionBlockForAttaching(block) && flag : !isExceptBlockForAttachWithPiston(block) && flag;
+        return direction == EnumFacing.UP ? iblockstate.isTopSolid() || !isExceptionBlockForAttaching(block) && flag : !isExceptBlockForAttachWithPiston(block) && flag;
     }
 
     @Override
