--- before/net/minecraft/block/BlockStoneBrick.java
+++ after/net/minecraft/block/BlockStoneBrick.java
@@ -1,6 +1,7 @@
 package net.minecraft.block;
 
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
@@ -11,7 +12,7 @@
 
 public class BlockStoneBrick extends Block
 {
-    public static final PropertyEnum<BlockStoneBrick.EnumType> field_176249_a = PropertyEnum.func_177709_a("variant", BlockStoneBrick.EnumType.class);
+    public static final PropertyEnum<BlockStoneBrick.EnumType> field_176249_a = PropertyEnum.<BlockStoneBrick.EnumType>func_177709_a("variant", BlockStoneBrick.EnumType.class);
     public static final int field_176248_b = BlockStoneBrick.EnumType.DEFAULT.func_176612_a();
     public static final int field_176250_M = BlockStoneBrick.EnumType.MOSSY.func_176612_a();
     public static final int field_176251_N = BlockStoneBrick.EnumType.CRACKED.func_176612_a();
@@ -26,7 +27,7 @@
 
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        return p_180651_1_.func_177229_b(field_176249_a).func_176612_a();
+        return ((BlockStoneBrick.EnumType)p_180651_1_.func_177229_b(field_176249_a)).func_176612_a();
     }
 
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
@@ -44,12 +45,12 @@
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176249_a).func_176612_a();
+        return ((BlockStoneBrick.EnumType)p_176201_1_.func_177229_b(field_176249_a)).func_176612_a();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176249_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176249_a});
     }
 
     public static enum EnumType implements IStringSerializable
