--- before/net/minecraft/block/BlockPumpkin.java
+++ after/net/minecraft/block/BlockPumpkin.java
@@ -153,8 +153,8 @@
     @Override
     public boolean canPlaceBlockAt(World worldIn, BlockPos pos)
     {
-        return worldIn.getBlockState(pos).getBlock().material.isReplaceable()
-               && worldIn.getBlockState(pos.down()).isTopSolid();
+        return worldIn.getBlockState(pos).getBlock().isReplaceable(worldIn, pos)
+               && worldIn.isSideSolid(pos.down(), EnumFacing.UP);
     }
 
     @Override
