--- before/net/minecraft/block/BlockNote.java
+++ after/net/minecraft/block/BlockNote.java
@@ -86,7 +86,9 @@
             if (tileentity instanceof TileEntityNote)
             {
                 TileEntityNote tileentitynote = (TileEntityNote)tileentity;
+                int old = tileentitynote.note;
                 tileentitynote.changePitch();
+                if (old == tileentitynote.note) return false;
                 tileentitynote.triggerNote(worldIn, pos);
                 playerIn.addStat(StatList.NOTEBLOCK_TUNED);
             }
@@ -129,6 +131,10 @@
     @Override
     public boolean eventReceived(IBlockState state, World worldIn, BlockPos pos, int id, int param)
     {
+        net.minecraftforge.event.world.NoteBlockEvent.Play e = new net.minecraftforge.event.world.NoteBlockEvent.Play(worldIn, pos, state, param, id);
+        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(e)) return false;
+        id = e.getInstrument().ordinal();
+        param = e.getVanillaNoteId();
         float f = (float)Math.pow(2.0, (param - 12) / 12.0);
         worldIn.playSound(null, pos, this.getInstrument(id), SoundCategory.RECORDS, 3.0F, f);
         worldIn.spawnParticle(
