--- before/net/minecraft/block/BlockPistonMoving.java
+++ after/net/minecraft/block/BlockPistonMoving.java
@@ -140,7 +140,7 @@
     @Override
     public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune)
     {
-        if (!worldIn.isRemote)
+        if (false && !worldIn.isRemote) // Forge: Noop this out
         {
             TileEntityPiston tileentitypiston = this.getTilePistonAt(worldIn, pos);
 
@@ -150,6 +150,7 @@
                 iblockstate.getBlock().dropBlockAsItem(worldIn, pos, iblockstate, 0);
             }
         }
+        super.dropBlockAsItemWithChance(worldIn, pos, state, 1, fortune); // mimic vanilla behavior from above and ignore chance
     }
 
     @Nullable
@@ -253,6 +254,17 @@
     protected BlockStateContainer createBlockState()
     {
         return new BlockStateContainer(this, FACING, TYPE);
+    }
+
+    @Override
+    public void getDrops(net.minecraft.util.NonNullList<net.minecraft.item.ItemStack> drops, IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
+        TileEntityPiston tileentitypiston = this.getTilePistonAt(world, pos);
+        if (tileentitypiston != null)
+        {
+            IBlockState pushed = tileentitypiston.getPistonState();
+            drops.addAll(pushed.getBlock().getDrops(world, pos, pushed, fortune)); // use the old method until it gets removed, for backward compatibility
+        }
     }
 
     @Override
