--- before/net/minecraft/block/BlockIce.java
+++ after/net/minecraft/block/BlockIce.java
@@ -31,23 +31,24 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public BlockRenderLayer func_180664_k()
     {
         return BlockRenderLayer.TRANSLUCENT;
     }
 
-    @Override
-    public void func_180657_a(
-        World p_180657_1_, EntityPlayer p_180657_2_, BlockPos p_180657_3_, IBlockState p_180657_4_, @Nullable TileEntity p_180657_5_, ItemStack p_180657_6_
-    )
+    public void func_180657_a(World p_180657_1_, EntityPlayer p_180657_2_, BlockPos p_180657_3_, IBlockState p_180657_4_, @Nullable TileEntity p_180657_5_, ItemStack p_180657_6_)
     {
         p_180657_2_.func_71029_a(StatList.func_188055_a(this));
         p_180657_2_.func_71020_j(0.005F);
 
-        if (this.func_149700_E() && EnchantmentHelper.func_77506_a(Enchantments.field_185306_r, p_180657_6_) > 0)
+        if (this.canSilkHarvest(p_180657_1_, p_180657_3_, p_180657_4_, p_180657_2_) && EnchantmentHelper.func_77506_a(Enchantments.field_185306_r, p_180657_6_) > 0)
         {
-            func_180635_a(p_180657_1_, p_180657_3_, this.func_180643_i(p_180657_4_));
+            java.util.List<ItemStack> items = new java.util.ArrayList<ItemStack>();
+            items.add(this.func_180643_i(p_180657_4_));
+
+            net.minecraftforge.event.ForgeEventFactory.fireBlockHarvesting(items, p_180657_1_, p_180657_3_, p_180657_4_, 0, 1.0f, true, p_180657_2_);
+            for (ItemStack is : items)
+                func_180635_a(p_180657_1_, p_180657_3_, is);
         }
         else
         {
@@ -58,7 +59,9 @@
             }
 
             int i = EnchantmentHelper.func_77506_a(Enchantments.field_185308_t, p_180657_6_);
+            harvesters.set(p_180657_2_);
             this.func_176226_b(p_180657_1_, p_180657_3_, p_180657_4_, i);
+            harvesters.set(null);
             Material material = p_180657_1_.func_180495_p(p_180657_3_.func_177977_b()).func_185904_a();
 
             if (material.func_76230_c() || material.func_76224_d())
@@ -68,13 +71,11 @@
         }
     }
 
-    @Override
     public int func_149745_a(Random p_149745_1_)
     {
         return 0;
     }
 
-    @Override
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
     {
         if (p_180650_1_.func_175642_b(EnumSkyBlock.BLOCK, p_180650_2_) > 11 - this.func_176223_P().func_185891_c())
@@ -97,7 +98,6 @@
         }
     }
 
-    @Override
     public EnumPushReaction func_149656_h(IBlockState p_149656_1_)
     {
         return EnumPushReaction.NORMAL;
