--- before/net/minecraft/block/BlockSilverfish.java
+++ after/net/minecraft/block/BlockSilverfish.java
@@ -2,6 +2,7 @@
 
 import java.util.Random;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
@@ -16,7 +17,7 @@
 
 public class BlockSilverfish extends Block
 {
-    public static final PropertyEnum<BlockSilverfish.EnumType> field_176378_a = PropertyEnum.func_177709_a("variant", BlockSilverfish.EnumType.class);
+    public static final PropertyEnum<BlockSilverfish.EnumType> field_176378_a = PropertyEnum.<BlockSilverfish.EnumType>func_177709_a("variant", BlockSilverfish.EnumType.class);
 
     public BlockSilverfish()
     {
@@ -87,12 +88,12 @@
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176378_a).func_176881_a();
+        return ((BlockSilverfish.EnumType)p_176201_1_.func_177229_b(field_176378_a)).func_176881_a();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176378_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176378_a});
     }
 
     public static enum EnumType implements IStringSerializable
