--- before/net/minecraft/block/BlockFarmland.java
+++ after/net/minecraft/block/BlockFarmland.java
@@ -76,11 +76,7 @@
     @Override
     public void onFallenUpon(World worldIn, BlockPos pos, Entity entityIn, float fallDistance)
     {
-        if (!worldIn.isRemote
-                && worldIn.rand.nextFloat() < fallDistance - 0.5F
-                && entityIn instanceof EntityLivingBase
-                && (entityIn instanceof EntityPlayer || worldIn.getGameRules().getBoolean("mobGriefing"))
-                && entityIn.width * entityIn.width * entityIn.height > 0.512F)
+        if (net.minecraftforge.common.ForgeHooks.onFarmlandTrample(worldIn, pos, Blocks.DIRT.getDefaultState(), fallDistance, entityIn)) // Forge: Move logic to Entity#canTrample
         {
             turnToDirt(worldIn, pos);
         }
@@ -103,7 +99,7 @@
     private boolean hasCrops(World worldIn, BlockPos pos)
     {
         Block block = worldIn.getBlockState(pos.up()).getBlock();
-        return block instanceof BlockCrops || block instanceof BlockStem;
+        return block instanceof net.minecraftforge.common.IPlantable && canSustainPlant(worldIn.getBlockState(pos), worldIn, pos, net.minecraft.util.EnumFacing.UP, (net.minecraftforge.common.IPlantable)block);
     }
 
     private boolean hasWater(World worldIn, BlockPos pos)
@@ -116,7 +112,7 @@
             }
         }
 
-        return false;
+        return net.minecraftforge.common.FarmlandWaterManager.hasBlockWaterTicket(worldIn, pos);
     }
 
     @Override
