--- before/net/minecraft/block/BlockDynamicLiquid.java
+++ after/net/minecraft/block/BlockDynamicLiquid.java
@@ -29,6 +29,7 @@
     @Override
     public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand)
     {
+        if (!worldIn.isAreaLoaded(pos, this.getSlopeFindDistance(worldIn))) return; // Forge: avoid loading unloaded chunks
         int i = state.getValue(LEVEL);
         int j = 1;
 
@@ -70,7 +71,7 @@
                 }
             }
 
-            if (this.adjacentSourceBlocks >= 2 && this.material == Material.WATER)
+            if (this.adjacentSourceBlocks >= 2 && net.minecraftforge.event.ForgeEventFactory.canCreateFluidSource(worldIn, pos, state, this.material == Material.WATER))
             {
                 IBlockState iblockstate = worldIn.getBlockState(pos.down());
 
@@ -122,7 +123,7 @@
             if (this.material == Material.LAVA
                     && worldIn.getBlockState(pos.down()).getMaterial() == Material.WATER)
             {
-                worldIn.setBlockState(pos.down(), Blocks.STONE.getDefaultState());
+                worldIn.setBlockState(pos.down(), net.minecraftforge.event.ForgeEventFactory.fireFluidPlaceBlockEvent(worldIn, pos.down(), pos, Blocks.STONE.getDefaultState()));
                 this.triggerMixEffects(worldIn, pos.down());
                 return;
             }
@@ -170,6 +171,7 @@
                 }
                 else
                 {
+                    if (state.getBlock() != Blocks.SNOW_LAYER) //Forge: Vanilla has a 'bug' where snowballs don't drop like every other block. So special case because ewww...
                     state.getBlock().dropBlockAsItem(worldIn, pos, state, 0);
                 }
             }
@@ -192,7 +194,7 @@
                 if (!this.isBlocked(worldIn, blockpos, iblockstate)
                         && (iblockstate.getMaterial() != this.material || iblockstate.getValue(LEVEL) > 0))
                 {
-                    if (!this.isBlocked(worldIn, blockpos.down(), iblockstate))
+                    if (!this.isBlocked(worldIn, blockpos.down(), worldIn.getBlockState(blockpos.down())))
                     {
                         return distance;
                     }
@@ -260,7 +262,8 @@
 
     private boolean isBlocked(World worldIn, BlockPos pos, IBlockState state)
     {
-        Block block = worldIn.getBlockState(pos).getBlock();
+        Block block = state.getBlock(); //Forge: state must be valid for position
+        Material mat = state.getMaterial();
 
         if (block instanceof BlockDoor || block == Blocks.STANDING_SIGN || block == Blocks.LADDER || block == Blocks.REEDS)
         {
@@ -268,7 +271,7 @@
         }
         else
         {
-            return block.material != Material.PORTAL && block.material != Material.STRUCTURE_VOID ? block.material.blocksMovement() : true;
+            return mat != Material.PORTAL && mat != Material.STRUCTURE_VOID ? mat.blocksMovement() : true;
         }
     }
 
