--- before/net/minecraft/block/BlockRedstoneComparator.java
+++ after/net/minecraft/block/BlockRedstoneComparator.java
@@ -5,6 +5,7 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockStateContainer;
@@ -35,12 +36,12 @@
 public class BlockRedstoneComparator extends BlockRedstoneDiode implements ITileEntityProvider
 {
     public static final PropertyBool field_176464_a = PropertyBool.func_177716_a("powered");
-    public static final PropertyEnum<BlockRedstoneComparator.Mode> field_176463_b = PropertyEnum.func_177709_a("mode", BlockRedstoneComparator.Mode.class);
+    public static final PropertyEnum<BlockRedstoneComparator.Mode> field_176463_b = PropertyEnum.<BlockRedstoneComparator.Mode>func_177709_a("mode", BlockRedstoneComparator.Mode.class);
 
     public BlockRedstoneComparator(boolean p_i45399_1_)
     {
         super(p_i45399_1_);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_185512_D, EnumFacing.NORTH).func_177226_a(field_176464_a, false).func_177226_a(field_176463_b, BlockRedstoneComparator.Mode.COMPARE));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_185512_D, EnumFacing.NORTH).func_177226_a(field_176464_a, Boolean.valueOf(false)).func_177226_a(field_176463_b, BlockRedstoneComparator.Mode.COMPARE));
         this.field_149758_A = true;
     }
 
@@ -66,23 +67,23 @@
 
     protected IBlockState func_180674_e(IBlockState p_180674_1_)
     {
-        Boolean obool = p_180674_1_.func_177229_b(field_176464_a);
-        BlockRedstoneComparator.Mode blockredstonecomparator$mode = p_180674_1_.func_177229_b(field_176463_b);
-        EnumFacing enumfacing = p_180674_1_.func_177229_b(field_185512_D);
+        Boolean obool = (Boolean)p_180674_1_.func_177229_b(field_176464_a);
+        BlockRedstoneComparator.Mode blockredstonecomparator$mode = (BlockRedstoneComparator.Mode)p_180674_1_.func_177229_b(field_176463_b);
+        EnumFacing enumfacing = (EnumFacing)p_180674_1_.func_177229_b(field_185512_D);
         return Blocks.field_150455_bV.func_176223_P().func_177226_a(field_185512_D, enumfacing).func_177226_a(field_176464_a, obool).func_177226_a(field_176463_b, blockredstonecomparator$mode);
     }
 
     protected IBlockState func_180675_k(IBlockState p_180675_1_)
     {
-        Boolean obool = p_180675_1_.func_177229_b(field_176464_a);
-        BlockRedstoneComparator.Mode blockredstonecomparator$mode = p_180675_1_.func_177229_b(field_176463_b);
-        EnumFacing enumfacing = p_180675_1_.func_177229_b(field_185512_D);
+        Boolean obool = (Boolean)p_180675_1_.func_177229_b(field_176464_a);
+        BlockRedstoneComparator.Mode blockredstonecomparator$mode = (BlockRedstoneComparator.Mode)p_180675_1_.func_177229_b(field_176463_b);
+        EnumFacing enumfacing = (EnumFacing)p_180675_1_.func_177229_b(field_185512_D);
         return Blocks.field_150441_bU.func_176223_P().func_177226_a(field_185512_D, enumfacing).func_177226_a(field_176464_a, obool).func_177226_a(field_176463_b, blockredstonecomparator$mode);
     }
 
     protected boolean func_176406_l(IBlockState p_176406_1_)
     {
-        return this.field_149914_a || p_176406_1_.func_177229_b(field_176464_a);
+        return this.field_149914_a || ((Boolean)p_176406_1_.func_177229_b(field_176464_a)).booleanValue();
     }
 
     protected int func_176408_a(IBlockAccess p_176408_1_, BlockPos p_176408_2_, IBlockState p_176408_3_)
@@ -126,7 +127,7 @@
     protected int func_176397_f(World p_176397_1_, BlockPos p_176397_2_, IBlockState p_176397_3_)
     {
         int i = super.func_176397_f(p_176397_1_, p_176397_2_, p_176397_3_);
-        EnumFacing enumfacing = p_176397_3_.func_177229_b(field_185512_D);
+        EnumFacing enumfacing = (EnumFacing)p_176397_3_.func_177229_b(field_185512_D);
         BlockPos blockpos = p_176397_2_.func_177972_a(enumfacing);
         IBlockState iblockstate = p_176397_1_.func_180495_p(blockpos);
 
@@ -160,14 +161,14 @@
     @Nullable
     private EntityItemFrame func_176461_a(World p_176461_1_, final EnumFacing p_176461_2_, BlockPos p_176461_3_)
     {
-        List<EntityItemFrame> list = p_176461_1_.func_175647_a(EntityItemFrame.class, new AxisAlignedBB((double)p_176461_3_.func_177958_n(), (double)p_176461_3_.func_177956_o(), (double)p_176461_3_.func_177952_p(), (double)(p_176461_3_.func_177958_n() + 1), (double)(p_176461_3_.func_177956_o() + 1), (double)(p_176461_3_.func_177952_p() + 1)), new Predicate<Entity>()
+        List<EntityItemFrame> list = p_176461_1_.<EntityItemFrame>func_175647_a(EntityItemFrame.class, new AxisAlignedBB((double)p_176461_3_.func_177958_n(), (double)p_176461_3_.func_177956_o(), (double)p_176461_3_.func_177952_p(), (double)(p_176461_3_.func_177958_n() + 1), (double)(p_176461_3_.func_177956_o() + 1), (double)(p_176461_3_.func_177952_p() + 1)), new Predicate<Entity>()
         {
             public boolean apply(@Nullable Entity p_apply_1_)
             {
                 return p_apply_1_ != null && p_apply_1_.func_174811_aO() == p_176461_2_;
             }
         });
-        return list.size() == 1 ? list.get(0) : null;
+        return list.size() == 1 ? (EntityItemFrame)list.get(0) : null;
     }
 
     public boolean func_180639_a(World p_180639_1_, BlockPos p_180639_2_, IBlockState p_180639_3_, EntityPlayer p_180639_4_, EnumHand p_180639_5_, EnumFacing p_180639_6_, float p_180639_7_, float p_180639_8_, float p_180639_9_)
@@ -229,11 +230,11 @@
 
             if (flag && !flag1)
             {
-                p_176462_1_.func_180501_a(p_176462_2_, p_176462_3_.func_177226_a(field_176464_a, false), 2);
+                p_176462_1_.func_180501_a(p_176462_2_, p_176462_3_.func_177226_a(field_176464_a, Boolean.valueOf(false)), 2);
             }
             else if (!flag && flag1)
             {
-                p_176462_1_.func_180501_a(p_176462_2_, p_176462_3_.func_177226_a(field_176464_a, true), 2);
+                p_176462_1_.func_180501_a(p_176462_2_, p_176462_3_.func_177226_a(field_176464_a, Boolean.valueOf(true)), 2);
             }
 
             this.func_176400_h(p_176462_1_, p_176462_2_, p_176462_3_);
@@ -244,7 +245,7 @@
     {
         if (this.field_149914_a)
         {
-            p_180650_1_.func_180501_a(p_180650_2_, this.func_180675_k(p_180650_3_).func_177226_a(field_176464_a, true), 4);
+            p_180650_1_.func_180501_a(p_180650_2_, this.func_180675_k(p_180650_3_).func_177226_a(field_176464_a, Boolean.valueOf(true)), 4);
         }
 
         this.func_176462_k(p_180650_1_, p_180650_2_, p_180650_3_);
@@ -277,15 +278,15 @@
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_185512_D, EnumFacing.func_176731_b(p_176203_1_)).func_177226_a(field_176464_a, (p_176203_1_ & 8) > 0).func_177226_a(field_176463_b, (p_176203_1_ & 4) > 0 ? BlockRedstoneComparator.Mode.SUBTRACT : BlockRedstoneComparator.Mode.COMPARE);
+        return this.func_176223_P().func_177226_a(field_185512_D, EnumFacing.func_176731_b(p_176203_1_)).func_177226_a(field_176464_a, Boolean.valueOf((p_176203_1_ & 8) > 0)).func_177226_a(field_176463_b, (p_176203_1_ & 4) > 0 ? BlockRedstoneComparator.Mode.SUBTRACT : BlockRedstoneComparator.Mode.COMPARE);
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_185512_D).func_176736_b();
+        i = i | ((EnumFacing)p_176201_1_.func_177229_b(field_185512_D)).func_176736_b();
 
-        if (p_176201_1_.func_177229_b(field_176464_a))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176464_a)).booleanValue())
         {
             i |= 8;
         }
@@ -300,22 +301,37 @@
 
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177226_a(field_185512_D, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_185512_D)));
+        return p_185499_1_.func_177226_a(field_185512_D, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_185512_D)));
     }
 
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_185512_D)));
+        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_185512_D)));
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_185512_D, field_176463_b, field_176464_a);
+        return new BlockStateContainer(this, new IProperty[] {field_185512_D, field_176463_b, field_176464_a});
     }
 
     public IBlockState func_180642_a(World p_180642_1_, BlockPos p_180642_2_, EnumFacing p_180642_3_, float p_180642_4_, float p_180642_5_, float p_180642_6_, int p_180642_7_, EntityLivingBase p_180642_8_)
     {
-        return this.func_176223_P().func_177226_a(field_185512_D, p_180642_8_.func_174811_aO().func_176734_d()).func_177226_a(field_176464_a, false).func_177226_a(field_176463_b, BlockRedstoneComparator.Mode.COMPARE);
+        return this.func_176223_P().func_177226_a(field_185512_D, p_180642_8_.func_174811_aO().func_176734_d()).func_177226_a(field_176464_a, Boolean.valueOf(false)).func_177226_a(field_176463_b, BlockRedstoneComparator.Mode.COMPARE);
+    }
+
+    @Override
+    public void onNeighborChange(IBlockAccess world, BlockPos pos, BlockPos neighbor)
+    {
+        if (pos.func_177956_o() == neighbor.func_177956_o() && world instanceof World && !((World) world).field_72995_K)
+        {
+            func_189540_a(world.func_180495_p(pos), (World)world, pos, world.func_180495_p(neighbor).func_177230_c(), neighbor);
+        }
+    }
+
+    @Override
+    public boolean getWeakChanges(IBlockAccess world, BlockPos pos)
+    {
+        return true;
     }
 
     public static enum Mode implements IStringSerializable
