--- before/net/minecraft/block/BlockShulkerBox.java
+++ after/net/minecraft/block/BlockShulkerBox.java
@@ -5,6 +5,7 @@
 import net.minecraft.block.material.EnumPushReaction;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockFaceShape;
@@ -100,7 +101,7 @@
 
             if (tileentity instanceof TileEntityShulkerBox)
             {
-                EnumFacing enumfacing = p_180639_3_.func_177229_b(field_190957_a);
+                EnumFacing enumfacing = (EnumFacing)p_180639_3_.func_177229_b(field_190957_a);
                 boolean flag;
 
                 if (((TileEntityShulkerBox)tileentity).func_190591_p() == TileEntityShulkerBox.AnimationStatus.CLOSED)
@@ -135,12 +136,12 @@
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_190957_a);
+        return new BlockStateContainer(this, new IProperty[] {field_190957_a});
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_190957_a).func_176745_a();
+        return ((EnumFacing)p_176201_1_.func_177229_b(field_190957_a)).func_176745_a();
     }
 
     public IBlockState func_176203_a(int p_176203_1_)
@@ -224,7 +225,7 @@
 
             if (nbttagcompound1.func_150297_b("Items", 9))
             {
-                NonNullList<ItemStack> nonnulllist = NonNullList.func_191197_a(27, ItemStack.field_190927_a);
+                NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191197_a(27, ItemStack.field_190927_a);
                 ItemStackHelper.func_191283_b(nbttagcompound1, nonnulllist);
                 int i = 0;
                 int j = 0;
@@ -345,18 +346,18 @@
 
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177226_a(field_190957_a, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_190957_a)));
+        return p_185499_1_.func_177226_a(field_190957_a, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_190957_a)));
     }
 
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_190957_a)));
+        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_190957_a)));
     }
 
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
     {
         p_193383_2_ = this.func_176221_a(p_193383_2_, p_193383_1_, p_193383_3_);
-        EnumFacing enumfacing = p_193383_2_.func_177229_b(field_190957_a);
+        EnumFacing enumfacing = (EnumFacing)p_193383_2_.func_177229_b(field_190957_a);
         TileEntityShulkerBox.AnimationStatus tileentityshulkerbox$animationstatus = ((TileEntityShulkerBox)p_193383_1_.func_175625_s(p_193383_3_)).func_190591_p();
         return tileentityshulkerbox$animationstatus != TileEntityShulkerBox.AnimationStatus.CLOSED && (tileentityshulkerbox$animationstatus != TileEntityShulkerBox.AnimationStatus.OPENED || enumfacing != p_193383_4_.func_176734_d() && enumfacing != p_193383_4_) ? BlockFaceShape.UNDEFINED : BlockFaceShape.SOLID;
     }
