--- before/net/minecraft/block/BlockNewLog.java
+++ after/net/minecraft/block/BlockNewLog.java
@@ -3,6 +3,7 @@
 import com.google.common.base.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.MapColor;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
@@ -15,7 +16,7 @@
 
 public class BlockNewLog extends BlockLog
 {
-    public static final PropertyEnum<BlockPlanks.EnumType> field_176300_b = PropertyEnum.func_177708_a("variant", BlockPlanks.EnumType.class, new Predicate<BlockPlanks.EnumType>()
+    public static final PropertyEnum<BlockPlanks.EnumType> field_176300_b = PropertyEnum.<BlockPlanks.EnumType>func_177708_a("variant", BlockPlanks.EnumType.class, new Predicate<BlockPlanks.EnumType>()
     {
         public boolean apply(@Nullable BlockPlanks.EnumType p_apply_1_)
         {
@@ -30,7 +31,7 @@
 
     public MapColor func_180659_g(IBlockState p_180659_1_, IBlockAccess p_180659_2_, BlockPos p_180659_3_)
     {
-        BlockPlanks.EnumType blockplanks$enumtype = p_180659_1_.func_177229_b(field_176300_b);
+        BlockPlanks.EnumType blockplanks$enumtype = (BlockPlanks.EnumType)p_180659_1_.func_177229_b(field_176300_b);
 
         switch ((BlockLog.EnumAxis)p_180659_1_.func_177229_b(field_176299_a))
         {
@@ -85,7 +86,7 @@
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_176300_b).func_176839_a() - 4;
+        i = i | ((BlockPlanks.EnumType)p_176201_1_.func_177229_b(field_176300_b)).func_176839_a() - 4;
 
         switch ((BlockLog.EnumAxis)p_176201_1_.func_177229_b(field_176299_a))
         {
@@ -104,16 +105,16 @@
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176300_b, field_176299_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176300_b, field_176299_a});
     }
 
     protected ItemStack func_180643_i(IBlockState p_180643_1_)
     {
-        return new ItemStack(Item.func_150898_a(this), 1, p_180643_1_.func_177229_b(field_176300_b).func_176839_a() - 4);
+        return new ItemStack(Item.func_150898_a(this), 1, ((BlockPlanks.EnumType)p_180643_1_.func_177229_b(field_176300_b)).func_176839_a() - 4);
     }
 
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        return p_180651_1_.func_177229_b(field_176300_b).func_176839_a() - 4;
+        return ((BlockPlanks.EnumType)p_180651_1_.func_177229_b(field_176300_b)).func_176839_a() - 4;
     }
 }
