--- before/net/minecraft/block/BlockDragonEgg.java
+++ after/net/minecraft/block/BlockDragonEgg.java
@@ -52,7 +52,7 @@
 
     private void checkFall(World worldIn, BlockPos pos)
     {
-        if (BlockFalling.canFallThrough(worldIn.getBlockState(pos.down())) && pos.getY() >= 0)
+        if (worldIn.isAirBlock(pos.down()) && BlockFalling.canFallThrough(worldIn.getBlockState(pos.down())) && pos.getY() >= 0)
         {
             int i = 32;
 
@@ -73,7 +73,7 @@
                 worldIn.setBlockToAir(pos);
                 BlockPos blockpos = pos;
 
-                while (BlockFalling.canFallThrough(worldIn.getBlockState(blockpos)) && blockpos.getY() > 0)
+                while (worldIn.isAirBlock(blockpos) && BlockFalling.canFallThrough(worldIn.getBlockState(blockpos)) && blockpos.getY() > 0)
                 {
                     blockpos = blockpos.down();
                 }
@@ -123,7 +123,7 @@
                                         worldIn.rand.nextInt(16) - worldIn.rand.nextInt(16)
                                     );
 
-                if (worldIn.getBlockState(blockpos).getBlock().material == Material.AIR)
+                if (worldIn.isAirBlock(blockpos))
                 {
                     if (worldIn.isRemote)
                     {
