--- before/net/minecraft/block/BlockRedSandstone.java
+++ after/net/minecraft/block/BlockRedSandstone.java
@@ -1,6 +1,7 @@
 package net.minecraft.block;
 
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
@@ -11,7 +12,7 @@
 
 public class BlockRedSandstone extends Block
 {
-    public static final PropertyEnum<BlockRedSandstone.EnumType> field_176336_a = PropertyEnum.func_177709_a("type", BlockRedSandstone.EnumType.class);
+    public static final PropertyEnum<BlockRedSandstone.EnumType> field_176336_a = PropertyEnum.<BlockRedSandstone.EnumType>func_177709_a("type", BlockRedSandstone.EnumType.class);
 
     public BlockRedSandstone()
     {
@@ -22,7 +23,7 @@
 
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        return p_180651_1_.func_177229_b(field_176336_a).func_176827_a();
+        return ((BlockRedSandstone.EnumType)p_180651_1_.func_177229_b(field_176336_a)).func_176827_a();
     }
 
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
@@ -40,12 +41,12 @@
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176336_a).func_176827_a();
+        return ((BlockRedSandstone.EnumType)p_176201_1_.func_177229_b(field_176336_a)).func_176827_a();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176336_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176336_a});
     }
 
     public static enum EnumType implements IStringSerializable
