--- before/net/minecraft/block/BlockDoor.java
+++ after/net/minecraft/block/BlockDoor.java
@@ -230,7 +230,7 @@
                 flag1 = true;
             }
 
-            if (!worldIn.getBlockState(pos.down()).isTopSolid())
+            if (!worldIn.getBlockState(pos.down()).isSideSolid(worldIn,  pos.down(), EnumFacing.UP))
             {
                 worldIn.setBlockToAir(pos);
                 flag1 = true;
@@ -276,11 +276,15 @@
     @Override
     public boolean canPlaceBlockAt(World worldIn, BlockPos pos)
     {
-        return pos.getY() >= 255
-               ? false
-               : worldIn.getBlockState(pos.down()).isTopSolid()
-               && super.canPlaceBlockAt(worldIn, pos)
-               && super.canPlaceBlockAt(worldIn, pos.up());
+        if (pos.getY() >= worldIn.getHeight() - 1)
+        {
+            return false;
+        }
+        else
+        {
+            IBlockState state = worldIn.getBlockState(pos.down());
+            return (state.isTopSolid() || state.getBlockFaceShape(worldIn, pos.down(), EnumFacing.UP) == BlockFaceShape.SOLID) && super.canPlaceBlockAt(worldIn, pos) && super.canPlaceBlockAt(worldIn, pos.up());
+        }
     }
 
     @Override
