--- before/net/minecraft/block/BlockCarpet.java
+++ after/net/minecraft/block/BlockCarpet.java
@@ -2,6 +2,7 @@
 
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
@@ -20,7 +21,7 @@
 
 public class BlockCarpet extends Block
 {
-    public static final PropertyEnum<EnumDyeColor> field_176330_a = PropertyEnum.func_177709_a("color", EnumDyeColor.class);
+    public static final PropertyEnum<EnumDyeColor> field_176330_a = PropertyEnum.<EnumDyeColor>func_177709_a("color", EnumDyeColor.class);
     protected static final AxisAlignedBB field_185758_b = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 1.0D, 0.0625D, 1.0D);
 
     protected BlockCarpet()
@@ -38,7 +39,7 @@
 
     public MapColor func_180659_g(IBlockState p_180659_1_, IBlockAccess p_180659_2_, BlockPos p_180659_3_)
     {
-        return MapColor.func_193558_a(p_180659_1_.func_177229_b(field_176330_a));
+        return MapColor.func_193558_a((EnumDyeColor)p_180659_1_.func_177229_b(field_176330_a));
     }
 
     public boolean func_149662_c(IBlockState p_149662_1_)
@@ -95,7 +96,7 @@
 
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        return p_180651_1_.func_177229_b(field_176330_a).func_176765_a();
+        return ((EnumDyeColor)p_180651_1_.func_177229_b(field_176330_a)).func_176765_a();
     }
 
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
@@ -113,12 +114,12 @@
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176330_a).func_176765_a();
+        return ((EnumDyeColor)p_176201_1_.func_177229_b(field_176330_a)).func_176765_a();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176330_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176330_a});
     }
 
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
