--- before/net/minecraft/block/BlockRedstoneComparator.java
+++ after/net/minecraft/block/BlockRedstoneComparator.java
@@ -385,6 +385,21 @@
                .withProperty(MODE, BlockRedstoneComparator.Mode.COMPARE);
     }
 
+    @Override
+    public void onNeighborChange(IBlockAccess world, BlockPos pos, BlockPos neighbor)
+    {
+        if (pos.getY() == neighbor.getY() && world instanceof World && !((World) world).isRemote)
+        {
+            neighborChanged(world.getBlockState(pos), (World)world, pos, world.getBlockState(neighbor).getBlock(), neighbor);
+        }
+    }
+
+    @Override
+    public boolean getWeakChanges(IBlockAccess world, BlockPos pos)
+    {
+        return true;
+    }
+
     public static enum Mode implements IStringSerializable
     {
         COMPARE("compare"),
