--- before/net/minecraft/block/BlockIce.java
+++ after/net/minecraft/block/BlockIce.java
@@ -31,23 +31,24 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public BlockRenderLayer getRenderLayer()
     {
         return BlockRenderLayer.TRANSLUCENT;
     }
 
-    @Override
-    public void harvestBlock(
-        World worldIn, EntityPlayer player, BlockPos pos, IBlockState state, @Nullable TileEntity te, ItemStack stack
-    )
+    public void harvestBlock(World worldIn, EntityPlayer player, BlockPos pos, IBlockState state, @Nullable TileEntity te, ItemStack stack)
     {
         player.addStat(StatList.getBlockStats(this));
         player.addExhaustion(0.005F);
 
-        if (this.canSilkHarvest() && EnchantmentHelper.getEnchantmentLevel(Enchantments.SILK_TOUCH, stack) > 0)
+        if (this.canSilkHarvest(worldIn, pos, state, player) && EnchantmentHelper.getEnchantmentLevel(Enchantments.SILK_TOUCH, stack) > 0)
         {
-            spawnAsEntity(worldIn, pos, this.getSilkTouchDrop(state));
+            java.util.List<ItemStack> items = new java.util.ArrayList<ItemStack>();
+            items.add(this.getSilkTouchDrop(state));
+
+            net.minecraftforge.event.ForgeEventFactory.fireBlockHarvesting(items, worldIn, pos, state, 0, 1.0f, true, player);
+            for (ItemStack is : items)
+                spawnAsEntity(worldIn, pos, is);
         }
         else
         {
@@ -58,7 +59,9 @@
             }
 
             int i = EnchantmentHelper.getEnchantmentLevel(Enchantments.FORTUNE, stack);
+            harvesters.set(player);
             this.dropBlockAsItem(worldIn, pos, state, i);
+            harvesters.set(null);
             Material material = worldIn.getBlockState(pos.down()).getMaterial();
 
             if (material.blocksMovement() || material.isLiquid())
@@ -68,13 +71,11 @@
         }
     }
 
-    @Override
     public int quantityDropped(Random random)
     {
         return 0;
     }
 
-    @Override
     public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand)
     {
         if (worldIn.getLightFor(EnumSkyBlock.BLOCK, pos) > 11 - this.getDefaultState().getLightOpacity())
@@ -97,7 +98,6 @@
         }
     }
 
-    @Override
     public EnumPushReaction getPushReaction(IBlockState state)
     {
         return EnumPushReaction.NORMAL;
