--- before/net/minecraft/block/BlockLeaves.java
+++ after/net/minecraft/block/BlockLeaves.java
@@ -17,7 +17,7 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-public abstract class BlockLeaves extends Block
+public abstract class BlockLeaves extends Block implements net.minecraftforge.common.IShearable
 {
     public static final PropertyBool field_176237_a = PropertyBool.func_177716_a("decayable");
     public static final PropertyBool field_176236_b = PropertyBool.func_177716_a("check_decay");
@@ -34,7 +34,6 @@
         this.func_149672_a(SoundType.field_185850_c);
     }
 
-    @Override
     public void func_180663_b(World p_180663_1_, BlockPos p_180663_2_, IBlockState p_180663_3_)
     {
         int i = 1;
@@ -45,18 +44,18 @@
 
         if (p_180663_1_.func_175707_a(new BlockPos(k - 2, l - 2, i1 - 2), new BlockPos(k + 2, l + 2, i1 + 2)))
         {
-            for (int j1 = -1; j1 <= 1; j1++)
+            for (int j1 = -1; j1 <= 1; ++j1)
             {
-                for (int k1 = -1; k1 <= 1; k1++)
+                for (int k1 = -1; k1 <= 1; ++k1)
                 {
-                    for (int l1 = -1; l1 <= 1; l1++)
+                    for (int l1 = -1; l1 <= 1; ++l1)
                     {
                         BlockPos blockpos = p_180663_2_.func_177982_a(j1, k1, l1);
                         IBlockState iblockstate = p_180663_1_.func_180495_p(blockpos);
 
-                        if (iblockstate.func_185904_a() == Material.field_151584_j && !iblockstate.func_177229_b(field_176236_b))
+                        if (iblockstate.func_177230_c().isLeaves(iblockstate, p_180663_1_, blockpos))
                         {
-                            p_180663_1_.func_180501_a(blockpos, iblockstate.func_177226_a(field_176236_b, true), 4);
+                            iblockstate.func_177230_c().beginLeavesDecay(iblockstate, p_180663_1_, blockpos);
                         }
                     }
                 }
@@ -64,12 +63,11 @@
         }
     }
 
-    @Override
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
     {
         if (!p_180650_1_.field_72995_K)
         {
-            if (p_180650_3_.func_177229_b(field_176236_b) && p_180650_3_.func_177229_b(field_176237_a))
+            if (((Boolean)p_180650_3_.func_177229_b(field_176236_b)).booleanValue() && ((Boolean)p_180650_3_.func_177229_b(field_176237_a)).booleanValue())
             {
                 int i = 4;
                 int j = 5;
@@ -85,22 +83,23 @@
                     this.field_150128_a = new int[32768];
                 }
 
-                if (p_180650_1_.func_175707_a(new BlockPos(k - 5, l - 5, i1 - 5), new BlockPos(k + 5, l + 5, i1 + 5)))
+                if (!p_180650_1_.func_175697_a(p_180650_2_, 1)) return; // Forge: prevent decaying leaves from updating neighbors and loading unloaded chunks
+                if (p_180650_1_.func_175697_a(p_180650_2_, 6)) // Forge: extend range from 5 to 6 to account for neighbor checks in world.markAndNotifyBlock -> world.updateObservingBlocksAt
                 {
                     BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
 
-                    for (int i2 = -4; i2 <= 4; i2++)
+                    for (int i2 = -4; i2 <= 4; ++i2)
                     {
-                        for (int j2 = -4; j2 <= 4; j2++)
+                        for (int j2 = -4; j2 <= 4; ++j2)
                         {
-                            for (int k2 = -4; k2 <= 4; k2++)
+                            for (int k2 = -4; k2 <= 4; ++k2)
                             {
                                 IBlockState iblockstate = p_180650_1_.func_180495_p(blockpos$mutableblockpos.func_181079_c(k + i2, l + j2, i1 + k2));
                                 Block block = iblockstate.func_177230_c();
 
-                                if (block != Blocks.field_150364_r && block != Blocks.field_150363_s)
+                                if (!block.canSustainLeaves(iblockstate, p_180650_1_, blockpos$mutableblockpos.func_181079_c(k + i2, l + j2, i1 + k2)))
                                 {
-                                    if (iblockstate.func_185904_a() == Material.field_151584_j)
+                                    if (block.isLeaves(iblockstate, p_180650_1_, blockpos$mutableblockpos.func_181079_c(k + i2, l + j2, i1 + k2)))
                                     {
                                         this.field_150128_a[(i2 + 16) * 1024 + (j2 + 16) * 32 + k2 + 16] = -2;
                                     }
@@ -117,13 +116,13 @@
                         }
                     }
 
-                    for (int i3 = 1; i3 <= 4; i3++)
+                    for (int i3 = 1; i3 <= 4; ++i3)
                     {
-                        for (int j3 = -4; j3 <= 4; j3++)
+                        for (int j3 = -4; j3 <= 4; ++j3)
                         {
-                            for (int k3 = -4; k3 <= 4; k3++)
+                            for (int k3 = -4; k3 <= 4; ++k3)
                             {
-                                for (int l3 = -4; l3 <= 4; l3++)
+                                for (int l3 = -4; l3 <= 4; ++l3)
                                 {
                                     if (this.field_150128_a[(j3 + 16) * 1024 + (k3 + 16) * 32 + l3 + 16] == i3 - 1)
                                     {
@@ -167,7 +166,7 @@
 
                 if (l2 >= 0)
                 {
-                    p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_.func_177226_a(field_176236_b, false), 4);
+                    p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_.func_177226_a(field_176236_b, Boolean.valueOf(false)), 4);
                 }
                 else
                 {
@@ -178,17 +177,14 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_180655_c(IBlockState p_180655_1_, World p_180655_2_, BlockPos p_180655_3_, Random p_180655_4_)
     {
-        if (p_180655_2_.func_175727_C(p_180655_3_.func_177984_a())
-                && !p_180655_2_.func_180495_p(p_180655_3_.func_177977_b()).func_185896_q()
-                && p_180655_4_.nextInt(15) == 1)
+        if (p_180655_2_.func_175727_C(p_180655_3_.func_177984_a()) && !p_180655_2_.func_180495_p(p_180655_3_.func_177977_b()).func_185896_q() && p_180655_4_.nextInt(15) == 1)
         {
             double d0 = (double)((float)p_180655_3_.func_177958_n() + p_180655_4_.nextFloat());
-            double d1 = (double)p_180655_3_.func_177956_o() - 0.05;
+            double d1 = (double)p_180655_3_.func_177956_o() - 0.05D;
             double d2 = (double)((float)p_180655_3_.func_177952_p() + p_180655_4_.nextFloat());
-            p_180655_2_.func_175688_a(EnumParticleTypes.DRIP_WATER, d0, d1, d2, 0.0, 0.0, 0.0);
+            p_180655_2_.func_175688_a(EnumParticleTypes.DRIP_WATER, d0, d1, d2, 0.0D, 0.0D, 0.0D);
         }
     }
 
@@ -198,55 +194,19 @@
         p_176235_1_.func_175698_g(p_176235_2_);
     }
 
-    @Override
     public int func_149745_a(Random p_149745_1_)
     {
         return p_149745_1_.nextInt(20) == 0 ? 1 : 0;
     }
 
-    @Override
     public Item func_180660_a(IBlockState p_180660_1_, Random p_180660_2_, int p_180660_3_)
     {
         return Item.func_150898_a(Blocks.field_150345_g);
     }
 
-    @Override
     public void func_180653_a(World p_180653_1_, BlockPos p_180653_2_, IBlockState p_180653_3_, float p_180653_4_, int p_180653_5_)
     {
-        if (!p_180653_1_.field_72995_K)
-        {
-            int i = this.func_176232_d(p_180653_3_);
-
-            if (p_180653_5_ > 0)
-            {
-                i -= 2 << p_180653_5_;
-
-                if (i < 10)
-                {
-                    i = 10;
-                }
-            }
-
-            if (p_180653_1_.field_73012_v.nextInt(i) == 0)
-            {
-                Item item = this.func_180660_a(p_180653_3_, p_180653_1_.field_73012_v, p_180653_5_);
-                func_180635_a(p_180653_1_, p_180653_2_, new ItemStack(item, 1, this.func_180651_a(p_180653_3_)));
-            }
-
-            i = 200;
-
-            if (p_180653_5_ > 0)
-            {
-                i -= 10 << p_180653_5_;
-
-                if (i < 40)
-                {
-                    i = 40;
-                }
-            }
-
-            this.func_176234_a(p_180653_1_, p_180653_2_, p_180653_3_, i);
-        }
+        super.func_180653_a(p_180653_1_, p_180653_2_, p_180653_3_, p_180653_4_, p_180653_5_);
     }
 
     protected void func_176234_a(World p_176234_1_, BlockPos p_176234_2_, IBlockState p_176234_3_, int p_176234_4_)
@@ -258,7 +218,6 @@
         return 20;
     }
 
-    @Override
     public boolean func_149662_c(IBlockState p_149662_1_)
     {
         return !this.field_185686_c;
@@ -271,13 +230,11 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public BlockRenderLayer func_180664_k()
     {
         return this.field_185686_c ? BlockRenderLayer.CUTOUT_MIPPED : BlockRenderLayer.SOLID;
     }
 
-    @Override
     public boolean func_176214_u(IBlockState p_176214_1_)
     {
         return false;
@@ -285,12 +242,54 @@
 
     public abstract BlockPlanks.EnumType func_176233_b(int p_176233_1_);
 
+    @Override public boolean isShearable(ItemStack item, IBlockAccess world, BlockPos pos){ return true; }
+    @Override public boolean isLeaves(IBlockState state, IBlockAccess world, BlockPos pos){ return true; }
+
+    @Override
+    public void beginLeavesDecay(IBlockState state, World world, BlockPos pos)
+    {
+        if (!(Boolean)state.func_177229_b(field_176236_b))
+        {
+            world.func_180501_a(pos, state.func_177226_a(field_176236_b, true), 4);
+        }
+    }
+
+    @Override
+    public void getDrops(net.minecraft.util.NonNullList<ItemStack> drops, IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
+        Random rand = world instanceof World ? ((World)world).field_73012_v : new Random();
+        int chance = this.func_176232_d(state);
+
+        if (fortune > 0)
+        {
+            chance -= 2 << fortune;
+            if (chance < 10) chance = 10;
+        }
+
+        if (rand.nextInt(chance) == 0)
+        {
+            ItemStack drop = new ItemStack(func_180660_a(state, rand, fortune), 1, func_180651_a(state));
+            if (!drop.func_190926_b())
+                drops.add(drop);
+        }
+
+        chance = 200;
+        if (fortune > 0)
+        {
+            chance -= 10 << fortune;
+            if (chance < 40) chance = 40;
+        }
+
+        this.captureDrops(true);
+        if (world instanceof World)
+            this.func_176234_a((World)world, pos, state, chance); // Dammet mojang
+        drops.addAll(this.captureDrops(false));
+    }
+
+
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_176225_a(IBlockState p_176225_1_, IBlockAccess p_176225_2_, BlockPos p_176225_3_, EnumFacing p_176225_4_)
     {
-        return !this.field_185686_c && p_176225_2_.func_180495_p(p_176225_3_.func_177972_a(p_176225_4_)).func_177230_c() == this
-               ? false
-               : super.func_176225_a(p_176225_1_, p_176225_2_, p_176225_3_, p_176225_4_);
+        return !this.field_185686_c && p_176225_2_.func_180495_p(p_176225_3_.func_177972_a(p_176225_4_)).func_177230_c() == this ? false : super.func_176225_a(p_176225_1_, p_176225_2_, p_176225_3_, p_176225_4_);
     }
 }
