--- before/net/minecraft/block/BlockVine.java
+++ after/net/minecraft/block/BlockVine.java
@@ -3,6 +3,7 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
@@ -27,7 +28,7 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class BlockVine extends Block
+public class BlockVine extends Block implements net.minecraftforge.common.IShearable
 {
     public static final PropertyBool field_176277_a = PropertyBool.func_177716_a("up");
     public static final PropertyBool field_176273_b = PropertyBool.func_177716_a("north");
@@ -35,67 +36,57 @@
     public static final PropertyBool field_176279_N = PropertyBool.func_177716_a("south");
     public static final PropertyBool field_176280_O = PropertyBool.func_177716_a("west");
     public static final PropertyBool[] field_176274_P = new PropertyBool[] {field_176277_a, field_176273_b, field_176279_N, field_176280_O, field_176278_M};
-    protected static final AxisAlignedBB field_185757_g = new AxisAlignedBB(0.0, 0.9375, 0.0, 1.0, 1.0, 1.0);
-    protected static final AxisAlignedBB field_185753_B = new AxisAlignedBB(0.0, 0.0, 0.0, 0.0625, 1.0, 1.0);
-    protected static final AxisAlignedBB field_185754_C = new AxisAlignedBB(0.9375, 0.0, 0.0, 1.0, 1.0, 1.0);
-    protected static final AxisAlignedBB field_185755_D = new AxisAlignedBB(0.0, 0.0, 0.0, 1.0, 1.0, 0.0625);
-    protected static final AxisAlignedBB field_185756_E = new AxisAlignedBB(0.0, 0.0, 0.9375, 1.0, 1.0, 1.0);
+    protected static final AxisAlignedBB field_185757_g = new AxisAlignedBB(0.0D, 0.9375D, 0.0D, 1.0D, 1.0D, 1.0D);
+    protected static final AxisAlignedBB field_185753_B = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 0.0625D, 1.0D, 1.0D);
+    protected static final AxisAlignedBB field_185754_C = new AxisAlignedBB(0.9375D, 0.0D, 0.0D, 1.0D, 1.0D, 1.0D);
+    protected static final AxisAlignedBB field_185755_D = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 1.0D, 1.0D, 0.0625D);
+    protected static final AxisAlignedBB field_185756_E = new AxisAlignedBB(0.0D, 0.0D, 0.9375D, 1.0D, 1.0D, 1.0D);
 
     public BlockVine()
     {
         super(Material.field_151582_l);
-        this.func_180632_j(
-            this.field_176227_L
-            .func_177621_b()
-            .func_177226_a(field_176277_a, false)
-            .func_177226_a(field_176273_b, false)
-            .func_177226_a(field_176278_M, false)
-            .func_177226_a(field_176279_N, false)
-            .func_177226_a(field_176280_O, false)
-        );
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176277_a, Boolean.valueOf(false)).func_177226_a(field_176273_b, Boolean.valueOf(false)).func_177226_a(field_176278_M, Boolean.valueOf(false)).func_177226_a(field_176279_N, Boolean.valueOf(false)).func_177226_a(field_176280_O, Boolean.valueOf(false)));
         this.func_149675_a(true);
         this.func_149647_a(CreativeTabs.field_78031_c);
     }
 
     @Nullable
-    @Override
     public AxisAlignedBB func_180646_a(IBlockState p_180646_1_, IBlockAccess p_180646_2_, BlockPos p_180646_3_)
     {
         return field_185506_k;
     }
 
-    @Override
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
         p_185496_1_ = p_185496_1_.func_185899_b(p_185496_2_, p_185496_3_);
         int i = 0;
         AxisAlignedBB axisalignedbb = field_185505_j;
 
-        if (p_185496_1_.func_177229_b(field_176277_a))
+        if (((Boolean)p_185496_1_.func_177229_b(field_176277_a)).booleanValue())
         {
             axisalignedbb = field_185757_g;
             ++i;
         }
 
-        if (p_185496_1_.func_177229_b(field_176273_b))
+        if (((Boolean)p_185496_1_.func_177229_b(field_176273_b)).booleanValue())
         {
             axisalignedbb = field_185755_D;
             ++i;
         }
 
-        if (p_185496_1_.func_177229_b(field_176278_M))
+        if (((Boolean)p_185496_1_.func_177229_b(field_176278_M)).booleanValue())
         {
             axisalignedbb = field_185754_C;
             ++i;
         }
 
-        if (p_185496_1_.func_177229_b(field_176279_N))
+        if (((Boolean)p_185496_1_.func_177229_b(field_176279_N)).booleanValue())
         {
             axisalignedbb = field_185756_E;
             ++i;
         }
 
-        if (p_185496_1_.func_177229_b(field_176280_O))
+        if (((Boolean)p_185496_1_.func_177229_b(field_176280_O)).booleanValue())
         {
             axisalignedbb = field_185753_B;
             ++i;
@@ -104,34 +95,27 @@
         return i == 1 ? axisalignedbb : field_185505_j;
     }
 
-    @Override
     public IBlockState func_176221_a(IBlockState p_176221_1_, IBlockAccess p_176221_2_, BlockPos p_176221_3_)
     {
         BlockPos blockpos = p_176221_3_.func_177984_a();
-        return p_176221_1_.func_177226_a(
-                   field_176277_a, p_176221_2_.func_180495_p(blockpos).func_193401_d(p_176221_2_, blockpos, EnumFacing.DOWN) == BlockFaceShape.SOLID
-               );
+        return p_176221_1_.func_177226_a(field_176277_a, Boolean.valueOf(p_176221_2_.func_180495_p(blockpos).func_193401_d(p_176221_2_, blockpos, EnumFacing.DOWN) == BlockFaceShape.SOLID));
     }
 
-    @Override
     public boolean func_149662_c(IBlockState p_149662_1_)
     {
         return false;
     }
 
-    @Override
     public boolean func_149686_d(IBlockState p_149686_1_)
     {
         return false;
     }
 
-    @Override
     public boolean func_176200_f(IBlockAccess p_176200_1_, BlockPos p_176200_2_)
     {
         return true;
     }
 
-    @Override
     public boolean func_176198_a(World p_176198_1_, BlockPos p_176198_2_, EnumFacing p_176198_3_)
     {
         return p_176198_3_ != EnumFacing.DOWN && p_176198_3_ != EnumFacing.UP && this.func_193395_a(p_176198_1_, p_176198_2_, p_176198_3_);
@@ -140,8 +124,7 @@
     public boolean func_193395_a(World p_193395_1_, BlockPos p_193395_2_, EnumFacing p_193395_3_)
     {
         Block block = p_193395_1_.func_180495_p(p_193395_2_.func_177984_a()).func_177230_c();
-        return this.func_193396_c(p_193395_1_, p_193395_2_.func_177972_a(p_193395_3_.func_176734_d()), p_193395_3_)
-               && (block == Blocks.field_150350_a || block == Blocks.field_150395_bd || this.func_193396_c(p_193395_1_, p_193395_2_.func_177984_a(), EnumFacing.UP));
+        return this.func_193396_c(p_193395_1_, p_193395_2_.func_177972_a(p_193395_3_.func_176734_d()), p_193395_3_) && (block == Blocks.field_150350_a || block == Blocks.field_150395_bd || this.func_193396_c(p_193395_1_, p_193395_2_.func_177984_a(), EnumFacing.UP));
     }
 
     private boolean func_193396_c(World p_193396_1_, BlockPos p_193396_2_, EnumFacing p_193396_3_)
@@ -152,15 +135,7 @@
 
     protected static boolean func_193397_e(Block p_193397_0_)
     {
-        return p_193397_0_ instanceof BlockShulkerBox
-               || p_193397_0_ == Blocks.field_150461_bJ
-               || p_193397_0_ == Blocks.field_150383_bp
-               || p_193397_0_ == Blocks.field_150359_w
-               || p_193397_0_ == Blocks.field_150399_cn
-               || p_193397_0_ == Blocks.field_150331_J
-               || p_193397_0_ == Blocks.field_150320_F
-               || p_193397_0_ == Blocks.field_150332_K
-               || p_193397_0_ == Blocks.field_150415_aT;
+        return p_193397_0_ instanceof BlockShulkerBox || p_193397_0_ == Blocks.field_150461_bJ || p_193397_0_ == Blocks.field_150383_bp || p_193397_0_ == Blocks.field_150359_w || p_193397_0_ == Blocks.field_150399_cn || p_193397_0_ == Blocks.field_150331_J || p_193397_0_ == Blocks.field_150320_F || p_193397_0_ == Blocks.field_150332_K || p_193397_0_ == Blocks.field_150415_aT;
     }
 
     private boolean func_176269_e(World p_176269_1_, BlockPos p_176269_2_, IBlockState p_176269_3_)
@@ -171,13 +146,13 @@
         {
             PropertyBool propertybool = func_176267_a(enumfacing);
 
-            if (p_176269_3_.func_177229_b(propertybool) && !this.func_193395_a(p_176269_1_, p_176269_2_, enumfacing.func_176734_d()))
+            if (((Boolean)p_176269_3_.func_177229_b(propertybool)).booleanValue() && !this.func_193395_a(p_176269_1_, p_176269_2_, enumfacing.func_176734_d()))
             {
                 IBlockState iblockstate1 = p_176269_1_.func_180495_p(p_176269_2_.func_177984_a());
 
-                if (iblockstate1.func_177230_c() != this || !iblockstate1.func_177229_b(propertybool))
+                if (iblockstate1.func_177230_c() != this || !((Boolean)iblockstate1.func_177229_b(propertybool)).booleanValue())
                 {
-                    p_176269_3_ = p_176269_3_.func_177226_a(propertybool, false);
+                    p_176269_3_ = p_176269_3_.func_177226_a(propertybool, Boolean.valueOf(false));
                 }
             }
         }
@@ -197,7 +172,6 @@
         }
     }
 
-    @Override
     public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
     {
         if (!p_189540_2_.field_72995_K && !this.func_176269_e(p_189540_2_, p_189540_3_, p_189540_1_))
@@ -207,17 +181,16 @@
         }
     }
 
-    @Override
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
     {
         if (!p_180650_1_.field_72995_K)
         {
-            if (p_180650_1_.field_73012_v.nextInt(4) == 0)
+            if (p_180650_1_.field_73012_v.nextInt(4) == 0 && p_180650_1_.func_175697_a(p_180650_2_, 4)) // Forge: check area to prevent loading unloaded chunks
             {
                 int i = 4;
                 int j = 5;
                 boolean flag = false;
-                label179:
+                label181:
 
                 for (int k = -4; k <= 4; ++k)
                 {
@@ -227,173 +200,149 @@
                         {
                             if (p_180650_1_.func_180495_p(p_180650_2_.func_177982_a(k, i1, l)).func_177230_c() == this)
                             {
-                                if (--j <= 0)
+                                --j;
+
+                                if (j <= 0)
                                 {
                                     flag = true;
-                                    break label179;
+                                    break label181;
                                 }
                             }
                         }
                     }
                 }
 
-                EnumFacing enumfacing2 = EnumFacing.func_176741_a(p_180650_4_);
+                EnumFacing enumfacing1 = EnumFacing.func_176741_a(p_180650_4_);
                 BlockPos blockpos2 = p_180650_2_.func_177984_a();
 
-                if (enumfacing2 == EnumFacing.UP && p_180650_2_.func_177956_o() < 255 && p_180650_1_.func_175623_d(blockpos2))
+                if (enumfacing1 == EnumFacing.UP && p_180650_2_.func_177956_o() < 255 && p_180650_1_.func_175623_d(blockpos2))
                 {
-                    IBlockState iblockstate1 = p_180650_3_;
+                    IBlockState iblockstate2 = p_180650_3_;
 
-                    for (EnumFacing enumfacing3 : EnumFacing.Plane.HORIZONTAL)
+                    for (EnumFacing enumfacing2 : EnumFacing.Plane.HORIZONTAL)
                     {
-                        if (p_180650_4_.nextBoolean() && this.func_193395_a(p_180650_1_, blockpos2, enumfacing3.func_176734_d()))
+                        if (p_180650_4_.nextBoolean() && this.func_193395_a(p_180650_1_, blockpos2, enumfacing2.func_176734_d()))
                         {
-                            iblockstate1 = iblockstate1.func_177226_a(func_176267_a(enumfacing3), true);
+                            iblockstate2 = iblockstate2.func_177226_a(func_176267_a(enumfacing2), Boolean.valueOf(true));
                         }
                         else
                         {
-                            iblockstate1 = iblockstate1.func_177226_a(func_176267_a(enumfacing3), false);
-                        }
-                    }
-
-                    if (iblockstate1.func_177229_b(field_176273_b)
-                            || iblockstate1.func_177229_b(field_176278_M)
-                            || iblockstate1.func_177229_b(field_176279_N)
-                            || iblockstate1.func_177229_b(field_176280_O))
-                    {
-                        p_180650_1_.func_180501_a(blockpos2, iblockstate1, 2);
-                    }
-                }
-                else if (!enumfacing2.func_176740_k().func_176722_c() || p_180650_3_.func_177229_b(func_176267_a(enumfacing2)))
+                            iblockstate2 = iblockstate2.func_177226_a(func_176267_a(enumfacing2), Boolean.valueOf(false));
+                        }
+                    }
+
+                    if (((Boolean)iblockstate2.func_177229_b(field_176273_b)).booleanValue() || ((Boolean)iblockstate2.func_177229_b(field_176278_M)).booleanValue() || ((Boolean)iblockstate2.func_177229_b(field_176279_N)).booleanValue() || ((Boolean)iblockstate2.func_177229_b(field_176280_O)).booleanValue())
+                    {
+                        p_180650_1_.func_180501_a(blockpos2, iblockstate2, 2);
+                    }
+                }
+                else if (enumfacing1.func_176740_k().func_176722_c() && !((Boolean)p_180650_3_.func_177229_b(func_176267_a(enumfacing1))).booleanValue())
+                {
+                    if (!flag)
+                    {
+                        BlockPos blockpos4 = p_180650_2_.func_177972_a(enumfacing1);
+                        IBlockState iblockstate3 = p_180650_1_.func_180495_p(blockpos4);
+                        Block block1 = iblockstate3.func_177230_c();
+
+                        if (block1.isAir(iblockstate3, p_180650_1_, blockpos4))
+                        {
+                            EnumFacing enumfacing3 = enumfacing1.func_176746_e();
+                            EnumFacing enumfacing4 = enumfacing1.func_176735_f();
+                            boolean flag1 = ((Boolean)p_180650_3_.func_177229_b(func_176267_a(enumfacing3))).booleanValue();
+                            boolean flag2 = ((Boolean)p_180650_3_.func_177229_b(func_176267_a(enumfacing4))).booleanValue();
+                            BlockPos blockpos = blockpos4.func_177972_a(enumfacing3);
+                            BlockPos blockpos1 = blockpos4.func_177972_a(enumfacing4);
+
+                            if (flag1 && this.func_193395_a(p_180650_1_, blockpos.func_177972_a(enumfacing3), enumfacing3))
+                            {
+                                p_180650_1_.func_180501_a(blockpos4, this.func_176223_P().func_177226_a(func_176267_a(enumfacing3), Boolean.valueOf(true)), 2);
+                            }
+                            else if (flag2 && this.func_193395_a(p_180650_1_, blockpos1.func_177972_a(enumfacing4), enumfacing4))
+                            {
+                                p_180650_1_.func_180501_a(blockpos4, this.func_176223_P().func_177226_a(func_176267_a(enumfacing4), Boolean.valueOf(true)), 2);
+                            }
+                            else if (flag1 && p_180650_1_.func_175623_d(blockpos) && this.func_193395_a(p_180650_1_, blockpos, enumfacing1))
+                            {
+                                p_180650_1_.func_180501_a(blockpos, this.func_176223_P().func_177226_a(func_176267_a(enumfacing1.func_176734_d()), Boolean.valueOf(true)), 2);
+                            }
+                            else if (flag2 && p_180650_1_.func_175623_d(blockpos1) && this.func_193395_a(p_180650_1_, blockpos1, enumfacing1))
+                            {
+                                p_180650_1_.func_180501_a(blockpos1, this.func_176223_P().func_177226_a(func_176267_a(enumfacing1.func_176734_d()), Boolean.valueOf(true)), 2);
+                            }
+                        }
+                        else if (iblockstate3.func_193401_d(p_180650_1_, blockpos4, enumfacing1) == BlockFaceShape.SOLID)
+                        {
+                            p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_.func_177226_a(func_176267_a(enumfacing1), Boolean.valueOf(true)), 2);
+                        }
+                    }
+                }
+                else
                 {
                     if (p_180650_2_.func_177956_o() > 1)
                     {
-                        BlockPos blockpos4 = p_180650_2_.func_177977_b();
-                        IBlockState iblockstate2 = p_180650_1_.func_180495_p(blockpos4);
-                        Block block1 = iblockstate2.func_177230_c();
+                        BlockPos blockpos3 = p_180650_2_.func_177977_b();
+                        IBlockState iblockstate = p_180650_1_.func_180495_p(blockpos3);
+                        Block block = iblockstate.func_177230_c();
 
-                        if (block1.field_149764_J == Material.field_151579_a)
+                        if (block.field_149764_J == Material.field_151579_a)
                         {
-                            IBlockState iblockstate3 = p_180650_3_;
+                            IBlockState iblockstate1 = p_180650_3_;
 
-                            for (EnumFacing enumfacing4 : EnumFacing.Plane.HORIZONTAL)
+                            for (EnumFacing enumfacing : EnumFacing.Plane.HORIZONTAL)
                             {
                                 if (p_180650_4_.nextBoolean())
                                 {
-                                    iblockstate3 = iblockstate3.func_177226_a(func_176267_a(enumfacing4), false);
+                                    iblockstate1 = iblockstate1.func_177226_a(func_176267_a(enumfacing), Boolean.valueOf(false));
                                 }
                             }
 
-                            if (iblockstate3.func_177229_b(field_176273_b)
-                                    || iblockstate3.func_177229_b(field_176278_M)
-                                    || iblockstate3.func_177229_b(field_176279_N)
-                                    || iblockstate3.func_177229_b(field_176280_O))
+                            if (((Boolean)iblockstate1.func_177229_b(field_176273_b)).booleanValue() || ((Boolean)iblockstate1.func_177229_b(field_176278_M)).booleanValue() || ((Boolean)iblockstate1.func_177229_b(field_176279_N)).booleanValue() || ((Boolean)iblockstate1.func_177229_b(field_176280_O)).booleanValue())
                             {
-                                p_180650_1_.func_180501_a(blockpos4, iblockstate3, 2);
+                                p_180650_1_.func_180501_a(blockpos3, iblockstate1, 2);
                             }
                         }
-                        else if (block1 == this)
+                        else if (block == this)
                         {
-                            IBlockState iblockstate4 = iblockstate2;
+                            IBlockState iblockstate4 = iblockstate;
 
                             for (EnumFacing enumfacing5 : EnumFacing.Plane.HORIZONTAL)
                             {
                                 PropertyBool propertybool = func_176267_a(enumfacing5);
 
-                                if (p_180650_4_.nextBoolean() && p_180650_3_.func_177229_b(propertybool))
+                                if (p_180650_4_.nextBoolean() && ((Boolean)p_180650_3_.func_177229_b(propertybool)).booleanValue())
                                 {
-                                    iblockstate4 = iblockstate4.func_177226_a(propertybool, true);
+                                    iblockstate4 = iblockstate4.func_177226_a(propertybool, Boolean.valueOf(true));
                                 }
                             }
 
-                            if (iblockstate4.func_177229_b(field_176273_b)
-                                    || iblockstate4.func_177229_b(field_176278_M)
-                                    || iblockstate4.func_177229_b(field_176279_N)
-                                    || iblockstate4.func_177229_b(field_176280_O))
+                            if (((Boolean)iblockstate4.func_177229_b(field_176273_b)).booleanValue() || ((Boolean)iblockstate4.func_177229_b(field_176278_M)).booleanValue() || ((Boolean)iblockstate4.func_177229_b(field_176279_N)).booleanValue() || ((Boolean)iblockstate4.func_177229_b(field_176280_O)).booleanValue())
                             {
-                                p_180650_1_.func_180501_a(blockpos4, iblockstate4, 2);
+                                p_180650_1_.func_180501_a(blockpos3, iblockstate4, 2);
                             }
                         }
-                    }
-                }
-                else if (!flag)
-                {
-                    BlockPos blockpos3 = p_180650_2_.func_177972_a(enumfacing2);
-                    IBlockState iblockstate = p_180650_1_.func_180495_p(blockpos3);
-                    Block block = iblockstate.func_177230_c();
-
-                    if (block.field_149764_J == Material.field_151579_a)
-                    {
-                        EnumFacing enumfacing = enumfacing2.func_176746_e();
-                        EnumFacing enumfacing1 = enumfacing2.func_176735_f();
-                        boolean flag1 = p_180650_3_.func_177229_b(func_176267_a(enumfacing));
-                        boolean flag2 = p_180650_3_.func_177229_b(func_176267_a(enumfacing1));
-                        BlockPos blockpos = blockpos3.func_177972_a(enumfacing);
-                        BlockPos blockpos1 = blockpos3.func_177972_a(enumfacing1);
-
-                        if (flag1 && this.func_193395_a(p_180650_1_, blockpos.func_177972_a(enumfacing), enumfacing))
-                        {
-                            p_180650_1_.func_180501_a(blockpos3, this.func_176223_P().func_177226_a(func_176267_a(enumfacing), true), 2);
-                        }
-                        else if (flag2 && this.func_193395_a(p_180650_1_, blockpos1.func_177972_a(enumfacing1), enumfacing1))
-                        {
-                            p_180650_1_.func_180501_a(blockpos3, this.func_176223_P().func_177226_a(func_176267_a(enumfacing1), true), 2);
-                        }
-                        else if (flag1 && p_180650_1_.func_175623_d(blockpos) && this.func_193395_a(p_180650_1_, blockpos, enumfacing2))
-                        {
-                            p_180650_1_.func_180501_a(blockpos, this.func_176223_P().func_177226_a(func_176267_a(enumfacing2.func_176734_d()), true), 2);
-                        }
-                        else if (flag2 && p_180650_1_.func_175623_d(blockpos1) && this.func_193395_a(p_180650_1_, blockpos1, enumfacing2))
-                        {
-                            p_180650_1_.func_180501_a(blockpos1, this.func_176223_P().func_177226_a(func_176267_a(enumfacing2.func_176734_d()), true), 2);
-                        }
-                    }
-                    else if (iblockstate.func_193401_d(p_180650_1_, blockpos3, enumfacing2) == BlockFaceShape.SOLID)
-                    {
-                        p_180650_1_.func_180501_a(p_180650_2_, p_180650_3_.func_177226_a(func_176267_a(enumfacing2), true), 2);
                     }
                 }
             }
         }
     }
 
-    @Override
-    public IBlockState func_180642_a(
-        World p_180642_1_,
-        BlockPos p_180642_2_,
-        EnumFacing p_180642_3_,
-        float p_180642_4_,
-        float p_180642_5_,
-        float p_180642_6_,
-        int p_180642_7_,
-        EntityLivingBase p_180642_8_
-    )
+    public IBlockState func_180642_a(World p_180642_1_, BlockPos p_180642_2_, EnumFacing p_180642_3_, float p_180642_4_, float p_180642_5_, float p_180642_6_, int p_180642_7_, EntityLivingBase p_180642_8_)
     {
-        IBlockState iblockstate = this.func_176223_P()
-                                  .func_177226_a(field_176277_a, false)
-                                  .func_177226_a(field_176273_b, false)
-                                  .func_177226_a(field_176278_M, false)
-                                  .func_177226_a(field_176279_N, false)
-                                  .func_177226_a(field_176280_O, false);
-        return p_180642_3_.func_176740_k().func_176722_c() ? iblockstate.func_177226_a(func_176267_a(p_180642_3_.func_176734_d()), true) : iblockstate;
+        IBlockState iblockstate = this.func_176223_P().func_177226_a(field_176277_a, Boolean.valueOf(false)).func_177226_a(field_176273_b, Boolean.valueOf(false)).func_177226_a(field_176278_M, Boolean.valueOf(false)).func_177226_a(field_176279_N, Boolean.valueOf(false)).func_177226_a(field_176280_O, Boolean.valueOf(false));
+        return p_180642_3_.func_176740_k().func_176722_c() ? iblockstate.func_177226_a(func_176267_a(p_180642_3_.func_176734_d()), Boolean.valueOf(true)) : iblockstate;
     }
 
-    @Override
     public Item func_180660_a(IBlockState p_180660_1_, Random p_180660_2_, int p_180660_3_)
     {
         return Items.field_190931_a;
     }
 
-    @Override
     public int func_149745_a(Random p_149745_1_)
     {
         return 0;
     }
 
-    @Override
-    public void func_180657_a(
-        World p_180657_1_, EntityPlayer p_180657_2_, BlockPos p_180657_3_, IBlockState p_180657_4_, @Nullable TileEntity p_180657_5_, ItemStack p_180657_6_
-    )
+    public void func_180657_a(World p_180657_1_, EntityPlayer p_180657_2_, BlockPos p_180657_3_, IBlockState p_180657_4_, @Nullable TileEntity p_180657_5_, ItemStack p_180657_6_)
     {
         if (!p_180657_1_.field_72995_K && p_180657_6_.func_77973_b() == Items.field_151097_aZ)
         {
@@ -406,44 +355,37 @@
         }
     }
 
+    public IBlockState func_176203_a(int p_176203_1_)
+    {
+        return this.func_176223_P().func_177226_a(field_176279_N, Boolean.valueOf((p_176203_1_ & 1) > 0)).func_177226_a(field_176280_O, Boolean.valueOf((p_176203_1_ & 2) > 0)).func_177226_a(field_176273_b, Boolean.valueOf((p_176203_1_ & 4) > 0)).func_177226_a(field_176278_M, Boolean.valueOf((p_176203_1_ & 8) > 0));
+    }
+
     @SideOnly(Side.CLIENT)
-    @Override
     public BlockRenderLayer func_180664_k()
     {
         return BlockRenderLayer.CUTOUT;
     }
 
-    @Override
-    public IBlockState func_176203_a(int p_176203_1_)
-    {
-        return this.func_176223_P()
-               .func_177226_a(field_176279_N, (p_176203_1_ & 1) > 0)
-               .func_177226_a(field_176280_O, (p_176203_1_ & 2) > 0)
-               .func_177226_a(field_176273_b, (p_176203_1_ & 4) > 0)
-               .func_177226_a(field_176278_M, (p_176203_1_ & 8) > 0);
-    }
-
-    @Override
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
 
-        if (p_176201_1_.func_177229_b(field_176279_N))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176279_N)).booleanValue())
         {
             i |= 1;
         }
 
-        if (p_176201_1_.func_177229_b(field_176280_O))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176280_O)).booleanValue())
         {
             i |= 2;
         }
 
-        if (p_176201_1_.func_177229_b(field_176273_b))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176273_b)).booleanValue())
         {
             i |= 4;
         }
 
-        if (p_176201_1_.func_177229_b(field_176278_M))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176278_M)).booleanValue())
         {
             i |= 8;
         }
@@ -451,48 +393,34 @@
         return i;
     }
 
-    @Override
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176277_a, field_176273_b, field_176278_M, field_176279_N, field_176280_O);
+        return new BlockStateContainer(this, new IProperty[] {field_176277_a, field_176273_b, field_176278_M, field_176279_N, field_176280_O});
     }
 
-    @Override
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
         switch (p_185499_2_)
         {
             case CLOCKWISE_180:
-                return p_185499_1_.func_177226_a(field_176273_b, p_185499_1_.func_177229_b(field_176279_N))
-                       .func_177226_a(field_176278_M, p_185499_1_.func_177229_b(field_176280_O))
-                       .func_177226_a(field_176279_N, p_185499_1_.func_177229_b(field_176273_b))
-                       .func_177226_a(field_176280_O, p_185499_1_.func_177229_b(field_176278_M));
+                return p_185499_1_.func_177226_a(field_176273_b, p_185499_1_.func_177229_b(field_176279_N)).func_177226_a(field_176278_M, p_185499_1_.func_177229_b(field_176280_O)).func_177226_a(field_176279_N, p_185499_1_.func_177229_b(field_176273_b)).func_177226_a(field_176280_O, p_185499_1_.func_177229_b(field_176278_M));
             case COUNTERCLOCKWISE_90:
-                return p_185499_1_.func_177226_a(field_176273_b, p_185499_1_.func_177229_b(field_176278_M))
-                       .func_177226_a(field_176278_M, p_185499_1_.func_177229_b(field_176279_N))
-                       .func_177226_a(field_176279_N, p_185499_1_.func_177229_b(field_176280_O))
-                       .func_177226_a(field_176280_O, p_185499_1_.func_177229_b(field_176273_b));
+                return p_185499_1_.func_177226_a(field_176273_b, p_185499_1_.func_177229_b(field_176278_M)).func_177226_a(field_176278_M, p_185499_1_.func_177229_b(field_176279_N)).func_177226_a(field_176279_N, p_185499_1_.func_177229_b(field_176280_O)).func_177226_a(field_176280_O, p_185499_1_.func_177229_b(field_176273_b));
             case CLOCKWISE_90:
-                return p_185499_1_.func_177226_a(field_176273_b, p_185499_1_.func_177229_b(field_176280_O))
-                       .func_177226_a(field_176278_M, p_185499_1_.func_177229_b(field_176273_b))
-                       .func_177226_a(field_176279_N, p_185499_1_.func_177229_b(field_176278_M))
-                       .func_177226_a(field_176280_O, p_185499_1_.func_177229_b(field_176279_N));
+                return p_185499_1_.func_177226_a(field_176273_b, p_185499_1_.func_177229_b(field_176280_O)).func_177226_a(field_176278_M, p_185499_1_.func_177229_b(field_176273_b)).func_177226_a(field_176279_N, p_185499_1_.func_177229_b(field_176278_M)).func_177226_a(field_176280_O, p_185499_1_.func_177229_b(field_176279_N));
             default:
                 return p_185499_1_;
         }
     }
 
-    @Override
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
         switch (p_185471_2_)
         {
             case LEFT_RIGHT:
-                return p_185471_1_.func_177226_a(field_176273_b, p_185471_1_.func_177229_b(field_176279_N))
-                       .func_177226_a(field_176279_N, p_185471_1_.func_177229_b(field_176273_b));
+                return p_185471_1_.func_177226_a(field_176273_b, p_185471_1_.func_177229_b(field_176279_N)).func_177226_a(field_176279_N, p_185471_1_.func_177229_b(field_176273_b));
             case FRONT_BACK:
-                return p_185471_1_.func_177226_a(field_176278_M, p_185471_1_.func_177229_b(field_176280_O))
-                       .func_177226_a(field_176280_O, p_185471_1_.func_177229_b(field_176278_M));
+                return p_185471_1_.func_177226_a(field_176278_M, p_185471_1_.func_177229_b(field_176280_O)).func_177226_a(field_176280_O, p_185471_1_.func_177229_b(field_176278_M));
             default:
                 return super.func_185471_a(p_185471_1_, p_185471_2_);
         }
@@ -523,7 +451,7 @@
 
         for (PropertyBool propertybool : field_176274_P)
         {
-            if (p_176268_0_.func_177229_b(propertybool))
+            if (((Boolean)p_176268_0_.func_177229_b(propertybool)).booleanValue())
             {
                 ++i;
             }
@@ -531,8 +459,17 @@
 
         return i;
     }
-
+    /*************************FORGE START***********************************/
+    @Override public boolean isLadder(IBlockState state, IBlockAccess world, BlockPos pos, EntityLivingBase entity){ return true; }
+    @Override public boolean isShearable(ItemStack item, IBlockAccess world, BlockPos pos){ return true; }
     @Override
+    public java.util.List<ItemStack> onSheared(ItemStack item, IBlockAccess world, BlockPos pos, int fortune)
+    {
+        return java.util.Arrays.asList(new ItemStack(this, 1));
+    }
+    /*************************FORGE END***********************************/
+
+
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
     {
         return BlockFaceShape.UNDEFINED;
