--- before/net/minecraft/block/BlockRailPowered.java
+++ after/net/minecraft/block/BlockRailPowered.java
@@ -29,10 +29,18 @@
     }
             );
     public static final PropertyBool POWERED = PropertyBool.create("powered");
+    
+    private final boolean isActivator;
 
     protected BlockRailPowered()
     {
+        this(false);
+    }
+    
+    protected BlockRailPowered(boolean isActivator)
+    {
         super(true);
+        this.isActivator = isActivator;
         this.setDefaultState(
             this.blockState.getBaseState().withProperty(SHAPE, BlockRailBase.EnumRailDirection.NORTH_SOUTH).withProperty(POWERED, false)
         );
@@ -149,7 +157,7 @@
     {
         IBlockState iblockstate = worldIn.getBlockState(pos);
 
-        if (iblockstate.getBlock() != this)
+        if (!(iblockstate.getBlock() instanceof BlockRailPowered) || isActivator != ((BlockRailPowered)iblockstate.getBlock()).isActivator)
         {
             return false;
         }
