--- before/net/minecraft/block/BlockRailPowered.java
+++ after/net/minecraft/block/BlockRailPowered.java
@@ -14,30 +14,30 @@
 
 public class BlockRailPowered extends BlockRailBase
 {
-    public static final PropertyEnum<BlockRailBase.EnumRailDirection> field_176568_b = PropertyEnum.func_177708_a(
-                    "shape",
-                    BlockRailBase.EnumRailDirection.class,
-                    new Predicate<BlockRailBase.EnumRailDirection>()
+    public static final PropertyEnum<BlockRailBase.EnumRailDirection> field_176568_b = PropertyEnum.<BlockRailBase.EnumRailDirection>func_177708_a("shape", BlockRailBase.EnumRailDirection.class, new Predicate<BlockRailBase.EnumRailDirection>()
     {
         public boolean apply(@Nullable BlockRailBase.EnumRailDirection p_apply_1_)
         {
-            return p_apply_1_ != BlockRailBase.EnumRailDirection.NORTH_EAST
-                   && p_apply_1_ != BlockRailBase.EnumRailDirection.NORTH_WEST
-                   && p_apply_1_ != BlockRailBase.EnumRailDirection.SOUTH_EAST
-                   && p_apply_1_ != BlockRailBase.EnumRailDirection.SOUTH_WEST;
+            return p_apply_1_ != BlockRailBase.EnumRailDirection.NORTH_EAST && p_apply_1_ != BlockRailBase.EnumRailDirection.NORTH_WEST && p_apply_1_ != BlockRailBase.EnumRailDirection.SOUTH_EAST && p_apply_1_ != BlockRailBase.EnumRailDirection.SOUTH_WEST;
         }
-    }
-            );
+    });
     public static final PropertyBool field_176569_M = PropertyBool.func_177716_a("powered");
 
+    private final boolean isActivator;
+
     protected BlockRailPowered()
     {
+        this(false);
+    }
+
+    protected BlockRailPowered(boolean isActivator)
+    {
         super(true);
-        this.func_180632_j(
-            this.field_176227_L.func_177621_b().func_177226_a(field_176568_b, BlockRailBase.EnumRailDirection.NORTH_SOUTH).func_177226_a(field_176569_M, false)
-        );
+        this.isActivator = isActivator;
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176568_b, BlockRailBase.EnumRailDirection.NORTH_SOUTH).func_177226_a(field_176569_M, Boolean.valueOf(false)));
     }
 
+    @SuppressWarnings("incomplete-switch")
     protected boolean func_176566_a(World p_176566_1_, BlockPos p_176566_2_, IBlockState p_176566_3_, boolean p_176566_4_, int p_176566_5_)
     {
         if (p_176566_5_ >= 8)
@@ -50,7 +50,7 @@
             int j = p_176566_2_.func_177956_o();
             int k = p_176566_2_.func_177952_p();
             boolean flag = true;
-            BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = p_176566_3_.func_177229_b(field_176568_b);
+            BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = (BlockRailBase.EnumRailDirection)p_176566_3_.func_177229_b(field_176568_b);
 
             switch (blockrailbase$enumraildirection)
             {
@@ -58,11 +58,11 @@
 
                     if (p_176566_4_)
                     {
-                        k++;
+                        ++k;
                     }
                     else
                     {
-                        k--;
+                        --k;
                     }
 
                     break;
@@ -70,11 +70,11 @@
 
                     if (p_176566_4_)
                     {
-                        i--;
+                        --i;
                     }
                     else
                     {
-                        i++;
+                        ++i;
                     }
 
                     break;
@@ -82,12 +82,12 @@
 
                     if (p_176566_4_)
                     {
-                        i--;
+                        --i;
                     }
                     else
                     {
-                        i++;
-                        j++;
+                        ++i;
+                        ++j;
                         flag = false;
                     }
 
@@ -97,13 +97,13 @@
 
                     if (p_176566_4_)
                     {
-                        i--;
-                        j++;
+                        --i;
+                        ++j;
                         flag = false;
                     }
                     else
                     {
-                        i++;
+                        ++i;
                     }
 
                     blockrailbase$enumraildirection = BlockRailBase.EnumRailDirection.EAST_WEST;
@@ -112,12 +112,12 @@
 
                     if (p_176566_4_)
                     {
-                        k++;
+                        ++k;
                     }
                     else
                     {
-                        k--;
-                        j++;
+                        --k;
+                        ++j;
                         flag = false;
                     }
 
@@ -127,21 +127,26 @@
 
                     if (p_176566_4_)
                     {
-                        k++;
-                        j++;
+                        ++k;
+                        ++j;
                         flag = false;
                     }
                     else
                     {
-                        k--;
+                        --k;
                     }
 
                     blockrailbase$enumraildirection = BlockRailBase.EnumRailDirection.NORTH_SOUTH;
             }
 
-            return this.func_176567_a(p_176566_1_, new BlockPos(i, j, k), p_176566_4_, p_176566_5_, blockrailbase$enumraildirection)
-                   ? true
-                   : flag && this.func_176567_a(p_176566_1_, new BlockPos(i, j - 1, k), p_176566_4_, p_176566_5_, blockrailbase$enumraildirection);
+            if (this.func_176567_a(p_176566_1_, new BlockPos(i, j, k), p_176566_4_, p_176566_5_, blockrailbase$enumraildirection))
+            {
+                return true;
+            }
+            else
+            {
+                return flag && this.func_176567_a(p_176566_1_, new BlockPos(i, j - 1, k), p_176566_4_, p_176566_5_, blockrailbase$enumraildirection);
+            }
         }
     }
 
@@ -149,34 +154,26 @@
     {
         IBlockState iblockstate = p_176567_1_.func_180495_p(p_176567_2_);
 
-        if (iblockstate.func_177230_c() != this)
+        if (!(iblockstate.func_177230_c() instanceof BlockRailPowered) || isActivator != ((BlockRailPowered)iblockstate.func_177230_c()).isActivator)
         {
             return false;
         }
         else
         {
-            BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = iblockstate.func_177229_b(field_176568_b);
+            BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = (BlockRailBase.EnumRailDirection)iblockstate.func_177229_b(field_176568_b);
 
-            if (p_176567_5_ != BlockRailBase.EnumRailDirection.EAST_WEST
-                    || blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.NORTH_SOUTH
-                    && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_NORTH
-                    && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_SOUTH)
+            if (p_176567_5_ != BlockRailBase.EnumRailDirection.EAST_WEST || blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.NORTH_SOUTH && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_NORTH && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_SOUTH)
             {
-                if (p_176567_5_ != BlockRailBase.EnumRailDirection.NORTH_SOUTH
-                        || blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.EAST_WEST
-                        && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_EAST
-                        && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_WEST)
+                if (p_176567_5_ != BlockRailBase.EnumRailDirection.NORTH_SOUTH || blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.EAST_WEST && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_EAST && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_WEST)
                 {
-                    if (!iblockstate.func_177229_b(field_176569_M))
+                    if (((Boolean)iblockstate.func_177229_b(field_176569_M)).booleanValue())
+                    {
+                        return p_176567_1_.func_175640_z(p_176567_2_) ? true : this.func_176566_a(p_176567_1_, p_176567_2_, iblockstate, p_176567_3_, p_176567_4_ + 1);
+                    }
+                    else
                     {
                         return false;
                     }
-                    else
-                    {
-                        return p_176567_1_.func_175640_z(p_176567_2_)
-                               ? true
-                               : this.func_176566_a(p_176567_1_, p_176567_2_, iblockstate, p_176567_3_, p_176567_4_ + 1);
-                    }
                 }
                 else
                 {
@@ -190,47 +187,39 @@
         }
     }
 
-    @Override
     protected void func_189541_b(IBlockState p_189541_1_, World p_189541_2_, BlockPos p_189541_3_, Block p_189541_4_)
     {
-        boolean flag = p_189541_1_.func_177229_b(field_176569_M);
-        boolean flag1 = p_189541_2_.func_175640_z(p_189541_3_)
-                        || this.func_176566_a(p_189541_2_, p_189541_3_, p_189541_1_, true, 0)
-                        || this.func_176566_a(p_189541_2_, p_189541_3_, p_189541_1_, false, 0);
+        boolean flag = ((Boolean)p_189541_1_.func_177229_b(field_176569_M)).booleanValue();
+        boolean flag1 = p_189541_2_.func_175640_z(p_189541_3_) || this.func_176566_a(p_189541_2_, p_189541_3_, p_189541_1_, true, 0) || this.func_176566_a(p_189541_2_, p_189541_3_, p_189541_1_, false, 0);
 
         if (flag1 != flag)
         {
-            p_189541_2_.func_180501_a(p_189541_3_, p_189541_1_.func_177226_a(field_176569_M, flag1), 3);
+            p_189541_2_.func_180501_a(p_189541_3_, p_189541_1_.func_177226_a(field_176569_M, Boolean.valueOf(flag1)), 3);
             p_189541_2_.func_175685_c(p_189541_3_.func_177977_b(), this, false);
 
-            if (p_189541_1_.func_177229_b(field_176568_b).func_177018_c())
+            if (((BlockRailBase.EnumRailDirection)p_189541_1_.func_177229_b(field_176568_b)).func_177018_c())
             {
                 p_189541_2_.func_175685_c(p_189541_3_.func_177984_a(), this, false);
             }
         }
     }
 
-    @Override
     public IProperty<BlockRailBase.EnumRailDirection> func_176560_l()
     {
         return field_176568_b;
     }
 
-    @Override
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P()
-               .func_177226_a(field_176568_b, BlockRailBase.EnumRailDirection.func_177016_a(p_176203_1_ & 7))
-               .func_177226_a(field_176569_M, (p_176203_1_ & 8) > 0);
+        return this.func_176223_P().func_177226_a(field_176568_b, BlockRailBase.EnumRailDirection.func_177016_a(p_176203_1_ & 7)).func_177226_a(field_176569_M, Boolean.valueOf((p_176203_1_ & 8) > 0));
     }
 
-    @Override
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_176568_b).func_177015_a();
+        i = i | ((BlockRailBase.EnumRailDirection)p_176201_1_.func_177229_b(field_176568_b)).func_177015_a();
 
-        if (p_176201_1_.func_177229_b(field_176569_M))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176569_M)).booleanValue())
         {
             i |= 8;
         }
@@ -238,7 +227,7 @@
         return i;
     }
 
-    @Override
+    @SuppressWarnings("incomplete-switch")
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
         switch (p_185499_2_)
@@ -322,10 +311,10 @@
         }
     }
 
-    @Override
+    @SuppressWarnings("incomplete-switch")
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = p_185471_1_.func_177229_b(field_176568_b);
+        BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = (BlockRailBase.EnumRailDirection)p_185471_1_.func_177229_b(field_176568_b);
 
         switch (p_185471_2_)
         {
@@ -375,9 +364,8 @@
         return super.func_185471_a(p_185471_1_, p_185471_2_);
     }
 
-    @Override
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176568_b, field_176569_M);
+        return new BlockStateContainer(this, new IProperty[] {field_176568_b, field_176569_M});
     }
 }
