--- before/net/minecraft/block/BlockPistonMoving.java
+++ after/net/minecraft/block/BlockPistonMoving.java
@@ -4,6 +4,7 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockFaceShape;
@@ -76,10 +77,10 @@
 
     public void func_176206_d(World p_176206_1_, BlockPos p_176206_2_, IBlockState p_176206_3_)
     {
-        BlockPos blockpos = p_176206_2_.func_177972_a(p_176206_3_.func_177229_b(field_176426_a).func_176734_d());
+        BlockPos blockpos = p_176206_2_.func_177972_a(((EnumFacing)p_176206_3_.func_177229_b(field_176426_a)).func_176734_d());
         IBlockState iblockstate = p_176206_1_.func_180495_p(blockpos);
 
-        if (iblockstate.func_177230_c() instanceof BlockPistonBase && iblockstate.func_177229_b(BlockPistonBase.field_176320_b))
+        if (iblockstate.func_177230_c() instanceof BlockPistonBase && ((Boolean)iblockstate.func_177229_b(BlockPistonBase.field_176320_b)).booleanValue())
         {
             p_176206_1_.func_175698_g(blockpos);
         }
@@ -115,7 +116,7 @@
 
     public void func_180653_a(World p_180653_1_, BlockPos p_180653_2_, IBlockState p_180653_3_, float p_180653_4_, int p_180653_5_)
     {
-        if (!p_180653_1_.field_72995_K)
+        if (false && !p_180653_1_.field_72995_K) // Forge: Noop this out
         {
             TileEntityPiston tileentitypiston = this.func_185589_c(p_180653_1_, p_180653_2_);
 
@@ -125,6 +126,7 @@
                 iblockstate.func_177230_c().func_176226_b(p_180653_1_, p_180653_2_, iblockstate, 0);
             }
         }
+        super.func_180653_a(p_180653_1_, p_180653_2_, p_180653_3_, 1, p_180653_5_); // mimic vanilla behavior from above and ignore chance
     }
 
     @Nullable
@@ -183,18 +185,18 @@
 
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177226_a(field_176426_a, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_176426_a)));
+        return p_185499_1_.func_177226_a(field_176426_a, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_176426_a)));
     }
 
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_176426_a)));
+        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_176426_a)));
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_176426_a).func_176745_a();
+        i = i | ((EnumFacing)p_176201_1_.func_177229_b(field_176426_a)).func_176745_a();
 
         if (p_176201_1_.func_177229_b(field_176425_b) == BlockPistonExtension.EnumPistonType.STICKY)
         {
@@ -206,7 +208,18 @@
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176426_a, field_176425_b);
+        return new BlockStateContainer(this, new IProperty[] {field_176426_a, field_176425_b});
+    }
+
+    @Override
+    public void getDrops(net.minecraft.util.NonNullList<net.minecraft.item.ItemStack> drops, IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
+        TileEntityPiston tileentitypiston = this.func_185589_c(world, pos);
+        if (tileentitypiston != null)
+        {
+            IBlockState pushed = tileentitypiston.func_174927_b();
+            drops.addAll(pushed.func_177230_c().getDrops(world, pos, pushed, fortune)); // use the old method until it gets removed, for backward compatibility
+        }
     }
 
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
