--- before/net/minecraft/block/BlockPortal.java
+++ after/net/minecraft/block/BlockPortal.java
@@ -4,6 +4,7 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
@@ -31,12 +32,10 @@
 
 public class BlockPortal extends BlockBreakable
 {
-    public static final PropertyEnum<EnumFacing.Axis> field_176550_a = PropertyEnum.func_177706_a(
-                    "axis", EnumFacing.Axis.class, EnumFacing.Axis.X, EnumFacing.Axis.Z
-                );
-    protected static final AxisAlignedBB field_185683_b = new AxisAlignedBB(0.0, 0.0, 0.375, 1.0, 1.0, 0.625);
-    protected static final AxisAlignedBB field_185684_c = new AxisAlignedBB(0.375, 0.0, 0.0, 0.625, 1.0, 1.0);
-    protected static final AxisAlignedBB field_185685_d = new AxisAlignedBB(0.375, 0.0, 0.375, 0.625, 1.0, 0.625);
+    public static final PropertyEnum<EnumFacing.Axis> field_176550_a = PropertyEnum.<EnumFacing.Axis>func_177706_a("axis", EnumFacing.Axis.class, EnumFacing.Axis.X, EnumFacing.Axis.Z);
+    protected static final AxisAlignedBB field_185683_b = new AxisAlignedBB(0.0D, 0.0D, 0.375D, 1.0D, 1.0D, 0.625D);
+    protected static final AxisAlignedBB field_185684_c = new AxisAlignedBB(0.375D, 0.0D, 0.0D, 0.625D, 1.0D, 1.0D);
+    protected static final AxisAlignedBB field_185685_d = new AxisAlignedBB(0.375D, 0.0D, 0.375D, 0.625D, 1.0D, 0.625D);
 
     public BlockPortal()
     {
@@ -45,7 +44,6 @@
         this.func_149675_a(true);
     }
 
-    @Override
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
         switch ((EnumFacing.Axis)p_185496_1_.func_177229_b(field_176550_a))
@@ -60,32 +58,23 @@
         }
     }
 
-    @Override
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
     {
         super.func_180650_b(p_180650_1_, p_180650_2_, p_180650_3_, p_180650_4_);
 
-        if (p_180650_1_.field_73011_w.func_76569_d()
-                && p_180650_1_.func_82736_K().func_82766_b("doMobSpawning")
-                && p_180650_4_.nextInt(2000) < p_180650_1_.func_175659_aa().func_151525_a())
+        if (p_180650_1_.field_73011_w.func_76569_d() && p_180650_1_.func_82736_K().func_82766_b("doMobSpawning") && p_180650_4_.nextInt(2000) < p_180650_1_.func_175659_aa().func_151525_a())
         {
             int i = p_180650_2_.func_177956_o();
-            BlockPos blockpos = p_180650_2_;
+            BlockPos blockpos;
 
-            while (!p_180650_1_.func_180495_p(blockpos).func_185896_q() && blockpos.func_177956_o() > 0)
+            for (blockpos = p_180650_2_; !p_180650_1_.func_180495_p(blockpos).func_185896_q() && blockpos.func_177956_o() > 0; blockpos = blockpos.func_177977_b())
             {
-                blockpos = blockpos.func_177977_b();
+                ;
             }
 
             if (i > 0 && !p_180650_1_.func_180495_p(blockpos.func_177984_a()).func_185915_l())
             {
-                Entity entity = ItemMonsterPlacer.func_77840_a(
-                                    p_180650_1_,
-                                    EntityList.func_191306_a(EntityPigZombie.class),
-                                    (double)blockpos.func_177958_n() + 0.5,
-                                    (double)blockpos.func_177956_o() + 1.1,
-                                    (double)blockpos.func_177952_p() + 0.5
-                                );
+                Entity entity = ItemMonsterPlacer.func_77840_a(p_180650_1_, EntityList.func_191306_a(EntityPigZombie.class), (double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 1.1D, (double)blockpos.func_177952_p() + 0.5D);
 
                 if (entity != null)
                 {
@@ -96,7 +85,6 @@
     }
 
     @Nullable
-    @Override
     public AxisAlignedBB func_180646_a(IBlockState p_180646_1_, IBlockAccess p_180646_2_, BlockPos p_180646_3_)
     {
         return field_185506_k;
@@ -114,7 +102,6 @@
         }
     }
 
-    @Override
     public boolean func_149686_d(IBlockState p_149686_1_)
     {
         return false;
@@ -124,7 +111,7 @@
     {
         BlockPortal.Size blockportal$size = new BlockPortal.Size(p_176548_1_, p_176548_2_, EnumFacing.Axis.X);
 
-        if (blockportal$size.func_150860_b() && blockportal$size.field_150864_e == 0)
+        if (blockportal$size.func_150860_b() && blockportal$size.field_150864_e == 0 && !net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_176548_1_, p_176548_2_, blockportal$size))
         {
             blockportal$size.func_150859_c();
             return true;
@@ -133,7 +120,7 @@
         {
             BlockPortal.Size blockportal$size1 = new BlockPortal.Size(p_176548_1_, p_176548_2_, EnumFacing.Axis.Z);
 
-            if (blockportal$size1.func_150860_b() && blockportal$size1.field_150864_e == 0)
+            if (blockportal$size1.func_150860_b() && blockportal$size1.field_150864_e == 0 && !net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_176548_1_, p_176548_2_, blockportal$size1))
             {
                 blockportal$size1.func_150859_c();
                 return true;
@@ -145,10 +132,9 @@
         }
     }
 
-    @Override
     public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
     {
-        EnumFacing.Axis enumfacing$axis = p_189540_1_.func_177229_b(field_176550_a);
+        EnumFacing.Axis enumfacing$axis = (EnumFacing.Axis)p_189540_1_.func_177229_b(field_176550_a);
 
         if (enumfacing$axis == EnumFacing.Axis.X)
         {
@@ -171,7 +157,6 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_176225_a(IBlockState p_176225_1_, IBlockAccess p_176225_2_, BlockPos p_176225_3_, EnumFacing p_176225_4_)
     {
         p_176225_3_ = p_176225_3_.func_177972_a(p_176225_4_);
@@ -179,7 +164,7 @@
 
         if (p_176225_1_.func_177230_c() == this)
         {
-            enumfacing$axis = p_176225_1_.func_177229_b(field_176550_a);
+            enumfacing$axis = (EnumFacing.Axis)p_176225_1_.func_177229_b(field_176550_a);
 
             if (enumfacing$axis == null)
             {
@@ -197,14 +182,10 @@
             }
         }
 
-        boolean flag = p_176225_2_.func_180495_p(p_176225_3_.func_177976_e()).func_177230_c() == this
-                       && p_176225_2_.func_180495_p(p_176225_3_.func_177985_f(2)).func_177230_c() != this;
-        boolean flag1 = p_176225_2_.func_180495_p(p_176225_3_.func_177974_f()).func_177230_c() == this
-                        && p_176225_2_.func_180495_p(p_176225_3_.func_177965_g(2)).func_177230_c() != this;
-        boolean flag2 = p_176225_2_.func_180495_p(p_176225_3_.func_177978_c()).func_177230_c() == this
-                        && p_176225_2_.func_180495_p(p_176225_3_.func_177964_d(2)).func_177230_c() != this;
-        boolean flag3 = p_176225_2_.func_180495_p(p_176225_3_.func_177968_d()).func_177230_c() == this
-                        && p_176225_2_.func_180495_p(p_176225_3_.func_177970_e(2)).func_177230_c() != this;
+        boolean flag = p_176225_2_.func_180495_p(p_176225_3_.func_177976_e()).func_177230_c() == this && p_176225_2_.func_180495_p(p_176225_3_.func_177985_f(2)).func_177230_c() != this;
+        boolean flag1 = p_176225_2_.func_180495_p(p_176225_3_.func_177974_f()).func_177230_c() == this && p_176225_2_.func_180495_p(p_176225_3_.func_177965_g(2)).func_177230_c() != this;
+        boolean flag2 = p_176225_2_.func_180495_p(p_176225_3_.func_177978_c()).func_177230_c() == this && p_176225_2_.func_180495_p(p_176225_3_.func_177964_d(2)).func_177230_c() != this;
+        boolean flag3 = p_176225_2_.func_180495_p(p_176225_3_.func_177968_d()).func_177230_c() == this && p_176225_2_.func_180495_p(p_176225_3_.func_177970_e(2)).func_177230_c() != this;
         boolean flag4 = flag || flag1 || enumfacing$axis == EnumFacing.Axis.X;
         boolean flag5 = flag2 || flag3 || enumfacing$axis == EnumFacing.Axis.Z;
 
@@ -226,20 +207,17 @@
         }
     }
 
-    @Override
     public int func_149745_a(Random p_149745_1_)
     {
         return 0;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public BlockRenderLayer func_180664_k()
     {
         return BlockRenderLayer.TRANSLUCENT;
     }
 
-    @Override
     public void func_180634_a(World p_180634_1_, BlockPos p_180634_2_, IBlockState p_180634_3_, Entity p_180634_4_)
     {
         if (!p_180634_4_.func_184218_aH() && !p_180634_4_.func_184207_aI() && p_180634_4_.func_184222_aU())
@@ -249,21 +227,11 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_180655_c(IBlockState p_180655_1_, World p_180655_2_, BlockPos p_180655_3_, Random p_180655_4_)
     {
         if (p_180655_4_.nextInt(100) == 0)
         {
-            p_180655_2_.func_184134_a(
-                (double)p_180655_3_.func_177958_n() + 0.5,
-                (double)p_180655_3_.func_177956_o() + 0.5,
-                (double)p_180655_3_.func_177952_p() + 0.5,
-                SoundEvents.field_187810_eg,
-                SoundCategory.BLOCKS,
-                0.5F,
-                p_180655_4_.nextFloat() * 0.4F + 0.8F,
-                false
-            );
+            p_180655_2_.func_184134_a((double)p_180655_3_.func_177958_n() + 0.5D, (double)p_180655_3_.func_177956_o() + 0.5D, (double)p_180655_3_.func_177952_p() + 0.5D, SoundEvents.field_187810_eg, SoundCategory.BLOCKS, 0.5F, p_180655_4_.nextFloat() * 0.4F + 0.8F, false);
         }
 
         for (int i = 0; i < 4; ++i)
@@ -271,20 +239,19 @@
             double d0 = (double)((float)p_180655_3_.func_177958_n() + p_180655_4_.nextFloat());
             double d1 = (double)((float)p_180655_3_.func_177956_o() + p_180655_4_.nextFloat());
             double d2 = (double)((float)p_180655_3_.func_177952_p() + p_180655_4_.nextFloat());
-            double d3 = ((double)p_180655_4_.nextFloat() - 0.5) * 0.5;
-            double d4 = ((double)p_180655_4_.nextFloat() - 0.5) * 0.5;
-            double d5 = ((double)p_180655_4_.nextFloat() - 0.5) * 0.5;
+            double d3 = ((double)p_180655_4_.nextFloat() - 0.5D) * 0.5D;
+            double d4 = ((double)p_180655_4_.nextFloat() - 0.5D) * 0.5D;
+            double d5 = ((double)p_180655_4_.nextFloat() - 0.5D) * 0.5D;
             int j = p_180655_4_.nextInt(2) * 2 - 1;
 
-            if (p_180655_2_.func_180495_p(p_180655_3_.func_177976_e()).func_177230_c() != this
-                    && p_180655_2_.func_180495_p(p_180655_3_.func_177974_f()).func_177230_c() != this)
+            if (p_180655_2_.func_180495_p(p_180655_3_.func_177976_e()).func_177230_c() != this && p_180655_2_.func_180495_p(p_180655_3_.func_177974_f()).func_177230_c() != this)
             {
-                d0 = (double)p_180655_3_.func_177958_n() + 0.5 + 0.25 * (double)j;
+                d0 = (double)p_180655_3_.func_177958_n() + 0.5D + 0.25D * (double)j;
                 d3 = (double)(p_180655_4_.nextFloat() * 2.0F * (float)j);
             }
             else
             {
-                d2 = (double)p_180655_3_.func_177952_p() + 0.5 + 0.25 * (double)j;
+                d2 = (double)p_180655_3_.func_177952_p() + 0.5D + 0.25D * (double)j;
                 d5 = (double)(p_180655_4_.nextFloat() * 2.0F * (float)j);
             }
 
@@ -292,25 +259,21 @@
         }
     }
 
-    @Override
     public ItemStack func_185473_a(World p_185473_1_, BlockPos p_185473_2_, IBlockState p_185473_3_)
     {
         return ItemStack.field_190927_a;
     }
 
-    @Override
     public IBlockState func_176203_a(int p_176203_1_)
     {
         return this.func_176223_P().func_177226_a(field_176550_a, (p_176203_1_ & 3) == 2 ? EnumFacing.Axis.Z : EnumFacing.Axis.X);
     }
 
-    @Override
     public int func_176201_c(IBlockState p_176201_1_)
     {
         return func_176549_a(p_176201_1_.func_177229_b(field_176550_a));
     }
 
-    @Override
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
         switch (p_185499_2_)
@@ -333,10 +296,9 @@
         }
     }
 
-    @Override
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176550_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176550_a});
     }
 
     public BlockPattern.PatternHelper func_181089_f(World p_181089_1_, BlockPos p_181089_2_)
@@ -363,17 +325,7 @@
 
             for (EnumFacing.AxisDirection enumfacing$axisdirection : EnumFacing.AxisDirection.values())
             {
-                BlockPattern.PatternHelper blockpattern$patternhelper = new BlockPattern.PatternHelper(
-                    enumfacing.func_176743_c() == enumfacing$axisdirection
-                    ? blockpos
-                    : blockpos.func_177967_a(blockportal$size.field_150866_c, blockportal$size.func_181101_b() - 1),
-                    EnumFacing.func_181076_a(enumfacing$axisdirection, enumfacing$axis),
-                    EnumFacing.UP,
-                    loadingcache,
-                    blockportal$size.func_181101_b(),
-                    blockportal$size.func_181100_a(),
-                    1
-                );
+                BlockPattern.PatternHelper blockpattern$patternhelper = new BlockPattern.PatternHelper(enumfacing.func_176743_c() == enumfacing$axisdirection ? blockpos : blockpos.func_177967_a(blockportal$size.field_150866_c, blockportal$size.func_181101_b() - 1), EnumFacing.func_181076_a(enumfacing$axisdirection, enumfacing$axis), EnumFacing.UP, loadingcache, blockportal$size.func_181101_b(), blockportal$size.func_181100_a(), 1);
 
                 for (int i = 0; i < blockportal$size.func_181101_b(); ++i)
                 {
@@ -383,7 +335,7 @@
 
                         if (blockworldstate.func_177509_a() != null && blockworldstate.func_177509_a().func_185904_a() != Material.field_151579_a)
                         {
-                            aint[enumfacing$axisdirection.ordinal()]++;
+                            ++aint[enumfacing$axisdirection.ordinal()];
                         }
                     }
                 }
@@ -399,21 +351,10 @@
                 }
             }
 
-            return new BlockPattern.PatternHelper(
-                       enumfacing.func_176743_c() == enumfacing$axisdirection1
-                       ? blockpos
-                       : blockpos.func_177967_a(blockportal$size.field_150866_c, blockportal$size.func_181101_b() - 1),
-                       EnumFacing.func_181076_a(enumfacing$axisdirection1, enumfacing$axis),
-                       EnumFacing.UP,
-                       loadingcache,
-                       blockportal$size.func_181101_b(),
-                       blockportal$size.func_181100_a(),
-                       1
-                   );
+            return new BlockPattern.PatternHelper(enumfacing.func_176743_c() == enumfacing$axisdirection1 ? blockpos : blockpos.func_177967_a(blockportal$size.field_150866_c, blockportal$size.func_181101_b() - 1), EnumFacing.func_181076_a(enumfacing$axisdirection1, enumfacing$axis), EnumFacing.UP, loadingcache, blockportal$size.func_181101_b(), blockportal$size.func_181100_a(), 1);
         }
     }
 
-    @Override
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
     {
         return BlockFaceShape.UNDEFINED;
@@ -446,15 +387,9 @@
                     this.field_150866_c = EnumFacing.SOUTH;
                 }
 
-                BlockPos blockpos = p_i45694_2_;
-
-                while (
-                    p_i45694_2_.func_177956_o() > blockpos.func_177956_o() - 21
-                    && p_i45694_2_.func_177956_o() > 0
-                    && this.func_150857_a(p_i45694_1_.func_180495_p(p_i45694_2_.func_177977_b()).func_177230_c())
-                )
+                for (BlockPos blockpos = p_i45694_2_; p_i45694_2_.func_177956_o() > blockpos.func_177956_o() - 21 && p_i45694_2_.func_177956_o() > 0 && this.func_150857_a(p_i45694_1_.func_180495_p(p_i45694_2_.func_177977_b()).func_177230_c()); p_i45694_2_ = p_i45694_2_.func_177977_b())
                 {
-                    p_i45694_2_ = p_i45694_2_.func_177977_b();
+                    ;
                 }
 
                 int i = this.func_180120_a(p_i45694_2_, this.field_150863_d) - 1;
@@ -485,8 +420,7 @@
                 {
                     BlockPos blockpos = p_180120_1_.func_177967_a(p_180120_2_, i);
 
-                    if (!this.func_150857_a(this.field_150867_a.func_180495_p(blockpos).func_177230_c())
-                            || this.field_150867_a.func_180495_p(blockpos.func_177977_b()).func_177230_c() != Blocks.field_150343_Z)
+                    if (!this.func_150857_a(this.field_150867_a.func_180495_p(blockpos).func_177230_c()) || this.field_150867_a.func_180495_p(blockpos.func_177977_b()).func_177230_c() != Blocks.field_150343_Z)
                     {
                         break;
                     }
@@ -550,8 +484,7 @@
 
                 for (int j = 0; j < this.field_150868_h; ++j)
                 {
-                    if (this.field_150867_a.func_180495_p(this.field_150861_f.func_177967_a(this.field_150866_c, j).func_177981_b(this.field_150862_g)).func_177230_c()
-                            != Blocks.field_150343_Z)
+                    if (this.field_150867_a.func_180495_p(this.field_150861_f.func_177967_a(this.field_150866_c, j).func_177981_b(this.field_150862_g)).func_177230_c() != Blocks.field_150343_Z)
                     {
                         this.field_150862_g = 0;
                         break;
@@ -589,10 +522,7 @@
 
                     for (int j = 0; j < this.field_150862_g; ++j)
                     {
-                        this.field_150867_a
-                        .func_180501_a(
-                            blockpos.func_177981_b(j), Blocks.field_150427_aO.func_176223_P().func_177226_a(BlockPortal.field_176550_a, this.field_150865_b), 2
-                        );
+                        this.field_150867_a.func_180501_a(blockpos.func_177981_b(j), Blocks.field_150427_aO.func_176223_P().func_177226_a(BlockPortal.field_176550_a, this.field_150865_b), 2);
                     }
                 }
             }
