--- before/net/minecraft/block/BlockStem.java
+++ after/net/minecraft/block/BlockStem.java
@@ -78,18 +78,19 @@
     {
         super.updateTick(worldIn, pos, state, rand);
 
+        if (!worldIn.isAreaLoaded(pos, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
         if (worldIn.getLightFromNeighbors(pos.up()) >= 9)
         {
             float f = BlockCrops.getGrowthChance(this, worldIn, pos);
 
-            if (rand.nextInt((int)(25.0F / f) + 1) == 0)
+            if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, pos, state, rand.nextInt((int)(25.0F / f) + 1) == 0))
             {
                 int i = state.getValue(AGE);
 
                 if (i < 7)
                 {
-                    state = state.withProperty(AGE, i + 1);
-                    worldIn.setBlockState(pos, state, 2);
+                    IBlockState newState = state.withProperty(AGE, i + 1);
+                    worldIn.setBlockState(pos, newState, 2);
                 }
                 else
                 {
@@ -102,14 +103,18 @@
                     }
 
                     pos = pos.offset(EnumFacing.Plane.HORIZONTAL.random(rand));
-                    Block block = worldIn.getBlockState(pos.down()).getBlock();
+                    IBlockState soil = worldIn.getBlockState(pos.down());
+                    Block block = soil.getBlock();
 
-                    if (worldIn.getBlockState(pos).getBlock().material == Material.AIR
-                            && (block == Blocks.FARMLAND || block == Blocks.DIRT || block == Blocks.GRASS))
+                    if (worldIn.isAirBlock(pos)
+                            && (block.canSustainPlant(soil, worldIn, pos.down(), EnumFacing.UP, this)
+                            || block == Blocks.DIRT
+                            || block == Blocks.GRASS))
                     {
                         worldIn.setBlockState(pos, this.crop.getDefaultState());
                     }
                 }
+                net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state, worldIn.getBlockState(pos));
             }
         }
     }
@@ -124,8 +129,11 @@
     public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune)
     {
         super.dropBlockAsItemWithChance(worldIn, pos, state, chance, fortune);
+    }
 
-        if (!worldIn.isRemote)
+    @Override
+    public void getDrops(net.minecraft.util.NonNullList<ItemStack> drops, IBlockAccess world, BlockPos pos, IBlockState state, int fortune)
+    {
         {
             Item item = this.getSeedItem();
 
@@ -135,9 +143,9 @@
 
                 for (int j = 0; j < 3; j++)
                 {
-                    if (worldIn.rand.nextInt(15) <= i)
+                    if (RANDOM.nextInt(15) <= i)
                     {
-                        spawnAsEntity(worldIn, pos, new ItemStack(item));
+                        drops.add(new ItemStack(item));
                     }
                 }
             }
