--- before/net/minecraft/block/BlockRedstoneWire.java
+++ after/net/minecraft/block/BlockRedstoneWire.java
@@ -133,16 +133,17 @@
         BlockPos blockpos = pos.offset(direction);
         IBlockState iblockstate = worldIn.getBlockState(pos.offset(direction));
 
-        if (!canConnectTo(worldIn.getBlockState(blockpos), direction)
-                && (iblockstate.isNormalCube() || !canConnectUpwardsTo(worldIn.getBlockState(blockpos.down()))))
+        if (!canConnectTo(worldIn.getBlockState(blockpos), direction, worldIn, blockpos)
+                && (iblockstate.isNormalCube() || !canConnectUpwardsTo(worldIn, blockpos.down())))
         {
             IBlockState iblockstate1 = worldIn.getBlockState(pos.up());
 
             if (!iblockstate1.isNormalCube())
             {
-                boolean flag = worldIn.getBlockState(blockpos).isTopSolid() || worldIn.getBlockState(blockpos).getBlock() == Blocks.GLOWSTONE;
+                boolean flag = worldIn.getBlockState(blockpos).isSideSolid(worldIn, blockpos, EnumFacing.UP)
+                               || worldIn.getBlockState(blockpos).getBlock() == Blocks.GLOWSTONE;
 
-                if (flag && canConnectUpwardsTo(worldIn.getBlockState(blockpos.up())))
+                if (flag && canConnectUpwardsTo(worldIn, blockpos.up()))
                 {
                     if (iblockstate.isBlockNormalCube())
                     {
@@ -183,7 +184,9 @@
     @Override
     public boolean canPlaceBlockAt(World worldIn, BlockPos pos)
     {
-        return worldIn.getBlockState(pos.down()).isTopSolid()
+        IBlockState downState = worldIn.getBlockState(pos.down());
+        return downState.isTopSolid()
+               || downState.getBlockFaceShape(worldIn, pos.down(), EnumFacing.UP) == BlockFaceShape.SOLID
                || worldIn.getBlockState(pos.down()).getBlock() == Blocks.GLOWSTONE;
     }
 
@@ -456,7 +459,7 @@
         {
             return true;
         }
-        else if (canConnectTo(iblockstate, side))
+        else if (canConnectTo(iblockstate, side, worldIn, pos))
         {
             return true;
         }
@@ -470,15 +473,10 @@
 
     protected static boolean canConnectUpwardsTo(IBlockAccess worldIn, BlockPos pos)
     {
-        return canConnectUpwardsTo(worldIn.getBlockState(pos));
-    }
-
-    protected static boolean canConnectUpwardsTo(IBlockState state)
-    {
-        return canConnectTo(state, null);
-    }
-
-    protected static boolean canConnectTo(IBlockState blockState, @Nullable EnumFacing side)
+        return canConnectTo(worldIn.getBlockState(pos), null, worldIn, pos);
+    }
+
+    protected static boolean canConnectTo(IBlockState blockState, @Nullable EnumFacing side, IBlockAccess world, BlockPos pos)
     {
         Block block = blockState.getBlock();
 
@@ -495,7 +493,7 @@
         {
             return Blocks.OBSERVER == blockState.getBlock()
                    ? side == blockState.getValue(BlockObserver.FACING)
-                   : blockState.canProvidePower() && side != null;
+                   : blockState.getBlock().canConnectRedstone(blockState, world, pos, side);
         }
     }
 
