--- before/net/minecraft/block/BlockLog.java
+++ after/net/minecraft/block/BlockLog.java
@@ -35,9 +35,9 @@
             {
                 IBlockState iblockstate = worldIn.getBlockState(blockpos);
 
-                if (iblockstate.getMaterial() == Material.LEAVES && !iblockstate.getValue(BlockLeaves.CHECK_DECAY))
+                if (iblockstate.getBlock().isLeaves(iblockstate, worldIn, blockpos))
                 {
-                    worldIn.setBlockState(blockpos, iblockstate.withProperty(BlockLeaves.CHECK_DECAY, true), 4);
+                    iblockstate.getBlock().beginLeavesDecay(iblockstate, worldIn, blockpos);
                 }
             }
         }
@@ -79,6 +79,18 @@
             default:
                 return state;
         }
+    }
+
+    @Override
+    public boolean canSustainLeaves(IBlockState state, net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return true;
+    }
+    
+    @Override
+    public boolean isWood(net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return true;
     }
 
     public static enum EnumAxis implements IStringSerializable
