--- before/net/minecraft/block/BlockCake.java
+++ after/net/minecraft/block/BlockCake.java
@@ -2,6 +2,7 @@
 
 import java.util.Random;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyInteger;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
@@ -29,13 +30,13 @@
     protected BlockCake()
     {
         super(Material.field_151568_F);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176589_a, 0));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176589_a, Integer.valueOf(0)));
         this.func_149675_a(true);
     }
 
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
-        return field_185595_b[p_185496_1_.func_177229_b(field_176589_a)];
+        return field_185595_b[((Integer)p_185496_1_.func_177229_b(field_176589_a)).intValue()];
     }
 
     public boolean func_149686_d(IBlockState p_149686_1_)
@@ -71,11 +72,11 @@
         {
             p_180682_4_.func_71029_a(StatList.field_188076_J);
             p_180682_4_.func_71024_bL().func_75122_a(2, 0.1F);
-            int i = p_180682_3_.func_177229_b(field_176589_a);
+            int i = ((Integer)p_180682_3_.func_177229_b(field_176589_a)).intValue();
 
             if (i < 6)
             {
-                p_180682_1_.func_180501_a(p_180682_2_, p_180682_3_.func_177226_a(field_176589_a, i + 1), 3);
+                p_180682_1_.func_180501_a(p_180682_2_, p_180682_3_.func_177226_a(field_176589_a, Integer.valueOf(i + 1)), 3);
             }
             else
             {
@@ -121,7 +122,7 @@
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176589_a, p_176203_1_);
+        return this.func_176223_P().func_177226_a(field_176589_a, Integer.valueOf(p_176203_1_));
     }
 
     @SideOnly(Side.CLIENT)
@@ -132,17 +133,17 @@
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176589_a);
+        return ((Integer)p_176201_1_.func_177229_b(field_176589_a)).intValue();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176589_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176589_a});
     }
 
     public int func_180641_l(IBlockState p_180641_1_, World p_180641_2_, BlockPos p_180641_3_)
     {
-        return (7 - p_180641_1_.func_177229_b(field_176589_a)) * 2;
+        return (7 - ((Integer)p_180641_1_.func_177229_b(field_176589_a)).intValue()) * 2;
     }
 
     public boolean func_149740_M(IBlockState p_149740_1_)
