--- before/net/minecraft/block/BlockRailPowered.java
+++ after/net/minecraft/block/BlockRailPowered.java
@@ -14,30 +14,30 @@
 
 public class BlockRailPowered extends BlockRailBase
 {
-    public static final PropertyEnum<BlockRailBase.EnumRailDirection> SHAPE = PropertyEnum.create(
-                    "shape",
-                    BlockRailBase.EnumRailDirection.class,
-                    new Predicate<BlockRailBase.EnumRailDirection>()
+    public static final PropertyEnum<BlockRailBase.EnumRailDirection> SHAPE = PropertyEnum.<BlockRailBase.EnumRailDirection>create("shape", BlockRailBase.EnumRailDirection.class, new Predicate<BlockRailBase.EnumRailDirection>()
     {
         public boolean apply(@Nullable BlockRailBase.EnumRailDirection p_apply_1_)
         {
-            return p_apply_1_ != BlockRailBase.EnumRailDirection.NORTH_EAST
-                   && p_apply_1_ != BlockRailBase.EnumRailDirection.NORTH_WEST
-                   && p_apply_1_ != BlockRailBase.EnumRailDirection.SOUTH_EAST
-                   && p_apply_1_ != BlockRailBase.EnumRailDirection.SOUTH_WEST;
+            return p_apply_1_ != BlockRailBase.EnumRailDirection.NORTH_EAST && p_apply_1_ != BlockRailBase.EnumRailDirection.NORTH_WEST && p_apply_1_ != BlockRailBase.EnumRailDirection.SOUTH_EAST && p_apply_1_ != BlockRailBase.EnumRailDirection.SOUTH_WEST;
         }
-    }
-            );
+    });
     public static final PropertyBool POWERED = PropertyBool.create("powered");
 
+    private final boolean isActivator;
+
     protected BlockRailPowered()
     {
+        this(false);
+    }
+
+    protected BlockRailPowered(boolean isActivator)
+    {
         super(true);
-        this.setDefaultState(
-            this.blockState.getBaseState().withProperty(SHAPE, BlockRailBase.EnumRailDirection.NORTH_SOUTH).withProperty(POWERED, false)
-        );
+        this.isActivator = isActivator;
+        this.setDefaultState(this.blockState.getBaseState().withProperty(SHAPE, BlockRailBase.EnumRailDirection.NORTH_SOUTH).withProperty(POWERED, Boolean.valueOf(false)));
     }
 
+    @SuppressWarnings("incomplete-switch")
     protected boolean findPoweredRailSignal(World worldIn, BlockPos pos, IBlockState state, boolean p_176566_4_, int p_176566_5_)
     {
         if (p_176566_5_ >= 8)
@@ -50,7 +50,7 @@
             int j = pos.getY();
             int k = pos.getZ();
             boolean flag = true;
-            BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = state.getValue(SHAPE);
+            BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = (BlockRailBase.EnumRailDirection)state.getValue(SHAPE);
 
             switch (blockrailbase$enumraildirection)
             {
@@ -58,11 +58,11 @@
 
                     if (p_176566_4_)
                     {
-                        k++;
+                        ++k;
                     }
                     else
                     {
-                        k--;
+                        --k;
                     }
 
                     break;
@@ -70,11 +70,11 @@
 
                     if (p_176566_4_)
                     {
-                        i--;
+                        --i;
                     }
                     else
                     {
-                        i++;
+                        ++i;
                     }
 
                     break;
@@ -82,12 +82,12 @@
 
                     if (p_176566_4_)
                     {
-                        i--;
+                        --i;
                     }
                     else
                     {
-                        i++;
-                        j++;
+                        ++i;
+                        ++j;
                         flag = false;
                     }
 
@@ -97,13 +97,13 @@
 
                     if (p_176566_4_)
                     {
-                        i--;
-                        j++;
+                        --i;
+                        ++j;
                         flag = false;
                     }
                     else
                     {
-                        i++;
+                        ++i;
                     }
 
                     blockrailbase$enumraildirection = BlockRailBase.EnumRailDirection.EAST_WEST;
@@ -112,12 +112,12 @@
 
                     if (p_176566_4_)
                     {
-                        k++;
+                        ++k;
                     }
                     else
                     {
-                        k--;
-                        j++;
+                        --k;
+                        ++j;
                         flag = false;
                     }
 
@@ -127,21 +127,26 @@
 
                     if (p_176566_4_)
                     {
-                        k++;
-                        j++;
+                        ++k;
+                        ++j;
                         flag = false;
                     }
                     else
                     {
-                        k--;
+                        --k;
                     }
 
                     blockrailbase$enumraildirection = BlockRailBase.EnumRailDirection.NORTH_SOUTH;
             }
 
-            return this.isSameRailWithPower(worldIn, new BlockPos(i, j, k), p_176566_4_, p_176566_5_, blockrailbase$enumraildirection)
-                   ? true
-                   : flag && this.isSameRailWithPower(worldIn, new BlockPos(i, j - 1, k), p_176566_4_, p_176566_5_, blockrailbase$enumraildirection);
+            if (this.isSameRailWithPower(worldIn, new BlockPos(i, j, k), p_176566_4_, p_176566_5_, blockrailbase$enumraildirection))
+            {
+                return true;
+            }
+            else
+            {
+                return flag && this.isSameRailWithPower(worldIn, new BlockPos(i, j - 1, k), p_176566_4_, p_176566_5_, blockrailbase$enumraildirection);
+            }
         }
     }
 
@@ -149,34 +154,26 @@
     {
         IBlockState iblockstate = worldIn.getBlockState(pos);
 
-        if (iblockstate.getBlock() != this)
+        if (!(iblockstate.getBlock() instanceof BlockRailPowered) || isActivator != ((BlockRailPowered)iblockstate.getBlock()).isActivator)
         {
             return false;
         }
         else
         {
-            BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = iblockstate.getValue(SHAPE);
+            BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = (BlockRailBase.EnumRailDirection)iblockstate.getValue(SHAPE);
 
-            if (p_176567_5_ != BlockRailBase.EnumRailDirection.EAST_WEST
-                    || blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.NORTH_SOUTH
-                    && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_NORTH
-                    && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_SOUTH)
+            if (p_176567_5_ != BlockRailBase.EnumRailDirection.EAST_WEST || blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.NORTH_SOUTH && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_NORTH && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_SOUTH)
             {
-                if (p_176567_5_ != BlockRailBase.EnumRailDirection.NORTH_SOUTH
-                        || blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.EAST_WEST
-                        && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_EAST
-                        && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_WEST)
+                if (p_176567_5_ != BlockRailBase.EnumRailDirection.NORTH_SOUTH || blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.EAST_WEST && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_EAST && blockrailbase$enumraildirection != BlockRailBase.EnumRailDirection.ASCENDING_WEST)
                 {
-                    if (!iblockstate.getValue(POWERED))
+                    if (((Boolean)iblockstate.getValue(POWERED)).booleanValue())
+                    {
+                        return worldIn.isBlockPowered(pos) ? true : this.findPoweredRailSignal(worldIn, pos, iblockstate, p_176567_3_, distance + 1);
+                    }
+                    else
                     {
                         return false;
                     }
-                    else
-                    {
-                        return worldIn.isBlockPowered(pos)
-                               ? true
-                               : this.findPoweredRailSignal(worldIn, pos, iblockstate, p_176567_3_, distance + 1);
-                    }
                 }
                 else
                 {
@@ -190,47 +187,39 @@
         }
     }
 
-    @Override
     protected void updateState(IBlockState state, World worldIn, BlockPos pos, Block blockIn)
     {
-        boolean flag = state.getValue(POWERED);
-        boolean flag1 = worldIn.isBlockPowered(pos)
-                        || this.findPoweredRailSignal(worldIn, pos, state, true, 0)
-                        || this.findPoweredRailSignal(worldIn, pos, state, false, 0);
+        boolean flag = ((Boolean)state.getValue(POWERED)).booleanValue();
+        boolean flag1 = worldIn.isBlockPowered(pos) || this.findPoweredRailSignal(worldIn, pos, state, true, 0) || this.findPoweredRailSignal(worldIn, pos, state, false, 0);
 
         if (flag1 != flag)
         {
-            worldIn.setBlockState(pos, state.withProperty(POWERED, flag1), 3);
+            worldIn.setBlockState(pos, state.withProperty(POWERED, Boolean.valueOf(flag1)), 3);
             worldIn.notifyNeighborsOfStateChange(pos.down(), this, false);
 
-            if (state.getValue(SHAPE).isAscending())
+            if (((BlockRailBase.EnumRailDirection)state.getValue(SHAPE)).isAscending())
             {
                 worldIn.notifyNeighborsOfStateChange(pos.up(), this, false);
             }
         }
     }
 
-    @Override
     public IProperty<BlockRailBase.EnumRailDirection> getShapeProperty()
     {
         return SHAPE;
     }
 
-    @Override
     public IBlockState getStateFromMeta(int meta)
     {
-        return this.getDefaultState()
-               .withProperty(SHAPE, BlockRailBase.EnumRailDirection.byMetadata(meta & 7))
-               .withProperty(POWERED, (meta & 8) > 0);
+        return this.getDefaultState().withProperty(SHAPE, BlockRailBase.EnumRailDirection.byMetadata(meta & 7)).withProperty(POWERED, Boolean.valueOf((meta & 8) > 0));
     }
 
-    @Override
     public int getMetaFromState(IBlockState state)
     {
         int i = 0;
-        i |= state.getValue(SHAPE).getMetadata();
+        i = i | ((BlockRailBase.EnumRailDirection)state.getValue(SHAPE)).getMetadata();
 
-        if (state.getValue(POWERED))
+        if (((Boolean)state.getValue(POWERED)).booleanValue())
         {
             i |= 8;
         }
@@ -238,7 +227,7 @@
         return i;
     }
 
-    @Override
+    @SuppressWarnings("incomplete-switch")
     public IBlockState withRotation(IBlockState state, Rotation rot)
     {
         switch (rot)
@@ -322,10 +311,10 @@
         }
     }
 
-    @Override
+    @SuppressWarnings("incomplete-switch")
     public IBlockState withMirror(IBlockState state, Mirror mirrorIn)
     {
-        BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = state.getValue(SHAPE);
+        BlockRailBase.EnumRailDirection blockrailbase$enumraildirection = (BlockRailBase.EnumRailDirection)state.getValue(SHAPE);
 
         switch (mirrorIn)
         {
@@ -375,9 +364,8 @@
         return super.withMirror(state, mirrorIn);
     }
 
-    @Override
     protected BlockStateContainer createBlockState()
     {
-        return new BlockStateContainer(this, SHAPE, POWERED);
+        return new BlockStateContainer(this, new IProperty[] {SHAPE, POWERED});
     }
 }
