--- before/net/minecraft/block/BlockCauldron.java
+++ after/net/minecraft/block/BlockCauldron.java
@@ -5,6 +5,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyInteger;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
@@ -43,7 +44,7 @@
     public BlockCauldron()
     {
         super(Material.field_151573_f, MapColor.field_151665_m);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176591_a, 0));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176591_a, Integer.valueOf(0)));
     }
 
     public void func_185477_a(IBlockState p_185477_1_, World p_185477_2_, BlockPos p_185477_3_, AxisAlignedBB p_185477_4_, List<AxisAlignedBB> p_185477_5_, @Nullable Entity p_185477_6_, boolean p_185477_7_)
@@ -72,7 +73,7 @@
 
     public void func_180634_a(World p_180634_1_, BlockPos p_180634_2_, IBlockState p_180634_3_, Entity p_180634_4_)
     {
-        int i = p_180634_3_.func_177229_b(field_176591_a);
+        int i = ((Integer)p_180634_3_.func_177229_b(field_176591_a)).intValue();
         float f = (float)p_180634_2_.func_177956_o() + (6.0F + (float)(3 * i)) / 16.0F;
 
         if (!p_180634_1_.field_72995_K && p_180634_4_.func_70027_ad() && i > 0 && p_180634_4_.func_174813_aQ().field_72338_b <= (double)f)
@@ -92,7 +93,7 @@
         }
         else
         {
-            int i = p_180639_3_.func_177229_b(field_176591_a);
+            int i = ((Integer)p_180639_3_.func_177229_b(field_176591_a)).intValue();
             Item item = itemstack.func_77973_b();
 
             if (item == Items.field_151131_as)
@@ -244,7 +245,7 @@
 
     public void func_176590_a(World p_176590_1_, BlockPos p_176590_2_, IBlockState p_176590_3_, int p_176590_4_)
     {
-        p_176590_1_.func_180501_a(p_176590_2_, p_176590_3_.func_177226_a(field_176591_a, MathHelper.func_76125_a(p_176590_4_, 0, 3)), 2);
+        p_176590_1_.func_180501_a(p_176590_2_, p_176590_3_.func_177226_a(field_176591_a, Integer.valueOf(MathHelper.func_76125_a(p_176590_4_, 0, 3))), 2);
         p_176590_1_.func_175666_e(p_176590_2_, this);
     }
 
@@ -254,11 +255,11 @@
         {
             float f = p_176224_1_.func_180494_b(p_176224_2_).func_180626_a(p_176224_2_);
 
-            if (!(p_176224_1_.func_72959_q().func_76939_a(f, p_176224_2_.func_177956_o()) < 0.15F))
+            if (p_176224_1_.func_72959_q().func_76939_a(f, p_176224_2_.func_177956_o()) >= 0.15F)
             {
                 IBlockState iblockstate = p_176224_1_.func_180495_p(p_176224_2_);
 
-                if (iblockstate.func_177229_b(field_176591_a) < 3)
+                if (((Integer)iblockstate.func_177229_b(field_176591_a)).intValue() < 3)
                 {
                     p_176224_1_.func_180501_a(p_176224_2_, iblockstate.func_177231_a(field_176591_a), 2);
                 }
@@ -283,22 +284,22 @@
 
     public int func_180641_l(IBlockState p_180641_1_, World p_180641_2_, BlockPos p_180641_3_)
     {
-        return p_180641_1_.func_177229_b(field_176591_a);
+        return ((Integer)p_180641_1_.func_177229_b(field_176591_a)).intValue();
     }
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176591_a, p_176203_1_);
+        return this.func_176223_P().func_177226_a(field_176591_a, Integer.valueOf(p_176203_1_));
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176591_a);
+        return ((Integer)p_176201_1_.func_177229_b(field_176591_a)).intValue();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176591_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176591_a});
     }
 
     public boolean func_176205_b(IBlockAccess p_176205_1_, BlockPos p_176205_2_)
