--- before/net/minecraft/block/BlockLadder.java
+++ after/net/minecraft/block/BlockLadder.java
@@ -1,6 +1,7 @@
 package net.minecraft.block;
 
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
@@ -21,10 +22,10 @@
 public class BlockLadder extends Block
 {
     public static final PropertyDirection field_176382_a = BlockHorizontal.field_185512_D;
-    protected static final AxisAlignedBB field_185687_b = new AxisAlignedBB(0.0, 0.0, 0.0, 0.1875, 1.0, 1.0);
-    protected static final AxisAlignedBB field_185688_c = new AxisAlignedBB(0.8125, 0.0, 0.0, 1.0, 1.0, 1.0);
-    protected static final AxisAlignedBB field_185689_d = new AxisAlignedBB(0.0, 0.0, 0.0, 1.0, 1.0, 0.1875);
-    protected static final AxisAlignedBB field_185690_e = new AxisAlignedBB(0.0, 0.0, 0.8125, 1.0, 1.0, 1.0);
+    protected static final AxisAlignedBB field_185687_b = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 0.1875D, 1.0D, 1.0D);
+    protected static final AxisAlignedBB field_185688_c = new AxisAlignedBB(0.8125D, 0.0D, 0.0D, 1.0D, 1.0D, 1.0D);
+    protected static final AxisAlignedBB field_185689_d = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 1.0D, 1.0D, 0.1875D);
+    protected static final AxisAlignedBB field_185690_e = new AxisAlignedBB(0.0D, 0.0D, 0.8125D, 1.0D, 1.0D, 1.0D);
 
     protected BlockLadder()
     {
@@ -33,7 +34,6 @@
         this.func_149647_a(CreativeTabs.field_78031_c);
     }
 
-    @Override
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
         switch ((EnumFacing)p_185496_1_.func_177229_b(field_176382_a))
@@ -50,19 +50,16 @@
         }
     }
 
-    @Override
     public boolean func_149662_c(IBlockState p_149662_1_)
     {
         return false;
     }
 
-    @Override
     public boolean func_149686_d(IBlockState p_149686_1_)
     {
         return false;
     }
 
-    @Override
     public boolean func_176198_a(World p_176198_1_, BlockPos p_176198_2_, EnumFacing p_176198_3_)
     {
         if (this.func_193392_c(p_176198_1_, p_176198_2_.func_177976_e(), p_176198_3_))
@@ -73,11 +70,13 @@
         {
             return true;
         }
+        else if (this.func_193392_c(p_176198_1_, p_176198_2_.func_177978_c(), p_176198_3_))
+        {
+            return true;
+        }
         else
         {
-            return this.func_193392_c(p_176198_1_, p_176198_2_.func_177978_c(), p_176198_3_)
-                   ? true
-                   : this.func_193392_c(p_176198_1_, p_176198_2_.func_177968_d(), p_176198_3_);
+            return this.func_193392_c(p_176198_1_, p_176198_2_.func_177968_d(), p_176198_3_);
         }
     }
 
@@ -88,17 +87,7 @@
         return !flag && iblockstate.func_193401_d(p_193392_1_, p_193392_2_, p_193392_3_) == BlockFaceShape.SOLID && !iblockstate.func_185897_m();
     }
 
-    @Override
-    public IBlockState func_180642_a(
-        World p_180642_1_,
-        BlockPos p_180642_2_,
-        EnumFacing p_180642_3_,
-        float p_180642_4_,
-        float p_180642_5_,
-        float p_180642_6_,
-        int p_180642_7_,
-        EntityLivingBase p_180642_8_
-    )
+    public IBlockState func_180642_a(World p_180642_1_, BlockPos p_180642_2_, EnumFacing p_180642_3_, float p_180642_4_, float p_180642_5_, float p_180642_6_, int p_180642_7_, EntityLivingBase p_180642_8_)
     {
         if (p_180642_3_.func_176740_k().func_176722_c() && this.func_193392_c(p_180642_1_, p_180642_2_.func_177972_a(p_180642_3_.func_176734_d()), p_180642_3_))
         {
@@ -118,10 +107,9 @@
         }
     }
 
-    @Override
     public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
     {
-        EnumFacing enumfacing = p_189540_1_.func_177229_b(field_176382_a);
+        EnumFacing enumfacing = (EnumFacing)p_189540_1_.func_177229_b(field_176382_a);
 
         if (!this.func_193392_c(p_189540_2_, p_189540_3_.func_177972_a(enumfacing.func_176734_d()), enumfacing))
         {
@@ -132,14 +120,6 @@
         super.func_189540_a(p_189540_1_, p_189540_2_, p_189540_3_, p_189540_4_, p_189540_5_);
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
-    public BlockRenderLayer func_180664_k()
-    {
-        return BlockRenderLayer.CUTOUT;
-    }
-
-    @Override
     public IBlockState func_176203_a(int p_176203_1_)
     {
         EnumFacing enumfacing = EnumFacing.func_82600_a(p_176203_1_);
@@ -152,31 +132,34 @@
         return this.func_176223_P().func_177226_a(field_176382_a, enumfacing);
     }
 
-    @Override
+    @SideOnly(Side.CLIENT)
+    public BlockRenderLayer func_180664_k()
+    {
+        return BlockRenderLayer.CUTOUT;
+    }
+
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176382_a).func_176745_a();
+        return ((EnumFacing)p_176201_1_.func_177229_b(field_176382_a)).func_176745_a();
     }
 
-    @Override
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177226_a(field_176382_a, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_176382_a)));
+        return p_185499_1_.func_177226_a(field_176382_a, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_176382_a)));
     }
 
-    @Override
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_176382_a)));
+        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_176382_a)));
     }
 
-    @Override
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176382_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176382_a});
     }
 
-    @Override
+    @Override public boolean isLadder(IBlockState state, IBlockAccess world, BlockPos pos, EntityLivingBase entity) { return true; }
+
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
     {
         return BlockFaceShape.UNDEFINED;
