--- before/net/minecraft/block/BlockColored.java
+++ after/net/minecraft/block/BlockColored.java
@@ -2,6 +2,7 @@
 
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyEnum;
 import net.minecraft.block.state.BlockStateContainer;
 import net.minecraft.block.state.IBlockState;
@@ -14,7 +15,7 @@
 
 public class BlockColored extends Block
 {
-    public static final PropertyEnum<EnumDyeColor> field_176581_a = PropertyEnum.func_177709_a("color", EnumDyeColor.class);
+    public static final PropertyEnum<EnumDyeColor> field_176581_a = PropertyEnum.<EnumDyeColor>func_177709_a("color", EnumDyeColor.class);
 
     public BlockColored(Material p_i45398_1_)
     {
@@ -25,7 +26,7 @@
 
     public int func_180651_a(IBlockState p_180651_1_)
     {
-        return p_180651_1_.func_177229_b(field_176581_a).func_176765_a();
+        return ((EnumDyeColor)p_180651_1_.func_177229_b(field_176581_a)).func_176765_a();
     }
 
     public void func_149666_a(CreativeTabs p_149666_1_, NonNullList<ItemStack> p_149666_2_)
@@ -38,7 +39,7 @@
 
     public MapColor func_180659_g(IBlockState p_180659_1_, IBlockAccess p_180659_2_, BlockPos p_180659_3_)
     {
-        return MapColor.func_193558_a(p_180659_1_.func_177229_b(field_176581_a));
+        return MapColor.func_193558_a((EnumDyeColor)p_180659_1_.func_177229_b(field_176581_a));
     }
 
     public IBlockState func_176203_a(int p_176203_1_)
@@ -48,11 +49,11 @@
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
-        return p_176201_1_.func_177229_b(field_176581_a).func_176765_a();
+        return ((EnumDyeColor)p_176201_1_.func_177229_b(field_176581_a)).func_176765_a();
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176581_a);
+        return new BlockStateContainer(this, new IProperty[] {field_176581_a});
     }
 }
