--- before/net/minecraft/block/BlockTripWireHook.java
+++ after/net/minecraft/block/BlockTripWireHook.java
@@ -4,6 +4,7 @@
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.state.BlockFaceShape;
@@ -40,7 +41,7 @@
     public BlockTripWireHook()
     {
         super(Material.field_151594_q);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176264_a, EnumFacing.NORTH).func_177226_a(field_176263_b, false).func_177226_a(field_176265_M, false));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176264_a, EnumFacing.NORTH).func_177226_a(field_176263_b, Boolean.valueOf(false)).func_177226_a(field_176265_M, Boolean.valueOf(false)));
         this.func_149647_a(CreativeTabs.field_78028_d);
         this.func_149675_a(true);
     }
@@ -101,7 +102,7 @@
 
     public IBlockState func_180642_a(World p_180642_1_, BlockPos p_180642_2_, EnumFacing p_180642_3_, float p_180642_4_, float p_180642_5_, float p_180642_6_, int p_180642_7_, EntityLivingBase p_180642_8_)
     {
-        IBlockState iblockstate = this.func_176223_P().func_177226_a(field_176263_b, false).func_177226_a(field_176265_M, false);
+        IBlockState iblockstate = this.func_176223_P().func_177226_a(field_176263_b, Boolean.valueOf(false)).func_177226_a(field_176265_M, Boolean.valueOf(false));
 
         if (p_180642_3_.func_176740_k().func_176722_c())
         {
@@ -122,7 +123,7 @@
         {
             if (this.func_176261_e(p_189540_2_, p_189540_3_, p_189540_1_))
             {
-                EnumFacing enumfacing = p_189540_1_.func_177229_b(field_176264_a);
+                EnumFacing enumfacing = (EnumFacing)p_189540_1_.func_177229_b(field_176264_a);
 
                 if (!this.func_176198_a(p_189540_2_, p_189540_3_, enumfacing))
                 {
@@ -135,9 +136,9 @@
 
     public void func_176260_a(World p_176260_1_, BlockPos p_176260_2_, IBlockState p_176260_3_, boolean p_176260_4_, boolean p_176260_5_, int p_176260_6_, @Nullable IBlockState p_176260_7_)
     {
-        EnumFacing enumfacing = p_176260_3_.func_177229_b(field_176264_a);
-        boolean flag = p_176260_3_.func_177229_b(field_176265_M);
-        boolean flag1 = p_176260_3_.func_177229_b(field_176263_b);
+        EnumFacing enumfacing = (EnumFacing)p_176260_3_.func_177229_b(field_176264_a);
+        boolean flag = ((Boolean)p_176260_3_.func_177229_b(field_176265_M)).booleanValue();
+        boolean flag1 = ((Boolean)p_176260_3_.func_177229_b(field_176263_b)).booleanValue();
         boolean flag2 = !p_176260_4_;
         boolean flag3 = false;
         int i = 0;
@@ -167,11 +168,11 @@
             {
                 if (j == p_176260_6_)
                 {
-                    iblockstate = MoreObjects.firstNonNull(p_176260_7_, iblockstate);
+                    iblockstate = (IBlockState)MoreObjects.firstNonNull(p_176260_7_, iblockstate);
                 }
 
-                boolean flag4 = !iblockstate.func_177229_b(BlockTripWire.field_176295_N);
-                boolean flag5 = iblockstate.func_177229_b(BlockTripWire.field_176293_a);
+                boolean flag4 = !((Boolean)iblockstate.func_177229_b(BlockTripWire.field_176295_N)).booleanValue();
+                boolean flag5 = ((Boolean)iblockstate.func_177229_b(BlockTripWire.field_176293_a)).booleanValue();
                 flag3 |= flag4 && flag5;
                 aiblockstate[j] = iblockstate;
 
@@ -183,9 +184,9 @@
             }
         }
 
-        flag2 &= i > 1;
-        flag3 &= flag2;
-        IBlockState iblockstate1 = this.func_176223_P().func_177226_a(field_176265_M, flag2).func_177226_a(field_176263_b, flag3);
+        flag2 = flag2 & i > 1;
+        flag3 = flag3 & flag2;
+        IBlockState iblockstate1 = this.func_176223_P().func_177226_a(field_176265_M, Boolean.valueOf(flag2)).func_177226_a(field_176263_b, Boolean.valueOf(flag3));
 
         if (i > 0)
         {
@@ -217,7 +218,7 @@
 
                 if (iblockstate2 != null && p_176260_1_.func_180495_p(blockpos2).func_185904_a() != Material.field_151579_a)
                 {
-                    p_176260_1_.func_180501_a(blockpos2, iblockstate2.func_177226_a(field_176265_M, flag2), 3);
+                    p_176260_1_.func_180501_a(blockpos2, iblockstate2.func_177226_a(field_176265_M, Boolean.valueOf(flag2)), 3);
                 }
             }
         }
@@ -274,8 +275,8 @@
 
     public void func_180663_b(World p_180663_1_, BlockPos p_180663_2_, IBlockState p_180663_3_)
     {
-        boolean flag = p_180663_3_.func_177229_b(field_176265_M);
-        boolean flag1 = p_180663_3_.func_177229_b(field_176263_b);
+        boolean flag = ((Boolean)p_180663_3_.func_177229_b(field_176265_M)).booleanValue();
+        boolean flag1 = ((Boolean)p_180663_3_.func_177229_b(field_176263_b)).booleanValue();
 
         if (flag || flag1)
         {
@@ -285,7 +286,7 @@
         if (flag1)
         {
             p_180663_1_.func_175685_c(p_180663_2_, this, false);
-            p_180663_1_.func_175685_c(p_180663_2_.func_177972_a(p_180663_3_.func_177229_b(field_176264_a).func_176734_d()), this, false);
+            p_180663_1_.func_175685_c(p_180663_2_.func_177972_a(((EnumFacing)p_180663_3_.func_177229_b(field_176264_a)).func_176734_d()), this, false);
         }
 
         super.func_180663_b(p_180663_1_, p_180663_2_, p_180663_3_);
@@ -293,12 +294,12 @@
 
     public int func_180656_a(IBlockState p_180656_1_, IBlockAccess p_180656_2_, BlockPos p_180656_3_, EnumFacing p_180656_4_)
     {
-        return p_180656_1_.func_177229_b(field_176263_b) ? 15 : 0;
+        return ((Boolean)p_180656_1_.func_177229_b(field_176263_b)).booleanValue() ? 15 : 0;
     }
 
     public int func_176211_b(IBlockState p_176211_1_, IBlockAccess p_176211_2_, BlockPos p_176211_3_, EnumFacing p_176211_4_)
     {
-        if (!p_176211_1_.func_177229_b(field_176263_b))
+        if (!((Boolean)p_176211_1_.func_177229_b(field_176263_b)).booleanValue())
         {
             return 0;
         }
@@ -321,20 +322,20 @@
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176264_a, EnumFacing.func_176731_b(p_176203_1_ & 3)).func_177226_a(field_176263_b, (p_176203_1_ & 8) > 0).func_177226_a(field_176265_M, (p_176203_1_ & 4) > 0);
+        return this.func_176223_P().func_177226_a(field_176264_a, EnumFacing.func_176731_b(p_176203_1_ & 3)).func_177226_a(field_176263_b, Boolean.valueOf((p_176203_1_ & 8) > 0)).func_177226_a(field_176265_M, Boolean.valueOf((p_176203_1_ & 4) > 0));
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_176264_a).func_176736_b();
+        i = i | ((EnumFacing)p_176201_1_.func_177229_b(field_176264_a)).func_176736_b();
 
-        if (p_176201_1_.func_177229_b(field_176263_b))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176263_b)).booleanValue())
         {
             i |= 8;
         }
 
-        if (p_176201_1_.func_177229_b(field_176265_M))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176265_M)).booleanValue())
         {
             i |= 4;
         }
@@ -344,17 +345,17 @@
 
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177226_a(field_176264_a, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_176264_a)));
+        return p_185499_1_.func_177226_a(field_176264_a, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_176264_a)));
     }
 
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_176264_a)));
+        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_176264_a)));
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176264_a, field_176263_b, field_176265_M);
+        return new BlockStateContainer(this, new IProperty[] {field_176264_a, field_176263_b, field_176265_M});
     }
 
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
