--- before/net/minecraft/block/BlockRedstoneDiode.java
+++ after/net/minecraft/block/BlockRedstoneDiode.java
@@ -18,7 +18,7 @@
 
 public abstract class BlockRedstoneDiode extends BlockHorizontal
 {
-    protected static final AxisAlignedBB field_185548_c = new AxisAlignedBB(0.0, 0.0, 0.0, 1.0, 0.125, 1.0);
+    protected static final AxisAlignedBB field_185548_c = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 1.0D, 0.125D, 1.0D);
     protected final boolean field_149914_a;
 
     protected BlockRedstoneDiode(boolean p_i45400_1_)
@@ -27,35 +27,32 @@
         this.field_149914_a = p_i45400_1_;
     }
 
-    @Override
     public AxisAlignedBB func_185496_a(IBlockState p_185496_1_, IBlockAccess p_185496_2_, BlockPos p_185496_3_)
     {
         return field_185548_c;
     }
 
-    @Override
     public boolean func_149686_d(IBlockState p_149686_1_)
     {
         return false;
     }
 
-    @Override
     public boolean func_176196_c(World p_176196_1_, BlockPos p_176196_2_)
     {
-        return p_176196_1_.func_180495_p(p_176196_2_.func_177977_b()).func_185896_q() ? super.func_176196_c(p_176196_1_, p_176196_2_) : false;
+        IBlockState downState = p_176196_1_.func_180495_p(p_176196_2_.func_177977_b());
+        return (downState.func_185896_q() || downState.func_193401_d(p_176196_1_, p_176196_2_.func_177977_b(), EnumFacing.UP) == BlockFaceShape.SOLID) ? super.func_176196_c(p_176196_1_, p_176196_2_) : false;
     }
 
     public boolean func_176409_d(World p_176409_1_, BlockPos p_176409_2_)
     {
-        return p_176409_1_.func_180495_p(p_176409_2_.func_177977_b()).func_185896_q();
+        IBlockState downState = p_176409_1_.func_180495_p(p_176409_2_.func_177977_b());
+        return downState.func_185896_q() || downState.func_193401_d(p_176409_1_, p_176409_2_.func_177977_b(), EnumFacing.UP) == BlockFaceShape.SOLID;
     }
 
-    @Override
     public void func_180645_a(World p_180645_1_, BlockPos p_180645_2_, IBlockState p_180645_3_, Random p_180645_4_)
     {
     }
 
-    @Override
     public void func_180650_b(World p_180650_1_, BlockPos p_180650_2_, IBlockState p_180650_3_, Random p_180650_4_)
     {
         if (!this.func_176405_b(p_180650_1_, p_180650_2_, p_180650_3_))
@@ -79,7 +76,6 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_176225_a(IBlockState p_176225_1_, IBlockAccess p_176225_2_, BlockPos p_176225_3_, EnumFacing p_176225_4_)
     {
         return p_176225_4_.func_176740_k() != EnumFacing.Axis.Y;
@@ -90,13 +86,11 @@
         return this.field_149914_a;
     }
 
-    @Override
     public int func_176211_b(IBlockState p_176211_1_, IBlockAccess p_176211_2_, BlockPos p_176211_3_, EnumFacing p_176211_4_)
     {
         return p_176211_1_.func_185911_a(p_176211_2_, p_176211_3_, p_176211_4_);
     }
 
-    @Override
     public int func_180656_a(IBlockState p_180656_1_, IBlockAccess p_180656_2_, BlockPos p_180656_3_, EnumFacing p_180656_4_)
     {
         if (!this.func_176406_l(p_180656_1_))
@@ -109,7 +103,6 @@
         }
     }
 
-    @Override
     public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
     {
         if (this.func_176409_d(p_189540_2_, p_189540_3_))
@@ -164,7 +157,7 @@
 
     protected int func_176397_f(World p_176397_1_, BlockPos p_176397_2_, IBlockState p_176397_3_)
     {
-        EnumFacing enumfacing = p_176397_3_.func_177229_b(field_185512_D);
+        EnumFacing enumfacing = (EnumFacing)p_176397_3_.func_177229_b(field_185512_D);
         BlockPos blockpos = p_176397_2_.func_177972_a(enumfacing);
         int i = p_176397_1_.func_175651_c(blockpos, enumfacing);
 
@@ -175,19 +168,16 @@
         else
         {
             IBlockState iblockstate = p_176397_1_.func_180495_p(blockpos);
-            return Math.max(i, iblockstate.func_177230_c() == Blocks.field_150488_af ? iblockstate.func_177229_b(BlockRedstoneWire.field_176351_O) : 0);
+            return Math.max(i, iblockstate.func_177230_c() == Blocks.field_150488_af ? ((Integer)iblockstate.func_177229_b(BlockRedstoneWire.field_176351_O)).intValue() : 0);
         }
     }
 
     protected int func_176407_c(IBlockAccess p_176407_1_, BlockPos p_176407_2_, IBlockState p_176407_3_)
     {
-        EnumFacing enumfacing = p_176407_3_.func_177229_b(field_185512_D);
+        EnumFacing enumfacing = (EnumFacing)p_176407_3_.func_177229_b(field_185512_D);
         EnumFacing enumfacing1 = enumfacing.func_176746_e();
         EnumFacing enumfacing2 = enumfacing.func_176735_f();
-        return Math.max(
-                   this.func_176401_c(p_176407_1_, p_176407_2_.func_177972_a(enumfacing1), enumfacing1),
-                   this.func_176401_c(p_176407_1_, p_176407_2_.func_177972_a(enumfacing2), enumfacing2)
-               );
+        return Math.max(this.func_176401_c(p_176407_1_, p_176407_2_.func_177972_a(enumfacing1), enumfacing1), this.func_176401_c(p_176407_1_, p_176407_2_.func_177972_a(enumfacing2), enumfacing2));
     }
 
     protected int func_176401_c(IBlockAccess p_176401_1_, BlockPos p_176401_2_, EnumFacing p_176401_3_)
@@ -203,9 +193,7 @@
             }
             else
             {
-                return block == Blocks.field_150488_af
-                       ? iblockstate.func_177229_b(BlockRedstoneWire.field_176351_O)
-                       : p_176401_1_.func_175627_a(p_176401_2_, p_176401_3_);
+                return block == Blocks.field_150488_af ? ((Integer)iblockstate.func_177229_b(BlockRedstoneWire.field_176351_O)).intValue() : p_176401_1_.func_175627_a(p_176401_2_, p_176401_3_);
             }
         }
         else
@@ -214,28 +202,16 @@
         }
     }
 
-    @Override
     public boolean func_149744_f(IBlockState p_149744_1_)
     {
         return true;
     }
 
-    @Override
-    public IBlockState func_180642_a(
-        World p_180642_1_,
-        BlockPos p_180642_2_,
-        EnumFacing p_180642_3_,
-        float p_180642_4_,
-        float p_180642_5_,
-        float p_180642_6_,
-        int p_180642_7_,
-        EntityLivingBase p_180642_8_
-    )
+    public IBlockState func_180642_a(World p_180642_1_, BlockPos p_180642_2_, EnumFacing p_180642_3_, float p_180642_4_, float p_180642_5_, float p_180642_6_, int p_180642_7_, EntityLivingBase p_180642_8_)
     {
         return this.func_176223_P().func_177226_a(field_185512_D, p_180642_8_.func_174811_aO().func_176734_d());
     }
 
-    @Override
     public void func_180633_a(World p_180633_1_, BlockPos p_180633_2_, IBlockState p_180633_3_, EntityLivingBase p_180633_4_, ItemStack p_180633_5_)
     {
         if (this.func_176404_e(p_180633_1_, p_180633_2_, p_180633_3_))
@@ -244,7 +220,6 @@
         }
     }
 
-    @Override
     public void func_176213_c(World p_176213_1_, BlockPos p_176213_2_, IBlockState p_176213_3_)
     {
         this.func_176400_h(p_176213_1_, p_176213_2_, p_176213_3_);
@@ -252,13 +227,14 @@
 
     protected void func_176400_h(World p_176400_1_, BlockPos p_176400_2_, IBlockState p_176400_3_)
     {
-        EnumFacing enumfacing = p_176400_3_.func_177229_b(field_185512_D);
+        EnumFacing enumfacing = (EnumFacing)p_176400_3_.func_177229_b(field_185512_D);
         BlockPos blockpos = p_176400_2_.func_177972_a(enumfacing.func_176734_d());
+        if(net.minecraftforge.event.ForgeEventFactory.onNeighborNotify(p_176400_1_, p_176400_2_, p_176400_1_.func_180495_p(p_176400_2_), java.util.EnumSet.of(enumfacing.func_176734_d()), false).isCanceled())
+            return;
         p_176400_1_.func_190524_a(blockpos, this, p_176400_2_);
         p_176400_1_.func_175695_a(blockpos, this, enumfacing);
     }
 
-    @Override
     public void func_176206_d(World p_176206_1_, BlockPos p_176206_2_, IBlockState p_176206_3_)
     {
         if (this.field_149914_a)
@@ -272,7 +248,6 @@
         super.func_176206_d(p_176206_1_, p_176206_2_, p_176206_3_);
     }
 
-    @Override
     public boolean func_149662_c(IBlockState p_149662_1_)
     {
         return false;
@@ -301,7 +276,7 @@
 
     public boolean func_176402_i(World p_176402_1_, BlockPos p_176402_2_, IBlockState p_176402_3_)
     {
-        EnumFacing enumfacing = p_176402_3_.func_177229_b(field_185512_D).func_176734_d();
+        EnumFacing enumfacing = ((EnumFacing)p_176402_3_.func_177229_b(field_185512_D)).func_176734_d();
         BlockPos blockpos = p_176402_2_.func_177972_a(enumfacing);
 
         if (func_185546_B(p_176402_1_.func_180495_p(blockpos)))
@@ -325,20 +300,36 @@
 
     protected abstract IBlockState func_180675_k(IBlockState p_180675_1_);
 
-    @Override
     public boolean func_149667_c(Block p_149667_1_)
     {
         return this.func_185547_C(p_149667_1_.func_176223_P());
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public BlockRenderLayer func_180664_k()
     {
         return BlockRenderLayer.CUTOUT;
     }
 
+    /* ======================================== FORGE START =====================================*/
     @Override
+    public boolean rotateBlock(World world, BlockPos pos, EnumFacing axis)
+    {
+        if (super.rotateBlock(world, pos, axis))
+        {
+            IBlockState state = world.func_180495_p(pos);
+            state = func_180675_k(state);
+            world.func_175656_a(pos, state);
+
+            if (func_176404_e(world, pos, state))
+            {
+                world.func_175684_a(pos, this, 1);
+            }
+            return true;
+        }
+        return false;
+    }
+
     public BlockFaceShape func_193383_a(IBlockAccess p_193383_1_, IBlockState p_193383_2_, BlockPos p_193383_3_, EnumFacing p_193383_4_)
     {
         return p_193383_4_ == EnumFacing.DOWN ? BlockFaceShape.SOLID : BlockFaceShape.UNDEFINED;
