--- before/net/minecraft/block/BlockDispenser.java
+++ after/net/minecraft/block/BlockDispenser.java
@@ -2,6 +2,7 @@
 
 import java.util.Random;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyBool;
 import net.minecraft.block.properties.PropertyDirection;
 import net.minecraft.block.state.BlockStateContainer;
@@ -35,13 +36,13 @@
 {
     public static final PropertyDirection field_176441_a = BlockDirectional.field_176387_N;
     public static final PropertyBool field_176440_b = PropertyBool.func_177716_a("triggered");
-    public static final RegistryDefaulted<Item, IBehaviorDispenseItem> field_149943_a = new RegistryDefaulted<>(new BehaviorDefaultDispenseItem());
+    public static final RegistryDefaulted<Item, IBehaviorDispenseItem> field_149943_a = new RegistryDefaulted<Item, IBehaviorDispenseItem>(new BehaviorDefaultDispenseItem());
     protected Random field_149942_b = new Random();
 
     protected BlockDispenser()
     {
         super(Material.field_151576_e);
-        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176441_a, EnumFacing.NORTH).func_177226_a(field_176440_b, false));
+        this.func_180632_j(this.field_176227_L.func_177621_b().func_177226_a(field_176441_a, EnumFacing.NORTH).func_177226_a(field_176440_b, Boolean.valueOf(false)));
         this.func_149647_a(CreativeTabs.field_78028_d);
     }
 
@@ -60,7 +61,7 @@
     {
         if (!p_176438_1_.field_72995_K)
         {
-            EnumFacing enumfacing = p_176438_3_.func_177229_b(field_176441_a);
+            EnumFacing enumfacing = (EnumFacing)p_176438_3_.func_177229_b(field_176441_a);
             boolean flag = p_176438_1_.func_180495_p(p_176438_2_.func_177978_c()).func_185913_b();
             boolean flag1 = p_176438_1_.func_180495_p(p_176438_2_.func_177968_d()).func_185913_b();
 
@@ -87,7 +88,7 @@
                 }
             }
 
-            p_176438_1_.func_180501_a(p_176438_2_, p_176438_3_.func_177226_a(field_176441_a, enumfacing).func_177226_a(field_176440_b, false), 2);
+            p_176438_1_.func_180501_a(p_176438_2_, p_176438_3_.func_177226_a(field_176441_a, enumfacing).func_177226_a(field_176440_b, Boolean.valueOf(false)), 2);
         }
     }
 
@@ -122,7 +123,7 @@
     protected void func_176439_d(World p_176439_1_, BlockPos p_176439_2_)
     {
         BlockSourceImpl blocksourceimpl = new BlockSourceImpl(p_176439_1_, p_176439_2_);
-        TileEntityDispenser tileentitydispenser = blocksourceimpl.func_150835_j();
+        TileEntityDispenser tileentitydispenser = (TileEntityDispenser)blocksourceimpl.func_150835_j();
 
         if (tileentitydispenser != null)
         {
@@ -153,16 +154,16 @@
     public void func_189540_a(IBlockState p_189540_1_, World p_189540_2_, BlockPos p_189540_3_, Block p_189540_4_, BlockPos p_189540_5_)
     {
         boolean flag = p_189540_2_.func_175640_z(p_189540_3_) || p_189540_2_.func_175640_z(p_189540_3_.func_177984_a());
-        boolean flag1 = p_189540_1_.func_177229_b(field_176440_b);
+        boolean flag1 = ((Boolean)p_189540_1_.func_177229_b(field_176440_b)).booleanValue();
 
         if (flag && !flag1)
         {
             p_189540_2_.func_175684_a(p_189540_3_, this, this.func_149738_a(p_189540_2_));
-            p_189540_2_.func_180501_a(p_189540_3_, p_189540_1_.func_177226_a(field_176440_b, true), 4);
+            p_189540_2_.func_180501_a(p_189540_3_, p_189540_1_.func_177226_a(field_176440_b, Boolean.valueOf(true)), 4);
         }
         else if (!flag && flag1)
         {
-            p_189540_2_.func_180501_a(p_189540_3_, p_189540_1_.func_177226_a(field_176440_b, false), 4);
+            p_189540_2_.func_180501_a(p_189540_3_, p_189540_1_.func_177226_a(field_176440_b, Boolean.valueOf(false)), 4);
         }
     }
 
@@ -181,7 +182,7 @@
 
     public IBlockState func_180642_a(World p_180642_1_, BlockPos p_180642_2_, EnumFacing p_180642_3_, float p_180642_4_, float p_180642_5_, float p_180642_6_, int p_180642_7_, EntityLivingBase p_180642_8_)
     {
-        return this.func_176223_P().func_177226_a(field_176441_a, EnumFacing.func_190914_a(p_180642_2_, p_180642_8_)).func_177226_a(field_176440_b, false);
+        return this.func_176223_P().func_177226_a(field_176441_a, EnumFacing.func_190914_a(p_180642_2_, p_180642_8_)).func_177226_a(field_176440_b, Boolean.valueOf(false));
     }
 
     public void func_180633_a(World p_180633_1_, BlockPos p_180633_2_, IBlockState p_180633_3_, EntityLivingBase p_180633_4_, ItemStack p_180633_5_)
@@ -214,7 +215,7 @@
 
     public static IPosition func_149939_a(IBlockSource p_149939_0_)
     {
-        EnumFacing enumfacing = p_149939_0_.func_189992_e().func_177229_b(field_176441_a);
+        EnumFacing enumfacing = (EnumFacing)p_149939_0_.func_189992_e().func_177229_b(field_176441_a);
         double d0 = p_149939_0_.func_82615_a() + 0.7D * (double)enumfacing.func_82601_c();
         double d1 = p_149939_0_.func_82617_b() + 0.7D * (double)enumfacing.func_96559_d();
         double d2 = p_149939_0_.func_82616_c() + 0.7D * (double)enumfacing.func_82599_e();
@@ -238,15 +239,15 @@
 
     public IBlockState func_176203_a(int p_176203_1_)
     {
-        return this.func_176223_P().func_177226_a(field_176441_a, EnumFacing.func_82600_a(p_176203_1_ & 7)).func_177226_a(field_176440_b, (p_176203_1_ & 8) > 0);
+        return this.func_176223_P().func_177226_a(field_176441_a, EnumFacing.func_82600_a(p_176203_1_ & 7)).func_177226_a(field_176440_b, Boolean.valueOf((p_176203_1_ & 8) > 0));
     }
 
     public int func_176201_c(IBlockState p_176201_1_)
     {
         int i = 0;
-        i |= p_176201_1_.func_177229_b(field_176441_a).func_176745_a();
+        i = i | ((EnumFacing)p_176201_1_.func_177229_b(field_176441_a)).func_176745_a();
 
-        if (p_176201_1_.func_177229_b(field_176440_b))
+        if (((Boolean)p_176201_1_.func_177229_b(field_176440_b)).booleanValue())
         {
             i |= 8;
         }
@@ -256,16 +257,16 @@
 
     public IBlockState func_185499_a(IBlockState p_185499_1_, Rotation p_185499_2_)
     {
-        return p_185499_1_.func_177226_a(field_176441_a, p_185499_2_.func_185831_a(p_185499_1_.func_177229_b(field_176441_a)));
+        return p_185499_1_.func_177226_a(field_176441_a, p_185499_2_.func_185831_a((EnumFacing)p_185499_1_.func_177229_b(field_176441_a)));
     }
 
     public IBlockState func_185471_a(IBlockState p_185471_1_, Mirror p_185471_2_)
     {
-        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a(p_185471_1_.func_177229_b(field_176441_a)));
+        return p_185471_1_.func_185907_a(p_185471_2_.func_185800_a((EnumFacing)p_185471_1_.func_177229_b(field_176441_a)));
     }
 
     protected BlockStateContainer func_180661_e()
     {
-        return new BlockStateContainer(this, field_176441_a, field_176440_b);
+        return new BlockStateContainer(this, new IProperty[] {field_176441_a, field_176440_b});
     }
 }
