--- before/net/minecraft/block/BlockReed.java
+++ after/net/minecraft/block/BlockReed.java
@@ -20,7 +20,7 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class BlockReed extends Block
+public class BlockReed extends Block implements net.minecraftforge.common.IPlantable
 {
     public static final PropertyInteger AGE = PropertyInteger.create("age", 0, 15);
     protected static final AxisAlignedBB REED_AABB = new AxisAlignedBB(0.125, 0.0, 0.125, 0.875, 1.0, 0.875);
@@ -57,6 +57,8 @@
                 {
                     int j = state.getValue(AGE);
 
+                    if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, pos, state, true))
+                    {
                     if (j == 15)
                     {
                         worldIn.setBlockState(pos.up(), this.getDefaultState());
@@ -66,6 +68,8 @@
                     {
                         worldIn.setBlockState(pos, state.withProperty(AGE, j + 1), 4);
                     }
+                    net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state, worldIn.getBlockState(pos));
+                    }
                 }
             }
         }
@@ -74,7 +78,9 @@
     @Override
     public boolean canPlaceBlockAt(World worldIn, BlockPos pos)
     {
-        Block block = worldIn.getBlockState(pos.down()).getBlock();
+        IBlockState state = worldIn.getBlockState(pos.down());
+        Block block = state.getBlock();
+        if (block.canSustainPlant(state, worldIn, pos.down(), EnumFacing.UP, this)) return true;
 
         if (block == this)
         {
@@ -175,6 +181,17 @@
     public int getMetaFromState(IBlockState state)
     {
         return state.getValue(AGE);
+    }
+
+    @Override
+    public net.minecraftforge.common.EnumPlantType getPlantType(IBlockAccess world, BlockPos pos)
+    {
+        return net.minecraftforge.common.EnumPlantType.Beach;
+    }
+    @Override
+    public IBlockState getPlant(IBlockAccess world, BlockPos pos)
+    {
+        return this.getDefaultState();
     }
 
     @Override
