--- before/net/minecraft/block/BlockMushroom.java
+++ after/net/minecraft/block/BlockMushroom.java
@@ -90,7 +90,7 @@
             {
                 return iblockstate.getBlock() == Blocks.DIRT && iblockstate.getValue(BlockDirt.VARIANT) == BlockDirt.DirtType.PODZOL
                        ? true
-                       : worldIn.getLight(pos) < 13 && this.canSustainBush(iblockstate);
+                       : worldIn.getLight(pos) < 13 && iblockstate.getBlock().canSustainPlant(iblockstate, worldIn, pos.down(), net.minecraft.util.EnumFacing.UP, this);
             }
         }
         else
