--- before/net/minecraft/block/BlockSkull.java
+++ after/net/minecraft/block/BlockSkull.java
@@ -148,11 +148,6 @@
     }
 
     @Override
-    public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune)
-    {
-    }
-
-    @Override
     public void onBlockHarvested(World worldIn, BlockPos pos, IBlockState state, EntityPlayer player)
     {
         if (player.capabilities.isCreativeMode)
@@ -160,6 +155,7 @@
             state = state.withProperty(NODROP, true);
             worldIn.setBlockState(pos, state, 4);
         }
+        this.dropBlockAsItem(worldIn, pos, state, 0);
 
         super.onBlockHarvested(worldIn, pos, state, player);
     }
@@ -167,7 +163,12 @@
     @Override
     public void breakBlock(World worldIn, BlockPos pos, IBlockState state)
     {
-        if (!worldIn.isRemote)
+        super.breakBlock(worldIn, pos, state);
+    }
+    
+    @Override
+    public void getDrops(net.minecraft.util.NonNullList<ItemStack> drops, IBlockAccess worldIn, BlockPos pos, IBlockState state, int fortune)
+    {
         {
             if (!state.getValue(NODROP))
             {
@@ -176,7 +177,7 @@
                 if (tileentity instanceof TileEntitySkull)
                 {
                     TileEntitySkull tileentityskull = (TileEntitySkull)tileentity;
-                    ItemStack itemstack = this.getItem(worldIn, pos, state);
+                    ItemStack itemstack = new ItemStack(Items.SKULL, 1, tileentityskull.getSkullType());
 
                     if (tileentityskull.getSkullType() == 3 && tileentityskull.getPlayerProfile() != null)
                     {
@@ -186,11 +187,9 @@
                         itemstack.getTagCompound().setTag("SkullOwner", nbttagcompound);
                     }
 
-                    spawnAsEntity(worldIn, pos, itemstack);
+                    drops.add(itemstack);
                 }
             }
-
-            super.breakBlock(worldIn, pos, state);
         }
     }
 
