--- before/net/minecraft/block/BlockDropper.java
+++ after/net/minecraft/block/BlockDropper.java
@@ -16,23 +16,20 @@
 {
     private final IBehaviorDispenseItem field_149947_P = new BehaviorDefaultDispenseItem();
 
-    @Override
     protected IBehaviorDispenseItem func_149940_a(ItemStack p_149940_1_)
     {
         return this.field_149947_P;
     }
 
-    @Override
     public TileEntity func_149915_a(World p_149915_1_, int p_149915_2_)
     {
         return new TileEntityDropper();
     }
 
-    @Override
     protected void func_176439_d(World p_176439_1_, BlockPos p_176439_2_)
     {
         BlockSourceImpl blocksourceimpl = new BlockSourceImpl(p_176439_1_, p_176439_2_);
-        TileEntityDispenser tileentitydispenser = blocksourceimpl.func_150835_j();
+        TileEntityDispenser tileentitydispenser = (TileEntityDispenser)blocksourceimpl.func_150835_j();
 
         if (tileentitydispenser != null)
         {
@@ -46,13 +43,11 @@
             {
                 ItemStack itemstack = tileentitydispenser.func_70301_a(i);
 
-                if (!itemstack.func_190926_b())
+                if (!itemstack.func_190926_b() && net.minecraftforge.items.VanillaInventoryCodeHooks.dropperInsertHook(p_176439_1_, p_176439_2_, tileentitydispenser, i, itemstack))
                 {
-                    EnumFacing enumfacing = p_176439_1_.func_180495_p(p_176439_2_).func_177229_b(field_176441_a);
+                    EnumFacing enumfacing = (EnumFacing)p_176439_1_.func_180495_p(p_176439_2_).func_177229_b(field_176441_a);
                     BlockPos blockpos = p_176439_2_.func_177972_a(enumfacing);
-                    IInventory iinventory = TileEntityHopper.func_145893_b(
-                                                p_176439_1_, (double)blockpos.func_177958_n(), (double)blockpos.func_177956_o(), (double)blockpos.func_177952_p()
-                                            );
+                    IInventory iinventory = TileEntityHopper.func_145893_b(p_176439_1_, (double)blockpos.func_177958_n(), (double)blockpos.func_177956_o(), (double)blockpos.func_177952_p());
                     ItemStack itemstack1;
 
                     if (iinventory == null)
@@ -61,9 +56,7 @@
                     }
                     else
                     {
-                        itemstack1 = TileEntityHopper.func_174918_a(
-                                         tileentitydispenser, iinventory, itemstack.func_77946_l().func_77979_a(1), enumfacing.func_176734_d()
-                                     );
+                        itemstack1 = TileEntityHopper.func_174918_a(tileentitydispenser, iinventory, itemstack.func_77946_l().func_77979_a(1), enumfacing.func_176734_d());
 
                         if (itemstack1.func_190926_b())
                         {
