--- original/net/minecraft/client/network/NetHandlerHandshakeMemory.java
+++ modified/net/minecraft/client/network/NetHandlerHandshakeMemory.java
@@ -22,6 +22,7 @@
     /** There are two recognized intentions for initiating a handshake: logging in and acquiring server status. The NetworkManager's protocol will be reconfigured according to the specified intention, although a login-intention must pass a versioncheck or receive a disconnect otherwise */
     @Override
     public void processHandshake(C00Handshake packetIn) {
+        if (!net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerHandshake(packetIn, this.networkManager)) return;
         this.networkManager.setConnectionState(packetIn.getRequestedState());
         this.networkManager.setNetHandler(new NetHandlerLoginServer(this.server, this.networkManager));
     }
