--- before/net/minecraft/nbt/NBTTagString.java
+++ after/net/minecraft/nbt/NBTTagString.java
@@ -20,27 +20,23 @@
         this.field_74751_a = p_i1389_1_;
     }
 
-    @Override
     void func_74734_a(DataOutput p_74734_1_) throws IOException
     {
         p_74734_1_.writeUTF(this.field_74751_a);
     }
 
-    @Override
     void func_152446_a(DataInput p_152446_1_, int p_152446_2_, NBTSizeTracker p_152446_3_) throws IOException
     {
         p_152446_3_.func_152450_a(288L);
         this.field_74751_a = p_152446_1_.readUTF();
-        p_152446_3_.func_152450_a((long)(16 * this.field_74751_a.length()));
+        NBTSizeTracker.readUTF(p_152446_3_, field_74751_a); // Forge: Correctly read String length including header.
     }
 
-    @Override
     public byte func_74732_a()
     {
         return 8;
     }
 
-    @Override
     public String toString()
     {
         return func_193588_a(this.field_74751_a);
@@ -51,13 +47,11 @@
         return new NBTTagString(this.field_74751_a);
     }
 
-    @Override
     public boolean func_82582_d()
     {
         return this.field_74751_a.isEmpty();
     }
 
-    @Override
     public boolean equals(Object p_equals_1_)
     {
         if (!super.equals(p_equals_1_))
@@ -71,13 +65,11 @@
         }
     }
 
-    @Override
     public int hashCode()
     {
         return super.hashCode() ^ this.field_74751_a.hashCode();
     }
 
-    @Override
     public String func_150285_a_()
     {
         return this.field_74751_a;
@@ -87,7 +79,7 @@
     {
         StringBuilder stringbuilder = new StringBuilder("\"");
 
-        for (int i = 0; i < p_193588_0_.length(); i++)
+        for (int i = 0; i < p_193588_0_.length(); ++i)
         {
             char c0 = p_193588_0_.charAt(i);
 
