--- before/net/minecraft/nbt/NBTTagCompound.java
+++ after/net/minecraft/nbt/NBTTagCompound.java
@@ -25,9 +25,8 @@
 {
     private static final Logger field_191551_b = LogManager.getLogger();
     private static final Pattern field_193583_c = Pattern.compile("[A-Za-z0-9._+-]+");
-    private final Map<String, NBTBase> field_74784_a = Maps.newHashMap();
+    private final Map<String, NBTBase> field_74784_a = Maps.<String, NBTBase>newHashMap();
 
-    @Override
     void func_74734_a(DataOutput p_74734_1_) throws IOException
     {
         for (String s : this.field_74784_a.keySet())
@@ -39,7 +38,6 @@
         p_74734_1_.writeByte(0);
     }
 
-    @Override
     void func_152446_a(DataInput p_152446_1_, int p_152446_2_, NBTSizeTracker p_152446_3_) throws IOException
     {
         p_152446_3_.func_152450_a(384L);
@@ -72,7 +70,6 @@
         return this.field_74784_a.keySet();
     }
 
-    @Override
     public byte func_74732_a()
     {
         return 10;
@@ -85,6 +82,7 @@
 
     public void func_74782_a(String p_74782_1_, NBTBase p_74782_2_)
     {
+        if (p_74782_2_ == null) throw new IllegalArgumentException("Invalid null NBT value with key " + p_74782_1_);
         this.field_74784_a.put(p_74782_1_, p_74782_2_);
     }
 
@@ -179,9 +177,13 @@
         {
             return true;
         }
+        else if (p_150297_2_ != 99)
+        {
+            return false;
+        }
         else
         {
-            return p_150297_2_ != 99 ? false : i == 1 || i == 2 || i == 3 || i == 4 || i == 5 || i == 6;
+            return i == 1 || i == 2 || i == 3 || i == 4 || i == 5 || i == 6;
         }
     }
 
@@ -194,8 +196,9 @@
                 return ((NBTPrimitive)this.field_74784_a.get(p_74771_1_)).func_150290_f();
             }
         }
-        catch (ClassCastException classcastexception)
+        catch (ClassCastException var3)
         {
+            ;
         }
 
         return 0;
@@ -210,8 +213,9 @@
                 return ((NBTPrimitive)this.field_74784_a.get(p_74765_1_)).func_150289_e();
             }
         }
-        catch (ClassCastException classcastexception)
+        catch (ClassCastException var3)
         {
+            ;
         }
 
         return 0;
@@ -226,8 +230,9 @@
                 return ((NBTPrimitive)this.field_74784_a.get(p_74762_1_)).func_150287_d();
             }
         }
-        catch (ClassCastException classcastexception)
+        catch (ClassCastException var3)
         {
+            ;
         }
 
         return 0;
@@ -242,8 +247,9 @@
                 return ((NBTPrimitive)this.field_74784_a.get(p_74763_1_)).func_150291_c();
             }
         }
-        catch (ClassCastException classcastexception)
+        catch (ClassCastException var3)
         {
+            ;
         }
 
         return 0L;
@@ -258,8 +264,9 @@
                 return ((NBTPrimitive)this.field_74784_a.get(p_74760_1_)).func_150288_h();
             }
         }
-        catch (ClassCastException classcastexception)
+        catch (ClassCastException var3)
         {
+            ;
         }
 
         return 0.0F;
@@ -274,11 +281,12 @@
                 return ((NBTPrimitive)this.field_74784_a.get(p_74769_1_)).func_150286_g();
             }
         }
-        catch (ClassCastException classcastexception)
+        catch (ClassCastException var3)
         {
+            ;
         }
 
-        return 0.0;
+        return 0.0D;
     }
 
     public String func_74779_i(String p_74779_1_)
@@ -287,11 +295,12 @@
         {
             if (this.func_150297_b(p_74779_1_, 8))
             {
-                return this.field_74784_a.get(p_74779_1_).func_150285_a_();
+                return ((NBTBase)this.field_74784_a.get(p_74779_1_)).func_150285_a_();
             }
         }
-        catch (ClassCastException classcastexception)
+        catch (ClassCastException var3)
         {
+            ;
         }
 
         return "";
@@ -382,7 +391,6 @@
         this.field_74784_a.remove(p_82580_1_);
     }
 
-    @Override
     public String toString()
     {
         StringBuilder stringbuilder = new StringBuilder("{");
@@ -408,7 +416,6 @@
         return stringbuilder.append('}').toString();
     }
 
-    @Override
     public boolean func_82582_d()
     {
         return this.field_74784_a.isEmpty();
@@ -422,7 +429,7 @@
         {
             public String call() throws Exception
             {
-                return NBTBase.field_82578_b[NBTTagCompound.this.field_74784_a.get(p_82581_1_).func_74732_a()];
+                return NBTBase.field_82578_b[((NBTBase)NBTTagCompound.this.field_74784_a.get(p_82581_1_)).func_74732_a()];
             }
         });
         crashreportcategory.func_189529_a("Tag type expected", new ICrashReportDetail<String>()
@@ -442,19 +449,17 @@
 
         for (String s : this.field_74784_a.keySet())
         {
-            nbttagcompound.func_74782_a(s, this.field_74784_a.get(s).func_74737_b());
+            nbttagcompound.func_74782_a(s, ((NBTBase)this.field_74784_a.get(s)).func_74737_b());
         }
 
         return nbttagcompound;
     }
 
-    @Override
     public boolean equals(Object p_equals_1_)
     {
         return super.equals(p_equals_1_) && Objects.equals(this.field_74784_a.entrySet(), ((NBTTagCompound)p_equals_1_).field_74784_a.entrySet());
     }
 
-    @Override
     public int hashCode()
     {
         return super.hashCode() ^ this.field_74784_a.hashCode();
@@ -473,6 +478,7 @@
 
     private static byte func_152447_a(DataInput p_152447_0_, NBTSizeTracker p_152447_1_) throws IOException
     {
+        p_152447_1_.func_152450_a(8);
         return p_152447_0_.readByte();
     }
 
@@ -483,6 +489,7 @@
 
     static NBTBase func_152449_a(byte p_152449_0_, String p_152449_1_, DataInput p_152449_2_, int p_152449_3_, NBTSizeTracker p_152449_4_) throws IOException
     {
+        p_152449_4_.func_152450_a(32); //Forge: 4 extra bytes for the object allocation.
         NBTBase nbtbase = NBTBase.func_150284_a(p_152449_0_);
 
         try
@@ -495,7 +502,7 @@
             CrashReport crashreport = CrashReport.func_85055_a(ioexception, "Loading NBT data");
             CrashReportCategory crashreportcategory = crashreport.func_85058_a("NBT Tag");
             crashreportcategory.func_71507_a("Tag name", p_152449_1_);
-            crashreportcategory.func_71507_a("Tag type", p_152449_0_);
+            crashreportcategory.func_71507_a("Tag type", Byte.valueOf(p_152449_0_));
             throw new ReportedException(crashreport);
         }
     }
