--- before/net/minecraft/nbt/NBTTagCompound.java
+++ after/net/minecraft/nbt/NBTTagCompound.java
@@ -85,6 +85,7 @@
 
     public void setTag(String key, NBTBase value)
     {
+        if (value == null) throw new IllegalArgumentException("Invalid null NBT value with key " + key);
         this.tagMap.put(key, value);
     }
 
@@ -473,6 +474,7 @@
 
     private static byte readType(DataInput input, NBTSizeTracker sizeTracker) throws IOException
     {
+        sizeTracker.read(8);
         return input.readByte();
     }
 
@@ -483,6 +485,7 @@
 
     static NBTBase readNBT(byte id, String key, DataInput input, int depth, NBTSizeTracker sizeTracker) throws IOException
     {
+        sizeTracker.read(32); //Forge: 4 extra bytes for the object allocation.
         NBTBase nbtbase = NBTBase.create(id);
 
         try
