--- before/net/minecraft/nbt/NBTTagList.java
+++ after/net/minecraft/nbt/NBTTagList.java
@@ -9,7 +9,7 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class NBTTagList extends NBTBase
+public class NBTTagList extends NBTBase implements java.lang.Iterable<NBTBase>
 {
     private static final Logger LOGGER = LogManager.getLogger();
     private List<NBTBase> tagList = Lists.newArrayList();
@@ -288,4 +288,5 @@
     {
         return this.tagType;
     }
+    @Override public java.util.Iterator<NBTBase> iterator() {return tagList.iterator();}
 }
