--- before/net/minecraft/nbt/NBTUtil.java
+++ after/net/minecraft/nbt/NBTUtil.java
@@ -2,10 +2,11 @@
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Optional;
+import com.google.common.collect.UnmodifiableIterator;
 import com.mojang.authlib.GameProfile;
 import com.mojang.authlib.properties.Property;
-import java.util.Map;
 import java.util.UUID;
+import java.util.Map.Entry;
 import javax.annotation.Nullable;
 import net.minecraft.block.Block;
 import net.minecraft.block.properties.IProperty;
@@ -46,7 +47,7 @@
             {
                 uuid = UUID.fromString(s1);
             }
-            catch (Throwable throwable)
+            catch (Throwable var12)
             {
                 uuid = null;
             }
@@ -80,7 +81,7 @@
 
             return gameprofile;
         }
-        catch (Throwable throwable1)
+        catch (Throwable var13)
         {
             return null;
         }
@@ -279,15 +280,17 @@
 
     public static NBTTagCompound func_190009_a(NBTTagCompound p_190009_0_, IBlockState p_190009_1_)
     {
-        p_190009_0_.func_74778_a("Name", Block.field_149771_c.func_177774_c(p_190009_1_.func_177230_c()).toString());
+        p_190009_0_.func_74778_a("Name", ((ResourceLocation)Block.field_149771_c.func_177774_c(p_190009_1_.func_177230_c())).toString());
 
         if (!p_190009_1_.func_177228_b().isEmpty())
         {
             NBTTagCompound nbttagcompound = new NBTTagCompound();
+            UnmodifiableIterator unmodifiableiterator = p_190009_1_.func_177228_b().entrySet().iterator();
 
-            for (Map.Entry < IProperty<?>, Comparable<? >> entry : p_190009_1_.func_177228_b().entrySet())
+            while (unmodifiableiterator.hasNext())
             {
-                IProperty<?> iproperty = entry.getKey();
+                Entry < IProperty<?>, Comparable<? >> entry = (Entry)unmodifiableiterator.next();
+                IProperty<?> iproperty = (IProperty)entry.getKey();
                 nbttagcompound.func_74778_a(iproperty.func_177701_a(), func_190010_a(iproperty, entry.getValue()));
             }
 
@@ -296,6 +299,7 @@
 
         return p_190009_0_;
     }
+
     @SuppressWarnings("unchecked")
     private static <T extends Comparable<T>> String func_190010_a(IProperty<T> p_190010_0_, Comparable<?> p_190010_1_)
     {
