--- before/net/minecraft/world/chunk/BlockStatePaletteHashMap.java
+++ after/net/minecraft/world/chunk/BlockStatePaletteHashMap.java
@@ -18,10 +18,9 @@
     {
         this.field_186048_c = p_i47089_1_;
         this.field_186047_b = p_i47089_2_;
-        this.field_186046_a = new IntIdentityHashBiMap<>(1 << p_i47089_1_);
+        this.field_186046_a = new IntIdentityHashBiMap<IBlockState>(1 << p_i47089_1_);
     }
 
-    @Override
     public int func_186041_a(IBlockState p_186041_1_)
     {
         int i = this.field_186046_a.func_186815_a(p_186041_1_);
@@ -40,14 +39,12 @@
     }
 
     @Nullable
-    @Override
     public IBlockState func_186039_a(int p_186039_1_)
     {
         return this.field_186046_a.func_186813_a(p_186039_1_);
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_186038_a(PacketBuffer p_186038_1_)
     {
         this.field_186046_a.func_186812_a();
@@ -59,7 +56,6 @@
         }
     }
 
-    @Override
     public void func_186037_b(PacketBuffer p_186037_1_)
     {
         int i = this.field_186046_a.func_186810_b();
@@ -71,7 +67,6 @@
         }
     }
 
-    @Override
     public int func_186040_a()
     {
         int i = PacketBuffer.func_150790_a(this.field_186046_a.func_186810_b());
