--- before/net/minecraft/world/storage/DerivedWorldInfo.java
+++ after/net/minecraft/world/storage/DerivedWorldInfo.java
@@ -20,255 +20,223 @@
         this.field_76115_a = p_i2145_1_;
     }
 
-    @Override
     public NBTTagCompound func_76082_a(@Nullable NBTTagCompound p_76082_1_)
     {
         return this.field_76115_a.func_76082_a(p_76082_1_);
     }
 
-    @Override
     public long func_76063_b()
     {
         return this.field_76115_a.func_76063_b();
     }
 
-    @Override
     public int func_76079_c()
     {
         return this.field_76115_a.func_76079_c();
     }
 
-    @Override
     public int func_76075_d()
     {
         return this.field_76115_a.func_76075_d();
     }
 
-    @Override
     public int func_76074_e()
     {
         return this.field_76115_a.func_76074_e();
     }
 
-    @Override
     public long func_82573_f()
     {
         return this.field_76115_a.func_82573_f();
     }
 
-    @Override
     public long func_76073_f()
     {
         return this.field_76115_a.func_76073_f();
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public long func_76092_g()
     {
         return this.field_76115_a.func_76092_g();
     }
 
-    @Override
     public NBTTagCompound func_76072_h()
     {
         return this.field_76115_a.func_76072_h();
     }
 
-    @Override
     public String func_76065_j()
     {
         return this.field_76115_a.func_76065_j();
     }
 
-    @Override
     public int func_76088_k()
     {
         return this.field_76115_a.func_76088_k();
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public long func_76057_l()
     {
         return this.field_76115_a.func_76057_l();
     }
 
-    @Override
     public boolean func_76061_m()
     {
         return this.field_76115_a.func_76061_m();
     }
 
-    @Override
     public int func_76071_n()
     {
         return this.field_76115_a.func_76071_n();
     }
 
-    @Override
     public boolean func_76059_o()
     {
         return this.field_76115_a.func_76059_o();
     }
 
-    @Override
     public int func_76083_p()
     {
         return this.field_76115_a.func_76083_p();
     }
 
-    @Override
     public GameType func_76077_q()
     {
         return this.field_76115_a.func_76077_q();
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_76058_a(int p_76058_1_)
     {
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_76056_b(int p_76056_1_)
     {
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
-    public void func_76087_c(int p_76087_1_)
-    {
-    }
-
-    @Override
     public void func_82572_b(long p_82572_1_)
     {
     }
 
-    @Override
+    @SideOnly(Side.CLIENT)
+    public void func_76087_c(int p_76087_1_)
+    {
+    }
+
     public void func_76068_b(long p_76068_1_)
     {
     }
 
-    @Override
     public void func_176143_a(BlockPos p_176143_1_)
     {
     }
 
-    @Override
     public void func_76062_a(String p_76062_1_)
     {
     }
 
-    @Override
     public void func_76078_e(int p_76078_1_)
     {
     }
 
-    @Override
     public void func_76069_a(boolean p_76069_1_)
     {
     }
 
-    @Override
     public void func_76090_f(int p_76090_1_)
     {
     }
 
-    @Override
     public void func_76084_b(boolean p_76084_1_)
     {
     }
 
-    @Override
     public void func_76080_g(int p_76080_1_)
     {
     }
 
-    @Override
     public boolean func_76089_r()
     {
         return this.field_76115_a.func_76089_r();
     }
 
-    @Override
     public boolean func_76093_s()
     {
         return this.field_76115_a.func_76093_s();
     }
 
-    @Override
     public WorldType func_76067_t()
     {
         return this.field_76115_a.func_76067_t();
     }
 
-    @Override
     public void func_76085_a(WorldType p_76085_1_)
     {
     }
 
-    @Override
     public boolean func_76086_u()
     {
         return this.field_76115_a.func_76086_u();
     }
 
-    @Override
     public void func_176121_c(boolean p_176121_1_)
     {
     }
 
-    @Override
     public boolean func_76070_v()
     {
         return this.field_76115_a.func_76070_v();
     }
 
-    @Override
     public void func_76091_d(boolean p_76091_1_)
     {
     }
 
-    @Override
     public GameRules func_82574_x()
     {
         return this.field_76115_a.func_82574_x();
     }
 
-    @Override
     public EnumDifficulty func_176130_y()
     {
         return this.field_76115_a.func_176130_y();
     }
 
-    @Override
     public void func_176144_a(EnumDifficulty p_176144_1_)
     {
     }
 
-    @Override
     public boolean func_176123_z()
     {
         return this.field_76115_a.func_176123_z();
     }
 
-    @Override
     public void func_180783_e(boolean p_180783_1_)
     {
     }
 
-    @Override
+    @Deprecated
     public void func_186345_a(DimensionType p_186345_1_, NBTTagCompound p_186345_2_)
     {
         this.field_76115_a.func_186345_a(p_186345_1_, p_186345_2_);
     }
 
-    @Override
+    @Deprecated
     public NBTTagCompound func_186347_a(DimensionType p_186347_1_)
     {
         return this.field_76115_a.func_186347_a(p_186347_1_);
+    }
+
+    public void setDimensionData(int dimensionID, NBTTagCompound compound)
+    {
+        this.field_76115_a.setDimensionData(dimensionID, compound);
+    }
+
+    public NBTTagCompound getDimensionData(int dimensionID)
+    {
+        return this.field_76115_a.getDimensionData(dimensionID);
     }
 }
