--- before/net/minecraft/world/storage/ISaveFormat.java
+++ after/net/minecraft/world/storage/ISaveFormat.java
@@ -18,6 +18,8 @@
     @SideOnly(Side.CLIENT)
     List<WorldSummary> func_75799_b() throws AnvilConverterException;
 
+    boolean func_75801_b(String p_75801_1_);
+
     @SideOnly(Side.CLIENT)
     void func_75800_d();
 
@@ -37,12 +39,10 @@
     @SideOnly(Side.CLIENT)
     boolean func_154334_a(String p_154334_1_);
 
-    boolean func_75801_b(String p_75801_1_);
-
     boolean func_75805_a(String p_75805_1_, IProgressUpdate p_75805_2_);
 
-    @SideOnly(Side.CLIENT)
-    boolean func_90033_f(String p_90033_1_);
-
     File func_186352_b(String p_186352_1_, String p_186352_2_);
+
+    @SideOnly(Side.CLIENT)
+    boolean func_90033_f(String p_90033_1_);
 }
