--- before/net/minecraft/world/storage/MapData.java
+++ after/net/minecraft/world/storage/MapData.java
@@ -20,7 +20,7 @@
 {
     public int xCenter;
     public int zCenter;
-    public byte dimension;
+    public int dimension; //FML byte -> int
     public boolean trackingPosition;
     public boolean unlimitedTracking;
     public byte scale;
@@ -46,7 +46,7 @@
     @Override
     public void readFromNBT(NBTTagCompound nbt)
     {
-        this.dimension = nbt.getByte("dimension");
+        this.dimension = nbt.getInteger("dimension");
         this.xCenter = nbt.getInteger("xCenter");
         this.zCenter = nbt.getInteger("zCenter");
         this.scale = nbt.getByte("scale");
@@ -99,7 +99,7 @@
     @Override
     public NBTTagCompound writeToNBT(NBTTagCompound compound)
     {
-        compound.setByte("dimension", this.dimension);
+        compound.setInteger("dimension", this.dimension);
         compound.setInteger("xCenter", this.xCenter);
         compound.setInteger("zCenter", this.zCenter);
         compound.setByte("scale", this.scale);
@@ -232,9 +232,9 @@
             rotationIn += rotationIn < 0.0 ? -8.0 : 8.0;
             b2 = (byte)(rotationIn * 16.0 / 360.0);
 
-            if (this.dimension < 0)
+            if (worldIn.provider.shouldMapSpin(decorationName, worldX, worldZ, rotationIn))
             {
-                int l = (int)(worldIn.getWorldInfo().getWorldTime() / 10L);
+                int l = (int)(worldIn.getWorldTime() / 10L);
                 b2 = (byte)(l * l * 34187121 + l * 121 >> 15 & 15);
             }
         }
