--- before/net/minecraft/world/storage/SaveFormatOld.java
+++ after/net/minecraft/world/storage/SaveFormatOld.java
@@ -44,7 +44,7 @@
     @SideOnly(Side.CLIENT)
     public List<WorldSummary> func_75799_b() throws AnvilConverterException
     {
-        List<WorldSummary> list = Lists.newArrayList();
+        List<WorldSummary> list = Lists.<WorldSummary>newArrayList();
 
         for (int i = 0; i < 5; ++i)
         {
@@ -109,6 +109,24 @@
         }
     }
 
+    //Forge: Special version of the above that runs during actual world loading and not metadata gathering.
+    public static WorldInfo loadAndFix(File file, DataFixer fixer, SaveHandler save)
+    {
+        try
+        {
+            NBTTagCompound nbt = CompressedStreamTools.func_74796_a(new FileInputStream(file));
+            WorldInfo info = new WorldInfo(fixer.func_188257_a(FixTypes.LEVEL, nbt.func_74775_l("Data")));
+            net.minecraftforge.fml.common.FMLCommonHandler.instance().handleWorldDataLoad(save, info, nbt);
+            return info;
+        }
+        catch (net.minecraftforge.fml.common.StartupQuery.AbortedException e) { throw e; }
+        catch (Exception exception)
+        {
+            field_151479_b.error((String)("Exception reading " + file), (Throwable)exception);
+            return null;
+        }
+    }
+
     @SideOnly(Side.CLIENT)
     public void func_75806_a(String p_75806_1_, String p_75806_2_)
     {
@@ -135,6 +153,11 @@
         }
     }
 
+    public ISaveHandler func_75804_a(String p_75804_1_, boolean p_75804_2_)
+    {
+        return new SaveHandler(this.field_75808_a, p_75804_1_, p_75804_2_, this.field_186354_b);
+    }
+
     @SideOnly(Side.CLIENT)
     public boolean func_154335_d(String p_154335_1_)
     {
@@ -190,8 +213,9 @@
                     {
                         Thread.sleep(500L);
                     }
-                    catch (InterruptedException interruptedexception)
+                    catch (InterruptedException var5)
                     {
+                        ;
                     }
                 }
             }
@@ -221,11 +245,6 @@
         }
 
         return true;
-    }
-
-    public ISaveHandler func_75804_a(String p_75804_1_, boolean p_75804_2_)
-    {
-        return new SaveHandler(this.field_75808_a, p_75804_1_, p_75804_2_, this.field_186354_b);
     }
 
     @SideOnly(Side.CLIENT)
