--- before/net/minecraft/world/storage/SaveFormatOld.java
+++ after/net/minecraft/world/storage/SaveFormatOld.java
@@ -36,19 +36,17 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public String getName()
     {
         return "Old Format";
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public List<WorldSummary> getSaveList() throws AnvilConverterException
     {
-        List<WorldSummary> list = Lists.newArrayList();
+        List<WorldSummary> list = Lists.<WorldSummary>newArrayList();
 
-        for (int i = 0; i < 5; i++)
+        for (int i = 0; i < 5; ++i)
         {
             String s = "World" + (i + 1);
             WorldInfo worldinfo = this.getWorldInfo(s);
@@ -63,13 +61,11 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void flushCache()
     {
     }
 
     @Nullable
-    @Override
     public WorldInfo getWorldInfo(String saveName)
     {
         File file1 = new File(this.savesDirectory, saveName);
@@ -113,8 +109,25 @@
         }
     }
 
+    //Forge: Special version of the above that runs during actual world loading and not metadata gathering.
+    public static WorldInfo loadAndFix(File file, DataFixer fixer, SaveHandler save)
+    {
+        try
+        {
+            NBTTagCompound nbt = CompressedStreamTools.readCompressed(new FileInputStream(file));
+            WorldInfo info = new WorldInfo(fixer.process(FixTypes.LEVEL, nbt.getCompoundTag("Data")));
+            net.minecraftforge.fml.common.FMLCommonHandler.instance().handleWorldDataLoad(save, info, nbt);
+            return info;
+        }
+        catch (net.minecraftforge.fml.common.StartupQuery.AbortedException e) { throw e; }
+        catch (Exception exception)
+        {
+            LOGGER.error((String)("Exception reading " + file), (Throwable)exception);
+            return null;
+        }
+    }
+
     @SideOnly(Side.CLIENT)
-    @Override
     public void renameWorld(String dirName, String newName)
     {
         File file1 = new File(this.savesDirectory, dirName);
@@ -140,8 +153,12 @@
         }
     }
 
+    public ISaveHandler getSaveLoader(String saveName, boolean storePlayerdata)
+    {
+        return new SaveHandler(this.savesDirectory, saveName, storePlayerdata, this.dataFixer);
+    }
+
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean isNewLevelIdAcceptable(String saveName)
     {
         File file1 = new File(this.savesDirectory, saveName);
@@ -167,7 +184,6 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean deleteWorldDirectory(String saveName)
     {
         File file1 = new File(this.savesDirectory, saveName);
@@ -178,11 +194,11 @@
         }
         else
         {
-            LOGGER.info("Deleting level {}", saveName);
+            LOGGER.info("Deleting level {}", (Object)saveName);
 
-            for (int i = 1; i <= 5; i++)
+            for (int i = 1; i <= 5; ++i)
             {
-                LOGGER.info("Attempt {}...", i);
+                LOGGER.info("Attempt {}...", (int)i);
 
                 if (deleteFiles(file1.listFiles()))
                 {
@@ -197,8 +213,9 @@
                     {
                         Thread.sleep(500L);
                     }
-                    catch (InterruptedException interruptedexception)
+                    catch (InterruptedException var5)
                     {
+                        ;
                     }
                 }
             }
@@ -212,17 +229,17 @@
     {
         for (File file1 : files)
         {
-            LOGGER.debug("Deleting {}", file1);
+            LOGGER.debug("Deleting {}", (Object)file1);
 
             if (file1.isDirectory() && !deleteFiles(file1.listFiles()))
             {
-                LOGGER.warn("Couldn't delete directory {}", file1);
+                LOGGER.warn("Couldn't delete directory {}", (Object)file1);
                 return false;
             }
 
             if (!file1.delete())
             {
-                LOGGER.warn("Couldn't delete file {}", file1);
+                LOGGER.warn("Couldn't delete file {}", (Object)file1);
                 return false;
             }
         }
@@ -230,40 +247,29 @@
         return true;
     }
 
-    @Override
-    public ISaveHandler getSaveLoader(String saveName, boolean storePlayerdata)
-    {
-        return new SaveHandler(this.savesDirectory, saveName, storePlayerdata, this.dataFixer);
-    }
-
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean isConvertible(String saveName)
     {
         return false;
     }
 
-    @Override
     public boolean isOldMapFormat(String saveName)
     {
         return false;
     }
 
-    @Override
     public boolean convertMapFormat(String filename, IProgressUpdate progressCallback)
     {
         return false;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean canLoadWorld(String saveName)
     {
         File file1 = new File(this.savesDirectory, saveName);
         return file1.isDirectory();
     }
 
-    @Override
     public File getFile(String p_186352_1_, String p_186352_2_)
     {
         return new File(new File(this.savesDirectory, p_186352_1_), p_186352_2_);
