--- before/net/minecraft/world/storage/ThreadedFileIOBase.java
+++ after/net/minecraft/world/storage/ThreadedFileIOBase.java
@@ -7,7 +7,7 @@
 public class ThreadedFileIOBase implements Runnable
 {
     private static final ThreadedFileIOBase field_75741_a = new ThreadedFileIOBase();
-    private final List<IThreadedFileIO> field_75739_b = Collections.synchronizedList(Lists.newArrayList());
+    private final List<IThreadedFileIO> field_75739_b = Collections.<IThreadedFileIO>synchronizedList(Lists.newArrayList());
     private volatile long field_75740_c;
     private volatile long field_75737_d;
     private volatile boolean field_75738_e;
