--- before/net/minecraft/world/storage/WorldInfo.java
+++ after/net/minecraft/world/storage/WorldInfo.java
@@ -63,8 +63,9 @@
     private double borderDamagePerBlock = 0.2;
     private int borderWarningDistance = 5;
     private int borderWarningTime = 15;
-    private final Map<DimensionType, NBTTagCompound> dimensionData = Maps.newEnumMap(DimensionType.class);
+    private final Map<Integer, NBTTagCompound> dimensionData = Maps.newHashMap();
     private GameRules gameRules = new GameRules();
+    private java.util.Map<String, net.minecraft.nbt.NBTBase> additionalProperties;
 
     protected WorldInfo()
     {
@@ -252,7 +253,7 @@
 
             for (String s : nbttagcompound1.getKeySet())
             {
-                this.dimensionData.put(DimensionType.getById(Integer.parseInt(s)), nbttagcompound1.getCompoundTag(s));
+                this.dimensionData.put(Integer.parseInt(s), nbttagcompound1.getCompoundTag(s));
             }
         }
     }
@@ -367,6 +368,7 @@
         nbt.setDouble("BorderSizeLerpTarget", this.borderSizeLerpTarget);
         nbt.setDouble("BorderWarningBlocks", this.borderWarningDistance);
         nbt.setDouble("BorderWarningTime", this.borderWarningTime);
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().getDataFixer().writeVersionData(nbt);
 
         if (this.difficulty != null)
         {
@@ -377,9 +379,10 @@
         nbt.setTag("GameRules", this.gameRules.writeToNBT());
         NBTTagCompound nbttagcompound1 = new NBTTagCompound();
 
-        for (Entry<DimensionType, NBTTagCompound> entry : this.dimensionData.entrySet())
+        for (Entry<Integer, NBTTagCompound> entry : this.dimensionData.entrySet())
         {
-            nbttagcompound1.setTag(String.valueOf(entry.getKey().getId()), entry.getValue());
+            if (entry.getValue() == null || entry.getValue().isEmpty()) continue;
+            nbttagcompound1.setTag(String.valueOf(entry.getKey()), entry.getValue());
         }
 
         nbt.setTag("DimensionData", nbttagcompound1);
@@ -709,6 +712,7 @@
 
     public void setDifficulty(EnumDifficulty newDifficulty)
     {
+        net.minecraftforge.common.ForgeHooks.onDifficultyChange(newDifficulty, this.difficulty);
         this.difficulty = newDifficulty;
     }
 
@@ -832,16 +836,47 @@
         }
         );
     }
-
+    /**
+     * Allow access to additional mod specific world based properties
+     * Used by FML to store mod list associated with a world, and maybe an id map
+     * Used by Forge to store the dimensions available to a world
+     * @param additionalProperties
+     */
+    public void setAdditionalProperties(java.util.Map<String,net.minecraft.nbt.NBTBase> additionalProperties)
+    {
+        // one time set for this
+        if (this.additionalProperties == null)
+        {
+            this.additionalProperties = additionalProperties;
+        }
+    }
+
+    public net.minecraft.nbt.NBTBase getAdditionalProperty(String additionalProperty)
+    {
+        return this.additionalProperties!=null? this.additionalProperties.get(additionalProperty) : null;
+    }
+
+    @Deprecated //Use the int version below, and pass in dimension id NOT TYPE id
     public NBTTagCompound getDimensionData(DimensionType dimensionIn)
     {
+        return getDimensionData(dimensionIn.getId());
+    }
+    public NBTTagCompound getDimensionData(int dimensionIn)
+    {
         NBTTagCompound nbttagcompound = this.dimensionData.get(dimensionIn);
         return nbttagcompound == null ? new NBTTagCompound() : nbttagcompound;
     }
 
+    @Deprecated //Use the int version below, and pass in dimension id NOT TYPE id
     public void setDimensionData(DimensionType dimensionIn, NBTTagCompound compound)
     {
-        this.dimensionData.put(dimensionIn, compound);
+        this.setDimensionData(dimensionIn.getId(), compound);
+    }
+
+    //Dimension numerical ID version of setter, as two dimensions could in theory have the same DimensionType. ID should be grabbed from the world NOT the Type
+    public void setDimensionData(int dimensionID, NBTTagCompound compound)
+    {
+        this.dimensionData.put(dimensionID, compound);
     }
 
     @SideOnly(Side.CLIENT)
