--- before/net/minecraft/world/storage/loot/functions/EnchantRandomly.java
+++ after/net/minecraft/world/storage/loot/functions/EnchantRandomly.java
@@ -42,7 +42,7 @@
 
         if (this.field_186558_b.isEmpty())
         {
-            List<Enchantment> list = Lists.newArrayList();
+            List<Enchantment> list = Lists.<Enchantment>newArrayList();
 
             for (Enchantment enchantment1 : Enchantment.field_185264_b)
             {
@@ -111,7 +111,7 @@
 
             public EnchantRandomly func_186530_b(JsonObject p_186530_1_, JsonDeserializationContext p_186530_2_, LootCondition[] p_186530_3_)
             {
-                List<Enchantment> list = Lists.newArrayList();
+                List<Enchantment> list = Lists.<Enchantment>newArrayList();
 
                 if (p_186530_1_.has("enchantments"))
                 {
