--- before/net/minecraft/world/storage/loot/functions/LootingEnchantBonus.java
+++ after/net/minecraft/world/storage/loot/functions/LootingEnchantBonus.java
@@ -33,7 +33,7 @@
 
         if (entity instanceof EntityLivingBase)
         {
-            int i = EnchantmentHelper.getLootingModifier((EntityLivingBase)entity);
+            int i = context.getLootingModifier();
 
             if (i == 0)
             {
