--- before/net/minecraft/world/storage/loot/functions/LootingEnchantBonus.java
+++ after/net/minecraft/world/storage/loot/functions/LootingEnchantBonus.java
@@ -32,7 +32,7 @@
 
         if (entity instanceof EntityLivingBase)
         {
-            int i = EnchantmentHelper.func_185283_h((EntityLivingBase)entity);
+            int i = p_186553_3_.getLootingModifier();
 
             if (i == 0)
             {
@@ -64,14 +64,14 @@
 
                 if (p_186532_2_.field_189971_b > 0)
                 {
-                    p_186532_1_.add("limit", p_186532_3_.serialize(p_186532_2_.field_189971_b));
+                    p_186532_1_.add("limit", p_186532_3_.serialize(Integer.valueOf(p_186532_2_.field_189971_b)));
                 }
             }
 
             public LootingEnchantBonus func_186530_b(JsonObject p_186530_1_, JsonDeserializationContext p_186530_2_, LootCondition[] p_186530_3_)
             {
                 int i = JsonUtils.func_151208_a(p_186530_1_, "limit", 0);
-                return new LootingEnchantBonus(p_186530_3_, JsonUtils.func_188174_a(p_186530_1_, "count", p_186530_2_, RandomValueRange.class), i);
+                return new LootingEnchantBonus(p_186530_3_, (RandomValueRange)JsonUtils.func_188174_a(p_186530_1_, "count", p_186530_2_, RandomValueRange.class), i);
             }
         }
 }
