--- before/net/minecraft/world/storage/loot/functions/SetAttributes.java
+++ after/net/minecraft/world/storage/loot/functions/SetAttributes.java
@@ -107,7 +107,7 @@
                 String s = JsonUtils.func_151200_h(p_186586_0_, "name");
                 String s1 = JsonUtils.func_151200_h(p_186586_0_, "attribute");
                 int i = func_186595_a(JsonUtils.func_151200_h(p_186586_0_, "operation"));
-                RandomValueRange randomvaluerange = JsonUtils.func_188174_a(p_186586_0_, "amount", p_186586_1_, RandomValueRange.class);
+                RandomValueRange randomvaluerange = (RandomValueRange)JsonUtils.func_188174_a(p_186586_0_, "amount", p_186586_1_, RandomValueRange.class);
                 UUID uuid = null;
                 EntityEquipmentSlot[] aentityequipmentslot;
 
@@ -145,7 +145,7 @@
                     {
                         uuid = UUID.fromString(s2);
                     }
-                    catch (IllegalArgumentException illegalargumentexception)
+                    catch (IllegalArgumentException var12)
                     {
                         throw new JsonSyntaxException("Invalid attribute modifier id '" + s2 + "' (must be UUID format, with dashes)");
                     }
