--- before/net/minecraft/world/storage/loot/functions/SetDamage.java
+++ after/net/minecraft/world/storage/loot/functions/SetDamage.java
@@ -54,7 +54,7 @@
 
             public SetDamage func_186530_b(JsonObject p_186530_1_, JsonDeserializationContext p_186530_2_, LootCondition[] p_186530_3_)
             {
-                return new SetDamage(p_186530_3_, JsonUtils.func_188174_a(p_186530_1_, "damage", p_186530_2_, RandomValueRange.class));
+                return new SetDamage(p_186530_3_, (RandomValueRange)JsonUtils.func_188174_a(p_186530_1_, "damage", p_186530_2_, RandomValueRange.class));
             }
         }
 }
