--- before/net/minecraft/world/gen/feature/WorldGenIceSpike.java
+++ after/net/minecraft/world/gen/feature/WorldGenIceSpike.java
@@ -53,7 +53,7 @@
                             IBlockState iblockstate = worldIn.getBlockState(position.add(i1, k, j1));
                             Block block = iblockstate.getBlock();
 
-                            if (iblockstate.getMaterial() == Material.AIR
+                            if (iblockstate.getBlock().isAir(iblockstate, worldIn, position.add(i1, k, j1))
                                     || block == Blocks.DIRT
                                     || block == Blocks.SNOW
                                     || block == Blocks.ICE)
@@ -66,7 +66,7 @@
                                 iblockstate = worldIn.getBlockState(position.add(i1, -k, j1));
                                 block = iblockstate.getBlock();
 
-                                if (iblockstate.getMaterial() == Material.AIR
+                                if (iblockstate.getBlock().isAir(iblockstate, worldIn, position.add(i1, -k, j1))
                                         || block == Blocks.DIRT
                                         || block == Blocks.SNOW
                                         || block == Blocks.ICE)
@@ -107,7 +107,7 @@
                         IBlockState iblockstate1 = worldIn.getBlockState(blockpos);
                         Block block1 = iblockstate1.getBlock();
 
-                        if (iblockstate1.getMaterial() != Material.AIR
+                        if (!iblockstate1.getBlock().isAir(iblockstate1, worldIn, blockpos)
                                 && block1 != Blocks.DIRT
                                 && block1 != Blocks.SNOW
                                 && block1 != Blocks.ICE
