--- before/net/minecraft/world/gen/feature/WorldGenSavannaTree.java
+++ after/net/minecraft/world/gen/feature/WorldGenSavannaTree.java
@@ -15,20 +15,14 @@
 
 public class WorldGenSavannaTree extends WorldGenAbstractTree
 {
-    private static final IBlockState TRUNK = Blocks.LOG2
-                .getDefaultState()
-                .withProperty(BlockNewLog.VARIANT, BlockPlanks.EnumType.ACACIA);
-    private static final IBlockState LEAF = Blocks.LEAVES2
-                .getDefaultState()
-                .withProperty(BlockNewLeaf.VARIANT, BlockPlanks.EnumType.ACACIA)
-                .withProperty(BlockLeaves.CHECK_DECAY, false);
+    private static final IBlockState TRUNK = Blocks.LOG2.getDefaultState().withProperty(BlockNewLog.VARIANT, BlockPlanks.EnumType.ACACIA);
+    private static final IBlockState LEAF = Blocks.LEAVES2.getDefaultState().withProperty(BlockNewLeaf.VARIANT, BlockPlanks.EnumType.ACACIA).withProperty(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
 
     public WorldGenSavannaTree(boolean doBlockNotify)
     {
         super(doBlockNotify);
     }
 
-    @Override
     public boolean generate(World worldIn, Random rand, BlockPos position)
     {
         int i = rand.nextInt(3) + rand.nextInt(3) + 5;
@@ -36,7 +30,7 @@
 
         if (position.getY() >= 1 && position.getY() + i + 1 <= 256)
         {
-            for (int j = position.getY(); j <= position.getY() + 1 + i; j++)
+            for (int j = position.getY(); j <= position.getY() + 1 + i; ++j)
             {
                 int k = 1;
 
@@ -52,15 +46,18 @@
 
                 BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
 
-                for (int l = position.getX() - k; l <= position.getX() + k && flag; l++)
+                for (int l = position.getX() - k; l <= position.getX() + k && flag; ++l)
                 {
-                    for (int i1 = position.getZ() - k; i1 <= position.getZ() + k && flag; i1++)
+                    for (int i1 = position.getZ() - k; i1 <= position.getZ() + k && flag; ++i1)
                     {
-                        if (j < 0 || j >= 256)
+                        if (j >= 0 && j < 256)
                         {
-                            flag = false;
+                            if (!this.isReplaceable(worldIn,blockpos$mutableblockpos.setPos(l, j, i1)))
+                            {
+                                flag = false;
+                            }
                         }
-                        else if (!this.canGrowInto(worldIn.getBlockState(blockpos$mutableblockpos.setPos(l, j, i1)).getBlock()))
+                        else
                         {
                             flag = false;
                         }
@@ -74,11 +71,13 @@
             }
             else
             {
-                Block block = worldIn.getBlockState(position.down()).getBlock();
+                BlockPos down = position.down();
+                IBlockState state = worldIn.getBlockState(down);
+                boolean isSoil = state.getBlock().canSustainPlant(state, worldIn, down, net.minecraft.util.EnumFacing.UP, ((net.minecraft.block.BlockSapling)Blocks.SAPLING));
 
-                if ((block == Blocks.GRASS || block == Blocks.DIRT) && position.getY() < 256 - i - 1)
+                if (isSoil && position.getY() < worldIn.getHeight() - i - 1)
                 {
-                    this.setDirtAt(worldIn, position.down());
+                    state.getBlock().onPlantGrow(state, worldIn, down, position);
                     EnumFacing enumfacing = EnumFacing.Plane.HORIZONTAL.random(rand);
                     int k2 = i - rand.nextInt(4) - 1;
                     int l2 = 3 - rand.nextInt(3);
@@ -86,7 +85,7 @@
                     int j1 = position.getZ();
                     int k1 = 0;
 
-                    for (int l1 = 0; l1 < i; l1++)
+                    for (int l1 = 0; l1 < i; ++l1)
                     {
                         int i2 = position.getY() + l1;
 
@@ -94,13 +93,13 @@
                         {
                             i3 += enumfacing.getXOffset();
                             j1 += enumfacing.getZOffset();
-                            l2--;
+                            --l2;
                         }
 
                         BlockPos blockpos = new BlockPos(i3, i2, j1);
-                        Material material = worldIn.getBlockState(blockpos).getMaterial();
+                        state = worldIn.getBlockState(blockpos);
 
-                        if (material == Material.AIR || material == Material.LEAVES)
+                        if (state.getBlock().isAir(state, worldIn, blockpos) || state.getBlock().isLeaves(state, worldIn, blockpos))
                         {
                             this.placeLogAt(worldIn, blockpos);
                             k1 = i2;
@@ -109,9 +108,9 @@
 
                     BlockPos blockpos2 = new BlockPos(i3, k1, j1);
 
-                    for (int j3 = -3; j3 <= 3; j3++)
+                    for (int j3 = -3; j3 <= 3; ++j3)
                     {
-                        for (int i4 = -3; i4 <= 3; i4++)
+                        for (int i4 = -3; i4 <= 3; ++i4)
                         {
                             if (Math.abs(j3) != 3 || Math.abs(i4) != 3)
                             {
@@ -122,9 +121,9 @@
 
                     blockpos2 = blockpos2.up();
 
-                    for (int k3 = -1; k3 <= 1; k3++)
+                    for (int k3 = -1; k3 <= 1; ++k3)
                     {
-                        for (int j4 = -1; j4 <= 1; j4++)
+                        for (int j4 = -1; j4 <= 1; ++j4)
                         {
                             this.placeLeafAt(worldIn, blockpos2.add(k3, 0, j4));
                         }
@@ -144,7 +143,7 @@
                         int k4 = 1 + rand.nextInt(3);
                         k1 = 0;
 
-                        for (int l4 = l3; l4 < i && k4 > 0; k4--)
+                        for (int l4 = l3; l4 < i && k4 > 0; --k4)
                         {
                             if (l4 >= 1)
                             {
@@ -152,25 +151,25 @@
                                 i3 += enumfacing1.getXOffset();
                                 j1 += enumfacing1.getZOffset();
                                 BlockPos blockpos1 = new BlockPos(i3, j2, j1);
-                                Material material1 = worldIn.getBlockState(blockpos1).getMaterial();
+                                state = worldIn.getBlockState(blockpos1);
 
-                                if (material1 == Material.AIR || material1 == Material.LEAVES)
+                                if (state.getBlock().isAir(state, worldIn, blockpos1) || state.getBlock().isLeaves(state, worldIn, blockpos1))
                                 {
                                     this.placeLogAt(worldIn, blockpos1);
                                     k1 = j2;
                                 }
                             }
 
-                            l4++;
+                            ++l4;
                         }
 
                         if (k1 > 0)
                         {
                             BlockPos blockpos3 = new BlockPos(i3, k1, j1);
 
-                            for (int i5 = -2; i5 <= 2; i5++)
+                            for (int i5 = -2; i5 <= 2; ++i5)
                             {
-                                for (int k5 = -2; k5 <= 2; k5++)
+                                for (int k5 = -2; k5 <= 2; ++k5)
                                 {
                                     if (Math.abs(i5) != 2 || Math.abs(k5) != 2)
                                     {
@@ -181,9 +180,9 @@
 
                             blockpos3 = blockpos3.up();
 
-                            for (int j5 = -1; j5 <= 1; j5++)
+                            for (int j5 = -1; j5 <= 1; ++j5)
                             {
-                                for (int l5 = -1; l5 <= 1; l5++)
+                                for (int l5 = -1; l5 <= 1; ++l5)
                                 {
                                     this.placeLeafAt(worldIn, blockpos3.add(j5, 0, l5));
                                 }
@@ -212,9 +211,9 @@
 
     private void placeLeafAt(World worldIn, BlockPos pos)
     {
-        Material material = worldIn.getBlockState(pos).getMaterial();
+        IBlockState state = worldIn.getBlockState(pos);
 
-        if (material == Material.AIR || material == Material.LEAVES)
+        if (state.getBlock().isAir(state, worldIn, pos) || state.getBlock().isLeaves(state, worldIn, pos))
         {
             this.setBlockAndNotifyAdequately(worldIn, pos, LEAF);
         }
