--- before/net/minecraft/world/gen/feature/WorldGenGlowStone2.java
+++ after/net/minecraft/world/gen/feature/WorldGenGlowStone2.java
@@ -9,7 +9,6 @@
 
 public class WorldGenGlowStone2 extends WorldGenerator
 {
-    @Override
     public boolean func_180709_b(World p_180709_1_, Random p_180709_2_, BlockPos p_180709_3_)
     {
         if (!p_180709_1_.func_175623_d(p_180709_3_))
@@ -24,13 +23,11 @@
         {
             p_180709_1_.func_180501_a(p_180709_3_, Blocks.field_150426_aN.func_176223_P(), 2);
 
-            for (int i = 0; i < 1500; i++)
+            for (int i = 0; i < 1500; ++i)
             {
-                BlockPos blockpos = p_180709_3_.func_177982_a(
-                                        p_180709_2_.nextInt(8) - p_180709_2_.nextInt(8), -p_180709_2_.nextInt(12), p_180709_2_.nextInt(8) - p_180709_2_.nextInt(8)
-                                    );
+                BlockPos blockpos = p_180709_3_.func_177982_a(p_180709_2_.nextInt(8) - p_180709_2_.nextInt(8), -p_180709_2_.nextInt(12), p_180709_2_.nextInt(8) - p_180709_2_.nextInt(8));
 
-                if (p_180709_1_.func_180495_p(blockpos).func_185904_a() == Material.field_151579_a)
+                if (p_180709_1_.func_175623_d(blockpos))
                 {
                     int j = 0;
 
@@ -38,7 +35,7 @@
                     {
                         if (p_180709_1_.func_180495_p(blockpos.func_177972_a(enumfacing)).func_177230_c() == Blocks.field_150426_aN)
                         {
-                            j++;
+                            ++j;
                         }
 
                         if (j > 1)
