--- before/net/minecraft/world/gen/feature/WorldGenVines.java
+++ after/net/minecraft/world/gen/feature/WorldGenVines.java
@@ -20,7 +20,7 @@
                 {
                     if (Blocks.field_150395_bd.func_176198_a(p_180709_1_, p_180709_3_, enumfacing))
                     {
-                        IBlockState iblockstate = Blocks.field_150395_bd.func_176223_P().func_177226_a(BlockVine.field_176273_b, enumfacing == EnumFacing.NORTH).func_177226_a(BlockVine.field_176278_M, enumfacing == EnumFacing.EAST).func_177226_a(BlockVine.field_176279_N, enumfacing == EnumFacing.SOUTH).func_177226_a(BlockVine.field_176280_O, enumfacing == EnumFacing.WEST);
+                        IBlockState iblockstate = Blocks.field_150395_bd.func_176223_P().func_177226_a(BlockVine.field_176273_b, Boolean.valueOf(enumfacing == EnumFacing.NORTH)).func_177226_a(BlockVine.field_176278_M, Boolean.valueOf(enumfacing == EnumFacing.EAST)).func_177226_a(BlockVine.field_176279_N, Boolean.valueOf(enumfacing == EnumFacing.SOUTH)).func_177226_a(BlockVine.field_176280_O, Boolean.valueOf(enumfacing == EnumFacing.WEST));
                         p_180709_1_.func_180501_a(p_180709_3_, iblockstate, 2);
                         break;
                     }
