--- before/net/minecraft/world/gen/feature/WorldGenMegaJungle.java
+++ after/net/minecraft/world/gen/feature/WorldGenMegaJungle.java
@@ -16,7 +16,6 @@
         super(p_i46448_1_, p_i46448_2_, p_i46448_3_, p_i46448_4_, p_i46448_5_);
     }
 
-    @Override
     public boolean func_180709_b(World p_180709_1_, Random p_180709_2_, BlockPos p_180709_3_)
     {
         int i = this.func_150533_a(p_180709_2_);
@@ -31,11 +30,11 @@
 
             for (int j = p_180709_3_.func_177956_o() + i - 2 - p_180709_2_.nextInt(4); j > p_180709_3_.func_177956_o() + i / 2; j -= 2 + p_180709_2_.nextInt(4))
             {
-                float f = p_180709_2_.nextFloat() * (float)(Math.PI * 2);
+                float f = p_180709_2_.nextFloat() * ((float)Math.PI * 2F);
                 int k = p_180709_3_.func_177958_n() + (int)(0.5F + MathHelper.func_76134_b(f) * 4.0F);
                 int l = p_180709_3_.func_177952_p() + (int)(0.5F + MathHelper.func_76126_a(f) * 4.0F);
 
-                for (int i1 = 0; i1 < 5; i1++)
+                for (int i1 = 0; i1 < 5; ++i1)
                 {
                     k = p_180709_3_.func_177958_n() + (int)(1.5F + MathHelper.func_76134_b(f) * (float)i1);
                     l = p_180709_3_.func_177952_p() + (int)(1.5F + MathHelper.func_76126_a(f) * (float)i1);
@@ -45,18 +44,18 @@
                 int j2 = 1 + p_180709_2_.nextInt(2);
                 int j1 = j;
 
-                for (int k1 = j - j2; k1 <= j1; k1++)
+                for (int k1 = j - j2; k1 <= j1; ++k1)
                 {
                     int l1 = k1 - j1;
                     this.func_175928_b(p_180709_1_, new BlockPos(k, k1, l), 1 - l1);
                 }
             }
 
-            for (int i2 = 0; i2 < i; i2++)
+            for (int i2 = 0; i2 < i; ++i2)
             {
                 BlockPos blockpos = p_180709_3_.func_177981_b(i2);
 
-                if (this.func_150523_a(p_180709_1_.func_180495_p(blockpos).func_177230_c()))
+                if (this.isAirLeaves(p_180709_1_,blockpos))
                 {
                     this.func_175903_a(p_180709_1_, blockpos, this.field_76520_b);
 
@@ -71,7 +70,7 @@
                 {
                     BlockPos blockpos1 = blockpos.func_177974_f();
 
-                    if (this.func_150523_a(p_180709_1_.func_180495_p(blockpos1).func_177230_c()))
+                    if (this.isAirLeaves(p_180709_1_,blockpos1))
                     {
                         this.func_175903_a(p_180709_1_, blockpos1, this.field_76520_b);
 
@@ -84,7 +83,7 @@
 
                     BlockPos blockpos2 = blockpos.func_177968_d().func_177974_f();
 
-                    if (this.func_150523_a(p_180709_1_.func_180495_p(blockpos2).func_177230_c()))
+                    if (this.isAirLeaves(p_180709_1_,blockpos2))
                     {
                         this.func_175903_a(p_180709_1_, blockpos2, this.field_76520_b);
 
@@ -97,7 +96,7 @@
 
                     BlockPos blockpos3 = blockpos.func_177968_d();
 
-                    if (this.func_150523_a(p_180709_1_.func_180495_p(blockpos3).func_177230_c()))
+                    if (this.isAirLeaves(p_180709_1_,blockpos3))
                     {
                         this.func_175903_a(p_180709_1_, blockpos3, this.field_76520_b);
 
@@ -118,7 +117,7 @@
     {
         if (p_181632_2_.nextInt(3) > 0 && p_181632_1_.func_175623_d(p_181632_3_))
         {
-            this.func_175903_a(p_181632_1_, p_181632_3_, Blocks.field_150395_bd.func_176223_P().func_177226_a(p_181632_4_, true));
+            this.func_175903_a(p_181632_1_, p_181632_3_, Blocks.field_150395_bd.func_176223_P().func_177226_a(p_181632_4_, Boolean.valueOf(true)));
         }
     }
 
@@ -126,9 +125,16 @@
     {
         int i = 2;
 
-        for (int j = -2; j <= 0; j++)
+        for (int j = -2; j <= 0; ++j)
         {
             this.func_175925_a(p_175930_1_, p_175930_2_.func_177981_b(j), p_175930_3_ + 1 - j);
         }
+    }
+
+    //Helper macro
+    private boolean isAirLeaves(World world, BlockPos pos)
+    {
+        IBlockState state = world.func_180495_p(pos);
+        return state.func_177230_c().isAir(state, world, pos) || state.func_177230_c().isLeaves(state, world, pos);
     }
 }
