--- before/net/minecraft/world/gen/feature/WorldGenTaiga1.java
+++ after/net/minecraft/world/gen/feature/WorldGenTaiga1.java
@@ -14,20 +14,14 @@
 
 public class WorldGenTaiga1 extends WorldGenAbstractTree
 {
-    private static final IBlockState field_181636_a = Blocks.field_150364_r
-                .func_176223_P()
-                .func_177226_a(BlockOldLog.field_176301_b, BlockPlanks.EnumType.SPRUCE);
-    private static final IBlockState field_181637_b = Blocks.field_150362_t
-                .func_176223_P()
-                .func_177226_a(BlockOldLeaf.field_176239_P, BlockPlanks.EnumType.SPRUCE)
-                .func_177226_a(BlockLeaves.field_176236_b, false);
+    private static final IBlockState field_181636_a = Blocks.field_150364_r.func_176223_P().func_177226_a(BlockOldLog.field_176301_b, BlockPlanks.EnumType.SPRUCE);
+    private static final IBlockState field_181637_b = Blocks.field_150362_t.func_176223_P().func_177226_a(BlockOldLeaf.field_176239_P, BlockPlanks.EnumType.SPRUCE).func_177226_a(BlockLeaves.field_176236_b, Boolean.valueOf(false));
 
     public WorldGenTaiga1()
     {
         super(false);
     }
 
-    @Override
     public boolean func_180709_b(World p_180709_1_, Random p_180709_2_, BlockPos p_180709_3_)
     {
         int i = p_180709_2_.nextInt(5) + 7;
@@ -39,7 +33,7 @@
         {
             boolean flag = true;
 
-            for (int i1 = p_180709_3_.func_177956_o(); i1 <= p_180709_3_.func_177956_o() + 1 + i && flag; i1++)
+            for (int i1 = p_180709_3_.func_177956_o(); i1 <= p_180709_3_.func_177956_o() + 1 + i && flag; ++i1)
             {
                 int j1 = 1;
 
@@ -54,15 +48,18 @@
 
                 BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
 
-                for (int k1 = p_180709_3_.func_177958_n() - j1; k1 <= p_180709_3_.func_177958_n() + j1 && flag; k1++)
+                for (int k1 = p_180709_3_.func_177958_n() - j1; k1 <= p_180709_3_.func_177958_n() + j1 && flag; ++k1)
                 {
-                    for (int l1 = p_180709_3_.func_177952_p() - j1; l1 <= p_180709_3_.func_177952_p() + j1 && flag; l1++)
+                    for (int l1 = p_180709_3_.func_177952_p() - j1; l1 <= p_180709_3_.func_177952_p() + j1 && flag; ++l1)
                     {
-                        if (i1 < 0 || i1 >= 256)
+                        if (i1 >= 0 && i1 < 256)
                         {
-                            flag = false;
+                            if (!this.isReplaceable(p_180709_1_,blockpos$mutableblockpos.func_181079_c(k1, i1, l1)))
+                            {
+                                flag = false;
+                            }
                         }
-                        else if (!this.func_150523_a(p_180709_1_.func_180495_p(blockpos$mutableblockpos.func_181079_c(k1, i1, l1)).func_177230_c()))
+                        else
                         {
                             flag = false;
                         }
@@ -76,28 +73,31 @@
             }
             else
             {
-                Block block = p_180709_1_.func_180495_p(p_180709_3_.func_177977_b()).func_177230_c();
+                BlockPos down = p_180709_3_.func_177977_b();
+                IBlockState state = p_180709_1_.func_180495_p(down);
+                boolean isSoil = state.func_177230_c().canSustainPlant(state, p_180709_1_, down, net.minecraft.util.EnumFacing.UP, (net.minecraft.block.BlockSapling)Blocks.field_150345_g);
 
-                if ((block == Blocks.field_150349_c || block == Blocks.field_150346_d) && p_180709_3_.func_177956_o() < 256 - i - 1)
+                if (isSoil && p_180709_3_.func_177956_o() < 256 - i - 1)
                 {
-                    this.func_175921_a(p_180709_1_, p_180709_3_.func_177977_b());
+                    state.func_177230_c().onPlantGrow(state, p_180709_1_, down, p_180709_3_);
                     int k2 = 0;
 
-                    for (int l2 = p_180709_3_.func_177956_o() + i; l2 >= p_180709_3_.func_177956_o() + j; l2--)
+                    for (int l2 = p_180709_3_.func_177956_o() + i; l2 >= p_180709_3_.func_177956_o() + j; --l2)
                     {
-                        for (int j3 = p_180709_3_.func_177958_n() - k2; j3 <= p_180709_3_.func_177958_n() + k2; j3++)
+                        for (int j3 = p_180709_3_.func_177958_n() - k2; j3 <= p_180709_3_.func_177958_n() + k2; ++j3)
                         {
                             int k3 = j3 - p_180709_3_.func_177958_n();
 
-                            for (int i2 = p_180709_3_.func_177952_p() - k2; i2 <= p_180709_3_.func_177952_p() + k2; i2++)
+                            for (int i2 = p_180709_3_.func_177952_p() - k2; i2 <= p_180709_3_.func_177952_p() + k2; ++i2)
                             {
                                 int j2 = i2 - p_180709_3_.func_177952_p();
 
                                 if (Math.abs(k3) != k2 || Math.abs(j2) != k2 || k2 <= 0)
                                 {
                                     BlockPos blockpos = new BlockPos(j3, l2, i2);
+                                    state = p_180709_1_.func_180495_p(blockpos);
 
-                                    if (!p_180709_1_.func_180495_p(blockpos).func_185913_b())
+                                    if (state.func_177230_c().canBeReplacedByLeaves(state, p_180709_1_, blockpos))
                                     {
                                         this.func_175903_a(p_180709_1_, blockpos, field_181637_b);
                                     }
@@ -107,19 +107,20 @@
 
                         if (k2 >= 1 && l2 == p_180709_3_.func_177956_o() + j + 1)
                         {
-                            k2--;
+                            --k2;
                         }
                         else if (k2 < l)
                         {
-                            k2++;
+                            ++k2;
                         }
                     }
 
-                    for (int i3 = 0; i3 < i - 1; i3++)
+                    for (int i3 = 0; i3 < i - 1; ++i3)
                     {
-                        Material material = p_180709_1_.func_180495_p(p_180709_3_.func_177981_b(i3)).func_185904_a();
+                        BlockPos upN = p_180709_3_.func_177981_b(i3);
+                        state = p_180709_1_.func_180495_p(upN);
 
-                        if (material == Material.field_151579_a || material == Material.field_151584_j)
+                        if (state.func_177230_c().isAir(state, p_180709_1_, upN) || state.func_177230_c().isLeaves(state, p_180709_1_, upN))
                         {
                             this.func_175903_a(p_180709_1_, p_180709_3_.func_177981_b(i3), field_181636_a);
                         }
