--- before/net/minecraft/world/gen/feature/WorldGenTallGrass.java
+++ after/net/minecraft/world/gen/feature/WorldGenTallGrass.java
@@ -21,7 +21,7 @@
     public boolean generate(World worldIn, Random rand, BlockPos position)
     {
         for (IBlockState iblockstate = worldIn.getBlockState(position);
-                (iblockstate.getMaterial() == Material.AIR || iblockstate.getMaterial() == Material.LEAVES) && position.getY() > 0;
+                (iblockstate.getBlock().isAir(iblockstate, worldIn, position) || iblockstate.getBlock().isLeaves(iblockstate, worldIn, position)) && position.getY() > 0;
                 iblockstate = worldIn.getBlockState(position)
             )
         {
