--- before/net/minecraft/world/gen/feature/WorldGenBirchTree.java
+++ after/net/minecraft/world/gen/feature/WorldGenBirchTree.java
@@ -13,13 +13,8 @@
 
 public class WorldGenBirchTree extends WorldGenAbstractTree
 {
-    private static final IBlockState field_181629_a = Blocks.field_150364_r
-                .func_176223_P()
-                .func_177226_a(BlockOldLog.field_176301_b, BlockPlanks.EnumType.BIRCH);
-    private static final IBlockState field_181630_b = Blocks.field_150362_t
-                .func_176223_P()
-                .func_177226_a(BlockOldLeaf.field_176239_P, BlockPlanks.EnumType.BIRCH)
-                .func_177226_a(BlockOldLeaf.field_176236_b, false);
+    private static final IBlockState field_181629_a = Blocks.field_150364_r.func_176223_P().func_177226_a(BlockOldLog.field_176301_b, BlockPlanks.EnumType.BIRCH);
+    private static final IBlockState field_181630_b = Blocks.field_150362_t.func_176223_P().func_177226_a(BlockOldLeaf.field_176239_P, BlockPlanks.EnumType.BIRCH).func_177226_a(BlockOldLeaf.field_176236_b, Boolean.valueOf(false));
     private final boolean field_150531_a;
 
     public WorldGenBirchTree(boolean p_i45449_1_, boolean p_i45449_2_)
@@ -28,7 +23,6 @@
         this.field_150531_a = p_i45449_2_;
     }
 
-    @Override
     public boolean func_180709_b(World p_180709_1_, Random p_180709_2_, BlockPos p_180709_3_)
     {
         int i = p_180709_2_.nextInt(3) + 5;
@@ -42,7 +36,7 @@
 
         if (p_180709_3_.func_177956_o() >= 1 && p_180709_3_.func_177956_o() + i + 1 <= 256)
         {
-            for (int j = p_180709_3_.func_177956_o(); j <= p_180709_3_.func_177956_o() + 1 + i; j++)
+            for (int j = p_180709_3_.func_177956_o(); j <= p_180709_3_.func_177956_o() + 1 + i; ++j)
             {
                 int k = 1;
 
@@ -58,15 +52,18 @@
 
                 BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
 
-                for (int l = p_180709_3_.func_177958_n() - k; l <= p_180709_3_.func_177958_n() + k && flag; l++)
+                for (int l = p_180709_3_.func_177958_n() - k; l <= p_180709_3_.func_177958_n() + k && flag; ++l)
                 {
-                    for (int i1 = p_180709_3_.func_177952_p() - k; i1 <= p_180709_3_.func_177952_p() + k && flag; i1++)
+                    for (int i1 = p_180709_3_.func_177952_p() - k; i1 <= p_180709_3_.func_177952_p() + k && flag; ++i1)
                     {
-                        if (j < 0 || j >= 256)
+                        if (j >= 0 && j < p_180709_1_.func_72800_K())
                         {
-                            flag = false;
+                            if (!this.isReplaceable(p_180709_1_, blockpos$mutableblockpos.func_181079_c(l, j, i1)))
+                            {
+                                flag = false;
+                            }
                         }
-                        else if (!this.func_150523_a(p_180709_1_.func_180495_p(blockpos$mutableblockpos.func_181079_c(l, j, i1)).func_177230_c()))
+                        else
                         {
                             flag = false;
                         }
@@ -80,32 +77,33 @@
             }
             else
             {
-                Block block = p_180709_1_.func_180495_p(p_180709_3_.func_177977_b()).func_177230_c();
+                BlockPos down = p_180709_3_.func_177977_b();
+                IBlockState state = p_180709_1_.func_180495_p(down);
+                boolean isSoil = state.func_177230_c().canSustainPlant(state, p_180709_1_, down, net.minecraft.util.EnumFacing.UP, (net.minecraft.block.BlockSapling)Blocks.field_150345_g);
 
-                if ((block == Blocks.field_150349_c || block == Blocks.field_150346_d || block == Blocks.field_150458_ak)
-                        && p_180709_3_.func_177956_o() < 256 - i - 1)
+                if (isSoil && p_180709_3_.func_177956_o() < p_180709_1_.func_72800_K() - i - 1)
                 {
-                    this.func_175921_a(p_180709_1_, p_180709_3_.func_177977_b());
+                    state.func_177230_c().onPlantGrow(state, p_180709_1_, down, p_180709_3_);
 
-                    for (int i2 = p_180709_3_.func_177956_o() - 3 + i; i2 <= p_180709_3_.func_177956_o() + i; i2++)
+                    for (int i2 = p_180709_3_.func_177956_o() - 3 + i; i2 <= p_180709_3_.func_177956_o() + i; ++i2)
                     {
                         int k2 = i2 - (p_180709_3_.func_177956_o() + i);
                         int l2 = 1 - k2 / 2;
 
-                        for (int i3 = p_180709_3_.func_177958_n() - l2; i3 <= p_180709_3_.func_177958_n() + l2; i3++)
+                        for (int i3 = p_180709_3_.func_177958_n() - l2; i3 <= p_180709_3_.func_177958_n() + l2; ++i3)
                         {
                             int j1 = i3 - p_180709_3_.func_177958_n();
 
-                            for (int k1 = p_180709_3_.func_177952_p() - l2; k1 <= p_180709_3_.func_177952_p() + l2; k1++)
+                            for (int k1 = p_180709_3_.func_177952_p() - l2; k1 <= p_180709_3_.func_177952_p() + l2; ++k1)
                             {
                                 int l1 = k1 - p_180709_3_.func_177952_p();
 
                                 if (Math.abs(j1) != l2 || Math.abs(l1) != l2 || p_180709_2_.nextInt(2) != 0 && k2 != 0)
                                 {
                                     BlockPos blockpos = new BlockPos(i3, i2, k1);
-                                    Material material = p_180709_1_.func_180495_p(blockpos).func_185904_a();
+                                    IBlockState state2 = p_180709_1_.func_180495_p(blockpos);
 
-                                    if (material == Material.field_151579_a || material == Material.field_151584_j)
+                                    if (state2.func_177230_c().isAir(state2, p_180709_1_, blockpos) || state2.func_177230_c().isAir(state2, p_180709_1_, blockpos))
                                     {
                                         this.func_175903_a(p_180709_1_, blockpos, field_181630_b);
                                     }
@@ -114,11 +112,12 @@
                         }
                     }
 
-                    for (int j2 = 0; j2 < i; j2++)
+                    for (int j2 = 0; j2 < i; ++j2)
                     {
-                        Material material1 = p_180709_1_.func_180495_p(p_180709_3_.func_177981_b(j2)).func_185904_a();
+                        BlockPos upN = p_180709_3_.func_177981_b(j2);
+                        IBlockState state2 = p_180709_1_.func_180495_p(upN);
 
-                        if (material1 == Material.field_151579_a || material1 == Material.field_151584_j)
+                        if (state2.func_177230_c().isAir(state2, p_180709_1_, upN) || state2.func_177230_c().isLeaves(state2, p_180709_1_, upN))
                         {
                             this.func_175903_a(p_180709_1_, p_180709_3_.func_177981_b(j2), field_181629_a);
                         }
