--- before/net/minecraft/world/gen/feature/WorldGenHellLava.java
+++ after/net/minecraft/world/gen/feature/WorldGenHellLava.java
@@ -26,7 +26,7 @@
         {
             return false;
         }
-        else if (worldIn.getBlockState(position).getMaterial() != Material.AIR
+        else if (!worldIn.isAirBlock(position)
                  && worldIn.getBlockState(position).getBlock() != Blocks.NETHERRACK)
         {
             return false;
