--- before/net/minecraft/world/gen/feature/WorldGenGlowStone1.java
+++ after/net/minecraft/world/gen/feature/WorldGenGlowStone1.java
@@ -30,7 +30,7 @@
                                         rand.nextInt(8) - rand.nextInt(8), -rand.nextInt(12), rand.nextInt(8) - rand.nextInt(8)
                                     );
 
-                if (worldIn.getBlockState(blockpos).getMaterial() == Material.AIR)
+                if (worldIn.isAirBlock(blockpos))
                 {
                     int j = 0;
 
