--- before/net/minecraft/world/gen/feature/WorldGenDungeons.java
+++ after/net/minecraft/world/gen/feature/WorldGenDungeons.java
@@ -21,15 +21,8 @@
 public class WorldGenDungeons extends WorldGenerator
 {
     private static final Logger LOGGER = LogManager.getLogger();
-    private static final ResourceLocation[] SPAWNERTYPES = new ResourceLocation[]
-        {
-            EntityList.getKey(EntitySkeleton.class),
-            EntityList.getKey(EntityZombie.class),
-            EntityList.getKey(EntityZombie.class),
-            EntityList.getKey(EntitySpider.class)
-        };
+    private static final ResourceLocation[] SPAWNERTYPES = new ResourceLocation[] {EntityList.getKey(EntitySkeleton.class), EntityList.getKey(EntityZombie.class), EntityList.getKey(EntityZombie.class), EntityList.getKey(EntitySpider.class)};
 
-    @Override
     public boolean generate(World worldIn, Random rand, BlockPos position)
     {
         int i = 3;
@@ -43,11 +36,11 @@
         int i2 = k1 + 1;
         int j2 = 0;
 
-        for (int k2 = k; k2 <= l; k2++)
+        for (int k2 = k; k2 <= l; ++k2)
         {
-            for (int l2 = -1; l2 <= 4; l2++)
+            for (int l2 = -1; l2 <= 4; ++l2)
             {
-                for (int i3 = l1; i3 <= i2; i3++)
+                for (int i3 = l1; i3 <= i2; ++i3)
                 {
                     BlockPos blockpos = position.add(k2, l2, i3);
                     Material material = worldIn.getBlockState(blockpos).getMaterial();
@@ -63,12 +56,9 @@
                         return false;
                     }
 
-                    if ((k2 == k || k2 == l || i3 == l1 || i3 == i2)
-                            && l2 == 0
-                            && worldIn.isAirBlock(blockpos)
-                            && worldIn.isAirBlock(blockpos.up()))
+                    if ((k2 == k || k2 == l || i3 == l1 || i3 == i2) && l2 == 0 && worldIn.isAirBlock(blockpos) && worldIn.isAirBlock(blockpos.up()))
                     {
-                        j2++;
+                        ++j2;
                     }
                 }
             }
@@ -76,11 +66,11 @@
 
         if (j2 >= 1 && j2 <= 5)
         {
-            for (int k3 = k; k3 <= l; k3++)
+            for (int k3 = k; k3 <= l; ++k3)
             {
-                for (int i4 = 3; i4 >= -1; i4--)
+                for (int i4 = 3; i4 >= -1; --i4)
                 {
-                    for (int k4 = l1; k4 <= i2; k4++)
+                    for (int k4 = l1; k4 <= i2; ++k4)
                     {
                         BlockPos blockpos1 = position.add(k3, i4, k4);
 
@@ -95,8 +85,7 @@
                         {
                             worldIn.setBlockToAir(blockpos1);
                         }
-                        else if (worldIn.getBlockState(blockpos1).getMaterial().isSolid()
-                                 && worldIn.getBlockState(blockpos1).getBlock() != Blocks.CHEST)
+                        else if (worldIn.getBlockState(blockpos1).getMaterial().isSolid() && worldIn.getBlockState(blockpos1).getBlock() != Blocks.CHEST)
                         {
                             if (i4 == -1 && rand.nextInt(4) != 0)
                             {
@@ -111,9 +100,9 @@
                 }
             }
 
-            for (int l3 = 0; l3 < 2; l3++)
+            for (int l3 = 0; l3 < 2; ++l3)
             {
-                for (int j4 = 0; j4 < 3; j4++)
+                for (int j4 = 0; j4 < 3; ++j4)
                 {
                     int l4 = position.getX() + rand.nextInt(j * 2 + 1) - j;
                     int i5 = position.getY();
@@ -128,15 +117,13 @@
                         {
                             if (worldIn.getBlockState(blockpos2.offset(enumfacing)).getMaterial().isSolid())
                             {
-                                j3++;
+                                ++j3;
                             }
                         }
 
                         if (j3 == 1)
                         {
-                            worldIn.setBlockState(
-                                blockpos2, Blocks.CHEST.correctFacing(worldIn, blockpos2, Blocks.CHEST.getDefaultState()), 2
-                            );
+                            worldIn.setBlockState(blockpos2, Blocks.CHEST.correctFacing(worldIn, blockpos2, Blocks.CHEST.getDefaultState()), 2);
                             TileEntity tileentity1 = worldIn.getTileEntity(blockpos2);
 
                             if (tileentity1 instanceof TileEntityChest)
@@ -159,9 +146,7 @@
             }
             else
             {
-                LOGGER.error(
-                    "Failed to fetch mob spawner entity at ({}, {}, {})", position.getX(), position.getY(), position.getZ()
-                );
+                LOGGER.error("Failed to fetch mob spawner entity at ({}, {}, {})", Integer.valueOf(position.getX()), Integer.valueOf(position.getY()), Integer.valueOf(position.getZ()));
             }
 
             return true;
@@ -174,6 +159,6 @@
 
     private ResourceLocation pickMobSpawner(Random rand)
     {
-        return SPAWNERTYPES[rand.nextInt(SPAWNERTYPES.length)];
+        return net.minecraftforge.common.DungeonHooks.getRandomDungeonMob(rand);
     }
 }
