--- before/net/minecraft/world/gen/feature/WorldGenMinable.java
+++ after/net/minecraft/world/gen/feature/WorldGenMinable.java
@@ -73,7 +73,8 @@
                                 {
                                     BlockPos blockpos = new BlockPos(l1, i2, j2);
 
-                                    if (this.field_175919_c.apply(p_180709_1_.func_180495_p(blockpos)))
+                                    IBlockState state = p_180709_1_.func_180495_p(blockpos);
+                                    if (state.func_177230_c().isReplaceableOreGen(state, p_180709_1_, blockpos, this.field_175919_c))
                                     {
                                         p_180709_1_.func_180501_a(blockpos, this.field_175920_a, 2);
                                     }
@@ -98,7 +99,7 @@
             {
                 if (p_apply_1_ != null && p_apply_1_.func_177230_c() == Blocks.field_150348_b)
                 {
-                    BlockStone.EnumType blockstone$enumtype = p_apply_1_.func_177229_b(BlockStone.field_176247_a);
+                    BlockStone.EnumType blockstone$enumtype = (BlockStone.EnumType)p_apply_1_.func_177229_b(BlockStone.field_176247_a);
                     return blockstone$enumtype.func_190912_e();
                 }
                 else
