--- before/net/minecraft/world/gen/feature/WorldGenBush.java
+++ after/net/minecraft/world/gen/feature/WorldGenBush.java
@@ -14,18 +14,13 @@
         this.block = blockIn;
     }
 
-    @Override
     public boolean generate(World worldIn, Random rand, BlockPos position)
     {
-        for (int i = 0; i < 64; i++)
+        for (int i = 0; i < 64; ++i)
         {
-            BlockPos blockpos = position.add(
-                                    rand.nextInt(8) - rand.nextInt(8), rand.nextInt(4) - rand.nextInt(4), rand.nextInt(8) - rand.nextInt(8)
-                                );
+            BlockPos blockpos = position.add(rand.nextInt(8) - rand.nextInt(8), rand.nextInt(4) - rand.nextInt(4), rand.nextInt(8) - rand.nextInt(8));
 
-            if (worldIn.isAirBlock(blockpos)
-                    && (!worldIn.provider.isNether() || blockpos.getY() < 255)
-                    && this.block.canBlockStay(worldIn, blockpos, this.block.getDefaultState()))
+            if (worldIn.isAirBlock(blockpos) && (!worldIn.provider.isNether() || blockpos.getY() < worldIn.getHeight() - 1) && this.block.canBlockStay(worldIn, blockpos, this.block.getDefaultState()))
             {
                 worldIn.setBlockState(blockpos, this.block.getDefaultState(), 2);
             }
