--- before/net/minecraft/world/gen/feature/WorldGenBigTree.java
+++ after/net/minecraft/world/gen/feature/WorldGenBigTree.java
@@ -20,10 +20,10 @@
     private BlockPos field_175947_m = BlockPos.field_177992_a;
     int field_76504_e;
     int field_76501_f;
-    double field_76502_g = 0.618;
-    double field_175944_d = 0.381;
-    double field_175945_e = 1.0;
-    double field_76513_k = 1.0;
+    double field_76502_g = 0.618D;
+    double field_175944_d = 0.381D;
+    double field_175945_e = 1.0D;
+    double field_76513_k = 1.0D;
     int field_175943_g = 1;
     int field_175950_h = 12;
     int field_76508_n = 4;
@@ -43,7 +43,7 @@
             this.field_76501_f = this.field_76504_e - 1;
         }
 
-        int i = (int)(1.382 + Math.pow(this.field_76513_k * (double)this.field_76504_e / 13.0, 2.0));
+        int i = (int)(1.382D + Math.pow(this.field_76513_k * (double)this.field_76504_e / 13.0D, 2.0D));
 
         if (i < 1)
         {
@@ -52,21 +52,21 @@
 
         int j = this.field_175947_m.func_177956_o() + this.field_76501_f;
         int k = this.field_76504_e - this.field_76508_n;
-        this.field_175948_j = Lists.newArrayList();
+        this.field_175948_j = Lists.<WorldGenBigTree.FoliageCoordinates>newArrayList();
         this.field_175948_j.add(new WorldGenBigTree.FoliageCoordinates(this.field_175947_m.func_177981_b(k), j));
 
         for (; k >= 0; --k)
         {
             float f = this.func_76490_a(k);
 
-            if (!(f < 0.0F))
+            if (f >= 0.0F)
             {
                 for (int l = 0; l < i; ++l)
                 {
-                    double d0 = this.field_175945_e * (double)f * ((double)this.field_175949_k.nextFloat() + 0.328);
+                    double d0 = this.field_175945_e * (double)f * ((double)this.field_175949_k.nextFloat() + 0.328D);
                     double d1 = (double)(this.field_175949_k.nextFloat() * 2.0F) * Math.PI;
-                    double d2 = d0 * Math.sin(d1) + 0.5;
-                    double d3 = d0 * Math.cos(d1) + 0.5;
+                    double d2 = d0 * Math.sin(d1) + 0.5D;
+                    double d3 = d0 * Math.cos(d1) + 0.5D;
                     BlockPos blockpos = this.field_175947_m.func_177963_a(d2, (double)(k - 1), d3);
                     BlockPos blockpos1 = blockpos.func_177981_b(this.field_76508_n);
 
@@ -90,18 +90,18 @@
 
     void func_181631_a(BlockPos p_181631_1_, float p_181631_2_, IBlockState p_181631_3_)
     {
-        int i = (int)((double)p_181631_2_ + 0.618);
+        int i = (int)((double)p_181631_2_ + 0.618D);
 
         for (int j = -i; j <= i; ++j)
         {
             for (int k = -i; k <= i; ++k)
             {
-                if (Math.pow((double)Math.abs(j) + 0.5, 2.0) + Math.pow((double)Math.abs(k) + 0.5, 2.0) <= (double)(p_181631_2_ * p_181631_2_))
+                if (Math.pow((double)Math.abs(j) + 0.5D, 2.0D) + Math.pow((double)Math.abs(k) + 0.5D, 2.0D) <= (double)(p_181631_2_ * p_181631_2_))
                 {
                     BlockPos blockpos = p_181631_1_.func_177982_a(j, 0, k);
-                    Material material = this.field_175946_l.func_180495_p(blockpos).func_185904_a();
+                    IBlockState state = this.field_175946_l.func_180495_p(blockpos);
 
-                    if (material == Material.field_151579_a || material == Material.field_151584_j)
+                    if (state.func_177230_c().isAir(state, field_175946_l, blockpos) || state.func_177230_c().isLeaves(state, field_175946_l, blockpos))
                     {
                         this.func_175903_a(this.field_175946_l, blockpos, p_181631_3_);
                     }
@@ -137,23 +137,21 @@
 
     float func_76495_b(int p_76495_1_)
     {
-        if (p_76495_1_ < 0 || p_76495_1_ >= this.field_76508_n)
-        {
-            return -1.0F;
-        }
-        else
+        if (p_76495_1_ >= 0 && p_76495_1_ < this.field_76508_n)
         {
             return p_76495_1_ != 0 && p_76495_1_ != this.field_76508_n - 1 ? 3.0F : 2.0F;
         }
+        else
+        {
+            return -1.0F;
+        }
     }
 
     void func_175940_a(BlockPos p_175940_1_)
     {
         for (int i = 0; i < this.field_76508_n; ++i)
         {
-            this.func_181631_a(
-                p_175940_1_.func_177981_b(i), this.func_76495_b(i), Blocks.field_150362_t.func_176223_P().func_177226_a(BlockLeaves.field_176236_b, false)
-            );
+            this.func_181631_a(p_175940_1_.func_177981_b(i), this.func_76495_b(i), Blocks.field_150362_t.func_176223_P().func_177226_a(BlockLeaves.field_176236_b, Boolean.valueOf(false)));
         }
     }
 
@@ -221,7 +219,7 @@
 
     boolean func_76493_c(int p_76493_1_)
     {
-        return (double)p_76493_1_ >= (double)this.field_76504_e * 0.2;
+        return (double)p_76493_1_ >= (double)this.field_76504_e * 0.2D;
     }
 
     void func_175942_c()
@@ -271,7 +269,7 @@
             {
                 BlockPos blockpos1 = p_175936_1_.func_177963_a((double)(0.5F + (float)j * f), (double)(0.5F + (float)j * f1), (double)(0.5F + (float)j * f2));
 
-                if (!this.func_150523_a(this.field_175946_l.func_180495_p(blockpos1).func_177230_c()))
+                if (!this.isReplaceable(field_175946_l, blockpos1))
                 {
                     return j;
                 }
@@ -281,13 +279,11 @@
         }
     }
 
-    @Override
     public void func_175904_e()
     {
         this.field_76508_n = 5;
     }
 
-    @Override
     public boolean func_180709_b(World p_180709_1_, Random p_180709_2_, BlockPos p_180709_3_)
     {
         this.field_175946_l = p_180709_1_;
@@ -301,6 +297,7 @@
 
         if (!this.func_76497_e())
         {
+            this.field_175946_l = null; //Fix vanilla Mem leak, holds latest world
             return false;
         }
         else
@@ -309,15 +306,18 @@
             this.func_175941_b();
             this.func_175942_c();
             this.func_175939_d();
+            this.field_175946_l = null; //Fix vanilla Mem leak, holds latest world
             return true;
         }
     }
 
     private boolean func_76497_e()
     {
-        Block block = this.field_175946_l.func_180495_p(this.field_175947_m.func_177977_b()).func_177230_c();
+        BlockPos down = this.field_175947_m.func_177977_b();
+        net.minecraft.block.state.IBlockState state = this.field_175946_l.func_180495_p(down);
+        boolean isSoil = state.func_177230_c().canSustainPlant(state, this.field_175946_l, down, net.minecraft.util.EnumFacing.UP, ((net.minecraft.block.BlockSapling)Blocks.field_150345_g));
 
-        if (block != Blocks.field_150346_d && block != Blocks.field_150349_c && block != Blocks.field_150458_ak)
+        if (!isSoil)
         {
             return false;
         }
