--- original/net/minecraft/util/EnumFacing.java
+++ modified/net/minecraft/util/EnumFacing.java
@@ -12,8 +12,6 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.Vec3i;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public enum EnumFacing implements IStringSerializable {
     DOWN(0, 1, -1, "down", EnumFacing.AxisDirection.NEGATIVE, EnumFacing.Axis.Y, new Vec3i(0, -1, 0)),
@@ -78,8 +76,6 @@
         return byIndex(this.opposite);
     }
 
-    /** Rotate this Facing around the given axis clockwise. If this facing cannot be rotated around the given axis, returns this facing without rotating. */
-    @SideOnly(Side.CLIENT)
     public EnumFacing rotateAround(EnumFacing.Axis axis) {
         switch (axis) {
             case X:
@@ -121,8 +117,6 @@
         }
     }
 
-    /** Rotate this Facing around the X axis (NORTH => DOWN => SOUTH => UP => NORTH) */
-    @SideOnly(Side.CLIENT)
     private EnumFacing rotateX() {
         switch (this) {
             case NORTH:
@@ -140,8 +134,6 @@
         }
     }
 
-    /** Rotate this Facing around the Z axis (EAST => DOWN => WEST => UP => EAST) */
-    @SideOnly(Side.CLIENT)
     private EnumFacing rotateZ() {
         switch (this) {
             case EAST:
@@ -200,7 +192,6 @@
 
     /** Get the facing specified by the given name */
     @Nullable
-    @SideOnly(Side.CLIENT)
     public static EnumFacing byName(String name) {
         return name == null ? null : NAME_LOOKUP.get(name.toLowerCase(Locale.ROOT));
     }
@@ -230,7 +221,6 @@
         return values()[rand.nextInt(values().length)];
     }
 
-    @SideOnly(Side.CLIENT)
     public static EnumFacing getFacingFromVector(float x, float y, float z) {
         EnumFacing enumfacing = NORTH;
         float f = Float.MIN_VALUE;
@@ -284,8 +274,6 @@
         return placer.getHorizontalFacing().getOpposite();
     }
 
-    /** Get a normalized Vector that points in the direction of this Facing. */
-    @SideOnly(Side.CLIENT)
     public Vec3i getDirectionVec() {
         return this.directionVec;
     }
@@ -317,7 +305,6 @@
 
         /** Get the axis specified by the given name */
         @Nullable
-        @SideOnly(Side.CLIENT)
         public static EnumFacing.Axis byName(String name) {
             return name == null ? null : NAME_LOOKUP.get(name.toLowerCase(Locale.ROOT));
         }
