--- original/net/minecraft/util/EnumFacing.java
+++ modified/net/minecraft/util/EnumFacing.java
@@ -76,7 +76,6 @@
         return byIndex(this.opposite);
     }
 
-    /** Rotate this Facing around the given axis clockwise. If this facing cannot be rotated around the given axis, returns this facing without rotating. */
     public EnumFacing rotateAround(EnumFacing.Axis axis) {
         switch (axis) {
             case X:
@@ -118,7 +117,6 @@
         }
     }
 
-    /** Rotate this Facing around the X axis (NORTH => DOWN => SOUTH => UP => NORTH) */
     private EnumFacing rotateX() {
         switch (this) {
             case NORTH:
@@ -136,7 +134,6 @@
         }
     }
 
-    /** Rotate this Facing around the Z axis (EAST => DOWN => WEST => UP => EAST) */
     private EnumFacing rotateZ() {
         switch (this) {
             case EAST:
@@ -277,7 +274,6 @@
         return placer.getHorizontalFacing().getOpposite();
     }
 
-    /** Get a normalized Vector that points in the direction of this Facing. */
     public Vec3i getDirectionVec() {
         return this.directionVec;
     }
