--- original/net/minecraft/util/IntIdentityHashBiMap.java
+++ modified/net/minecraft/util/IntIdentityHashBiMap.java
@@ -6,6 +6,8 @@
 import java.util.Iterator;
 import javax.annotation.Nullable;
 import net.minecraft.util.math.MathHelper;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class IntIdentityHashBiMap<K> implements IObjectIntIterable<K> {
     private static final Object EMPTY = null;
