--- before/net/minecraft/crash/CrashReport.java
+++ after/net/minecraft/crash/CrashReport.java
@@ -116,6 +116,7 @@
                 return IntCache.getCacheSizes();
             }
         });
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().enhanceCrashReport(this, this.systemDetailsCategory);
     }
 
     public String getDescription()
@@ -198,13 +199,14 @@
             IOUtils.closeQuietly((Writer)printwriter);
         }
 
-        return s;
+        return s + net.minecraftforge.common.ForgeHooks.gatherMixinInfo(throwable);
     }
 
     public String getCompleteReport()
     {
         StringBuilder stringbuilder = new StringBuilder();
         stringbuilder.append("---- Minecraft Crash Report ----\n");
+        net.minecraftforge.fml.relauncher.CoreModManager.onCrash(stringbuilder);
         stringbuilder.append("// ");
         stringbuilder.append(getWittyComment());
         stringbuilder.append("\n\n");
