--- before/net/minecraft/command/AdvancementCommand.java
+++ after/net/minecraft/command/AdvancementCommand.java
@@ -34,7 +34,7 @@
     {
         if (p_184881_3_.length < 1)
         {
-            throw new WrongUsageException("commands.advancement.usage");
+            throw new WrongUsageException("commands.advancement.usage", new Object[0]);
         }
         else
         {
@@ -61,7 +61,7 @@
             {
                 if (!"test".equals(p_184881_3_[0]))
                 {
-                    throw new WrongUsageException("commands.advancement.usage");
+                    throw new WrongUsageException("commands.advancement.usage", new Object[0]);
                 }
 
                 if (p_184881_3_.length == 3)
@@ -72,7 +72,7 @@
                 {
                     if (p_184881_3_.length != 4)
                     {
-                        throw new WrongUsageException("commands.advancement.test.usage");
+                        throw new WrongUsageException("commands.advancement.test.usage", new Object[0]);
                     }
 
                     this.func_192554_c(p_184881_2_, func_184888_a(p_184881_1_, p_184881_2_, p_184881_3_[1]), func_192551_a(p_184881_1_, p_184881_3_[2]), p_184881_3_[3]);
@@ -117,12 +117,12 @@
 
                 if (!advancement.func_192073_f().keySet().contains(s))
                 {
-                    throw new CommandException("commands.advancement.criterionNotFound", advancement.func_192067_g(), p_193516_3_[4]);
+                    throw new CommandException("commands.advancement.criterionNotFound", new Object[] {advancement.func_192067_g(), p_193516_3_[4]});
                 }
 
                 if (!p_193516_5_.func_193535_a(p_193516_4_, advancement, s))
                 {
-                    throw new CommandException(p_193516_5_.field_193541_d + ".criterion.failed", advancement.func_192067_g(), p_193516_4_.func_70005_c_(), s);
+                    throw new CommandException(p_193516_5_.field_193541_d + ".criterion.failed", new Object[] {advancement.func_192067_g(), p_193516_4_.func_70005_c_(), s});
                 }
 
                 func_152373_a(p_193516_2_, this, p_193516_5_.field_193541_d + ".criterion.success", new Object[] {advancement.func_192067_g(), p_193516_4_.func_70005_c_(), s});
@@ -158,7 +158,7 @@
 
     private List<Advancement> func_193514_a(Advancement p_193514_1_, AdvancementCommand.Mode p_193514_2_)
     {
-        List<Advancement> list = Lists.newArrayList();
+        List<Advancement> list = Lists.<Advancement>newArrayList();
 
         if (p_193514_2_.field_193555_h)
         {
@@ -185,11 +185,11 @@
 
         if (criterionprogress == null)
         {
-            throw new CommandException("commands.advancement.criterionNotFound", p_192554_3_.func_192067_g(), p_192554_4_);
+            throw new CommandException("commands.advancement.criterionNotFound", new Object[] {p_192554_3_.func_192067_g(), p_192554_4_});
         }
         else if (!criterionprogress.func_192151_a())
         {
-            throw new CommandException("commands.advancement.test.criterion.notDone", p_192554_2_.func_70005_c_(), p_192554_3_.func_192067_g(), p_192554_4_);
+            throw new CommandException("commands.advancement.test.criterion.notDone", new Object[] {p_192554_2_.func_70005_c_(), p_192554_3_.func_192067_g(), p_192554_4_});
         }
         else
         {
@@ -203,7 +203,7 @@
 
         if (!advancementprogress.func_192105_a())
         {
-            throw new CommandException("commands.advancement.test.advancement.notDone", p_192552_2_.func_70005_c_(), p_192552_3_.func_192067_g());
+            throw new CommandException("commands.advancement.test.advancement.notDone", new Object[] {p_192552_2_.func_70005_c_(), p_192552_3_.func_192067_g()});
         }
         else
         {
@@ -277,13 +277,13 @@
                 }
             }
 
-            return Collections.emptyList();
+            return Collections.<String>emptyList();
         }
     }
 
     private List<ResourceLocation> func_193517_a(MinecraftServer p_193517_1_)
     {
-        List<ResourceLocation> list = Lists.newArrayList();
+        List<ResourceLocation> list = Lists.<ResourceLocation>newArrayList();
 
         for (Advancement advancement : p_193517_1_.func_191949_aK().func_192780_b())
         {
@@ -304,7 +304,7 @@
 
         if (advancement == null)
         {
-            throw new CommandException("commands.advancement.advancementNotFound", p_192551_1_);
+            throw new CommandException("commands.advancement.advancementNotFound", new Object[] {p_192551_1_});
         }
         else
         {
@@ -390,7 +390,7 @@
 
         CommandException func_193534_a()
         {
-            return new CommandException(this.field_193541_d + ".usage");
+            return new CommandException(this.field_193541_d + ".usage", new Object[0]);
         }
 
         public int func_193532_a(EntityPlayerMP p_193532_1_, Iterable<Advancement> p_193532_2_)
@@ -440,7 +440,7 @@
 
         CommandException func_193544_a(AdvancementCommand.ActionType p_193544_1_)
         {
-            return new CommandException(p_193544_1_.field_193541_d + "." + this.field_193554_g + ".usage");
+            return new CommandException(p_193544_1_.field_193541_d + "." + this.field_193554_g + ".usage", new Object[0]);
         }
 
         void func_193546_a(ICommandSender p_193546_1_, AdvancementCommand p_193546_2_, AdvancementCommand.ActionType p_193546_3_, Object... p_193546_4_)
