--- before/net/minecraft/command/CommandBase.java
+++ after/net/minecraft/command/CommandBase.java
@@ -220,18 +220,21 @@
 
     public static boolean parseBoolean(String input) throws CommandException
     {
-        if ("true".equals(input) || "1".equals(input))
+        if (!"true".equals(input) && !"1".equals(input))
+        {
+            if (!"false".equals(input) && !"0".equals(input))
+            {
+                throw new CommandException("commands.generic.boolean.invalid", new Object[] {input});
+            }
+            else
+            {
+                return false;
+            }
+        }
+        else
         {
             return true;
         }
-        else if (!"false".equals(input) && !"0".equals(input))
-        {
-            throw new CommandException("commands.generic.boolean.invalid", input);
-        }
-        else
-        {
-            return false;
-        }
     }
 
     public static EntityPlayerMP getCommandSenderAsPlayer(ICommandSender sender) throws PlayerNotFoundException
@@ -391,14 +394,14 @@
     {
         ITextComponent itextcomponent = new TextComponentString("");
 
-        for (int i = index; i < args.length; i++)
+        for (int i = index; i < args.length; ++i)
         {
             if (i > index)
             {
                 itextcomponent.appendText(" ");
             }
 
-            ITextComponent itextcomponent1 = new TextComponentString(args[i]);
+            ITextComponent itextcomponent1 = net.minecraftforge.common.ForgeHooks.newChatWithLinks(args[i]); // Forge: links for messages
 
             if (p_147176_3_)
             {
@@ -427,7 +430,7 @@
     {
         StringBuilder stringbuilder = new StringBuilder();
 
-        for (int i = startPos; i < args.length; i++)
+        for (int i = startPos; i < args.length; ++i)
         {
             if (i > startPos)
             {
@@ -702,7 +705,7 @@
                     break;
                 }
 
-                Comparable<?> comparable = getValueHelper((IProperty<?>)iproperty, iterator1.next());
+                Comparable<?> comparable = getValueHelper(iproperty, iterator1.next());
 
                 if (comparable == null)
                 {
@@ -726,7 +729,7 @@
     {
         StringBuilder stringbuilder = new StringBuilder();
 
-        for (int i = 0; i < elements.length; i++)
+        for (int i = 0; i < elements.length; ++i)
         {
             String s = elements[i].toString();
 
@@ -752,7 +755,7 @@
     {
         ITextComponent itextcomponent = new TextComponentString("");
 
-        for (int i = 0; i < components.size(); i++)
+        for (int i = 0; i < components.size(); ++i)
         {
             if (i > 0)
             {
@@ -856,7 +859,7 @@
 
         if (!possibleCompletions.isEmpty())
         {
-            for (String s1 : Iterables.transform((Iterable<Object>)possibleCompletions, Functions.toStringFunction()))
+            for (String s1 : Iterables.transform(possibleCompletions, Functions.toStringFunction()))
             {
                 if (doesStringStartWith(s, s1))
                 {
