--- before/net/minecraft/command/CommandEntityData.java
+++ after/net/minecraft/command/CommandEntityData.java
@@ -29,7 +29,7 @@
     {
         if (p_184881_3_.length < 2)
         {
-            throw new WrongUsageException("commands.entitydata.usage");
+            throw new WrongUsageException("commands.entitydata.usage", new Object[0]);
         }
         else
         {
@@ -37,7 +37,7 @@
 
             if (entity instanceof EntityPlayer)
             {
-                throw new CommandException("commands.entitydata.noPlayers", entity.func_145748_c_());
+                throw new CommandException("commands.entitydata.noPlayers", new Object[] {entity.func_145748_c_()});
             }
             else
             {
@@ -51,7 +51,7 @@
                 }
                 catch (NBTException nbtexception)
                 {
-                    throw new CommandException("commands.entitydata.tagError", nbtexception.getMessage());
+                    throw new CommandException("commands.entitydata.tagError", new Object[] {nbtexception.getMessage()});
                 }
 
                 UUID uuid = entity.func_110124_au();
@@ -60,7 +60,7 @@
 
                 if (nbttagcompound.equals(nbttagcompound1))
                 {
-                    throw new CommandException("commands.entitydata.failed", nbttagcompound.toString());
+                    throw new CommandException("commands.entitydata.failed", new Object[] {nbttagcompound.toString()});
                 }
                 else
                 {
