--- before/net/minecraft/command/CommandFill.java
+++ after/net/minecraft/command/CommandFill.java
@@ -96,7 +96,7 @@
                 NBTTagCompound nbttagcompound = new NBTTagCompound();
                 boolean flag = false;
 
-                if (p_184881_3_.length >= 10 && block.func_149716_u())
+                if (p_184881_3_.length >= 10 && block.hasTileEntity(iblockstate))
                 {
                     String s = func_180529_a(p_184881_3_, 9);
 
@@ -137,7 +137,7 @@
                                             continue;
                                         }
                                     }
-                                    else if ("replace".equals(p_184881_3_[8]) && !block.func_149716_u() && p_184881_3_.length > 9)
+                                    else if ("replace".equals(p_184881_3_[8]) && !block.hasTileEntity(iblockstate) && p_184881_3_.length > 9)
                                     {
                                         Block block1 = CommandBase.func_147180_g(p_184881_2_, p_184881_3_[9]);
 
