--- before/net/minecraft/command/CommandGameMode.java
+++ after/net/minecraft/command/CommandGameMode.java
@@ -32,18 +32,18 @@
     {
         if (p_184881_3_.length <= 0)
         {
-            throw new WrongUsageException("commands.gamemode.usage");
+            throw new WrongUsageException("commands.gamemode.usage", new Object[0]);
         }
         else
         {
             GameType gametype = this.func_71539_b(p_184881_2_, p_184881_3_[0]);
             EntityPlayer entityplayer = p_184881_3_.length >= 2 ? func_184888_a(p_184881_1_, p_184881_2_, p_184881_3_[1]) : func_71521_c(p_184881_2_);
             entityplayer.func_71033_a(gametype);
-            ITextComponent itextcomponent = new TextComponentTranslation("gameMode." + gametype.func_77149_b());
+            ITextComponent itextcomponent = new TextComponentTranslation("gameMode." + gametype.func_77149_b(), new Object[0]);
 
             if (p_184881_2_.func_130014_f_().func_82736_K().func_82766_b("sendCommandFeedback"))
             {
-                entityplayer.func_145747_a(new TextComponentTranslation("gameMode.changed", itextcomponent));
+                entityplayer.func_145747_a(new TextComponentTranslation("gameMode.changed", new Object[] {itextcomponent}));
             }
 
             if (entityplayer == p_184881_2_)
