--- before/net/minecraft/command/CommandStats.java
+++ after/net/minecraft/command/CommandStats.java
@@ -35,7 +35,7 @@
     {
         if (p_184881_3_.length < 1)
         {
-            throw new WrongUsageException("commands.stats.usage");
+            throw new WrongUsageException("commands.stats.usage", new Object[0]);
         }
         else
         {
@@ -49,7 +49,7 @@
             {
                 if (!"block".equals(p_184881_3_[0]))
                 {
-                    throw new WrongUsageException("commands.stats.usage");
+                    throw new WrongUsageException("commands.stats.usage", new Object[0]);
                 }
 
                 flag = true;
@@ -61,7 +61,7 @@
             {
                 if (p_184881_3_.length < 5)
                 {
-                    throw new WrongUsageException("commands.stats.block.usage");
+                    throw new WrongUsageException("commands.stats.block.usage", new Object[0]);
                 }
 
                 i = 4;
@@ -70,7 +70,7 @@
             {
                 if (p_184881_3_.length < 3)
                 {
-                    throw new WrongUsageException("commands.stats.entity.usage");
+                    throw new WrongUsageException("commands.stats.entity.usage", new Object[0]);
                 }
 
                 i = 2;
@@ -84,27 +84,27 @@
                 {
                     if (i == 5)
                     {
-                        throw new WrongUsageException("commands.stats.block.set.usage");
+                        throw new WrongUsageException("commands.stats.block.set.usage", new Object[0]);
                     }
 
-                    throw new WrongUsageException("commands.stats.entity.set.usage");
+                    throw new WrongUsageException("commands.stats.entity.set.usage", new Object[0]);
                 }
             }
             else
             {
                 if (!"clear".equals(s))
                 {
-                    throw new WrongUsageException("commands.stats.usage");
+                    throw new WrongUsageException("commands.stats.usage", new Object[0]);
                 }
 
                 if (p_184881_3_.length < i + 1)
                 {
                     if (i == 5)
                     {
-                        throw new WrongUsageException("commands.stats.block.clear.usage");
+                        throw new WrongUsageException("commands.stats.block.clear.usage", new Object[0]);
                     }
 
-                    throw new WrongUsageException("commands.stats.entity.clear.usage");
+                    throw new WrongUsageException("commands.stats.entity.clear.usage", new Object[0]);
                 }
             }
 
@@ -112,7 +112,7 @@
 
             if (commandresultstats$type == null)
             {
-                throw new CommandException("commands.stats.failed");
+                throw new CommandException("commands.stats.failed", new Object[0]);
             }
             else
             {
@@ -126,7 +126,7 @@
 
                     if (tileentity == null)
                     {
-                        throw new CommandException("commands.stats.noCompatibleBlock", blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p());
+                        throw new CommandException("commands.stats.noCompatibleBlock", new Object[] {blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p()});
                     }
 
                     if (tileentity instanceof TileEntityCommandBlock)
@@ -137,7 +137,7 @@
                     {
                         if (!(tileentity instanceof TileEntitySign))
                         {
-                            throw new CommandException("commands.stats.noCompatibleBlock", blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p());
+                            throw new CommandException("commands.stats.noCompatibleBlock", new Object[] {blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p()});
                         }
 
                         commandresultstats = ((TileEntitySign)tileentity).func_174880_d();
@@ -156,7 +156,7 @@
 
                     if (s1.isEmpty() || s2.isEmpty())
                     {
-                        throw new CommandException("commands.stats.failed");
+                        throw new CommandException("commands.stats.failed", new Object[0]);
                     }
 
                     CommandResultStats.func_179667_a(commandresultstats, commandresultstats$type, s1, s2);
@@ -212,7 +212,7 @@
     protected List<String> func_184927_a(MinecraftServer p_184927_1_)
     {
         Collection<ScoreObjective> collection = p_184927_1_.func_71218_a(0).func_96441_U().func_96514_c();
-        List<String> list = Lists.newArrayList();
+        List<String> list = Lists.<String>newArrayList();
 
         for (ScoreObjective scoreobjective : collection)
         {
