--- before/net/minecraft/command/CommandTime.java
+++ after/net/minecraft/command/CommandTime.java
@@ -86,7 +86,7 @@
             }
         }
 
-        throw new WrongUsageException("commands.time.usage");
+        throw new WrongUsageException("commands.time.usage", new Object[0]);
     }
 
     public List<String> func_184883_a(MinecraftServer p_184883_1_, ICommandSender p_184883_2_, String[] p_184883_3_, @Nullable BlockPos p_184883_4_)
