--- before/net/minecraft/command/CommandXP.java
+++ after/net/minecraft/command/CommandXP.java
@@ -28,7 +28,7 @@
     {
         if (p_184881_3_.length <= 0)
         {
-            throw new WrongUsageException("commands.xp.usage");
+            throw new WrongUsageException("commands.xp.usage", new Object[0]);
         }
         else
         {
@@ -71,7 +71,7 @@
 
                 if (flag1)
                 {
-                    throw new CommandException("commands.xp.failure.widthdrawXp");
+                    throw new CommandException("commands.xp.failure.widthdrawXp", new Object[0]);
                 }
 
                 entityplayer.func_71023_q(i);
