--- before/net/minecraft/command/CommandTrigger.java
+++ after/net/minecraft/command/CommandTrigger.java
@@ -15,30 +15,26 @@
 
 public class CommandTrigger extends CommandBase
 {
-    @Override
     public String func_71517_b()
     {
         return "trigger";
     }
 
-    @Override
     public int func_82362_a()
     {
         return 0;
     }
 
-    @Override
     public String func_71518_a(ICommandSender p_71518_1_)
     {
         return "commands.trigger.usage";
     }
 
-    @Override
     public void func_184881_a(MinecraftServer p_184881_1_, ICommandSender p_184881_2_, String[] p_184881_3_) throws CommandException
     {
         if (p_184881_3_.length < 3)
         {
-            throw new WrongUsageException("commands.trigger.usage");
+            throw new WrongUsageException("commands.trigger.usage", new Object[0]);
         }
         else
         {
@@ -54,7 +50,7 @@
 
                 if (!(entity instanceof EntityPlayerMP))
                 {
-                    throw new CommandException("commands.trigger.invalidPlayer");
+                    throw new CommandException("commands.trigger.invalidPlayer", new Object[0]);
                 }
 
                 entityplayermp = (EntityPlayerMP)entity;
@@ -69,7 +65,7 @@
 
                 if (!scoreboard.func_178819_b(entityplayermp.func_70005_c_(), scoreobjective))
                 {
-                    throw new CommandException("commands.trigger.invalidObjective", p_184881_3_[0]);
+                    throw new CommandException("commands.trigger.invalidObjective", new Object[] {p_184881_3_[0]});
                 }
                 else
                 {
@@ -77,7 +73,7 @@
 
                     if (score.func_178816_g())
                     {
-                        throw new CommandException("commands.trigger.disabled", p_184881_3_[0]);
+                        throw new CommandException("commands.trigger.disabled", new Object[] {p_184881_3_[0]});
                     }
                     else
                     {
@@ -89,7 +85,7 @@
                         {
                             if (!"add".equals(p_184881_3_[1]))
                             {
-                                throw new CommandException("commands.trigger.invalidMode", p_184881_3_[1]);
+                                throw new CommandException("commands.trigger.invalidMode", new Object[] {p_184881_3_[1]});
                             }
 
                             score.func_96649_a(i);
@@ -106,18 +102,17 @@
             }
             else
             {
-                throw new CommandException("commands.trigger.invalidObjective", p_184881_3_[0]);
+                throw new CommandException("commands.trigger.invalidObjective", new Object[] {p_184881_3_[0]});
             }
         }
     }
 
-    @Override
     public List<String> func_184883_a(MinecraftServer p_184883_1_, ICommandSender p_184883_2_, String[] p_184883_3_, @Nullable BlockPos p_184883_4_)
     {
         if (p_184883_3_.length == 1)
         {
             Scoreboard scoreboard = p_184883_1_.func_71218_a(0).func_96441_U();
-            List<String> list = Lists.newArrayList();
+            List<String> list = Lists.<String>newArrayList();
 
             for (ScoreObjective scoreobjective : scoreboard.func_96514_c())
             {
@@ -127,7 +122,7 @@
                 }
             }
 
-            return func_71530_a(p_184883_3_, list.toArray(new String[list.size()]));
+            return func_71530_a(p_184883_3_, (String[])list.toArray(new String[list.size()]));
         }
         else
         {
