--- before/net/minecraft/command/CommandCompare.java
+++ after/net/minecraft/command/CommandCompare.java
@@ -33,7 +33,7 @@
     {
         if (p_184881_3_.length < 9)
         {
-            throw new WrongUsageException("commands.compare.usage");
+            throw new WrongUsageException("commands.compare.usage", new Object[0]);
         }
         else
         {
@@ -47,7 +47,7 @@
 
             if (i > 524288)
             {
-                throw new CommandException("commands.compare.tooManyBlocks", i, 524288);
+                throw new CommandException("commands.compare.tooManyBlocks", new Object[] {i, 524288});
             }
             else if (structureboundingbox.field_78895_b >= 0 && structureboundingbox.field_78894_e < 256 && structureboundingbox1.field_78895_b >= 0 && structureboundingbox1.field_78894_e < 256)
             {
@@ -115,7 +115,7 @@
 
                                     if (flag1)
                                     {
-                                        throw new CommandException("commands.compare.failed");
+                                        throw new CommandException("commands.compare.failed", new Object[0]);
                                     }
                                 }
                             }
@@ -127,12 +127,12 @@
                 }
                 else
                 {
-                    throw new CommandException("commands.compare.outOfWorld");
+                    throw new CommandException("commands.compare.outOfWorld", new Object[0]);
                 }
             }
             else
             {
-                throw new CommandException("commands.compare.outOfWorld");
+                throw new CommandException("commands.compare.outOfWorld", new Object[0]);
             }
         }
     }
