--- before/net/minecraft/command/CommandExecuteAt.java
+++ after/net/minecraft/command/CommandExecuteAt.java
@@ -32,7 +32,7 @@
     {
         if (p_184881_3_.length < 5)
         {
-            throw new WrongUsageException("commands.execute.usage");
+            throw new WrongUsageException("commands.execute.usage", new Object[0]);
         }
         else
         {
@@ -54,19 +54,19 @@
 
                 if (!world.func_175667_e(blockpos))
                 {
-                    throw new CommandException("commands.execute.failed", "detect", entity.func_70005_c_());
+                    throw new CommandException("commands.execute.failed", new Object[] {"detect", entity.func_70005_c_()});
                 }
 
                 IBlockState iblockstate = world.func_180495_p(blockpos);
 
                 if (iblockstate.func_177230_c() != block)
                 {
-                    throw new CommandException("commands.execute.failed", "detect", entity.func_70005_c_());
+                    throw new CommandException("commands.execute.failed", new Object[] {"detect", entity.func_70005_c_()});
                 }
 
                 if (!CommandBase.func_190791_b(block, p_184881_3_[9]).apply(iblockstate))
                 {
-                    throw new CommandException("commands.execute.failed", "detect", entity.func_70005_c_());
+                    throw new CommandException("commands.execute.failed", new Object[] {"detect", entity.func_70005_c_()});
                 }
 
                 i = 10;
@@ -82,12 +82,12 @@
 
                 if (j < 1)
                 {
-                    throw new CommandException("commands.execute.allInvocationsFailed", s);
+                    throw new CommandException("commands.execute.allInvocationsFailed", new Object[] {s});
                 }
             }
-            catch (Throwable throwable)
+            catch (Throwable var23)
             {
-                throw new CommandException("commands.execute.failed", s, entity.func_70005_c_());
+                throw new CommandException("commands.execute.failed", new Object[] {s, entity.func_70005_c_()});
             }
         }
     }
