--- before/net/minecraft/command/CommandDifficulty.java
+++ after/net/minecraft/command/CommandDifficulty.java
@@ -29,13 +29,13 @@
     {
         if (p_184881_3_.length <= 0)
         {
-            throw new WrongUsageException("commands.difficulty.usage");
+            throw new WrongUsageException("commands.difficulty.usage", new Object[0]);
         }
         else
         {
             EnumDifficulty enumdifficulty = this.func_180531_e(p_184881_3_[0]);
             p_184881_1_.func_147139_a(enumdifficulty);
-            func_152373_a(p_184881_2_, this, "commands.difficulty.success", new Object[] {new TextComponentTranslation(enumdifficulty.func_151526_b())});
+            func_152373_a(p_184881_2_, this, "commands.difficulty.success", new Object[] {new TextComponentTranslation(enumdifficulty.func_151526_b(), new Object[0])});
         }
     }
 
