--- before/net/minecraft/command/CommandSenderWrapper.java
+++ after/net/minecraft/command/CommandSenderWrapper.java
@@ -45,12 +45,12 @@
 
     public CommandSenderWrapper func_193999_a(int p_193999_1_)
     {
-        return this.field_194004_d != null && this.field_194004_d <= p_193999_1_ ? this : new CommandSenderWrapper(this.field_193043_a, this.field_194002_b, this.field_194003_c, p_193999_1_, this.field_194005_e, this.field_194006_f);
+        return this.field_194004_d != null && this.field_194004_d.intValue() <= p_193999_1_ ? this : new CommandSenderWrapper(this.field_193043_a, this.field_194002_b, this.field_194003_c, p_193999_1_, this.field_194005_e, this.field_194006_f);
     }
 
     public CommandSenderWrapper func_194001_a(boolean p_194001_1_)
     {
-        return this.field_194006_f == null || this.field_194006_f && !p_194001_1_ ? new CommandSenderWrapper(this.field_193043_a, this.field_194002_b, this.field_194003_c, this.field_194004_d, this.field_194005_e, p_194001_1_) : this;
+        return this.field_194006_f == null || this.field_194006_f.booleanValue() && !p_194001_1_ ? new CommandSenderWrapper(this.field_193043_a, this.field_194002_b, this.field_194003_c, this.field_194004_d, this.field_194005_e, p_194001_1_) : this;
     }
 
     public CommandSenderWrapper func_194000_i()
@@ -70,7 +70,7 @@
 
     public void func_145747_a(ITextComponent p_145747_1_)
     {
-        if (this.field_194006_f == null || this.field_194006_f)
+        if (this.field_194006_f == null || this.field_194006_f.booleanValue())
         {
             this.field_193043_a.func_145747_a(p_145747_1_);
         }
@@ -78,7 +78,7 @@
 
     public boolean func_70003_b(int p_70003_1_, String p_70003_2_)
     {
-        return this.field_194004_d != null && this.field_194004_d < p_70003_1_ ? false : this.field_193043_a.func_70003_b(p_70003_1_, p_70003_2_);
+        return this.field_194004_d != null && this.field_194004_d.intValue() < p_70003_1_ ? false : this.field_193043_a.func_70003_b(p_70003_1_, p_70003_2_);
     }
 
     public BlockPos func_180425_c()
@@ -118,7 +118,7 @@
 
     public boolean func_174792_t_()
     {
-        return this.field_194006_f != null ? this.field_194006_f : this.field_193043_a.func_174792_t_();
+        return this.field_194006_f != null ? this.field_194006_f.booleanValue() : this.field_193043_a.func_174792_t_();
     }
 
     public void func_174794_a(CommandResultStats.Type p_174794_1_, int p_174794_2_)
