--- before/net/minecraft/command/CommandResultStats.java
+++ after/net/minecraft/command/CommandResultStats.java
@@ -16,8 +16,14 @@
 {
     private static final int field_179676_a = CommandResultStats.Type.values().length;
     private static final String[] field_179674_b = new String[field_179676_a];
-    private String[] field_179675_c = field_179674_b;
-    private String[] field_179673_d = field_179674_b;
+    private String[] field_179675_c;
+    private String[] field_179673_d;
+
+    public CommandResultStats()
+    {
+        this.field_179675_c = field_179674_b;
+        this.field_179673_d = field_179674_b;
+    }
 
     public void func_184932_a(MinecraftServer p_184932_1_, final ICommandSender p_184932_2_, CommandResultStats.Type p_184932_3_, int p_184932_4_)
     {
@@ -78,7 +84,7 @@
             {
                 s1 = CommandBase.func_184891_e(p_184932_1_, icommandsender, s);
             }
-            catch (CommandException commandexception)
+            catch (CommandException var12)
             {
                 return;
             }
