--- before/net/minecraft/command/CommandGive.java
+++ after/net/minecraft/command/CommandGive.java
@@ -35,7 +35,7 @@
     {
         if (p_184881_3_.length < 2)
         {
-            throw new WrongUsageException("commands.give.usage");
+            throw new WrongUsageException("commands.give.usage", new Object[0]);
         }
         else
         {
@@ -55,7 +55,7 @@
                 }
                 catch (NBTException nbtexception)
                 {
-                    throw new CommandException("commands.give.tagError", nbtexception.getMessage());
+                    throw new CommandException("commands.give.tagError", new Object[] {nbtexception.getMessage()});
                 }
             }
 
