--- before/net/minecraft/command/CommandShowSeed.java
+++ after/net/minecraft/command/CommandShowSeed.java
@@ -30,6 +30,6 @@
     public void func_184881_a(MinecraftServer p_184881_1_, ICommandSender p_184881_2_, String[] p_184881_3_) throws CommandException
     {
         World world = (World)(p_184881_2_ instanceof EntityPlayer ? ((EntityPlayer)p_184881_2_).field_70170_p : p_184881_1_.func_71218_a(0));
-        p_184881_2_.func_145747_a(new TextComponentTranslation("commands.seed.success", world.func_72905_C()));
+        p_184881_2_.func_145747_a(new TextComponentTranslation("commands.seed.success", new Object[] {world.func_72905_C()}));
     }
 }
