--- before/net/minecraft/command/CommandSetSpawnpoint.java
+++ after/net/minecraft/command/CommandSetSpawnpoint.java
@@ -28,7 +28,7 @@
     {
         if (p_184881_3_.length > 1 && p_184881_3_.length < 4)
         {
-            throw new WrongUsageException("commands.spawnpoint.usage");
+            throw new WrongUsageException("commands.spawnpoint.usage", new Object[0]);
         }
         else
         {
