--- before/net/minecraft/command/CommandBase.java
+++ after/net/minecraft/command/CommandBase.java
@@ -220,18 +220,21 @@
 
     public static boolean func_180527_d(String p_180527_0_) throws CommandException
     {
-        if ("true".equals(p_180527_0_) || "1".equals(p_180527_0_))
+        if (!"true".equals(p_180527_0_) && !"1".equals(p_180527_0_))
+        {
+            if (!"false".equals(p_180527_0_) && !"0".equals(p_180527_0_))
+            {
+                throw new CommandException("commands.generic.boolean.invalid", new Object[] {p_180527_0_});
+            }
+            else
+            {
+                return false;
+            }
+        }
+        else
         {
             return true;
         }
-        else if (!"false".equals(p_180527_0_) && !"0".equals(p_180527_0_))
-        {
-            throw new CommandException("commands.generic.boolean.invalid", p_180527_0_);
-        }
-        else
-        {
-            return false;
-        }
     }
 
     public static EntityPlayerMP func_71521_c(ICommandSender p_71521_0_) throws PlayerNotFoundException
@@ -398,7 +401,7 @@
                 itextcomponent.func_150258_a(" ");
             }
 
-            ITextComponent itextcomponent1 = new TextComponentString(p_147176_1_[i]);
+            ITextComponent itextcomponent1 = net.minecraftforge.common.ForgeHooks.newChatWithLinks(p_147176_1_[i]); // Forge: links for messages
 
             if (p_147176_3_)
             {
