--- before/net/minecraft/command/CommandEffect.java
+++ after/net/minecraft/command/CommandEffect.java
@@ -31,17 +31,17 @@
     {
         if (p_184881_3_.length < 2)
         {
-            throw new WrongUsageException("commands.effect.usage");
+            throw new WrongUsageException("commands.effect.usage", new Object[0]);
         }
         else
         {
-            EntityLivingBase entitylivingbase = func_184884_a(p_184881_1_, p_184881_2_, p_184881_3_[0], EntityLivingBase.class);
+            EntityLivingBase entitylivingbase = (EntityLivingBase)func_184884_a(p_184881_1_, p_184881_2_, p_184881_3_[0], EntityLivingBase.class);
 
             if ("clear".equals(p_184881_3_[1]))
             {
                 if (entitylivingbase.func_70651_bq().isEmpty())
                 {
-                    throw new CommandException("commands.effect.failure.notActive.all", entitylivingbase.func_70005_c_());
+                    throw new CommandException("commands.effect.failure.notActive.all", new Object[] {entitylivingbase.func_70005_c_()});
                 }
                 else
                 {
@@ -57,14 +57,14 @@
                 {
                     potion = Potion.func_188412_a(func_180528_a(p_184881_3_[1], 1));
                 }
-                catch (NumberInvalidException numberinvalidexception)
+                catch (NumberInvalidException var11)
                 {
                     potion = Potion.func_180142_b(p_184881_3_[1]);
                 }
 
                 if (potion == null)
                 {
-                    throw new NumberInvalidException("commands.effect.notFound", p_184881_3_[1]);
+                    throw new NumberInvalidException("commands.effect.notFound", new Object[] {p_184881_3_[1]});
                 }
                 else
                 {
@@ -106,16 +106,16 @@
                     {
                         PotionEffect potioneffect = new PotionEffect(potion, i, k, false, flag);
                         entitylivingbase.func_70690_d(potioneffect);
-                        func_152373_a(p_184881_2_, this, "commands.effect.success", new Object[] {new TextComponentTranslation(potioneffect.func_76453_d()), Potion.func_188409_a(potion), k, entitylivingbase.func_70005_c_(), j});
+                        func_152373_a(p_184881_2_, this, "commands.effect.success", new Object[] {new TextComponentTranslation(potioneffect.func_76453_d(), new Object[0]), Potion.func_188409_a(potion), k, entitylivingbase.func_70005_c_(), j});
                     }
                     else if (entitylivingbase.func_70644_a(potion))
                     {
                         entitylivingbase.func_184589_d(potion);
-                        func_152373_a(p_184881_2_, this, "commands.effect.success.removed", new Object[] {new TextComponentTranslation(potion.func_76393_a()), entitylivingbase.func_70005_c_()});
+                        func_152373_a(p_184881_2_, this, "commands.effect.success.removed", new Object[] {new TextComponentTranslation(potion.func_76393_a(), new Object[0]), entitylivingbase.func_70005_c_()});
                     }
                     else
                     {
-                        throw new CommandException("commands.effect.failure.notActive", new TextComponentTranslation(potion.func_76393_a()), entitylivingbase.func_70005_c_());
+                        throw new CommandException("commands.effect.failure.notActive", new Object[] {new TextComponentTranslation(potion.func_76393_a(), new Object[0]), entitylivingbase.func_70005_c_()});
                     }
                 }
             }
