--- before/net/minecraft/command/EntitySelector.java
+++ after/net/minecraft/command/EntitySelector.java
@@ -13,6 +13,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.Map.Entry;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import javax.annotation.Nullable;
@@ -42,7 +43,7 @@
     private static final Pattern field_82389_a = Pattern.compile("^@([pares])(?:\\[([^ ]*)\\])?$");
     private static final Splitter field_190828_b = Splitter.on(',').omitEmptyStrings();
     private static final Splitter field_190829_c = Splitter.on('=').limit(2);
-    private static final Set<String> field_190830_d = Sets.newHashSet();
+    private static final Set<String> field_190830_d = Sets.<String>newHashSet();
     private static final String field_190831_e = func_190826_c("r");
     private static final String field_190832_f = func_190826_c("rm");
     private static final String field_190833_g = func_190826_c("l");
@@ -81,7 +82,7 @@
     @Nullable
     public static EntityPlayerMP func_82386_a(ICommandSender p_82386_0_, String p_82386_1_) throws CommandException
     {
-        return func_179652_a(p_82386_0_, p_82386_1_, EntityPlayerMP.class);
+        return (EntityPlayerMP)func_179652_a(p_82386_0_, p_82386_1_, EntityPlayerMP.class);
     }
 
     public static List<EntityPlayerMP> func_193531_b(ICommandSender p_193531_0_, String p_193531_1_) throws CommandException
@@ -93,7 +94,7 @@
     public static <T extends Entity> T func_179652_a(ICommandSender p_179652_0_, String p_179652_1_, Class <? extends T > p_179652_2_) throws CommandException
     {
         List<T> list = func_179656_b(p_179652_0_, p_179652_1_, p_179652_2_);
-        return (T)(list.size() == 1 ? list.get(0) : null);
+        return (T)(list.size() == 1 ? (Entity)list.get(0) : null);
     }
 
     @Nullable
@@ -107,7 +108,7 @@
         }
         else
         {
-            List<ITextComponent> list1 = Lists.newArrayList();
+            List<ITextComponent> list1 = Lists.<ITextComponent>newArrayList();
 
             for (Entity entity : list)
             {
@@ -120,6 +121,11 @@
 
     public static <T extends Entity> List<T> func_179656_b(ICommandSender p_179656_0_, String p_179656_1_, Class <? extends T > p_179656_2_) throws CommandException
     {
+        return net.minecraftforge.common.command.SelectorHandlerManager.matchEntities(p_179656_0_, p_179656_1_, p_179656_2_);
+    }
+
+    public static <T extends Entity> List<T> matchEntitiesDefault(ICommandSender p_179656_0_, String p_179656_1_, Class <? extends T > p_179656_2_) throws CommandException
+    {
         Matcher matcher = field_82389_a.matcher(p_179656_1_);
 
         if (matcher.matches() && p_179656_0_.func_70003_b(1, "@"))
@@ -128,7 +134,7 @@
 
             if (!func_179655_b(p_179656_0_, map))
             {
-                return Collections.emptyList();
+                return Collections.<T>emptyList();
             }
             else
             {
@@ -136,13 +142,13 @@
                 BlockPos blockpos = func_179664_b(map, p_179656_0_.func_180425_c());
                 Vec3d vec3d = func_189210_b(map, p_179656_0_.func_174791_d());
                 List<World> list = func_179654_a(p_179656_0_, map);
-                List<T> list1 = Lists.newArrayList();
+                List<T> list1 = Lists.<T>newArrayList();
 
                 for (World world : list)
                 {
                     if (world != null)
                     {
-                        List<Predicate<Entity>> list2 = Lists.newArrayList();
+                        List<Predicate<Entity>> list2 = Lists.<Predicate<Entity>>newArrayList();
                         list2.addAll(func_179663_a(map, s));
                         list2.addAll(func_179648_b(map));
                         list2.addAll(func_179649_c(map));
@@ -152,6 +158,7 @@
                         list2.addAll(func_184951_f(map));
                         list2.addAll(func_180698_a(map, vec3d));
                         list2.addAll(func_179662_g(map));
+                        list2.addAll(net.minecraftforge.fml.common.registry.GameRegistry.createEntitySelectors(map, s, p_179656_0_, vec3d));
 
                         if ("s".equalsIgnoreCase(s))
                         {
@@ -168,7 +175,7 @@
 
                                     if (!axisalignedbb.func_72326_a(entity.func_174813_aQ()))
                                     {
-                                        return Collections.emptyList();
+                                        return Collections.<T>emptyList();
                                     }
                                 }
 
@@ -176,14 +183,14 @@
                                 {
                                     if (!predicate.apply(entity))
                                     {
-                                        return Collections.emptyList();
+                                        return Collections.<T>emptyList();
                                     }
                                 }
 
-                                return Lists.newArrayList((T)(entity));
+                                return Lists.newArrayList((T)entity);
                             }
 
-                            return Collections.emptyList();
+                            return Collections.<T>emptyList();
                         }
 
                         list1.addAll(func_179660_a(map, p_179656_2_, list2, s, world, blockpos));
@@ -195,13 +202,13 @@
         }
         else
         {
-            return Collections.emptyList();
+            return Collections.<T>emptyList();
         }
     }
 
     private static List<World> func_179654_a(ICommandSender p_179654_0_, Map<String, String> p_179654_1_)
     {
-        List<World> list = Lists.newArrayList();
+        List<World> list = Lists.<World>newArrayList();
 
         if (func_179665_h(p_179654_1_))
         {
@@ -233,7 +240,7 @@
             }
             else
             {
-                TextComponentTranslation textcomponenttranslation = new TextComponentTranslation("commands.generic.entity.invalidType", resourcelocation);
+                TextComponentTranslation textcomponenttranslation = new TextComponentTranslation("commands.generic.entity.invalidType", new Object[] {resourcelocation});
                 textcomponenttranslation.func_150256_b().func_150238_a(TextFormatting.RED);
                 p_179655_0_.func_145747_a(textcomponenttranslation);
                 return false;
@@ -271,7 +278,7 @@
 
     private static List<Predicate<Entity>> func_179648_b(Map<String, String> p_179648_0_)
     {
-        List<Predicate<Entity>> list = Lists.newArrayList();
+        List<Predicate<Entity>> list = Lists.<Predicate<Entity>>newArrayList();
         final int i = func_179653_a(p_179648_0_, field_190834_h, -1);
         final int j = func_179653_a(p_179648_0_, field_190833_g, -1);
 
@@ -299,7 +306,7 @@
 
     private static List<Predicate<Entity>> func_179649_c(Map<String, String> p_179649_0_)
     {
-        List<Predicate<Entity>> list = Lists.newArrayList();
+        List<Predicate<Entity>> list = Lists.<Predicate<Entity>>newArrayList();
         String s = func_179651_b(p_179649_0_, field_190846_t);
 
         if (s == null)
@@ -322,12 +329,12 @@
                 int i = Integer.parseInt(s);
                 gametype = GameType.func_185329_a(i, GameType.NOT_SET);
             }
-            catch (Throwable throwable)
+            catch (Throwable var6)
             {
                 gametype = GameType.func_185328_a(s, GameType.NOT_SET);
             }
 
-            final GameType gametype1 = gametype;
+            final GameType type = gametype;
             list.add(new Predicate<Entity>()
             {
                 public boolean apply(@Nullable Entity p_apply_1_)
@@ -339,8 +346,8 @@
                     else
                     {
                         EntityPlayerMP entityplayermp = (EntityPlayerMP)p_apply_1_;
-                        GameType gametype2 = entityplayermp.field_71134_c.func_73081_b();
-                        return flag ? gametype2 != gametype1 : gametype2 == gametype1;
+                        GameType gametype1 = entityplayermp.field_71134_c.func_73081_b();
+                        return flag ? gametype1 != type : gametype1 == type;
                     }
                 }
             });
@@ -350,7 +357,7 @@
 
     private static List<Predicate<Entity>> func_179659_d(Map<String, String> p_179659_0_)
     {
-        List<Predicate<Entity>> list = Lists.newArrayList();
+        List<Predicate<Entity>> list = Lists.<Predicate<Entity>>newArrayList();
         String s = func_179651_b(p_179659_0_, field_190847_u);
         final boolean flag = s != null && s.startsWith("!");
 
@@ -361,7 +368,7 @@
 
         if (s != null)
         {
-            final String s2 = s;
+            final String s_f_ = s;
             list.add(new Predicate<Entity>()
             {
                 public boolean apply(@Nullable Entity p_apply_1_)
@@ -375,7 +382,7 @@
                         EntityLivingBase entitylivingbase = (EntityLivingBase)p_apply_1_;
                         Team team = entitylivingbase.func_96124_cp();
                         String s1 = team == null ? "" : team.func_96661_b();
-                        return s1.equals(s2) != flag;
+                        return s1.equals(s_f_) != flag;
                     }
                 }
             });
@@ -399,7 +406,7 @@
                 {
                     Scoreboard scoreboard = p_184952_0_.func_184102_h().func_71218_a(0).func_96441_U();
 
-                    for (Map.Entry<String, Integer> entry : map.entrySet())
+                    for (Entry<String, Integer> entry : map.entrySet())
                     {
                         String s = entry.getKey();
                         boolean flag = false;
@@ -427,12 +434,12 @@
                         Score score = scoreboard.func_96529_a(s1, scoreobjective);
                         int i = score.func_96652_c();
 
-                        if (i < entry.getValue() && flag)
+                        if (i < ((Integer)entry.getValue()).intValue() && flag)
                         {
                             return false;
                         }
 
-                        if (i > entry.getValue() && !flag)
+                        if (i > ((Integer)entry.getValue()).intValue() && !flag)
                         {
                             return false;
                         }
@@ -446,7 +453,7 @@
 
     private static List<Predicate<Entity>> func_179647_f(Map<String, String> p_179647_0_)
     {
-        List<Predicate<Entity>> list = Lists.newArrayList();
+        List<Predicate<Entity>> list = Lists.<Predicate<Entity>>newArrayList();
         String s = func_179651_b(p_179647_0_, field_190848_v);
         final boolean flag = s != null && s.startsWith("!");
 
@@ -457,12 +464,12 @@
 
         if (s != null)
         {
-            final String s1 = s;
+            final String s_f_ = s;
             list.add(new Predicate<Entity>()
             {
                 public boolean apply(@Nullable Entity p_apply_1_)
                 {
-                    return p_apply_1_ != null && p_apply_1_.func_70005_c_().equals(s1) != flag;
+                    return p_apply_1_ != null && p_apply_1_.func_70005_c_().equals(s_f_) != flag;
                 }
             });
         }
@@ -472,7 +479,7 @@
 
     private static List<Predicate<Entity>> func_184951_f(Map<String, String> p_184951_0_)
     {
-        List<Predicate<Entity>> list = Lists.newArrayList();
+        List<Predicate<Entity>> list = Lists.<Predicate<Entity>>newArrayList();
         String s = func_179651_b(p_184951_0_, field_190850_x);
         final boolean flag = s != null && s.startsWith("!");
 
@@ -483,7 +490,7 @@
 
         if (s != null)
         {
-            final String s1 = s;
+            final String s_f_ = s;
             list.add(new Predicate<Entity>()
             {
                 public boolean apply(@Nullable Entity p_apply_1_)
@@ -492,13 +499,13 @@
                     {
                         return false;
                     }
-                    else if ("".equals(s1))
+                    else if ("".equals(s_f_))
                     {
                         return p_apply_1_.func_184216_O().isEmpty() != flag;
                     }
                     else
                     {
-                        return p_apply_1_.func_184216_O().contains(s1) != flag;
+                        return p_apply_1_.func_184216_O().contains(s_f_) != flag;
                     }
                 }
             });
@@ -516,7 +523,7 @@
 
         if (flag && flag1)
         {
-            return Collections.emptyList();
+            return Collections.<Predicate<Entity>>emptyList();
         }
         else
         {
@@ -544,7 +551,7 @@
 
     private static List<Predicate<Entity>> func_179662_g(Map<String, String> p_179662_0_)
     {
-        List<Predicate<Entity>> list = Lists.newArrayList();
+        List<Predicate<Entity>> list = Lists.<Predicate<Entity>>newArrayList();
 
         if (p_179662_0_.containsKey(field_190844_r) || p_179662_0_.containsKey(field_190843_q))
         {
@@ -609,7 +616,7 @@
 
     private static <T extends Entity> List<T> func_179660_a(Map<String, String> p_179660_0_, Class <? extends T > p_179660_1_, List<Predicate<Entity>> p_179660_2_, String p_179660_3_, World p_179660_4_, BlockPos p_179660_5_)
     {
-        List<T> list = Lists.newArrayList();
+        List<T> list = Lists.<T>newArrayList();
         String s = func_179651_b(p_179660_0_, field_190849_w);
         s = s != null && s.startsWith("!") ? s.substring(1) : s;
         boolean flag = !p_179660_3_.equals("e");
@@ -619,7 +626,7 @@
         int k = func_179653_a(p_179660_0_, field_190840_n, 0);
         int l = func_179653_a(p_179660_0_, field_190831_e, -1);
         Predicate<Entity> predicate = Predicates.and(p_179660_2_);
-        Predicate<Entity> predicate1 = Predicates.and(EntitySelectors.field_94557_a, predicate);
+        Predicate<Entity> predicate1 = Predicates.<Entity>and(EntitySelectors.field_94557_a, predicate);
 
         if (!p_179660_0_.containsKey(field_190838_l) && !p_179660_0_.containsKey(field_190839_m) && !p_179660_0_.containsKey(field_190840_n))
         {
@@ -699,7 +706,7 @@
 
         if (entity != null && p_179658_3_.isAssignableFrom(entity.getClass()) && i == 1 && p_179658_0_.contains(entity) && !"r".equals(p_179658_4_))
         {
-            p_179658_0_ = Lists.newArrayList((T)(entity));
+            p_179658_0_ = Lists.newArrayList((T)entity);
         }
 
         if (i != 0)
@@ -770,13 +777,13 @@
 
     public static Map<String, Integer> func_96560_a(Map<String, String> p_96560_0_)
     {
-        Map<String, Integer> map = Maps.newHashMap();
+        Map<String, Integer> map = Maps.<String, Integer>newHashMap();
 
         for (String s : p_96560_0_.keySet())
         {
             if (s.startsWith("score_") && s.length() > "score_".length())
             {
-                map.put(s.substring("score_".length()), MathHelper.func_82715_a(p_96560_0_.get(s), 1));
+                map.put(s.substring("score_".length()), Integer.valueOf(MathHelper.func_82715_a(p_96560_0_.get(s), 1)));
             }
         }
 
@@ -785,6 +792,11 @@
 
     public static boolean func_82377_a(String p_82377_0_) throws CommandException
     {
+        return net.minecraftforge.common.command.SelectorHandlerManager.matchesMultiplePlayers(p_82377_0_);
+    }
+
+    public static boolean matchesMultiplePlayersDefault(String p_82377_0_) throws CommandException
+    {
         Matcher matcher = field_82389_a.matcher(p_82377_0_);
 
         if (!matcher.matches())
@@ -802,12 +814,17 @@
 
     public static boolean func_82378_b(String p_82378_0_)
     {
+        return net.minecraftforge.common.command.SelectorHandlerManager.isSelector(p_82378_0_);
+    }
+
+    public static boolean isSelectorDefault(String p_82378_0_)
+    {
         return field_82389_a.matcher(p_82378_0_).matches();
     }
 
     private static Map<String, String> func_82381_h(@Nullable String p_82381_0_) throws CommandException
     {
-        Map<String, String> map = Maps.newHashMap();
+        Map<String, String> map = Maps.<String, String>newHashMap();
 
         if (p_82381_0_ == null)
         {
@@ -822,10 +839,10 @@
 
                 if (!field_190851_y.apply(s1))
                 {
-                    throw new CommandException("commands.generic.selector_argument", s);
+                    throw new CommandException("commands.generic.selector_argument", new Object[] {s});
                 }
 
-                map.put(s1, iterator.hasNext() ? iterator.next() : "");
+                map.put(s1, iterator.hasNext() ? (String)iterator.next() : "");
             }
 
             return map;
