--- before/net/minecraft/command/ServerCommandManager.java
+++ after/net/minecraft/command/ServerCommandManager.java
@@ -116,7 +116,6 @@
         CommandBase.setCommandListener(this);
     }
 
-    @Override
     public void notifyListener(ICommandSender sender, ICommand command, int flags, String translationKey, Object... translationArgs)
     {
         boolean flag = true;
@@ -127,19 +126,15 @@
             flag = false;
         }
 
-        ITextComponent itextcomponent = new TextComponentTranslation(
-            "chat.type.admin", sender.getName(), new TextComponentTranslation(translationKey, translationArgs)
-        );
+        ITextComponent itextcomponent = new TextComponentTranslation("chat.type.admin", new Object[] {sender.getName(), new TextComponentTranslation(translationKey, translationArgs)});
         itextcomponent.getStyle().setColor(TextFormatting.GRAY);
-        itextcomponent.getStyle().setItalic(true);
+        itextcomponent.getStyle().setItalic(Boolean.valueOf(true));
 
         if (flag)
         {
             for (EntityPlayer entityplayer : minecraftserver.getPlayerList().getPlayers())
             {
-                if (entityplayer != sender
-                        && minecraftserver.getPlayerList().canSendCommands(entityplayer.getGameProfile())
-                        && command.checkPermission(this.server, sender))
+                if (entityplayer != sender && minecraftserver.getPlayerList().canSendCommands(entityplayer.getGameProfile()) && command.checkPermission(this.server, sender))
                 {
                     boolean flag1 = sender instanceof MinecraftServer && this.server.shouldBroadcastConsoleToOps();
                     boolean flag2 = sender instanceof RConConsoleSource && this.server.shouldBroadcastRconToOps();
@@ -170,7 +165,6 @@
         }
     }
 
-    @Override
     protected MinecraftServer getServer()
     {
         return this.server;
