--- original/net/minecraft/world/storage/loot/conditions/RandomChanceWithLooting.java
+++ modified/net/minecraft/world/storage/loot/conditions/RandomChanceWithLooting.java
@@ -21,10 +21,7 @@
 
     @Override
     public boolean testCondition(Random rand, LootContext context) {
-        int i = 0;
-        if (context.getKiller() instanceof EntityLivingBase) {
-            i = EnchantmentHelper.getLootingModifier((EntityLivingBase)context.getKiller());
-        }
+        int i = context.getLootingModifier();
 
         return rand.nextFloat() < this.chance + i * this.lootingMultiplier;
     }
