--- original/net/minecraft/entity/projectile/EntitySmallFireball.java
+++ modified/net/minecraft/entity/projectile/EntitySmallFireball.java
@@ -46,7 +46,7 @@
             } else {
                 boolean flag1 = true;
                 if (this.shootingEntity != null && this.shootingEntity instanceof EntityLiving) {
-                    flag1 = this.world.getGameRules().getBoolean("mobGriefing");
+                    flag1 = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.shootingEntity);
                 }
 
                 if (flag1) {
