--- original/net/minecraft/entity/projectile/EntityWitherSkull.java
+++ modified/net/minecraft/entity/projectile/EntityWitherSkull.java
@@ -61,7 +61,8 @@
     public float getExplosionResistance(Explosion explosionIn, World worldIn, BlockPos pos, IBlockState blockStateIn) {
         float f = super.getExplosionResistance(explosionIn, worldIn, pos, blockStateIn);
         Block block = blockStateIn.getBlock();
-        if (this.isInvulnerable() && EntityWither.canDestroyBlock(block)) {
+
+        if (this.isInvulnerable() && block.canEntityDestroy(blockStateIn, worldIn, pos, this) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this.shootingEntity, pos, blockStateIn)) {
             f = Math.min(0.8F, f);
         }
 
@@ -107,7 +108,7 @@
                     this.posZ,
                     1.0F,
                     false,
-                    this.world.getGameRules().getBoolean("mobGriefing")
+                    net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.shootingEntity)
                 );
             this.setDead();
         }
