--- before/net/minecraft/client/gui/recipebook/GhostRecipe.java
+++ after/net/minecraft/client/gui/recipebook/GhostRecipe.java
@@ -20,7 +20,7 @@
 public class GhostRecipe
 {
     private IRecipe field_192687_a;
-    private final List<GhostRecipe.GhostIngredient> field_192688_b = Lists.newArrayList();
+    private final List<GhostRecipe.GhostIngredient> field_192688_b = Lists.<GhostRecipe.GhostIngredient>newArrayList();
     private float field_194190_c;
 
     public void func_192682_a()
