--- before/net/minecraft/client/gui/recipebook/GuiButtonRecipe.java
+++ after/net/minecraft/client/gui/recipebook/GuiButtonRecipe.java
@@ -100,7 +100,7 @@
             this.func_73729_b(this.field_146128_h, this.field_146129_i, i, j, this.field_146120_f, this.field_146121_g);
             List<IRecipe> list = this.func_193927_f();
             this.field_193932_t = MathHelper.func_76141_d(this.field_193931_r / 30.0F) % list.size();
-            ItemStack itemstack = list.get(this.field_193932_t).func_77571_b();
+            ItemStack itemstack = ((IRecipe)list.get(this.field_193932_t)).func_77571_b();
             int k = 4;
 
             if (this.field_191774_p.func_194211_e() && this.func_193927_f().size() > 1)
@@ -146,7 +146,7 @@
 
     public List<String> func_191772_a(GuiScreen p_191772_1_)
     {
-        ItemStack itemstack = this.func_193927_f().get(this.field_193932_t).func_77571_b();
+        ItemStack itemstack = ((IRecipe)this.func_193927_f().get(this.field_193932_t)).func_77571_b();
         List<String> list = p_191772_1_.func_191927_a(itemstack);
 
         if (this.field_191774_p.func_194208_a(this.field_193930_p.func_192815_c()).size() > 1)
