--- before/net/minecraft/client/gui/recipebook/GuiButtonRecipeTab.java
+++ after/net/minecraft/client/gui/recipebook/GuiButtonRecipeTab.java
@@ -1,5 +1,6 @@
 package net.minecraft.client.gui.recipebook;
 
+import java.util.Iterator;
 import java.util.List;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.GuiButtonToggle;
@@ -30,17 +31,29 @@
     public void func_193918_a(Minecraft p_193918_1_)
     {
         RecipeBook recipebook = p_193918_1_.field_71439_g.func_192035_E();
+        label21:
 
         for (RecipeList recipelist : RecipeBookClient.field_194086_e.get(this.field_193921_u))
         {
-            for (IRecipe irecipe : recipelist.func_194208_a(recipebook.func_192815_c()))
+            Iterator iterator = recipelist.func_194208_a(recipebook.func_192815_c()).iterator();
+
+            while (true)
             {
+                if (!iterator.hasNext())
+                {
+                    continue label21;
+                }
+
+                IRecipe irecipe = (IRecipe)iterator.next();
+
                 if (recipebook.func_194076_e(irecipe))
                 {
-                    this.field_193922_v = 15.0F;
-                    return;
+                    break;
                 }
             }
+
+            this.field_193922_v = 15.0F;
+            return;
         }
     }
 
@@ -124,7 +137,7 @@
 
     public boolean func_193919_e()
     {
-        List<RecipeList> list = RecipeBookClient.field_194086_e.get(this.field_193921_u);
+        List<RecipeList> list = (List)RecipeBookClient.field_194086_e.get(this.field_193921_u);
         this.field_146125_m = false;
 
         for (RecipeList recipelist : list)
