--- before/net/minecraft/client/gui/recipebook/GuiRecipeOverlay.java
+++ after/net/minecraft/client/gui/recipebook/GuiRecipeOverlay.java
@@ -23,7 +23,7 @@
 public class GuiRecipeOverlay extends Gui
 {
     private static final ResourceLocation RECIPE_BOOK_TEXTURE = new ResourceLocation("textures/gui/recipe_book.png");
-    private final List<GuiRecipeOverlay.Button> buttonList = Lists.newArrayList();
+    private final List<GuiRecipeOverlay.Button> buttonList = Lists.<GuiRecipeOverlay.Button>newArrayList();
     private boolean visible;
     private int x;
     private int y;
@@ -32,16 +32,7 @@
     private IRecipe lastRecipeClicked;
     private float time;
 
-    public void init(
-        Minecraft mcIn,
-        RecipeList recipeListIn,
-        int p_191845_3_,
-        int p_191845_4_,
-        int p_191845_5_,
-        int p_191845_6_,
-        float p_191845_7_,
-        RecipeBook p_191845_8_
-    )
+    public void init(Minecraft mcIn, RecipeList recipeListIn, int p_191845_3_, int p_191845_4_, int p_191845_5_, int p_191845_6_, float p_191845_7_, RecipeBook p_191845_8_)
     {
         this.mc = mcIn;
         this.recipeList = recipeListIn;
@@ -82,15 +73,10 @@
         this.visible = true;
         this.buttonList.clear();
 
-        for (int j1 = 0; j1 < j; j1++)
+        for (int j1 = 0; j1 < j; ++j1)
         {
             boolean flag1 = j1 < i;
-            this.buttonList
-            .add(
-                new GuiRecipeOverlay.Button(
-                    this.x + 4 + 25 * (j1 % k), this.y + 5 + 25 * (j1 / k), flag1 ? list.get(j1) : list1.get(j1 - i), flag1
-                )
-            );
+            this.buttonList.add(new GuiRecipeOverlay.Button(this.x + 4 + 25 * (j1 % k), this.y + 5 + 25 * (j1 / k), flag1 ? (IRecipe)list.get(j1) : (IRecipe)list1.get(j1 - i), flag1));
         }
 
         this.lastRecipeClicked = null;
@@ -161,127 +147,37 @@
     private void nineInchSprite(int p_191846_1_, int p_191846_2_, int p_191846_3_, int p_191846_4_, int p_191846_5_, int p_191846_6_)
     {
         this.drawTexturedModalRect(this.x, this.y, p_191846_5_, p_191846_6_, p_191846_4_, p_191846_4_);
-        this.drawTexturedModalRect(
-            this.x + p_191846_4_ * 2 + p_191846_1_ * p_191846_3_,
-            this.y,
-            p_191846_5_ + p_191846_3_ + p_191846_4_,
-            p_191846_6_,
-            p_191846_4_,
-            p_191846_4_
-        );
-        this.drawTexturedModalRect(
-            this.x,
-            this.y + p_191846_4_ * 2 + p_191846_2_ * p_191846_3_,
-            p_191846_5_,
-            p_191846_6_ + p_191846_3_ + p_191846_4_,
-            p_191846_4_,
-            p_191846_4_
-        );
-        this.drawTexturedModalRect(
-            this.x + p_191846_4_ * 2 + p_191846_1_ * p_191846_3_,
-            this.y + p_191846_4_ * 2 + p_191846_2_ * p_191846_3_,
-            p_191846_5_ + p_191846_3_ + p_191846_4_,
-            p_191846_6_ + p_191846_3_ + p_191846_4_,
-            p_191846_4_,
-            p_191846_4_
-        );
+        this.drawTexturedModalRect(this.x + p_191846_4_ * 2 + p_191846_1_ * p_191846_3_, this.y, p_191846_5_ + p_191846_3_ + p_191846_4_, p_191846_6_, p_191846_4_, p_191846_4_);
+        this.drawTexturedModalRect(this.x, this.y + p_191846_4_ * 2 + p_191846_2_ * p_191846_3_, p_191846_5_, p_191846_6_ + p_191846_3_ + p_191846_4_, p_191846_4_, p_191846_4_);
+        this.drawTexturedModalRect(this.x + p_191846_4_ * 2 + p_191846_1_ * p_191846_3_, this.y + p_191846_4_ * 2 + p_191846_2_ * p_191846_3_, p_191846_5_ + p_191846_3_ + p_191846_4_, p_191846_6_ + p_191846_3_ + p_191846_4_, p_191846_4_, p_191846_4_);
 
-        for (int i = 0; i < p_191846_1_; i++)
+        for (int i = 0; i < p_191846_1_; ++i)
         {
-            this.drawTexturedModalRect(
-                this.x + p_191846_4_ + i * p_191846_3_, this.y, p_191846_5_ + p_191846_4_, p_191846_6_, p_191846_3_, p_191846_4_
-            );
-            this.drawTexturedModalRect(
-                this.x + p_191846_4_ + (i + 1) * p_191846_3_, this.y, p_191846_5_ + p_191846_4_, p_191846_6_, p_191846_4_, p_191846_4_
-            );
+            this.drawTexturedModalRect(this.x + p_191846_4_ + i * p_191846_3_, this.y, p_191846_5_ + p_191846_4_, p_191846_6_, p_191846_3_, p_191846_4_);
+            this.drawTexturedModalRect(this.x + p_191846_4_ + (i + 1) * p_191846_3_, this.y, p_191846_5_ + p_191846_4_, p_191846_6_, p_191846_4_, p_191846_4_);
 
-            for (int j = 0; j < p_191846_2_; j++)
+            for (int j = 0; j < p_191846_2_; ++j)
             {
                 if (i == 0)
                 {
-                    this.drawTexturedModalRect(
-                        this.x, this.y + p_191846_4_ + j * p_191846_3_, p_191846_5_, p_191846_6_ + p_191846_4_, p_191846_4_, p_191846_3_
-                    );
-                    this.drawTexturedModalRect(
-                        this.x,
-                        this.y + p_191846_4_ + (j + 1) * p_191846_3_,
-                        p_191846_5_,
-                        p_191846_6_ + p_191846_4_,
-                        p_191846_4_,
-                        p_191846_4_
-                    );
+                    this.drawTexturedModalRect(this.x, this.y + p_191846_4_ + j * p_191846_3_, p_191846_5_, p_191846_6_ + p_191846_4_, p_191846_4_, p_191846_3_);
+                    this.drawTexturedModalRect(this.x, this.y + p_191846_4_ + (j + 1) * p_191846_3_, p_191846_5_, p_191846_6_ + p_191846_4_, p_191846_4_, p_191846_4_);
                 }
 
-                this.drawTexturedModalRect(
-                    this.x + p_191846_4_ + i * p_191846_3_,
-                    this.y + p_191846_4_ + j * p_191846_3_,
-                    p_191846_5_ + p_191846_4_,
-                    p_191846_6_ + p_191846_4_,
-                    p_191846_3_,
-                    p_191846_3_
-                );
-                this.drawTexturedModalRect(
-                    this.x + p_191846_4_ + (i + 1) * p_191846_3_,
-                    this.y + p_191846_4_ + j * p_191846_3_,
-                    p_191846_5_ + p_191846_4_,
-                    p_191846_6_ + p_191846_4_,
-                    p_191846_4_,
-                    p_191846_3_
-                );
-                this.drawTexturedModalRect(
-                    this.x + p_191846_4_ + i * p_191846_3_,
-                    this.y + p_191846_4_ + (j + 1) * p_191846_3_,
-                    p_191846_5_ + p_191846_4_,
-                    p_191846_6_ + p_191846_4_,
-                    p_191846_3_,
-                    p_191846_4_
-                );
-                this.drawTexturedModalRect(
-                    this.x + p_191846_4_ + (i + 1) * p_191846_3_ - 1,
-                    this.y + p_191846_4_ + (j + 1) * p_191846_3_ - 1,
-                    p_191846_5_ + p_191846_4_,
-                    p_191846_6_ + p_191846_4_,
-                    p_191846_4_ + 1,
-                    p_191846_4_ + 1
-                );
+                this.drawTexturedModalRect(this.x + p_191846_4_ + i * p_191846_3_, this.y + p_191846_4_ + j * p_191846_3_, p_191846_5_ + p_191846_4_, p_191846_6_ + p_191846_4_, p_191846_3_, p_191846_3_);
+                this.drawTexturedModalRect(this.x + p_191846_4_ + (i + 1) * p_191846_3_, this.y + p_191846_4_ + j * p_191846_3_, p_191846_5_ + p_191846_4_, p_191846_6_ + p_191846_4_, p_191846_4_, p_191846_3_);
+                this.drawTexturedModalRect(this.x + p_191846_4_ + i * p_191846_3_, this.y + p_191846_4_ + (j + 1) * p_191846_3_, p_191846_5_ + p_191846_4_, p_191846_6_ + p_191846_4_, p_191846_3_, p_191846_4_);
+                this.drawTexturedModalRect(this.x + p_191846_4_ + (i + 1) * p_191846_3_ - 1, this.y + p_191846_4_ + (j + 1) * p_191846_3_ - 1, p_191846_5_ + p_191846_4_, p_191846_6_ + p_191846_4_, p_191846_4_ + 1, p_191846_4_ + 1);
 
                 if (i == p_191846_1_ - 1)
                 {
-                    this.drawTexturedModalRect(
-                        this.x + p_191846_4_ * 2 + p_191846_1_ * p_191846_3_,
-                        this.y + p_191846_4_ + j * p_191846_3_,
-                        p_191846_5_ + p_191846_3_ + p_191846_4_,
-                        p_191846_6_ + p_191846_4_,
-                        p_191846_4_,
-                        p_191846_3_
-                    );
-                    this.drawTexturedModalRect(
-                        this.x + p_191846_4_ * 2 + p_191846_1_ * p_191846_3_,
-                        this.y + p_191846_4_ + (j + 1) * p_191846_3_,
-                        p_191846_5_ + p_191846_3_ + p_191846_4_,
-                        p_191846_6_ + p_191846_4_,
-                        p_191846_4_,
-                        p_191846_4_
-                    );
+                    this.drawTexturedModalRect(this.x + p_191846_4_ * 2 + p_191846_1_ * p_191846_3_, this.y + p_191846_4_ + j * p_191846_3_, p_191846_5_ + p_191846_3_ + p_191846_4_, p_191846_6_ + p_191846_4_, p_191846_4_, p_191846_3_);
+                    this.drawTexturedModalRect(this.x + p_191846_4_ * 2 + p_191846_1_ * p_191846_3_, this.y + p_191846_4_ + (j + 1) * p_191846_3_, p_191846_5_ + p_191846_3_ + p_191846_4_, p_191846_6_ + p_191846_4_, p_191846_4_, p_191846_4_);
                 }
             }
 
-            this.drawTexturedModalRect(
-                this.x + p_191846_4_ + i * p_191846_3_,
-                this.y + p_191846_4_ * 2 + p_191846_2_ * p_191846_3_,
-                p_191846_5_ + p_191846_4_,
-                p_191846_6_ + p_191846_3_ + p_191846_4_,
-                p_191846_3_,
-                p_191846_4_
-            );
-            this.drawTexturedModalRect(
-                this.x + p_191846_4_ + (i + 1) * p_191846_3_,
-                this.y + p_191846_4_ * 2 + p_191846_2_ * p_191846_3_,
-                p_191846_5_ + p_191846_4_,
-                p_191846_6_ + p_191846_3_ + p_191846_4_,
-                p_191846_4_,
-                p_191846_4_
-            );
+            this.drawTexturedModalRect(this.x + p_191846_4_ + i * p_191846_3_, this.y + p_191846_4_ * 2 + p_191846_2_ * p_191846_3_, p_191846_5_ + p_191846_4_, p_191846_6_ + p_191846_3_ + p_191846_4_, p_191846_3_, p_191846_4_);
+            this.drawTexturedModalRect(this.x + p_191846_4_ + (i + 1) * p_191846_3_, this.y + p_191846_4_ * 2 + p_191846_2_ * p_191846_3_, p_191846_5_ + p_191846_4_, p_191846_6_ + p_191846_3_ + p_191846_4_, p_191846_4_, p_191846_4_);
         }
     }
 
@@ -310,16 +206,12 @@
             this.isCraftable = p_i47594_5_;
         }
 
-        @Override
         public void drawButton(Minecraft mc, int mouseX, int mouseY, float partialTicks)
         {
             RenderHelper.enableGUIStandardItemLighting();
             GlStateManager.enableAlpha();
             mc.getTextureManager().bindTexture(GuiRecipeOverlay.RECIPE_BOOK_TEXTURE);
-            this.hovered = mouseX >= this.x
-                                  && mouseY >= this.y
-                                  && mouseX < this.x + this.width
-                                  && mouseY < this.y + this.height;
+            this.hovered = mouseX >= this.x && mouseY >= this.y && mouseX < this.x + this.width && mouseY < this.y + this.height;
             int i = 152;
 
             if (!this.isCraftable)
@@ -338,24 +230,24 @@
             int k = 3;
             int l = 3;
 
-            if (this.recipe instanceof ShapedRecipes)
+            if (this.recipe instanceof net.minecraftforge.common.crafting.IShapedRecipe)
             {
-                ShapedRecipes shapedrecipes = (ShapedRecipes)this.recipe;
-                k = shapedrecipes.getWidth();
-                l = shapedrecipes.getHeight();
+                net.minecraftforge.common.crafting.IShapedRecipe shapedrecipes = (net.minecraftforge.common.crafting.IShapedRecipe)this.recipe;
+                k = shapedrecipes.getRecipeWidth();
+                l = shapedrecipes.getRecipeHeight();
             }
 
             Iterator<Ingredient> iterator = this.recipe.getIngredients().iterator();
 
-            for (int i1 = 0; i1 < l; i1++)
+            for (int i1 = 0; i1 < l; ++i1)
             {
                 int j1 = 3 + i1 * 7;
 
-                for (int k1 = 0; k1 < k; k1++)
+                for (int k1 = 0; k1 < k; ++k1)
                 {
                     if (iterator.hasNext())
                     {
-                        ItemStack[] aitemstack = iterator.next().getMatchingStacks();
+                        ItemStack[] aitemstack = ((Ingredient)iterator.next()).getMatchingStacks();
 
                         if (aitemstack.length != 0)
                         {
@@ -366,8 +258,7 @@
                             int j2 = (int)((float)(this.y + j1) / 0.42F - 3.0F);
                             GlStateManager.scale(0.42F, 0.42F, 1.0F);
                             GlStateManager.enableLighting();
-                            mc.getRenderItem()
-                            .renderItemAndEffectIntoGUI(aitemstack[MathHelper.floor(GuiRecipeOverlay.this.time / 30.0F) % aitemstack.length], i2, j2);
+                            mc.getRenderItem().renderItemAndEffectIntoGUI(aitemstack[MathHelper.floor(GuiRecipeOverlay.this.time / 30.0F) % aitemstack.length], i2, j2);
                             GlStateManager.disableLighting();
                             GlStateManager.popMatrix();
                         }
