--- before/net/minecraft/client/gui/recipebook/RecipeList.java
+++ after/net/minecraft/client/gui/recipebook/RecipeList.java
@@ -14,7 +14,7 @@
 @SideOnly(Side.CLIENT)
 public class RecipeList
 {
-    private List<IRecipe> field_192713_b = Lists.newArrayList();
+    private List<IRecipe> field_192713_b = Lists.<IRecipe>newArrayList();
     private final BitSet field_194215_b = new BitSet();
     private final BitSet field_194216_c = new BitSet();
     private final BitSet field_194217_d = new BitSet();
@@ -66,7 +66,7 @@
 
     public List<IRecipe> func_194208_a(boolean p_194208_1_)
     {
-        List<IRecipe> list = Lists.newArrayList();
+        List<IRecipe> list = Lists.<IRecipe>newArrayList();
 
         for (int i = this.field_194217_d.nextSetBit(0); i >= 0; i = this.field_194217_d.nextSetBit(i + 1))
         {
@@ -81,7 +81,7 @@
 
     public List<IRecipe> func_194207_b(boolean p_194207_1_)
     {
-        List<IRecipe> list = Lists.newArrayList();
+        List<IRecipe> list = Lists.<IRecipe>newArrayList();
 
         for (int i = this.field_194217_d.nextSetBit(0); i >= 0; i = this.field_194217_d.nextSetBit(i + 1))
         {
@@ -100,7 +100,7 @@
 
         if (this.field_194218_e)
         {
-            ItemStack itemstack = this.field_192713_b.get(0).func_77571_b();
+            ItemStack itemstack = ((IRecipe)this.field_192713_b.get(0)).func_77571_b();
             ItemStack itemstack1 = p_192709_1_.func_77571_b();
             this.field_194218_e = ItemStack.func_179545_c(itemstack, itemstack1) && ItemStack.func_77970_a(itemstack, itemstack1);
         }
