--- before/net/minecraft/util/datafix/walkers/BlockEntityTag.java
+++ after/net/minecraft/util/datafix/walkers/BlockEntityTag.java
@@ -14,18 +14,15 @@
 public class BlockEntityTag implements IDataWalker
 {
     private static final Logger LOGGER = LogManager.getLogger();
-    private static final Map<String, String> NEW_TO_OLD_ID_MAP = Maps.newHashMap();
-    private static final Map<String, String> ITEM_ID_TO_BLOCK_ENTITY_ID = Maps.newHashMap();
+    private static final Map<String, String> NEW_TO_OLD_ID_MAP = Maps.<String, String>newHashMap();
+    private static final Map<String, String> ITEM_ID_TO_BLOCK_ENTITY_ID = Maps.<String, String>newHashMap();
 
     @Nullable
     private static String getBlockEntityID(int blockID, String p_188267_1_)
     {
-        return blockID < 515
-               ? NEW_TO_OLD_ID_MAP.get(new ResourceLocation(p_188267_1_).toString())
-               : ITEM_ID_TO_BLOCK_ENTITY_ID.get(new ResourceLocation(p_188267_1_).toString());
+        return blockID < 515 ? (String)NEW_TO_OLD_ID_MAP.get((new ResourceLocation(p_188267_1_)).toString()) : (String)ITEM_ID_TO_BLOCK_ENTITY_ID.get((new ResourceLocation(p_188267_1_)).toString());
     }
 
-    @Override
     public NBTTagCompound process(IDataFixer fixer, NBTTagCompound compound, int versionIn)
     {
         if (!compound.hasKey("tag", 10))
@@ -45,7 +42,7 @@
 
                 if (s1 == null)
                 {
-                    LOGGER.warn("Unable to resolve BlockEntity for ItemInstance: {}", s);
+                    LOGGER.warn("Unable to resolve BlockEntity for ItemInstance: {}", (Object)s);
                     flag = false;
                 }
                 else
