--- before/net/minecraft/client/renderer/entity/RenderMinecart.java
+++ after/net/minecraft/client/renderer/entity/RenderMinecart.java
@@ -39,14 +39,14 @@
         double d0 = p_76986_1_.field_70142_S + (p_76986_1_.field_70165_t - p_76986_1_.field_70142_S) * (double)p_76986_9_;
         double d1 = p_76986_1_.field_70137_T + (p_76986_1_.field_70163_u - p_76986_1_.field_70137_T) * (double)p_76986_9_;
         double d2 = p_76986_1_.field_70136_U + (p_76986_1_.field_70161_v - p_76986_1_.field_70136_U) * (double)p_76986_9_;
-        double d3 = (double)0.3F;
+        double d3 = 0.30000001192092896D;
         Vec3d vec3d = p_76986_1_.func_70489_a(d0, d1, d2);
         float f3 = p_76986_1_.field_70127_C + (p_76986_1_.field_70125_A - p_76986_1_.field_70127_C) * p_76986_9_;
 
         if (vec3d != null)
         {
-            Vec3d vec3d1 = p_76986_1_.func_70495_a(d0, d1, d2, (double)0.3F);
-            Vec3d vec3d2 = p_76986_1_.func_70495_a(d0, d1, d2, (double) - 0.3F);
+            Vec3d vec3d1 = p_76986_1_.func_70495_a(d0, d1, d2, 0.30000001192092896D);
+            Vec3d vec3d2 = p_76986_1_.func_70495_a(d0, d1, d2, -0.30000001192092896D);
 
             if (vec3d1 == null)
             {
