--- before/net/minecraft/client/renderer/entity/RenderManager.java
+++ after/net/minecraft/client/renderer/entity/RenderManager.java
@@ -106,6 +106,7 @@
 import net.minecraft.entity.projectile.EntityWitherSkull;
 import net.minecraft.init.Blocks;
 import net.minecraft.init.Items;
+import net.minecraft.util.EnumFacing;
 import net.minecraft.util.ReportedException;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
@@ -117,8 +118,8 @@
 @SideOnly(Side.CLIENT)
 public class RenderManager
 {
-    public final Map < Class <? extends Entity > , Render <? extends Entity >> field_78729_o = Maps.newHashMap();
-    private final Map<String, RenderPlayer> field_178636_l = Maps.newHashMap();
+    public final Map < Class <? extends Entity > , Render <? extends Entity >> field_78729_o = Maps. < Class <? extends Entity > , Render <? extends Entity >> newHashMap();
+    private final Map<String, RenderPlayer> field_178636_l = Maps.<String, RenderPlayer>newHashMap();
     private final RenderPlayer field_178637_m;
     private FontRenderer field_78736_p;
     private double field_78725_b;
@@ -191,13 +192,13 @@
         this.field_78729_o.put(EntityLeashKnot.class, new RenderLeashKnot(this));
         this.field_78729_o.put(EntityTippedArrow.class, new RenderTippedArrow(this));
         this.field_78729_o.put(EntitySpectralArrow.class, new RenderSpectralArrow(this));
-        this.field_78729_o.put(EntitySnowball.class, new RenderSnowball<>(this, Items.field_151126_ay, p_i46180_2_));
-        this.field_78729_o.put(EntityEnderPearl.class, new RenderSnowball<>(this, Items.field_151079_bi, p_i46180_2_));
-        this.field_78729_o.put(EntityEnderEye.class, new RenderSnowball<>(this, Items.field_151061_bv, p_i46180_2_));
-        this.field_78729_o.put(EntityEgg.class, new RenderSnowball<>(this, Items.field_151110_aK, p_i46180_2_));
+        this.field_78729_o.put(EntitySnowball.class, new RenderSnowball(this, Items.field_151126_ay, p_i46180_2_));
+        this.field_78729_o.put(EntityEnderPearl.class, new RenderSnowball(this, Items.field_151079_bi, p_i46180_2_));
+        this.field_78729_o.put(EntityEnderEye.class, new RenderSnowball(this, Items.field_151061_bv, p_i46180_2_));
+        this.field_78729_o.put(EntityEgg.class, new RenderSnowball(this, Items.field_151110_aK, p_i46180_2_));
         this.field_78729_o.put(EntityPotion.class, new RenderPotion(this, p_i46180_2_));
-        this.field_78729_o.put(EntityExpBottle.class, new RenderSnowball<>(this, Items.field_151062_by, p_i46180_2_));
-        this.field_78729_o.put(EntityFireworkRocket.class, new RenderSnowball<>(this, Items.field_151152_bP, p_i46180_2_));
+        this.field_78729_o.put(EntityExpBottle.class, new RenderSnowball(this, Items.field_151062_by, p_i46180_2_));
+        this.field_78729_o.put(EntityFireworkRocket.class, new RenderSnowball(this, Items.field_151152_bP, p_i46180_2_));
         this.field_78729_o.put(EntityLargeFireball.class, new RenderFireball(this, 2.0F));
         this.field_78729_o.put(EntitySmallFireball.class, new RenderFireball(this, 0.5F));
         this.field_78729_o.put(EntityDragonFireball.class, new RenderDragonFireball(this));
@@ -226,6 +227,11 @@
         this.field_178637_m = new RenderPlayer(this);
         this.field_178636_l.put("default", this.field_178637_m);
         this.field_178636_l.put("slim", new RenderPlayer(this, true));
+        net.minecraftforge.fml.client.registry.RenderingRegistry.loadEntityRenderers(this, this.field_78729_o);
+    }
+
+    public Map<String, RenderPlayer> getSkinMap() {
+        return (Map<String, RenderPlayer>) java.util.Collections.unmodifiableMap(field_178636_l);
     }
 
     public void func_178628_a(double p_178628_1_, double p_178628_3_, double p_178628_5_)
@@ -259,7 +265,7 @@
         }
         else
         {
-            return this.func_78715_a(p_78713_1_.getClass());
+            return this.<T>func_78715_a(p_78713_1_.getClass());
         }
     }
 
@@ -276,9 +282,9 @@
             IBlockState iblockstate = p_180597_1_.func_180495_p(new BlockPos(p_180597_3_));
             Block block = iblockstate.func_177230_c();
 
-            if (block == Blocks.field_150324_C)
+            if (block.isBed(iblockstate, p_180597_1_, new BlockPos(p_180597_3_), (EntityLivingBase)p_180597_3_))
             {
-                int i = iblockstate.func_177229_b(BlockBed.field_185512_D).func_176736_b();
+                int i = block.getBedDirection(iblockstate, p_180597_1_, new BlockPos(p_180597_3_)).func_176736_b();
                 this.field_78735_i = (float)(i * 90 + 180);
                 this.field_78732_j = 0.0F;
             }
@@ -331,7 +337,7 @@
 
     public boolean func_178635_a(Entity p_178635_1_, ICamera p_178635_2_, double p_178635_3_, double p_178635_5_, double p_178635_7_)
     {
-        Render<Entity> render = this.func_78713_a(p_178635_1_);
+        Render<Entity> render = this.<Entity>func_78713_a(p_178635_1_);
         return render != null && render.func_177071_a(p_178635_1_, p_178635_2_, p_178635_3_, p_178635_5_, p_178635_7_);
     }
 
@@ -368,7 +374,7 @@
 
         try
         {
-            render = this.func_78713_a(p_188391_1_);
+            render = this.<Entity>func_78713_a(p_188391_1_);
 
             if (render != null && this.field_78724_e != null)
             {
@@ -415,8 +421,8 @@
             CrashReportCategory crashreportcategory1 = crashreport.func_85058_a("Renderer details");
             crashreportcategory1.func_71507_a("Assigned renderer", render);
             crashreportcategory1.func_71507_a("Location", CrashReportCategory.func_85074_a(p_188391_2_, p_188391_4_, p_188391_6_));
-            crashreportcategory1.func_71507_a("Rotation", p_188391_8_);
-            crashreportcategory1.func_71507_a("Delta", p_188391_9_);
+            crashreportcategory1.func_71507_a("Rotation", Float.valueOf(p_188391_8_));
+            crashreportcategory1.func_71507_a("Delta", Float.valueOf(p_188391_9_));
             throw new ReportedException(crashreport);
         }
     }
@@ -445,7 +451,7 @@
         int k = i / 65536;
         OpenGlHelper.func_77475_a(OpenGlHelper.field_77476_b, (float)j, (float)k);
         GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
-        Render<Entity> render = this.func_78713_a(p_188389_1_);
+        Render<Entity> render = this.<Entity>func_78713_a(p_188389_1_);
 
         if (render != null && this.field_78724_e != null)
         {
@@ -480,7 +486,7 @@
         if (p_85094_1_ instanceof EntityLivingBase)
         {
             float f1 = 0.01F;
-            RenderGlobal.func_189694_a(p_85094_2_ - (double)f, p_85094_4_ + (double)p_85094_1_.func_70047_e() - (double)0.01F, p_85094_6_ - (double)f, p_85094_2_ + (double)f, p_85094_4_ + (double)p_85094_1_.func_70047_e() + (double)0.01F, p_85094_6_ + (double)f, 1.0F, 0.0F, 0.0F, 1.0F);
+            RenderGlobal.func_189694_a(p_85094_2_ - (double)f, p_85094_4_ + (double)p_85094_1_.func_70047_e() - 0.009999999776482582D, p_85094_6_ - (double)f, p_85094_2_ + (double)f, p_85094_4_ + (double)p_85094_1_.func_70047_e() + 0.009999999776482582D, p_85094_6_ + (double)f, 1.0F, 0.0F, 0.0F, 1.0F);
         }
 
         Tessellator tessellator = Tessellator.func_178181_a();
