--- before/net/minecraft/client/renderer/entity/RenderLiving.java
+++ after/net/minecraft/client/renderer/entity/RenderLiving.java
@@ -72,11 +72,11 @@
 
         if (entity != null)
         {
-            p_110827_4_ -= (1.6D - (double)p_110827_1_.field_70131_O) * 0.5D;
+            p_110827_4_ = p_110827_4_ - (1.6D - (double)p_110827_1_.field_70131_O) * 0.5D;
             Tessellator tessellator = Tessellator.func_178181_a();
             BufferBuilder bufferbuilder = tessellator.func_178180_c();
-            double d0 = this.func_110828_a((double)entity.field_70126_B, (double)entity.field_70177_z, (double)(p_110827_9_ * 0.5F)) * (double)((float)Math.PI / 180F);
-            double d1 = this.func_110828_a((double)entity.field_70127_C, (double)entity.field_70125_A, (double)(p_110827_9_ * 0.5F)) * (double)((float)Math.PI / 180F);
+            double d0 = this.func_110828_a((double)entity.field_70126_B, (double)entity.field_70177_z, (double)(p_110827_9_ * 0.5F)) * 0.01745329238474369D;
+            double d1 = this.func_110828_a((double)entity.field_70127_C, (double)entity.field_70125_A, (double)(p_110827_9_ * 0.5F)) * 0.01745329238474369D;
             double d2 = Math.cos(d0);
             double d3 = Math.sin(d0);
             double d4 = Math.sin(d1);
@@ -92,14 +92,14 @@
             double d6 = this.func_110828_a(entity.field_70169_q, entity.field_70165_t, (double)p_110827_9_) - d2 * 0.7D - d3 * 0.5D * d5;
             double d7 = this.func_110828_a(entity.field_70167_r + (double)entity.func_70047_e() * 0.7D, entity.field_70163_u + (double)entity.func_70047_e() * 0.7D, (double)p_110827_9_) - d4 * 0.5D - 0.25D;
             double d8 = this.func_110828_a(entity.field_70166_s, entity.field_70161_v, (double)p_110827_9_) - d3 * 0.7D + d2 * 0.5D * d5;
-            double d9 = this.func_110828_a((double)p_110827_1_.field_70760_ar, (double)p_110827_1_.field_70761_aq, (double)p_110827_9_) * (double)((float)Math.PI / 180F) + (Math.PI / 2D);
+            double d9 = this.func_110828_a((double)p_110827_1_.field_70760_ar, (double)p_110827_1_.field_70761_aq, (double)p_110827_9_) * 0.01745329238474369D + (Math.PI / 2D);
             d2 = Math.cos(d9) * (double)p_110827_1_.field_70130_N * 0.4D;
             d3 = Math.sin(d9) * (double)p_110827_1_.field_70130_N * 0.4D;
             double d10 = this.func_110828_a(p_110827_1_.field_70169_q, p_110827_1_.field_70165_t, (double)p_110827_9_) + d2;
             double d11 = this.func_110828_a(p_110827_1_.field_70167_r, p_110827_1_.field_70163_u, (double)p_110827_9_);
             double d12 = this.func_110828_a(p_110827_1_.field_70166_s, p_110827_1_.field_70161_v, (double)p_110827_9_) + d3;
-            p_110827_2_ += d2;
-            p_110827_6_ += d3;
+            p_110827_2_ = p_110827_2_ + d2;
+            p_110827_6_ = p_110827_6_ + d3;
             double d13 = (double)((float)(d6 - d10));
             double d14 = (double)((float)(d7 - d11));
             double d15 = (double)((float)(d8 - d12));
