--- before/net/minecraft/client/renderer/entity/Render.java
+++ after/net/minecraft/client/renderer/entity/Render.java
@@ -243,7 +243,7 @@
             BufferBuilder bufferbuilder = tessellator.func_178180_c();
             double d0 = ((double)p_188299_9_ - (p_188299_4_ - ((double)p_188299_8_.func_177956_o() + p_188299_13_)) / 2.0D) * 0.5D * (double)this.func_76982_b().func_175724_o(p_188299_8_);
 
-            if (!(d0 < 0.0D))
+            if (d0 >= 0.0D)
             {
                 if (d0 > 1.0D)
                 {
@@ -336,7 +336,7 @@
     {
         double d0 = p_147906_1_.func_70068_e(this.field_76990_c.field_78734_h);
 
-        if (!(d0 > (double)(p_147906_9_ * p_147906_9_)))
+        if (d0 <= (double)(p_147906_9_ * p_147906_9_))
         {
             boolean flag = p_147906_1_.func_70093_af();
             float f = this.field_76990_c.field_78735_i;
