--- before/net/minecraft/client/renderer/entity/RenderLivingBase.java
+++ after/net/minecraft/client/renderer/entity/RenderLivingBase.java
@@ -1,7 +1,6 @@
 package net.minecraft.client.renderer.entity;
 
 import com.google.common.collect.Lists;
-import java.nio.Buffer;
 import java.nio.FloatBuffer;
 import java.util.List;
 import net.minecraft.client.Minecraft;
@@ -30,9 +29,12 @@
     private static final DynamicTexture field_177096_e = new DynamicTexture(16, 16);
     protected ModelBase field_77045_g;
     protected FloatBuffer field_177095_g = GLAllocation.func_74529_h(4);
-    protected List<LayerRenderer<T>> field_177097_h = Lists.newArrayList();
+    protected List<LayerRenderer<T>> field_177097_h = Lists.<LayerRenderer<T>>newArrayList();
     protected boolean field_188323_j;
 
+    public static float NAME_TAG_RANGE = 64.0f;
+    public static float NAME_TAG_RANGE_SNEAK = 32.0f;
+
     public RenderLivingBase(RenderManager p_i46156_1_, ModelBase p_i46156_2_, float p_i46156_3_)
     {
         super(p_i46156_1_);
@@ -52,11 +54,11 @@
 
     protected float func_77034_a(float p_77034_1_, float p_77034_2_, float p_77034_3_)
     {
-        float f = p_77034_2_ - p_77034_1_;
+        float f;
 
-        while (f < -180.0F)
+        for (f = p_77034_2_ - p_77034_1_; f < -180.0F; f += 360.0F)
         {
-            f += 360.0F;
+            ;
         }
 
         while (f >= 180.0F)
@@ -73,10 +75,12 @@
 
     public void func_76986_a(T p_76986_1_, double p_76986_2_, double p_76986_4_, double p_76986_6_, float p_76986_8_, float p_76986_9_)
     {
+        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Pre<T>(p_76986_1_, this, p_76986_9_, p_76986_2_, p_76986_4_, p_76986_6_))) return;
         GlStateManager.func_179094_E();
         GlStateManager.func_179129_p();
         this.field_77045_g.field_78095_p = this.func_77040_d(p_76986_1_, p_76986_9_);
-        this.field_77045_g.field_78093_q = p_76986_1_.func_184218_aH();
+        boolean shouldSit = p_76986_1_.func_184218_aH() && (p_76986_1_.func_184187_bx() != null && p_76986_1_.func_184187_bx().shouldRiderSit());
+        this.field_77045_g.field_78093_q = shouldSit;
         this.field_77045_g.field_78091_s = p_76986_1_.func_70631_g_();
 
         try
@@ -85,7 +89,7 @@
             float f1 = this.func_77034_a(p_76986_1_.field_70758_at, p_76986_1_.field_70759_as, p_76986_9_);
             float f2 = f1 - f;
 
-            if (p_76986_1_.func_184218_aH() && p_76986_1_.func_184187_bx() instanceof EntityLivingBase)
+            if (shouldSit && p_76986_1_.func_184187_bx() instanceof EntityLivingBase)
             {
                 EntityLivingBase entitylivingbase = (EntityLivingBase)p_76986_1_.func_184187_bx();
                 f = this.func_77034_a(entitylivingbase.field_70760_ar, entitylivingbase.field_70761_aq, p_76986_9_);
@@ -134,6 +138,7 @@
                 {
                     f5 = 1.0F;
                 }
+                f2 = f1 - f; // Forge: Fix MC-1207
             }
 
             GlStateManager.func_179141_d();
@@ -195,6 +200,7 @@
         GlStateManager.func_179089_o();
         GlStateManager.func_179121_F();
         super.func_76986_a(p_76986_1_, p_76986_2_, p_76986_4_, p_76986_6_, p_76986_8_, p_76986_9_);
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Post<T>(p_76986_1_, this, p_76986_9_, p_76986_2_, p_76986_4_, p_76986_6_));
     }
 
     public float func_188322_c(T p_188322_1_, float p_188322_2_)
@@ -264,7 +270,7 @@
     {
         float f = p_177092_1_.func_70013_c();
         int i = this.func_77030_a(p_177092_1_, f, p_177092_2_);
-        boolean flag = (i >> 24 & 0xFF) > 0;
+        boolean flag = (i >> 24 & 255) > 0;
         boolean flag1 = p_177092_1_.field_70737_aN > 0 || p_177092_1_.field_70725_aQ > 0;
 
         if (!flag && !flag1)
@@ -301,7 +307,7 @@
             GlStateManager.func_187399_a(8960, OpenGlHelper.field_176077_E, 7681);
             GlStateManager.func_187399_a(8960, OpenGlHelper.field_176078_F, OpenGlHelper.field_176091_w);
             GlStateManager.func_187399_a(8960, OpenGlHelper.field_176085_I, 770);
-            ((Buffer)this.field_177095_g).position(0);
+            this.field_177095_g.position(0);
 
             if (flag1)
             {
@@ -312,17 +318,17 @@
             }
             else
             {
-                float f1 = (float)(i >> 24 & 0xFF) / 255.0F;
-                float f2 = (float)(i >> 16 & 0xFF) / 255.0F;
-                float f3 = (float)(i >> 8 & 0xFF) / 255.0F;
-                float f4 = (float)(i & 0xFF) / 255.0F;
+                float f1 = (float)(i >> 24 & 255) / 255.0F;
+                float f2 = (float)(i >> 16 & 255) / 255.0F;
+                float f3 = (float)(i >> 8 & 255) / 255.0F;
+                float f4 = (float)(i & 255) / 255.0F;
                 this.field_177095_g.put(f2);
                 this.field_177095_g.put(f3);
                 this.field_177095_g.put(f4);
                 this.field_177095_g.put(1.0F - f1);
             }
 
-            ((Buffer)this.field_177095_g).flip();
+            this.field_177095_g.flip();
             GlStateManager.func_187448_b(8960, 8705, this.field_177095_g);
             GlStateManager.func_179138_g(OpenGlHelper.field_176096_r);
             GlStateManager.func_179098_w();
@@ -407,9 +413,7 @@
         {
             String s = TextFormatting.func_110646_a(p_77043_1_.func_70005_c_());
 
-            if (s != null
-                    && ("Dinnerbone".equals(s) || "Grumm".equals(s))
-                    && (!(p_77043_1_ instanceof EntityPlayer) || ((EntityPlayer)p_77043_1_).func_175148_a(EnumPlayerModelParts.CAPE)))
+            if (s != null && ("Dinnerbone".equals(s) || "Grumm".equals(s)) && (!(p_77043_1_ instanceof EntityPlayer) || ((EntityPlayer)p_77043_1_).func_175148_a(EnumPlayerModelParts.CAPE)))
             {
                 GlStateManager.func_179109_b(0.0F, p_77043_1_.field_70131_O + 0.1F, 0.0F);
                 GlStateManager.func_179114_b(180.0F, 0.0F, 0.0F, 1.0F);
@@ -427,9 +431,7 @@
         return (float)p_77044_1_.field_70173_aa + p_77044_2_;
     }
 
-    protected void func_177093_a(
-        T p_177093_1_, float p_177093_2_, float p_177093_3_, float p_177093_4_, float p_177093_5_, float p_177093_6_, float p_177093_7_, float p_177093_8_
-    )
+    protected void func_177093_a(T p_177093_1_, float p_177093_2_, float p_177093_3_, float p_177093_4_, float p_177093_5_, float p_177093_6_, float p_177093_7_, float p_177093_8_)
     {
         for (LayerRenderer<T> layerrenderer : this.field_177097_h)
         {
@@ -459,18 +461,20 @@
 
     public void func_177067_a(T p_177067_1_, double p_177067_2_, double p_177067_4_, double p_177067_6_)
     {
+        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Specials.Pre<T>(p_177067_1_, this, p_177067_2_, p_177067_4_, p_177067_6_))) return;
         if (this.func_177070_b(p_177067_1_))
         {
             double d0 = p_177067_1_.func_70068_e(this.field_76990_c.field_78734_h);
-            float f = p_177067_1_.func_70093_af() ? 32.0F : 64.0F;
+            float f = p_177067_1_.func_70093_af() ? NAME_TAG_RANGE_SNEAK : NAME_TAG_RANGE;
 
-            if (!(d0 >= (double)(f * f)))
+            if (d0 < (double)(f * f))
             {
                 String s = p_177067_1_.func_145748_c_().func_150254_d();
                 GlStateManager.func_179092_a(516, 0.1F);
                 this.func_188296_a(p_177067_1_, p_177067_2_, p_177067_4_, p_177067_6_, s, d0);
             }
         }
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderLivingEvent.Specials.Post<T>(p_177067_1_, this, p_177067_2_, p_177067_4_, p_177067_6_));
     }
 
     protected boolean func_177070_b(T p_177070_1_)
