--- before/net/minecraft/client/renderer/entity/RenderItemFrame.java
+++ after/net/minecraft/client/renderer/entity/RenderItemFrame.java
@@ -42,14 +42,14 @@
         double d0 = (double)blockpos.func_177958_n() - p_76986_1_.field_70165_t + p_76986_2_;
         double d1 = (double)blockpos.func_177956_o() - p_76986_1_.field_70163_u + p_76986_4_;
         double d2 = (double)blockpos.func_177952_p() - p_76986_1_.field_70161_v + p_76986_6_;
-        GlStateManager.func_179137_b(d0 + 0.5, d1 + 0.5, d2 + 0.5);
+        GlStateManager.func_179137_b(d0 + 0.5D, d1 + 0.5D, d2 + 0.5D);
         GlStateManager.func_179114_b(180.0F - p_76986_1_.field_70177_z, 0.0F, 1.0F, 0.0F);
         this.field_76990_c.field_78724_e.func_110577_a(TextureMap.field_110575_b);
         BlockRendererDispatcher blockrendererdispatcher = this.field_147917_g.func_175602_ab();
         ModelManager modelmanager = blockrendererdispatcher.func_175023_a().func_178126_b();
         IBakedModel ibakedmodel;
 
-        if (p_76986_1_.func_82335_i().func_77973_b() == Items.field_151098_aY)
+        if (p_76986_1_.func_82335_i().func_77973_b() instanceof net.minecraft.item.ItemMap)
         {
             ibakedmodel = modelmanager.func_174953_a(this.field_177073_g);
         }
@@ -79,12 +79,7 @@
         GlStateManager.func_179109_b(0.0F, 0.0F, 0.4375F);
         this.func_82402_b(p_76986_1_);
         GlStateManager.func_179121_F();
-        this.func_177067_a(
-            p_76986_1_,
-            p_76986_2_ + (double)((float)p_76986_1_.field_174860_b.func_82601_c() * 0.3F),
-            p_76986_4_ - 0.25,
-            p_76986_6_ + (double)((float)p_76986_1_.field_174860_b.func_82599_e() * 0.3F)
-        );
+        this.func_177067_a(p_76986_1_, p_76986_2_ + (double)((float)p_76986_1_.field_174860_b.func_82601_c() * 0.3F), p_76986_4_ - 0.25D, p_76986_6_ + (double)((float)p_76986_1_.field_174860_b.func_82599_e() * 0.3F));
     }
 
     @Nullable
@@ -101,10 +96,13 @@
         {
             GlStateManager.func_179094_E();
             GlStateManager.func_179140_f();
-            boolean flag = itemstack.func_77973_b() == Items.field_151098_aY;
+            boolean flag = itemstack.func_77973_b() instanceof net.minecraft.item.ItemMap;
             int i = flag ? p_82402_1_.func_82333_j() % 4 * 2 : p_82402_1_.func_82333_j();
             GlStateManager.func_179114_b((float)i * 360.0F / 8.0F, 0.0F, 0.0F, 1.0F);
 
+            net.minecraftforge.client.event.RenderItemInFrameEvent event = new net.minecraftforge.client.event.RenderItemInFrameEvent(p_82402_1_, this);
+            if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event))
+            {
             if (flag)
             {
                 this.field_76990_c.field_78724_e.func_110577_a(field_110789_a);
@@ -112,7 +110,7 @@
                 float f = 0.0078125F;
                 GlStateManager.func_179152_a(0.0078125F, 0.0078125F, 0.0078125F);
                 GlStateManager.func_179109_b(-64.0F, -64.0F, 0.0F);
-                MapData mapdata = Items.field_151098_aY.func_77873_a(itemstack, p_82402_1_.field_70170_p);
+                MapData mapdata = ((net.minecraft.item.ItemMap) itemstack.func_77973_b()).func_77873_a(itemstack, p_82402_1_.field_70170_p);
                 GlStateManager.func_179109_b(0.0F, 0.0F, -1.0F);
 
                 if (mapdata != null)
@@ -129,6 +127,7 @@
                 RenderHelper.func_74518_a();
                 GlStateManager.func_179099_b();
             }
+            }
 
             GlStateManager.func_179145_e();
             GlStateManager.func_179121_F();
@@ -137,15 +136,12 @@
 
     protected void func_177067_a(EntityItemFrame p_177067_1_, double p_177067_2_, double p_177067_4_, double p_177067_6_)
     {
-        if (Minecraft.func_71382_s()
-                && !p_177067_1_.func_82335_i().func_190926_b()
-                && p_177067_1_.func_82335_i().func_82837_s()
-                && this.field_76990_c.field_147941_i == p_177067_1_)
+        if (Minecraft.func_71382_s() && !p_177067_1_.func_82335_i().func_190926_b() && p_177067_1_.func_82335_i().func_82837_s() && this.field_76990_c.field_147941_i == p_177067_1_)
         {
             double d0 = p_177067_1_.func_70068_e(this.field_76990_c.field_78734_h);
             float f = p_177067_1_.func_70093_af() ? 32.0F : 64.0F;
 
-            if (!(d0 >= (double)(f * f)))
+            if (d0 < (double)(f * f))
             {
                 String s = p_177067_1_.func_82335_i().func_82833_r();
                 this.func_147906_a(p_177067_1_, s, p_177067_2_, p_177067_4_, p_177067_6_, 64);
