--- before/net/minecraft/client/renderer/entity/RenderPlayer.java
+++ after/net/minecraft/client/renderer/entity/RenderPlayer.java
@@ -56,6 +56,7 @@
 
     public void doRender(AbstractClientPlayer entity, double x, double y, double z, float entityYaw, float partialTicks)
     {
+        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderPlayerEvent.Pre(entity, this, partialTicks, x, y, z))) return;
         if (!entity.isUser() || this.renderManager.renderViewEntity == entity)
         {
             double d0 = y;
@@ -70,6 +71,7 @@
             super.doRender(entity, x, d0, z, entityYaw, partialTicks);
             GlStateManager.disableBlendProfile(GlStateManager.Profile.PLAYER_SKIN);
         }
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.client.event.RenderPlayerEvent.Post(entity, this, partialTicks, x, y, z));
     }
 
     private void setModelVisibilities(AbstractClientPlayer clientPlayer)
@@ -128,6 +130,11 @@
                     {
                         modelbiped$armpose1 = ModelBiped.ArmPose.BLOCK;
                     }
+                    // FORGE: fix MC-88356 allow offhand to use bow and arrow animation
+                    else if (enumaction1 == EnumAction.BOW)
+                    {
+                        modelbiped$armpose1 = ModelBiped.ArmPose.BOW_AND_ARROW;
+                    }
                 }
             }
 
@@ -183,6 +190,7 @@
 
     public void renderRightArm(AbstractClientPlayer clientPlayer)
     {
+        if (net.minecraftforge.client.ForgeHooksClient.renderSpecificFirstPersonArm(this, clientPlayer, EnumHandSide.RIGHT)) return;
         float f = 1.0F;
         GlStateManager.color(1.0F, 1.0F, 1.0F);
         float f1 = 0.0625F;
@@ -201,6 +209,7 @@
 
     public void renderLeftArm(AbstractClientPlayer clientPlayer)
     {
+        if (net.minecraftforge.client.ForgeHooksClient.renderSpecificFirstPersonArm(this, clientPlayer, EnumHandSide.LEFT)) return;
         float f = 1.0F;
         GlStateManager.color(1.0F, 1.0F, 1.0F);
         float f1 = 0.0625F;
