--- before/net/minecraft/client/renderer/entity/RenderZombieVillager.java
+++ after/net/minecraft/client/renderer/entity/RenderZombieVillager.java
@@ -25,6 +25,7 @@
 
     protected ResourceLocation getEntityTexture(EntityZombieVillager entity)
     {
+        if (true) return entity.getForgeProfession().getZombieSkin();
         switch (entity.getProfession())
         {
             case 0:
@@ -47,7 +48,7 @@
     {
         if (entityLiving.isConverting())
         {
-            rotationYaw += (float)(Math.cos((double)entityLiving.ticksExisted * 3.25) * Math.PI * 0.25);
+            rotationYaw += (float)(Math.cos((double)entityLiving.ticksExisted * 3.25D) * Math.PI * 0.25D);
         }
 
         super.applyRotations(entityLiving, ageInTicks, rotationYaw, partialTicks);
