--- before/net/minecraft/client/renderer/entity/RenderShulker.java
+++ after/net/minecraft/client/renderer/entity/RenderShulker.java
@@ -38,7 +38,7 @@
             BlockPos blockpos = p_76986_1_.func_184699_da();
             BlockPos blockpos1 = p_76986_1_.func_184692_dd();
             double d0 = (double)((float)i - p_76986_9_) / 6.0D;
-            d0 *= d0;
+            d0 = d0 * d0;
             double d1 = (double)(blockpos.func_177958_n() - blockpos1.func_177958_n()) * d0;
             double d2 = (double)(blockpos.func_177956_o() - blockpos1.func_177956_o()) * d0;
             double d3 = (double)(blockpos.func_177952_p() - blockpos1.func_177952_p()) * d0;
@@ -163,8 +163,8 @@
             }
 
             ModelRenderer modelrenderer = RenderShulker.this.func_177087_b().field_187066_a;
-            modelrenderer.field_78796_g = p_177141_6_ * ((float)Math.PI / 180F);
-            modelrenderer.field_78795_f = p_177141_7_ * ((float)Math.PI / 180F);
+            modelrenderer.field_78796_g = p_177141_6_ * 0.017453292F;
+            modelrenderer.field_78795_f = p_177141_7_ * 0.017453292F;
             RenderShulker.this.func_110776_a(RenderShulker.field_188342_a[p_177141_1_.func_190769_dn().func_176765_a()]);
             modelrenderer.func_78785_a(p_177141_8_);
             GlStateManager.func_179121_F();
