--- before/net/minecraft/client/renderer/entity/RenderGuardian.java
+++ after/net/minecraft/client/renderer/entity/RenderGuardian.java
@@ -119,8 +119,8 @@
             double d15 = 0.0D + Math.sin(d1 + 0.0D) * 0.2D;
             double d16 = 0.0D + Math.cos(d1 + (Math.PI / 2D)) * 0.2D;
             double d17 = 0.0D + Math.sin(d1 + (Math.PI / 2D)) * 0.2D;
-            double d18 = 0.0D + Math.cos(d1 + (Math.PI * 1.5D)) * 0.2D;
-            double d19 = 0.0D + Math.sin(d1 + (Math.PI * 1.5D)) * 0.2D;
+            double d18 = 0.0D + Math.cos(d1 + (Math.PI * 3D / 2D)) * 0.2D;
+            double d19 = 0.0D + Math.sin(d1 + (Math.PI * 3D / 2D)) * 0.2D;
             double d20 = 0.0D;
             double d21 = 0.4999D;
             double d22 = (double)(-1.0F + f3);
