--- before/net/minecraft/client/renderer/entity/RenderVex.java
+++ after/net/minecraft/client/renderer/entity/RenderVex.java
@@ -12,11 +12,12 @@
 {
     private static final ResourceLocation field_191343_a = new ResourceLocation("textures/entity/illager/vex.png");
     private static final ResourceLocation field_191344_j = new ResourceLocation("textures/entity/illager/vex_charging.png");
-    private int field_191345_k = ((ModelVex)this.field_77045_g).func_191228_a();
+    private int field_191345_k;
 
     public RenderVex(RenderManager p_i47190_1_)
     {
         super(p_i47190_1_, new ModelVex(), 0.3F);
+        this.field_191345_k = ((ModelVex)this.field_77045_g).func_191228_a();
     }
 
     protected ResourceLocation func_110775_a(EntityVex p_110775_1_)
