--- before/net/minecraft/client/renderer/entity/RenderCreeper.java
+++ after/net/minecraft/client/renderer/entity/RenderCreeper.java
@@ -25,8 +25,8 @@
         float f = p_77041_1_.func_70831_j(p_77041_2_);
         float f1 = 1.0F + MathHelper.func_76126_a(f * 100.0F) * f * 0.01F;
         f = MathHelper.func_76131_a(f, 0.0F, 1.0F);
-        f *= f;
-        f *= f;
+        f = f * f;
+        f = f * f;
         float f2 = (1.0F + f * 0.4F) * f1;
         float f3 = (1.0F + f * 0.1F) / f1;
         GlStateManager.func_179152_a(f2, f3, f2);
