--- before/net/minecraft/client/renderer/entity/RenderZombieVillager.java
+++ after/net/minecraft/client/renderer/entity/RenderZombieVillager.java
@@ -25,6 +25,7 @@
 
     protected ResourceLocation func_110775_a(EntityZombieVillager p_110775_1_)
     {
+        if (true) return p_110775_1_.getForgeProfession().getZombieSkin();
         switch (p_110775_1_.func_190736_dl())
         {
             case 0:
@@ -47,7 +48,7 @@
     {
         if (p_77043_1_.func_82230_o())
         {
-            p_77043_3_ += (float)(Math.cos((double)p_77043_1_.field_70173_aa * 3.25) * Math.PI * 0.25);
+            p_77043_3_ += (float)(Math.cos((double)p_77043_1_.field_70173_aa * 3.25D) * Math.PI * 0.25D);
         }
 
         super.func_77043_a(p_77043_1_, p_77043_2_, p_77043_3_, p_77043_4_);
