--- before/net/minecraft/client/renderer/entity/RenderTNTPrimed.java
+++ after/net/minecraft/client/renderer/entity/RenderTNTPrimed.java
@@ -30,8 +30,8 @@
         {
             float f = 1.0F - ((float)p_76986_1_.func_184536_l() - p_76986_9_ + 1.0F) / 10.0F;
             f = MathHelper.func_76131_a(f, 0.0F, 1.0F);
-            f *= f;
-            f *= f;
+            f = f * f;
+            f = f * f;
             float f1 = 1.0F + f * 0.3F;
             GlStateManager.func_179152_a(f1, f1, f1);
         }
