--- before/net/minecraft/util/text/translation/LanguageMap.java
+++ after/net/minecraft/util/text/translation/LanguageMap.java
@@ -18,35 +18,64 @@
     private static final Pattern field_111053_a = Pattern.compile("%(\\d+\\$)?[\\d\\.]*[df]");
     private static final Splitter field_135065_b = Splitter.on('=').limit(2);
     private static final LanguageMap field_74817_a = new LanguageMap();
-    private final Map<String, String> field_74816_c = Maps.newHashMap();
+    private final Map<String, String> field_74816_c = Maps.<String, String>newHashMap();
     private long field_150511_e;
 
     public LanguageMap()
     {
-        try
-        {
-            InputStream inputstream = LanguageMap.class.getResourceAsStream("/assets/minecraft/lang/en_us.lang");
+        InputStream inputstream = LanguageMap.class.getResourceAsStream("/assets/minecraft/lang/en_us.lang");
+        try
+        {
+            inject(this, inputstream);
+        }
+        finally
+        {
+            IOUtils.closeQuietly(inputstream); // Forge: close stream after use (MC-153470)
+        }
+    }
+
+    public static void inject(InputStream inputstream)
+    {
+        inject(field_74817_a, inputstream);
+    }
+
+    private static void inject(LanguageMap inst, InputStream inputstream)
+    {
+        Map<String, String> map = parseLangFile(inputstream);
+        inst.field_74816_c.putAll(map);
+        inst.field_150511_e = System.currentTimeMillis();
+    }
+
+    public static Map<String, String> parseLangFile(InputStream inputstream)
+    {
+        Map<String, String> table = Maps.newHashMap();
+        try
+        {
+            inputstream = net.minecraftforge.fml.common.FMLCommonHandler.instance().loadLanguage(table, inputstream);
+            if (inputstream == null) return table;
 
             for (String s : IOUtils.readLines(inputstream, StandardCharsets.UTF_8))
             {
                 if (!s.isEmpty() && s.charAt(0) != '#')
                 {
-                    String[] astring = Iterables.toArray(field_135065_b.split(s), String.class);
+                    String[] astring = (String[])Iterables.toArray(field_135065_b.split(s), String.class);
 
                     if (astring != null && astring.length == 2)
                     {
                         String s1 = astring[0];
                         String s2 = field_111053_a.matcher(astring[1]).replaceAll("%$1s");
-                        this.field_74816_c.put(s1, s2);
+                        table.put(s1, s2);
                     }
                 }
             }
 
-            this.field_150511_e = System.currentTimeMillis();
         }
-        catch (IOException ioexception)
+        catch (IOException var7)
         {
+            ;
         }
+        catch (Exception ex) {}
+        return table;
     }
 
     static LanguageMap func_74808_a()
@@ -76,7 +105,7 @@
         {
             return String.format(s, p_74803_2_);
         }
-        catch (IllegalFormatException illegalformatexception)
+        catch (IllegalFormatException var5)
         {
             return "Format error: " + s;
         }
