--- before/net/minecraft/tileentity/BannerPattern.java
+++ after/net/minecraft/tileentity/BannerPattern.java
@@ -52,11 +52,13 @@
 
     private final String fileName;
     private final String hashname;
-    private final String[] patterns = new String[3];
-    private ItemStack patternItem = ItemStack.EMPTY;
+    private final String[] patterns;
+    private ItemStack patternItem;
 
     private BannerPattern(String p_i47245_3_, String p_i47245_4_)
     {
+        this.patterns = new String[3];
+        this.patternItem = ItemStack.EMPTY;
         this.fileName = p_i47245_3_;
         this.hashname = p_i47245_4_;
     }
