--- before/net/minecraft/tileentity/BannerPattern.java
+++ after/net/minecraft/tileentity/BannerPattern.java
@@ -52,11 +52,13 @@
 
     private final String field_191014_N;
     private final String field_191015_O;
-    private final String[] field_191016_P = new String[3];
-    private ItemStack field_191017_Q = ItemStack.field_190927_a;
+    private final String[] field_191016_P;
+    private ItemStack field_191017_Q;
 
     private BannerPattern(String p_i47245_3_, String p_i47245_4_)
     {
+        this.field_191016_P = new String[3];
+        this.field_191017_Q = ItemStack.field_190927_a;
         this.field_191014_N = p_i47245_3_;
         this.field_191015_O = p_i47245_4_;
     }
