--- before/net/minecraft/tileentity/MobSpawnerBaseLogic.java
+++ after/net/minecraft/tileentity/MobSpawnerBaseLogic.java
@@ -5,6 +5,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLiving;
+import net.minecraft.entity.IEntityLivingData;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTTagList;
 import net.minecraft.util.EnumParticleTypes;
@@ -22,7 +23,7 @@
 public abstract class MobSpawnerBaseLogic
 {
     private int field_98286_b = 20;
-    private final List<WeightedSpawnerEntity> field_98285_e = Lists.newArrayList();
+    private final List<WeightedSpawnerEntity> field_98285_e = Lists.<WeightedSpawnerEntity>newArrayList();
     private WeightedSpawnerEntity field_98282_f = new WeightedSpawnerEntity();
     private double field_98287_c;
     private double field_98284_d;
@@ -52,10 +53,7 @@
     private boolean func_98279_f()
     {
         BlockPos blockpos = this.func_177221_b();
-        return this.func_98271_a()
-               .func_175636_b(
-                   (double)blockpos.func_177958_n() + 0.5, (double)blockpos.func_177956_o() + 0.5, (double)blockpos.func_177952_p() + 0.5, (double)this.field_98289_l
-               );
+        return this.func_98271_a().func_175636_b((double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 0.5D, (double)blockpos.func_177952_p() + 0.5D, (double)this.field_98289_l);
     }
 
     public void func_98278_g()
@@ -70,11 +68,11 @@
 
             if (this.func_98271_a().field_72995_K)
             {
-                double d0 = (double)((float)blockpos.func_177958_n() + this.func_98271_a().field_73012_v.nextFloat());
-                double d1 = (double)((float)blockpos.func_177956_o() + this.func_98271_a().field_73012_v.nextFloat());
-                double d2 = (double)((float)blockpos.func_177952_p() + this.func_98271_a().field_73012_v.nextFloat());
-                this.func_98271_a().func_175688_a(EnumParticleTypes.SMOKE_NORMAL, d0, d1, d2, 0.0, 0.0, 0.0);
-                this.func_98271_a().func_175688_a(EnumParticleTypes.FLAME, d0, d1, d2, 0.0, 0.0, 0.0);
+                double d3 = (double)((float)blockpos.func_177958_n() + this.func_98271_a().field_73012_v.nextFloat());
+                double d4 = (double)((float)blockpos.func_177956_o() + this.func_98271_a().field_73012_v.nextFloat());
+                double d5 = (double)((float)blockpos.func_177952_p() + this.func_98271_a().field_73012_v.nextFloat());
+                this.func_98271_a().func_175688_a(EnumParticleTypes.SMOKE_NORMAL, d3, d4, d5, 0.0D, 0.0D, 0.0D);
+                this.func_98271_a().func_175688_a(EnumParticleTypes.FLAME, d3, d4, d5, 0.0D, 0.0D, 0.0D);
 
                 if (this.field_98286_b > 0)
                 {
@@ -82,7 +80,7 @@
                 }
 
                 this.field_98284_d = this.field_98287_c;
-                this.field_98287_c = (this.field_98287_c + (double)(1000.0F / ((float)this.field_98286_b + 200.0F))) % 360.0;
+                this.field_98287_c = (this.field_98287_c + (double)(1000.0F / ((float)this.field_98286_b + 200.0F))) % 360.0D;
             }
             else
             {
@@ -105,33 +103,17 @@
                     NBTTagList nbttaglist = nbttagcompound.func_150295_c("Pos", 6);
                     World world = this.func_98271_a();
                     int j = nbttaglist.func_74745_c();
-                    double d3 = j >= 1
-                                ? nbttaglist.func_150309_d(0)
-                                : (double)blockpos.func_177958_n() + (world.field_73012_v.nextDouble() - world.field_73012_v.nextDouble()) * (double)this.field_98290_m + 0.5;
-                    double d4 = j >= 2 ? nbttaglist.func_150309_d(1) : (double)(blockpos.func_177956_o() + world.field_73012_v.nextInt(3) - 1);
-                    double d5 = j >= 3
-                                ? nbttaglist.func_150309_d(2)
-                                : (double)blockpos.func_177952_p() + (world.field_73012_v.nextDouble() - world.field_73012_v.nextDouble()) * (double)this.field_98290_m + 0.5;
-                    Entity entity = AnvilChunkLoader.func_186054_a(nbttagcompound, world, d3, d4, d5, false);
+                    double d0 = j >= 1 ? nbttaglist.func_150309_d(0) : (double)blockpos.func_177958_n() + (world.field_73012_v.nextDouble() - world.field_73012_v.nextDouble()) * (double)this.field_98290_m + 0.5D;
+                    double d1 = j >= 2 ? nbttaglist.func_150309_d(1) : (double)(blockpos.func_177956_o() + world.field_73012_v.nextInt(3) - 1);
+                    double d2 = j >= 3 ? nbttaglist.func_150309_d(2) : (double)blockpos.func_177952_p() + (world.field_73012_v.nextDouble() - world.field_73012_v.nextDouble()) * (double)this.field_98290_m + 0.5D;
+                    Entity entity = AnvilChunkLoader.func_186054_a(nbttagcompound, world, d0, d1, d2, false);
 
                     if (entity == null)
                     {
                         return;
                     }
 
-                    int k = world.func_72872_a(
-                                entity.getClass(),
-                                new AxisAlignedBB(
-                                    (double)blockpos.func_177958_n(),
-                                    (double)blockpos.func_177956_o(),
-                                    (double)blockpos.func_177952_p(),
-                                    (double)(blockpos.func_177958_n() + 1),
-                                    (double)(blockpos.func_177956_o() + 1),
-                                    (double)(blockpos.func_177952_p() + 1)
-                                )
-                                .func_186662_g((double)this.field_98290_m)
-                            )
-                            .size();
+                    int k = world.func_72872_a(entity.getClass(), (new AxisAlignedBB((double)blockpos.func_177958_n(), (double)blockpos.func_177956_o(), (double)blockpos.func_177952_p(), (double)(blockpos.func_177958_n() + 1), (double)(blockpos.func_177956_o() + 1), (double)(blockpos.func_177952_p() + 1))).func_186662_g((double)this.field_98290_m)).size();
 
                     if (k >= this.field_98292_k)
                     {
@@ -142,13 +124,12 @@
                     EntityLiving entityliving = entity instanceof EntityLiving ? (EntityLiving)entity : null;
                     entity.func_70012_b(entity.field_70165_t, entity.field_70163_u, entity.field_70161_v, world.field_73012_v.nextFloat() * 360.0F, 0.0F);
 
-                    if (entityliving == null || entityliving.func_70601_bi() && entityliving.func_70058_J())
+                    if (entityliving == null || net.minecraftforge.event.ForgeEventFactory.canEntitySpawnSpawner(entityliving, func_98271_a(), (float)entity.field_70165_t, (float)entity.field_70163_u, (float)entity.field_70161_v, this))
                     {
-                        if (this.field_98282_f.func_185277_b().func_186856_d() == 1
-                                && this.field_98282_f.func_185277_b().func_150297_b("id", 8)
-                                && entity instanceof EntityLiving)
+                        if (this.field_98282_f.func_185277_b().func_186856_d() == 1 && this.field_98282_f.func_185277_b().func_150297_b("id", 8) && entity instanceof EntityLiving)
                         {
-                            ((EntityLiving)entity).func_180482_a(world.func_175649_E(new BlockPos(entity)), null);
+                            if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(entityliving, this.func_98271_a(), (float)entity.field_70165_t, (float)entity.field_70163_u, (float)entity.field_70161_v, this))
+                            ((EntityLiving)entity).func_180482_a(world.func_175649_E(new BlockPos(entity)), (IEntityLivingData)null);
                         }
 
                         AnvilChunkLoader.func_186052_a(entity, world);
@@ -179,12 +160,13 @@
         }
         else
         {
-            this.field_98286_b = this.field_98283_g + this.func_98271_a().field_73012_v.nextInt(this.field_98293_h - this.field_98283_g);
+            int i = this.field_98293_h - this.field_98283_g;
+            this.field_98286_b = this.field_98283_g + this.func_98271_a().field_73012_v.nextInt(i);
         }
 
         if (!this.field_98285_e.isEmpty())
         {
-            this.func_184993_a(WeightedRandom.func_76271_a(this.func_98271_a().field_73012_v, this.field_98285_e));
+            this.func_184993_a((WeightedSpawnerEntity)WeightedRandom.func_76271_a(this.func_98271_a().field_73012_v, this.field_98285_e));
         }
 
         this.func_98267_a(1);
@@ -211,7 +193,7 @@
         }
         else if (!this.field_98285_e.isEmpty())
         {
-            this.func_184993_a(WeightedRandom.func_76271_a(this.func_98271_a().field_73012_v, this.field_98285_e));
+            this.func_184993_a((WeightedSpawnerEntity)WeightedRandom.func_76271_a(this.func_98271_a().field_73012_v, this.field_98285_e));
         }
 
         if (p_98270_1_.func_150297_b("MinSpawnDelay", 99))
@@ -275,24 +257,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
-    public Entity func_184994_d()
-    {
-        if (this.field_98291_j == null)
-        {
-            this.field_98291_j = AnvilChunkLoader.func_186051_a(this.field_98282_f.func_185277_b(), this.func_98271_a(), false);
-
-            if (this.field_98282_f.func_185277_b().func_186856_d() == 1
-                    && this.field_98282_f.func_185277_b().func_150297_b("id", 8)
-                    && this.field_98291_j instanceof EntityLiving)
-            {
-                ((EntityLiving)this.field_98291_j).func_180482_a(this.func_98271_a().func_175649_E(new BlockPos(this.field_98291_j)), null);
-            }
-        }
-
-        return this.field_98291_j;
-    }
-
     public boolean func_98268_b(int p_98268_1_)
     {
         if (p_98268_1_ == 1 && this.func_98271_a().field_72995_K)
@@ -306,6 +270,22 @@
         }
     }
 
+    @SideOnly(Side.CLIENT)
+    public Entity func_184994_d()
+    {
+        if (this.field_98291_j == null)
+        {
+            this.field_98291_j = AnvilChunkLoader.func_186051_a(this.field_98282_f.func_185277_b(), this.func_98271_a(), false);
+
+            if (this.field_98282_f.func_185277_b().func_186856_d() == 1 && this.field_98282_f.func_185277_b().func_150297_b("id", 8) && this.field_98291_j instanceof EntityLiving)
+            {
+                ((EntityLiving)this.field_98291_j).func_180482_a(this.func_98271_a().func_175649_E(new BlockPos(this.field_98291_j)), (IEntityLivingData)null);
+            }
+        }
+
+        return this.field_98291_j;
+    }
+
     public void func_184993_a(WeightedSpawnerEntity p_184993_1_)
     {
         this.field_98282_f = p_184993_1_;
@@ -328,4 +308,7 @@
     {
         return this.field_98284_d;
     }
+
+    /* ======================================== FORGE START =====================================*/
+    @Nullable public Entity getSpawnerEntity() { return null; }
 }
