--- before/net/minecraft/tileentity/MobSpawnerBaseLogic.java
+++ after/net/minecraft/tileentity/MobSpawnerBaseLogic.java
@@ -140,12 +140,13 @@
                     EntityLiving entityliving = entity instanceof EntityLiving ? (EntityLiving)entity : null;
                     entity.setLocationAndAngles(entity.posX, entity.posY, entity.posZ, world.rand.nextFloat() * 360.0F, 0.0F);
 
-                    if (entityliving == null || entityliving.getCanSpawnHere() && entityliving.isNotColliding())
+                    if (entityliving == null || net.minecraftforge.event.ForgeEventFactory.canEntitySpawnSpawner(entityliving, getSpawnerWorld(), (float)entity.posX, (float)entity.posY, (float)entity.posZ, this))
                     {
                         if (this.spawnData.getNbt().getSize() == 1
                                 && this.spawnData.getNbt().hasKey("id", 8)
                                 && entity instanceof EntityLiving)
                         {
+                            if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(entityliving, this.getSpawnerWorld(), (float)entity.posX, (float)entity.posY, (float)entity.posZ, this))
                             ((EntityLiving)entity).onInitialSpawn(world.getDifficultyForLocation(new BlockPos(entity)), null);
                         }
 
@@ -326,4 +327,7 @@
     {
         return this.prevMobRotation;
     }
+
+    /* ======================================== FORGE START =====================================*/
+    @Nullable public Entity getSpawnerEntity() { return null; }
 }
