--- before/net/minecraft/tileentity/TileEntityFurnace.java
+++ after/net/minecraft/tileentity/TileEntityFurnace.java
@@ -32,7 +32,7 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class TileEntityFurnace extends TileEntityLockable implements ITickable, ISidedInventory
+public class TileEntityFurnace extends TileEntityLockable implements ITickable, ISidedInventory, net.minecraftforge.items.IFurnace
 {
     private static final int[] SLOTS_TOP = new int[] {0};
     private static final int[] SLOTS_BOTTOM = new int[] {2, 1};
@@ -43,6 +43,7 @@
     private int cookTime;
     private int totalCookTime;
     private String furnaceCustomName;
+    private float storageExp;
 
     @Override
     public int getSizeInventory()
@@ -130,11 +131,11 @@
         super.readFromNBT(compound);
         this.furnaceItemStacks = NonNullList.withSize(this.getSizeInventory(), ItemStack.EMPTY);
         ItemStackHelper.loadAllItems(compound, this.furnaceItemStacks);
-        this.furnaceBurnTime = compound.getShort("BurnTime");
-        this.cookTime = compound.getShort("CookTime");
-        this.totalCookTime = compound.getShort("CookTimeTotal");
+        this.furnaceBurnTime = compound.getInteger("BurnTime");
+        this.cookTime = compound.getInteger("CookTime");
+        this.totalCookTime = compound.getInteger("CookTimeTotal");
         this.currentItemBurnTime = getItemBurnTime(this.furnaceItemStacks.get(1));
-
+        this.storageExp = compound.getFloat("StorageExp");
         if (compound.hasKey("CustomName", 8))
         {
             this.furnaceCustomName = compound.getString("CustomName");
@@ -145,11 +146,11 @@
     public NBTTagCompound writeToNBT(NBTTagCompound compound)
     {
         super.writeToNBT(compound);
-        compound.setShort("BurnTime", (short)this.furnaceBurnTime);
-        compound.setShort("CookTime", (short)this.cookTime);
-        compound.setShort("CookTimeTotal", (short)this.totalCookTime);
+        compound.setInteger("BurnTime", this.furnaceBurnTime);
+        compound.setInteger("CookTime", this.cookTime);
+        compound.setInteger("CookTimeTotal", this.totalCookTime);
         ItemStackHelper.saveAllItems(compound, this.furnaceItemStacks);
-
+        compound.setFloat("StorageExp", this.storageExp);
         if (this.hasCustomName())
         {
             compound.setString("CustomName", this.furnaceCustomName);
@@ -208,8 +209,8 @@
 
                             if (itemstack.isEmpty())
                             {
-                                Item item1 = item.getContainerItem();
-                                this.furnaceItemStacks.set(1, item1 == null ? ItemStack.EMPTY : new ItemStack(item1));
+                                ItemStack item1 = item.getContainerItem(itemstack);
+                                this.furnaceItemStacks.set(1, item1);
                             }
                         }
                     }
@@ -250,9 +251,22 @@
         }
     }
 
+    public void addExp(float experience)
+    {
+        this.storageExp += experience;
+    }
+
+    @Override
+    public float extractExp()
+    {
+        float exp = this.storageExp;
+        this.storageExp = 0;
+        return exp;
+    }
+
     public int getCookTime(ItemStack stack)
     {
-        return 200;
+        return net.minecraftforge.common.crafting.FurnaceRecipeManager.instance().getCookTime(stack);
     }
 
     private boolean canSmelt()
@@ -281,11 +295,13 @@
                 {
                     return false;
                 }
+                else if (itemstack1.getCount() + itemstack.getCount() <= this.getInventoryStackLimit() && itemstack1.getCount() + itemstack.getCount() <= itemstack1.getMaxStackSize())  // Forge fix: make furnace respect stack sizes in furnace recipes
+                {
+                    return true;
+                }
                 else
                 {
-                    return itemstack1.getCount() < this.getInventoryStackLimit() && itemstack1.getCount() < itemstack1.getMaxStackSize()
-                           ? true
-                           : itemstack1.getCount() < itemstack.getMaxStackSize();
+                    return itemstack1.getCount() + itemstack.getCount() <= itemstack.getMaxStackSize(); // Forge fix: make furnace respect stack sizes in furnace recipes
                 }
             }
         }
@@ -305,7 +321,7 @@
             }
             else if (itemstack2.getItem() == itemstack1.getItem())
             {
-                itemstack2.grow(1);
+                itemstack2.grow(itemstack1.getCount());
             }
 
             if (itemstack.getItem() == Item.getItemFromBlock(Blocks.SPONGE)
@@ -315,7 +331,7 @@
             {
                 this.furnaceItemStacks.set(1, new ItemStack(Items.WATER_BUCKET));
             }
-
+            this.addExp(net.minecraftforge.common.crafting.FurnaceRecipeManager.instance().getExperience(itemstack));
             itemstack.shrink(1);
         }
     }
@@ -328,6 +344,8 @@
         }
         else
         {
+            int burnTime = net.minecraftforge.event.ForgeEventFactory.getItemBurnTime(stack);
+            if (burnTime >= 0) return burnTime;
             Item item = stack.getItem();
 
             if (item == Item.getItemFromBlock(Blocks.WOODEN_SLAB))
@@ -549,5 +567,24 @@
     public void clear()
     {
         this.furnaceItemStacks.clear();
+    }
+
+    net.minecraftforge.items.IItemHandler handlerTop = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.UP);
+    net.minecraftforge.items.IItemHandler handlerBottom = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.DOWN);
+    net.minecraftforge.items.IItemHandler handlerSide = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.WEST);
+
+    @SuppressWarnings("unchecked")
+    @Override
+    @javax.annotation.Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @javax.annotation.Nullable net.minecraft.util.EnumFacing facing)
+    {
+        if (facing != null && capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+            if (facing == EnumFacing.DOWN)
+                return (T) handlerBottom;
+            else if (facing == EnumFacing.UP)
+                return (T) handlerTop;
+            else
+                return (T) handlerSide;
+        return super.getCapability(capability, facing);
     }
 }
