--- before/net/minecraft/tileentity/TileEntityEnchantmentTable.java
+++ after/net/minecraft/tileentity/TileEntityEnchantmentTable.java
@@ -67,11 +67,15 @@
             {
                 float f1 = this.field_145932_k;
 
-                do
+                while (true)
                 {
                     this.field_145932_k += (float)(field_145923_r.nextInt(4) - field_145923_r.nextInt(4));
+
+                    if (f1 != this.field_145932_k)
+                    {
+                        break;
+                    }
                 }
-                while (f1 == this.field_145932_k);
             }
         }
         else
@@ -104,6 +108,7 @@
 
         for (f2 = this.field_145924_q - this.field_145928_o; f2 >= (float)Math.PI; f2 -= ((float)Math.PI * 2F))
         {
+            ;
         }
 
         while (f2 < -(float)Math.PI)
@@ -139,7 +144,7 @@
 
     public ITextComponent func_145748_c_()
     {
-        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_()));
+        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_(), new Object[0]));
     }
 
     public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
