--- before/net/minecraft/tileentity/TileEntityLockableLoot.java
+++ after/net/minecraft/tileentity/TileEntityLockableLoot.java
@@ -73,14 +73,13 @@
 
             if (p_184281_1_ != null)
             {
-                lootcontext$builder.func_186469_a(p_184281_1_.func_184817_da());
+                lootcontext$builder.func_186469_a(p_184281_1_.func_184817_da()).func_186470_a(p_184281_1_); // Forge: add player to LootContext
             }
 
             loottable.func_186460_a(this, random, lootcontext$builder.func_186471_a());
         }
     }
 
-    @Override
     public ResourceLocation func_184276_b()
     {
         return this.field_184284_m;
@@ -92,7 +91,6 @@
         this.field_184285_n = p_189404_2_;
     }
 
-    @Override
     public boolean func_145818_k_()
     {
         return this.field_190577_o != null && !this.field_190577_o.isEmpty();
@@ -103,17 +101,15 @@
         this.field_190577_o = p_190575_1_;
     }
 
-    @Override
     public ItemStack func_70301_a(int p_70301_1_)
     {
-        this.func_184281_d(null);
-        return this.func_190576_q().get(p_70301_1_);
+        this.func_184281_d((EntityPlayer)null);
+        return (ItemStack)this.func_190576_q().get(p_70301_1_);
     }
 
-    @Override
     public ItemStack func_70298_a(int p_70298_1_, int p_70298_2_)
     {
-        this.func_184281_d(null);
+        this.func_184281_d((EntityPlayer)null);
         ItemStack itemstack = ItemStackHelper.func_188382_a(this.func_190576_q(), p_70298_1_, p_70298_2_);
 
         if (!itemstack.func_190926_b())
@@ -124,17 +120,15 @@
         return itemstack;
     }
 
-    @Override
     public ItemStack func_70304_b(int p_70304_1_)
     {
-        this.func_184281_d(null);
+        this.func_184281_d((EntityPlayer)null);
         return ItemStackHelper.func_188383_a(this.func_190576_q(), p_70304_1_);
     }
 
-    @Override
     public void func_70299_a(int p_70299_1_, @Nullable ItemStack p_70299_2_)
     {
-        this.func_184281_d(null);
+        this.func_184281_d((EntityPlayer)null);
         this.func_190576_q().set(p_70299_1_, p_70299_2_);
 
         if (p_70299_2_.func_190916_E() > this.func_70297_j_())
@@ -145,7 +139,6 @@
         this.func_70296_d();
     }
 
-    @Override
     public boolean func_70300_a(EntityPlayer p_70300_1_)
     {
         if (this.field_145850_b.func_175625_s(this.field_174879_c) != this)
@@ -154,54 +147,40 @@
         }
         else
         {
-            return !(
-                       p_70300_1_.func_70092_e(
-                           (double)this.field_174879_c.func_177958_n() + 0.5,
-                           (double)this.field_174879_c.func_177956_o() + 0.5,
-                           (double)this.field_174879_c.func_177952_p() + 0.5
-                       )
-                       > 64.0
-                   );
+            return p_70300_1_.func_70092_e((double)this.field_174879_c.func_177958_n() + 0.5D, (double)this.field_174879_c.func_177956_o() + 0.5D, (double)this.field_174879_c.func_177952_p() + 0.5D) <= 64.0D;
         }
     }
 
-    @Override
     public void func_174889_b(EntityPlayer p_174889_1_)
     {
     }
 
-    @Override
     public void func_174886_c(EntityPlayer p_174886_1_)
     {
     }
 
-    @Override
     public boolean func_94041_b(int p_94041_1_, ItemStack p_94041_2_)
     {
         return true;
     }
 
-    @Override
     public int func_174887_a_(int p_174887_1_)
     {
         return 0;
     }
 
-    @Override
     public void func_174885_b(int p_174885_1_, int p_174885_2_)
     {
     }
 
-    @Override
     public int func_174890_g()
     {
         return 0;
     }
 
-    @Override
     public void func_174888_l()
     {
-        this.func_184281_d(null);
+        this.func_184281_d((EntityPlayer)null);
         this.func_190576_q().clear();
     }
 
