--- before/net/minecraft/tileentity/TileEntitySkull.java
+++ after/net/minecraft/tileentity/TileEntitySkull.java
@@ -156,7 +156,7 @@
                 }
                 else
                 {
-                    Property property = Iterables.getFirst(gameprofile.getProperties().get("textures"), (Property)null);
+                    Property property = (Property)Iterables.getFirst(gameprofile.getProperties().get("textures"), (Object)null);
 
                     if (property == null)
                     {
