--- before/net/minecraft/tileentity/TileEntityEnderChest.java
+++ after/net/minecraft/tileentity/TileEntityEnderChest.java
@@ -106,7 +106,7 @@
         }
         else
         {
-            return !(p_145971_1_.func_70092_e((double)this.field_174879_c.func_177958_n() + 0.5D, (double)this.field_174879_c.func_177956_o() + 0.5D, (double)this.field_174879_c.func_177952_p() + 0.5D) > 64.0D);
+            return p_145971_1_.func_70092_e((double)this.field_174879_c.func_177958_n() + 0.5D, (double)this.field_174879_c.func_177956_o() + 0.5D, (double)this.field_174879_c.func_177952_p() + 0.5D) <= 64.0D;
         }
     }
 }
