--- before/net/minecraft/tileentity/TileEntityChest.java
+++ after/net/minecraft/tileentity/TileEntityChest.java
@@ -25,7 +25,7 @@
 
 public class TileEntityChest extends TileEntityLockableLoot implements ITickable
 {
-    private NonNullList<ItemStack> field_145985_p = NonNullList.func_191197_a(27, ItemStack.field_190927_a);
+    private NonNullList<ItemStack> field_145985_p = NonNullList.<ItemStack>func_191197_a(27, ItemStack.field_190927_a);
     public boolean field_145984_a;
     public TileEntityChest field_145992_i;
     public TileEntityChest field_145990_j;
@@ -46,13 +46,11 @@
         this.field_145982_r = p_i46677_1_;
     }
 
-    @Override
     public int func_70302_i_()
     {
         return 27;
     }
 
-    @Override
     public boolean func_191420_l()
     {
         for (ItemStack itemstack : this.field_145985_p)
@@ -66,7 +64,6 @@
         return true;
     }
 
-    @Override
     public String func_70005_c_()
     {
         return this.func_145818_k_() ? this.field_190577_o : "container.chest";
@@ -74,14 +71,13 @@
 
     public static void func_189677_a(DataFixer p_189677_0_)
     {
-        p_189677_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityChest.class, "Items"));
+        p_189677_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityChest.class, new String[] {"Items"}));
     }
 
-    @Override
     public void func_145839_a(NBTTagCompound p_145839_1_)
     {
         super.func_145839_a(p_145839_1_);
-        this.field_145985_p = NonNullList.func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
+        this.field_145985_p = NonNullList.<ItemStack>func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
 
         if (!this.func_184283_b(p_145839_1_))
         {
@@ -94,7 +90,6 @@
         }
     }
 
-    @Override
     public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
     {
         super.func_189515_b(p_189515_1_);
@@ -112,17 +107,16 @@
         return p_189515_1_;
     }
 
-    @Override
     public int func_70297_j_()
     {
         return 64;
     }
 
-    @Override
     public void func_145836_u()
     {
         super.func_145836_u();
         this.field_145984_a = false;
+        doubleChestHandler = null;
     }
 
     @SuppressWarnings("incomplete-switch")
@@ -174,6 +168,7 @@
     {
         if (!this.field_145984_a)
         {
+            if (this.field_145850_b == null || !this.field_145850_b.func_175697_a(this.field_174879_c, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbors
             this.field_145984_a = true;
             this.field_145991_k = this.func_174911_a(EnumFacing.WEST);
             this.field_145990_j = this.func_174911_a(EnumFacing.EAST);
@@ -215,7 +210,6 @@
         }
     }
 
-    @Override
     public void func_73660_a()
     {
         this.func_145979_i();
@@ -229,18 +223,7 @@
             this.field_145987_o = 0;
             float f = 5.0F;
 
-            for (EntityPlayer entityplayer : this.field_145850_b
-                    .func_72872_a(
-                        EntityPlayer.class,
-                        new AxisAlignedBB(
-                            (double)((float)i - 5.0F),
-                            (double)((float)j - 5.0F),
-                            (double)((float)k - 5.0F),
-                            (double)((float)(i + 1) + 5.0F),
-                            (double)((float)(j + 1) + 5.0F),
-                            (double)((float)(k + 1) + 5.0F)
-                        )
-                    ))
+            for (EntityPlayer entityplayer : this.field_145850_b.func_72872_a(EntityPlayer.class, new AxisAlignedBB((double)((float)i - 5.0F), (double)((float)j - 5.0F), (double)((float)k - 5.0F), (double)((float)(i + 1) + 5.0F), (double)((float)(j + 1) + 5.0F), (double)((float)(k + 1) + 5.0F))))
             {
                 if (entityplayer.field_71070_bA instanceof ContainerChest)
                 {
@@ -259,30 +242,20 @@
 
         if (this.field_145987_o > 0 && this.field_145989_m == 0.0F && this.field_145992_i == null && this.field_145991_k == null)
         {
-            double d1 = (double)i + 0.5;
-            double d2 = (double)k + 0.5;
+            double d1 = (double)i + 0.5D;
+            double d2 = (double)k + 0.5D;
 
             if (this.field_145988_l != null)
             {
-                d2 += 0.5;
+                d2 += 0.5D;
             }
 
             if (this.field_145990_j != null)
             {
-                d1 += 0.5;
+                d1 += 0.5D;
             }
 
-            this.field_145850_b
-            .func_184148_a(
-                null,
-                d1,
-                (double)j + 0.5,
-                d2,
-                SoundEvents.field_187657_V,
-                SoundCategory.BLOCKS,
-                0.5F,
-                this.field_145850_b.field_73012_v.nextFloat() * 0.1F + 0.9F
-            );
+            this.field_145850_b.func_184148_a((EntityPlayer)null, d1, (double)j + 0.5D, d2, SoundEvents.field_187657_V, SoundCategory.BLOCKS, 0.5F, this.field_145850_b.field_73012_v.nextFloat() * 0.1F + 0.9F);
         }
 
         if (this.field_145987_o == 0 && this.field_145989_m > 0.0F || this.field_145987_o > 0 && this.field_145989_m < 1.0F)
@@ -307,30 +280,20 @@
 
             if (this.field_145989_m < 0.5F && f2 >= 0.5F && this.field_145992_i == null && this.field_145991_k == null)
             {
-                double d3 = (double)i + 0.5;
-                double d0 = (double)k + 0.5;
+                double d3 = (double)i + 0.5D;
+                double d0 = (double)k + 0.5D;
 
                 if (this.field_145988_l != null)
                 {
-                    d0 += 0.5;
+                    d0 += 0.5D;
                 }
 
                 if (this.field_145990_j != null)
                 {
-                    d3 += 0.5;
+                    d3 += 0.5D;
                 }
 
-                this.field_145850_b
-                .func_184148_a(
-                    null,
-                    d3,
-                    (double)j + 0.5,
-                    d0,
-                    SoundEvents.field_187651_T,
-                    SoundCategory.BLOCKS,
-                    0.5F,
-                    this.field_145850_b.field_73012_v.nextFloat() * 0.1F + 0.9F
-                );
+                this.field_145850_b.func_184148_a((EntityPlayer)null, d3, (double)j + 0.5D, d0, SoundEvents.field_187651_T, SoundCategory.BLOCKS, 0.5F, this.field_145850_b.field_73012_v.nextFloat() * 0.1F + 0.9F);
             }
 
             if (this.field_145989_m < 0.0F)
@@ -340,7 +303,6 @@
         }
     }
 
-    @Override
     public boolean func_145842_c(int p_145842_1_, int p_145842_2_)
     {
         if (p_145842_1_ == 1)
@@ -354,7 +316,6 @@
         }
     }
 
-    @Override
     public void func_174889_b(EntityPlayer p_174889_1_)
     {
         if (!p_174889_1_.func_175149_v())
@@ -375,7 +336,6 @@
         }
     }
 
-    @Override
     public void func_174886_c(EntityPlayer p_174886_1_)
     {
         if (!p_174886_1_.func_175149_v() && this.func_145838_q() instanceof BlockChest)
@@ -391,7 +351,29 @@
         }
     }
 
+    public net.minecraftforge.items.VanillaDoubleChestItemHandler doubleChestHandler;
+
+    @SuppressWarnings("unchecked")
     @Override
+    @Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable net.minecraft.util.EnumFacing facing)
+    {
+        if (capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+        {
+            if(doubleChestHandler == null || doubleChestHandler.needsRefresh())
+                doubleChestHandler = net.minecraftforge.items.VanillaDoubleChestItemHandler.get(this);
+            if (doubleChestHandler != null && doubleChestHandler != net.minecraftforge.items.VanillaDoubleChestItemHandler.NO_ADJACENT_CHESTS_INSTANCE)
+                return (T) doubleChestHandler;
+        }
+        return super.getCapability(capability, facing);
+    }
+
+    public net.minecraftforge.items.IItemHandler getSingleChestHandler()
+    {
+        return super.getCapability(net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, null);
+    }
+
+
     public void func_145843_s()
     {
         super.func_145843_s();
@@ -414,20 +396,17 @@
         return this.field_145982_r;
     }
 
-    @Override
     public String func_174875_k()
     {
         return "minecraft:chest";
     }
 
-    @Override
     public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
     {
         this.func_184281_d(p_174876_2_);
         return new ContainerChest(p_174876_1_, this, p_174876_2_);
     }
 
-    @Override
     protected NonNullList<ItemStack> func_190576_q()
     {
         return this.field_145985_p;
