--- before/net/minecraft/tileentity/TileEntitySign.java
+++ after/net/minecraft/tileentity/TileEntitySign.java
@@ -67,7 +67,7 @@
             @Override
             public boolean canUseCommand(int permLevel, String commandName)
             {
-                return true;
+                return permLevel <= 2; //Forge: Fixes  MC-75630 - Exploit with signs and command blocks
             }
             @Override
             public BlockPos getPosition()
