--- before/net/minecraft/tileentity/TileEntityFurnace.java
+++ after/net/minecraft/tileentity/TileEntityFurnace.java
@@ -34,23 +34,35 @@
 
 public class TileEntityFurnace extends TileEntityLockable implements ITickable, ISidedInventory
 {
+    public static final int FUEL = 1;
+    public static final int OUTPUT = 2;
+
+    protected ItemStack recipeKey = ItemStack.field_190927_a;
+    protected ItemStack recipeOutput = ItemStack.field_190927_a;
+    protected ItemStack failedMatch = ItemStack.field_190927_a;
+
+    @net.minecraftforge.fml.common.registry.GameRegistry.ItemStackHolder(value = "minecraft:sponge", meta = 1)
+    public static final ItemStack WET_SPONGE = ItemStack.field_190927_a;
+
     private static final int[] field_145962_k = new int[] {0};
     private static final int[] field_145959_l = new int[] {2, 1};
     private static final int[] field_145960_m = new int[] {1};
-    private NonNullList<ItemStack> field_145957_n = NonNullList.func_191197_a(3, ItemStack.field_190927_a);
+    private NonNullList<ItemStack> field_145957_n = NonNullList.<ItemStack>func_191197_a(3, ItemStack.field_190927_a);
     private int field_145956_a;
     private int field_145963_i;
     private int field_174906_k;
     private int field_174905_l;
     private String field_145958_o;
 
-    @Override
+    public TileEntityFurnace() {
+        this.field_174905_l = 200;
+    }
+
     public int func_70302_i_()
     {
         return this.field_145957_n.size();
     }
 
-    @Override
     public boolean func_191420_l()
     {
         for (ItemStack itemstack : this.field_145957_n)
@@ -64,25 +76,21 @@
         return true;
     }
 
-    @Override
     public ItemStack func_70301_a(int p_70301_1_)
     {
         return this.field_145957_n.get(p_70301_1_);
     }
 
-    @Override
     public ItemStack func_70298_a(int p_70298_1_, int p_70298_2_)
     {
         return ItemStackHelper.func_188382_a(this.field_145957_n, p_70298_1_, p_70298_2_);
     }
 
-    @Override
     public ItemStack func_70304_b(int p_70304_1_)
     {
         return ItemStackHelper.func_188383_a(this.field_145957_n, p_70304_1_);
     }
 
-    @Override
     public void func_70299_a(int p_70299_1_, ItemStack p_70299_2_)
     {
         ItemStack itemstack = this.field_145957_n.get(p_70299_1_);
@@ -102,13 +110,11 @@
         }
     }
 
-    @Override
     public String func_70005_c_()
     {
         return this.func_145818_k_() ? this.field_145958_o : "container.furnace";
     }
 
-    @Override
     public boolean func_145818_k_()
     {
         return this.field_145958_o != null && !this.field_145958_o.isEmpty();
@@ -121,18 +127,17 @@
 
     public static void func_189676_a(DataFixer p_189676_0_)
     {
-        p_189676_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityFurnace.class, "Items"));
+        p_189676_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityFurnace.class, new String[] {"Items"}));
     }
 
-    @Override
     public void func_145839_a(NBTTagCompound p_145839_1_)
     {
         super.func_145839_a(p_145839_1_);
-        this.field_145957_n = NonNullList.func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
+        this.field_145957_n = NonNullList.<ItemStack>func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
         ItemStackHelper.func_191283_b(p_145839_1_, this.field_145957_n);
-        this.field_145956_a = p_145839_1_.func_74765_d("BurnTime");
-        this.field_174906_k = p_145839_1_.func_74765_d("CookTime");
-        this.field_174905_l = p_145839_1_.func_74765_d("CookTimeTotal");
+        this.field_145956_a = p_145839_1_.func_74762_e("BurnTime");
+        this.field_174906_k = p_145839_1_.func_74762_e("CookTime");
+        this.field_174905_l = p_145839_1_.func_74762_e("CookTimeTotal");
         this.field_145963_i = func_145952_a(this.field_145957_n.get(1));
 
         if (p_145839_1_.func_150297_b("CustomName", 8))
@@ -141,13 +146,12 @@
         }
     }
 
-    @Override
     public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
     {
         super.func_189515_b(p_189515_1_);
-        p_189515_1_.func_74777_a("BurnTime", (short)this.field_145956_a);
-        p_189515_1_.func_74777_a("CookTime", (short)this.field_174906_k);
-        p_189515_1_.func_74777_a("CookTimeTotal", (short)this.field_174905_l);
+        p_189515_1_.func_74768_a("BurnTime", this.field_145956_a);
+        p_189515_1_.func_74768_a("CookTime", this.field_174906_k);
+        p_189515_1_.func_74768_a("CookTimeTotal", this.field_174905_l);
         ItemStackHelper.func_191282_a(p_189515_1_, this.field_145957_n);
 
         if (this.func_145818_k_())
@@ -158,7 +162,6 @@
         return p_189515_1_;
     }
 
-    @Override
     public int func_70297_j_()
     {
         return 64;
@@ -175,78 +178,55 @@
         return p_174903_0_.func_174887_a_(0) > 0;
     }
 
-    @Override
     public void func_73660_a()
     {
-        boolean flag = this.func_145950_i();
-        boolean flag1 = false;
-
-        if (this.func_145950_i())
-        {
-            --this.field_145956_a;
-        }
-
-        if (!this.field_145850_b.field_72995_K)
-        {
-            ItemStack itemstack = this.field_145957_n.get(1);
-
-            if (this.func_145950_i() || !itemstack.func_190926_b() && !this.field_145957_n.get(0).func_190926_b())
-            {
-                if (!this.func_145950_i() && this.func_145948_k())
-                {
-                    this.field_145956_a = func_145952_a(itemstack);
-                    this.field_145963_i = this.field_145956_a;
-
-                    if (this.func_145950_i())
-                    {
-                        flag1 = true;
-
-                        if (!itemstack.func_190926_b())
-                        {
-                            Item item = itemstack.func_77973_b();
-                            itemstack.func_190918_g(1);
-
-                            if (itemstack.func_190926_b())
-                            {
-                                Item item1 = item.func_77668_q();
-                                this.field_145957_n.set(1, item1 == null ? ItemStack.field_190927_a : new ItemStack(item1));
-                            }
-                        }
-                    }
-                }
-
-                if (this.func_145950_i() && this.func_145948_k())
-                {
-                    ++this.field_174906_k;
-
-                    if (this.field_174906_k == this.field_174905_l)
-                    {
-                        this.field_174906_k = 0;
-                        this.field_174905_l = this.func_174904_a(this.field_145957_n.get(0));
-                        this.func_145949_j();
-                        flag1 = true;
-                    }
-                }
-                else
-                {
-                    this.field_174906_k = 0;
-                }
-            }
-            else if (!this.func_145950_i() && this.field_174906_k > 0)
-            {
-                this.field_174906_k = MathHelper.func_76125_a(this.field_174906_k - 2, 0, this.field_174905_l);
-            }
-
-            if (flag != this.func_145950_i())
-            {
-                flag1 = true;
-                BlockFurnace.func_176446_a(this.func_145950_i(), this.field_145850_b, this.field_174879_c);
-            }
-        }
-
-        if (flag1)
-        {
-            this.func_70296_d();
+        if (field_145850_b.field_72995_K && func_145950_i()) {
+            field_145956_a--;
+            return;
+        } else if (field_145850_b.field_72995_K) return;
+
+        ItemStack fuel = ItemStack.field_190927_a;
+        boolean canSmelt = func_145948_k();
+
+        if (!this.func_145950_i() && !(fuel = field_145957_n.get(FUEL)).func_190926_b()) {
+            if (canSmelt) burnFuel(fuel, false);
+        }
+
+        boolean wasBurning = func_145950_i();
+
+        if (this.func_145950_i()) {
+            field_145956_a--;
+            if (canSmelt) smelt();
+            else field_174906_k = 0;
+        }
+
+        if (!this.func_145950_i()) {
+            if (!(fuel = field_145957_n.get(FUEL)).func_190926_b()) {
+                if (func_145948_k()) burnFuel(fuel, wasBurning);
+            } else if (field_174906_k > 0) {
+                field_174906_k = MathHelper.func_76125_a(field_174906_k - 2, 0, field_174905_l);
+            }
+        }
+
+        if (wasBurning && !func_145950_i()) BlockFurnace.func_176446_a(false, field_145850_b, field_174879_c);
+    }
+
+    protected void smelt() {
+        field_174906_k++;
+        if (this.field_174906_k == this.field_174905_l) {
+            this.field_174906_k = 0;
+            this.field_174905_l = this.func_174904_a(this.field_145957_n.getFirst());
+            this.func_145949_j();
+        }
+    }
+
+    protected void burnFuel(ItemStack fuel, boolean burnedThisTick) {
+        field_145963_i = (field_145956_a = func_145952_a(fuel));
+        if (this.func_145950_i()) {
+            Item item = fuel.func_77973_b();
+            fuel.func_190918_g(1);
+            if (fuel.func_190926_b()) field_145957_n.set(FUEL, item.getContainerItem(fuel));
+            if (!burnedThisTick) BlockFurnace.func_176446_a(true, field_145850_b, field_174879_c);
         }
     }
 
@@ -257,69 +237,55 @@
 
     private boolean func_145948_k()
     {
-        if (this.field_145957_n.get(0).func_190926_b())
-        {
-            return false;
-        }
-        else
-        {
-            ItemStack itemstack = FurnaceRecipes.func_77602_a().func_151395_a(this.field_145957_n.get(0));
+        ItemStack input = field_145957_n.getFirst();
+        ItemStack output = field_145957_n.get(OUTPUT);
+        if (input.func_190926_b() || input == failedMatch) return false;
 
-            if (itemstack.func_190926_b())
-            {
+        if (recipeKey.func_190926_b() || !net.minecraftforge.oredict.OreDictionary.itemMatches(recipeKey, input, false)) {
+            boolean matched = false;
+            for (java.util.Map.Entry<ItemStack, ItemStack> e : FurnaceRecipes.func_77602_a().func_77599_b().entrySet()) {
+                if (net.minecraftforge.oredict.OreDictionary.itemMatches(e.getKey(), input, false)) {
+                    recipeKey = e.getKey();
+                    recipeOutput = e.getValue();
+                    matched = true;
+                    failedMatch = ItemStack.field_190927_a;
+                    break;
+                }
+            }
+            if (!matched) {
+                recipeKey = ItemStack.field_190927_a;
+                recipeOutput = ItemStack.field_190927_a;
+                failedMatch = input;
                 return false;
             }
-            else
-            {
-                ItemStack itemstack1 = this.field_145957_n.get(2);
-
-                if (itemstack1.func_190926_b())
-                {
-                    return true;
-                }
-                else if (!itemstack1.func_77969_a(itemstack))
-                {
-                    return false;
-                }
-                else if (itemstack1.func_190916_E() < this.func_70297_j_() && itemstack1.func_190916_E() < itemstack1.func_77976_d())
-                {
-                    return true;
-                }
-                else
-                {
-                    return itemstack1.func_190916_E() < itemstack.func_77976_d();
-                }
-            }
         }
+
+        return !recipeOutput.func_190926_b() && (output.func_190926_b() || (itemsMatch(recipeOutput, output) && (recipeOutput.func_190916_E() + output.func_190916_E() <= output.func_77976_d())));
+    }
+
+    boolean itemsMatch(ItemStack a, ItemStack b) {
+        return net.minecraftforge.items.ItemHandlerHelper.canItemStacksStack(a, b);
     }
 
     public void func_145949_j()
     {
-        if (this.func_145948_k())
-        {
-            ItemStack itemstack = this.field_145957_n.get(0);
-            ItemStack itemstack1 = FurnaceRecipes.func_77602_a().func_151395_a(itemstack);
-            ItemStack itemstack2 = this.field_145957_n.get(2);
-
-            if (itemstack2.func_190926_b())
-            {
-                this.field_145957_n.set(2, itemstack1.func_77946_l());
-            }
-            else if (itemstack2.func_77973_b() == itemstack1.func_77973_b())
-            {
-                itemstack2.func_190917_f(1);
-            }
-
-            if (itemstack.func_77973_b() == Item.func_150898_a(Blocks.field_150360_v)
-                    && itemstack.func_77960_j() == 1
-                    && !this.field_145957_n.get(1).func_190926_b()
-                    && this.field_145957_n.get(1).func_77973_b() == Items.field_151133_ar)
-            {
-                this.field_145957_n.set(1, new ItemStack(Items.field_151131_as));
-            }
-
-            itemstack.func_190918_g(1);
-        }
+        ItemStack input = this.field_145957_n.getFirst();
+        ItemStack recipeOutput = FurnaceRecipes.func_77602_a().func_77599_b().get(recipeKey);
+        ItemStack output = this.field_145957_n.get(OUTPUT);
+
+        if (output.func_190926_b()) this.field_145957_n.set(OUTPUT, recipeOutput.func_77946_l());
+        else if (itemsMatch(output, recipeOutput)) output.func_190917_f(recipeOutput.func_190916_E());
+
+        if (input.func_77969_a(WET_SPONGE) && this.field_145957_n.get(FUEL).func_77973_b() == Items.field_151133_ar) this.field_145957_n.set(FUEL, new ItemStack(Items.field_151131_as));
+
+        input.func_190918_g(1);
+    }
+
+    @Override
+    public boolean shouldRefresh(net.minecraft.world.World world, net.minecraft.util.math.BlockPos pos, net.minecraft.block.state.IBlockState oldState, net.minecraft.block.state.IBlockState newState) {
+        if (oldState.func_177230_c() == Blocks.field_150460_al && newState.func_177230_c() == Blocks.field_150470_am) return false;
+        else if (oldState.func_177230_c() == Blocks.field_150470_am && newState.func_177230_c() == Blocks.field_150460_al) return false;
+        else return oldState.func_177230_c() != newState.func_177230_c();
     }
 
     public static int func_145952_a(ItemStack p_145952_0_)
@@ -330,6 +296,8 @@
         }
         else
         {
+            int burnTime = net.minecraftforge.event.ForgeEventFactory.getItemBurnTime(p_145952_0_);
+            if (burnTime >= 0) return burnTime;
             Item item = p_145952_0_.func_77973_b();
 
             if (item == Item.func_150898_a(Blocks.field_150376_bx))
@@ -376,37 +344,43 @@
             {
                 return 100;
             }
-            else if (item == Items.field_151031_f || item == Items.field_151112_aM)
-            {
-                return 300;
-            }
-            else if (item == Items.field_151155_ap)
-            {
-                return 200;
-            }
-            else if (item == Items.field_151044_h)
-            {
-                return 1600;
-            }
-            else if (item == Items.field_151129_at)
-            {
-                return 20000;
-            }
-            else if (item == Item.func_150898_a(Blocks.field_150345_g) || item == Items.field_151054_z)
-            {
-                return 100;
-            }
-            else if (item == Items.field_151072_bj)
-            {
-                return 2400;
-            }
-            else if (item instanceof ItemDoor && item != Items.field_151139_aw)
-            {
-                return 200;
+            else if (item != Items.field_151031_f && item != Items.field_151112_aM)
+            {
+                if (item == Items.field_151155_ap)
+                {
+                    return 200;
+                }
+                else if (item == Items.field_151044_h)
+                {
+                    return 1600;
+                }
+                else if (item == Items.field_151129_at)
+                {
+                    return 20000;
+                }
+                else if (item != Item.func_150898_a(Blocks.field_150345_g) && item != Items.field_151054_z)
+                {
+                    if (item == Items.field_151072_bj)
+                    {
+                        return 2400;
+                    }
+                    else if (item instanceof ItemDoor && item != Items.field_151139_aw)
+                    {
+                        return 200;
+                    }
+                    else
+                    {
+                        return item instanceof ItemBoat ? 400 : 0;
+                    }
+                }
+                else
+                {
+                    return 100;
+                }
             }
             else
             {
-                return item instanceof ItemBoat ? 400 : 0;
+                return 300;
             }
         }
     }
@@ -416,7 +390,6 @@
         return func_145952_a(p_145954_0_) > 0;
     }
 
-    @Override
     public boolean func_70300_a(EntityPlayer p_70300_1_)
     {
         if (this.field_145850_b.func_175625_s(this.field_174879_c) != this)
@@ -425,28 +398,18 @@
         }
         else
         {
-            return !(
-                       p_70300_1_.func_70092_e(
-                           (double)this.field_174879_c.func_177958_n() + 0.5,
-                           (double)this.field_174879_c.func_177956_o() + 0.5,
-                           (double)this.field_174879_c.func_177952_p() + 0.5
-                       )
-                       > 64.0
-                   );
+            return p_70300_1_.func_70092_e((double)this.field_174879_c.func_177958_n() + 0.5D, (double)this.field_174879_c.func_177956_o() + 0.5D, (double)this.field_174879_c.func_177952_p() + 0.5D) <= 64.0D;
         }
     }
 
-    @Override
     public void func_174889_b(EntityPlayer p_174889_1_)
     {
     }
 
-    @Override
     public void func_174886_c(EntityPlayer p_174886_1_)
     {
     }
 
-    @Override
     public boolean func_94041_b(int p_94041_1_, ItemStack p_94041_2_)
     {
         if (p_94041_1_ == 2)
@@ -464,7 +427,6 @@
         }
     }
 
-    @Override
     public int[] func_180463_a(EnumFacing p_180463_1_)
     {
         if (p_180463_1_ == EnumFacing.DOWN)
@@ -477,13 +439,11 @@
         }
     }
 
-    @Override
     public boolean func_180462_a(int p_180462_1_, ItemStack p_180462_2_, EnumFacing p_180462_3_)
     {
         return this.func_94041_b(p_180462_1_, p_180462_2_);
     }
 
-    @Override
     public boolean func_180461_b(int p_180461_1_, ItemStack p_180461_2_, EnumFacing p_180461_3_)
     {
         if (p_180461_3_ == EnumFacing.DOWN && p_180461_1_ == 1)
@@ -499,19 +459,16 @@
         return true;
     }
 
-    @Override
     public String func_174875_k()
     {
         return "minecraft:furnace";
     }
 
-    @Override
     public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
     {
         return new ContainerFurnace(p_174876_1_, this);
     }
 
-    @Override
     public int func_174887_a_(int p_174887_1_)
     {
         switch (p_174887_1_)
@@ -529,7 +486,6 @@
         }
     }
 
-    @Override
     public void func_174885_b(int p_174885_1_, int p_174885_2_)
     {
         switch (p_174885_1_)
@@ -548,15 +504,32 @@
         }
     }
 
-    @Override
     public int func_174890_g()
     {
         return 4;
     }
 
-    @Override
     public void func_174888_l()
     {
         this.field_145957_n.clear();
+    }
+
+    net.minecraftforge.items.IItemHandler handlerTop = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.UP);
+    net.minecraftforge.items.IItemHandler handlerBottom = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.DOWN);
+    net.minecraftforge.items.IItemHandler handlerSide = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.WEST);
+
+    @SuppressWarnings("unchecked")
+    @Override
+    @javax.annotation.Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @javax.annotation.Nullable net.minecraft.util.EnumFacing facing)
+    {
+        if (facing != null && capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+            if (facing == EnumFacing.DOWN)
+                return (T) handlerBottom;
+            else if (facing == EnumFacing.UP)
+                return (T) handlerTop;
+            else
+                return (T) handlerSide;
+        return super.getCapability(capability, facing);
     }
 }
