--- before/net/minecraft/tileentity/TileEntityLockable.java
+++ after/net/minecraft/tileentity/TileEntityLockable.java
@@ -11,14 +11,12 @@
 {
     private LockCode field_174901_a = LockCode.field_180162_a;
 
-    @Override
     public void func_145839_a(NBTTagCompound p_145839_1_)
     {
         super.func_145839_a(p_145839_1_);
         this.field_174901_a = LockCode.func_180158_b(p_145839_1_);
     }
 
-    @Override
     public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
     {
         super.func_189515_b(p_189515_1_);
@@ -31,27 +29,46 @@
         return p_189515_1_;
     }
 
-    @Override
     public boolean func_174893_q_()
     {
         return this.field_174901_a != null && !this.field_174901_a.func_180160_a();
     }
 
-    @Override
     public LockCode func_174891_i()
     {
         return this.field_174901_a;
     }
 
-    @Override
     public void func_174892_a(LockCode p_174892_1_)
     {
         this.field_174901_a = p_174892_1_;
     }
 
-    @Override
     public ITextComponent func_145748_c_()
     {
-        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_()));
+        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_(), new Object[0]));
+    }
+
+    private net.minecraftforge.items.IItemHandler itemHandler;
+
+    protected net.minecraftforge.items.IItemHandler createUnSidedHandler()
+    {
+        return new net.minecraftforge.items.wrapper.InvWrapper(this);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    @javax.annotation.Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @javax.annotation.Nullable net.minecraft.util.EnumFacing facing)
+    {
+        if (capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+            return (T) (itemHandler == null ? (itemHandler = createUnSidedHandler()) : itemHandler);
+        return super.getCapability(capability, facing);
+    }
+
+    @Override
+    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, @javax.annotation.Nullable net.minecraft.util.EnumFacing facing)
+    {
+        return capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY || super.hasCapability(capability, facing);
     }
 }
