--- before/net/minecraft/tileentity/TileEntityHopper.java
+++ after/net/minecraft/tileentity/TileEntityHopper.java
@@ -88,7 +88,8 @@
     public ItemStack decrStackSize(int index, int count)
     {
         this.fillWithLoot(null);
-        return ItemStackHelper.getAndSplit(this.getItems(), index, count);
+        ItemStack itemstack = ItemStackHelper.getAndSplit(this.getItems(), index, count);
+        return itemstack;
     }
 
     @Override
@@ -199,6 +200,7 @@
 
     private boolean transferItemsOut()
     {
+        if (net.minecraftforge.items.VanillaInventoryCodeHooks.insertHook(this)) { return true; }
         IInventory iinventory = this.getInventoryForHopperTransfer();
 
         if (iinventory == null)
@@ -305,6 +307,8 @@
 
     public static boolean pullItems(IHopper hopper)
     {
+        Boolean ret = net.minecraftforge.items.VanillaInventoryCodeHooks.extractHook(hopper);
+        if (ret != null) return ret;
         IInventory iinventory = getSourceInventory(hopper);
 
         if (iinventory != null)
@@ -406,6 +410,12 @@
         }
     }
 
+
+    protected net.minecraftforge.items.IItemHandler createUnSidedHandler()
+    {
+        return new net.minecraftforge.items.VanillaHopperItemHandler(this);
+    }
+
     public static ItemStack putStackInInventoryAllSlots(IInventory source, IInventory destination, ItemStack stack, @Nullable EnumFacing direction)
     {
         if (destination instanceof ISidedInventory && direction != null)
@@ -530,9 +540,10 @@
         int j = MathHelper.floor(y);
         int k = MathHelper.floor(z);
         BlockPos blockpos = new BlockPos(i, j, k);
-        Block block = worldIn.getBlockState(blockpos).getBlock();
+        net.minecraft.block.state.IBlockState state = worldIn.getBlockState(blockpos);
+        Block block = state.getBlock();
 
-        if (block.hasTileEntity())
+        if (block.hasTileEntity(state))
         {
             TileEntity tileentity = worldIn.getTileEntity(blockpos);
 
@@ -631,4 +642,6 @@
     {
         return this.inventory;
     }
+
+    public long getLastUpdateTime() { return tickedGameTime; } // Forge
 }
