--- before/net/minecraft/tileentity/TileEntityStructure.java
+++ after/net/minecraft/tileentity/TileEntityStructure.java
@@ -94,7 +94,7 @@
         {
             this.field_184426_k = Rotation.valueOf(p_145839_1_.func_74779_i("rotation"));
         }
-        catch (IllegalArgumentException illegalargumentexception2)
+        catch (IllegalArgumentException var11)
         {
             this.field_184426_k = Rotation.NONE;
         }
@@ -103,7 +103,7 @@
         {
             this.field_184425_j = Mirror.valueOf(p_145839_1_.func_74779_i("mirror"));
         }
-        catch (IllegalArgumentException illegalargumentexception1)
+        catch (IllegalArgumentException var10)
         {
             this.field_184425_j = Mirror.NONE;
         }
@@ -112,7 +112,7 @@
         {
             this.field_184427_l = TileEntityStructure.Mode.valueOf(p_145839_1_.func_74779_i("mode"));
         }
-        catch (IllegalArgumentException illegalargumentexception)
+        catch (IllegalArgumentException var9)
         {
             this.field_184427_l = TileEntityStructure.Mode.DATA;
         }
@@ -235,28 +235,28 @@
         this.field_184425_j = p_184411_1_;
     }
 
-    @SideOnly(Side.CLIENT)
-    public Rotation func_189726_i()
-    {
-        return this.field_184426_k;
-    }
-
     public void func_184408_a(Rotation p_184408_1_)
     {
         this.field_184426_k = p_184408_1_;
     }
 
-    @SideOnly(Side.CLIENT)
-    public String func_189708_j()
-    {
-        return this.field_184422_g;
-    }
-
     public void func_184410_b(String p_184410_1_)
     {
         this.field_184422_g = p_184410_1_;
     }
 
+    @SideOnly(Side.CLIENT)
+    public Rotation func_189726_i()
+    {
+        return this.field_184426_k;
+    }
+
+    @SideOnly(Side.CLIENT)
+    public String func_189708_j()
+    {
+        return this.field_184422_g;
+    }
+
     public TileEntityStructure.Mode func_189700_k()
     {
         return this.field_184427_l;
@@ -273,6 +273,21 @@
         }
     }
 
+    public void func_184406_a(boolean p_184406_1_)
+    {
+        this.field_184428_m = p_184406_1_;
+    }
+
+    public void func_189718_a(float p_189718_1_)
+    {
+        this.field_189730_q = p_189718_1_;
+    }
+
+    public void func_189725_a(long p_189725_1_)
+    {
+        this.field_189731_r = p_189725_1_;
+    }
+
     @SideOnly(Side.CLIENT)
     public void func_189724_l()
     {
@@ -298,33 +313,18 @@
         return this.field_184428_m;
     }
 
-    public void func_184406_a(boolean p_184406_1_)
-    {
-        this.field_184428_m = p_184406_1_;
-    }
-
     @SideOnly(Side.CLIENT)
     public float func_189702_n()
     {
         return this.field_189730_q;
     }
 
-    public void func_189718_a(float p_189718_1_)
-    {
-        this.field_189730_q = p_189718_1_;
-    }
-
     @SideOnly(Side.CLIENT)
     public long func_189719_o()
     {
         return this.field_189731_r;
     }
 
-    public void func_189725_a(long p_189725_1_)
-    {
-        this.field_189731_r = p_189725_1_;
-    }
-
     public boolean func_184417_l()
     {
         if (this.field_184427_l != TileEntityStructure.Mode.SAVE)
@@ -379,7 +379,7 @@
 
     private List<TileEntityStructure> func_184418_a(BlockPos p_184418_1_, BlockPos p_184418_2_)
     {
-        List<TileEntityStructure> list = Lists.newArrayList();
+        List<TileEntityStructure> list = Lists.<TileEntityStructure>newArrayList();
 
         for (BlockPos.MutableBlockPos blockpos$mutableblockpos : BlockPos.func_177975_b(p_184418_1_, p_184418_2_))
         {
@@ -405,7 +405,7 @@
 
         if (p_184416_2_.size() > 1)
         {
-            BlockPos blockpos = p_184416_2_.get(0).func_174877_v();
+            BlockPos blockpos = ((TileEntityStructure)p_184416_2_.get(0)).func_174877_v();
             structureboundingbox = new StructureBoundingBox(blockpos, blockpos);
         }
         else
@@ -528,7 +528,7 @@
 
                     if (this.field_189730_q < 1.0F)
                     {
-                        placementsettings.func_189946_a(MathHelper.func_76131_a(this.field_189730_q, 0.0F, 1.0F)).func_189949_a(this.field_189731_r);
+                        placementsettings.func_189946_a(MathHelper.func_76131_a(this.field_189730_q, 0.0F, 1.0F)).func_189949_a(Long.valueOf(this.field_189731_r));
                     }
 
                     template.func_186260_a(this.field_145850_b, blockpos1, placementsettings);
@@ -599,7 +599,7 @@
     @Nullable
     public ITextComponent func_145748_c_()
     {
-        return new TextComponentTranslation("structure_block.hover." + this.field_184427_l.field_185116_f, this.field_184427_l == TileEntityStructure.Mode.DATA ? this.field_184422_g : this.field_184420_a);
+        return new TextComponentTranslation("structure_block.hover." + this.field_184427_l.field_185116_f, new Object[] {this.field_184427_l == TileEntityStructure.Mode.DATA ? this.field_184422_g : this.field_184420_a});
     }
 
     public static enum Mode implements IStringSerializable
