--- before/net/minecraft/tileentity/TileEntityShulkerBox.java
+++ after/net/minecraft/tileentity/TileEntityShulkerBox.java
@@ -33,10 +33,10 @@
 public class TileEntityShulkerBox extends TileEntityLockableLoot implements ITickable, ISidedInventory
 {
     private static final int[] field_190595_a = new int[27];
-    private NonNullList<ItemStack> field_190596_f = NonNullList.func_191197_a(27, ItemStack.field_190927_a);
+    private NonNullList<ItemStack> field_190596_f;
     private boolean field_190597_g;
     private int field_190598_h;
-    private TileEntityShulkerBox.AnimationStatus field_190599_i = TileEntityShulkerBox.AnimationStatus.CLOSED;
+    private TileEntityShulkerBox.AnimationStatus field_190599_i;
     private float field_190600_j;
     private float field_190601_k;
     private EnumDyeColor field_190602_l;
@@ -49,6 +49,8 @@
 
     public TileEntityShulkerBox(@Nullable EnumDyeColor p_i47242_1_)
     {
+        this.field_190596_f = NonNullList.<ItemStack>func_191197_a(27, ItemStack.field_190927_a);
+        this.field_190599_i = TileEntityShulkerBox.AnimationStatus.CLOSED;
         this.field_190602_l = p_i47242_1_;
     }
 
@@ -104,7 +106,7 @@
 
     public AxisAlignedBB func_190584_a(IBlockState p_190584_1_)
     {
-        return this.func_190587_b(p_190584_1_.func_177229_b(BlockShulkerBox.field_190957_a));
+        return this.func_190587_b((EnumFacing)p_190584_1_.func_177229_b(BlockShulkerBox.field_190957_a));
     }
 
     public AxisAlignedBB func_190587_b(EnumFacing p_190587_1_)
@@ -124,7 +126,7 @@
 
         if (iblockstate.func_177230_c() instanceof BlockShulkerBox)
         {
-            EnumFacing enumfacing = iblockstate.func_177229_b(BlockShulkerBox.field_190957_a);
+            EnumFacing enumfacing = (EnumFacing)iblockstate.func_177229_b(BlockShulkerBox.field_190957_a);
             AxisAlignedBB axisalignedbb = this.func_190588_c(enumfacing).func_186670_a(this.field_174879_c);
             List<Entity> list = this.field_145850_b.func_72839_b((Entity)null, axisalignedbb);
 
@@ -154,7 +156,7 @@
                                     d0 = axisalignedbb1.field_72336_d - axisalignedbb.field_72340_a;
                                 }
 
-                                d0 += 0.01D;
+                                d0 = d0 + 0.01D;
                                 break;
                             case Y:
 
@@ -167,7 +169,7 @@
                                     d1 = axisalignedbb1.field_72337_e - axisalignedbb.field_72338_b;
                                 }
 
-                                d1 += 0.01D;
+                                d1 = d1 + 0.01D;
                                 break;
                             case Z:
 
@@ -180,7 +182,7 @@
                                     d2 = axisalignedbb1.field_72334_f - axisalignedbb.field_72339_c;
                                 }
 
-                                d2 += 0.01D;
+                                d2 = d2 + 0.01D;
                         }
 
                         entity.func_70091_d(MoverType.SHULKER_BOX, d0 * (double)enumfacing.func_82601_c(), d1 * (double)enumfacing.func_96559_d(), d2 * (double)enumfacing.func_82599_e());
@@ -274,7 +276,7 @@
 
     public static void func_190593_a(DataFixer p_190593_0_)
     {
-        p_190593_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityShulkerBox.class, "Items"));
+        p_190593_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityShulkerBox.class, new String[] {"Items"}));
     }
 
     public void func_145839_a(NBTTagCompound p_145839_1_)
@@ -291,7 +293,7 @@
 
     public void func_190586_e(NBTTagCompound p_190586_1_)
     {
-        this.field_190596_f = NonNullList.func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
+        this.field_190596_f = NonNullList.<ItemStack>func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
 
         if (!this.func_184283_b(p_190586_1_) && p_190586_1_.func_150297_b("Items", 9))
         {
@@ -409,6 +411,7 @@
     {
         for (int i = 0; i < field_190595_a.length; field_190595_a[i] = i++)
         {
+            ;
         }
     }
 
@@ -418,5 +421,10 @@
         OPENING,
         OPENED,
         CLOSING;
+    }
+
+    protected net.minecraftforge.items.IItemHandler createUnSidedHandler()
+    {
+        return new net.minecraftforge.items.wrapper.SidedInvWrapper(this, EnumFacing.UP);
     }
 }
