--- before/net/minecraft/tileentity/TileEntitySign.java
+++ after/net/minecraft/tileentity/TileEntitySign.java
@@ -22,21 +22,17 @@
 
 public class TileEntitySign extends TileEntity
 {
-    public final ITextComponent[] field_145915_a = new ITextComponent[]
-    {
-        new TextComponentString(""), new TextComponentString(""), new TextComponentString(""), new TextComponentString("")
-    };
+    public final ITextComponent[] field_145915_a = new ITextComponent[] {new TextComponentString(""), new TextComponentString(""), new TextComponentString(""), new TextComponentString("")};
     public int field_145918_i = -1;
     private boolean field_145916_j = true;
     private EntityPlayer field_145917_k;
     private final CommandResultStats field_174883_i = new CommandResultStats();
 
-    @Override
     public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
     {
         super.func_189515_b(p_189515_1_);
 
-        for (int i = 0; i < 4; i++)
+        for (int i = 0; i < 4; ++i)
         {
             String s = ITextComponent.Serializer.func_150696_a(this.field_145915_a[i]);
             p_189515_1_.func_74778_a("Text" + (i + 1), s);
@@ -46,65 +42,53 @@
         return p_189515_1_;
     }
 
-    @Override
     protected void func_190201_b(World p_190201_1_)
     {
         this.func_145834_a(p_190201_1_);
     }
 
-    @Override
     public void func_145839_a(NBTTagCompound p_145839_1_)
     {
         this.field_145916_j = false;
         super.func_145839_a(p_145839_1_);
         ICommandSender icommandsender = new ICommandSender()
         {
-            @Override
             public String func_70005_c_()
             {
                 return "Sign";
             }
-            @Override
             public boolean func_70003_b(int p_70003_1_, String p_70003_2_)
             {
-                return true;
+                return p_70003_1_ <= 2; //Forge: Fixes  MC-75630 - Exploit with signs and command blocks
             }
-            @Override
             public BlockPos func_180425_c()
             {
                 return TileEntitySign.this.field_174879_c;
             }
-            @Override
             public Vec3d func_174791_d()
             {
-                return new Vec3d(
-                           (double)TileEntitySign.this.field_174879_c.func_177958_n() + 0.5,
-                           (double)TileEntitySign.this.field_174879_c.func_177956_o() + 0.5,
-                           (double)TileEntitySign.this.field_174879_c.func_177952_p() + 0.5
-                       );
+                return new Vec3d((double)TileEntitySign.this.field_174879_c.func_177958_n() + 0.5D, (double)TileEntitySign.this.field_174879_c.func_177956_o() + 0.5D, (double)TileEntitySign.this.field_174879_c.func_177952_p() + 0.5D);
             }
-            @Override
             public World func_130014_f_()
             {
                 return TileEntitySign.this.field_145850_b;
             }
-            @Override
             public MinecraftServer func_184102_h()
             {
                 return TileEntitySign.this.field_145850_b.func_73046_m();
             }
         };
 
-        for (int i = 0; i < 4; i++)
+        for (int i = 0; i < 4; ++i)
         {
             String s = p_145839_1_.func_74779_i("Text" + (i + 1));
             ITextComponent itextcomponent = ITextComponent.Serializer.func_150699_a(s);
 
             try
             {
-                this.field_145915_a[i] = TextComponentUtils.func_179985_a(icommandsender, itextcomponent, null);
+                this.field_145915_a[i] = TextComponentUtils.func_179985_a(icommandsender, itextcomponent, (Entity)null);
             }
-            catch (CommandException commandexception)
+            catch (CommandException var7)
             {
                 this.field_145915_a[i] = itextcomponent;
             }
@@ -114,19 +98,16 @@
     }
 
     @Nullable
-    @Override
     public SPacketUpdateTileEntity func_189518_D_()
     {
         return new SPacketUpdateTileEntity(this.field_174879_c, 9, this.func_189517_E_());
     }
 
-    @Override
     public NBTTagCompound func_189517_E_()
     {
         return this.func_189515_b(new NBTTagCompound());
     }
 
-    @Override
     public boolean func_183000_F()
     {
         return true;
@@ -162,55 +143,41 @@
     {
         ICommandSender icommandsender = new ICommandSender()
         {
-            @Override
             public String func_70005_c_()
             {
                 return p_174882_1_.func_70005_c_();
             }
-            @Override
             public ITextComponent func_145748_c_()
             {
                 return p_174882_1_.func_145748_c_();
             }
-            @Override
             public void func_145747_a(ITextComponent p_145747_1_)
             {
             }
-            @Override
             public boolean func_70003_b(int p_70003_1_, String p_70003_2_)
             {
                 return p_70003_1_ <= 2;
             }
-            @Override
             public BlockPos func_180425_c()
             {
                 return TileEntitySign.this.field_174879_c;
             }
-            @Override
             public Vec3d func_174791_d()
             {
-                return new Vec3d(
-                           (double)TileEntitySign.this.field_174879_c.func_177958_n() + 0.5,
-                           (double)TileEntitySign.this.field_174879_c.func_177956_o() + 0.5,
-                           (double)TileEntitySign.this.field_174879_c.func_177952_p() + 0.5
-                       );
+                return new Vec3d((double)TileEntitySign.this.field_174879_c.func_177958_n() + 0.5D, (double)TileEntitySign.this.field_174879_c.func_177956_o() + 0.5D, (double)TileEntitySign.this.field_174879_c.func_177952_p() + 0.5D);
             }
-            @Override
             public World func_130014_f_()
             {
                 return p_174882_1_.func_130014_f_();
             }
-            @Override
             public Entity func_174793_f()
             {
                 return p_174882_1_;
             }
-            @Override
             public boolean func_174792_t_()
             {
                 return false;
             }
-            @Override
             public void func_174794_a(CommandResultStats.Type p_174794_1_, int p_174794_2_)
             {
                 if (TileEntitySign.this.field_145850_b != null && !TileEntitySign.this.field_145850_b.field_72995_K)
@@ -218,7 +185,6 @@
                     TileEntitySign.this.field_174883_i.func_184932_a(TileEntitySign.this.field_145850_b.func_73046_m(), this, p_174794_1_, p_174794_2_);
                 }
             }
-            @Override
             public MinecraftServer func_184102_h()
             {
                 return p_174882_1_.func_184102_h();
