--- before/net/minecraft/tileentity/TileEntityShulkerBox.java
+++ after/net/minecraft/tileentity/TileEntityShulkerBox.java
@@ -33,10 +33,10 @@
 public class TileEntityShulkerBox extends TileEntityLockableLoot implements ITickable, ISidedInventory
 {
     private static final int[] field_190595_a = new int[27];
-    private NonNullList<ItemStack> field_190596_f = NonNullList.func_191197_a(27, ItemStack.field_190927_a);
+    private NonNullList<ItemStack> field_190596_f;
     private boolean field_190597_g;
     private int field_190598_h;
-    private TileEntityShulkerBox.AnimationStatus field_190599_i = TileEntityShulkerBox.AnimationStatus.CLOSED;
+    private TileEntityShulkerBox.AnimationStatus field_190599_i;
     private float field_190600_j;
     private float field_190601_k;
     private EnumDyeColor field_190602_l;
@@ -44,15 +44,16 @@
 
     public TileEntityShulkerBox()
     {
-        this(null);
+        this((EnumDyeColor)null);
     }
 
     public TileEntityShulkerBox(@Nullable EnumDyeColor p_i47242_1_)
     {
+        this.field_190596_f = NonNullList.<ItemStack>func_191197_a(27, ItemStack.field_190927_a);
+        this.field_190599_i = TileEntityShulkerBox.AnimationStatus.CLOSED;
         this.field_190602_l = p_i47242_1_;
     }
 
-    @Override
     public void func_73660_a()
     {
         this.func_190583_o();
@@ -105,24 +106,18 @@
 
     public AxisAlignedBB func_190584_a(IBlockState p_190584_1_)
     {
-        return this.func_190587_b(p_190584_1_.func_177229_b(BlockShulkerBox.field_190957_a));
+        return this.func_190587_b((EnumFacing)p_190584_1_.func_177229_b(BlockShulkerBox.field_190957_a));
     }
 
     public AxisAlignedBB func_190587_b(EnumFacing p_190587_1_)
     {
-        return Block.field_185505_j
-               .func_72321_a(
-                   (double)(0.5F * this.func_190585_a(1.0F) * (float)p_190587_1_.func_82601_c()),
-                   (double)(0.5F * this.func_190585_a(1.0F) * (float)p_190587_1_.func_96559_d()),
-                   (double)(0.5F * this.func_190585_a(1.0F) * (float)p_190587_1_.func_82599_e())
-               );
+        return Block.field_185505_j.func_72321_a((double)(0.5F * this.func_190585_a(1.0F) * (float)p_190587_1_.func_82601_c()), (double)(0.5F * this.func_190585_a(1.0F) * (float)p_190587_1_.func_96559_d()), (double)(0.5F * this.func_190585_a(1.0F) * (float)p_190587_1_.func_82599_e()));
     }
 
     private AxisAlignedBB func_190588_c(EnumFacing p_190588_1_)
     {
         EnumFacing enumfacing = p_190588_1_.func_176734_d();
-        return this.func_190587_b(p_190588_1_)
-               .func_191195_a((double)enumfacing.func_82601_c(), (double)enumfacing.func_96559_d(), (double)enumfacing.func_82599_e());
+        return this.func_190587_b(p_190588_1_).func_191195_a((double)enumfacing.func_82601_c(), (double)enumfacing.func_96559_d(), (double)enumfacing.func_82599_e());
     }
 
     private void func_190589_G()
@@ -131,21 +126,21 @@
 
         if (iblockstate.func_177230_c() instanceof BlockShulkerBox)
         {
-            EnumFacing enumfacing = iblockstate.func_177229_b(BlockShulkerBox.field_190957_a);
+            EnumFacing enumfacing = (EnumFacing)iblockstate.func_177229_b(BlockShulkerBox.field_190957_a);
             AxisAlignedBB axisalignedbb = this.func_190588_c(enumfacing).func_186670_a(this.field_174879_c);
-            List<Entity> list = this.field_145850_b.func_72839_b(null, axisalignedbb);
+            List<Entity> list = this.field_145850_b.func_72839_b((Entity)null, axisalignedbb);
 
             if (!list.isEmpty())
             {
-                for (int i = 0; i < list.size(); i++)
+                for (int i = 0; i < list.size(); ++i)
                 {
                     Entity entity = list.get(i);
 
                     if (entity.func_184192_z() != EnumPushReaction.IGNORE)
                     {
-                        double d0 = 0.0;
-                        double d1 = 0.0;
-                        double d2 = 0.0;
+                        double d0 = 0.0D;
+                        double d1 = 0.0D;
+                        double d2 = 0.0D;
                         AxisAlignedBB axisalignedbb1 = entity.func_174813_aQ();
 
                         switch (enumfacing.func_176740_k())
@@ -161,7 +156,7 @@
                                     d0 = axisalignedbb1.field_72336_d - axisalignedbb.field_72340_a;
                                 }
 
-                                d0 += 0.01;
+                                d0 = d0 + 0.01D;
                                 break;
                             case Y:
 
@@ -174,7 +169,7 @@
                                     d1 = axisalignedbb1.field_72337_e - axisalignedbb.field_72338_b;
                                 }
 
-                                d1 += 0.01;
+                                d1 = d1 + 0.01D;
                                 break;
                             case Z:
 
@@ -187,34 +182,26 @@
                                     d2 = axisalignedbb1.field_72334_f - axisalignedbb.field_72339_c;
                                 }
 
-                                d2 += 0.01;
+                                d2 = d2 + 0.01D;
                         }
 
-                        entity.func_70091_d(
-                            MoverType.SHULKER_BOX,
-                            d0 * (double)enumfacing.func_82601_c(),
-                            d1 * (double)enumfacing.func_96559_d(),
-                            d2 * (double)enumfacing.func_82599_e()
-                        );
+                        entity.func_70091_d(MoverType.SHULKER_BOX, d0 * (double)enumfacing.func_82601_c(), d1 * (double)enumfacing.func_96559_d(), d2 * (double)enumfacing.func_82599_e());
                     }
                 }
             }
         }
     }
 
-    @Override
     public int func_70302_i_()
     {
         return this.field_190596_f.size();
     }
 
-    @Override
     public int func_70297_j_()
     {
         return 64;
     }
 
-    @Override
     public boolean func_145842_c(int p_145842_1_, int p_145842_2_)
     {
         if (p_145842_1_ == 1)
@@ -239,7 +226,6 @@
         }
     }
 
-    @Override
     public void func_174889_b(EntityPlayer p_174889_1_)
     {
         if (!p_174889_1_.func_175149_v())
@@ -249,60 +235,40 @@
                 this.field_190598_h = 0;
             }
 
-            this.field_190598_h++;
+            ++this.field_190598_h;
             this.field_145850_b.func_175641_c(this.field_174879_c, this.func_145838_q(), 1, this.field_190598_h);
 
             if (this.field_190598_h == 1)
             {
-                this.field_145850_b
-                .func_184133_a(
-                    null,
-                    this.field_174879_c,
-                    SoundEvents.field_191262_fB,
-                    SoundCategory.BLOCKS,
-                    0.5F,
-                    this.field_145850_b.field_73012_v.nextFloat() * 0.1F + 0.9F
-                );
+                this.field_145850_b.func_184133_a((EntityPlayer)null, this.field_174879_c, SoundEvents.field_191262_fB, SoundCategory.BLOCKS, 0.5F, this.field_145850_b.field_73012_v.nextFloat() * 0.1F + 0.9F);
             }
         }
     }
 
-    @Override
     public void func_174886_c(EntityPlayer p_174886_1_)
     {
         if (!p_174886_1_.func_175149_v())
         {
-            this.field_190598_h--;
+            --this.field_190598_h;
             this.field_145850_b.func_175641_c(this.field_174879_c, this.func_145838_q(), 1, this.field_190598_h);
 
             if (this.field_190598_h <= 0)
             {
-                this.field_145850_b
-                .func_184133_a(
-                    null,
-                    this.field_174879_c,
-                    SoundEvents.field_191261_fA,
-                    SoundCategory.BLOCKS,
-                    0.5F,
-                    this.field_145850_b.field_73012_v.nextFloat() * 0.1F + 0.9F
-                );
+                this.field_145850_b.func_184133_a((EntityPlayer)null, this.field_174879_c, SoundEvents.field_191261_fA, SoundCategory.BLOCKS, 0.5F, this.field_145850_b.field_73012_v.nextFloat() * 0.1F + 0.9F);
             }
         }
     }
 
-    @Override
     public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
     {
         return new ContainerShulkerBox(p_174876_1_, this, p_174876_2_);
     }
 
-    @Override
     public String func_174875_k()
     {
         return "minecraft:shulker_box";
     }
 
-    @Override
     public String func_70005_c_()
     {
         return this.func_145818_k_() ? this.field_190577_o : "container.shulkerBox";
@@ -310,17 +276,15 @@
 
     public static void func_190593_a(DataFixer p_190593_0_)
     {
-        p_190593_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityShulkerBox.class, "Items"));
+        p_190593_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityShulkerBox.class, new String[] {"Items"}));
     }
 
-    @Override
     public void func_145839_a(NBTTagCompound p_145839_1_)
     {
         super.func_145839_a(p_145839_1_);
         this.func_190586_e(p_145839_1_);
     }
 
-    @Override
     public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
     {
         super.func_189515_b(p_189515_1_);
@@ -329,7 +293,7 @@
 
     public void func_190586_e(NBTTagCompound p_190586_1_)
     {
-        this.field_190596_f = NonNullList.func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
+        this.field_190596_f = NonNullList.<ItemStack>func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
 
         if (!this.func_184283_b(p_190586_1_) && p_190586_1_.func_150297_b("Items", 9))
         {
@@ -362,13 +326,11 @@
         return p_190580_1_;
     }
 
-    @Override
     protected NonNullList<ItemStack> func_190576_q()
     {
         return this.field_190596_f;
     }
 
-    @Override
     public boolean func_191420_l()
     {
         for (ItemStack itemstack : this.field_190596_f)
@@ -382,25 +344,21 @@
         return true;
     }
 
-    @Override
     public int[] func_180463_a(EnumFacing p_180463_1_)
     {
         return field_190595_a;
     }
 
-    @Override
     public boolean func_180462_a(int p_180462_1_, ItemStack p_180462_2_, EnumFacing p_180462_3_)
     {
         return !(Block.func_149634_a(p_180462_2_.func_77973_b()) instanceof BlockShulkerBox);
     }
 
-    @Override
     public boolean func_180461_b(int p_180461_1_, ItemStack p_180461_2_, EnumFacing p_180461_3_)
     {
         return true;
     }
 
-    @Override
     public void func_174888_l()
     {
         this.field_190597_g = true;
@@ -429,7 +387,6 @@
     }
 
     @Nullable
-    @Override
     public SPacketUpdateTileEntity func_189518_D_()
     {
         return new SPacketUpdateTileEntity(this.field_174879_c, 10, this.func_189517_E_());
@@ -452,11 +409,9 @@
 
     static
     {
-        int i = 0;
-
-        while (i < field_190595_a.length)
+        for (int i = 0; i < field_190595_a.length; field_190595_a[i] = i++)
         {
-            field_190595_a[i] = i++;
+            ;
         }
     }
 
@@ -466,5 +421,10 @@
         OPENING,
         OPENED,
         CLOSING;
+    }
+
+    protected net.minecraftforge.items.IItemHandler createUnSidedHandler()
+    {
+        return new net.minecraftforge.items.wrapper.SidedInvWrapper(this, EnumFacing.UP);
     }
 }
