--- before/net/minecraft/tileentity/TileEntityHopper.java
+++ after/net/minecraft/tileentity/TileEntityHopper.java
@@ -30,20 +30,19 @@
 
 public class TileEntityHopper extends TileEntityLockableLoot implements IHopper, ITickable
 {
-    private NonNullList<ItemStack> field_145900_a = NonNullList.func_191197_a(5, ItemStack.field_190927_a);
+    private NonNullList<ItemStack> field_145900_a = NonNullList.<ItemStack>func_191197_a(5, ItemStack.field_190927_a);
     private int field_145901_j = -1;
     private long field_190578_g;
 
     public static void func_189683_a(DataFixer p_189683_0_)
     {
-        p_189683_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityHopper.class, "Items"));
+        p_189683_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityHopper.class, new String[] {"Items"}));
     }
 
-    @Override
     public void func_145839_a(NBTTagCompound p_145839_1_)
     {
         super.func_145839_a(p_145839_1_);
-        this.field_145900_a = NonNullList.func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
+        this.field_145900_a = NonNullList.<ItemStack>func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
 
         if (!this.func_184283_b(p_145839_1_))
         {
@@ -58,7 +57,6 @@
         this.field_145901_j = p_145839_1_.func_74762_e("TransferCooldown");
     }
 
-    @Override
     public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
     {
         super.func_189515_b(p_189515_1_);
@@ -78,23 +76,21 @@
         return p_189515_1_;
     }
 
-    @Override
     public int func_70302_i_()
     {
         return this.field_145900_a.size();
     }
 
-    @Override
     public ItemStack func_70298_a(int p_70298_1_, int p_70298_2_)
     {
-        this.func_184281_d(null);
-        return ItemStackHelper.func_188382_a(this.func_190576_q(), p_70298_1_, p_70298_2_);
+        this.func_184281_d((EntityPlayer)null);
+        ItemStack itemstack = ItemStackHelper.func_188382_a(this.func_190576_q(), p_70298_1_, p_70298_2_);
+        return itemstack;
     }
 
-    @Override
     public void func_70299_a(int p_70299_1_, ItemStack p_70299_2_)
     {
-        this.func_184281_d(null);
+        this.func_184281_d((EntityPlayer)null);
         this.func_190576_q().set(p_70299_1_, p_70299_2_);
 
         if (p_70299_2_.func_190916_E() > this.func_70297_j_())
@@ -103,24 +99,21 @@
         }
     }
 
-    @Override
     public String func_70005_c_()
     {
         return this.func_145818_k_() ? this.field_190577_o : "container.hopper";
     }
 
-    @Override
     public int func_70297_j_()
     {
         return 64;
     }
 
-    @Override
     public void func_73660_a()
     {
         if (this.field_145850_b != null && !this.field_145850_b.field_72995_K)
         {
-            this.field_145901_j--;
+            --this.field_145901_j;
             this.field_190578_g = this.field_145850_b.func_82737_E();
 
             if (!this.func_145888_j())
@@ -178,7 +171,6 @@
         return true;
     }
 
-    @Override
     public boolean func_191420_l()
     {
         return this.func_152104_k();
@@ -199,6 +191,7 @@
 
     private boolean func_145883_k()
     {
+        if (net.minecraftforge.items.VanillaInventoryCodeHooks.insertHook(this)) { return true; }
         IInventory iinventory = this.func_145895_l();
 
         if (iinventory == null)
@@ -215,7 +208,7 @@
             }
             else
             {
-                for (int i = 0; i < this.func_70302_i_(); i++)
+                for (int i = 0; i < this.func_70302_i_(); ++i)
                 {
                     if (!this.func_70301_a(i).func_190926_b())
                     {
@@ -244,23 +237,9 @@
             ISidedInventory isidedinventory = (ISidedInventory)p_174919_1_;
             int[] aint = isidedinventory.func_180463_a(p_174919_2_);
 
-            for (int i : aint)
-            {
-                ItemStack itemstack = isidedinventory.func_70301_a(i);
-
-                if (itemstack.func_190926_b() || itemstack.func_190916_E() != itemstack.func_77976_d())
-                {
-                    return false;
-                }
-            }
-        }
-        else
-        {
-            int j = p_174919_1_.func_70302_i_();
-
-            for (int k = 0; k < j; k++)
-            {
-                ItemStack itemstack1 = p_174919_1_.func_70301_a(k);
+            for (int k : aint)
+            {
+                ItemStack itemstack1 = isidedinventory.func_70301_a(k);
 
                 if (itemstack1.func_190926_b() || itemstack1.func_190916_E() != itemstack1.func_77976_d())
                 {
@@ -268,6 +247,20 @@
                 }
             }
         }
+        else
+        {
+            int i = p_174919_1_.func_70302_i_();
+
+            for (int j = 0; j < i; ++j)
+            {
+                ItemStack itemstack = p_174919_1_.func_70301_a(j);
+
+                if (itemstack.func_190926_b() || itemstack.func_190916_E() != itemstack.func_77976_d())
+                {
+                    return false;
+                }
+            }
+        }
 
         return true;
     }
@@ -291,7 +284,7 @@
         {
             int j = p_174917_0_.func_70302_i_();
 
-            for (int k = 0; k < j; k++)
+            for (int k = 0; k < j; ++k)
             {
                 if (!p_174917_0_.func_70301_a(k).func_190926_b())
                 {
@@ -305,6 +298,8 @@
 
     public static boolean func_145891_a(IHopper p_145891_0_)
     {
+        Boolean ret = net.minecraftforge.items.VanillaInventoryCodeHooks.extractHook(p_145891_0_);
+        if (ret != null) return ret;
         IInventory iinventory = func_145884_b(p_145891_0_);
 
         if (iinventory != null)
@@ -333,7 +328,7 @@
             {
                 int j = iinventory.func_70302_i_();
 
-                for (int k = 0; k < j; k++)
+                for (int k = 0; k < j; ++k)
                 {
                     if (func_174915_a(p_145891_0_, iinventory, k, enumfacing))
                     {
@@ -344,11 +339,9 @@
         }
         else
         {
-            for (EntityItem entityitem : func_184292_a(
-                        p_145891_0_.func_145831_w(), p_145891_0_.func_96107_aA(), p_145891_0_.func_96109_aB(), p_145891_0_.func_96108_aC()
-                    ))
+            for (EntityItem entityitem : func_184292_a(p_145891_0_.func_145831_w(), p_145891_0_.func_96107_aA(), p_145891_0_.func_96109_aB(), p_145891_0_.func_96108_aC()))
             {
-                if (func_145898_a(null, p_145891_0_, entityitem))
+                if (func_145898_a((IInventory)null, p_145891_0_, entityitem))
                 {
                     return true;
                 }
@@ -365,7 +358,7 @@
         if (!itemstack.func_190926_b() && func_174921_b(p_174915_1_, itemstack, p_174915_2_, p_174915_3_))
         {
             ItemStack itemstack1 = itemstack.func_77946_l();
-            ItemStack itemstack2 = func_174918_a(p_174915_1_, p_174915_0_, p_174915_1_.func_70298_a(p_174915_2_, 1), null);
+            ItemStack itemstack2 = func_174918_a(p_174915_1_, p_174915_0_, p_174915_1_.func_70298_a(p_174915_2_, 1), (EnumFacing)null);
 
             if (itemstack2.func_190926_b())
             {
@@ -390,7 +383,7 @@
         else
         {
             ItemStack itemstack = p_145898_2_.func_92059_d().func_77946_l();
-            ItemStack itemstack1 = func_174918_a(p_145898_0_, p_145898_1_, itemstack, null);
+            ItemStack itemstack1 = func_174918_a(p_145898_0_, p_145898_1_, itemstack, (EnumFacing)null);
 
             if (itemstack1.func_190926_b())
             {
@@ -406,6 +399,12 @@
         }
     }
 
+
+    protected net.minecraftforge.items.IItemHandler createUnSidedHandler()
+    {
+        return new net.minecraftforge.items.VanillaHopperItemHandler(this);
+    }
+
     public static ItemStack func_174918_a(IInventory p_174918_0_, IInventory p_174918_1_, ItemStack p_174918_2_, @Nullable EnumFacing p_174918_3_)
     {
         if (p_174918_1_ instanceof ISidedInventory && p_174918_3_ != null)
@@ -413,7 +412,7 @@
             ISidedInventory isidedinventory = (ISidedInventory)p_174918_1_;
             int[] aint = isidedinventory.func_180463_a(p_174918_3_);
 
-            for (int k = 0; k < aint.length && !p_174918_2_.func_190926_b(); k++)
+            for (int k = 0; k < aint.length && !p_174918_2_.func_190926_b(); ++k)
             {
                 p_174918_2_ = func_174916_c(p_174918_0_, p_174918_1_, p_174918_2_, aint[k], p_174918_3_);
             }
@@ -422,7 +421,7 @@
         {
             int i = p_174918_1_.func_70302_i_();
 
-            for (int j = 0; j < i && !p_174918_2_.func_190926_b(); j++)
+            for (int j = 0; j < i && !p_174918_2_.func_190926_b(); ++j)
             {
                 p_174918_2_ = func_174916_c(p_174918_0_, p_174918_1_, p_174918_2_, j, p_174918_3_);
             }
@@ -433,9 +432,14 @@
 
     private static boolean func_174920_a(IInventory p_174920_0_, ItemStack p_174920_1_, int p_174920_2_, EnumFacing p_174920_3_)
     {
-        return !p_174920_0_.func_94041_b(p_174920_2_, p_174920_1_)
-               ? false
-               : !(p_174920_0_ instanceof ISidedInventory) || ((ISidedInventory)p_174920_0_).func_180462_a(p_174920_2_, p_174920_1_, p_174920_3_);
+        if (!p_174920_0_.func_94041_b(p_174920_2_, p_174920_1_))
+        {
+            return false;
+        }
+        else
+        {
+            return !(p_174920_0_ instanceof ISidedInventory) || ((ISidedInventory)p_174920_0_).func_180462_a(p_174920_2_, p_174920_1_, p_174920_3_);
+        }
     }
 
     private static boolean func_174921_b(IInventory p_174921_0_, ItemStack p_174921_1_, int p_174921_2_, EnumFacing p_174921_3_)
@@ -501,26 +505,17 @@
     private IInventory func_145895_l()
     {
         EnumFacing enumfacing = BlockHopper.func_176428_b(this.func_145832_p());
-        return func_145893_b(
-                   this.func_145831_w(),
-                   this.func_96107_aA() + (double)enumfacing.func_82601_c(),
-                   this.func_96109_aB() + (double)enumfacing.func_96559_d(),
-                   this.func_96108_aC() + (double)enumfacing.func_82599_e()
-               );
+        return func_145893_b(this.func_145831_w(), this.func_96107_aA() + (double)enumfacing.func_82601_c(), this.func_96109_aB() + (double)enumfacing.func_96559_d(), this.func_96108_aC() + (double)enumfacing.func_82599_e());
     }
 
     public static IInventory func_145884_b(IHopper p_145884_0_)
     {
-        return func_145893_b(p_145884_0_.func_145831_w(), p_145884_0_.func_96107_aA(), p_145884_0_.func_96109_aB() + 1.0, p_145884_0_.func_96108_aC());
+        return func_145893_b(p_145884_0_.func_145831_w(), p_145884_0_.func_96107_aA(), p_145884_0_.func_96109_aB() + 1.0D, p_145884_0_.func_96108_aC());
     }
 
     public static List<EntityItem> func_184292_a(World p_184292_0_, double p_184292_1_, double p_184292_3_, double p_184292_5_)
     {
-        return p_184292_0_.func_175647_a(
-                   EntityItem.class,
-                   new AxisAlignedBB(p_184292_1_ - 0.5, p_184292_3_, p_184292_5_ - 0.5, p_184292_1_ + 0.5, p_184292_3_ + 1.5, p_184292_5_ + 0.5),
-                   EntitySelectors.field_94557_a
-               );
+        return p_184292_0_.<EntityItem>func_175647_a(EntityItem.class, new AxisAlignedBB(p_184292_1_ - 0.5D, p_184292_3_, p_184292_5_ - 0.5D, p_184292_1_ + 0.5D, p_184292_3_ + 1.5D, p_184292_5_ + 0.5D), EntitySelectors.field_94557_a);
     }
 
     public static IInventory func_145893_b(World p_145893_0_, double p_145893_1_, double p_145893_3_, double p_145893_5_)
@@ -530,9 +525,10 @@
         int j = MathHelper.func_76128_c(p_145893_3_);
         int k = MathHelper.func_76128_c(p_145893_5_);
         BlockPos blockpos = new BlockPos(i, j, k);
-        Block block = p_145893_0_.func_180495_p(blockpos).func_177230_c();
+        net.minecraft.block.state.IBlockState state = p_145893_0_.func_180495_p(blockpos);
+        Block block = state.func_177230_c();
 
-        if (block.func_149716_u())
+        if (block.hasTileEntity(state))
         {
             TileEntity tileentity = p_145893_0_.func_175625_s(blockpos);
 
@@ -549,11 +545,7 @@
 
         if (iinventory == null)
         {
-            List<Entity> list = p_145893_0_.func_175674_a(
-                                    null,
-                                    new AxisAlignedBB(p_145893_1_ - 0.5, p_145893_3_ - 0.5, p_145893_5_ - 0.5, p_145893_1_ + 0.5, p_145893_3_ + 0.5, p_145893_5_ + 0.5),
-                                    EntitySelectors.field_96566_b
-                                );
+            List<Entity> list = p_145893_0_.func_175674_a((Entity)null, new AxisAlignedBB(p_145893_1_ - 0.5D, p_145893_3_ - 0.5D, p_145893_5_ - 0.5D, p_145893_1_ + 0.5D, p_145893_3_ + 0.5D, p_145893_5_ + 0.5D), EntitySelectors.field_96566_b);
 
             if (!list.isEmpty())
             {
@@ -574,28 +566,29 @@
         {
             return false;
         }
+        else if (p_145894_0_.func_190916_E() > p_145894_0_.func_77976_d())
+        {
+            return false;
+        }
         else
         {
-            return p_145894_0_.func_190916_E() > p_145894_0_.func_77976_d() ? false : ItemStack.func_77970_a(p_145894_0_, p_145894_1_);
+            return ItemStack.func_77970_a(p_145894_0_, p_145894_1_);
         }
     }
 
-    @Override
     public double func_96107_aA()
     {
-        return (double)this.field_174879_c.func_177958_n() + 0.5;
+        return (double)this.field_174879_c.func_177958_n() + 0.5D;
     }
 
-    @Override
     public double func_96109_aB()
     {
-        return (double)this.field_174879_c.func_177956_o() + 0.5;
+        return (double)this.field_174879_c.func_177956_o() + 0.5D;
     }
 
-    @Override
     public double func_96108_aC()
     {
-        return (double)this.field_174879_c.func_177952_p() + 0.5;
+        return (double)this.field_174879_c.func_177952_p() + 0.5D;
     }
 
     public void func_145896_c(int p_145896_1_)
@@ -613,22 +606,21 @@
         return this.field_145901_j > 8;
     }
 
-    @Override
     public String func_174875_k()
     {
         return "minecraft:hopper";
     }
 
-    @Override
     public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
     {
         this.func_184281_d(p_174876_2_);
         return new ContainerHopper(p_174876_1_, this, p_174876_2_);
     }
 
-    @Override
     protected NonNullList<ItemStack> func_190576_q()
     {
         return this.field_145900_a;
     }
+
+    public long getLastUpdateTime() { return field_190578_g; } // Forge
 }
