--- before/net/minecraft/tileentity/TileEntityPiston.java
+++ after/net/minecraft/tileentity/TileEntityPiston.java
@@ -55,13 +55,11 @@
         return this.field_174932_a;
     }
 
-    @Override
     public NBTTagCompound func_189517_E_()
     {
         return this.func_189515_b(new NBTTagCompound());
     }
 
-    @Override
     public int func_145832_p()
     {
         return 0;
@@ -125,46 +123,31 @@
     {
         p_184319_3_ = this.func_184320_e(p_184319_3_);
         IBlockState iblockstate = this.func_190606_j();
-        return iblockstate.func_185900_c(p_184319_1_, p_184319_2_)
-               .func_72317_d(
-                   (double)(p_184319_3_ * (float)this.field_174931_f.func_82601_c()),
-                   (double)(p_184319_3_ * (float)this.field_174931_f.func_96559_d()),
-                   (double)(p_184319_3_ * (float)this.field_174931_f.func_82599_e())
-               );
+        return iblockstate.func_185900_c(p_184319_1_, p_184319_2_).func_72317_d((double)(p_184319_3_ * (float)this.field_174931_f.func_82601_c()), (double)(p_184319_3_ * (float)this.field_174931_f.func_96559_d()), (double)(p_184319_3_ * (float)this.field_174931_f.func_82599_e()));
     }
 
     private IBlockState func_190606_j()
     {
-        return !this.func_145868_b() && this.func_145867_d()
-               ? Blocks.field_150332_K
-               .func_176223_P()
-               .func_177226_a(
-                   BlockPistonExtension.field_176325_b,
-                   this.field_174932_a.func_177230_c() == Blocks.field_150320_F
-                   ? BlockPistonExtension.EnumPistonType.STICKY
-                   : BlockPistonExtension.EnumPistonType.DEFAULT
-               )
-               .func_177226_a(BlockPistonExtension.field_176387_N, this.field_174932_a.func_177229_b(BlockPistonBase.field_176387_N))
-               : this.field_174932_a;
+        return !this.func_145868_b() && this.func_145867_d() ? Blocks.field_150332_K.func_176223_P().func_177226_a(BlockPistonExtension.field_176325_b, this.field_174932_a.func_177230_c() == Blocks.field_150320_F ? BlockPistonExtension.EnumPistonType.STICKY : BlockPistonExtension.EnumPistonType.DEFAULT).func_177226_a(BlockPistonExtension.field_176387_N, this.field_174932_a.func_177229_b(BlockPistonBase.field_176387_N)) : this.field_174932_a;
     }
 
     private void func_184322_i(float p_184322_1_)
     {
         EnumFacing enumfacing = this.field_145875_k ? this.field_174931_f : this.field_174931_f.func_176734_d();
         double d0 = (double)(p_184322_1_ - this.field_145873_m);
-        List<AxisAlignedBB> list = Lists.newArrayList();
-        this.func_190606_j().func_185908_a(this.field_145850_b, BlockPos.field_177992_a, new AxisAlignedBB(BlockPos.field_177992_a), list, null, true);
+        List<AxisAlignedBB> list = Lists.<AxisAlignedBB>newArrayList();
+        this.func_190606_j().func_185908_a(this.field_145850_b, BlockPos.field_177992_a, new AxisAlignedBB(BlockPos.field_177992_a), list, (Entity)null, true);
 
         if (!list.isEmpty())
         {
             AxisAlignedBB axisalignedbb = this.func_190607_a(this.func_191515_a(list));
-            List<Entity> list1 = this.field_145850_b.func_72839_b(null, this.func_190610_a(axisalignedbb, enumfacing, d0).func_111270_a(axisalignedbb));
+            List<Entity> list1 = this.field_145850_b.func_72839_b((Entity)null, this.func_190610_a(axisalignedbb, enumfacing, d0).func_111270_a(axisalignedbb));
 
             if (!list1.isEmpty())
             {
-                boolean flag = this.field_174932_a.func_177230_c() == Blocks.field_180399_cE;
+                boolean flag = this.field_174932_a.func_177230_c().isStickyBlock(this.field_174932_a);
 
-                for (int i = 0; i < list1.size(); i++)
+                for (int i = 0; i < list1.size(); ++i)
                 {
                     Entity entity = list1.get(i);
 
@@ -185,9 +168,9 @@
                             }
                         }
 
-                        double d1 = 0.0;
+                        double d1 = 0.0D;
 
-                        for (int j = 0; j < list.size(); j++)
+                        for (int j = 0; j < list.size(); ++j)
                         {
                             AxisAlignedBB axisalignedbb1 = this.func_190610_a(this.func_190607_a(list.get(j)), enumfacing, d0);
                             AxisAlignedBB axisalignedbb2 = entity.func_174813_aQ();
@@ -203,16 +186,11 @@
                             }
                         }
 
-                        if (!(d1 <= 0.0))
+                        if (d1 > 0.0D)
                         {
-                            d1 = Math.min(d1, d0) + 0.01;
+                            d1 = Math.min(d1, d0) + 0.01D;
                             field_190613_i.set(enumfacing);
-                            entity.func_70091_d(
-                                MoverType.PISTON,
-                                d1 * (double)enumfacing.func_82601_c(),
-                                d1 * (double)enumfacing.func_96559_d(),
-                                d1 * (double)enumfacing.func_82599_e()
-                            );
+                            entity.func_70091_d(MoverType.PISTON, d1 * (double)enumfacing.func_82601_c(), d1 * (double)enumfacing.func_96559_d(), d1 * (double)enumfacing.func_82599_e());
                             field_190613_i.set(null);
 
                             if (!this.field_145875_k && this.field_145872_l)
@@ -228,12 +206,12 @@
 
     private AxisAlignedBB func_191515_a(List<AxisAlignedBB> p_191515_1_)
     {
-        double d0 = 0.0;
-        double d1 = 0.0;
-        double d2 = 0.0;
-        double d3 = 1.0;
-        double d4 = 1.0;
-        double d5 = 1.0;
+        double d0 = 0.0D;
+        double d1 = 0.0D;
+        double d2 = 0.0D;
+        double d3 = 1.0D;
+        double d4 = 1.0D;
+        double d5 = 1.0D;
 
         for (AxisAlignedBB axisalignedbb : p_191515_1_)
         {
@@ -265,76 +243,30 @@
     private AxisAlignedBB func_190607_a(AxisAlignedBB p_190607_1_)
     {
         double d0 = (double)this.func_184320_e(this.field_145873_m);
-        return p_190607_1_.func_72317_d(
-                   (double)this.field_174879_c.func_177958_n() + d0 * (double)this.field_174931_f.func_82601_c(),
-                   (double)this.field_174879_c.func_177956_o() + d0 * (double)this.field_174931_f.func_96559_d(),
-                   (double)this.field_174879_c.func_177952_p() + d0 * (double)this.field_174931_f.func_82599_e()
-               );
+        return p_190607_1_.func_72317_d((double)this.field_174879_c.func_177958_n() + d0 * (double)this.field_174931_f.func_82601_c(), (double)this.field_174879_c.func_177956_o() + d0 * (double)this.field_174931_f.func_96559_d(), (double)this.field_174879_c.func_177952_p() + d0 * (double)this.field_174931_f.func_82599_e());
     }
 
     private AxisAlignedBB func_190610_a(AxisAlignedBB p_190610_1_, EnumFacing p_190610_2_, double p_190610_3_)
     {
         double d0 = p_190610_3_ * (double)p_190610_2_.func_176743_c().func_179524_a();
-        double d1 = Math.min(d0, 0.0);
-        double d2 = Math.max(d0, 0.0);
+        double d1 = Math.min(d0, 0.0D);
+        double d2 = Math.max(d0, 0.0D);
 
         switch (p_190610_2_)
         {
             case WEST:
-                return new AxisAlignedBB(
-                           p_190610_1_.field_72340_a + d1,
-                           p_190610_1_.field_72338_b,
-                           p_190610_1_.field_72339_c,
-                           p_190610_1_.field_72340_a + d2,
-                           p_190610_1_.field_72337_e,
-                           p_190610_1_.field_72334_f
-                       );
+                return new AxisAlignedBB(p_190610_1_.field_72340_a + d1, p_190610_1_.field_72338_b, p_190610_1_.field_72339_c, p_190610_1_.field_72340_a + d2, p_190610_1_.field_72337_e, p_190610_1_.field_72334_f);
             case EAST:
-                return new AxisAlignedBB(
-                           p_190610_1_.field_72336_d + d1,
-                           p_190610_1_.field_72338_b,
-                           p_190610_1_.field_72339_c,
-                           p_190610_1_.field_72336_d + d2,
-                           p_190610_1_.field_72337_e,
-                           p_190610_1_.field_72334_f
-                       );
+                return new AxisAlignedBB(p_190610_1_.field_72336_d + d1, p_190610_1_.field_72338_b, p_190610_1_.field_72339_c, p_190610_1_.field_72336_d + d2, p_190610_1_.field_72337_e, p_190610_1_.field_72334_f);
             case DOWN:
-                return new AxisAlignedBB(
-                           p_190610_1_.field_72340_a,
-                           p_190610_1_.field_72338_b + d1,
-                           p_190610_1_.field_72339_c,
-                           p_190610_1_.field_72336_d,
-                           p_190610_1_.field_72338_b + d2,
-                           p_190610_1_.field_72334_f
-                       );
+                return new AxisAlignedBB(p_190610_1_.field_72340_a, p_190610_1_.field_72338_b + d1, p_190610_1_.field_72339_c, p_190610_1_.field_72336_d, p_190610_1_.field_72338_b + d2, p_190610_1_.field_72334_f);
             case UP:
             default:
-                return new AxisAlignedBB(
-                           p_190610_1_.field_72340_a,
-                           p_190610_1_.field_72337_e + d1,
-                           p_190610_1_.field_72339_c,
-                           p_190610_1_.field_72336_d,
-                           p_190610_1_.field_72337_e + d2,
-                           p_190610_1_.field_72334_f
-                       );
+                return new AxisAlignedBB(p_190610_1_.field_72340_a, p_190610_1_.field_72337_e + d1, p_190610_1_.field_72339_c, p_190610_1_.field_72336_d, p_190610_1_.field_72337_e + d2, p_190610_1_.field_72334_f);
             case NORTH:
-                return new AxisAlignedBB(
-                           p_190610_1_.field_72340_a,
-                           p_190610_1_.field_72338_b,
-                           p_190610_1_.field_72339_c + d1,
-                           p_190610_1_.field_72336_d,
-                           p_190610_1_.field_72337_e,
-                           p_190610_1_.field_72339_c + d2
-                       );
+                return new AxisAlignedBB(p_190610_1_.field_72340_a, p_190610_1_.field_72338_b, p_190610_1_.field_72339_c + d1, p_190610_1_.field_72336_d, p_190610_1_.field_72337_e, p_190610_1_.field_72339_c + d2);
             case SOUTH:
-                return new AxisAlignedBB(
-                           p_190610_1_.field_72340_a,
-                           p_190610_1_.field_72338_b,
-                           p_190610_1_.field_72334_f + d1,
-                           p_190610_1_.field_72336_d,
-                           p_190610_1_.field_72337_e,
-                           p_190610_1_.field_72334_f + d2
-                       );
+                return new AxisAlignedBB(p_190610_1_.field_72340_a, p_190610_1_.field_72338_b, p_190610_1_.field_72334_f + d1, p_190610_1_.field_72336_d, p_190610_1_.field_72337_e, p_190610_1_.field_72334_f + d2);
         }
     }
 
@@ -346,16 +278,14 @@
         if (axisalignedbb.func_72326_a(axisalignedbb1))
         {
             EnumFacing enumfacing = p_190605_2_.func_176734_d();
-            double d0 = this.func_190612_a(axisalignedbb1, enumfacing, axisalignedbb) + 0.01;
-            double d1 = this.func_190612_a(axisalignedbb1, enumfacing, axisalignedbb.func_191500_a(axisalignedbb1)) + 0.01;
+            double d0 = this.func_190612_a(axisalignedbb1, enumfacing, axisalignedbb) + 0.01D;
+            double d1 = this.func_190612_a(axisalignedbb1, enumfacing, axisalignedbb.func_191500_a(axisalignedbb1)) + 0.01D;
 
-            if (Math.abs(d0 - d1) < 0.01)
+            if (Math.abs(d0 - d1) < 0.01D)
             {
-                d0 = Math.min(d0, p_190605_3_) + 0.01;
+                d0 = Math.min(d0, p_190605_3_) + 0.01D;
                 field_190613_i.set(p_190605_2_);
-                p_190605_1_.func_70091_d(
-                    MoverType.PISTON, d0 * (double)enumfacing.func_82601_c(), d0 * (double)enumfacing.func_96559_d(), d0 * (double)enumfacing.func_82599_e()
-                );
+                p_190605_1_.func_70091_d(MoverType.PISTON, d0 * (double)enumfacing.func_82601_c(), d0 * (double)enumfacing.func_96559_d(), d0 * (double)enumfacing.func_82599_e());
                 field_190613_i.set(null);
             }
         }
@@ -363,23 +293,17 @@
 
     private static double func_190611_b(AxisAlignedBB p_190611_0_, EnumFacing p_190611_1_, AxisAlignedBB p_190611_2_)
     {
-        return p_190611_1_.func_176743_c() == EnumFacing.AxisDirection.POSITIVE
-               ? p_190611_0_.field_72336_d - p_190611_2_.field_72340_a
-               : p_190611_2_.field_72336_d - p_190611_0_.field_72340_a;
+        return p_190611_1_.func_176743_c() == EnumFacing.AxisDirection.POSITIVE ? p_190611_0_.field_72336_d - p_190611_2_.field_72340_a : p_190611_2_.field_72336_d - p_190611_0_.field_72340_a;
     }
 
     private static double func_190608_c(AxisAlignedBB p_190608_0_, EnumFacing p_190608_1_, AxisAlignedBB p_190608_2_)
     {
-        return p_190608_1_.func_176743_c() == EnumFacing.AxisDirection.POSITIVE
-               ? p_190608_0_.field_72337_e - p_190608_2_.field_72338_b
-               : p_190608_2_.field_72337_e - p_190608_0_.field_72338_b;
+        return p_190608_1_.func_176743_c() == EnumFacing.AxisDirection.POSITIVE ? p_190608_0_.field_72337_e - p_190608_2_.field_72338_b : p_190608_2_.field_72337_e - p_190608_0_.field_72338_b;
     }
 
     private static double func_190604_d(AxisAlignedBB p_190604_0_, EnumFacing p_190604_1_, AxisAlignedBB p_190604_2_)
     {
-        return p_190604_1_.func_176743_c() == EnumFacing.AxisDirection.POSITIVE
-               ? p_190604_0_.field_72334_f - p_190604_2_.field_72339_c
-               : p_190604_2_.field_72334_f - p_190604_0_.field_72339_c;
+        return p_190604_1_.func_176743_c() == EnumFacing.AxisDirection.POSITIVE ? p_190604_0_.field_72334_f - p_190604_2_.field_72339_c : p_190604_2_.field_72334_f - p_190604_0_.field_72339_c;
     }
 
     public void func_145866_f()
@@ -399,7 +323,6 @@
         }
     }
 
-    @Override
     public void func_73660_a()
     {
         this.field_145870_n = this.field_145873_m;
@@ -432,7 +355,6 @@
     {
     }
 
-    @Override
     public void func_145839_a(NBTTagCompound p_145839_1_)
     {
         super.func_145839_a(p_145839_1_);
@@ -444,7 +366,6 @@
         this.field_145872_l = p_145839_1_.func_74767_n("source");
     }
 
-    @Override
     public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
     {
         super.func_189515_b(p_189515_1_);
@@ -461,24 +382,19 @@
     {
         if (!this.field_145875_k && this.field_145872_l)
         {
-            this.field_174932_a
-            .func_177226_a(BlockPistonBase.field_176320_b, true)
-            .func_185908_a(p_190609_1_, p_190609_2_, p_190609_3_, p_190609_4_, p_190609_5_, false);
+            this.field_174932_a.func_177226_a(BlockPistonBase.field_176320_b, Boolean.valueOf(true)).func_185908_a(p_190609_1_, p_190609_2_, p_190609_3_, p_190609_4_, p_190609_5_, false);
         }
 
         EnumFacing enumfacing = field_190613_i.get();
 
-        if (!((double)this.field_145873_m < 1.0) || enumfacing != (this.field_145875_k ? this.field_174931_f : this.field_174931_f.func_176734_d()))
+        if ((double)this.field_145873_m >= 1.0D || enumfacing != (this.field_145875_k ? this.field_174931_f : this.field_174931_f.func_176734_d()))
         {
             int i = p_190609_4_.size();
             IBlockState iblockstate;
 
             if (this.func_145867_d())
             {
-                iblockstate = Blocks.field_150332_K
-                              .func_176223_P()
-                              .func_177226_a(BlockPistonExtension.field_176387_N, this.field_174931_f)
-                              .func_177226_a(BlockPistonExtension.field_176327_M, this.field_145875_k != 1.0F - this.field_145873_m < 0.25F);
+                iblockstate = Blocks.field_150332_K.func_176223_P().func_177226_a(BlockPistonExtension.field_176387_N, this.field_174931_f).func_177226_a(BlockPistonExtension.field_176327_M, Boolean.valueOf(this.field_145875_k != 1.0F - this.field_145873_m < 0.25F));
             }
             else
             {
@@ -491,9 +407,9 @@
             double d2 = (double)((float)this.field_174931_f.func_82599_e() * f);
             iblockstate.func_185908_a(p_190609_1_, p_190609_2_, p_190609_3_.func_72317_d(-d0, -d1, -d2), p_190609_4_, p_190609_5_, true);
 
-            for (int j = i; j < p_190609_4_.size(); j++)
+            for (int j = i; j < p_190609_4_.size(); ++j)
             {
-                p_190609_4_.set(j, p_190609_4_.get(j).func_72317_d(d0, d1, d2));
+                p_190609_4_.set(j, ((AxisAlignedBB)p_190609_4_.get(j)).func_72317_d(d0, d1, d2));
             }
         }
     }
