--- before/net/minecraft/client/renderer/block/model/BakedQuadRetextured.java
+++ after/net/minecraft/client/renderer/block/model/BakedQuadRetextured.java
@@ -16,7 +16,9 @@
             Arrays.copyOf(quad.getVertexData(), quad.getVertexData().length),
             quad.tintIndex,
             FaceBakery.getFacingFromVertexData(quad.getVertexData()),
-            quad.getSprite()
+            quad.getSprite(),
+            quad.applyDiffuseLighting,
+            quad.format
         );
         this.texture = textureIn;
         this.remapQuad();
@@ -26,13 +28,20 @@
     {
         for (int i = 0; i < 4; i++)
         {
-            int j = 7 * i;
-            this.vertexData[j + 4] = Float.floatToRawIntBits(
-                                             this.texture.getInterpolatedU(this.sprite.getUnInterpolatedU(Float.intBitsToFloat(this.vertexData[j + 4])))
+            int j = format.getIntegerSize() * i;
+            int uvIndex = format.getUvOffsetById(0) / 4;
+            this.vertexData[j + uvIndex] = Float.floatToRawIntBits(
+                                             this.texture.getInterpolatedU(this.sprite.getUnInterpolatedU(Float.intBitsToFloat(this.vertexData[j + uvIndex])))
                                          );
-            this.vertexData[j + 4 + 1] = Float.floatToRawIntBits(
-                                                 this.texture.getInterpolatedV(this.sprite.getUnInterpolatedV(Float.intBitsToFloat(this.vertexData[j + 4 + 1])))
+            this.vertexData[j + uvIndex + 1] = Float.floatToRawIntBits(
+                                                 this.texture.getInterpolatedV(this.sprite.getUnInterpolatedV(Float.intBitsToFloat(this.vertexData[j + uvIndex + 1])))
                                              );
         }
+    }
+
+    @Override
+    public TextureAtlasSprite getSprite()
+    {
+        return texture;
     }
 }
