--- before/net/minecraft/client/renderer/block/model/BlockPart.java
+++ after/net/minecraft/client/renderer/block/model/BlockPart.java
@@ -10,6 +10,7 @@
 import java.lang.reflect.Type;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Map.Entry;
 import javax.annotation.Nullable;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.JsonUtils;
@@ -39,7 +40,7 @@
 
     private void func_178235_a()
     {
-        for (Map.Entry<EnumFacing, BlockPartFace> entry : this.field_178240_c.entrySet())
+        for (Entry<EnumFacing, BlockPartFace> entry : this.field_178240_c.entrySet())
         {
             float[] afloat = this.func_178236_a(entry.getKey());
             (entry.getValue()).field_178243_e.func_178349_a(afloat);
@@ -155,10 +156,10 @@
                 Map<EnumFacing, BlockPartFace> map = Maps.newEnumMap(EnumFacing.class);
                 JsonObject jsonobject = JsonUtils.func_152754_s(p_178253_2_, "faces");
 
-                for (Map.Entry<String, JsonElement> entry : jsonobject.entrySet())
+                for (Entry<String, JsonElement> entry : jsonobject.entrySet())
                 {
                     EnumFacing enumfacing = this.func_178248_a(entry.getKey());
-                    map.put(enumfacing, p_178253_1_.deserialize(entry.getValue(), BlockPartFace.class));
+                    map.put(enumfacing, (BlockPartFace)p_178253_1_.deserialize(entry.getValue(), BlockPartFace.class));
                 }
 
                 return map;
@@ -182,7 +183,7 @@
             {
                 Vector3f vector3f = this.func_178251_a(p_178247_1_, "to");
 
-                if (!(vector3f.x < -16.0F) && !(vector3f.y < -16.0F) && !(vector3f.z < -16.0F) && !(vector3f.x > 32.0F) && !(vector3f.y > 32.0F) && !(vector3f.z > 32.0F))
+                if (vector3f.x >= -16.0F && vector3f.y >= -16.0F && vector3f.z >= -16.0F && vector3f.x <= 32.0F && vector3f.y <= 32.0F && vector3f.z <= 32.0F)
                 {
                     return vector3f;
                 }
@@ -196,7 +197,7 @@
             {
                 Vector3f vector3f = this.func_178251_a(p_178249_1_, "from");
 
-                if (!(vector3f.x < -16.0F) && !(vector3f.y < -16.0F) && !(vector3f.z < -16.0F) && !(vector3f.x > 32.0F) && !(vector3f.y > 32.0F) && !(vector3f.z > 32.0F))
+                if (vector3f.x >= -16.0F && vector3f.y >= -16.0F && vector3f.z >= -16.0F && vector3f.x <= 32.0F && vector3f.y <= 32.0F && vector3f.z <= 32.0F)
                 {
                     return vector3f;
                 }
