--- before/net/minecraft/client/renderer/block/model/BlockPartRotation.java
+++ after/net/minecraft/client/renderer/block/model/BlockPartRotation.java
@@ -3,7 +3,7 @@
 import net.minecraft.util.EnumFacing;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
-import org.lwjgl.util.vector.Vector3f;
+import org.lwjglx.util.vector.Vector3f;
 
 @SideOnly(Side.CLIENT)
 public class BlockPartRotation
@@ -13,11 +13,11 @@
     public final float field_178343_c;
     public final boolean field_178341_d;
 
-    public BlockPartRotation(Vector3f p_i46229_1_, EnumFacing.Axis p_i46229_2_, float p_i46229_3_, boolean p_i46229_4_)
+    public BlockPartRotation(Vector3f originIn, EnumFacing.Axis axisIn, float angleIn, boolean rescaleIn)
     {
-        this.field_178344_a = p_i46229_1_;
-        this.field_178342_b = p_i46229_2_;
-        this.field_178343_c = p_i46229_3_;
-        this.field_178341_d = p_i46229_4_;
+        this.field_178344_a = originIn;
+        this.field_178342_b = axisIn;
+        this.field_178343_c = angleIn;
+        this.field_178341_d = rescaleIn;
     }
 }
