--- before/net/minecraft/client/renderer/block/model/BlockPartFace.java
+++ after/net/minecraft/client/renderer/block/model/BlockPartFace.java
@@ -38,7 +38,7 @@
                 EnumFacing enumfacing = this.func_178339_c(jsonobject);
                 int i = this.func_178337_a(jsonobject);
                 String s = this.func_178340_b(jsonobject);
-                BlockFaceUV blockfaceuv = p_deserialize_3_.deserialize(jsonobject, BlockFaceUV.class);
+                BlockFaceUV blockfaceuv = (BlockFaceUV)p_deserialize_3_.deserialize(jsonobject, BlockFaceUV.class);
                 return new BlockPartFace(enumfacing, i, s, blockfaceuv);
             }
 
