--- before/net/minecraft/client/renderer/block/model/BlockPart.java
+++ after/net/minecraft/client/renderer/block/model/BlockPart.java
@@ -17,7 +17,7 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
-import org.lwjgl.util.vector.Vector3f;
+import org.lwjglx.util.vector.Vector3f;
 
 @SideOnly(Side.CLIENT)
 public class BlockPart
