--- before/net/minecraft/client/renderer/block/model/ItemModelGenerator.java
+++ after/net/minecraft/client/renderer/block/model/ItemModelGenerator.java
@@ -11,7 +11,7 @@
 import net.minecraft.util.ResourceLocation;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
-import org.lwjgl.util.vector.Vector3f;
+import org.lwjglx.util.vector.Vector3f;
 
 @SideOnly(Side.CLIENT)
 public class ItemModelGenerator
