--- before/net/minecraft/client/renderer/block/model/ItemOverride.java
+++ after/net/minecraft/client/renderer/block/model/ItemOverride.java
@@ -8,6 +8,7 @@
 import com.google.gson.JsonParseException;
 import java.lang.reflect.Type;
 import java.util.Map;
+import java.util.Map.Entry;
 import javax.annotation.Nullable;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.item.IItemPropertyGetter;
@@ -40,11 +41,11 @@
     {
         Item item = p_188027_1_.func_77973_b();
 
-        for (Map.Entry<ResourceLocation, Float> entry : this.field_188029_b.entrySet())
+        for (Entry<ResourceLocation, Float> entry : this.field_188029_b.entrySet())
         {
             IItemPropertyGetter iitempropertygetter = item.func_185045_a(entry.getKey());
 
-            if (iitempropertygetter == null || iitempropertygetter.func_185085_a(p_188027_1_, p_188027_2_, p_188027_3_) < entry.getValue())
+            if (iitempropertygetter == null || iitempropertygetter.func_185085_a(p_188027_1_, p_188027_2_, p_188027_3_) < ((Float)entry.getValue()).floatValue())
             {
                 return false;
             }
@@ -66,12 +67,12 @@
 
             protected Map<ResourceLocation, Float> func_188025_a(JsonObject p_188025_1_)
             {
-                Map<ResourceLocation, Float> map = Maps.newLinkedHashMap();
+                Map<ResourceLocation, Float> map = Maps.<ResourceLocation, Float>newLinkedHashMap();
                 JsonObject jsonobject = JsonUtils.func_152754_s(p_188025_1_, "predicate");
 
-                for (Map.Entry<String, JsonElement> entry : jsonobject.entrySet())
+                for (Entry<String, JsonElement> entry : jsonobject.entrySet())
                 {
-                    map.put(new ResourceLocation(entry.getKey()), JsonUtils.func_151220_d(entry.getValue(), entry.getKey()));
+                    map.put(new ResourceLocation(entry.getKey()), Float.valueOf(JsonUtils.func_151220_d(entry.getValue(), entry.getKey())));
                 }
 
                 return map;
