--- before/net/minecraft/client/renderer/block/model/ItemCameraTransforms.java
+++ after/net/minecraft/client/renderer/block/model/ItemCameraTransforms.java
@@ -12,6 +12,9 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.lwjgl.util.vector.Quaternion;
 
+/*
+ * @deprecated use {@link net.minecraftforge.client.model.IPerspectiveAwareModel} instead
+ */
 @SideOnly(Side.CLIENT)
 public class ItemCameraTransforms
 {
@@ -48,6 +51,7 @@
         );
     }
 
+    @Deprecated
     public ItemCameraTransforms(ItemCameraTransforms transforms)
     {
         this.thirdperson_left = transforms.thirdperson_left;
@@ -60,6 +64,7 @@
         this.fixed = transforms.fixed;
     }
 
+    @Deprecated
     public ItemCameraTransforms(
         ItemTransformVec3f thirdperson_leftIn,
         ItemTransformVec3f thirdperson_rightIn,
@@ -125,6 +130,7 @@
         return new Quaternion(f3 * f6 * f8 + f4 * f5 * f7, f4 * f5 * f8 - f3 * f6 * f7, f3 * f5 * f8 + f4 * f6 * f7, f4 * f6 * f8 - f3 * f5 * f7);
     }
 
+    @Deprecated
     public ItemTransformVec3f getTransform(ItemCameraTransforms.TransformType type)
     {
         switch (type)
@@ -202,7 +208,7 @@
         }
 
     @SideOnly(Side.CLIENT)
-    public static enum TransformType
+    public static enum TransformType implements net.minecraftforge.common.model.IModelPart
     {
         NONE,
         THIRD_PERSON_LEFT_HAND,
