--- before/net/minecraft/client/renderer/block/model/ModelBlockDefinition.java
+++ after/net/minecraft/client/renderer/block/model/ModelBlockDefinition.java
@@ -40,7 +40,11 @@
 
     public static ModelBlockDefinition parseFromReader(Reader reader)
     {
-        return JsonUtils.fromJson(GSON, reader, ModelBlockDefinition.class);
+        return parseFromReader(reader, null);
+    }
+
+    public static ModelBlockDefinition parseFromReader(Reader reader, net.minecraft.util.ResourceLocation location) {
+        return net.minecraftforge.client.model.BlockStateLoader.load(reader, location, GSON);
     }
 
     public ModelBlockDefinition(Map<String, VariantList> variants, Multipart multipartIn)
