--- before/net/minecraft/client/renderer/block/model/SimpleBakedModel.java
+++ after/net/minecraft/client/renderer/block/model/SimpleBakedModel.java
@@ -24,15 +24,7 @@
     protected final ItemCameraTransforms field_177558_f;
     protected final ItemOverrideList field_188620_g;
 
-    public SimpleBakedModel(
-        List<BakedQuad> p_i46535_1_,
-        Map<EnumFacing, List<BakedQuad>> p_i46535_2_,
-        boolean p_i46535_3_,
-        boolean p_i46535_4_,
-        TextureAtlasSprite p_i46535_5_,
-        ItemCameraTransforms p_i46535_6_,
-        ItemOverrideList p_i46535_7_
-    )
+    public SimpleBakedModel(List<BakedQuad> p_i46535_1_, Map<EnumFacing, List<BakedQuad>> p_i46535_2_, boolean p_i46535_3_, boolean p_i46535_4_, TextureAtlasSprite p_i46535_5_, ItemCameraTransforms p_i46535_6_, ItemOverrideList p_i46535_7_)
     {
         this.field_177563_a = p_i46535_1_;
         this.field_177561_b = p_i46535_2_;
@@ -43,43 +35,36 @@
         this.field_188620_g = p_i46535_7_;
     }
 
-    @Override
     public List<BakedQuad> func_188616_a(@Nullable IBlockState p_188616_1_, @Nullable EnumFacing p_188616_2_, long p_188616_3_)
     {
-        return p_188616_2_ == null ? this.field_177563_a : this.field_177561_b.get(p_188616_2_);
+        return p_188616_2_ == null ? this.field_177563_a : (List)this.field_177561_b.get(p_188616_2_);
     }
 
-    @Override
     public boolean func_177555_b()
     {
         return this.field_177562_c;
     }
 
-    @Override
     public boolean func_177556_c()
     {
         return this.field_177559_d;
     }
 
-    @Override
     public boolean func_188618_c()
     {
         return false;
     }
 
-    @Override
     public TextureAtlasSprite func_177554_e()
     {
         return this.field_177560_e;
     }
 
-    @Override
     public ItemCameraTransforms func_177552_f()
     {
         return this.field_177558_f;
     }
 
-    @Override
     public ItemOverrideList func_188617_f()
     {
         return this.field_188620_g;
@@ -88,8 +73,8 @@
     @SideOnly(Side.CLIENT)
     public static class Builder
         {
-            private final List<BakedQuad> field_177656_a = Lists.newArrayList();
-            private final Map<EnumFacing, List<BakedQuad>> field_177654_b = Maps.newEnumMap(EnumFacing.class);
+            private final List<BakedQuad> field_177656_a;
+            private final Map<EnumFacing, List<BakedQuad>> field_177654_b;
             private final ItemOverrideList field_188646_c;
             private final boolean field_177655_c;
             private TextureAtlasSprite field_177652_d;
@@ -103,7 +88,7 @@
 
             public Builder(IBlockState p_i46989_1_, IBakedModel p_i46989_2_, TextureAtlasSprite p_i46989_3_, BlockPos p_i46989_4_)
             {
-                this(p_i46989_2_.func_177555_b(), p_i46989_2_.func_177556_c(), p_i46989_2_.func_177552_f(), p_i46989_2_.func_188617_f());
+                this(p_i46989_2_.isAmbientOcclusion(p_i46989_1_), p_i46989_2_.func_177556_c(), p_i46989_2_.func_177552_f(), p_i46989_2_.func_188617_f());
                 this.field_177652_d = p_i46989_2_.func_177554_e();
                 long i = MathHelper.func_180186_a(p_i46989_4_);
 
@@ -117,6 +102,9 @@
 
             private Builder(boolean p_i46990_1_, boolean p_i46990_2_, ItemCameraTransforms p_i46990_3_, ItemOverrideList p_i46990_4_)
             {
+                this.field_177656_a = Lists.<BakedQuad>newArrayList();
+                this.field_177654_b = Maps.newEnumMap(EnumFacing.class);
+
                 for (EnumFacing enumfacing : EnumFacing.values())
                 {
                     this.field_177654_b.put(enumfacing, Lists.newArrayList());
@@ -138,7 +126,7 @@
 
             private void func_188645_a(IBlockState p_188645_1_, IBakedModel p_188645_2_, TextureAtlasSprite p_188645_3_, long p_188645_4_)
             {
-                for (BakedQuad bakedquad : p_188645_2_.func_188616_a(p_188645_1_, null, p_188645_4_))
+                for (BakedQuad bakedquad : p_188645_2_.func_188616_a(p_188645_1_, (EnumFacing)null, p_188645_4_))
                 {
                     this.func_177648_a(new BakedQuadRetextured(bakedquad, p_188645_3_));
                 }
@@ -146,7 +134,7 @@
 
             public SimpleBakedModel.Builder func_177650_a(EnumFacing p_177650_1_, BakedQuad p_177650_2_)
             {
-                this.field_177654_b.get(p_177650_1_).add(p_177650_2_);
+                (this.field_177654_b.get(p_177650_1_)).add(p_177650_2_);
                 return this;
             }
 
@@ -170,15 +158,7 @@
                 }
                 else
                 {
-                    return new SimpleBakedModel(
-                               this.field_177656_a,
-                               this.field_177654_b,
-                               this.field_177655_c,
-                               this.field_177653_e,
-                               this.field_177652_d,
-                               this.field_177651_f,
-                               this.field_188646_c
-                           );
+                    return new SimpleBakedModel(this.field_177656_a, this.field_177654_b, this.field_177655_c, this.field_177653_e, this.field_177652_d, this.field_177651_f, this.field_188646_c);
                 }
             }
         }
