--- before/net/minecraft/client/renderer/block/model/SimpleBakedModel.java
+++ after/net/minecraft/client/renderer/block/model/SimpleBakedModel.java
@@ -37,7 +37,7 @@
 
     public List<BakedQuad> func_188616_a(@Nullable IBlockState p_188616_1_, @Nullable EnumFacing p_188616_2_, long p_188616_3_)
     {
-        return p_188616_2_ == null ? this.field_177563_a : this.field_177561_b.get(p_188616_2_);
+        return p_188616_2_ == null ? this.field_177563_a : (List)this.field_177561_b.get(p_188616_2_);
     }
 
     public boolean func_177555_b()
@@ -73,8 +73,8 @@
     @SideOnly(Side.CLIENT)
     public static class Builder
         {
-            private final List<BakedQuad> field_177656_a = Lists.newArrayList();
-            private final Map<EnumFacing, List<BakedQuad>> field_177654_b = Maps.newEnumMap(EnumFacing.class);
+            private final List<BakedQuad> field_177656_a;
+            private final Map<EnumFacing, List<BakedQuad>> field_177654_b;
             private final ItemOverrideList field_188646_c;
             private final boolean field_177655_c;
             private TextureAtlasSprite field_177652_d;
@@ -88,7 +88,7 @@
 
             public Builder(IBlockState p_i46989_1_, IBakedModel p_i46989_2_, TextureAtlasSprite p_i46989_3_, BlockPos p_i46989_4_)
             {
-                this(p_i46989_2_.func_177555_b(), p_i46989_2_.func_177556_c(), p_i46989_2_.func_177552_f(), p_i46989_2_.func_188617_f());
+                this(p_i46989_2_.isAmbientOcclusion(p_i46989_1_), p_i46989_2_.func_177556_c(), p_i46989_2_.func_177552_f(), p_i46989_2_.func_188617_f());
                 this.field_177652_d = p_i46989_2_.func_177554_e();
                 long i = MathHelper.func_180186_a(p_i46989_4_);
 
@@ -102,6 +102,9 @@
 
             private Builder(boolean p_i46990_1_, boolean p_i46990_2_, ItemCameraTransforms p_i46990_3_, ItemOverrideList p_i46990_4_)
             {
+                this.field_177656_a = Lists.<BakedQuad>newArrayList();
+                this.field_177654_b = Maps.newEnumMap(EnumFacing.class);
+
                 for (EnumFacing enumfacing : EnumFacing.values())
                 {
                     this.field_177654_b.put(enumfacing, Lists.newArrayList());
@@ -131,7 +134,7 @@
 
             public SimpleBakedModel.Builder func_177650_a(EnumFacing p_177650_1_, BakedQuad p_177650_2_)
             {
-                this.field_177654_b.get(p_177650_1_).add(p_177650_2_);
+                (this.field_177654_b.get(p_177650_1_)).add(p_177650_2_);
                 return this;
             }
 
