--- before/net/minecraft/client/renderer/block/model/VariantList.java
+++ after/net/minecraft/client/renderer/block/model/VariantList.java
@@ -53,7 +53,7 @@
         {
             public VariantList deserialize(JsonElement p_deserialize_1_, Type p_deserialize_2_, JsonDeserializationContext p_deserialize_3_) throws JsonParseException
             {
-                List<Variant> list = Lists.newArrayList();
+                List<Variant> list = Lists.<Variant>newArrayList();
 
                 if (p_deserialize_1_.isJsonArray())
                 {
@@ -66,12 +66,12 @@
 
                     for (JsonElement jsonelement : jsonarray)
                     {
-                        list.add(p_deserialize_3_.deserialize(jsonelement, Variant.class));
+                        list.add((Variant)p_deserialize_3_.deserialize(jsonelement, Variant.class));
                     }
                 }
                 else
                 {
-                    list.add(p_deserialize_3_.deserialize(p_deserialize_1_, Variant.class));
+                    list.add((Variant)p_deserialize_3_.deserialize(p_deserialize_1_, Variant.class));
                 }
 
                 return new VariantList(list);
