--- before/net/minecraft/client/renderer/block/model/Variant.java
+++ after/net/minecraft/client/renderer/block/model/Variant.java
@@ -12,7 +12,7 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class Variant
+public class Variant implements net.minecraftforge.client.model.ISmartVariant
 {
     private final ResourceLocation field_188050_a;
     private final ModelRotation field_188051_b;
@@ -32,11 +32,17 @@
         return this.field_188050_a;
     }
 
+    @Deprecated
     public ModelRotation func_188048_b()
     {
         return this.field_188051_b;
     }
 
+    public net.minecraftforge.common.model.IModelState getState()
+    {
+        return this.field_188051_b;
+    }
+
     public boolean func_188049_c()
     {
         return this.field_188052_c;
@@ -47,21 +53,11 @@
         return this.field_188053_d;
     }
 
-    @Override
     public String toString()
     {
-        return "Variant{modelLocation="
-               + this.field_188050_a
-               + ", rotation="
-               + this.field_188051_b
-               + ", uvLock="
-               + this.field_188052_c
-               + ", weight="
-               + this.field_188053_d
-               + '}';
+        return "Variant{modelLocation=" + this.field_188050_a + ", rotation=" + this.field_188051_b + ", uvLock=" + this.field_188052_c + ", weight=" + this.field_188053_d + '}';
     }
 
-    @Override
     public boolean equals(Object p_equals_1_)
     {
         if (this == p_equals_1_)
@@ -75,20 +71,17 @@
         else
         {
             Variant variant = (Variant)p_equals_1_;
-            return this.field_188050_a.equals(variant.field_188050_a)
-                   && this.field_188051_b == variant.field_188051_b
-                   && this.field_188052_c == variant.field_188052_c
-                   && this.field_188053_d == variant.field_188053_d;
+            return this.field_188050_a.equals(variant.field_188050_a) && this.field_188051_b == variant.field_188051_b && this.field_188052_c == variant.field_188052_c && this.field_188053_d == variant.field_188053_d;
         }
     }
 
-    @Override
     public int hashCode()
     {
         int i = this.field_188050_a.hashCode();
         i = 31 * i + this.field_188051_b.hashCode();
         i = 31 * i + Boolean.valueOf(this.field_188052_c).hashCode();
-        return 31 * i + this.field_188053_d;
+        i = 31 * i + this.field_188053_d;
+        return i;
     }
 
     @SideOnly(Side.CLIENT)
@@ -107,7 +100,8 @@
             private ResourceLocation func_188041_a(String p_188041_1_)
             {
                 ResourceLocation resourcelocation = new ResourceLocation(p_188041_1_);
-                return new ResourceLocation(resourcelocation.func_110624_b(), "block/" + resourcelocation.func_110623_a());
+                resourcelocation = new ResourceLocation(resourcelocation.func_110624_b(), "block/" + resourcelocation.func_110623_a());
+                return resourcelocation;
             }
 
             private boolean func_188044_d(JsonObject p_188044_1_)
@@ -150,4 +144,9 @@
                 }
             }
         }
+
+    public net.minecraftforge.client.model.IModel process(net.minecraftforge.client.model.IModel base)
+    {
+        return base.uvlock(func_188049_c());
+    }
 }
