--- before/net/minecraft/client/renderer/block/model/WeightedBakedModel.java
+++ after/net/minecraft/client/renderer/block/model/WeightedBakedModel.java
@@ -44,6 +44,12 @@
     }
 
     @Override
+    public boolean isAmbientOcclusion(IBlockState state)
+    {
+        return this.baseModel.isAmbientOcclusion(state);
+    }
+
+    @Override
     public boolean isGui3d()
     {
         return this.baseModel.isGui3d();
