--- before/net/minecraft/client/gui/inventory/GuiCrafting.java
+++ after/net/minecraft/client/gui/inventory/GuiCrafting.java
@@ -22,7 +22,7 @@
 {
     private static final ResourceLocation field_147019_u = new ResourceLocation("textures/gui/container/crafting_table.png");
     private GuiButtonImage field_192049_w;
-    private final GuiRecipeBook field_192050_x = new GuiRecipeBook();
+    private final GuiRecipeBook field_192050_x;
     private boolean field_193112_y;
 
     public GuiCrafting(InventoryPlayer p_i45504_1_, World p_i45504_2_)
@@ -33,6 +33,7 @@
     public GuiCrafting(InventoryPlayer p_i45505_1_, World p_i45505_2_, BlockPos p_i45505_3_)
     {
         super(new ContainerWorkbench(p_i45505_1_, p_i45505_2_, p_i45505_3_));
+        this.field_192050_x = new GuiRecipeBook();
     }
 
     public void func_73866_w_()
