--- before/net/minecraft/client/gui/inventory/GuiEditStructure.java
+++ after/net/minecraft/client/gui/inventory/GuiEditStructure.java
@@ -22,7 +22,7 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import org.lwjgl.input.Keyboard;
+import org.lwjglx.input.Keyboard;
 
 @SideOnly(Side.CLIENT)
 public class GuiEditStructure extends GuiScreen
