--- before/net/minecraft/client/gui/inventory/GuiEditCommandBlockMinecart.java
+++ after/net/minecraft/client/gui/inventory/GuiEditCommandBlockMinecart.java
@@ -16,7 +16,7 @@
 import net.minecraft.util.text.ITextComponent;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
-import org.lwjgl.input.Keyboard;
+import org.lwjglx.input.Keyboard;
 
 @SideOnly(Side.CLIENT)
 public class GuiEditCommandBlockMinecart extends GuiScreen implements ITabCompleter
