--- before/net/minecraft/client/gui/inventory/GuiEditSign.java
+++ after/net/minecraft/client/gui/inventory/GuiEditSign.java
@@ -15,7 +15,7 @@
 import net.minecraft.util.text.TextComponentString;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
-import org.lwjgl.input.Keyboard;
+import org.lwjglx.input.Keyboard;
 
 @SideOnly(Side.CLIENT)
 public class GuiEditSign extends GuiScreen
