--- before/net/minecraft/client/audio/PositionedSound.java
+++ after/net/minecraft/client/audio/PositionedSound.java
@@ -15,14 +15,14 @@
     private SoundEventAccessor field_184369_l;
     protected SoundCategory field_184368_b;
     protected ResourceLocation field_147664_a;
-    protected float field_147662_b = 1.0F;
-    protected float field_147663_c = 1.0F;
+    protected float field_147662_b;
+    protected float field_147663_c;
     protected float field_147660_d;
     protected float field_147661_e;
     protected float field_147658_f;
     protected boolean field_147659_g;
     protected int field_147665_h;
-    protected ISound.AttenuationType field_147666_i = ISound.AttenuationType.LINEAR;
+    protected ISound.AttenuationType field_147666_i;
 
     protected PositionedSound(SoundEvent p_i46533_1_, SoundCategory p_i46533_2_)
     {
@@ -31,6 +31,9 @@
 
     protected PositionedSound(ResourceLocation p_i46534_1_, SoundCategory p_i46534_2_)
     {
+        this.field_147662_b = 1.0F;
+        this.field_147663_c = 1.0F;
+        this.field_147666_i = ISound.AttenuationType.LINEAR;
         this.field_147664_a = p_i46534_1_;
         this.field_184368_b = p_i46534_2_;
     }
