--- before/net/minecraft/client/audio/SoundHandler.java
+++ after/net/minecraft/client/audio/SoundHandler.java
@@ -2,6 +2,7 @@
 
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
+import java.io.Closeable;
 import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
@@ -36,23 +37,17 @@
 {
     public static final Sound field_147700_a = new Sound("meta:missing_sound", 1.0F, 1.0F, 1, Sound.Type.FILE, false);
     private static final Logger field_147698_b = LogManager.getLogger();
-    private static final Gson field_147699_c = new GsonBuilder()
-        .registerTypeHierarchyAdapter(ITextComponent.class, new ITextComponent.Serializer())
-        .registerTypeAdapter(SoundList.class, new SoundListSerializer())
-        .create();
+    private static final Gson field_147699_c = (new GsonBuilder()).registerTypeHierarchyAdapter(ITextComponent.class, new ITextComponent.Serializer()).registerTypeAdapter(SoundList.class, new SoundListSerializer()).create();
     private static final ParameterizedType field_147696_d = new ParameterizedType()
     {
-        @Override
         public Type[] getActualTypeArguments()
         {
             return new Type[] {String.class, SoundList.class};
         }
-        @Override
         public Type getRawType()
         {
             return Map.class;
         }
-        @Override
         public Type getOwnerType()
         {
             return null;
@@ -68,11 +63,11 @@
         this.field_147694_f = new SoundManager(this, p_i45122_2_);
     }
 
-    @Override
     public void func_110549_a(IResourceManager p_110549_1_)
     {
         this.field_147697_e.func_148763_c();
 
+        java.util.List<net.minecraft.util.Tuple<ResourceLocation, SoundList>> resources = new java.util.LinkedList<>();
         for (String s : p_110549_1_.func_135055_a())
         {
             try
@@ -85,7 +80,7 @@
 
                         for (Entry<String, SoundList> entry : map.entrySet())
                         {
-                            this.func_147693_a(new ResourceLocation(s, entry.getKey()), entry.getValue());
+                            resources.add(new net.minecraft.util.Tuple<>(new ResourceLocation(s, entry.getKey()), entry.getValue()));
                         }
                     }
                     catch (RuntimeException runtimeexception)
@@ -94,14 +89,29 @@
                     }
                 }
             }
-            catch (IOException ioexception)
+            catch (IOException var11)
             {
+                ;
             }
         }
 
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar resourcesBar = net.minecraftforge.fml.common.ProgressManager.push("Loading sounds", resources.size());
+        resources.forEach(entry ->
+        {
+            resourcesBar.step(entry.func_76341_a().toString());
+            try
+            {
+                this.func_147693_a(entry.func_76341_a(), entry.func_76340_b());
+            }
+            catch (RuntimeException e)
+            {
+                field_147698_b.warn("Invalid sounds.json", e);
+            }
+        });
+        net.minecraftforge.fml.common.ProgressManager.pop(resourcesBar);
         for (ResourceLocation resourcelocation : this.field_147697_e.func_148742_b())
         {
-            SoundEventAccessor soundeventaccessor = this.field_147697_e.func_82594_a(resourcelocation);
+            SoundEventAccessor soundeventaccessor = (SoundEventAccessor)this.field_147697_e.func_82594_a(resourcelocation);
 
             if (soundeventaccessor.func_188712_c() instanceof TextComponentTranslation)
             {
@@ -118,7 +128,7 @@
         {
             if (SoundEvent.field_187505_a.func_82594_a(resourcelocation1) == null)
             {
-                field_147698_b.debug("Not having sound event for: {}", resourcelocation1);
+                field_147698_b.debug("Not having sound event for: {}", (Object)resourcelocation1);
             }
         }
 
@@ -132,7 +142,7 @@
 
         try
         {
-            map = JsonUtils.func_193841_a(field_147699_c, new InputStreamReader(p_175085_1_, StandardCharsets.UTF_8), field_147696_d);
+            map = (Map)JsonUtils.func_193841_a(field_147699_c, new InputStreamReader(p_175085_1_, StandardCharsets.UTF_8), field_147696_d);
         }
         finally
         {
@@ -144,14 +154,14 @@
 
     private void func_147693_a(ResourceLocation p_147693_1_, SoundList p_147693_2_)
     {
-        SoundEventAccessor soundeventaccessor = this.field_147697_e.func_82594_a(p_147693_1_);
+        SoundEventAccessor soundeventaccessor = (SoundEventAccessor)this.field_147697_e.func_82594_a(p_147693_1_);
         boolean flag = soundeventaccessor == null;
 
         if (flag || p_147693_2_.func_148574_b())
         {
             if (!flag)
             {
-                field_147698_b.debug("Replaced sound event location {}", p_147693_1_);
+                field_147698_b.debug("Replaced sound event location {}", (Object)p_147693_1_);
             }
 
             soundeventaccessor = new SoundEventAccessor(p_147693_1_, p_147693_2_.func_188701_c());
@@ -177,15 +187,14 @@
                 case SOUND_EVENT:
                     isoundeventaccessor = new ISoundEventAccessor<Sound>()
                     {
-                        @Override
                         public int func_148721_a()
                         {
-                            SoundEventAccessor soundeventaccessor1 = SoundHandler.this.field_147697_e.func_82594_a(resourcelocation);
+                            SoundEventAccessor soundeventaccessor1 = (SoundEventAccessor)SoundHandler.this.field_147697_e.func_82594_a(resourcelocation);
                             return soundeventaccessor1 == null ? 0 : soundeventaccessor1.func_148721_a();
                         }
                         public Sound func_148720_g()
                         {
-                            SoundEventAccessor soundeventaccessor1 = SoundHandler.this.field_147697_e.func_82594_a(resourcelocation);
+                            SoundEventAccessor soundeventaccessor1 = (SoundEventAccessor)SoundHandler.this.field_147697_e.func_82594_a(resourcelocation);
 
                             if (soundeventaccessor1 == null)
                             {
@@ -194,14 +203,7 @@
                             else
                             {
                                 Sound sound1 = soundeventaccessor1.func_148720_g();
-                                return new Sound(
-                                           sound1.func_188719_a().toString(),
-                                           sound1.func_188724_c() * sound.func_188724_c(),
-                                           sound1.func_188725_d() * sound.func_188725_d(),
-                                           sound.func_148721_a(),
-                                           Sound.Type.FILE,
-                                           sound1.func_188723_h() || sound.func_188723_h()
-                                       );
+                                return new Sound(sound1.func_188719_a().toString(), sound1.func_188724_c() * sound.func_188724_c(), sound1.func_188725_d() * sound.func_188725_d(), sound.func_148721_a(), Sound.Type.FILE, sound1.func_188723_h() || sound.func_188723_h());
                             }
                         }
                     };
@@ -226,19 +228,20 @@
             iresource.func_110527_b();
             return true;
         }
-        catch (FileNotFoundException filenotfoundexception)
+        catch (FileNotFoundException var11)
         {
             field_147698_b.warn("File {} does not exist, cannot add it to event {}", resourcelocation, p_184401_2_);
-            return false;
+            flag = false;
         }
         catch (IOException ioexception)
         {
             field_147698_b.warn("Could not load sound file {}, cannot add it to event {}", resourcelocation, p_184401_2_, ioexception);
             flag = false;
+            return flag;
         }
         finally
         {
-            IOUtils.closeQuietly(iresource);
+            IOUtils.closeQuietly((Closeable)iresource);
         }
 
         return flag;
@@ -247,7 +250,7 @@
     @Nullable
     public SoundEventAccessor func_184398_a(ResourceLocation p_184398_1_)
     {
-        return this.field_147697_e.func_82594_a(p_184398_1_);
+        return (SoundEventAccessor)this.field_147697_e.func_82594_a(p_184398_1_);
     }
 
     public void func_147682_a(ISound p_147682_1_)
@@ -265,6 +268,11 @@
         this.field_147694_f.func_148615_a(p_147691_1_, p_147691_2_);
     }
 
+    public void setListener(net.minecraft.entity.Entity entity, float partialTicks)
+    {
+        this.field_147694_f.setListener(entity, partialTicks);
+    }
+
     public void func_147689_b()
     {
         this.field_147694_f.func_148610_e();
@@ -280,7 +288,6 @@
         this.field_147694_f.func_148613_b();
     }
 
-    @Override
     public void func_73660_a()
     {
         this.field_147694_f.func_148605_d();
