--- before/net/minecraft/village/MerchantRecipe.java
+++ after/net/minecraft/village/MerchantRecipe.java
@@ -8,15 +8,18 @@
 
 public class MerchantRecipe
 {
-    private ItemStack field_77403_a = ItemStack.field_190927_a;
-    private ItemStack field_77401_b = ItemStack.field_190927_a;
-    private ItemStack field_77402_c = ItemStack.field_190927_a;
+    private ItemStack field_77403_a;
+    private ItemStack field_77401_b;
+    private ItemStack field_77402_c;
     private int field_77400_d;
     private int field_82786_e;
     private boolean field_180323_f;
 
     public MerchantRecipe(NBTTagCompound p_i1940_1_)
     {
+        this.field_77403_a = ItemStack.field_190927_a;
+        this.field_77401_b = ItemStack.field_190927_a;
+        this.field_77402_c = ItemStack.field_190927_a;
         this.func_77390_a(p_i1940_1_);
     }
 
@@ -27,6 +30,9 @@
 
     public MerchantRecipe(ItemStack p_i45760_1_, ItemStack p_i45760_2_, ItemStack p_i45760_3_, int p_i45760_4_, int p_i45760_5_)
     {
+        this.field_77403_a = ItemStack.field_190927_a;
+        this.field_77401_b = ItemStack.field_190927_a;
+        this.field_77402_c = ItemStack.field_190927_a;
         this.field_77403_a = p_i45760_1_;
         this.field_77401_b = p_i45760_2_;
         this.field_77402_c = p_i45760_3_;
