--- before/net/minecraft/village/VillageSiege.java
+++ after/net/minecraft/village/VillageSiege.java
@@ -136,6 +136,7 @@
 
                     if (vec3d != null)
                     {
+                        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.village.VillageSiegeEvent(this, world, entityplayer, village, vec3d))) return false;
                         this.nextSpawnTime = 0;
                         this.siegeCount = 20;
                         return true;
