--- before/net/minecraft/village/VillageCollection.java
+++ after/net/minecraft/village/VillageCollection.java
@@ -19,9 +19,9 @@
 public class VillageCollection extends WorldSavedData
 {
     private World field_75556_a;
-    private final List<BlockPos> field_75554_b = Lists.newArrayList();
-    private final List<VillageDoorInfo> field_75555_c = Lists.newArrayList();
-    private final List<Village> field_75552_d = Lists.newArrayList();
+    private final List<BlockPos> field_75554_b = Lists.<BlockPos>newArrayList();
+    private final List<VillageDoorInfo> field_75555_c = Lists.<VillageDoorInfo>newArrayList();
+    private final List<Village> field_75552_d = Lists.<Village>newArrayList();
     private int field_75553_e;
 
     public VillageCollection(String p_i1677_1_)
@@ -59,7 +59,7 @@
 
     public void func_75544_a()
     {
-        this.field_75553_e++;
+        ++this.field_75553_e;
 
         for (Village village : this.field_75552_d)
         {
@@ -100,17 +100,17 @@
     public Village func_176056_a(BlockPos p_176056_1_, int p_176056_2_)
     {
         Village village = null;
-        double d0 = Float.MAX_VALUE;
+        double d0 = 3.4028234663852886E38D;
 
         for (Village village1 : this.field_75552_d)
         {
             double d1 = village1.func_180608_a().func_177951_i(p_176056_1_);
 
-            if (!(d1 >= d0))
+            if (d1 < d0)
             {
                 float f = (float)(p_176056_2_ + village1.func_75568_b());
 
-                if (!(d1 > (double)(f * f)))
+                if (d1 <= (double)(f * f))
                 {
                     village = village1;
                     d0 = d1;
@@ -131,7 +131,7 @@
 
     private void func_75545_e()
     {
-        for (int i = 0; i < this.field_75555_c.size(); i++)
+        for (int i = 0; i < this.field_75555_c.size(); ++i)
         {
             VillageDoorInfo villagedoorinfo = this.field_75555_c.get(i);
             Village village = this.func_176056_a(villagedoorinfo.func_179852_d(), 32);
@@ -151,15 +151,16 @@
 
     private void func_180609_b(BlockPos p_180609_1_)
     {
+        if (!this.field_75556_a.func_175697_a(p_180609_1_, 16)) return; // Forge: prevent loading unloaded chunks when checking for doors
         int i = 16;
         int j = 4;
         int k = 16;
 
-        for (int l = -16; l < 16; l++)
+        for (int l = -16; l < 16; ++l)
         {
-            for (int i1 = -4; i1 < 4; i1++)
+            for (int i1 = -4; i1 < 4; ++i1)
             {
-                for (int j1 = -16; j1 < 16; j1++)
+                for (int j1 = -16; j1 < 16; ++j1)
                 {
                     BlockPos blockpos = p_180609_1_.func_177982_a(l, i1, j1);
 
@@ -186,9 +187,7 @@
     {
         for (VillageDoorInfo villagedoorinfo : this.field_75555_c)
         {
-            if (villagedoorinfo.func_179852_d().func_177958_n() == p_176055_1_.func_177958_n()
-                    && villagedoorinfo.func_179852_d().func_177952_p() == p_176055_1_.func_177952_p()
-                    && Math.abs(villagedoorinfo.func_179852_d().func_177956_o() - p_176055_1_.func_177956_o()) <= 1)
+            if (villagedoorinfo.func_179852_d().func_177958_n() == p_176055_1_.func_177958_n() && villagedoorinfo.func_179852_d().func_177952_p() == p_176055_1_.func_177952_p() && Math.abs(villagedoorinfo.func_179852_d().func_177956_o() - p_176055_1_.func_177956_o()) <= 1)
             {
                 return villagedoorinfo;
             }
@@ -224,11 +223,13 @@
     {
         int i = 0;
 
-        for (int j = 1; j <= 5; j++)
+        for (int j = 1; j <= 5; ++j)
         {
             if (this.field_75556_a.func_175678_i(p_176061_1_.func_177967_a(p_176061_2_, j)))
             {
-                if (++i >= p_176061_3_)
+                ++i;
+
+                if (i >= p_176061_3_)
                 {
                     return i;
                 }
@@ -255,16 +256,23 @@
     {
         IBlockState iblockstate = this.field_75556_a.func_180495_p(p_176058_1_);
         Block block = iblockstate.func_177230_c();
-        return block instanceof BlockDoor ? iblockstate.func_185904_a() == Material.field_151575_d : false;
+
+        if (block instanceof BlockDoor)
+        {
+            return iblockstate.func_185904_a() == Material.field_151575_d;
+        }
+        else
+        {
+            return false;
+        }
     }
 
-    @Override
     public void func_76184_a(NBTTagCompound p_76184_1_)
     {
         this.field_75553_e = p_76184_1_.func_74762_e("Tick");
         NBTTagList nbttaglist = p_76184_1_.func_150295_c("Villages", 10);
 
-        for (int i = 0; i < nbttaglist.func_74745_c(); i++)
+        for (int i = 0; i < nbttaglist.func_74745_c(); ++i)
         {
             NBTTagCompound nbttagcompound = nbttaglist.func_150305_b(i);
             Village village = new Village();
@@ -273,7 +281,6 @@
         }
     }
 
-    @Override
     public NBTTagCompound func_189551_b(NBTTagCompound p_189551_1_)
     {
         p_189551_1_.func_74768_a("Tick", this.field_75553_e);
