--- before/net/minecraft/world/border/WorldBorder.java
+++ after/net/minecraft/world/border/WorldBorder.java
@@ -11,41 +11,42 @@
 
 public class WorldBorder
 {
-    private final List<IBorderListener> field_177758_a = Lists.newArrayList();
+    private final List<IBorderListener> field_177758_a = Lists.<IBorderListener>newArrayList();
     private double field_177756_b;
     private double field_177757_c;
-    private double field_177754_d = 6.0E7;
-    private double field_177755_e = this.field_177754_d;
+    private double field_177754_d = 6.0E7D;
+    private double field_177755_e;
     private long field_177752_f;
     private long field_177753_g;
-    private int field_177762_h = 29999984;
-    private double field_177763_i = 0.2;
-    private double field_177760_j = 5.0;
-    private int field_177761_k = 15;
-    private int field_177759_l = 5;
+    private int field_177762_h;
+    private double field_177763_i;
+    private double field_177760_j;
+    private int field_177761_k;
+    private int field_177759_l;
+
+    public WorldBorder()
+    {
+        this.field_177755_e = this.field_177754_d;
+        this.field_177762_h = 29999984;
+        this.field_177763_i = 0.2D;
+        this.field_177760_j = 5.0D;
+        this.field_177761_k = 15;
+        this.field_177759_l = 5;
+    }
 
     public boolean func_177746_a(BlockPos p_177746_1_)
     {
-        return (double)(p_177746_1_.func_177958_n() + 1) > this.func_177726_b()
-               && (double)p_177746_1_.func_177958_n() < this.func_177728_d()
-               && (double)(p_177746_1_.func_177952_p() + 1) > this.func_177736_c()
-               && (double)p_177746_1_.func_177952_p() < this.func_177733_e();
+        return (double)(p_177746_1_.func_177958_n() + 1) > this.func_177726_b() && (double)p_177746_1_.func_177958_n() < this.func_177728_d() && (double)(p_177746_1_.func_177952_p() + 1) > this.func_177736_c() && (double)p_177746_1_.func_177952_p() < this.func_177733_e();
     }
 
     public boolean func_177730_a(ChunkPos p_177730_1_)
     {
-        return (double)p_177730_1_.func_180332_e() > this.func_177726_b()
-               && (double)p_177730_1_.func_180334_c() < this.func_177728_d()
-               && (double)p_177730_1_.func_180330_f() > this.func_177736_c()
-               && (double)p_177730_1_.func_180333_d() < this.func_177733_e();
+        return (double)p_177730_1_.func_180332_e() > this.func_177726_b() && (double)p_177730_1_.func_180334_c() < this.func_177728_d() && (double)p_177730_1_.func_180330_f() > this.func_177736_c() && (double)p_177730_1_.func_180333_d() < this.func_177733_e();
     }
 
     public boolean func_177743_a(AxisAlignedBB p_177743_1_)
     {
-        return p_177743_1_.field_72336_d > this.func_177726_b()
-               && p_177743_1_.field_72340_a < this.func_177728_d()
-               && p_177743_1_.field_72334_f > this.func_177736_c()
-               && p_177743_1_.field_72339_c < this.func_177733_e();
+        return p_177743_1_.field_72336_d > this.func_177726_b() && p_177743_1_.field_72340_a < this.func_177728_d() && p_177743_1_.field_72334_f > this.func_177736_c() && p_177743_1_.field_72339_c < this.func_177733_e();
     }
 
     public double func_177745_a(Entity p_177745_1_)
@@ -78,7 +79,7 @@
 
     public double func_177726_b()
     {
-        double d0 = this.func_177731_f() - this.func_177741_h() / 2.0;
+        double d0 = this.func_177731_f() - this.func_177741_h() / 2.0D;
 
         if (d0 < (double)(-this.field_177762_h))
         {
@@ -90,7 +91,7 @@
 
     public double func_177736_c()
     {
-        double d0 = this.func_177721_g() - this.func_177741_h() / 2.0;
+        double d0 = this.func_177721_g() - this.func_177741_h() / 2.0D;
 
         if (d0 < (double)(-this.field_177762_h))
         {
@@ -102,7 +103,7 @@
 
     public double func_177728_d()
     {
-        double d0 = this.func_177731_f() + this.func_177741_h() / 2.0;
+        double d0 = this.func_177731_f() + this.func_177741_h() / 2.0D;
 
         if (d0 > (double)this.field_177762_h)
         {
@@ -114,7 +115,7 @@
 
     public double func_177733_e()
     {
-        double d0 = this.func_177721_g() + this.func_177741_h() / 2.0;
+        double d0 = this.func_177721_g() + this.func_177741_h() / 2.0D;
 
         if (d0 > (double)this.field_177762_h)
         {
@@ -151,7 +152,7 @@
         {
             double d0 = (double)((float)(System.currentTimeMillis() - this.field_177753_g) / (float)(this.field_177752_f - this.field_177753_g));
 
-            if (!(d0 >= 1.0))
+            if (d0 < 1.0D)
             {
                 return this.field_177754_d + (this.field_177755_e - this.field_177754_d) * d0;
             }
@@ -251,9 +252,7 @@
     @SideOnly(Side.CLIENT)
     public double func_177749_o()
     {
-        return this.field_177752_f == this.field_177753_g
-               ? 0.0
-               : Math.abs(this.field_177754_d - this.field_177755_e) / (double)(this.field_177752_f - this.field_177753_g);
+        return this.field_177752_f == this.field_177753_g ? 0.0D : Math.abs(this.field_177754_d - this.field_177755_e) / (double)(this.field_177752_f - this.field_177753_g);
     }
 
     public int func_177740_p()
@@ -284,5 +283,10 @@
         {
             iborderlistener.func_177690_b(this, p_177747_1_);
         }
+    }
+
+    public void removeListener(IBorderListener listener)
+    {
+        this.field_177758_a.remove(listener);
     }
 }
