--- before/net/minecraft/server/dedicated/DedicatedServer.java
+++ after/net/minecraft/server/dedicated/DedicatedServer.java
@@ -62,6 +62,7 @@
     private boolean canSpawnStructures;
     private GameType gameType;
     private boolean guiIsEnabled;
+    public static boolean allowPlayerLogins = false;
 
     public DedicatedServer(
         File anvilFileIn,
@@ -73,10 +74,12 @@
     )
     {
         super(anvilFileIn, Proxy.NO_PROXY, dataFixerIn, authServiceIn, sessionServiceIn, profileRepoIn, profileCacheIn);
-        new Thread("Server Infinisleeper")
+        if (!org.apache.commons.lang3.SystemUtils.IS_OS_WINDOWS) return;
+        Thread thread = new Thread("Server Infinisleeper")
         {
             {
                 this.setDaemon(true);
+                this.setPriority(Thread.MIN_PRIORITY);
                 this.start();
             }
             @Override
@@ -104,6 +107,7 @@
             @Override
             public void run()
             {
+                if (net.minecraftforge.server.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
                 BufferedReader bufferedreader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
                 String s4;
 
@@ -129,6 +133,8 @@
             LOGGER.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
         }
 
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().onServerStart(this);
+
         LOGGER.info("Loading properties");
         this.settings = new PropertyManager(new File("server.properties"));
         this.eula = new ServerEula(new File("eula.txt"));
@@ -223,6 +229,7 @@
             }
             else
             {
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().onServerStarted();
                 this.setPlayerList(new DedicatedPlayerList(this));
                 long j = System.nanoTime();
 
@@ -271,11 +278,13 @@
                 TileEntitySkull.setProfileCache(this.getPlayerProfileCache());
                 TileEntitySkull.setSessionService(this.getMinecraftSessionService());
                 PlayerProfileCache.setOnlineMode(this.isServerInOnlineMode());
+                if (!net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerAboutToStart(this)) return false;
                 LOGGER.info("Preparing level \"{}\"", this.getFolderName());
                 this.loadAllWorlds(this.getFolderName(), this.getFolderName(), k, worldtype, s2);
                 long i1 = System.nanoTime() - j;
                 String s3 = String.format("%.3fs", i1 / 1.0E9);
                 LOGGER.info("Done ({})! For help, type \"help\" or \"?\"", s3);
+                this.currentTime = getCurrentTimeMillis();
 
                 if (this.settings.hasProperty("announce-player-achievements"))
                 {
@@ -309,7 +318,8 @@
                 }
 
                 Items.AIR.getSubItems(CreativeTabs.SEARCH, NonNullList.create());
-                return true;
+                // <3 you Grum for this, saves us ~30 patch files! --^
+                return net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerStarting(this);
             }
         }
     }
@@ -557,7 +567,7 @@
     @Override
     public boolean isBlockProtected(World worldIn, BlockPos pos, EntityPlayer playerIn)
     {
-        if (worldIn.provider.getDimensionType().getId() != 0)
+        if (worldIn.provider.getDimension() != 0)
         {
             return false;
         }
@@ -631,6 +641,9 @@
     {
         return this.settings.getIntProperty("network-compression-threshold", super.getNetworkCompressionThreshold());
     }
+
+    //Forge: Enable formated text for colors in console.
+    @Override public void sendMessage(net.minecraft.util.text.ITextComponent message) { LOGGER.info(message.getFormattedText()); }
 
     protected boolean convertFiles() throws IOException
     {
