--- before/net/minecraft/server/dedicated/DedicatedPlayerList.java
+++ after/net/minecraft/server/dedicated/DedicatedPlayerList.java
@@ -43,7 +43,7 @@
     public void func_72371_a(boolean p_72371_1_)
     {
         super.func_72371_a(p_72371_1_);
-        this.func_72365_p().func_71328_a("white-list", p_72371_1_);
+        this.func_72365_p().func_71328_a("white-list", Boolean.valueOf(p_72371_1_));
         this.func_72365_p().func_71326_a();
     }
 
