--- before/net/minecraft/server/dedicated/DedicatedServer.java
+++ after/net/minecraft/server/dedicated/DedicatedServer.java
@@ -53,7 +53,7 @@
 {
     private static final Logger field_155771_h = LogManager.getLogger();
     private static final Pattern field_189647_l = Pattern.compile("^[a-fA-F0-9]{40}$");
-    public final List<PendingCommand> field_71341_l = Collections.synchronizedList(Lists.newArrayList());
+    public final List<PendingCommand> field_71341_l = Collections.<PendingCommand>synchronizedList(Lists.newArrayList());
     private RConThreadQuery field_71342_m;
     private final RConConsoleSource field_184115_n = new RConConsoleSource(this);
     private RConThreadMain field_71339_n;
@@ -62,24 +62,17 @@
     private boolean field_71338_p;
     private GameType field_71337_q;
     private boolean field_71335_s;
+    public static boolean allowPlayerLogins = false;
 
-    public DedicatedServer(
-        File p_i47086_1_,
-        DataFixer p_i47086_2_,
-        YggdrasilAuthenticationService p_i47086_3_,
-        MinecraftSessionService p_i47086_4_,
-        GameProfileRepository p_i47086_5_,
-        PlayerProfileCache p_i47086_6_
-    )
+    public DedicatedServer(File p_i47086_1_, DataFixer p_i47086_2_, YggdrasilAuthenticationService p_i47086_3_, MinecraftSessionService p_i47086_4_, GameProfileRepository p_i47086_5_, PlayerProfileCache p_i47086_6_)
     {
         super(p_i47086_1_, Proxy.NO_PROXY, p_i47086_2_, p_i47086_3_, p_i47086_4_, p_i47086_5_, p_i47086_6_);
-        new Thread("Server Infinisleeper")
+        Thread thread = new Thread("Server Infinisleeper")
         {
             {
                 this.setDaemon(true);
                 this.start();
             }
-            @Override
             public void run()
             {
                 while (true)
@@ -88,22 +81,22 @@
                     {
                         Thread.sleep(2147483647L);
                     }
-                    catch (InterruptedException interruptedexception)
+                    catch (InterruptedException var2)
                     {
+                        ;
                     }
                 }
             }
         };
     }
 
-    @Override
     public boolean func_71197_b() throws IOException
     {
         Thread thread = new Thread("Server console handler")
         {
-            @Override
             public void run()
             {
+                if (net.minecraftforge.server.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
                 BufferedReader bufferedreader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
                 String s4;
 
@@ -129,6 +122,8 @@
             field_155771_h.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
         }
 
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().onServerStart(this);
+
         field_155771_h.info("Loading properties");
         this.field_71340_o = new PropertyManager(new File("server.properties"));
         this.field_154332_n = new ServerEula(new File("eula.txt"));
@@ -163,17 +158,17 @@
 
             if (this.field_71340_o.func_73669_a("difficulty", 1) < 0)
             {
-                this.field_71340_o.func_73667_a("difficulty", 0);
+                this.field_71340_o.func_73667_a("difficulty", Integer.valueOf(0));
             }
             else if (this.field_71340_o.func_73669_a("difficulty", 1) > 3)
             {
-                this.field_71340_o.func_73667_a("difficulty", 3);
+                this.field_71340_o.func_73667_a("difficulty", Integer.valueOf(3));
             }
 
             this.field_71338_p = this.field_71340_o.func_73670_a("generate-structures", true);
             int i = this.field_71340_o.func_73669_a("gamemode", GameType.SURVIVAL.func_77148_a());
             this.field_71337_q = WorldSettings.func_77161_a(i);
-            field_155771_h.info("Default game type: {}", this.field_71337_q);
+            field_155771_h.info("Default game type: {}", (Object)this.field_71337_q);
             InetAddress inetaddress = null;
 
             if (!this.func_71211_k().isEmpty())
@@ -188,7 +183,7 @@
 
             field_155771_h.info("Generating keypair");
             this.func_71253_a(CryptManager.func_75891_b());
-            field_155771_h.info("Starting Minecraft server on {}:{}", this.func_71211_k().isEmpty() ? "*" : this.func_71211_k(), this.func_71215_F());
+            field_155771_h.info("Starting Minecraft server on {}:{}", this.func_71211_k().isEmpty() ? "*" : this.func_71211_k(), Integer.valueOf(this.func_71215_F()));
 
             try
             {
@@ -197,7 +192,7 @@
             catch (IOException ioexception)
             {
                 field_155771_h.warn("**** FAILED TO BIND TO PORT!");
-                field_155771_h.warn("The exception was: {}", ioexception.toString());
+                field_155771_h.warn("The exception was: {}", (Object)ioexception.toString());
                 field_155771_h.warn("Perhaps a server is already running on that port?");
                 return false;
             }
@@ -206,9 +201,7 @@
             {
                 field_155771_h.warn("**** SERVER IS RUNNING IN OFFLINE/INSECURE MODE!");
                 field_155771_h.warn("The server will make no attempt to authenticate usernames. Beware.");
-                field_155771_h.warn(
-                    "While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose."
-                );
+                field_155771_h.warn("While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose.");
                 field_155771_h.warn("To change this, set \"online-mode\" to \"true\" in the server.properties file.");
             }
 
@@ -223,6 +216,7 @@
             }
             else
             {
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().onServerStarted();
                 this.func_184105_a(new DedicatedPlayerList(this));
                 long j = System.nanoTime();
 
@@ -234,7 +228,7 @@
                 String s = this.field_71340_o.func_73671_a("level-seed", "");
                 String s1 = this.field_71340_o.func_73671_a("level-type", "DEFAULT");
                 String s2 = this.field_71340_o.func_73671_a("generator-settings", "");
-                long k = new Random().nextLong();
+                long k = (new Random()).nextLong();
 
                 if (!s.isEmpty())
                 {
@@ -247,7 +241,7 @@
                             k = l;
                         }
                     }
-                    catch (NumberFormatException numberformatexception)
+                    catch (NumberFormatException var16)
                     {
                         k = (long)s.hashCode();
                     }
@@ -267,21 +261,21 @@
                 this.func_71191_d(this.field_71340_o.func_73669_a("max-build-height", 256));
                 this.func_71191_d((this.func_71207_Z() + 8) / 16 * 16);
                 this.func_71191_d(MathHelper.func_76125_a(this.func_71207_Z(), 64, 256));
-                this.field_71340_o.func_73667_a("max-build-height", this.func_71207_Z());
+                this.field_71340_o.func_73667_a("max-build-height", Integer.valueOf(this.func_71207_Z()));
                 TileEntitySkull.func_184293_a(this.func_152358_ax());
                 TileEntitySkull.func_184294_a(this.func_147130_as());
                 PlayerProfileCache.func_187320_a(this.func_71266_T());
-                field_155771_h.info("Preparing level \"{}\"", this.func_71270_I());
+                if (!net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerAboutToStart(this)) return false;
+                field_155771_h.info("Preparing level \"{}\"", (Object)this.func_71270_I());
                 this.func_71247_a(this.func_71270_I(), this.func_71270_I(), k, worldtype, s2);
                 long i1 = System.nanoTime() - j;
-                String s3 = String.format("%.3fs", (double)i1 / 1.0E9);
-                field_155771_h.info("Done ({})! For help, type \"help\" or \"?\"", s3);
+                String s3 = String.format("%.3fs", (double)i1 / 1.0E9D);
+                field_155771_h.info("Done ({})! For help, type \"help\" or \"?\"", (Object)s3);
+                this.field_175591_ab = func_130071_aq();
 
                 if (this.field_71340_o.func_187239_a("announce-player-achievements"))
                 {
-                    this.field_71305_c[0]
-                    .func_82736_K()
-                    .func_82764_b("announceAdvancements", this.field_71340_o.func_73670_a("announce-player-achievements", true) ? "true" : "false");
+                    this.field_71305_c[0].func_82736_K().func_82764_b("announceAdvancements", this.field_71340_o.func_73670_a("announce-player-achievements", true) ? "true" : "false");
                     this.field_71340_o.func_187238_b("announce-player-achievements");
                     this.field_71340_o.func_73668_b();
                 }
@@ -309,7 +303,8 @@
                 }
 
                 Items.field_190931_a.func_150895_a(CreativeTabs.field_78027_g, NonNullList.func_191196_a());
-                return true;
+                // <3 you Grum for this, saves us ~30 patch files! --^
+                return net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerStarting(this);
             }
         }
     }
@@ -339,46 +334,38 @@
 
         if (!this.field_71340_o.func_73671_a("resource-pack", "").isEmpty() && s.isEmpty())
         {
-            field_155771_h.warn(
-                "You specified a resource pack without providing a sha1 hash. Pack will be updated on the client only if you change the name of the pack."
-            );
+            field_155771_h.warn("You specified a resource pack without providing a sha1 hash. Pack will be updated on the client only if you change the name of the pack.");
         }
 
         return s;
     }
 
-    @Override
     public void func_71235_a(GameType p_71235_1_)
     {
         super.func_71235_a(p_71235_1_);
         this.field_71337_q = p_71235_1_;
     }
 
-    @Override
     public boolean func_71225_e()
     {
         return this.field_71338_p;
     }
 
-    @Override
     public GameType func_71265_f()
     {
         return this.field_71337_q;
     }
 
-    @Override
     public EnumDifficulty func_147135_j()
     {
         return EnumDifficulty.func_151523_a(this.field_71340_o.func_73669_a("difficulty", EnumDifficulty.NORMAL.func_151525_a()));
     }
 
-    @Override
     public boolean func_71199_h()
     {
         return this.field_71340_o.func_73670_a("hardcore", false);
     }
 
-    @Override
     public CrashReport func_71230_b(CrashReport p_71230_1_)
     {
         p_71230_1_ = super.func_71230_b(p_71230_1_);
@@ -400,40 +387,34 @@
         return p_71230_1_;
     }
 
-    @Override
     public void func_71240_o()
     {
         System.exit(0);
     }
 
-    @Override
     public void func_71190_q()
     {
         super.func_71190_q();
         this.func_71333_ah();
     }
 
-    @Override
     public boolean func_71255_r()
     {
         return this.field_71340_o.func_73670_a("allow-nether", true);
     }
 
-    @Override
     public boolean func_71193_K()
     {
         return this.field_71340_o.func_73670_a("spawn-monsters", true);
     }
 
-    @Override
     public void func_70000_a(Snooper p_70000_1_)
     {
-        p_70000_1_.func_152768_a("whitelist_enabled", this.func_184103_al().func_72383_n());
-        p_70000_1_.func_152768_a("whitelist_count", this.func_184103_al().func_152598_l().length);
+        p_70000_1_.func_152768_a("whitelist_enabled", Boolean.valueOf(this.func_184103_al().func_72383_n()));
+        p_70000_1_.func_152768_a("whitelist_count", Integer.valueOf(this.func_184103_al().func_152598_l().length));
         super.func_70000_a(p_70000_1_);
     }
 
-    @Override
     public boolean func_70002_Q()
     {
         return this.field_71340_o.func_73670_a("snooper-enabled", true);
@@ -453,13 +434,11 @@
         }
     }
 
-    @Override
     public boolean func_71262_S()
     {
         return true;
     }
 
-    @Override
     public boolean func_181035_ah()
     {
         return this.field_71340_o.func_73670_a("use-native-transport", true);
@@ -470,13 +449,11 @@
         return (DedicatedPlayerList)super.func_184103_al();
     }
 
-    @Override
     public int func_71327_a(String p_71327_1_, int p_71327_2_)
     {
         return this.field_71340_o.func_73669_a(p_71327_1_, p_71327_2_);
     }
 
-    @Override
     public String func_71330_a(String p_71330_1_, String p_71330_2_)
     {
         return this.field_71340_o.func_73671_a(p_71330_1_, p_71330_2_);
@@ -487,38 +464,32 @@
         return this.field_71340_o.func_73670_a(p_71332_1_, p_71332_2_);
     }
 
-    @Override
     public void func_71328_a(String p_71328_1_, Object p_71328_2_)
     {
         this.field_71340_o.func_73667_a(p_71328_1_, p_71328_2_);
     }
 
-    @Override
     public void func_71326_a()
     {
         this.field_71340_o.func_73668_b();
     }
 
-    @Override
     public String func_71329_c()
     {
         File file1 = this.field_71340_o.func_73665_c();
         return file1 != null ? file1.getAbsolutePath() : "No settings file";
     }
 
-    @Override
     public String func_71277_t()
     {
         return this.func_71211_k();
     }
 
-    @Override
     public int func_71234_u()
     {
         return this.func_71215_F();
     }
 
-    @Override
     public String func_71274_v()
     {
         return this.func_71273_Y();
@@ -530,34 +501,29 @@
         this.field_71335_s = true;
     }
 
-    @Override
     public boolean func_71279_ae()
     {
         return this.field_71335_s;
     }
 
-    @Override
     public String func_71206_a(GameType p_71206_1_, boolean p_71206_2_)
     {
         return "";
     }
 
-    @Override
     public boolean func_82356_Z()
     {
         return this.field_71340_o.func_73670_a("enable-command-block", false);
     }
 
-    @Override
     public int func_82357_ak()
     {
         return this.field_71340_o.func_73669_a("spawn-protection", super.func_82357_ak());
     }
 
-    @Override
     public boolean func_175579_a(World p_175579_1_, BlockPos p_175579_2_, EntityPlayer p_175579_3_)
     {
-        if (p_175579_1_.field_73011_w.func_186058_p().func_186068_a() != 0)
+        if (p_175579_1_.field_73011_w.getDimension() != 0)
         {
             return false;
         }
@@ -583,33 +549,28 @@
         }
     }
 
-    @Override
     public int func_110455_j()
     {
         return this.field_71340_o.func_73669_a("op-permission-level", 4);
     }
 
-    @Override
     public void func_143006_e(int p_143006_1_)
     {
         super.func_143006_e(p_143006_1_);
-        this.field_71340_o.func_73667_a("player-idle-timeout", p_143006_1_);
+        this.field_71340_o.func_73667_a("player-idle-timeout", Integer.valueOf(p_143006_1_));
         this.func_71326_a();
     }
 
-    @Override
     public boolean func_181034_q()
     {
         return this.field_71340_o.func_73670_a("broadcast-rcon-to-ops", true);
     }
 
-    @Override
     public boolean func_183002_r()
     {
         return this.field_71340_o.func_73670_a("broadcast-console-to-ops", true);
     }
 
-    @Override
     public int func_175580_aG()
     {
         int i = this.field_71340_o.func_73669_a("max-world-size", super.func_175580_aG());
@@ -626,12 +587,14 @@
         return i;
     }
 
-    @Override
     public int func_175577_aI()
     {
         return this.field_71340_o.func_73669_a("network-compression-threshold", super.func_175577_aI());
     }
 
+    //Forge: Enable formated text for colors in console.
+    @Override public void func_145747_a(net.minecraft.util.text.ITextComponent message) { field_155771_h.info(message.func_150254_d()); }
+
     protected boolean func_152368_aE() throws IOException
     {
         boolean flag = false;
@@ -708,8 +671,9 @@
         {
             Thread.sleep(5000L);
         }
-        catch (InterruptedException interruptedexception)
+        catch (InterruptedException var2)
         {
+            ;
         }
     }
 
@@ -718,13 +682,11 @@
         return this.field_71340_o.func_179885_a("max-tick-time", TimeUnit.MINUTES.toMillis(1L));
     }
 
-    @Override
     public String func_71258_A()
     {
         return "";
     }
 
-    @Override
     public String func_71252_i(String p_71252_1_)
     {
         this.field_184115_n.func_70007_b();
