--- before/net/minecraft/server/dedicated/PropertyManager.java
+++ after/net/minecraft/server/dedicated/PropertyManager.java
@@ -43,8 +43,9 @@
                     {
                         fileinputstream.close();
                     }
-                    catch (IOException ioexception)
+                    catch (IOException var11)
                     {
+                        ;
                     }
                 }
             }
@@ -84,8 +85,9 @@
                 {
                     fileoutputstream.close();
                 }
-                catch (IOException ioexception)
+                catch (IOException var10)
                 {
+                    ;
                 }
             }
         }
@@ -114,7 +116,7 @@
         {
             return Integer.parseInt(this.func_73671_a(p_73669_1_, "" + p_73669_2_));
         }
-        catch (Exception exception)
+        catch (Exception var4)
         {
             this.field_73672_b.setProperty(p_73669_1_, "" + p_73669_2_);
             this.func_73668_b();
@@ -128,7 +130,7 @@
         {
             return Long.parseLong(this.func_73671_a(p_179885_1_, "" + p_179885_2_));
         }
-        catch (Exception exception)
+        catch (Exception var5)
         {
             this.field_73672_b.setProperty(p_179885_1_, "" + p_179885_2_);
             this.func_73668_b();
@@ -142,7 +144,7 @@
         {
             return Boolean.parseBoolean(this.func_73671_a(p_73670_1_, "" + p_73670_2_));
         }
-        catch (Exception exception)
+        catch (Exception var4)
         {
             this.field_73672_b.setProperty(p_73670_1_, "" + p_73670_2_);
             this.func_73668_b();
