--- before/net/minecraft/network/rcon/RConConsoleSource.java
+++ after/net/minecraft/network/rcon/RConConsoleSource.java
@@ -17,18 +17,6 @@
         this.field_184171_b = p_i46835_1_;
     }
 
-    @SideOnly(Side.SERVER)
-    public void func_70007_b()
-    {
-        this.field_70009_b.setLength(0);
-    }
-
-    @SideOnly(Side.SERVER)
-    public String func_70008_c()
-    {
-        return this.field_70009_b.toString();
-    }
-
     public String func_70005_c_()
     {
         return "Rcon";
@@ -57,5 +45,17 @@
     public MinecraftServer func_184102_h()
     {
         return this.field_184171_b;
+    }
+
+    @SideOnly(Side.SERVER)
+    public void func_70007_b()
+    {
+        this.field_70009_b.setLength(0);
+    }
+
+    @SideOnly(Side.SERVER)
+    public String func_70008_c()
+    {
+        return this.field_70009_b.toString();
     }
 }
