--- before/net/minecraft/entity/monster/AbstractSkeleton.java
+++ after/net/minecraft/entity/monster/AbstractSkeleton.java
@@ -205,7 +205,7 @@
             this.tasks.removeTask(this.aiArrowAttack);
             ItemStack itemstack = this.getHeldItemMainhand();
 
-            if (itemstack.getItem() == Items.BOW)
+            if (itemstack.getItem() instanceof net.minecraft.item.ItemBow)
             {
                 int i = 20;
 
@@ -228,6 +228,8 @@
     public void attackEntityWithRangedAttack(EntityLivingBase target, float distanceFactor)
     {
         EntityArrow entityarrow = this.getArrow(distanceFactor);
+        if (this.getHeldItemMainhand().getItem() instanceof net.minecraft.item.ItemBow)
+            entityarrow = ((net.minecraft.item.ItemBow) this.getHeldItemMainhand().getItem()).customizeArrow(entityarrow);
         double d0 = target.posX - this.posX;
         double d1 = target.getEntityBoundingBox().minY + target.height / 3.0F - entityarrow.posY;
         double d2 = target.posZ - this.posZ;
