--- before/net/minecraft/entity/monster/AbstractIllager.java
+++ after/net/minecraft/entity/monster/AbstractIllager.java
@@ -10,7 +10,7 @@
 
 public abstract class AbstractIllager extends EntityMob
 {
-    protected static final DataParameter<Byte> field_193080_a = EntityDataManager.func_187226_a(AbstractIllager.class, DataSerializers.field_187191_a);
+    protected static final DataParameter<Byte> field_193080_a = EntityDataManager.<Byte>func_187226_a(AbstractIllager.class, DataSerializers.field_187191_a);
 
     public AbstractIllager(World p_i47509_1_)
     {
@@ -20,30 +20,30 @@
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_193080_a, (byte)0);
+        this.field_70180_af.func_187214_a(field_193080_a, Byte.valueOf((byte)0));
     }
 
     @SideOnly(Side.CLIENT)
     protected boolean func_193078_a(int p_193078_1_)
     {
-        int i = this.field_70180_af.func_187225_a(field_193080_a);
+        int i = ((Byte)this.field_70180_af.func_187225_a(field_193080_a)).byteValue();
         return (i & p_193078_1_) != 0;
     }
 
     protected void func_193079_a(int p_193079_1_, boolean p_193079_2_)
     {
-        int i = this.field_70180_af.func_187225_a(field_193080_a);
+        int i = ((Byte)this.field_70180_af.func_187225_a(field_193080_a)).byteValue();
 
         if (p_193079_2_)
         {
-            i |= p_193079_1_;
+            i = i | p_193079_1_;
         }
         else
         {
-            i &= ~p_193079_1_;
+            i = i & ~p_193079_1_;
         }
 
-        this.field_70180_af.func_187227_b(field_193080_a, (byte)(i & 255));
+        this.field_70180_af.func_187227_b(field_193080_a, Byte.valueOf((byte)(i & 255)));
     }
 
     public EnumCreatureAttribute func_70668_bt()
