--- before/net/minecraft/entity/monster/EntityGhast.java
+++ after/net/minecraft/entity/monster/EntityGhast.java
@@ -33,7 +33,7 @@
 
 public class EntityGhast extends EntityFlying implements IMob
 {
-    private static final DataParameter<Boolean> field_184683_a = EntityDataManager.func_187226_a(EntityGhast.class, DataSerializers.field_187198_h);
+    private static final DataParameter<Boolean> field_184683_a = EntityDataManager.<Boolean>func_187226_a(EntityGhast.class, DataSerializers.field_187198_h);
     private int field_92014_j = 1;
 
     public EntityGhast(World p_i1735_1_)
@@ -56,12 +56,12 @@
     @SideOnly(Side.CLIENT)
     public boolean func_110182_bF()
     {
-        return this.field_70180_af.func_187225_a(field_184683_a);
+        return ((Boolean)this.field_70180_af.func_187225_a(field_184683_a)).booleanValue();
     }
 
     public void func_175454_a(boolean p_175454_1_)
     {
-        this.field_70180_af.func_187227_b(field_184683_a, p_175454_1_);
+        this.field_70180_af.func_187227_b(field_184683_a, Boolean.valueOf(p_175454_1_));
     }
 
     public int func_175453_cd()
@@ -99,7 +99,7 @@
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184683_a, false);
+        this.field_70180_af.func_187214_a(field_184683_a, Boolean.valueOf(false));
     }
 
     protected void func_110147_ax()
