--- before/net/minecraft/entity/monster/EntityEnderman.java
+++ after/net/minecraft/entity/monster/EntityEnderman.java
@@ -306,7 +306,9 @@
 
     private boolean teleportTo(double x, double y, double z)
     {
-        boolean flag = this.attemptTeleport(x, y, z);
+        net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(this, x, y, z, 0);
+        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return false;
+        boolean flag = this.attemptTeleport(event.getTargetX(), event.getTargetY(), event.getTargetZ());
 
         if (flag)
         {
@@ -400,6 +402,18 @@
         }
     }
 
+    /*===================================== Forge Start ==============================*/
+    public static void setCarriable(Block block, boolean canCarry)
+    {
+        if (canCarry) CARRIABLE_BLOCKS.add(block);
+        else          CARRIABLE_BLOCKS.remove(block);
+    }
+    public static boolean getCarriable(Block block)
+    {
+        return CARRIABLE_BLOCKS.contains(block);
+    }
+    /*===================================== Forge End ==============================*/
+
     public boolean isScreaming()
     {
         return this.dataManager.get(SCREAMING);
@@ -552,7 +566,7 @@
                 }
                 else
                 {
-                    return !this.enderman.world.getGameRules().getBoolean("mobGriefing")
+                    return !net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.enderman.world, this.enderman)
                            ? false
                            : this.enderman.getRNG().nextInt(2000) == 0;
                 }
@@ -571,7 +585,16 @@
                 IBlockState iblockstate1 = world.getBlockState(blockpos.down());
                 IBlockState iblockstate2 = this.enderman.getHeldBlockState();
 
-                if (iblockstate2 != null && this.canPlaceBlock(world, blockpos, iblockstate2.getBlock(), iblockstate, iblockstate1))
+                if (iblockstate2 != null && this.canPlaceBlock(world, blockpos, iblockstate2.getBlock(), iblockstate, iblockstate1)
+                    && net.minecraftforge.event.ForgeEventFactory.onBlockPlace(
+                        enderman,
+                        new net.minecraftforge.common.util.BlockSnapshot(
+                            world,
+                            blockpos,
+                            iblockstate2
+                        ),
+                        net.minecraft.util.EnumFacing.UP
+                    ).isCanceled())
                 {
                     world.setBlockState(blockpos, iblockstate2, 3);
                     this.enderman.setHeldBlockState(null);
@@ -613,7 +636,7 @@
                 }
                 else
                 {
-                    return !this.enderman.world.getGameRules().getBoolean("mobGriefing") ? false : this.enderman.getRNG().nextInt(20) == 0;
+                    return !net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.enderman.world, this.enderman) ? false : this.enderman.getRNG().nextInt(20) == 0;
                 }
             }
 
