--- before/net/minecraft/entity/monster/EntityPolarBear.java
+++ after/net/minecraft/entity/monster/EntityPolarBear.java
@@ -38,7 +38,7 @@
 
 public class EntityPolarBear extends EntityAnimal
 {
-    private static final DataParameter<Boolean> field_189798_bx = EntityDataManager.func_187226_a(EntityPolarBear.class, DataSerializers.field_187198_h);
+    private static final DataParameter<Boolean> field_189798_bx = EntityDataManager.<Boolean>func_187226_a(EntityPolarBear.class, DataSerializers.field_187198_h);
     private float field_189799_by;
     private float field_189800_bz;
     private int field_189797_bB;
@@ -121,7 +121,7 @@
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_189798_bx, false);
+        this.field_70180_af.func_187214_a(field_189798_bx, Boolean.valueOf(false));
     }
 
     public void func_70071_h_()
@@ -162,12 +162,12 @@
 
     public boolean func_189793_df()
     {
-        return this.field_70180_af.func_187225_a(field_189798_bx);
+        return ((Boolean)this.field_70180_af.func_187225_a(field_189798_bx)).booleanValue();
     }
 
     public void func_189794_p(boolean p_189794_1_)
     {
-        this.field_70180_af.func_187227_b(field_189798_bx, p_189794_1_);
+        this.field_70180_af.func_187227_b(field_189798_bx, Boolean.valueOf(p_189794_1_));
     }
 
     @SideOnly(Side.CLIENT)
@@ -204,7 +204,7 @@
     {
         public AIAttackPlayer()
         {
-            super(EntityPolarBear.this, EntityPlayer.class, 20, true, true, (Predicate <? super EntityPlayer >)null);
+            super(EntityPolarBear.this, EntityPlayer.class, 20, true, true, (Predicate)null);
         }
 
         public boolean func_75250_a()
