--- before/net/minecraft/entity/monster/EntitySnowman.java
+++ after/net/minecraft/entity/monster/EntitySnowman.java
@@ -31,9 +31,9 @@
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
 
-public class EntitySnowman extends EntityGolem implements IRangedAttackMob
+public class EntitySnowman extends EntityGolem implements IRangedAttackMob, net.minecraftforge.common.IShearable
 {
-    private static final DataParameter<Byte> field_184749_a = EntityDataManager.func_187226_a(EntitySnowman.class, DataSerializers.field_187191_a);
+    private static final DataParameter<Byte> field_184749_a = EntityDataManager.<Byte>func_187226_a(EntitySnowman.class, DataSerializers.field_187191_a);
 
     public EntitySnowman(World p_i1692_1_)
     {
@@ -52,20 +52,20 @@
         this.field_70714_bg.func_75776_a(2, new EntityAIWanderAvoidWater(this, 1.0D, 1.0000001E-5F));
         this.field_70714_bg.func_75776_a(3, new EntityAIWatchClosest(this, EntityPlayer.class, 6.0F));
         this.field_70714_bg.func_75776_a(4, new EntityAILookIdle(this));
-        this.field_70715_bh.func_75776_a(1, new EntityAINearestAttackableTarget<>(this, EntityLiving.class, 10, true, false, IMob.field_82192_a));
+        this.field_70715_bh.func_75776_a(1, new EntityAINearestAttackableTarget(this, EntityLiving.class, 10, true, false, IMob.field_82192_a));
     }
 
     protected void func_110147_ax()
     {
         super.func_110147_ax();
         this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(4.0D);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a((double)0.2F);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.20000000298023224D);
     }
 
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184749_a, (byte)16);
+        this.field_70180_af.func_187214_a(field_184749_a, Byte.valueOf((byte)16));
     }
 
     public void func_70014_b(NBTTagCompound p_70014_1_)
@@ -104,7 +104,7 @@
                 this.func_70097_a(DamageSource.field_76370_b, 1.0F);
             }
 
-            if (!this.field_70170_p.func_82736_K().func_82766_b("mobGriefing"))
+            if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this))
             {
                 return;
             }
@@ -133,7 +133,7 @@
     public void func_82196_d(EntityLivingBase p_82196_1_, float p_82196_2_)
     {
         EntitySnowball entitysnowball = new EntitySnowball(this.field_70170_p, this);
-        double d0 = p_82196_1_.field_70163_u + (double)p_82196_1_.func_70047_e() - (double)1.1F;
+        double d0 = p_82196_1_.field_70163_u + (double)p_82196_1_.func_70047_e() - 1.100000023841858D;
         double d1 = p_82196_1_.field_70165_t - this.field_70165_t;
         double d2 = d0 - entitysnowball.field_70163_u;
         double d3 = p_82196_1_.field_70161_v - this.field_70161_v;
@@ -152,7 +152,7 @@
     {
         ItemStack itemstack = p_184645_1_.func_184586_b(p_184645_2_);
 
-        if (itemstack.func_77973_b() == Items.field_151097_aZ && this.func_184748_o() && !this.field_70170_p.field_72995_K)
+        if (false && itemstack.func_77973_b() == Items.field_151097_aZ && this.func_184748_o() && !this.field_70170_p.field_72995_K) //Forge: Moved to onSheared
         {
             this.func_184747_a(false);
             itemstack.func_77972_a(1, p_184645_1_);
@@ -163,20 +163,20 @@
 
     public boolean func_184748_o()
     {
-        return (this.field_70180_af.func_187225_a(field_184749_a) & 16) != 0;
+        return (((Byte)this.field_70180_af.func_187225_a(field_184749_a)).byteValue() & 16) != 0;
     }
 
     public void func_184747_a(boolean p_184747_1_)
     {
-        byte b0 = this.field_70180_af.func_187225_a(field_184749_a);
+        byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184749_a)).byteValue();
 
         if (p_184747_1_)
         {
-            this.field_70180_af.func_187227_b(field_184749_a, (byte)(b0 | 16));
+            this.field_70180_af.func_187227_b(field_184749_a, Byte.valueOf((byte)(b0 | 16)));
         }
         else
         {
-            this.field_70180_af.func_187227_b(field_184749_a, (byte)(b0 & -17));
+            this.field_70180_af.func_187227_b(field_184749_a, Byte.valueOf((byte)(b0 & -17)));
         }
     }
 
@@ -196,6 +196,14 @@
     protected SoundEvent func_184615_bR()
     {
         return SoundEvents.field_187801_fC;
+    }
+
+    @Override public boolean isShearable(ItemStack item, net.minecraft.world.IBlockAccess world, BlockPos pos) { return this.func_184748_o(); }
+    @Override
+    public java.util.List<ItemStack> onSheared(ItemStack item, net.minecraft.world.IBlockAccess world, BlockPos pos, int fortune)
+    {
+        this.func_184747_a(false);
+        return com.google.common.collect.Lists.newArrayList();
     }
 
     public void func_184724_a(boolean p_184724_1_)
