--- before/net/minecraft/entity/monster/EntitySlime.java
+++ after/net/minecraft/entity/monster/EntitySlime.java
@@ -38,7 +38,7 @@
 
 public class EntitySlime extends EntityLiving implements IMob
 {
-    private static final DataParameter<Integer> field_184711_bt = EntityDataManager.func_187226_a(EntitySlime.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Integer> field_184711_bt = EntityDataManager.<Integer>func_187226_a(EntitySlime.class, DataSerializers.field_187192_b);
     public float field_70813_a;
     public float field_70811_b;
     public float field_70812_c;
@@ -63,12 +63,12 @@
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184711_bt, 1);
+        this.field_70180_af.func_187214_a(field_184711_bt, Integer.valueOf(1));
     }
 
     protected void func_70799_a(int p_70799_1_, boolean p_70799_2_)
     {
-        this.field_70180_af.func_187227_b(field_184711_bt, p_70799_1_);
+        this.field_70180_af.func_187227_b(field_184711_bt, Integer.valueOf(p_70799_1_));
         this.func_70105_a(0.51000005F * (float)p_70799_1_, 0.51000005F * (float)p_70799_1_);
         this.func_70107_b(this.field_70165_t, this.field_70163_u, this.field_70161_v);
         this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a((double)(p_70799_1_ * p_70799_1_));
@@ -84,7 +84,7 @@
 
     public int func_70809_q()
     {
-        return this.field_70180_af.func_187225_a(field_184711_bt);
+        return ((Integer)this.field_70180_af.func_187225_a(field_184711_bt)).intValue();
     }
 
     public static void func_189758_c(DataFixer p_189758_0_)
@@ -137,7 +137,7 @@
         if (this.field_70122_E && !this.field_175452_bi)
         {
             int i = this.func_70809_q();
-
+            if (spawnCustomParticles()) { i = 0; } // don't spawn particles if it's handled by the implementation itself
             for (int j = 0; j < i * 8; ++j)
             {
                 float f = this.field_70146_Z.nextFloat() * ((float)Math.PI * 2F);
@@ -304,7 +304,7 @@
         BlockPos blockpos = new BlockPos(MathHelper.func_76128_c(this.field_70165_t), 0, MathHelper.func_76128_c(this.field_70161_v));
         Chunk chunk = this.field_70170_p.func_175726_f(blockpos);
 
-        if (this.field_70170_p.func_72912_H().func_76067_t() == WorldType.field_77138_c && this.field_70146_Z.nextInt(4) != 1)
+        if (this.field_70170_p.func_72912_H().func_76067_t().handleSlimeSpawnReduction(field_70146_Z, field_70170_p))
         {
             return false;
         }
@@ -346,7 +346,7 @@
 
     protected void func_70664_aZ()
     {
-        this.field_70181_x = (double)0.42F;
+        this.field_70181_x = 0.41999998688697815D;
         this.field_70160_al = true;
     }
 
@@ -369,6 +369,14 @@
     {
         return this.func_189101_db() ? SoundEvents.field_189110_fE : SoundEvents.field_187882_fq;
     }
+
+    /* ======================================== FORGE START =====================================*/
+    /**
+     * Called when the slime spawns particles on landing, see onUpdate.
+     * Return true to prevent the spawning of the default particles.
+     */
+    protected boolean spawnCustomParticles() { return false; }
+    /* ======================================== FORGE END   =====================================*/
 
     static class AISlimeAttack extends EntityAIBase
         {
