--- before/net/minecraft/entity/monster/EntitySlime.java
+++ after/net/minecraft/entity/monster/EntitySlime.java
@@ -142,7 +142,7 @@
         if (this.onGround && !this.wasOnGround)
         {
             int i = this.getSlimeSize();
-
+            if (spawnCustomParticles()) { i = 0; } // don't spawn particles if it's handled by the implementation itself
             for (int j = 0; j < i * 8; j++)
             {
                 float f = this.rand.nextFloat() * (float)(Math.PI * 2);
@@ -325,7 +325,7 @@
         BlockPos blockpos = new BlockPos(MathHelper.floor(this.posX), 0, MathHelper.floor(this.posZ));
         Chunk chunk = this.world.getChunk(blockpos);
 
-        if (this.world.getWorldInfo().getTerrainType() == WorldType.FLAT && this.rand.nextInt(4) != 1)
+        if (this.world.getWorldInfo().getTerrainType().handleSlimeSpawnReduction(rand, world))
         {
             return false;
         }
@@ -377,6 +377,7 @@
     {
         this.motionY = 0.42F;
         this.isAirBorne = true;
+        net.minecraftforge.common.ForgeHooks.onLivingJump(this);
     }
 
     @Nullable
@@ -399,6 +400,14 @@
     {
         return this.isSmallSlime() ? SoundEvents.ENTITY_SMALL_SLIME_JUMP : SoundEvents.ENTITY_SLIME_JUMP;
     }
+
+    /* ======================================== FORGE START =====================================*/
+    /**
+     * Called when the slime spawns particles on landing, see onUpdate.
+     * Return true to prevent the spawning of the default particles.
+     */
+    protected boolean spawnCustomParticles() { return false; }
+    /* ======================================== FORGE END   =====================================*/
 
     static class AISlimeAttack extends EntityAIBase
         {
