--- before/net/minecraft/entity/monster/EntityCreeper.java
+++ after/net/minecraft/entity/monster/EntityCreeper.java
@@ -39,9 +39,9 @@
 
 public class EntityCreeper extends EntityMob
 {
-    private static final DataParameter<Integer> field_184713_a = EntityDataManager.func_187226_a(EntityCreeper.class, DataSerializers.field_187192_b);
-    private static final DataParameter<Boolean> field_184714_b = EntityDataManager.func_187226_a(EntityCreeper.class, DataSerializers.field_187198_h);
-    private static final DataParameter<Boolean> field_184715_c = EntityDataManager.func_187226_a(EntityCreeper.class, DataSerializers.field_187198_h);
+    private static final DataParameter<Integer> field_184713_a = EntityDataManager.<Integer>func_187226_a(EntityCreeper.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Boolean> field_184714_b = EntityDataManager.<Boolean>func_187226_a(EntityCreeper.class, DataSerializers.field_187198_h);
+    private static final DataParameter<Boolean> field_184715_c = EntityDataManager.<Boolean>func_187226_a(EntityCreeper.class, DataSerializers.field_187198_h);
     private int field_70834_e;
     private int field_70833_d;
     private int field_82225_f = 30;
@@ -54,34 +54,30 @@
         this.func_70105_a(0.6F, 1.7F);
     }
 
-    @Override
     protected void func_184651_r()
     {
         this.field_70714_bg.func_75776_a(1, new EntityAISwimming(this));
         this.field_70714_bg.func_75776_a(2, new EntityAICreeperSwell(this));
-        this.field_70714_bg.func_75776_a(3, new EntityAIAvoidEntity<>(this, EntityOcelot.class, 6.0F, 1.0, 1.2));
-        this.field_70714_bg.func_75776_a(4, new EntityAIAttackMelee(this, 1.0, false));
-        this.field_70714_bg.func_75776_a(5, new EntityAIWanderAvoidWater(this, 0.8));
+        this.field_70714_bg.func_75776_a(3, new EntityAIAvoidEntity(this, EntityOcelot.class, 6.0F, 1.0D, 1.2D));
+        this.field_70714_bg.func_75776_a(4, new EntityAIAttackMelee(this, 1.0D, false));
+        this.field_70714_bg.func_75776_a(5, new EntityAIWanderAvoidWater(this, 0.8D));
         this.field_70714_bg.func_75776_a(6, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
         this.field_70714_bg.func_75776_a(6, new EntityAILookIdle(this));
-        this.field_70715_bh.func_75776_a(1, new EntityAINearestAttackableTarget<>(this, EntityPlayer.class, true));
-        this.field_70715_bh.func_75776_a(2, new EntityAIHurtByTarget(this, false));
+        this.field_70715_bh.func_75776_a(1, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
+        this.field_70715_bh.func_75776_a(2, new EntityAIHurtByTarget(this, false, new Class[0]));
     }
 
-    @Override
     protected void func_110147_ax()
     {
         super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.25);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.25D);
     }
 
-    @Override
     public int func_82143_as()
     {
         return this.func_70638_az() == null ? 3 : 3 + (int)(this.func_110143_aJ() - 1.0F);
     }
 
-    @Override
     public void func_180430_e(float p_180430_1_, float p_180430_2_)
     {
         super.func_180430_e(p_180430_1_, p_180430_2_);
@@ -93,13 +89,12 @@
         }
     }
 
-    @Override
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184713_a, -1);
-        this.field_70180_af.func_187214_a(field_184714_b, false);
-        this.field_70180_af.func_187214_a(field_184715_c, false);
+        this.field_70180_af.func_187214_a(field_184713_a, Integer.valueOf(-1));
+        this.field_70180_af.func_187214_a(field_184714_b, Boolean.valueOf(false));
+        this.field_70180_af.func_187214_a(field_184715_c, Boolean.valueOf(false));
     }
 
     public static void func_189762_b(DataFixer p_189762_0_)
@@ -107,12 +102,11 @@
         EntityLiving.func_189752_a(p_189762_0_, EntityCreeper.class);
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
 
-        if (this.field_70180_af.func_187225_a(field_184714_b))
+        if (((Boolean)this.field_70180_af.func_187225_a(field_184714_b)).booleanValue())
         {
             p_70014_1_.func_74757_a("powered", true);
         }
@@ -122,11 +116,10 @@
         p_70014_1_.func_74757_a("ignited", this.func_146078_ca());
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
-        this.field_70180_af.func_187227_b(field_184714_b, p_70037_1_.func_74767_n("powered"));
+        this.field_70180_af.func_187227_b(field_184714_b, Boolean.valueOf(p_70037_1_.func_74767_n("powered")));
 
         if (p_70037_1_.func_150297_b("Fuse", 99))
         {
@@ -144,7 +137,6 @@
         }
     }
 
-    @Override
     public void func_70071_h_()
     {
         if (this.func_70089_S())
@@ -180,19 +172,16 @@
         super.func_70071_h_();
     }
 
-    @Override
     protected SoundEvent func_184601_bQ(DamageSource p_184601_1_)
     {
         return SoundEvents.field_187570_aq;
     }
 
-    @Override
     protected SoundEvent func_184615_bR()
     {
         return SoundEvents.field_187568_ap;
     }
 
-    @Override
     public void func_70645_a(DamageSource p_70645_1_)
     {
         super.func_70645_a(p_70645_1_);
@@ -206,10 +195,7 @@
                 int k = i + this.field_70146_Z.nextInt(j - i + 1);
                 this.func_145779_a(Item.func_150899_d(k), 1);
             }
-            else if (p_70645_1_.func_76346_g() instanceof EntityCreeper
-                     && p_70645_1_.func_76346_g() != this
-                     && ((EntityCreeper)p_70645_1_.func_76346_g()).func_70830_n()
-                     && ((EntityCreeper)p_70645_1_.func_76346_g()).func_70650_aV())
+            else if (p_70645_1_.func_76346_g() instanceof EntityCreeper && p_70645_1_.func_76346_g() != this && ((EntityCreeper)p_70645_1_.func_76346_g()).func_70830_n() && ((EntityCreeper)p_70645_1_.func_76346_g()).func_70650_aV())
             {
                 ((EntityCreeper)p_70645_1_.func_76346_g()).func_175493_co();
                 this.func_70099_a(new ItemStack(Items.field_151144_bL, 1, 4), 0.0F);
@@ -217,7 +203,6 @@
         }
     }
 
-    @Override
     public boolean func_70652_k(Entity p_70652_1_)
     {
         return true;
@@ -225,7 +210,7 @@
 
     public boolean func_70830_n()
     {
-        return this.field_70180_af.func_187225_a(field_184714_b);
+        return ((Boolean)this.field_70180_af.func_187225_a(field_184714_b)).booleanValue();
     }
 
     @SideOnly(Side.CLIENT)
@@ -235,7 +220,6 @@
     }
 
     @Nullable
-    @Override
     protected ResourceLocation func_184647_J()
     {
         return LootTableList.field_186434_p;
@@ -243,39 +227,27 @@
 
     public int func_70832_p()
     {
-        return this.field_70180_af.func_187225_a(field_184713_a);
+        return ((Integer)this.field_70180_af.func_187225_a(field_184713_a)).intValue();
     }
 
     public void func_70829_a(int p_70829_1_)
     {
-        this.field_70180_af.func_187227_b(field_184713_a, p_70829_1_);
+        this.field_70180_af.func_187227_b(field_184713_a, Integer.valueOf(p_70829_1_));
     }
 
-    @Override
     public void func_70077_a(EntityLightningBolt p_70077_1_)
     {
         super.func_70077_a(p_70077_1_);
-        this.field_70180_af.func_187227_b(field_184714_b, true);
+        this.field_70180_af.func_187227_b(field_184714_b, Boolean.valueOf(true));
     }
 
-    @Override
     protected boolean func_184645_a(EntityPlayer p_184645_1_, EnumHand p_184645_2_)
     {
         ItemStack itemstack = p_184645_1_.func_184586_b(p_184645_2_);
 
         if (itemstack.func_77973_b() == Items.field_151033_d)
         {
-            this.field_70170_p
-            .func_184148_a(
-                p_184645_1_,
-                this.field_70165_t,
-                this.field_70163_u,
-                this.field_70161_v,
-                SoundEvents.field_187649_bu,
-                this.func_184176_by(),
-                1.0F,
-                this.field_70146_Z.nextFloat() * 0.4F + 0.8F
-            );
+            this.field_70170_p.func_184148_a(p_184645_1_, this.field_70165_t, this.field_70163_u, this.field_70161_v, SoundEvents.field_187649_bu, this.func_184176_by(), 1.0F, this.field_70146_Z.nextFloat() * 0.4F + 0.8F);
             p_184645_1_.func_184609_a(p_184645_2_);
 
             if (!this.field_70170_p.field_72995_K)
@@ -293,7 +265,7 @@
     {
         if (!this.field_70170_p.field_72995_K)
         {
-            boolean flag = this.field_70170_p.func_82736_K().func_82766_b("mobGriefing");
+            boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this);
             float f = this.func_70830_n() ? 2.0F : 1.0F;
             this.field_70729_aU = true;
             this.field_70170_p.func_72876_a(this, this.field_70165_t, this.field_70163_u, this.field_70161_v, (float)this.field_82226_g * f, flag);
@@ -326,12 +298,12 @@
 
     public boolean func_146078_ca()
     {
-        return this.field_70180_af.func_187225_a(field_184715_c);
+        return ((Boolean)this.field_70180_af.func_187225_a(field_184715_c)).booleanValue();
     }
 
     public void func_146079_cb()
     {
-        this.field_70180_af.func_187227_b(field_184715_c, true);
+        this.field_70180_af.func_187227_b(field_184715_c, Boolean.valueOf(true));
     }
 
     public boolean func_70650_aV()
