--- before/net/minecraft/entity/monster/EntityMob.java
+++ after/net/minecraft/entity/monster/EntityMob.java
@@ -138,14 +138,14 @@
 
                 if (!itemstack.isEmpty()
                         && !itemstack1.isEmpty()
-                        && itemstack.getItem() instanceof ItemAxe
-                        && itemstack1.getItem() == Items.SHIELD)
+                        && itemstack.getItem().canDisableShield(itemstack, itemstack1, entityplayer, this)
+                        && itemstack1.getItem().isShield(itemstack1, entityplayer))
                 {
                     float f1 = 0.25F + (float)EnchantmentHelper.getEfficiencyModifier(this) * 0.05F;
 
                     if (this.rand.nextFloat() < f1)
                     {
-                        entityplayer.getCooldownTracker().setCooldown(Items.SHIELD, 100);
+                        entityplayer.getCooldownTracker().setCooldown(itemstack1.getItem(), 100);
                         this.world.setEntityState(entityplayer, (byte)30);
                     }
                 }
