--- before/net/minecraft/entity/monster/EntityZombie.java
+++ after/net/minecraft/entity/monster/EntityZombie.java
@@ -54,10 +54,9 @@
 
 public class EntityZombie extends EntityMob
 {
-    protected static final IAttribute field_110186_bp = new RangedAttribute(null, "zombie.spawnReinforcements", 0.0, 0.0, 1.0)
-        .func_111117_a("Spawn Reinforcements Chance");
+    protected static final IAttribute field_110186_bp = (new RangedAttribute((IAttribute)null, "zombie.spawnReinforcements", 0.0D, 0.0D, 1.0D)).func_111117_a("Spawn Reinforcements Chance");
     private static final UUID field_110187_bq = UUID.fromString("B9766B59-9566-4402-BC1F-2EE2A276D836");
-    private static final AttributeModifier field_110188_br = new AttributeModifier(field_110187_bq, "Baby speed boost", 0.5, 1);
+    private static final AttributeModifier field_110188_br = new AttributeModifier(field_110187_bq, "Baby speed boost", 0.5D, 1);
     private static final DataParameter<Boolean> field_184737_bv = EntityDataManager.func_187226_a(EntityZombie.class, DataSerializers.field_187198_h);
     private static final DataParameter<Integer> field_184738_bw = EntityDataManager.func_187226_a(EntityZombie.class, DataSerializers.field_187192_b);
     private static final DataParameter<Boolean> field_184740_by = EntityDataManager.func_187226_a(EntityZombie.class, DataSerializers.field_187198_h);
@@ -72,13 +71,12 @@
         this.func_70105_a(0.6F, 1.95F);
     }
 
-    @Override
     protected void func_184651_r()
     {
         this.field_70714_bg.func_75776_a(0, new EntityAISwimming(this));
-        this.field_70714_bg.func_75776_a(2, new EntityAIZombieAttack(this, 1.0, false));
-        this.field_70714_bg.func_75776_a(5, new EntityAIMoveTowardsRestriction(this, 1.0));
-        this.field_70714_bg.func_75776_a(7, new EntityAIWanderAvoidWater(this, 1.0));
+        this.field_70714_bg.func_75776_a(2, new EntityAIZombieAttack(this, 1.0D, false));
+        this.field_70714_bg.func_75776_a(5, new EntityAIMoveTowardsRestriction(this, 1.0D));
+        this.field_70714_bg.func_75776_a(7, new EntityAIWanderAvoidWater(this, 1.0D));
         this.field_70714_bg.func_75776_a(8, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
         this.field_70714_bg.func_75776_a(8, new EntityAILookIdle(this));
         this.func_175456_n();
@@ -86,25 +84,23 @@
 
     protected void func_175456_n()
     {
-        this.field_70714_bg.func_75776_a(6, new EntityAIMoveThroughVillage(this, 1.0, false));
+        this.field_70714_bg.func_75776_a(6, new EntityAIMoveThroughVillage(this, 1.0D, false));
         this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, true, EntityPigZombie.class));
         this.field_70715_bh.func_75776_a(2, new EntityAINearestAttackableTarget<>(this, EntityPlayer.class, true));
         this.field_70715_bh.func_75776_a(3, new EntityAINearestAttackableTarget<>(this, EntityVillager.class, false));
         this.field_70715_bh.func_75776_a(3, new EntityAINearestAttackableTarget<>(this, EntityIronGolem.class, true));
     }
 
-    @Override
     protected void func_110147_ax()
     {
         super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111265_b).func_111128_a(35.0);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.23F);
-        this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111128_a(3.0);
-        this.func_110148_a(SharedMonsterAttributes.field_188791_g).func_111128_a(2.0);
-        this.func_110140_aT().func_111150_b(field_110186_bp).func_111128_a(this.field_70146_Z.nextDouble() * 0.1F);
+        this.func_110148_a(SharedMonsterAttributes.field_111265_b).func_111128_a(35.0D);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a((double)0.23F);
+        this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111128_a(3.0D);
+        this.func_110148_a(SharedMonsterAttributes.field_188791_g).func_111128_a(2.0D);
+        this.func_110140_aT().func_111150_b(field_110186_bp).func_111128_a(this.field_70146_Z.nextDouble() * (double)0.1F);
     }
 
-    @Override
     protected void func_70088_a()
     {
         super.func_70088_a();
@@ -147,13 +143,11 @@
         }
     }
 
-    @Override
     public boolean func_70631_g_()
     {
         return this.func_184212_Q().func_187225_a(field_184737_bv);
     }
 
-    @Override
     protected int func_70693_a(EntityPlayer p_70693_1_)
     {
         if (this.func_70631_g_())
@@ -182,7 +176,6 @@
         this.func_146071_k(p_82227_1_);
     }
 
-    @Override
     public void func_184206_a(DataParameter<?> p_184206_1_)
     {
         if (field_184737_bv.equals(p_184206_1_))
@@ -193,16 +186,13 @@
         super.func_184206_a(p_184206_1_);
     }
 
-    @Override
     public void func_70636_d()
     {
         if (this.field_70170_p.func_72935_r() && !this.field_70170_p.field_72995_K && !this.func_70631_g_() && this.func_190730_o())
         {
             float f = this.func_70013_c();
 
-            if (f > 0.5F
-                    && this.field_70146_Z.nextFloat() * 30.0F < (f - 0.4F) * 2.0F
-                    && this.field_70170_p.func_175678_i(new BlockPos(this.field_70165_t, this.field_70163_u + (double)this.func_70047_e(), this.field_70161_v)))
+            if (f > 0.5F && this.field_70146_Z.nextFloat() * 30.0F < (f - 0.4F) * 2.0F && this.field_70170_p.func_175678_i(new BlockPos(this.field_70165_t, this.field_70163_u + (double)this.func_70047_e(), this.field_70161_v)))
             {
                 boolean flag = true;
                 ItemStack itemstack = this.func_184582_a(EntityEquipmentSlot.HEAD);
@@ -238,7 +228,6 @@
         return true;
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         if (super.func_70097_a(p_70097_1_, p_70097_2_))
@@ -250,37 +239,42 @@
                 entitylivingbase = (EntityLivingBase)p_70097_1_.func_76346_g();
             }
 
-            if (entitylivingbase != null
-                    && this.field_70170_p.func_175659_aa() == EnumDifficulty.HARD
-                    && (double)this.field_70146_Z.nextFloat() < this.func_110148_a(field_110186_bp).func_111126_e()
-                    && this.field_70170_p.func_82736_K().func_82766_b("doMobSpawning"))
+            int i = MathHelper.func_76128_c(this.field_70165_t);
+            int j = MathHelper.func_76128_c(this.field_70163_u);
+            int k = MathHelper.func_76128_c(this.field_70161_v);
+            net.minecraftforge.event.entity.living.ZombieEvent.SummonAidEvent summonAid = net.minecraftforge.event.ForgeEventFactory.fireZombieSummonAid(this, field_70170_p, i, j, k, entitylivingbase, this.func_110148_a(field_110186_bp).func_111126_e());
+            if (summonAid.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) return true;
+
+            if (summonAid.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW  ||
+                entitylivingbase != null && this.field_70170_p.func_175659_aa() == EnumDifficulty.HARD && (double)this.field_70146_Z.nextFloat() < this.func_110148_a(field_110186_bp).func_111126_e() && this.field_70170_p.func_82736_K().func_82766_b("doMobSpawning"))
             {
-                int i = MathHelper.func_76128_c(this.field_70165_t);
-                int j = MathHelper.func_76128_c(this.field_70163_u);
-                int k = MathHelper.func_76128_c(this.field_70161_v);
-                EntityZombie entityzombie = new EntityZombie(this.field_70170_p);
+                EntityZombie entityzombie;
+                if (summonAid.getCustomSummonedAid() != null && summonAid.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW)
+                {
+                    entityzombie = summonAid.getCustomSummonedAid();
+                }
+                else
+                {
+                    entityzombie = new EntityZombie(this.field_70170_p);
+                }
 
-                for (int l = 0; l < 50; l++)
+                for (int l = 0; l < 50; ++l)
                 {
                     int i1 = i + MathHelper.func_76136_a(this.field_70146_Z, 7, 40) * MathHelper.func_76136_a(this.field_70146_Z, -1, 1);
                     int j1 = j + MathHelper.func_76136_a(this.field_70146_Z, 7, 40) * MathHelper.func_76136_a(this.field_70146_Z, -1, 1);
                     int k1 = k + MathHelper.func_76136_a(this.field_70146_Z, 7, 40) * MathHelper.func_76136_a(this.field_70146_Z, -1, 1);
 
-                    if (this.field_70170_p.func_180495_p(new BlockPos(i1, j1 - 1, k1)).func_185896_q()
-                            && this.field_70170_p.func_175671_l(new BlockPos(i1, j1, k1)) < 10)
+                    if (this.field_70170_p.func_180495_p(new BlockPos(i1, j1 - 1, k1)).isSideSolid(this.field_70170_p, new BlockPos(i1, j1 - 1, k1), net.minecraft.util.EnumFacing.UP) && this.field_70170_p.func_175671_l(new BlockPos(i1, j1, k1)) < 10)
                     {
                         entityzombie.func_70107_b((double)i1, (double)j1, (double)k1);
 
-                        if (!this.field_70170_p.func_175636_b((double)i1, (double)j1, (double)k1, 7.0)
-                                && this.field_70170_p.func_72917_a(entityzombie.func_174813_aQ(), entityzombie)
-                                && this.field_70170_p.func_184144_a(entityzombie, entityzombie.func_174813_aQ()).isEmpty()
-                                && !this.field_70170_p.func_72953_d(entityzombie.func_174813_aQ()))
+                        if (!this.field_70170_p.func_175636_b((double)i1, (double)j1, (double)k1, 7.0D) && this.field_70170_p.func_72917_a(entityzombie.func_174813_aQ(), entityzombie) && this.field_70170_p.func_184144_a(entityzombie, entityzombie.func_174813_aQ()).isEmpty() && !this.field_70170_p.func_72953_d(entityzombie.func_174813_aQ()))
                         {
                             this.field_70170_p.func_72838_d(entityzombie);
-                            entityzombie.func_70624_b(entitylivingbase);
-                            entityzombie.func_180482_a(this.field_70170_p.func_175649_E(new BlockPos(entityzombie)), null);
-                            this.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Zombie reinforcement caller charge", -0.05F, 0));
-                            entityzombie.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Zombie reinforcement callee charge", -0.05F, 0));
+                            if (entitylivingbase != null) entityzombie.func_70624_b(entitylivingbase);
+                            entityzombie.func_180482_a(this.field_70170_p.func_175649_E(new BlockPos(entityzombie)), (IEntityLivingData)null);
+                            this.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Zombie reinforcement caller charge", -0.05000000074505806D, 0));
+                            entityzombie.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Zombie reinforcement callee charge", -0.05000000074505806D, 0));
                             break;
                         }
                     }
@@ -295,7 +289,6 @@
         }
     }
 
-    @Override
     public boolean func_70652_k(Entity p_70652_1_)
     {
         boolean flag = super.func_70652_k(p_70652_1_);
@@ -313,19 +306,16 @@
         return flag;
     }
 
-    @Override
     protected SoundEvent func_184639_G()
     {
         return SoundEvents.field_187899_gZ;
     }
 
-    @Override
     protected SoundEvent func_184601_bQ(DamageSource p_184601_1_)
     {
         return SoundEvents.field_187934_hh;
     }
 
-    @Override
     protected SoundEvent func_184615_bR()
     {
         return SoundEvents.field_187930_hd;
@@ -336,26 +326,22 @@
         return SoundEvents.field_187939_hm;
     }
 
-    @Override
     protected void func_180429_a(BlockPos p_180429_1_, Block p_180429_2_)
     {
         this.func_184185_a(this.func_190731_di(), 0.15F, 1.0F);
     }
 
-    @Override
     public EnumCreatureAttribute func_70668_bt()
     {
         return EnumCreatureAttribute.UNDEAD;
     }
 
     @Nullable
-    @Override
     protected ResourceLocation func_184647_J()
     {
         return LootTableList.field_186383_ah;
     }
 
-    @Override
     protected void func_180481_a(DifficultyInstance p_180481_1_)
     {
         super.func_180481_a(p_180481_1_);
@@ -380,7 +366,6 @@
         EntityLiving.func_189752_a(p_189779_0_, EntityZombie.class);
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
@@ -393,7 +378,6 @@
         p_70014_1_.func_74757_a("CanBreakDoors", this.func_146072_bX());
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
@@ -406,13 +390,11 @@
         this.func_146070_a(p_70037_1_.func_74767_n("CanBreakDoors"));
     }
 
-    @Override
     public void func_70074_a(EntityLivingBase p_70074_1_)
     {
         super.func_70074_a(p_70074_1_);
 
-        if ((this.field_70170_p.func_175659_aa() == EnumDifficulty.NORMAL || this.field_70170_p.func_175659_aa() == EnumDifficulty.HARD)
-                && p_70074_1_ instanceof EntityVillager)
+        if ((this.field_70170_p.func_175659_aa() == EnumDifficulty.NORMAL || this.field_70170_p.func_175659_aa() == EnumDifficulty.HARD) && p_70074_1_ instanceof EntityVillager)
         {
             if (this.field_70170_p.func_175659_aa() != EnumDifficulty.HARD && this.field_70146_Z.nextBoolean())
             {
@@ -435,31 +417,28 @@
             }
 
             this.field_70170_p.func_72838_d(entityzombievillager);
-            this.field_70170_p.func_180498_a(null, 1026, new BlockPos(this), 0);
+            this.field_70170_p.func_180498_a((EntityPlayer)null, 1026, new BlockPos(this), 0);
         }
     }
 
-    @Override
     public float func_70047_e()
     {
         float f = 1.74F;
 
         if (this.func_70631_g_())
         {
-            f = (float)((double)f - 0.81);
+            f = (float)((double)f - 0.81D);
         }
 
         return f;
     }
 
-    @Override
     protected boolean func_175448_a(ItemStack p_175448_1_)
     {
         return p_175448_1_.func_77973_b() == Items.field_151110_aK && this.func_70631_g_() && this.func_184218_aH() ? false : super.func_175448_a(p_175448_1_);
     }
 
     @Nullable
-    @Override
     public IEntityLivingData func_180482_a(DifficultyInstance p_180482_1_, @Nullable IEntityLivingData p_180482_2_)
     {
         p_180482_2_ = super.func_180482_a(p_180482_1_, p_180482_2_);
@@ -468,7 +447,7 @@
 
         if (p_180482_2_ == null)
         {
-            p_180482_2_ = new EntityZombie.GroupData(this.field_70170_p.field_73012_v.nextFloat() < 0.05F);
+            p_180482_2_ = new EntityZombie.GroupData(this.field_70170_p.field_73012_v.nextFloat() < net.minecraftforge.common.ForgeModContainer.zombieBabyChance);
         }
 
         if (p_180482_2_ instanceof EntityZombie.GroupData)
@@ -479,10 +458,9 @@
             {
                 this.func_82227_f(true);
 
-                if ((double)this.field_70170_p.field_73012_v.nextFloat() < 0.05)
+                if ((double)this.field_70170_p.field_73012_v.nextFloat() < 0.05D)
                 {
-                    List<EntityChicken> list = this.field_70170_p
-                                               .func_175647_a(EntityChicken.class, this.func_174813_aQ().func_72314_b(5.0, 3.0, 5.0), EntitySelectors.field_152785_b);
+                    List<EntityChicken> list = this.field_70170_p.func_175647_a(EntityChicken.class, this.func_174813_aQ().func_72314_b(5.0D, 3.0D, 5.0D), EntitySelectors.field_152785_b);
 
                     if (!list.isEmpty())
                     {
@@ -491,11 +469,11 @@
                         this.func_184220_m(entitychicken);
                     }
                 }
-                else if ((double)this.field_70170_p.field_73012_v.nextFloat() < 0.05)
+                else if ((double)this.field_70170_p.field_73012_v.nextFloat() < 0.05D)
                 {
                     EntityChicken entitychicken1 = new EntityChicken(this.field_70170_p);
                     entitychicken1.func_70012_b(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70177_z, 0.0F);
-                    entitychicken1.func_180482_a(p_180482_1_, null);
+                    entitychicken1.func_180482_a(p_180482_1_, (IEntityLivingData)null);
                     entitychicken1.func_152117_i(true);
                     this.field_70170_p.func_72838_d(entitychicken1);
                     this.func_184220_m(entitychicken1);
@@ -518,20 +496,18 @@
             }
         }
 
-        this.func_110148_a(SharedMonsterAttributes.field_111266_c)
-        .func_111121_a(new AttributeModifier("Random spawn bonus", this.field_70146_Z.nextDouble() * 0.05F, 0));
-        double d0 = this.field_70146_Z.nextDouble() * 1.5 * (double)f;
+        this.func_110148_a(SharedMonsterAttributes.field_111266_c).func_111121_a(new AttributeModifier("Random spawn bonus", this.field_70146_Z.nextDouble() * (double)0.05F, 0));
+        double d0 = this.field_70146_Z.nextDouble() * 1.5D * (double)f;
 
-        if (d0 > 1.0)
+        if (d0 > 1.0D)
         {
             this.func_110148_a(SharedMonsterAttributes.field_111265_b).func_111121_a(new AttributeModifier("Random zombie-spawn bonus", d0, 2));
         }
 
         if (this.field_70146_Z.nextFloat() < f * 0.05F)
         {
-            this.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Leader zombie bonus", this.field_70146_Z.nextDouble() * 0.25 + 0.5, 0));
-            this.func_110148_a(SharedMonsterAttributes.field_111267_a)
-            .func_111121_a(new AttributeModifier("Leader zombie bonus", this.field_70146_Z.nextDouble() * 3.0 + 1.0, 2));
+            this.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Leader zombie bonus", this.field_70146_Z.nextDouble() * 0.25D + 0.5D, 0));
+            this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111121_a(new AttributeModifier("Leader zombie bonus", this.field_70146_Z.nextDouble() * 3.0D + 1.0D, 2));
             this.func_146070_a(true);
         }
 
@@ -543,7 +519,6 @@
         this.func_146069_a(p_146071_1_ ? 0.5F : 1.0F);
     }
 
-    @Override
     protected final void func_70105_a(float p_70105_1_, float p_70105_2_)
     {
         boolean flag = this.field_146074_bv > 0.0F && this.field_146073_bw > 0.0F;
@@ -561,13 +536,11 @@
         super.func_70105_a(this.field_146074_bv * p_146069_1_, this.field_146073_bw * p_146069_1_);
     }
 
-    @Override
     public double func_70033_W()
     {
-        return this.func_70631_g_() ? 0.0 : -0.45;
+        return this.func_70631_g_() ? 0.0D : -0.45D;
     }
 
-    @Override
     public void func_70645_a(DamageSource p_70645_1_)
     {
         super.func_70645_a(p_70645_1_);
