--- before/net/minecraft/entity/monster/EntityZombieVillager.java
+++ after/net/minecraft/entity/monster/EntityZombieVillager.java
@@ -33,8 +33,8 @@
 
 public class EntityZombieVillager extends EntityZombie
 {
-    private static final DataParameter<Boolean> field_184739_bx = EntityDataManager.func_187226_a(EntityZombieVillager.class, DataSerializers.field_187198_h);
-    private static final DataParameter<Integer> field_190739_c = EntityDataManager.func_187226_a(EntityZombieVillager.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Boolean> field_184739_bx = EntityDataManager.<Boolean>func_187226_a(EntityZombieVillager.class, DataSerializers.field_187198_h);
+    private static final DataParameter<Integer> field_190739_c = EntityDataManager.<Integer>func_187226_a(EntityZombieVillager.class, DataSerializers.field_187192_b);
     private int field_82234_d;
     private UUID field_191992_by;
 
@@ -43,22 +43,24 @@
         super(p_i47277_1_);
     }
 
-    @Override
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184739_bx, false);
-        this.field_70180_af.func_187214_a(field_190739_c, 0);
+        this.field_70180_af.func_187214_a(field_184739_bx, Boolean.valueOf(false));
+        this.field_70180_af.func_187214_a(field_190739_c, Integer.valueOf(0));
     }
 
     public void func_190733_a(int p_190733_1_)
     {
-        this.field_70180_af.func_187227_b(field_190739_c, p_190733_1_);
+        this.field_70180_af.func_187227_b(field_190739_c, Integer.valueOf(p_190733_1_));
+        net.minecraftforge.fml.common.registry.VillagerRegistry.onSetProfession(this, p_190733_1_);
     }
 
+    //Use Forge Variant below
+    @Deprecated
     public int func_190736_dl()
     {
-        return Math.max(this.field_70180_af.func_187225_a(field_190739_c) % 6, 0);
+        return Math.max(((Integer)this.field_70180_af.func_187225_a(field_190739_c)).intValue(), 0);
     }
 
     public static void func_190737_b(DataFixer p_190737_0_)
@@ -66,11 +68,11 @@
         EntityLiving.func_189752_a(p_190737_0_, EntityZombieVillager.class);
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
         p_70014_1_.func_74768_a("Profession", this.func_190736_dl());
+        p_70014_1_.func_74778_a("ProfessionName", this.getForgeProfession().getRegistryName().toString());
         p_70014_1_.func_74768_a("ConversionTime", this.func_82230_o() ? this.field_82234_d : -1);
 
         if (this.field_191992_by != null)
@@ -79,29 +81,30 @@
         }
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
         this.func_190733_a(p_70037_1_.func_74762_e("Profession"));
+        if (p_70037_1_.func_74764_b("ProfessionName"))
+        {
+            net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession p = net.minecraftforge.fml.common.registry.ForgeRegistries.VILLAGER_PROFESSIONS.getValue(new net.minecraft.util.ResourceLocation(p_70037_1_.func_74779_i("ProfessionName")));
+            if (p == null) p = net.minecraftforge.fml.common.registry.VillagerRegistry.FARMER;
+            this.setForgeProfession(p);
+        }
 
         if (p_70037_1_.func_150297_b("ConversionTime", 99) && p_70037_1_.func_74762_e("ConversionTime") > -1)
         {
-            this.func_191991_a(
-                p_70037_1_.func_186855_b("ConversionPlayer") ? p_70037_1_.func_186857_a("ConversionPlayer") : null, p_70037_1_.func_74762_e("ConversionTime")
-            );
+            this.func_191991_a(p_70037_1_.func_186855_b("ConversionPlayer") ? p_70037_1_.func_186857_a("ConversionPlayer") : null, p_70037_1_.func_74762_e("ConversionTime"));
         }
     }
 
     @Nullable
-    @Override
     public IEntityLivingData func_180482_a(DifficultyInstance p_180482_1_, @Nullable IEntityLivingData p_180482_2_)
     {
-        this.func_190733_a(this.field_70170_p.field_73012_v.nextInt(6));
+        net.minecraftforge.fml.common.registry.VillagerRegistry.setRandomProfession(this, this.field_70170_p.field_73012_v);
         return super.func_180482_a(p_180482_1_, p_180482_2_);
     }
 
-    @Override
     public void func_70071_h_()
     {
         if (!this.field_70170_p.field_72995_K && this.func_82230_o())
@@ -118,7 +121,6 @@
         super.func_70071_h_();
     }
 
-    @Override
     public boolean func_184645_a(EntityPlayer p_184645_1_, EnumHand p_184645_2_)
     {
         ItemStack itemstack = p_184645_1_.func_184586_b(p_184645_2_);
@@ -143,7 +145,6 @@
         }
     }
 
-    @Override
     protected boolean func_70692_ba()
     {
         return !this.func_82230_o();
@@ -151,38 +152,27 @@
 
     public boolean func_82230_o()
     {
-        return this.func_184212_Q().func_187225_a(field_184739_bx);
+        return ((Boolean)this.func_184212_Q().func_187225_a(field_184739_bx)).booleanValue();
     }
 
     protected void func_191991_a(@Nullable UUID p_191991_1_, int p_191991_2_)
     {
         this.field_191992_by = p_191991_1_;
         this.field_82234_d = p_191991_2_;
-        this.func_184212_Q().func_187227_b(field_184739_bx, true);
+        this.func_184212_Q().func_187227_b(field_184739_bx, Boolean.valueOf(true));
         this.func_184589_d(MobEffects.field_76437_t);
         this.func_70690_d(new PotionEffect(MobEffects.field_76420_g, p_191991_2_, Math.min(this.field_70170_p.func_175659_aa().func_151525_a() - 1, 0)));
         this.field_70170_p.func_72960_a(this, (byte)16);
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_70103_a(byte p_70103_1_)
     {
         if (p_70103_1_ == 16)
         {
             if (!this.func_174814_R())
             {
-                this.field_70170_p
-                .func_184134_a(
-                    this.field_70165_t + 0.5,
-                    this.field_70163_u + 0.5,
-                    this.field_70161_v + 0.5,
-                    SoundEvents.field_187942_hp,
-                    this.func_184176_by(),
-                    1.0F + this.field_70146_Z.nextFloat(),
-                    this.field_70146_Z.nextFloat() * 0.7F + 0.3F,
-                    false
-                );
+                this.field_70170_p.func_184134_a(this.field_70165_t + 0.5D, this.field_70163_u + 0.5D, this.field_70161_v + 0.5D, SoundEvents.field_187942_hp, this.func_184176_by(), 1.0F + this.field_70146_Z.nextFloat(), this.field_70146_Z.nextFloat() * 0.7F + 0.3F, false);
             }
         }
         else
@@ -195,8 +185,8 @@
     {
         EntityVillager entityvillager = new EntityVillager(this.field_70170_p);
         entityvillager.func_82149_j(this);
-        entityvillager.func_70938_b(this.func_190736_dl());
-        entityvillager.func_190672_a(this.field_70170_p.func_175649_E(new BlockPos(entityvillager)), null, false);
+        entityvillager.setProfession(this.getForgeProfession());
+        entityvillager.func_190672_a(this.field_70170_p.func_175649_E(new BlockPos(entityvillager)), (IEntityLivingData)null, false);
         entityvillager.func_82187_q();
 
         if (this.func_70631_g_())
@@ -226,7 +216,7 @@
         }
 
         entityvillager.func_70690_d(new PotionEffect(MobEffects.field_76431_k, 200, 0));
-        this.field_70170_p.func_180498_a(null, 1027, new BlockPos((int)this.field_70165_t, (int)this.field_70163_u, (int)this.field_70161_v), 0);
+        this.field_70170_p.func_180498_a((EntityPlayer)null, 1027, new BlockPos((int)this.field_70165_t, (int)this.field_70163_u, (int)this.field_70161_v), 0);
     }
 
     protected int func_190735_dq()
@@ -263,48 +253,72 @@
         return i;
     }
 
-    @Override
     protected float func_70647_i()
     {
-        return this.func_70631_g_()
-               ? (this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 2.0F
-               : (this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 1.0F;
+        return this.func_70631_g_() ? (this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 2.0F : (this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 1.0F;
     }
 
-    @Override
     public SoundEvent func_184639_G()
     {
         return SoundEvents.field_187940_hn;
     }
 
-    @Override
     public SoundEvent func_184601_bQ(DamageSource p_184601_1_)
     {
         return SoundEvents.field_187944_hr;
     }
 
-    @Override
     public SoundEvent func_184615_bR()
     {
         return SoundEvents.field_187943_hq;
     }
 
-    @Override
     public SoundEvent func_190731_di()
     {
         return SoundEvents.field_187946_ht;
     }
 
     @Nullable
-    @Override
     protected ResourceLocation func_184647_J()
     {
         return LootTableList.field_191183_as;
     }
 
-    @Override
     protected ItemStack func_190732_dj()
     {
         return ItemStack.field_190927_a;
     }
+
+    /* ======================================== FORGE START =====================================*/
+
+    @Nullable
+    private net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession prof;
+    public void setForgeProfession(net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession prof)
+    {
+        this.prof = prof;
+        this.func_190733_a(net.minecraftforge.fml.common.registry.VillagerRegistry.getId(prof));
+    }
+
+    public net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession getForgeProfession()
+    {
+        if (this.prof == null)
+        {
+            this.prof = net.minecraftforge.fml.common.registry.VillagerRegistry.getById(this.func_190736_dl());
+            if (this.prof == null)
+                return net.minecraftforge.fml.common.registry.VillagerRegistry.FARMER;
+        }
+        return this.prof;
+    }
+
+    @Override
+    public void func_184206_a(DataParameter<?> key)
+    {
+        super.func_184206_a(key);
+        if (key.equals(field_190739_c))
+        {
+            net.minecraftforge.fml.common.registry.VillagerRegistry.onSetProfession(this, this.field_70180_af.func_187225_a(field_190739_c));
+        }
+    }
+
+    /* ======================================== FORGE END =====================================*/
 }
