--- before/net/minecraft/entity/monster/EntityVex.java
+++ after/net/minecraft/entity/monster/EntityVex.java
@@ -38,7 +38,7 @@
 
 public class EntityVex extends EntityMob
 {
-    protected static final DataParameter<Byte> field_190664_a = EntityDataManager.func_187226_a(EntityVex.class, DataSerializers.field_187191_a);
+    protected static final DataParameter<Byte> field_190664_a = EntityDataManager.<Byte>func_187226_a(EntityVex.class, DataSerializers.field_187191_a);
     private EntityLiving field_190665_b;
     @Nullable
     private BlockPos field_190666_c;
@@ -82,9 +82,9 @@
         this.field_70714_bg.func_75776_a(8, new EntityVex.AIMoveRandom());
         this.field_70714_bg.func_75776_a(9, new EntityAIWatchClosest(this, EntityPlayer.class, 3.0F, 1.0F));
         this.field_70714_bg.func_75776_a(10, new EntityAIWatchClosest(this, EntityLiving.class, 8.0F));
-        this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, true, EntityVex.class));
+        this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, true, new Class[] {EntityVex.class}));
         this.field_70715_bh.func_75776_a(2, new EntityVex.AICopyOwnerTarget(this));
-        this.field_70715_bh.func_75776_a(3, new EntityAINearestAttackableTarget<>(this, EntityPlayer.class, true));
+        this.field_70715_bh.func_75776_a(3, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
     }
 
     protected void func_110147_ax()
@@ -97,7 +97,7 @@
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_190664_a, (byte)0);
+        this.field_70180_af.func_187214_a(field_190664_a, Byte.valueOf((byte)0));
     }
 
     public static void func_190663_b(DataFixer p_190663_0_)
@@ -155,24 +155,24 @@
 
     private boolean func_190656_b(int p_190656_1_)
     {
-        int i = this.field_70180_af.func_187225_a(field_190664_a);
+        int i = ((Byte)this.field_70180_af.func_187225_a(field_190664_a)).byteValue();
         return (i & p_190656_1_) != 0;
     }
 
     private void func_190660_a(int p_190660_1_, boolean p_190660_2_)
     {
-        int i = this.field_70180_af.func_187225_a(field_190664_a);
+        int i = ((Byte)this.field_70180_af.func_187225_a(field_190664_a)).byteValue();
 
         if (p_190660_2_)
         {
-            i |= p_190660_1_;
+            i = i | p_190660_1_;
         }
         else
         {
-            i &= ~p_190660_1_;
+            i = i & ~p_190660_1_;
         }
 
-        this.field_70180_af.func_187227_b(field_190664_a, (byte)(i & 255));
+        this.field_70180_af.func_187227_b(field_190664_a, Byte.valueOf((byte)(i & 255)));
     }
 
     public boolean func_190647_dj()
