--- before/net/minecraft/entity/monster/EntityMagmaCube.java
+++ after/net/minecraft/entity/monster/EntityMagmaCube.java
@@ -28,28 +28,22 @@
         EntityLiving.func_189752_a(p_189759_0_, EntityMagmaCube.class);
     }
 
-    @Override
     protected void func_110147_ax()
     {
         super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.2F);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.20000000298023224D);
     }
 
-    @Override
     public boolean func_70601_bi()
     {
         return this.field_70170_p.func_175659_aa() != EnumDifficulty.PEACEFUL;
     }
 
-    @Override
     public boolean func_70058_J()
     {
-        return this.field_70170_p.func_72917_a(this.func_174813_aQ(), this)
-               && this.field_70170_p.func_184144_a(this, this.func_174813_aQ()).isEmpty()
-               && !this.field_70170_p.func_72953_d(this.func_174813_aQ());
+        return this.field_70170_p.func_72917_a(this.func_174813_aQ(), this) && this.field_70170_p.func_184144_a(this, this.func_174813_aQ()).isEmpty() && !this.field_70170_p.func_72953_d(this.func_174813_aQ());
     }
 
-    @Override
     protected void func_70799_a(int p_70799_1_, boolean p_70799_2_)
     {
         super.func_70799_a(p_70799_1_, p_70799_2_);
@@ -57,105 +51,89 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public int func_70070_b()
     {
         return 15728880;
     }
 
-    @Override
     public float func_70013_c()
     {
         return 1.0F;
     }
 
-    @Override
     protected EnumParticleTypes func_180487_n()
     {
         return EnumParticleTypes.FLAME;
     }
 
-    @Override
     protected EntitySlime func_70802_j()
     {
         return new EntityMagmaCube(this.field_70170_p);
     }
 
     @Nullable
-    @Override
     protected ResourceLocation func_184647_J()
     {
         return this.func_189101_db() ? LootTableList.field_186419_a : LootTableList.field_186379_ad;
     }
 
-    @Override
     public boolean func_70027_ad()
     {
         return false;
     }
 
-    @Override
     protected int func_70806_k()
     {
         return super.func_70806_k() * 4;
     }
 
-    @Override
     protected void func_70808_l()
     {
         this.field_70813_a *= 0.9F;
     }
 
-    @Override
     protected void func_70664_aZ()
     {
         this.field_70181_x = (double)(0.42F + (float)this.func_70809_q() * 0.1F);
         this.field_70160_al = true;
+        net.minecraftforge.common.ForgeHooks.onLivingJump(this);
     }
 
-    @Override
     protected void func_180466_bG()
     {
         this.field_70181_x = (double)(0.22F + (float)this.func_70809_q() * 0.05F);
         this.field_70160_al = true;
     }
 
-    @Override
     public void func_180430_e(float p_180430_1_, float p_180430_2_)
     {
     }
 
-    @Override
     protected boolean func_70800_m()
     {
         return true;
     }
 
-    @Override
     protected int func_70805_n()
     {
         return super.func_70805_n() + 2;
     }
 
-    @Override
     protected SoundEvent func_184601_bQ(DamageSource p_184601_1_)
     {
         return this.func_189101_db() ? SoundEvents.field_187892_fv : SoundEvents.field_187760_dh;
     }
 
-    @Override
     protected SoundEvent func_184615_bR()
     {
         return this.func_189101_db() ? SoundEvents.field_187890_fu : SoundEvents.field_187758_dg;
     }
 
-    @Override
     protected SoundEvent func_184709_cY()
     {
         return this.func_189101_db() ? SoundEvents.field_187894_fw : SoundEvents.field_187764_dj;
     }
 
-    @Override
     protected SoundEvent func_184710_cZ()
     {
         return SoundEvents.field_187762_di;
