--- before/net/minecraft/entity/monster/EntityShulker.java
+++ after/net/minecraft/entity/monster/EntityShulker.java
@@ -51,13 +51,11 @@
 public class EntityShulker extends EntityGolem implements IMob
 {
     private static final UUID field_184703_bv = UUID.fromString("7E0292F2-9434-48D5-A29F-9583AF7DF27F");
-    private static final AttributeModifier field_184704_bw = new AttributeModifier(field_184703_bv, "Covered armor bonus", 20.0, 0).func_111168_a(false);
-    protected static final DataParameter<EnumFacing> field_184700_a = EntityDataManager.func_187226_a(EntityShulker.class, DataSerializers.field_187202_l);
-    protected static final DataParameter<Optional<BlockPos>> field_184701_b = EntityDataManager.func_187226_a(
-                    EntityShulker.class, DataSerializers.field_187201_k
-                );
-    protected static final DataParameter<Byte> field_184702_c = EntityDataManager.func_187226_a(EntityShulker.class, DataSerializers.field_187191_a);
-    protected static final DataParameter<Byte> field_190770_bw = EntityDataManager.func_187226_a(EntityShulker.class, DataSerializers.field_187191_a);
+    private static final AttributeModifier field_184704_bw = (new AttributeModifier(field_184703_bv, "Covered armor bonus", 20.0D, 0)).func_111168_a(false);
+    protected static final DataParameter<EnumFacing> field_184700_a = EntityDataManager.<EnumFacing>func_187226_a(EntityShulker.class, DataSerializers.field_187202_l);
+    protected static final DataParameter<Optional<BlockPos>> field_184701_b = EntityDataManager.<Optional<BlockPos>>func_187226_a(EntityShulker.class, DataSerializers.field_187201_k);
+    protected static final DataParameter<Byte> field_184702_c = EntityDataManager.<Byte>func_187226_a(EntityShulker.class, DataSerializers.field_187191_a);
+    protected static final DataParameter<Byte> field_190770_bw = EntityDataManager.<Byte>func_187226_a(EntityShulker.class, DataSerializers.field_187191_a);
     public static final EnumDyeColor field_190771_bx = EnumDyeColor.PURPLE;
     private float field_184705_bx;
     private float field_184706_by;
@@ -76,7 +74,6 @@
     }
 
     @Nullable
-    @Override
     public IEntityLivingData func_180482_a(DifficultyInstance p_180482_1_, @Nullable IEntityLivingData p_180482_2_)
     {
         this.field_70761_aq = 180.0F;
@@ -88,37 +85,32 @@
         return super.func_180482_a(p_180482_1_, p_180482_2_);
     }
 
-    @Override
     protected void func_184651_r()
     {
         this.field_70714_bg.func_75776_a(1, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
         this.field_70714_bg.func_75776_a(4, new EntityShulker.AIAttack());
         this.field_70714_bg.func_75776_a(7, new EntityShulker.AIPeek());
         this.field_70714_bg.func_75776_a(8, new EntityAILookIdle(this));
-        this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, true));
+        this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, true, new Class[0]));
         this.field_70715_bh.func_75776_a(2, new EntityShulker.AIAttackNearest(this));
         this.field_70715_bh.func_75776_a(3, new EntityShulker.AIDefenseAttack(this));
     }
 
-    @Override
     protected boolean func_70041_e_()
     {
         return false;
     }
 
-    @Override
     public SoundCategory func_184176_by()
     {
         return SoundCategory.HOSTILE;
     }
 
-    @Override
     protected SoundEvent func_184639_G()
     {
         return SoundEvents.field_187773_eO;
     }
 
-    @Override
     public void func_70642_aH()
     {
         if (!this.func_184686_df())
@@ -127,36 +119,31 @@
         }
     }
 
-    @Override
     protected SoundEvent func_184615_bR()
     {
         return SoundEvents.field_187781_eS;
     }
 
-    @Override
     protected SoundEvent func_184601_bQ(DamageSource p_184601_1_)
     {
         return this.func_184686_df() ? SoundEvents.field_187785_eU : SoundEvents.field_187783_eT;
     }
 
-    @Override
     protected void func_70088_a()
     {
         super.func_70088_a();
         this.field_70180_af.func_187214_a(field_184700_a, EnumFacing.DOWN);
         this.field_70180_af.func_187214_a(field_184701_b, Optional.absent());
-        this.field_70180_af.func_187214_a(field_184702_c, (byte)0);
-        this.field_70180_af.func_187214_a(field_190770_bw, (byte)field_190771_bx.func_176765_a());
+        this.field_70180_af.func_187214_a(field_184702_c, Byte.valueOf((byte)0));
+        this.field_70180_af.func_187214_a(field_190770_bw, Byte.valueOf((byte)field_190771_bx.func_176765_a()));
     }
 
-    @Override
     protected void func_110147_ax()
     {
         super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(30.0);
+        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(30.0D);
     }
 
-    @Override
     protected EntityBodyHelper func_184650_s()
     {
         return new EntityShulker.BodyHelper(this);
@@ -167,13 +154,12 @@
         EntityLiving.func_189752_a(p_189757_0_, EntityShulker.class);
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
         this.field_70180_af.func_187227_b(field_184700_a, EnumFacing.func_82600_a(p_70037_1_.func_74771_c("AttachFace")));
-        this.field_70180_af.func_187227_b(field_184702_c, p_70037_1_.func_74771_c("Peek"));
-        this.field_70180_af.func_187227_b(field_190770_bw, p_70037_1_.func_74771_c("Color"));
+        this.field_70180_af.func_187227_b(field_184702_c, Byte.valueOf(p_70037_1_.func_74771_c("Peek")));
+        this.field_70180_af.func_187227_b(field_190770_bw, Byte.valueOf(p_70037_1_.func_74771_c("Color")));
 
         if (p_70037_1_.func_74764_b("APX"))
         {
@@ -188,13 +174,12 @@
         }
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
-        p_70014_1_.func_74774_a("AttachFace", (byte)this.field_70180_af.func_187225_a(field_184700_a).func_176745_a());
-        p_70014_1_.func_74774_a("Peek", this.field_70180_af.func_187225_a(field_184702_c));
-        p_70014_1_.func_74774_a("Color", this.field_70180_af.func_187225_a(field_190770_bw));
+        p_70014_1_.func_74774_a("AttachFace", (byte)((EnumFacing)this.field_70180_af.func_187225_a(field_184700_a)).func_176745_a());
+        p_70014_1_.func_74774_a("Peek", ((Byte)this.field_70180_af.func_187225_a(field_184702_c)).byteValue());
+        p_70014_1_.func_74774_a("Color", ((Byte)this.field_70180_af.func_187225_a(field_190770_bw)).byteValue());
         BlockPos blockpos = this.func_184699_da();
 
         if (blockpos != null)
@@ -205,11 +190,10 @@
         }
     }
 
-    @Override
     public void func_70071_h_()
     {
         super.func_70071_h_();
-        BlockPos blockpos = this.field_70180_af.func_187225_a(field_184701_b).orNull();
+        BlockPos blockpos = (BlockPos)((Optional)this.field_70180_af.func_187225_a(field_184701_b)).orNull();
 
         if (blockpos == null && !this.field_70170_p.field_72995_K)
         {
@@ -234,7 +218,7 @@
             {
                 if (iblockstate.func_177230_c() == Blocks.field_180384_M)
                 {
-                    EnumFacing enumfacing = iblockstate.func_177229_b(BlockPistonBase.field_176387_N);
+                    EnumFacing enumfacing = (EnumFacing)iblockstate.func_177229_b(BlockPistonBase.field_176387_N);
 
                     if (this.field_70170_p.func_175623_d(blockpos.func_177972_a(enumfacing)))
                     {
@@ -248,7 +232,7 @@
                 }
                 else if (iblockstate.func_177230_c() == Blocks.field_150332_K)
                 {
-                    EnumFacing enumfacing3 = iblockstate.func_177229_b(BlockPistonExtension.field_176387_N);
+                    EnumFacing enumfacing3 = (EnumFacing)iblockstate.func_177229_b(BlockPistonExtension.field_176387_N);
 
                     if (this.field_70170_p.func_175623_d(blockpos.func_177972_a(enumfacing3)))
                     {
@@ -324,105 +308,52 @@
                 }
             }
 
-            this.field_70165_t = (double)blockpos.func_177958_n() + 0.5;
+            this.field_70165_t = (double)blockpos.func_177958_n() + 0.5D;
             this.field_70163_u = (double)blockpos.func_177956_o();
-            this.field_70161_v = (double)blockpos.func_177952_p() + 0.5;
+            this.field_70161_v = (double)blockpos.func_177952_p() + 0.5D;
+            if (this.isAddedToWorld() && !this.field_70170_p.field_72995_K) this.field_70170_p.func_72866_a(this, false); // Forge - Process chunk registration after moving.
             this.field_70169_q = this.field_70165_t;
             this.field_70167_r = this.field_70163_u;
             this.field_70166_s = this.field_70161_v;
             this.field_70142_S = this.field_70165_t;
             this.field_70137_T = this.field_70163_u;
             this.field_70136_U = this.field_70161_v;
-            double d3 = 0.5 - (double)MathHelper.func_76126_a((0.5F + this.field_184706_by) * (float) Math.PI) * 0.5;
-            double d4 = 0.5 - (double)MathHelper.func_76126_a((0.5F + this.field_184705_bx) * (float) Math.PI) * 0.5;
+            double d3 = 0.5D - (double)MathHelper.func_76126_a((0.5F + this.field_184706_by) * (float)Math.PI) * 0.5D;
+            double d4 = 0.5D - (double)MathHelper.func_76126_a((0.5F + this.field_184705_bx) * (float)Math.PI) * 0.5D;
             double d5 = d3 - d4;
-            double d0 = 0.0;
-            double d1 = 0.0;
-            double d2 = 0.0;
+            double d0 = 0.0D;
+            double d1 = 0.0D;
+            double d2 = 0.0D;
             EnumFacing enumfacing2 = this.func_184696_cZ();
 
             switch (enumfacing2)
             {
                 case DOWN:
-                    this.func_174826_a(
-                        new AxisAlignedBB(
-                            this.field_70165_t - 0.5,
-                            this.field_70163_u,
-                            this.field_70161_v - 0.5,
-                            this.field_70165_t + 0.5,
-                            this.field_70163_u + 1.0 + d3,
-                            this.field_70161_v + 0.5
-                        )
-                    );
+                    this.func_174826_a(new AxisAlignedBB(this.field_70165_t - 0.5D, this.field_70163_u, this.field_70161_v - 0.5D, this.field_70165_t + 0.5D, this.field_70163_u + 1.0D + d3, this.field_70161_v + 0.5D));
                     d1 = d5;
                     break;
                 case UP:
-                    this.func_174826_a(
-                        new AxisAlignedBB(
-                            this.field_70165_t - 0.5,
-                            this.field_70163_u - d3,
-                            this.field_70161_v - 0.5,
-                            this.field_70165_t + 0.5,
-                            this.field_70163_u + 1.0,
-                            this.field_70161_v + 0.5
-                        )
-                    );
+                    this.func_174826_a(new AxisAlignedBB(this.field_70165_t - 0.5D, this.field_70163_u - d3, this.field_70161_v - 0.5D, this.field_70165_t + 0.5D, this.field_70163_u + 1.0D, this.field_70161_v + 0.5D));
                     d1 = -d5;
                     break;
                 case NORTH:
-                    this.func_174826_a(
-                        new AxisAlignedBB(
-                            this.field_70165_t - 0.5,
-                            this.field_70163_u,
-                            this.field_70161_v - 0.5,
-                            this.field_70165_t + 0.5,
-                            this.field_70163_u + 1.0,
-                            this.field_70161_v + 0.5 + d3
-                        )
-                    );
+                    this.func_174826_a(new AxisAlignedBB(this.field_70165_t - 0.5D, this.field_70163_u, this.field_70161_v - 0.5D, this.field_70165_t + 0.5D, this.field_70163_u + 1.0D, this.field_70161_v + 0.5D + d3));
                     d2 = d5;
                     break;
                 case SOUTH:
-                    this.func_174826_a(
-                        new AxisAlignedBB(
-                            this.field_70165_t - 0.5,
-                            this.field_70163_u,
-                            this.field_70161_v - 0.5 - d3,
-                            this.field_70165_t + 0.5,
-                            this.field_70163_u + 1.0,
-                            this.field_70161_v + 0.5
-                        )
-                    );
+                    this.func_174826_a(new AxisAlignedBB(this.field_70165_t - 0.5D, this.field_70163_u, this.field_70161_v - 0.5D - d3, this.field_70165_t + 0.5D, this.field_70163_u + 1.0D, this.field_70161_v + 0.5D));
                     d2 = -d5;
                     break;
                 case WEST:
-                    this.func_174826_a(
-                        new AxisAlignedBB(
-                            this.field_70165_t - 0.5,
-                            this.field_70163_u,
-                            this.field_70161_v - 0.5,
-                            this.field_70165_t + 0.5 + d3,
-                            this.field_70163_u + 1.0,
-                            this.field_70161_v + 0.5
-                        )
-                    );
+                    this.func_174826_a(new AxisAlignedBB(this.field_70165_t - 0.5D, this.field_70163_u, this.field_70161_v - 0.5D, this.field_70165_t + 0.5D + d3, this.field_70163_u + 1.0D, this.field_70161_v + 0.5D));
                     d0 = d5;
                     break;
                 case EAST:
-                    this.func_174826_a(
-                        new AxisAlignedBB(
-                            this.field_70165_t - 0.5 - d3,
-                            this.field_70163_u,
-                            this.field_70161_v - 0.5,
-                            this.field_70165_t + 0.5,
-                            this.field_70163_u + 1.0,
-                            this.field_70161_v + 0.5
-                        )
-                    );
+                    this.func_174826_a(new AxisAlignedBB(this.field_70165_t - 0.5D - d3, this.field_70163_u, this.field_70161_v - 0.5D, this.field_70165_t + 0.5D, this.field_70163_u + 1.0D, this.field_70161_v + 0.5D));
                     d0 = -d5;
             }
 
-            if (d5 > 0.0)
+            if (d5 > 0.0D)
             {
                 List<Entity> list = this.field_70170_p.func_72839_b(this, this.func_174813_aQ());
 
@@ -440,7 +371,6 @@
         }
     }
 
-    @Override
     public void func_70091_d(MoverType p_70091_1_, double p_70091_2_, double p_70091_4_, double p_70091_6_)
     {
         if (p_70091_1_ == MoverType.SHULKER_BOX)
@@ -453,20 +383,19 @@
         }
     }
 
-    @Override
     public void func_70107_b(double p_70107_1_, double p_70107_3_, double p_70107_5_)
     {
         super.func_70107_b(p_70107_1_, p_70107_3_, p_70107_5_);
 
         if (this.field_70180_af != null && this.field_70173_aa != 0)
         {
-            Optional<BlockPos> optional = this.field_70180_af.func_187225_a(field_184701_b);
-            Optional<BlockPos> optional1 = Optional.of(new BlockPos(p_70107_1_, p_70107_3_, p_70107_5_));
+            Optional<BlockPos> optional = (Optional)this.field_70180_af.func_187225_a(field_184701_b);
+            Optional<BlockPos> optional1 = Optional.<BlockPos>of(new BlockPos(p_70107_1_, p_70107_3_, p_70107_5_));
 
             if (!optional1.equals(optional))
             {
                 this.field_70180_af.func_187227_b(field_184701_b, optional1);
-                this.field_70180_af.func_187227_b(field_184702_c, (byte)0);
+                this.field_70180_af.func_187227_b(field_184702_c, Byte.valueOf((byte)0));
                 this.field_70160_al = true;
             }
         }
@@ -480,14 +409,9 @@
 
             for (int i = 0; i < 5; ++i)
             {
-                BlockPos blockpos1 = blockpos.func_177982_a(
-                                         8 - this.field_70146_Z.nextInt(17), 8 - this.field_70146_Z.nextInt(17), 8 - this.field_70146_Z.nextInt(17)
-                                     );
+                BlockPos blockpos1 = blockpos.func_177982_a(8 - this.field_70146_Z.nextInt(17), 8 - this.field_70146_Z.nextInt(17), 8 - this.field_70146_Z.nextInt(17));
 
-                if (blockpos1.func_177956_o() > 0
-                        && this.field_70170_p.func_175623_d(blockpos1)
-                        && this.field_70170_p.func_191503_g(this)
-                        && this.field_70170_p.func_184144_a(this, new AxisAlignedBB(blockpos1)).isEmpty())
+                if (blockpos1.func_177956_o() > 0 && this.field_70170_p.func_175623_d(blockpos1) && this.field_70170_p.func_191503_g(this) && this.field_70170_p.func_184144_a(this, new AxisAlignedBB(blockpos1)).isEmpty())
                 {
                     boolean flag = false;
 
@@ -503,10 +427,17 @@
 
                     if (flag)
                     {
+                        net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(this, blockpos1.func_177958_n(), blockpos1.func_177956_o(), blockpos1.func_177952_p(), 0);
+                        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) flag = false;
+                        blockpos1 = new BlockPos(event.getTargetX(), event.getTargetY(), event.getTargetZ());
+                    }
+
+                    if (flag)
+                    {
                         this.func_184185_a(SoundEvents.field_187791_eX, 1.0F, 1.0F);
                         this.field_70180_af.func_187227_b(field_184701_b, Optional.of(blockpos1));
-                        this.field_70180_af.func_187227_b(field_184702_c, (byte)0);
-                        this.func_70624_b(null);
+                        this.field_70180_af.func_187227_b(field_184702_c, Byte.valueOf((byte)0));
+                        this.func_70624_b((EntityLivingBase)null);
                         return true;
                     }
                 }
@@ -520,19 +451,17 @@
         }
     }
 
-    @Override
     public void func_70636_d()
     {
         super.func_70636_d();
-        this.field_70159_w = 0.0;
-        this.field_70181_x = 0.0;
-        this.field_70179_y = 0.0;
+        this.field_70159_w = 0.0D;
+        this.field_70181_x = 0.0D;
+        this.field_70179_y = 0.0D;
         this.field_70760_ar = 180.0F;
         this.field_70761_aq = 180.0F;
         this.field_70177_z = 180.0F;
     }
 
-    @Override
     public void func_184206_a(DataParameter<?> p_184206_1_)
     {
         if (field_184701_b.equals(p_184206_1_) && this.field_70170_p.field_72995_K && !this.func_184218_aH())
@@ -550,9 +479,9 @@
                     this.field_184708_bA = 6;
                 }
 
-                this.field_70165_t = (double)blockpos.func_177958_n() + 0.5;
+                this.field_70165_t = (double)blockpos.func_177958_n() + 0.5D;
                 this.field_70163_u = (double)blockpos.func_177956_o();
-                this.field_70161_v = (double)blockpos.func_177952_p() + 0.5;
+                this.field_70161_v = (double)blockpos.func_177952_p() + 0.5D;
                 this.field_70169_q = this.field_70165_t;
                 this.field_70167_r = this.field_70163_u;
                 this.field_70166_s = this.field_70161_v;
@@ -566,15 +495,11 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
-    public void func_180426_a(
-        double p_180426_1_, double p_180426_3_, double p_180426_5_, float p_180426_7_, float p_180426_8_, int p_180426_9_, boolean p_180426_10_
-    )
+    public void func_180426_a(double p_180426_1_, double p_180426_3_, double p_180426_5_, float p_180426_7_, float p_180426_8_, int p_180426_9_, boolean p_180426_10_)
     {
         this.field_70716_bi = 0;
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         if (this.func_184686_df())
@@ -589,7 +514,7 @@
 
         if (super.func_70097_a(p_70097_1_, p_70097_2_))
         {
-            if ((double)this.func_110143_aJ() < (double)this.func_110138_aP() * 0.5 && this.field_70146_Z.nextInt(4) == 0)
+            if ((double)this.func_110143_aJ() < (double)this.func_110138_aP() * 0.5D && this.field_70146_Z.nextInt(4) == 0)
             {
                 this.func_184689_o();
             }
@@ -608,7 +533,6 @@
     }
 
     @Nullable
-    @Override
     public AxisAlignedBB func_70046_E()
     {
         return this.func_70089_S() ? this.func_174813_aQ() : null;
@@ -616,13 +540,13 @@
 
     public EnumFacing func_184696_cZ()
     {
-        return this.field_70180_af.func_187225_a(field_184700_a);
+        return (EnumFacing)this.field_70180_af.func_187225_a(field_184700_a);
     }
 
     @Nullable
     public BlockPos func_184699_da()
     {
-        return this.field_70180_af.func_187225_a(field_184701_b).orNull();
+        return (BlockPos)((Optional)this.field_70180_af.func_187225_a(field_184701_b)).orNull();
     }
 
     public void func_184694_g(@Nullable BlockPos p_184694_1_)
@@ -632,7 +556,7 @@
 
     public int func_184684_db()
     {
-        return this.field_70180_af.func_187225_a(field_184702_c);
+        return ((Byte)this.field_70180_af.func_187225_a(field_184702_c)).byteValue();
     }
 
     public void func_184691_a(int p_184691_1_)
@@ -652,7 +576,7 @@
             }
         }
 
-        this.field_70180_af.func_187227_b(field_184702_c, (byte)p_184691_1_);
+        this.field_70180_af.func_187227_b(field_184702_c, Byte.valueOf((byte)p_184691_1_));
     }
 
     @SideOnly(Side.CLIENT)
@@ -673,30 +597,25 @@
         return this.field_184707_bz;
     }
 
-    @Override
     public float func_70047_e()
     {
         return 0.5F;
     }
 
-    @Override
     public int func_70646_bf()
     {
         return 180;
     }
 
-    @Override
     public int func_184649_cE()
     {
         return 180;
     }
 
-    @Override
     public void func_70108_f(Entity p_70108_1_)
     {
     }
 
-    @Override
     public float func_70111_Y()
     {
         return 0.0F;
@@ -709,7 +628,6 @@
     }
 
     @Nullable
-    @Override
     protected ResourceLocation func_184647_J()
     {
         return LootTableList.field_186442_x;
@@ -718,7 +636,7 @@
     @SideOnly(Side.CLIENT)
     public EnumDyeColor func_190769_dn()
     {
-        return EnumDyeColor.func_176764_b(this.field_70180_af.func_187225_a(field_190770_bw));
+        return EnumDyeColor.func_176764_b(((Byte)this.field_70180_af.func_187225_a(field_190770_bw)).byteValue());
     }
 
     class AIAttack extends EntityAIBase
@@ -730,7 +648,6 @@
             this.func_75248_a(3);
         }
 
-        @Override
         public boolean func_75250_a()
         {
             EntityLivingBase entitylivingbase = EntityShulker.this.func_70638_az();
@@ -745,20 +662,17 @@
             }
         }
 
-        @Override
         public void func_75249_e()
         {
             this.field_188520_b = 20;
             EntityShulker.this.func_184691_a(100);
         }
 
-        @Override
         public void func_75251_c()
         {
             EntityShulker.this.func_184691_a(0);
         }
 
-        @Override
         public void func_75246_d()
         {
             if (EntityShulker.this.field_70170_p.func_175659_aa() != EnumDifficulty.PEACEFUL)
@@ -768,25 +682,19 @@
                 EntityShulker.this.func_70671_ap().func_75651_a(entitylivingbase, 180.0F, 180.0F);
                 double d0 = EntityShulker.this.func_70068_e(entitylivingbase);
 
-                if (d0 < 400.0)
+                if (d0 < 400.0D)
                 {
                     if (this.field_188520_b <= 0)
                     {
                         this.field_188520_b = 20 + EntityShulker.this.field_70146_Z.nextInt(10) * 20 / 2;
-                        EntityShulkerBullet entityshulkerbullet = new EntityShulkerBullet(
-                            EntityShulker.this.field_70170_p, EntityShulker.this, entitylivingbase, EntityShulker.this.func_184696_cZ().func_176740_k()
-                        );
+                        EntityShulkerBullet entityshulkerbullet = new EntityShulkerBullet(EntityShulker.this.field_70170_p, EntityShulker.this, entitylivingbase, EntityShulker.this.func_184696_cZ().func_176740_k());
                         EntityShulker.this.field_70170_p.func_72838_d(entityshulkerbullet);
-                        EntityShulker.this.func_184185_a(
-                            SoundEvents.field_187789_eW,
-                            2.0F,
-                            (EntityShulker.this.field_70146_Z.nextFloat() - EntityShulker.this.field_70146_Z.nextFloat()) * 0.2F + 1.0F
-                        );
+                        EntityShulker.this.func_184185_a(SoundEvents.field_187789_eW, 2.0F, (EntityShulker.this.field_70146_Z.nextFloat() - EntityShulker.this.field_70146_Z.nextFloat()) * 0.2F + 1.0F);
                     }
                 }
                 else
                 {
-                    EntityShulker.this.func_70624_b(null);
+                    EntityShulker.this.func_70624_b((EntityLivingBase)null);
                 }
 
                 super.func_75246_d();
@@ -801,26 +709,22 @@
             super(p_i47060_2_, EntityPlayer.class, true);
         }
 
-        @Override
         public boolean func_75250_a()
         {
             return EntityShulker.this.field_70170_p.func_175659_aa() == EnumDifficulty.PEACEFUL ? false : super.func_75250_a();
         }
 
-        @Override
         protected AxisAlignedBB func_188511_a(double p_188511_1_)
         {
             EnumFacing enumfacing = ((EntityShulker)this.field_75299_d).func_184696_cZ();
 
             if (enumfacing.func_176740_k() == EnumFacing.Axis.X)
             {
-                return this.field_75299_d.func_174813_aQ().func_72314_b(4.0, p_188511_1_, p_188511_1_);
+                return this.field_75299_d.func_174813_aQ().func_72314_b(4.0D, p_188511_1_, p_188511_1_);
             }
             else
             {
-                return enumfacing.func_176740_k() == EnumFacing.Axis.Z
-                       ? this.field_75299_d.func_174813_aQ().func_72314_b(p_188511_1_, p_188511_1_, 4.0)
-                       : this.field_75299_d.func_174813_aQ().func_72314_b(p_188511_1_, 4.0, p_188511_1_);
+                return enumfacing.func_176740_k() == EnumFacing.Axis.Z ? this.field_75299_d.func_174813_aQ().func_72314_b(p_188511_1_, p_188511_1_, 4.0D) : this.field_75299_d.func_174813_aQ().func_72314_b(p_188511_1_, 4.0D, p_188511_1_);
             }
         }
     }
@@ -838,26 +742,22 @@
                 });
             }
 
-            @Override
             public boolean func_75250_a()
             {
                 return this.field_75299_d.func_96124_cp() == null ? false : super.func_75250_a();
             }
 
-            @Override
             protected AxisAlignedBB func_188511_a(double p_188511_1_)
             {
                 EnumFacing enumfacing = ((EntityShulker)this.field_75299_d).func_184696_cZ();
 
                 if (enumfacing.func_176740_k() == EnumFacing.Axis.X)
                 {
-                    return this.field_75299_d.func_174813_aQ().func_72314_b(4.0, p_188511_1_, p_188511_1_);
+                    return this.field_75299_d.func_174813_aQ().func_72314_b(4.0D, p_188511_1_, p_188511_1_);
                 }
                 else
                 {
-                    return enumfacing.func_176740_k() == EnumFacing.Axis.Z
-                           ? this.field_75299_d.func_174813_aQ().func_72314_b(p_188511_1_, p_188511_1_, 4.0)
-                           : this.field_75299_d.func_174813_aQ().func_72314_b(p_188511_1_, 4.0, p_188511_1_);
+                    return enumfacing.func_176740_k() == EnumFacing.Axis.Z ? this.field_75299_d.func_174813_aQ().func_72314_b(p_188511_1_, p_188511_1_, 4.0D) : this.field_75299_d.func_174813_aQ().func_72314_b(p_188511_1_, 4.0D, p_188511_1_);
                 }
             }
         }
@@ -870,26 +770,22 @@
         {
         }
 
-        @Override
         public boolean func_75250_a()
         {
             return EntityShulker.this.func_70638_az() == null && EntityShulker.this.field_70146_Z.nextInt(40) == 0;
         }
 
-        @Override
         public boolean func_75253_b()
         {
             return EntityShulker.this.func_70638_az() == null && this.field_188522_b > 0;
         }
 
-        @Override
         public void func_75249_e()
         {
             this.field_188522_b = 20 * (1 + EntityShulker.this.field_70146_Z.nextInt(3));
             EntityShulker.this.func_184691_a(30);
         }
 
-        @Override
         public void func_75251_c()
         {
             if (EntityShulker.this.func_70638_az() == null)
@@ -898,7 +794,6 @@
             }
         }
 
-        @Override
         public void func_75246_d()
         {
             --this.field_188522_b;
@@ -912,7 +807,6 @@
             super(p_i47062_2_);
         }
 
-        @Override
         public void func_75664_a()
         {
         }
