--- before/net/minecraft/entity/monster/EntityEvoker.java
+++ after/net/minecraft/entity/monster/EntityEvoker.java
@@ -409,7 +409,7 @@
             {
                 return false;
             }
-            else if (!EntityEvoker.this.world.getGameRules().getBoolean("mobGriefing"))
+            else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(EntityEvoker.this.world, EntityEvoker.this))
             {
                 return false;
             }
