--- before/net/minecraft/entity/monster/EntitySpider.java
+++ after/net/minecraft/entity/monster/EntitySpider.java
@@ -154,7 +154,13 @@
     @Override
     public boolean isPotionApplicable(PotionEffect potioneffectIn)
     {
-        return potioneffectIn.getPotion() == MobEffects.POISON ? false : super.isPotionApplicable(potioneffectIn);
+        if(potioneffectIn.getPotion() == MobEffects.POISON)
+        {
+            net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent event = new net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent(this, potioneffectIn);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+            return event.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW;
+        }
+        return super.isPotionApplicable(potioneffectIn);
     }
 
     public boolean isBesideClimbableBlock()
