--- before/net/minecraft/entity/monster/EntityZombieVillager.java
+++ after/net/minecraft/entity/monster/EntityZombieVillager.java
@@ -54,11 +54,14 @@
     public void setProfession(int profession)
     {
         this.dataManager.set(PROFESSION, profession);
+        net.minecraftforge.fml.common.registry.VillagerRegistry.onSetProfession(this, profession);
     }
 
+    //Use Forge Variant below
+    @Deprecated
     public int getProfession()
     {
-        return Math.max(this.dataManager.get(PROFESSION) % 6, 0);
+        return Math.max(this.dataManager.get(PROFESSION), 0);
     }
 
     public static void registerFixesZombieVillager(DataFixer fixer)
@@ -71,6 +74,7 @@
     {
         super.writeEntityToNBT(compound);
         compound.setInteger("Profession", this.getProfession());
+        compound.setString("ProfessionName", this.getForgeProfession().getRegistryName().toString());
         compound.setInteger("ConversionTime", this.isConverting() ? this.conversionTime : -1);
 
         if (this.converstionStarter != null)
@@ -84,6 +88,12 @@
     {
         super.readEntityFromNBT(compound);
         this.setProfession(compound.getInteger("Profession"));
+        if (compound.hasKey("ProfessionName"))
+        {
+            net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession p = net.minecraftforge.fml.common.registry.ForgeRegistries.VILLAGER_PROFESSIONS.getValue(new net.minecraft.util.ResourceLocation(compound.getString("ProfessionName")));
+            if (p == null) p = net.minecraftforge.fml.common.registry.VillagerRegistry.FARMER;
+            this.setForgeProfession(p);
+        }
 
         if (compound.hasKey("ConversionTime", 99) && compound.getInteger("ConversionTime") > -1)
         {
@@ -97,7 +107,7 @@
     @Override
     public IEntityLivingData onInitialSpawn(DifficultyInstance difficulty, @Nullable IEntityLivingData livingdata)
     {
-        this.setProfession(this.world.rand.nextInt(6));
+        net.minecraftforge.fml.common.registry.VillagerRegistry.setRandomProfession(this, this.world.rand);
         return super.onInitialSpawn(difficulty, livingdata);
     }
 
@@ -195,7 +205,7 @@
     {
         EntityVillager entityvillager = new EntityVillager(this.world);
         entityvillager.copyLocationAndAnglesFrom(this);
-        entityvillager.setProfession(this.getProfession());
+        entityvillager.setProfession(this.getForgeProfession());
         entityvillager.finalizeMobSpawn(this.world.getDifficultyForLocation(new BlockPos(entityvillager)), null, false);
         entityvillager.setLookingForHome();
 
@@ -307,4 +317,37 @@
     {
         return ItemStack.EMPTY;
     }
+
+    /* ======================================== FORGE START =====================================*/
+
+    @Nullable
+    private net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession prof;
+    public void setForgeProfession(net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession prof)
+    {
+        this.prof = prof;
+        this.setProfession(net.minecraftforge.fml.common.registry.VillagerRegistry.getId(prof));
+    }
+
+    public net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession getForgeProfession()
+    {
+        if (this.prof == null)
+        {
+            this.prof = net.minecraftforge.fml.common.registry.VillagerRegistry.getById(this.getProfession());
+            if (this.prof == null)
+                return net.minecraftforge.fml.common.registry.VillagerRegistry.FARMER;
+        }
+        return this.prof;
+    }
+
+    @Override
+    public void notifyDataManagerChange(DataParameter<?> key)
+    {
+        super.notifyDataManagerChange(key);
+        if (key.equals(PROFESSION))
+        {
+            net.minecraftforge.fml.common.registry.VillagerRegistry.onSetProfession(this, this.dataManager.get(PROFESSION));
+        }
+    }
+
+    /* ======================================== FORGE END =====================================*/
 }
