--- before/net/minecraft/entity/monster/EntityPigZombie.java
+++ after/net/minecraft/entity/monster/EntityPigZombie.java
@@ -61,7 +61,7 @@
     {
         super.func_110147_ax();
         this.func_110148_a(field_110186_bp).func_111128_a(0.0D);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a((double)0.23F);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.23000000417232513D);
         this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111128_a(5.0D);
     }
 
