--- before/net/minecraft/entity/monster/EntityElderGuardian.java
+++ after/net/minecraft/entity/monster/EntityElderGuardian.java
@@ -38,7 +38,7 @@
     protected void func_110147_ax()
     {
         super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a((double)0.3F);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.30000001192092896D);
         this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111128_a(8.0D);
         this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(80.0D);
     }
@@ -94,7 +94,7 @@
         if ((this.field_70173_aa + this.func_145782_y()) % 1200 == 0)
         {
             Potion potion = MobEffects.field_76419_f;
-            List<EntityPlayerMP> list = this.field_70170_p.func_175661_b(EntityPlayerMP.class, new Predicate<EntityPlayerMP>()
+            List<EntityPlayerMP> list = this.field_70170_p.<EntityPlayerMP>func_175661_b(EntityPlayerMP.class, new Predicate<EntityPlayerMP>()
             {
                 public boolean apply(@Nullable EntityPlayerMP p_apply_1_)
                 {
