--- before/net/minecraft/entity/monster/EntityIronGolem.java
+++ after/net/minecraft/entity/monster/EntityIronGolem.java
@@ -321,7 +321,7 @@
     {
         if (!this.isPlayerCreated() && this.attackingPlayer != null && this.village != null)
         {
-            this.village.modifyPlayerReputation(this.attackingPlayer.getName(), -5);
+            this.village.modifyPlayerReputation(this.attackingPlayer.getUniqueID(), -5);
         }
 
         super.onDeath(cause);
