--- before/net/minecraft/entity/monster/EntitySilverfish.java
+++ after/net/minecraft/entity/monster/EntitySilverfish.java
@@ -202,7 +202,7 @@
                 {
                     Random random = this.entity.getRNG();
 
-                    if (this.entity.world.getGameRules().getBoolean("mobGriefing") && random.nextInt(10) == 0)
+                    if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.entity.world, this.entity) && random.nextInt(10) == 0)
                     {
                         this.facing = EnumFacing.random(random);
                         BlockPos blockpos = new BlockPos(this.entity.posX, this.entity.posY + 0.5, this.entity.posZ)
@@ -301,7 +301,7 @@
 
                                 if (iblockstate.getBlock() == Blocks.MONSTER_EGG)
                                 {
-                                    if (world.getGameRules().getBoolean("mobGriefing"))
+                                    if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(world, this.silverfish))
                                     {
                                         world.destroyBlock(blockpos1, true);
                                     }
