--- before/net/minecraft/client/renderer/chunk/ChunkRenderWorker.java
+++ after/net/minecraft/client/renderer/chunk/ChunkRenderWorker.java
@@ -149,7 +149,7 @@
             }
 
             final CompiledChunk compiledchunk = p_178474_1_.func_178544_c();
-            ArrayList arraylist = Lists.newArrayList();
+            ArrayList<ListenableFuture<Object>> arraylist = Lists.newArrayList();
 
             if (chunkcompiletaskgenerator$type == ChunkCompileTaskGenerator.Type.REBUILD_CHUNK)
             {
@@ -213,7 +213,7 @@
                         Minecraft.func_71410_x().func_71404_a(CrashReport.func_85055_a(p_onFailure_1_, "Rendering chunk"));
                     }
                 }
-            });
+            }, Runnable::run);
         }
     }
 
