--- before/net/minecraft/client/renderer/chunk/ChunkRenderWorker.java
+++ after/net/minecraft/client/renderer/chunk/ChunkRenderWorker.java
@@ -239,7 +239,7 @@
                         Minecraft.getMinecraft().crashed(CrashReport.makeCrashReport(p_onFailure_1_, "Rendering chunk"));
                     }
                 }
-            }
+            }, Runnable::run
             );
         }
     }
