--- before/net/minecraft/client/renderer/chunk/SetVisibility.java
+++ after/net/minecraft/client/renderer/chunk/SetVisibility.java
@@ -10,7 +10,12 @@
 public class SetVisibility
 {
     private static final int field_178623_a = EnumFacing.values().length;
-    private final BitSet field_178622_b = new BitSet(field_178623_a * field_178623_a);
+    private final BitSet field_178622_b;
+
+    public SetVisibility()
+    {
+        this.field_178622_b = new BitSet(field_178623_a * field_178623_a);
+    }
 
     public void func_178620_a(Set<EnumFacing> p_178620_1_)
     {
