--- before/net/minecraft/client/renderer/block/statemap/BlockStateMapper.java
+++ after/net/minecraft/client/renderer/block/statemap/BlockStateMapper.java
@@ -16,8 +16,8 @@
 @SideOnly(Side.CLIENT)
 public class BlockStateMapper
 {
-    private final Map<Block, IStateMapper> field_178450_a = Maps.newIdentityHashMap();
-    private final Set<Block> field_178449_b = Sets.newIdentityHashSet();
+    private final Map<Block, IStateMapper> field_178450_a = Maps.<Block, IStateMapper>newIdentityHashMap();
+    private final Set<Block> field_178449_b = Sets.<Block>newIdentityHashSet();
 
     public void func_178447_a(Block p_178447_1_, IStateMapper p_178447_2_)
     {
@@ -31,7 +31,7 @@
 
     public Map<IBlockState, ModelResourceLocation> func_178446_a()
     {
-        Map<IBlockState, ModelResourceLocation> map = Maps.newIdentityHashMap();
+        Map<IBlockState, ModelResourceLocation> map = Maps.<IBlockState, ModelResourceLocation>newIdentityHashMap();
 
         for (Block block : Block.field_149771_c)
         {
@@ -45,7 +45,7 @@
     {
         if (this.field_178449_b.contains(p_188182_1_))
         {
-            return Collections.emptySet();
+            return Collections.<ResourceLocation>emptySet();
         }
         else
         {
@@ -53,11 +53,11 @@
 
             if (istatemapper == null)
             {
-                return Collections.singleton(Block.field_149771_c.func_177774_c(p_188182_1_));
+                return Collections.<ResourceLocation>singleton(Block.field_149771_c.func_177774_c(p_188182_1_));
             }
             else
             {
-                Set<ResourceLocation> set = Sets.newHashSet();
+                Set<ResourceLocation> set = Sets.<ResourceLocation>newHashSet();
 
                 for (ModelResourceLocation modelresourcelocation : istatemapper.func_178130_a(p_188182_1_).values())
                 {
@@ -71,6 +71,6 @@
 
     public Map<IBlockState, ModelResourceLocation> func_188181_b(Block p_188181_1_)
     {
-        return this.field_178449_b.contains(p_188181_1_) ? Collections.emptyMap() : MoreObjects.firstNonNull(this.field_178450_a.get(p_188181_1_), new DefaultStateMapper()).func_178130_a(p_188181_1_);
+        return this.field_178449_b.contains(p_188181_1_) ? Collections.emptyMap() : ((IStateMapper)MoreObjects.firstNonNull(this.field_178450_a.get(p_188181_1_), new DefaultStateMapper())).func_178130_a(p_188181_1_);
     }
 }
