--- before/net/minecraft/client/resources/DefaultResourcePack.java
+++ after/net/minecraft/client/resources/DefaultResourcePack.java
@@ -20,7 +20,7 @@
 @SideOnly(Side.CLIENT)
 public class DefaultResourcePack implements IResourcePack
 {
-    public static final Set<String> field_110608_a = ImmutableSet.of("minecraft", "realms");
+    public static final Set<String> field_110608_a = ImmutableSet.<String>of("minecraft", "realms");
     private final ResourceIndex field_188549_b;
 
     public DefaultResourcePack(ResourceIndex p_i46541_1_)
@@ -68,7 +68,7 @@
             URL url = DefaultResourcePack.class.getResource(s);
             return url != null && FolderResourcePack.func_191384_a(new File(url.getFile()), s) ? DefaultResourcePack.class.getResourceAsStream(s) : null;
         }
-        catch (IOException ioexception)
+        catch (IOException var4)
         {
             return DefaultResourcePack.class.getResourceAsStream(s);
         }
@@ -90,13 +90,13 @@
         try
         {
             InputStream inputstream = new FileInputStream(this.field_188549_b.func_188546_a());
-            return AbstractResourcePack.func_110596_a(p_135058_1_, inputstream, p_135058_2_);
+            return (T)AbstractResourcePack.func_110596_a(p_135058_1_, inputstream, p_135058_2_);
         }
-        catch (RuntimeException runtimeexception)
+        catch (RuntimeException var4)
         {
             return (T)null;
         }
-        catch (FileNotFoundException filenotfoundexception)
+        catch (FileNotFoundException var5)
         {
             return (T)null;
         }
