--- before/net/minecraft/client/resources/FileResourcePack.java
+++ after/net/minecraft/client/resources/FileResourcePack.java
@@ -58,7 +58,7 @@
         {
             return this.func_110599_c().getEntry(p_110593_1_) != null;
         }
-        catch (IOException ioexception)
+        catch (IOException var3)
         {
             return false;
         }
@@ -72,13 +72,13 @@
         {
             zipfile = this.func_110599_c();
         }
-        catch (IOException ioexception)
+        catch (IOException var8)
         {
-            return Collections.emptySet();
+            return Collections.<String>emptySet();
         }
 
         Enumeration <? extends ZipEntry > enumeration = zipfile.entries();
-        Set<String> set = Sets.newHashSet();
+        Set<String> set = Sets.<String>newHashSet();
 
         while (enumeration.hasMoreElements())
         {
