--- before/net/minecraft/client/resources/FileResourcePack.java
+++ after/net/minecraft/client/resources/FileResourcePack.java
@@ -17,7 +17,7 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
-public class FileResourcePack extends AbstractResourcePack implements Closeable
+public class FileResourcePack extends AbstractResourcePack implements AutoCloseable
 {
     public static final Splitter field_110601_c = Splitter.on('/').omitEmptyStrings().limit(3);
     private ZipFile field_110600_d;
@@ -112,15 +112,9 @@
     }
 
     @Override
-    protected void finalize() throws Throwable
-    {
-        this.close();
-        super.finalize();
-    }
-
-    @Override
     public void close() throws IOException
     {
+        this.close();
         if (this.field_110600_d != null)
         {
             this.field_110600_d.close();
