--- ../src-base/minecraft/net/minecraft/client/resources/FolderResourcePack.java
+++ ../src-work/minecraft/net/minecraft/client/resources/FolderResourcePack.java
@@ -77,7 +77,7 @@
         return null;
     }
 
-    public Set<String> func_110587_b()
+    public Set<String> getNamespaces()
     {
         Set<String> set = Sets.<String>newHashSet();
         File file1 = new File(this.field_110597_b, "assets/");
