--- before/net/minecraft/client/resources/IResourcePack.java
+++ after/net/minecraft/client/resources/IResourcePack.java
@@ -18,7 +18,7 @@
 
     boolean func_110589_b(ResourceLocation p_110589_1_);
 
-    Set<String> func_110587_b();
+    Set<String> getNamespaces();
 
     @Nullable
     <T extends IMetadataSection> T func_135058_a(MetadataSerializer p_135058_1_, String p_135058_2_) throws IOException;
