--- before/net/minecraft/client/resources/LanguageManager.java
+++ after/net/minecraft/client/resources/LanguageManager.java
@@ -22,7 +22,7 @@
     private final MetadataSerializer field_135047_b;
     private String field_135048_c;
     protected static final Locale field_135049_a = new Locale();
-    private final Map<String, Language> field_135046_d = Maps.newHashMap();
+    private final Map<String, Language> field_135046_d = Maps.<String, Language>newHashMap();
 
     public LanguageManager(MetadataSerializer p_i1304_1_, String p_i1304_2_)
     {
@@ -39,7 +39,7 @@
         {
             try
             {
-                LanguageMetadataSection languagemetadatasection = iresourcepack.func_135058_a(this.field_135047_b, "language");
+                LanguageMetadataSection languagemetadatasection = (LanguageMetadataSection)iresourcepack.func_135058_a(this.field_135047_b, "language");
 
                 if (languagemetadatasection != null)
                 {
