--- before/net/minecraft/client/resources/LegacyV2Adapter.java
+++ after/net/minecraft/client/resources/LegacyV2Adapter.java
@@ -63,7 +63,7 @@
     @Nullable
     public <T extends IMetadataSection> T func_135058_a(MetadataSerializer p_135058_1_, String p_135058_2_) throws IOException
     {
-        return this.field_191383_a.func_135058_a(p_135058_1_, p_135058_2_);
+        return (T)this.field_191383_a.func_135058_a(p_135058_1_, p_135058_2_);
     }
 
     public BufferedImage func_110586_a() throws IOException
