--- before/net/minecraft/client/resources/LegacyV2Adapter.java
+++ after/net/minecraft/client/resources/LegacyV2Adapter.java
@@ -55,9 +55,9 @@
         return this.field_191383_a.func_110589_b(this.func_191382_c(p_110589_1_));
     }
 
-    public Set<String> func_110587_b()
+    public Set<String> getNamespaces()
     {
-        return this.field_191383_a.func_110587_b();
+        return this.field_191383_a.getNamespaces();
     }
 
     @Nullable
