--- before/net/minecraft/client/resources/Locale.java
+++ after/net/minecraft/client/resources/Locale.java
@@ -53,26 +53,6 @@
 
     private void checkUnicode()
     {
-        this.unicode = false;
-        int i = 0;
-        int j = 0;
-
-        for (String s : this.properties.values())
-        {
-            int k = s.length();
-            j += k;
-
-            for (int l = 0; l < k; l++)
-            {
-                if (s.charAt(l) >= 256)
-                {
-                    i++;
-                }
-            }
-        }
-
-        float f = (float)i / j;
-        this.unicode = f > 0.1;
     }
 
     private void loadLocaleData(List<IResource> resourcesList) throws IOException
@@ -94,6 +74,8 @@
 
     private void loadLocaleData(InputStream inputStreamIn) throws IOException
     {
+        inputStreamIn = net.minecraftforge.fml.common.FMLCommonHandler.instance().loadLanguage(properties, inputStreamIn);
+        if (inputStreamIn == null) return;
         for (String s : IOUtils.readLines(inputStreamIn, StandardCharsets.UTF_8))
         {
             if (!s.isEmpty() && s.charAt(0) != '#')
