--- before/net/minecraft/client/resources/ResourcePackRepository.java
+++ after/net/minecraft/client/resources/ResourcePackRepository.java
@@ -297,7 +297,7 @@
                     FileUtils.deleteQuietly(file1);
                     settablefuture.setException(p_onFailure_1_);
                 }
-            });
+            }, Runnable::run);
             return this.downloadingPacks;
         }
         finally
@@ -308,9 +308,10 @@
 
     private boolean checkHash(String p_190113_1_, File p_190113_2_)
     {
+        java.io.InputStream is = null;
         try
         {
-            String s = DigestUtils.sha1Hex(new FileInputStream(p_190113_2_));
+            String s = DigestUtils.sha1Hex(is = new FileInputStream(p_190113_2_));
 
             if (p_190113_1_.isEmpty())
             {
@@ -330,6 +331,7 @@
         {
             LOGGER.warn("File {} couldn't be hashed.", p_190113_2_, ioexception);
         }
+        finally { IOUtils.closeQuietly(is); } // Forge: close stream after use
 
         return false;
     }
