--- before/net/minecraft/client/resources/ResourcePackListEntry.java
+++ after/net/minecraft/client/resources/ResourcePackListEntry.java
@@ -18,9 +18,9 @@
 public abstract class ResourcePackListEntry implements GuiListExtended.IGuiListEntry
 {
     private static final ResourceLocation field_148316_c = new ResourceLocation("textures/gui/resource_packs.png");
-    private static final ITextComponent field_183020_d = new TextComponentTranslation("resourcePack.incompatible");
-    private static final ITextComponent field_183021_e = new TextComponentTranslation("resourcePack.incompatible.old");
-    private static final ITextComponent field_183022_f = new TextComponentTranslation("resourcePack.incompatible.new");
+    private static final ITextComponent field_183020_d = new TextComponentTranslation("resourcePack.incompatible", new Object[0]);
+    private static final ITextComponent field_183021_e = new TextComponentTranslation("resourcePack.incompatible.old", new Object[0]);
+    private static final ITextComponent field_183022_f = new TextComponentTranslation("resourcePack.incompatible.new", new Object[0]);
     protected final Minecraft field_148317_a;
     protected final GuiScreenResourcePacks field_148315_b;
 
@@ -159,14 +159,14 @@
     {
         List<ResourcePackListEntry> list = this.field_148315_b.func_146962_b(this);
         int i = list.indexOf(this);
-        return i > 0 && list.get(i - 1).func_148310_d();
+        return i > 0 && ((ResourcePackListEntry)list.get(i - 1)).func_148310_d();
     }
 
     protected boolean func_148307_h()
     {
         List<ResourcePackListEntry> list = this.field_148315_b.func_146962_b(this);
         int i = list.indexOf(this);
-        return i >= 0 && i < list.size() - 1 && list.get(i + 1).func_148310_d();
+        return i >= 0 && i < list.size() - 1 && ((ResourcePackListEntry)list.get(i + 1)).func_148310_d();
     }
 
     public boolean func_148278_a(int p_148278_1_, int p_148278_2_, int p_148278_3_, int p_148278_4_, int p_148278_5_, int p_148278_6_)
@@ -176,7 +176,7 @@
             if (this.func_148309_e())
             {
                 this.field_148315_b.func_175288_g();
-                final int j = this.field_148315_b.func_146963_h().get(0).func_186768_j() ? 1 : 0;
+                final int j = ((ResourcePackListEntry)this.field_148315_b.func_146963_h().get(0)).func_186768_j() ? 1 : 0;
                 int l = this.func_183019_a();
 
                 if (l == 3)
