--- before/net/minecraft/client/resources/ResourceIndex.java
+++ after/net/minecraft/client/resources/ResourceIndex.java
@@ -41,7 +41,7 @@
         try
         {
             bufferedreader = Files.newReader(file2, StandardCharsets.UTF_8);
-            JsonObject jsonobject = new JsonParser().parse(bufferedreader).getAsJsonObject();
+            JsonObject jsonobject = JsonParser.parseReader(bufferedreader).getAsJsonObject();
             JsonObject jsonobject1 = JsonUtils.getJsonObject(jsonobject, "objects", null);
 
             if (jsonobject1 != null)
