--- before/net/minecraft/client/resources/SimpleResource.java
+++ after/net/minecraft/client/resources/SimpleResource.java
@@ -21,7 +21,7 @@
 @SideOnly(Side.CLIENT)
 public class SimpleResource implements IResource
 {
-    private final Map<String, IMetadataSection> field_110535_a = Maps.newHashMap();
+    private final Map<String, IMetadataSection> field_110535_a = Maps.<String, IMetadataSection>newHashMap();
     private final String field_177242_b;
     private final ResourceLocation field_110533_b;
     private final InputStream field_110534_c;
@@ -140,7 +140,8 @@
     public int hashCode()
     {
         int i = this.field_177242_b != null ? this.field_177242_b.hashCode() : 0;
-        return 31 * i + (this.field_110533_b != null ? this.field_110533_b.hashCode() : 0);
+        i = 31 * i + (this.field_110533_b != null ? this.field_110533_b.hashCode() : 0);
+        return i;
     }
 
     public void close() throws IOException
