--- before/net/minecraft/client/resources/SkinManager.java
+++ after/net/minecraft/client/resources/SkinManager.java
@@ -30,39 +30,39 @@
 @SideOnly(Side.CLIENT)
 public class SkinManager
 {
-    private static final ExecutorService field_152794_b = new ThreadPoolExecutor(0, 2, 1L, TimeUnit.MINUTES, new LinkedBlockingQueue<>());
+    private static final ExecutorService field_152794_b = new ThreadPoolExecutor(0, 2, 1L, TimeUnit.MINUTES, new LinkedBlockingQueue());
     private final TextureManager field_152795_c;
     private final File field_152796_d;
     private final MinecraftSessionService field_152797_e;
-    private final LoadingCache<GameProfile, Map<MinecraftProfileTexture.Type, MinecraftProfileTexture>> field_152798_f;
+    private final LoadingCache<GameProfile, Map<Type, MinecraftProfileTexture>> field_152798_f;
 
     public SkinManager(TextureManager p_i1044_1_, File p_i1044_2_, MinecraftSessionService p_i1044_3_)
     {
         this.field_152795_c = p_i1044_1_;
         this.field_152796_d = p_i1044_2_;
         this.field_152797_e = p_i1044_3_;
-        this.field_152798_f = CacheBuilder.newBuilder().expireAfterAccess(15L, TimeUnit.SECONDS).build(new CacheLoader<GameProfile, Map<MinecraftProfileTexture.Type, MinecraftProfileTexture>>()
+        this.field_152798_f = CacheBuilder.newBuilder().expireAfterAccess(15L, TimeUnit.SECONDS).<GameProfile, Map<Type, MinecraftProfileTexture>>build(new CacheLoader<GameProfile, Map<Type, MinecraftProfileTexture>>()
         {
-            public Map<MinecraftProfileTexture.Type, MinecraftProfileTexture> load(GameProfile p_load_1_) throws Exception
+            public Map<Type, MinecraftProfileTexture> load(GameProfile p_load_1_) throws Exception
             {
                 try
                 {
                     return Minecraft.func_71410_x().func_152347_ac().getTextures(p_load_1_, false);
                 }
-                catch (Throwable throwable)
+                catch (Throwable var3)
                 {
-                    return Maps.newHashMap();
+                    return Maps.<Type, MinecraftProfileTexture>newHashMap();
                 }
             }
         });
     }
 
-    public ResourceLocation func_152792_a(MinecraftProfileTexture p_152792_1_, MinecraftProfileTexture.Type p_152792_2_)
+    public ResourceLocation func_152792_a(MinecraftProfileTexture p_152792_1_, Type p_152792_2_)
     {
         return this.func_152789_a(p_152792_1_, p_152792_2_, (SkinManager.SkinAvailableCallback)null);
     }
 
-    public ResourceLocation func_152789_a(final MinecraftProfileTexture p_152789_1_, final MinecraftProfileTexture.Type p_152789_2_, @Nullable final SkinManager.SkinAvailableCallback p_152789_3_)
+    public ResourceLocation func_152789_a(final MinecraftProfileTexture p_152789_1_, final Type p_152789_2_, @Nullable final SkinManager.SkinAvailableCallback p_152789_3_)
     {
         final ResourceLocation resourcelocation = new ResourceLocation("skins/" + p_152789_1_.getHash());
         ITextureObject itextureobject = this.field_152795_c.func_110581_b(resourcelocation);
@@ -115,14 +115,15 @@
         {
             public void run()
             {
-                final Map<MinecraftProfileTexture.Type, MinecraftProfileTexture> map = Maps.newHashMap();
+                final Map<Type, MinecraftProfileTexture> map = Maps.<Type, MinecraftProfileTexture>newHashMap();
 
                 try
                 {
                     map.putAll(SkinManager.this.field_152797_e.getTextures(p_152790_1_, p_152790_3_));
                 }
-                catch (InsecureTextureException insecuretextureexception)
+                catch (InsecureTextureException var3)
                 {
+                    ;
                 }
 
                 if (map.isEmpty() && p_152790_1_.getId().equals(Minecraft.func_71410_x().func_110432_I().func_148256_e().getId()))
@@ -151,14 +152,14 @@
         });
     }
 
-    public Map<MinecraftProfileTexture.Type, MinecraftProfileTexture> func_152788_a(GameProfile p_152788_1_)
+    public Map<Type, MinecraftProfileTexture> func_152788_a(GameProfile p_152788_1_)
     {
-        return this.field_152798_f.getUnchecked(p_152788_1_);
+        return (Map)this.field_152798_f.getUnchecked(p_152788_1_);
     }
 
     @SideOnly(Side.CLIENT)
     public interface SkinAvailableCallback
     {
-        void func_180521_a(MinecraftProfileTexture.Type p_180521_1_, ResourceLocation p_180521_2_, MinecraftProfileTexture p_180521_3_);
+        void func_180521_a(Type p_180521_1_, ResourceLocation p_180521_2_, MinecraftProfileTexture p_180521_3_);
     }
 }
