--- before/net/minecraft/client/renderer/tileentity/TileEntityBedRenderer.java
+++ after/net/minecraft/client/renderer/tileentity/TileEntityBedRenderer.java
@@ -14,7 +14,12 @@
 {
     private static final ResourceLocation[] field_193848_a;
     private ModelBed field_193849_d = new ModelBed();
-    private int field_193850_e = this.field_193849_d.func_193770_a();
+    private int field_193850_e;
+
+    public TileEntityBedRenderer()
+    {
+        this.field_193850_e = this.field_193849_d.func_193770_a();
+    }
 
     public void func_192841_a(TileEntityBed p_192841_1_, double p_192841_2_, double p_192841_4_, double p_192841_6_, float p_192841_8_, int p_192841_9_, float p_192841_10_)
     {
