--- before/net/minecraft/client/renderer/tileentity/TileEntityItemStackRenderer.java
+++ after/net/minecraft/client/renderer/tileentity/TileEntityItemStackRenderer.java
@@ -1,6 +1,7 @@
 package net.minecraft.client.renderer.tileentity;
 
 import com.mojang.authlib.GameProfile;
+import java.util.UUID;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockChest;
 import net.minecraft.block.BlockShulkerBox;
@@ -51,24 +52,19 @@
         if (item == Items.field_179564_cE)
         {
             this.field_179024_e.func_175112_a(p_192838_1_, false);
-            TileEntityRendererDispatcher.field_147556_a.func_192855_a(this.field_179024_e, 0.0, 0.0, 0.0, 0.0F, p_192838_2_);
+            TileEntityRendererDispatcher.field_147556_a.func_192855_a(this.field_179024_e, 0.0D, 0.0D, 0.0D, 0.0F, p_192838_2_);
         }
         else if (item == Items.field_151104_aV)
         {
             this.field_193843_g.func_193051_a(p_192838_1_);
-            TileEntityRendererDispatcher.field_147556_a.func_147549_a(this.field_193843_g, 0.0, 0.0, 0.0, 0.0F);
+            TileEntityRendererDispatcher.field_147556_a.func_147549_a(this.field_193843_g, 0.0D, 0.0D, 0.0D, 0.0F);
         }
         else if (item == Items.field_185159_cQ)
         {
             if (p_192838_1_.func_179543_a("BlockEntityTag") != null)
             {
                 this.field_179024_e.func_175112_a(p_192838_1_, true);
-                Minecraft.func_71410_x()
-                .func_110434_K()
-                .func_110577_a(
-                    BannerTextures.field_187485_b
-                    .func_187478_a(this.field_179024_e.func_175116_e(), this.field_179024_e.func_175114_c(), this.field_179024_e.func_175110_d())
-                );
+                Minecraft.func_71410_x().func_110434_K().func_110577_a(BannerTextures.field_187485_b.func_187478_a(this.field_179024_e.func_175116_e(), this.field_179024_e.func_175114_c(), this.field_179024_e.func_175110_d()));
             }
             else
             {
@@ -94,7 +90,7 @@
                 }
                 else if (nbttagcompound.func_150297_b("SkullOwner", 8) && !StringUtils.isBlank(nbttagcompound.func_74779_i("SkullOwner")))
                 {
-                    GameProfile gameprofile1 = new GameProfile(null, nbttagcompound.func_74779_i("SkullOwner"));
+                    GameProfile gameprofile1 = new GameProfile((UUID)null, nbttagcompound.func_74779_i("SkullOwner"));
                     gameprofile = TileEntitySkull.func_174884_b(gameprofile1);
                     nbttagcompound.func_82580_o("SkullOwner");
                     nbttagcompound.func_74782_a("SkullOwner", NBTUtil.func_180708_a(new NBTTagCompound(), gameprofile));
@@ -112,20 +108,20 @@
         }
         else if (item == Item.func_150898_a(Blocks.field_150477_bB))
         {
-            TileEntityRendererDispatcher.field_147556_a.func_192855_a(this.field_147716_d, 0.0, 0.0, 0.0, 0.0F, p_192838_2_);
+            TileEntityRendererDispatcher.field_147556_a.func_192855_a(this.field_147716_d, 0.0D, 0.0D, 0.0D, 0.0F, p_192838_2_);
         }
         else if (item == Item.func_150898_a(Blocks.field_150447_bR))
         {
-            TileEntityRendererDispatcher.field_147556_a.func_192855_a(this.field_147718_c, 0.0, 0.0, 0.0, 0.0F, p_192838_2_);
+            TileEntityRendererDispatcher.field_147556_a.func_192855_a(this.field_147718_c, 0.0D, 0.0D, 0.0D, 0.0F, p_192838_2_);
         }
         else if (Block.func_149634_a(item) instanceof BlockShulkerBox)
         {
-            TileEntityRendererDispatcher.field_147556_a
-            .func_192855_a(field_191274_b[BlockShulkerBox.func_190955_b(item).func_176765_a()], 0.0, 0.0, 0.0, 0.0F, p_192838_2_);
+            TileEntityRendererDispatcher.field_147556_a.func_192855_a(field_191274_b[BlockShulkerBox.func_190955_b(item).func_176765_a()], 0.0D, 0.0D, 0.0D, 0.0F, p_192838_2_);
         }
+        else if (Block.func_149634_a(item) != Blocks.field_150486_ae) net.minecraftforge.client.ForgeHooksClient.renderTileItem(p_192838_1_.func_77973_b(), p_192838_1_.func_77960_j());
         else
         {
-            TileEntityRendererDispatcher.field_147556_a.func_192855_a(this.field_147717_b, 0.0, 0.0, 0.0, 0.0F, p_192838_2_);
+            TileEntityRendererDispatcher.field_147556_a.func_192855_a(this.field_147717_b, 0.0D, 0.0D, 0.0D, 0.0F, p_192838_2_);
         }
     }
 
