--- before/net/minecraft/client/renderer/tileentity/TileEntitySkullRenderer.java
+++ after/net/minecraft/client/renderer/tileentity/TileEntitySkullRenderer.java
@@ -82,7 +82,7 @@
                     if (p_188190_7_ != null)
                     {
                         Minecraft minecraft = Minecraft.func_71410_x();
-                        Map<MinecraftProfileTexture.Type, MinecraftProfileTexture> map = minecraft.func_152342_ad().func_152788_a(p_188190_7_);
+                        Map<Type, MinecraftProfileTexture> map = minecraft.func_152342_ad().func_152788_a(p_188190_7_);
 
                         if (map.containsKey(Type.SKIN))
                         {
