--- before/net/minecraft/client/renderer/tileentity/TileEntityEnchantmentTableRenderer.java
+++ after/net/minecraft/client/renderer/tileentity/TileEntityEnchantmentTableRenderer.java
@@ -25,6 +25,7 @@
 
         for (f1 = p_192841_1_.field_145928_o - p_192841_1_.field_145925_p; f1 >= (float)Math.PI; f1 -= ((float)Math.PI * 2F))
         {
+            ;
         }
 
         while (f1 < -(float)Math.PI)
