--- before/net/minecraft/network/status/server/SPacketServerInfo.java
+++ after/net/minecraft/network/status/server/SPacketServerInfo.java
@@ -30,7 +30,7 @@
 
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
-        this.field_149296_b = JsonUtils.func_188178_a(field_149297_a, p_148837_1_.func_150789_c(32767), ServerStatusResponse.class);
+        this.field_149296_b = (ServerStatusResponse)JsonUtils.func_188178_a(field_149297_a, p_148837_1_.func_150789_c(32767), ServerStatusResponse.class);
     }
 
     public void func_148840_b(PacketBuffer p_148840_1_) throws IOException
