--- before/net/minecraft/client/multiplayer/ChunkProviderClient.java
+++ after/net/minecraft/client/multiplayer/ChunkProviderClient.java
@@ -61,6 +61,7 @@
     {
         Chunk chunk = new Chunk(this.world, chunkX, chunkZ);
         this.loadedChunks.put(ChunkPos.asLong(chunkX, chunkZ), chunk);
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(chunk));
         chunk.markLoaded(true);
         return chunk;
     }
