--- before/net/minecraft/client/multiplayer/ChunkProviderClient.java
+++ after/net/minecraft/client/multiplayer/ChunkProviderClient.java
@@ -3,6 +3,7 @@
 import com.google.common.base.MoreObjects;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectIterator;
 import javax.annotation.Nullable;
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.world.World;
@@ -21,7 +22,6 @@
     private final Chunk blankChunk;
     private final Long2ObjectMap<Chunk> loadedChunks = new Long2ObjectOpenHashMap<Chunk>(8192)
     {
-        @Override
         protected void rehash(int p_rehash_1_)
         {
             if (p_rehash_1_ > this.key.length)
@@ -51,51 +51,49 @@
     }
 
     @Nullable
-    @Override
     public Chunk getLoadedChunk(int x, int z)
     {
-        return this.loadedChunks.get(ChunkPos.asLong(x, z));
+        return (Chunk)this.loadedChunks.get(ChunkPos.asLong(x, z));
     }
 
     public Chunk loadChunk(int chunkX, int chunkZ)
     {
         Chunk chunk = new Chunk(this.world, chunkX, chunkZ);
         this.loadedChunks.put(ChunkPos.asLong(chunkX, chunkZ), chunk);
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(chunk));
         chunk.markLoaded(true);
         return chunk;
     }
 
-    @Override
     public Chunk provideChunk(int x, int z)
     {
-        return MoreObjects.firstNonNull(this.getLoadedChunk(x, z), this.blankChunk);
+        return (Chunk)MoreObjects.firstNonNull(this.getLoadedChunk(x, z), this.blankChunk);
     }
 
-    @Override
     public boolean tick()
     {
         long i = System.currentTimeMillis();
+        ObjectIterator objectiterator = this.loadedChunks.values().iterator();
 
-        for (Chunk chunk : this.loadedChunks.values())
+        while (objectiterator.hasNext())
         {
+            Chunk chunk = (Chunk)objectiterator.next();
             chunk.onTick(System.currentTimeMillis() - i > 5L);
         }
 
         if (System.currentTimeMillis() - i > 100L)
         {
-            LOGGER.info("Warning: Clientside chunk ticking took {} ms", System.currentTimeMillis() - i);
+            LOGGER.info("Warning: Clientside chunk ticking took {} ms", (long)(System.currentTimeMillis() - i));
         }
 
         return false;
     }
 
-    @Override
     public String makeString()
     {
         return "MultiplayerChunkCache: " + this.loadedChunks.size() + ", " + this.loadedChunks.size();
     }
 
-    @Override
     public boolean isChunkGeneratedAt(int x, int z)
     {
         return this.loadedChunks.containsKey(ChunkPos.asLong(x, z));
