--- before/net/minecraft/client/multiplayer/ChunkProviderClient.java
+++ after/net/minecraft/client/multiplayer/ChunkProviderClient.java
@@ -3,6 +3,7 @@
 import com.google.common.base.MoreObjects;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectIterator;
 import javax.annotation.Nullable;
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.world.World;
@@ -21,7 +22,6 @@
     private final Chunk field_73238_a;
     private final Long2ObjectMap<Chunk> field_73236_b = new Long2ObjectOpenHashMap<Chunk>(8192)
     {
-        @Override
         protected void rehash(int p_rehash_1_)
         {
             if (p_rehash_1_ > this.key.length)
@@ -32,6 +32,14 @@
     };
     private final World field_73235_d;
 
+    public java.util.function.BiConsumer<Integer, Integer> loadChunkCallback = null;
+    public java.util.function.BiConsumer<Integer, Integer> unloadChunkCallback = null;
+
+    public Long2ObjectMap<Chunk> getLoadedChunks()
+    {
+        return field_73236_b;
+    }
+
     public ChunkProviderClient(World p_i1184_1_)
     {
         this.field_73238_a = new EmptyChunk(p_i1184_1_, 0, 0);
@@ -48,54 +56,60 @@
         }
 
         this.field_73236_b.remove(ChunkPos.func_77272_a(p_73234_1_, p_73234_2_));
+        if (unloadChunkCallback != null)
+        {
+            unloadChunkCallback.accept(p_73234_1_, p_73234_2_);
+        }
     }
 
     @Nullable
-    @Override
     public Chunk func_186026_b(int p_186026_1_, int p_186026_2_)
     {
-        return this.field_73236_b.get(ChunkPos.func_77272_a(p_186026_1_, p_186026_2_));
+        return (Chunk)this.field_73236_b.get(ChunkPos.func_77272_a(p_186026_1_, p_186026_2_));
     }
 
     public Chunk func_73158_c(int p_73158_1_, int p_73158_2_)
     {
         Chunk chunk = new Chunk(this.field_73235_d, p_73158_1_, p_73158_2_);
         this.field_73236_b.put(ChunkPos.func_77272_a(p_73158_1_, p_73158_2_), chunk);
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(chunk));
         chunk.func_177417_c(true);
+        if (loadChunkCallback != null)
+        {
+            loadChunkCallback.accept(p_73158_1_, p_73158_2_);
+        }
         return chunk;
     }
 
-    @Override
     public Chunk func_186025_d(int p_186025_1_, int p_186025_2_)
     {
-        return MoreObjects.firstNonNull(this.func_186026_b(p_186025_1_, p_186025_2_), this.field_73238_a);
+        return (Chunk)MoreObjects.firstNonNull(this.func_186026_b(p_186025_1_, p_186025_2_), this.field_73238_a);
     }
 
-    @Override
     public boolean func_73156_b()
     {
         long i = System.currentTimeMillis();
+        ObjectIterator objectiterator = this.field_73236_b.values().iterator();
 
-        for (Chunk chunk : this.field_73236_b.values())
+        while (objectiterator.hasNext())
         {
+            Chunk chunk = (Chunk)objectiterator.next();
             chunk.func_150804_b(System.currentTimeMillis() - i > 5L);
         }
 
         if (System.currentTimeMillis() - i > 100L)
         {
-            field_147436_a.info("Warning: Clientside chunk ticking took {} ms", System.currentTimeMillis() - i);
+            field_147436_a.info("Warning: Clientside chunk ticking took {} ms", (long)(System.currentTimeMillis() - i));
         }
 
         return false;
     }
 
-    @Override
     public String func_73148_d()
     {
         return "MultiplayerChunkCache: " + this.field_73236_b.size() + ", " + this.field_73236_b.size();
     }
 
-    @Override
     public boolean func_191062_e(int p_191062_1_, int p_191062_2_)
     {
         return this.field_73236_b.containsKey(ChunkPos.func_77272_a(p_191062_1_, p_191062_2_));
