--- before/net/minecraft/client/multiplayer/ClientAdvancementManager.java
+++ after/net/minecraft/client/multiplayer/ClientAdvancementManager.java
@@ -2,6 +2,7 @@
 
 import com.google.common.collect.Maps;
 import java.util.Map;
+import java.util.Map.Entry;
 import javax.annotation.Nullable;
 import net.minecraft.advancements.Advancement;
 import net.minecraft.advancements.AdvancementList;
@@ -23,7 +24,7 @@
     private static final Logger field_192800_a = LogManager.getLogger();
     private final Minecraft field_192801_b;
     private final AdvancementList field_192802_c = new AdvancementList();
-    private final Map<Advancement, AdvancementProgress> field_192803_d = Maps.newHashMap();
+    private final Map<Advancement, AdvancementProgress> field_192803_d = Maps.<Advancement, AdvancementProgress>newHashMap();
     @Nullable
     private ClientAdvancementManager.IListener field_192804_e;
     @Nullable
@@ -45,7 +46,7 @@
         this.field_192802_c.func_192085_a(p_192799_1_.func_192600_b());
         this.field_192802_c.func_192083_a(p_192799_1_.func_192603_a());
 
-        for (Map.Entry<ResourceLocation, AdvancementProgress> entry : p_192799_1_.func_192604_c().entrySet())
+        for (Entry<ResourceLocation, AdvancementProgress> entry : p_192799_1_.func_192604_c().entrySet())
         {
             Advancement advancement = this.field_192802_c.func_192084_a(entry.getKey());
 
@@ -104,7 +105,7 @@
 
         if (p_192798_1_ != null)
         {
-            for (Map.Entry<Advancement, AdvancementProgress> entry : this.field_192803_d.entrySet())
+            for (Entry<Advancement, AdvancementProgress> entry : this.field_192803_d.entrySet())
             {
                 p_192798_1_.func_191933_a(entry.getKey(), entry.getValue());
             }
