--- before/net/minecraft/client/multiplayer/GuiConnecting.java
+++ after/net/minecraft/client/multiplayer/GuiConnecting.java
@@ -35,7 +35,7 @@
         this.field_146297_k = p_i1181_2_;
         this.field_146374_i = p_i1181_1_;
         ServerAddress serveraddress = ServerAddress.func_78860_a(p_i1181_3_.field_78845_b);
-        p_i1181_2_.func_71403_a(null);
+        p_i1181_2_.func_71403_a((WorldClient)null);
         p_i1181_2_.func_71351_a(p_i1181_3_);
         this.func_146367_a(serveraddress.func_78861_a(), serveraddress.func_78864_b());
     }
@@ -44,16 +44,15 @@
     {
         this.field_146297_k = p_i1182_2_;
         this.field_146374_i = p_i1182_1_;
-        p_i1182_2_.func_71403_a(null);
+        p_i1182_2_.func_71403_a((WorldClient)null);
         this.func_146367_a(p_i1182_3_, p_i1182_4_);
     }
 
     private void func_146367_a(final String p_146367_1_, final int p_146367_2_)
     {
-        field_146370_f.info("Connecting to {}, {}", p_146367_1_, p_146367_2_);
+        field_146370_f.info("Connecting to {}, {}", p_146367_1_, Integer.valueOf(p_146367_2_));
         (new Thread("Server Connector #" + field_146372_a.incrementAndGet())
         {
-            @Override
             public void run()
             {
                 InetAddress inetaddress = null;
@@ -66,14 +65,9 @@
                     }
 
                     inetaddress = InetAddress.getByName(p_146367_1_);
-                    GuiConnecting.this.field_146371_g = NetworkManager.func_181124_a(
-                                                            inetaddress, p_146367_2_, GuiConnecting.this.field_146297_k.field_71474_y.func_181148_f()
-                                                        );
-                    GuiConnecting.this.field_146371_g
-                    .func_150719_a(
-                        new NetHandlerLoginClient(GuiConnecting.this.field_146371_g, GuiConnecting.this.field_146297_k, GuiConnecting.this.field_146374_i)
-                    );
-                    GuiConnecting.this.field_146371_g.func_179290_a(new C00Handshake(p_146367_1_, p_146367_2_, EnumConnectionState.LOGIN));
+                    GuiConnecting.this.field_146371_g = NetworkManager.func_181124_a(inetaddress, p_146367_2_, GuiConnecting.this.field_146297_k.field_71474_y.func_181148_f());
+                    GuiConnecting.this.field_146371_g.func_150719_a(new NetHandlerLoginClient(GuiConnecting.this.field_146371_g, GuiConnecting.this.field_146297_k, GuiConnecting.this.field_146374_i));
+                    GuiConnecting.this.field_146371_g.func_179290_a(new C00Handshake(p_146367_1_, p_146367_2_, EnumConnectionState.LOGIN, true));
                     GuiConnecting.this.field_146371_g.func_179290_a(new CPacketLoginStart(GuiConnecting.this.field_146297_k.func_110432_I().func_148256_e()));
                 }
                 catch (UnknownHostException unknownhostexception)
@@ -84,12 +78,7 @@
                     }
 
                     GuiConnecting.field_146370_f.error("Couldn't connect to server", (Throwable)unknownhostexception);
-                    GuiConnecting.this.field_146297_k
-                    .func_147108_a(
-                        new GuiDisconnected(
-                            GuiConnecting.this.field_146374_i, "connect.failed", new TextComponentTranslation("disconnect.genericReason", "Unknown host")
-                        )
-                    );
+                    GuiConnecting.this.field_146297_k.func_147108_a(new GuiDisconnected(GuiConnecting.this.field_146374_i, "connect.failed", new TextComponentTranslation("disconnect.genericReason", new Object[] {"Unknown host"})));
                 }
                 catch (Exception exception)
                 {
@@ -107,17 +96,12 @@
                         s = s.replaceAll(s1, "");
                     }
 
-                    GuiConnecting.this.field_146297_k
-                    .func_147108_a(
-                        new GuiDisconnected(GuiConnecting.this.field_146374_i, "connect.failed", new TextComponentTranslation("disconnect.genericReason", s))
-                    );
+                    GuiConnecting.this.field_146297_k.func_147108_a(new GuiDisconnected(GuiConnecting.this.field_146374_i, "connect.failed", new TextComponentTranslation("disconnect.genericReason", new Object[] {s})));
                 }
             }
-        })
-        .start();
+        }).start();
     }
 
-    @Override
     public void func_73876_c()
     {
         if (this.field_146371_g != null)
@@ -133,19 +117,16 @@
         }
     }
 
-    @Override
     protected void func_73869_a(char p_73869_1_, int p_73869_2_) throws IOException
     {
     }
 
-    @Override
     public void func_73866_w_()
     {
         this.field_146292_n.clear();
         this.field_146292_n.add(new GuiButton(0, this.field_146294_l / 2 - 100, this.field_146295_m / 4 + 120 + 12, I18n.func_135052_a("gui.cancel")));
     }
 
-    @Override
     protected void func_146284_a(GuiButton p_146284_1_) throws IOException
     {
         if (p_146284_1_.field_146127_k == 0)
@@ -161,7 +142,6 @@
         }
     }
 
-    @Override
     public void func_73863_a(int p_73863_1_, int p_73863_2_, float p_73863_3_)
     {
         this.func_146276_q_();
