--- before/net/minecraft/client/multiplayer/ServerData.java
+++ after/net/minecraft/client/multiplayer/ServerData.java
@@ -125,7 +125,7 @@
 
         private ServerResourceMode(String p_i1053_3_)
         {
-            this.field_152594_d = new TextComponentTranslation("addServer.resourcePack." + p_i1053_3_);
+            this.field_152594_d = new TextComponentTranslation("addServer.resourcePack." + p_i1053_3_, new Object[0]);
         }
 
         public ITextComponent func_152589_a()
