--- before/net/minecraft/util/datafix/fixes/BookPagesStrictJSON.java
+++ after/net/minecraft/util/datafix/fixes/BookPagesStrictJSON.java
@@ -38,15 +38,16 @@
                         {
                             try
                             {
-                                itextcomponent = JsonUtils.func_188176_a(SignStrictJSON.field_188225_a, s, ITextComponent.class, true);
+                                itextcomponent = (ITextComponent)JsonUtils.func_188176_a(SignStrictJSON.field_188225_a, s, ITextComponent.class, true);
 
                                 if (itextcomponent == null)
                                 {
                                     itextcomponent = new TextComponentString("");
                                 }
                             }
-                            catch (JsonParseException jsonparseexception2)
+                            catch (JsonParseException var10)
                             {
+                                ;
                             }
 
                             if (itextcomponent == null)
@@ -55,8 +56,9 @@
                                 {
                                     itextcomponent = ITextComponent.Serializer.func_150699_a(s);
                                 }
-                                catch (JsonParseException jsonparseexception1)
+                                catch (JsonParseException var9)
                                 {
+                                    ;
                                 }
                             }
 
@@ -66,8 +68,9 @@
                                 {
                                     itextcomponent = ITextComponent.Serializer.func_186877_b(s);
                                 }
-                                catch (JsonParseException jsonparseexception)
+                                catch (JsonParseException var8)
                                 {
+                                    ;
                                 }
                             }
 
