--- before/net/minecraft/util/datafix/fixes/ZombieProfToType.java
+++ after/net/minecraft/util/datafix/fixes/ZombieProfToType.java
@@ -27,8 +27,9 @@
                     {
                         i = this.func_191277_a(p_188217_1_.func_74762_e("VillagerProfession"));
                     }
-                    catch (RuntimeException runtimeexception)
+                    catch (RuntimeException var4)
                     {
+                        ;
                     }
                 }
 
