--- original/net/minecraft/network/PacketBuffer.java
+++ modified/net/minecraft/network/PacketBuffer.java
@@ -267,7 +267,7 @@
             this.writeShort(stack.getMetadata());
             NBTTagCompound nbttagcompound = null;
             if (stack.getItem().isDamageable() || stack.getItem().getShareTag()) {
-                nbttagcompound = stack.getTagCompound();
+                nbttagcompound = stack.getItem().getNBTShareTag(stack);
             }
 
             this.writeCompoundTag(nbttagcompound);
@@ -286,7 +286,7 @@
         int j = this.readByte();
         int k = this.readShort();
         ItemStack itemstack = new ItemStack(Item.getItemById(i), j, k);
-        itemstack.setTagCompound(this.readCompoundTag());
+        itemstack.getItem().readNBTShareTag(itemstack, this.readCompoundTag());
         return itemstack;
     }
 
