--- before/net/minecraft/world/storage/loot/conditions/LootConditionManager.java
+++ after/net/minecraft/world/storage/loot/conditions/LootConditionManager.java
@@ -19,8 +19,8 @@
 
 public class LootConditionManager
 {
-    private static final Map < ResourceLocation, LootCondition.Serializer<? >> NAME_TO_SERIALIZER_MAP = Maps.newHashMap();
-    private static final Map < Class <? extends LootCondition > , LootCondition.Serializer<? >> CLASS_TO_SERIALIZER_MAP = Maps.newHashMap();
+    private static final Map < ResourceLocation, LootCondition.Serializer<? >> NAME_TO_SERIALIZER_MAP = Maps. < ResourceLocation, LootCondition.Serializer<? >> newHashMap();
+    private static final Map < Class <? extends LootCondition > , LootCondition.Serializer<? >> CLASS_TO_SERIALIZER_MAP = Maps. < Class <? extends LootCondition > , LootCondition.Serializer<? >> newHashMap();
 
     public static <T extends LootCondition> void registerCondition(LootCondition.Serializer <? extends T > condition)
     {
@@ -42,6 +42,15 @@
         }
     }
 
+    public static boolean testAllConditions(Iterable<LootCondition> conditions, Random rand, LootContext context)
+    {
+        if (conditions == null) return true;
+        for (LootCondition cond : conditions)
+           if (!cond.testCondition(rand, context))
+                return false;
+        return true;
+    }
+
     public static boolean testAllConditions(@Nullable LootCondition[] conditions, Random rand, LootContext context)
     {
         if (conditions == null)
@@ -64,7 +73,7 @@
 
     public static LootCondition.Serializer<?> getSerializerForName(ResourceLocation location)
     {
-        LootCondition.Serializer<?> serializer = NAME_TO_SERIALIZER_MAP.get(location);
+        LootCondition.Serializer<?> serializer = (LootCondition.Serializer)NAME_TO_SERIALIZER_MAP.get(location);
 
         if (serializer == null)
         {
@@ -78,7 +87,7 @@
 
     public static <T extends LootCondition> LootCondition.Serializer<T> getSerializerFor(T conditionClass)
     {
-        LootCondition.Serializer<T> serializer = (LootCondition.Serializer<T>)CLASS_TO_SERIALIZER_MAP.get(conditionClass.getClass());
+        LootCondition.Serializer<T> serializer = (LootCondition.Serializer)CLASS_TO_SERIALIZER_MAP.get(conditionClass.getClass());
 
         if (serializer == null)
         {
@@ -111,7 +120,7 @@
                 {
                     serializer = LootConditionManager.getSerializerForName(resourcelocation);
                 }
-                catch (IllegalArgumentException illegalargumentexception)
+                catch (IllegalArgumentException var8)
                 {
                     throw new JsonSyntaxException("Unknown condition '" + resourcelocation + "'");
                 }
@@ -121,7 +130,7 @@
 
             public JsonElement serialize(LootCondition p_serialize_1_, Type p_serialize_2_, JsonSerializationContext p_serialize_3_)
             {
-                LootCondition.Serializer<LootCondition> serializer = LootConditionManager.getSerializerFor(p_serialize_1_);
+                LootCondition.Serializer<LootCondition> serializer = LootConditionManager.<LootCondition>getSerializerFor(p_serialize_1_);
                 JsonObject jsonobject = new JsonObject();
                 serializer.serialize(jsonobject, p_serialize_1_, p_serialize_3_);
                 jsonobject.addProperty("condition", serializer.getLootTableLocation().toString());
