--- original/net/minecraft/util/text/TextComponentBase.java
+++ modified/net/minecraft/util/text/TextComponentBase.java
@@ -6,6 +6,8 @@
 import java.util.Iterator;
 import java.util.List;
 import javax.annotation.Nullable;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
 
 public abstract class TextComponentBase implements ITextComponent {
     /** The later siblings of this component.  If this component turns the text bold, that will apply to all the siblings until a later sibling turns the text something else. */
@@ -89,7 +91,6 @@
         return stringbuilder.toString();
     }
 
-    /** Gets the text of this component <em>and all sibling components</em>, with formatting codes added for rendering. */
     @Override
     public final String getFormattedText() {
         StringBuilder stringbuilder = new StringBuilder();
