--- before/net/minecraft/world/gen/structure/template/PlacementSettings.java
+++ after/net/minecraft/world/gen/structure/template/PlacementSettings.java
@@ -121,7 +121,7 @@
         }
         else if (this.field_189953_j != null)
         {
-            return this.field_189953_j == 0L ? new Random(System.currentTimeMillis()) : new Random(this.field_189953_j);
+            return this.field_189953_j.longValue() == 0L ? new Random(System.currentTimeMillis()) : new Random(this.field_189953_j.longValue());
         }
         else if (p_189947_1_ == null)
         {
