--- before/net/minecraft/world/gen/structure/template/Template.java
+++ after/net/minecraft/world/gen/structure/template/Template.java
@@ -38,8 +38,8 @@
 
 public class Template
 {
-    private final List<Template.BlockInfo> field_186270_a = Lists.newArrayList();
-    private final List<Template.EntityInfo> field_186271_b = Lists.newArrayList();
+    private final List<Template.BlockInfo> field_186270_a = Lists.<Template.BlockInfo>newArrayList();
+    private final List<Template.EntityInfo> field_186271_b = Lists.<Template.EntityInfo>newArrayList();
     private BlockPos field_186272_c = BlockPos.field_177992_a;
     private String field_186273_d = "?";
 
@@ -63,19 +63,11 @@
         if (p_186254_3_.func_177958_n() >= 1 && p_186254_3_.func_177956_o() >= 1 && p_186254_3_.func_177952_p() >= 1)
         {
             BlockPos blockpos = p_186254_2_.func_177971_a(p_186254_3_).func_177982_a(-1, -1, -1);
-            List<Template.BlockInfo> list = Lists.newArrayList();
-            List<Template.BlockInfo> list1 = Lists.newArrayList();
-            List<Template.BlockInfo> list2 = Lists.newArrayList();
-            BlockPos blockpos1 = new BlockPos(
-                Math.min(p_186254_2_.func_177958_n(), blockpos.func_177958_n()),
-                Math.min(p_186254_2_.func_177956_o(), blockpos.func_177956_o()),
-                Math.min(p_186254_2_.func_177952_p(), blockpos.func_177952_p())
-            );
-            BlockPos blockpos2 = new BlockPos(
-                Math.max(p_186254_2_.func_177958_n(), blockpos.func_177958_n()),
-                Math.max(p_186254_2_.func_177956_o(), blockpos.func_177956_o()),
-                Math.max(p_186254_2_.func_177952_p(), blockpos.func_177952_p())
-            );
+            List<Template.BlockInfo> list = Lists.<Template.BlockInfo>newArrayList();
+            List<Template.BlockInfo> list1 = Lists.<Template.BlockInfo>newArrayList();
+            List<Template.BlockInfo> list2 = Lists.<Template.BlockInfo>newArrayList();
+            BlockPos blockpos1 = new BlockPos(Math.min(p_186254_2_.func_177958_n(), blockpos.func_177958_n()), Math.min(p_186254_2_.func_177956_o(), blockpos.func_177956_o()), Math.min(p_186254_2_.func_177952_p(), blockpos.func_177952_p()));
+            BlockPos blockpos2 = new BlockPos(Math.max(p_186254_2_.func_177958_n(), blockpos.func_177958_n()), Math.max(p_186254_2_.func_177956_o(), blockpos.func_177956_o()), Math.max(p_186254_2_.func_177952_p(), blockpos.func_177952_p()));
             this.field_186272_c = p_186254_3_;
 
             for (BlockPos.MutableBlockPos blockpos$mutableblockpos : BlockPos.func_177975_b(blockpos1, blockpos2))
@@ -97,11 +89,11 @@
                     }
                     else if (!iblockstate.func_185913_b() && !iblockstate.func_185917_h())
                     {
-                        list2.add(new Template.BlockInfo(blockpos3, iblockstate, null));
+                        list2.add(new Template.BlockInfo(blockpos3, iblockstate, (NBTTagCompound)null));
                     }
                     else
                     {
-                        list.add(new Template.BlockInfo(blockpos3, iblockstate, null));
+                        list.add(new Template.BlockInfo(blockpos3, iblockstate, (NBTTagCompound)null));
                     }
                 }
             }
@@ -124,7 +116,7 @@
 
     private void func_186255_a(World p_186255_1_, BlockPos p_186255_2_, BlockPos p_186255_3_)
     {
-        List<Entity> list = p_186255_1_.func_175647_a(Entity.class, new AxisAlignedBB(p_186255_2_, p_186255_3_), new Predicate<Entity>()
+        List<Entity> list = p_186255_1_.<Entity>func_175647_a(Entity.class, new AxisAlignedBB(p_186255_2_, p_186255_3_), new Predicate<Entity>()
         {
             public boolean apply(@Nullable Entity p_apply_1_)
             {
@@ -135,11 +127,7 @@
 
         for (Entity entity : list)
         {
-            Vec3d vec3d = new Vec3d(
-                entity.field_70165_t - (double)p_186255_2_.func_177958_n(),
-                entity.field_70163_u - (double)p_186255_2_.func_177956_o(),
-                entity.field_70161_v - (double)p_186255_2_.func_177952_p()
-            );
+            Vec3d vec3d = new Vec3d(entity.field_70165_t - (double)p_186255_2_.func_177958_n(), entity.field_70163_u - (double)p_186255_2_.func_177956_o(), entity.field_70161_v - (double)p_186255_2_.func_177952_p());
             NBTTagCompound nbttagcompound = new NBTTagCompound();
             entity.func_70039_c(nbttagcompound);
             BlockPos blockpos;
@@ -159,7 +147,7 @@
 
     public Map<BlockPos, String> func_186258_a(BlockPos p_186258_1_, PlacementSettings p_186258_2_)
     {
-        Map<BlockPos, String> map = Maps.newHashMap();
+        Map<BlockPos, String> map = Maps.<BlockPos, String>newHashMap();
         StructureBoundingBox structureboundingbox = p_186258_2_.func_186213_g();
 
         for (Template.BlockInfo template$blockinfo : this.field_186270_a)
@@ -215,10 +203,7 @@
 
     public void func_189960_a(World p_189960_1_, BlockPos p_189960_2_, @Nullable ITemplateProcessor p_189960_3_, PlacementSettings p_189960_4_, int p_189960_5_)
     {
-        if ((!this.field_186270_a.isEmpty() || !p_189960_4_.func_186221_e() && !this.field_186271_b.isEmpty())
-                && this.field_186272_c.func_177958_n() >= 1
-                && this.field_186272_c.func_177956_o() >= 1
-                && this.field_186272_c.func_177952_p() >= 1)
+        if ((!this.field_186270_a.isEmpty() || !p_189960_4_.func_186221_e() && !this.field_186271_b.isEmpty()) && this.field_186272_c.func_177958_n() >= 1 && this.field_186272_c.func_177956_o() >= 1 && this.field_186272_c.func_177952_p() >= 1)
         {
             Block block = p_189960_4_.func_186219_f();
             StructureBoundingBox structureboundingbox = p_189960_4_.func_186213_g();
@@ -226,17 +211,15 @@
             for (Template.BlockInfo template$blockinfo : this.field_186270_a)
             {
                 BlockPos blockpos = func_186266_a(p_189960_4_, template$blockinfo.field_186242_a).func_177971_a(p_189960_2_);
-                Template.BlockInfo template$blockinfo1 = p_189960_3_ != null
-                        ? p_189960_3_.func_189943_a(p_189960_1_, blockpos, template$blockinfo)
-                        : template$blockinfo;
+                // Forge: skip processing blocks outside BB to prevent cascading worldgen issues
+                if (structureboundingbox != null && !structureboundingbox.func_175898_b(blockpos)) continue;
+                Template.BlockInfo template$blockinfo1 = p_189960_3_ != null ? p_189960_3_.func_189943_a(p_189960_1_, blockpos, template$blockinfo) : template$blockinfo;
 
                 if (template$blockinfo1 != null)
                 {
                     Block block1 = template$blockinfo1.field_186243_b.func_177230_c();
 
-                    if ((block == null || block != block1)
-                            && (!p_189960_4_.func_186227_h() || block1 != Blocks.field_185779_df)
-                            && (structureboundingbox == null || structureboundingbox.func_175898_b(blockpos)))
+                    if ((block == null || block != block1) && (!p_189960_4_.func_186227_h() || block1 != Blocks.field_185779_df) && (structureboundingbox == null || structureboundingbox.func_175898_b(blockpos)))
                     {
                         IBlockState iblockstate = template$blockinfo1.field_186243_b.func_185902_a(p_189960_4_.func_186212_b());
                         IBlockState iblockstate1 = iblockstate.func_185907_a(p_189960_4_.func_186215_c());
@@ -327,7 +310,7 @@
                 {
                     entity = EntityList.func_75615_a(nbttagcompound, p_186263_1_);
                 }
-                catch (Exception exception)
+                catch (Exception var15)
                 {
                     entity = null;
                 }
@@ -335,7 +318,7 @@
                 if (entity != null)
                 {
                     float f = entity.func_184217_a(p_186263_3_);
-                    f += entity.field_70177_z - entity.func_184229_a(p_186263_4_);
+                    f = f + (entity.field_70177_z - entity.func_184229_a(p_186263_4_));
                     entity.func_70012_b(vec3d1.field_72450_a, vec3d1.field_72448_b, vec3d1.field_72449_c, f, entity.field_70125_A);
                     p_186263_1_.func_72838_d(entity);
                 }
@@ -397,10 +380,10 @@
         switch (p_186269_1_)
         {
             case LEFT_RIGHT:
-                d2 = 1.0 - d2;
+                d2 = 1.0D - d2;
                 break;
             case FRONT_BACK:
-                d0 = 1.0 - d0;
+                d0 = 1.0D - d0;
                 break;
             default:
                 flag = false;
@@ -409,11 +392,11 @@
         switch (p_186269_2_)
         {
             case COUNTERCLOCKWISE_90:
-                return new Vec3d(d2, d1, 1.0 - d0);
+                return new Vec3d(d2, d1, 1.0D - d0);
             case CLOCKWISE_90:
-                return new Vec3d(1.0 - d2, d1, d0);
+                return new Vec3d(1.0D - d2, d1, d0);
             case CLOCKWISE_180:
-                return new Vec3d(1.0 - d0, d1, 1.0 - d2);
+                return new Vec3d(1.0D - d0, d1, 1.0D - d2);
             default:
                 return flag ? new Vec3d(d0, d1, d2) : p_186269_0_;
         }
@@ -454,7 +437,6 @@
     {
         p_191158_0_.func_188258_a(FixTypes.STRUCTURE, new IDataWalker()
         {
-            @Override
             public NBTTagCompound func_188266_a(IDataFixer p_188266_1_, NBTTagCompound p_188266_2_, int p_188266_3_)
             {
                 if (p_188266_2_.func_150297_b("entities", 9))
@@ -500,14 +482,7 @@
         for (Template.BlockInfo template$blockinfo : this.field_186270_a)
         {
             NBTTagCompound nbttagcompound = new NBTTagCompound();
-            nbttagcompound.func_74782_a(
-                "pos",
-                this.func_186267_a(
-                    template$blockinfo.field_186242_a.func_177958_n(),
-                    template$blockinfo.field_186242_a.func_177956_o(),
-                    template$blockinfo.field_186242_a.func_177952_p()
-                )
-            );
+            nbttagcompound.func_74782_a("pos", this.func_186267_a(template$blockinfo.field_186242_a.func_177958_n(), template$blockinfo.field_186242_a.func_177956_o(), template$blockinfo.field_186242_a.func_177952_p()));
             nbttagcompound.func_74768_a("state", template$basicpalette.func_189954_a(template$blockinfo.field_186243_b));
 
             if (template$blockinfo.field_186244_c != null)
@@ -523,22 +498,8 @@
         for (Template.EntityInfo template$entityinfo : this.field_186271_b)
         {
             NBTTagCompound nbttagcompound1 = new NBTTagCompound();
-            nbttagcompound1.func_74782_a(
-                "pos",
-                this.func_186264_a(
-                    template$entityinfo.field_186247_a.field_72450_a,
-                    template$entityinfo.field_186247_a.field_72448_b,
-                    template$entityinfo.field_186247_a.field_72449_c
-                )
-            );
-            nbttagcompound1.func_74782_a(
-                "blockPos",
-                this.func_186267_a(
-                    template$entityinfo.field_186248_b.func_177958_n(),
-                    template$entityinfo.field_186248_b.func_177956_o(),
-                    template$entityinfo.field_186248_b.func_177952_p()
-                )
-            );
+            nbttagcompound1.func_74782_a("pos", this.func_186264_a(template$entityinfo.field_186247_a.field_72450_a, template$entityinfo.field_186247_a.field_72448_b, template$entityinfo.field_186247_a.field_72449_c));
+            nbttagcompound1.func_74782_a("blockPos", this.func_186267_a(template$entityinfo.field_186248_b.func_177958_n(), template$entityinfo.field_186248_b.func_177956_o(), template$entityinfo.field_186248_b.func_177952_p()));
 
             if (template$entityinfo.field_186249_c != null)
             {
@@ -555,12 +516,11 @@
             nbttaglist2.func_74742_a(NBTUtil.func_190009_a(new NBTTagCompound(), iblockstate));
         }
 
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().getDataFixer().writeVersionData(p_189552_1_); //Moved up for MC updating reasons.
         p_189552_1_.func_74782_a("palette", nbttaglist2);
         p_189552_1_.func_74782_a("blocks", nbttaglist);
         p_189552_1_.func_74782_a("entities", nbttaglist1);
-        p_189552_1_.func_74782_a(
-            "size", this.func_186267_a(this.field_186272_c.func_177958_n(), this.field_186272_c.func_177956_o(), this.field_186272_c.func_177952_p())
-        );
+        p_189552_1_.func_74782_a("size", this.func_186267_a(this.field_186272_c.func_177958_n(), this.field_186272_c.func_177956_o(), this.field_186272_c.func_177952_p()));
         p_189552_1_.func_74778_a("author", this.field_186273_d);
         p_189552_1_.func_74768_a("DataVersion", 1343);
         return p_189552_1_;
@@ -648,11 +608,12 @@
     static class BasicPalette implements Iterable<IBlockState>
         {
             public static final IBlockState field_189957_a = Blocks.field_150350_a.func_176223_P();
-            final ObjectIntIdentityMap<IBlockState> field_189958_b = new ObjectIntIdentityMap<>(16);
+            final ObjectIntIdentityMap<IBlockState> field_189958_b;
             private int field_189959_c;
 
             private BasicPalette()
             {
+                this.field_189958_b = new ObjectIntIdentityMap<IBlockState>(16);
             }
 
             public int func_189954_a(IBlockState p_189954_1_)
@@ -675,7 +636,6 @@
                 return iblockstate == null ? field_189957_a : iblockstate;
             }
 
-            @Override
             public Iterator<IBlockState> iterator()
             {
                 return this.field_189958_b.iterator();
