--- before/net/minecraft/world/gen/structure/template/TemplateManager.java
+++ after/net/minecraft/world/gen/structure/template/TemplateManager.java
@@ -19,7 +19,7 @@
 
 public class TemplateManager
 {
-    private final Map<String, Template> field_186240_a = Maps.newHashMap();
+    private final Map<String, Template> field_186240_a = Maps.<String, Template>newHashMap();
     private final String field_186241_b;
     private final DataFixer field_191154_c;
 
@@ -62,7 +62,7 @@
                 this.func_186235_b(p_189942_2_);
             }
 
-            return this.field_186240_a.containsKey(s) ? this.field_186240_a.get(s) : null;
+            return this.field_186240_a.containsKey(s) ? (Template)this.field_186240_a.get(s) : null;
         }
     }
 
@@ -86,7 +86,7 @@
                 this.func_186239_a(s, inputstream);
                 return true;
             }
-            catch (Throwable throwable)
+            catch (Throwable var10)
             {
                 flag = false;
             }
@@ -112,7 +112,7 @@
             this.func_186239_a(s1, inputstream);
             return true;
         }
-        catch (Throwable throwable)
+        catch (Throwable var10)
         {
             flag = false;
         }
@@ -170,7 +170,7 @@
                 CompressedStreamTools.func_74799_a(nbttagcompound, outputstream);
                 return true;
             }
-            catch (Throwable throwable)
+            catch (Throwable var13)
             {
                 flag = false;
             }
