--- original/net/minecraft/world/gen/feature/WorldGenTaiga2.java
+++ modified/net/minecraft/world/gen/feature/WorldGenTaiga2.java
@@ -32,7 +32,8 @@
         int k = i - j;
         int l = 2 + rand.nextInt(2);
         boolean flag = true;
-        if (position.getY() >= 1 && position.getY() + i + 1 <= 256) {
+
+        if (position.getY() >= 1 && position.getY() + i + 1 <= worldIn.getHeight()) {
             for (int i1 = position.getY(); i1 <= position.getY() + 1 + i && flag; i1++) {
                 int j1;
                 if (i1 - position.getY() < j) {
@@ -45,9 +46,10 @@
 
                 for (int k1 = position.getX() - j1; k1 <= position.getX() + j1 && flag; k1++) {
                     for (int l1 = position.getZ() - j1; l1 <= position.getZ() + j1 && flag; l1++) {
-                        if (i1 >= 0 && i1 < 256) {
-                            Material material = worldIn.getBlockState(blockpos$mutableblockpos.setPos(k1, i1, l1)).getMaterial();
-                            if (material != Material.AIR && material != Material.LEAVES) {
+                        if (i1 >= 0 && i1 < worldIn.getHeight()) {
+                            IBlockState state = worldIn.getBlockState(blockpos$mutableblockpos.setPos(k1, i1, l1));
+
+                            if (!state.getBlock().isAir(state, worldIn, blockpos$mutableblockpos.setPos(k1, i1, l1)) && !state.getBlock().isLeaves(state, worldIn, blockpos$mutableblockpos.setPos(k1, i1, l1))) {
                                 flag = false;
                             }
                         } else {
@@ -61,10 +63,12 @@
                 return false;
             }
 
-            Block block = worldIn.getBlockState(position.down()).getBlock();
-            if ((block == Blocks.GRASS || block == Blocks.DIRT || block == Blocks.FARMLAND)
-                && position.getY() < 256 - i - 1) {
-                this.setDirtAt(worldIn, position.down());
+            BlockPos down = position.down();
+            IBlockState state = worldIn.getBlockState(down);
+            boolean isSoil = state.getBlock().canSustainPlant(state, worldIn, down, net.minecraft.util.EnumFacing.UP, (net.minecraft.block.BlockSapling)Blocks.SAPLING);
+            if (isSoil
+                && position.getY() < worldIn.getHeight() - i - 1) {
+                state.getBlock().onPlantGrow(state, worldIn, down, position);
                 int i3 = rand.nextInt(2);
                 int j3 = 1;
                 int k3 = 0;
@@ -79,7 +83,7 @@
                             int l2 = k2 - position.getZ();
                             if (Math.abs(j2) != i3 || Math.abs(l2) != i3 || i3 <= 0) {
                                 BlockPos blockpos = new BlockPos(i2, j4, k2);
-                                if (!worldIn.getBlockState(blockpos).isFullBlock()) {
+                                if (state.getBlock().canBeReplacedByLeaves(state, worldIn, blockpos)) {
                                     this.setBlockAndNotifyAdequately(worldIn, blockpos, LEAF);
                                 }
                             }
@@ -100,8 +104,8 @@
                 int i4 = rand.nextInt(3);
 
                 for (int k4 = 0; k4 < i - i4; k4++) {
-                    Material material1 = worldIn.getBlockState(position.up(k4)).getMaterial();
-                    if (material1 == Material.AIR || material1 == Material.LEAVES) {
+                    BlockPos upN = position.up(k4); state = worldIn.getBlockState(upN);
+                    if (state.getBlock().isAir(state, worldIn, upN) || state.getBlock().isLeaves(state, worldIn, upN)) {
                         this.setBlockAndNotifyAdequately(worldIn, position.up(k4), TRUNK);
                     }
                 }
