--- original/net/minecraft/world/gen/feature/WorldGeneratorBonusChest.java
+++ modified/net/minecraft/world/gen/feature/WorldGeneratorBonusChest.java
@@ -14,7 +14,7 @@
     @Override
     public boolean generate(World worldIn, Random rand, BlockPos position) {
         for (IBlockState iblockstate = worldIn.getBlockState(position);
-            (iblockstate.getMaterial() == Material.AIR || iblockstate.getMaterial() == Material.LEAVES)
+            (iblockstate.getBlock().isAir(iblockstate, worldIn, position) || iblockstate.getBlock().isLeaves(iblockstate, worldIn, position))
                 && position.getY() > 1;
             iblockstate = worldIn.getBlockState(position)
         ) {
@@ -44,19 +44,19 @@
                 BlockPos blockpos2 = blockpos.west();
                 BlockPos blockpos3 = blockpos.north();
                 BlockPos blockpos4 = blockpos.south();
-                if (worldIn.isAirBlock(blockpos2) && worldIn.getBlockState(blockpos2.down()).isTopSolid()) {
+                if (worldIn.isAirBlock(blockpos2) && worldIn.getBlockState(blockpos2.down()).isSideSolid(worldIn, blockpos2.down(), net.minecraft.util.EnumFacing.UP)) {
                     worldIn.setBlockState(blockpos2, Blocks.TORCH.getDefaultState(), 2);
                 }
 
-                if (worldIn.isAirBlock(blockpos1) && worldIn.getBlockState(blockpos1.down()).isTopSolid()) {
+                if (worldIn.isAirBlock(blockpos1) && worldIn.getBlockState(blockpos1.down()).isSideSolid(worldIn, blockpos1.down(), net.minecraft.util.EnumFacing.UP)) {
                     worldIn.setBlockState(blockpos1, Blocks.TORCH.getDefaultState(), 2);
                 }
 
-                if (worldIn.isAirBlock(blockpos3) && worldIn.getBlockState(blockpos3.down()).isTopSolid()) {
+                if (worldIn.isAirBlock(blockpos3) && worldIn.getBlockState(blockpos3.down()).isSideSolid(worldIn, blockpos3.down(), net.minecraft.util.EnumFacing.UP)) {
                     worldIn.setBlockState(blockpos3, Blocks.TORCH.getDefaultState(), 2);
                 }
 
-                if (worldIn.isAirBlock(blockpos4) && worldIn.getBlockState(blockpos4.down()).isTopSolid()) {
+                if (worldIn.isAirBlock(blockpos4) && worldIn.getBlockState(blockpos4.down()).isSideSolid(worldIn, blockpos4.down(), net.minecraft.util.EnumFacing.UP)) {
                     worldIn.setBlockState(blockpos4, Blocks.TORCH.getDefaultState(), 2);
                 }
 
