--- original/net/minecraft/world/gen/feature/WorldGenHugeTrees.java
+++ modified/net/minecraft/world/gen/feature/WorldGenHugeTrees.java
@@ -50,8 +50,8 @@
                 for (int k = -j; k <= j && flag; k++) {
                     for (int l = -j; l <= j && flag; l++) {
                         if (leavesPos.getY() + i < 0
-                            || leavesPos.getY() + i >= 256
-                            || !this.canGrowInto(worldIn.getBlockState(leavesPos.add(k, i, l)).getBlock())) {
+                                || leavesPos.getY() + i >= 256
+                                || !this.isReplaceable(worldIn,leavesPos.add(k, i, l))) {
                             flag = false;
                         }
                     }
@@ -70,12 +70,14 @@
      */
     private boolean ensureDirtsUnderneath(BlockPos pos, World worldIn) {
         BlockPos blockpos = pos.down();
-        Block block = worldIn.getBlockState(blockpos).getBlock();
-        if ((block == Blocks.GRASS || block == Blocks.DIRT) && pos.getY() >= 2) {
-            this.setDirtAt(worldIn, blockpos);
-            this.setDirtAt(worldIn, blockpos.east());
-            this.setDirtAt(worldIn, blockpos.south());
-            this.setDirtAt(worldIn, blockpos.south().east());
+        IBlockState state = worldIn.getBlockState(blockpos);
+        boolean isSoil = state.getBlock().canSustainPlant(state, worldIn, blockpos, net.minecraft.util.EnumFacing.UP, ((net.minecraft.block.BlockSapling)Blocks.SAPLING));
+
+        if (isSoil && pos.getY() >= 2) {
+            this.onPlantGrow(worldIn, blockpos, pos);
+            this.onPlantGrow(worldIn, blockpos.east(), pos);
+            this.onPlantGrow(worldIn, blockpos.south(), pos);
+            this.onPlantGrow(worldIn, blockpos.south().east(), pos);
             return true;
         } else {
             return false;
@@ -100,8 +102,9 @@
                 int i1 = k - 1;
                 if (j * j + k * k <= i || l * l + i1 * i1 <= i || j * j + i1 * i1 <= i || l * l + k * k <= i) {
                     BlockPos blockpos = layerCenter.add(j, 0, k);
-                    Material material = worldIn.getBlockState(blockpos).getMaterial();
-                    if (material == Material.AIR || material == Material.LEAVES) {
+                    IBlockState state = worldIn.getBlockState(blockpos);
+
+                    if (state.getBlock().isAir(state, worldIn, blockpos) || state.getBlock().isLeaves(state, worldIn, blockpos)) {
                         this.setBlockAndNotifyAdequately(worldIn, blockpos, this.leavesMetadata);
                     }
                 }
@@ -117,12 +120,19 @@
             for (int k = -width; k <= width; k++) {
                 if (j * j + k * k <= i) {
                     BlockPos blockpos = layerCenter.add(j, 0, k);
-                    Material material = worldIn.getBlockState(blockpos).getMaterial();
-                    if (material == Material.AIR || material == Material.LEAVES) {
+                    IBlockState state = worldIn.getBlockState(blockpos);
+
+                    if (state.getBlock().isAir(state, worldIn, blockpos) || state.getBlock().isLeaves(state, worldIn, blockpos)) {
                         this.setBlockAndNotifyAdequately(worldIn, blockpos, this.leavesMetadata);
                     }
                 }
             }
         }
     }
+
+    //Just a helper macro
+    private void onPlantGrow(World world, BlockPos pos, BlockPos source) {
+        IBlockState state = world.getBlockState(pos);
+        state.getBlock().onPlantGrow(state, world, pos, source);
+    }
 }
