--- original/net/minecraft/world/gen/feature/WorldGenLakes.java
+++ modified/net/minecraft/world/gen/feature/WorldGenLakes.java
@@ -142,7 +142,8 @@
                 for (int l3 = 0; l3 < 16; l3++) {
                     int l4 = 4;
                     if (worldIn.canBlockFreezeWater(position.add(k2, 4, l3))) {
-                        worldIn.setBlockState(position.add(k2, 4, l3), Blocks.ICE.getDefaultState(), 2);
+                        int flag = net.minecraftforge.common.ForgeModContainer.fixVanillaCascading ? 2| 16 : 2; //Forge: With bit 5 unset, it will notify neighbors and load adjacent chunks.
+                        worldIn.setBlockState(position.add(k2, 4, l3), Blocks.ICE.getDefaultState(), flag); //Forge
                     }
                 }
             }
