--- original/net/minecraft/server/dedicated/DedicatedServer.java
+++ modified/net/minecraft/server/dedicated/DedicatedServer.java
@@ -61,6 +61,7 @@
     private boolean canSpawnStructures;
     private GameType gameType;
     private boolean guiIsEnabled;
+    public static boolean allowPlayerLogins = false;
 
     public DedicatedServer(
         File anvilFileIn,
@@ -71,9 +72,11 @@
         PlayerProfileCache profileCacheIn
     ) {
         super(anvilFileIn, Proxy.NO_PROXY, dataFixerIn, authServiceIn, sessionServiceIn, profileRepoIn, profileCacheIn);
-        new Thread("Server Infinisleeper") {
+        if (!org.apache.commons.lang3.SystemUtils.IS_OS_WINDOWS) return;
+        Thread thread = new Thread("Server Infinisleeper") {
             {
                 this.setDaemon(true);
+                this.setPriority(Thread.MIN_PRIORITY);
                 this.start();
             }
 
@@ -95,6 +98,7 @@
         Thread thread = new Thread("Server console handler") {
             @Override
             public void run() {
+                if (net.minecraftforge.server.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
                 BufferedReader bufferedreader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
 
                 String s4;
@@ -114,6 +118,8 @@
             LOGGER.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
         }
 
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().onServerStart(this);
+
         LOGGER.info("Loading properties");
         this.settings = new PropertyManager(new File("server.properties"));
         this.eula = new ServerEula(new File("eula.txt"));
@@ -188,6 +194,7 @@
             return false;
         }
 
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().onServerStarted();
         this.setPlayerList(new DedicatedPlayerList(this));
         long j = System.nanoTime();
         if (this.getFolderName() == null) {
@@ -258,7 +265,8 @@
         }
 
         Items.AIR.getSubItems(CreativeTabs.SEARCH, NonNullList.create());
-        return true;
+        // <3 you Grum for this, saves us ~30 patch files! --^
+        return net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerStarting(this);
     }
 
     public String loadResourcePackSHA() {
@@ -478,7 +486,7 @@
 
     @Override
     public boolean isBlockProtected(World worldIn, BlockPos pos, EntityPlayer playerIn) {
-        if (worldIn.provider.getDimensionType().getId() != 0) {
+        if (worldIn.provider.getDimension() != 0) {
             return false;
         }
 
@@ -543,6 +551,9 @@
         return this.settings.getIntProperty("network-compression-threshold", super.getNetworkCompressionThreshold());
     }
 
+    //Forge: Enable formated text for colors in console.
+    @Override public void sendMessage(net.minecraft.util.text.ITextComponent message) { LOGGER.info(message.getFormattedText()); }
+
     protected boolean convertFiles() throws IOException {
         boolean flag = false;
 
