--- original/net/minecraft/world/biome/BiomeSnow.java
+++ modified/net/minecraft/world/biome/BiomeSnow.java
@@ -50,7 +50,7 @@
 
     @Override
     public void decorate(World worldIn, Random rand, BlockPos pos) {
-        if (this.superIcy) {
+        if (this.superIcy && net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.ICE)) {
             for (int i = 0; i < 3; i++) {
                 int j = rand.nextInt(16) + 8;
                 int k = rand.nextInt(16) + 8;
