--- original/net/minecraft/world/biome/BiomeSavanna.java
+++ modified/net/minecraft/world/biome/BiomeSavanna.java
@@ -35,6 +35,7 @@
     public void decorate(World worldIn, Random rand, BlockPos pos) {
         DOUBLE_PLANT_GENERATOR.setPlantType(BlockDoublePlant.EnumPlantType.GRASS);
 
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.GRASS))
         for (int i = 0; i < 7; i++) {
             int j = rand.nextInt(16) + 8;
             int k = rand.nextInt(16) + 8;
