--- before/net/minecraft/client/shader/ShaderManager.java
+++ after/net/minecraft/client/shader/ShaderManager.java
@@ -52,15 +52,15 @@
 
     public ShaderManager(IResourceManager resourceManager, String programName) throws JsonException, IOException
     {
-        JsonParser jsonparser = new JsonParser();
-        ResourceLocation resourcelocation = new ResourceLocation("shaders/program/" + programName + ".json");
+        String[] rl = ResourceLocation.splitObjectName(programName);
+        ResourceLocation resourcelocation = new ResourceLocation(rl[0], "shaders/program/" + rl[1] + ".json");
         this.programFilename = programName;
         IResource iresource = null;
 
         try
         {
             iresource = resourceManager.getResource(resourcelocation);
-            JsonObject jsonobject = jsonparser.parse(IOUtils.toString(iresource.getInputStream(), StandardCharsets.UTF_8)).getAsJsonObject();
+            JsonObject jsonobject = JsonParser.parseString(IOUtils.toString(iresource.getInputStream(), StandardCharsets.UTF_8)).getAsJsonObject();
             String s = JsonUtils.getString(jsonobject, "vertex");
             String s1 = JsonUtils.getString(jsonobject, "fragment");
             JsonArray jsonarray = JsonUtils.getJsonArray(jsonobject, "samplers", null);
