--- before/net/minecraft/client/shader/ShaderUniform.java
+++ after/net/minecraft/client/shader/ShaderUniform.java
@@ -1,6 +1,5 @@
 package net.minecraft.client.shader;
 
-import java.nio.Buffer;
 import java.nio.FloatBuffer;
 import java.nio.IntBuffer;
 import net.minecraft.client.renderer.OpenGlHelper;
@@ -99,14 +98,14 @@
 
     public void func_148090_a(float p_148090_1_)
     {
-        ((Buffer)this.field_148098_f).position(0);
+        this.field_148098_f.position(0);
         this.field_148098_f.put(0, p_148090_1_);
         this.func_148096_h();
     }
 
     public void func_148087_a(float p_148087_1_, float p_148087_2_)
     {
-        ((Buffer)this.field_148098_f).position(0);
+        this.field_148098_f.position(0);
         this.field_148098_f.put(0, p_148087_1_);
         this.field_148098_f.put(1, p_148087_2_);
         this.func_148096_h();
@@ -114,7 +113,7 @@
 
     public void func_148095_a(float p_148095_1_, float p_148095_2_, float p_148095_3_)
     {
-        ((Buffer)this.field_148098_f).position(0);
+        this.field_148098_f.position(0);
         this.field_148098_f.put(0, p_148095_1_);
         this.field_148098_f.put(1, p_148095_2_);
         this.field_148098_f.put(2, p_148095_3_);
@@ -123,18 +122,18 @@
 
     public void func_148081_a(float p_148081_1_, float p_148081_2_, float p_148081_3_, float p_148081_4_)
     {
-        ((Buffer)this.field_148098_f).position(0);
+        this.field_148098_f.position(0);
         this.field_148098_f.put(p_148081_1_);
         this.field_148098_f.put(p_148081_2_);
         this.field_148098_f.put(p_148081_3_);
         this.field_148098_f.put(p_148081_4_);
-        ((Buffer)this.field_148098_f).flip();
+        this.field_148098_f.flip();
         this.func_148096_h();
     }
 
     public void func_148092_b(float p_148092_1_, float p_148092_2_, float p_148092_3_, float p_148092_4_)
     {
-        ((Buffer)this.field_148098_f).position(0);
+        this.field_148098_f.position(0);
 
         if (this.field_148100_d >= 4)
         {
@@ -161,7 +160,7 @@
 
     public void func_148083_a(int p_148083_1_, int p_148083_2_, int p_148083_3_, int p_148083_4_)
     {
-        ((Buffer)this.field_148101_e).position(0);
+        this.field_148101_e.position(0);
 
         if (this.field_148100_d >= 0)
         {
@@ -190,20 +189,20 @@
     {
         if (p_148097_1_.length < this.field_148103_c)
         {
-            field_148104_a.warn("Uniform.set called with a too-small value array (expected {}, got {}). Ignoring.", this.field_148103_c, p_148097_1_.length);
+            field_148104_a.warn("Uniform.set called with a too-small value array (expected {}, got {}). Ignoring.", Integer.valueOf(this.field_148103_c), Integer.valueOf(p_148097_1_.length));
         }
         else
         {
-            ((Buffer)this.field_148098_f).position(0);
+            this.field_148098_f.position(0);
             this.field_148098_f.put(p_148097_1_);
-            ((Buffer)this.field_148098_f).position(0);
+            this.field_148098_f.position(0);
             this.func_148096_h();
         }
     }
 
     public void func_148094_a(float p_148094_1_, float p_148094_2_, float p_148094_3_, float p_148094_4_, float p_148094_5_, float p_148094_6_, float p_148094_7_, float p_148094_8_, float p_148094_9_, float p_148094_10_, float p_148094_11_, float p_148094_12_, float p_148094_13_, float p_148094_14_, float p_148094_15_, float p_148094_16_)
     {
-        ((Buffer)this.field_148098_f).position(0);
+        this.field_148098_f.position(0);
         this.field_148098_f.put(0, p_148094_1_);
         this.field_148098_f.put(1, p_148094_2_);
         this.field_148098_f.put(2, p_148094_3_);
@@ -232,6 +231,7 @@
     {
         if (!this.field_148105_h)
         {
+            ;
         }
 
         this.field_148105_h = false;
