--- before/net/minecraft/client/shader/ShaderDefault.java
+++ after/net/minecraft/client/shader/ShaderDefault.java
@@ -2,7 +2,7 @@
 
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
-import org.lwjgl.util.vector.Matrix4f;
+import org.lwjglx.util.vector.Matrix4f;
 
 @SideOnly(Side.CLIENT)
 public class ShaderDefault extends ShaderUniform
@@ -44,7 +44,7 @@
     {
     }
 
-    public void func_148088_a(Matrix4f p_148088_1_)
+    public void set(Matrix4f matrix)
     {
     }
 }
