--- before/net/minecraft/client/gui/FontRenderer.java
+++ after/net/minecraft/client/gui/FontRenderer.java
@@ -56,7 +56,7 @@
         this.locationFontTexture = location;
         this.renderEngine = textureManagerIn;
         this.unicodeFlag = unicode;
-        textureManagerIn.bindTexture(this.locationFontTexture);
+        bindTexture(this.locationFontTexture);
 
         for (int i = 0; i < 32; i++)
         {
@@ -107,7 +107,7 @@
 
         try
         {
-            iresource = Minecraft.getMinecraft().getResourceManager().getResource(this.locationFontTexture);
+            iresource = getResource(this.locationFontTexture);
             bufferedimage = TextureUtil.readBufferedImage(iresource.getInputStream());
         }
         catch (IOException ioexception)
@@ -171,7 +171,7 @@
 
         try
         {
-            iresource = Minecraft.getMinecraft().getResourceManager().getResource(new ResourceLocation("font/glyph_sizes.bin"));
+            iresource = getResource(new ResourceLocation("font/glyph_sizes.bin"));
             iresource.getInputStream().read(this.glyphWidth);
         }
         catch (IOException ioexception)
@@ -186,6 +186,7 @@
 
     private float renderChar(char ch, boolean italic)
     {
+        if (ch == 160) return 4.0F; // forge: display nbsp as space. MC-2595
         if (ch == ' ')
         {
             return 4.0F;
@@ -203,7 +204,7 @@
         int i = ch % 16 * 8;
         int j = ch / 16 * 8;
         int k = italic ? 1 : 0;
-        this.renderEngine.bindTexture(this.locationFontTexture);
+        bindTexture(this.locationFontTexture);
         int l = this.charWidth[ch];
         float f = l - 0.01F;
         GlStateManager.glBegin(5);
@@ -231,7 +232,7 @@
 
     private void loadGlyphTexture(int page)
     {
-        this.renderEngine.bindTexture(this.getUnicodePageLocation(page));
+        bindTexture(this.getUnicodePageLocation(page));
     }
 
     protected float renderUnicodeChar(char ch, boolean italic)
@@ -280,7 +281,7 @@
 
     public int drawString(String text, float x, float y, int color, boolean dropShadow)
     {
-        GlStateManager.enableAlpha();
+        enableAlpha();
         this.resetStyles();
         int i;
 
@@ -350,7 +351,7 @@
 
                     int j1 = this.colorCode[i1];
                     this.textColor = j1;
-                    GlStateManager.color((j1 >> 16) / 255.0F, (j1 >> 8 & 0xFF) / 255.0F, (j1 & 0xFF) / 255.0F, this.alpha);
+                    setColor((j1 >> 16) / 255.0F, (j1 >> 8 & 255) / 255.0F, (j1 & 255) / 255.0F, this.alpha);
                 }
                 else if (i1 == 16)
                 {
@@ -379,7 +380,7 @@
                     this.strikethroughStyle = false;
                     this.underlineStyle = false;
                     this.italicStyle = false;
-                    GlStateManager.color(this.red, this.blue, this.green, this.alpha);
+                    setColor(this.red, this.blue, this.green, this.alpha);
                 }
 
                 i++;
@@ -409,7 +410,7 @@
                     c0 = c1;
                 }
 
-                float f1 = this.unicodeFlag ? 0.5F : 1.0F;
+                float f1 = j == -1 || this.unicodeFlag ? 0.5f : 1f;
                 boolean flag = (c0 == 0 || j == -1 || this.unicodeFlag) && shadow;
 
                 if (flag)
@@ -447,6 +448,15 @@
 
                     f++;
                 }
+                doDraw(f);
+            }
+        }
+    }
+
+    protected void doDraw(float f)
+    {
+        {
+            {
 
                 if (this.strikethroughStyle)
                 {
@@ -520,7 +530,7 @@
             this.blue = (color >> 8 & 0xFF) / 255.0F;
             this.green = (color & 0xFF) / 255.0F;
             this.alpha = (color >> 24 & 0xFF) / 255.0F;
-            GlStateManager.color(this.red, this.blue, this.green, this.alpha);
+            setColor(this.red, this.blue, this.green, this.alpha);
             this.posX = x;
             this.posY = y;
             this.renderStringAtPos(text, dropShadow);
@@ -574,6 +584,7 @@
 
     public int getCharWidth(char character)
     {
+        if (character == 160) return 4; // forge: display nbsp as space. MC-2595
         if (character == 167)
         {
             return -1;
@@ -672,7 +683,7 @@
 
     private String trimStringNewline(String text)
     {
-        while (text != null && text.endsWith("\n"))
+        while (text.endsWith("\n"))
         {
             text = text.substring(0, text.length() - 1);
         }
@@ -719,9 +730,88 @@
 
     public List<String> listFormattedStringToWidth(String str, int wrapWidth)
     {
-        return Arrays.asList(this.wrapFormattedStringToWidth(str, wrapWidth).split("\n"));
+        if (str.isEmpty())
+        {
+            return java.util.Collections.singletonList("");
+        }
+        com.ibm.icu.segmenter.Segments segments = com.cleanroommc.client.SegmenterHolder.SEGMENTER.segment(str);
+        List<String> list = new java.util.ArrayList<>();
+        StringBuilder line = new StringBuilder();
+        StringBuilder seg = new StringBuilder();
+        StringBuilder format = new StringBuilder();
+        StringBuilder segFormat = new StringBuilder();
+        int formatSize = 0;
+        boolean bold  = false;
+        int lineWidth = 0;
+        int segmentWidth;
+        out:
+        for (java.util.Iterator<com.ibm.icu.segmenter.Segment> it = segments.segments().iterator(); it.hasNext(); ) {
+            com.ibm.icu.segmenter.Segment segment = it.next();
+            segmentWidth = 0;
+            int size = segment.limit - segment.start;
+            CharSequence sequence = segment.getSubSequence();
+            segFormat.setLength(0);
+            segFormat.append(format);
+            for (int i = 0; i < size; i++) {
+                char c = sequence.charAt(i);
+                switch (c) {
+                    case '\n':
+                        line.append(seg);
+                        seg.setLength(0);
+                        list.add(line.toString());
+                        line.setLength(0);
+                        line.append(format);
+                        formatSize = format.length();
+                        lineWidth = 0;
+                        continue out;
+                    case '§':
+                        if (++i < size) {
+                            char next = sequence.charAt(i);
+                            if (next == 'l' || next == 'L') {
+                                bold = true;
+                            } else if (next == 'r' || next == 'R') {
+                                bold = false;
+                                format.setLength(0);
+                            } else if (isFormatColor(next)) {
+                                bold = false;
+                            }
+                            format.append('§').append(next);
+                            seg.append('§').append(next);
+                            continue;
+                        }
+                    default:
+                        int cWidth = getCharWidth(c);
+                        if (bold) cWidth++;
+                        if (lineWidth + cWidth > wrapWidth) {
+                            if (line.length() == formatSize) {
+                                line.append(seg);
+                                seg.setLength(0);
+                                lineWidth = 0;
+                                segmentWidth = 0;
+                            } else {
+                                lineWidth = segmentWidth;
+                                format.setLength(0);
+                                format.append(segFormat);
+                            }
+                            list.add(line.toString());
+                            line.setLength(0);
+                            line.append(format);
+                            formatSize = format.length();
+                        }
+                        seg.append(c);
+                        lineWidth += cWidth;
+                        segmentWidth += cWidth;
+                }
+            }
+        }
+        line.append(seg);
+        if (!line.isEmpty()) {
+            list.add(line.toString());
+        }
+        return list;
     }
 
+    @Deprecated
     String wrapFormattedStringToWidth(String str, int wrapWidth)
     {
         int i = this.sizeStringToWidth(str, wrapWidth);
@@ -740,6 +830,7 @@
         }
     }
 
+    @Deprecated
     private int sizeStringToWidth(String str, int wrapWidth)
     {
         int i = str.length();
@@ -838,6 +929,26 @@
     public boolean getBidiFlag()
     {
         return this.bidiFlag;
+    }
+
+    protected void setColor(float r, float g, float b, float a)
+    {
+        GlStateManager.color(r,g,b,a);
+    }
+
+    protected void enableAlpha()
+    {
+        GlStateManager.enableAlpha();
+    }
+
+    protected void bindTexture(ResourceLocation location)
+    {
+        renderEngine.bindTexture(location);
+    }
+
+    protected IResource getResource(ResourceLocation location) throws IOException
+    {
+        return Minecraft.getMinecraft().getResourceManager().getResource(location);
     }
 
     public int getColorCode(char character)
