--- before/net/minecraft/client/gui/FontRenderer.java
+++ after/net/minecraft/client/gui/FontRenderer.java
@@ -56,7 +56,7 @@
         this.locationFontTexture = location;
         this.renderEngine = textureManagerIn;
         this.unicodeFlag = unicode;
-        textureManagerIn.bindTexture(this.locationFontTexture);
+        bindTexture(this.locationFontTexture);
 
         for (int i = 0; i < 32; i++)
         {
@@ -107,7 +107,7 @@
 
         try
         {
-            iresource = Minecraft.getMinecraft().getResourceManager().getResource(this.locationFontTexture);
+            iresource = getResource(this.locationFontTexture);
             bufferedimage = TextureUtil.readBufferedImage(iresource.getInputStream());
         }
         catch (IOException ioexception)
@@ -171,7 +171,7 @@
 
         try
         {
-            iresource = Minecraft.getMinecraft().getResourceManager().getResource(new ResourceLocation("font/glyph_sizes.bin"));
+            iresource = getResource(new ResourceLocation("font/glyph_sizes.bin"));
             iresource.getInputStream().read(this.glyphWidth);
         }
         catch (IOException ioexception)
@@ -186,6 +186,7 @@
 
     private float renderChar(char ch, boolean italic)
     {
+        if (ch == 160) return 4.0F; // forge: display nbsp as space. MC-2595
         if (ch == ' ')
         {
             return 4.0F;
@@ -203,7 +204,7 @@
         int i = ch % 16 * 8;
         int j = ch / 16 * 8;
         int k = italic ? 1 : 0;
-        this.renderEngine.bindTexture(this.locationFontTexture);
+        bindTexture(this.locationFontTexture);
         int l = this.charWidth[ch];
         float f = l - 0.01F;
         GlStateManager.glBegin(5);
@@ -231,7 +232,7 @@
 
     private void loadGlyphTexture(int page)
     {
-        this.renderEngine.bindTexture(this.getUnicodePageLocation(page));
+        bindTexture(this.getUnicodePageLocation(page));
     }
 
     protected float renderUnicodeChar(char ch, boolean italic)
@@ -280,7 +281,7 @@
 
     public int drawString(String text, float x, float y, int color, boolean dropShadow)
     {
-        GlStateManager.enableAlpha();
+        enableAlpha();
         this.resetStyles();
         int i;
 
@@ -326,7 +327,7 @@
         {
             char c0 = text.charAt(i);
 
-            if (c0 == 167 && i + 1 < text.length())
+            if (c0 == 167 && i + 1 < text.length() && (isFormatColor(text.charAt(i + 1)) || isFormatSpecial(text.charAt(i + 1))))
             {
                 int i1 = "0123456789abcdefklmnor".indexOf(String.valueOf(text.charAt(i + 1)).toLowerCase(Locale.ROOT).charAt(0));
 
@@ -338,11 +339,6 @@
                     this.underlineStyle = false;
                     this.italicStyle = false;
 
-                    if (i1 < 0 || i1 > 15)
-                    {
-                        i1 = 15;
-                    }
-
                     if (shadow)
                     {
                         i1 += 16;
@@ -350,7 +346,7 @@
 
                     int j1 = this.colorCode[i1];
                     this.textColor = j1;
-                    GlStateManager.color((j1 >> 16) / 255.0F, (j1 >> 8 & 0xFF) / 255.0F, (j1 & 0xFF) / 255.0F, this.alpha);
+                    setColor((j1 >> 16) / 255.0F, (j1 >> 8 & 255) / 255.0F, (j1 & 255) / 255.0F, this.alpha);
                 }
                 else if (i1 == 16)
                 {
@@ -379,7 +375,7 @@
                     this.strikethroughStyle = false;
                     this.underlineStyle = false;
                     this.italicStyle = false;
-                    GlStateManager.color(this.red, this.blue, this.green, this.alpha);
+                    setColor(this.red, this.blue, this.green, this.alpha);
                 }
 
                 i++;
@@ -409,7 +405,7 @@
                     c0 = c1;
                 }
 
-                float f1 = this.unicodeFlag ? 0.5F : 1.0F;
+                float f1 = j == -1 || this.unicodeFlag ? 0.5f : 1f;
                 boolean flag = (c0 == 0 || j == -1 || this.unicodeFlag) && shadow;
 
                 if (flag)
@@ -447,6 +443,15 @@
 
                     f++;
                 }
+                doDraw(f);
+            }
+        }
+    }
+
+    protected void doDraw(float f)
+    {
+        {
+            {
 
                 if (this.strikethroughStyle)
                 {
@@ -520,7 +525,7 @@
             this.blue = (color >> 8 & 0xFF) / 255.0F;
             this.green = (color & 0xFF) / 255.0F;
             this.alpha = (color >> 24 & 0xFF) / 255.0F;
-            GlStateManager.color(this.red, this.blue, this.green, this.alpha);
+            setColor(this.red, this.blue, this.green, this.alpha);
             this.posX = x;
             this.posY = y;
             this.renderStringAtPos(text, dropShadow);
@@ -574,6 +579,7 @@
 
     public int getCharWidth(char character)
     {
+        if (character == 160) return 4; // forge: display nbsp as space. MC-2595
         if (character == 167)
         {
             return -1;
@@ -672,7 +678,7 @@
 
     private String trimStringNewline(String text)
     {
-        while (text != null && text.endsWith("\n"))
+        while (text.endsWith("\n"))
         {
             text = text.substring(0, text.length() - 1);
         }
@@ -719,9 +725,96 @@
 
     public List<String> listFormattedStringToWidth(String str, int wrapWidth)
     {
-        return Arrays.asList(this.wrapFormattedStringToWidth(str, wrapWidth).split("\n"));
+        if (str.isEmpty()){
+            return java.util.Collections.singletonList("");
+        }
+        com.cleanroommc.client.BreakIteratorHolder.BREAK_ITERATOR.setText(str);
+        List<String> list = new java.util.ArrayList<>();
+        int lineWidth = 0, fed = 0, icui, d, prevFormat = 0;
+        StringBuilder format = new StringBuilder(); // For next line's format since it should use format of previous line
+        int[] widths = new int[str.length()];
+        String[] formats = new String[str.length()];
+        StringBuilder line = new StringBuilder();
+        String temp;
+        char[] chars = str.toCharArray();
+        char f;
+        boolean bold = false;
+        for (int i = 0; i < chars.length; i++) {
+            char current = chars[i];
+            switch (current) {
+                case '\n':
+                    list.add(line.toString());
+                    fed = i + 1;
+                    line.delete(0, line.length()).append(format);
+                    prevFormat = format.length();
+                    lineWidth = 0;
+                    widths[0] = lineWidth;
+                    formats[0] = format.toString();
+                    continue;
+                case '§':
+                    if (i + 1 < chars.length) { // Prevent out of bound
+                        f = chars[i + 1];
+                        boolean isC = isFormatColor(f);
+                        if (isC || isFormatSpecial(f)) {
+                            if (f != 'l' && f != 'L') { // Check start of bold style
+                                if (f == 'r' || f == 'R') { // Not Bold, check end of style
+                                    bold = false;
+                                    format.delete(0, format.length()); // Clear the format
+                                } else if (isC) {
+                                    bold = false;
+                                }
+                            } else {
+                                bold = true;
+                            }
+                            format.append('§').append(f); // Add to current format code
+                            line.append('§').append(f);
+                            widths[i - fed] = lineWidth;
+                            widths[i - fed + 1] = lineWidth;
+                            formats[i - fed] = format.toString();
+                            formats[i - fed + 1] = format.toString();
+                            i++;
+                            continue;
+                        }
+                    }
+                default:
+                    line.append(current);
+                    lineWidth += getCharWidth(current);
+                    if (bold) {
+                        lineWidth++; // Bold style is one pixel wider
+                    }
+                    break;
+            }
+            widths[i - fed] = lineWidth;
+            formats[i - fed] = format.toString();
+            if (lineWidth > wrapWidth) {
+                if (com.cleanroommc.client.BreakIteratorHolder.BREAK_ITERATOR.isBoundary(i)) {
+                    icui = i;
+                } else {
+                    icui = com.cleanroommc.client.BreakIteratorHolder.BREAK_ITERATOR.preceding(i);
+                }
+                if (icui <= fed + 1 || i == icui) {
+                    list.add(line.substring(0,line.length() - 1));
+                    fed = i;
+                    line.delete(0, line.length()).append(format).append(current);
+                    prevFormat = format.length();
+                    lineWidth = getCharWidth(current);
+                } else {
+                    d = icui - fed;
+                    if (line.charAt(d + prevFormat - 2) == '§') d++;
+                    list.add(line.substring(0, d + prevFormat));
+                    temp = line.substring(d + prevFormat);
+                    fed += d;
+                    line.delete(0, line.length()).append(formats[d]).append(temp);
+                    prevFormat = formats[d].length();
+                    lineWidth = lineWidth - widths[d - 1];
+                }
+            }
+        }
+        list.add(line.toString());
+        return list;
     }
 
+    @Deprecated
     String wrapFormattedStringToWidth(String str, int wrapWidth)
     {
         int i = this.sizeStringToWidth(str, wrapWidth);
@@ -740,6 +833,7 @@
         }
     }
 
+    @Deprecated
     private int sizeStringToWidth(String str, int wrapWidth)
     {
         int i = str.length();
@@ -838,6 +932,26 @@
     public boolean getBidiFlag()
     {
         return this.bidiFlag;
+    }
+
+    protected void setColor(float r, float g, float b, float a)
+    {
+        GlStateManager.color(r,g,b,a);
+    }
+
+    protected void enableAlpha()
+    {
+        GlStateManager.enableAlpha();
+    }
+
+    protected void bindTexture(ResourceLocation location)
+    {
+        renderEngine.bindTexture(location);
+    }
+
+    protected IResource getResource(ResourceLocation location) throws IOException
+    {
+        return Minecraft.getMinecraft().getResourceManager().getResource(location);
     }
 
     public int getColorCode(char character)
