--- before/net/minecraft/client/gui/GuiCreateFlatWorld.java
+++ after/net/minecraft/client/gui/GuiCreateFlatWorld.java
@@ -194,7 +194,7 @@
 
         protected void func_192637_a(int p_192637_1_, int p_192637_2_, int p_192637_3_, int p_192637_4_, int p_192637_5_, int p_192637_6_, float p_192637_7_)
         {
-            FlatLayerInfo flatlayerinfo = GuiCreateFlatWorld.this.field_146387_g.func_82650_c().get(GuiCreateFlatWorld.this.field_146387_g.func_82650_c().size() - p_192637_1_ - 1);
+            FlatLayerInfo flatlayerinfo = (FlatLayerInfo)GuiCreateFlatWorld.this.field_146387_g.func_82650_c().get(GuiCreateFlatWorld.this.field_146387_g.func_82650_c().size() - p_192637_1_ - 1);
             IBlockState iblockstate = flatlayerinfo.func_175900_c();
             Block block = iblockstate.func_177230_c();
             Item item = Item.func_150898_a(block);
