--- before/net/minecraft/client/gui/GuiControls.java
+++ after/net/minecraft/client/gui/GuiControls.java
@@ -11,10 +11,7 @@
 @SideOnly(Side.CLIENT)
 public class GuiControls extends GuiScreen
 {
-    private static final GameSettings.Options[] field_146492_g = new GameSettings.Options[]
-        {
-            GameSettings.Options.INVERT_MOUSE, GameSettings.Options.SENSITIVITY, GameSettings.Options.TOUCHSCREEN, GameSettings.Options.AUTO_JUMP
-        };
+    private static final GameSettings.Options[] field_146492_g = new GameSettings.Options[] {GameSettings.Options.INVERT_MOUSE, GameSettings.Options.SENSITIVITY, GameSettings.Options.TOUCHSCREEN, GameSettings.Options.AUTO_JUMP};
     private final GuiScreen field_146496_h;
     protected String field_146495_a = "Controls";
     private final GameSettings field_146497_i;
@@ -29,14 +26,11 @@
         this.field_146497_i = p_i1027_2_;
     }
 
-    @Override
     public void func_73866_w_()
     {
         this.field_146494_r = new GuiKeyBindingList(this, this.field_146297_k);
         this.field_146292_n.add(new GuiButton(200, this.field_146294_l / 2 - 155 + 160, this.field_146295_m - 29, 150, 20, I18n.func_135052_a("gui.done")));
-        this.field_146493_s = this.func_189646_b(
-                                  new GuiButton(201, this.field_146294_l / 2 - 155, this.field_146295_m - 29, 150, 20, I18n.func_135052_a("controls.resetAll"))
-                              );
+        this.field_146493_s = this.func_189646_b(new GuiButton(201, this.field_146294_l / 2 - 155, this.field_146295_m - 29, 150, 20, I18n.func_135052_a("controls.resetAll")));
         this.field_146495_a = I18n.func_135052_a("controls.title");
         int i = 0;
 
@@ -44,39 +38,23 @@
         {
             if (gamesettings$options.func_74380_a())
             {
-                this.field_146292_n
-                .add(
-                    new GuiOptionSlider(
-                        gamesettings$options.func_74381_c(), this.field_146294_l / 2 - 155 + i % 2 * 160, 18 + 24 * (i >> 1), gamesettings$options
-                    )
-                );
+                this.field_146292_n.add(new GuiOptionSlider(gamesettings$options.func_74381_c(), this.field_146294_l / 2 - 155 + i % 2 * 160, 18 + 24 * (i >> 1), gamesettings$options));
             }
             else
             {
-                this.field_146292_n
-                .add(
-                    new GuiOptionButton(
-                        gamesettings$options.func_74381_c(),
-                        this.field_146294_l / 2 - 155 + i % 2 * 160,
-                        18 + 24 * (i >> 1),
-                        gamesettings$options,
-                        this.field_146497_i.func_74297_c(gamesettings$options)
-                    )
-                );
+                this.field_146292_n.add(new GuiOptionButton(gamesettings$options.func_74381_c(), this.field_146294_l / 2 - 155 + i % 2 * 160, 18 + 24 * (i >> 1), gamesettings$options, this.field_146497_i.func_74297_c(gamesettings$options)));
             }
 
-            i++;
+            ++i;
         }
     }
 
-    @Override
     public void func_146274_d() throws IOException
     {
         super.func_146274_d();
         this.field_146494_r.func_178039_p();
     }
 
-    @Override
     protected void func_146284_a(GuiButton p_146284_1_) throws IOException
     {
         if (p_146284_1_.field_146127_k == 200)
@@ -87,7 +65,7 @@
         {
             for (KeyBinding keybinding : this.field_146297_k.field_71474_y.field_74324_K)
             {
-                keybinding.func_151462_b(keybinding.func_151469_h());
+                keybinding.setToDefault();
             }
 
             KeyBinding.func_74508_b();
@@ -99,11 +77,11 @@
         }
     }
 
-    @Override
     protected void func_73864_a(int p_73864_1_, int p_73864_2_, int p_73864_3_) throws IOException
     {
         if (this.field_146491_f != null)
         {
+            this.field_146491_f.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.getActiveModifier(), -100 + p_73864_3_);
             this.field_146497_i.func_151440_a(this.field_146491_f, -100 + p_73864_3_);
             this.field_146491_f = null;
             KeyBinding.func_74508_b();
@@ -114,7 +92,6 @@
         }
     }
 
-    @Override
     protected void func_146286_b(int p_146286_1_, int p_146286_2_, int p_146286_3_)
     {
         if (p_146286_3_ != 0 || !this.field_146494_r.func_148181_b(p_146286_1_, p_146286_2_, p_146286_3_))
@@ -123,24 +100,27 @@
         }
     }
 
-    @Override
     protected void func_73869_a(char p_73869_1_, int p_73869_2_) throws IOException
     {
         if (this.field_146491_f != null)
         {
             if (p_73869_2_ == 1)
             {
+                this.field_146491_f.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.NONE, 0);
                 this.field_146497_i.func_151440_a(this.field_146491_f, 0);
             }
             else if (p_73869_2_ != 0)
             {
+                this.field_146491_f.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.getActiveModifier(), p_73869_2_);
                 this.field_146497_i.func_151440_a(this.field_146491_f, p_73869_2_);
             }
             else if (p_73869_1_ > 0)
             {
+                this.field_146491_f.setKeyModifierAndCode(net.minecraftforge.client.settings.KeyModifier.getActiveModifier(), p_73869_1_ + 256);
                 this.field_146497_i.func_151440_a(this.field_146491_f, p_73869_1_ + 256);
             }
 
+            if (!net.minecraftforge.client.settings.KeyModifier.isKeyCodeModifier(p_73869_2_))
             this.field_146491_f = null;
             this.field_152177_g = Minecraft.func_71386_F();
             KeyBinding.func_74508_b();
@@ -151,7 +131,6 @@
         }
     }
 
-    @Override
     public void func_73863_a(int p_73863_1_, int p_73863_2_, float p_73863_3_)
     {
         this.func_146276_q_();
@@ -161,7 +140,7 @@
 
         for (KeyBinding keybinding : this.field_146497_i.field_74324_K)
         {
-            if (keybinding.func_151463_i() != keybinding.func_151469_h())
+            if (!keybinding.isSetToDefaultValue())
             {
                 flag = true;
                 break;
