--- before/net/minecraft/client/gui/GuiCreateWorld.java
+++ after/net/minecraft/client/gui/GuiCreateWorld.java
@@ -43,33 +43,7 @@
     private String field_146330_J;
     private int field_146331_K;
     public String field_146334_a = "";
-    private static final String[] field_146327_L = new String[]
-        {
-            "CON",
-            "COM",
-            "PRN",
-            "AUX",
-            "CLOCK$",
-            "NUL",
-            "COM1",
-            "COM2",
-            "COM3",
-            "COM4",
-            "COM5",
-            "COM6",
-            "COM7",
-            "COM8",
-            "COM9",
-            "LPT1",
-            "LPT2",
-            "LPT3",
-            "LPT4",
-            "LPT5",
-            "LPT6",
-            "LPT7",
-            "LPT8",
-            "LPT9"
-        };
+    private static final String[] field_146327_L = new String[] {"CON", "COM", "PRN", "AUX", "CLOCK$", "NUL", "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9", "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9"};
 
     public GuiCreateWorld(GuiScreen p_i46320_1_)
     {
@@ -78,14 +52,12 @@
         this.field_146330_J = I18n.func_135052_a("selectWorld.newWorld");
     }
 
-    @Override
     public void func_73876_c()
     {
         this.field_146333_g.func_146178_a();
         this.field_146335_h.func_146178_a();
     }
 
-    @Override
     public void func_73866_w_()
     {
         Keyboard.enableRepeatEvents(true);
@@ -133,9 +105,7 @@
 
     private void func_146319_h()
     {
-        this.field_146343_z.field_146126_j = I18n.func_135052_a("selectWorld.gameMode")
-                                             + ": "
-                                             + I18n.func_135052_a("selectWorld.gameMode." + this.field_146342_r);
+        this.field_146343_z.field_146126_j = I18n.func_135052_a("selectWorld.gameMode") + ": " + I18n.func_135052_a("selectWorld.gameMode." + this.field_146342_r);
         this.field_146323_G = I18n.func_135052_a("selectWorld.gameMode." + this.field_146342_r + ".line1");
         this.field_146328_H = I18n.func_135052_a("selectWorld.gameMode." + this.field_146342_r + ".line2");
         this.field_146325_B.field_146126_j = I18n.func_135052_a("selectWorld.mapFeatures") + " ";
@@ -160,9 +130,7 @@
             this.field_146326_C.field_146126_j = this.field_146326_C.field_146126_j + I18n.func_135052_a("options.off");
         }
 
-        this.field_146320_D.field_146126_j = I18n.func_135052_a("selectWorld.mapType")
-                                             + " "
-                                             + I18n.func_135052_a(WorldType.field_77139_a[this.field_146331_K].func_77128_b());
+        this.field_146320_D.field_146126_j = I18n.func_135052_a("selectWorld.mapType") + " " + I18n.func_135052_a(WorldType.field_77139_a[this.field_146331_K].func_77128_b());
         this.field_146321_E.field_146126_j = I18n.func_135052_a("selectWorld.allowCommands") + " ";
 
         if (this.field_146340_t && !this.field_146337_w)
@@ -195,13 +163,11 @@
         return p_146317_1_;
     }
 
-    @Override
     public void func_146281_b()
     {
         Keyboard.enableRepeatEvents(false);
     }
 
-    @Override
     protected void func_146284_a(GuiButton p_146284_1_) throws IOException
     {
         if (p_146284_1_.field_146124_l)
@@ -212,7 +178,7 @@
             }
             else if (p_146284_1_.field_146127_k == 0)
             {
-                this.field_146297_k.func_147108_a(null);
+                this.field_146297_k.func_147108_a((GuiScreen)null);
 
                 if (this.field_146345_x)
                 {
@@ -220,7 +186,7 @@
                 }
 
                 this.field_146345_x = true;
-                long i = new Random().nextLong();
+                long i = (new Random()).nextLong();
                 String s = this.field_146335_h.func_146179_b();
 
                 if (!StringUtils.isEmpty(s))
@@ -234,15 +200,15 @@
                             i = j;
                         }
                     }
-                    catch (NumberFormatException numberformatexception)
+                    catch (NumberFormatException var7)
                     {
                         i = (long)s.hashCode();
                     }
                 }
 
-                WorldSettings worldsettings = new WorldSettings(
-                    i, GameType.func_77142_a(this.field_146342_r), this.field_146341_s, this.field_146337_w, WorldType.field_77139_a[this.field_146331_K]
-                );
+                WorldType.field_77139_a[this.field_146331_K].onGUICreateWorldPress();
+
+                WorldSettings worldsettings = new WorldSettings(i, GameType.func_77142_a(this.field_146342_r), this.field_146341_s, this.field_146337_w, WorldType.field_77139_a[this.field_146331_K]);
                 worldsettings.func_82750_a(this.field_146334_a);
 
                 if (this.field_146338_v && !this.field_146337_w)
@@ -348,14 +314,7 @@
             }
             else if (p_146284_1_.field_146127_k == 8)
             {
-                if (WorldType.field_77139_a[this.field_146331_K] == WorldType.field_77138_c)
-                {
-                    this.field_146297_k.func_147108_a(new GuiCreateFlatWorld(this, this.field_146334_a));
-                }
-                else
-                {
-                    this.field_146297_k.func_147108_a(new GuiCustomizeWorldScreen(this, this.field_146334_a));
-                }
+                WorldType.field_77139_a[this.field_146331_K].onCustomizeButton(field_146297_k, this);
             }
         }
     }
@@ -364,14 +323,14 @@
     {
         WorldType worldtype = WorldType.field_77139_a[this.field_146331_K];
 
-        if (worldtype == null || !worldtype.func_77126_d())
-        {
-            return false;
-        }
-        else
+        if (worldtype != null && worldtype.func_77126_d())
         {
             return worldtype == WorldType.field_180272_g ? func_146272_n() : true;
         }
+        else
+        {
+            return false;
+        }
     }
 
     private void func_146315_i()
@@ -415,11 +374,7 @@
             this.field_146326_C.field_146125_m = this.field_146344_y;
             this.field_146320_D.field_146125_m = this.field_146344_y;
             this.field_146321_E.field_146125_m = this.field_146344_y;
-            this.field_146322_F.field_146125_m = this.field_146344_y
-                                                 && (
-                                                     WorldType.field_77139_a[this.field_146331_K] == WorldType.field_77138_c
-                                                     || WorldType.field_77139_a[this.field_146331_K] == WorldType.field_180271_f
-                                                 );
+            this.field_146322_F.field_146125_m = this.field_146344_y && WorldType.field_77139_a[this.field_146331_K].isCustomizable();
         }
 
         this.func_146319_h();
@@ -434,7 +389,6 @@
         }
     }
 
-    @Override
     protected void func_73869_a(char p_73869_1_, int p_73869_2_) throws IOException
     {
         if (this.field_146333_g.func_146206_l() && !this.field_146344_y)
@@ -453,11 +407,10 @@
             this.func_146284_a(this.field_146292_n.get(0));
         }
 
-        this.field_146292_n.get(0).field_146124_l = !this.field_146333_g.func_146179_b().isEmpty();
+        (this.field_146292_n.get(0)).field_146124_l = !this.field_146333_g.func_146179_b().isEmpty();
         this.func_146314_g();
     }
 
-    @Override
     protected void func_73864_a(int p_73864_1_, int p_73864_2_, int p_73864_3_) throws IOException
     {
         super.func_73864_a(p_73864_1_, p_73864_2_, p_73864_3_);
@@ -472,7 +425,6 @@
         }
     }
 
-    @Override
     public void func_73863_a(int p_73863_1_, int p_73863_2_, float p_73863_3_)
     {
         this.func_146276_q_();
@@ -497,22 +449,13 @@
 
             if (WorldType.field_77139_a[this.field_146331_K].func_151357_h())
             {
-                this.field_146289_q
-                .func_78279_b(
-                    I18n.func_135052_a(WorldType.field_77139_a[this.field_146331_K].func_151359_c()),
-                    this.field_146320_D.field_146128_h + 2,
-                    this.field_146320_D.field_146129_i + 22,
-                    this.field_146320_D.func_146117_b(),
-                    10526880
-                );
+                this.field_146289_q.func_78279_b(I18n.func_135052_a(WorldType.field_77139_a[this.field_146331_K].func_151359_c()), this.field_146320_D.field_146128_h + 2, this.field_146320_D.field_146129_i + 22, this.field_146320_D.func_146117_b(), 10526880);
             }
         }
         else
         {
             this.func_73731_b(this.field_146289_q, I18n.func_135052_a("selectWorld.enterName"), this.field_146294_l / 2 - 100, 47, -6250336);
-            this.func_73731_b(
-                this.field_146289_q, I18n.func_135052_a("selectWorld.resultFolder") + " " + this.field_146336_i, this.field_146294_l / 2 - 100, 85, -6250336
-            );
+            this.func_73731_b(this.field_146289_q, I18n.func_135052_a("selectWorld.resultFolder") + " " + this.field_146336_i, this.field_146294_l / 2 - 100, 85, -6250336);
             this.field_146333_g.func_146194_f();
             this.func_73731_b(this.field_146289_q, this.field_146323_G, this.field_146294_l / 2 - 100, 137, -6250336);
             this.func_73731_b(this.field_146289_q, this.field_146328_H, this.field_146294_l / 2 - 100, 149, -6250336);
