--- before/net/minecraft/client/gui/GuiGameOver.java
+++ after/net/minecraft/client/gui/GuiGameOver.java
@@ -74,7 +74,7 @@
 
                 if (this.field_146297_k.field_71441_e.func_72912_H().func_76093_s())
                 {
-                    this.field_146297_k.func_147108_a(new GuiMainMenu());
+                    func_73878_a(true, 0); // Forge : Fix MC-30646
                 }
                 else
                 {
