--- before/net/minecraft/client/gui/GuiGameOver.java
+++ after/net/minecraft/client/gui/GuiGameOver.java
@@ -74,7 +74,7 @@
 
                 if (this.mc.world.getWorldInfo().isHardcoreModeEnabled())
                 {
-                    this.mc.displayGuiScreen(new GuiMainMenu());
+                    confirmClicked(true, 0); // Forge : Fix MC-30646
                 }
                 else
                 {
@@ -195,7 +195,7 @@
     public void updateScreen()
     {
         super.updateScreen();
-        this.enableButtonsTimer++;
+        ++this.enableButtonsTimer;
 
         if (this.enableButtonsTimer == 20)
         {
