--- before/net/minecraft/client/gui/GuiIngameMenu.java
+++ after/net/minecraft/client/gui/GuiIngameMenu.java
@@ -3,6 +3,7 @@
 import java.io.IOException;
 import net.minecraft.client.gui.achievement.GuiStats;
 import net.minecraft.client.gui.advancements.GuiScreenAdvancements;
+import net.minecraft.client.multiplayer.WorldClient;
 import net.minecraft.client.resources.I18n;
 import net.minecraft.realms.RealmsBridge;
 import net.minecraftforge.fml.relauncher.Side;
@@ -14,7 +15,6 @@
     private int saveStep;
     private int visibleTime;
 
-    @Override
     public void initGui()
     {
         this.saveStep = 0;
@@ -25,21 +25,18 @@
 
         if (!this.mc.isIntegratedServerRunning())
         {
-            this.buttonList.get(0).displayString = I18n.format("menu.disconnect");
+            (this.buttonList.get(0)).displayString = I18n.format("menu.disconnect");
         }
 
         this.buttonList.add(new GuiButton(4, this.width / 2 - 100, this.height / 4 + 24 + -16, I18n.format("menu.returnToGame")));
         this.buttonList.add(new GuiButton(0, this.width / 2 - 100, this.height / 4 + 96 + -16, 98, 20, I18n.format("menu.options")));
-        GuiButton guibutton = this.addButton(
-                                  new GuiButton(7, this.width / 2 + 2, this.height / 4 + 96 + -16, 98, 20, I18n.format("menu.shareToLan"))
-                              );
+        this.buttonList.add(new GuiButton(12, this.width / 2 + 2, this.height / 4 + 96 + i, 98, 20, I18n.format("fml.menu.modoptions")));
+        GuiButton guibutton = this.addButton(new GuiButton(7, this.width / 2 - 100, this.height / 4 + 72 + -16, 200, 20, I18n.format("menu.shareToLan", new Object[0])));
         guibutton.enabled = this.mc.isSingleplayer() && !this.mc.getIntegratedServer().getPublic();
-        this.buttonList
-        .add(new GuiButton(5, this.width / 2 - 100, this.height / 4 + 48 + -16, 98, 20, I18n.format("gui.advancements")));
+        this.buttonList.add(new GuiButton(5, this.width / 2 - 100, this.height / 4 + 48 + -16, 98, 20, I18n.format("gui.advancements")));
         this.buttonList.add(new GuiButton(6, this.width / 2 + 2, this.height / 4 + 48 + -16, 98, 20, I18n.format("gui.stats")));
     }
 
-    @Override
     protected void actionPerformed(GuiButton button) throws IOException
     {
         switch (button.id)
@@ -52,7 +49,7 @@
                 boolean flag1 = this.mc.isConnectedToRealms();
                 button.enabled = false;
                 this.mc.world.sendQuittingDisconnectingPacket();
-                this.mc.loadWorld(null);
+                this.mc.loadWorld((WorldClient)null);
 
                 if (flag)
                 {
@@ -73,28 +70,32 @@
             default:
                 break;
             case 4:
-                this.mc.displayGuiScreen(null);
+                this.mc.displayGuiScreen((GuiScreen)null);
                 this.mc.setIngameFocus();
                 break;
             case 5:
+                if (this.mc.player != null)
                 this.mc.displayGuiScreen(new GuiScreenAdvancements(this.mc.player.connection.getAdvancementManager()));
                 break;
             case 6:
+                if (this.mc.player != null)
                 this.mc.displayGuiScreen(new GuiStats(this, this.mc.player.getStatFileWriter()));
                 break;
             case 7:
                 this.mc.displayGuiScreen(new GuiShareToLan(this));
+                break;
+            case 12:
+                net.minecraftforge.fml.client.FMLClientHandler.instance().showInGameModOptions(this);
+                break;
         }
     }
 
-    @Override
     public void updateScreen()
     {
         super.updateScreen();
-        this.visibleTime++;
+        ++this.visibleTime;
     }
 
-    @Override
     public void drawScreen(int mouseX, int mouseY, float partialTicks)
     {
         this.drawDefaultBackground();
