--- before/net/minecraft/client/gui/GuiListWorldSelection.java
+++ after/net/minecraft/client/gui/GuiListWorldSelection.java
@@ -19,7 +19,7 @@
 {
     private static final Logger field_186797_u = LogManager.getLogger();
     private final GuiWorldSelection field_186798_v;
-    private final List<GuiListWorldSelectionEntry> field_186799_w = Lists.newArrayList();
+    private final List<GuiListWorldSelectionEntry> field_186799_w = Lists.<GuiListWorldSelectionEntry>newArrayList();
     private int field_186800_x = -1;
 
     public GuiListWorldSelection(GuiWorldSelection p_i46590_1_, Minecraft p_i46590_2_, int p_i46590_3_, int p_i46590_4_, int p_i46590_5_, int p_i46590_6_, int p_i46590_7_)
