--- before/net/minecraft/client/gui/GuiListWorldSelectionEntry.java
+++ after/net/minecraft/client/gui/GuiListWorldSelectionEntry.java
@@ -300,7 +300,7 @@
 
         if (this.client.getSaveLoader().canLoadWorld(this.worldSummary.getFileName()))
         {
-            this.client.launchIntegratedServer(this.worldSummary.getFileName(), this.worldSummary.getDisplayName(), null);
+            net.minecraftforge.fml.client.FMLClientHandler.instance().tryLoadExistingWorld(worldSelScreen, this.worldSummary);
         }
     }
 
