--- before/net/minecraft/client/gui/GuiListWorldSelectionEntry.java
+++ after/net/minecraft/client/gui/GuiListWorldSelectionEntry.java
@@ -14,6 +14,7 @@
 import net.minecraft.init.SoundEvents;
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.util.text.TextFormatting;
+import net.minecraft.world.WorldSettings;
 import net.minecraft.world.storage.ISaveFormat;
 import net.minecraft.world.storage.ISaveHandler;
 import net.minecraft.world.storage.WorldInfo;
@@ -58,18 +59,7 @@
         this.loadServerIcon();
     }
 
-    @Override
-    public void drawEntry(
-        int slotIndex,
-        int x,
-        int y,
-        int listWidth,
-        int slotHeight,
-        int mouseX,
-        int mouseY,
-        boolean isSelected,
-        float partialTicks
-    )
+    public void drawEntry(int slotIndex, int x, int y, int listWidth, int slotHeight, int mouseX, int mouseY, boolean isSelected, float partialTicks)
     {
         String s = this.worldSummary.getDisplayName();
         String s1 = this.worldSummary.getFileName() + " (" + DATE_FORMAT.format(new Date(this.worldSummary.getLastTimePlayed())) + ")";
@@ -119,14 +109,7 @@
 
         this.client.fontRenderer.drawString(s, x + 32 + 3, y + 1, 16777215);
         this.client.fontRenderer.drawString(s1, x + 32 + 3, y + this.client.fontRenderer.FONT_HEIGHT + 3, 8421504);
-        this.client
-        .fontRenderer
-        .drawString(
-            s2,
-            x + 32 + 3,
-            y + this.client.fontRenderer.FONT_HEIGHT + this.client.fontRenderer.FONT_HEIGHT + 3,
-            8421504
-        );
+        this.client.fontRenderer.drawString(s2, x + 32 + 3, y + this.client.fontRenderer.FONT_HEIGHT + this.client.fontRenderer.FONT_HEIGHT + 3, 8421504);
         GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
         this.client.getTextureManager().bindTexture(this.icon != null ? this.iconLocation : ICON_MISSING);
         GlStateManager.enableBlend();
@@ -151,14 +134,7 @@
 
                     if (j < 32)
                     {
-                        this.worldSelScreen
-                        .setVersionTooltip(
-                            TextFormatting.RED
-                            + I18n.format("selectWorld.tooltip.fromNewerVersion1")
-                            + "\n"
-                            + TextFormatting.RED
-                            + I18n.format("selectWorld.tooltip.fromNewerVersion2")
-                        );
+                        this.worldSelScreen.setVersionTooltip(TextFormatting.RED + I18n.format("selectWorld.tooltip.fromNewerVersion1") + "\n" + TextFormatting.RED + I18n.format("selectWorld.tooltip.fromNewerVersion2"));
                     }
                 }
                 else
@@ -167,14 +143,7 @@
 
                     if (j < 32)
                     {
-                        this.worldSelScreen
-                        .setVersionTooltip(
-                            TextFormatting.GOLD
-                            + I18n.format("selectWorld.tooltip.snapshot1")
-                            + "\n"
-                            + TextFormatting.GOLD
-                            + I18n.format("selectWorld.tooltip.snapshot2")
-                        );
+                        this.worldSelScreen.setVersionTooltip(TextFormatting.GOLD + I18n.format("selectWorld.tooltip.snapshot1") + "\n" + TextFormatting.GOLD + I18n.format("selectWorld.tooltip.snapshot2"));
                     }
                 }
             }
@@ -185,7 +154,6 @@
         }
     }
 
-    @Override
     public boolean mousePressed(int slotIndex, int mouseX, int mouseY, int mouseEvent, int relativeX, int relativeY)
     {
         this.containingListSel.selectWorld(slotIndex);
@@ -211,12 +179,8 @@
     {
         if (this.worldSummary.askToOpenWorld())
         {
-            this.client
-            .displayGuiScreen(
-                new GuiYesNo(
-                    new GuiYesNoCallback()
+            this.client.displayGuiScreen(new GuiYesNo(new GuiYesNoCallback()
             {
-                @Override
                 public void confirmClicked(boolean result, int id)
                 {
                     if (result)
@@ -228,14 +192,7 @@
                         GuiListWorldSelectionEntry.this.client.displayGuiScreen(GuiListWorldSelectionEntry.this.worldSelScreen);
                     }
                 }
-            },
-            I18n.format("selectWorld.versionQuestion"),
-            I18n.format("selectWorld.versionWarning", this.worldSummary.getVersionName()),
-            I18n.format("selectWorld.versionJoinButton"),
-            I18n.format("gui.cancel"),
-            0
-                )
-            );
+            }, I18n.format("selectWorld.versionQuestion"), I18n.format("selectWorld.versionWarning", this.worldSummary.getVersionName()), I18n.format("selectWorld.versionJoinButton"), I18n.format("gui.cancel"), 0));
         }
         else
         {
@@ -245,12 +202,8 @@
 
     public void deleteWorld()
     {
-        this.client
-        .displayGuiScreen(
-            new GuiYesNo(
-                new GuiYesNoCallback()
+        this.client.displayGuiScreen(new GuiYesNo(new GuiYesNoCallback()
         {
-            @Override
             public void confirmClicked(boolean result, int id)
             {
                 if (result)
@@ -264,14 +217,7 @@
 
                 GuiListWorldSelectionEntry.this.client.displayGuiScreen(GuiListWorldSelectionEntry.this.worldSelScreen);
             }
-        },
-        I18n.format("selectWorld.deleteQuestion"),
-        "'" + this.worldSummary.getDisplayName() + "' " + I18n.format("selectWorld.deleteWarning"),
-        I18n.format("selectWorld.deleteButton"),
-        I18n.format("gui.cancel"),
-        0
-            )
-        );
+        }, I18n.format("selectWorld.deleteQuestion"), "'" + this.worldSummary.getDisplayName() + "' " + I18n.format("selectWorld.deleteWarning"), I18n.format("selectWorld.deleteButton"), I18n.format("gui.cancel"), 0));
     }
 
     public void editWorld()
@@ -300,7 +246,7 @@
 
         if (this.client.getSaveLoader().canLoadWorld(this.worldSummary.getFileName()))
         {
-            this.client.launchIntegratedServer(this.worldSummary.getFileName(), this.worldSummary.getDisplayName(), null);
+            net.minecraftforge.fml.client.FMLClientHandler.instance().tryLoadExistingWorld(worldSelScreen, this.worldSummary);
         }
     }
 
@@ -341,12 +287,10 @@
         }
     }
 
-    @Override
     public void mouseReleased(int slotIndex, int x, int y, int mouseEvent, int relativeX, int relativeY)
     {
     }
 
-    @Override
     public void updatePosition(int slotIndex, int x, int y, float partialTicks)
     {
     }
