--- before/net/minecraft/client/gui/GuiLabel.java
+++ after/net/minecraft/client/gui/GuiLabel.java
@@ -35,7 +35,7 @@
         this.field_146174_h = p_i45540_4_;
         this.field_146167_a = p_i45540_5_;
         this.field_146161_f = p_i45540_6_;
-        this.field_146173_k = Lists.newArrayList();
+        this.field_146173_k = Lists.<String>newArrayList();
         this.field_146170_l = false;
         this.field_146171_m = false;
         this.field_146168_n = p_i45540_7_;
