--- before/net/minecraft/client/gui/GuiMerchant.java
+++ after/net/minecraft/client/gui/GuiMerchant.java
@@ -46,8 +46,8 @@
         super.func_73866_w_();
         int i = (this.field_146294_l - this.field_146999_f) / 2;
         int j = (this.field_146295_m - this.field_147000_g) / 2;
-        this.field_147043_x = this.func_189646_b(new GuiMerchant.MerchantButton(1, i + 120 + 27, j + 24 - 1, true));
-        this.field_147042_y = this.func_189646_b(new GuiMerchant.MerchantButton(2, i + 36 - 19, j + 24 - 1, false));
+        this.field_147043_x = (GuiMerchant.MerchantButton)this.func_189646_b(new GuiMerchant.MerchantButton(1, i + 120 + 27, j + 24 - 1, true));
+        this.field_147042_y = (GuiMerchant.MerchantButton)this.func_189646_b(new GuiMerchant.MerchantButton(2, i + 36 - 19, j + 24 - 1, false));
         this.field_147043_x.field_146124_l = false;
         this.field_147042_y.field_146124_l = false;
     }
@@ -126,7 +126,7 @@
                 return;
             }
 
-            MerchantRecipe merchantrecipe = merchantrecipelist.get(k);
+            MerchantRecipe merchantrecipe = (MerchantRecipe)merchantrecipelist.get(k);
 
             if (merchantrecipe.func_82784_g())
             {
@@ -150,7 +150,7 @@
             int i = (this.field_146294_l - this.field_146999_f) / 2;
             int j = (this.field_146295_m - this.field_147000_g) / 2;
             int k = this.field_147041_z;
-            MerchantRecipe merchantrecipe = merchantrecipelist.get(k);
+            MerchantRecipe merchantrecipe = (MerchantRecipe)merchantrecipelist.get(k);
             ItemStack itemstack = merchantrecipe.func_77394_a();
             ItemStack itemstack1 = merchantrecipe.func_77396_b();
             ItemStack itemstack2 = merchantrecipe.func_77397_d();
