--- before/net/minecraft/client/gui/GuiMainMenu.java
+++ after/net/minecraft/client/gui/GuiMainMenu.java
@@ -3,6 +3,7 @@
 import com.google.common.collect.Lists;
 import com.google.common.util.concurrent.Runnables;
 import java.io.BufferedReader;
+import java.io.Closeable;
 import java.io.IOException;
 import java.io.InputStreamReader;
 import java.net.URI;
@@ -57,35 +58,30 @@
     private int field_92020_v;
     private int field_92019_w;
     private String field_92025_p;
-    private String field_146972_A = field_96138_a;
+    private String field_146972_A;
     private String field_104024_v;
     private static final ResourceLocation field_110353_x = new ResourceLocation("texts/splashes.txt");
     private static final ResourceLocation field_110352_y = new ResourceLocation("textures/gui/title/minecraft.png");
     private static final ResourceLocation field_194400_H = new ResourceLocation("textures/gui/title/edition.png");
-    private static final ResourceLocation[] field_73978_o = new ResourceLocation[]
-        {
-            new ResourceLocation("textures/gui/title/background/panorama_0.png"),
-            new ResourceLocation("textures/gui/title/background/panorama_1.png"),
-            new ResourceLocation("textures/gui/title/background/panorama_2.png"),
-            new ResourceLocation("textures/gui/title/background/panorama_3.png"),
-            new ResourceLocation("textures/gui/title/background/panorama_4.png"),
-            new ResourceLocation("textures/gui/title/background/panorama_5.png")
-        };
+    private static final ResourceLocation[] field_73978_o = new ResourceLocation[] {new ResourceLocation("textures/gui/title/background/panorama_0.png"), new ResourceLocation("textures/gui/title/background/panorama_1.png"), new ResourceLocation("textures/gui/title/background/panorama_2.png"), new ResourceLocation("textures/gui/title/background/panorama_3.png"), new ResourceLocation("textures/gui/title/background/panorama_4.png"), new ResourceLocation("textures/gui/title/background/panorama_5.png")};
     private ResourceLocation field_110351_G;
     private GuiButton field_175372_K;
     private boolean field_183502_L;
     private GuiScreen field_183503_M;
     private int field_193978_M;
     private int field_193979_N;
+    private GuiButton modButton;
+    private net.minecraftforge.client.gui.NotificationModUpdateScreen modUpdateNotification;
 
     public GuiMainMenu()
     {
+        this.field_146972_A = field_96138_a;
         this.field_73975_c = "missingno";
         IResource iresource = null;
 
         try
         {
-            List<String> list = Lists.newArrayList();
+            List<String> list = Lists.<String>newArrayList();
             iresource = Minecraft.func_71410_x().func_110442_L().func_110536_a(field_110353_x);
             BufferedReader bufferedreader = new BufferedReader(new InputStreamReader(iresource.func_110527_b(), StandardCharsets.UTF_8));
             String s;
@@ -102,19 +98,24 @@
 
             if (!list.isEmpty())
             {
-                do
+                while (true)
                 {
                     this.field_73975_c = list.get(field_175374_h.nextInt(list.size()));
+
+                    if (this.field_73975_c.hashCode() != 125780783)
+                    {
+                        break;
+                    }
                 }
-                while (this.field_73975_c.hashCode() == 125780783);
             }
         }
-        catch (IOException ioexception)
+        catch (IOException var8)
         {
+            ;
         }
         finally
         {
-            IOUtils.closeQuietly(iresource);
+            IOUtils.closeQuietly((Closeable)iresource);
         }
 
         this.field_73974_b = field_175374_h.nextFloat();
@@ -130,30 +131,18 @@
 
     private boolean func_183501_a()
     {
-        return Minecraft.func_71410_x().field_71474_y.func_74308_b(GameSettings.Options.REALMS_NOTIFICATIONS) && this.field_183503_M != null;
-    }
-
-    @Override
-    public void func_73876_c()
-    {
-        if (this.func_183501_a())
-        {
-            this.field_183503_M.func_73876_c();
-        }
-    }
-
-    @Override
+        return false;
+    }
+
     public boolean func_73868_f()
     {
         return false;
     }
 
-    @Override
     protected void func_73869_a(char p_73869_1_, int p_73869_2_) throws IOException
     {
     }
 
-    @Override
     public void func_73866_w_()
     {
         this.field_73977_n = new DynamicTexture(256, 256);
@@ -198,40 +187,27 @@
             this.field_92024_r = this.field_146289_q.func_78256_a(this.field_146972_A);
             int k = Math.max(this.field_92023_s, this.field_92024_r);
             this.field_92022_t = (this.field_146294_l - k) / 2;
-            this.field_92021_u = this.field_146292_n.get(0).field_146129_i - 24;
+            this.field_92021_u = (this.field_146292_n.get(0)).field_146129_i - 24;
             this.field_92020_v = this.field_92022_t + k;
             this.field_92019_w = this.field_92021_u + 24;
         }
 
         this.field_146297_k.func_181537_a(false);
 
-        if (Minecraft.func_71410_x().field_71474_y.func_74308_b(GameSettings.Options.REALMS_NOTIFICATIONS) && !this.field_183502_L)
-        {
-            RealmsBridge realmsbridge = new RealmsBridge();
-            this.field_183503_M = realmsbridge.getNotificationScreen(this);
-            this.field_183502_L = true;
-        }
-
-        if (this.func_183501_a())
-        {
-            this.field_183503_M.func_183500_a(this.field_146294_l, this.field_146295_m);
-            this.field_183503_M.func_73866_w_();
-        }
+        modUpdateNotification = net.minecraftforge.client.gui.NotificationModUpdateScreen.init(this, modButton);
     }
 
     private void func_73969_a(int p_73969_1_, int p_73969_2_)
     {
         this.field_146292_n.add(new GuiButton(1, this.field_146294_l / 2 - 100, p_73969_1_, I18n.func_135052_a("menu.singleplayer")));
         this.field_146292_n.add(new GuiButton(2, this.field_146294_l / 2 - 100, p_73969_1_ + p_73969_2_ * 1, I18n.func_135052_a("menu.multiplayer")));
-        this.field_175372_K = this.func_189646_b(new GuiButton(14, this.field_146294_l / 2 - 100, p_73969_1_ + p_73969_2_ * 2, I18n.func_135052_a("menu.online")));
+        this.field_146292_n.add(modButton = new GuiButton(6, this.field_146294_l / 2 - 100, p_73969_1_ + p_73969_2_ * 2, I18n.func_135052_a("fml.menu.mods")));
     }
 
     private void func_73972_b(int p_73972_1_, int p_73972_2_)
     {
         this.field_146292_n.add(new GuiButton(11, this.field_146294_l / 2 - 100, p_73972_1_, I18n.func_135052_a("menu.playdemo")));
-        this.field_73973_d = this.func_189646_b(
-                                 new GuiButton(12, this.field_146294_l / 2 - 100, p_73972_1_ + p_73972_2_ * 1, I18n.func_135052_a("menu.resetdemo"))
-                             );
+        this.field_73973_d = this.func_189646_b(new GuiButton(12, this.field_146294_l / 2 - 100, p_73972_1_ + p_73972_2_ * 1, I18n.func_135052_a("menu.resetdemo")));
         ISaveFormat isaveformat = this.field_146297_k.func_71359_d();
         WorldInfo worldinfo = isaveformat.func_75803_c("Demo_World");
 
@@ -241,7 +217,6 @@
         }
     }
 
-    @Override
     protected void func_146284_a(GuiButton p_146284_1_) throws IOException
     {
         if (p_146284_1_.field_146127_k == 0)
@@ -264,16 +239,16 @@
             this.field_146297_k.func_147108_a(new GuiMultiplayer(this));
         }
 
-        if (p_146284_1_.field_146127_k == 14 && this.field_175372_K.field_146125_m)
-        {
-            this.func_140005_i();
-        }
-
         if (p_146284_1_.field_146127_k == 4)
         {
             this.field_146297_k.func_71400_g();
         }
 
+        if (p_146284_1_.field_146127_k == 6)
+        {
+            this.field_146297_k.func_147108_a(new net.minecraftforge.fml.client.GuiModList(this));
+        }
+
         if (p_146284_1_.field_146127_k == 11)
         {
             this.field_146297_k.func_71371_a("Demo_World", "Demo_World", WorldServerDemo.field_73071_a);
@@ -286,28 +261,15 @@
 
             if (worldinfo != null)
             {
-                this.field_146297_k
-                .func_147108_a(
-                    new GuiYesNo(
-                        this,
-                        I18n.func_135052_a("selectWorld.deleteQuestion"),
-                        "'" + worldinfo.func_76065_j() + "' " + I18n.func_135052_a("selectWorld.deleteWarning"),
-                        I18n.func_135052_a("selectWorld.deleteButton"),
-                        I18n.func_135052_a("gui.cancel"),
-                        12
-                    )
-                );
+                this.field_146297_k.func_147108_a(new GuiYesNo(this, I18n.func_135052_a("selectWorld.deleteQuestion"), "'" + worldinfo.func_76065_j() + "' " + I18n.func_135052_a("selectWorld.deleteWarning"), I18n.func_135052_a("selectWorld.deleteButton"), I18n.func_135052_a("gui.cancel"), 12));
             }
         }
     }
 
     private void func_140005_i()
     {
-        RealmsBridge realmsbridge = new RealmsBridge();
-        realmsbridge.switchToRealms(this);
     }
 
-    @Override
     public void func_73878_a(boolean p_73878_1_, int p_73878_2_)
     {
         if (p_73878_1_ && p_73878_2_ == 12)
@@ -328,7 +290,7 @@
                 try
                 {
                     Class<?> oclass = Class.forName("java.awt.Desktop");
-                    Object object = oclass.getMethod("getDesktop").invoke(null);
+                    Object object = oclass.getMethod("getDesktop").invoke((Object)null);
                     oclass.getMethod("browse", URI.class).invoke(object, new URI(this.field_104024_v));
                 }
                 catch (Throwable throwable)
@@ -359,9 +321,7 @@
         GlStateManager.func_179118_c();
         GlStateManager.func_179129_p();
         GlStateManager.func_179132_a(false);
-        GlStateManager.func_187428_a(
-            GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO
-        );
+        GlStateManager.func_187428_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
         int i = 8;
 
         for (int j = 0; j < 64; ++j)
@@ -407,10 +367,10 @@
                 bufferbuilder.func_181668_a(7, DefaultVertexFormats.field_181709_i);
                 int l = 255 / (j + 1);
                 float f3 = 0.0F;
-                bufferbuilder.func_181662_b(-1.0, -1.0, 1.0).func_187315_a(0.0, 0.0).func_181669_b(255, 255, 255, l).func_181675_d();
-                bufferbuilder.func_181662_b(1.0, -1.0, 1.0).func_187315_a(1.0, 0.0).func_181669_b(255, 255, 255, l).func_181675_d();
-                bufferbuilder.func_181662_b(1.0, 1.0, 1.0).func_187315_a(1.0, 1.0).func_181669_b(255, 255, 255, l).func_181675_d();
-                bufferbuilder.func_181662_b(-1.0, 1.0, 1.0).func_187315_a(0.0, 1.0).func_181669_b(255, 255, 255, l).func_181675_d();
+                bufferbuilder.func_181662_b(-1.0D, -1.0D, 1.0D).func_187315_a(0.0D, 0.0D).func_181669_b(255, 255, 255, l).func_181675_d();
+                bufferbuilder.func_181662_b(1.0D, -1.0D, 1.0D).func_187315_a(1.0D, 0.0D).func_181669_b(255, 255, 255, l).func_181675_d();
+                bufferbuilder.func_181662_b(1.0D, 1.0D, 1.0D).func_187315_a(1.0D, 1.0D).func_181669_b(255, 255, 255, l).func_181675_d();
+                bufferbuilder.func_181662_b(-1.0D, 1.0D, 1.0D).func_187315_a(0.0D, 1.0D).func_181669_b(255, 255, 255, l).func_181675_d();
                 tessellator.func_78381_a();
                 GlStateManager.func_179121_F();
             }
@@ -419,7 +379,7 @@
             GlStateManager.func_179135_a(true, true, true, false);
         }
 
-        bufferbuilder.func_178969_c(0.0, 0.0, 0.0);
+        bufferbuilder.func_178969_c(0.0D, 0.0D, 0.0D);
         GlStateManager.func_179135_a(true, true, true, true);
         GlStateManager.func_179128_n(5889);
         GlStateManager.func_179121_F();
@@ -437,9 +397,7 @@
         GlStateManager.func_187421_b(3553, 10240, 9729);
         GlStateManager.func_187443_a(3553, 0, 0, 0, 0, 0, 256, 256);
         GlStateManager.func_179147_l();
-        GlStateManager.func_187428_a(
-            GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO
-        );
+        GlStateManager.func_187428_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
         GlStateManager.func_179135_a(true, true, true, false);
         Tessellator tessellator = Tessellator.func_178181_a();
         BufferBuilder bufferbuilder = tessellator.func_178180_c();
@@ -453,22 +411,10 @@
             int k = this.field_146294_l;
             int l = this.field_146295_m;
             float f1 = (float)(j - 1) / 256.0F;
-            bufferbuilder.func_181662_b((double)k, (double)l, (double)this.field_73735_i)
-            .func_187315_a((double)(0.0F + f1), 1.0)
-            .func_181666_a(1.0F, 1.0F, 1.0F, f)
-            .func_181675_d();
-            bufferbuilder.func_181662_b((double)k, 0.0, (double)this.field_73735_i)
-            .func_187315_a((double)(1.0F + f1), 1.0)
-            .func_181666_a(1.0F, 1.0F, 1.0F, f)
-            .func_181675_d();
-            bufferbuilder.func_181662_b(0.0, 0.0, (double)this.field_73735_i)
-            .func_187315_a((double)(1.0F + f1), 0.0)
-            .func_181666_a(1.0F, 1.0F, 1.0F, f)
-            .func_181675_d();
-            bufferbuilder.func_181662_b(0.0, (double)l, (double)this.field_73735_i)
-            .func_187315_a((double)(0.0F + f1), 0.0)
-            .func_181666_a(1.0F, 1.0F, 1.0F, f)
-            .func_181675_d();
+            bufferbuilder.func_181662_b((double)k, (double)l, (double)this.field_73735_i).func_187315_a((double)(0.0F + f1), 1.0D).func_181666_a(1.0F, 1.0F, 1.0F, f).func_181675_d();
+            bufferbuilder.func_181662_b((double)k, 0.0D, (double)this.field_73735_i).func_187315_a((double)(1.0F + f1), 1.0D).func_181666_a(1.0F, 1.0F, 1.0F, f).func_181675_d();
+            bufferbuilder.func_181662_b(0.0D, 0.0D, (double)this.field_73735_i).func_187315_a((double)(1.0F + f1), 0.0D).func_181666_a(1.0F, 1.0F, 1.0F, f).func_181675_d();
+            bufferbuilder.func_181662_b(0.0D, (double)l, (double)this.field_73735_i).func_187315_a((double)(0.0F + f1), 0.0D).func_181666_a(1.0F, 1.0F, 1.0F, f).func_181675_d();
         }
 
         tessellator.func_78381_a();
@@ -498,26 +444,13 @@
         Tessellator tessellator = Tessellator.func_178181_a();
         BufferBuilder bufferbuilder = tessellator.func_178180_c();
         bufferbuilder.func_181668_a(7, DefaultVertexFormats.field_181709_i);
-        bufferbuilder.func_181662_b(0.0, (double)j, (double)this.field_73735_i)
-        .func_187315_a((double)(0.5F - f1), (double)(0.5F + f2))
-        .func_181666_a(1.0F, 1.0F, 1.0F, 1.0F)
-        .func_181675_d();
-        bufferbuilder.func_181662_b((double)i, (double)j, (double)this.field_73735_i)
-        .func_187315_a((double)(0.5F - f1), (double)(0.5F - f2))
-        .func_181666_a(1.0F, 1.0F, 1.0F, 1.0F)
-        .func_181675_d();
-        bufferbuilder.func_181662_b((double)i, 0.0, (double)this.field_73735_i)
-        .func_187315_a((double)(0.5F + f1), (double)(0.5F - f2))
-        .func_181666_a(1.0F, 1.0F, 1.0F, 1.0F)
-        .func_181675_d();
-        bufferbuilder.func_181662_b(0.0, 0.0, (double)this.field_73735_i)
-        .func_187315_a((double)(0.5F + f1), (double)(0.5F + f2))
-        .func_181666_a(1.0F, 1.0F, 1.0F, 1.0F)
-        .func_181675_d();
+        bufferbuilder.func_181662_b(0.0D, (double)j, (double)this.field_73735_i).func_187315_a((double)(0.5F - f1), (double)(0.5F + f2)).func_181666_a(1.0F, 1.0F, 1.0F, 1.0F).func_181675_d();
+        bufferbuilder.func_181662_b((double)i, (double)j, (double)this.field_73735_i).func_187315_a((double)(0.5F - f1), (double)(0.5F - f2)).func_181666_a(1.0F, 1.0F, 1.0F, 1.0F).func_181675_d();
+        bufferbuilder.func_181662_b((double)i, 0.0D, (double)this.field_73735_i).func_187315_a((double)(0.5F + f1), (double)(0.5F - f2)).func_181666_a(1.0F, 1.0F, 1.0F, 1.0F).func_181675_d();
+        bufferbuilder.func_181662_b(0.0D, 0.0D, (double)this.field_73735_i).func_187315_a((double)(0.5F + f1), (double)(0.5F + f2)).func_181666_a(1.0F, 1.0F, 1.0F, 1.0F).func_181675_d();
         tessellator.func_78381_a();
     }
 
-    @Override
     public void func_73863_a(int p_73863_1_, int p_73863_2_, float p_73863_3_)
     {
         this.field_73979_m += p_73863_3_;
@@ -532,7 +465,7 @@
         this.field_146297_k.func_110434_K().func_110577_a(field_110352_y);
         GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
 
-        if ((double)this.field_73974_b < 1.0E-4)
+        if ((double)this.field_73974_b < 1.0E-4D)
         {
             this.func_73729_b(j + 0, 30, 0, 0, 99, 44);
             this.func_73729_b(j + 99, 30, 129, 0, 27, 44);
@@ -548,10 +481,13 @@
 
         this.field_146297_k.func_110434_K().func_110577_a(field_194400_H);
         func_146110_a(j + 88, 67, 0.0F, 0.0F, 98, 14, 128.0F, 16.0F);
+
+        this.field_73975_c = net.minecraftforge.client.ForgeHooksClient.renderMainMenu(this, this.field_146289_q, this.field_146294_l, this.field_146295_m, this.field_73975_c);
+
         GlStateManager.func_179094_E();
         GlStateManager.func_179109_b((float)(this.field_146294_l / 2 + 90), 70.0F, 0.0F);
         GlStateManager.func_179114_b(-20.0F, 0.0F, 0.0F, 1.0F);
-        float f = 1.8F - MathHelper.func_76135_e(MathHelper.func_76126_a((float)(Minecraft.func_71386_F() % 1000L) / 1000.0F * (float)(Math.PI * 2)) * 0.1F);
+        float f = 1.8F - MathHelper.func_76135_e(MathHelper.func_76126_a((float)(Minecraft.func_71386_F() % 1000L) / 1000.0F * ((float)Math.PI * 2F)) * 0.1F);
         f = f * 100.0F / (float)(this.field_146289_q.func_78256_a(this.field_73975_c) + 32);
         GlStateManager.func_179152_a(f, f, f);
         this.func_73732_a(this.field_146289_q, this.field_73975_c, 0, -8, -256);
@@ -567,14 +503,19 @@
             s = s + ("release".equalsIgnoreCase(this.field_146297_k.func_184123_d()) ? "" : "/" + this.field_146297_k.func_184123_d());
         }
 
-        this.func_73731_b(this.field_146289_q, s, 2, this.field_146295_m - 10, -1);
+        java.util.List<String> brandings = com.google.common.collect.Lists.reverse(net.minecraftforge.fml.common.FMLCommonHandler.instance().getBrandings(true));
+        for (int brdline = 0; brdline < brandings.size(); brdline++)
+        {
+            String brd = brandings.get(brdline);
+            if (!com.google.common.base.Strings.isNullOrEmpty(brd))
+            {
+                this.func_73731_b(this.field_146289_q, brd, 2, this.field_146295_m - ( 10 + brdline * (this.field_146289_q.field_78288_b + 1)), 16777215);
+            }
+        }
+
         this.func_73731_b(this.field_146289_q, "Copyright Mojang AB. Do not distribute!", this.field_193979_N, this.field_146295_m - 10, -1);
 
-        if (p_73863_1_ > this.field_193979_N
-                && p_73863_1_ < this.field_193979_N + this.field_193978_M
-                && p_73863_2_ > this.field_146295_m - 10
-                && p_73863_2_ < this.field_146295_m
-                && Mouse.isInsideWindow())
+        if (p_73863_1_ > this.field_193979_N && p_73863_1_ < this.field_193979_N + this.field_193978_M && p_73863_2_ > this.field_146295_m - 10 && p_73863_2_ < this.field_146295_m && Mouse.isInsideWindow())
         {
             func_73734_a(this.field_193979_N, this.field_146295_m - 1, this.field_193979_N + this.field_193978_M, this.field_146295_m, -1);
         }
@@ -583,32 +524,21 @@
         {
             func_73734_a(this.field_92022_t - 2, this.field_92021_u - 2, this.field_92020_v + 2, this.field_92019_w - 1, 1428160512);
             this.func_73731_b(this.field_146289_q, this.field_92025_p, this.field_92022_t, this.field_92021_u, -1);
-            this.func_73731_b(
-                this.field_146289_q, this.field_146972_A, (this.field_146294_l - this.field_92024_r) / 2, this.field_146292_n.get(0).field_146129_i - 12, -1
-            );
+            this.func_73731_b(this.field_146289_q, this.field_146972_A, (this.field_146294_l - this.field_92024_r) / 2, (this.field_146292_n.get(0)).field_146129_i - 12, -1);
         }
 
         super.func_73863_a(p_73863_1_, p_73863_2_, p_73863_3_);
 
-        if (this.func_183501_a())
-        {
-            this.field_183503_M.func_73863_a(p_73863_1_, p_73863_2_, p_73863_3_);
-        }
+        modUpdateNotification.func_73863_a(p_73863_1_, p_73863_2_, p_73863_3_);
     }
 
-    @Override
     protected void func_73864_a(int p_73864_1_, int p_73864_2_, int p_73864_3_) throws IOException
     {
         super.func_73864_a(p_73864_1_, p_73864_2_, p_73864_3_);
 
         synchronized (this.field_104025_t)
         {
-            if (!this.field_92025_p.isEmpty()
-                    && !StringUtils.func_151246_b(this.field_104024_v)
-                    && p_73864_1_ >= this.field_92022_t
-                    && p_73864_1_ <= this.field_92020_v
-                    && p_73864_2_ >= this.field_92021_u
-                    && p_73864_2_ <= this.field_92019_w)
+            if (!this.field_92025_p.isEmpty() && !StringUtils.func_151246_b(this.field_104024_v) && p_73864_1_ >= this.field_92022_t && p_73864_1_ <= this.field_92020_v && p_73864_2_ >= this.field_92021_u && p_73864_2_ <= this.field_92019_w)
             {
                 GuiConfirmOpenLink guiconfirmopenlink = new GuiConfirmOpenLink(this, this.field_104024_v, 13, true);
                 guiconfirmopenlink.func_146358_g();
@@ -616,26 +546,10 @@
             }
         }
 
-        if (this.func_183501_a())
-        {
-            this.field_183503_M.func_73864_a(p_73864_1_, p_73864_2_, p_73864_3_);
-        }
-
-        if (p_73864_1_ > this.field_193979_N
-                && p_73864_1_ < this.field_193979_N + this.field_193978_M
-                && p_73864_2_ > this.field_146295_m - 10
-                && p_73864_2_ < this.field_146295_m)
+        if (p_73864_1_ > this.field_193979_N && p_73864_1_ < this.field_193979_N + this.field_193978_M && p_73864_2_ > this.field_146295_m - 10 && p_73864_2_ < this.field_146295_m)
         {
             this.field_146297_k.func_147108_a(new GuiWinGame(false, Runnables.doNothing()));
         }
     }
 
-    @Override
-    public void func_146281_b()
-    {
-        if (this.field_183503_M != null)
-        {
-            this.field_183503_M.func_146281_b();
-        }
-    }
 }
