--- before/net/minecraft/client/gui/GuiMultiplayer.java
+++ after/net/minecraft/client/gui/GuiMultiplayer.java
@@ -15,7 +15,7 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import org.lwjgl.input.Keyboard;
+import org.lwjglx.input.Keyboard;
 
 @SideOnly(Side.CLIENT)
 public class GuiMultiplayer extends GuiScreen
@@ -41,6 +41,7 @@
     public GuiMultiplayer(GuiScreen p_i1040_1_)
     {
         this.field_146798_g = p_i1040_1_;
+        net.minecraftforge.fml.client.FMLClientHandler.instance().setupServerList();
     }
 
     public void func_73866_w_()
@@ -375,7 +376,7 @@
 
     private void func_146791_a(ServerData p_146791_1_)
     {
-        this.field_146297_k.func_147108_a(new GuiConnecting(this, this.field_146297_k, p_146791_1_));
+        net.minecraftforge.fml.client.FMLClientHandler.instance().connectToServer(this, p_146791_1_);
     }
 
     public void func_146790_a(int p_146790_1_)
