--- before/net/minecraft/client/gui/GuiNewChat.java
+++ after/net/minecraft/client/gui/GuiNewChat.java
@@ -20,9 +20,9 @@
 {
     private static final Logger field_146249_a = LogManager.getLogger();
     private final Minecraft field_146247_f;
-    private final List<String> field_146248_g = Lists.newArrayList();
-    private final List<ChatLine> field_146252_h = Lists.newArrayList();
-    private final List<ChatLine> field_146253_i = Lists.newArrayList();
+    private final List<String> field_146248_g = Lists.<String>newArrayList();
+    private final List<ChatLine> field_146252_h = Lists.<ChatLine>newArrayList();
+    private final List<ChatLine> field_146253_i = Lists.<ChatLine>newArrayList();
     private int field_146250_j;
     private boolean field_146251_k;
 
@@ -67,9 +67,9 @@
                         {
                             double d0 = (double)j1 / 200.0D;
                             d0 = 1.0D - d0;
-                            d0 *= 10.0D;
+                            d0 = d0 * 10.0D;
                             d0 = MathHelper.func_151237_a(d0, 0.0D, 1.0D);
-                            d0 *= d0;
+                            d0 = d0 * d0;
                             int l1 = (int)(255.0D * d0);
 
                             if (flag)
@@ -197,7 +197,7 @@
 
     public void func_146239_a(String p_146239_1_)
     {
-        if (this.field_146248_g.isEmpty() || !this.field_146248_g.get(this.field_146248_g.size() - 1).equals(p_146239_1_))
+        if (this.field_146248_g.isEmpty() || !((String)this.field_146248_g.get(this.field_146248_g.size() - 1)).equals(p_146239_1_))
         {
             this.field_146248_g.add(p_146239_1_);
         }
