--- before/net/minecraft/client/gui/GuiOptionSlider.java
+++ after/net/minecraft/client/gui/GuiOptionSlider.java
@@ -10,7 +10,7 @@
 @SideOnly(Side.CLIENT)
 public class GuiOptionSlider extends GuiButton
 {
-    private float field_146134_p = 1.0F;
+    private float field_146134_p;
     public boolean field_146135_o;
     private final GameSettings.Options field_146133_q;
     private final float field_146132_r;
@@ -24,6 +24,7 @@
     public GuiOptionSlider(int p_i45017_1_, int p_i45017_2_, int p_i45017_3_, GameSettings.Options p_i45017_4_, float p_i45017_5_, float p_i45017_6_)
     {
         super(p_i45017_1_, p_i45017_2_, p_i45017_3_, 150, 20, "");
+        this.field_146134_p = 1.0F;
         this.field_146133_q = p_i45017_4_;
         this.field_146132_r = p_i45017_5_;
         this.field_146131_s = p_i45017_6_;
