--- before/net/minecraft/client/gui/GuiPlayerTabOverlay.java
+++ after/net/minecraft/client/gui/GuiPlayerTabOverlay.java
@@ -58,7 +58,7 @@
     public void func_175249_a(int p_175249_1_, Scoreboard p_175249_2_, @Nullable ScoreObjective p_175249_3_)
     {
         NetHandlerPlayClient nethandlerplayclient = this.field_175250_f.field_71439_g.field_71174_a;
-        List<NetworkPlayerInfo> list = field_175252_a.sortedCopy(nethandlerplayclient.func_175106_d());
+        List<NetworkPlayerInfo> list = field_175252_a.<NetworkPlayerInfo>sortedCopy(nethandlerplayclient.func_175106_d());
         int i = 0;
         int j = 0;
 
@@ -211,7 +211,7 @@
 
         if (list2 != null)
         {
-            k1 += i4 * 9 + 1;
+            k1 = k1 + i4 * 9 + 1;
             func_73734_a(p_175249_1_ / 2 - l1 / 2 - 1, k1 - 1, p_175249_1_ / 2 + l1 / 2 + 1, k1 + list2.size() * this.field_175250_f.field_71466_p.field_78288_b, Integer.MIN_VALUE);
 
             for (String s3 : list2)
