--- before/net/minecraft/client/gui/GuiPageButtonList.java
+++ after/net/minecraft/client/gui/GuiPageButtonList.java
@@ -14,9 +14,9 @@
 @SideOnly(Side.CLIENT)
 public class GuiPageButtonList extends GuiListExtended
 {
-    private final List<GuiPageButtonList.GuiEntry> field_178074_u = Lists.newArrayList();
-    private final IntHashMap<Gui> field_178073_v = new IntHashMap<>();
-    private final List<GuiTextField> field_178072_w = Lists.newArrayList();
+    private final List<GuiPageButtonList.GuiEntry> field_178074_u = Lists.<GuiPageButtonList.GuiEntry>newArrayList();
+    private final IntHashMap<Gui> field_178073_v = new IntHashMap<Gui>();
+    private final List<GuiTextField> field_178072_w = Lists.<GuiTextField>newArrayList();
     private final GuiPageButtonList.GuiListEntry[][] field_178078_x;
     private int field_178077_y;
     private final GuiPageButtonList.GuiResponder field_178076_z;
@@ -77,7 +77,7 @@
             GuiPageButtonList.GuiListEntry guipagebuttonlist$guilistentry = this.field_178078_x[this.field_178077_y][i];
             GuiPageButtonList.GuiListEntry guipagebuttonlist$guilistentry1 = i < this.field_178078_x[this.field_178077_y].length - 1 ? this.field_178078_x[this.field_178077_y][i + 1] : null;
             Gui gui = this.field_178073_v.func_76041_a(guipagebuttonlist$guilistentry.func_178935_b());
-            Gui gui1 = guipagebuttonlist$guilistentry1 != null ? this.field_178073_v.func_76041_a(guipagebuttonlist$guilistentry1.func_178935_b()) : null;
+            Gui gui1 = guipagebuttonlist$guilistentry1 != null ? (Gui)this.field_178073_v.func_76041_a(guipagebuttonlist$guilistentry1.func_178935_b()) : null;
             GuiPageButtonList.GuiEntry guipagebuttonlist$guientry = new GuiPageButtonList.GuiEntry(gui, gui1);
             this.field_178074_u.add(guipagebuttonlist$guientry);
         }
@@ -378,7 +378,7 @@
             public EditBoxEntry(int p_i45534_1_, String p_i45534_2_, boolean p_i45534_3_, Predicate<String> p_i45534_4_)
             {
                 super(p_i45534_1_, p_i45534_2_, p_i45534_3_);
-                this.field_178951_a = MoreObjects.firstNonNull(p_i45534_4_, Predicates.alwaysTrue());
+                this.field_178951_a = (Predicate)MoreObjects.firstNonNull(p_i45534_4_, Predicates.alwaysTrue());
             }
 
             public Predicate<String> func_178950_a()
