--- before/net/minecraft/client/gui/GuiSpectator.java
+++ after/net/minecraft/client/gui/GuiSpectator.java
@@ -109,7 +109,7 @@
             GlStateManager.func_179131_c(f, f, f, p_175266_4_);
             p_175266_5_.func_178663_a(f, i);
             GlStateManager.func_179121_F();
-            String s = String.valueOf((Object)GameSettings.func_74298_c(this.field_175268_g.field_71474_y.field_151456_ac[p_175266_1_].func_151463_i()));
+            String s = String.valueOf(this.field_175268_g.field_71474_y.field_151456_ac[p_175266_1_].getDisplayName());
 
             if (i > 3 && p_175266_5_.func_178662_A_())
             {
@@ -158,6 +158,7 @@
 
         for (i = this.field_175271_i.func_178648_e() + p_175259_1_; i >= 0 && i <= 8 && (this.field_175271_i.func_178643_a(i) == SpectatorMenu.field_178657_a || !this.field_175271_i.func_178643_a(i).func_178662_A_()); i += p_175259_1_)
         {
+            ;
         }
 
         if (i >= 0 && i <= 8)
