--- before/net/minecraft/client/gui/GuiSleepMP.java
+++ after/net/minecraft/client/gui/GuiSleepMP.java
@@ -10,14 +10,12 @@
 @SideOnly(Side.CLIENT)
 public class GuiSleepMP extends GuiChat
 {
-    @Override
     public void func_73866_w_()
     {
         super.func_73866_w_();
         this.field_146292_n.add(new GuiButton(1, this.field_146294_l / 2 - 100, this.field_146295_m - 40, I18n.func_135052_a("multiplayer.stopSleeping")));
     }
 
-    @Override
     protected void func_73869_a(char p_73869_1_, int p_73869_2_) throws IOException
     {
         if (p_73869_2_ == 1)
@@ -34,7 +32,7 @@
 
             if (!s.isEmpty())
             {
-                this.field_146297_k.field_71439_g.func_71165_d(s);
+                this.func_175275_f(s); // Forge: fix vanilla not adding messages to the sent list while sleeping
             }
 
             this.field_146415_a.func_146180_a("");
@@ -42,7 +40,6 @@
         }
     }
 
-    @Override
     protected void func_146284_a(GuiButton p_146284_1_) throws IOException
     {
         if (p_146284_1_.field_146127_k == 1)
