--- before/net/minecraft/client/gui/GuiSpectator.java
+++ after/net/minecraft/client/gui/GuiSpectator.java
@@ -74,9 +74,7 @@
     {
         GlStateManager.enableRescaleNormal();
         GlStateManager.enableBlend();
-        GlStateManager.tryBlendFuncSeparate(
-            GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO
-        );
+        GlStateManager.tryBlendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
         GlStateManager.color(1.0F, 1.0F, 1.0F, p_175258_2_);
         this.mc.getTextureManager().bindTexture(WIDGETS);
         this.drawTexturedModalRect((float)(p_175258_3_ - 91), p_175258_4_, 0, 0, 182, 22);
@@ -88,7 +86,7 @@
 
         RenderHelper.enableGUIStandardItemLighting();
 
-        for (int i = 0; i < 9; i++)
+        for (int i = 0; i < 9; ++i)
         {
             this.renderSlot(i, p_175258_1_.getScaledWidth() / 2 - 90 + i * 20 + 2, p_175258_4_ + 3.0F, p_175258_2_, p_175258_5_.getObject(i));
         }
@@ -111,15 +109,11 @@
             GlStateManager.color(f, f, f, p_175266_4_);
             p_175266_5_.renderIcon(f, i);
             GlStateManager.popMatrix();
-            String s = String.valueOf(GameSettings.getKeyDisplayString(this.mc.gameSettings.keyBindsHotbar[p_175266_1_].getKeyCode()));
+            String s = String.valueOf(this.mc.gameSettings.keyBindsHotbar[p_175266_1_].getDisplayName());
 
             if (i > 3 && p_175266_5_.isEnabled())
             {
-                this.mc
-                .fontRenderer
-                .drawStringWithShadow(
-                    s, (float)(p_175266_2_ + 19 - 2 - this.mc.fontRenderer.getStringWidth(s)), p_175266_3_ + 6.0F + 3.0F, 16777215 + (i << 24)
-                );
+                this.mc.fontRenderer.drawStringWithShadow(s, (float)(p_175266_2_ + 19 - 2 - this.mc.fontRenderer.getStringWidth(s)), p_175266_3_ + 6.0F + 3.0F, 16777215 + (i << 24));
             }
         }
     }
@@ -131,9 +125,7 @@
         if (i > 3 && this.menu != null)
         {
             ISpectatorMenuObject ispectatormenuobject = this.menu.getSelectedItem();
-            String s = ispectatormenuobject == SpectatorMenu.EMPTY_SLOT
-                       ? this.menu.getSelectedCategory().getPrompt().getFormattedText()
-                       : ispectatormenuobject.getSpectatorName().getFormattedText();
+            String s = ispectatormenuobject == SpectatorMenu.EMPTY_SLOT ? this.menu.getSelectedCategory().getPrompt().getFormattedText() : ispectatormenuobject.getSpectatorName().getFormattedText();
 
             if (s != null)
             {
@@ -141,12 +133,7 @@
                 int k = p_175263_1_.getScaledHeight() - 35;
                 GlStateManager.pushMatrix();
                 GlStateManager.enableBlend();
-                GlStateManager.tryBlendFuncSeparate(
-                    GlStateManager.SourceFactor.SRC_ALPHA,
-                    GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA,
-                    GlStateManager.SourceFactor.ONE,
-                    GlStateManager.DestFactor.ZERO
-                );
+                GlStateManager.tryBlendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
                 this.mc.fontRenderer.drawStringWithShadow(s, (float)j, (float)k, 16777215 + (i << 24));
                 GlStateManager.disableBlend();
                 GlStateManager.popMatrix();
@@ -154,7 +141,6 @@
         }
     }
 
-    @Override
     public void onSpectatorMenuClosed(SpectatorMenu menu)
     {
         this.menu = null;
@@ -168,13 +154,11 @@
 
     public void onMouseScroll(int p_175259_1_)
     {
-        int i = this.menu.getSelectedSlot() + p_175259_1_;
+        int i;
 
-        while (
-            i >= 0 && i <= 8 && (this.menu.getItem(i) == SpectatorMenu.EMPTY_SLOT || !this.menu.getItem(i).isEnabled())
-        )
+        for (i = this.menu.getSelectedSlot() + p_175259_1_; i >= 0 && i <= 8 && (this.menu.getItem(i) == SpectatorMenu.EMPTY_SLOT || !this.menu.getItem(i).isEnabled()); i += p_175259_1_)
         {
-            i += p_175259_1_;
+            ;
         }
 
         if (i >= 0 && i <= 8)
