--- before/net/minecraft/client/gui/GuiScreenBook.java
+++ after/net/minecraft/client/gui/GuiScreenBook.java
@@ -69,6 +69,7 @@
 
             if (this.field_146476_w < 1)
             {
+                this.field_146483_y.func_74742_a(new NBTTagString("")); // Forge: fix MC-1685
                 this.field_146476_w = 1;
             }
         }
@@ -81,14 +82,12 @@
         }
     }
 
-    @Override
     public void func_73876_c()
     {
         super.func_73876_c();
         ++this.field_146479_t;
     }
 
-    @Override
     public void func_73866_w_()
     {
         this.field_146292_n.clear();
@@ -108,12 +107,11 @@
 
         int i = (this.field_146294_l - 192) / 2;
         int j = 2;
-        this.field_146470_A = this.func_189646_b(new GuiScreenBook.NextPageButton(1, i + 120, 156, true));
-        this.field_146471_B = this.func_189646_b(new GuiScreenBook.NextPageButton(2, i + 38, 156, false));
+        this.field_146470_A = (GuiScreenBook.NextPageButton)this.func_189646_b(new GuiScreenBook.NextPageButton(1, i + 120, 156, true));
+        this.field_146471_B = (GuiScreenBook.NextPageButton)this.func_189646_b(new GuiScreenBook.NextPageButton(2, i + 38, 156, false));
         this.func_146464_h();
     }
 
-    @Override
     public void func_146281_b()
     {
         Keyboard.enableRepeatEvents(false);
@@ -178,14 +176,13 @@
         }
     }
 
-    @Override
     protected void func_146284_a(GuiButton p_146284_1_) throws IOException
     {
         if (p_146284_1_.field_146124_l)
         {
             if (p_146284_1_.field_146127_k == 0)
             {
-                this.field_146297_k.func_147108_a(null);
+                this.field_146297_k.func_147108_a((GuiScreen)null);
                 this.func_146462_a(false);
             }
             else if (p_146284_1_.field_146127_k == 3 && this.field_146475_i)
@@ -218,7 +215,7 @@
             else if (p_146284_1_.field_146127_k == 5 && this.field_146480_s)
             {
                 this.func_146462_a(true);
-                this.field_146297_k.func_147108_a(null);
+                this.field_146297_k.func_147108_a((GuiScreen)null);
             }
             else if (p_146284_1_.field_146127_k == 4 && this.field_146480_s)
             {
@@ -239,7 +236,6 @@
         }
     }
 
-    @Override
     protected void func_73869_a(char p_73869_1_, int p_73869_2_) throws IOException
     {
         super.func_73869_a(p_73869_1_, p_73869_2_);
@@ -309,7 +305,7 @@
                 if (!this.field_146482_z.isEmpty())
                 {
                     this.func_146462_a(true);
-                    this.field_146297_k.func_147108_a(null);
+                    this.field_146297_k.func_147108_a((GuiScreen)null);
                 }
 
                 return;
@@ -326,9 +322,7 @@
 
     private String func_146456_p()
     {
-        return this.field_146483_y != null && this.field_146484_x >= 0 && this.field_146484_x < this.field_146483_y.func_74745_c()
-               ? this.field_146483_y.func_150307_f(this.field_146484_x)
-               : "";
+        return this.field_146483_y != null && this.field_146484_x >= 0 && this.field_146484_x < this.field_146483_y.func_74745_c() ? this.field_146483_y.func_150307_f(this.field_146484_x) : "";
     }
 
     private void func_146457_a(String p_146457_1_)
@@ -352,7 +346,6 @@
         }
     }
 
-    @Override
     public void func_73863_a(int p_73863_1_, int p_73863_2_, float p_73863_3_)
     {
         GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
@@ -420,11 +413,9 @@
                     try
                     {
                         ITextComponent itextcomponent = ITextComponent.Serializer.func_150699_a(s5);
-                        this.field_175386_A = itextcomponent != null
-                                              ? GuiUtilRenderComponents.func_178908_a(itextcomponent, 116, this.field_146289_q, true, true)
-                                              : null;
+                        this.field_175386_A = itextcomponent != null ? GuiUtilRenderComponents.func_178908_a(itextcomponent, 116, this.field_146289_q, true, true) : null;
                     }
-                    catch (JsonParseException jsonparseexception)
+                    catch (JsonParseException var13)
                     {
                         this.field_175386_A = null;
                     }
@@ -467,7 +458,6 @@
         super.func_73863_a(p_73863_1_, p_73863_2_, p_73863_3_);
     }
 
-    @Override
     protected void func_73864_a(int p_73864_1_, int p_73864_2_, int p_73864_3_) throws IOException
     {
         if (p_73864_3_ == 0)
@@ -483,7 +473,6 @@
         super.func_73864_a(p_73864_1_, p_73864_2_, p_73864_3_);
     }
 
-    @Override
     public boolean func_175276_a(ITextComponent p_175276_1_)
     {
         ClickEvent clickevent = p_175276_1_.func_150256_b().func_150235_h();
@@ -507,8 +496,9 @@
                     return true;
                 }
             }
-            catch (Throwable throwable)
+            catch (Throwable var5)
             {
+                ;
             }
 
             return false;
@@ -519,7 +509,7 @@
 
             if (flag && clickevent.func_150669_a() == ClickEvent.Action.RUN_COMMAND)
             {
-                this.field_146297_k.func_147108_a(null);
+                this.field_146297_k.func_147108_a((GuiScreen)null);
             }
 
             return flag;
@@ -590,15 +580,11 @@
                 this.field_146151_o = p_i46316_4_;
             }
 
-            @Override
             public void func_191745_a(Minecraft p_191745_1_, int p_191745_2_, int p_191745_3_, float p_191745_4_)
             {
                 if (this.field_146125_m)
                 {
-                    boolean flag = p_191745_2_ >= this.field_146128_h
-                                   && p_191745_3_ >= this.field_146129_i
-                                   && p_191745_2_ < this.field_146128_h + this.field_146120_f
-                                   && p_191745_3_ < this.field_146129_i + this.field_146121_g;
+                    boolean flag = p_191745_2_ >= this.field_146128_h && p_191745_3_ >= this.field_146129_i && p_191745_2_ < this.field_146128_h + this.field_146120_f && p_191745_3_ < this.field_146129_i + this.field_146121_g;
                     GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
                     p_191745_1_.func_110434_K().func_110577_a(GuiScreenBook.field_146466_f);
                     int i = 0;
