--- before/net/minecraft/client/gui/GuiSleepMP.java
+++ after/net/minecraft/client/gui/GuiSleepMP.java
@@ -10,14 +10,12 @@
 @SideOnly(Side.CLIENT)
 public class GuiSleepMP extends GuiChat
 {
-    @Override
     public void initGui()
     {
         super.initGui();
         this.buttonList.add(new GuiButton(1, this.width / 2 - 100, this.height - 40, I18n.format("multiplayer.stopSleeping")));
     }
 
-    @Override
     protected void keyTyped(char typedChar, int keyCode) throws IOException
     {
         if (keyCode == 1)
@@ -34,7 +32,7 @@
 
             if (!s.isEmpty())
             {
-                this.mc.player.sendChatMessage(s);
+                this.sendChatMessage(s); // Forge: fix vanilla not adding messages to the sent list while sleeping
             }
 
             this.inputField.setText("");
@@ -42,7 +40,6 @@
         }
     }
 
-    @Override
     protected void actionPerformed(GuiButton button) throws IOException
     {
         if (button.id == 1)
