--- before/net/minecraft/client/gui/GuiScreenAddServer.java
+++ after/net/minecraft/client/gui/GuiScreenAddServer.java
@@ -9,7 +9,7 @@
 import net.minecraft.util.StringUtils;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
-import org.lwjgl.input.Keyboard;
+import org.lwjglx.input.Keyboard;
 
 @SideOnly(Side.CLIENT)
 public class GuiScreenAddServer extends GuiScreen
