--- before/net/minecraft/client/gui/GuiSubtitleOverlay.java
+++ after/net/minecraft/client/gui/GuiSubtitleOverlay.java
@@ -17,7 +17,7 @@
 public class GuiSubtitleOverlay extends Gui implements ISoundEventListener
 {
     private final Minecraft field_184069_a;
-    private final List<GuiSubtitleOverlay.Subtitle> field_184070_f = Lists.newArrayList();
+    private final List<GuiSubtitleOverlay.Subtitle> field_184070_f = Lists.<GuiSubtitleOverlay.Subtitle>newArrayList();
     private boolean field_184071_g;
 
     public GuiSubtitleOverlay(Minecraft p_i46603_1_)
@@ -44,8 +44,8 @@
             GlStateManager.func_179147_l();
             GlStateManager.func_187428_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
             Vec3d vec3d = new Vec3d(this.field_184069_a.field_71439_g.field_70165_t, this.field_184069_a.field_71439_g.field_70163_u + (double)this.field_184069_a.field_71439_g.func_70047_e(), this.field_184069_a.field_71439_g.field_70161_v);
-            Vec3d vec3d1 = (new Vec3d(0.0D, 0.0D, -1.0D)).func_178789_a(-this.field_184069_a.field_71439_g.field_70125_A * ((float)Math.PI / 180F)).func_178785_b(-this.field_184069_a.field_71439_g.field_70177_z * ((float)Math.PI / 180F));
-            Vec3d vec3d2 = (new Vec3d(0.0D, 1.0D, 0.0D)).func_178789_a(-this.field_184069_a.field_71439_g.field_70125_A * ((float)Math.PI / 180F)).func_178785_b(-this.field_184069_a.field_71439_g.field_70177_z * ((float)Math.PI / 180F));
+            Vec3d vec3d1 = (new Vec3d(0.0D, 0.0D, -1.0D)).func_178789_a(-this.field_184069_a.field_71439_g.field_70125_A * 0.017453292F).func_178785_b(-this.field_184069_a.field_71439_g.field_70177_z * 0.017453292F);
+            Vec3d vec3d2 = (new Vec3d(0.0D, 1.0D, 0.0D)).func_178789_a(-this.field_184069_a.field_71439_g.field_70125_A * 0.017453292F).func_178785_b(-this.field_184069_a.field_71439_g.field_70177_z * 0.017453292F);
             Vec3d vec3d3 = vec3d1.func_72431_c(vec3d2);
             int i = 0;
             int j = 0;
@@ -65,7 +65,7 @@
                 }
             }
 
-            j += this.field_184069_a.field_71466_p.func_78256_a("<") + this.field_184069_a.field_71466_p.func_78256_a(" ") + this.field_184069_a.field_71466_p.func_78256_a(">") + this.field_184069_a.field_71466_p.func_78256_a(" ");
+            j = j + this.field_184069_a.field_71466_p.func_78256_a("<") + this.field_184069_a.field_71466_p.func_78256_a(" ") + this.field_184069_a.field_71466_p.func_78256_a(">") + this.field_184069_a.field_71466_p.func_78256_a(" ");
 
             for (GuiSubtitleOverlay.Subtitle guisubtitleoverlay$subtitle1 : this.field_184070_f)
             {
