--- before/net/minecraft/client/gui/GuiShareToLan.java
+++ after/net/minecraft/client/gui/GuiShareToLan.java
@@ -88,7 +88,7 @@
 
             if (s != null)
             {
-                itextcomponent = new TextComponentTranslation("commands.publish.started", s);
+                itextcomponent = new TextComponentTranslation("commands.publish.started", new Object[] {s});
             }
             else
             {
