--- before/net/minecraft/client/gui/GuiScreenDemo.java
+++ after/net/minecraft/client/gui/GuiScreenDemo.java
@@ -17,18 +17,14 @@
     private static final Logger field_146349_a = LogManager.getLogger();
     private static final ResourceLocation field_146348_f = new ResourceLocation("textures/gui/demo_background.png");
 
-    @Override
     public void func_73866_w_()
     {
         this.field_146292_n.clear();
         int i = -16;
-        this.field_146292_n
-        .add(new GuiButton(1, this.field_146294_l / 2 - 116, this.field_146295_m / 2 + 62 + -16, 114, 20, I18n.func_135052_a("demo.help.buy")));
-        this.field_146292_n
-        .add(new GuiButton(2, this.field_146294_l / 2 + 2, this.field_146295_m / 2 + 62 + -16, 114, 20, I18n.func_135052_a("demo.help.later")));
+        this.field_146292_n.add(new GuiButton(1, this.field_146294_l / 2 - 116, this.field_146295_m / 2 + 62 + -16, 114, 20, I18n.func_135052_a("demo.help.buy")));
+        this.field_146292_n.add(new GuiButton(2, this.field_146294_l / 2 + 2, this.field_146295_m / 2 + 62 + -16, 114, 20, I18n.func_135052_a("demo.help.later")));
     }
 
-    @Override
     protected void func_146284_a(GuiButton p_146284_1_) throws IOException
     {
         switch (p_146284_1_.field_146127_k)
@@ -39,7 +35,7 @@
                 try
                 {
                     Class<?> oclass = Class.forName("java.awt.Desktop");
-                    Object object = oclass.getMethod("getDesktop").invoke(null);
+                    Object object = oclass.getMethod("getDesktop").invoke((Object)null);
                     oclass.getMethod("browse", URI.class).invoke(object, new URI("http://www.minecraft.net/store?source=demo"));
                 }
                 catch (Throwable throwable)
@@ -49,12 +45,11 @@
 
                 break;
             case 2:
-                this.field_146297_k.func_147108_a(null);
+                this.field_146297_k.func_147108_a((GuiScreen)null);
                 this.field_146297_k.func_71381_h();
         }
     }
 
-    @Override
     public void func_146276_q_()
     {
         super.func_146276_q_();
@@ -65,33 +60,18 @@
         this.func_73729_b(i, j, 0, 0, 248, 166);
     }
 
-    @Override
     public void func_73863_a(int p_73863_1_, int p_73863_2_, float p_73863_3_)
     {
         this.func_146276_q_();
         int i = (this.field_146294_l - 248) / 2 + 10;
         int j = (this.field_146295_m - 166) / 2 + 8;
         this.field_146289_q.func_78276_b(I18n.func_135052_a("demo.help.title"), i, j, 2039583);
-        j += 12;
+        j = j + 12;
         GameSettings gamesettings = this.field_146297_k.field_71474_y;
-        this.field_146289_q
-        .func_78276_b(
-            I18n.func_135052_a(
-                "demo.help.movementShort",
-                GameSettings.func_74298_c(gamesettings.field_74351_w.func_151463_i()),
-                GameSettings.func_74298_c(gamesettings.field_74370_x.func_151463_i()),
-                GameSettings.func_74298_c(gamesettings.field_74368_y.func_151463_i()),
-                GameSettings.func_74298_c(gamesettings.field_74366_z.func_151463_i())
-            ),
-            i,
-            j,
-            5197647
-        );
+        this.field_146289_q.func_78276_b(I18n.func_135052_a("demo.help.movementShort", gamesettings.field_74351_w.getDisplayName(), gamesettings.field_74370_x.getDisplayName(), gamesettings.field_74368_y.getDisplayName(), gamesettings.field_74366_z.getDisplayName()), i, j, 5197647);
         this.field_146289_q.func_78276_b(I18n.func_135052_a("demo.help.movementMouse"), i, j + 12, 5197647);
-        this.field_146289_q
-        .func_78276_b(I18n.func_135052_a("demo.help.jump", GameSettings.func_74298_c(gamesettings.field_74314_A.func_151463_i())), i, j + 24, 5197647);
-        this.field_146289_q
-        .func_78276_b(I18n.func_135052_a("demo.help.inventory", GameSettings.func_74298_c(gamesettings.field_151445_Q.func_151463_i())), i, j + 36, 5197647);
+        this.field_146289_q.func_78276_b(I18n.func_135052_a("demo.help.jump", gamesettings.field_74314_A.getDisplayName()), i, j + 24, 5197647);
+        this.field_146289_q.func_78276_b(I18n.func_135052_a("demo.help.inventory", gamesettings.field_151445_Q.getDisplayName()), i, j + 36, 5197647);
         this.field_146289_q.func_78279_b(I18n.func_135052_a("demo.help.fullWrapped"), i, j + 68, 218, 2039583);
         super.func_73863_a(p_73863_1_, p_73863_2_, p_73863_3_);
     }
