--- before/net/minecraft/client/gui/GuiSlot.java
+++ after/net/minecraft/client/gui/GuiSlot.java
@@ -186,27 +186,8 @@
             GlStateManager.disableFog();
             Tessellator tessellator = Tessellator.getInstance();
             BufferBuilder bufferbuilder = tessellator.getBuffer();
-            this.mc.getTextureManager().bindTexture(Gui.OPTIONS_BACKGROUND);
-            GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
-            float f = 32.0F;
-            bufferbuilder.begin(7, DefaultVertexFormats.POSITION_TEX_COLOR);
-            bufferbuilder.pos(this.left, this.bottom, 0.0)
-            .tex(this.left / 32.0F, (this.bottom + (int)this.amountScrolled) / 32.0F)
-            .color(32, 32, 32, 255)
-            .endVertex();
-            bufferbuilder.pos(this.right, this.bottom, 0.0)
-            .tex(this.right / 32.0F, (this.bottom + (int)this.amountScrolled) / 32.0F)
-            .color(32, 32, 32, 255)
-            .endVertex();
-            bufferbuilder.pos(this.right, this.top, 0.0)
-            .tex(this.right / 32.0F, (this.top + (int)this.amountScrolled) / 32.0F)
-            .color(32, 32, 32, 255)
-            .endVertex();
-            bufferbuilder.pos(this.left, this.top, 0.0)
-            .tex(this.left / 32.0F, (this.top + (int)this.amountScrolled) / 32.0F)
-            .color(32, 32, 32, 255)
-            .endVertex();
-            tessellator.draw();
+            // Forge: background rendering moved into separate method.
+            this.drawContainerBackground(tessellator);
             int k = this.left + this.width / 2 - this.getListWidth() / 2 + 2;
             int l = this.top + 4 - (int)this.amountScrolled;
 
@@ -491,5 +472,31 @@
     public int getSlotHeight()
     {
         return this.slotHeight;
+    }
+
+    protected void drawContainerBackground(Tessellator tessellator)
+    {
+        BufferBuilder buffer = tessellator.getBuffer();
+        this.mc.getTextureManager().bindTexture(Gui.OPTIONS_BACKGROUND);
+        GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+        float f = 32.0F;
+        buffer.begin(7, DefaultVertexFormats.POSITION_TEX_COLOR);
+        buffer.pos(this.left, this.bottom, 0.0)
+        .tex((float)this.left / f, (float)(this.bottom + (int)this.amountScrolled) / f)
+        .color(32, 32, 32, 255)
+        .endVertex();
+        buffer.pos(this.right, this.bottom, 0.0)
+        .tex((float)this.right / f, (float)(this.bottom + (int)this.amountScrolled) / f)
+        .color(32, 32, 32, 255)
+        .endVertex();
+        buffer.pos(this.right, this.top, 0.0)
+        .tex((float)this.right / f, (float)(this.top + (int)this.amountScrolled) / f)
+        .color(32, 32, 32, 255)
+        .endVertex();
+        buffer.pos(this.left, this.top, 0.0)
+        .tex((float)this.left / f, (float)(this.top + (int)this.amountScrolled) / f)
+        .color(32, 32, 32, 255)
+        .endVertex();
+        tessellator.draw();
     }
 }
