--- before/net/minecraft/client/gui/GuiTextField.java
+++ after/net/minecraft/client/gui/GuiTextField.java
@@ -35,7 +35,7 @@
     private int field_146221_u = 7368816;
     private boolean field_146220_v = true;
     private GuiPageButtonList.GuiResponder field_175210_x;
-    private Predicate<String> field_175209_y = Predicates.alwaysTrue();
+    private Predicate<String> field_175209_y = Predicates.<String>alwaysTrue();
 
     public GuiTextField(int p_i45542_1_, FontRenderer p_i45542_2_, int p_i45542_3_, int p_i45542_4_, int p_i45542_5_, int p_i45542_6_)
     {
