--- before/net/minecraft/client/gui/GuiUtilRenderComponents.java
+++ after/net/minecraft/client/gui/GuiUtilRenderComponents.java
@@ -21,7 +21,7 @@
     {
         int i = 0;
         ITextComponent itextcomponent = new TextComponentString("");
-        List<ITextComponent> list = Lists.newArrayList();
+        List<ITextComponent> list = Lists.<ITextComponent>newArrayList();
         List<ITextComponent> list1 = Lists.newArrayList(p_178908_0_);
 
         for (int j = 0; j < list1.size(); ++j)
@@ -49,18 +49,20 @@
 
             if (i + i1 > p_178908_1_)
             {
+                FontRenderer.BREAK_ITERATOR.setText(s4);
                 String s2 = p_178908_2_.func_78262_a(s4, p_178908_1_ - i, false);
                 String s3 = s2.length() < s4.length() ? s4.substring(s2.length()) : null;
 
                 if (s3 != null && !s3.isEmpty())
                 {
-                    int l = s2.lastIndexOf(32);
+                    FontRenderer.BREAK_ITERATOR.following(s2.length());
+                    int l = FontRenderer.BREAK_ITERATOR.previous();
 
                     if (l >= 0 && p_178908_2_.func_78256_a(s4.substring(0, l)) > 0)
                     {
                         s2 = s4.substring(0, l);
 
-                        if (p_178908_3_)
+                        if (p_178908_3_ && s4.charAt(l) == ' ')
                         {
                             ++l;
                         }
