--- before/net/minecraft/client/gui/GuiUtilRenderComponents.java
+++ after/net/minecraft/client/gui/GuiUtilRenderComponents.java
@@ -1,9 +1,11 @@
 package net.minecraft.client.gui;
 
+import com.cleanroommc.client.BreakIteratorHolder;
 import com.google.common.collect.Lists;
 import java.util.List;
 import net.minecraft.client.Minecraft;
 import net.minecraft.util.text.ITextComponent;
+import net.minecraft.util.text.Style;
 import net.minecraft.util.text.TextComponentString;
 import net.minecraft.util.text.TextFormatting;
 import net.minecraftforge.fml.relauncher.Side;
@@ -22,19 +24,21 @@
         )
     {
         int i = 0;
-        ITextComponent itextcomponent = new TextComponentString("");
         List<ITextComponent> list = Lists.newArrayList();
+        List<String> cache = Lists.newArrayList();
+        List<Style> styleCache = Lists.newArrayList();
+        List<Boolean> newLineCache = Lists.newArrayList();
         List<ITextComponent> list1 = Lists.newArrayList(textComponent);
 
+        boolean flag;
         for (int j = 0; j < list1.size(); j++)
         {
             ITextComponent itextcomponent1 = list1.get(j);
             String s = itextcomponent1.getUnformattedComponentText();
-            boolean flag = false;
 
             if (s.contains("\n"))
             {
-                int k = s.indexOf(10);
+                int k = s.indexOf('\n');
                 String s1 = s.substring(k + 1);
                 s = s.substring(0, k + 1);
                 ITextComponent itextcomponent2 = new TextComponentString(s1);
@@ -42,27 +46,45 @@
                 list1.add(j + 1, itextcomponent2);
                 flag = true;
             }
+            else
+            {
+                flag = false;
+            }
 
             String s4 = removeTextColorsIfConfigured(itextcomponent1.getStyle().getFormattingCode() + s, forceTextColor);
             String s5 = s4.endsWith("\n") ? s4.substring(0, s4.length() - 1) : s4;
-            int i1 = fontRendererIn.getStringWidth(s5);
-            TextComponentString textcomponentstring = new TextComponentString(s5);
-            textcomponentstring.setStyle(itextcomponent1.getStyle().createShallowCopy());
+            cache.add(s5);
+            styleCache.add(itextcomponent1.getStyle());
+            newLineCache.add(flag);
+        }
+        BreakIteratorHolder.BREAK_ITERATOR.setText(String.join("\n", cache));
+        int fed = 0;
+        ITextComponent itextcomponent = new TextComponentString("");
+        String s4;
+        Style style;
+        for (int j = 0; j < cache.size(); j++)
+        {
+            s4 = cache.get(j);
+            style = styleCache.get(j);
+            flag = newLineCache.get(j);
+            int i1 = fontRendererIn.getStringWidth(s4);
+            TextComponentString textcomponentstring = new TextComponentString(s4);
+            textcomponentstring.setStyle(style.createShallowCopy());
 
             if (i + i1 > maxTextLenght)
             {
                 String s2 = fontRendererIn.trimStringToWidth(s4, maxTextLenght - i, false);
                 String s3 = s2.length() < s4.length() ? s4.substring(s2.length()) : null;
 
-                if (s3 != null && !s3.isEmpty())
+                if (s3 != null)
                 {
-                    int l = s2.lastIndexOf(32);
+                    int l = BreakIteratorHolder.BREAK_ITERATOR.preceding(fed + s2.length()) - fed;
 
                     if (l >= 0 && fontRendererIn.getStringWidth(s4.substring(0, l)) > 0)
                     {
                         s2 = s4.substring(0, l);
 
-                        if (p_178908_3_)
+                        if (p_178908_3_ && s4.charAt(l) == ' ')
                         {
                             l++;
                         }
@@ -75,14 +97,15 @@
                         s3 = s4;
                     }
 
-                    TextComponentString textcomponentstring1 = new TextComponentString(s3);
-                    textcomponentstring1.setStyle(itextcomponent1.getStyle().createShallowCopy());
-                    list1.add(j + 1, textcomponentstring1);
+                    s3 = FontRenderer.getFormatFromString(s2) + s3; //Forge: Fix chat formatting not surviving line wrapping.
+
+                    cache.set(j, s3);
+                    j--;
                 }
 
                 i1 = fontRendererIn.getStringWidth(s2);
                 textcomponentstring = new TextComponentString(s2);
-                textcomponentstring.setStyle(itextcomponent1.getStyle().createShallowCopy());
+                textcomponentstring.setStyle(style.createShallowCopy());
                 flag = true;
             }
 
@@ -90,6 +113,7 @@
             {
                 i += i1;
                 itextcomponent.appendSibling(textcomponentstring);
+                fed += textcomponentstring.getText().length();
             }
             else
             {
