--- before/net/minecraft/client/gui/GuiWorldEdit.java
+++ after/net/minecraft/client/gui/GuiWorldEdit.java
@@ -8,7 +8,7 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.commons.io.FileUtils;
-import org.lwjgl.input.Keyboard;
+import org.lwjglx.input.Keyboard;
 
 @SideOnly(Side.CLIENT)
 public class GuiWorldEdit extends GuiScreen
