--- before/net/minecraft/client/gui/ServerListEntryNormal.java
+++ after/net/minecraft/client/gui/ServerListEntryNormal.java
@@ -29,9 +29,7 @@
 public class ServerListEntryNormal implements GuiListExtended.IGuiListEntry
 {
     private static final Logger field_148304_a = LogManager.getLogger();
-    private static final ThreadPoolExecutor field_148302_b = new ScheduledThreadPoolExecutor(
-            5, new ThreadFactoryBuilder().setNameFormat("Server Pinger #%d").setDaemon(true).build()
-        );
+    private static final ThreadPoolExecutor field_148302_b = new ScheduledThreadPoolExecutor(5, (new ThreadFactoryBuilder()).setNameFormat("Server Pinger #%d").setDaemon(true).build());
     private static final ResourceLocation field_178015_c = new ResourceLocation("textures/misc/unknown_server.png");
     private static final ResourceLocation field_178014_d = new ResourceLocation("textures/gui/server_selection.png");
     private final GuiMultiplayer field_148303_c;
@@ -51,18 +49,7 @@
         this.field_148305_h = (DynamicTexture)this.field_148300_d.func_110434_K().func_110581_b(this.field_148306_i);
     }
 
-    @Override
-    public void func_192634_a(
-        int p_192634_1_,
-        int p_192634_2_,
-        int p_192634_3_,
-        int p_192634_4_,
-        int p_192634_5_,
-        int p_192634_6_,
-        int p_192634_7_,
-        boolean p_192634_8_,
-        float p_192634_9_
-    )
+    public void func_192634_a(int p_192634_1_, int p_192634_2_, int p_192634_3_, int p_192634_4_, int p_192634_5_, int p_192634_6_, int p_192634_7_, boolean p_192634_8_, float p_192634_9_)
     {
         if (!this.field_148301_e.field_78841_f)
         {
@@ -72,19 +59,18 @@
             this.field_148301_e.field_78846_c = "";
             field_148302_b.submit(new Runnable()
             {
-                @Override
                 public void run()
                 {
                     try
                     {
                         ServerListEntryNormal.this.field_148303_c.func_146789_i().func_147224_a(ServerListEntryNormal.this.field_148301_e);
                     }
-                    catch (UnknownHostException unknownhostexception)
+                    catch (UnknownHostException var2)
                     {
                         ServerListEntryNormal.this.field_148301_e.field_78844_e = -1L;
                         ServerListEntryNormal.this.field_148301_e.field_78843_d = TextFormatting.DARK_RED + I18n.func_135052_a("multiplayer.status.cannot_resolve");
                     }
-                    catch (Exception exception)
+                    catch (Exception var3)
                     {
                         ServerListEntryNormal.this.field_148301_e.field_78844_e = -1L;
                         ServerListEntryNormal.this.field_148301_e.field_78843_d = TextFormatting.DARK_RED + I18n.func_135052_a("multiplayer.status.cannot_connect");
@@ -97,13 +83,11 @@
         boolean flag1 = this.field_148301_e.field_82821_f < 340;
         boolean flag2 = flag || flag1;
         this.field_148300_d.field_71466_p.func_78276_b(this.field_148301_e.field_78847_a, p_192634_2_ + 32 + 3, p_192634_3_ + 1, 16777215);
-        List<String> list = this.field_148300_d.field_71466_p.func_78271_c(this.field_148301_e.field_78843_d, p_192634_4_ - 32 - 2);
+        List<String> list = this.field_148300_d.field_71466_p.func_78271_c(net.minecraftforge.fml.client.FMLClientHandler.instance().fixDescription(this.field_148301_e.field_78843_d), p_192634_4_ - 32 - 2);
 
-        for (int i = 0; i < Math.min(list.size(), 2); i++)
+        for (int i = 0; i < Math.min(list.size(), 2); ++i)
         {
-            this.field_148300_d
-            .field_71466_p
-            .func_78276_b(list.get(i), p_192634_2_ + 32 + 3, p_192634_3_ + 12 + this.field_148300_d.field_71466_p.field_78288_b * i, 8421504);
+            this.field_148300_d.field_71466_p.func_78276_b(list.get(i), p_192634_2_ + 32 + 3, p_192634_3_ + 12 + this.field_148300_d.field_71466_p.field_78288_b * i, 8421504);
         }
 
         String s2 = flag2 ? TextFormatting.DARK_RED + this.field_148301_e.field_82822_g : this.field_148301_e.field_78846_c;
@@ -193,6 +177,11 @@
         int i1 = p_192634_6_ - p_192634_2_;
         int j1 = p_192634_7_ - p_192634_3_;
 
+        String tooltip = net.minecraftforge.fml.client.FMLClientHandler.instance().enhanceServerListEntry(this, this.field_148301_e, p_192634_2_, p_192634_4_, p_192634_3_, i1, j1);
+        if (tooltip != null)
+        {
+            this.field_148303_c.func_146793_a(tooltip);
+        } else
         if (i1 >= p_192634_4_ - 15 && i1 <= p_192634_4_ - 5 && j1 >= 0 && j1 <= 8)
         {
             this.field_148303_c.func_146793_a(s1);
@@ -270,10 +259,10 @@
         }
         else
         {
-            ByteBuf bytebuf = Unpooled.copiedBuffer(this.field_148301_e.func_147409_e(), StandardCharsets.UTF_8);
+            ByteBuf bytebuf = Unpooled.copiedBuffer((CharSequence)this.field_148301_e.func_147409_e(), StandardCharsets.UTF_8);
             ByteBuf bytebuf1 = null;
             BufferedImage bufferedimage;
-            label80:
+            label99:
             {
                 try
                 {
@@ -281,12 +270,12 @@
                     bufferedimage = TextureUtil.func_177053_a(new ByteBufInputStream(bytebuf1));
                     Validate.validState(bufferedimage.getWidth() == 64, "Must be 64 pixels wide");
                     Validate.validState(bufferedimage.getHeight() == 64, "Must be 64 pixels high");
-                    break label80;
+                    break label99;
                 }
                 catch (Throwable throwable)
                 {
                     field_148304_a.error("Invalid icon for server {} ({})", this.field_148301_e.field_78847_a, this.field_148301_e.field_78845_b, throwable);
-                    this.field_148301_e.func_147407_a(null);
+                    this.field_148301_e.func_147407_a((String)null);
                 }
                 finally
                 {
@@ -312,7 +301,6 @@
         }
     }
 
-    @Override
     public boolean func_148278_a(int p_148278_1_, int p_148278_2_, int p_148278_3_, int p_148278_4_, int p_148278_5_, int p_148278_6_)
     {
         if (p_148278_5_ <= 32)
@@ -348,12 +336,10 @@
         return false;
     }
 
-    @Override
     public void func_192633_a(int p_192633_1_, int p_192633_2_, int p_192633_3_, float p_192633_4_)
     {
     }
 
-    @Override
     public void func_148277_b(int p_148277_1_, int p_148277_2_, int p_148277_3_, int p_148277_4_, int p_148277_5_, int p_148277_6_)
     {
     }
