--- before/net/minecraft/client/particle/ParticleDragonBreath.java
+++ after/net/minecraft/client/particle/ParticleDragonBreath.java
@@ -61,12 +61,12 @@
                 this.field_187131_k *= 1.1D;
             }
 
-            this.field_187129_i *= (double)0.96F;
-            this.field_187131_k *= (double)0.96F;
+            this.field_187129_i *= 0.9599999785423279D;
+            this.field_187131_k *= 0.9599999785423279D;
 
             if (this.field_187144_G)
             {
-                this.field_187130_j *= (double)0.96F;
+                this.field_187130_j *= 0.9599999785423279D;
             }
         }
     }
