--- before/net/minecraft/client/particle/ParticleFlame.java
+++ after/net/minecraft/client/particle/ParticleFlame.java
@@ -15,9 +15,9 @@
     protected ParticleFlame(World p_i1209_1_, double p_i1209_2_, double p_i1209_4_, double p_i1209_6_, double p_i1209_8_, double p_i1209_10_, double p_i1209_12_)
     {
         super(p_i1209_1_, p_i1209_2_, p_i1209_4_, p_i1209_6_, p_i1209_8_, p_i1209_10_, p_i1209_12_);
-        this.field_187129_i = this.field_187129_i * (double)0.01F + p_i1209_8_;
-        this.field_187130_j = this.field_187130_j * (double)0.01F + p_i1209_10_;
-        this.field_187131_k = this.field_187131_k * (double)0.01F + p_i1209_12_;
+        this.field_187129_i = this.field_187129_i * 0.009999999776482582D + p_i1209_8_;
+        this.field_187130_j = this.field_187130_j * 0.009999999776482582D + p_i1209_10_;
+        this.field_187131_k = this.field_187131_k * 0.009999999776482582D + p_i1209_12_;
         this.field_187126_f += (double)((this.field_187136_p.nextFloat() - this.field_187136_p.nextFloat()) * 0.05F);
         this.field_187127_g += (double)((this.field_187136_p.nextFloat() - this.field_187136_p.nextFloat()) * 0.05F);
         this.field_187128_h += (double)((this.field_187136_p.nextFloat() - this.field_187136_p.nextFloat()) * 0.05F);
@@ -49,7 +49,7 @@
         int i = super.func_189214_a(p_189214_1_);
         int j = i & 255;
         int k = i >> 16 & 255;
-        j += (int)(f * 15.0F * 16.0F);
+        j = j + (int)(f * 15.0F * 16.0F);
 
         if (j > 240)
         {
@@ -71,14 +71,14 @@
         }
 
         this.func_187110_a(this.field_187129_i, this.field_187130_j, this.field_187131_k);
-        this.field_187129_i *= (double)0.96F;
-        this.field_187130_j *= (double)0.96F;
-        this.field_187131_k *= (double)0.96F;
+        this.field_187129_i *= 0.9599999785423279D;
+        this.field_187130_j *= 0.9599999785423279D;
+        this.field_187131_k *= 0.9599999785423279D;
 
         if (this.field_187132_l)
         {
-            this.field_187129_i *= (double)0.7F;
-            this.field_187131_k *= (double)0.7F;
+            this.field_187129_i *= 0.699999988079071D;
+            this.field_187131_k *= 0.699999988079071D;
         }
     }
 
