--- before/net/minecraft/client/particle/ParticleItemPickup.java
+++ after/net/minecraft/client/particle/ParticleItemPickup.java
@@ -71,7 +71,9 @@
         d7 -= interpPosY;
         d8 -= interpPosZ;
         GlStateManager.enableLighting();
+        GlStateManager.enableCull();
         this.renderManager.renderEntity(this.item, d6, d7, d8, this.item.rotationYaw, partialTicks, false);
+        GlStateManager.disableCull();
     }
 
     @Override
