--- before/net/minecraft/client/particle/ParticleSnowShovel.java
+++ after/net/minecraft/client/particle/ParticleSnowShovel.java
@@ -20,13 +20,13 @@
     protected ParticleSnowShovel(World p_i1228_1_, double p_i1228_2_, double p_i1228_4_, double p_i1228_6_, double p_i1228_8_, double p_i1228_10_, double p_i1228_12_, float p_i1228_14_)
     {
         super(p_i1228_1_, p_i1228_2_, p_i1228_4_, p_i1228_6_, p_i1228_8_, p_i1228_10_, p_i1228_12_);
-        this.field_187129_i *= (double)0.1F;
-        this.field_187130_j *= (double)0.1F;
-        this.field_187131_k *= (double)0.1F;
+        this.field_187129_i *= 0.10000000149011612D;
+        this.field_187130_j *= 0.10000000149011612D;
+        this.field_187131_k *= 0.10000000149011612D;
         this.field_187129_i += p_i1228_8_;
         this.field_187130_j += p_i1228_10_;
         this.field_187131_k += p_i1228_12_;
-        float f = 1.0F - (float)(Math.random() * (double)0.3F);
+        float f = 1.0F - (float)(Math.random() * 0.30000001192092896D);
         this.field_70552_h = f;
         this.field_70553_i = f;
         this.field_70551_j = f;
@@ -59,14 +59,14 @@
         this.func_70536_a(7 - this.field_70546_d * 8 / this.field_70547_e);
         this.field_187130_j -= 0.03D;
         this.func_187110_a(this.field_187129_i, this.field_187130_j, this.field_187131_k);
-        this.field_187129_i *= (double)0.99F;
-        this.field_187130_j *= (double)0.99F;
-        this.field_187131_k *= (double)0.99F;
+        this.field_187129_i *= 0.9900000095367432D;
+        this.field_187130_j *= 0.9900000095367432D;
+        this.field_187131_k *= 0.9900000095367432D;
 
         if (this.field_187132_l)
         {
-            this.field_187129_i *= (double)0.7F;
-            this.field_187131_k *= (double)0.7F;
+            this.field_187129_i *= 0.699999988079071D;
+            this.field_187131_k *= 0.699999988079071D;
         }
     }
 
