--- before/net/minecraft/client/particle/Particle.java
+++ after/net/minecraft/client/particle/Particle.java
@@ -16,7 +16,7 @@
 @SideOnly(Side.CLIENT)
 public class Particle
 {
-    private static final AxisAlignedBB field_187121_a = new AxisAlignedBB(0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
+    private static final AxisAlignedBB field_187121_a = new AxisAlignedBB(0.0D, 0.0D, 0.0D, 0.0D, 0.0D, 0.0D);
     protected World field_187122_b;
     protected double field_187123_c;
     protected double field_187124_d;
@@ -27,13 +27,13 @@
     protected double field_187129_i;
     protected double field_187130_j;
     protected double field_187131_k;
-    private AxisAlignedBB field_187120_G = field_187121_a;
+    private AxisAlignedBB field_187120_G;
     protected boolean field_187132_l;
     protected boolean field_190017_n;
     protected boolean field_187133_m;
-    protected float field_187134_n = 0.6F;
-    protected float field_187135_o = 1.8F;
-    protected Random field_187136_p = new Random();
+    protected float field_187134_n;
+    protected float field_187135_o;
+    protected Random field_187136_p;
     protected int field_94054_b;
     protected int field_94055_c;
     protected float field_70548_b;
@@ -45,7 +45,7 @@
     protected float field_70552_h;
     protected float field_70553_i;
     protected float field_70551_j;
-    protected float field_82339_as = 1.0F;
+    protected float field_82339_as;
     protected TextureAtlasSprite field_187119_C;
     protected float field_190014_F;
     protected float field_190015_G;
@@ -56,6 +56,11 @@
 
     protected Particle(World p_i46352_1_, double p_i46352_2_, double p_i46352_4_, double p_i46352_6_)
     {
+        this.field_187120_G = field_187121_a;
+        this.field_187134_n = 0.6F;
+        this.field_187135_o = 1.8F;
+        this.field_187136_p = new Random();
+        this.field_82339_as = 1.0F;
         this.field_187122_b = p_i46352_1_;
         this.func_187115_a(0.2F, 0.2F);
         this.func_187109_b(p_i46352_2_, p_i46352_4_, p_i46352_6_);
@@ -76,22 +81,20 @@
     public Particle(World p_i1219_1_, double p_i1219_2_, double p_i1219_4_, double p_i1219_6_, double p_i1219_8_, double p_i1219_10_, double p_i1219_12_)
     {
         this(p_i1219_1_, p_i1219_2_, p_i1219_4_, p_i1219_6_);
-        this.field_187129_i = p_i1219_8_ + (Math.random() * 2.0 - 1.0) * 0.4F;
-        this.field_187130_j = p_i1219_10_ + (Math.random() * 2.0 - 1.0) * 0.4F;
-        this.field_187131_k = p_i1219_12_ + (Math.random() * 2.0 - 1.0) * 0.4F;
-        float f = (float)(Math.random() + Math.random() + 1.0) * 0.15F;
-        float f1 = MathHelper.func_76133_a(
-                       this.field_187129_i * this.field_187129_i + this.field_187130_j * this.field_187130_j + this.field_187131_k * this.field_187131_k
-                   );
-        this.field_187129_i = this.field_187129_i / (double)f1 * (double)f * 0.4F;
-        this.field_187130_j = this.field_187130_j / (double)f1 * (double)f * 0.4F + 0.1F;
-        this.field_187131_k = this.field_187131_k / (double)f1 * (double)f * 0.4F;
+        this.field_187129_i = p_i1219_8_ + (Math.random() * 2.0D - 1.0D) * 0.4000000059604645D;
+        this.field_187130_j = p_i1219_10_ + (Math.random() * 2.0D - 1.0D) * 0.4000000059604645D;
+        this.field_187131_k = p_i1219_12_ + (Math.random() * 2.0D - 1.0D) * 0.4000000059604645D;
+        float f = (float)(Math.random() + Math.random() + 1.0D) * 0.15F;
+        float f1 = MathHelper.func_76133_a(this.field_187129_i * this.field_187129_i + this.field_187130_j * this.field_187130_j + this.field_187131_k * this.field_187131_k);
+        this.field_187129_i = this.field_187129_i / (double)f1 * (double)f * 0.4000000059604645D;
+        this.field_187130_j = this.field_187130_j / (double)f1 * (double)f * 0.4000000059604645D + 0.10000000149011612D;
+        this.field_187131_k = this.field_187131_k / (double)f1 * (double)f * 0.4000000059604645D;
     }
 
     public Particle func_70543_e(float p_70543_1_)
     {
         this.field_187129_i *= (double)p_70543_1_;
-        this.field_187130_j = (this.field_187130_j - 0.1F) * (double)p_70543_1_ + 0.1F;
+        this.field_187130_j = (this.field_187130_j - 0.10000000149011612D) * (double)p_70543_1_ + 0.10000000149011612D;
         this.field_187131_k *= (double)p_70543_1_;
         return this;
     }
@@ -151,29 +154,20 @@
             this.func_187112_i();
         }
 
-        this.field_187130_j = this.field_187130_j - 0.04 * (double)this.field_70545_g;
+        this.field_187130_j -= 0.04D * (double)this.field_70545_g;
         this.func_187110_a(this.field_187129_i, this.field_187130_j, this.field_187131_k);
-        this.field_187129_i *= 0.98F;
-        this.field_187130_j *= 0.98F;
-        this.field_187131_k *= 0.98F;
+        this.field_187129_i *= 0.9800000190734863D;
+        this.field_187130_j *= 0.9800000190734863D;
+        this.field_187131_k *= 0.9800000190734863D;
 
         if (this.field_187132_l)
         {
-            this.field_187129_i *= 0.7F;
-            this.field_187131_k *= 0.7F;
+            this.field_187129_i *= 0.699999988079071D;
+            this.field_187131_k *= 0.699999988079071D;
         }
     }
 
-    public void func_180434_a(
-        BufferBuilder p_180434_1_,
-        Entity p_180434_2_,
-        float p_180434_3_,
-        float p_180434_4_,
-        float p_180434_5_,
-        float p_180434_6_,
-        float p_180434_7_,
-        float p_180434_8_
-    )
+    public void func_180434_a(BufferBuilder p_180434_1_, Entity p_180434_2_, float p_180434_3_, float p_180434_4_, float p_180434_5_, float p_180434_6_, float p_180434_7_, float p_180434_8_)
     {
         float f = (float)this.field_94054_b / 16.0F;
         float f1 = f + 0.0624375F;
@@ -195,13 +189,7 @@
         int i = this.func_189214_a(p_180434_3_);
         int j = i >> 16 & 65535;
         int k = i & 65535;
-        Vec3d[] avec3d = new Vec3d[]
-        {
-            new Vec3d((double)(-p_180434_4_ * f4 - p_180434_7_ * f4), (double)(-p_180434_5_ * f4), (double)(-p_180434_6_ * f4 - p_180434_8_ * f4)),
-            new Vec3d((double)(-p_180434_4_ * f4 + p_180434_7_ * f4), (double)(p_180434_5_ * f4), (double)(-p_180434_6_ * f4 + p_180434_8_ * f4)),
-            new Vec3d((double)(p_180434_4_ * f4 + p_180434_7_ * f4), (double)(p_180434_5_ * f4), (double)(p_180434_6_ * f4 + p_180434_8_ * f4)),
-            new Vec3d((double)(p_180434_4_ * f4 - p_180434_7_ * f4), (double)(-p_180434_5_ * f4), (double)(p_180434_6_ * f4 - p_180434_8_ * f4))
-        };
+        Vec3d[] avec3d = new Vec3d[] {new Vec3d((double)(-p_180434_4_ * f4 - p_180434_7_ * f4), (double)(-p_180434_5_ * f4), (double)(-p_180434_6_ * f4 - p_180434_8_ * f4)), new Vec3d((double)(-p_180434_4_ * f4 + p_180434_7_ * f4), (double)(p_180434_5_ * f4), (double)(-p_180434_6_ * f4 + p_180434_8_ * f4)), new Vec3d((double)(p_180434_4_ * f4 + p_180434_7_ * f4), (double)(p_180434_5_ * f4), (double)(p_180434_6_ * f4 + p_180434_8_ * f4)), new Vec3d((double)(p_180434_4_ * f4 - p_180434_7_ * f4), (double)(-p_180434_5_ * f4), (double)(p_180434_6_ * f4 - p_180434_8_ * f4))};
 
         if (this.field_190014_F != 0.0F)
         {
@@ -212,34 +200,16 @@
             float f12 = MathHelper.func_76126_a(f8 * 0.5F) * (float)field_190016_K.field_72449_c;
             Vec3d vec3d = new Vec3d((double)f10, (double)f11, (double)f12);
 
-            for (int l = 0; l < 4; l++)
+            for (int l = 0; l < 4; ++l)
             {
-                avec3d[l] = vec3d.func_186678_a(2.0 * avec3d[l].func_72430_b(vec3d))
-                            .func_178787_e(avec3d[l].func_186678_a((double)(f9 * f9) - vec3d.func_72430_b(vec3d)))
-                            .func_178787_e(vec3d.func_72431_c(avec3d[l]).func_186678_a((double)(2.0F * f9)));
+                avec3d[l] = vec3d.func_186678_a(2.0D * avec3d[l].func_72430_b(vec3d)).func_178787_e(avec3d[l].func_186678_a((double)(f9 * f9) - vec3d.func_72430_b(vec3d))).func_178787_e(vec3d.func_72431_c(avec3d[l]).func_186678_a((double)(2.0F * f9)));
             }
         }
 
-        p_180434_1_.func_181662_b((double)f5 + avec3d[0].field_72450_a, (double)f6 + avec3d[0].field_72448_b, (double)f7 + avec3d[0].field_72449_c)
-        .func_187315_a((double)f1, (double)f3)
-        .func_181666_a(this.field_70552_h, this.field_70553_i, this.field_70551_j, this.field_82339_as)
-        .func_187314_a(j, k)
-        .func_181675_d();
-        p_180434_1_.func_181662_b((double)f5 + avec3d[1].field_72450_a, (double)f6 + avec3d[1].field_72448_b, (double)f7 + avec3d[1].field_72449_c)
-        .func_187315_a((double)f1, (double)f2)
-        .func_181666_a(this.field_70552_h, this.field_70553_i, this.field_70551_j, this.field_82339_as)
-        .func_187314_a(j, k)
-        .func_181675_d();
-        p_180434_1_.func_181662_b((double)f5 + avec3d[2].field_72450_a, (double)f6 + avec3d[2].field_72448_b, (double)f7 + avec3d[2].field_72449_c)
-        .func_187315_a((double)f, (double)f2)
-        .func_181666_a(this.field_70552_h, this.field_70553_i, this.field_70551_j, this.field_82339_as)
-        .func_187314_a(j, k)
-        .func_181675_d();
-        p_180434_1_.func_181662_b((double)f5 + avec3d[3].field_72450_a, (double)f6 + avec3d[3].field_72448_b, (double)f7 + avec3d[3].field_72449_c)
-        .func_187315_a((double)f, (double)f3)
-        .func_181666_a(this.field_70552_h, this.field_70553_i, this.field_70551_j, this.field_82339_as)
-        .func_187314_a(j, k)
-        .func_181675_d();
+        p_180434_1_.func_181662_b((double)f5 + avec3d[0].field_72450_a, (double)f6 + avec3d[0].field_72448_b, (double)f7 + avec3d[0].field_72449_c).func_187315_a((double)f1, (double)f3).func_181666_a(this.field_70552_h, this.field_70553_i, this.field_70551_j, this.field_82339_as).func_187314_a(j, k).func_181675_d();
+        p_180434_1_.func_181662_b((double)f5 + avec3d[1].field_72450_a, (double)f6 + avec3d[1].field_72448_b, (double)f7 + avec3d[1].field_72449_c).func_187315_a((double)f1, (double)f2).func_181666_a(this.field_70552_h, this.field_70553_i, this.field_70551_j, this.field_82339_as).func_187314_a(j, k).func_181675_d();
+        p_180434_1_.func_181662_b((double)f5 + avec3d[2].field_72450_a, (double)f6 + avec3d[2].field_72448_b, (double)f7 + avec3d[2].field_72449_c).func_187315_a((double)f, (double)f2).func_181666_a(this.field_70552_h, this.field_70553_i, this.field_70551_j, this.field_82339_as).func_187314_a(j, k).func_181675_d();
+        p_180434_1_.func_181662_b((double)f5 + avec3d[3].field_72450_a, (double)f6 + avec3d[3].field_72448_b, (double)f7 + avec3d[3].field_72449_c).func_187315_a((double)f, (double)f3).func_181666_a(this.field_70552_h, this.field_70553_i, this.field_70551_j, this.field_82339_as).func_187314_a(j, k).func_181675_d();
     }
 
     public int func_70537_b()
@@ -276,29 +246,12 @@
 
     public void func_94053_h()
     {
-        this.field_94054_b++;
+        ++this.field_94054_b;
     }
 
-    @Override
     public String toString()
     {
-        return this.getClass().getSimpleName()
-               + ", Pos ("
-               + this.field_187126_f
-               + ","
-               + this.field_187127_g
-               + ","
-               + this.field_187128_h
-               + "), RGBA ("
-               + this.field_70552_h
-               + ","
-               + this.field_70553_i
-               + ","
-               + this.field_70551_j
-               + ","
-               + this.field_82339_as
-               + "), Age "
-               + this.field_70546_d;
+        return this.getClass().getSimpleName() + ", Pos (" + this.field_187126_f + "," + this.field_187127_g + "," + this.field_187128_h + "), RGBA (" + this.field_70552_h + "," + this.field_70553_i + "," + this.field_70551_j + "," + this.field_82339_as + "), Age " + this.field_70546_d;
     }
 
     public void func_187112_i()
@@ -312,17 +265,8 @@
         {
             this.field_187134_n = p_187115_1_;
             this.field_187135_o = p_187115_2_;
-            AxisAlignedBB axisalignedbb = this.func_187116_l();
-            this.func_187108_a(
-                new AxisAlignedBB(
-                    axisalignedbb.field_72340_a,
-                    axisalignedbb.field_72338_b,
-                    axisalignedbb.field_72339_c,
-                    axisalignedbb.field_72340_a + (double)this.field_187134_n,
-                    axisalignedbb.field_72338_b + (double)this.field_187135_o,
-                    axisalignedbb.field_72339_c + (double)this.field_187134_n
-                )
-            );
+            // FORGE: Fix MC-12269 - Glitchy movement when setSize is called without setPosition
+            func_187109_b(field_187126_f, field_187127_g, field_187128_h);
         }
     }
 
@@ -333,43 +277,39 @@
         this.field_187128_h = p_187109_5_;
         float f = this.field_187134_n / 2.0F;
         float f1 = this.field_187135_o;
-        this.func_187108_a(
-            new AxisAlignedBB(
-                p_187109_1_ - (double)f, p_187109_3_, p_187109_5_ - (double)f, p_187109_1_ + (double)f, p_187109_3_ + (double)f1, p_187109_5_ + (double)f
-            )
-        );
+        this.func_187108_a(new AxisAlignedBB(p_187109_1_ - (double)f, p_187109_3_, p_187109_5_ - (double)f, p_187109_1_ + (double)f, p_187109_3_ + (double)f1, p_187109_5_ + (double)f));
     }
 
     public void func_187110_a(double p_187110_1_, double p_187110_3_, double p_187110_5_)
     {
-        double d0 = p_187110_1_;
-        double d1 = p_187110_3_;
-        double d2 = p_187110_5_;
+        double d0 = p_187110_3_;
+        double origX = p_187110_1_;
+        double origZ = p_187110_5_;
 
         if (this.field_190017_n)
         {
-            List<AxisAlignedBB> list = this.field_187122_b.func_184144_a(null, this.func_187116_l().func_72321_a(p_187110_1_, p_187110_3_, p_187110_5_));
+            List<AxisAlignedBB> list = this.field_187122_b.func_184144_a((Entity)null, this.func_187116_l().func_72321_a(p_187110_1_, p_187110_3_, p_187110_5_));
 
             for (AxisAlignedBB axisalignedbb : list)
             {
                 p_187110_3_ = axisalignedbb.func_72323_b(this.func_187116_l(), p_187110_3_);
             }
 
-            this.func_187108_a(this.func_187116_l().func_72317_d(0.0, p_187110_3_, 0.0));
+            this.func_187108_a(this.func_187116_l().func_72317_d(0.0D, p_187110_3_, 0.0D));
 
             for (AxisAlignedBB axisalignedbb1 : list)
             {
                 p_187110_1_ = axisalignedbb1.func_72316_a(this.func_187116_l(), p_187110_1_);
             }
 
-            this.func_187108_a(this.func_187116_l().func_72317_d(p_187110_1_, 0.0, 0.0));
+            this.func_187108_a(this.func_187116_l().func_72317_d(p_187110_1_, 0.0D, 0.0D));
 
             for (AxisAlignedBB axisalignedbb2 : list)
             {
                 p_187110_5_ = axisalignedbb2.func_72322_c(this.func_187116_l(), p_187110_5_);
             }
 
-            this.func_187108_a(this.func_187116_l().func_72317_d(0.0, 0.0, p_187110_5_));
+            this.func_187108_a(this.func_187116_l().func_72317_d(0.0D, 0.0D, p_187110_5_));
         }
         else
         {
@@ -377,25 +317,25 @@
         }
 
         this.func_187118_j();
-        this.field_187132_l = d1 != p_187110_3_ && d1 < 0.0;
+        this.field_187132_l = d0 != p_187110_3_ && d0 < 0.0D;
 
-        if (d0 != p_187110_1_)
+        if (origX != p_187110_1_)
         {
-            this.field_187129_i = 0.0;
+            this.field_187129_i = 0.0D;
         }
 
-        if (d2 != p_187110_5_)
+        if (origZ != p_187110_5_)
         {
-            this.field_187131_k = 0.0;
+            this.field_187131_k = 0.0D;
         }
     }
 
     protected void func_187118_j()
     {
         AxisAlignedBB axisalignedbb = this.func_187116_l();
-        this.field_187126_f = (axisalignedbb.field_72340_a + axisalignedbb.field_72336_d) / 2.0;
+        this.field_187126_f = (axisalignedbb.field_72340_a + axisalignedbb.field_72336_d) / 2.0D;
         this.field_187127_g = axisalignedbb.field_72338_b;
-        this.field_187128_h = (axisalignedbb.field_72339_c + axisalignedbb.field_72334_f) / 2.0;
+        this.field_187128_h = (axisalignedbb.field_72339_c + axisalignedbb.field_72334_f) / 2.0D;
     }
 
     public int func_189214_a(float p_189214_1_)
