--- before/net/minecraft/client/particle/ParticleFirework.java
+++ after/net/minecraft/client/particle/ParticleFirework.java
@@ -279,7 +279,7 @@
             private boolean func_92037_i()
             {
                 Minecraft minecraft = Minecraft.func_71410_x();
-                return minecraft == null || minecraft.func_175606_aa() == null || !(minecraft.func_175606_aa().func_70092_e(this.field_187126_f, this.field_187127_g, this.field_187128_h) < 256.0D);
+                return minecraft == null || minecraft.func_175606_aa() == null || minecraft.func_175606_aa().func_70092_e(this.field_187126_f, this.field_187127_g, this.field_187128_h) >= 256.0D;
             }
 
             private void func_92034_a(double p_92034_1_, double p_92034_3_, double p_92034_5_, double p_92034_7_, double p_92034_9_, double p_92034_11_, int[] p_92034_13_, int[] p_92034_14_, boolean p_92034_15_, boolean p_92034_16_)
@@ -350,7 +350,7 @@
                             double d9 = (d4 + (d6 - d4) * d8) * p_92038_1_;
                             double d10 = (d5 + (d7 - d5) * d8) * p_92038_1_;
                             double d11 = d9 * Math.sin(d3);
-                            d9 *= Math.cos(d3);
+                            d9 = d9 * Math.cos(d3);
 
                             for (double d12 = -1.0D; d12 <= 1.0D; d12 += 2.0D)
                             {
