--- before/net/minecraft/client/particle/Particle.java
+++ after/net/minecraft/client/particle/Particle.java
@@ -312,17 +312,8 @@
         {
             this.width = particleWidth;
             this.height = particleHeight;
-            AxisAlignedBB axisalignedbb = this.getBoundingBox();
-            this.setBoundingBox(
-                new AxisAlignedBB(
-                    axisalignedbb.minX,
-                    axisalignedbb.minY,
-                    axisalignedbb.minZ,
-                    axisalignedbb.minX + (double)this.width,
-                    axisalignedbb.minY + (double)this.height,
-                    axisalignedbb.minZ + (double)this.width
-                )
-            );
+            // FORGE: Fix MC-12269 - Glitchy movement when setSize is called without setPosition
+            setPosition(posX, posY, posZ);
         }
     }
 
