--- before/net/minecraft/client/particle/ParticleCrit.java
+++ after/net/minecraft/client/particle/ParticleCrit.java
@@ -20,13 +20,13 @@
     protected ParticleCrit(World p_i46285_1_, double p_i46285_2_, double p_i46285_4_, double p_i46285_6_, double p_i46285_8_, double p_i46285_10_, double p_i46285_12_, float p_i46285_14_)
     {
         super(p_i46285_1_, p_i46285_2_, p_i46285_4_, p_i46285_6_, 0.0D, 0.0D, 0.0D);
-        this.field_187129_i *= (double)0.1F;
-        this.field_187130_j *= (double)0.1F;
-        this.field_187131_k *= (double)0.1F;
+        this.field_187129_i *= 0.10000000149011612D;
+        this.field_187130_j *= 0.10000000149011612D;
+        this.field_187131_k *= 0.10000000149011612D;
         this.field_187129_i += p_i46285_8_ * 0.4D;
         this.field_187130_j += p_i46285_10_ * 0.4D;
         this.field_187131_k += p_i46285_12_ * 0.4D;
-        float f = (float)(Math.random() * (double)0.3F + (double)0.6F);
+        float f = (float)(Math.random() * 0.30000001192092896D + 0.6000000238418579D);
         this.field_70552_h = f;
         this.field_70553_i = f;
         this.field_70551_j = f;
@@ -61,15 +61,15 @@
         this.func_187110_a(this.field_187129_i, this.field_187130_j, this.field_187131_k);
         this.field_70553_i = (float)((double)this.field_70553_i * 0.96D);
         this.field_70551_j = (float)((double)this.field_70551_j * 0.9D);
-        this.field_187129_i *= (double)0.7F;
-        this.field_187130_j *= (double)0.7F;
-        this.field_187131_k *= (double)0.7F;
-        this.field_187130_j -= (double)0.02F;
+        this.field_187129_i *= 0.699999988079071D;
+        this.field_187130_j *= 0.699999988079071D;
+        this.field_187131_k *= 0.699999988079071D;
+        this.field_187130_j -= 0.019999999552965164D;
 
         if (this.field_187132_l)
         {
-            this.field_187129_i *= (double)0.7F;
-            this.field_187131_k *= (double)0.7F;
+            this.field_187129_i *= 0.699999988079071D;
+            this.field_187131_k *= 0.699999988079071D;
         }
     }
 
