--- before/net/minecraft/client/particle/ParticleDrip.java
+++ after/net/minecraft/client/particle/ParticleDrip.java
@@ -86,9 +86,9 @@
         }
 
         this.func_187110_a(this.field_187129_i, this.field_187130_j, this.field_187131_k);
-        this.field_187129_i *= (double)0.98F;
-        this.field_187130_j *= (double)0.98F;
-        this.field_187131_k *= (double)0.98F;
+        this.field_187129_i *= 0.9800000190734863D;
+        this.field_187130_j *= 0.9800000190734863D;
+        this.field_187131_k *= 0.9800000190734863D;
 
         if (this.field_70547_e-- <= 0)
         {
@@ -107,8 +107,8 @@
                 this.func_70536_a(114);
             }
 
-            this.field_187129_i *= (double)0.7F;
-            this.field_187131_k *= (double)0.7F;
+            this.field_187129_i *= 0.699999988079071D;
+            this.field_187131_k *= 0.699999988079071D;
         }
 
         BlockPos blockpos = new BlockPos(this.field_187126_f, this.field_187127_g, this.field_187128_h);
@@ -121,7 +121,7 @@
 
             if (iblockstate.func_177230_c() instanceof BlockLiquid)
             {
-                d0 = (double)BlockLiquid.func_149801_b(iblockstate.func_177229_b(BlockLiquid.field_176367_b));
+                d0 = (double)BlockLiquid.func_149801_b(((Integer)iblockstate.func_177229_b(BlockLiquid.field_176367_b)).intValue());
             }
 
             double d1 = (double)(MathHelper.func_76128_c(this.field_187127_g) + 1) - d0;
