--- before/net/minecraft/client/particle/ParticleCloud.java
+++ after/net/minecraft/client/particle/ParticleCloud.java
@@ -18,13 +18,13 @@
     {
         super(p_i1221_1_, p_i1221_2_, p_i1221_4_, p_i1221_6_, 0.0D, 0.0D, 0.0D);
         float f = 2.5F;
-        this.field_187129_i *= (double)0.1F;
-        this.field_187130_j *= (double)0.1F;
-        this.field_187131_k *= (double)0.1F;
+        this.field_187129_i *= 0.10000000149011612D;
+        this.field_187130_j *= 0.10000000149011612D;
+        this.field_187131_k *= 0.10000000149011612D;
         this.field_187129_i += p_i1221_8_;
         this.field_187130_j += p_i1221_10_;
         this.field_187131_k += p_i1221_12_;
-        float f1 = 1.0F - (float)(Math.random() * (double)0.3F);
+        float f1 = 1.0F - (float)(Math.random() * 0.30000001192092896D);
         this.field_70552_h = f1;
         this.field_70553_i = f1;
         this.field_70551_j = f1;
@@ -56,9 +56,9 @@
 
         this.func_70536_a(7 - this.field_70546_d * 8 / this.field_70547_e);
         this.func_187110_a(this.field_187129_i, this.field_187130_j, this.field_187131_k);
-        this.field_187129_i *= (double)0.96F;
-        this.field_187130_j *= (double)0.96F;
-        this.field_187131_k *= (double)0.96F;
+        this.field_187129_i *= 0.9599999785423279D;
+        this.field_187130_j *= 0.9599999785423279D;
+        this.field_187131_k *= 0.9599999785423279D;
         EntityPlayer entityplayer = this.field_187122_b.func_184137_a(this.field_187126_f, this.field_187127_g, this.field_187128_h, 2.0D, false);
 
         if (entityplayer != null)
@@ -75,8 +75,8 @@
 
         if (this.field_187132_l)
         {
-            this.field_187129_i *= (double)0.7F;
-            this.field_187131_k *= (double)0.7F;
+            this.field_187129_i *= 0.699999988079071D;
+            this.field_187131_k *= 0.699999988079071D;
         }
     }
 
