--- before/net/minecraft/client/particle/ParticleSmokeNormal.java
+++ after/net/minecraft/client/particle/ParticleSmokeNormal.java
@@ -20,13 +20,13 @@
     protected ParticleSmokeNormal(World p_i46348_1_, double p_i46348_2_, double p_i46348_4_, double p_i46348_6_, double p_i46348_8_, double p_i46348_10_, double p_i46348_12_, float p_i46348_14_)
     {
         super(p_i46348_1_, p_i46348_2_, p_i46348_4_, p_i46348_6_, 0.0D, 0.0D, 0.0D);
-        this.field_187129_i *= (double)0.1F;
-        this.field_187130_j *= (double)0.1F;
-        this.field_187131_k *= (double)0.1F;
+        this.field_187129_i *= 0.10000000149011612D;
+        this.field_187130_j *= 0.10000000149011612D;
+        this.field_187131_k *= 0.10000000149011612D;
         this.field_187129_i += p_i46348_8_;
         this.field_187130_j += p_i46348_10_;
         this.field_187131_k += p_i46348_12_;
-        float f = (float)(Math.random() * (double)0.3F);
+        float f = (float)(Math.random() * 0.30000001192092896D);
         this.field_70552_h = f;
         this.field_70553_i = f;
         this.field_70551_j = f;
@@ -66,14 +66,14 @@
             this.field_187131_k *= 1.1D;
         }
 
-        this.field_187129_i *= (double)0.96F;
-        this.field_187130_j *= (double)0.96F;
-        this.field_187131_k *= (double)0.96F;
+        this.field_187129_i *= 0.9599999785423279D;
+        this.field_187130_j *= 0.9599999785423279D;
+        this.field_187131_k *= 0.9599999785423279D;
 
         if (this.field_187132_l)
         {
-            this.field_187129_i *= (double)0.7F;
-            this.field_187131_k *= (double)0.7F;
+            this.field_187129_i *= 0.699999988079071D;
+            this.field_187131_k *= 0.699999988079071D;
         }
     }
 
