--- before/net/minecraft/client/particle/ParticlePortal.java
+++ after/net/minecraft/client/particle/ParticlePortal.java
@@ -46,7 +46,7 @@
     {
         float f = ((float)this.field_70546_d + p_180434_3_) / (float)this.field_70547_e;
         f = 1.0F - f;
-        f *= f;
+        f = f * f;
         f = 1.0F - f;
         this.field_70544_f = this.field_70571_a * f;
         super.func_180434_a(p_180434_1_, p_180434_2_, p_180434_3_, p_180434_4_, p_180434_5_, p_180434_6_, p_180434_7_, p_180434_8_);
@@ -56,11 +56,11 @@
     {
         int i = super.func_189214_a(p_189214_1_);
         float f = (float)this.field_70546_d / (float)this.field_70547_e;
-        f *= f;
-        f *= f;
+        f = f * f;
+        f = f * f;
         int j = i & 255;
         int k = i >> 16 & 255;
-        k += (int)(f * 15.0F * 16.0F);
+        k = k + (int)(f * 15.0F * 16.0F);
 
         if (k > 240)
         {
