--- before/net/minecraft/client/particle/ParticleBreaking.java
+++ after/net/minecraft/client/particle/ParticleBreaking.java
@@ -21,9 +21,9 @@
     protected ParticleBreaking(World p_i1197_1_, double p_i1197_2_, double p_i1197_4_, double p_i1197_6_, double p_i1197_8_, double p_i1197_10_, double p_i1197_12_, Item p_i1197_14_, int p_i1197_15_)
     {
         this(p_i1197_1_, p_i1197_2_, p_i1197_4_, p_i1197_6_, p_i1197_14_, p_i1197_15_);
-        this.field_187129_i *= (double)0.1F;
-        this.field_187130_j *= (double)0.1F;
-        this.field_187131_k *= (double)0.1F;
+        this.field_187129_i *= 0.10000000149011612D;
+        this.field_187130_j *= 0.10000000149011612D;
+        this.field_187131_k *= 0.10000000149011612D;
         this.field_187129_i += p_i1197_8_;
         this.field_187130_j += p_i1197_10_;
         this.field_187131_k += p_i1197_12_;
