--- before/net/minecraft/client/particle/ParticleEmitter.java
+++ after/net/minecraft/client/particle/ParticleEmitter.java
@@ -41,7 +41,7 @@
             double d1 = (double)(this.field_187136_p.nextFloat() * 2.0F - 1.0F);
             double d2 = (double)(this.field_187136_p.nextFloat() * 2.0F - 1.0F);
 
-            if (!(d0 * d0 + d1 * d1 + d2 * d2 > 1.0D))
+            if (d0 * d0 + d1 * d1 + d2 * d2 <= 1.0D)
             {
                 double d3 = this.field_174851_a.field_70165_t + d0 * (double)this.field_174851_a.field_70130_N / 4.0D;
                 double d4 = this.field_174851_a.func_174813_aQ().field_72338_b + (double)(this.field_174851_a.field_70131_O / 2.0F) + d1 * (double)this.field_174851_a.field_70131_O / 4.0D;
