--- before/net/minecraft/client/particle/ParticleRedstone.java
+++ after/net/minecraft/client/particle/ParticleRedstone.java
@@ -20,9 +20,9 @@
     protected ParticleRedstone(World p_i46350_1_, double p_i46350_2_, double p_i46350_4_, double p_i46350_6_, float p_i46350_8_, float p_i46350_9_, float p_i46350_10_, float p_i46350_11_)
     {
         super(p_i46350_1_, p_i46350_2_, p_i46350_4_, p_i46350_6_, 0.0D, 0.0D, 0.0D);
-        this.field_187129_i *= (double)0.1F;
-        this.field_187130_j *= (double)0.1F;
-        this.field_187131_k *= (double)0.1F;
+        this.field_187129_i *= 0.10000000149011612D;
+        this.field_187130_j *= 0.10000000149011612D;
+        this.field_187131_k *= 0.10000000149011612D;
 
         if (p_i46350_9_ == 0.0F)
         {
@@ -30,9 +30,9 @@
         }
 
         float f = (float)Math.random() * 0.4F + 0.6F;
-        this.field_70552_h = ((float)(Math.random() * (double)0.2F) + 0.8F) * p_i46350_9_ * f;
-        this.field_70553_i = ((float)(Math.random() * (double)0.2F) + 0.8F) * p_i46350_10_ * f;
-        this.field_70551_j = ((float)(Math.random() * (double)0.2F) + 0.8F) * p_i46350_11_ * f;
+        this.field_70552_h = ((float)(Math.random() * 0.20000000298023224D) + 0.8F) * p_i46350_9_ * f;
+        this.field_70553_i = ((float)(Math.random() * 0.20000000298023224D) + 0.8F) * p_i46350_10_ * f;
+        this.field_70551_j = ((float)(Math.random() * 0.20000000298023224D) + 0.8F) * p_i46350_11_ * f;
         this.field_70544_f *= 0.75F;
         this.field_70544_f *= p_i46350_8_;
         this.field_70570_a = this.field_70544_f;
@@ -68,14 +68,14 @@
             this.field_187131_k *= 1.1D;
         }
 
-        this.field_187129_i *= (double)0.96F;
-        this.field_187130_j *= (double)0.96F;
-        this.field_187131_k *= (double)0.96F;
+        this.field_187129_i *= 0.9599999785423279D;
+        this.field_187130_j *= 0.9599999785423279D;
+        this.field_187131_k *= 0.9599999785423279D;
 
         if (this.field_187132_l)
         {
-            this.field_187129_i *= (double)0.7F;
-            this.field_187131_k *= (double)0.7F;
+            this.field_187129_i *= 0.699999988079071D;
+            this.field_187131_k *= 0.699999988079071D;
         }
     }
 
