--- before/net/minecraft/client/particle/ParticleBubble.java
+++ after/net/minecraft/client/particle/ParticleBubble.java
@@ -18,9 +18,9 @@
         this.func_70536_a(32);
         this.func_187115_a(0.02F, 0.02F);
         this.field_70544_f *= this.field_187136_p.nextFloat() * 0.6F + 0.2F;
-        this.field_187129_i = p_i1198_8_ * (double)0.2F + (Math.random() * 2.0D - 1.0D) * (double)0.02F;
-        this.field_187130_j = p_i1198_10_ * (double)0.2F + (Math.random() * 2.0D - 1.0D) * (double)0.02F;
-        this.field_187131_k = p_i1198_12_ * (double)0.2F + (Math.random() * 2.0D - 1.0D) * (double)0.02F;
+        this.field_187129_i = p_i1198_8_ * 0.20000000298023224D + (Math.random() * 2.0D - 1.0D) * 0.019999999552965164D;
+        this.field_187130_j = p_i1198_10_ * 0.20000000298023224D + (Math.random() * 2.0D - 1.0D) * 0.019999999552965164D;
+        this.field_187131_k = p_i1198_12_ * 0.20000000298023224D + (Math.random() * 2.0D - 1.0D) * 0.019999999552965164D;
         this.field_70547_e = (int)(8.0D / (Math.random() * 0.8D + 0.2D));
     }
 
@@ -31,9 +31,9 @@
         this.field_187125_e = this.field_187128_h;
         this.field_187130_j += 0.002D;
         this.func_187110_a(this.field_187129_i, this.field_187130_j, this.field_187131_k);
-        this.field_187129_i *= (double)0.85F;
-        this.field_187130_j *= (double)0.85F;
-        this.field_187131_k *= (double)0.85F;
+        this.field_187129_i *= 0.8500000238418579D;
+        this.field_187130_j *= 0.8500000238418579D;
+        this.field_187131_k *= 0.8500000238418579D;
 
         if (this.field_187122_b.func_180495_p(new BlockPos(this.field_187126_f, this.field_187127_g, this.field_187128_h)).func_185904_a() != Material.field_151586_h)
         {
