--- before/net/minecraft/client/particle/ParticleFallingDust.java
+++ after/net/minecraft/client/particle/ParticleFallingDust.java
@@ -69,8 +69,8 @@
 
         this.func_70536_a(7 - this.field_70546_d * 8 / this.field_70547_e);
         this.func_187110_a(this.field_187129_i, this.field_187130_j, this.field_187131_k);
-        this.field_187130_j -= (double)0.003F;
-        this.field_187130_j = Math.max(this.field_187130_j, (double) - 0.14F);
+        this.field_187130_j -= 0.003000000026077032D;
+        this.field_187130_j = Math.max(this.field_187130_j, -0.14000000059604645D);
     }
 
     @SideOnly(Side.CLIENT)
