--- before/net/minecraft/client/particle/ParticleHeart.java
+++ after/net/minecraft/client/particle/ParticleHeart.java
@@ -20,9 +20,9 @@
     protected ParticleHeart(World p_i46354_1_, double p_i46354_2_, double p_i46354_4_, double p_i46354_6_, double p_i46354_8_, double p_i46354_10_, double p_i46354_12_, float p_i46354_14_)
     {
         super(p_i46354_1_, p_i46354_2_, p_i46354_4_, p_i46354_6_, 0.0D, 0.0D, 0.0D);
-        this.field_187129_i *= (double)0.01F;
-        this.field_187130_j *= (double)0.01F;
-        this.field_187131_k *= (double)0.01F;
+        this.field_187129_i *= 0.009999999776482582D;
+        this.field_187130_j *= 0.009999999776482582D;
+        this.field_187131_k *= 0.009999999776482582D;
         this.field_187130_j += 0.1D;
         this.field_70544_f *= 0.75F;
         this.field_70544_f *= p_i46354_14_;
@@ -58,14 +58,14 @@
             this.field_187131_k *= 1.1D;
         }
 
-        this.field_187129_i *= (double)0.86F;
-        this.field_187130_j *= (double)0.86F;
-        this.field_187131_k *= (double)0.86F;
+        this.field_187129_i *= 0.8600000143051147D;
+        this.field_187130_j *= 0.8600000143051147D;
+        this.field_187131_k *= 0.8600000143051147D;
 
         if (this.field_187132_l)
         {
-            this.field_187129_i *= (double)0.7F;
-            this.field_187131_k *= (double)0.7F;
+            this.field_187129_i *= 0.699999988079071D;
+            this.field_187131_k *= 0.699999988079071D;
         }
     }
 
