--- before/net/minecraft/world/biome/BiomeProvider.java
+++ after/net/minecraft/world/biome/BiomeProvider.java
@@ -21,7 +21,8 @@
     private GenLayer genBiomes;
     private GenLayer biomeIndexLayer;
     private final BiomeCache biomeCache = new BiomeCache(this);
-    private final List<Biome> biomesToSpawnIn = Lists.newArrayList(
+    private final List<Biome> biomesToSpawnIn = Lists.newArrayList(allowedBiomes);
+    public static List<Biome> allowedBiomes = Lists.newArrayList(
                 Biomes.FOREST, Biomes.PLAINS, Biomes.TAIGA, Biomes.TAIGA_HILLS, Biomes.FOREST_HILLS, Biomes.JUNGLE, Biomes.JUNGLE_HILLS
             );
 
@@ -39,6 +40,7 @@
         }
 
         GenLayer[] agenlayer = GenLayer.initializeAllBiomeGenerators(seed, worldTypeIn, this.settings);
+        agenlayer = getModdedBiomeGenerators(worldTypeIn, seed, agenlayer);
         this.genBiomes = agenlayer[0];
         this.biomeIndexLayer = agenlayer[1];
     }
@@ -205,6 +207,13 @@
     public void cleanupCache()
     {
         this.biomeCache.cleanupCache();
+    }
+
+    public GenLayer[] getModdedBiomeGenerators(WorldType worldType, long seed, GenLayer[] original)
+    {
+        net.minecraftforge.event.terraingen.WorldTypeEvent.InitBiomeGens event = new net.minecraftforge.event.terraingen.WorldTypeEvent.InitBiomeGens(worldType, seed, original);
+        net.minecraftforge.common.MinecraftForge.TERRAIN_GEN_BUS.post(event);
+        return event.getNewBiomeGens();
     }
 
     public boolean isFixedBiome()
