--- before/net/minecraft/world/biome/BiomeJungle.java
+++ after/net/minecraft/world/biome/BiomeJungle.java
@@ -25,17 +25,9 @@
 public class BiomeJungle extends Biome
 {
     private final boolean field_150614_aC;
-    private static final IBlockState field_181620_aE = Blocks.field_150364_r
-                .func_176223_P()
-                .func_177226_a(BlockOldLog.field_176301_b, BlockPlanks.EnumType.JUNGLE);
-    private static final IBlockState field_181621_aF = Blocks.field_150362_t
-                .func_176223_P()
-                .func_177226_a(BlockOldLeaf.field_176239_P, BlockPlanks.EnumType.JUNGLE)
-                .func_177226_a(BlockLeaves.field_176236_b, false);
-    private static final IBlockState field_181622_aG = Blocks.field_150362_t
-                .func_176223_P()
-                .func_177226_a(BlockOldLeaf.field_176239_P, BlockPlanks.EnumType.OAK)
-                .func_177226_a(BlockLeaves.field_176236_b, false);
+    private static final IBlockState field_181620_aE = Blocks.field_150364_r.func_176223_P().func_177226_a(BlockOldLog.field_176301_b, BlockPlanks.EnumType.JUNGLE);
+    private static final IBlockState field_181621_aF = Blocks.field_150362_t.func_176223_P().func_177226_a(BlockOldLeaf.field_176239_P, BlockPlanks.EnumType.JUNGLE).func_177226_a(BlockLeaves.field_176236_b, Boolean.valueOf(false));
+    private static final IBlockState field_181622_aG = Blocks.field_150362_t.func_176223_P().func_177226_a(BlockOldLeaf.field_176239_P, BlockPlanks.EnumType.OAK).func_177226_a(BlockLeaves.field_176236_b, Boolean.valueOf(false));
 
     public BiomeJungle(boolean p_i46705_1_, Biome.BiomeProperties p_i46705_2_)
     {
@@ -63,7 +55,6 @@
         this.field_76762_K.add(new Biome.SpawnListEntry(EntityChicken.class, 10, 4, 4));
     }
 
-    @Override
     public WorldGenAbstractTree func_150567_a(Random p_150567_1_)
     {
         if (p_150567_1_.nextInt(10) == 0)
@@ -76,29 +67,29 @@
         }
         else
         {
-            return (WorldGenAbstractTree)(!this.field_150614_aC && p_150567_1_.nextInt(3) == 0
-                                          ? new WorldGenMegaJungle(false, 10, 20, field_181620_aE, field_181621_aF)
-                                          : new WorldGenTrees(false, 4 + p_150567_1_.nextInt(7), field_181620_aE, field_181621_aF, true));
+            return (WorldGenAbstractTree)(!this.field_150614_aC && p_150567_1_.nextInt(3) == 0 ? new WorldGenMegaJungle(false, 10, 20, field_181620_aE, field_181621_aF) : new WorldGenTrees(false, 4 + p_150567_1_.nextInt(7), field_181620_aE, field_181621_aF, true));
         }
     }
 
-    @Override
     public WorldGenerator func_76730_b(Random p_76730_1_)
     {
         return p_76730_1_.nextInt(4) == 0 ? new WorldGenTallGrass(BlockTallGrass.EnumType.FERN) : new WorldGenTallGrass(BlockTallGrass.EnumType.GRASS);
     }
 
-    @Override
     public void func_180624_a(World p_180624_1_, Random p_180624_2_, BlockPos p_180624_3_)
     {
         super.func_180624_a(p_180624_1_, p_180624_2_, p_180624_3_);
         int i = p_180624_2_.nextInt(16) + 8;
         int j = p_180624_2_.nextInt(16) + 8;
-        int k = p_180624_2_.nextInt(p_180624_1_.func_175645_m(p_180624_3_.func_177982_a(i, 0, j)).func_177956_o() * 2);
-        new WorldGenMelon().func_180709_b(p_180624_1_, p_180624_2_, p_180624_3_.func_177982_a(i, k, j));
+        int height = p_180624_1_.func_175645_m(p_180624_3_.func_177982_a(i, 0, j)).func_177956_o() * 2; // could == 0, which crashes nextInt
+        if (height < 1) height = 1;
+        int k = p_180624_2_.nextInt(height);
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, new net.minecraft.util.math.ChunkPos(p_180624_3_), p_180624_3_.func_177982_a(i, k, j), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.PUMPKIN))
+        (new WorldGenMelon()).func_180709_b(p_180624_1_, p_180624_2_, p_180624_3_.func_177982_a(i, k, j));
         WorldGenVines worldgenvines = new WorldGenVines();
 
-        for (int j1 = 0; j1 < 50; j1++)
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, new net.minecraft.util.math.ChunkPos(p_180624_3_), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.GRASS))
+        for (int j1 = 0; j1 < 50; ++j1)
         {
             k = p_180624_2_.nextInt(16) + 8;
             int l = 128;
