--- before/net/minecraft/world/biome/BiomeDesert.java
+++ after/net/minecraft/world/biome/BiomeDesert.java
@@ -48,6 +48,7 @@
     {
         super.decorate(worldIn, rand, pos);
 
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.DESERT_WELL))
         if (rand.nextInt(1000) == 0)
         {
             int i = rand.nextInt(16) + 8;
@@ -56,6 +57,7 @@
             new WorldGenDesertWells().generate(worldIn, rand, blockpos);
         }
 
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.FOSSIL))
         if (rand.nextInt(64) == 0)
         {
             new WorldGenFossils().generate(worldIn, rand, pos);
