--- before/net/minecraft/world/biome/BiomeTaiga.java
+++ after/net/minecraft/world/biome/BiomeTaiga.java
@@ -48,7 +48,6 @@
         }
     }
 
-    @Override
     public WorldGenAbstractTree func_150567_a(Random p_150567_1_)
     {
         if ((this.field_150644_aH == BiomeTaiga.Type.MEGA || this.field_150644_aH == BiomeTaiga.Type.MEGA_SPRUCE) && p_150567_1_.nextInt(3) == 0)
@@ -61,16 +60,14 @@
         }
     }
 
-    @Override
     public WorldGenerator func_76730_b(Random p_76730_1_)
     {
         return p_76730_1_.nextInt(5) > 0 ? new WorldGenTallGrass(BlockTallGrass.EnumType.FERN) : new WorldGenTallGrass(BlockTallGrass.EnumType.GRASS);
     }
 
-    @Override
     public void func_180624_a(World p_180624_1_, Random p_180624_2_, BlockPos p_180624_3_)
     {
-        if (this.field_150644_aH == BiomeTaiga.Type.MEGA || this.field_150644_aH == BiomeTaiga.Type.MEGA_SPRUCE)
+        if ((this.field_150644_aH == BiomeTaiga.Type.MEGA || this.field_150644_aH == BiomeTaiga.Type.MEGA_SPRUCE) && net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, new net.minecraft.util.math.ChunkPos(p_180624_3_), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.ROCK))
         {
             int i = p_180624_2_.nextInt(3);
 
@@ -85,6 +82,7 @@
 
         field_180280_ag.func_180710_a(BlockDoublePlant.EnumPlantType.FERN);
 
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(p_180624_1_, p_180624_2_, new net.minecraft.util.math.ChunkPos(p_180624_3_), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.FLOWERS))
         for (int i1 = 0; i1 < 7; ++i1)
         {
             int j1 = p_180624_2_.nextInt(16) + 8;
@@ -96,7 +94,6 @@
         super.func_180624_a(p_180624_1_, p_180624_2_, p_180624_3_);
     }
 
-    @Override
     public void func_180622_a(World p_180622_1_, Random p_180622_2_, ChunkPrimer p_180622_3_, int p_180622_4_, int p_180622_5_, double p_180622_6_)
     {
         if (this.field_150644_aH == BiomeTaiga.Type.MEGA || this.field_150644_aH == BiomeTaiga.Type.MEGA_SPRUCE)
@@ -104,11 +101,11 @@
             this.field_76752_A = Blocks.field_150349_c.func_176223_P();
             this.field_76753_B = Blocks.field_150346_d.func_176223_P();
 
-            if (p_180622_6_ > 1.75)
+            if (p_180622_6_ > 1.75D)
             {
                 this.field_76752_A = Blocks.field_150346_d.func_176223_P().func_177226_a(BlockDirt.field_176386_a, BlockDirt.DirtType.COARSE_DIRT);
             }
-            else if (p_180622_6_ > -0.95)
+            else if (p_180622_6_ > -0.95D)
             {
                 this.field_76752_A = Blocks.field_150346_d.func_176223_P().func_177226_a(BlockDirt.field_176386_a, BlockDirt.DirtType.PODZOL);
             }
