--- before/net/minecraft/world/biome/BiomeVoidDecorator.java
+++ after/net/minecraft/world/biome/BiomeVoidDecorator.java
@@ -13,7 +13,7 @@
         int i = 16;
         double d0 = blockpos.func_177951_i(p_180292_4_.func_177982_a(8, blockpos.func_177956_o(), 8));
 
-        if (!(d0 > 1024.0D))
+        if (d0 <= 1024.0D)
         {
             BlockPos blockpos1 = new BlockPos(blockpos.func_177958_n() - 16, blockpos.func_177956_o() - 1, blockpos.func_177952_p() - 16);
             BlockPos blockpos2 = new BlockPos(blockpos.func_177958_n() + 16, blockpos.func_177956_o() - 1, blockpos.func_177952_p() + 16);
