--- before/net/minecraft/world/biome/BiomeTaiga.java
+++ after/net/minecraft/world/biome/BiomeTaiga.java
@@ -48,7 +48,6 @@
         }
     }
 
-    @Override
     public WorldGenAbstractTree getRandomTreeFeature(Random rand)
     {
         if ((this.type == BiomeTaiga.Type.MEGA || this.type == BiomeTaiga.Type.MEGA_SPRUCE) && rand.nextInt(3) == 0)
@@ -61,20 +60,18 @@
         }
     }
 
-    @Override
     public WorldGenerator getRandomWorldGenForGrass(Random rand)
     {
         return rand.nextInt(5) > 0 ? new WorldGenTallGrass(BlockTallGrass.EnumType.FERN) : new WorldGenTallGrass(BlockTallGrass.EnumType.GRASS);
     }
 
-    @Override
     public void decorate(World worldIn, Random rand, BlockPos pos)
     {
-        if (this.type == BiomeTaiga.Type.MEGA || this.type == BiomeTaiga.Type.MEGA_SPRUCE)
+        if ((this.type == BiomeTaiga.Type.MEGA || this.type == BiomeTaiga.Type.MEGA_SPRUCE) && net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.ROCK))
         {
             int i = rand.nextInt(3);
 
-            for (int j = 0; j < i; j++)
+            for (int j = 0; j < i; ++j)
             {
                 int k = rand.nextInt(16) + 8;
                 int l = rand.nextInt(16) + 8;
@@ -85,7 +82,8 @@
 
         DOUBLE_PLANT_GENERATOR.setPlantType(BlockDoublePlant.EnumPlantType.FERN);
 
-        for (int i1 = 0; i1 < 7; i1++)
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.FLOWERS))
+        for (int i1 = 0; i1 < 7; ++i1)
         {
             int j1 = rand.nextInt(16) + 8;
             int k1 = rand.nextInt(16) + 8;
@@ -96,7 +94,6 @@
         super.decorate(worldIn, rand, pos);
     }
 
-    @Override
     public void genTerrainBlocks(World worldIn, Random rand, ChunkPrimer chunkPrimerIn, int x, int z, double noiseVal)
     {
         if (this.type == BiomeTaiga.Type.MEGA || this.type == BiomeTaiga.Type.MEGA_SPRUCE)
@@ -104,11 +101,11 @@
             this.topBlock = Blocks.GRASS.getDefaultState();
             this.fillerBlock = Blocks.DIRT.getDefaultState();
 
-            if (noiseVal > 1.75)
+            if (noiseVal > 1.75D)
             {
                 this.topBlock = Blocks.DIRT.getDefaultState().withProperty(BlockDirt.VARIANT, BlockDirt.DirtType.COARSE_DIRT);
             }
-            else if (noiseVal > -0.95)
+            else if (noiseVal > -0.95D)
             {
                 this.topBlock = Blocks.DIRT.getDefaultState().withProperty(BlockDirt.VARIANT, BlockDirt.DirtType.PODZOL);
             }
