--- before/net/minecraft/world/biome/BiomeSnow.java
+++ after/net/minecraft/world/biome/BiomeSnow.java
@@ -58,7 +58,7 @@
     @Override
     public void decorate(World worldIn, Random rand, BlockPos pos)
     {
-        if (this.superIcy)
+        if (this.superIcy && net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.ICE))
         {
             for (int i = 0; i < 3; i++)
             {
