--- before/net/minecraft/world/biome/BiomeMesa.java
+++ after/net/minecraft/world/biome/BiomeMesa.java
@@ -74,14 +74,14 @@
     @Override
     public int getFoliageColorAtPos(BlockPos pos)
     {
-        return 10387789;
+        return getModdedBiomeFoliageColor(10387789);
     }
 
     @SideOnly(Side.CLIENT)
     @Override
     public int getGrassColorAtPos(BlockPos pos)
     {
-        return 9470285;
+        return getModdedBiomeGrassColor(9470285);
     }
 
     @Override
@@ -340,6 +340,7 @@
         protected void generateOres(World worldIn, Random random)
         {
             super.generateOres(worldIn, random);
+            if (net.minecraftforge.event.terraingen.TerrainGen.generateOre(worldIn, random, goldGen, chunkPos, net.minecraftforge.event.terraingen.OreGenEvent.GenerateMinable.EventType.GOLD))
             this.genStandardOre1(worldIn, random, 20, this.goldGen, 32, 80);
         }
     }
