--- before/net/minecraft/world/biome/BiomeDesert.java
+++ after/net/minecraft/world/biome/BiomeDesert.java
@@ -43,22 +43,23 @@
         this.spawnableMonsterList.add(new Biome.SpawnListEntry(EntityHusk.class, 80, 4, 4));
     }
 
-    @Override
     public void decorate(World worldIn, Random rand, BlockPos pos)
     {
         super.decorate(worldIn, rand, pos);
 
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.DESERT_WELL))
         if (rand.nextInt(1000) == 0)
         {
             int i = rand.nextInt(16) + 8;
             int j = rand.nextInt(16) + 8;
             BlockPos blockpos = worldIn.getHeight(pos.add(i, 0, j)).up();
-            new WorldGenDesertWells().generate(worldIn, rand, blockpos);
+            (new WorldGenDesertWells()).generate(worldIn, rand, blockpos);
         }
 
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.FOSSIL))
         if (rand.nextInt(64) == 0)
         {
-            new WorldGenFossils().generate(worldIn, rand, pos);
+            (new WorldGenFossils()).generate(worldIn, rand, pos);
         }
     }
 }
