--- before/net/minecraft/world/biome/BiomeTaiga.java
+++ after/net/minecraft/world/biome/BiomeTaiga.java
@@ -70,7 +70,7 @@
     @Override
     public void decorate(World worldIn, Random rand, BlockPos pos)
     {
-        if (this.type == BiomeTaiga.Type.MEGA || this.type == BiomeTaiga.Type.MEGA_SPRUCE)
+        if ((this.type == BiomeTaiga.Type.MEGA || this.type == BiomeTaiga.Type.MEGA_SPRUCE) && net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.ROCK))
         {
             int i = rand.nextInt(3);
 
@@ -85,6 +85,7 @@
 
         DOUBLE_PLANT_GENERATOR.setPlantType(BlockDoublePlant.EnumPlantType.FERN);
 
+        if(net.minecraftforge.event.terraingen.TerrainGen.decorate(worldIn, rand, new net.minecraft.util.math.ChunkPos(pos), net.minecraftforge.event.terraingen.DecorateBiomeEvent.Decorate.EventType.FLOWERS))
         for (int i1 = 0; i1 < 7; i1++)
         {
             int j1 = rand.nextInt(16) + 8;
