--- before/net/minecraft/world/biome/BiomeMesa.java
+++ after/net/minecraft/world/biome/BiomeMesa.java
@@ -19,16 +19,12 @@
 
 public class BiomeMesa extends Biome
 {
-    protected static final IBlockState COARSE_DIRT = Blocks.DIRT
-                .getDefaultState()
-                .withProperty(BlockDirt.VARIANT, BlockDirt.DirtType.COARSE_DIRT);
+    protected static final IBlockState COARSE_DIRT = Blocks.DIRT.getDefaultState().withProperty(BlockDirt.VARIANT, BlockDirt.DirtType.COARSE_DIRT);
     protected static final IBlockState GRASS = Blocks.GRASS.getDefaultState();
     protected static final IBlockState HARDENED_CLAY = Blocks.HARDENED_CLAY.getDefaultState();
     protected static final IBlockState STAINED_HARDENED_CLAY = Blocks.STAINED_HARDENED_CLAY.getDefaultState();
     protected static final IBlockState ORANGE_STAINED_HARDENED_CLAY = STAINED_HARDENED_CLAY.withProperty(BlockColored.COLOR, EnumDyeColor.ORANGE);
-    protected static final IBlockState RED_SAND = Blocks.SAND
-                .getDefaultState()
-                .withProperty(BlockSand.VARIANT, BlockSand.EnumType.RED_SAND);
+    protected static final IBlockState RED_SAND = Blocks.SAND.getDefaultState().withProperty(BlockSand.VARIANT, BlockSand.EnumType.RED_SAND);
     private IBlockState[] clayBands;
     private long worldSeed;
     private NoiseGeneratorPerlin pillarNoise;
@@ -58,33 +54,16 @@
         }
     }
 
-    @Override
     public BiomeDecorator createBiomeDecorator()
     {
         return new BiomeMesa.Decorator();
     }
 
-    @Override
     public WorldGenAbstractTree getRandomTreeFeature(Random rand)
     {
         return TREE_FEATURE;
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
-    public int getFoliageColorAtPos(BlockPos pos)
-    {
-        return 10387789;
-    }
-
-    @SideOnly(Side.CLIENT)
-    @Override
-    public int getGrassColorAtPos(BlockPos pos)
-    {
-        return 9470285;
-    }
-
-    @Override
     public void genTerrainBlocks(World worldIn, Random rand, ChunkPrimer chunkPrimerIn, int x, int z, double noiseVal)
     {
         if (this.clayBands == null || this.worldSeed != worldIn.getSeed())
@@ -100,27 +79,27 @@
         }
 
         this.worldSeed = worldIn.getSeed();
-        double d4 = 0.0;
+        double d4 = 0.0D;
 
         if (this.brycePillars)
         {
             int i = (x & -16) + (z & 15);
             int j = (z & -16) + (x & 15);
-            double d0 = Math.min(Math.abs(noiseVal), this.pillarNoise.getValue(i * 0.25, j * 0.25));
+            double d0 = Math.min(Math.abs(noiseVal), this.pillarNoise.getValue((double)i * 0.25D, (double)j * 0.25D));
 
-            if (d0 > 0.0)
+            if (d0 > 0.0D)
             {
-                double d1 = 0.001953125;
-                double d2 = Math.abs(this.pillarRoofNoise.getValue(i * 0.001953125, j * 0.001953125));
-                d4 = d0 * d0 * 2.5;
-                double d3 = Math.ceil(d2 * 50.0) + 14.0;
+                double d1 = 0.001953125D;
+                double d2 = Math.abs(this.pillarRoofNoise.getValue((double)i * 0.001953125D, (double)j * 0.001953125D));
+                d4 = d0 * d0 * 2.5D;
+                double d3 = Math.ceil(d2 * 50.0D) + 14.0D;
 
                 if (d4 > d3)
                 {
                     d4 = d3;
                 }
 
-                d4 += 64.0;
+                d4 = d4 + 64.0D;
             }
         }
 
@@ -129,13 +108,13 @@
         int i2 = worldIn.getSeaLevel();
         IBlockState iblockstate = STAINED_HARDENED_CLAY;
         IBlockState iblockstate3 = this.fillerBlock;
-        int k = (int)(noiseVal / 3.0 + 3.0 + rand.nextDouble() * 0.25);
-        boolean flag = Math.cos(noiseVal / 3.0 * Math.PI) > 0.0;
+        int k = (int)(noiseVal / 3.0D + 3.0D + rand.nextDouble() * 0.25D);
+        boolean flag = Math.cos(noiseVal / 3.0D * Math.PI) > 0.0D;
         int l = -1;
         boolean flag1 = false;
         int i1 = 0;
 
-        for (int j1 = 255; j1 >= 0; j1--)
+        for (int j1 = 255; j1 >= 0; --j1)
         {
             if (chunkPrimerIn.getBlockState(l1, j1, k1).getMaterial() == Material.AIR && j1 < (int)d4)
             {
@@ -178,54 +157,60 @@
 
                         l = k + Math.max(0, j1 - i2);
 
-                        if (j1 < i2 - 1)
-                        {
-                            chunkPrimerIn.setBlockState(l1, j1, k1, iblockstate3);
-
-                            if (iblockstate3.getBlock() == Blocks.STAINED_HARDENED_CLAY)
-                            {
-                                chunkPrimerIn.setBlockState(l1, j1, k1, ORANGE_STAINED_HARDENED_CLAY);
-                            }
-                        }
-                        else if (!this.hasForest || j1 <= 86 + k * 2)
-                        {
-                            if (j1 <= i2 + 3 + k)
-                            {
-                                chunkPrimerIn.setBlockState(l1, j1, k1, this.topBlock);
-                                flag1 = true;
-                            }
-                            else
+                        if (j1 >= i2 - 1)
+                        {
+                            if (this.hasForest && j1 > 86 + k * 2)
+                            {
+                                if (flag)
+                                {
+                                    chunkPrimerIn.setBlockState(l1, j1, k1, COARSE_DIRT);
+                                }
+                                else
+                                {
+                                    chunkPrimerIn.setBlockState(l1, j1, k1, GRASS);
+                                }
+                            }
+                            else if (j1 > i2 + 3 + k)
                             {
                                 IBlockState iblockstate2;
 
-                                if (j1 < 64 || j1 > 127)
+                                if (j1 >= 64 && j1 <= 127)
+                                {
+                                    if (flag)
+                                    {
+                                        iblockstate2 = HARDENED_CLAY;
+                                    }
+                                    else
+                                    {
+                                        iblockstate2 = this.getBand(x, j1, z);
+                                    }
+                                }
+                                else
                                 {
                                     iblockstate2 = ORANGE_STAINED_HARDENED_CLAY;
                                 }
-                                else if (flag)
-                                {
-                                    iblockstate2 = HARDENED_CLAY;
-                                }
-                                else
-                                {
-                                    iblockstate2 = this.getBand(x, j1, z);
-                                }
 
                                 chunkPrimerIn.setBlockState(l1, j1, k1, iblockstate2);
                             }
-                        }
-                        else if (flag)
-                        {
-                            chunkPrimerIn.setBlockState(l1, j1, k1, COARSE_DIRT);
+                            else
+                            {
+                                chunkPrimerIn.setBlockState(l1, j1, k1, this.topBlock);
+                                flag1 = true;
+                            }
                         }
                         else
                         {
-                            chunkPrimerIn.setBlockState(l1, j1, k1, GRASS);
+                            chunkPrimerIn.setBlockState(l1, j1, k1, iblockstate3);
+
+                            if (iblockstate3.getBlock() == Blocks.STAINED_HARDENED_CLAY)
+                            {
+                                chunkPrimerIn.setBlockState(l1, j1, k1, ORANGE_STAINED_HARDENED_CLAY);
+                            }
                         }
                     }
                     else if (l > 0)
                     {
-                        l--;
+                        --l;
 
                         if (flag1)
                         {
@@ -237,7 +222,7 @@
                         }
                     }
 
-                    i1++;
+                    ++i1;
                 }
             }
         }
@@ -250,7 +235,7 @@
         Random random = new Random(p_150619_1_);
         this.clayBandsOffsetNoise = new NoiseGeneratorPerlin(random, 1);
 
-        for (int l1 = 0; l1 < 64; l1++)
+        for (int l1 = 0; l1 < 64; ++l1)
         {
             l1 += random.nextInt(5) + 1;
 
@@ -262,12 +247,12 @@
 
         int i2 = random.nextInt(4) + 2;
 
-        for (int i = 0; i < i2; i++)
+        for (int i = 0; i < i2; ++i)
         {
             int j = random.nextInt(3) + 1;
             int k = random.nextInt(64);
 
-            for (int l = 0; k + l < 64 && l < j; l++)
+            for (int l = 0; k + l < 64 && l < j; ++l)
             {
                 this.clayBands[k + l] = STAINED_HARDENED_CLAY.withProperty(BlockColored.COLOR, EnumDyeColor.YELLOW);
             }
@@ -275,12 +260,12 @@
 
         int j2 = random.nextInt(4) + 2;
 
-        for (int k2 = 0; k2 < j2; k2++)
+        for (int k2 = 0; k2 < j2; ++k2)
         {
             int i3 = random.nextInt(3) + 2;
             int l3 = random.nextInt(64);
 
-            for (int i1 = 0; l3 + i1 < 64 && i1 < i3; i1++)
+            for (int i1 = 0; l3 + i1 < 64 && i1 < i3; ++i1)
             {
                 this.clayBands[l3 + i1] = STAINED_HARDENED_CLAY.withProperty(BlockColored.COLOR, EnumDyeColor.BROWN);
             }
@@ -288,12 +273,12 @@
 
         int l2 = random.nextInt(4) + 2;
 
-        for (int j3 = 0; j3 < l2; j3++)
+        for (int j3 = 0; j3 < l2; ++j3)
         {
             int i4 = random.nextInt(3) + 1;
             int k4 = random.nextInt(64);
 
-            for (int j1 = 0; k4 + j1 < 64 && j1 < i4; j1++)
+            for (int j1 = 0; k4 + j1 < 64 && j1 < i4; ++j1)
             {
                 this.clayBands[k4 + j1] = STAINED_HARDENED_CLAY.withProperty(BlockColored.COLOR, EnumDyeColor.RED);
             }
@@ -302,12 +287,12 @@
         int k3 = random.nextInt(3) + 3;
         int j4 = 0;
 
-        for (int l4 = 0; l4 < k3; l4++)
+        for (int l4 = 0; l4 < k3; ++l4)
         {
             int i5 = 1;
             j4 += random.nextInt(16) + 4;
 
-            for (int k1 = 0; j4 + k1 < 64 && k1 < 1; k1++)
+            for (int k1 = 0; j4 + k1 < 64 && k1 < 1; ++k1)
             {
                 this.clayBands[j4 + k1] = STAINED_HARDENED_CLAY.withProperty(BlockColored.COLOR, EnumDyeColor.WHITE);
 
@@ -326,20 +311,32 @@
 
     public IBlockState getBand(int p_180629_1_, int p_180629_2_, int p_180629_3_)
     {
-        int i = (int)Math.round(this.clayBandsOffsetNoise.getValue(p_180629_1_ / 512.0, p_180629_1_ / 512.0) * 2.0);
+        int i = (int)Math.round(this.clayBandsOffsetNoise.getValue((double)p_180629_1_ / 512.0D, (double)p_180629_1_ / 512.0D) * 2.0D);
         return this.clayBands[(p_180629_2_ + i + 64) % 64];
     }
 
+    @SideOnly(Side.CLIENT)
+    public int getFoliageColorAtPos(BlockPos pos)
+    {
+        return getModdedBiomeFoliageColor(10387789);
+    }
+
+    @SideOnly(Side.CLIENT)
+    public int getGrassColorAtPos(BlockPos pos)
+    {
+        return getModdedBiomeGrassColor(9470285);
+    }
+
     class Decorator extends BiomeDecorator
     {
         private Decorator()
         {
         }
 
-        @Override
         protected void generateOres(World worldIn, Random random)
         {
             super.generateOres(worldIn, random);
+            if (net.minecraftforge.event.terraingen.TerrainGen.generateOre(worldIn, random, goldGen, chunkPos, net.minecraftforge.event.terraingen.OreGenEvent.GenerateMinable.EventType.GOLD))
             this.genStandardOre1(worldIn, random, 20, this.goldGen, 32, 80);
         }
     }
