--- before/net/minecraft/world/biome/BiomeMesa.java
+++ after/net/minecraft/world/biome/BiomeMesa.java
@@ -19,16 +19,12 @@
 
 public class BiomeMesa extends Biome
 {
-    protected static final IBlockState field_185385_y = Blocks.field_150346_d
-                .func_176223_P()
-                .func_177226_a(BlockDirt.field_176386_a, BlockDirt.DirtType.COARSE_DIRT);
+    protected static final IBlockState field_185385_y = Blocks.field_150346_d.func_176223_P().func_177226_a(BlockDirt.field_176386_a, BlockDirt.DirtType.COARSE_DIRT);
     protected static final IBlockState field_185386_z = Blocks.field_150349_c.func_176223_P();
     protected static final IBlockState field_185381_A = Blocks.field_150405_ch.func_176223_P();
     protected static final IBlockState field_185382_B = Blocks.field_150406_ce.func_176223_P();
     protected static final IBlockState field_185383_C = field_185382_B.func_177226_a(BlockColored.field_176581_a, EnumDyeColor.ORANGE);
-    protected static final IBlockState field_185384_D = Blocks.field_150354_m
-                .func_176223_P()
-                .func_177226_a(BlockSand.field_176504_a, BlockSand.EnumType.RED_SAND);
+    protected static final IBlockState field_185384_D = Blocks.field_150354_m.func_176223_P().func_177226_a(BlockSand.field_176504_a, BlockSand.EnumType.RED_SAND);
     private IBlockState[] field_150621_aC;
     private long field_150622_aD;
     private NoiseGeneratorPerlin field_150623_aE;
@@ -58,33 +54,16 @@
         }
     }
 
-    @Override
     public BiomeDecorator func_76729_a()
     {
         return new BiomeMesa.Decorator();
     }
 
-    @Override
     public WorldGenAbstractTree func_150567_a(Random p_150567_1_)
     {
         return field_76757_N;
     }
 
-    @SideOnly(Side.CLIENT)
-    @Override
-    public int func_180625_c(BlockPos p_180625_1_)
-    {
-        return 10387789;
-    }
-
-    @SideOnly(Side.CLIENT)
-    @Override
-    public int func_180627_b(BlockPos p_180627_1_)
-    {
-        return 9470285;
-    }
-
-    @Override
     public void func_180622_a(World p_180622_1_, Random p_180622_2_, ChunkPrimer p_180622_3_, int p_180622_4_, int p_180622_5_, double p_180622_6_)
     {
         if (this.field_150621_aC == null || this.field_150622_aD != p_180622_1_.func_72905_C())
@@ -100,27 +79,27 @@
         }
 
         this.field_150622_aD = p_180622_1_.func_72905_C();
-        double d4 = 0.0;
+        double d4 = 0.0D;
 
         if (this.field_150626_aH)
         {
             int i = (p_180622_4_ & -16) + (p_180622_5_ & 15);
             int j = (p_180622_5_ & -16) + (p_180622_4_ & 15);
-            double d0 = Math.min(Math.abs(p_180622_6_), this.field_150623_aE.func_151601_a((double)i * 0.25, (double)j * 0.25));
+            double d0 = Math.min(Math.abs(p_180622_6_), this.field_150623_aE.func_151601_a((double)i * 0.25D, (double)j * 0.25D));
 
-            if (d0 > 0.0)
+            if (d0 > 0.0D)
             {
-                double d1 = 0.001953125;
-                double d2 = Math.abs(this.field_150624_aF.func_151601_a((double)i * 0.001953125, (double)j * 0.001953125));
-                d4 = d0 * d0 * 2.5;
-                double d3 = Math.ceil(d2 * 50.0) + 14.0;
+                double d1 = 0.001953125D;
+                double d2 = Math.abs(this.field_150624_aF.func_151601_a((double)i * 0.001953125D, (double)j * 0.001953125D));
+                d4 = d0 * d0 * 2.5D;
+                double d3 = Math.ceil(d2 * 50.0D) + 14.0D;
 
                 if (d4 > d3)
                 {
                     d4 = d3;
                 }
 
-                d4 += 64.0;
+                d4 = d4 + 64.0D;
             }
         }
 
@@ -129,8 +108,8 @@
         int i2 = p_180622_1_.func_181545_F();
         IBlockState iblockstate = field_185382_B;
         IBlockState iblockstate3 = this.field_76753_B;
-        int k = (int)(p_180622_6_ / 3.0 + 3.0 + p_180622_2_.nextDouble() * 0.25);
-        boolean flag = Math.cos(p_180622_6_ / 3.0 * Math.PI) > 0.0;
+        int k = (int)(p_180622_6_ / 3.0D + 3.0D + p_180622_2_.nextDouble() * 0.25D);
+        boolean flag = Math.cos(p_180622_6_ / 3.0D * Math.PI) > 0.0D;
         int l = -1;
         boolean flag1 = false;
         int i1 = 0;
@@ -178,49 +157,55 @@
 
                         l = k + Math.max(0, j1 - i2);
 
-                        if (j1 < i2 - 1)
-                        {
-                            p_180622_3_.func_177855_a(l1, j1, k1, iblockstate3);
-
-                            if (iblockstate3.func_177230_c() == Blocks.field_150406_ce)
-                            {
-                                p_180622_3_.func_177855_a(l1, j1, k1, field_185383_C);
-                            }
-                        }
-                        else if (!this.field_150620_aI || j1 <= 86 + k * 2)
-                        {
-                            if (j1 <= i2 + 3 + k)
-                            {
-                                p_180622_3_.func_177855_a(l1, j1, k1, this.field_76752_A);
-                                flag1 = true;
-                            }
-                            else
+                        if (j1 >= i2 - 1)
+                        {
+                            if (this.field_150620_aI && j1 > 86 + k * 2)
+                            {
+                                if (flag)
+                                {
+                                    p_180622_3_.func_177855_a(l1, j1, k1, field_185385_y);
+                                }
+                                else
+                                {
+                                    p_180622_3_.func_177855_a(l1, j1, k1, field_185386_z);
+                                }
+                            }
+                            else if (j1 > i2 + 3 + k)
                             {
                                 IBlockState iblockstate2;
 
-                                if (j1 < 64 || j1 > 127)
+                                if (j1 >= 64 && j1 <= 127)
+                                {
+                                    if (flag)
+                                    {
+                                        iblockstate2 = field_185381_A;
+                                    }
+                                    else
+                                    {
+                                        iblockstate2 = this.func_180629_a(p_180622_4_, j1, p_180622_5_);
+                                    }
+                                }
+                                else
                                 {
                                     iblockstate2 = field_185383_C;
                                 }
-                                else if (flag)
-                                {
-                                    iblockstate2 = field_185381_A;
-                                }
-                                else
-                                {
-                                    iblockstate2 = this.func_180629_a(p_180622_4_, j1, p_180622_5_);
-                                }
 
                                 p_180622_3_.func_177855_a(l1, j1, k1, iblockstate2);
                             }
-                        }
-                        else if (flag)
-                        {
-                            p_180622_3_.func_177855_a(l1, j1, k1, field_185385_y);
+                            else
+                            {
+                                p_180622_3_.func_177855_a(l1, j1, k1, this.field_76752_A);
+                                flag1 = true;
+                            }
                         }
                         else
                         {
-                            p_180622_3_.func_177855_a(l1, j1, k1, field_185386_z);
+                            p_180622_3_.func_177855_a(l1, j1, k1, iblockstate3);
+
+                            if (iblockstate3.func_177230_c() == Blocks.field_150406_ce)
+                            {
+                                p_180622_3_.func_177855_a(l1, j1, k1, field_185383_C);
+                            }
                         }
                     }
                     else if (l > 0)
@@ -326,20 +311,32 @@
 
     public IBlockState func_180629_a(int p_180629_1_, int p_180629_2_, int p_180629_3_)
     {
-        int i = (int)Math.round(this.field_150625_aG.func_151601_a((double)p_180629_1_ / 512.0, (double)p_180629_1_ / 512.0) * 2.0);
+        int i = (int)Math.round(this.field_150625_aG.func_151601_a((double)p_180629_1_ / 512.0D, (double)p_180629_1_ / 512.0D) * 2.0D);
         return this.field_150621_aC[(p_180629_2_ + i + 64) % 64];
     }
 
+    @SideOnly(Side.CLIENT)
+    public int func_180625_c(BlockPos p_180625_1_)
+    {
+        return getModdedBiomeFoliageColor(10387789);
+    }
+
+    @SideOnly(Side.CLIENT)
+    public int func_180627_b(BlockPos p_180627_1_)
+    {
+        return getModdedBiomeGrassColor(9470285);
+    }
+
     class Decorator extends BiomeDecorator
     {
         private Decorator()
         {
         }
 
-        @Override
         protected void func_76797_b(World p_76797_1_, Random p_76797_2_)
         {
             super.func_76797_b(p_76797_1_, p_76797_2_);
+            if (net.minecraftforge.event.terraingen.TerrainGen.generateOre(p_76797_1_, p_76797_2_, field_76819_m, field_180294_c, net.minecraftforge.event.terraingen.OreGenEvent.GenerateMinable.EventType.GOLD))
             this.func_76795_a(p_76797_1_, p_76797_2_, 20, this.field_76819_m, 32, 80);
         }
     }
