--- before/net/minecraft/client/resources/data/AnimationMetadataSection.java
+++ after/net/minecraft/client/resources/data/AnimationMetadataSection.java
@@ -62,21 +62,21 @@
 
     public boolean func_110470_b(int p_110470_1_)
     {
-        return !this.field_110478_a.get(p_110470_1_).func_110495_a();
+        return !((AnimationFrame)this.field_110478_a.get(p_110470_1_)).func_110495_a();
     }
 
     public int func_110468_c(int p_110468_1_)
     {
-        return this.field_110478_a.get(p_110468_1_).func_110496_c();
+        return ((AnimationFrame)this.field_110478_a.get(p_110468_1_)).func_110496_c();
     }
 
     public Set<Integer> func_130073_e()
     {
-        Set<Integer> set = Sets.newHashSet();
+        Set<Integer> set = Sets.<Integer>newHashSet();
 
         for (AnimationFrame animationframe : this.field_110478_a)
         {
-            set.add(animationframe.func_110496_c());
+            set.add(Integer.valueOf(animationframe.func_110496_c()));
         }
 
         return set;
