--- before/net/minecraft/client/resources/data/AnimationMetadataSectionSerializer.java
+++ after/net/minecraft/client/resources/data/AnimationMetadataSectionSerializer.java
@@ -21,7 +21,7 @@
 {
     public AnimationMetadataSection deserialize(JsonElement p_deserialize_1_, Type p_deserialize_2_, JsonDeserializationContext p_deserialize_3_) throws JsonParseException
     {
-        List<AnimationFrame> list = Lists.newArrayList();
+        List<AnimationFrame> list = Lists.<AnimationFrame>newArrayList();
         JsonObject jsonobject = JsonUtils.func_151210_l(p_deserialize_1_, "metadata section");
         int i = JsonUtils.func_151208_a(jsonobject, "frametime", 1);
 
@@ -99,16 +99,16 @@
     public JsonElement serialize(AnimationMetadataSection p_serialize_1_, Type p_serialize_2_, JsonSerializationContext p_serialize_3_)
     {
         JsonObject jsonobject = new JsonObject();
-        jsonobject.addProperty("frametime", p_serialize_1_.func_110469_d());
+        jsonobject.addProperty("frametime", Integer.valueOf(p_serialize_1_.func_110469_d()));
 
         if (p_serialize_1_.func_110474_b() != -1)
         {
-            jsonobject.addProperty("width", p_serialize_1_.func_110474_b());
+            jsonobject.addProperty("width", Integer.valueOf(p_serialize_1_.func_110474_b()));
         }
 
         if (p_serialize_1_.func_110471_a() != -1)
         {
-            jsonobject.addProperty("height", p_serialize_1_.func_110471_a());
+            jsonobject.addProperty("height", Integer.valueOf(p_serialize_1_.func_110471_a()));
         }
 
         if (p_serialize_1_.func_110473_c() > 0)
@@ -120,8 +120,8 @@
                 if (p_serialize_1_.func_110470_b(i))
                 {
                     JsonObject jsonobject1 = new JsonObject();
-                    jsonobject1.addProperty("index", p_serialize_1_.func_110468_c(i));
-                    jsonobject1.addProperty("time", p_serialize_1_.func_110472_a(i));
+                    jsonobject1.addProperty("index", Integer.valueOf(p_serialize_1_.func_110468_c(i)));
+                    jsonobject1.addProperty("time", Integer.valueOf(p_serialize_1_.func_110472_a(i)));
                     jsonarray.add(jsonobject1);
                 }
                 else
