--- before/net/minecraft/client/resources/data/LanguageMetadataSectionSerializer.java
+++ after/net/minecraft/client/resources/data/LanguageMetadataSectionSerializer.java
@@ -6,8 +6,8 @@
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
 import java.lang.reflect.Type;
-import java.util.Map;
 import java.util.Set;
+import java.util.Map.Entry;
 import net.minecraft.client.resources.Language;
 import net.minecraft.util.JsonUtils;
 import net.minecraftforge.fml.relauncher.Side;
@@ -19,9 +19,9 @@
     public LanguageMetadataSection deserialize(JsonElement p_deserialize_1_, Type p_deserialize_2_, JsonDeserializationContext p_deserialize_3_) throws JsonParseException
     {
         JsonObject jsonobject = p_deserialize_1_.getAsJsonObject();
-        Set<Language> set = Sets.newHashSet();
+        Set<Language> set = Sets.<Language>newHashSet();
 
-        for (Map.Entry<String, JsonElement> entry : jsonobject.entrySet())
+        for (Entry<String, JsonElement> entry : jsonobject.entrySet())
         {
             String s = entry.getKey();
 
