--- before/net/minecraft/network/datasync/DataSerializers.java
+++ after/net/minecraft/network/datasync/DataSerializers.java
@@ -17,21 +17,20 @@
 
 public class DataSerializers
 {
-    private static final IntIdentityHashBiMap < DataSerializer<? >> field_187204_n = new IntIdentityHashBiMap<>(16);
+    private static final IntIdentityHashBiMap < DataSerializer<? >> field_187204_n = new IntIdentityHashBiMap < DataSerializer<? >> (16);
     public static final DataSerializer<Byte> field_187191_a = new DataSerializer<Byte>()
     {
         public void func_187160_a(PacketBuffer p_187160_1_, Byte p_187160_2_)
         {
-            p_187160_1_.writeByte(p_187160_2_);
+            p_187160_1_.writeByte(p_187160_2_.byteValue());
         }
         public Byte func_187159_a(PacketBuffer p_187159_1_) throws IOException
         {
             return p_187159_1_.readByte();
         }
-        @Override
         public DataParameter<Byte> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<Byte>(p_187161_1_, this);
         }
         public Byte func_192717_a(Byte p_192717_1_)
         {
@@ -42,16 +41,15 @@
     {
         public void func_187160_a(PacketBuffer p_187160_1_, Integer p_187160_2_)
         {
-            p_187160_1_.func_150787_b(p_187160_2_);
+            p_187160_1_.func_150787_b(p_187160_2_.intValue());
         }
         public Integer func_187159_a(PacketBuffer p_187159_1_) throws IOException
         {
             return p_187159_1_.func_150792_a();
         }
-        @Override
         public DataParameter<Integer> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<Integer>(p_187161_1_, this);
         }
         public Integer func_192717_a(Integer p_192717_1_)
         {
@@ -62,16 +60,15 @@
     {
         public void func_187160_a(PacketBuffer p_187160_1_, Float p_187160_2_)
         {
-            p_187160_1_.writeFloat(p_187160_2_);
+            p_187160_1_.writeFloat(p_187160_2_.floatValue());
         }
         public Float func_187159_a(PacketBuffer p_187159_1_) throws IOException
         {
             return p_187159_1_.readFloat();
         }
-        @Override
         public DataParameter<Float> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<Float>(p_187161_1_, this);
         }
         public Float func_192717_a(Float p_192717_1_)
         {
@@ -88,10 +85,9 @@
         {
             return p_187159_1_.func_150789_c(32767);
         }
-        @Override
         public DataParameter<String> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<String>(p_187161_1_, this);
         }
         public String func_192717_a(String p_192717_1_)
         {
@@ -108,10 +104,9 @@
         {
             return p_187159_1_.func_179258_d();
         }
-        @Override
         public DataParameter<ITextComponent> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<ITextComponent>(p_187161_1_, this);
         }
         public ITextComponent func_192717_a(ITextComponent p_192717_1_)
         {
@@ -128,10 +123,9 @@
         {
             return p_187159_1_.func_150791_c();
         }
-        @Override
         public DataParameter<ItemStack> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<ItemStack>(p_187161_1_, this);
         }
         public ItemStack func_192717_a(ItemStack p_192717_1_)
         {
@@ -156,10 +150,9 @@
             int i = p_187159_1_.func_150792_a();
             return i == 0 ? Optional.absent() : Optional.of(Block.func_176220_d(i));
         }
-        @Override
         public DataParameter<Optional<IBlockState>> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<Optional<IBlockState>>(p_187161_1_, this);
         }
         public Optional<IBlockState> func_192717_a(Optional<IBlockState> p_192717_1_)
         {
@@ -170,16 +163,15 @@
     {
         public void func_187160_a(PacketBuffer p_187160_1_, Boolean p_187160_2_)
         {
-            p_187160_1_.writeBoolean(p_187160_2_);
+            p_187160_1_.writeBoolean(p_187160_2_.booleanValue());
         }
         public Boolean func_187159_a(PacketBuffer p_187159_1_) throws IOException
         {
             return p_187159_1_.readBoolean();
         }
-        @Override
         public DataParameter<Boolean> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<Boolean>(p_187161_1_, this);
         }
         public Boolean func_192717_a(Boolean p_192717_1_)
         {
@@ -198,10 +190,9 @@
         {
             return new Rotations(p_187159_1_.readFloat(), p_187159_1_.readFloat(), p_187159_1_.readFloat());
         }
-        @Override
         public DataParameter<Rotations> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<Rotations>(p_187161_1_, this);
         }
         public Rotations func_192717_a(Rotations p_192717_1_)
         {
@@ -218,10 +209,9 @@
         {
             return p_187159_1_.func_179259_c();
         }
-        @Override
         public DataParameter<BlockPos> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<BlockPos>(p_187161_1_, this);
         }
         public BlockPos func_192717_a(BlockPos p_192717_1_)
         {
@@ -243,10 +233,9 @@
         {
             return !p_187159_1_.readBoolean() ? Optional.absent() : Optional.of(p_187159_1_.func_179259_c());
         }
-        @Override
         public DataParameter<Optional<BlockPos>> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<Optional<BlockPos>>(p_187161_1_, this);
         }
         public Optional<BlockPos> func_192717_a(Optional<BlockPos> p_192717_1_)
         {
@@ -261,12 +250,11 @@
         }
         public EnumFacing func_187159_a(PacketBuffer p_187159_1_) throws IOException
         {
-            return p_187159_1_.func_179257_a(EnumFacing.class);
+            return (EnumFacing)p_187159_1_.func_179257_a(EnumFacing.class);
         }
-        @Override
         public DataParameter<EnumFacing> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<EnumFacing>(p_187161_1_, this);
         }
         public EnumFacing func_192717_a(EnumFacing p_192717_1_)
         {
@@ -288,10 +276,9 @@
         {
             return !p_187159_1_.readBoolean() ? Optional.absent() : Optional.of(p_187159_1_.func_179253_g());
         }
-        @Override
         public DataParameter<Optional<UUID>> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<Optional<UUID>>(p_187161_1_, this);
         }
         public Optional<UUID> func_192717_a(Optional<UUID> p_192717_1_)
         {
@@ -308,10 +295,9 @@
         {
             return p_187159_1_.func_150793_b();
         }
-        @Override
         public DataParameter<NBTTagCompound> func_187161_a(int p_187161_1_)
         {
-            return new DataParameter<>(p_187161_1_, this);
+            return new DataParameter<NBTTagCompound>(p_187161_1_, this);
         }
         public NBTTagCompound func_192717_a(NBTTagCompound p_192717_1_)
         {
@@ -319,20 +305,21 @@
         }
     };
 
+    @Deprecated // Forge: ONLY FOR VANILLA - mods should use the Forge registry
     public static void func_187189_a(DataSerializer<?> p_187189_0_)
     {
-        field_187204_n.func_186808_c(p_187189_0_);
+        if (field_187204_n.func_186808_c(p_187189_0_) >= 256) throw new RuntimeException("Vanilla DataSerializer ID limit exceeded");
     }
 
     @Nullable
     public static DataSerializer<?> func_187190_a(int p_187190_0_)
     {
-        return field_187204_n.func_186813_a(p_187190_0_);
+        return net.minecraftforge.common.ForgeHooks.getSerializer(p_187190_0_, field_187204_n);
     }
 
     public static int func_187188_b(DataSerializer<?> p_187188_0_)
     {
-        return field_187204_n.func_186815_a(p_187188_0_);
+        return net.minecraftforge.common.ForgeHooks.getSerializerId(p_187188_0_, field_187204_n);
     }
 
     static
