--- before/net/minecraft/network/datasync/EntityDataManager.java
+++ after/net/minecraft/network/datasync/EntityDataManager.java
@@ -24,9 +24,9 @@
 public class EntityDataManager
 {
     private static final Logger LOGGER = LogManager.getLogger();
-    private static final Map < Class <? extends Entity > , Integer > NEXT_ID_MAP = Maps.newHashMap();
+    private static final Map < Class <? extends Entity > , Integer > NEXT_ID_MAP = Maps. < Class <? extends Entity > , Integer > newHashMap();
     private final Entity entity;
-    private final Map < Integer, EntityDataManager.DataEntry<? >> entries = Maps.newHashMap();
+    private final Map < Integer, EntityDataManager.DataEntry<? >> entries = Maps. < Integer, EntityDataManager.DataEntry<? >> newHashMap();
     private final ReadWriteLock lock = new ReentrantReadWriteLock();
     private boolean empty = true;
     private boolean dirty;
@@ -38,7 +38,7 @@
 
     public static <T> DataParameter<T> createKey(Class <? extends Entity > clazz, DataSerializer<T> serializer)
     {
-        if (LOGGER.isDebugEnabled())
+        if (true || LOGGER.isDebugEnabled()) //Forge: This is very useful for mods that register keys on classes that are not their own
         {
             try
             {
@@ -46,11 +46,14 @@
 
                 if (!oclass.equals(clazz))
                 {
-                    LOGGER.debug("defineId called for: {} from {}", clazz, oclass, new RuntimeException());
+                    //Forge: log at warn, mods should not add to classes that they don't own, and only add stacktrace when in debug is enabled as it is mostly not needed and consumes time
+                    if (LOGGER.isDebugEnabled()) LOGGER.warn("defineId called for: {} from {}", clazz, oclass, new RuntimeException());
+                    else LOGGER.warn("defineId called for: {} from {}", clazz, oclass);
                 }
             }
-            catch (ClassNotFoundException classnotfoundexception)
+            catch (ClassNotFoundException var5)
             {
+                ;
             }
         }
 
@@ -58,7 +61,7 @@
 
         if (NEXT_ID_MAP.containsKey(clazz))
         {
-            j = NEXT_ID_MAP.get(clazz) + 1;
+            j = ((Integer)NEXT_ID_MAP.get(clazz)).intValue() + 1;
         }
         else
         {
@@ -71,7 +74,7 @@
 
                 if (NEXT_ID_MAP.containsKey(oclass1))
                 {
-                    i = NEXT_ID_MAP.get(oclass1) + 1;
+                    i = ((Integer)NEXT_ID_MAP.get(oclass1)).intValue() + 1;
                     break;
                 }
             }
@@ -85,7 +88,7 @@
         }
         else
         {
-            NEXT_ID_MAP.put(clazz, j);
+            NEXT_ID_MAP.put(clazz, Integer.valueOf(j));
             return serializer.createKey(j);
         }
     }
@@ -98,7 +101,7 @@
         {
             throw new IllegalArgumentException("Data value id is too big with " + i + "! (Max is " + 254 + ")");
         }
-        else if (this.entries.containsKey(i))
+        else if (this.entries.containsKey(Integer.valueOf(i)))
         {
             throw new IllegalArgumentException("Duplicate id value for " + i + "!");
         }
@@ -114,9 +117,9 @@
 
     private <T> void setEntry(DataParameter<T> key, T value)
     {
-        EntityDataManager.DataEntry<T> dataentry = new EntityDataManager.DataEntry<>(key, value);
+        EntityDataManager.DataEntry<T> dataentry = new EntityDataManager.DataEntry<T>(key, value);
         this.lock.writeLock().lock();
-        this.entries.put(key.getId(), dataentry);
+        this.entries.put(Integer.valueOf(key.getId()), dataentry);
         this.empty = false;
         this.lock.writeLock().unlock();
     }
@@ -124,11 +127,11 @@
     private <T> EntityDataManager.DataEntry<T> getEntry(DataParameter<T> key)
     {
         this.lock.readLock().lock();
-        EntityDataManager.DataEntry dataentry;
+        EntityDataManager.DataEntry<T> dataentry;
 
         try
         {
-            dataentry = (EntityDataManager.DataEntry)this.entries.get(key.getId());
+            dataentry = (EntityDataManager.DataEntry)this.entries.get(Integer.valueOf(key.getId()));
         }
         catch (Throwable throwable)
         {
@@ -144,12 +147,12 @@
 
     public <T> T get(DataParameter<T> key)
     {
-        return this.getEntry(key).getValue();
+        return (T)this.getEntry(key).getValue();
     }
 
     public <T> void set(DataParameter<T> key, T value)
     {
-        EntityDataManager.DataEntry<T> dataentry = this.getEntry(key);
+        EntityDataManager.DataEntry<T> dataentry = this.<T>getEntry(key);
 
         if (ObjectUtils.notEqual(value, dataentry.getValue()))
         {
@@ -177,9 +180,9 @@
         {
             int i = 0;
 
-            for (int j = entriesIn.size(); i < j; i++)
+            for (int j = entriesIn.size(); i < j; ++i)
             {
-                EntityDataManager.DataEntry<?> dataentry = entriesIn.get(i);
+                EntityDataManager.DataEntry<?> dataentry = (EntityDataManager.DataEntry)entriesIn.get(i);
                 writeEntry(buf, dataentry);
             }
         }
@@ -204,7 +207,7 @@
 
                     if (list == null)
                     {
-                        list = Lists.newArrayList();
+                        list = Lists. < EntityDataManager.DataEntry<? >> newArrayList();
                     }
 
                     list.add(dataentry.copy());
@@ -241,7 +244,7 @@
         {
             if (list == null)
             {
-                list = Lists.newArrayList();
+                list = Lists. < EntityDataManager.DataEntry<? >> newArrayList();
             }
 
             list.add(dataentry.copy());
@@ -278,7 +281,7 @@
         {
             if (list == null)
             {
-                list = Lists.newArrayList();
+                list = Lists. < EntityDataManager.DataEntry<? >> newArrayList();
             }
 
             int j = buf.readVarInt();
@@ -302,7 +305,7 @@
 
         for (EntityDataManager.DataEntry<?> dataentry : entriesIn)
         {
-            EntityDataManager.DataEntry<?> dataentry1 = this.entries.get(dataentry.getKey().getId());
+            EntityDataManager.DataEntry<?> dataentry1 = (EntityDataManager.DataEntry)this.entries.get(Integer.valueOf(dataentry.getKey().getId()));
 
             if (dataentry1 != null)
             {
@@ -379,7 +382,7 @@
 
             public EntityDataManager.DataEntry<T> copy()
             {
-                return new EntityDataManager.DataEntry<>(this.key, this.key.getSerializer().copyValue(this.value));
+                return new EntityDataManager.DataEntry<T>(this.key, this.key.getSerializer().copyValue(this.value));
             }
         }
 }
