--- before/net/minecraft/client/settings/HotbarSnapshot.java
+++ after/net/minecraft/client/settings/HotbarSnapshot.java
@@ -29,7 +29,7 @@
 
         for (int i = 0; i < field_192835_a; ++i)
         {
-            nbttaglist.func_74742_a(this.get(i).func_77955_b(new NBTTagCompound()));
+            nbttaglist.func_74742_a(((ItemStack)this.get(i)).func_77955_b(new NBTTagCompound()));
         }
 
         return nbttaglist;
@@ -47,7 +47,7 @@
     {
         for (int i = 0; i < field_192835_a; ++i)
         {
-            if (!this.get(i).func_190926_b())
+            if (!((ItemStack)this.get(i)).func_190926_b())
             {
                 return false;
             }
