--- before/net/minecraft/advancements/AdvancementTreeNode.java
+++ after/net/minecraft/advancements/AdvancementTreeNode.java
@@ -10,7 +10,7 @@
     private final AdvancementTreeNode field_192329_b;
     private final AdvancementTreeNode field_192330_c;
     private final int field_192331_d;
-    private final List<AdvancementTreeNode> field_192332_e = Lists.newArrayList();
+    private final List<AdvancementTreeNode> field_192332_e = Lists.<AdvancementTreeNode>newArrayList();
     private AdvancementTreeNode field_192333_f;
     private AdvancementTreeNode field_192334_g;
     private int field_192335_h;
@@ -152,7 +152,7 @@
         }
         else
         {
-            return !this.field_192332_e.isEmpty() ? this.field_192332_e.get(0) : null;
+            return !this.field_192332_e.isEmpty() ? (AdvancementTreeNode)this.field_192332_e.get(0) : null;
         }
     }
 
@@ -165,7 +165,7 @@
         }
         else
         {
-            return !this.field_192332_e.isEmpty() ? this.field_192332_e.get(this.field_192332_e.size() - 1) : null;
+            return !this.field_192332_e.isEmpty() ? (AdvancementTreeNode)this.field_192332_e.get(this.field_192332_e.size() - 1) : null;
         }
     }
 
