--- before/net/minecraft/advancements/AdvancementRewards.java
+++ after/net/minecraft/advancements/AdvancementRewards.java
@@ -14,6 +14,7 @@
 import net.minecraft.command.ICommandSender;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.item.EntityItem;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.init.SoundEvents;
 import net.minecraft.item.ItemStack;
@@ -50,7 +51,7 @@
     public void apply(final EntityPlayerMP player)
     {
         player.addExperience(this.experience);
-        LootContext lootcontext = new LootContext.Builder(player.getServerWorld()).withLootedEntity(player).build();
+        LootContext lootcontext = new LootContext.Builder(player.getServerWorld()).withLootedEntity(player).withPlayer(player).withLuck(player.getLuck()).build(); // Forge: add player & luck to LootContext
         boolean flag = false;
 
         for (ResourceLocation resourcelocation : this.loot)
