--- before/net/minecraft/advancements/AdvancementManager.java
+++ after/net/minecraft/advancements/AdvancementManager.java
@@ -74,6 +74,7 @@
         ADVANCEMENT_LIST.clear();
         Map<ResourceLocation, Advancement.Builder> map = this.loadCustomAdvancements();
         this.loadBuiltInAdvancements(map);
+        this.hasErrored |= net.minecraftforge.common.ForgeHooks.loadAdvancements(map);
         ADVANCEMENT_LIST.loadAdvancements(map);
 
         for (Advancement advancement : ADVANCEMENT_LIST.getRoots())
