--- before/net/minecraft/advancements/Criterion.java
+++ after/net/minecraft/advancements/Criterion.java
@@ -6,6 +6,7 @@
 import com.google.gson.JsonObject;
 import com.google.gson.JsonSyntaxException;
 import java.util.Map;
+import java.util.Map.Entry;
 import javax.annotation.Nullable;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.util.JsonUtils;
@@ -52,9 +53,9 @@
 
     public static Map<String, Criterion> func_192144_b(JsonObject p_192144_0_, JsonDeserializationContext p_192144_1_)
     {
-        Map<String, Criterion> map = Maps.newHashMap();
+        Map<String, Criterion> map = Maps.<String, Criterion>newHashMap();
 
-        for (Map.Entry<String, JsonElement> entry : p_192144_0_.entrySet())
+        for (Entry<String, JsonElement> entry : p_192144_0_.entrySet())
         {
             map.put(entry.getKey(), func_192145_a(JsonUtils.func_151210_l(entry.getValue(), "criterion"), p_192144_1_));
         }
@@ -64,7 +65,7 @@
 
     public static Map<String, Criterion> func_192142_c(PacketBuffer p_192142_0_)
     {
-        Map<String, Criterion> map = Maps.newHashMap();
+        Map<String, Criterion> map = Maps.<String, Criterion>newHashMap();
         int i = p_192142_0_.func_150792_a();
 
         for (int j = 0; j < i; ++j)
@@ -79,10 +80,10 @@
     {
         p_192141_1_.func_150787_b(p_192141_0_.size());
 
-        for (Map.Entry<String, Criterion> entry : p_192141_0_.entrySet())
+        for (Entry<String, Criterion> entry : p_192141_0_.entrySet())
         {
             p_192141_1_.func_180714_a(entry.getKey());
-            entry.getValue().func_192140_a(p_192141_1_);
+            ((Criterion)entry.getValue()).func_192140_a(p_192141_1_);
         }
     }
 
