--- before/net/minecraft/advancements/DisplayInfo.java
+++ after/net/minecraft/advancements/DisplayInfo.java
@@ -103,8 +103,8 @@
 
     public static DisplayInfo func_192294_a(JsonObject p_192294_0_, JsonDeserializationContext p_192294_1_)
     {
-        ITextComponent itextcomponent = JsonUtils.func_188174_a(p_192294_0_, "title", p_192294_1_, ITextComponent.class);
-        ITextComponent itextcomponent1 = JsonUtils.func_188174_a(p_192294_0_, "description", p_192294_1_, ITextComponent.class);
+        ITextComponent itextcomponent = (ITextComponent)JsonUtils.func_188174_a(p_192294_0_, "title", p_192294_1_, ITextComponent.class);
+        ITextComponent itextcomponent1 = (ITextComponent)JsonUtils.func_188174_a(p_192294_0_, "description", p_192294_1_, ITextComponent.class);
 
         if (itextcomponent != null && itextcomponent1 != null)
         {
@@ -132,7 +132,9 @@
         {
             Item item = JsonUtils.func_188180_i(p_193221_0_, "item");
             int i = JsonUtils.func_151208_a(p_193221_0_, "data", 0);
-            return new ItemStack(item, 1, i);
+            ItemStack ret = new ItemStack(item, 1, i);
+            ret.func_77982_d(net.minecraftforge.common.util.JsonUtils.readNBT(p_193221_0_, "nbt"));
+            return ret;
         }
     }
 
@@ -175,7 +177,7 @@
         ITextComponent itextcomponent = p_192295_0_.func_179258_d();
         ITextComponent itextcomponent1 = p_192295_0_.func_179258_d();
         ItemStack itemstack = p_192295_0_.func_150791_c();
-        FrameType frametype = p_192295_0_.func_179257_a(FrameType.class);
+        FrameType frametype = (FrameType)p_192295_0_.func_179257_a(FrameType.class);
         int i = p_192295_0_.readInt();
         ResourceLocation resourcelocation = (i & 1) != 0 ? p_192295_0_.func_192575_l() : null;
         boolean flag = (i & 2) != 0;
