--- before/net/minecraft/advancements/ICriterionTrigger.java
+++ after/net/minecraft/advancements/ICriterionTrigger.java
@@ -68,7 +68,8 @@
             {
                 int i = this.field_192160_a.hashCode();
                 i = 31 * i + this.field_192161_b.hashCode();
-                return 31 * i + this.field_192162_c.hashCode();
+                i = 31 * i + this.field_192162_c.hashCode();
+                return i;
             }
         }
 }
