--- before/net/minecraft/advancements/PlayerAdvancements.java
+++ after/net/minecraft/advancements/PlayerAdvancements.java
@@ -16,6 +16,7 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.Map.Entry;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
@@ -32,20 +33,16 @@
 public class PlayerAdvancements
 {
     private static final Logger field_192753_a = LogManager.getLogger();
-    private static final Gson field_192754_b = new GsonBuilder()
-        .registerTypeAdapter(AdvancementProgress.class, new AdvancementProgress.Serializer())
-        .registerTypeAdapter(ResourceLocation.class, new ResourceLocation.Serializer())
-        .setPrettyPrinting()
-        .create();
+    private static final Gson field_192754_b = (new GsonBuilder()).registerTypeAdapter(AdvancementProgress.class, new AdvancementProgress.Serializer()).registerTypeAdapter(ResourceLocation.class, new ResourceLocation.Serializer()).setPrettyPrinting().create();
     private static final TypeToken<Map<ResourceLocation, AdvancementProgress>> field_192755_c = new TypeToken<Map<ResourceLocation, AdvancementProgress>>()
     {
     };
     private final MinecraftServer field_192756_d;
     private final File field_192757_e;
-    private final Map<Advancement, AdvancementProgress> field_192758_f = Maps.newLinkedHashMap();
-    private final Set<Advancement> field_192759_g = Sets.newLinkedHashSet();
-    private final Set<Advancement> field_192760_h = Sets.newLinkedHashSet();
-    private final Set<Advancement> field_192761_i = Sets.newLinkedHashSet();
+    private final Map<Advancement, AdvancementProgress> field_192758_f = Maps.<Advancement, AdvancementProgress>newLinkedHashMap();
+    private final Set<Advancement> field_192759_g = Sets.<Advancement>newLinkedHashSet();
+    private final Set<Advancement> field_192760_h = Sets.<Advancement>newLinkedHashSet();
+    private final Set<Advancement> field_192761_i = Sets.<Advancement>newLinkedHashSet();
     private EntityPlayerMP field_192762_j;
     @Nullable
     private Advancement field_194221_k;
@@ -199,6 +196,9 @@
 
     public boolean func_192750_a(Advancement p_192750_1_, String p_192750_2_)
     {
+        // Forge: don't grant advancements for fake players
+        if (this.field_192762_j instanceof net.minecraftforge.common.util.FakePlayer) return false;
+
         boolean flag = false;
         AdvancementProgress advancementprogress = this.func_192747_a(p_192750_1_);
         boolean flag1 = advancementprogress.func_192105_a();
@@ -213,20 +213,12 @@
             {
                 p_192750_1_.func_192072_d().func_192113_a(this.field_192762_j);
 
-                if (p_192750_1_.func_192068_c() != null
-                        && p_192750_1_.func_192068_c().func_193220_i()
-                        && this.field_192762_j.field_70170_p.func_82736_K().func_82766_b("announceAdvancements"))
+                if (p_192750_1_.func_192068_c() != null && p_192750_1_.func_192068_c().func_193220_i() && this.field_192762_j.field_70170_p.func_82736_K().func_82766_b("announceAdvancements"))
                 {
-                    this.field_192756_d
-                    .func_184103_al()
-                    .func_148539_a(
-                        new TextComponentTranslation(
-                            "chat.type.advancement." + p_192750_1_.func_192068_c().func_192291_d().func_192307_a(),
-                            this.field_192762_j.func_145748_c_(),
-                            p_192750_1_.func_193123_j()
-                        )
-                    );
+                    this.field_192756_d.func_184103_al().func_148539_a(new TextComponentTranslation("chat.type.advancement." + p_192750_1_.func_192068_c().func_192291_d().func_192307_a(), new Object[] {this.field_192762_j.func_145748_c_(), p_192750_1_.func_193123_j()}));
                 }
+
+                net.minecraftforge.common.ForgeHooks.onAdvancement(this.field_192762_j, p_192750_1_);
             }
         }
 
@@ -365,9 +357,7 @@
 
         if (advancement != this.field_194221_k)
         {
-            this.field_192762_j
-            .field_71135_a
-            .func_147359_a(new SPacketSelectAdvancementsTab(this.field_194221_k == null ? null : this.field_194221_k.func_192067_g()));
+            this.field_192762_j.field_71135_a.func_147359_a(new SPacketSelectAdvancementsTab(this.field_194221_k == null ? null : this.field_194221_k.func_192067_g()));
         }
     }
 
