--- before/net/minecraft/advancements/PlayerAdvancements.java
+++ after/net/minecraft/advancements/PlayerAdvancements.java
@@ -199,6 +199,9 @@
 
     public boolean grantCriterion(Advancement p_192750_1_, String p_192750_2_)
     {
+        // Forge: don't grant advancements for fake players
+        if (this.player instanceof net.minecraftforge.common.util.FakePlayer) return false;
+
         boolean flag = false;
         AdvancementProgress advancementprogress = this.getProgress(p_192750_1_);
         boolean flag1 = advancementprogress.isDone();
@@ -227,6 +230,8 @@
                         )
                     );
                 }
+
+                net.minecraftforge.common.ForgeHooks.onAdvancement(this.player, p_192750_1_);
             }
         }
 
