--- original/net/minecraft/client/renderer/vertex/VertexFormat.java
+++ modified/net/minecraft/client/renderer/vertex/VertexFormat.java
@@ -17,6 +17,7 @@
     private int colorElementOffset = -1;
     private final List<Integer> uvOffsetsById = Lists.newArrayList();
     private int normalElementOffset = -1;
+    private int hashCode;
 
     public VertexFormat(VertexFormat vertexFormatIn) {
         this();
@@ -38,6 +39,7 @@
         this.uvOffsetsById.clear();
         this.normalElementOffset = -1;
         this.vertexSize = 0;
+        this.hashCode = 0;
     }
 
     public VertexFormat addElement(VertexFormatElement element) {
@@ -60,6 +62,7 @@
         }
 
         this.vertexSize = this.vertexSize + element.getSize();
+        this.hashCode = 0;
         return this;
     }
 
@@ -158,8 +161,11 @@
 
     @Override
     public int hashCode() {
+        if (this.hashCode != 0) return this.hashCode;
         int i = this.elements.hashCode();
         i = 31 * i + this.offsets.hashCode();
-        return 31 * i + this.vertexSize;
+        i = 31 * i + this.vertexSize;
+        this.hashCode = i;
+        return i;
     }
 }
