--- before/net/minecraft/server/management/DemoPlayerInteractionManager.java
+++ after/net/minecraft/server/management/DemoPlayerInteractionManager.java
@@ -46,7 +46,7 @@
         {
             if (j <= 6L)
             {
-                this.field_73090_b.func_145747_a(new TextComponentTranslation("demo.day." + j));
+                this.field_73090_b.func_145747_a(new TextComponentTranslation("demo.day." + j, new Object[0]));
             }
         }
         else if (j == 1L)
@@ -66,7 +66,7 @@
         }
         else if (j == 5L && i % 24000L == 22000L)
         {
-            this.field_73090_b.func_145747_a(new TextComponentTranslation("demo.day.warning"));
+            this.field_73090_b.func_145747_a(new TextComponentTranslation("demo.day.warning", new Object[0]));
         }
     }
 
@@ -74,7 +74,7 @@
     {
         if (this.field_73104_e > 100)
         {
-            this.field_73090_b.func_145747_a(new TextComponentTranslation("demo.reminder"));
+            this.field_73090_b.func_145747_a(new TextComponentTranslation("demo.reminder", new Object[0]));
             this.field_73104_e = 0;
         }
     }
