--- before/net/minecraft/server/management/PlayerProfileCache.java
+++ after/net/minecraft/server/management/PlayerProfileCache.java
@@ -45,9 +45,9 @@
 {
     public static final SimpleDateFormat field_152659_a = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss Z");
     private static boolean field_187322_c;
-    private final Map<String, PlayerProfileCache.ProfileEntry> field_152661_c = Maps.newHashMap();
-    private final Map<UUID, PlayerProfileCache.ProfileEntry> field_152662_d = Maps.newHashMap();
-    private final Deque<GameProfile> field_152663_e = Lists.newLinkedList();
+    private final Map<String, PlayerProfileCache.ProfileEntry> field_152661_c = Maps.<String, PlayerProfileCache.ProfileEntry>newHashMap();
+    private final Map<UUID, PlayerProfileCache.ProfileEntry> field_152662_d = Maps.<UUID, PlayerProfileCache.ProfileEntry>newHashMap();
+    private final Deque<GameProfile> field_152663_e = Lists.<GameProfile>newLinkedList();
     private final GameProfileRepository field_187323_g;
     protected final Gson field_152660_b;
     private final File field_152665_g;
@@ -184,7 +184,7 @@
     public String[] func_152654_a()
     {
         List<String> list = Lists.newArrayList(this.field_152661_c.keySet());
-        return list.toArray(new String[list.size()]);
+        return (String[])list.toArray(new String[list.size()]);
     }
 
     @Nullable
@@ -215,7 +215,7 @@
         try
         {
             bufferedreader = Files.newReader(this.field_152665_g, StandardCharsets.UTF_8);
-            List<PlayerProfileCache.ProfileEntry> list = JsonUtils.func_193841_a(this.field_152660_b, bufferedreader, field_152666_h);
+            List<PlayerProfileCache.ProfileEntry> list = (List)JsonUtils.func_193841_a(this.field_152660_b, bufferedreader, field_152666_h);
             this.field_152661_c.clear();
             this.field_152662_d.clear();
             this.field_152663_e.clear();
@@ -231,11 +231,13 @@
                 }
             }
         }
-        catch (FileNotFoundException filenotfoundexception)
+        catch (FileNotFoundException var9)
         {
+            ;
         }
-        catch (JsonParseException jsonparseexception)
+        catch (JsonParseException var10)
         {
+            ;
         }
         finally
         {
@@ -254,10 +256,11 @@
             bufferedwriter.write(s);
             return;
         }
-        catch (FileNotFoundException filenotfoundexception)
+        catch (FileNotFoundException var8)
         {
+            ;
         }
-        catch (IOException ioexception)
+        catch (IOException var9)
         {
             return;
         }
@@ -269,7 +272,7 @@
 
     private List<PlayerProfileCache.ProfileEntry> func_152656_a(int p_152656_1_)
     {
-        List<PlayerProfileCache.ProfileEntry> list = Lists.newArrayList();
+        List<PlayerProfileCache.ProfileEntry> list = Lists.<PlayerProfileCache.ProfileEntry>newArrayList();
 
         for (GameProfile gameprofile : Lists.newArrayList(Iterators.limit(this.field_152663_e.iterator(), p_152656_1_)))
         {
@@ -343,7 +346,7 @@
                         {
                             date = PlayerProfileCache.field_152659_a.parse(jsonelement2.getAsString());
                         }
-                        catch (ParseException parseexception)
+                        catch (ParseException var14)
                         {
                             date = null;
                         }
@@ -357,7 +360,7 @@
                         {
                             uuid = UUID.fromString(s);
                         }
-                        catch (Throwable throwable)
+                        catch (Throwable var13)
                         {
                             return null;
                         }
