--- before/net/minecraft/server/management/PlayerChunkMap.java
+++ after/net/minecraft/server/management/PlayerChunkMap.java
@@ -38,12 +38,12 @@
         }
     };
     private final WorldServer field_72701_a;
-    private final List<EntityPlayerMP> field_72699_b = Lists.newArrayList();
-    private final Long2ObjectMap<PlayerChunkMapEntry> field_72700_c = new Long2ObjectOpenHashMap<>(4096);
-    private final Set<PlayerChunkMapEntry> field_72697_d = Sets.newHashSet();
-    private final List<PlayerChunkMapEntry> field_187310_g = Lists.newLinkedList();
-    private final List<PlayerChunkMapEntry> field_187311_h = Lists.newLinkedList();
-    private final List<PlayerChunkMapEntry> field_111193_e = Lists.newArrayList();
+    private final List<EntityPlayerMP> field_72699_b = Lists.<EntityPlayerMP>newArrayList();
+    private final Long2ObjectMap<PlayerChunkMapEntry> field_72700_c = new Long2ObjectOpenHashMap<PlayerChunkMapEntry>(4096);
+    private final Set<PlayerChunkMapEntry> field_72697_d = Sets.<PlayerChunkMapEntry>newHashSet();
+    private final List<PlayerChunkMapEntry> field_187310_g = Lists.<PlayerChunkMapEntry>newLinkedList();
+    private final List<PlayerChunkMapEntry> field_187311_h = Lists.<PlayerChunkMapEntry>newLinkedList();
+    private final List<PlayerChunkMapEntry> field_111193_e = Lists.<PlayerChunkMapEntry>newArrayList();
     private int field_72698_e;
     private long field_111192_g;
     private boolean field_187312_l = true;
@@ -97,7 +97,7 @@
                         return chunk;
                     }
 
-                    return this.endOfData();
+                    return (Chunk)this.endOfData();
                 }
             }
         };
@@ -229,13 +229,13 @@
     @Nullable
     public PlayerChunkMapEntry func_187301_b(int p_187301_1_, int p_187301_2_)
     {
-        return this.field_72700_c.get(func_187307_d(p_187301_1_, p_187301_2_));
+        return (PlayerChunkMapEntry)this.field_72700_c.get(func_187307_d(p_187301_1_, p_187301_2_));
     }
 
     private PlayerChunkMapEntry func_187302_c(int p_187302_1_, int p_187302_2_)
     {
         long i = func_187307_d(p_187302_1_, p_187302_2_);
-        PlayerChunkMapEntry playerchunkmapentry = this.field_72700_c.get(i);
+        PlayerChunkMapEntry playerchunkmapentry = (PlayerChunkMapEntry)this.field_72700_c.get(i);
 
         if (playerchunkmapentry == null)
         {
@@ -333,7 +333,7 @@
         double d1 = p_72685_1_.field_71132_e - p_72685_1_.field_70161_v;
         double d2 = d0 * d0 + d1 * d1;
 
-        if (!(d2 < 64.0D))
+        if (d2 >= 64.0D)
         {
             int k = (int)p_72685_1_.field_71131_d >> 4;
             int l = (int)p_72685_1_.field_71132_e >> 4;
