--- before/net/minecraft/server/management/PlayerList.java
+++ after/net/minecraft/server/management/PlayerList.java
@@ -100,7 +100,7 @@
         this.maxPlayers = 8;
     }
 
-    public void initializeConnectionToPlayer(NetworkManager netManager, EntityPlayerMP playerIn)
+    public void initializeConnectionToPlayer(NetworkManager netManager, EntityPlayerMP playerIn, NetHandlerPlayServer nethandlerplayserver)
     {
         GameProfile gameprofile = playerIn.getGameProfile();
         PlayerProfileCache playerprofilecache = this.server.getPlayerProfileCache();
@@ -109,6 +109,17 @@
         playerprofilecache.addEntry(gameprofile);
         NBTTagCompound nbttagcompound = this.readPlayerDataFromFile(playerIn);
         playerIn.setWorld(this.server.getWorld(playerIn.dimension));
+
+        World playerWorld = this.server.getWorld(playerIn.dimension);
+        if (playerWorld == null)
+        {
+            playerIn.dimension = 0;
+            playerWorld = this.server.getWorld(0);
+            BlockPos spawnPoint = playerWorld.provider.getRandomizedSpawnPoint();
+            playerIn.setPosition(spawnPoint.getX(), spawnPoint.getY(), spawnPoint.getZ());
+        }
+
+        playerIn.setWorld(playerWorld);
         playerIn.interactionManager.setWorld((WorldServer)playerIn.world);
         String s1 = "local";
 
@@ -129,13 +140,14 @@
         WorldServer worldserver = this.server.getWorld(playerIn.dimension);
         WorldInfo worldinfo = worldserver.getWorldInfo();
         this.setPlayerGameTypeBasedOnOther(playerIn, null, worldserver);
-        NetHandlerPlayServer nethandlerplayserver = new NetHandlerPlayServer(this.server, netManager, playerIn);
+        playerIn.connection = nethandlerplayserver;
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().fireServerConnectionEvent(netManager);
         nethandlerplayserver.sendPacket(
             new SPacketJoinGame(
                 playerIn.getEntityId(),
                 playerIn.interactionManager.getGameType(),
                 worldinfo.isHardcoreModeEnabled(),
-                worldserver.provider.getDimensionType().getId(),
+                worldserver.provider.getDimension(),
                 worldserver.getDifficulty(),
                 this.getMaxPlayers(),
                 worldinfo.getTerrainType(),
@@ -221,6 +233,7 @@
         }
 
         playerIn.addSelfToInternalCraftingInventory();
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerLoggedIn(playerIn);
     }
 
     protected void sendScoreboard(ServerScoreboard scoreboardIn, EntityPlayerMP playerIn)
@@ -330,6 +343,7 @@
             nbttagcompound1 = nbttagcompound;
             playerIn.readFromNBT(nbttagcompound);
             LOGGER.debug("loading single player");
+            net.minecraftforge.event.ForgeEventFactory.firePlayerLoadingEvent(playerIn, this.playerDataManager, playerIn.getUniqueID().toString());
         }
         else
         {
@@ -339,8 +353,24 @@
         return nbttagcompound1;
     }
 
+    public NBTTagCompound getPlayerNBT(EntityPlayerMP player)
+    {
+        // Hacky method to allow loading the NBT for a player prior to login
+        NBTTagCompound nbttagcompound = this.server.worlds[0].getWorldInfo().getPlayerNBTTagCompound();
+        if (player.getName().equals(this.server.getServerOwner()) && nbttagcompound != null)
+        {
+            return nbttagcompound;
+        }
+        else
+        {
+            return ((net.minecraft.world.storage.SaveHandler)this.playerDataManager).getPlayerNBT(player);
+        }
+    }
+
     protected void writePlayerData(EntityPlayerMP playerIn)
     {
+        if (playerIn.connection == null) return;
+
         this.playerDataManager.writePlayerData(playerIn);
         StatisticsManagerServer statisticsmanagerserver = this.playerStatFiles.get(playerIn.getUniqueID());
 
@@ -369,6 +399,7 @@
             playerIn.connection.sendPacket(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, this.playerEntityList.get(i)));
         }
 
+        net.minecraftforge.common.chunkio.ChunkIOExecutor.adjustPoolSize(this.getCurrentPlayerCount());
         worldserver.spawnEntity(playerIn);
         this.preparePlayer(playerIn, null);
     }
@@ -380,6 +411,7 @@
 
     public void playerLoggedOut(EntityPlayerMP playerIn)
     {
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerLoggedOut(playerIn);
         WorldServer worldserver = playerIn.getServerWorld();
         playerIn.addStat(StatList.LEAVE_GAME);
         this.writePlayerData(playerIn);
@@ -402,6 +434,7 @@
                 worldserver.getChunk(playerIn.chunkCoordX, playerIn.chunkCoordZ).markDirty();
             }
         }
+        net.minecraftforge.common.chunkio.ChunkIOExecutor.adjustPoolSize(this.getCurrentPlayerCount());
 
         worldserver.removeEntity(playerIn);
         worldserver.getPlayerChunkMap().removePlayer(playerIn);
@@ -499,13 +532,24 @@
 
     public EntityPlayerMP recreatePlayerEntity(EntityPlayerMP playerIn, int dimension, boolean conqueredEnd)
     {
+        World world = server.getWorld(dimension);
+        if (world == null)
+        {
+            dimension = playerIn.getSpawnDimension();
+        }
+        else if (!world.provider.canRespawnHere())
+        {
+            dimension = world.provider.getRespawnDimension(playerIn);
+        }
+        if (server.getWorld(dimension) == null) dimension = 0;
+
         playerIn.getServerWorld().getEntityTracker().removePlayerFromTrackers(playerIn);
         playerIn.getServerWorld().getEntityTracker().untrack(playerIn);
         playerIn.getServerWorld().getPlayerChunkMap().removePlayer(playerIn);
         this.playerEntityList.remove(playerIn);
         this.server.getWorld(playerIn.dimension).removeEntityDangerously(playerIn);
-        BlockPos blockpos = playerIn.getBedLocation();
-        boolean flag = playerIn.isSpawnForced();
+        BlockPos blockpos = playerIn.getBedLocation(dimension);
+        boolean flag = playerIn.isSpawnForced(dimension);
         playerIn.dimension = dimension;
         PlayerInteractionManager playerinteractionmanager;
 
@@ -523,6 +567,7 @@
         );
         entityplayermp.connection = playerIn.connection;
         entityplayermp.copyFrom(playerIn, conqueredEnd);
+        entityplayermp.dimension = dimension;
         entityplayermp.setEntityId(playerIn.getEntityId());
         entityplayermp.setCommandStats(playerIn);
         entityplayermp.setPrimaryHand(playerIn.getPrimaryHand());
@@ -586,6 +631,7 @@
         this.uuidToPlayerMap.put(entityplayermp.getUniqueID(), entityplayermp);
         entityplayermp.addSelfToInternalCraftingInventory();
         entityplayermp.setHealth(entityplayermp.getHealth());
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerRespawnEvent(entityplayermp, conqueredEnd);
         return entityplayermp;
     }
 
@@ -600,6 +646,17 @@
 
     public void changePlayerDimension(EntityPlayerMP player, int dimensionIn)
     {
+        transferPlayerToDimension(player, dimensionIn, server.getWorld(dimensionIn).getDefaultTeleporter());
+    }
+
+    // TODO: Remove (1.13)
+    public void transferPlayerToDimension(EntityPlayerMP player, int dimensionIn, net.minecraft.world.Teleporter teleporter)
+    {
+        transferPlayerToDimension(player, dimensionIn, (net.minecraftforge.common.util.ITeleporter) teleporter);
+    }
+
+    public void transferPlayerToDimension(EntityPlayerMP player, int dimensionIn, net.minecraftforge.common.util.ITeleporter teleporter)
+    {
         int i = player.dimension;
         WorldServer worldserver = this.server.getWorld(player.dimension);
         player.dimension = dimensionIn;
@@ -608,15 +665,15 @@
         .sendPacket(
             new SPacketRespawn(
                 player.dimension,
-                player.world.getDifficulty(),
-                player.world.getWorldInfo().getTerrainType(),
+                worldserver1.getDifficulty(),
+                worldserver1.getWorldInfo().getTerrainType(),
                 player.interactionManager.getGameType()
             )
-        );
+        ); // Forge: Use new dimensions information
         this.updatePermissionLevel(player);
         worldserver.removeEntityDangerously(player);
         player.isDead = false;
-        this.transferEntityToWorld(player, i, worldserver, worldserver1);
+        this.transferEntityToWorld(player, i, worldserver, worldserver1, teleporter);
         this.preparePlayer(player, worldserver);
         player.connection
         .setPlayerLocation(player.posX, player.posY, player.posZ, player.rotationYaw, player.rotationPitch);
@@ -629,17 +686,34 @@
         {
             player.connection.sendPacket(new SPacketEntityEffect(player.getEntityId(), potioneffect));
         }
+        // Fix MC-88179: on non-death SPacketRespawn, also resend attributes
+        net.minecraft.entity.ai.attributes.AttributeMap attributemap = (net.minecraft.entity.ai.attributes.AttributeMap) player.getAttributeMap();
+        java.util.Collection<net.minecraft.entity.ai.attributes.IAttributeInstance> watchedAttribs = attributemap.getWatchedAttributes();
+        if (!watchedAttribs.isEmpty()) player.connection.sendPacket(new net.minecraft.network.play.server.SPacketEntityProperties(player.getEntityId(), watchedAttribs));
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerChangedDimensionEvent(player, i, dimensionIn);
     }
 
     public void transferEntityToWorld(Entity entityIn, int lastDimension, WorldServer oldWorldIn, WorldServer toWorldIn)
     {
-        double d0 = entityIn.posX;
-        double d1 = entityIn.posZ;
+        transferEntityToWorld(entityIn, lastDimension, oldWorldIn, toWorldIn, toWorldIn.getDefaultTeleporter());
+    }
+
+    // TODO: Remove (1.13)
+    public void transferEntityToWorld(Entity entityIn, int lastDimension, WorldServer oldWorldIn, WorldServer toWorldIn, net.minecraft.world.Teleporter teleporter)
+    {
+        transferEntityToWorld(entityIn, lastDimension, oldWorldIn, toWorldIn, (net.minecraftforge.common.util.ITeleporter) teleporter);
+    }
+
+    public void transferEntityToWorld(Entity entityIn, int lastDimension, WorldServer oldWorldIn, WorldServer toWorldIn, net.minecraftforge.common.util.ITeleporter teleporter)
+    {
+        double moveFactor = oldWorldIn.provider.getMovementFactor() / toWorldIn.provider.getMovementFactor();
+        double d0 = MathHelper.clamp(entityIn.posX * moveFactor, toWorldIn.getWorldBorder().minX() + 16.0, toWorldIn.getWorldBorder().maxX() - 16.0);
+        double d1 = MathHelper.clamp(entityIn.posZ * moveFactor, toWorldIn.getWorldBorder().minZ() + 16.0, toWorldIn.getWorldBorder().maxZ() - 16.0);
         double d2 = 8.0;
         float f = entityIn.rotationYaw;
         oldWorldIn.profiler.startSection("moving");
 
-        if (entityIn.dimension == -1)
+        if (false && entityIn.dimension == -1)
         {
             d0 = MathHelper.clamp(d0 / 8.0, toWorldIn.getWorldBorder().minX() + 16.0, toWorldIn.getWorldBorder().maxX() - 16.0);
             d1 = MathHelper.clamp(d1 / 8.0, toWorldIn.getWorldBorder().minZ() + 16.0, toWorldIn.getWorldBorder().maxZ() - 16.0);
@@ -650,7 +724,7 @@
                 oldWorldIn.updateEntityWithOptionalForce(entityIn, false);
             }
         }
-        else if (entityIn.dimension == 0)
+        else if (false && entityIn.dimension == 0)
         {
             d0 = MathHelper.clamp(d0 * 8.0, toWorldIn.getWorldBorder().minX() + 16.0, toWorldIn.getWorldBorder().maxX() - 16.0);
             d1 = MathHelper.clamp(d1 * 8.0, toWorldIn.getWorldBorder().minZ() + 16.0, toWorldIn.getWorldBorder().maxZ() - 16.0);
@@ -661,7 +735,7 @@
                 oldWorldIn.updateEntityWithOptionalForce(entityIn, false);
             }
         }
-        else
+        if (entityIn.dimension == 1 && teleporter.isVanilla())
         {
             BlockPos blockpos;
 
@@ -687,7 +761,7 @@
 
         oldWorldIn.profiler.endSection();
 
-        if (lastDimension != 1)
+        if (lastDimension != 1 || !teleporter.isVanilla())
         {
             oldWorldIn.profiler.startSection("placing");
             d0 = MathHelper.clamp((int)d0, -29999872, 29999872);
@@ -696,7 +770,8 @@
             if (entityIn.isEntityAlive())
             {
                 entityIn.setLocationAndAngles(d0, entityIn.posY, d1, entityIn.rotationYaw, entityIn.rotationPitch);
-                toWorldIn.getDefaultTeleporter().placeInPortal(entityIn, f);
+                oldWorldIn.updateEntityWithOptionalForce(entityIn, false);
+                teleporter.placeEntity(toWorldIn, entityIn, f);
                 toWorldIn.spawnEntity(entityIn);
                 toWorldIn.updateEntityWithOptionalForce(entityIn, false);
             }
