--- before/net/minecraft/server/management/UserListBans.java
+++ after/net/minecraft/server/management/UserListBans.java
@@ -28,7 +28,7 @@
 
         for (UserListBansEntry userlistbansentry : this.func_152688_e().values())
         {
-            astring[i++] = userlistbansentry.func_152640_f().getName();
+            astring[i++] = ((GameProfile)userlistbansentry.func_152640_f()).getName();
         }
 
         return astring;
@@ -43,9 +43,9 @@
     {
         for (UserListBansEntry userlistbansentry : this.func_152688_e().values())
         {
-            if (p_152703_1_.equalsIgnoreCase(userlistbansentry.func_152640_f().getName()))
+            if (p_152703_1_.equalsIgnoreCase(((GameProfile)userlistbansentry.func_152640_f()).getName()))
             {
-                return userlistbansentry.func_152640_f();
+                return (GameProfile)userlistbansentry.func_152640_f();
             }
         }
 
