--- before/net/minecraft/server/management/UserList.java
+++ after/net/minecraft/server/management/UserList.java
@@ -37,7 +37,7 @@
     protected static final Logger field_152693_a = LogManager.getLogger();
     protected final Gson field_152694_b;
     private final File field_152695_c;
-    private final Map<String, V> field_152696_d = Maps.newHashMap();
+    private final Map<String, V> field_152696_d = Maps.<String, V>newHashMap();
     private boolean field_152697_e = true;
     private static final ParameterizedType field_152698_f = new ParameterizedType()
     {
@@ -90,7 +90,7 @@
     public V func_152683_b(K p_152683_1_)
     {
         this.func_152680_h();
-        return this.field_152696_d.get(this.func_152681_a(p_152683_1_));
+        return (V)(this.field_152696_d.get(this.func_152681_a(p_152683_1_)));
     }
 
     public void func_152684_c(K p_152684_1_)
@@ -115,7 +115,7 @@
 
     public String[] func_152685_a()
     {
-        return this.field_152696_d.keySet().toArray(new String[this.field_152696_d.size()]);
+        return (String[])this.field_152696_d.keySet().toArray(new String[this.field_152696_d.size()]);
     }
 
     protected String func_152681_a(K p_152681_1_)
@@ -130,7 +130,7 @@
 
     private void func_152680_h()
     {
-        List<K> list = Lists.newArrayList();
+        List<K> list = Lists.<K>newArrayList();
 
         for (V v : this.field_152696_d.values())
         {
@@ -148,7 +148,7 @@
 
     protected UserListEntry<K> func_152682_a(JsonObject p_152682_1_)
     {
-        return new UserListEntry<>((K)null, p_152682_1_);
+        return new UserListEntry<K>(null, p_152682_1_);
     }
 
     protected Map<String, V> func_152688_e()
@@ -190,7 +190,7 @@
             try
             {
                 bufferedreader = Files.newReader(this.field_152695_c, StandardCharsets.UTF_8);
-                collection = JsonUtils.func_193841_a(this.field_152694_b, bufferedreader, field_152698_f);
+                collection = (Collection)JsonUtils.func_193841_a(this.field_152694_b, bufferedreader, field_152698_f);
             }
             finally
             {
