--- before/net/minecraft/server/management/UserListOps.java
+++ after/net/minecraft/server/management/UserListOps.java
@@ -23,7 +23,7 @@
 
         for (UserListOpsEntry userlistopsentry : this.func_152688_e().values())
         {
-            astring[i++] = userlistopsentry.func_152640_f().getName();
+            astring[i++] = ((GameProfile)userlistopsentry.func_152640_f()).getName();
         }
 
         return astring;
@@ -31,13 +31,13 @@
 
     public int func_187452_a(GameProfile p_187452_1_)
     {
-        UserListOpsEntry userlistopsentry = this.func_152683_b(p_187452_1_);
+        UserListOpsEntry userlistopsentry = (UserListOpsEntry)this.func_152683_b(p_187452_1_);
         return userlistopsentry != null ? userlistopsentry.func_152644_a() : 0;
     }
 
     public boolean func_183026_b(GameProfile p_183026_1_)
     {
-        UserListOpsEntry userlistopsentry = this.func_152683_b(p_183026_1_);
+        UserListOpsEntry userlistopsentry = (UserListOpsEntry)this.func_152683_b(p_183026_1_);
         return userlistopsentry != null ? userlistopsentry.func_183024_b() : false;
     }
 
@@ -50,9 +50,9 @@
     {
         for (UserListOpsEntry userlistopsentry : this.func_152688_e().values())
         {
-            if (p_152700_1_.equalsIgnoreCase(userlistopsentry.func_152640_f().getName()))
+            if (p_152700_1_.equalsIgnoreCase(((GameProfile)userlistopsentry.func_152640_f()).getName()))
             {
-                return userlistopsentry.func_152640_f();
+                return (GameProfile)userlistopsentry.func_152640_f();
             }
         }
 
