--- before/net/minecraft/server/management/UserListWhitelist.java
+++ after/net/minecraft/server/management/UserListWhitelist.java
@@ -18,12 +18,6 @@
         return new UserListWhitelistEntry(p_152682_1_);
     }
 
-    @SideOnly(Side.SERVER)
-    public boolean func_152705_a(GameProfile p_152705_1_)
-    {
-        return this.func_152692_d(p_152705_1_);
-    }
-
     public String[] func_152685_a()
     {
         String[] astring = new String[this.func_152688_e().size()];
@@ -31,12 +25,18 @@
 
         for (UserListWhitelistEntry userlistwhitelistentry : this.func_152688_e().values())
         {
-            astring[i++] = userlistwhitelistentry.func_152640_f().getName();
+            astring[i++] = ((GameProfile)userlistwhitelistentry.func_152640_f()).getName();
         }
 
         return astring;
     }
 
+    @SideOnly(Side.SERVER)
+    public boolean func_152705_a(GameProfile p_152705_1_)
+    {
+        return this.func_152692_d(p_152705_1_);
+    }
+
     protected String func_152681_a(GameProfile p_152681_1_)
     {
         return p_152681_1_.getId().toString();
@@ -46,9 +46,9 @@
     {
         for (UserListWhitelistEntry userlistwhitelistentry : this.func_152688_e().values())
         {
-            if (p_152706_1_.equalsIgnoreCase(userlistwhitelistentry.func_152640_f().getName()))
+            if (p_152706_1_.equalsIgnoreCase(((GameProfile)userlistwhitelistentry.func_152640_f()).getName()))
             {
-                return userlistwhitelistentry.func_152640_f();
+                return (GameProfile)userlistwhitelistentry.func_152640_f();
             }
         }
 
