--- before/net/minecraft/server/management/UserListIPBansEntry.java
+++ after/net/minecraft/server/management/UserListIPBansEntry.java
@@ -29,7 +29,7 @@
     {
         if (this.func_152640_f() != null)
         {
-            p_152641_1_.addProperty("ip", this.func_152640_f());
+            p_152641_1_.addProperty("ip", (String)this.func_152640_f());
             super.func_152641_a(p_152641_1_);
         }
     }
