--- before/net/minecraft/server/management/UserListOpsEntry.java
+++ after/net/minecraft/server/management/UserListOpsEntry.java
@@ -37,11 +37,11 @@
     {
         if (this.func_152640_f() != null)
         {
-            p_152641_1_.addProperty("uuid", this.func_152640_f().getId() == null ? "" : this.func_152640_f().getId().toString());
-            p_152641_1_.addProperty("name", this.func_152640_f().getName());
+            p_152641_1_.addProperty("uuid", ((GameProfile)this.func_152640_f()).getId() == null ? "" : ((GameProfile)this.func_152640_f()).getId().toString());
+            p_152641_1_.addProperty("name", ((GameProfile)this.func_152640_f()).getName());
             super.func_152641_a(p_152641_1_);
-            p_152641_1_.addProperty("level", this.field_152645_a);
-            p_152641_1_.addProperty("bypassesPlayerLimit", this.field_183025_b);
+            p_152641_1_.addProperty("level", Integer.valueOf(this.field_152645_a));
+            p_152641_1_.addProperty("bypassesPlayerLimit", Boolean.valueOf(this.field_183025_b));
         }
     }
 
@@ -56,7 +56,7 @@
             {
                 uuid = UUID.fromString(s);
             }
-            catch (Throwable throwable)
+            catch (Throwable var4)
             {
                 return null;
             }
