--- before/net/minecraft/client/renderer/color/BlockColors.java
+++ after/net/minecraft/client/renderer/color/BlockColors.java
@@ -27,7 +27,8 @@
 @SideOnly(Side.CLIENT)
 public class BlockColors
 {
-    private final ObjectIntIdentityMap<IBlockColor> field_186725_a = new ObjectIntIdentityMap<>(32);
+    // FORGE: Use RegistryDelegates as non-Vanilla block ids are not constant
+    private final java.util.Map<net.minecraftforge.registries.IRegistryDelegate<Block>, IBlockColor> blockColorMap = com.google.common.collect.Maps.newHashMap();
 
     public static BlockColors func_186723_a()
     {
@@ -36,7 +37,7 @@
         {
             public int func_186720_a(IBlockState p_186720_1_, @Nullable IBlockAccess p_186720_2_, @Nullable BlockPos p_186720_3_, int p_186720_4_)
             {
-                BlockDoublePlant.EnumPlantType blockdoubleplant$enumplanttype = p_186720_1_.func_177229_b(BlockDoublePlant.field_176493_a);
+                BlockDoublePlant.EnumPlantType blockdoubleplant$enumplanttype = (BlockDoublePlant.EnumPlantType)p_186720_1_.func_177229_b(BlockDoublePlant.field_176493_a);
                 return p_186720_2_ != null && p_186720_3_ != null && (blockdoubleplant$enumplanttype == BlockDoublePlant.EnumPlantType.GRASS || blockdoubleplant$enumplanttype == BlockDoublePlant.EnumPlantType.FERN) ? BiomeColorHelper.func_180286_a(p_186720_2_, p_186720_1_.func_177229_b(BlockDoublePlant.field_176492_b) == BlockDoublePlant.EnumBlockHalf.UPPER ? p_186720_3_.func_177977_b() : p_186720_3_) : -1;
             }
         }, Blocks.field_150398_cm);
@@ -76,7 +77,7 @@
         {
             public int func_186720_a(IBlockState p_186720_1_, @Nullable IBlockAccess p_186720_2_, @Nullable BlockPos p_186720_3_, int p_186720_4_)
             {
-                BlockPlanks.EnumType blockplanks$enumtype = p_186720_1_.func_177229_b(BlockOldLeaf.field_176239_P);
+                BlockPlanks.EnumType blockplanks$enumtype = (BlockPlanks.EnumType)p_186720_1_.func_177229_b(BlockOldLeaf.field_176239_P);
 
                 if (blockplanks$enumtype == BlockPlanks.EnumType.SPRUCE)
                 {
@@ -110,7 +111,7 @@
         {
             public int func_186720_a(IBlockState p_186720_1_, @Nullable IBlockAccess p_186720_2_, @Nullable BlockPos p_186720_3_, int p_186720_4_)
             {
-                return BlockRedstoneWire.func_176337_b(p_186720_1_.func_177229_b(BlockRedstoneWire.field_176351_O));
+                return BlockRedstoneWire.func_176337_b(((Integer)p_186720_1_.func_177229_b(BlockRedstoneWire.field_176351_O)).intValue());
             }
         }, Blocks.field_150488_af);
         blockcolors.func_186722_a(new IBlockColor()
@@ -124,7 +125,7 @@
         {
             public int func_186720_a(IBlockState p_186720_1_, @Nullable IBlockAccess p_186720_2_, @Nullable BlockPos p_186720_3_, int p_186720_4_)
             {
-                int i = p_186720_1_.func_177229_b(BlockStem.field_176484_a);
+                int i = ((Integer)p_186720_1_.func_177229_b(BlockStem.field_176484_a)).intValue();
                 int j = i * 32;
                 int k = 255 - i * 8;
                 int l = i * 4;
@@ -159,12 +160,13 @@
                 return p_186720_2_ != null && p_186720_3_ != null ? 2129968 : 7455580;
             }
         }, Blocks.field_150392_bi);
+        net.minecraftforge.client.ForgeHooksClient.onBlockColorsInit(blockcolors);
         return blockcolors;
     }
 
     public int func_189991_a(IBlockState p_189991_1_, World p_189991_2_, BlockPos p_189991_3_)
     {
-        IBlockColor iblockcolor = this.field_186725_a.func_148745_a(Block.func_149682_b(p_189991_1_.func_177230_c()));
+        IBlockColor iblockcolor = this.blockColorMap.get(p_189991_1_.func_177230_c().delegate);
 
         if (iblockcolor != null)
         {
@@ -179,7 +181,7 @@
 
     public int func_186724_a(IBlockState p_186724_1_, @Nullable IBlockAccess p_186724_2_, @Nullable BlockPos p_186724_3_, int p_186724_4_)
     {
-        IBlockColor iblockcolor = this.field_186725_a.func_148745_a(Block.func_149682_b(p_186724_1_.func_177230_c()));
+        IBlockColor iblockcolor = this.blockColorMap.get(p_186724_1_.func_177230_c().delegate);
         return iblockcolor == null ? -1 : iblockcolor.func_186720_a(p_186724_1_, p_186724_2_, p_186724_3_, p_186724_4_);
     }
 
@@ -187,7 +189,9 @@
     {
         for (Block block : p_186722_2_)
         {
-            this.field_186725_a.func_148746_a(p_186722_1_, Block.func_149682_b(block));
+            if (block == null) throw new IllegalArgumentException("Block registered to block color handler cannot be null!");
+            if (block.getRegistryName() == null) throw new IllegalArgumentException("Block must be registered before assigning color handler.");
+            this.blockColorMap.put(block.delegate, p_186722_1_);
         }
     }
 }
