--- before/net/minecraft/client/renderer/color/ItemColors.java
+++ after/net/minecraft/client/renderer/color/ItemColors.java
@@ -17,44 +17,38 @@
 import net.minecraft.nbt.NBTTagIntArray;
 import net.minecraft.potion.PotionUtils;
 import net.minecraft.util.ObjectIntIdentityMap;
+import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.ColorizerGrass;
+import net.minecraft.world.IBlockAccess;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 @SideOnly(Side.CLIENT)
 public class ItemColors
 {
-    private final ObjectIntIdentityMap<IItemColor> field_186732_a = new ObjectIntIdentityMap<>(32);
+    // FORGE: Use RegistryDelegates as non-Vanilla item ids are not constant
+    private final java.util.Map<net.minecraftforge.registries.IRegistryDelegate<Item>, IItemColor> itemColorMap = com.google.common.collect.Maps.newHashMap();
 
     public static ItemColors func_186729_a(final BlockColors p_186729_0_)
     {
         ItemColors itemcolors = new ItemColors();
         itemcolors.func_186730_a(new IItemColor()
         {
-            @Override
             public int func_186726_a(ItemStack p_186726_1_, int p_186726_2_)
             {
                 return p_186726_2_ > 0 ? -1 : ((ItemArmor)p_186726_1_.func_77973_b()).func_82814_b(p_186726_1_);
             }
         }, Items.field_151024_Q, Items.field_151027_R, Items.field_151026_S, Items.field_151021_T);
-        itemcolors.func_186731_a(
-            new IItemColor()
+        itemcolors.func_186731_a(new IItemColor()
         {
-            @Override
             public int func_186726_a(ItemStack p_186726_1_, int p_186726_2_)
             {
                 BlockDoublePlant.EnumPlantType blockdoubleplant$enumplanttype = BlockDoublePlant.EnumPlantType.func_176938_a(p_186726_1_.func_77960_j());
-                return blockdoubleplant$enumplanttype != BlockDoublePlant.EnumPlantType.GRASS
-                       && blockdoubleplant$enumplanttype != BlockDoublePlant.EnumPlantType.FERN
-                       ? -1
-                       : ColorizerGrass.func_77480_a(0.5, 1.0);
+                return blockdoubleplant$enumplanttype != BlockDoublePlant.EnumPlantType.GRASS && blockdoubleplant$enumplanttype != BlockDoublePlant.EnumPlantType.FERN ? -1 : ColorizerGrass.func_77480_a(0.5D, 1.0D);
             }
-        },
-        Blocks.field_150398_cm
-        );
+        }, Blocks.field_150398_cm);
         itemcolors.func_186730_a(new IItemColor()
         {
-            @Override
             public int func_186726_a(ItemStack p_186726_1_, int p_186726_2_)
             {
                 if (p_186726_2_ != 1)
@@ -85,14 +79,14 @@
 
                             for (int l : aint)
                             {
-                                i += (l & 0xFF0000) >> 16;
-                                j += (l & 0xFF00) >> 8;
-                                k += (l & 0xFF) >> 0;
+                                i += (l & 16711680) >> 16;
+                                j += (l & 65280) >> 8;
+                                k += (l & 255) >> 0;
                             }
 
-                            i /= aint.length;
-                            j /= aint.length;
-                            k /= aint.length;
+                            i = i / aint.length;
+                            j = j / aint.length;
+                            k = k / aint.length;
                             return i << 16 | j << 8 | k;
                         }
                     }
@@ -101,7 +95,6 @@
         }, Items.field_151154_bQ);
         itemcolors.func_186730_a(new IItemColor()
         {
-            @Override
             public int func_186726_a(ItemStack p_186726_1_, int p_186726_2_)
             {
                 return p_186726_2_ > 0 ? -1 : PotionUtils.func_190932_c(p_186726_1_);
@@ -109,7 +102,6 @@
         }, Items.field_151068_bn, Items.field_185155_bH, Items.field_185156_bI);
         itemcolors.func_186730_a(new IItemColor()
         {
-            @Override
             public int func_186726_a(ItemStack p_186726_1_, int p_186726_2_)
             {
                 EntityList.EntityEggInfo entitylist$entityegginfo = EntityList.field_75627_a.get(ItemMonsterPlacer.func_190908_h(p_186726_1_));
@@ -126,16 +118,14 @@
         }, Items.field_151063_bx);
         itemcolors.func_186731_a(new IItemColor()
         {
-            @Override
             public int func_186726_a(ItemStack p_186726_1_, int p_186726_2_)
             {
                 IBlockState iblockstate = ((ItemBlock)p_186726_1_.func_77973_b()).func_179223_d().func_176203_a(p_186726_1_.func_77960_j());
-                return p_186729_0_.func_186724_a(iblockstate, null, null, p_186726_2_);
+                return p_186729_0_.func_186724_a(iblockstate, (IBlockAccess)null, (BlockPos)null, p_186726_2_);
             }
         }, Blocks.field_150349_c, Blocks.field_150329_H, Blocks.field_150395_bd, Blocks.field_150362_t, Blocks.field_150361_u, Blocks.field_150392_bi);
         itemcolors.func_186730_a(new IItemColor()
         {
-            @Override
             public int func_186726_a(ItemStack p_186726_1_, int p_186726_2_)
             {
                 return p_186726_2_ == 0 ? PotionUtils.func_190932_c(p_186726_1_) : -1;
@@ -143,18 +133,18 @@
         }, Items.field_185167_i);
         itemcolors.func_186730_a(new IItemColor()
         {
-            @Override
             public int func_186726_a(ItemStack p_186726_1_, int p_186726_2_)
             {
                 return p_186726_2_ == 0 ? -1 : ItemMap.func_190907_h(p_186726_1_);
             }
         }, Items.field_151098_aY);
+        net.minecraftforge.client.ForgeHooksClient.onItemColorsInit(itemcolors, p_186729_0_);
         return itemcolors;
     }
 
     public int func_186728_a(ItemStack p_186728_1_, int p_186728_2_)
     {
-        IItemColor iitemcolor = this.field_186732_a.func_148745_a(Item.field_150901_e.func_148757_b(p_186728_1_.func_77973_b()));
+        IItemColor iitemcolor = this.itemColorMap.get(p_186728_1_.func_77973_b().delegate);
         return iitemcolor == null ? -1 : iitemcolor.func_186726_a(p_186728_1_, p_186728_2_);
     }
 
@@ -162,7 +152,9 @@
     {
         for (Block block : p_186731_2_)
         {
-            this.field_186732_a.func_148746_a(p_186731_1_, Item.func_150891_b(Item.func_150898_a(block)));
+            if (block == null) throw new IllegalArgumentException("Block registered to item color handler cannot be null!");
+            if (block.getRegistryName() == null) throw new IllegalArgumentException("Block must be registered before assigning color handler.");
+            this.itemColorMap.put(Item.func_150898_a(block).delegate, p_186731_1_);
         }
     }
 
@@ -170,7 +162,9 @@
     {
         for (Item item : p_186730_2_)
         {
-            this.field_186732_a.func_148746_a(p_186730_1_, Item.func_150891_b(item));
+            if (item == null) throw new IllegalArgumentException("Item registered to item color handler cannot be null!");
+            if (item.getRegistryName() == null) throw new IllegalArgumentException("Item must be registered before assigning color handler.");
+            this.itemColorMap.put(item.delegate, p_186730_1_);
         }
     }
 }
