--- before/net/minecraft/client/util/JsonBlendingMode.java
+++ after/net/minecraft/client/util/JsonBlendingMode.java
@@ -128,7 +128,8 @@
         i = 31 * i + this.field_148115_e;
         i = 31 * i + this.field_148112_f;
         i = 31 * i + (this.field_148113_g ? 1 : 0);
-        return 31 * i + (this.field_148119_h ? 1 : 0);
+        i = 31 * i + (this.field_148119_h ? 1 : 0);
+        return i;
     }
 
     public boolean func_148111_b()
