--- before/net/minecraft/client/util/RecipeItemHelper.java
+++ after/net/minecraft/client/util/RecipeItemHelper.java
@@ -8,8 +8,10 @@
 import it.unimi.dsi.fastutil.ints.IntCollection;
 import it.unimi.dsi.fastutil.ints.IntIterator;
 import it.unimi.dsi.fastutil.ints.IntList;
+import it.unimi.dsi.fastutil.ints.IntListIterator;
 import java.util.BitSet;
 import java.util.List;
+import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
@@ -22,10 +24,15 @@
 
     public void func_194112_a(ItemStack p_194112_1_)
     {
+        this.accountStack(p_194112_1_, -1);
+    }
+
+    public void accountStack(ItemStack p_194112_1_, int forceCount)
+    {
         if (!p_194112_1_.func_190926_b() && !p_194112_1_.func_77951_h() && !p_194112_1_.func_77948_v() && !p_194112_1_.func_82837_s())
         {
             int i = func_194113_b(p_194112_1_);
-            int j = p_194112_1_.func_190916_E();
+            int j = forceCount == -1 ? p_194112_1_.func_190916_E() : forceCount;
             this.func_194117_b(i, j);
         }
     }
@@ -69,7 +76,7 @@
 
     public boolean func_194118_a(IRecipe p_194118_1_, @Nullable IntList p_194118_2_, int p_194118_3_)
     {
-        return new RecipeItemHelper.RecipePicker(p_194118_1_).func_194092_a(p_194118_3_, p_194118_2_);
+        return (new RecipeItemHelper.RecipePicker(p_194118_1_)).func_194092_a(p_194118_3_, p_194118_2_);
     }
 
     public int func_194114_b(IRecipe p_194114_1_, @Nullable IntList p_194114_2_)
@@ -79,7 +86,7 @@
 
     public int func_194121_a(IRecipe p_194121_1_, int p_194121_2_, @Nullable IntList p_194121_3_)
     {
-        return new RecipeItemHelper.RecipePicker(p_194121_1_).func_194102_b(p_194121_2_, p_194121_3_);
+        return (new RecipeItemHelper.RecipePicker(p_194121_1_)).func_194102_b(p_194121_2_, p_194121_3_);
     }
 
     public static ItemStack func_194115_b(int p_194115_0_)
@@ -106,17 +113,20 @@
         {
             this.field_194105_b = p_i47608_2_;
             this.field_194106_c.addAll(p_i47608_2_.func_192400_c());
-            this.field_194106_c.removeIf(p_194103_0_ -> p_194103_0_ == Ingredient.field_193370_a);
+            this.field_194106_c.removeIf((p_194103_0_) ->
+            {
+                return p_194103_0_ == Ingredient.field_193370_a;
+            });
             this.field_194107_d = this.field_194106_c.size();
             this.field_194108_e = this.func_194097_a();
             this.field_194109_f = this.field_194108_e.length;
             this.field_194110_g = new BitSet(this.field_194107_d + this.field_194109_f + this.field_194107_d + this.field_194107_d * this.field_194109_f);
 
-            for (int i = 0; i < this.field_194106_c.size(); i++)
+            for (int i = 0; i < this.field_194106_c.size(); ++i)
             {
                 IntList intlist = this.field_194106_c.get(i).func_194139_b();
 
-                for (int j = 0; j < this.field_194109_f; j++)
+                for (int j = 0; j < this.field_194109_f; ++j)
                 {
                     if (intlist.contains(this.field_194108_e[j]))
                     {
@@ -136,13 +146,13 @@
             {
                 int i;
 
-                for (i = 0; this.func_194098_a(p_194092_1_); i++)
+                for (i = 0; this.func_194098_a(p_194092_1_); ++i)
                 {
                     RecipeItemHelper.this.func_194122_a(this.field_194108_e[this.field_194111_h.getInt(0)], p_194092_1_);
                     int j = this.field_194111_h.size() - 1;
                     this.func_194096_c(this.field_194111_h.getInt(j));
 
-                    for (int k = 0; k < j; k++)
+                    for (int k = 0; k < j; ++k)
                     {
                         this.func_194089_c((k & 1) == 0, this.field_194111_h.get(k), this.field_194111_h.get(k + 1));
                     }
@@ -163,7 +173,7 @@
                 int l = 0;
                 List<Ingredient> list = this.field_194105_b.func_192400_c();
 
-                for (int i1 = 0; i1 < list.size(); i1++)
+                for (int i1 = 0; i1 < list.size(); ++i1)
                 {
                     if (flag1 && list.get(i1) == Ingredient.field_193370_a)
                     {
@@ -171,7 +181,7 @@
                     }
                     else
                     {
-                        for (int j1 = 0; j1 < this.field_194109_f; j1++)
+                        for (int j1 = 0; j1 < this.field_194109_f; ++j1)
                         {
                             if (this.func_194100_b(false, l, j1))
                             {
@@ -185,7 +195,7 @@
                             }
                         }
 
-                        l++;
+                        ++l;
                     }
                 }
 
@@ -219,7 +229,7 @@
         {
             int i = this.field_194109_f;
 
-            for (int j = 0; j < i; j++)
+            for (int j = 0; j < i; ++j)
             {
                 if (RecipeItemHelper.this.field_194124_a.get(this.field_194108_e[j]) >= p_194098_1_)
                 {
@@ -237,29 +247,21 @@
                         }
 
                         int i1 = flag ? this.field_194107_d : i;
-                        int j1 = 0;
 
-                        while (true)
+                        for (int j1 = 0; j1 < i1; ++j1)
                         {
-                            if (j1 < i1)
+                            if (!this.func_194101_b(flag, j1) && this.func_194093_a(flag, l, j1) && this.func_194100_b(flag, l, j1))
                             {
-                                if (this.func_194101_b(flag, j1) || !this.func_194093_a(flag, l, j1) || !this.func_194100_b(flag, l, j1))
-                                {
-                                    j1++;
-                                    continue;
-                                }
-
                                 this.func_194088_a(flag, j1);
-                            }
-
-                            j1 = this.field_194111_h.size();
-
-                            if (j1 == k)
-                            {
-                                this.field_194111_h.removeInt(j1 - 1);
-                            }
-
-                            break;
+                                break;
+                            }
+                        }
+
+                        int k1 = this.field_194111_h.size();
+
+                        if (k1 == k)
+                        {
+                            this.field_194111_h.removeInt(k1 - 1);
                         }
                     }
 
@@ -334,7 +336,7 @@
             {
                 int k = (i + j) / 2;
 
-                if (this.func_194092_a(k, null))
+                if (this.func_194092_a(k, (IntList)null))
                 {
                     if (j - i <= 1)
                     {
