--- before/net/minecraft/client/util/RecipeItemHelper.java
+++ after/net/minecraft/client/util/RecipeItemHelper.java
@@ -22,10 +22,15 @@
 
     public void accountStack(ItemStack stack)
     {
+        this.accountStack(stack, -1);
+    }
+
+    public void accountStack(ItemStack stack, int forceCount)
+    {
         if (!stack.isEmpty() && !stack.isItemDamaged() && !stack.isItemEnchanted() && !stack.hasDisplayName())
         {
             int i = pack(stack);
-            int j = stack.getCount();
+            int j = forceCount == -1 ? stack.getCount() : forceCount;
             this.increment(i, j);
         }
     }
@@ -237,29 +242,21 @@
                         }
 
                         int i1 = flag ? this.ingredientCount : i;
-                        int j1 = 0;
 
-                        while (true)
+                        for (int j1 = 0; j1 < i1; ++j1)
                         {
-                            if (j1 < i1)
+                            if (!this.hasVisited(flag, j1) && this.hasConnection(flag, l, j1) && this.hasResidual(flag, l, j1))
                             {
-                                if (this.hasVisited(flag, j1) || !this.hasConnection(flag, l, j1) || !this.hasResidual(flag, l, j1))
-                                {
-                                    j1++;
-                                    continue;
-                                }
-
                                 this.visit(flag, j1);
-                            }
-
-                            j1 = this.path.size();
-
-                            if (j1 == k)
-                            {
-                                this.path.removeInt(j1 - 1);
-                            }
-
-                            break;
+                                break;
+                            }
+                        }
+
+                        int k1 = this.path.size();
+
+                        if (k1 == k)
+                        {
+                            this.path.removeInt(k1 - 1);
                         }
                     }
 
