--- before/net/minecraft/client/util/RecipeBookClient.java
+++ after/net/minecraft/client/util/RecipeBookClient.java
@@ -7,6 +7,7 @@
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Function;
 import net.minecraft.client.gui.recipebook.RecipeList;
 import net.minecraft.creativetab.CreativeTabs;
 import net.minecraft.item.ItemStack;
@@ -19,15 +20,21 @@
 @SideOnly(Side.CLIENT)
 public class RecipeBookClient extends RecipeBook
 {
-    public static final Map<CreativeTabs, List<RecipeList>> field_194086_e = Maps.newHashMap();
-    public static final List<RecipeList> field_194087_f = Lists.newArrayList();
+    public static final Map<CreativeTabs, List<RecipeList>> field_194086_e = Maps.<CreativeTabs, List<RecipeList>>newHashMap();
+    public static final List<RecipeList> field_194087_f = Lists.<RecipeList>newArrayList();
 
     private static RecipeList func_194082_a(CreativeTabs p_194082_0_)
     {
         RecipeList recipelist = new RecipeList();
         field_194087_f.add(recipelist);
-        field_194086_e.computeIfAbsent(p_194082_0_, (p_194085_0_) -> new ArrayList()).add(recipelist);
-        field_194086_e.computeIfAbsent(CreativeTabs.field_78027_g, (p_194083_0_) -> new ArrayList()).add(recipelist);
+        (field_194086_e.computeIfAbsent(p_194082_0_, (p_194085_0_) ->
+        {
+            return new ArrayList();
+        })).add(recipelist);
+        (field_194086_e.computeIfAbsent(CreativeTabs.field_78027_g, (p_194083_0_) ->
+        {
+            return new ArrayList();
+        })).add(recipelist);
         return recipelist;
     }
 
@@ -47,7 +54,15 @@
 
     static
     {
-        Table<CreativeTabs, String, RecipeList> table = HashBasedTable.create();
+        rebuildTable();
+    }
+
+    public static void rebuildTable()
+    {
+        field_194086_e.clear();
+        field_194087_f.clear();
+
+        Table<CreativeTabs, String, RecipeList> table = HashBasedTable.<CreativeTabs, String, RecipeList>create();
 
         for (IRecipe irecipe : CraftingManager.field_193380_a)
         {