--- before/net/minecraft/entity/EntityAgeable.java
+++ after/net/minecraft/entity/EntityAgeable.java
@@ -15,7 +15,7 @@
 
 public abstract class EntityAgeable extends EntityCreature
 {
-    private static final DataParameter<Boolean> field_184751_bv = EntityDataManager.func_187226_a(EntityAgeable.class, DataSerializers.field_187198_h);
+    private static final DataParameter<Boolean> field_184751_bv = EntityDataManager.<Boolean>func_187226_a(EntityAgeable.class, DataSerializers.field_187198_h);
     protected int field_175504_a;
     protected int field_175502_b;
     protected int field_175503_c;
@@ -30,7 +30,6 @@
     @Nullable
     public abstract EntityAgeable func_90011_a(EntityAgeable p_90011_1_);
 
-    @Override
     public boolean func_184645_a(EntityPlayer p_184645_1_, EnumHand p_184645_2_)
     {
         ItemStack itemstack = p_184645_1_.func_184586_b(p_184645_2_);
@@ -39,7 +38,7 @@
         {
             if (!this.field_70170_p.field_72995_K)
             {
-                Class <? extends Entity > oclass = EntityList.field_191308_b.func_82594_a(ItemMonsterPlacer.func_190908_h(itemstack));
+                Class <? extends Entity > oclass = EntityList.getClass(ItemMonsterPlacer.func_190908_h(itemstack));
 
                 if (oclass != null && this.getClass() == oclass)
                 {
@@ -80,23 +79,22 @@
         }
         else
         {
-            Class <? extends Entity > oclass = EntityList.field_191308_b.func_82594_a(ItemMonsterPlacer.func_190908_h(p_190669_1_));
+            Class <? extends Entity > oclass = EntityList.getClass(ItemMonsterPlacer.func_190908_h(p_190669_1_));
             return oclass != null && p_190669_2_ == oclass;
         }
     }
 
-    @Override
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184751_bv, false);
+        this.field_70180_af.func_187214_a(field_184751_bv, Boolean.valueOf(false));
     }
 
     public int func_70874_b()
     {
         if (this.field_70170_p.field_72995_K)
         {
-            return this.field_70180_af.func_187225_a(field_184751_bv) ? -1 : 1;
+            return ((Boolean)this.field_70180_af.func_187225_a(field_184751_bv)).booleanValue() ? -1 : 1;
         }
         else
         {
@@ -107,24 +105,25 @@
     public void func_175501_a(int p_175501_1_, boolean p_175501_2_)
     {
         int i = this.func_70874_b();
-        i += p_175501_1_ * 20;
+        int j = i;
+        i = i + p_175501_1_ * 20;
 
         if (i > 0)
         {
             i = 0;
 
-            if (i < 0)
+            if (j < 0)
             {
                 this.func_175500_n();
             }
         }
 
-        int j = i - i;
+        int k = i - j;
         this.func_70873_a(i);
 
         if (p_175501_2_)
         {
-            this.field_175502_b += j;
+            this.field_175502_b += k;
 
             if (this.field_175503_c == 0)
             {
@@ -145,12 +144,11 @@
 
     public void func_70873_a(int p_70873_1_)
     {
-        this.field_70180_af.func_187227_b(field_184751_bv, p_70873_1_ < 0);
+        this.field_70180_af.func_187227_b(field_184751_bv, Boolean.valueOf(p_70873_1_ < 0));
         this.field_175504_a = p_70873_1_;
         this.func_98054_a(this.func_70631_g_());
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
@@ -158,7 +156,6 @@
         p_70014_1_.func_74768_a("ForcedAge", this.field_175502_b);
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
@@ -166,7 +163,6 @@
         this.field_175502_b = p_70037_1_.func_74762_e("ForcedAge");
     }
 
-    @Override
     public void func_184206_a(DataParameter<?> p_184206_1_)
     {
         if (field_184751_bv.equals(p_184206_1_))
@@ -177,7 +173,6 @@
         super.func_184206_a(p_184206_1_);
     }
 
-    @Override
     public void func_70636_d()
     {
         super.func_70636_d();
@@ -188,19 +183,10 @@
             {
                 if (this.field_175503_c % 4 == 0)
                 {
-                    this.field_70170_p
-                    .func_175688_a(
-                        EnumParticleTypes.VILLAGER_HAPPY,
-                        this.field_70165_t + (double)(this.field_70146_Z.nextFloat() * this.field_70130_N * 2.0F) - (double)this.field_70130_N,
-                        this.field_70163_u + 0.5 + (double)(this.field_70146_Z.nextFloat() * this.field_70131_O),
-                        this.field_70161_v + (double)(this.field_70146_Z.nextFloat() * this.field_70130_N * 2.0F) - (double)this.field_70130_N,
-                        0.0,
-                        0.0,
-                        0.0
-                    );
+                    this.field_70170_p.func_175688_a(EnumParticleTypes.VILLAGER_HAPPY, this.field_70165_t + (double)(this.field_70146_Z.nextFloat() * this.field_70130_N * 2.0F) - (double)this.field_70130_N, this.field_70163_u + 0.5D + (double)(this.field_70146_Z.nextFloat() * this.field_70131_O), this.field_70161_v + (double)(this.field_70146_Z.nextFloat() * this.field_70130_N * 2.0F) - (double)this.field_70130_N, 0.0D, 0.0D, 0.0D);
                 }
 
-                this.field_175503_c--;
+                --this.field_175503_c;
             }
         }
         else
@@ -209,7 +195,8 @@
 
             if (i < 0)
             {
-                this.func_70873_a(++i);
+                ++i;
+                this.func_70873_a(i);
 
                 if (i == 0)
                 {
@@ -218,7 +205,8 @@
             }
             else if (i > 0)
             {
-                this.func_70873_a(--i);
+                --i;
+                this.func_70873_a(i);
             }
         }
     }
@@ -227,7 +215,6 @@
     {
     }
 
-    @Override
     public boolean func_70631_g_()
     {
         return this.func_70874_b() < 0;
@@ -238,7 +225,6 @@
         this.func_98055_j(p_98054_1_ ? 0.5F : 1.0F);
     }
 
-    @Override
     protected final void func_70105_a(float p_70105_1_, float p_70105_2_)
     {
         boolean flag = this.field_98056_d > 0.0F;
