--- before/net/minecraft/entity/EntityLiving.java
+++ after/net/minecraft/entity/EntityLiving.java
@@ -60,7 +60,7 @@
 
 public abstract class EntityLiving extends EntityLivingBase
 {
-    private static final DataParameter<Byte> field_184654_a = EntityDataManager.func_187226_a(EntityLiving.class, DataSerializers.field_187191_a);
+    private static final DataParameter<Byte> field_184654_a = EntityDataManager.<Byte>func_187226_a(EntityLiving.class, DataSerializers.field_187191_a);
     public int field_70757_a;
     protected int field_70728_aV;
     private final EntityLookHelper field_70749_g;
@@ -72,9 +72,9 @@
     public final EntityAITasks field_70715_bh;
     private EntityLivingBase field_70696_bz;
     private final EntitySenses field_70723_bA;
-    private final NonNullList<ItemStack> field_184656_bv = NonNullList.func_191197_a(2, ItemStack.field_190927_a);
+    private final NonNullList<ItemStack> field_184656_bv = NonNullList.<ItemStack>func_191197_a(2, ItemStack.field_190927_a);
     protected float[] field_82174_bp = new float[2];
-    private final NonNullList<ItemStack> field_184657_bw = NonNullList.func_191197_a(4, ItemStack.field_190927_a);
+    private final NonNullList<ItemStack> field_184657_bw = NonNullList.<ItemStack>func_191197_a(4, ItemStack.field_190927_a);
     protected float[] field_184655_bs = new float[4];
     private boolean field_82172_bs;
     private boolean field_82179_bU;
@@ -123,12 +123,12 @@
     public float func_184643_a(PathNodeType p_184643_1_)
     {
         Float f = this.field_184658_bz.get(p_184643_1_);
-        return f == null ? p_184643_1_.func_186289_a() : f;
+        return f == null ? p_184643_1_.func_186289_a() : f.floatValue();
     }
 
     public void func_184644_a(PathNodeType p_184644_1_, float p_184644_2_)
     {
-        this.field_184658_bz.put(p_184644_1_, p_184644_2_);
+        this.field_184658_bz.put(p_184644_1_, Float.valueOf(p_184644_2_));
     }
 
     protected EntityBodyHelper func_184650_s()
@@ -170,6 +170,7 @@
     public void func_70624_b(@Nullable EntityLivingBase p_70624_1_)
     {
         this.field_70696_bz = p_70624_1_;
+        net.minecraftforge.common.ForgeHooks.onLivingSetAttackTarget(this, p_70624_1_);
     }
 
     public boolean func_70686_a(Class <? extends EntityLivingBase > p_70686_1_)
@@ -184,7 +185,7 @@
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184654_a, (byte)0);
+        this.field_70180_af.func_187214_a(field_184654_a, Byte.valueOf((byte)0));
     }
 
     public int func_70627_aG()
@@ -235,7 +236,7 @@
 
             for (int j = 0; j < this.field_184657_bw.size(); ++j)
             {
-                if (!this.field_184657_bw.get(j).func_190926_b() && this.field_184655_bs[j] <= 1.0F)
+                if (!((ItemStack)this.field_184657_bw.get(j)).func_190926_b() && this.field_184655_bs[j] <= 1.0F)
                 {
                     i += 1 + this.field_70146_Z.nextInt(3);
                 }
@@ -243,7 +244,7 @@
 
             for (int k = 0; k < this.field_184656_bv.size(); ++k)
             {
-                if (!this.field_184656_bv.get(k).func_190926_b() && this.field_82174_bp[k] <= 1.0F)
+                if (!((ItemStack)this.field_184656_bv.get(k)).func_190926_b() && this.field_82174_bp[k] <= 1.0F)
                 {
                     i += 1 + this.field_70146_Z.nextInt(3);
                 }
@@ -348,7 +349,7 @@
 
     public static void func_189752_a(DataFixer p_189752_0_, Class<?> p_189752_1_)
     {
-        p_189752_0_.func_188258_a(FixTypes.ENTITY, new ItemStackDataLists(p_189752_1_, "ArmorItems", "HandItems"));
+        p_189752_0_.func_188258_a(FixTypes.ENTITY, new ItemStackDataLists(p_189752_1_, new String[] {"ArmorItems", "HandItems"}));
     }
 
     public void func_70014_b(NBTTagCompound p_70014_1_)
@@ -576,7 +577,7 @@
         super.func_70636_d();
         this.field_70170_p.field_72984_F.func_76320_a("looting");
 
-        if (!this.field_70170_p.field_72995_K && this.func_98052_bS() && !this.field_70729_aU && this.field_70170_p.func_82736_K().func_82766_b("mobGriefing"))
+        if (!this.field_70170_p.field_72995_K && this.func_98052_bS() && !this.field_70729_aU && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_70170_p, this))
         {
             for (EntityItem entityitem : this.field_70170_p.func_72872_a(EntityItem.class, this.func_174813_aQ().func_72314_b(1.0D, 0.0D, 1.0D)))
             {
@@ -702,10 +703,22 @@
 
     protected void func_70623_bb()
     {
+        net.minecraftforge.fml.common.eventhandler.Event.Result result = null;
         if (this.field_82179_bU)
         {
             this.field_70708_bq = 0;
         }
+        else if ((this.field_70708_bq & 0x1F) == 0x1F && (result = net.minecraftforge.event.ForgeEventFactory.canEntityDespawn(this)) != net.minecraftforge.fml.common.eventhandler.Event.Result.DEFAULT)
+        {
+            if (result == net.minecraftforge.fml.common.eventhandler.Event.Result.DENY)
+            {
+                this.field_70708_bq = 0;
+            }
+            else
+            {
+                this.func_70106_y();
+            }
+        }
         else
         {
             Entity entity = this.field_70170_p.func_72890_a(this, -1.0D);
@@ -805,8 +818,8 @@
         }
 
         double d3 = (double)MathHelper.func_76133_a(d0 * d0 + d2 * d2);
-        float f = (float)(MathHelper.func_181159_b(d2, d0) * (double)(180F / (float)Math.PI)) - 90.0F;
-        float f1 = (float)(-(MathHelper.func_181159_b(d1, d3) * (double)(180F / (float)Math.PI)));
+        float f = (float)(MathHelper.func_181159_b(d2, d0) * (180D / Math.PI)) - 90.0F;
+        float f1 = (float)(-(MathHelper.func_181159_b(d1, d3) * (180D / Math.PI)));
         this.field_70125_A = this.func_70663_b(this.field_70125_A, f1, p_70625_3_);
         this.field_70177_z = this.func_70663_b(this.field_70177_z, f, p_70625_2_);
     }
@@ -839,7 +852,6 @@
         return !this.field_70170_p.func_72953_d(this.func_174813_aQ()) && this.field_70170_p.func_184144_a(this, this.func_174813_aQ()).isEmpty() && this.field_70170_p.func_72917_a(this.func_174813_aQ(), this);
     }
 
-    @SideOnly(Side.CLIENT)
     public float func_70603_bj()
     {
         return 1.0F;
@@ -859,7 +871,7 @@
         else
         {
             int i = (int)(this.func_110143_aJ() - this.func_110138_aP() * 0.33F);
-            i -= (3 - this.field_70170_p.func_175659_aa().func_151525_a()) * 4;
+            i = i - (3 - this.field_70170_p.func_175659_aa().func_151525_a()) * 4;
 
             if (i < 0)
             {
@@ -991,6 +1003,8 @@
 
     public static EntityEquipmentSlot func_184640_d(ItemStack p_184640_0_)
     {
+        final EntityEquipmentSlot slot = p_184640_0_.func_77973_b().getEquipmentSlot(p_184640_0_);
+        if (slot != null) return slot; // FORGE: Allow modders to set a non-default equipment slot for a stack; e.g. a non-armor chestplate-slot item
         if (p_184640_0_.func_77973_b() != Item.func_150898_a(Blocks.field_150423_aK) && p_184640_0_.func_77973_b() != Items.field_151144_bL)
         {
             if (p_184640_0_.func_77973_b() instanceof ItemArmor)
@@ -1003,7 +1017,7 @@
             }
             else
             {
-                return p_184640_0_.func_77973_b() == Items.field_185159_cQ ? EntityEquipmentSlot.OFFHAND : EntityEquipmentSlot.MAINHAND;
+                return p_184640_0_.func_77973_b().isShield(p_184640_0_, null) ? EntityEquipmentSlot.OFFHAND : EntityEquipmentSlot.MAINHAND;
             }
         }
         else
@@ -1404,24 +1418,24 @@
 
     public void func_94061_f(boolean p_94061_1_)
     {
-        byte b0 = this.field_70180_af.func_187225_a(field_184654_a);
-        this.field_70180_af.func_187227_b(field_184654_a, p_94061_1_ ? (byte)(b0 | 1) : (byte)(b0 & -2));
+        byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184654_a)).byteValue();
+        this.field_70180_af.func_187227_b(field_184654_a, Byte.valueOf(p_94061_1_ ? (byte)(b0 | 1) : (byte)(b0 & -2)));
     }
 
     public void func_184641_n(boolean p_184641_1_)
     {
-        byte b0 = this.field_70180_af.func_187225_a(field_184654_a);
-        this.field_70180_af.func_187227_b(field_184654_a, p_184641_1_ ? (byte)(b0 | 2) : (byte)(b0 & -3));
+        byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184654_a)).byteValue();
+        this.field_70180_af.func_187227_b(field_184654_a, Byte.valueOf(p_184641_1_ ? (byte)(b0 | 2) : (byte)(b0 & -3)));
     }
 
     public boolean func_175446_cd()
     {
-        return (this.field_70180_af.func_187225_a(field_184654_a) & 1) != 0;
+        return (((Byte)this.field_70180_af.func_187225_a(field_184654_a)).byteValue() & 1) != 0;
     }
 
     public boolean func_184638_cS()
     {
-        return (this.field_70180_af.func_187225_a(field_184654_a) & 2) != 0;
+        return (((Byte)this.field_70180_af.func_187225_a(field_184654_a)).byteValue() & 2) != 0;
     }
 
     public EnumHandSide func_184591_cq()
@@ -1434,5 +1448,19 @@
         ON_GROUND,
         IN_AIR,
         IN_WATER;
+
+        private final java.util.function.BiPredicate<net.minecraft.world.IBlockAccess, BlockPos> spawnPredicate;
+
+        SpawnPlacementType() { this.spawnPredicate = null; }
+
+        SpawnPlacementType(java.util.function.BiPredicate<net.minecraft.world.IBlockAccess, BlockPos> spawnPredicate)
+        {
+            this.spawnPredicate = spawnPredicate;
+        }
+
+        public boolean canSpawnAt(World world, BlockPos pos)
+        {
+            return this.spawnPredicate != null ? this.spawnPredicate.test(world, pos) : net.minecraft.world.WorldEntitySpawner.canCreatureTypeSpawnBody(this, world, pos);
+        }
     }
 }
