--- before/net/minecraft/entity/EntityLeashKnot.java
+++ after/net/minecraft/entity/EntityLeashKnot.java
@@ -25,94 +25,70 @@
     public EntityLeashKnot(World p_i45851_1_, BlockPos p_i45851_2_)
     {
         super(p_i45851_1_, p_i45851_2_);
-        this.func_70107_b((double)p_i45851_2_.func_177958_n() + 0.5, (double)p_i45851_2_.func_177956_o() + 0.5, (double)p_i45851_2_.func_177952_p() + 0.5);
+        this.func_70107_b((double)p_i45851_2_.func_177958_n() + 0.5D, (double)p_i45851_2_.func_177956_o() + 0.5D, (double)p_i45851_2_.func_177952_p() + 0.5D);
         float f = 0.125F;
         float f1 = 0.1875F;
         float f2 = 0.25F;
-        this.func_174826_a(
-            new AxisAlignedBB(
-                this.field_70165_t - 0.1875,
-                this.field_70163_u - 0.25 + 0.125,
-                this.field_70161_v - 0.1875,
-                this.field_70165_t + 0.1875,
-                this.field_70163_u + 0.25 + 0.125,
-                this.field_70161_v + 0.1875
-            )
-        );
+        this.func_174826_a(new AxisAlignedBB(this.field_70165_t - 0.1875D, this.field_70163_u - 0.25D + 0.125D, this.field_70161_v - 0.1875D, this.field_70165_t + 0.1875D, this.field_70163_u + 0.25D + 0.125D, this.field_70161_v + 0.1875D));
         this.field_98038_p = true;
     }
 
-    @Override
     public void func_70107_b(double p_70107_1_, double p_70107_3_, double p_70107_5_)
     {
-        super.func_70107_b(
-            (double)MathHelper.func_76128_c(p_70107_1_) + 0.5,
-            (double)MathHelper.func_76128_c(p_70107_3_) + 0.5,
-            (double)MathHelper.func_76128_c(p_70107_5_) + 0.5
-        );
+        super.func_70107_b((double)MathHelper.func_76128_c(p_70107_1_) + 0.5D, (double)MathHelper.func_76128_c(p_70107_3_) + 0.5D, (double)MathHelper.func_76128_c(p_70107_5_) + 0.5D);
     }
 
-    @Override
     protected void func_174856_o()
     {
-        this.field_70165_t = (double)this.field_174861_a.func_177958_n() + 0.5;
-        this.field_70163_u = (double)this.field_174861_a.func_177956_o() + 0.5;
-        this.field_70161_v = (double)this.field_174861_a.func_177952_p() + 0.5;
+        this.field_70165_t = (double)this.field_174861_a.func_177958_n() + 0.5D;
+        this.field_70163_u = (double)this.field_174861_a.func_177956_o() + 0.5D;
+        this.field_70161_v = (double)this.field_174861_a.func_177952_p() + 0.5D;
+        if (this.isAddedToWorld() && !this.field_70170_p.field_72995_K) this.field_70170_p.func_72866_a(this, false); // Forge - Process chunk registration after moving.
     }
 
-    @Override
     public void func_174859_a(EnumFacing p_174859_1_)
     {
     }
 
-    @Override
     public int func_82329_d()
     {
         return 9;
     }
 
-    @Override
     public int func_82330_g()
     {
         return 9;
     }
 
-    @Override
     public float func_70047_e()
     {
         return -0.0625F;
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public boolean func_70112_a(double p_70112_1_)
     {
-        return p_70112_1_ < 1024.0;
+        return p_70112_1_ < 1024.0D;
     }
 
-    @Override
     public void func_110128_b(@Nullable Entity p_110128_1_)
     {
         this.func_184185_a(SoundEvents.field_187746_da, 1.0F, 1.0F);
     }
 
-    @Override
     public boolean func_70039_c(NBTTagCompound p_70039_1_)
     {
         return false;
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
     }
 
-    @Override
     public boolean func_184230_a(EntityPlayer p_184230_1_, EnumHand p_184230_2_)
     {
         if (this.field_70170_p.field_72995_K)
@@ -122,19 +98,8 @@
         else
         {
             boolean flag = false;
-            double d0 = 7.0;
-            List<EntityLiving> list = this.field_70170_p
-                                      .func_72872_a(
-                                          EntityLiving.class,
-                                          new AxisAlignedBB(
-                                              this.field_70165_t - 7.0,
-                                              this.field_70163_u - 7.0,
-                                              this.field_70161_v - 7.0,
-                                              this.field_70165_t + 7.0,
-                                              this.field_70163_u + 7.0,
-                                              this.field_70161_v + 7.0
-                                          )
-                                      );
+            double d0 = 7.0D;
+            List<EntityLiving> list = this.field_70170_p.<EntityLiving>func_72872_a(EntityLiving.class, new AxisAlignedBB(this.field_70165_t - 7.0D, this.field_70163_u - 7.0D, this.field_70161_v - 7.0D, this.field_70165_t + 7.0D, this.field_70163_u + 7.0D, this.field_70161_v + 7.0D));
 
             for (EntityLiving entityliving : list)
             {
@@ -165,7 +130,6 @@
         }
     }
 
-    @Override
     public boolean func_70518_d()
     {
         return this.field_70170_p.func_180495_p(this.field_174861_a).func_177230_c() instanceof BlockFence;
@@ -186,9 +150,7 @@
         int j = p_174863_1_.func_177956_o();
         int k = p_174863_1_.func_177952_p();
 
-        for (EntityLeashKnot entityleashknot : p_174863_0_.func_72872_a(
-                    EntityLeashKnot.class, new AxisAlignedBB((double)i - 1.0, (double)j - 1.0, (double)k - 1.0, (double)i + 1.0, (double)j + 1.0, (double)k + 1.0)
-                ))
+        for (EntityLeashKnot entityleashknot : p_174863_0_.func_72872_a(EntityLeashKnot.class, new AxisAlignedBB((double)i - 1.0D, (double)j - 1.0D, (double)k - 1.0D, (double)i + 1.0D, (double)j + 1.0D, (double)k + 1.0D)))
         {
             if (entityleashknot.func_174857_n().equals(p_174863_1_))
             {
@@ -199,7 +161,6 @@
         return null;
     }
 
-    @Override
     public void func_184523_o()
     {
         this.func_184185_a(SoundEvents.field_187748_db, 1.0F, 1.0F);
