--- before/net/minecraft/entity/EntitySpawnPlacementRegistry.java
+++ after/net/minecraft/entity/EntitySpawnPlacementRegistry.java
@@ -46,11 +46,16 @@
 
 public class EntitySpawnPlacementRegistry
 {
-    private static final Map < Class<?>, EntityLiving.SpawnPlacementType > field_180110_a = Maps.newHashMap();
+    private static final Map < Class<?>, EntityLiving.SpawnPlacementType > field_180110_a = Maps. < Class<?>, EntityLiving.SpawnPlacementType > newHashMap();
 
     public static EntityLiving.SpawnPlacementType func_180109_a(Class<?> p_180109_0_)
     {
-        return field_180110_a.get(p_180109_0_);
+        return field_180110_a.getOrDefault(p_180109_0_, EntityLiving.SpawnPlacementType.ON_GROUND);
+    }
+
+    public static void setPlacementType(Class<? extends Entity> entityClass, EntityLiving.SpawnPlacementType placementType)
+    {
+        field_180110_a.putIfAbsent(entityClass, placementType);
     }
 
     static
