--- before/net/minecraft/entity/EntityHanging.java
+++ after/net/minecraft/entity/EntityHanging.java
@@ -45,7 +45,6 @@
         this.field_174861_a = p_i45853_2_;
     }
 
-    @Override
     protected void func_70088_a()
     {
     }
@@ -64,18 +63,18 @@
     {
         if (this.field_174860_b != null)
         {
-            double d0 = (double)this.field_174861_a.func_177958_n() + 0.5;
-            double d1 = (double)this.field_174861_a.func_177956_o() + 0.5;
-            double d2 = (double)this.field_174861_a.func_177952_p() + 0.5;
-            double d3 = 0.46875;
+            double d0 = (double)this.field_174861_a.func_177958_n() + 0.5D;
+            double d1 = (double)this.field_174861_a.func_177956_o() + 0.5D;
+            double d2 = (double)this.field_174861_a.func_177952_p() + 0.5D;
+            double d3 = 0.46875D;
             double d4 = this.func_190202_a(this.func_82329_d());
             double d5 = this.func_190202_a(this.func_82330_g());
-            d0 -= (double)this.field_174860_b.func_82601_c() * 0.46875;
-            d2 -= (double)this.field_174860_b.func_82599_e() * 0.46875;
-            d1 += d5;
+            d0 = d0 - (double)this.field_174860_b.func_82601_c() * 0.46875D;
+            d2 = d2 - (double)this.field_174860_b.func_82599_e() * 0.46875D;
+            d1 = d1 + d5;
             EnumFacing enumfacing = this.field_174860_b.func_176735_f();
-            d0 += d4 * (double)enumfacing.func_82601_c();
-            d2 += d4 * (double)enumfacing.func_82599_e();
+            d0 = d0 + d4 * (double)enumfacing.func_82601_c();
+            d2 = d2 + d4 * (double)enumfacing.func_82599_e();
             this.field_70165_t = d0;
             this.field_70163_u = d1;
             this.field_70161_v = d2;
@@ -85,26 +84,25 @@
 
             if (this.field_174860_b.func_176740_k() == EnumFacing.Axis.Z)
             {
-                d8 = 1.0;
+                d8 = 1.0D;
             }
             else
             {
-                d6 = 1.0;
+                d6 = 1.0D;
             }
 
-            d6 /= 32.0;
-            d7 /= 32.0;
-            d8 /= 32.0;
+            d6 = d6 / 32.0D;
+            d7 = d7 / 32.0D;
+            d8 = d8 / 32.0D;
             this.func_174826_a(new AxisAlignedBB(d0 - d6, d1 - d7, d2 - d8, d0 + d6, d1 + d7, d2 + d8));
         }
     }
 
     private double func_190202_a(int p_190202_1_)
     {
-        return p_190202_1_ % 32 == 0 ? 0.5 : 0.0;
+        return p_190202_1_ % 32 == 0 ? 0.5D : 0.0D;
     }
 
-    @Override
     public void func_70071_h_()
     {
         this.field_70169_q = this.field_70165_t;
@@ -118,7 +116,7 @@
             if (!this.field_70128_L && !this.func_70518_d())
             {
                 this.func_70106_y();
-                this.func_110128_b(null);
+                this.func_110128_b((Entity)null);
             }
         }
     }
@@ -146,6 +144,9 @@
                     blockpos$mutableblockpos.func_189533_g(blockpos).func_189534_c(enumfacing, k + i1).func_189534_c(EnumFacing.UP, l + j1);
                     IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos$mutableblockpos);
 
+                    if (iblockstate.isSideSolid(this.field_70170_p, blockpos$mutableblockpos, this.field_174860_b))
+                        continue;
+
                     if (!iblockstate.func_185904_a().func_76220_a() && !BlockRedstoneDiode.func_185546_B(iblockstate))
                     {
                         return false;
@@ -157,25 +158,21 @@
         }
     }
 
-    @Override
     public boolean func_70067_L()
     {
         return true;
     }
 
-    @Override
     public boolean func_85031_j(Entity p_85031_1_)
     {
         return p_85031_1_ instanceof EntityPlayer ? this.func_70097_a(DamageSource.func_76365_a((EntityPlayer)p_85031_1_), 0.0F) : false;
     }
 
-    @Override
     public EnumFacing func_174811_aO()
     {
         return this.field_174860_b;
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         if (this.func_180431_b(p_70097_1_))
@@ -195,27 +192,24 @@
         }
     }
 
-    @Override
     public void func_70091_d(MoverType p_70091_1_, double p_70091_2_, double p_70091_4_, double p_70091_6_)
     {
-        if (!this.field_70170_p.field_72995_K && !this.field_70128_L && p_70091_2_ * p_70091_2_ + p_70091_4_ * p_70091_4_ + p_70091_6_ * p_70091_6_ > 0.0)
+        if (!this.field_70170_p.field_72995_K && !this.field_70128_L && p_70091_2_ * p_70091_2_ + p_70091_4_ * p_70091_4_ + p_70091_6_ * p_70091_6_ > 0.0D)
         {
             this.func_70106_y();
-            this.func_110128_b(null);
+            this.func_110128_b((Entity)null);
         }
     }
 
-    @Override
     public void func_70024_g(double p_70024_1_, double p_70024_3_, double p_70024_5_)
     {
-        if (!this.field_70170_p.field_72995_K && !this.field_70128_L && p_70024_1_ * p_70024_1_ + p_70024_3_ * p_70024_3_ + p_70024_5_ * p_70024_5_ > 0.0)
+        if (!this.field_70170_p.field_72995_K && !this.field_70128_L && p_70024_1_ * p_70024_1_ + p_70024_3_ * p_70024_3_ + p_70024_5_ * p_70024_5_ > 0.0D)
         {
             this.func_70106_y();
-            this.func_110128_b(null);
+            this.func_110128_b((Entity)null);
         }
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         p_70014_1_.func_74774_a("Facing", (byte)this.field_174860_b.func_176736_b());
@@ -225,7 +219,6 @@
         p_70014_1_.func_74768_a("TileZ", blockpos.func_177952_p());
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         this.field_174861_a = new BlockPos(p_70037_1_.func_74762_e("TileX"), p_70037_1_.func_74762_e("TileY"), p_70037_1_.func_74762_e("TileZ"));
@@ -240,28 +233,19 @@
 
     public abstract void func_184523_o();
 
-    @Override
     public EntityItem func_70099_a(ItemStack p_70099_1_, float p_70099_2_)
     {
-        EntityItem entityitem = new EntityItem(
-            this.field_70170_p,
-            this.field_70165_t + (double)((float)this.field_174860_b.func_82601_c() * 0.15F),
-            this.field_70163_u + (double)p_70099_2_,
-            this.field_70161_v + (double)((float)this.field_174860_b.func_82599_e() * 0.15F),
-            p_70099_1_
-        );
+        EntityItem entityitem = new EntityItem(this.field_70170_p, this.field_70165_t + (double)((float)this.field_174860_b.func_82601_c() * 0.15F), this.field_70163_u + (double)p_70099_2_, this.field_70161_v + (double)((float)this.field_174860_b.func_82599_e() * 0.15F), p_70099_1_);
         entityitem.func_174869_p();
         this.field_70170_p.func_72838_d(entityitem);
         return entityitem;
     }
 
-    @Override
     protected boolean func_142008_O()
     {
         return false;
     }
 
-    @Override
     public void func_70107_b(double p_70107_1_, double p_70107_3_, double p_70107_5_)
     {
         this.field_174861_a = new BlockPos(p_70107_1_, p_70107_3_, p_70107_5_);
@@ -275,7 +259,6 @@
     }
 
     @SuppressWarnings("incomplete-switch")
-    @Override
     public float func_184229_a(Rotation p_184229_1_)
     {
         if (this.field_174860_b != null && this.field_174860_b.func_176740_k() != EnumFacing.Axis.Y)
@@ -308,13 +291,11 @@
         }
     }
 
-    @Override
     public float func_184217_a(Mirror p_184217_1_)
     {
         return this.func_184229_a(p_184217_1_.func_185800_a(this.field_174860_b));
     }
 
-    @Override
     public void func_70077_a(EntityLightningBolt p_70077_1_)
     {
     }
