--- before/net/minecraft/entity/EntityLiving.java
+++ after/net/minecraft/entity/EntityLiving.java
@@ -171,6 +171,7 @@
     public void setAttackTarget(@Nullable EntityLivingBase entitylivingbaseIn)
     {
         this.attackTarget = entitylivingbaseIn;
+        this.attackTarget = net.minecraftforge.common.ForgeHooks.onLivingSetAttackTarget(this, entitylivingbaseIn);
     }
 
     public boolean canAttackClass(Class <? extends EntityLivingBase > cls)
@@ -601,7 +602,7 @@
         super.onLivingUpdate();
         this.world.profiler.startSection("looting");
 
-        if (!this.world.isRemote && this.canPickUpLoot() && !this.dead && this.world.getGameRules().getBoolean("mobGriefing"))
+        if (!this.world.isRemote && this.canPickUpLoot() && !this.dead && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this))
         {
             for (EntityItem entityitem : this.world.getEntitiesWithinAABB(EntityItem.class, this.getEntityBoundingBox().grow(1.0, 0.0, 1.0)))
             {
@@ -729,10 +730,22 @@
 
     protected void despawnEntity()
     {
+        net.minecraftforge.fml.common.eventhandler.Event.Result result = null;
         if (this.persistenceRequired)
         {
             this.idleTime = 0;
         }
+        else if ((this.idleTime & 0x1F) == 0x1F && (result = net.minecraftforge.event.ForgeEventFactory.canEntityDespawn(this)) != net.minecraftforge.fml.common.eventhandler.Event.Result.DEFAULT)
+        {
+            if (result == net.minecraftforge.fml.common.eventhandler.Event.Result.DENY)
+            {
+                this.idleTime = 0;
+            }
+            else
+            {
+                this.setDead();
+            }
+        }
         else
         {
             Entity entity = this.world.getClosestPlayerToEntity(this, -1.0);
@@ -869,7 +882,6 @@
                && this.world.checkNoEntityCollision(this.getEntityBoundingBox(), this);
     }
 
-    @SideOnly(Side.CLIENT)
     public float getRenderSizeModifier()
     {
         return 1.0F;
@@ -1032,6 +1044,8 @@
 
     public static EntityEquipmentSlot getSlotForItemStack(ItemStack stack)
     {
+        final EntityEquipmentSlot slot = stack.getItem().getEquipmentSlot(stack);
+        if (slot != null) return slot; // FORGE: Allow modders to set a non-default equipment slot for a stack; e.g. a non-armor chestplate-slot item
         if (stack.getItem() == Item.getItemFromBlock(Blocks.PUMPKIN) || stack.getItem() == Items.SKULL)
         {
             return EntityEquipmentSlot.HEAD;
@@ -1046,7 +1060,7 @@
         }
         else
         {
-            return stack.getItem() == Items.SHIELD ? EntityEquipmentSlot.OFFHAND : EntityEquipmentSlot.MAINHAND;
+            return stack.getItem().isShield(stack, null) ? EntityEquipmentSlot.OFFHAND : EntityEquipmentSlot.MAINHAND;
         }
     }
 
@@ -1488,5 +1502,19 @@
         ON_GROUND,
         IN_AIR,
         IN_WATER;
+
+        private final java.util.function.BiPredicate<net.minecraft.world.IBlockAccess, BlockPos> spawnPredicate;
+
+        SpawnPlacementType() { this.spawnPredicate = null; }
+
+        SpawnPlacementType(java.util.function.BiPredicate<net.minecraft.world.IBlockAccess, BlockPos> spawnPredicate)
+        {
+            this.spawnPredicate = spawnPredicate;
+        }
+
+        public boolean canSpawnAt(World world, BlockPos pos)
+        {
+            return this.spawnPredicate != null ? this.spawnPredicate.test(world, pos) : net.minecraft.world.WorldEntitySpawner.canCreatureTypeSpawnBody(this, world, pos);
+        }
     }
 }
