--- before/net/minecraft/entity/EntityHanging.java
+++ after/net/minecraft/entity/EntityHanging.java
@@ -146,6 +146,9 @@
                     blockpos$mutableblockpos.setPos(blockpos).move(enumfacing, k + i1).move(EnumFacing.UP, l + j1);
                     IBlockState iblockstate = this.world.getBlockState(blockpos$mutableblockpos);
 
+                    if (iblockstate.isSideSolid(this.world, blockpos$mutableblockpos, this.facingDirection))
+                        continue;
+
                     if (!iblockstate.getMaterial().isSolid() && !BlockRedstoneDiode.isDiode(iblockstate))
                     {
                         return false;
