--- before/net/minecraft/entity/EntityTrackerEntry.java
+++ after/net/minecraft/entity/EntityTrackerEntry.java
@@ -90,11 +90,11 @@
     private boolean field_73144_s;
     private final boolean field_73143_t;
     private int field_73142_u;
-    private List<Entity> field_187263_w = Collections.emptyList();
+    private List<Entity> field_187263_w = Collections.<Entity>emptyList();
     private boolean field_73141_v;
     private boolean field_180234_y;
     public boolean field_73133_n;
-    public final Set<EntityPlayerMP> field_73134_o = Sets.newHashSet();
+    public final Set<EntityPlayerMP> field_73134_o = Sets.<EntityPlayerMP>newHashSet();
 
     public EntityTrackerEntry(Entity p_i46837_1_, int p_i46837_2_, int p_i46837_3_, int p_i46837_4_, boolean p_i46837_5_)
     {
@@ -158,13 +158,13 @@
 
             if (itemstack.func_77973_b() instanceof ItemMap)
             {
-                MapData mapdata = Items.field_151098_aY.func_77873_a(itemstack, this.field_73132_a.field_70170_p);
+                MapData mapdata = ((ItemMap) itemstack.func_77973_b()).func_77873_a(itemstack, this.field_73132_a.field_70170_p);
 
                 for (EntityPlayer entityplayer : p_73122_1_)
                 {
                     EntityPlayerMP entityplayermp = (EntityPlayerMP)entityplayer;
                     mapdata.func_76191_a(entityplayermp, itemstack);
-                    Packet<?> packet = Items.field_151098_aY.func_150911_c(itemstack, this.field_73132_a.field_70170_p, entityplayermp);
+                    Packet<?> packet = ((ItemMap) itemstack.func_77973_b()).func_150911_c(itemstack, this.field_73132_a.field_70170_p, entityplayermp);
 
                     if (packet != null)
                     {
@@ -457,6 +457,7 @@
 
                     this.field_73132_a.func_184178_b(p_73117_1_);
                     p_73117_1_.func_184848_d(this.field_73132_a);
+                    net.minecraftforge.event.ForgeEventFactory.onStartEntityTracking(field_73132_a, p_73117_1_);
                 }
             }
             else if (this.field_73134_o.contains(p_73117_1_))
@@ -464,6 +465,7 @@
                 this.field_73134_o.remove(p_73117_1_);
                 this.field_73132_a.func_184203_c(p_73117_1_);
                 p_73117_1_.func_152339_d(this.field_73132_a);
+                net.minecraftforge.event.ForgeEventFactory.onStopEntityTracking(field_73132_a, p_73117_1_);
             }
         }
     }
@@ -495,6 +497,9 @@
         {
             field_151262_p.warn("Fetching addPacket for removed entity");
         }
+
+        Packet pkt = net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.getEntitySpawningPacket(this.field_73132_a);
+        if (pkt != null) return pkt;
 
         if (this.field_73132_a instanceof EntityPlayerMP)
         {
