--- before/net/minecraft/entity/EntityTrackerEntry.java
+++ after/net/minecraft/entity/EntityTrackerEntry.java
@@ -90,11 +90,11 @@
     private boolean field_73144_s;
     private final boolean field_73143_t;
     private int field_73142_u;
-    private List<Entity> field_187263_w = Collections.emptyList();
+    private List<Entity> field_187263_w = Collections.<Entity>emptyList();
     private boolean field_73141_v;
     private boolean field_180234_y;
     public boolean field_73133_n;
-    public final Set<EntityPlayerMP> field_73134_o = Sets.newHashSet();
+    public final Set<EntityPlayerMP> field_73134_o = Sets.<EntityPlayerMP>newHashSet();
 
     public EntityTrackerEntry(Entity p_i46837_1_, int p_i46837_2_, int p_i46837_3_, int p_i46837_4_, boolean p_i46837_5_)
     {
@@ -112,15 +112,18 @@
         this.field_180234_y = p_i46837_1_.field_70122_E;
     }
 
-    @Override
     public boolean equals(Object p_equals_1_)
     {
-        return p_equals_1_ instanceof EntityTrackerEntry
-               ? ((EntityTrackerEntry)p_equals_1_).field_73132_a.func_145782_y() == this.field_73132_a.func_145782_y()
-               : false;
+        if (p_equals_1_ instanceof EntityTrackerEntry)
+        {
+            return ((EntityTrackerEntry)p_equals_1_).field_73132_a.func_145782_y() == this.field_73132_a.func_145782_y();
+        }
+        else
+        {
+            return false;
+        }
     }
 
-    @Override
     public int hashCode()
     {
         return this.field_73132_a.func_145782_y();
@@ -130,7 +133,7 @@
     {
         this.field_73133_n = false;
 
-        if (!this.field_73144_s || this.field_73132_a.func_70092_e(this.field_73147_p, this.field_73146_q, this.field_73145_r) > 16.0)
+        if (!this.field_73144_s || this.field_73132_a.func_70092_e(this.field_73147_p, this.field_73146_q, this.field_73145_r) > 16.0D)
         {
             this.field_73147_p = this.field_73132_a.field_70165_t;
             this.field_73146_q = this.field_73132_a.field_70163_u;
@@ -155,13 +158,13 @@
 
             if (itemstack.func_77973_b() instanceof ItemMap)
             {
-                MapData mapdata = Items.field_151098_aY.func_77873_a(itemstack, this.field_73132_a.field_70170_p);
+                MapData mapdata = ((ItemMap) itemstack.func_77973_b()).func_77873_a(itemstack, this.field_73132_a.field_70170_p);
 
                 for (EntityPlayer entityplayer : p_73122_1_)
                 {
                     EntityPlayerMP entityplayermp = (EntityPlayerMP)entityplayer;
                     mapdata.func_76191_a(entityplayermp, itemstack);
-                    Packet<?> packet = Items.field_151098_aY.func_150911_c(itemstack, this.field_73132_a.field_70170_p, entityplayermp);
+                    Packet<?> packet = ((ItemMap) itemstack.func_77973_b()).func_150911_c(itemstack, this.field_73132_a.field_70170_p, entityplayermp);
 
                     if (packet != null)
                     {
@@ -183,9 +186,7 @@
 
                 if (flag3)
                 {
-                    this.func_151259_a(
-                        new SPacketEntity.S16PacketEntityLook(this.field_73132_a.func_145782_y(), (byte)j1, (byte)l1, this.field_73132_a.field_70122_E)
-                    );
+                    this.func_151259_a(new SPacketEntity.S16PacketEntityLook(this.field_73132_a.func_145782_y(), (byte)j1, (byte)l1, this.field_73132_a.field_70122_E));
                     this.field_73127_g = j1;
                     this.field_73139_h = l1;
                 }
@@ -198,7 +199,7 @@
             }
             else
             {
-                this.field_73142_u++;
+                ++this.field_73142_u;
                 long i1 = EntityTracker.func_187253_a(this.field_73132_a.field_70165_t);
                 long i2 = EntityTracker.func_187253_a(this.field_73132_a.field_70163_u);
                 long j2 = EntityTracker.func_187253_a(this.field_73132_a.field_70161_v);
@@ -213,15 +214,7 @@
 
                 if (this.field_73136_m > 0 || this.field_73132_a instanceof EntityArrow)
                 {
-                    if (j >= -32768L
-                            && j < 32768L
-                            && k >= -32768L
-                            && k < 32768L
-                            && l >= -32768L
-                            && l < 32768L
-                            && this.field_73142_u <= 400
-                            && !this.field_73141_v
-                            && this.field_180234_y == this.field_73132_a.field_70122_E)
+                    if (j >= -32768L && j < 32768L && k >= -32768L && k < 32768L && l >= -32768L && l < 32768L && this.field_73142_u <= 400 && !this.field_73141_v && this.field_180234_y == this.field_73132_a.field_70122_E)
                     {
                         if ((!flag || !flag1) && !(this.field_73132_a instanceof EntityArrow))
                         {
@@ -236,9 +229,7 @@
                         }
                         else
                         {
-                            packet1 = new SPacketEntity.S17PacketEntityLookMove(
-                                this.field_73132_a.func_145782_y(), j, k, l, (byte)k2, (byte)i, this.field_73132_a.field_70122_E
-                            );
+                            packet1 = new SPacketEntity.S17PacketEntityLookMove(this.field_73132_a.func_145782_y(), j, k, l, (byte)k2, (byte)i, this.field_73132_a.field_70122_E);
                         }
                     }
                     else
@@ -262,11 +253,10 @@
                     double d0 = this.field_73132_a.field_70159_w - this.field_73137_j;
                     double d1 = this.field_73132_a.field_70181_x - this.field_73138_k;
                     double d2 = this.field_73132_a.field_70179_y - this.field_73135_l;
-                    double d3 = 0.02;
+                    double d3 = 0.02D;
                     double d4 = d0 * d0 + d1 * d1 + d2 * d2;
 
-                    if (d4 > 4.0E-4
-                            || d4 > 0.0 && this.field_73132_a.field_70159_w == 0.0 && this.field_73132_a.field_70181_x == 0.0 && this.field_73132_a.field_70179_y == 0.0)
+                    if (d4 > 4.0E-4D || d4 > 0.0D && this.field_73132_a.field_70159_w == 0.0D && this.field_73132_a.field_70181_x == 0.0D && this.field_73132_a.field_70179_y == 0.0D)
                     {
                         this.field_73137_j = this.field_73132_a.field_70159_w;
                         this.field_73138_k = this.field_73132_a.field_70181_x;
@@ -309,7 +299,7 @@
             this.field_73132_a.field_70160_al = false;
         }
 
-        this.field_73136_m++;
+        ++this.field_73136_m;
 
         if (this.field_73132_a.field_70133_I)
         {
@@ -392,8 +382,7 @@
 
                     if (!this.field_73132_a.func_184212_Q().func_187228_d())
                     {
-                        p_73117_1_.field_71135_a
-                        .func_147359_a(new SPacketEntityMetadata(this.field_73132_a.func_145782_y(), this.field_73132_a.func_184212_Q(), true));
+                        p_73117_1_.field_71135_a.func_147359_a(new SPacketEntityMetadata(this.field_73132_a.func_145782_y(), this.field_73132_a.func_184212_Q(), true));
                     }
 
                     boolean flag = this.field_73143_t;
@@ -420,15 +409,7 @@
 
                     if (flag && !(packet instanceof SPacketSpawnMob))
                     {
-                        p_73117_1_.field_71135_a
-                        .func_147359_a(
-                            new SPacketEntityVelocity(
-                                this.field_73132_a.func_145782_y(),
-                                this.field_73132_a.field_70159_w,
-                                this.field_73132_a.field_70181_x,
-                                this.field_73132_a.field_70179_y
-                            )
-                        );
+                        p_73117_1_.field_71135_a.func_147359_a(new SPacketEntityVelocity(this.field_73132_a.func_145782_y(), this.field_73132_a.field_70159_w, this.field_73132_a.field_70181_x, this.field_73132_a.field_70179_y));
                     }
 
                     if (this.field_73132_a instanceof EntityLivingBase)
@@ -476,6 +457,7 @@
 
                     this.field_73132_a.func_184178_b(p_73117_1_);
                     p_73117_1_.func_184848_d(this.field_73132_a);
+                    net.minecraftforge.event.ForgeEventFactory.onStartEntityTracking(field_73132_a, p_73117_1_);
                 }
             }
             else if (this.field_73134_o.contains(p_73117_1_))
@@ -483,14 +465,15 @@
                 this.field_73134_o.remove(p_73117_1_);
                 this.field_73132_a.func_184203_c(p_73117_1_);
                 p_73117_1_.func_152339_d(this.field_73132_a);
+                net.minecraftforge.event.ForgeEventFactory.onStopEntityTracking(field_73132_a, p_73117_1_);
             }
         }
     }
 
     public boolean func_180233_c(EntityPlayerMP p_180233_1_)
     {
-        double d0 = p_180233_1_.field_70165_t - (double)this.field_73128_d / 4096.0;
-        double d1 = p_180233_1_.field_70161_v - (double)this.field_73126_f / 4096.0;
+        double d0 = p_180233_1_.field_70165_t - (double)this.field_73128_d / 4096.0D;
+        double d1 = p_180233_1_.field_70161_v - (double)this.field_73126_f / 4096.0D;
         int i = Math.min(this.field_73130_b, this.field_187262_f);
         return d0 >= (double)(-i) && d0 <= (double)i && d1 >= (double)(-i) && d1 <= (double)i && this.field_73132_a.func_174827_a(p_180233_1_);
     }
@@ -502,7 +485,7 @@
 
     public void func_73125_b(List<EntityPlayer> p_73125_1_)
     {
-        for (int i = 0; i < p_73125_1_.size(); i++)
+        for (int i = 0; i < p_73125_1_.size(); ++i)
         {
             this.func_73117_b((EntityPlayerMP)p_73125_1_.get(i));
         }
@@ -515,6 +498,9 @@
             field_151262_p.warn("Fetching addPacket for removed entity");
         }
 
+        Packet pkt = net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.getEntitySpawningPacket(this.field_73132_a);
+        if (pkt != null) return pkt;
+
         if (this.field_73132_a instanceof EntityPlayerMP)
         {
             return new SPacketSpawnPlayer((EntityPlayer)this.field_73132_a);
@@ -616,17 +602,17 @@
                 spacketspawnobject = new SPacketSpawnObject(this.field_73132_a, i, 0);
             }
 
-            spacketspawnobject.func_149003_d((int)(entityfireball.field_70232_b * 8000.0));
-            spacketspawnobject.func_149000_e((int)(entityfireball.field_70233_c * 8000.0));
-            spacketspawnobject.func_149007_f((int)(entityfireball.field_70230_d * 8000.0));
+            spacketspawnobject.func_149003_d((int)(entityfireball.field_70232_b * 8000.0D));
+            spacketspawnobject.func_149000_e((int)(entityfireball.field_70233_c * 8000.0D));
+            spacketspawnobject.func_149007_f((int)(entityfireball.field_70230_d * 8000.0D));
             return spacketspawnobject;
         }
         else if (this.field_73132_a instanceof EntityShulkerBullet)
         {
             SPacketSpawnObject spacketspawnobject1 = new SPacketSpawnObject(this.field_73132_a, 67, 0);
-            spacketspawnobject1.func_149003_d((int)(this.field_73132_a.field_70159_w * 8000.0));
-            spacketspawnobject1.func_149000_e((int)(this.field_73132_a.field_70181_x * 8000.0));
-            spacketspawnobject1.func_149007_f((int)(this.field_73132_a.field_70179_y * 8000.0));
+            spacketspawnobject1.func_149003_d((int)(this.field_73132_a.field_70159_w * 8000.0D));
+            spacketspawnobject1.func_149000_e((int)(this.field_73132_a.field_70181_x * 8000.0D));
+            spacketspawnobject1.func_149007_f((int)(this.field_73132_a.field_70179_y * 8000.0D));
             return spacketspawnobject1;
         }
         else if (this.field_73132_a instanceof EntityEgg)
