--- before/net/minecraft/entity/EntityFlying.java
+++ after/net/minecraft/entity/EntityFlying.java
@@ -47,17 +47,14 @@
 
             if (this.onGround)
             {
-                f = this.world
-                    .getBlockState(
+                BlockPos underPos =
                         new BlockPos(
                             MathHelper.floor(this.posX),
                             MathHelper.floor(this.getEntityBoundingBox().minY) - 1,
                             MathHelper.floor(this.posZ)
-                        )
-                    )
-                    .getBlock()
-                    .slipperiness
-                    * 0.91F;
+                        );
+                IBlockState underState = this.world.getBlockState(underPos);
+                f = underState.getBlock().getSlipperiness(underState, this.world, underPos, this) * 0.91F;
             }
 
             float f1 = 0.16277136F / (f * f * f);
@@ -66,17 +63,14 @@
 
             if (this.onGround)
             {
-                f = this.world
-                    .getBlockState(
+                BlockPos underPos =
                         new BlockPos(
                             MathHelper.floor(this.posX),
                             MathHelper.floor(this.getEntityBoundingBox().minY) - 1,
                             MathHelper.floor(this.posZ)
-                        )
-                    )
-                    .getBlock()
-                    .slipperiness
-                    * 0.91F;
+                        );
+                IBlockState underState = this.world.getBlockState(underPos);
+                f = underState.getBlock().getSlipperiness(underState, this.world, underPos, this) * 0.91F;
             }
 
             this.move(MoverType.SELF, this.motionX, this.motionY, this.motionZ);
