--- before/net/minecraft/entity/EntitySpawnPlacementRegistry.java
+++ after/net/minecraft/entity/EntitySpawnPlacementRegistry.java
@@ -46,11 +46,16 @@
 
 public class EntitySpawnPlacementRegistry
 {
-    private static final Map < Class<?>, EntityLiving.SpawnPlacementType > ENTITY_PLACEMENTS = Maps.newHashMap();
+    private static final Map < Class<?>, EntityLiving.SpawnPlacementType > ENTITY_PLACEMENTS = Maps. < Class<?>, EntityLiving.SpawnPlacementType > newHashMap();
 
     public static EntityLiving.SpawnPlacementType getPlacementForEntity(Class<?> entityClass)
     {
-        return ENTITY_PLACEMENTS.get(entityClass);
+        return ENTITY_PLACEMENTS.getOrDefault(entityClass, EntityLiving.SpawnPlacementType.ON_GROUND);
+    }
+
+    public static void setPlacementType(Class<? extends Entity> entityClass, EntityLiving.SpawnPlacementType placementType)
+    {
+        ENTITY_PLACEMENTS.putIfAbsent(entityClass, placementType);
     }
 
     static
