--- before/net/minecraft/entity/EntityAgeable.java
+++ after/net/minecraft/entity/EntityAgeable.java
@@ -39,7 +39,7 @@
         {
             if (!this.world.isRemote)
             {
-                Class <? extends Entity > oclass = EntityList.REGISTRY.getObject(ItemMonsterPlacer.getNamedIdFrom(itemstack));
+                Class <? extends Entity > oclass = EntityList.getClass(ItemMonsterPlacer.getNamedIdFrom(itemstack));
 
                 if (oclass != null && this.getClass() == oclass)
                 {
@@ -80,7 +80,7 @@
         }
         else
         {
-            Class <? extends Entity > oclass = EntityList.REGISTRY.getObject(ItemMonsterPlacer.getNamedIdFrom(stack));
+            Class <? extends Entity > oclass = EntityList.getClass(ItemMonsterPlacer.getNamedIdFrom(stack));
             return oclass != null && entityClass == oclass;
         }
     }
