--- before/net/minecraft/entity/NpcMerchant.java
+++ after/net/minecraft/entity/NpcMerchant.java
@@ -29,54 +29,45 @@
     }
 
     @Nullable
-    @Override
     public EntityPlayer func_70931_l_()
     {
         return this.field_70935_b;
     }
 
-    @Override
     public void func_70932_a_(@Nullable EntityPlayer p_70932_1_)
     {
     }
 
     @Nullable
-    @Override
     public MerchantRecipeList func_70934_b(EntityPlayer p_70934_1_)
     {
-        return this.field_70936_c;
+        return net.minecraftforge.event.ForgeEventFactory.listTradeOffers(this, p_70934_1_, this.field_70936_c);
     }
 
-    @Override
     public void func_70930_a(@Nullable MerchantRecipeList p_70930_1_)
     {
         this.field_70936_c = p_70930_1_;
     }
 
-    @Override
     public void func_70933_a(MerchantRecipe p_70933_1_)
     {
         p_70933_1_.func_77399_f();
     }
 
-    @Override
     public void func_110297_a_(ItemStack p_110297_1_)
     {
     }
 
-    @Override
     public ITextComponent func_145748_c_()
     {
-        return (ITextComponent)(this.field_175548_d != null ? this.field_175548_d : new TextComponentTranslation("entity.Villager.name"));
+        return (ITextComponent)(this.field_175548_d != null ? this.field_175548_d : new TextComponentTranslation("entity.Villager.name", new Object[0]));
     }
 
-    @Override
     public World func_190670_t_()
     {
         return this.field_70935_b.field_70170_p;
     }
 
-    @Override
     public BlockPos func_190671_u_()
     {
         return new BlockPos(this.field_70935_b);
