--- before/net/minecraft/entity/NpcMerchant.java
+++ after/net/minecraft/entity/NpcMerchant.java
@@ -41,7 +41,7 @@
     @Nullable
     public MerchantRecipeList func_70934_b(EntityPlayer p_70934_1_)
     {
-        return this.field_70936_c;
+        return net.minecraftforge.event.ForgeEventFactory.listTradeOffers(this, p_70934_1_, this.field_70936_c);
     }
 
     public void func_70930_a(@Nullable MerchantRecipeList p_70930_1_)
@@ -60,7 +60,7 @@
 
     public ITextComponent func_145748_c_()
     {
-        return (ITextComponent)(this.field_175548_d != null ? this.field_175548_d : new TextComponentTranslation("entity.Villager.name"));
+        return (ITextComponent)(this.field_175548_d != null ? this.field_175548_d : new TextComponentTranslation("entity.Villager.name", new Object[0]));
     }
 
     public World func_190670_t_()
