--- before/net/minecraft/entity/SharedMonsterAttributes.java
+++ after/net/minecraft/entity/SharedMonsterAttributes.java
@@ -16,7 +16,7 @@
 public class SharedMonsterAttributes
 {
     private static final Logger LOGGER = LogManager.getLogger();
-    public static final IAttribute MAX_HEALTH = new RangedAttribute(null, "generic.maxHealth", 20.0, 0.0, 1024.0)
+    public static final IAttribute MAX_HEALTH = (new RangedAttribute(null, "generic.maxHealth", 20.0, Float.MIN_VALUE, 1024.0)) // Forge: set smallest max-health value to fix MC-119183. This gets rounded to float so we use the smallest positive float value.
         .setDescription("Max Health")
         .setShouldWatch(true);
     public static final IAttribute FOLLOW_RANGE = new RangedAttribute(null, "generic.followRange", 32.0, 0.0, 2048.0).setDescription("Follow Range");
