--- before/net/minecraft/network/play/client/CPacketClickWindow.java
+++ after/net/minecraft/network/play/client/CPacketClickWindow.java
@@ -57,7 +57,7 @@
         buf.writeByte(this.packedClickData);
         buf.writeShort(this.actionNumber);
         buf.writeEnumValue(this.mode);
-        buf.writeItemStack(this.clickedItem);
+        net.minecraftforge.common.util.PacketUtil.writeItemStackFromClientToServer(buf, this.clickedItem);
     }
 
     public int getWindowId()
