--- before/net/minecraft/network/play/client/CPacketUseEntity.java
+++ after/net/minecraft/network/play/client/CPacketUseEntity.java
@@ -49,7 +49,7 @@
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
         this.field_149567_a = p_148837_1_.func_150792_a();
-        this.field_149566_b = p_148837_1_.func_179257_a(CPacketUseEntity.Action.class);
+        this.field_149566_b = (CPacketUseEntity.Action)p_148837_1_.func_179257_a(CPacketUseEntity.Action.class);
 
         if (this.field_149566_b == CPacketUseEntity.Action.INTERACT_AT)
         {
@@ -58,7 +58,7 @@
 
         if (this.field_149566_b == CPacketUseEntity.Action.INTERACT || this.field_149566_b == CPacketUseEntity.Action.INTERACT_AT)
         {
-            this.field_186995_d = p_148837_1_.func_179257_a(EnumHand.class);
+            this.field_186995_d = (EnumHand)p_148837_1_.func_179257_a(EnumHand.class);
         }
     }
 
