--- before/net/minecraft/network/play/client/CPacketAnimation.java
+++ after/net/minecraft/network/play/client/CPacketAnimation.java
@@ -21,7 +21,7 @@
 
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
-        this.field_187019_a = p_148837_1_.func_179257_a(EnumHand.class);
+        this.field_187019_a = (EnumHand)p_148837_1_.func_179257_a(EnumHand.class);
     }
 
     public void func_148840_b(PacketBuffer p_148840_1_) throws IOException
