--- before/net/minecraft/network/play/client/CPacketSeenAdvancements.java
+++ after/net/minecraft/network/play/client/CPacketSeenAdvancements.java
@@ -40,7 +40,7 @@
 
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
-        this.field_194166_a = p_148837_1_.func_179257_a(CPacketSeenAdvancements.Action.class);
+        this.field_194166_a = (CPacketSeenAdvancements.Action)p_148837_1_.func_179257_a(CPacketSeenAdvancements.Action.class);
 
         if (this.field_194166_a == CPacketSeenAdvancements.Action.OPENED_TAB)
         {
