--- before/net/minecraft/network/play/client/CPacketCreativeInventoryAction.java
+++ after/net/minecraft/network/play/client/CPacketCreativeInventoryAction.java
@@ -40,7 +40,7 @@
     public void writePacketData(PacketBuffer buf) throws IOException
     {
         buf.writeShort(this.slotId);
-        buf.writeItemStack(this.stack);
+        net.minecraftforge.common.util.PacketUtil.writeItemStackFromClientToServer(buf, this.stack);
     }
 
     public int getSlotId()
