--- before/net/minecraft/network/play/client/CPacketPlayerDigging.java
+++ after/net/minecraft/network/play/client/CPacketPlayerDigging.java
@@ -29,7 +29,7 @@
 
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
-        this.field_149508_e = p_148837_1_.func_179257_a(CPacketPlayerDigging.Action.class);
+        this.field_149508_e = (CPacketPlayerDigging.Action)p_148837_1_.func_179257_a(CPacketPlayerDigging.Action.class);
         this.field_179717_a = p_148837_1_.func_179259_c();
         this.field_179716_b = EnumFacing.func_82600_a(p_148837_1_.readUnsignedByte());
     }
