--- before/net/minecraft/network/play/client/CPacketPlayerTryUseItemOnBlock.java
+++ after/net/minecraft/network/play/client/CPacketPlayerTryUseItemOnBlock.java
@@ -37,8 +37,8 @@
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
         this.field_179725_b = p_148837_1_.func_179259_c();
-        this.field_149579_d = p_148837_1_.func_179257_a(EnumFacing.class);
-        this.field_187027_c = p_148837_1_.func_179257_a(EnumHand.class);
+        this.field_149579_d = (EnumFacing)p_148837_1_.func_179257_a(EnumFacing.class);
+        this.field_187027_c = (EnumHand)p_148837_1_.func_179257_a(EnumHand.class);
         this.field_149577_f = p_148837_1_.readFloat();
         this.field_149578_g = p_148837_1_.readFloat();
         this.field_149584_h = p_148837_1_.readFloat();
