--- before/net/minecraft/client/renderer/entity/layers/LayerBipedArmor.java
+++ after/net/minecraft/client/renderer/entity/layers/LayerBipedArmor.java
@@ -14,13 +14,13 @@
         super(rendererIn);
     }
 
-    @Override
     protected void initArmor()
     {
         this.modelLeggings = new ModelBiped(0.5F);
         this.modelArmor = new ModelBiped(1.0F);
     }
 
+    @SuppressWarnings("incomplete-switch")
     protected void setModelSlotVisible(ModelBiped p_188359_1_, EntityEquipmentSlot slotIn)
     {
         this.setModelVisible(p_188359_1_);
@@ -50,5 +50,11 @@
     protected void setModelVisible(ModelBiped model)
     {
         model.setVisible(false);
+    }
+
+    @Override
+    protected ModelBiped getArmorModelHook(net.minecraft.entity.EntityLivingBase entity, net.minecraft.item.ItemStack itemStack, EntityEquipmentSlot slot, ModelBiped model)
+    {
+        return net.minecraftforge.client.ForgeHooksClient.getArmorModel(entity, itemStack, slot, model);
     }
 }
