--- before/net/minecraft/client/renderer/entity/layers/LayerCape.java
+++ after/net/minecraft/client/renderer/entity/layers/LayerCape.java
@@ -21,21 +21,9 @@
         this.field_177167_a = p_i46123_1_;
     }
 
-    public void func_177141_a(
-        AbstractClientPlayer p_177141_1_,
-        float p_177141_2_,
-        float p_177141_3_,
-        float p_177141_4_,
-        float p_177141_5_,
-        float p_177141_6_,
-        float p_177141_7_,
-        float p_177141_8_
-    )
+    public void func_177141_a(AbstractClientPlayer p_177141_1_, float p_177141_2_, float p_177141_3_, float p_177141_4_, float p_177141_5_, float p_177141_6_, float p_177141_7_, float p_177141_8_)
     {
-        if (p_177141_1_.func_152122_n()
-                && !p_177141_1_.func_82150_aj()
-                && p_177141_1_.func_175148_a(EnumPlayerModelParts.CAPE)
-                && p_177141_1_.func_110303_q() != null)
+        if (p_177141_1_.func_152122_n() && !p_177141_1_.func_82150_aj() && p_177141_1_.func_175148_a(EnumPlayerModelParts.CAPE) && p_177141_1_.func_110303_q() != null)
         {
             ItemStack itemstack = p_177141_1_.func_184582_a(EntityEquipmentSlot.CHEST);
 
@@ -45,18 +33,12 @@
                 this.field_177167_a.func_110776_a(p_177141_1_.func_110303_q());
                 GlStateManager.func_179094_E();
                 GlStateManager.func_179109_b(0.0F, 0.0F, 0.125F);
-                double d0 = p_177141_1_.field_71091_bM
-                            + (p_177141_1_.field_71094_bP - p_177141_1_.field_71091_bM) * (double)p_177141_4_
-                            - (p_177141_1_.field_70169_q + (p_177141_1_.field_70165_t - p_177141_1_.field_70169_q) * (double)p_177141_4_);
-                double d1 = p_177141_1_.field_71096_bN
-                            + (p_177141_1_.field_71095_bQ - p_177141_1_.field_71096_bN) * (double)p_177141_4_
-                            - (p_177141_1_.field_70167_r + (p_177141_1_.field_70163_u - p_177141_1_.field_70167_r) * (double)p_177141_4_);
-                double d2 = p_177141_1_.field_71097_bO
-                            + (p_177141_1_.field_71085_bR - p_177141_1_.field_71097_bO) * (double)p_177141_4_
-                            - (p_177141_1_.field_70166_s + (p_177141_1_.field_70161_v - p_177141_1_.field_70166_s) * (double)p_177141_4_);
+                double d0 = p_177141_1_.field_71091_bM + (p_177141_1_.field_71094_bP - p_177141_1_.field_71091_bM) * (double)p_177141_4_ - (p_177141_1_.field_70169_q + (p_177141_1_.field_70165_t - p_177141_1_.field_70169_q) * (double)p_177141_4_);
+                double d1 = p_177141_1_.field_71096_bN + (p_177141_1_.field_71095_bQ - p_177141_1_.field_71096_bN) * (double)p_177141_4_ - (p_177141_1_.field_70167_r + (p_177141_1_.field_70163_u - p_177141_1_.field_70167_r) * (double)p_177141_4_);
+                double d2 = p_177141_1_.field_71097_bO + (p_177141_1_.field_71085_bR - p_177141_1_.field_71097_bO) * (double)p_177141_4_ - (p_177141_1_.field_70166_s + (p_177141_1_.field_70161_v - p_177141_1_.field_70166_s) * (double)p_177141_4_);
                 float f = p_177141_1_.field_70760_ar + (p_177141_1_.field_70761_aq - p_177141_1_.field_70760_ar) * p_177141_4_;
-                double d3 = (double)MathHelper.func_76126_a(f * (float)(Math.PI / 180.0));
-                double d4 = (double)(-MathHelper.func_76134_b(f * (float)(Math.PI / 180.0)));
+                double d3 = (double)MathHelper.func_76126_a(f * 0.017453292F);
+                double d4 = (double)(-MathHelper.func_76134_b(f * 0.017453292F));
                 float f1 = (float)d1 * 10.0F;
                 f1 = MathHelper.func_76131_a(f1, -6.0F, 32.0F);
                 float f2 = (float)(d0 * d3 + d2 * d4) * 100.0F;
@@ -68,9 +50,7 @@
                 }
 
                 float f4 = p_177141_1_.field_71107_bF + (p_177141_1_.field_71109_bG - p_177141_1_.field_71107_bF) * p_177141_4_;
-                f1 += MathHelper.func_76126_a((p_177141_1_.field_70141_P + (p_177141_1_.field_70140_Q - p_177141_1_.field_70141_P) * p_177141_4_) * 6.0F)
-                      * 32.0F
-                      * f4;
+                f1 = f1 + MathHelper.func_76126_a((p_177141_1_.field_70141_P + (p_177141_1_.field_70140_Q - p_177141_1_.field_70141_P) * p_177141_4_) * 6.0F) * 32.0F * f4;
 
                 if (p_177141_1_.func_70093_af())
                 {
@@ -87,7 +67,6 @@
         }
     }
 
-    @Override
     public boolean func_177142_b()
     {
         return false;
