--- before/net/minecraft/block/state/BlockStateBase.java
+++ after/net/minecraft/block/state/BlockStateBase.java
@@ -24,7 +24,7 @@
             }
             else
             {
-                IProperty<?> iproperty = p_apply_1_.getKey();
+                IProperty<?> iproperty = (IProperty)p_apply_1_.getKey();
                 return iproperty.func_177701_a() + "=" + this.func_185886_a(iproperty, p_apply_1_.getValue());
             }
         }
@@ -34,7 +34,6 @@
         }
     };
 
-    @Override
     public <T extends Comparable<T>> IBlockState func_177231_a(IProperty<T> p_177231_1_)
     {
         return this.func_177226_a(p_177231_1_, func_177232_a(p_177231_1_.func_177700_c(), this.func_177229_b(p_177231_1_)));
@@ -60,7 +59,6 @@
         return iterator.next();
     }
 
-    @Override
     public String toString()
     {
         StringBuilder stringbuilder = new StringBuilder();
@@ -74,5 +72,11 @@
         }
 
         return stringbuilder.toString();
+    }
+
+    @Nullable
+    public com.google.common.collect.ImmutableTable<IProperty<?>, Comparable<?>, IBlockState> getPropertyValueTable()
+    {
+        return null;
     }
 }
