--- before/net/minecraft/block/state/BlockPistonStructureHelper.java
+++ after/net/minecraft/block/state/BlockPistonStructureHelper.java
@@ -17,8 +17,8 @@
     private final BlockPos field_177259_b;
     private final BlockPos field_177260_c;
     private final EnumFacing field_177257_d;
-    private final List<BlockPos> field_177258_e = Lists.newArrayList();
-    private final List<BlockPos> field_177256_f = Lists.newArrayList();
+    private final List<BlockPos> field_177258_e = Lists.<BlockPos>newArrayList();
+    private final List<BlockPos> field_177256_f = Lists.<BlockPos>newArrayList();
 
     public BlockPistonStructureHelper(World p_i45664_1_, BlockPos p_i45664_2_, EnumFacing p_i45664_3_, boolean p_i45664_4_)
     {
@@ -61,11 +61,11 @@
         }
         else
         {
-            for (int i = 0; i < this.field_177258_e.size(); i++)
+            for (int i = 0; i < this.field_177258_e.size(); ++i)
             {
                 BlockPos blockpos = this.field_177258_e.get(i);
 
-                if (this.field_177261_a.func_180495_p(blockpos).func_177230_c() == Blocks.field_180399_cE && !this.func_177250_b(blockpos))
+                if (this.field_177261_a.func_180495_p(blockpos).func_177230_c().isStickyBlock(this.field_177261_a.func_180495_p(blockpos)) && !this.func_177250_b(blockpos))
                 {
                     return false;
                 }
@@ -80,7 +80,7 @@
         IBlockState iblockstate = this.field_177261_a.func_180495_p(p_177251_1_);
         Block block = iblockstate.func_177230_c();
 
-        if (iblockstate.func_185904_a() == Material.field_151579_a)
+        if (iblockstate.func_177230_c().isAir(iblockstate, this.field_177261_a, p_177251_1_))
         {
             return true;
         }
@@ -106,20 +106,20 @@
             }
             else
             {
-                while (block == Blocks.field_180399_cE)
+                while (block.isStickyBlock(iblockstate))
                 {
                     BlockPos blockpos = p_177251_1_.func_177967_a(this.field_177257_d.func_176734_d(), i);
                     iblockstate = this.field_177261_a.func_180495_p(blockpos);
                     block = iblockstate.func_177230_c();
 
-                    if (iblockstate.func_185904_a() == Material.field_151579_a
-                            || !BlockPistonBase.func_185646_a(iblockstate, this.field_177261_a, blockpos, this.field_177257_d, false, this.field_177257_d.func_176734_d())
-                            || blockpos.equals(this.field_177259_b))
+                    if (iblockstate.func_177230_c().isAir(iblockstate, this.field_177261_a, blockpos) || !BlockPistonBase.func_185646_a(iblockstate, this.field_177261_a, blockpos, this.field_177257_d, false, this.field_177257_d.func_176734_d()) || blockpos.equals(this.field_177259_b))
                     {
                         break;
                     }
 
-                    if (++i + this.field_177258_e.size() > 12)
+                    ++i;
+
+                    if (i + this.field_177258_e.size() > 12)
                     {
                         return false;
                     }
@@ -127,10 +127,10 @@
 
                 int i1 = 0;
 
-                for (int j = i - 1; j >= 0; j--)
+                for (int j = i - 1; j >= 0; --j)
                 {
                     this.field_177258_e.add(p_177251_1_.func_177967_a(this.field_177257_d.func_176734_d(), j));
-                    i1++;
+                    ++i1;
                 }
 
                 int j1 = 1;
@@ -144,11 +144,11 @@
                     {
                         this.func_177255_a(i1, k);
 
-                        for (int l = 0; l <= k + i1; l++)
+                        for (int l = 0; l <= k + i1; ++l)
                         {
                             BlockPos blockpos2 = this.field_177258_e.get(l);
 
-                            if (this.field_177261_a.func_180495_p(blockpos2).func_177230_c() == Blocks.field_180399_cE && !this.func_177250_b(blockpos2))
+                            if (this.field_177261_a.func_180495_p(blockpos2).func_177230_c().isStickyBlock(this.field_177261_a.func_180495_p(blockpos2)) && !this.func_177250_b(blockpos2))
                             {
                                 return false;
                             }
@@ -159,13 +159,12 @@
 
                     iblockstate = this.field_177261_a.func_180495_p(blockpos1);
 
-                    if (iblockstate.func_185904_a() == Material.field_151579_a)
+                    if (iblockstate.func_177230_c().isAir(iblockstate, this.field_177261_a, blockpos1))
                     {
                         return true;
                     }
 
-                    if (!BlockPistonBase.func_185646_a(iblockstate, this.field_177261_a, blockpos1, this.field_177257_d, true, this.field_177257_d)
-                            || blockpos1.equals(this.field_177259_b))
+                    if (!BlockPistonBase.func_185646_a(iblockstate, this.field_177261_a, blockpos1, this.field_177257_d, true, this.field_177257_d) || blockpos1.equals(this.field_177259_b))
                     {
                         return false;
                     }
@@ -182,8 +181,8 @@
                     }
 
                     this.field_177258_e.add(blockpos1);
-                    i1++;
-                    j1++;
+                    ++i1;
+                    ++j1;
                 }
             }
         }
@@ -191,9 +190,9 @@
 
     private void func_177255_a(int p_177255_1_, int p_177255_2_)
     {
-        List<BlockPos> list = Lists.newArrayList();
-        List<BlockPos> list1 = Lists.newArrayList();
-        List<BlockPos> list2 = Lists.newArrayList();
+        List<BlockPos> list = Lists.<BlockPos>newArrayList();
+        List<BlockPos> list1 = Lists.<BlockPos>newArrayList();
+        List<BlockPos> list2 = Lists.<BlockPos>newArrayList();
         list.addAll(this.field_177258_e.subList(0, p_177255_2_));
         list1.addAll(this.field_177258_e.subList(this.field_177258_e.size() - p_177255_1_, this.field_177258_e.size()));
         list2.addAll(this.field_177258_e.subList(p_177255_2_, this.field_177258_e.size() - p_177255_1_));
