--- before/net/minecraft/block/state/IBlockProperties.java
+++ after/net/minecraft/block/state/IBlockProperties.java
@@ -28,9 +28,13 @@
 
     boolean func_189884_a(Entity p_189884_1_);
 
+    @Deprecated //Forge location aware version below
     int func_185891_c();
+    int getLightOpacity(IBlockAccess world, BlockPos pos);
 
+    @Deprecated //Forge location aware version below
     int func_185906_d();
+    int getLightValue(IBlockAccess world, BlockPos pos);
 
     @SideOnly(Side.CLIENT)
     boolean func_185895_e();
@@ -89,15 +93,19 @@
     @Nullable
     AxisAlignedBB func_185890_d(IBlockAccess p_185890_1_, BlockPos p_185890_2_);
 
-    void func_185908_a(
-        World p_185908_1_, BlockPos p_185908_2_, AxisAlignedBB p_185908_3_, List<AxisAlignedBB> p_185908_4_, @Nullable Entity p_185908_5_, boolean p_185908_6_
-    );
+    void func_185908_a(World p_185908_1_, BlockPos p_185908_2_, AxisAlignedBB p_185908_3_, List<AxisAlignedBB> p_185908_4_, @Nullable Entity p_185908_5_, boolean p_185908_6_);
 
     AxisAlignedBB func_185900_c(IBlockAccess p_185900_1_, BlockPos p_185900_2_);
 
     RayTraceResult func_185910_a(World p_185910_1_, BlockPos p_185910_2_, Vec3d p_185910_3_, Vec3d p_185910_4_);
 
+    @Deprecated // Forge: Use isSideSolid(IBlockAccess, BlockPos, EnumFacing.UP) instead
     boolean func_185896_q();
+
+    //Forge added functions
+    boolean doesSideBlockRendering(IBlockAccess world, BlockPos pos, EnumFacing side);
+    boolean isSideSolid(IBlockAccess world, BlockPos pos, EnumFacing side);
+    boolean doesSideBlockChestOpening(IBlockAccess world, BlockPos pos, EnumFacing side);
 
     Vec3d func_191059_e(IBlockAccess p_191059_1_, BlockPos p_191059_2_);
 
