--- before/net/minecraft/world/gen/ChunkGeneratorFlat.java
+++ after/net/minecraft/world/gen/ChunkGeneratorFlat.java
@@ -166,6 +166,7 @@
     @Override
     public void populate(int x, int z)
     {
+        net.minecraft.block.BlockFalling.fallInstantly = true;
         int i = x * 16;
         int j = z * 16;
         BlockPos blockpos = new BlockPos(i, 0, j);
@@ -177,6 +178,8 @@
         this.random.setSeed(x * k + z * l ^ this.world.getSeed());
         ChunkPos chunkpos = new ChunkPos(x, z);
 
+        net.minecraftforge.event.ForgeEventFactory.onChunkPopulate(true, this, this.world, this.random, x, z, flag);
+
         for (MapGenStructure mapgenstructure : this.structureGenerators.values())
         {
             boolean flag1 = mapgenstructure.generateStructure(this.world, this.random, chunkpos);
@@ -226,6 +229,9 @@
         {
             biome.decorate(this.world, this.random, blockpos);
         }
+
+        net.minecraftforge.event.ForgeEventFactory.onChunkPopulate(false, this, this.world, this.random, x, z, flag);
+        net.minecraft.block.BlockFalling.fallInstantly = false;
     }
 
     @Override
