--- before/net/minecraft/world/gen/ChunkGeneratorDebug.java
+++ after/net/minecraft/world/gen/ChunkGeneratorDebug.java
@@ -16,7 +16,7 @@
 
 public class ChunkGeneratorDebug implements IChunkGenerator
 {
-    private static final List<IBlockState> field_177464_a = Lists.newArrayList();
+    private static final List<IBlockState> field_177464_a = Lists.<IBlockState>newArrayList();
     private static final int field_177462_b;
     private static final int field_181039_c;
     protected static final IBlockState field_185934_a = Blocks.field_150350_a.func_176223_P();
@@ -68,8 +68,8 @@
 
         if (p_177461_0_ > 0 && p_177461_1_ > 0 && p_177461_0_ % 2 != 0 && p_177461_1_ % 2 != 0)
         {
-            p_177461_0_ /= 2;
-            p_177461_1_ /= 2;
+            p_177461_0_ = p_177461_0_ / 2;
+            p_177461_1_ = p_177461_1_ / 2;
 
             if (p_177461_0_ <= field_177462_b && p_177461_1_ <= field_181039_c)
             {
