--- before/net/minecraft/world/gen/ChunkGeneratorSettings.java
+++ after/net/minecraft/world/gen/ChunkGeneratorSettings.java
@@ -274,9 +274,9 @@
                 {
                     try
                     {
-                        return JsonUtils.func_188178_a(field_177901_a, p_177865_0_, ChunkGeneratorSettings.Factory.class);
+                        return (ChunkGeneratorSettings.Factory)JsonUtils.func_188178_a(field_177901_a, p_177865_0_, ChunkGeneratorSettings.Factory.class);
                     }
-                    catch (Exception exception)
+                    catch (Exception var2)
                     {
                         return new ChunkGeneratorSettings.Factory();
                     }
@@ -789,7 +789,8 @@
                 i = 31 * i + this.field_177910_ax;
                 i = 31 * i + this.field_177897_ay;
                 i = 31 * i + this.field_177895_az;
-                return 31 * i + this.field_177889_aA;
+                i = 31 * i + this.field_177889_aA;
+                return i;
             }
 
             public ChunkGeneratorSettings func_177864_b()
@@ -900,8 +901,9 @@
                     chunkgeneratorsettings$factory.field_177895_az = JsonUtils.func_151208_a(jsonobject, "lapisCenterHeight", chunkgeneratorsettings$factory.field_177895_az);
                     chunkgeneratorsettings$factory.field_177889_aA = JsonUtils.func_151208_a(jsonobject, "lapisSpread", chunkgeneratorsettings$factory.field_177889_aA);
                 }
-                catch (Exception exception)
+                catch (Exception var7)
                 {
+                    ;
                 }
 
                 return chunkgeneratorsettings$factory;
@@ -910,85 +912,85 @@
             public JsonElement serialize(ChunkGeneratorSettings.Factory p_serialize_1_, Type p_serialize_2_, JsonSerializationContext p_serialize_3_)
             {
                 JsonObject jsonobject = new JsonObject();
-                jsonobject.addProperty("coordinateScale", p_serialize_1_.field_177899_b);
-                jsonobject.addProperty("heightScale", p_serialize_1_.field_177900_c);
-                jsonobject.addProperty("lowerLimitScale", p_serialize_1_.field_177898_e);
-                jsonobject.addProperty("upperLimitScale", p_serialize_1_.field_177896_d);
-                jsonobject.addProperty("depthNoiseScaleX", p_serialize_1_.field_177893_f);
-                jsonobject.addProperty("depthNoiseScaleZ", p_serialize_1_.field_177894_g);
-                jsonobject.addProperty("depthNoiseScaleExponent", p_serialize_1_.field_177915_h);
-                jsonobject.addProperty("mainNoiseScaleX", p_serialize_1_.field_177917_i);
-                jsonobject.addProperty("mainNoiseScaleY", p_serialize_1_.field_177911_j);
-                jsonobject.addProperty("mainNoiseScaleZ", p_serialize_1_.field_177913_k);
-                jsonobject.addProperty("baseSize", p_serialize_1_.field_177907_l);
-                jsonobject.addProperty("stretchY", p_serialize_1_.field_177909_m);
-                jsonobject.addProperty("biomeDepthWeight", p_serialize_1_.field_177903_n);
-                jsonobject.addProperty("biomeDepthOffset", p_serialize_1_.field_177905_o);
-                jsonobject.addProperty("biomeScaleWeight", p_serialize_1_.field_177933_p);
-                jsonobject.addProperty("biomeScaleOffset", p_serialize_1_.field_177931_q);
-                jsonobject.addProperty("seaLevel", p_serialize_1_.field_177929_r);
-                jsonobject.addProperty("useCaves", p_serialize_1_.field_177927_s);
-                jsonobject.addProperty("useDungeons", p_serialize_1_.field_177925_t);
-                jsonobject.addProperty("dungeonChance", p_serialize_1_.field_177923_u);
-                jsonobject.addProperty("useStrongholds", p_serialize_1_.field_177921_v);
-                jsonobject.addProperty("useVillages", p_serialize_1_.field_177919_w);
-                jsonobject.addProperty("useMineShafts", p_serialize_1_.field_177944_x);
-                jsonobject.addProperty("useTemples", p_serialize_1_.field_177942_y);
-                jsonobject.addProperty("useMonuments", p_serialize_1_.field_177940_z);
-                jsonobject.addProperty("useMansions", p_serialize_1_.field_191076_A);
-                jsonobject.addProperty("useRavines", p_serialize_1_.field_177870_A);
-                jsonobject.addProperty("useWaterLakes", p_serialize_1_.field_177871_B);
-                jsonobject.addProperty("waterLakeChance", p_serialize_1_.field_177872_C);
-                jsonobject.addProperty("useLavaLakes", p_serialize_1_.field_177866_D);
-                jsonobject.addProperty("lavaLakeChance", p_serialize_1_.field_177867_E);
-                jsonobject.addProperty("useLavaOceans", p_serialize_1_.field_177868_F);
-                jsonobject.addProperty("fixedBiome", p_serialize_1_.field_177869_G);
-                jsonobject.addProperty("biomeSize", p_serialize_1_.field_177877_H);
-                jsonobject.addProperty("riverSize", p_serialize_1_.field_177878_I);
-                jsonobject.addProperty("dirtSize", p_serialize_1_.field_177879_J);
-                jsonobject.addProperty("dirtCount", p_serialize_1_.field_177880_K);
-                jsonobject.addProperty("dirtMinHeight", p_serialize_1_.field_177873_L);
-                jsonobject.addProperty("dirtMaxHeight", p_serialize_1_.field_177874_M);
-                jsonobject.addProperty("gravelSize", p_serialize_1_.field_177875_N);
-                jsonobject.addProperty("gravelCount", p_serialize_1_.field_177876_O);
-                jsonobject.addProperty("gravelMinHeight", p_serialize_1_.field_177886_P);
-                jsonobject.addProperty("gravelMaxHeight", p_serialize_1_.field_177885_Q);
-                jsonobject.addProperty("graniteSize", p_serialize_1_.field_177888_R);
-                jsonobject.addProperty("graniteCount", p_serialize_1_.field_177887_S);
-                jsonobject.addProperty("graniteMinHeight", p_serialize_1_.field_177882_T);
-                jsonobject.addProperty("graniteMaxHeight", p_serialize_1_.field_177881_U);
-                jsonobject.addProperty("dioriteSize", p_serialize_1_.field_177884_V);
-                jsonobject.addProperty("dioriteCount", p_serialize_1_.field_177883_W);
-                jsonobject.addProperty("dioriteMinHeight", p_serialize_1_.field_177891_X);
-                jsonobject.addProperty("dioriteMaxHeight", p_serialize_1_.field_177890_Y);
-                jsonobject.addProperty("andesiteSize", p_serialize_1_.field_177892_Z);
-                jsonobject.addProperty("andesiteCount", p_serialize_1_.field_177936_aa);
-                jsonobject.addProperty("andesiteMinHeight", p_serialize_1_.field_177937_ab);
-                jsonobject.addProperty("andesiteMaxHeight", p_serialize_1_.field_177934_ac);
-                jsonobject.addProperty("coalSize", p_serialize_1_.field_177935_ad);
-                jsonobject.addProperty("coalCount", p_serialize_1_.field_177941_ae);
-                jsonobject.addProperty("coalMinHeight", p_serialize_1_.field_177943_af);
-                jsonobject.addProperty("coalMaxHeight", p_serialize_1_.field_177938_ag);
-                jsonobject.addProperty("ironSize", p_serialize_1_.field_177939_ah);
-                jsonobject.addProperty("ironCount", p_serialize_1_.field_177922_ai);
-                jsonobject.addProperty("ironMinHeight", p_serialize_1_.field_177924_aj);
-                jsonobject.addProperty("ironMaxHeight", p_serialize_1_.field_177918_ak);
-                jsonobject.addProperty("goldSize", p_serialize_1_.field_177920_al);
-                jsonobject.addProperty("goldCount", p_serialize_1_.field_177930_am);
-                jsonobject.addProperty("goldMinHeight", p_serialize_1_.field_177932_an);
-                jsonobject.addProperty("goldMaxHeight", p_serialize_1_.field_177926_ao);
-                jsonobject.addProperty("redstoneSize", p_serialize_1_.field_177928_ap);
-                jsonobject.addProperty("redstoneCount", p_serialize_1_.field_177908_aq);
-                jsonobject.addProperty("redstoneMinHeight", p_serialize_1_.field_177906_ar);
-                jsonobject.addProperty("redstoneMaxHeight", p_serialize_1_.field_177904_as);
-                jsonobject.addProperty("diamondSize", p_serialize_1_.field_177902_at);
-                jsonobject.addProperty("diamondCount", p_serialize_1_.field_177916_au);
-                jsonobject.addProperty("diamondMinHeight", p_serialize_1_.field_177914_av);
-                jsonobject.addProperty("diamondMaxHeight", p_serialize_1_.field_177912_aw);
-                jsonobject.addProperty("lapisSize", p_serialize_1_.field_177910_ax);
-                jsonobject.addProperty("lapisCount", p_serialize_1_.field_177897_ay);
-                jsonobject.addProperty("lapisCenterHeight", p_serialize_1_.field_177895_az);
-                jsonobject.addProperty("lapisSpread", p_serialize_1_.field_177889_aA);
+                jsonobject.addProperty("coordinateScale", Float.valueOf(p_serialize_1_.field_177899_b));
+                jsonobject.addProperty("heightScale", Float.valueOf(p_serialize_1_.field_177900_c));
+                jsonobject.addProperty("lowerLimitScale", Float.valueOf(p_serialize_1_.field_177898_e));
+                jsonobject.addProperty("upperLimitScale", Float.valueOf(p_serialize_1_.field_177896_d));
+                jsonobject.addProperty("depthNoiseScaleX", Float.valueOf(p_serialize_1_.field_177893_f));
+                jsonobject.addProperty("depthNoiseScaleZ", Float.valueOf(p_serialize_1_.field_177894_g));
+                jsonobject.addProperty("depthNoiseScaleExponent", Float.valueOf(p_serialize_1_.field_177915_h));
+                jsonobject.addProperty("mainNoiseScaleX", Float.valueOf(p_serialize_1_.field_177917_i));
+                jsonobject.addProperty("mainNoiseScaleY", Float.valueOf(p_serialize_1_.field_177911_j));
+                jsonobject.addProperty("mainNoiseScaleZ", Float.valueOf(p_serialize_1_.field_177913_k));
+                jsonobject.addProperty("baseSize", Float.valueOf(p_serialize_1_.field_177907_l));
+                jsonobject.addProperty("stretchY", Float.valueOf(p_serialize_1_.field_177909_m));
+                jsonobject.addProperty("biomeDepthWeight", Float.valueOf(p_serialize_1_.field_177903_n));
+                jsonobject.addProperty("biomeDepthOffset", Float.valueOf(p_serialize_1_.field_177905_o));
+                jsonobject.addProperty("biomeScaleWeight", Float.valueOf(p_serialize_1_.field_177933_p));
+                jsonobject.addProperty("biomeScaleOffset", Float.valueOf(p_serialize_1_.field_177931_q));
+                jsonobject.addProperty("seaLevel", Integer.valueOf(p_serialize_1_.field_177929_r));
+                jsonobject.addProperty("useCaves", Boolean.valueOf(p_serialize_1_.field_177927_s));
+                jsonobject.addProperty("useDungeons", Boolean.valueOf(p_serialize_1_.field_177925_t));
+                jsonobject.addProperty("dungeonChance", Integer.valueOf(p_serialize_1_.field_177923_u));
+                jsonobject.addProperty("useStrongholds", Boolean.valueOf(p_serialize_1_.field_177921_v));
+                jsonobject.addProperty("useVillages", Boolean.valueOf(p_serialize_1_.field_177919_w));
+                jsonobject.addProperty("useMineShafts", Boolean.valueOf(p_serialize_1_.field_177944_x));
+                jsonobject.addProperty("useTemples", Boolean.valueOf(p_serialize_1_.field_177942_y));
+                jsonobject.addProperty("useMonuments", Boolean.valueOf(p_serialize_1_.field_177940_z));
+                jsonobject.addProperty("useMansions", Boolean.valueOf(p_serialize_1_.field_191076_A));
+                jsonobject.addProperty("useRavines", Boolean.valueOf(p_serialize_1_.field_177870_A));
+                jsonobject.addProperty("useWaterLakes", Boolean.valueOf(p_serialize_1_.field_177871_B));
+                jsonobject.addProperty("waterLakeChance", Integer.valueOf(p_serialize_1_.field_177872_C));
+                jsonobject.addProperty("useLavaLakes", Boolean.valueOf(p_serialize_1_.field_177866_D));
+                jsonobject.addProperty("lavaLakeChance", Integer.valueOf(p_serialize_1_.field_177867_E));
+                jsonobject.addProperty("useLavaOceans", Boolean.valueOf(p_serialize_1_.field_177868_F));
+                jsonobject.addProperty("fixedBiome", Integer.valueOf(p_serialize_1_.field_177869_G));
+                jsonobject.addProperty("biomeSize", Integer.valueOf(p_serialize_1_.field_177877_H));
+                jsonobject.addProperty("riverSize", Integer.valueOf(p_serialize_1_.field_177878_I));
+                jsonobject.addProperty("dirtSize", Integer.valueOf(p_serialize_1_.field_177879_J));
+                jsonobject.addProperty("dirtCount", Integer.valueOf(p_serialize_1_.field_177880_K));
+                jsonobject.addProperty("dirtMinHeight", Integer.valueOf(p_serialize_1_.field_177873_L));
+                jsonobject.addProperty("dirtMaxHeight", Integer.valueOf(p_serialize_1_.field_177874_M));
+                jsonobject.addProperty("gravelSize", Integer.valueOf(p_serialize_1_.field_177875_N));
+                jsonobject.addProperty("gravelCount", Integer.valueOf(p_serialize_1_.field_177876_O));
+                jsonobject.addProperty("gravelMinHeight", Integer.valueOf(p_serialize_1_.field_177886_P));
+                jsonobject.addProperty("gravelMaxHeight", Integer.valueOf(p_serialize_1_.field_177885_Q));
+                jsonobject.addProperty("graniteSize", Integer.valueOf(p_serialize_1_.field_177888_R));
+                jsonobject.addProperty("graniteCount", Integer.valueOf(p_serialize_1_.field_177887_S));
+                jsonobject.addProperty("graniteMinHeight", Integer.valueOf(p_serialize_1_.field_177882_T));
+                jsonobject.addProperty("graniteMaxHeight", Integer.valueOf(p_serialize_1_.field_177881_U));
+                jsonobject.addProperty("dioriteSize", Integer.valueOf(p_serialize_1_.field_177884_V));
+                jsonobject.addProperty("dioriteCount", Integer.valueOf(p_serialize_1_.field_177883_W));
+                jsonobject.addProperty("dioriteMinHeight", Integer.valueOf(p_serialize_1_.field_177891_X));
+                jsonobject.addProperty("dioriteMaxHeight", Integer.valueOf(p_serialize_1_.field_177890_Y));
+                jsonobject.addProperty("andesiteSize", Integer.valueOf(p_serialize_1_.field_177892_Z));
+                jsonobject.addProperty("andesiteCount", Integer.valueOf(p_serialize_1_.field_177936_aa));
+                jsonobject.addProperty("andesiteMinHeight", Integer.valueOf(p_serialize_1_.field_177937_ab));
+                jsonobject.addProperty("andesiteMaxHeight", Integer.valueOf(p_serialize_1_.field_177934_ac));
+                jsonobject.addProperty("coalSize", Integer.valueOf(p_serialize_1_.field_177935_ad));
+                jsonobject.addProperty("coalCount", Integer.valueOf(p_serialize_1_.field_177941_ae));
+                jsonobject.addProperty("coalMinHeight", Integer.valueOf(p_serialize_1_.field_177943_af));
+                jsonobject.addProperty("coalMaxHeight", Integer.valueOf(p_serialize_1_.field_177938_ag));
+                jsonobject.addProperty("ironSize", Integer.valueOf(p_serialize_1_.field_177939_ah));
+                jsonobject.addProperty("ironCount", Integer.valueOf(p_serialize_1_.field_177922_ai));
+                jsonobject.addProperty("ironMinHeight", Integer.valueOf(p_serialize_1_.field_177924_aj));
+                jsonobject.addProperty("ironMaxHeight", Integer.valueOf(p_serialize_1_.field_177918_ak));
+                jsonobject.addProperty("goldSize", Integer.valueOf(p_serialize_1_.field_177920_al));
+                jsonobject.addProperty("goldCount", Integer.valueOf(p_serialize_1_.field_177930_am));
+                jsonobject.addProperty("goldMinHeight", Integer.valueOf(p_serialize_1_.field_177932_an));
+                jsonobject.addProperty("goldMaxHeight", Integer.valueOf(p_serialize_1_.field_177926_ao));
+                jsonobject.addProperty("redstoneSize", Integer.valueOf(p_serialize_1_.field_177928_ap));
+                jsonobject.addProperty("redstoneCount", Integer.valueOf(p_serialize_1_.field_177908_aq));
+                jsonobject.addProperty("redstoneMinHeight", Integer.valueOf(p_serialize_1_.field_177906_ar));
+                jsonobject.addProperty("redstoneMaxHeight", Integer.valueOf(p_serialize_1_.field_177904_as));
+                jsonobject.addProperty("diamondSize", Integer.valueOf(p_serialize_1_.field_177902_at));
+                jsonobject.addProperty("diamondCount", Integer.valueOf(p_serialize_1_.field_177916_au));
+                jsonobject.addProperty("diamondMinHeight", Integer.valueOf(p_serialize_1_.field_177914_av));
+                jsonobject.addProperty("diamondMaxHeight", Integer.valueOf(p_serialize_1_.field_177912_aw));
+                jsonobject.addProperty("lapisSize", Integer.valueOf(p_serialize_1_.field_177910_ax));
+                jsonobject.addProperty("lapisCount", Integer.valueOf(p_serialize_1_.field_177897_ay));
+                jsonobject.addProperty("lapisCenterHeight", Integer.valueOf(p_serialize_1_.field_177895_az));
+                jsonobject.addProperty("lapisSpread", Integer.valueOf(p_serialize_1_.field_177889_aA));
                 return jsonobject;
             }
         }
