--- before/net/minecraft/world/gen/ChunkProviderServer.java
+++ after/net/minecraft/world/gen/ChunkProviderServer.java
@@ -34,6 +34,7 @@
     public final IChunkLoader chunkLoader;
     public final Long2ObjectMap<Chunk> loadedChunks = new Long2ObjectOpenHashMap<>(8192);
     public final WorldServer world;
+    private final Set<Long> loadingChunks = com.google.common.collect.Sets.newHashSet();
 
     public ChunkProviderServer(WorldServer worldObjIn, IChunkLoader chunkLoaderIn, IChunkGenerator chunkGeneratorIn)
     {
@@ -82,20 +83,49 @@
     @Nullable
     public Chunk loadChunk(int x, int z)
     {
+        return loadChunk(x, z, null);
+    }
+
+    @Nullable
+    public Chunk loadChunk(int x, int z, @Nullable Runnable runnable)
+    {
         Chunk chunk = this.getLoadedChunk(x, z);
 
         if (chunk == null)
         {
-            chunk = this.loadChunkFromFile(x, z);
+            long pos = ChunkPos.asLong(x, z);
 
-            if (chunk != null)
+            chunk = net.minecraftforge.common.ForgeChunkManager.fetchDormantChunk(pos, this.world);
+            if (chunk != null || !(this.chunkLoader instanceof net.minecraft.world.chunk.storage.AnvilChunkLoader))
             {
+                if (!loadingChunks.add(pos)) net.minecraftforge.fml.common.FMLLog.bigWarning("There is an attempt to load a chunk ({},{}) in dimension {} that is already being loaded. This will cause weird chunk breakages.", x, z, this.world.provider.getDimension());
+                if (chunk == null) chunk = this.loadChunkFromFile(x, z);
+
+                if (chunk != null)
+                {
                 this.loadedChunks.put(ChunkPos.asLong(x, z), chunk);
                 chunk.onLoad();
                 chunk.populate(this, this.chunkGenerator);
+                }
+
+                loadingChunks.remove(pos);
+            }
+            else
+            {
+                net.minecraft.world.chunk.storage.AnvilChunkLoader loader = (net.minecraft.world.chunk.storage.AnvilChunkLoader) this.chunkLoader;
+                if (runnable == null || !net.minecraftforge.common.ForgeChunkManager.asyncChunkLoading)
+                    chunk = net.minecraftforge.common.chunkio.ChunkIOExecutor.syncChunkLoad(this.world, loader, this, x, z);
+                else if (loader.isChunkGeneratedAt(x, z))
+                {
+                    // We can only use the async queue for already generated chunks
+                    net.minecraftforge.common.chunkio.ChunkIOExecutor.queueChunkLoad(this.world, loader, this, x, z, runnable);
+                    return null;
+                }
             }
         }
 
+        // If we didn't load the chunk async and have a callback run it now
+        if (runnable != null) runnable.run();
         return chunk;
     }
 
@@ -222,6 +252,11 @@
         {
             if (!this.droppedChunks.isEmpty())
             {
+                for (ChunkPos forced : this.world.getPersistentChunks().keySet())
+                {
+                    this.droppedChunks.remove(ChunkPos.asLong(forced.x, forced.z));
+                }
+
                 Iterator<Long> iterator = this.droppedChunks.iterator();
 
                 for (int i = 0; i < 100 && iterator.hasNext(); iterator.remove())
@@ -232,6 +267,7 @@
                     if (chunk != null && chunk.unloadQueued)
                     {
                         chunk.onUnload();
+                        net.minecraftforge.common.ForgeChunkManager.putDormantChunk(ChunkPos.asLong(chunk.x, chunk.z), chunk);
                         this.saveChunkData(chunk);
                         this.saveChunkExtraData(chunk);
                         this.loadedChunks.remove(olong);
@@ -239,6 +275,8 @@
                     }
                 }
             }
+
+            if (this.loadedChunks.isEmpty()) net.minecraftforge.common.DimensionManager.unloadWorld(this.world.provider.getDimension());
 
             this.chunkLoader.chunkTick();
         }
