--- before/net/minecraft/world/gen/FlatLayerInfo.java
+++ after/net/minecraft/world/gen/FlatLayerInfo.java
@@ -8,7 +8,7 @@
 {
     private final int field_175902_a;
     private IBlockState field_175901_b;
-    private int field_82664_a = 1;
+    private int field_82664_a;
     private int field_82661_d;
 
     public FlatLayerInfo(int p_i45467_1_, Block p_i45467_2_)
@@ -18,6 +18,7 @@
 
     public FlatLayerInfo(int p_i45627_1_, int p_i45627_2_, Block p_i45627_3_)
     {
+        this.field_82664_a = 1;
         this.field_175902_a = p_i45627_1_;
         this.field_82664_a = p_i45627_2_;
         this.field_175901_b = p_i45627_3_.func_176223_P();
