--- before/net/minecraft/world/gen/NoiseGeneratorImproved.java
+++ after/net/minecraft/world/gen/NoiseGeneratorImproved.java
@@ -4,7 +4,7 @@
 
 public class NoiseGeneratorImproved extends NoiseGenerator
 {
-    private final int[] field_76312_d = new int[512];
+    private final int[] field_76312_d;
     public double field_76315_a;
     public double field_76313_b;
     public double field_76314_c;
@@ -21,12 +21,14 @@
 
     public NoiseGeneratorImproved(Random p_i45469_1_)
     {
+        this.field_76312_d = new int[512];
         this.field_76315_a = p_i45469_1_.nextDouble() * 256.0D;
         this.field_76313_b = p_i45469_1_.nextDouble() * 256.0D;
         this.field_76314_c = p_i45469_1_.nextDouble() * 256.0D;
 
         for (int i = 0; i < 256; this.field_76312_d[i] = i++)
         {
+            ;
         }
 
         for (int l = 0; l < 256; ++l)
@@ -80,7 +82,7 @@
                 }
 
                 int k2 = i6 & 255;
-                d17 -= (double)i6;
+                d17 = d17 - (double)i6;
                 double d18 = d17 * d17 * d17 * (d17 * (d17 * 6.0D - 15.0D) + 10.0D);
 
                 for (int j6 = 0; j6 < p_76308_10_; ++j6)
@@ -94,7 +96,7 @@
                     }
 
                     int l6 = k6 & 255;
-                    d19 -= (double)k6;
+                    d19 = d19 - (double)k6;
                     double d20 = d19 * d19 * d19 * (d19 * (d19 * 6.0D - 15.0D) + 10.0D);
                     i5 = this.field_76312_d[k2] + 0;
                     j5 = this.field_76312_d[i5] + l6;
@@ -135,7 +137,7 @@
                 }
 
                 int j3 = i3 & 255;
-                d5 -= (double)i3;
+                d5 = d5 - (double)i3;
                 double d6 = d5 * d5 * d5 * (d5 * (d5 * 6.0D - 15.0D) + 10.0D);
 
                 for (int k3 = 0; k3 < p_76308_10_; ++k3)
@@ -149,7 +151,7 @@
                     }
 
                     int i4 = l3 & 255;
-                    d7 -= (double)l3;
+                    d7 = d7 - (double)l3;
                     double d8 = d7 * d7 * d7 * (d7 * (d7 * 6.0D - 15.0D) + 10.0D);
 
                     for (int j4 = 0; j4 < p_76308_9_; ++j4)
@@ -163,7 +165,7 @@
                         }
 
                         int l4 = k4 & 255;
-                        d9 -= (double)k4;
+                        d9 = d9 - (double)k4;
                         double d10 = d9 * d9 * d9 * (d9 * (d9 * 6.0D - 15.0D) + 10.0D);
 
                         if (j4 == 0 || l4 != k)
