--- before/net/minecraft/world/gen/NoiseGeneratorOctaves.java
+++ after/net/minecraft/world/gen/NoiseGeneratorOctaves.java
@@ -42,12 +42,12 @@
             double d2 = (double)p_76304_4_ * d3 * p_76304_12_;
             long k = MathHelper.func_76124_d(d0);
             long l = MathHelper.func_76124_d(d2);
-            d0 -= (double)k;
-            d2 -= (double)l;
-            k %= 16777216L;
-            l %= 16777216L;
-            d0 += (double)k;
-            d2 += (double)l;
+            d0 = d0 - (double)k;
+            d2 = d2 - (double)l;
+            k = k % 16777216L;
+            l = l % 16777216L;
+            d0 = d0 + (double)k;
+            d2 = d2 + (double)l;
             this.field_76307_a[j].func_76308_a(p_76304_1_, d0, d1, d2, p_76304_5_, p_76304_6_, p_76304_7_, p_76304_8_ * d3, p_76304_10_ * d3, p_76304_12_ * d3, d3);
             d3 /= 2.0D;
         }
