--- before/net/minecraft/command/server/CommandWhitelist.java
+++ after/net/minecraft/command/server/CommandWhitelist.java
@@ -15,30 +15,26 @@
 
 public class CommandWhitelist extends CommandBase
 {
-    @Override
     public String func_71517_b()
     {
         return "whitelist";
     }
 
-    @Override
     public int func_82362_a()
     {
         return 3;
     }
 
-    @Override
     public String func_71518_a(ICommandSender p_71518_1_)
     {
         return "commands.whitelist.usage";
     }
 
-    @Override
     public void func_184881_a(MinecraftServer p_184881_1_, ICommandSender p_184881_2_, String[] p_184881_3_) throws CommandException
     {
         if (p_184881_3_.length < 1)
         {
-            throw new WrongUsageException("commands.whitelist.usage");
+            throw new WrongUsageException("commands.whitelist.usage", new Object[0]);
         }
         else
         {
@@ -54,11 +50,7 @@
             }
             else if ("list".equals(p_184881_3_[0]))
             {
-                p_184881_2_.func_145747_a(
-                    new TextComponentTranslation(
-                        "commands.whitelist.list", p_184881_1_.func_184103_al().func_152598_l().length, p_184881_1_.func_184103_al().func_72373_m().length
-                    )
-                );
+                p_184881_2_.func_145747_a(new TextComponentTranslation("commands.whitelist.list", new Object[] {p_184881_1_.func_184103_al().func_152598_l().length, p_184881_1_.func_184103_al().func_72373_m().length}));
                 String[] astring = p_184881_1_.func_184103_al().func_152598_l();
                 p_184881_2_.func_145747_a(new TextComponentString(func_71527_a(astring)));
             }
@@ -66,14 +58,14 @@
             {
                 if (p_184881_3_.length < 2)
                 {
-                    throw new WrongUsageException("commands.whitelist.add.usage");
+                    throw new WrongUsageException("commands.whitelist.add.usage", new Object[0]);
                 }
 
                 GameProfile gameprofile = p_184881_1_.func_152358_ax().func_152655_a(p_184881_3_[1]);
 
                 if (gameprofile == null)
                 {
-                    throw new CommandException("commands.whitelist.add.failed", p_184881_3_[1]);
+                    throw new CommandException("commands.whitelist.add.failed", new Object[] {p_184881_3_[1]});
                 }
 
                 p_184881_1_.func_184103_al().func_152601_d(gameprofile);
@@ -83,14 +75,14 @@
             {
                 if (p_184881_3_.length < 2)
                 {
-                    throw new WrongUsageException("commands.whitelist.remove.usage");
+                    throw new WrongUsageException("commands.whitelist.remove.usage", new Object[0]);
                 }
 
                 GameProfile gameprofile1 = p_184881_1_.func_184103_al().func_152599_k().func_152706_a(p_184881_3_[1]);
 
                 if (gameprofile1 == null)
                 {
-                    throw new CommandException("commands.whitelist.remove.failed", p_184881_3_[1]);
+                    throw new CommandException("commands.whitelist.remove.failed", new Object[] {p_184881_3_[1]});
                 }
 
                 p_184881_1_.func_184103_al().func_152597_c(gameprofile1);
@@ -104,7 +96,6 @@
         }
     }
 
-    @Override
     public List<String> func_184883_a(MinecraftServer p_184883_1_, ICommandSender p_184883_2_, String[] p_184883_3_, @Nullable BlockPos p_184883_4_)
     {
         if (p_184883_3_.length == 1)
@@ -126,7 +117,7 @@
                 }
             }
 
-            return Collections.emptyList();
+            return Collections.<String>emptyList();
         }
     }
 }
