--- before/net/minecraft/command/server/CommandPardonIp.java
+++ after/net/minecraft/command/server/CommandPardonIp.java
@@ -47,12 +47,12 @@
             }
             else
             {
-                throw new SyntaxErrorException("commands.unbanip.invalid");
+                throw new SyntaxErrorException("commands.unbanip.invalid", new Object[0]);
             }
         }
         else
         {
-            throw new WrongUsageException("commands.unbanip.usage");
+            throw new WrongUsageException("commands.unbanip.usage", new Object[0]);
         }
     }
 
