--- before/net/minecraft/command/server/CommandTestForBlock.java
+++ after/net/minecraft/command/server/CommandTestForBlock.java
@@ -41,7 +41,7 @@
     {
         if (p_184881_3_.length < 4)
         {
-            throw new WrongUsageException("commands.testforblock.usage");
+            throw new WrongUsageException("commands.testforblock.usage", new Object[0]);
         }
         else
         {
@@ -51,7 +51,7 @@
 
             if (block == null)
             {
-                throw new NumberInvalidException("commands.setblock.notFound", p_184881_3_[3]);
+                throw new NumberInvalidException("commands.setblock.notFound", new Object[] {p_184881_3_[3]});
             }
             else
             {
@@ -59,7 +59,7 @@
 
                 if (!world.func_175667_e(blockpos))
                 {
-                    throw new CommandException("commands.testforblock.outOfWorld");
+                    throw new CommandException("commands.testforblock.outOfWorld", new Object[0]);
                 }
                 else
                 {
@@ -77,7 +77,7 @@
                         }
                         catch (NBTException nbtexception)
                         {
-                            throw new CommandException("commands.setblock.tagError", nbtexception.getMessage());
+                            throw new CommandException("commands.setblock.tagError", new Object[] {nbtexception.getMessage()});
                         }
                     }
 
@@ -86,18 +86,18 @@
 
                     if (block1 != block)
                     {
-                        throw new CommandException("commands.testforblock.failed.tile", blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p(), block1.func_149732_F(), block.func_149732_F());
+                        throw new CommandException("commands.testforblock.failed.tile", new Object[] {blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p(), block1.func_149732_F(), block.func_149732_F()});
                     }
                     else if (p_184881_3_.length >= 5 && !CommandBase.func_190791_b(block, p_184881_3_[4]).apply(iblockstate))
                     {
                         try
                         {
                             int i = iblockstate.func_177230_c().func_176201_c(iblockstate);
-                            throw new CommandException("commands.testforblock.failed.data", blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p(), i, Integer.parseInt(p_184881_3_[4]));
+                            throw new CommandException("commands.testforblock.failed.data", new Object[] {blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p(), i, Integer.parseInt(p_184881_3_[4])});
                         }
-                        catch (NumberFormatException numberformatexception)
+                        catch (NumberFormatException var13)
                         {
-                            throw new CommandException("commands.testforblock.failed.data", blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p(), iblockstate.toString(), p_184881_3_[4]);
+                            throw new CommandException("commands.testforblock.failed.data", new Object[] {blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p(), iblockstate.toString(), p_184881_3_[4]});
                         }
                     }
                     else
@@ -108,14 +108,14 @@
 
                             if (tileentity == null)
                             {
-                                throw new CommandException("commands.testforblock.failed.tileEntity", blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p());
+                                throw new CommandException("commands.testforblock.failed.tileEntity", new Object[] {blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p()});
                             }
 
                             NBTTagCompound nbttagcompound1 = tileentity.func_189515_b(new NBTTagCompound());
 
                             if (!NBTUtil.func_181123_a(nbttagcompound, nbttagcompound1, true))
                             {
-                                throw new CommandException("commands.testforblock.failed.nbt", blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p());
+                                throw new CommandException("commands.testforblock.failed.nbt", new Object[] {blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p()});
                             }
                         }
 
