--- before/net/minecraft/command/server/CommandWhitelist.java
+++ after/net/minecraft/command/server/CommandWhitelist.java
@@ -15,30 +15,26 @@
 
 public class CommandWhitelist extends CommandBase
 {
-    @Override
     public String getName()
     {
         return "whitelist";
     }
 
-    @Override
     public int getRequiredPermissionLevel()
     {
         return 3;
     }
 
-    @Override
     public String getUsage(ICommandSender sender)
     {
         return "commands.whitelist.usage";
     }
 
-    @Override
     public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException
     {
         if (args.length < 1)
         {
-            throw new WrongUsageException("commands.whitelist.usage");
+            throw new WrongUsageException("commands.whitelist.usage", new Object[0]);
         }
         else
         {
@@ -54,11 +50,7 @@
             }
             else if ("list".equals(args[0]))
             {
-                sender.sendMessage(
-                    new TextComponentTranslation(
-                        "commands.whitelist.list", server.getPlayerList().getWhitelistedPlayerNames().length, server.getPlayerList().getAvailablePlayerDat().length
-                    )
-                );
+                sender.sendMessage(new TextComponentTranslation("commands.whitelist.list", new Object[] {server.getPlayerList().getWhitelistedPlayerNames().length, server.getPlayerList().getAvailablePlayerDat().length}));
                 String[] astring = server.getPlayerList().getWhitelistedPlayerNames();
                 sender.sendMessage(new TextComponentString(joinNiceString(astring)));
             }
@@ -66,14 +58,14 @@
             {
                 if (args.length < 2)
                 {
-                    throw new WrongUsageException("commands.whitelist.add.usage");
+                    throw new WrongUsageException("commands.whitelist.add.usage", new Object[0]);
                 }
 
                 GameProfile gameprofile = server.getPlayerProfileCache().getGameProfileForUsername(args[1]);
 
                 if (gameprofile == null)
                 {
-                    throw new CommandException("commands.whitelist.add.failed", args[1]);
+                    throw new CommandException("commands.whitelist.add.failed", new Object[] {args[1]});
                 }
 
                 server.getPlayerList().addWhitelistedPlayer(gameprofile);
@@ -83,14 +75,14 @@
             {
                 if (args.length < 2)
                 {
-                    throw new WrongUsageException("commands.whitelist.remove.usage");
+                    throw new WrongUsageException("commands.whitelist.remove.usage", new Object[0]);
                 }
 
                 GameProfile gameprofile1 = server.getPlayerList().getWhitelistedPlayers().getByName(args[1]);
 
                 if (gameprofile1 == null)
                 {
-                    throw new CommandException("commands.whitelist.remove.failed", args[1]);
+                    throw new CommandException("commands.whitelist.remove.failed", new Object[] {args[1]});
                 }
 
                 server.getPlayerList().removePlayerFromWhitelist(gameprofile1);
@@ -104,7 +96,6 @@
         }
     }
 
-    @Override
     public List<String> getTabCompletions(MinecraftServer server, ICommandSender sender, String[] args, @Nullable BlockPos targetPos)
     {
         if (args.length == 1)
@@ -126,7 +117,7 @@
                 }
             }
 
-            return Collections.emptyList();
+            return Collections.<String>emptyList();
         }
     }
 }
