--- before/net/minecraft/command/server/CommandSetBlock.java
+++ after/net/minecraft/command/server/CommandSetBlock.java
@@ -74,7 +74,7 @@
                 NBTTagCompound nbttagcompound = new NBTTagCompound();
                 boolean flag = false;
 
-                if (args.length >= 7 && block.hasTileEntity())
+                if (args.length >= 7 && block.hasTileEntity(iblockstate))
                 {
                     String s = buildString(args, 6);
 
