--- before/net/minecraft/command/server/CommandBroadcast.java
+++ after/net/minecraft/command/server/CommandBroadcast.java
@@ -34,11 +34,11 @@
         if (p_184881_3_.length > 0 && p_184881_3_[0].length() > 0)
         {
             ITextComponent itextcomponent = func_147176_a(p_184881_2_, p_184881_3_, 0, true);
-            p_184881_1_.func_184103_al().func_148539_a(new TextComponentTranslation("chat.type.announcement", p_184881_2_.func_145748_c_(), itextcomponent));
+            p_184881_1_.func_184103_al().func_148539_a(new TextComponentTranslation("chat.type.announcement", new Object[] {p_184881_2_.func_145748_c_(), itextcomponent}));
         }
         else
         {
-            throw new WrongUsageException("commands.say.usage");
+            throw new WrongUsageException("commands.say.usage", new Object[0]);
         }
     }
 
