--- before/net/minecraft/command/server/CommandDeOp.java
+++ after/net/minecraft/command/server/CommandDeOp.java
@@ -36,7 +36,7 @@
 
             if (gameprofile == null)
             {
-                throw new CommandException("commands.deop.failed", p_184881_3_[0]);
+                throw new CommandException("commands.deop.failed", new Object[] {p_184881_3_[0]});
             }
             else
             {
@@ -46,7 +46,7 @@
         }
         else
         {
-            throw new WrongUsageException("commands.deop.usage");
+            throw new WrongUsageException("commands.deop.usage", new Object[0]);
         }
     }
 
