--- before/net/minecraft/command/server/CommandOp.java
+++ after/net/minecraft/command/server/CommandOp.java
@@ -37,7 +37,7 @@
 
             if (gameprofile == null)
             {
-                throw new CommandException("commands.op.failed", p_184881_3_[0]);
+                throw new CommandException("commands.op.failed", new Object[] {p_184881_3_[0]});
             }
             else
             {
@@ -47,7 +47,7 @@
         }
         else
         {
-            throw new WrongUsageException("commands.op.usage");
+            throw new WrongUsageException("commands.op.usage", new Object[0]);
         }
     }
 
@@ -56,7 +56,7 @@
         if (p_184883_3_.length == 1)
         {
             String s = p_184883_3_[p_184883_3_.length - 1];
-            List<String> list = Lists.newArrayList();
+            List<String> list = Lists.<String>newArrayList();
 
             for (GameProfile gameprofile : p_184883_1_.func_152357_F())
             {
@@ -70,7 +70,7 @@
         }
         else
         {
-            return Collections.emptyList();
+            return Collections.<String>emptyList();
         }
     }
 }
