--- before/net/minecraft/server/DebugLoggingPrintStream.java
+++ after/net/minecraft/server/DebugLoggingPrintStream.java
@@ -14,6 +14,6 @@
     {
         StackTraceElement[] astacktraceelement = Thread.currentThread().getStackTrace();
         StackTraceElement stacktraceelement = astacktraceelement[Math.min(3, astacktraceelement.length)];
-        field_179884_a.info("[{}]@.({}:{}): {}", this.field_179883_b, stacktraceelement.getFileName(), stacktraceelement.getLineNumber(), p_179882_1_);
+        field_179884_a.info("[{}]@.({}:{}): {}", this.field_179883_b, stacktraceelement.getFileName(), Integer.valueOf(stacktraceelement.getLineNumber()), p_179882_1_);
     }
 }
