--- before/net/minecraft/server/MinecraftServer.java
+++ after/net/minecraft/server/MinecraftServer.java
@@ -103,7 +103,7 @@
     @SideOnly(Side.SERVER)
     private String hostname;
     private int serverPort = -1;
-    public WorldServer[] worlds;
+    public WorldServer[] worlds = new WorldServer[0];
     private PlayerList playerList;
     private boolean serverRunning = true;
     private boolean serverStopped;
@@ -121,7 +121,8 @@
     private int buildLimit;
     private int maxPlayerIdleMinutes;
     public final long[] tickTimeArray = new long[100];
-    public long[][] timeOfLastDimensionTick;
+    //public long[][] timeOfLastDimensionTick;
+    public java.util.Hashtable<Integer, long[]> worldTickTimes = new java.util.Hashtable<>();
     private KeyPair serverKeyPair;
     private String serverOwner;
     private String folderName;
@@ -184,7 +185,7 @@
             this.setUserMessage("menu.convertingLevel");
             this.getActiveAnvilConverter().convertMapFormat(worldNameIn, new IProgressUpdate()
             {
-                private long startTime = MinecraftServer.getCurrentTimeMillis();
+                private long startTime = System.currentTimeMillis();
                 @Override
                 public void displaySavingString(String message)
                 {
@@ -197,9 +198,9 @@
                 @Override
                 public void setLoadingProgress(int progress)
                 {
-                    if (MinecraftServer.getCurrentTimeMillis() - this.startTime >= 1000L)
+                    if (System.currentTimeMillis() - this.startTime >= 1000L)
                     {
-                        this.startTime = MinecraftServer.getCurrentTimeMillis();
+                        this.startTime = System.currentTimeMillis();
                         MinecraftServer.LOGGER.info("Converting... {}%", progress);
                     }
                 }
@@ -233,8 +234,6 @@
     {
         this.convertMapIfNeeded(saveName);
         this.setUserMessage("menu.loadingLevel");
-        this.worlds = new WorldServer[3];
-        this.timeOfLastDimensionTick = new long[this.worlds.length][100];
         ISaveHandler isavehandler = this.anvilConverterForAnvilFile.getSaveLoader(saveName, true);
         this.setResourcePackFromWorld(this.getFolderName(), isavehandler);
         WorldInfo worldinfo = isavehandler.loadWorldInfo();
@@ -265,6 +264,7 @@
             worldsettings = new WorldSettings(worldinfo);
         }
 
+        if (false) { //Forge Dead code, reimplemented below
         for (int i = 0; i < this.worlds.length; i++)
         {
             int j = 0;
@@ -304,8 +304,23 @@
                 this.worlds[i].getWorldInfo().setGameType(this.getGameType());
             }
         }
-
-        this.playerList.setPlayerManager(this.worlds);
+        } //Forge: End dead code
+
+        WorldServer overWorld = (WorldServer)(isDemo() ? new WorldServerDemo(this, isavehandler, worldinfo, 0, profiler).init() : new WorldServer(this, isavehandler, worldinfo, 0, profiler).init());
+        overWorld.initialize(worldsettings);
+        for (int dim : net.minecraftforge.common.DimensionManager.getStaticDimensionIDs())
+        {
+            WorldServer world = (dim == 0 ? overWorld : (WorldServer)new WorldServerMulti(this, isavehandler, dim, overWorld, profiler).init());
+            world.addEventListener(new ServerWorldEventHandler(this, world));
+
+            if (!this.isSinglePlayer())
+            {
+                world.getWorldInfo().setGameType(this.getGameType());
+            }
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Load(world));
+        }
+
+        this.playerList.setPlayerManager(new WorldServer[]{ overWorld });
         this.setDifficultyForAllWorlds(this.getDifficulty());
         this.initialWorldChunkLoad();
     }
@@ -320,7 +335,7 @@
         this.setUserMessage("menu.generatingTerrain");
         int j1 = 0;
         LOGGER.info("Preparing start region for level 0");
-        WorldServer worldserver = this.worlds[0];
+        WorldServer worldserver = net.minecraftforge.common.DimensionManager.getWorld(j1);
         BlockPos blockpos = worldserver.getSpawnPoint();
         long k1 = getCurrentTimeMillis();
 
@@ -447,15 +462,24 @@
             {
                 if (worldserver1 != null)
                 {
+                    net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Unload(worldserver1));
                     worldserver1.flush();
                 }
             }
+
+            WorldServer[] tmp = worlds;
+            for (WorldServer world : tmp)
+            {
+                net.minecraftforge.common.DimensionManager.setWorld(world.provider.getDimension(), null, this);
+            }
         }
 
         if (this.usageSnooper.isSnooperRunning())
         {
             this.usageSnooper.stopSnooper();
         }
+
+        CommandBase.setCommandListener(null); // Forge: fix MC-128561
     }
 
     @SideOnly(Side.SERVER)
@@ -487,6 +511,7 @@
         {
             if (this.init())
             {
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerStarted();
                 this.currentTime = getCurrentTimeMillis();
                 long i = 0L;
                 this.statusResponse.setServerDescription(new TextComponentString(this.motd));
@@ -533,12 +558,20 @@
                     Thread.sleep(Math.max(1L, 50L - i));
                     this.serverIsRunning = true;
                 }
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerStopping();
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().expectServerStopped(); // has to come before finalTick to avoid race conditions
             }
             else
             {
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().expectServerStopped(); // has to come before finalTick to avoid race conditions
                 this.finalTick(null);
             }
         }
+        catch (net.minecraftforge.fml.common.StartupQuery.AbortedException e)
+        {
+            // ignore silently
+            net.minecraftforge.fml.common.FMLCommonHandler.instance().expectServerStopped(); // has to come before finalTick to avoid race conditions
+        }
         catch (Throwable throwable1)
         {
             LOGGER.error("Encountered an unexpected exception", throwable1);
@@ -566,13 +599,13 @@
                 LOGGER.error("We were unable to save this crash report to disk.");
             }
 
+            net.minecraftforge.fml.common.FMLCommonHandler.instance().expectServerStopped(); // has to come before finalTick to avoid race conditions
             this.finalTick(crashreport);
         }
         finally
         {
             try
             {
-                this.serverStopped = true;
                 this.stopServer();
             }
             catch (Throwable throwable)
@@ -581,6 +614,8 @@
             }
             finally
             {
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerStopped();
+                this.serverStopped = true;
                 this.systemExitNow();
             }
         }
@@ -607,6 +642,7 @@
                 ImageIO.write(bufferedimage, "PNG", new ByteBufOutputStream(bytebuf));
                 ByteBuf bytebuf1 = Base64.encode(bytebuf);
                 response.setFavicon("data:image/png;base64," + bytebuf1.toString(StandardCharsets.UTF_8));
+                bytebuf1.release(); // Forge: fix MC-122085
             }
             catch (Exception exception)
             {
@@ -648,6 +684,7 @@
     public void tick()
     {
         long i = System.nanoTime();
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().onPreServerTick();
         this.tickCounter++;
 
         if (this.startProfiling)
@@ -674,6 +711,7 @@
 
             Collections.shuffle(Arrays.asList(agameprofile));
             this.statusResponse.getPlayers().setPlayers(agameprofile);
+            this.statusResponse.invalidateJson();
         }
 
         if (this.tickCounter % 900 == 0)
@@ -701,6 +739,7 @@
 
         this.profiler.endSection();
         this.profiler.endSection();
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().onPostServerTick();
     }
 
     public void updateTimeLightAndEntities()
@@ -716,28 +755,31 @@
         }
 
         this.profiler.endStartSection("levels");
+        net.minecraftforge.common.chunkio.ChunkIOExecutor.tick();
 
-        for (int j = 0; j < this.worlds.length; j++)
+        Integer[] ids = net.minecraftforge.common.DimensionManager.getIDs(this.tickCounter % 200 == 0);
+        for (int x = 0; x < ids.length; x++)
         {
+            int id = ids[x];
             long i = System.nanoTime();
 
-            if (j == 0 || this.getAllowNether())
+            if (id == 0 || this.getAllowNether())
             {
-                WorldServer worldserver = this.worlds[j];
+                WorldServer worldserver = net.minecraftforge.common.DimensionManager.getWorld(id);
                 this.profiler.func_194340_a(() -> worldserver.getWorldInfo().getWorldName());
 
                 if (this.tickCounter % 20 == 0)
                 {
                     this.profiler.startSection("timeSync");
-                    this.playerList
-                    .sendPacketToAllPlayersInDimension(
-                        new SPacketTimeUpdate(worldserver.getTotalWorldTime(), worldserver.getWorldTime(), worldserver.getGameRules().getBoolean("doDaylightCycle")),
-                        worldserver.provider.getDimensionType().getId()
+                    this.playerList.sendPacketToAllPlayersInDimension(
+                        new SPacketTimeUpdate(worldserver.getTotalWorldTime(), worldserver.getWorldTime(),
+                            worldserver.getGameRules().getBoolean("doDaylightCycle")), worldserver.provider.getDimension()
                     );
                     this.profiler.endSection();
                 }
 
                 this.profiler.startSection("tick");
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().onPreWorldTick(worldserver);
 
                 try
                 {
@@ -761,6 +803,7 @@
                     throw new ReportedException(crashreport1);
                 }
 
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().onPostWorldTick(worldserver);
                 this.profiler.endSection();
                 this.profiler.startSection("tracker");
                 worldserver.getEntityTracker().tick();
@@ -768,9 +811,11 @@
                 this.profiler.endSection();
             }
 
-            this.timeOfLastDimensionTick[j][this.tickCounter % 100] = System.nanoTime() - i;
+            worldTickTimes.get(id)[this.tickCounter % 100] = System.nanoTime() - i;
         }
 
+        this.profiler.endStartSection("dim_unloading");
+        net.minecraftforge.common.DimensionManager.unloadWorlds(worldTickTimes);
         this.profiler.endStartSection("connection");
         this.getNetworkSystem().networkTick();
         this.profiler.endStartSection("players");
@@ -801,6 +846,16 @@
     @SideOnly(Side.SERVER)
     public static void main(String[] p_main_0_)
     {
+        //Forge: Copied from DedicatedServer.init as to run as early as possible, Old code left in place intentionally.
+        //Done in good faith with permission: https://github.com/MinecraftForge/MinecraftForge/issues/3659#issuecomment-390467028
+        ServerEula eula = new ServerEula(new File("eula.txt"));
+        if (!eula.hasAcceptedEULA())
+        {
+            LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+            eula.createEULAFile();
+            return;
+        }
+
         Bootstrap.register();
 
         try
@@ -921,7 +976,8 @@
 
     public void startServerThread()
     {
-        this.serverThread = new Thread(this, "Server thread");
+        net.minecraftforge.fml.common.StartupQuery.reset();
+        this.serverThread = new Thread(net.minecraftforge.fml.common.thread.SidedThreadGroups.SERVER, this, "Server thread");
         this.serverThread.start();
     }
 
@@ -943,14 +999,13 @@
 
     public WorldServer getWorld(int dimension)
     {
-        if (dimension == -1)
-        {
-            return this.worlds[1];
-        }
-        else
-        {
-            return dimension == 1 ? this.worlds[2] : this.worlds[0];
-        }
+        WorldServer ret = net.minecraftforge.common.DimensionManager.getWorld(dimension, true);
+        if (ret == null)
+        {
+            net.minecraftforge.common.DimensionManager.initDimension(dimension);
+            ret = net.minecraftforge.common.DimensionManager.getWorld(dimension);
+        }
+        return ret;
     }
 
     public String getMinecraftVersion()
@@ -1001,7 +1056,7 @@
 
     public String getServerModName()
     {
-        return "vanilla";
+        return net.minecraftforge.fml.common.FMLCommonHandler.instance().getModName();
     }
 
     public CrashReport addServerInfoToCrashReport(CrashReport report)
@@ -1570,7 +1625,7 @@
             }
             catch (Exception exception)
             {
-                return Futures.immediateFailedCheckedFuture(exception);
+                return Futures.immediateFailedFuture(exception);
             }
         }
     }
@@ -1634,5 +1689,10 @@
         {
             this.addScheduledTask(this::reload);
         }
+    }
+
+    public DataFixer getDataFixer()
+    {
+        return this.dataFixer;
     }
 }
