--- before/net/minecraft/client/renderer/BlockFluidRenderer.java
+++ after/net/minecraft/client/renderer/BlockFluidRenderer.java
@@ -226,9 +226,9 @@
 
                 if (!flag)
                 {
-                    Block block = blockAccess.getBlockState(blockpos).getBlock();
+                    IBlockState state = blockAccess.getBlockState(blockpos);
 
-                    if (block == Blocks.GLASS || block == Blocks.STAINED_GLASS)
+                    if (state.getBlockFaceShape(blockAccess, blockpos, EnumFacing.VALUES[i1+2].getOpposite()) == net.minecraft.block.state.BlockFaceShape.SOLID)
                     {
                         textureatlassprite1 = this.atlasSpriteWaterOverlay;
                     }
