--- before/net/minecraft/client/renderer/OpenGlHelper.java
+++ after/net/minecraft/client/renderer/OpenGlHelper.java
@@ -33,6 +33,7 @@
 import org.lwjgl.opengl.GLContext;
 import oshi.SystemInfo;
 import oshi.hardware.Processor;
+import oshi.hardware.CentralProcessor;
 
 @SideOnly(Side.CLIENT)
 public class OpenGlHelper
@@ -93,6 +94,10 @@
     public static int GL_ARRAY_BUFFER;
     public static int GL_STATIC_DRAW;
 
+    /* Stores the last values sent into setLightmapTextureCoords */
+    public static float lastBrightnessX = 0.0f;
+    public static float lastBrightnessY = 0.0f;
+
     public static void initializeTextures()
     {
         ContextCapabilities contextcapabilities = GLContext.getCapabilities();
@@ -296,8 +301,8 @@
 
         try
         {
-            Processor[] aprocessor = new SystemInfo().getHardware().getProcessors();
-            cpu = String.format("%dx %s", aprocessor.length, aprocessor[0]).replaceAll("\\s+", " ");
+            CentralProcessor aprocessor = (new SystemInfo()).getHardware().getProcessor();
+            cpu = String.format("%dx %s", aprocessor.getLogicalProcessorCount(), aprocessor.getProcessorIdentifier().getName()).replaceAll("\\s+", " ");
         }
         catch (Throwable throwable)
         {
@@ -853,6 +858,12 @@
         else
         {
             GL13.glMultiTexCoord2f(target, x, y);
+        }
+
+        if (target == lightmapTexUnit)
+        {
+            lastBrightnessX = x;
+            lastBrightnessY = y;
         }
     }
 
