--- before/net/minecraft/client/renderer/OpenGlHelper.java
+++ after/net/minecraft/client/renderer/OpenGlHelper.java
@@ -15,24 +15,25 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import org.lwjgl.Sys;
-import org.lwjgl.opengl.ARBFramebufferObject;
-import org.lwjgl.opengl.ARBMultitexture;
-import org.lwjgl.opengl.ARBShaderObjects;
-import org.lwjgl.opengl.ARBVertexBufferObject;
-import org.lwjgl.opengl.ARBVertexShader;
-import org.lwjgl.opengl.ContextCapabilities;
-import org.lwjgl.opengl.EXTBlendFuncSeparate;
-import org.lwjgl.opengl.EXTFramebufferObject;
-import org.lwjgl.opengl.GL11;
-import org.lwjgl.opengl.GL13;
-import org.lwjgl.opengl.GL14;
-import org.lwjgl.opengl.GL15;
-import org.lwjgl.opengl.GL20;
-import org.lwjgl.opengl.GL30;
-import org.lwjgl.opengl.GLContext;
+import org.lwjglx.Sys;
+import org.lwjglx.opengl.ARBFramebufferObject;
+import org.lwjglx.opengl.ARBMultitexture;
+import org.lwjglx.opengl.ARBShaderObjects;
+import org.lwjglx.opengl.ARBVertexBufferObject;
+import org.lwjglx.opengl.ARBVertexShader;
+import org.lwjglx.opengl.ContextCapabilities;
+import org.lwjglx.opengl.EXTBlendFuncSeparate;
+import org.lwjglx.opengl.EXTFramebufferObject;
+import org.lwjglx.opengl.GL11;
+import org.lwjglx.opengl.GL13;
+import org.lwjglx.opengl.GL14;
+import org.lwjglx.opengl.GL15;
+import org.lwjglx.opengl.GL20;
+import org.lwjglx.opengl.GL30;
+import org.lwjglx.opengl.GLContext;
 import oshi.SystemInfo;
 import oshi.hardware.Processor;
+import oshi.hardware.CentralProcessor;
 
 @SideOnly(Side.CLIENT)
 public class OpenGlHelper
@@ -93,6 +94,10 @@
     public static int GL_ARRAY_BUFFER;
     public static int GL_STATIC_DRAW;
 
+    /* Stores the last values sent into setLightmapTextureCoords */
+    public static float lastBrightnessX = 0.0f;
+    public static float lastBrightnessY = 0.0f;
+
     public static void initializeTextures()
     {
         ContextCapabilities contextcapabilities = GLContext.getCapabilities();
@@ -296,8 +301,8 @@
 
         try
         {
-            Processor[] aprocessor = new SystemInfo().getHardware().getProcessors();
-            cpu = String.format("%dx %s", aprocessor.length, aprocessor[0]).replaceAll("\\s+", " ");
+            CentralProcessor aprocessor = (new SystemInfo()).getHardware().getProcessor();
+            cpu = String.format("%dx %s", aprocessor.getLogicalProcessorCount(), aprocessor.getProcessorIdentifier().getName()).replaceAll("\\s+", " ");
         }
         catch (Throwable throwable)
         {
@@ -853,6 +858,12 @@
         else
         {
             GL13.glMultiTexCoord2f(target, x, y);
+        }
+
+        if (target == lightmapTexUnit)
+        {
+            lastBrightnessX = x;
+            lastBrightnessY = y;
         }
     }
 
