--- before/net/minecraft/client/renderer/OpenGlHelper.java
+++ after/net/minecraft/client/renderer/OpenGlHelper.java
@@ -15,22 +15,22 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import org.lwjgl.Sys;
-import org.lwjgl.opengl.ARBFramebufferObject;
-import org.lwjgl.opengl.ARBMultitexture;
-import org.lwjgl.opengl.ARBShaderObjects;
-import org.lwjgl.opengl.ARBVertexBufferObject;
-import org.lwjgl.opengl.ARBVertexShader;
-import org.lwjgl.opengl.ContextCapabilities;
-import org.lwjgl.opengl.EXTBlendFuncSeparate;
-import org.lwjgl.opengl.EXTFramebufferObject;
-import org.lwjgl.opengl.GL11;
-import org.lwjgl.opengl.GL13;
-import org.lwjgl.opengl.GL14;
-import org.lwjgl.opengl.GL15;
-import org.lwjgl.opengl.GL20;
-import org.lwjgl.opengl.GL30;
-import org.lwjgl.opengl.GLContext;
+import org.lwjglx.Sys;
+import org.lwjglx.opengl.ARBFramebufferObject;
+import org.lwjglx.opengl.ARBMultitexture;
+import org.lwjglx.opengl.ARBShaderObjects;
+import org.lwjglx.opengl.ARBVertexBufferObject;
+import org.lwjglx.opengl.ARBVertexShader;
+import org.lwjglx.opengl.ContextCapabilities;
+import org.lwjglx.opengl.EXTBlendFuncSeparate;
+import org.lwjglx.opengl.EXTFramebufferObject;
+import org.lwjglx.opengl.GL11;
+import org.lwjglx.opengl.GL13;
+import org.lwjglx.opengl.GL14;
+import org.lwjglx.opengl.GL15;
+import org.lwjglx.opengl.GL20;
+import org.lwjglx.opengl.GL30;
+import org.lwjglx.opengl.GLContext;
 import oshi.SystemInfo;
 import oshi.hardware.Processor;
 
@@ -93,6 +93,10 @@
     public static int field_176089_P;
     public static int field_148826_e;
 
+    /* Stores the last values sent into setLightmapTextureCoords */
+    public static float lastBrightnessX = 0.0f;
+    public static float lastBrightnessY = 0.0f;
+
     public static void func_77474_a()
     {
         ContextCapabilities contextcapabilities = GLContext.getCapabilities();
@@ -852,6 +856,12 @@
         else
         {
             GL13.glMultiTexCoord2f(p_77475_0_, p_77475_1_, p_77475_2_);
+        }
+
+        if (p_77475_0_ == field_77476_b)
+        {
+            lastBrightnessX = p_77475_1_;
+            lastBrightnessY = p_77475_2_;
         }
     }
 
