--- before/net/minecraft/client/renderer/OpenGlHelper.java
+++ after/net/minecraft/client/renderer/OpenGlHelper.java
@@ -31,8 +31,8 @@
 import org.lwjgl.opengl.GL20;
 import org.lwjgl.opengl.GL30;
 import org.lwjgl.opengl.GLContext;
-import oshi.SystemInfo;
-import oshi.hardware.Processor;
+import oshi.ffm.SystemInfo;
+import oshi.hardware.CentralProcessor;
 
 @SideOnly(Side.CLIENT)
 public class OpenGlHelper
@@ -93,6 +93,10 @@
     public static int GL_ARRAY_BUFFER;
     public static int GL_STATIC_DRAW;
 
+    /* Stores the last values sent into setLightmapTextureCoords */
+    public static float lastBrightnessX = 0.0f;
+    public static float lastBrightnessY = 0.0f;
+
     public static void initializeTextures()
     {
         ContextCapabilities contextcapabilities = GLContext.getCapabilities();
@@ -296,8 +300,8 @@
 
         try
         {
-            Processor[] aprocessor = new SystemInfo().getHardware().getProcessors();
-            cpu = String.format("%dx %s", aprocessor.length, aprocessor[0]).replaceAll("\\s+", " ");
+            CentralProcessor aprocessor = (new SystemInfo()).getHardware().getProcessor();
+            cpu = String.format("%dx %s", aprocessor.getLogicalProcessorCount(), aprocessor.getProcessorIdentifier().getName()).replaceAll("\\s+", " ");
         }
         catch (Throwable throwable)
         {
@@ -854,6 +858,12 @@
         {
             GL13.glMultiTexCoord2f(target, x, y);
         }
+
+        if (target == lightmapTexUnit)
+        {
+            lastBrightnessX = x;
+            lastBrightnessY = y;
+        }
     }
 
     public static void glBlendFunc(int sFactorRGB, int dFactorRGB, int sfactorAlpha, int dfactorAlpha)
@@ -916,54 +926,23 @@
 
     public static void openFile(File fileIn)
     {
-        String s = fileIn.getAbsolutePath();
-
-        if (Util.getOSType() == Util.EnumOS.OSX)
-        {
-            try
-            {
-                LOGGER.info(s);
-                Runtime.getRuntime().exec(new String[] {"/usr/bin/open", s});
-                return;
-            }
-            catch (IOException ioexception1)
-            {
-                LOGGER.error("Couldn't open file", (Throwable)ioexception1);
-            }
-        }
-        else if (Util.getOSType() == Util.EnumOS.WINDOWS)
-        {
-            String s1 = String.format("cmd.exe /C start \"Open file\" \"%s\"", s);
-
-            try
-            {
-                Runtime.getRuntime().exec(s1);
-                return;
-            }
-            catch (IOException ioexception)
-            {
-                LOGGER.error("Couldn't open file", (Throwable)ioexception);
-            }
-        }
-
-        boolean flag = false;
-
         try
         {
-            Class<?> oclass = Class.forName("java.awt.Desktop");
-            Object object = oclass.getMethod("getDesktop").invoke(null);
-            oclass.getMethod("browse", URI.class).invoke(object, fileIn.toURI());
-        }
-        catch (Throwable throwable)
-        {
-            LOGGER.error("Couldn't open link", throwable);
-            flag = true;
-        }
-
-        if (flag)
-        {
-            LOGGER.info("Opening via system class!");
-            Sys.openURL("file://" + s);
+            if (!java.awt.Desktop.isDesktopSupported())
+            {
+                LOGGER.warn("Desktop API is not supported");
+                return;
+            }
+            java.awt.Desktop desktop = java.awt.Desktop.getDesktop();
+            if (!desktop.isSupported(java.awt.Desktop.Action.OPEN))
+            {
+                LOGGER.warn("Opening files is not supported.");
+                return;
+            }
+            desktop.open(fileIn);
+        } catch (IOException e)
+        {
+            LOGGER.error("Could not open: " + fileIn.getAbsolutePath(), e);
         }
     }
 
