--- before/net/minecraft/client/renderer/ViewFrustum.java
+++ after/net/minecraft/client/renderer/ViewFrustum.java
@@ -153,14 +153,14 @@
 
         if (j >= 0 && j < this.field_178168_c)
         {
-            i %= this.field_178165_d;
+            i = i % this.field_178165_d;
 
             if (i < 0)
             {
                 i += this.field_178165_d;
             }
 
-            k %= this.field_178166_e;
+            k = k % this.field_178166_e;
 
             if (k < 0)
             {
