--- before/net/minecraft/util/text/ITextComponent.java
+++ after/net/minecraft/util/text/ITextComponent.java
@@ -34,7 +34,6 @@
 
     String func_150260_c();
 
-    @SideOnly(Side.CLIENT)
     String func_150254_d();
 
     List<ITextComponent> func_150253_a();
@@ -97,7 +96,7 @@
                             JsonArray jsonarray = jsonobject.getAsJsonArray("with");
                             Object[] aobject = new Object[jsonarray.size()];
 
-                            for (int i = 0; i < aobject.length; i++)
+                            for (int i = 0; i < aobject.length; ++i)
                             {
                                 aobject[i] = this.deserialize(jsonarray.get(i), p_deserialize_2_, p_deserialize_3_);
 
@@ -116,7 +115,7 @@
                         }
                         else
                         {
-                            itextcomponent = new TextComponentTranslation(s);
+                            itextcomponent = new TextComponentTranslation(s, new Object[0]);
                         }
                     }
                     else if (jsonobject.has("score"))
@@ -158,13 +157,13 @@
                             throw new JsonParseException("Unexpected empty array of components");
                         }
 
-                        for (int j = 0; j < jsonarray2.size(); j++)
+                        for (int j = 0; j < jsonarray2.size(); ++j)
                         {
                             itextcomponent.func_150257_a(this.deserialize(jsonarray2.get(j), p_deserialize_2_, p_deserialize_3_));
                         }
                     }
 
-                    itextcomponent.func_150255_a(p_deserialize_3_.deserialize(p_deserialize_1_, Style.class));
+                    itextcomponent.func_150255_a((Style)p_deserialize_3_.deserialize(p_deserialize_1_, Style.class));
                     return itextcomponent;
                 }
             }
@@ -269,13 +268,13 @@
             @Nullable
             public static ITextComponent func_150699_a(String p_150699_0_)
             {
-                return JsonUtils.func_188176_a(field_150700_a, p_150699_0_, ITextComponent.class, false);
+                return (ITextComponent)JsonUtils.func_188176_a(field_150700_a, p_150699_0_, ITextComponent.class, false);
             }
 
             @Nullable
             public static ITextComponent func_186877_b(String p_186877_0_)
             {
-                return JsonUtils.func_188176_a(field_150700_a, p_186877_0_, ITextComponent.class, true);
+                return (ITextComponent)JsonUtils.func_188176_a(field_150700_a, p_186877_0_, ITextComponent.class, true);
             }
 
             static
