--- before/net/minecraft/util/text/Style.java
+++ after/net/minecraft/util/text/Style.java
@@ -138,7 +138,6 @@
         {
             return this;
         }
-        @SideOnly(Side.CLIENT)
         @Override
         public String getFormattingCode()
         {
@@ -268,7 +267,6 @@
         return this;
     }
 
-    @SideOnly(Side.CLIENT)
     public String getFormattingCode()
     {
         if (this.isEmpty())
