--- before/net/minecraft/util/text/TextComponentTranslation.java
+++ after/net/minecraft/util/text/TextComponentTranslation.java
@@ -18,7 +18,7 @@
     private final Object field_150274_f = new Object();
     private long field_150275_g = -1L;
     @VisibleForTesting
-    List<ITextComponent> field_150278_b = Lists.newArrayList();
+    List<ITextComponent> field_150278_b = Lists.<ITextComponent>newArrayList();
     public static final Pattern field_150279_c = Pattern.compile("%(?:(\\d+)\\$)?([A-Za-z%]|$)");
 
     public TextComponentTranslation(String p_i45160_1_, Object... p_i45160_2_)
@@ -56,7 +56,7 @@
         {
             this.func_150269_b(I18n.func_74838_a(this.field_150276_d));
         }
-        catch (TextComponentTranslationFormatException textcomponenttranslationformatexception1)
+        catch (TextComponentTranslationFormatException textcomponenttranslationformatexception)
         {
             this.field_150278_b.clear();
 
@@ -64,9 +64,9 @@
             {
                 this.func_150269_b(I18n.func_150826_b(this.field_150276_d));
             }
-            catch (TextComponentTranslationFormatException textcomponenttranslationformatexception)
+            catch (TextComponentTranslationFormatException var5)
             {
-                throw textcomponenttranslationformatexception1;
+                throw textcomponenttranslationformatexception;
             }
         }
     }
@@ -184,7 +184,7 @@
     public Iterator<ITextComponent> iterator()
     {
         this.func_150270_g();
-        return Iterators.concat(func_150262_a(this.field_150278_b), func_150262_a(this.field_150264_a));
+        return Iterators.<ITextComponent>concat(func_150262_a(this.field_150278_b), func_150262_a(this.field_150264_a));
     }
 
     public String func_150261_e()
@@ -248,7 +248,8 @@
     {
         int i = super.hashCode();
         i = 31 * i + this.field_150276_d.hashCode();
-        return 31 * i + Arrays.hashCode(this.field_150277_e);
+        i = 31 * i + Arrays.hashCode(this.field_150277_e);
+        return i;
     }
 
     public String toString()
