--- before/net/minecraft/util/text/TextComponentUtils.java
+++ after/net/minecraft/util/text/TextComponentUtils.java
@@ -21,11 +21,11 @@
 
             if (EntitySelector.func_82378_b(s))
             {
-                List<Entity> list = EntitySelector.func_179656_b(p_179985_0_, s, Entity.class);
+                List<Entity> list = EntitySelector.<Entity>func_179656_b(p_179985_0_, s, Entity.class);
 
                 if (list.size() != 1)
                 {
-                    throw new EntityNotFoundException("commands.generic.selector.notFound", s);
+                    throw new EntityNotFoundException("commands.generic.selector.notFound", new Object[] {s});
                 }
 
                 Entity entity = list.get(0);
