--- before/net/minecraft/network/play/server/SPacketStatistics.java
+++ after/net/minecraft/network/play/server/SPacketStatistics.java
@@ -3,6 +3,7 @@
 import com.google.common.collect.Maps;
 import java.io.IOException;
 import java.util.Map;
+import java.util.Map.Entry;
 import net.minecraft.network.Packet;
 import net.minecraft.network.PacketBuffer;
 import net.minecraft.network.play.INetHandlerPlayClient;
@@ -32,7 +33,7 @@
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
         int i = p_148837_1_.func_150792_a();
-        this.field_148976_a = Maps.newHashMap();
+        this.field_148976_a = Maps.<StatBase, Integer>newHashMap();
 
         for (int j = 0; j < i; ++j)
         {
@@ -41,7 +42,7 @@
 
             if (statbase != null)
             {
-                this.field_148976_a.put(statbase, k);
+                this.field_148976_a.put(statbase, Integer.valueOf(k));
             }
         }
     }
@@ -50,10 +51,10 @@
     {
         p_148840_1_.func_150787_b(this.field_148976_a.size());
 
-        for (Map.Entry<StatBase, Integer> entry : this.field_148976_a.entrySet())
+        for (Entry<StatBase, Integer> entry : this.field_148976_a.entrySet())
         {
             p_148840_1_.func_180714_a((entry.getKey()).field_75975_e);
-            p_148840_1_.func_150787_b(entry.getValue());
+            p_148840_1_.func_150787_b(((Integer)entry.getValue()).intValue());
         }
     }
 
