--- before/net/minecraft/network/play/server/SPacketParticles.java
+++ after/net/minecraft/network/play/server/SPacketParticles.java
@@ -94,6 +94,11 @@
         return this.field_179751_a;
     }
 
+    public void func_148833_a(INetHandlerPlayClient p_148833_1_)
+    {
+        p_148833_1_.func_147289_a(this);
+    }
+
     @SideOnly(Side.CLIENT)
     public boolean func_179750_b()
     {
@@ -152,10 +157,5 @@
     public int[] func_179748_k()
     {
         return this.field_179753_k;
-    }
-
-    public void func_148833_a(INetHandlerPlayClient p_148833_1_)
-    {
-        p_148833_1_.func_147289_a(this);
     }
 }
