--- before/net/minecraft/network/play/server/SPacketMaps.java
+++ after/net/minecraft/network/play/server/SPacketMaps.java
@@ -31,7 +31,7 @@
         this.field_149191_a = p_i46937_1_;
         this.field_179739_b = p_i46937_2_;
         this.field_186950_c = p_i46937_3_;
-        this.field_179740_c = p_i46937_4_.toArray(new MapDecoration[p_i46937_4_.size()]);
+        this.field_179740_c = (MapDecoration[])p_i46937_4_.toArray(new MapDecoration[p_i46937_4_.size()]);
         this.field_179737_d = p_i46937_6_;
         this.field_179738_e = p_i46937_7_;
         this.field_179735_f = p_i46937_8_;
