--- before/net/minecraft/network/play/server/SPacketTeams.java
+++ after/net/minecraft/network/play/server/SPacketTeams.java
@@ -17,19 +17,27 @@
     private String field_149318_b = "";
     private String field_149319_c = "";
     private String field_149316_d = "";
-    private String field_179816_e = Team.EnumVisible.ALWAYS.field_178830_e;
-    private String field_186976_f = Team.CollisionRule.ALWAYS.field_186693_e;
-    private int field_179815_f = -1;
-    private final Collection<String> field_149317_e = Lists.newArrayList();
+    private String field_179816_e;
+    private String field_186976_f;
+    private int field_179815_f;
+    private final Collection<String> field_149317_e;
     private int field_149314_f;
     private int field_149315_g;
 
     public SPacketTeams()
     {
+        this.field_179816_e = Team.EnumVisible.ALWAYS.field_178830_e;
+        this.field_186976_f = Team.CollisionRule.ALWAYS.field_186693_e;
+        this.field_179815_f = -1;
+        this.field_149317_e = Lists.<String>newArrayList();
     }
 
     public SPacketTeams(ScorePlayerTeam p_i46907_1_, int p_i46907_2_)
     {
+        this.field_179816_e = Team.EnumVisible.ALWAYS.field_178830_e;
+        this.field_186976_f = Team.CollisionRule.ALWAYS.field_186693_e;
+        this.field_179815_f = -1;
+        this.field_149317_e = Lists.<String>newArrayList();
         this.field_149320_a = p_i46907_1_.func_96661_b();
         this.field_149314_f = p_i46907_2_;
 
@@ -52,6 +60,11 @@
 
     public SPacketTeams(ScorePlayerTeam p_i46908_1_, Collection<String> p_i46908_2_, int p_i46908_3_)
     {
+        this.field_179816_e = Team.EnumVisible.ALWAYS.field_178830_e;
+        this.field_186976_f = Team.CollisionRule.ALWAYS.field_186693_e;
+        this.field_179815_f = -1;
+        this.field_149317_e = Lists.<String>newArrayList();
+
         if (p_i46908_3_ != 3 && p_i46908_3_ != 4)
         {
             throw new IllegalArgumentException("Method must be join or leave for player constructor");
