--- before/net/minecraft/network/play/server/SPacketJoinGame.java
+++ after/net/minecraft/network/play/server/SPacketJoinGame.java
@@ -42,7 +42,7 @@
         this.field_149206_a = p_148837_1_.readInt();
         int i = p_148837_1_.readUnsignedByte();
         this.field_149204_b = (i & 8) == 8;
-        i &= -9;
+        i = i & -9;
         this.field_149205_c = GameType.func_77146_a(i);
         this.field_149202_d = p_148837_1_.readInt();
         this.field_149203_e = EnumDifficulty.func_151523_a(p_148837_1_.readUnsignedByte());
