--- before/net/minecraft/network/play/server/SPacketEntityEquipment.java
+++ after/net/minecraft/network/play/server/SPacketEntityEquipment.java
@@ -29,7 +29,7 @@
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
         this.field_149394_a = p_148837_1_.func_150792_a();
-        this.field_149392_b = p_148837_1_.func_179257_a(EntityEquipmentSlot.class);
+        this.field_149392_b = (EntityEquipmentSlot)p_148837_1_.func_179257_a(EntityEquipmentSlot.class);
         this.field_149393_c = p_148837_1_.func_150791_c();
     }
 
