--- before/net/minecraft/network/play/server/SPacketExplosion.java
+++ after/net/minecraft/network/play/server/SPacketExplosion.java
@@ -49,7 +49,7 @@
         this.field_149157_c = (double)p_148837_1_.readFloat();
         this.field_149154_d = p_148837_1_.readFloat();
         int i = p_148837_1_.readInt();
-        this.field_149155_e = Lists.newArrayListWithCapacity(i);
+        this.field_149155_e = Lists.<BlockPos>newArrayListWithCapacity(i);
         int j = (int)this.field_149158_a;
         int k = (int)this.field_149156_b;
         int l = (int)this.field_149157_c;
