--- before/net/minecraft/network/play/server/SPacketPlayerPosLook.java
+++ after/net/minecraft/network/play/server/SPacketPlayerPosLook.java
@@ -130,7 +130,7 @@
 
         public static Set<SPacketPlayerPosLook.EnumFlags> func_187044_a(int p_187044_0_)
         {
-            Set<SPacketPlayerPosLook.EnumFlags> set = EnumSet.noneOf(SPacketPlayerPosLook.EnumFlags.class);
+            Set<SPacketPlayerPosLook.EnumFlags> set = EnumSet.<SPacketPlayerPosLook.EnumFlags>noneOf(SPacketPlayerPosLook.EnumFlags.class);
 
             for (SPacketPlayerPosLook.EnumFlags spacketplayerposlook$enumflags : values())
             {
