--- before/net/minecraft/network/play/server/SPacketEntityProperties.java
+++ after/net/minecraft/network/play/server/SPacketEntityProperties.java
@@ -16,7 +16,7 @@
 public class SPacketEntityProperties implements Packet<INetHandlerPlayClient>
 {
     private int field_149445_a;
-    private final List<SPacketEntityProperties.Snapshot> field_149444_b = Lists.newArrayList();
+    private final List<SPacketEntityProperties.Snapshot> field_149444_b = Lists.<SPacketEntityProperties.Snapshot>newArrayList();
 
     public SPacketEntityProperties()
     {
@@ -41,7 +41,7 @@
         {
             String s = p_148837_1_.func_150789_c(64);
             double d0 = p_148837_1_.readDouble();
-            List<AttributeModifier> list = Lists.newArrayList();
+            List<AttributeModifier> list = Lists.<AttributeModifier>newArrayList();
             int k = p_148837_1_.func_150792_a();
 
             for (int l = 0; l < k; ++l)
