--- before/net/minecraft/network/play/server/SPacketWindowItems.java
+++ after/net/minecraft/network/play/server/SPacketWindowItems.java
@@ -22,7 +22,7 @@
     public SPacketWindowItems(int p_i47317_1_, NonNullList<ItemStack> p_i47317_2_)
     {
         this.field_148914_a = p_i47317_1_;
-        this.field_148913_b = NonNullList.func_191197_a(p_i47317_2_.size(), ItemStack.field_190927_a);
+        this.field_148913_b = NonNullList.<ItemStack>func_191197_a(p_i47317_2_.size(), ItemStack.field_190927_a);
 
         for (int i = 0; i < this.field_148913_b.size(); ++i)
         {
@@ -35,7 +35,7 @@
     {
         this.field_148914_a = p_148837_1_.readUnsignedByte();
         int i = p_148837_1_.readShort();
-        this.field_148913_b = NonNullList.func_191197_a(i, ItemStack.field_190927_a);
+        this.field_148913_b = NonNullList.<ItemStack>func_191197_a(i, ItemStack.field_190927_a);
 
         for (int j = 0; j < i; ++j)
         {
