--- before/net/minecraft/network/play/server/SPacketUpdateBossInfo.java
+++ after/net/minecraft/network/play/server/SPacketUpdateBossInfo.java
@@ -42,15 +42,15 @@
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
         this.field_186911_a = p_148837_1_.func_179253_g();
-        this.field_186912_b = p_148837_1_.func_179257_a(SPacketUpdateBossInfo.Operation.class);
+        this.field_186912_b = (SPacketUpdateBossInfo.Operation)p_148837_1_.func_179257_a(SPacketUpdateBossInfo.Operation.class);
 
         switch (this.field_186912_b)
         {
             case ADD:
                 this.field_186913_c = p_148837_1_.func_179258_d();
                 this.field_186914_d = p_148837_1_.readFloat();
-                this.field_186915_e = p_148837_1_.func_179257_a(BossInfo.Color.class);
-                this.field_186916_f = p_148837_1_.func_179257_a(BossInfo.Overlay.class);
+                this.field_186915_e = (BossInfo.Color)p_148837_1_.func_179257_a(BossInfo.Color.class);
+                this.field_186916_f = (BossInfo.Overlay)p_148837_1_.func_179257_a(BossInfo.Overlay.class);
                 this.func_186903_a(p_148837_1_.readUnsignedByte());
             case REMOVE:
             default:
@@ -62,8 +62,8 @@
                 this.field_186913_c = p_148837_1_.func_179258_d();
                 break;
             case UPDATE_STYLE:
-                this.field_186915_e = p_148837_1_.func_179257_a(BossInfo.Color.class);
-                this.field_186916_f = p_148837_1_.func_179257_a(BossInfo.Overlay.class);
+                this.field_186915_e = (BossInfo.Color)p_148837_1_.func_179257_a(BossInfo.Color.class);
+                this.field_186916_f = (BossInfo.Overlay)p_148837_1_.func_179257_a(BossInfo.Overlay.class);
                 break;
             case UPDATE_PROPERTIES:
                 this.func_186903_a(p_148837_1_.readUnsignedByte());
