--- before/net/minecraft/network/play/server/SPacketAdvancementInfo.java
+++ after/net/minecraft/network/play/server/SPacketAdvancementInfo.java
@@ -6,6 +6,7 @@
 import java.util.Collection;
 import java.util.Map;
 import java.util.Set;
+import java.util.Map.Entry;
 import net.minecraft.advancements.Advancement;
 import net.minecraft.advancements.AdvancementProgress;
 import net.minecraft.network.Packet;
@@ -29,7 +30,7 @@
     public SPacketAdvancementInfo(boolean p_i47519_1_, Collection<Advancement> p_i47519_2_, Set<ResourceLocation> p_i47519_3_, Map<ResourceLocation, AdvancementProgress> p_i47519_4_)
     {
         this.field_192605_a = p_i47519_1_;
-        this.field_192606_b = Maps.newHashMap();
+        this.field_192606_b = Maps.<ResourceLocation, Advancement.Builder>newHashMap();
 
         for (Advancement advancement : p_i47519_2_)
         {
@@ -37,7 +38,7 @@
         }
 
         this.field_192607_c = p_i47519_3_;
-        this.field_192608_d = Maps.newHashMap(p_i47519_4_);
+        this.field_192608_d = Maps.<ResourceLocation, AdvancementProgress>newHashMap(p_i47519_4_);
     }
 
     public void func_148833_a(INetHandlerPlayClient p_148833_1_)
@@ -48,9 +49,9 @@
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
         this.field_192605_a = p_148837_1_.readBoolean();
-        this.field_192606_b = Maps.newHashMap();
-        this.field_192607_c = Sets.newLinkedHashSet();
-        this.field_192608_d = Maps.newHashMap();
+        this.field_192606_b = Maps.<ResourceLocation, Advancement.Builder>newHashMap();
+        this.field_192607_c = Sets.<ResourceLocation>newLinkedHashSet();
+        this.field_192608_d = Maps.<ResourceLocation, AdvancementProgress>newHashMap();
         int i = p_148837_1_.func_150792_a();
 
         for (int j = 0; j < i; ++j)
@@ -82,7 +83,7 @@
         p_148840_1_.writeBoolean(this.field_192605_a);
         p_148840_1_.func_150787_b(this.field_192606_b.size());
 
-        for (Map.Entry<ResourceLocation, Advancement.Builder> entry : this.field_192606_b.entrySet())
+        for (Entry<ResourceLocation, Advancement.Builder> entry : this.field_192606_b.entrySet())
         {
             ResourceLocation resourcelocation = entry.getKey();
             Advancement.Builder advancement$builder = entry.getValue();
@@ -99,10 +100,10 @@
 
         p_148840_1_.func_150787_b(this.field_192608_d.size());
 
-        for (Map.Entry<ResourceLocation, AdvancementProgress> entry1 : this.field_192608_d.entrySet())
+        for (Entry<ResourceLocation, AdvancementProgress> entry1 : this.field_192608_d.entrySet())
         {
             p_148840_1_.func_192572_a(entry1.getKey());
-            entry1.getValue().func_192104_a(p_148840_1_);
+            ((AdvancementProgress)entry1.getValue()).func_192104_a(p_148840_1_);
         }
     }
 
