--- before/net/minecraft/network/play/server/SPacketPlayerListItem.java
+++ after/net/minecraft/network/play/server/SPacketPlayerListItem.java
@@ -19,7 +19,7 @@
 public class SPacketPlayerListItem implements Packet<INetHandlerPlayClient>
 {
     private SPacketPlayerListItem.Action field_179770_a;
-    private final List<SPacketPlayerListItem.AddPlayerData> field_179769_b = Lists.newArrayList();
+    private final List<SPacketPlayerListItem.AddPlayerData> field_179769_b = Lists.<SPacketPlayerListItem.AddPlayerData>newArrayList();
 
     public SPacketPlayerListItem()
     {
@@ -47,7 +47,7 @@
 
     public void func_148837_a(PacketBuffer p_148837_1_) throws IOException
     {
-        this.field_179770_a = p_148837_1_.func_179257_a(SPacketPlayerListItem.Action.class);
+        this.field_179770_a = (SPacketPlayerListItem.Action)p_148837_1_.func_179257_a(SPacketPlayerListItem.Action.class);
         int i = p_148837_1_.func_150792_a();
 
         for (int j = 0; j < i; ++j)
