--- before/net/minecraft/client/gui/toasts/GuiToast.java
+++ after/net/minecraft/client/gui/toasts/GuiToast.java
@@ -18,7 +18,7 @@
 {
     private final Minecraft field_191790_f;
     private final GuiToast.ToastInstance<?>[] field_191791_g = new GuiToast.ToastInstance[5];
-    private final Deque<IToast> field_191792_h = Queues.newArrayDeque();
+    private final Deque<IToast> field_191792_h = Queues.<IToast>newArrayDeque();
 
     public GuiToast(Minecraft p_i47388_1_)
     {
@@ -90,12 +90,15 @@
     class ToastInstance<T extends IToast>
     {
         private final T field_193688_b;
-        private long field_193689_c = -1L;
-        private long field_193690_d = -1L;
-        private IToast.Visibility field_193691_e = IToast.Visibility.SHOW;
+        private long field_193689_c;
+        private long field_193690_d;
+        private IToast.Visibility field_193691_e;
 
         private ToastInstance(T p_i47483_2_)
         {
+            this.field_193689_c = -1L;
+            this.field_193690_d = -1L;
+            this.field_193691_e = IToast.Visibility.SHOW;
             this.field_193688_b = p_i47483_2_;
         }
 
@@ -107,7 +110,7 @@
         private float func_193686_a(long p_193686_1_)
         {
             float f = MathHelper.func_76131_a((float)(p_193686_1_ - this.field_193689_c) / 600.0F, 0.0F, 1.0F);
-            f *= f;
+            f = f * f;
             return this.field_193691_e == IToast.Visibility.HIDE ? 1.0F - f : f;
         }
 
