--- before/net/minecraft/client/gui/toasts/SystemToast.java
+++ after/net/minecraft/client/gui/toasts/SystemToast.java
@@ -61,7 +61,7 @@
 
     public static void func_193657_a(GuiToast p_193657_0_, SystemToast.Type p_193657_1_, ITextComponent p_193657_2_, @Nullable ITextComponent p_193657_3_)
     {
-        SystemToast systemtoast = p_193657_0_.func_192990_a(SystemToast.class, p_193657_1_);
+        SystemToast systemtoast = (SystemToast)p_193657_0_.func_192990_a(SystemToast.class, p_193657_1_);
 
         if (systemtoast == null)
         {
