--- before/net/minecraft/dispenser/BehaviorProjectileDispense.java
+++ after/net/minecraft/dispenser/BehaviorProjectileDispense.java
@@ -13,7 +13,7 @@
     {
         World world = p_82487_1_.func_82618_k();
         IPosition iposition = BlockDispenser.func_149939_a(p_82487_1_);
-        EnumFacing enumfacing = p_82487_1_.func_189992_e().func_177229_b(BlockDispenser.field_176441_a);
+        EnumFacing enumfacing = (EnumFacing)p_82487_1_.func_189992_e().func_177229_b(BlockDispenser.field_176441_a);
         IProjectile iprojectile = this.func_82499_a(world, iposition, p_82487_2_);
         iprojectile.func_70186_c((double)enumfacing.func_82601_c(), (double)((float)enumfacing.func_96559_d() + 0.1F), (double)enumfacing.func_82599_e(), this.func_82500_b(), this.func_82498_a());
         world.func_72838_d((Entity)iprojectile);
