--- before/net/minecraft/client/network/NetHandlerLoginClient.java
+++ after/net/minecraft/client/network/NetHandlerLoginClient.java
@@ -49,13 +49,12 @@
         this.field_147395_c = p_i45059_3_;
     }
 
-    @Override
     public void func_147389_a(SPacketEncryptionRequest p_147389_1_)
     {
         final SecretKey secretkey = CryptManager.func_75890_a();
         String s = p_147389_1_.func_149609_c();
         PublicKey publickey = p_147389_1_.func_149608_d();
-        String s1 = new BigInteger(CryptManager.func_75895_a(s, publickey, secretkey)).toString(16);
+        String s1 = (new BigInteger(CryptManager.func_75895_a(s, publickey, secretkey))).toString(16);
 
         if (this.field_147394_b.func_147104_D() != null && this.field_147394_b.func_147104_D().func_181041_d())
         {
@@ -63,7 +62,7 @@
             {
                 this.func_147391_c().joinServer(this.field_147394_b.func_110432_I().func_148256_e(), this.field_147394_b.func_110432_I().func_148254_d(), s1);
             }
-            catch (AuthenticationException authenticationexception1)
+            catch (AuthenticationException var10)
             {
                 field_147396_a.warn("Couldn't connect to auth servers but will continue to join LAN");
             }
@@ -74,33 +73,25 @@
             {
                 this.func_147391_c().joinServer(this.field_147394_b.func_110432_I().func_148256_e(), this.field_147394_b.func_110432_I().func_148254_d(), s1);
             }
-            catch (AuthenticationUnavailableException authenticationunavailableexception)
+            catch (AuthenticationUnavailableException var7)
             {
-                this.field_147393_d
-                .func_150718_a(
-                    new TextComponentTranslation("disconnect.loginFailedInfo", new TextComponentTranslation("disconnect.loginFailedInfo.serversUnavailable"))
-                );
+                this.field_147393_d.func_150718_a(new TextComponentTranslation("disconnect.loginFailedInfo", new Object[] {new TextComponentTranslation("disconnect.loginFailedInfo.serversUnavailable", new Object[0])}));
                 return;
             }
-            catch (InvalidCredentialsException invalidcredentialsexception)
+            catch (InvalidCredentialsException var8)
             {
-                this.field_147393_d
-                .func_150718_a(
-                    new TextComponentTranslation("disconnect.loginFailedInfo", new TextComponentTranslation("disconnect.loginFailedInfo.invalidSession"))
-                );
+                this.field_147393_d.func_150718_a(new TextComponentTranslation("disconnect.loginFailedInfo", new Object[] {new TextComponentTranslation("disconnect.loginFailedInfo.invalidSession", new Object[0])}));
                 return;
             }
             catch (AuthenticationException authenticationexception)
             {
-                this.field_147393_d.func_150718_a(new TextComponentTranslation("disconnect.loginFailedInfo", authenticationexception.getMessage()));
+                this.field_147393_d.func_150718_a(new TextComponentTranslation("disconnect.loginFailedInfo", new Object[] {authenticationexception.getMessage()}));
                 return;
             }
         }
 
-        this.field_147393_d
-        .func_179288_a(new CPacketEncryptionResponse(secretkey, publickey, p_147389_1_.func_149607_e()), new GenericFutureListener < Future <? super Void >> ()
+        this.field_147393_d.func_179288_a(new CPacketEncryptionResponse(secretkey, publickey, p_147389_1_.func_149607_e()), new GenericFutureListener < Future <? super Void >> ()
         {
-            @Override
             public void operationComplete(Future <? super Void > p_operationComplete_1_) throws Exception
             {
                 NetHandlerLoginClient.this.field_147393_d.func_150727_a(secretkey);
@@ -113,21 +104,21 @@
         return this.field_147394_b.func_152347_ac();
     }
 
-    @Override
     public void func_147390_a(SPacketLoginSuccess p_147390_1_)
     {
         this.field_175091_e = p_147390_1_.func_179730_a();
         this.field_147393_d.func_150723_a(EnumConnectionState.PLAY);
-        this.field_147393_d.func_150719_a(new NetHandlerPlayClient(this.field_147394_b, this.field_147395_c, this.field_147393_d, this.field_175091_e));
+        NetHandlerPlayClient nhpc = new NetHandlerPlayClient(this.field_147394_b, this.field_147395_c, this.field_147393_d, this.field_175091_e);
+        this.field_147393_d.func_150719_a(nhpc);
+        net.minecraftforge.fml.client.FMLClientHandler.instance().setPlayClient(nhpc);
+        net.minecraftforge.fml.common.network.internal.FMLNetworkHandler.fmlClientHandshake(this.field_147393_d);
     }
 
-    @Override
     public void func_147231_a(ITextComponent p_147231_1_)
     {
         if (this.field_147395_c != null && this.field_147395_c instanceof GuiScreenRealmsProxy)
         {
-            this.field_147394_b
-            .func_147108_a(new DisconnectedRealmsScreen(((GuiScreenRealmsProxy)this.field_147395_c).func_154321_a(), "connect.failed", p_147231_1_).getProxy());
+            this.field_147394_b.func_147108_a((new DisconnectedRealmsScreen(((GuiScreenRealmsProxy)this.field_147395_c).func_154321_a(), "connect.failed", p_147231_1_)).getProxy());
         }
         else
         {
@@ -135,13 +126,11 @@
         }
     }
 
-    @Override
     public void func_147388_a(SPacketDisconnect p_147388_1_)
     {
         this.field_147393_d.func_150718_a(p_147388_1_.func_149603_c());
     }
 
-    @Override
     public void func_180464_a(SPacketEnableCompression p_180464_1_)
     {
         if (!this.field_147393_d.func_150731_c())
