--- before/net/minecraft/client/network/ServerPinger.java
+++ after/net/minecraft/client/network/ServerPinger.java
@@ -166,6 +166,7 @@
                         server.setBase64EncodedIconData(null);
                     }
 
+                    net.minecraftforge.fml.client.FMLClientHandler.instance().bindServerListData(server, serverstatusresponse);
                     this.pingSentAt = Minecraft.getSystemTime();
                     networkmanager.sendPacket(new CPacketPing(this.pingSentAt));
                     this.successful = true;
@@ -195,7 +196,7 @@
 
         try
         {
-            networkmanager.sendPacket(new C00Handshake(serveraddress.getIP(), serveraddress.getPort(), EnumConnectionState.STATUS));
+            networkmanager.sendPacket(new C00Handshake(serveraddress.getIP(), serveraddress.getPort(), EnumConnectionState.STATUS, true));
             networkmanager.sendPacket(new CPacketServerQuery());
         }
         catch (Throwable throwable)
