--- before/net/minecraft/block/properties/PropertyBool.java
+++ after/net/minecraft/block/properties/PropertyBool.java
@@ -6,7 +6,7 @@
 
 public class PropertyBool extends PropertyHelper<Boolean>
 {
-    private final ImmutableSet<Boolean> field_177717_a = ImmutableSet.of(true, false);
+    private final ImmutableSet<Boolean> field_177717_a = ImmutableSet.<Boolean>of(Boolean.valueOf(true), Boolean.valueOf(false));
 
     protected PropertyBool(String p_i45651_1_)
     {
