--- original/net/minecraft/client/model/ModelRenderer.java
+++ modified/net/minecraft/client/model/ModelRenderer.java
@@ -123,6 +123,7 @@
         this.rotationPointZ = rotationPointZIn;
     }
 
+    @SideOnly(Side.CLIENT)
     public void render(float scale) {
         if (!this.isHidden) {
             if (this.showModel) {
@@ -178,6 +179,7 @@
         }
     }
 
+    @SideOnly(Side.CLIENT)
     public void renderWithRotation(float scale) {
         if (!this.isHidden) {
             if (this.showModel) {
@@ -205,7 +207,7 @@
         }
     }
 
-    /** Allows the changing of Angles after a box has been rendered */
+    @SideOnly(Side.CLIENT)
     public void postRender(float scale) {
         if (!this.isHidden) {
             if (this.showModel) {
@@ -233,7 +235,7 @@
         }
     }
 
-    /** Compiles a GL display list for this model */
+    @SideOnly(Side.CLIENT)
     private void compileDisplayList(float scale) {
         this.displayList = GLAllocation.generateDisplayLists(1);
         GlStateManager.glNewList(this.displayList, 4864);
