--- original/net/minecraft/client/model/TexturedQuad.java
+++ modified/net/minecraft/client/model/TexturedQuad.java
@@ -36,6 +36,7 @@
     }
 
     /** Draw this primitve. This is typically called only once as the generated drawing instructions are saved by the renderer and reused later. */
+    @net.minecraftforge.fml.relauncher.SideOnly(net.minecraftforge.fml.relauncher.Side.CLIENT)
     public void draw(BufferBuilder renderer, float scale) {
         Vec3d vec3d = this.vertexPositions[1].vector3D.subtractReverse(this.vertexPositions[0].vector3D);
         Vec3d vec3d1 = this.vertexPositions[1].vector3D.subtractReverse(this.vertexPositions[2].vector3D);
