--- before/net/minecraft/util/math/BlockPos.java
+++ after/net/minecraft/util/math/BlockPos.java
@@ -205,7 +205,7 @@
                         }
                         else if (this.field_191535_c == p_191532_3_ && this.field_191536_d == p_191532_4_ && this.field_191537_e == p_191532_5_)
                         {
-                            return this.endOfData();
+                            return (BlockPos)this.endOfData();
                         }
                         else
                         {
@@ -261,7 +261,7 @@
                         }
                         else if (this.field_179314_b.field_177997_b == p_191531_3_ && this.field_179314_b.field_177998_c == p_191531_4_ && this.field_179314_b.field_177996_d == p_191531_5_)
                         {
-                            return this.endOfData();
+                            return (BlockPos.MutableBlockPos)this.endOfData();
                         }
                         else
                         {
@@ -356,17 +356,17 @@
                 return this;
             }
 
+            public BlockPos.MutableBlockPos func_189532_c(double p_189532_1_, double p_189532_3_, double p_189532_5_)
+            {
+                return this.func_181079_c(MathHelper.func_76128_c(p_189532_1_), MathHelper.func_76128_c(p_189532_3_), MathHelper.func_76128_c(p_189532_5_));
+            }
+
             @SideOnly(Side.CLIENT)
             public BlockPos.MutableBlockPos func_189535_a(Entity p_189535_1_)
             {
                 return this.func_189532_c(p_189535_1_.field_70165_t, p_189535_1_.field_70163_u, p_189535_1_.field_70161_v);
             }
 
-            public BlockPos.MutableBlockPos func_189532_c(double p_189532_1_, double p_189532_3_, double p_189532_5_)
-            {
-                return this.func_181079_c(MathHelper.func_76128_c(p_189532_1_), MathHelper.func_76128_c(p_189532_3_), MathHelper.func_76128_c(p_189532_5_));
-            }
-
             public BlockPos.MutableBlockPos func_189533_g(Vec3i p_189533_1_)
             {
                 return this.func_181079_c(p_189533_1_.func_177958_n(), p_189533_1_.func_177956_o(), p_189533_1_.func_177952_p());
@@ -396,7 +396,7 @@
     public static final class PooledMutableBlockPos extends BlockPos.MutableBlockPos
         {
             private boolean field_185350_f;
-            private static final List<BlockPos.PooledMutableBlockPos> field_185351_g = Lists.newArrayList();
+            private static final List<BlockPos.PooledMutableBlockPos> field_185351_g = Lists.<BlockPos.PooledMutableBlockPos>newArrayList();
 
             private PooledMutableBlockPos(int p_i46586_1_, int p_i46586_2_, int p_i46586_3_)
             {
