--- before/net/minecraft/util/math/MathHelper.java
+++ after/net/minecraft/util/math/MathHelper.java
@@ -209,7 +209,7 @@
 
     public static float func_76142_g(float p_76142_0_)
     {
-        p_76142_0_ %= 360.0F;
+        p_76142_0_ = p_76142_0_ % 360.0F;
 
         if (p_76142_0_ >= 180.0F)
         {
@@ -226,7 +226,7 @@
 
     public static double func_76138_g(double p_76138_0_)
     {
-        p_76138_0_ %= 360.0D;
+        p_76138_0_ = p_76138_0_ % 360.0D;
 
         if (p_76138_0_ >= 180.0D)
         {
@@ -243,7 +243,7 @@
 
     public static int func_188209_b(int p_188209_0_)
     {
-        p_188209_0_ %= 360;
+        p_188209_0_ = p_188209_0_ % 360;
 
         if (p_188209_0_ >= 180)
         {
@@ -264,7 +264,7 @@
         {
             return Integer.parseInt(p_82715_0_);
         }
-        catch (Throwable throwable)
+        catch (Throwable var3)
         {
             return p_82715_1_;
         }
@@ -281,7 +281,7 @@
         {
             return Double.parseDouble(p_82712_0_);
         }
-        catch (Throwable throwable)
+        catch (Throwable var4)
         {
             return p_82712_1_;
         }
@@ -295,11 +295,11 @@
     public static int func_151236_b(int p_151236_0_)
     {
         int i = p_151236_0_ - 1;
-        i |= i >> 1;
-        i |= i >> 2;
-        i |= i >> 4;
-        i |= i >> 8;
-        i |= i >> 16;
+        i = i | i >> 1;
+        i = i | i >> 2;
+        i = i | i >> 4;
+        i = i | i >> 8;
+        i = i | i >> 16;
         return i + 1;
     }
 
@@ -341,6 +341,13 @@
         }
     }
 
+    public static long func_180187_c(int p_180187_0_, int p_180187_1_, int p_180187_2_)
+    {
+        long i = (long)(p_180187_0_ * 3129871) ^ (long)p_180187_2_ * 116129781L ^ (long)p_180187_1_;
+        i = i * i * 42317861L + i * 11L;
+        return i;
+    }
+
     @SideOnly(Side.CLIENT)
     public static int func_180183_b(float p_180183_0_, float p_180183_1_, float p_180183_2_)
     {
@@ -351,7 +358,8 @@
     public static int func_180181_b(int p_180181_0_, int p_180181_1_, int p_180181_2_)
     {
         int lvt_3_1_ = (p_180181_0_ << 8) + p_180181_1_;
-        return (lvt_3_1_ << 8) + p_180181_2_;
+        lvt_3_1_ = (lvt_3_1_ << 8) + p_180181_2_;
+        return lvt_3_1_;
     }
 
     @SideOnly(Side.CLIENT)
@@ -381,12 +389,6 @@
         return func_180187_c(p_180186_0_.func_177958_n(), p_180186_0_.func_177956_o(), p_180186_0_.func_177952_p());
     }
 
-    public static long func_180187_c(int p_180187_0_, int p_180187_1_, int p_180187_2_)
-    {
-        long i = (long)(p_180187_0_ * 3129871) ^ (long)p_180187_2_ * 116129781L ^ (long)p_180187_1_;
-        return i * i * 42317861L + i * 11L;
-    }
-
     public static UUID func_180182_a(Random p_180182_0_)
     {
         long i = p_180182_0_.nextLong() & -61441L | 16384L;
@@ -438,8 +440,8 @@
             }
 
             double d9 = func_181161_i(d0);
-            p_181159_2_ *= d9;
-            p_181159_0_ *= d9;
+            p_181159_2_ = p_181159_2_ * d9;
+            p_181159_0_ = p_181159_0_ * d9;
             double d2 = field_181163_d + p_181159_0_;
             int i = (int)Double.doubleToRawLongBits(d2);
             double d3 = field_181164_e[i];
@@ -474,7 +476,8 @@
         long i = Double.doubleToRawLongBits(p_181161_0_);
         i = 6910469410427058090L - (i >> 1);
         p_181161_0_ = Double.longBitsToDouble(i);
-        return p_181161_0_ * (1.5D - d0 * p_181161_0_ * p_181161_0_);
+        p_181161_0_ = p_181161_0_ * (1.5D - d0 * p_181161_0_ * p_181161_0_);
+        return p_181161_0_;
     }
 
     @SideOnly(Side.CLIENT)
@@ -533,11 +536,12 @@
 
     public static int func_188208_f(int p_188208_0_)
     {
-        p_188208_0_ ^= p_188208_0_ >>> 16;
-        p_188208_0_ *= -2048144789;
-        p_188208_0_ ^= p_188208_0_ >>> 13;
-        p_188208_0_ *= -1028477387;
-        return p_188208_0_ ^ p_188208_0_ >>> 16;
+        p_188208_0_ = p_188208_0_ ^ p_188208_0_ >>> 16;
+        p_188208_0_ = p_188208_0_ * -2048144789;
+        p_188208_0_ = p_188208_0_ ^ p_188208_0_ >>> 13;
+        p_188208_0_ = p_188208_0_ * -1028477387;
+        p_188208_0_ = p_188208_0_ ^ p_188208_0_ >>> 16;
+        return p_188208_0_;
     }
 
     static
