--- before/net/minecraft/util/math/Vec3d.java
+++ after/net/minecraft/util/math/Vec3d.java
@@ -1,9 +1,6 @@
 package net.minecraft.util.math;
 
 import javax.annotation.Nullable;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
-
 public class Vec3d
 {
     public static final Vec3d ZERO = new Vec3d(0.0, 0.0, 0.0);
@@ -58,7 +55,6 @@
         return this.x * vec.x + this.y * vec.y + this.z * vec.z;
     }
 
-    @SideOnly(Side.CLIENT)
     public Vec3d crossProduct(Vec3d vec)
     {
         return new Vec3d(
@@ -124,7 +120,6 @@
                );
     }
 
-    @SideOnly(Side.CLIENT)
     public double lengthSquared()
     {
         return this.x * this.x + this.y * this.y + this.z * this.z;
@@ -247,13 +242,11 @@
         return new Vec3d(d0, d1, d2);
     }
 
-    @SideOnly(Side.CLIENT)
     public static Vec3d fromPitchYaw(Vec2f p_189984_0_)
     {
         return fromPitchYaw(p_189984_0_.x, p_189984_0_.y);
     }
 
-    @SideOnly(Side.CLIENT)
     public static Vec3d fromPitchYaw(float p_189986_0_, float p_189986_1_)
     {
         float f = MathHelper.cos(-p_189986_1_ * (float)(Math.PI / 180.0) - (float) Math.PI);
