--- before/net/minecraft/util/text/event/ClickEvent.java
+++ after/net/minecraft/util/text/event/ClickEvent.java
@@ -69,7 +69,8 @@
     public int hashCode()
     {
         int i = this.field_150671_a.hashCode();
-        return 31 * i + (this.field_150670_b != null ? this.field_150670_b.hashCode() : 0);
+        i = 31 * i + (this.field_150670_b != null ? this.field_150670_b.hashCode() : 0);
+        return i;
     }
 
     public static enum Action
@@ -80,7 +81,7 @@
         SUGGEST_COMMAND("suggest_command", true),
         CHANGE_PAGE("change_page", true);
 
-        private static final Map<String, ClickEvent.Action> field_150679_e = Maps.newHashMap();
+        private static final Map<String, ClickEvent.Action> field_150679_e = Maps.<String, ClickEvent.Action>newHashMap();
         private final boolean field_150676_f;
         private final String field_150677_g;
 
