--- original/net/minecraft/client/renderer/entity/layers/LayerBipedArmor.java
+++ modified/net/minecraft/client/renderer/entity/layers/LayerBipedArmor.java
@@ -18,6 +18,7 @@
         this.modelArmor = new ModelBiped(1.0F);
     }
 
+    @SuppressWarnings("incomplete-switch")
     protected void setModelSlotVisible(ModelBiped p_188359_1_, EntityEquipmentSlot slotIn) {
         this.setModelVisible(p_188359_1_);
         switch (slotIn) {
@@ -44,4 +45,9 @@
     protected void setModelVisible(ModelBiped model) {
         model.setVisible(false);
     }
+
+    @Override
+    protected ModelBiped getArmorModelHook(net.minecraft.entity.EntityLivingBase entity, net.minecraft.item.ItemStack itemStack, EntityEquipmentSlot slot, ModelBiped model) {
+        return net.minecraftforge.client.ForgeHooksClient.getArmorModel(entity, itemStack, slot, model);
+    }
 }
