--- before/net/minecraft/entity/passive/AbstractHorse.java
+++ after/net/minecraft/entity/passive/AbstractHorse.java
@@ -343,6 +343,7 @@
 
         this.horseChest.addInventoryChangeListener(this);
         this.updateHorseSlots();
+        this.itemHandler = new net.minecraftforge.items.wrapper.InvWrapper(this.horseChest);
     }
 
     protected void updateHorseSlots()
@@ -1333,5 +1334,23 @@
         }
 
         return livingdata;
+    }
+
+    // FORGE
+    private net.minecraftforge.items.IItemHandler itemHandler = null; // Initialized by initHorseChest above.
+
+    @SuppressWarnings("unchecked")
+    @Override
+    @Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable net.minecraft.util.EnumFacing facing)
+    {
+        if (capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) return (T) itemHandler;
+        return super.getCapability(capability, facing);
+    }
+
+    @Override
+    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, @Nullable net.minecraft.util.EnumFacing facing)
+    {
+        return capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY || super.hasCapability(capability, facing);
     }
 }
