--- before/net/minecraft/entity/passive/AbstractChestHorse.java
+++ after/net/minecraft/entity/passive/AbstractChestHorse.java
@@ -22,7 +22,7 @@
 
 public abstract class AbstractChestHorse extends AbstractHorse
 {
-    private static final DataParameter<Boolean> field_190698_bG = EntityDataManager.func_187226_a(AbstractChestHorse.class, DataSerializers.field_187198_h);
+    private static final DataParameter<Boolean> field_190698_bG = EntityDataManager.<Boolean>func_187226_a(AbstractChestHorse.class, DataSerializers.field_187198_h);
 
     public AbstractChestHorse(World p_i47300_1_)
     {
@@ -33,25 +33,25 @@
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_190698_bG, false);
+        this.field_70180_af.func_187214_a(field_190698_bG, Boolean.valueOf(false));
     }
 
     protected void func_110147_ax()
     {
         super.func_110147_ax();
         this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a((double)this.func_110267_cL());
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a((double)0.175F);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.17499999701976776D);
         this.func_110148_a(field_110271_bv).func_111128_a(0.5D);
     }
 
     public boolean func_190695_dh()
     {
-        return this.field_70180_af.func_187225_a(field_190698_bG);
+        return ((Boolean)this.field_70180_af.func_187225_a(field_190698_bG)).booleanValue();
     }
 
     public void func_110207_m(boolean p_110207_1_)
     {
-        this.field_70180_af.func_187227_b(field_190698_bG, p_110207_1_);
+        this.field_70180_af.func_187227_b(field_190698_bG, Boolean.valueOf(p_110207_1_));
     }
 
     protected int func_190686_di()
@@ -88,7 +88,7 @@
     public static void func_190694_b(DataFixer p_190694_0_, Class<?> p_190694_1_)
     {
         AbstractHorse.func_190683_c(p_190694_0_, p_190694_1_);
-        p_190694_0_.func_188258_a(FixTypes.ENTITY, new ItemStackDataLists(p_190694_1_, "Items"));
+        p_190694_0_.func_188258_a(FixTypes.ENTITY, new ItemStackDataLists(p_190694_1_, new String[] {"Items"}));
     }
 
     public void func_70014_b(NBTTagCompound p_70014_1_)
