--- before/net/minecraft/entity/passive/EntityCow.java
+++ after/net/minecraft/entity/passive/EntityCow.java
@@ -55,7 +55,7 @@
     {
         super.func_110147_ax();
         this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(10.0D);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a((double)0.2F);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.20000000298023224D);
     }
 
     protected SoundEvent func_184639_G()
