--- before/net/minecraft/entity/passive/EntityWolf.java
+++ after/net/minecraft/entity/passive/EntityWolf.java
@@ -54,9 +54,9 @@
 
 public class EntityWolf extends EntityTameable
 {
-    private static final DataParameter<Float> field_184759_bz = EntityDataManager.func_187226_a(EntityWolf.class, DataSerializers.field_187193_c);
-    private static final DataParameter<Boolean> field_184760_bA = EntityDataManager.func_187226_a(EntityWolf.class, DataSerializers.field_187198_h);
-    private static final DataParameter<Integer> field_184758_bB = EntityDataManager.func_187226_a(EntityWolf.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Float> field_184759_bz = EntityDataManager.<Float>func_187226_a(EntityWolf.class, DataSerializers.field_187193_c);
+    private static final DataParameter<Boolean> field_184760_bA = EntityDataManager.<Boolean>func_187226_a(EntityWolf.class, DataSerializers.field_187198_h);
+    private static final DataParameter<Integer> field_184758_bB = EntityDataManager.<Integer>func_187226_a(EntityWolf.class, DataSerializers.field_187192_b);
     private float field_70926_e;
     private float field_70924_f;
     private boolean field_70925_g;
@@ -71,53 +71,50 @@
         this.func_70903_f(false);
     }
 
-    @Override
     protected void func_184651_r()
     {
         this.field_70911_d = new EntityAISit(this);
         this.field_70714_bg.func_75776_a(1, new EntityAISwimming(this));
         this.field_70714_bg.func_75776_a(2, this.field_70911_d);
-        this.field_70714_bg.func_75776_a(3, new EntityWolf.AIAvoidEntity<>(this, EntityLlama.class, 24.0F, 1.5, 1.5));
+        this.field_70714_bg.func_75776_a(3, new EntityWolf.AIAvoidEntity(this, EntityLlama.class, 24.0F, 1.5D, 1.5D));
         this.field_70714_bg.func_75776_a(4, new EntityAILeapAtTarget(this, 0.4F));
-        this.field_70714_bg.func_75776_a(5, new EntityAIAttackMelee(this, 1.0, true));
-        this.field_70714_bg.func_75776_a(6, new EntityAIFollowOwner(this, 1.0, 10.0F, 2.0F));
-        this.field_70714_bg.func_75776_a(7, new EntityAIMate(this, 1.0));
-        this.field_70714_bg.func_75776_a(8, new EntityAIWanderAvoidWater(this, 1.0));
+        this.field_70714_bg.func_75776_a(5, new EntityAIAttackMelee(this, 1.0D, true));
+        this.field_70714_bg.func_75776_a(6, new EntityAIFollowOwner(this, 1.0D, 10.0F, 2.0F));
+        this.field_70714_bg.func_75776_a(7, new EntityAIMate(this, 1.0D));
+        this.field_70714_bg.func_75776_a(8, new EntityAIWanderAvoidWater(this, 1.0D));
         this.field_70714_bg.func_75776_a(9, new EntityAIBeg(this, 8.0F));
         this.field_70714_bg.func_75776_a(10, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
         this.field_70714_bg.func_75776_a(10, new EntityAILookIdle(this));
         this.field_70715_bh.func_75776_a(1, new EntityAIOwnerHurtByTarget(this));
         this.field_70715_bh.func_75776_a(2, new EntityAIOwnerHurtTarget(this));
-        this.field_70715_bh.func_75776_a(3, new EntityAIHurtByTarget(this, true));
-        this.field_70715_bh.func_75776_a(4, new EntityAITargetNonTamed<>(this, EntityAnimal.class, false, new Predicate<Entity>()
+        this.field_70715_bh.func_75776_a(3, new EntityAIHurtByTarget(this, true, new Class[0]));
+        this.field_70715_bh.func_75776_a(4, new EntityAITargetNonTamed(this, EntityAnimal.class, false, new Predicate<Entity>()
         {
             public boolean apply(@Nullable Entity p_apply_1_)
             {
                 return p_apply_1_ instanceof EntitySheep || p_apply_1_ instanceof EntityRabbit;
             }
         }));
-        this.field_70715_bh.func_75776_a(5, new EntityAINearestAttackableTarget<>(this, AbstractSkeleton.class, false));
+        this.field_70715_bh.func_75776_a(5, new EntityAINearestAttackableTarget(this, AbstractSkeleton.class, false));
     }
 
-    @Override
     protected void func_110147_ax()
     {
         super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.3F);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.30000001192092896D);
 
         if (this.func_70909_n())
         {
-            this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(20.0);
+            this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(20.0D);
         }
         else
         {
-            this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(8.0);
+            this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(8.0D);
         }
 
-        this.func_110140_aT().func_111150_b(SharedMonsterAttributes.field_111264_e).func_111128_a(2.0);
+        this.func_110140_aT().func_111150_b(SharedMonsterAttributes.field_111264_e).func_111128_a(2.0D);
     }
 
-    @Override
     public void func_70624_b(@Nullable EntityLivingBase p_70624_1_)
     {
         super.func_70624_b(p_70624_1_);
@@ -132,22 +129,19 @@
         }
     }
 
-    @Override
     protected void func_70619_bc()
     {
-        this.field_70180_af.func_187227_b(field_184759_bz, this.func_110143_aJ());
+        this.field_70180_af.func_187227_b(field_184759_bz, Float.valueOf(this.func_110143_aJ()));
     }
 
-    @Override
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184759_bz, this.func_110143_aJ());
-        this.field_70180_af.func_187214_a(field_184760_bA, false);
-        this.field_70180_af.func_187214_a(field_184758_bB, EnumDyeColor.RED.func_176767_b());
+        this.field_70180_af.func_187214_a(field_184759_bz, Float.valueOf(this.func_110143_aJ()));
+        this.field_70180_af.func_187214_a(field_184760_bA, Boolean.valueOf(false));
+        this.field_70180_af.func_187214_a(field_184758_bB, Integer.valueOf(EnumDyeColor.RED.func_176767_b()));
     }
 
-    @Override
     protected void func_180429_a(BlockPos p_180429_1_, Block p_180429_2_)
     {
         this.func_184185_a(SoundEvents.field_187869_gK, 0.15F, 1.0F);
@@ -158,7 +152,6 @@
         EntityLiving.func_189752_a(p_189788_0_, EntityWolf.class);
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
@@ -166,7 +159,6 @@
         p_70014_1_.func_74774_a("CollarColor", (byte)this.func_175546_cu().func_176767_b());
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
@@ -178,7 +170,6 @@
         }
     }
 
-    @Override
     protected SoundEvent func_184639_G()
     {
         if (this.func_70919_bu())
@@ -187,7 +178,7 @@
         }
         else if (this.field_70146_Z.nextInt(3) == 0)
         {
-            return this.func_70909_n() && this.field_70180_af.func_187225_a(field_184759_bz) < 10.0F ? SoundEvents.field_187871_gL : SoundEvents.field_187865_gI;
+            return this.func_70909_n() && ((Float)this.field_70180_af.func_187225_a(field_184759_bz)).floatValue() < 10.0F ? SoundEvents.field_187871_gL : SoundEvents.field_187865_gI;
         }
         else
         {
@@ -195,32 +186,27 @@
         }
     }
 
-    @Override
     protected SoundEvent func_184601_bQ(DamageSource p_184601_1_)
     {
         return SoundEvents.field_187863_gH;
     }
 
-    @Override
     protected SoundEvent func_184615_bR()
     {
         return SoundEvents.field_187859_gF;
     }
 
-    @Override
     protected float func_70599_aP()
     {
         return 0.4F;
     }
 
     @Nullable
-    @Override
     protected ResourceLocation func_184647_J()
     {
         return LootTableList.field_186401_I;
     }
 
-    @Override
     public void func_70636_d()
     {
         super.func_70636_d();
@@ -239,7 +225,6 @@
         }
     }
 
-    @Override
     public void func_70071_h_()
     {
         super.func_70071_h_();
@@ -247,11 +232,11 @@
 
         if (this.func_70922_bv())
         {
-            this.field_70926_e = this.field_70926_e + (1.0F - this.field_70926_e) * 0.4F;
+            this.field_70926_e += (1.0F - this.field_70926_e) * 0.4F;
         }
         else
         {
-            this.field_70926_e = this.field_70926_e + (0.0F - this.field_70926_e) * 0.4F;
+            this.field_70926_e += (0.0F - this.field_70926_e) * 0.4F;
         }
 
         if (this.func_70026_G())
@@ -265,9 +250,7 @@
         {
             if (this.field_70929_i == 0.0F)
             {
-                this.func_184185_a(
-                    SoundEvents.field_187867_gJ, this.func_70599_aP(), (this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 1.0F
-                );
+                this.func_184185_a(SoundEvents.field_187867_gJ, this.func_70599_aP(), (this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 1.0F);
             }
 
             this.field_70927_j = this.field_70929_i;
@@ -284,22 +267,13 @@
             if (this.field_70929_i > 0.4F)
             {
                 float f = (float)this.func_174813_aQ().field_72338_b;
-                int i = (int)(MathHelper.func_76126_a((this.field_70929_i - 0.4F) * (float) Math.PI) * 7.0F);
+                int i = (int)(MathHelper.func_76126_a((this.field_70929_i - 0.4F) * (float)Math.PI) * 7.0F);
 
-                for (int j = 0; j < i; j++)
+                for (int j = 0; j < i; ++j)
                 {
                     float f1 = (this.field_70146_Z.nextFloat() * 2.0F - 1.0F) * this.field_70130_N * 0.5F;
                     float f2 = (this.field_70146_Z.nextFloat() * 2.0F - 1.0F) * this.field_70130_N * 0.5F;
-                    this.field_70170_p
-                    .func_175688_a(
-                        EnumParticleTypes.WATER_SPLASH,
-                        this.field_70165_t + (double)f1,
-                        (double)(f + 0.8F),
-                        this.field_70161_v + (double)f2,
-                        this.field_70159_w,
-                        this.field_70181_x,
-                        this.field_70179_y
-                    );
+                    this.field_70170_p.func_175688_a(EnumParticleTypes.WATER_SPLASH, this.field_70165_t + (double)f1, (double)(f + 0.8F), this.field_70161_v + (double)f2, this.field_70159_w, this.field_70181_x, this.field_70179_y);
                 }
             }
         }
@@ -331,28 +305,25 @@
             f = 1.0F;
         }
 
-        return MathHelper.func_76126_a(f * (float) Math.PI) * MathHelper.func_76126_a(f * (float) Math.PI * 11.0F) * 0.15F * (float) Math.PI;
+        return MathHelper.func_76126_a(f * (float)Math.PI) * MathHelper.func_76126_a(f * (float)Math.PI * 11.0F) * 0.15F * (float)Math.PI;
     }
 
     @SideOnly(Side.CLIENT)
     public float func_70917_k(float p_70917_1_)
     {
-        return (this.field_70924_f + (this.field_70926_e - this.field_70924_f) * p_70917_1_) * 0.15F * (float) Math.PI;
+        return (this.field_70924_f + (this.field_70926_e - this.field_70924_f) * p_70917_1_) * 0.15F * (float)Math.PI;
     }
 
-    @Override
     public float func_70047_e()
     {
         return this.field_70131_O * 0.8F;
     }
 
-    @Override
     public int func_70646_bf()
     {
         return this.func_70906_o() ? 20 : super.func_70646_bf();
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         if (this.func_180431_b(p_70097_1_))
@@ -377,12 +348,9 @@
         }
     }
 
-    @Override
     public boolean func_70652_k(Entity p_70652_1_)
     {
-        boolean flag = p_70652_1_.func_70097_a(
-                           DamageSource.func_76358_a(this), (float)((int)this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111126_e())
-                       );
+        boolean flag = p_70652_1_.func_70097_a(DamageSource.func_76358_a(this), (float)((int)this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111126_e()));
 
         if (flag)
         {
@@ -392,24 +360,22 @@
         return flag;
     }
 
-    @Override
     public void func_70903_f(boolean p_70903_1_)
     {
         super.func_70903_f(p_70903_1_);
 
         if (p_70903_1_)
         {
-            this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(20.0);
+            this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(20.0D);
         }
         else
         {
-            this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(8.0);
+            this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(8.0D);
         }
 
-        this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111128_a(4.0);
+        this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111128_a(4.0D);
     }
 
-    @Override
     public boolean func_184645_a(EntityPlayer p_184645_1_, EnumHand p_184645_2_)
     {
         ItemStack itemstack = p_184645_1_.func_184586_b(p_184645_2_);
@@ -422,7 +388,7 @@
                 {
                     ItemFood itemfood = (ItemFood)itemstack.func_77973_b();
 
-                    if (itemfood.func_77845_h() && this.field_70180_af.func_187225_a(field_184759_bz) < 20.0F)
+                    if (itemfood.func_77845_h() && ((Float)this.field_70180_af.func_187225_a(field_184759_bz)).floatValue() < 20.0F)
                     {
                         if (!p_184645_1_.field_71075_bZ.field_75098_d)
                         {
@@ -456,7 +422,7 @@
                 this.field_70911_d.func_75270_a(!this.func_70906_o());
                 this.field_70703_bu = false;
                 this.field_70699_by.func_75499_g();
-                this.func_70624_b(null);
+                this.func_70624_b((EntityLivingBase)null);
             }
         }
         else if (itemstack.func_77973_b() == Items.field_151103_aS && !this.func_70919_bu())
@@ -468,11 +434,11 @@
 
             if (!this.field_70170_p.field_72995_K)
             {
-                if (this.field_70146_Z.nextInt(3) == 0)
+                if (this.field_70146_Z.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_184645_1_))
                 {
                     this.func_193101_c(p_184645_1_);
                     this.field_70699_by.func_75499_g();
-                    this.func_70624_b(null);
+                    this.func_70624_b((EntityLivingBase)null);
                     this.field_70911_d.func_75270_a(true);
                     this.func_70606_j(20.0F);
                     this.func_70908_e(true);
@@ -492,7 +458,6 @@
     }
 
     @SideOnly(Side.CLIENT)
-    @Override
     public void func_70103_a(byte p_70103_1_)
     {
         if (p_70103_1_ == 8)
@@ -516,19 +481,15 @@
         }
         else
         {
-            return this.func_70909_n()
-                   ? (0.55F - (this.func_110138_aP() - this.field_70180_af.func_187225_a(field_184759_bz)) * 0.02F) * (float) Math.PI
-                   : (float)(Math.PI / 5);
+            return this.func_70909_n() ? (0.55F - (this.func_110138_aP() - ((Float)this.field_70180_af.func_187225_a(field_184759_bz)).floatValue()) * 0.02F) * (float)Math.PI : ((float)Math.PI / 5F);
         }
     }
 
-    @Override
     public boolean func_70877_b(ItemStack p_70877_1_)
     {
         return p_70877_1_.func_77973_b() instanceof ItemFood && ((ItemFood)p_70877_1_.func_77973_b()).func_77845_h();
     }
 
-    @Override
     public int func_70641_bl()
     {
         return 8;
@@ -536,31 +497,31 @@
 
     public boolean func_70919_bu()
     {
-        return (this.field_70180_af.func_187225_a(field_184755_bv) & 2) != 0;
+        return (((Byte)this.field_70180_af.func_187225_a(field_184755_bv)).byteValue() & 2) != 0;
     }
 
     public void func_70916_h(boolean p_70916_1_)
     {
-        byte b0 = this.field_70180_af.func_187225_a(field_184755_bv);
+        byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184755_bv)).byteValue();
 
         if (p_70916_1_)
         {
-            this.field_70180_af.func_187227_b(field_184755_bv, (byte)(b0 | 2));
+            this.field_70180_af.func_187227_b(field_184755_bv, Byte.valueOf((byte)(b0 | 2)));
         }
         else
         {
-            this.field_70180_af.func_187227_b(field_184755_bv, (byte)(b0 & -3));
+            this.field_70180_af.func_187227_b(field_184755_bv, Byte.valueOf((byte)(b0 & -3)));
         }
     }
 
     public EnumDyeColor func_175546_cu()
     {
-        return EnumDyeColor.func_176766_a(this.field_70180_af.func_187225_a(field_184758_bB) & 15);
+        return EnumDyeColor.func_176766_a(((Integer)this.field_70180_af.func_187225_a(field_184758_bB)).intValue() & 15);
     }
 
     public void func_175547_a(EnumDyeColor p_175547_1_)
     {
-        this.field_70180_af.func_187227_b(field_184758_bB, p_175547_1_.func_176767_b());
+        this.field_70180_af.func_187227_b(field_184758_bB, Integer.valueOf(p_175547_1_.func_176767_b()));
     }
 
     public EntityWolf func_90011_a(EntityAgeable p_90011_1_)
@@ -579,10 +540,9 @@
 
     public void func_70918_i(boolean p_70918_1_)
     {
-        this.field_70180_af.func_187227_b(field_184760_bA, p_70918_1_);
+        this.field_70180_af.func_187227_b(field_184760_bA, Boolean.valueOf(p_70918_1_));
     }
 
-    @Override
     public boolean func_70878_b(EntityAnimal p_70878_1_)
     {
         if (p_70878_1_ == this)
@@ -605,19 +565,22 @@
             {
                 return false;
             }
+            else if (entitywolf.func_70906_o())
+            {
+                return false;
+            }
             else
             {
-                return entitywolf.func_70906_o() ? false : this.func_70880_s() && entitywolf.func_70880_s();
+                return this.func_70880_s() && entitywolf.func_70880_s();
             }
         }
     }
 
     public boolean func_70922_bv()
     {
-        return this.field_70180_af.func_187225_a(field_184760_bA);
+        return ((Boolean)this.field_70180_af.func_187225_a(field_184760_bA)).booleanValue();
     }
 
-    @Override
     public boolean func_142018_a(EntityLivingBase p_142018_1_, EntityLivingBase p_142018_2_)
     {
         if (!(p_142018_1_ instanceof EntityCreeper) && !(p_142018_1_ instanceof EntityGhast))
@@ -632,11 +595,14 @@
                 }
             }
 
-            return p_142018_1_ instanceof EntityPlayer
-                   && p_142018_2_ instanceof EntityPlayer
-                   && !((EntityPlayer)p_142018_2_).func_96122_a((EntityPlayer)p_142018_1_)
-                   ? false
-                   : !(p_142018_1_ instanceof AbstractHorse) || !((AbstractHorse)p_142018_1_).func_110248_bS();
+            if (p_142018_1_ instanceof EntityPlayer && p_142018_2_ instanceof EntityPlayer && !((EntityPlayer)p_142018_2_).func_96122_a((EntityPlayer)p_142018_1_))
+            {
+                return false;
+            }
+            else
+            {
+                return !(p_142018_1_ instanceof AbstractHorse) || !((AbstractHorse)p_142018_1_).func_110248_bS();
+            }
         }
         else
         {
@@ -644,7 +610,6 @@
         }
     }
 
-    @Override
     public boolean func_184652_a(EntityPlayer p_184652_1_)
     {
         return !this.func_70919_bu() && super.func_184652_a(p_184652_1_);
@@ -660,12 +625,16 @@
             this.field_190856_d = p_i47251_2_;
         }
 
-        @Override
         public boolean func_75250_a()
         {
-            return super.func_75250_a() && this.field_75376_d instanceof EntityLlama
-                   ? !this.field_190856_d.func_70909_n() && this.func_190854_a((EntityLlama)this.field_75376_d)
-                   : false;
+            if (super.func_75250_a() && this.field_75376_d instanceof EntityLlama)
+            {
+                return !this.field_190856_d.func_70909_n() && this.func_190854_a((EntityLlama)this.field_75376_d);
+            }
+            else
+            {
+                return false;
+            }
         }
 
         private boolean func_190854_a(EntityLlama p_190854_1_)
@@ -673,17 +642,15 @@
             return p_190854_1_.func_190707_dL() >= EntityWolf.this.field_70146_Z.nextInt(5);
         }
 
-        @Override
         public void func_75249_e()
         {
-            EntityWolf.this.func_70624_b(null);
+            EntityWolf.this.func_70624_b((EntityLivingBase)null);
             super.func_75249_e();
         }
 
-        @Override
         public void func_75246_d()
         {
-            EntityWolf.this.func_70624_b(null);
+            EntityWolf.this.func_70624_b((EntityLivingBase)null);
             super.func_75246_d();
         }
     }
