--- before/net/minecraft/entity/passive/EntitySheep.java
+++ after/net/minecraft/entity/passive/EntitySheep.java
@@ -46,7 +46,7 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class EntitySheep extends EntityAnimal
+public class EntitySheep extends EntityAnimal implements net.minecraftforge.common.IShearable
 {
     private static final DataParameter<Byte> DYE_COLOR = EntityDataManager.createKey(EntitySheep.class, DataSerializers.BYTE);
     private final InventoryCrafting inventoryCrafting = new InventoryCrafting(new Container()
@@ -229,7 +229,7 @@
     {
         ItemStack itemstack = player.getHeldItem(hand);
 
-        if (itemstack.getItem() == Items.SHEARS && !this.getSheared() && !this.isChild())
+        if (false && itemstack.getItem() == Items.SHEARS && !this.getSheared() && !this.isChild())   //Forge: Moved to onSheared
         {
             if (!this.world.isRemote)
             {
@@ -381,6 +381,26 @@
         livingdata = super.onInitialSpawn(difficulty, livingdata);
         this.setFleeceColor(getRandomSheepColor(this.world.rand));
         return livingdata;
+    }
+
+    @Override
+    public boolean isShearable(ItemStack item, net.minecraft.world.IBlockAccess world, BlockPos pos)
+    {
+        return !this.getSheared() && !this.isChild();
+    }
+    
+    @Override
+    public java.util.List<ItemStack> onSheared(ItemStack item, net.minecraft.world.IBlockAccess world, BlockPos pos, int fortune)
+    {
+        this.setSheared(true);
+        int i = 1 + this.rand.nextInt(3);
+
+        java.util.List<ItemStack> ret = new java.util.ArrayList<ItemStack>();
+        for (int j = 0; j < i; ++j)
+            ret.add(new ItemStack(Item.getItemFromBlock(Blocks.WOOL), 1, this.getFleeceColor().getMetadata()));
+
+        this.playSound(SoundEvents.ENTITY_SHEEP_SHEAR, 1.0F, 1.0F);
+        return ret;
     }
 
     private EnumDyeColor getDyeColorMixFromParents(EntityAnimal father, EntityAnimal mother)
