--- before/net/minecraft/entity/passive/EntityWolf.java
+++ after/net/minecraft/entity/passive/EntityWolf.java
@@ -466,7 +466,7 @@
 
             if (!this.world.isRemote)
             {
-                if (this.rand.nextInt(3) == 0)
+                if (this.rand.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, player))
                 {
                     this.setTamedBy(player);
                     this.navigator.clearPath();
