--- before/net/minecraft/entity/passive/EntitySkeletonHorse.java
+++ after/net/minecraft/entity/passive/EntitySkeletonHorse.java
@@ -32,7 +32,7 @@
     {
         super.func_110147_ax();
         this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(15.0D);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a((double)0.2F);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.20000000298023224D);
         this.func_110148_a(field_110271_bv).func_111128_a(this.func_110245_cM());
     }
 
