--- before/net/minecraft/entity/passive/EntityHorse.java
+++ after/net/minecraft/entity/passive/EntityHorse.java
@@ -34,6 +34,7 @@
     private static final UUID ARMOR_MODIFIER_UUID = UUID.fromString("556E1665-8B10-40C8-8F9D-CF9B1667F295");
     private static final DataParameter<Integer> HORSE_VARIANT = EntityDataManager.createKey(EntityHorse.class, DataSerializers.VARINT);
     private static final DataParameter<Integer> HORSE_ARMOR = EntityDataManager.createKey(EntityHorse.class, DataSerializers.VARINT);
+    private static final DataParameter<ItemStack> HORSE_ARMOR_STACK = EntityDataManager.createKey(EntityHorse.class, DataSerializers.ITEM_STACK);
     private static final String[] HORSE_TEXTURES = new String[]
         {
             "textures/entity/horse/horse_white.png",
@@ -68,6 +69,7 @@
         super.entityInit();
         this.dataManager.register(HORSE_VARIANT, 0);
         this.dataManager.register(HORSE_ARMOR, HorseArmorType.NONE.getOrdinal());
+        this.dataManager.register(HORSE_ARMOR_STACK, ItemStack.EMPTY);
     }
 
     public static void registerFixesHorse(DataFixer fixer)
@@ -98,7 +100,7 @@
         {
             ItemStack itemstack = new ItemStack(compound.getCompoundTag("ArmorItem"));
 
-            if (!itemstack.isEmpty() && HorseArmorType.isHorseArmor(itemstack.getItem()))
+            if (!itemstack.isEmpty() && isArmor(itemstack))
             {
                 this.horseChest.setInventorySlotContents(1, itemstack);
             }
@@ -129,11 +131,12 @@
         int i = this.getHorseVariant();
         int j = (i & 0xFF) % 7;
         int k = ((i & 0xFF00) >> 8) % 5;
-        HorseArmorType horsearmortype = this.getHorseArmorType();
+        ItemStack armorStack = this.dataManager.get(HORSE_ARMOR_STACK);
+        String texture = !armorStack.isEmpty() ? armorStack.getItem().getHorseArmorTexture(this, armorStack) : HorseArmorType.getByOrdinal(this.dataManager.get(HORSE_ARMOR)).getTextureName(); //If armorStack is empty, the server is vanilla so the texture should be determined the vanilla way
         this.horseTexturesArray[0] = HORSE_TEXTURES[j];
         this.horseTexturesArray[1] = HORSE_MARKING_TEXTURES[k];
-        this.horseTexturesArray[2] = horsearmortype.getTextureName();
-        this.texturePrefix = "horse/" + HORSE_TEXTURES_ABBR[j] + HORSE_MARKING_TEXTURES_ABBR[k] + horsearmortype.getHash();
+        this.horseTexturesArray[2] = texture;
+        this.texturePrefix = "horse/" + HORSE_TEXTURES_ABBR[j] + HORSE_MARKING_TEXTURES_ABBR[k] + texture;
     }
 
     @SideOnly(Side.CLIENT)
@@ -169,6 +172,7 @@
     {
         HorseArmorType horsearmortype = HorseArmorType.getByItemStack(itemStackIn);
         this.dataManager.set(HORSE_ARMOR, horsearmortype.getOrdinal());
+        this.dataManager.set(HORSE_ARMOR_STACK, itemStackIn);
         this.resetTexturePrefix();
 
         if (!this.world.isRemote)
@@ -179,14 +183,16 @@
             if (i != 0)
             {
                 this.getEntityAttribute(SharedMonsterAttributes.ARMOR)
-                .applyModifier(new AttributeModifier(ARMOR_MODIFIER_UUID, "Horse armor bonus", (double)i, 0).setSaved(false));
+                .applyModifier((new AttributeModifier(ARMOR_MODIFIER_UUID, "Horse armor bonus", (double)i, 0)).setSaved(false));
             }
         }
     }
 
     public HorseArmorType getHorseArmorType()
     {
-        return HorseArmorType.getByOrdinal(this.dataManager.get(HORSE_ARMOR));
+        HorseArmorType armor = HorseArmorType.getByItemStack(this.dataManager.get(HORSE_ARMOR_STACK)); //First check the Forge armor DataParameter
+        if (armor == HorseArmorType.NONE) armor = HorseArmorType.getByOrdinal(this.dataManager.get(HORSE_ARMOR)); //If the Forge armor DataParameter returns NONE, fallback to the vanilla armor DataParameter. This is necessary to prevent issues with Forge clients connected to vanilla servers.
+        return armor;
     }
 
     @Override
@@ -232,6 +238,8 @@
             this.dataManager.setClean();
             this.resetTexturePrefix();
         }
+        ItemStack armor = this.horseChest.getStackInSlot(1);
+        if (isArmor(armor)) armor.getItem().onHorseArmorTick(world, this, armor);
     }
 
     @Override
@@ -414,7 +422,7 @@
     @Override
     public boolean isArmor(ItemStack stack)
     {
-        return HorseArmorType.isHorseArmor(stack.getItem());
+        return HorseArmorType.isHorseArmor(stack);
     }
 
     @Nullable
