--- before/net/minecraft/entity/passive/EntityOcelot.java
+++ after/net/minecraft/entity/passive/EntityOcelot.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.passive;
 
+import com.google.common.base.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.block.Block;
 import net.minecraft.block.material.Material;
@@ -43,7 +44,7 @@
 
 public class EntityOcelot extends EntityTameable
 {
-    private static final DataParameter<Integer> field_184757_bz = EntityDataManager.func_187226_a(EntityOcelot.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Integer> field_184757_bz = EntityDataManager.<Integer>func_187226_a(EntityOcelot.class, DataSerializers.field_187192_b);
     private EntityAIAvoidEntity<EntityPlayer> field_175545_bm;
     private EntityAITempt field_70914_e;
 
@@ -53,44 +54,41 @@
         this.func_70105_a(0.6F, 0.7F);
     }
 
-    @Override
     protected void func_184651_r()
     {
         this.field_70911_d = new EntityAISit(this);
-        this.field_70914_e = new EntityAITempt(this, 0.6, Items.field_151115_aP, true);
+        this.field_70914_e = new EntityAITempt(this, 0.6D, Items.field_151115_aP, true);
         this.field_70714_bg.func_75776_a(1, new EntityAISwimming(this));
         this.field_70714_bg.func_75776_a(2, this.field_70911_d);
         this.field_70714_bg.func_75776_a(3, this.field_70914_e);
-        this.field_70714_bg.func_75776_a(5, new EntityAIFollowOwner(this, 1.0, 10.0F, 5.0F));
-        this.field_70714_bg.func_75776_a(6, new EntityAIOcelotSit(this, 0.8));
+        this.field_70714_bg.func_75776_a(5, new EntityAIFollowOwner(this, 1.0D, 10.0F, 5.0F));
+        this.field_70714_bg.func_75776_a(6, new EntityAIOcelotSit(this, 0.8D));
         this.field_70714_bg.func_75776_a(7, new EntityAILeapAtTarget(this, 0.3F));
         this.field_70714_bg.func_75776_a(8, new EntityAIOcelotAttack(this));
-        this.field_70714_bg.func_75776_a(9, new EntityAIMate(this, 0.8));
-        this.field_70714_bg.func_75776_a(10, new EntityAIWanderAvoidWater(this, 0.8, 1.0000001E-5F));
+        this.field_70714_bg.func_75776_a(9, new EntityAIMate(this, 0.8D));
+        this.field_70714_bg.func_75776_a(10, new EntityAIWanderAvoidWater(this, 0.8D, 1.0000001E-5F));
         this.field_70714_bg.func_75776_a(11, new EntityAIWatchClosest(this, EntityPlayer.class, 10.0F));
-        this.field_70715_bh.func_75776_a(1, new EntityAITargetNonTamed<>(this, EntityChicken.class, false, null));
+        this.field_70715_bh.func_75776_a(1, new EntityAITargetNonTamed(this, EntityChicken.class, false, (Predicate)null));
     }
 
-    @Override
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184757_bz, 0);
+        this.field_70180_af.func_187214_a(field_184757_bz, Integer.valueOf(0));
     }
 
-    @Override
     public void func_70619_bc()
     {
         if (this.func_70605_aq().func_75640_a())
         {
             double d0 = this.func_70605_aq().func_75638_b();
 
-            if (d0 == 0.6)
+            if (d0 == 0.6D)
             {
                 this.func_70095_a(true);
                 this.func_70031_b(false);
             }
-            else if (d0 == 1.33)
+            else if (d0 == 1.33D)
             {
                 this.func_70095_a(false);
                 this.func_70031_b(true);
@@ -108,21 +106,18 @@
         }
     }
 
-    @Override
     protected boolean func_70692_ba()
     {
         return !this.func_70909_n() && this.field_70173_aa > 2400;
     }
 
-    @Override
     protected void func_110147_ax()
     {
         super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(10.0);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.3F);
+        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(10.0D);
+        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.30000001192092896D);
     }
 
-    @Override
     public void func_180430_e(float p_180430_1_, float p_180430_2_)
     {
     }
@@ -132,14 +127,12 @@
         EntityLiving.func_189752_a(p_189787_0_, EntityOcelot.class);
     }
 
-    @Override
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
         super.func_70014_b(p_70014_1_);
         p_70014_1_.func_74768_a("CatType", this.func_70913_u());
     }
 
-    @Override
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
         super.func_70037_a(p_70037_1_);
@@ -147,7 +140,6 @@
     }
 
     @Nullable
-    @Override
     protected SoundEvent func_184639_G()
     {
         if (this.func_70909_n())
@@ -167,31 +159,26 @@
         }
     }
 
-    @Override
     protected SoundEvent func_184601_bQ(DamageSource p_184601_1_)
     {
         return SoundEvents.field_187642_Q;
     }
 
-    @Override
     protected SoundEvent func_184615_bR()
     {
         return SoundEvents.field_187639_P;
     }
 
-    @Override
     protected float func_70599_aP()
     {
         return 0.4F;
     }
 
-    @Override
     public boolean func_70652_k(Entity p_70652_1_)
     {
         return p_70652_1_.func_70097_a(DamageSource.func_76358_a(this), 3.0F);
     }
 
-    @Override
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
         if (this.func_180431_b(p_70097_1_))
@@ -210,13 +197,11 @@
     }
 
     @Nullable
-    @Override
     protected ResourceLocation func_184647_J()
     {
         return LootTableList.field_186402_J;
     }
 
-    @Override
     public boolean func_184645_a(EntityPlayer p_184645_1_, EnumHand p_184645_2_)
     {
         ItemStack itemstack = p_184645_1_.func_184586_b(p_184645_2_);
@@ -228,9 +213,7 @@
                 this.field_70911_d.func_75270_a(!this.func_70906_o());
             }
         }
-        else if ((this.field_70914_e == null || this.field_70914_e.func_75277_f())
-                 && itemstack.func_77973_b() == Items.field_151115_aP
-                 && p_184645_1_.func_70068_e(this) < 9.0)
+        else if ((this.field_70914_e == null || this.field_70914_e.func_75277_f()) && itemstack.func_77973_b() == Items.field_151115_aP && p_184645_1_.func_70068_e(this) < 9.0D)
         {
             if (!p_184645_1_.field_71075_bZ.field_75098_d)
             {
@@ -239,7 +222,7 @@
 
             if (!this.field_70170_p.field_72995_K)
             {
-                if (this.field_70146_Z.nextInt(3) == 0)
+                if (this.field_70146_Z.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_184645_1_))
                 {
                     this.func_193101_c(p_184645_1_);
                     this.func_70912_b(1 + this.field_70170_p.field_73012_v.nextInt(3));
@@ -274,13 +257,11 @@
         return entityocelot;
     }
 
-    @Override
     public boolean func_70877_b(ItemStack p_70877_1_)
     {
         return p_70877_1_.func_77973_b() == Items.field_151115_aP;
     }
 
-    @Override
     public boolean func_70878_b(EntityAnimal p_70878_1_)
     {
         if (p_70878_1_ == this)
@@ -312,26 +293,22 @@
 
     public int func_70913_u()
     {
-        return this.field_70180_af.func_187225_a(field_184757_bz);
+        return ((Integer)this.field_70180_af.func_187225_a(field_184757_bz)).intValue();
     }
 
     public void func_70912_b(int p_70912_1_)
     {
-        this.field_70180_af.func_187227_b(field_184757_bz, p_70912_1_);
+        this.field_70180_af.func_187227_b(field_184757_bz, Integer.valueOf(p_70912_1_));
     }
 
-    @Override
     public boolean func_70601_bi()
     {
         return this.field_70170_p.field_73012_v.nextInt(3) != 0;
     }
 
-    @Override
     public boolean func_70058_J()
     {
-        if (this.field_70170_p.func_72917_a(this.func_174813_aQ(), this)
-                && this.field_70170_p.func_184144_a(this, this.func_174813_aQ()).isEmpty()
-                && !this.field_70170_p.func_72953_d(this.func_174813_aQ()))
+        if (this.field_70170_p.func_72917_a(this.func_174813_aQ(), this) && this.field_70170_p.func_184144_a(this, this.func_174813_aQ()).isEmpty() && !this.field_70170_p.func_72953_d(this.func_174813_aQ()))
         {
             BlockPos blockpos = new BlockPos(this.field_70165_t, this.func_174813_aQ().field_72338_b, this.field_70161_v);
 
@@ -343,7 +320,7 @@
             IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos.func_177977_b());
             Block block = iblockstate.func_177230_c();
 
-            if (block == Blocks.field_150349_c || iblockstate.func_185904_a() == Material.field_151584_j)
+            if (block == Blocks.field_150349_c || block.isLeaves(iblockstate, this.field_70170_p, blockpos.func_177977_b()))
             {
                 return true;
             }
@@ -352,7 +329,6 @@
         return false;
     }
 
-    @Override
     public String func_70005_c_()
     {
         if (this.func_145818_k_())
@@ -365,12 +341,11 @@
         }
     }
 
-    @Override
     protected void func_175544_ck()
     {
         if (this.field_175545_bm == null)
         {
-            this.field_175545_bm = new EntityAIAvoidEntity<>(this, EntityPlayer.class, 16.0F, 0.8, 1.33);
+            this.field_175545_bm = new EntityAIAvoidEntity<EntityPlayer>(this, EntityPlayer.class, 16.0F, 0.8D, 1.33D);
         }
 
         this.field_70714_bg.func_85156_a(this.field_175545_bm);
@@ -382,7 +357,6 @@
     }
 
     @Nullable
-    @Override
     public IEntityLivingData func_180482_a(DifficultyInstance p_180482_1_, @Nullable IEntityLivingData p_180482_2_)
     {
         p_180482_2_ = super.func_180482_a(p_180482_1_, p_180482_2_);
