--- before/net/minecraft/entity/passive/EntityRabbit.java
+++ after/net/minecraft/entity/passive/EntityRabbit.java
@@ -592,7 +592,7 @@
             {
                 if (this.runDelay <= 0)
                 {
-                    if (!this.rabbit.world.getGameRules().getBoolean("mobGriefing"))
+                    if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.rabbit.world, this.rabbit))
                     {
                         return false;
                     }
