--- before/net/minecraft/entity/passive/EntityChicken.java
+++ after/net/minecraft/entity/passive/EntityChicken.java
@@ -182,8 +182,8 @@
     public void func_184232_k(Entity p_184232_1_)
     {
         super.func_184232_k(p_184232_1_);
-        float f = MathHelper.func_76126_a(this.field_70761_aq * ((float)Math.PI / 180F));
-        float f1 = MathHelper.func_76134_b(this.field_70761_aq * ((float)Math.PI / 180F));
+        float f = MathHelper.func_76126_a(this.field_70761_aq * 0.017453292F);
+        float f1 = MathHelper.func_76134_b(this.field_70761_aq * 0.017453292F);
         float f2 = 0.1F;
         float f3 = 0.0F;
         p_184232_1_.func_70107_b(this.field_70165_t + (double)(0.1F * f), this.field_70163_u + (double)(this.field_70131_O * 0.5F) + p_184232_1_.func_70033_W() + 0.0D, this.field_70161_v - (double)(0.1F * f1));
