--- before/net/minecraft/entity/passive/EntityPig.java
+++ after/net/minecraft/entity/passive/EntityPig.java
@@ -40,8 +40,8 @@
 
 public class EntityPig extends EntityAnimal
 {
-    private static final DataParameter<Boolean> field_184763_bv = EntityDataManager.func_187226_a(EntityPig.class, DataSerializers.field_187198_h);
-    private static final DataParameter<Integer> field_191520_bx = EntityDataManager.func_187226_a(EntityPig.class, DataSerializers.field_187192_b);
+    private static final DataParameter<Boolean> field_184763_bv = EntityDataManager.<Boolean>func_187226_a(EntityPig.class, DataSerializers.field_187198_h);
+    private static final DataParameter<Integer> field_191520_bx = EntityDataManager.<Integer>func_187226_a(EntityPig.class, DataSerializers.field_187192_b);
     private static final Set<Item> field_184764_bw = Sets.newHashSet(Items.field_151172_bF, Items.field_151174_bG, Items.field_185164_cV);
     private boolean field_184765_bx;
     private int field_184766_bz;
@@ -76,7 +76,7 @@
     @Nullable
     public Entity func_184179_bs()
     {
-        return this.func_184188_bt().isEmpty() ? null : this.func_184188_bt().get(0);
+        return this.func_184188_bt().isEmpty() ? null : (Entity)this.func_184188_bt().get(0);
     }
 
     public boolean func_82171_bF()
@@ -100,7 +100,7 @@
         {
             this.field_184765_bx = true;
             this.field_184766_bz = 0;
-            this.field_184767_bA = this.field_70180_af.func_187225_a(field_191520_bx);
+            this.field_184767_bA = ((Integer)this.field_70180_af.func_187225_a(field_191520_bx)).intValue();
         }
 
         super.func_184206_a(p_184206_1_);
@@ -109,8 +109,8 @@
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184763_bv, false);
-        this.field_70180_af.func_187214_a(field_191520_bx, 0);
+        this.field_70180_af.func_187214_a(field_184763_bv, Boolean.valueOf(false));
+        this.field_70180_af.func_187214_a(field_191520_bx, Integer.valueOf(0));
     }
 
     public static void func_189792_b(DataFixer p_189792_0_)
@@ -207,18 +207,18 @@
 
     public boolean func_70901_n()
     {
-        return this.field_70180_af.func_187225_a(field_184763_bv);
+        return ((Boolean)this.field_70180_af.func_187225_a(field_184763_bv)).booleanValue();
     }
 
     public void func_70900_e(boolean p_70900_1_)
     {
         if (p_70900_1_)
         {
-            this.field_70180_af.func_187227_b(field_184763_bv, true);
+            this.field_70180_af.func_187227_b(field_184763_bv, Boolean.valueOf(true));
         }
         else
         {
-            this.field_70180_af.func_187227_b(field_184763_bv, false);
+            this.field_70180_af.func_187227_b(field_184763_bv, Boolean.valueOf(false));
         }
     }
 
@@ -244,7 +244,7 @@
 
     public void func_191986_a(float p_191986_1_, float p_191986_2_, float p_191986_3_)
     {
-        Entity entity = this.func_184188_bt().isEmpty() ? null : this.func_184188_bt().get(0);
+        Entity entity = this.func_184188_bt().isEmpty() ? null : (Entity)this.func_184188_bt().get(0);
 
         if (this.func_184207_aI() && this.func_82171_bF())
         {
@@ -313,7 +313,7 @@
             this.field_184765_bx = true;
             this.field_184766_bz = 0;
             this.field_184767_bA = this.func_70681_au().nextInt(841) + 140;
-            this.func_184212_Q().func_187227_b(field_191520_bx, this.field_184767_bA);
+            this.func_184212_Q().func_187227_b(field_191520_bx, Integer.valueOf(this.field_184767_bA));
             return true;
         }
     }
