--- before/net/minecraft/entity/passive/EntityVillager.java
+++ after/net/minecraft/entity/passive/EntityVillager.java
@@ -109,7 +109,7 @@
     private boolean needsInitilization;
     private boolean isWillingToMate;
     private int wealth;
-    private String lastBuyingPlayer;
+    private java.util.UUID lastBuyingPlayer;
     private int careerId;
     private int careerLevel;
     private boolean isLookingForHome;
@@ -487,7 +487,8 @@
         else if (!this.holdingSpawnEggOfClass(itemstack, (Class <? extends Entity >)this.getClass())
                  && this.isEntityAlive()
                  && !this.isTrading()
-                 && !this.isChild())
+                 && !this.isChild()
+                 && !player.isSneaking())
         {
             if (this.buyingList == null)
             {
@@ -566,6 +567,7 @@
     {
         super.writeEntityToNBT(compound);
         compound.setInteger("Profession", this.getProfession());
+        compound.setString("ProfessionName", this.getProfessionForge().getRegistryName().toString());
         compound.setInteger("Riches", this.wealth);
         compound.setInteger("Career", this.careerId);
         compound.setInteger("CareerLevel", this.careerLevel);
@@ -596,6 +598,14 @@
     {
         super.readEntityFromNBT(compound);
         this.setProfession(compound.getInteger("Profession"));
+        if (compound.hasKey("ProfessionName"))
+        {
+            net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession p =
+                net.minecraftforge.fml.common.registry.ForgeRegistries.VILLAGER_PROFESSIONS.getValue(new net.minecraft.util.ResourceLocation(compound.getString("ProfessionName")));
+            if (p == null)
+                p = net.minecraftforge.fml.common.registry.ForgeRegistries.VILLAGER_PROFESSIONS.getValue(new net.minecraft.util.ResourceLocation("minecraft:farmer"));
+            this.setProfession(p);
+        }
         this.wealth = compound.getInteger("Riches");
         this.careerId = compound.getInteger("Career");
         this.careerLevel = compound.getInteger("CareerLevel");
@@ -656,12 +666,42 @@
 
     public void setProfession(int professionId)
     {
-        this.dataManager.set(PROFESSION, professionId);
+        this.dataManager.set(PROFESSION, Integer.valueOf(professionId));
+        net.minecraftforge.fml.common.registry.VillagerRegistry.onSetProfession(this, professionId);
     }
 
+    @Deprecated //Use Forge Variant below
     public int getProfession()
     {
-        return Math.max(this.dataManager.get(PROFESSION) % 6, 0);
+        return Math.max(((Integer)this.dataManager.get(PROFESSION)).intValue(), 0);
+    }
+
+    private net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession prof;
+    public void setProfession(net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession prof)
+    {
+        this.prof = prof;
+        this.setProfession(net.minecraftforge.fml.common.registry.VillagerRegistry.getId(prof));
+    }
+
+    public net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession getProfessionForge()
+    {
+        if (this.prof == null)
+        {
+            this.prof = net.minecraftforge.fml.common.registry.VillagerRegistry.getById(this.getProfession());
+            if (this.prof == null)
+                return net.minecraftforge.fml.common.registry.VillagerRegistry.getById(0); //Farmer
+        }
+        return this.prof;
+    }
+
+    @Override
+    public void notifyDataManagerChange(DataParameter<?> key)
+    {
+        super.notifyDataManagerChange(key);
+        if (key.equals(PROFESSION))
+        {
+            net.minecraftforge.fml.common.registry.VillagerRegistry.onSetProfession(this, this.dataManager.get(PROFESSION));
+        }
     }
 
     public boolean isMating()
@@ -702,7 +742,7 @@
                     i = -3;
                 }
 
-                this.village.modifyPlayerReputation(livingBase.getName(), i);
+                this.village.modifyPlayerReputation(livingBase.getUniqueID(), i);
 
                 if (this.isEntityAlive())
                 {
@@ -723,7 +763,7 @@
             {
                 if (entity instanceof EntityPlayer)
                 {
-                    this.village.modifyPlayerReputation(entity.getName(), -2);
+                    this.village.modifyPlayerReputation(entity.getUniqueID(), -2);
                 }
                 else if (entity instanceof IMob)
                 {
@@ -820,7 +860,7 @@
 
             if (this.buyingPlayer != null)
             {
-                this.lastBuyingPlayer = this.buyingPlayer.getName();
+                this.lastBuyingPlayer = this.buyingPlayer.getUniqueID();
             }
             else
             {
@@ -865,20 +905,18 @@
             this.populateBuyingList();
         }
 
-        return this.buyingList;
+        return net.minecraftforge.event.ForgeEventFactory.listTradeOffers(this, player, buyingList);
     }
 
     private void populateBuyingList()
     {
-        EntityVillager.ITradeList[][][] aentityvillager$itradelist = DEFAULT_TRADE_LIST_MAP[this.getProfession()];
-
         if (this.careerId != 0 && this.careerLevel != 0)
         {
             this.careerLevel++;
         }
         else
         {
-            this.careerId = this.rand.nextInt(aentityvillager$itradelist.length) + 1;
+            this.careerId = this.getProfessionForge().getRandomCareer(this.rand) + 1;
             this.careerLevel = 1;
         }
 
@@ -889,19 +927,13 @@
 
         int i = this.careerId - 1;
         int j = this.careerLevel - 1;
+        java.util.List<EntityVillager.ITradeList> trades = this.getProfessionForge().getCareer(i).getTrades(j);
 
-        if (i >= 0 && i < aentityvillager$itradelist.length)
+        if (trades != null)
         {
-            EntityVillager.ITradeList[][] aentityvillager$itradelist1 = aentityvillager$itradelist[i];
-
-            if (j >= 0 && j < aentityvillager$itradelist1.length)
+            for (EntityVillager.ITradeList entityvillager$itradelist : trades)
             {
-                EntityVillager.ITradeList[] aentityvillager$itradelist2 = aentityvillager$itradelist1[j];
-
-                for (EntityVillager.ITradeList entityvillager$itradelist : aentityvillager$itradelist2)
-                {
-                    entityvillager$itradelist.addMerchantRecipe(this, this.buyingList, this.rand);
-                }
+                entityvillager$itradelist.addMerchantRecipe(this, this.buyingList, this.rand);
             }
         }
     }
@@ -1015,7 +1047,7 @@
                     s1 = "nitwit";
             }
 
-            if (s1 != null)
+            s1 = this.getProfessionForge().getCareer(this.careerId-1).getName();
             {
                 ITextComponent itextcomponent = new TextComponentTranslation("entity.Villager." + s1);
                 itextcomponent.getStyle().setHoverEvent(this.getHoverEvent());
@@ -1028,10 +1060,6 @@
 
                 return itextcomponent;
             }
-            else
-            {
-                return super.getDisplayName();
-            }
         }
     }
 
@@ -1097,7 +1125,7 @@
 
         if (p_190672_3_)
         {
-            this.setProfession(this.world.rand.nextInt(6));
+            net.minecraftforge.fml.common.registry.VillagerRegistry.setRandomProfession(this, this.world.rand);
         }
 
         this.setAdditionalAItasks();
@@ -1194,7 +1222,15 @@
     public boolean wantsMoreFood()
     {
         boolean flag = this.getProfession() == 0;
-        return flag ? !this.hasEnoughItems(5) : !this.hasEnoughItems(1);
+
+        if (flag)
+        {
+            return !this.hasEnoughItems(5);
+        }
+        else
+        {
+            return !this.hasEnoughItems(1);
+        }
     }
 
     private boolean hasEnoughItems(int multiplier)
@@ -1381,6 +1417,9 @@
                 recipeList.add(new MerchantRecipe(itemstack, itemstack1));
             }
         }
+    //MODDERS DO NOT USE OR EDIT THIS IN ANY WAY IT WILL HAVE NO EFFECT, THIS IS JUST IN HERE TO ALLOW FORGE TO ACCESS IT
+    @Deprecated
+    public static ITradeList[][][][] GET_TRADES_DONT_USE(){ return DEFAULT_TRADE_LIST_MAP; }
 
     public static class ListItemForEmeralds implements EntityVillager.ITradeList
         {
