--- before/net/minecraft/entity/passive/EntityTameable.java
+++ after/net/minecraft/entity/passive/EntityTameable.java
@@ -24,8 +24,8 @@
 
 public abstract class EntityTameable extends EntityAnimal implements IEntityOwnable
 {
-    protected static final DataParameter<Byte> field_184755_bv = EntityDataManager.func_187226_a(EntityTameable.class, DataSerializers.field_187191_a);
-    protected static final DataParameter<Optional<UUID>> field_184756_bw = EntityDataManager.func_187226_a(EntityTameable.class, DataSerializers.field_187203_m);
+    protected static final DataParameter<Byte> field_184755_bv = EntityDataManager.<Byte>func_187226_a(EntityTameable.class, DataSerializers.field_187191_a);
+    protected static final DataParameter<Optional<UUID>> field_184756_bw = EntityDataManager.<Optional<UUID>>func_187226_a(EntityTameable.class, DataSerializers.field_187203_m);
     protected EntityAISit field_70911_d;
 
     public EntityTameable(World p_i1604_1_)
@@ -37,7 +37,7 @@
     protected void func_70088_a()
     {
         super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184755_bv, (byte)0);
+        this.field_70180_af.func_187214_a(field_184755_bv, Byte.valueOf((byte)0));
         this.field_70180_af.func_187214_a(field_184756_bw, Optional.absent());
     }
 
@@ -79,7 +79,7 @@
                 this.func_184754_b(UUID.fromString(s));
                 this.func_70903_f(true);
             }
-            catch (Throwable throwable)
+            catch (Throwable var4)
             {
                 this.func_70903_f(false);
             }
@@ -135,20 +135,20 @@
 
     public boolean func_70909_n()
     {
-        return (this.field_70180_af.func_187225_a(field_184755_bv) & 4) != 0;
+        return (((Byte)this.field_70180_af.func_187225_a(field_184755_bv)).byteValue() & 4) != 0;
     }
 
     public void func_70903_f(boolean p_70903_1_)
     {
-        byte b0 = this.field_70180_af.func_187225_a(field_184755_bv);
+        byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184755_bv)).byteValue();
 
         if (p_70903_1_)
         {
-            this.field_70180_af.func_187227_b(field_184755_bv, (byte)(b0 | 4));
+            this.field_70180_af.func_187227_b(field_184755_bv, Byte.valueOf((byte)(b0 | 4)));
         }
         else
         {
-            this.field_70180_af.func_187227_b(field_184755_bv, (byte)(b0 & -5));
+            this.field_70180_af.func_187227_b(field_184755_bv, Byte.valueOf((byte)(b0 & -5)));
         }
 
         this.func_175544_ck();
@@ -160,27 +160,27 @@
 
     public boolean func_70906_o()
     {
-        return (this.field_70180_af.func_187225_a(field_184755_bv) & 1) != 0;
+        return (((Byte)this.field_70180_af.func_187225_a(field_184755_bv)).byteValue() & 1) != 0;
     }
 
     public void func_70904_g(boolean p_70904_1_)
     {
-        byte b0 = this.field_70180_af.func_187225_a(field_184755_bv);
+        byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184755_bv)).byteValue();
 
         if (p_70904_1_)
         {
-            this.field_70180_af.func_187227_b(field_184755_bv, (byte)(b0 | 1));
+            this.field_70180_af.func_187227_b(field_184755_bv, Byte.valueOf((byte)(b0 | 1)));
         }
         else
         {
-            this.field_70180_af.func_187227_b(field_184755_bv, (byte)(b0 & -2));
+            this.field_70180_af.func_187227_b(field_184755_bv, Byte.valueOf((byte)(b0 & -2)));
         }
     }
 
     @Nullable
     public UUID func_184753_b()
     {
-        return this.field_70180_af.func_187225_a(field_184756_bw).orNull();
+        return (UUID)((Optional)this.field_70180_af.func_187225_a(field_184756_bw)).orNull();
     }
 
     public void func_184754_b(@Nullable UUID p_184754_1_)
@@ -207,7 +207,7 @@
             UUID uuid = this.func_184753_b();
             return uuid == null ? null : this.field_70170_p.func_152378_a(uuid);
         }
-        catch (IllegalArgumentException illegalargumentexception)
+        catch (IllegalArgumentException var2)
         {
             return null;
         }
