--- before/net/minecraft/server/gui/MinecraftServerGui.java
+++ after/net/minecraft/server/gui/MinecraftServerGui.java
@@ -2,6 +2,7 @@
 
 import com.mojang.util.QueueLogAppender;
 import java.awt.BorderLayout;
+import java.awt.Component;
 import java.awt.Dimension;
 import java.awt.Font;
 import java.awt.event.ActionEvent;
@@ -22,6 +23,7 @@
 import javax.swing.UIManager;
 import javax.swing.border.EtchedBorder;
 import javax.swing.border.TitledBorder;
+import javax.swing.text.AttributeSet;
 import javax.swing.text.BadLocationException;
 import javax.swing.text.Document;
 import net.minecraft.server.dedicated.DedicatedServer;
@@ -51,11 +53,10 @@
         JFrame jframe = new JFrame("Minecraft server");
         jframe.add(minecraftservergui);
         jframe.pack();
-        jframe.setLocationRelativeTo(null);
+        jframe.setLocationRelativeTo((Component)null);
         jframe.setVisible(true);
         jframe.addWindowListener(new WindowAdapter()
         {
-            @Override
             public void windowClosing(WindowEvent p_windowClosing_1_)
             {
                 p_120016_0_.func_71263_m();
@@ -75,6 +76,7 @@
                 System.exit(0);
             }
         });
+        minecraftservergui.latch.countDown();
     }
 
     public MinecraftServerGui(DedicatedServer p_i2362_1_)
@@ -121,7 +123,6 @@
         final JTextField jtextfield = new JTextField();
         jtextfield.addActionListener(new ActionListener()
         {
-            @Override
             public void actionPerformed(ActionEvent p_actionPerformed_1_)
             {
                 String s = jtextfield.getText().trim();
@@ -136,7 +137,6 @@
         });
         jtextarea.addFocusListener(new FocusAdapter()
         {
-            @Override
             public void focusGained(FocusEvent p_focusGained_1_)
             {
             }
@@ -146,7 +146,6 @@
         jpanel.setBorder(new TitledBorder(new EtchedBorder(), "Log and chat"));
         Thread thread = new Thread(new Runnable()
         {
-            @Override
             public void run()
             {
                 String s;
@@ -162,13 +161,17 @@
         return jpanel;
     }
 
+    private java.util.concurrent.CountDownLatch latch = new java.util.concurrent.CountDownLatch(1);
     public void func_164247_a(final JTextArea p_164247_1_, final JScrollPane p_164247_2_, final String p_164247_3_)
     {
+        try
+        {
+            latch.await();
+        } catch (InterruptedException e){} //Prevent logging until after constructor has ended.
         if (!SwingUtilities.isEventDispatchThread())
         {
             SwingUtilities.invokeLater(new Runnable()
             {
-                @Override
                 public void run()
                 {
                     MinecraftServerGui.this.func_164247_a(p_164247_1_, p_164247_2_, p_164247_3_);
@@ -188,7 +191,7 @@
 
             try
             {
-                document.insertString(document.getLength(), p_164247_3_, null);
+                document.insertString(document.getLength(), p_164247_3_, (AttributeSet)null);
             }
             catch (BadLocationException badlocationexception)
             {
