--- before/net/minecraft/world/storage/loot/LootEntryEmpty.java
+++ after/net/minecraft/world/storage/loot/LootEntryEmpty.java
@@ -10,25 +10,21 @@
 
 public class LootEntryEmpty extends LootEntry
 {
-    public LootEntryEmpty(int weightIn, int qualityIn, LootCondition[] conditionsIn)
+    public LootEntryEmpty(int weightIn, int qualityIn, LootCondition[] conditionsIn, String entryName)
     {
-        super(weightIn, qualityIn, conditionsIn);
+        super(weightIn, qualityIn, conditionsIn, entryName);
     }
 
-    @Override
     public void addLoot(Collection<ItemStack> stacks, Random rand, LootContext context)
     {
     }
 
-    @Override
     protected void serialize(JsonObject json, JsonSerializationContext context)
     {
     }
 
-    public static LootEntryEmpty deserialize(
-            JsonObject object, JsonDeserializationContext deserializationContext, int weightIn, int qualityIn, LootCondition[] conditionsIn
-        )
+    public static LootEntryEmpty deserialize(JsonObject object, JsonDeserializationContext deserializationContext, int weightIn, int qualityIn, LootCondition[] conditionsIn)
     {
-        return new LootEntryEmpty(weightIn, qualityIn, conditionsIn);
+        return new LootEntryEmpty(weightIn, qualityIn, conditionsIn, net.minecraftforge.common.ForgeHooks.readLootEntryName(object, "empty"));
     }
 }
