--- before/net/minecraft/world/storage/loot/LootContext.java
+++ after/net/minecraft/world/storage/loot/LootContext.java
@@ -96,6 +96,16 @@
         }
     }
 
+    public WorldServer getWorld()
+    {
+        return world;
+    }
+
+    public int getLootingModifier()
+    {
+        return net.minecraftforge.common.ForgeHooks.getLootingLevel(getLootedEntity(), getKiller(), damageSource);
+    }
+
     public static class Builder
         {
             private final WorldServer world;
