--- before/net/minecraft/world/storage/loot/LootTableList.java
+++ after/net/minecraft/world/storage/loot/LootTableList.java
@@ -8,8 +8,8 @@
 
 public class LootTableList
 {
-    private static final Set<ResourceLocation> field_186391_ap = Sets.newHashSet();
-    private static final Set<ResourceLocation> field_186392_aq = Collections.unmodifiableSet(field_186391_ap);
+    private static final Set<ResourceLocation> field_186391_ap = Sets.<ResourceLocation>newHashSet();
+    private static final Set<ResourceLocation> field_186392_aq = Collections.<ResourceLocation>unmodifiableSet(field_186391_ap);
     public static final ResourceLocation field_186419_a = func_186373_a("empty");
     public static final ResourceLocation field_186420_b = func_186373_a("chests/spawn_bonus_chest");
     public static final ResourceLocation field_186421_c = func_186373_a("chests/end_city_treasure");
