--- before/net/minecraft/world/storage/loot/LootContext.java
+++ after/net/minecraft/world/storage/loot/LootContext.java
@@ -23,16 +23,9 @@
     private final EntityPlayer player;
     @Nullable
     private final DamageSource damageSource;
-    private final Set<LootTable> lootTables = Sets.newLinkedHashSet();
+    private final Set<LootTable> lootTables = Sets.<LootTable>newLinkedHashSet();
 
-    public LootContext(
-        float luckIn,
-        WorldServer worldIn,
-        LootTableManager lootTableManagerIn,
-        @Nullable Entity lootedEntityIn,
-        @Nullable EntityPlayer playerIn,
-        @Nullable DamageSource damageSourceIn
-    )
+    public LootContext(float luckIn, WorldServer worldIn, LootTableManager lootTableManagerIn, @Nullable Entity lootedEntityIn, @Nullable EntityPlayer playerIn, @Nullable DamageSource damageSourceIn)
     {
         this.luck = luckIn;
         this.world = worldIn;
@@ -96,6 +89,16 @@
         }
     }
 
+    public WorldServer getWorld()
+    {
+        return world;
+    }
+
+    public int getLootingModifier()
+    {
+        return net.minecraftforge.common.ForgeHooks.getLootingLevel(getLootedEntity(), getKiller(), damageSource);
+    }
+
     public static class Builder
         {
             private final WorldServer world;
@@ -135,9 +138,7 @@
 
             public LootContext build()
             {
-                return new LootContext(
-                           this.luck, this.world, this.world.getLootTableManager(), this.lootedEntity, this.player, this.damageSource
-                       );
+                return new LootContext(this.luck, this.world, this.world.getLootTableManager(), this.lootedEntity, this.player, this.damageSource);
             }
         }
 
