--- before/net/minecraft/world/storage/loot/LootTableManager.java
+++ after/net/minecraft/world/storage/loot/LootTableManager.java
@@ -119,7 +119,7 @@
 
                         try
                         {
-                            return JsonUtils.gsonDeserialize(LootTableManager.GSON_INSTANCE, s, LootTable.class);
+                            return net.minecraftforge.common.ForgeHooks.loadLootTable(LootTableManager.GSON_INSTANCE, resource, s, true, LootTableManager.this);
                         }
                         catch (IllegalArgumentException | JsonParseException jsonparseexception)
                         {
@@ -161,7 +161,7 @@
 
                 try
                 {
-                    return JsonUtils.gsonDeserialize(LootTableManager.GSON_INSTANCE, s, LootTable.class);
+                    return net.minecraftforge.common.ForgeHooks.loadLootTable(LootTableManager.GSON_INSTANCE, resource, s, false, LootTableManager.this);
                 }
                 catch (JsonParseException jsonparseexception)
                 {
