--- before/net/minecraft/world/storage/loot/RandomValueRange.java
+++ after/net/minecraft/world/storage/loot/RandomValueRange.java
@@ -81,8 +81,8 @@
                 else
                 {
                     JsonObject jsonobject = new JsonObject();
-                    jsonobject.addProperty("min", p_serialize_1_.field_186514_a);
-                    jsonobject.addProperty("max", p_serialize_1_.field_186515_b);
+                    jsonobject.addProperty("min", Float.valueOf(p_serialize_1_.field_186514_a));
+                    jsonobject.addProperty("max", Float.valueOf(p_serialize_1_.field_186515_b));
                     return jsonobject;
                 }
             }
