--- before/net/minecraft/inventory/ContainerPlayer.java
+++ after/net/minecraft/inventory/ContainerPlayer.java
@@ -49,7 +49,7 @@
                 @Override
                 public boolean isItemValid(ItemStack stack)
                 {
-                    return entityequipmentslot == EntityLiving.getSlotForItemStack(stack);
+                    return stack.getItem().isValidArmor(stack, entityequipmentslot, player);
                 }
                 @Override
                 public boolean canTakeStack(EntityPlayer playerIn)
