--- before/net/minecraft/inventory/Container.java
+++ after/net/minecraft/inventory/Container.java
@@ -81,9 +81,11 @@
 
             if (!ItemStack.areItemStacksEqual(itemstack1, itemstack))
             {
+                boolean clientStackChanged = !ItemStack.areItemStacksEqualUsingNBTShareTag(itemstack1, itemstack);
                 itemstack1 = itemstack.isEmpty() ? ItemStack.EMPTY : itemstack.copy();
                 this.inventoryItemStacks.set(i, itemstack1);
 
+                if (clientStackChanged)
                 for (int j = 0; j < this.listeners.size(); j++)
                 {
                     this.listeners.get(j).sendSlotContents(this, i, itemstack1);
@@ -600,18 +602,19 @@
                         && ItemStack.areItemStackTagsEqual(stack, itemstack))
                 {
                     int j = itemstack.getCount() + stack.getCount();
+                    int maxSize = Math.min(slot.getSlotStackLimit(), stack.getMaxStackSize());
 
-                    if (j <= stack.getMaxStackSize())
+                    if (j <= maxSize)
                     {
                         stack.setCount(0);
                         itemstack.setCount(j);
                         slot.onSlotChanged();
                         flag = true;
                     }
-                    else if (itemstack.getCount() < stack.getMaxStackSize())
+                    else if (itemstack.getCount() < maxSize)
                     {
-                        stack.shrink(stack.getMaxStackSize() - itemstack.getCount());
-                        itemstack.setCount(stack.getMaxStackSize());
+                        stack.shrink(maxSize - itemstack.getCount());
+                        itemstack.setCount(maxSize);
                         slot.onSlotChanged();
                         flag = true;
                     }
@@ -727,7 +730,7 @@
                 stack.setCount(1);
                 break;
             case 2:
-                stack.setCount(stack.getItem().getItemStackLimit());
+                stack.setCount(stack.getMaxStackSize());
         }
 
         stack.grow(slotStackSize);
