--- before/net/minecraft/inventory/ContainerBeacon.java
+++ after/net/minecraft/inventory/ContainerBeacon.java
@@ -95,12 +95,9 @@
 
                 slot.onSlotChange(itemstack1, itemstack);
             }
-            else if (!this.beaconSlot.getHasStack() && this.beaconSlot.isItemValid(itemstack1) && itemstack1.getCount() == 1)
+            else if (this.mergeItemStack(itemstack1, 0, 1, false)) //Forge Fix Shift Clicking in beacons with stacks larger then 1.
             {
-                if (!this.mergeItemStack(itemstack1, 0, 1, false))
-                {
-                    return ItemStack.EMPTY;
-                }
+                return ItemStack.EMPTY;
             }
             else if (index >= 1 && index < 28)
             {
@@ -151,8 +148,7 @@
         @Override
         public boolean isItemValid(ItemStack stack)
         {
-            Item item = stack.getItem();
-            return item == Items.EMERALD || item == Items.DIAMOND || item == Items.GOLD_INGOT || item == Items.IRON_INGOT;
+            return stack.getItem().isBeaconPayment(stack);
         }
 
         @Override
