--- before/net/minecraft/inventory/ContainerHorseInventory.java
+++ after/net/minecraft/inventory/ContainerHorseInventory.java
@@ -38,14 +38,14 @@
             {
                 return p_i45791_3_.func_190682_f(p_75214_1_);
             }
+            public int func_75219_a()
+            {
+                return 1;
+            }
             @SideOnly(Side.CLIENT)
             public boolean func_111238_b()
             {
                 return p_i45791_3_.func_190677_dK();
-            }
-            public int func_75219_a()
-            {
-                return 1;
             }
         });
 
