--- before/net/minecraft/inventory/ContainerRepair.java
+++ after/net/minecraft/inventory/ContainerRepair.java
@@ -75,6 +75,8 @@
                     thePlayer.addExperienceLevel(-ContainerRepair.this.maximumCost);
                 }
 
+                float breakChance = net.minecraftforge.common.ForgeHooks.onAnvilRepair(thePlayer, stack, ContainerRepair.this.inputSlots.getStackInSlot(0), ContainerRepair.this.inputSlots.getStackInSlot(1));
+
                 ContainerRepair.this.inputSlots.setInventorySlotContents(0, ItemStack.EMPTY);
 
                 if (ContainerRepair.this.materialCost > 0)
@@ -102,7 +104,7 @@
                 if (!thePlayer.capabilities.isCreativeMode
                         && !worldIn.isRemote
                         && iblockstate.getBlock() == Blocks.ANVIL
-                        && thePlayer.getRNG().nextFloat() < 0.12F)
+                        && thePlayer.getRNG().nextFloat() < breakChance)
                 {
                     int l = iblockstate.getValue(BlockAnvil.DAMAGE);
 
@@ -172,10 +174,12 @@
             Map<Enchantment, Integer> map = EnchantmentHelper.getEnchantments(itemstack1);
             j += itemstack.getRepairCost() + (itemstack2.isEmpty() ? 0 : itemstack2.getRepairCost());
             this.materialCost = 0;
+            boolean flag = false;
 
             if (!itemstack2.isEmpty())
             {
-                boolean flag = itemstack2.getItem() == Items.ENCHANTED_BOOK && !ItemEnchantedBook.getEnchantments(itemstack2).isEmpty();
+                if (!net.minecraftforge.common.ForgeHooks.onAnvilChange(this, itemstack, itemstack2, outputSlot, repairedItemName, j, player)) return;
+                flag = itemstack2.getItem() == Items.ENCHANTED_BOOK && !ItemEnchantedBook.getEnchantments(itemstack2).isEmpty();
 
                 if (itemstack1.isItemStackDamageable() && itemstack1.getItem().getIsRepairable(itemstack, itemstack2))
                 {
@@ -222,7 +226,7 @@
                             l1 = 0;
                         }
 
-                        if (l1 < itemstack1.getMetadata())
+                        if (l1 < itemstack1.getItemDamage()) // vanilla uses metadata here instead of damage.
                         {
                             itemstack1.setItemDamage(l1);
                             i += 2;
@@ -326,6 +330,7 @@
                 i += k;
                 itemstack1.setStackDisplayName(this.repairedItemName);
             }
+            if (flag && !itemstack1.getItem().isBookEnchantable(itemstack1, itemstack2)) itemstack1 = ItemStack.EMPTY;
 
             this.maximumCost = j + i;
 
