--- before/net/minecraft/inventory/ContainerBrewingStand.java
+++ after/net/minecraft/inventory/ContainerBrewingStand.java
@@ -205,7 +205,7 @@
             @Override
             public boolean isItemValid(ItemStack stack)
             {
-                return PotionHelper.isReagent(stack);
+                return net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidIngredient(stack);
             }
 
             @Override
@@ -241,6 +241,7 @@
 
                 if (thePlayer instanceof EntityPlayerMP)
                 {
+                    net.minecraftforge.event.ForgeEventFactory.onPlayerBrewedPotion(thePlayer, stack);
                     CriteriaTriggers.BREWED_POTION.trigger((EntityPlayerMP)thePlayer, potiontype);
                 }
 
@@ -250,8 +251,7 @@
 
             public static boolean canHoldPotion(ItemStack stack)
             {
-                Item item = stack.getItem();
-                return item == Items.POTIONITEM || item == Items.SPLASH_POTION || item == Items.LINGERING_POTION || item == Items.GLASS_BOTTLE;
+                return net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidInput(stack);
             }
         }
 }
