--- before/net/minecraft/inventory/InventoryBasic.java
+++ after/net/minecraft/inventory/InventoryBasic.java
@@ -24,7 +24,7 @@
         this.field_70483_a = p_i1561_1_;
         this.field_94051_e = p_i1561_2_;
         this.field_70481_b = p_i1561_3_;
-        this.field_70482_c = NonNullList.func_191197_a(p_i1561_3_, ItemStack.field_190927_a);
+        this.field_70482_c = NonNullList.<ItemStack>func_191197_a(p_i1561_3_, ItemStack.field_190927_a);
     }
 
     @SideOnly(Side.CLIENT)
@@ -37,7 +37,7 @@
     {
         if (this.field_70480_d == null)
         {
-            this.field_70480_d = Lists.newArrayList();
+            this.field_70480_d = Lists.<IInventoryChangedListener>newArrayList();
         }
 
         this.field_70480_d.add(p_110134_1_);
@@ -50,7 +50,7 @@
 
     public ItemStack func_70301_a(int p_70301_1_)
     {
-        return p_70301_1_ >= 0 && p_70301_1_ < this.field_70482_c.size() ? this.field_70482_c.get(p_70301_1_) : ItemStack.field_190927_a;
+        return p_70301_1_ >= 0 && p_70301_1_ < this.field_70482_c.size() ? (ItemStack)this.field_70482_c.get(p_70301_1_) : ItemStack.field_190927_a;
     }
 
     public ItemStack func_70298_a(int p_70298_1_, int p_70298_2_)
@@ -170,7 +170,7 @@
 
     public ITextComponent func_145748_c_()
     {
-        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_()));
+        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_(), new Object[0]));
     }
 
     public int func_70297_j_()
@@ -184,7 +184,7 @@
         {
             for (int i = 0; i < this.field_70480_d.size(); ++i)
             {
-                this.field_70480_d.get(i).func_76316_a(this);
+                ((IInventoryChangedListener)this.field_70480_d.get(i)).func_76316_a(this);
             }
         }
     }
