--- before/net/minecraft/inventory/InventoryLargeChest.java
+++ after/net/minecraft/inventory/InventoryLargeChest.java
@@ -76,7 +76,7 @@
 
     public ITextComponent func_145748_c_()
     {
-        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_()));
+        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_(), new Object[0]));
     }
 
     public ItemStack func_70301_a(int p_70301_1_)
