--- before/net/minecraft/inventory/InventoryMerchant.java
+++ after/net/minecraft/inventory/InventoryMerchant.java
@@ -13,7 +13,7 @@
 public class InventoryMerchant implements IInventory
 {
     private final IMerchant field_70476_a;
-    private final NonNullList<ItemStack> field_70474_b = NonNullList.func_191197_a(3, ItemStack.field_190927_a);
+    private final NonNullList<ItemStack> field_70474_b = NonNullList.<ItemStack>func_191197_a(3, ItemStack.field_190927_a);
     private final EntityPlayer field_70475_c;
     private MerchantRecipe field_70472_d;
     private int field_70473_e;
@@ -105,7 +105,7 @@
 
     public ITextComponent func_145748_c_()
     {
-        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_()));
+        return (ITextComponent)(this.func_145818_k_() ? new TextComponentString(this.func_70005_c_()) : new TextComponentTranslation(this.func_70005_c_(), new Object[0]));
     }
 
     public int func_70297_j_()
