--- before/net/minecraft/inventory/SlotFurnaceOutput.java
+++ after/net/minecraft/inventory/SlotFurnaceOutput.java
@@ -17,24 +17,21 @@
         this.player = player;
     }
 
-    @Override
     public boolean isItemValid(ItemStack stack)
     {
         return false;
     }
 
-    @Override
     public ItemStack decrStackSize(int amount)
     {
         if (this.getHasStack())
         {
-            this.removeCount = this.removeCount + Math.min(amount, this.getStack().getCount());
+            this.removeCount += Math.min(amount, this.getStack().getCount());
         }
 
         return super.decrStackSize(amount);
     }
 
-    @Override
     public ItemStack onTake(EntityPlayer thePlayer, ItemStack stack)
     {
         this.onCrafting(stack);
@@ -42,14 +39,12 @@
         return stack;
     }
 
-    @Override
     protected void onCrafting(ItemStack stack, int amount)
     {
         this.removeCount += amount;
         this.onCrafting(stack);
     }
 
-    @Override
     protected void onCrafting(ItemStack stack)
     {
         stack.onCrafting(this.player.world, this.player, this.removeCount);
@@ -65,11 +60,11 @@
             }
             else if (f < 1.0F)
             {
-                int j = MathHelper.floor(i * f);
+                int j = MathHelper.floor((float)i * f);
 
-                if (j < MathHelper.ceil(i * f) && Math.random() < i * f - j)
+                if (j < MathHelper.ceil((float)i * f) && Math.random() < (double)((float)i * f - (float)j))
                 {
-                    j++;
+                    ++j;
                 }
 
                 i = j;
@@ -79,20 +74,11 @@
             {
                 int k = EntityXPOrb.getXPSplit(i);
                 i -= k;
-                this.player
-                .world
-                .spawnEntity(
-                    new EntityXPOrb(
-                        this.player.world,
-                        this.player.posX,
-                        this.player.posY + 0.5,
-                        this.player.posZ + 0.5,
-                        k
-                    )
-                );
+                this.player.world.spawnEntity(new EntityXPOrb(this.player.world, this.player.posX, this.player.posY + 0.5D, this.player.posZ + 0.5D, k));
             }
         }
 
         this.removeCount = 0;
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerSmeltedEvent(player, stack);
     }
 }
