--- before/net/minecraft/entity/ai/EntityAIFollowGolem.java
+++ after/net/minecraft/entity/ai/EntityAIFollowGolem.java
@@ -29,7 +29,7 @@
         }
         else
         {
-            List<EntityIronGolem> list = this.field_75294_a.field_70170_p.func_72872_a(EntityIronGolem.class, this.field_75294_a.func_174813_aQ().func_72314_b(6.0D, 2.0D, 6.0D));
+            List<EntityIronGolem> list = this.field_75294_a.field_70170_p.<EntityIronGolem>func_72872_a(EntityIronGolem.class, this.field_75294_a.func_174813_aQ().func_72314_b(6.0D, 2.0D, 6.0D));
 
             if (list.isEmpty())
             {
