--- before/net/minecraft/entity/ai/EntityAIAttackMelee.java
+++ after/net/minecraft/entity/ai/EntityAIAttackMelee.java
@@ -21,6 +21,8 @@
     private double targetY;
     private double targetZ;
     protected final int attackInterval = 20;
+    private int failedPathFindingPenalty = 0;
+    private boolean canPenalize = false;
 
     public EntityAIAttackMelee(EntityCreature creature, double speedIn, boolean useLongMemory)
     {
@@ -46,6 +48,19 @@
         }
         else
         {
+            if (canPenalize)
+            {
+                if (--this.delayCounter <= 0)
+                {
+                    this.path = this.attacker.getNavigator().getPathToEntityLiving(entitylivingbase);
+                    this.delayCounter = 4 + this.attacker.getRNG().nextInt(7);
+                    return this.path != null;
+                }
+                else
+                {
+                    return true;
+                }
+            }
             this.path = this.attacker.getNavigator().getPathToEntityLiving(entitylivingbase);
             return this.path != null
                    ? true
@@ -122,6 +137,23 @@
             this.targetY = entitylivingbase.getEntityBoundingBox().minY;
             this.targetZ = entitylivingbase.posZ;
             this.delayCounter = 4 + this.attacker.getRNG().nextInt(7);
+
+            if (this.canPenalize)
+            {
+                this.delayCounter += failedPathFindingPenalty;
+                if (this.attacker.getNavigator().getPath() != null)
+                {
+                    net.minecraft.pathfinding.PathPoint finalPathPoint = this.attacker.getNavigator().getPath().getFinalPathPoint();
+                    if (finalPathPoint != null && entitylivingbase.getDistanceSq(finalPathPoint.x, finalPathPoint.y, finalPathPoint.z) < 1)
+                        failedPathFindingPenalty = 0;
+                    else
+                        failedPathFindingPenalty += 10;
+                }
+                else
+                {
+                    failedPathFindingPenalty += 10;
+                }
+            }
 
             if (d0 > 1024.0)
             {
