--- before/net/minecraft/entity/ai/EntityAIPlay.java
+++ after/net/minecraft/entity/ai/EntityAIPlay.java
@@ -31,7 +31,7 @@
         }
         else
         {
-            List<EntityVillager> list = this.field_75262_a.field_70170_p.func_72872_a(EntityVillager.class, this.field_75262_a.func_174813_aQ().func_72314_b(6.0D, 3.0D, 6.0D));
+            List<EntityVillager> list = this.field_75262_a.field_70170_p.<EntityVillager>func_72872_a(EntityVillager.class, this.field_75262_a.func_174813_aQ().func_72314_b(6.0D, 3.0D, 6.0D));
             double d0 = Double.MAX_VALUE;
 
             for (EntityVillager entityvillager : list)
@@ -40,7 +40,7 @@
                 {
                     double d1 = entityvillager.func_70068_e(this.field_75262_a);
 
-                    if (!(d1 > d0))
+                    if (d1 <= d0)
                     {
                         d0 = d1;
                         this.field_75260_b = entityvillager;
