--- before/net/minecraft/entity/ai/EntityAIEatGrass.java
+++ after/net/minecraft/entity/ai/EntityAIEatGrass.java
@@ -13,8 +13,7 @@
 
 public class EntityAIEatGrass extends EntityAIBase
 {
-    private static final Predicate<IBlockState> IS_TALL_GRASS = BlockStateMatcher.forBlock(Blocks.TALLGRASS)
-                .where(BlockTallGrass.TYPE, Predicates.equalTo(BlockTallGrass.EnumType.GRASS));
+    private static final Predicate<IBlockState> IS_TALL_GRASS = BlockStateMatcher.forBlock(Blocks.TALLGRASS).where(BlockTallGrass.TYPE, Predicates.equalTo(BlockTallGrass.EnumType.GRASS));
     private final EntityLiving grassEaterEntity;
     private final World entityWorld;
     int eatingGrassTimer;
@@ -26,7 +25,6 @@
         this.setMutexBits(7);
     }
 
-    @Override
     public boolean shouldExecute()
     {
         if (this.grassEaterEntity.getRNG().nextInt(this.grassEaterEntity.isChild() ? 50 : 1000) != 0)
@@ -36,13 +34,18 @@
         else
         {
             BlockPos blockpos = new BlockPos(this.grassEaterEntity.posX, this.grassEaterEntity.posY, this.grassEaterEntity.posZ);
-            return IS_TALL_GRASS.apply(this.entityWorld.getBlockState(blockpos))
-                   ? true
-                   : this.entityWorld.getBlockState(blockpos.down()).getBlock() == Blocks.GRASS;
+
+            if (IS_TALL_GRASS.apply(this.entityWorld.getBlockState(blockpos)))
+            {
+                return true;
+            }
+            else
+            {
+                return this.entityWorld.getBlockState(blockpos.down()).getBlock() == Blocks.GRASS;
+            }
         }
     }
 
-    @Override
     public void startExecuting()
     {
         this.eatingGrassTimer = 40;
@@ -50,13 +53,11 @@
         this.grassEaterEntity.getNavigator().clearPath();
     }
 
-    @Override
     public void resetTask()
     {
         this.eatingGrassTimer = 0;
     }
 
-    @Override
     public boolean shouldContinueExecuting()
     {
         return this.eatingGrassTimer > 0;
@@ -67,7 +68,6 @@
         return this.eatingGrassTimer;
     }
 
-    @Override
     public void updateTask()
     {
         this.eatingGrassTimer = Math.max(0, this.eatingGrassTimer - 1);
@@ -78,7 +78,7 @@
 
             if (IS_TALL_GRASS.apply(this.entityWorld.getBlockState(blockpos)))
             {
-                if (this.entityWorld.getGameRules().getBoolean("mobGriefing"))
+                if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.entityWorld, this.grassEaterEntity))
                 {
                     this.entityWorld.destroyBlock(blockpos, false);
                 }
@@ -91,7 +91,7 @@
 
                 if (this.entityWorld.getBlockState(blockpos1).getBlock() == Blocks.GRASS)
                 {
-                    if (this.entityWorld.getGameRules().getBoolean("mobGriefing"))
+                    if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.entityWorld, this.grassEaterEntity))
                     {
                         this.entityWorld.playEvent(2001, blockpos1, Block.getIdFromBlock(Blocks.GRASS));
                         this.entityWorld.setBlockState(blockpos1, Blocks.DIRT.getDefaultState(), 2);
