--- before/net/minecraft/entity/ai/EntityAIOcelotAttack.java
+++ after/net/minecraft/entity/ai/EntityAIOcelotAttack.java
@@ -74,7 +74,7 @@
         this.field_75409_b.func_70661_as().func_75497_a(this.field_75410_c, d2);
         this.field_75408_d = Math.max(this.field_75408_d - 1, 0);
 
-        if (!(d1 > d0))
+        if (d1 <= d0)
         {
             if (this.field_75408_d <= 0)
             {
