--- before/net/minecraft/entity/ai/EntityAIFindEntityNearestPlayer.java
+++ after/net/minecraft/entity/ai/EntityAIFindEntityNearestPlayer.java
@@ -51,7 +51,7 @@
 
                     if (p_apply_1_.func_70093_af())
                     {
-                        d0 *= (double)0.8F;
+                        d0 *= 0.800000011920929D;
                     }
 
                     if (p_apply_1_.func_82150_aj())
@@ -76,7 +76,7 @@
     public boolean func_75250_a()
     {
         double d0 = this.func_179431_f();
-        List<EntityPlayer> list = this.field_179434_b.field_70170_p.func_175647_a(EntityPlayer.class, this.field_179434_b.func_174813_aQ().func_72314_b(d0, 4.0D, d0), this.field_179435_c);
+        List<EntityPlayer> list = this.field_179434_b.field_70170_p.<EntityPlayer>func_175647_a(EntityPlayer.class, this.field_179434_b.func_174813_aQ().func_72314_b(d0, 4.0D, d0), this.field_179435_c);
         Collections.sort(list, this.field_179432_d);
 
         if (list.isEmpty())
