--- before/net/minecraft/entity/ai/EntityAIFollowParent.java
+++ after/net/minecraft/entity/ai/EntityAIFollowParent.java
@@ -24,7 +24,7 @@
         }
         else
         {
-            List<EntityAnimal> list = this.field_75348_a.field_70170_p.func_72872_a(this.field_75348_a.getClass(), this.field_75348_a.func_174813_aQ().func_72314_b(8.0D, 4.0D, 8.0D));
+            List<EntityAnimal> list = this.field_75348_a.field_70170_p.<EntityAnimal>func_72872_a(this.field_75348_a.getClass(), this.field_75348_a.func_174813_aQ().func_72314_b(8.0D, 4.0D, 8.0D));
             EntityAnimal entityanimal = null;
             double d0 = Double.MAX_VALUE;
 
@@ -34,7 +34,7 @@
                 {
                     double d1 = this.field_75348_a.func_70068_e(entityanimal1);
 
-                    if (!(d1 > d0))
+                    if (d1 <= d0)
                     {
                         d0 = d1;
                         entityanimal = entityanimal1;
@@ -71,7 +71,7 @@
         else
         {
             double d0 = this.field_75348_a.func_70068_e(this.field_75346_b);
-            return !(d0 < 9.0D) && !(d0 > 256.0D);
+            return d0 >= 9.0D && d0 <= 256.0D;
         }
     }
 
