--- before/net/minecraft/entity/ai/EntityAIMate.java
+++ after/net/minecraft/entity/ai/EntityAIMate.java
@@ -96,6 +96,18 @@
     {
         EntityAgeable entityageable = this.animal.createChild(this.targetMate);
 
+        final net.minecraftforge.event.entity.living.BabyEntitySpawnEvent event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(animal, targetMate, entityageable);
+        final boolean cancelled = net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+        entityageable = event.getChild();
+        if (cancelled) {
+            //Reset the "inLove" state for the animals
+            this.animal.setGrowingAge(6000);
+            this.targetMate.setGrowingAge(6000);
+            this.animal.resetInLove();
+            this.targetMate.resetInLove();
+            return;
+        }
+        
         if (entityageable != null)
         {
             EntityPlayerMP entityplayermp = this.animal.getLoveCause();
