--- before/net/minecraft/entity/ai/EntityAIEatGrass.java
+++ after/net/minecraft/entity/ai/EntityAIEatGrass.java
@@ -78,7 +78,7 @@
 
             if (IS_TALL_GRASS.apply(this.entityWorld.getBlockState(blockpos)))
             {
-                if (this.entityWorld.getGameRules().getBoolean("mobGriefing"))
+                if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.entityWorld, this.grassEaterEntity))
                 {
                     this.entityWorld.destroyBlock(blockpos, false);
                 }
@@ -91,7 +91,7 @@
 
                 if (this.entityWorld.getBlockState(blockpos1).getBlock() == Blocks.GRASS)
                 {
-                    if (this.entityWorld.getGameRules().getBoolean("mobGriefing"))
+                    if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.entityWorld, this.grassEaterEntity))
                     {
                         this.entityWorld.playEvent(2001, blockpos1, Block.getIdFromBlock(Blocks.GRASS));
                         this.entityWorld.setBlockState(blockpos1, Blocks.DIRT.getDefaultState(), 2);
