--- before/net/minecraft/entity/ai/EntityAITasks.java
+++ after/net/minecraft/entity/ai/EntityAITasks.java
@@ -11,8 +11,8 @@
 public class EntityAITasks
 {
     private static final Logger field_151506_a = LogManager.getLogger();
-    public final Set<EntityAITasks.EntityAITaskEntry> field_75782_a = Sets.newLinkedHashSet();
-    private final Set<EntityAITasks.EntityAITaskEntry> field_75780_b = Sets.newLinkedHashSet();
+    public final Set<EntityAITasks.EntityAITaskEntry> field_75782_a = Sets.<EntityAITasks.EntityAITaskEntry>newLinkedHashSet();
+    private final Set<EntityAITasks.EntityAITaskEntry> field_75780_b = Sets.<EntityAITasks.EntityAITaskEntry>newLinkedHashSet();
     private final Profiler field_75781_c;
     private int field_75778_d;
     private int field_75779_e = 3;
