--- before/net/minecraft/entity/ai/EntityMoveHelper.java
+++ after/net/minecraft/entity/ai/EntityMoveHelper.java
@@ -77,10 +77,10 @@
             }
 
             f4 = f1 / f4;
-            f2 *= f4;
-            f3 *= f4;
-            float f5 = MathHelper.func_76126_a(this.field_75648_a.field_70177_z * ((float)Math.PI / 180F));
-            float f6 = MathHelper.func_76134_b(this.field_75648_a.field_70177_z * ((float)Math.PI / 180F));
+            f2 = f2 * f4;
+            f3 = f3 * f4;
+            float f5 = MathHelper.func_76126_a(this.field_75648_a.field_70177_z * 0.017453292F);
+            float f6 = MathHelper.func_76134_b(this.field_75648_a.field_70177_z * 0.017453292F);
             float f7 = f2 * f6 - f3 * f5;
             float f8 = f3 * f6 + f2 * f5;
             PathNavigate pathnavigate = this.field_75648_a.func_70661_as();
@@ -110,13 +110,13 @@
             double d2 = this.field_75647_c - this.field_75648_a.field_70163_u;
             double d3 = d0 * d0 + d2 * d2 + d1 * d1;
 
-            if (d3 < (double)2.5000003E-7F)
+            if (d3 < 2.500000277905201E-7D)
             {
                 this.field_75648_a.func_191989_p(0.0F);
                 return;
             }
 
-            float f9 = (float)(MathHelper.func_181159_b(d1, d0) * (double)(180F / (float)Math.PI)) - 90.0F;
+            float f9 = (float)(MathHelper.func_181159_b(d1, d0) * (180D / Math.PI)) - 90.0F;
             this.field_75648_a.field_70177_z = this.func_75639_a(this.field_75648_a.field_70177_z, f9, 90.0F);
             this.field_75648_a.func_70659_e((float)(this.field_75645_e * this.field_75648_a.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111126_e()));
 
