--- before/net/minecraft/entity/ai/EntityAILlamaFollowCaravan.java
+++ after/net/minecraft/entity/ai/EntityAILlamaFollowCaravan.java
@@ -21,7 +21,7 @@
     {
         if (!this.field_190859_a.func_110167_bD() && !this.field_190859_a.func_190718_dR())
         {
-            List<EntityLlama> list = this.field_190859_a.field_70170_p.func_72872_a(this.field_190859_a.getClass(), this.field_190859_a.func_174813_aQ().func_72314_b(9.0D, 4.0D, 9.0D));
+            List<EntityLlama> list = this.field_190859_a.field_70170_p.<EntityLlama>func_72872_a(this.field_190859_a.getClass(), this.field_190859_a.func_174813_aQ().func_72314_b(9.0D, 4.0D, 9.0D));
             EntityLlama entityllama = null;
             double d0 = Double.MAX_VALUE;
 
@@ -31,7 +31,7 @@
                 {
                     double d1 = this.field_190859_a.func_70068_e(entityllama1);
 
-                    if (!(d1 > d0))
+                    if (d1 <= d0)
                     {
                         d0 = d1;
                         entityllama = entityllama1;
@@ -47,7 +47,7 @@
                     {
                         double d2 = this.field_190859_a.func_70068_e(entityllama2);
 
-                        if (!(d2 > d0))
+                        if (d2 <= d0)
                         {
                             d0 = d2;
                             entityllama = entityllama2;
