--- before/net/minecraft/entity/ai/EntityAIVillagerInteract.java
+++ after/net/minecraft/entity/ai/EntityAIVillagerInteract.java
@@ -76,14 +76,14 @@
 
                     if (!itemstack1.func_190926_b())
                     {
-                        double d0 = this.field_179477_f.field_70163_u - (double)0.3F + (double)this.field_179477_f.func_70047_e();
+                        double d0 = this.field_179477_f.field_70163_u - 0.30000001192092896D + (double)this.field_179477_f.func_70047_e();
                         EntityItem entityitem = new EntityItem(this.field_179477_f.field_70170_p, this.field_179477_f.field_70165_t, d0, this.field_179477_f.field_70161_v, itemstack1);
                         float f = 0.3F;
                         float f1 = this.field_179477_f.field_70759_as;
                         float f2 = this.field_179477_f.field_70125_A;
-                        entityitem.field_70159_w = (double)(-MathHelper.func_76126_a(f1 * ((float)Math.PI / 180F)) * MathHelper.func_76134_b(f2 * ((float)Math.PI / 180F)) * 0.3F);
-                        entityitem.field_70179_y = (double)(MathHelper.func_76134_b(f1 * ((float)Math.PI / 180F)) * MathHelper.func_76134_b(f2 * ((float)Math.PI / 180F)) * 0.3F);
-                        entityitem.field_70181_x = (double)(-MathHelper.func_76126_a(f2 * ((float)Math.PI / 180F)) * 0.3F + 0.1F);
+                        entityitem.field_70159_w = (double)(-MathHelper.func_76126_a(f1 * 0.017453292F) * MathHelper.func_76134_b(f2 * 0.017453292F) * 0.3F);
+                        entityitem.field_70179_y = (double)(MathHelper.func_76134_b(f1 * 0.017453292F) * MathHelper.func_76134_b(f2 * 0.017453292F) * 0.3F);
+                        entityitem.field_70181_x = (double)(-MathHelper.func_76126_a(f2 * 0.017453292F) * 0.3F + 0.1F);
                         entityitem.func_174869_p();
                         this.field_179477_f.field_70170_p.func_72838_d(entityitem);
                         break;
