--- before/net/minecraft/entity/ai/EntityAIBreakDoor.java
+++ after/net/minecraft/entity/ai/EntityAIBreakDoor.java
@@ -15,45 +15,55 @@
         super(p_i1618_1_);
     }
 
-    @Override
     public boolean func_75250_a()
     {
         if (!super.func_75250_a())
         {
             return false;
         }
-        else if (!this.field_75356_a.field_70170_p.func_82736_K().func_82766_b("mobGriefing"))
+        else if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_75356_a.field_70170_p, this.field_75356_a) || !this.field_75356_a.field_70170_p.func_180495_p(this.field_179507_b).func_177230_c().canEntityDestroy(this.field_75356_a.field_70170_p.func_180495_p(this.field_179507_b), this.field_75356_a.field_70170_p, this.field_179507_b, this.field_75356_a) || !net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this.field_75356_a, this.field_179507_b, this.field_75356_a.field_70170_p.func_180495_p(this.field_179507_b)))
         {
             return false;
         }
         else
         {
+            BlockDoor blockdoor = this.field_151504_e;
             return !BlockDoor.func_176514_f(this.field_75356_a.field_70170_p, this.field_179507_b);
         }
     }
 
-    @Override
     public void func_75249_e()
     {
         super.func_75249_e();
         this.field_75359_i = 0;
     }
 
-    @Override
     public boolean func_75253_b()
     {
         double d0 = this.field_75356_a.func_174818_b(this.field_179507_b);
-        return this.field_75359_i <= 240 && !BlockDoor.func_176514_f(this.field_75356_a.field_70170_p, this.field_179507_b) && d0 < 4.0;
+        boolean flag;
+
+        if (this.field_75359_i <= 240)
+        {
+            BlockDoor blockdoor = this.field_151504_e;
+
+            if (!BlockDoor.func_176514_f(this.field_75356_a.field_70170_p, this.field_179507_b) && d0 < 4.0D)
+            {
+                flag = true;
+                return flag;
+            }
+        }
+
+        flag = false;
+        return flag;
     }
 
-    @Override
     public void func_75251_c()
     {
         super.func_75251_c();
         this.field_75356_a.field_70170_p.func_175715_c(this.field_75356_a.func_145782_y(), this.field_179507_b, -1);
     }
 
-    @Override
     public void func_75246_d()
     {
         super.func_75246_d();
