--- before/net/minecraft/entity/ai/EntityAIWanderAvoidWaterFlying.java
+++ after/net/minecraft/entity/ai/EntityAIWanderAvoidWaterFlying.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.ai;
 
+import java.util.Iterator;
 import javax.annotation.Nullable;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockLeaves;
@@ -41,9 +42,19 @@
         BlockPos blockpos = new BlockPos(this.field_75457_a);
         BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
         BlockPos.MutableBlockPos blockpos$mutableblockpos1 = new BlockPos.MutableBlockPos();
+        Iterable<BlockPos.MutableBlockPos> iterable = BlockPos.MutableBlockPos.func_191531_b(MathHelper.func_76128_c(this.field_75457_a.field_70165_t - 3.0D), MathHelper.func_76128_c(this.field_75457_a.field_70163_u - 6.0D), MathHelper.func_76128_c(this.field_75457_a.field_70161_v - 3.0D), MathHelper.func_76128_c(this.field_75457_a.field_70165_t + 3.0D), MathHelper.func_76128_c(this.field_75457_a.field_70163_u + 6.0D), MathHelper.func_76128_c(this.field_75457_a.field_70161_v + 3.0D));
+        Iterator iterator = iterable.iterator();
+        BlockPos blockpos1;
 
-        for (BlockPos blockpos1 : BlockPos.MutableBlockPos.func_191531_b(MathHelper.func_76128_c(this.field_75457_a.field_70165_t - 3.0D), MathHelper.func_76128_c(this.field_75457_a.field_70163_u - 6.0D), MathHelper.func_76128_c(this.field_75457_a.field_70161_v - 3.0D), MathHelper.func_76128_c(this.field_75457_a.field_70165_t + 3.0D), MathHelper.func_76128_c(this.field_75457_a.field_70163_u + 6.0D), MathHelper.func_76128_c(this.field_75457_a.field_70161_v + 3.0D)))
+        while (true)
         {
+            if (!iterator.hasNext())
+            {
+                return null;
+            }
+
+            blockpos1 = (BlockPos)iterator.next();
+
             if (!blockpos.equals(blockpos1))
             {
                 Block block = this.field_75457_a.field_70170_p.func_180495_p(blockpos$mutableblockpos1.func_189533_g(blockpos1).func_189536_c(EnumFacing.DOWN)).func_177230_c();
@@ -51,11 +62,11 @@
 
                 if (flag && this.field_75457_a.field_70170_p.func_175623_d(blockpos1) && this.field_75457_a.field_70170_p.func_175623_d(blockpos$mutableblockpos.func_189533_g(blockpos1).func_189536_c(EnumFacing.UP)))
                 {
-                    return new Vec3d((double)blockpos1.func_177958_n(), (double)blockpos1.func_177956_o(), (double)blockpos1.func_177952_p());
+                    break;
                 }
             }
         }
 
-        return null;
+        return new Vec3d((double)blockpos1.func_177958_n(), (double)blockpos1.func_177956_o(), (double)blockpos1.func_177952_p());
     }
 }
