--- before/net/minecraft/entity/ai/EntitySenses.java
+++ after/net/minecraft/entity/ai/EntitySenses.java
@@ -8,8 +8,8 @@
 public class EntitySenses
 {
     EntityLiving field_75526_a;
-    List<Entity> field_75524_b = Lists.newArrayList();
-    List<Entity> field_75525_c = Lists.newArrayList();
+    List<Entity> field_75524_b = Lists.<Entity>newArrayList();
+    List<Entity> field_75525_c = Lists.<Entity>newArrayList();
 
     public EntitySenses(EntityLiving p_i1672_1_)
     {
