--- before/net/minecraft/entity/ai/EntityAIEatGrass.java
+++ after/net/minecraft/entity/ai/EntityAIEatGrass.java
@@ -13,8 +13,7 @@
 
 public class EntityAIEatGrass extends EntityAIBase
 {
-    private static final Predicate<IBlockState> field_179505_b = BlockStateMatcher.func_177638_a(Blocks.field_150329_H)
-                .func_177637_a(BlockTallGrass.field_176497_a, Predicates.equalTo(BlockTallGrass.EnumType.GRASS));
+    private static final Predicate<IBlockState> field_179505_b = BlockStateMatcher.func_177638_a(Blocks.field_150329_H).func_177637_a(BlockTallGrass.field_176497_a, Predicates.equalTo(BlockTallGrass.EnumType.GRASS));
     private final EntityLiving field_151500_b;
     private final World field_151501_c;
     int field_151502_a;
@@ -26,7 +25,6 @@
         this.func_75248_a(7);
     }
 
-    @Override
     public boolean func_75250_a()
     {
         if (this.field_151500_b.func_70681_au().nextInt(this.field_151500_b.func_70631_g_() ? 50 : 1000) != 0)
@@ -36,13 +34,18 @@
         else
         {
             BlockPos blockpos = new BlockPos(this.field_151500_b.field_70165_t, this.field_151500_b.field_70163_u, this.field_151500_b.field_70161_v);
-            return field_179505_b.apply(this.field_151501_c.func_180495_p(blockpos))
-                   ? true
-                   : this.field_151501_c.func_180495_p(blockpos.func_177977_b()).func_177230_c() == Blocks.field_150349_c;
+
+            if (field_179505_b.apply(this.field_151501_c.func_180495_p(blockpos)))
+            {
+                return true;
+            }
+            else
+            {
+                return this.field_151501_c.func_180495_p(blockpos.func_177977_b()).func_177230_c() == Blocks.field_150349_c;
+            }
         }
     }
 
-    @Override
     public void func_75249_e()
     {
         this.field_151502_a = 40;
@@ -50,13 +53,11 @@
         this.field_151500_b.func_70661_as().func_75499_g();
     }
 
-    @Override
     public void func_75251_c()
     {
         this.field_151502_a = 0;
     }
 
-    @Override
     public boolean func_75253_b()
     {
         return this.field_151502_a > 0;
@@ -67,7 +68,6 @@
         return this.field_151502_a;
     }
 
-    @Override
     public void func_75246_d()
     {
         this.field_151502_a = Math.max(0, this.field_151502_a - 1);
@@ -78,7 +78,7 @@
 
             if (field_179505_b.apply(this.field_151501_c.func_180495_p(blockpos)))
             {
-                if (this.field_151501_c.func_82736_K().func_82766_b("mobGriefing"))
+                if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_151501_c, this.field_151500_b))
                 {
                     this.field_151501_c.func_175655_b(blockpos, false);
                 }
@@ -91,7 +91,7 @@
 
                 if (this.field_151501_c.func_180495_p(blockpos1).func_177230_c() == Blocks.field_150349_c)
                 {
-                    if (this.field_151501_c.func_82736_K().func_82766_b("mobGriefing"))
+                    if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.field_151501_c, this.field_151500_b))
                     {
                         this.field_151501_c.func_175718_b(2001, blockpos1, Block.func_149682_b(Blocks.field_150349_c));
                         this.field_151501_c.func_180501_a(blockpos1, Blocks.field_150346_d.func_176223_P(), 2);
