--- before/net/minecraft/entity/ai/EntityAIFollow.java
+++ after/net/minecraft/entity/ai/EntityAIFollow.java
@@ -45,7 +45,7 @@
 
     public boolean func_75250_a()
     {
-        List<EntityLiving> list = this.field_192372_a.field_70170_p.func_175647_a(EntityLiving.class, this.field_192372_a.func_174813_aQ().func_186662_g((double)this.field_192380_i), this.field_192373_b);
+        List<EntityLiving> list = this.field_192372_a.field_70170_p.<EntityLiving>func_175647_a(EntityLiving.class, this.field_192372_a.func_174813_aQ().func_186662_g((double)this.field_192380_i), this.field_192373_b);
 
         if (!list.isEmpty())
         {
@@ -95,7 +95,7 @@
                 double d2 = this.field_192372_a.field_70161_v - this.field_192374_c.field_70161_v;
                 double d3 = d0 * d0 + d1 * d1 + d2 * d2;
 
-                if (!(d3 <= (double)(this.field_192378_g * this.field_192378_g)))
+                if (d3 > (double)(this.field_192378_g * this.field_192378_g))
                 {
                     this.field_192376_e.func_75497_a(this.field_192374_c, this.field_192375_d);
                 }
