--- before/net/minecraft/entity/ai/EntityAILookAtVillager.java
+++ after/net/minecraft/entity/ai/EntityAILookAtVillager.java
@@ -27,7 +27,7 @@
         }
         else
         {
-            this.field_75395_b = (EntityVillager) this.field_75397_a.field_70170_p.func_72857_a(EntityVillager.class, this.field_75397_a.func_174813_aQ().func_72314_b(6.0D, 2.0D, 6.0D), this.field_75397_a);
+            this.field_75395_b = (EntityVillager)this.field_75397_a.field_70170_p.func_72857_a(EntityVillager.class, this.field_75397_a.func_174813_aQ().func_72314_b(6.0D, 2.0D, 6.0D), this.field_75397_a);
             return this.field_75395_b != null;
         }
     }
