--- before/net/minecraft/entity/ai/EntityLookHelper.java
+++ after/net/minecraft/entity/ai/EntityLookHelper.java
@@ -60,8 +60,8 @@
             double d1 = this.field_75653_f - (this.field_75659_a.field_70163_u + (double)this.field_75659_a.func_70047_e());
             double d2 = this.field_75654_g - this.field_75659_a.field_70161_v;
             double d3 = (double)MathHelper.func_76133_a(d0 * d0 + d2 * d2);
-            float f = (float)(MathHelper.func_181159_b(d2, d0) * (double)(180F / (float)Math.PI)) - 90.0F;
-            float f1 = (float)(-(MathHelper.func_181159_b(d1, d3) * (double)(180F / (float)Math.PI)));
+            float f = (float)(MathHelper.func_181159_b(d2, d0) * (180D / Math.PI)) - 90.0F;
+            float f1 = (float)(-(MathHelper.func_181159_b(d1, d3) * (180D / Math.PI)));
             this.field_75659_a.field_70125_A = this.func_75652_a(this.field_75659_a.field_70125_A, f1, this.field_75658_c);
             this.field_75659_a.field_70759_as = this.func_75652_a(this.field_75659_a.field_70759_as, f, this.field_75657_b);
         }
