--- before/net/minecraft/entity/ai/EntityAIBreakDoor.java
+++ after/net/minecraft/entity/ai/EntityAIBreakDoor.java
@@ -24,7 +24,21 @@
         }
         else
         {
-            return !this.entity.world.getGameRules().getBoolean("mobGriefing")
+            return !net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.entity.world, this.entity)
+                   || !this.entity.world
+                       .getBlockState(this.doorPosition)
+                       .getBlock()
+                       .canEntityDestroy(
+                           this.entity.world.getBlockState(this.doorPosition),
+                           this.entity.world,
+                           this.doorPosition,
+                           this.entity
+                       )
+                   || !net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(
+                       this.entity,
+                       this.doorPosition,
+                       this.entity.world.getBlockState(this.doorPosition)
+                   )
                    ? false
                    : !BlockDoor.isOpen(this.entity.world, this.doorPosition);
         }
