--- before/net/minecraft/entity/ai/EntityAIMoveThroughVillage.java
+++ after/net/minecraft/entity/ai/EntityAIMoveThroughVillage.java
@@ -18,7 +18,7 @@
     private Path field_75419_c;
     private VillageDoorInfo field_75416_d;
     private final boolean field_75417_e;
-    private final List<VillageDoorInfo> field_75415_f = Lists.newArrayList();
+    private final List<VillageDoorInfo> field_75415_f = Lists.<VillageDoorInfo>newArrayList();
 
     public EntityAIMoveThroughVillage(EntityCreature p_i1638_1_, double p_i1638_2_, boolean p_i1638_4_)
     {
