--- before/net/minecraft/entity/ai/EntityAIAttackRangedBow.java
+++ after/net/minecraft/entity/ai/EntityAIAttackRangedBow.java
@@ -41,7 +41,7 @@
 
     protected boolean isBowInMainhand()
     {
-        return !this.entity.getHeldItemMainhand().isEmpty() && this.entity.getHeldItemMainhand().getItem() == Items.BOW;
+        return !this.entity.getHeldItemMainhand().isEmpty() && this.entity.getHeldItemMainhand().getItem() instanceof ItemBow;
     }
 
     @Override
