--- before/net/minecraft/client/gui/chat/NarratorChatListener.java
+++ after/net/minecraft/client/gui/chat/NarratorChatListener.java
@@ -40,23 +40,23 @@
     public void func_193641_a(int p_193641_1_)
     {
         this.field_192580_a.clear();
-        this.field_192580_a.say((new TextComponentTranslation("options.narrator")).func_150260_c() + " : " + (new TextComponentTranslation(GameSettings.field_193632_b[p_193641_1_])).func_150260_c());
+        this.field_192580_a.say((new TextComponentTranslation("options.narrator", new Object[0])).func_150260_c() + " : " + (new TextComponentTranslation(GameSettings.field_193632_b[p_193641_1_], new Object[0])).func_150260_c());
         GuiToast guitoast = Minecraft.func_71410_x().func_193033_an();
 
         if (this.field_192580_a.active())
         {
             if (p_193641_1_ == 0)
             {
-                SystemToast.func_193657_a(guitoast, SystemToast.Type.NARRATOR_TOGGLE, new TextComponentTranslation("narrator.toast.disabled"), (ITextComponent)null);
+                SystemToast.func_193657_a(guitoast, SystemToast.Type.NARRATOR_TOGGLE, new TextComponentTranslation("narrator.toast.disabled", new Object[0]), (ITextComponent)null);
             }
             else
             {
-                SystemToast.func_193657_a(guitoast, SystemToast.Type.NARRATOR_TOGGLE, new TextComponentTranslation("narrator.toast.enabled"), new TextComponentTranslation(GameSettings.field_193632_b[p_193641_1_]));
+                SystemToast.func_193657_a(guitoast, SystemToast.Type.NARRATOR_TOGGLE, new TextComponentTranslation("narrator.toast.enabled", new Object[0]), new TextComponentTranslation(GameSettings.field_193632_b[p_193641_1_], new Object[0]));
             }
         }
         else
         {
-            SystemToast.func_193657_a(guitoast, SystemToast.Type.NARRATOR_TOGGLE, new TextComponentTranslation("narrator.toast.disabled"), new TextComponentTranslation("options.narrator.notavailable"));
+            SystemToast.func_193657_a(guitoast, SystemToast.Type.NARRATOR_TOGGLE, new TextComponentTranslation("narrator.toast.disabled", new Object[0]), new TextComponentTranslation("options.narrator.notavailable", new Object[0]));
         }
     }
 
