--- original/net/minecraft/block/BlockPistonBase.java
+++ modified/net/minecraft/block/BlockPistonBase.java
@@ -245,7 +245,7 @@
                 }
 
                 if (!flag1
-                    && iblockstate.getMaterial() != Material.AIR
+                    && !iblockstate.getBlock().isAir(iblockstate, worldIn, blockpos)
                     && canPush(iblockstate, worldIn, blockpos, enumfacing.getOpposite(), false, enumfacing)
                     && (iblockstate.getPushReaction() == EnumPushReaction.NORMAL || block == Blocks.PISTON || block == Blocks.STICKY_PISTON)) {
                     this.doMove(worldIn, pos, enumfacing, false);
@@ -307,7 +307,7 @@
                     return false;
                 }
 
-                return !block.hasTileEntity();
+                return !block.hasTileEntity(blockStateIn);
             } else {
                 return false;
             }
@@ -342,7 +342,9 @@
         for (int j = list2.size() - 1; j >= 0; j--) {
             BlockPos blockpos1 = list2.get(j);
             IBlockState iblockstate = worldIn.getBlockState(blockpos1);
-            iblockstate.getBlock().dropBlockAsItem(worldIn, blockpos1, iblockstate, 0);
+            // Forge: With our change to how snowballs are dropped this needs to disallow to mimic vanilla behavior.
+            float chance = iblockstate.getBlock() instanceof BlockSnow ? -1.0f : 1.0f;
+            iblockstate.getBlock().dropBlockAsItemWithChance(worldIn, blockpos1, iblockstate, chance, 0);
             worldIn.setBlockState(blockpos1, Blocks.AIR.getDefaultState(), 4);
             aiblockstate[--k] = iblockstate;
         }
@@ -441,6 +443,13 @@
      * @return an approximation of the form of the given face
      * @deprecated call via {@link IBlockState#getBlockFaceShape(IBlockAccess,BlockPos,EnumFacing)} whenever possible. Implementing/overriding is fine.
      */
+    /* ======================================== FORGE START =====================================*/
+    public boolean rotateBlock(World world, BlockPos pos, EnumFacing axis)
+    {
+        IBlockState state = world.getBlockState(pos);
+        return !state.getValue(EXTENDED) && super.rotateBlock(world, pos, axis);
+    }
+    /* ======================================== FORGE END =======================================*/
     @Override
     public BlockFaceShape getBlockFaceShape(IBlockAccess worldIn, IBlockState state, BlockPos pos, EnumFacing face) {
         state = this.getActualState(state, worldIn, pos);
