--- original/net/minecraft/block/BlockMycelium.java
+++ modified/net/minecraft/block/BlockMycelium.java
@@ -36,7 +36,8 @@
     @Override
     public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand) {
         if (!worldIn.isRemote) {
-            if (worldIn.getLightFromNeighbors(pos.up()) < 4 && worldIn.getBlockState(pos.up()).getLightOpacity() > 2) {
+            if (!worldIn.isAreaLoaded(pos, 2)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light and spreading
+            if (worldIn.getLightFromNeighbors(pos.up()) < 4 && worldIn.getBlockState(pos.up()).getLightOpacity(worldIn, pos.up()) > 2) {
                 worldIn.setBlockState(pos, Blocks.DIRT.getDefaultState().withProperty(BlockDirt.VARIANT, BlockDirt.DirtType.DIRT));
             } else {
                 if (worldIn.getLightFromNeighbors(pos.up()) >= 9) {
@@ -45,9 +46,9 @@
                         IBlockState iblockstate = worldIn.getBlockState(blockpos);
                         IBlockState iblockstate1 = worldIn.getBlockState(blockpos.up());
                         if (iblockstate.getBlock() == Blocks.DIRT
-                            && iblockstate.getValue(BlockDirt.VARIANT) == BlockDirt.DirtType.DIRT
-                            && worldIn.getLightFromNeighbors(blockpos.up()) >= 4
-                            && iblockstate1.getLightOpacity() <= 2) {
+                                && iblockstate.getValue(BlockDirt.VARIANT) == BlockDirt.DirtType.DIRT
+                                && worldIn.getLightFromNeighbors(blockpos.up()) >= 4
+                                && iblockstate1.getLightOpacity(worldIn, blockpos.up()) <= 2) {
                             worldIn.setBlockState(blockpos, this.getDefaultState());
                         }
                     }
