--- original/net/minecraft/block/BlockPortal.java
+++ modified/net/minecraft/block/BlockPortal.java
@@ -108,12 +108,14 @@
 
     public boolean trySpawnPortal(World worldIn, BlockPos pos) {
         BlockPortal.Size blockportal$size = new BlockPortal.Size(worldIn, pos, EnumFacing.Axis.X);
-        if (blockportal$size.isValid() && blockportal$size.portalBlockCount == 0) {
+
+        if (blockportal$size.isValid() && blockportal$size.portalBlockCount == 0 && !net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(worldIn, pos, blockportal$size)) {
             blockportal$size.placePortalBlocks();
             return true;
         } else {
             BlockPortal.Size blockportal$size1 = new BlockPortal.Size(worldIn, pos, EnumFacing.Axis.Z);
-            if (blockportal$size1.isValid() && blockportal$size1.portalBlockCount == 0) {
+
+            if (blockportal$size1.isValid() && blockportal$size1.portalBlockCount == 0 && !net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(worldIn, pos, blockportal$size1)) {
                 blockportal$size1.placePortalBlocks();
                 return true;
             } else {
