--- original/net/minecraft/block/BlockRedstoneWire.java
+++ modified/net/minecraft/block/BlockRedstoneWire.java
@@ -123,13 +123,14 @@
     private BlockRedstoneWire.EnumAttachPosition getAttachPosition(IBlockAccess worldIn, BlockPos pos, EnumFacing direction) {
         BlockPos blockpos = pos.offset(direction);
         IBlockState iblockstate = worldIn.getBlockState(pos.offset(direction));
-        if (!canConnectTo(worldIn.getBlockState(blockpos), direction)
-            && (iblockstate.isNormalCube() || !canConnectUpwardsTo(worldIn.getBlockState(blockpos.down())))) {
+
+        if (!canConnectTo(worldIn.getBlockState(blockpos), direction, worldIn, blockpos)
+                && (iblockstate.isNormalCube() || !canConnectUpwardsTo(worldIn, blockpos.down()))) {
             IBlockState iblockstate1 = worldIn.getBlockState(pos.up());
             if (!iblockstate1.isNormalCube()) {
-                boolean flag = worldIn.getBlockState(blockpos).isTopSolid()
+                boolean flag = worldIn.getBlockState(blockpos).isSideSolid(worldIn, blockpos, EnumFacing.UP)
                     || worldIn.getBlockState(blockpos).getBlock() == Blocks.GLOWSTONE;
-                if (flag && canConnectUpwardsTo(worldIn.getBlockState(blockpos.up()))) {
+                if (flag && canConnectUpwardsTo(worldIn, blockpos.up())) {
                     if (iblockstate.isBlockNormalCube()) {
                         return BlockRedstoneWire.EnumAttachPosition.UP;
                     }
@@ -169,8 +170,10 @@
     /** Checks if this block can be placed exactly at the given position. */
     @Override
     public boolean canPlaceBlockAt(World worldIn, BlockPos pos) {
-        return worldIn.getBlockState(pos.down()).isTopSolid()
-            || worldIn.getBlockState(pos.down()).getBlock() == Blocks.GLOWSTONE;
+        IBlockState downState = worldIn.getBlockState(pos.down());
+        return downState.isTopSolid()
+               || downState.getBlockFaceShape(worldIn, pos.down(), EnumFacing.UP) == BlockFaceShape.SOLID
+               || worldIn.getBlockState(pos.down()).getBlock() == Blocks.GLOWSTONE;
     }
 
     private IBlockState updateSurroundingRedstone(World worldIn, BlockPos pos, IBlockState state) {
@@ -377,7 +380,7 @@
         boolean flag1 = worldIn.getBlockState(pos.up()).isNormalCube();
         if (!flag1 && flag && canConnectUpwardsTo(worldIn, blockpos.up())) {
             return true;
-        } else if (canConnectTo(iblockstate, side)) {
+        } else if (canConnectTo(iblockstate, side, worldIn, pos)) {
             return true;
         } else {
             return iblockstate.getBlock() == Blocks.POWERED_REPEATER && iblockstate.getValue(BlockRedstoneDiode.FACING) == side
@@ -387,14 +390,10 @@
     }
 
     protected static boolean canConnectUpwardsTo(IBlockAccess worldIn, BlockPos pos) {
-        return canConnectUpwardsTo(worldIn.getBlockState(pos));
+        return canConnectTo(worldIn.getBlockState(pos), null, worldIn, pos);
     }
 
-    protected static boolean canConnectUpwardsTo(IBlockState state) {
-        return canConnectTo(state, null);
-    }
-
-    protected static boolean canConnectTo(IBlockState blockState, @Nullable EnumFacing side) {
+    protected static boolean canConnectTo(IBlockState blockState, @Nullable EnumFacing side, IBlockAccess world, BlockPos pos) {
         Block block = blockState.getBlock();
         if (block == Blocks.REDSTONE_WIRE) {
             return true;
@@ -403,8 +402,8 @@
             return enumfacing == side || enumfacing.getOpposite() == side;
         } else {
             return Blocks.OBSERVER == blockState.getBlock()
-                ? side == blockState.getValue(BlockObserver.FACING)
-                : blockState.canProvidePower() && side != null;
+                   ? side == blockState.getValue(BlockObserver.FACING)
+                   : blockState.getBlock().canConnectRedstone(blockState, world, pos, side);
         }
     }
 
