--- original/net/minecraft/block/BlockRedstoneTorch.java
+++ modified/net/minecraft/block/BlockRedstoneTorch.java
@@ -20,7 +20,7 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class BlockRedstoneTorch extends BlockTorch {
-    private static final Map<World, List<BlockRedstoneTorch.Toggle>> toggles = Maps.newHashMap();
+    private static final Map<World, List<BlockRedstoneTorch.Toggle>> toggles = new java.util.WeakHashMap<World, List<Toggle>>(); // FORGE - fix vanilla MC-101233
     private final boolean isOn;
 
     private boolean isBurnedOut(World worldIn, BlockPos pos, boolean turnOff) {
