--- original/net/minecraft/block/BlockSnow.java
+++ modified/net/minecraft/block/BlockSnow.java
@@ -119,8 +119,8 @@
         if (block != Blocks.ICE && block != Blocks.PACKED_ICE && block != Blocks.BARRIER) {
             BlockFaceShape blockfaceshape = iblockstate.getBlockFaceShape(worldIn, pos.down(), EnumFacing.UP);
             return blockfaceshape == BlockFaceShape.SOLID
-                || iblockstate.getMaterial() == Material.LEAVES
-                || block == this && iblockstate.getValue(LAYERS) == 8;
+                   || iblockstate.getBlock().isLeaves(iblockstate, worldIn, pos.down())
+                   || block == this && iblockstate.getValue(LAYERS) == 8;
         } else {
             return false;
         }
@@ -134,7 +134,6 @@
 
     private boolean checkAndDropBlock(World worldIn, BlockPos pos, IBlockState state) {
         if (!this.canPlaceBlockAt(worldIn, pos)) {
-            this.dropBlockAsItem(worldIn, pos, state, 0);
             worldIn.setBlockToAir(pos);
             return false;
         } else {
@@ -147,9 +146,8 @@
     public void harvestBlock(
         World worldIn, EntityPlayer player, BlockPos pos, IBlockState state, @Nullable TileEntity te, ItemStack stack
     ) {
-        spawnAsEntity(worldIn, pos, new ItemStack(Items.SNOWBALL, state.getValue(LAYERS) + 1, 0));
+        super.harvestBlock(worldIn, player, pos, state, te, stack);
         worldIn.setBlockToAir(pos);
-        player.addStat(StatList.getBlockStats(this));
     }
 
     /** Get the Item that this Block should drop when harvested. */
@@ -161,13 +159,12 @@
     /** Returns the quantity of items to drop on block destruction. */
     @Override
     public int quantityDropped(Random random) {
-        return 0;
+        return 1;
     }
 
     @Override
     public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand) {
         if (worldIn.getLightFor(EnumSkyBlock.BLOCK, pos) > 11) {
-            this.dropBlockAsItem(worldIn, pos, worldIn.getBlockState(pos), 0);
             worldIn.setBlockToAir(pos);
         }
     }
@@ -205,6 +202,11 @@
     }
 
     @Override
+    public int quantityDropped(IBlockState state, int fortune, Random random) {
+        return state.getValue(LAYERS) + 1;
+    }
+
+    @Override
     protected BlockStateContainer createBlockState() {
         return new BlockStateContainer(this, LAYERS);
     }
