--- before/net/minecraft/entity/ai/attributes/AttributeMap.java
+++ after/net/minecraft/entity/ai/attributes/AttributeMap.java
@@ -8,8 +8,8 @@
 
 public class AttributeMap extends AbstractAttributeMap
 {
-    private final Set<IAttributeInstance> field_111162_d = Sets.newHashSet();
-    protected final Map<String, IAttributeInstance> field_111163_c = new LowerStringMap<>();
+    private final Set<IAttributeInstance> field_111162_d = Sets.<IAttributeInstance>newHashSet();
+    protected final Map<String, IAttributeInstance> field_111163_c = new LowerStringMap();
 
     public ModifiableAttributeInstance func_111151_a(IAttribute p_111151_1_)
     {
@@ -70,7 +70,7 @@
 
     public Collection<IAttributeInstance> func_111160_c()
     {
-        Set<IAttributeInstance> set = Sets.newHashSet();
+        Set<IAttributeInstance> set = Sets.<IAttributeInstance>newHashSet();
 
         for (IAttributeInstance iattributeinstance : this.func_111146_a())
         {
