--- before/net/minecraft/entity/ai/attributes/RangedAttribute.java
+++ after/net/minecraft/entity/ai/attributes/RangedAttribute.java
@@ -42,6 +42,7 @@
 
     public double func_111109_a(double p_111109_1_)
     {
-        return MathHelper.func_151237_a(p_111109_1_, this.field_111120_a, this.field_111118_b);
+        p_111109_1_ = MathHelper.func_151237_a(p_111109_1_, this.field_111120_a, this.field_111118_b);
+        return p_111109_1_;
     }
 }
