--- before/net/minecraft/client/renderer/texture/LayeredColorMaskTexture.java
+++ after/net/minecraft/client/renderer/texture/LayeredColorMaskTexture.java
@@ -67,13 +67,11 @@
                     try
                     {
                         String s = this.field_174949_h.get(j);
-                        int k = this.field_174950_i.get(j).func_193350_e();
+                        int k = ((EnumDyeColor)this.field_174950_i.get(j)).func_193350_e();
 
                         if (s != null)
                         {
-                            iresource1 = p_110551_1_.func_110536_a(new ResourceLocation(s));
-                            BufferedImage bufferedimage2 = TextureUtil.func_177053_a(iresource1.func_110527_b());
-
+                            BufferedImage bufferedimage2 = net.minecraftforge.client.MinecraftForgeClient.getImageLayer(new ResourceLocation(s), p_110551_1_);
                             if (bufferedimage2.getWidth() == bufferedimage.getWidth() && bufferedimage2.getHeight() == bufferedimage.getHeight() && bufferedimage2.getType() == 6)
                             {
                                 for (int l = 0; l < bufferedimage2.getHeight(); ++l)
