--- before/net/minecraft/client/renderer/texture/TextureManager.java
+++ after/net/minecraft/client/renderer/texture/TextureManager.java
@@ -138,6 +138,7 @@
 
         if (itextureobject != null)
         {
+            this.mapTextureObjects.remove(textureLocation); // Forge: fix MC-98707
             TextureUtil.deleteTexture(itextureobject.getGlTextureId());
         }
     }
@@ -145,11 +146,13 @@
     @Override
     public void onResourceManagerReload(IResourceManager resourceManager)
     {
+        net.minecraftforge.fml.common.ProgressManager.ProgressBar bar = net.minecraftforge.fml.common.ProgressManager.push("Reloading Texture Manager", this.mapTextureObjects.keySet().size(), true);
         Iterator<Entry<ResourceLocation, ITextureObject>> iterator = this.mapTextureObjects.entrySet().iterator();
 
         while (iterator.hasNext())
         {
             Entry<ResourceLocation, ITextureObject> entry = iterator.next();
+            bar.step(entry.getKey().toString());
             ITextureObject itextureobject = entry.getValue();
 
             if (itextureobject == TextureUtil.MISSING_TEXTURE)
@@ -161,5 +164,6 @@
                 this.loadTexture(entry.getKey(), itextureobject);
             }
         }
+        net.minecraftforge.fml.common.ProgressManager.pop(bar);
     }
 }
