--- original/net/minecraft/enchantment/EnchantmentFrostWalker.java
+++ modified/net/minecraft/enchantment/EnchantmentFrostWalker.java
@@ -57,8 +57,9 @@
                     if (iblockstate.getMaterial() == Material.AIR) {
                         IBlockState iblockstate1 = worldIn.getBlockState(blockpos$mutableblockpos1);
                         if (iblockstate1.getMaterial() == Material.WATER
-                            && iblockstate1.getValue(BlockLiquid.LEVEL) == 0
-                            && worldIn.mayPlace(Blocks.FROSTED_ICE, blockpos$mutableblockpos1, false, EnumFacing.DOWN, null)) {
+                                && (iblockstate1.getBlock() == net.minecraft.init.Blocks.WATER || iblockstate1.getBlock() == net.minecraft.init.Blocks.FLOWING_WATER)
+                                && iblockstate1.getValue(BlockLiquid.LEVEL) == 0
+                                && worldIn.mayPlace(Blocks.FROSTED_ICE, blockpos$mutableblockpos1, false, EnumFacing.DOWN, null)) {
                             worldIn.setBlockState(blockpos$mutableblockpos1, Blocks.FROSTED_ICE.getDefaultState());
                             worldIn.scheduleUpdate(
                                 blockpos$mutableblockpos1.toImmutable(),
