--- before/net/minecraft/util/datafix/DataFixesManager.java
+++ after/net/minecraft/util/datafix/DataFixesManager.java
@@ -165,6 +165,7 @@
     public static DataFixer createFixer()
     {
         DataFixer datafixer = new DataFixer(1343);
+        datafixer = new net.minecraftforge.common.util.CompoundDataFixer(datafixer);
         WorldInfo.registerFixes(datafixer);
         EntityPlayerMP.registerFixesPlayerMP(datafixer);
         EntityPlayer.registerFixesPlayer(datafixer);
@@ -272,7 +273,7 @@
         {
             NBTTagList nbttaglist = compound.getTagList(key, 10);
 
-            for (int i = 0; i < nbttaglist.tagCount(); i++)
+            for (int i = 0; i < nbttaglist.tagCount(); ++i)
             {
                 nbttaglist.set(i, fixer.process(FixTypes.ITEM_INSTANCE, nbttaglist.getCompoundTagAt(i), version));
             }
