--- before/net/minecraft/util/datafix/DataFixer.java
+++ after/net/minecraft/util/datafix/DataFixer.java
@@ -12,8 +12,8 @@
 public class DataFixer implements IDataFixer
 {
     private static final Logger field_188259_a = LogManager.getLogger();
-    private final Map<IFixType, List<IDataWalker>> field_188260_b = Maps.newHashMap();
-    private final Map<IFixType, List<IFixableData>> field_188261_c = Maps.newHashMap();
+    private final Map<IFixType, List<IDataWalker>> field_188260_b = Maps.<IFixType, List<IDataWalker>>newHashMap();
+    private final Map<IFixType, List<IFixableData>> field_188261_c = Maps.<IFixType, List<IFixableData>>newHashMap();
     public final int field_188262_d;
 
     public DataFixer(int p_i46828_1_)
@@ -40,7 +40,7 @@
 
     private NBTTagCompound func_188252_b(IFixType p_188252_1_, NBTTagCompound p_188252_2_, int p_188252_3_)
     {
-        List<IFixableData> list = this.field_188261_c.get(p_188252_1_);
+        List<IFixableData> list = (List)this.field_188261_c.get(p_188252_1_);
 
         if (list != null)
         {
@@ -60,13 +60,13 @@
 
     private NBTTagCompound func_188253_c(IFixType p_188253_1_, NBTTagCompound p_188253_2_, int p_188253_3_)
     {
-        List<IDataWalker> list = this.field_188260_b.get(p_188253_1_);
+        List<IDataWalker> list = (List)this.field_188260_b.get(p_188253_1_);
 
         if (list != null)
         {
             for (int i = 0; i < list.size(); ++i)
             {
-                p_188253_2_ = list.get(i).func_188266_a(this, p_188253_2_, p_188253_3_);
+                p_188253_2_ = ((IDataWalker)list.get(i)).func_188266_a(this, p_188253_2_, p_188253_3_);
             }
         }
 
@@ -85,20 +85,20 @@
 
     public void func_188256_a(IFixType p_188256_1_, IFixableData p_188256_2_)
     {
-        List<IFixableData> list = this.func_188254_a(this.field_188261_c, p_188256_1_);
+        List<IFixableData> list = this.<IFixableData>func_188254_a(this.field_188261_c, p_188256_1_);
         int i = p_188256_2_.func_188216_a();
 
         if (i > this.field_188262_d)
         {
-            field_188259_a.warn("Ignored fix registered for version: {} as the DataVersion of the game is: {}", i, this.field_188262_d);
+            field_188259_a.warn("Ignored fix registered for version: {} as the DataVersion of the game is: {}", Integer.valueOf(i), Integer.valueOf(this.field_188262_d));
         }
         else
         {
-            if (!list.isEmpty() && Util.func_184878_a(list).func_188216_a() > i)
+            if (!list.isEmpty() && ((IFixableData)Util.func_184878_a(list)).func_188216_a() > i)
             {
                 for (int j = 0; j < list.size(); ++j)
                 {
-                    if (list.get(j).func_188216_a() > i)
+                    if (((IFixableData)list.get(j)).func_188216_a() > i)
                     {
                         list.add(j, p_188256_2_);
                         break;
@@ -114,11 +114,11 @@
 
     private <V> List<V> func_188254_a(Map<IFixType, List<V>> p_188254_1_, IFixType p_188254_2_)
     {
-        List<V> list = p_188254_1_.get(p_188254_2_);
+        List<V> list = (List)p_188254_1_.get(p_188254_2_);
 
         if (list == null)
         {
-            list = Lists.newArrayList();
+            list = Lists.<V>newArrayList();
             p_188254_1_.put(p_188254_2_, list);
         }
 
